// igniteui-angular-charts/esm5/lib/AbsoluteVolumeOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { AbsoluteVolumeOscillatorIndicatorStrategy } from "./AbsoluteVolumeOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var AbsoluteVolumeOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AbsoluteVolumeOscillatorIndicator, _super);
    function AbsoluteVolumeOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AbsoluteVolumeOscillatorIndicator.prototype.bi = function () {
        return new AbsoluteVolumeOscillatorIndicator();
    };
    AbsoluteVolumeOscillatorIndicator.prototype.get_abn = function () {
        return new AbsoluteVolumeOscillatorIndicatorStrategy();
    };
    Object.defineProperty(AbsoluteVolumeOscillatorIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    AbsoluteVolumeOscillatorIndicator.prototype.get_ab0 = function () {
        return AbsoluteVolumeOscillatorIndicator.$;
    };
    Object.defineProperty(AbsoluteVolumeOscillatorIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AbsoluteVolumeOscillatorIndicator.prototype, "shortPeriod", {
        get: function () {
            return typeGetValue(this.c(AbsoluteVolumeOscillatorIndicator.ab9));
        },
        set: function (a) {
            this.h(AbsoluteVolumeOscillatorIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    AbsoluteVolumeOscillatorIndicator.prototype.abv = function () {
        return this.shortPeriod;
    };
    Object.defineProperty(AbsoluteVolumeOscillatorIndicator.prototype, "longPeriod", {
        get: function () {
            return typeGetValue(this.c(AbsoluteVolumeOscillatorIndicator.ab8));
        },
        set: function (a) {
            this.h(AbsoluteVolumeOscillatorIndicator.ab8, a);
        },
        enumerable: false,
        configurable: true
    });
    AbsoluteVolumeOscillatorIndicator.prototype.abs = function () {
        return this.longPeriod;
    };
    AbsoluteVolumeOscillatorIndicator.$t = markType(AbsoluteVolumeOscillatorIndicator, 'AbsoluteVolumeOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    AbsoluteVolumeOscillatorIndicator.ab9 = StrategyBasedIndicator.ab7(10, AbsoluteVolumeOscillatorIndicator.$);
    AbsoluteVolumeOscillatorIndicator.ab8 = StrategyBasedIndicator.ab3(30, AbsoluteVolumeOscillatorIndicator.$);
    return AbsoluteVolumeOscillatorIndicator;
}(StrategyBasedIndicator));
export { AbsoluteVolumeOscillatorIndicator };
