// igniteui-angular-charts/esm5/lib/AnnotationLayerProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, typeCast, delegateCombine, delegateRemove, markType } from "igniteui-angular-core";
import { IAnnotationLayerProxy_$type } from "./IAnnotationLayerProxy";
import { CrosshairLayer } from "./CrosshairLayer";
import { ValueLayer } from "./ValueLayer";
import { CalloutLayer } from "./CalloutLayer";
import { FinalValueLayer } from "./FinalValueLayer";
import { DataToolTipLayer } from "./DataToolTipLayer";
import { DoubleCollection } from "igniteui-angular-core";
/**
 * @hidden
 */
var AnnotationLayerProxy = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AnnotationLayerProxy, _super);
    function AnnotationLayerProxy() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._container = null;
        return _this;
    }
    Object.defineProperty(AnnotationLayerProxy.prototype, "container", {
        get: function () {
            return this._container;
        },
        set: function (a) {
            this._container = a;
        },
        enumerable: false,
        configurable: true
    });
    AnnotationLayerProxy.prototype.matchesType = function (a) {
        if (!a.eq) {
            return false;
        }
        return true;
    };
    AnnotationLayerProxy.prototype.createCrosshairLayer = function () {
        var a = new CrosshairLayer();
        var b = this.container.getExternal(a, null, null);
        return a;
    };
    AnnotationLayerProxy.prototype.createValueLayer = function () {
        var a = new ValueLayer();
        var b = this.container.getExternal(a, null, null);
        return a;
    };
    AnnotationLayerProxy.prototype.createCalloutLayer = function () {
        var a = new CalloutLayer();
        var b = this.container.getExternal(a, null, null);
        return a;
    };
    AnnotationLayerProxy.prototype.createFinalValueLayer = function () {
        var a = new FinalValueLayer();
        var b = this.container.getExternal(a, null, null);
        return a;
    };
    AnnotationLayerProxy.prototype.setCrosshairHorizontalLineVisibility = function (a, b) {
        a.acg = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairVerticalLineVisibility = function (a, b) {
        a.ach = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairHorizontalLineStroke = function (a, b) {
        a.ab4 = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairVerticalLineStroke = function (a, b) {
        a.ab5 = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairXAxisForeground = function (a, b) {
        a.ab8 = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairYAxisForeground = function (a, b) {
        a.acb = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairXAxisBackground = function (a, b) {
        a.ab6 = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairYAxisBackground = function (a, b) {
        a.ab9 = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairYAxisPrecision = function (a, b) {
        a.aad = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairXAxisPrecision = function (a, b) {
        a.aac = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairOnAxesEnabled = function (a, b) {
        a.zv = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairUseInterpolation = function (a, b) {
        a.zx = b;
    };
    AnnotationLayerProxy.prototype.setCrosshairThickness = function (a, b) {
        a.jn = b;
    };
    AnnotationLayerProxy.prototype.isCrosshairLayer = function (a) {
        return typeCast(CrosshairLayer.$, a) !== null;
    };
    AnnotationLayerProxy.prototype.setFinalValuePrecision = function (a, b) {
        a.z2 = b;
    };
    AnnotationLayerProxy.prototype.setFinalValueBackground = function (a, b) {
        a.aaw = b;
    };
    AnnotationLayerProxy.prototype.setFinalValueForeground = function (a, b) {
        a.aay = b;
    };
    AnnotationLayerProxy.prototype.setContentMemberPath = function (a, b) {
        a.acd = b;
    };
    AnnotationLayerProxy.prototype.setCalloutsAllowedPositions = function (a, b) {
        a.z2 = b;
    };
    AnnotationLayerProxy.prototype.setLabelMemberPath = function (a, b) {
        a.acr = b;
    };
    AnnotationLayerProxy.prototype.setXMemberPath = function (a, b) {
        a.ac5 = b;
    };
    AnnotationLayerProxy.prototype.setYMemberPath = function (a, b) {
        a.ac8 = b;
    };
    AnnotationLayerProxy.prototype.setTargetSeries = function (a, b) {
        a.z7 = b;
    };
    AnnotationLayerProxy.prototype.setCollisionChannel = function (a, b) {
        a.aca = b;
    };
    AnnotationLayerProxy.prototype.setIsAutoCalloutBehaviorEnabled = function (a, b) {
        a.aav = b;
    };
    AnnotationLayerProxy.prototype.setGreedyCollisionMode = function (a, b) {
        a.zx = b ? 2 : 0;
    };
    AnnotationLayerProxy.prototype.setCollisionMode = function (a, b) {
        a.zx = b;
    };
    AnnotationLayerProxy.prototype.setUseItemColorForFill = function (a, b) {
        a.aa1 = b;
    };
    AnnotationLayerProxy.prototype.setUseItemColorForOutline = function (a, b) {
        a.aa2 = b;
    };
    AnnotationLayerProxy.prototype.setCalloutStrokeThickness = function (a, b) {
        a.abk = b;
    };
    AnnotationLayerProxy.prototype.setUseAutoContrastingLabelColors = function (a, b) {
        a.aaz = b;
    };
    AnnotationLayerProxy.prototype.setCalloutBackground = function (a, b) {
        a.aet = b;
    };
    AnnotationLayerProxy.prototype.setAutoCalloutLabelPrecision = function (a, b) {
        a.abn = b;
    };
    AnnotationLayerProxy.prototype.setCalloutLightColor = function (a, b) {
        a.aew = b;
    };
    AnnotationLayerProxy.prototype.setCalloutDarkColor = function (a, b) {
        a.aeu = b;
    };
    AnnotationLayerProxy.prototype.setUseDedicatedLanes = function (a, b) {
        a.zv = b ? 2 : 0;
    };
    AnnotationLayerProxy.prototype.isCalloutLayer = function (a) {
        if (typeCast(CalloutLayer.$, a) !== null) {
            return true;
        }
        return false;
    };
    AnnotationLayerProxy.prototype.isDataToolTipLayer = function (a) {
        if (typeCast(DataToolTipLayer.$, a) !== null) {
            return true;
        }
        return false;
    };
    AnnotationLayerProxy.prototype.isFinalValueLayer = function (a) {
        if (typeCast(FinalValueLayer.$, a) !== null) {
            return true;
        }
        return false;
    };
    AnnotationLayerProxy.prototype.addCalloutStyleUpdating = function (a, b) {
        var c = a;
        c.calloutStyleUpdating = delegateCombine(c.calloutStyleUpdating, b);
    };
    AnnotationLayerProxy.prototype.removeCalloutStyleUpdating = function (a, b) {
        var c = a;
        c.calloutStyleUpdating = delegateRemove(c.calloutStyleUpdating, b);
    };
    AnnotationLayerProxy.prototype.addCalloutRenderStyleUpdating = function (a, b) {
        var c = a;
        c.calloutRenderStyleUpdating = delegateCombine(c.calloutRenderStyleUpdating, b);
    };
    AnnotationLayerProxy.prototype.removeCalloutRenderStyleUpdating = function (a, b) {
        var c = a;
        c.calloutRenderStyleUpdating = delegateRemove(c.calloutRenderStyleUpdating, b);
    };
    AnnotationLayerProxy.prototype.setCalloutStyleUpdatingEventEnabled = function (a, b) {
        a.aay = b;
    };
    AnnotationLayerProxy.prototype.setUseValueForAutoCalloutLabels = function (a, b) {
        a.aa4 = b;
    };
    AnnotationLayerProxy.prototype.setCalloutRenderStyleUpdatingEventEnabled = function (a, b) {
        a.aax = b;
    };
    AnnotationLayerProxy.prototype.updateDataToolTipPropertyByName = function (a, b, c) {
        a.afc(b, c);
    };
    AnnotationLayerProxy.prototype.getIsAutoCalloutBehaviorEnabled = function (a) {
        return a.aav;
    };
    AnnotationLayerProxy.prototype.setValueMode = function (a, b) {
        a.zu = b;
    };
    AnnotationLayerProxy.prototype.setValueAxisAnnotationEnabled = function (a, b) {
        a.zw = b;
    };
    AnnotationLayerProxy.prototype.getValueMode = function (a) {
        return a.zu;
    };
    AnnotationLayerProxy.prototype.isValueLayer = function (a) {
        return typeCast(ValueLayer.$, a) !== null;
    };
    AnnotationLayerProxy.prototype.setBrush = function (a, b) {
        a.u9 = b;
    };
    AnnotationLayerProxy.prototype.setDashArray = function (a, b) {
        var c = new DoubleCollection();
        for (var d = 0; d < b.length; d++) {
            c.add(b[d]);
        }
        a.vw = c;
    };
    AnnotationLayerProxy.$t = markType(AnnotationLayerProxy, 'AnnotationLayerProxy', Base.$, [IAnnotationLayerProxy_$type]);
    return AnnotationLayerProxy;
}(Base));
export { AnnotationLayerProxy };
