// igniteui-angular-charts/esm5/lib/AverageDirectionalIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AverageDirectionalIndexIndicatorStrategy } from "./AverageDirectionalIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var AverageDirectionalIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AverageDirectionalIndexIndicator, _super);
    function AverageDirectionalIndexIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AverageDirectionalIndexIndicator.prototype.bi = function () {
        return new AverageDirectionalIndexIndicator();
    };
    AverageDirectionalIndexIndicator.prototype.get_abn = function () {
        return new AverageDirectionalIndexIndicatorStrategy();
    };
    Object.defineProperty(AverageDirectionalIndexIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    AverageDirectionalIndexIndicator.prototype.get_ab0 = function () {
        return AverageDirectionalIndexIndicator.$;
    };
    Object.defineProperty(AverageDirectionalIndexIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AverageDirectionalIndexIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(AverageDirectionalIndexIndicator.ab9));
        },
        set: function (a) {
            this.h(AverageDirectionalIndexIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    AverageDirectionalIndexIndicator.prototype.abt = function () {
        return this.ab8;
    };
    AverageDirectionalIndexIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    AverageDirectionalIndexIndicator.$t = markType(AverageDirectionalIndexIndicator, 'AverageDirectionalIndexIndicator', StrategyBasedIndicator.$);
    AverageDirectionalIndexIndicator.ab9 = StrategyBasedIndicator.ab5(14, AverageDirectionalIndexIndicator.$);
    return AverageDirectionalIndexIndicator;
}(StrategyBasedIndicator));
export { AverageDirectionalIndexIndicator };
