// igniteui-angular-charts/esm5/lib/AverageTrueRangeIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AverageTrueRangeIndicatorStrategy } from "./AverageTrueRangeIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var AverageTrueRangeIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AverageTrueRangeIndicator, _super);
    function AverageTrueRangeIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AverageTrueRangeIndicator.prototype.bi = function () {
        return new AverageTrueRangeIndicator();
    };
    AverageTrueRangeIndicator.prototype.get_abn = function () {
        return new AverageTrueRangeIndicatorStrategy();
    };
    Object.defineProperty(AverageTrueRangeIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    AverageTrueRangeIndicator.prototype.get_ab0 = function () {
        return AverageTrueRangeIndicator.$;
    };
    Object.defineProperty(AverageTrueRangeIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AverageTrueRangeIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(AverageTrueRangeIndicator.ab9));
        },
        set: function (a) {
            this.h(AverageTrueRangeIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    AverageTrueRangeIndicator.prototype.abt = function () {
        return this.ab8;
    };
    AverageTrueRangeIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    AverageTrueRangeIndicator.$t = markType(AverageTrueRangeIndicator, 'AverageTrueRangeIndicator', StrategyBasedIndicator.$);
    AverageTrueRangeIndicator.ab9 = StrategyBasedIndicator.ab5(14, AverageTrueRangeIndicator.$);
    return AverageTrueRangeIndicator;
}(StrategyBasedIndicator));
export { AverageTrueRangeIndicator };
