// igniteui-angular-charts/esm5/lib/BollingerBandWidthIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { BollingerBandWidthIndicatorStrategy } from "./BollingerBandWidthIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var BollingerBandWidthIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BollingerBandWidthIndicator, _super);
    function BollingerBandWidthIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BollingerBandWidthIndicator.prototype.bi = function () {
        return new BollingerBandWidthIndicator();
    };
    BollingerBandWidthIndicator.prototype.get_abn = function () {
        return new BollingerBandWidthIndicatorStrategy();
    };
    Object.defineProperty(BollingerBandWidthIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    BollingerBandWidthIndicator.prototype.get_ab0 = function () {
        return BollingerBandWidthIndicator.$;
    };
    Object.defineProperty(BollingerBandWidthIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BollingerBandWidthIndicator.prototype, "ab9", {
        get: function () {
            return typeGetValue(this.c(BollingerBandWidthIndicator.acc));
        },
        set: function (a) {
            this.h(BollingerBandWidthIndicator.acc, a);
        },
        enumerable: false,
        configurable: true
    });
    BollingerBandWidthIndicator.prototype.abt = function () {
        return this.ab9;
    };
    Object.defineProperty(BollingerBandWidthIndicator.prototype, "ab8", {
        get: function () {
            return this.c(BollingerBandWidthIndicator.acb);
        },
        set: function (a) {
            this.h(BollingerBandWidthIndicator.acb, a);
        },
        enumerable: false,
        configurable: true
    });
    BollingerBandWidthIndicator.prototype.abr = function () {
        return this.ab8;
    };
    BollingerBandWidthIndicator.$t = markType(BollingerBandWidthIndicator, 'BollingerBandWidthIndicator', StrategyBasedIndicator.$);
    BollingerBandWidthIndicator.acc = StrategyBasedIndicator.ab5(20, BollingerBandWidthIndicator.$);
    BollingerBandWidthIndicator.acb = StrategyBasedIndicator.ab4(2, BollingerBandWidthIndicator.$);
    return BollingerBandWidthIndicator;
}(StrategyBasedIndicator));
export { BollingerBandWidthIndicator };
