// igniteui-angular-charts/esm5/lib/BollingerBandsOverlay_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { FinancialOverlay } from "./FinancialOverlay";
import { DependencyProperty } from "igniteui-angular-core";
import { Series } from "./Series";
import { List$1 } from "igniteui-angular-core";
import { Number_$type, typeCast, runOn, markType, markDep, getEnumerator, Point_$type, fromEnum } from "igniteui-angular-core";
import { CategoryFrame } from "./CategoryFrame";
import { SafeEnumerable } from "./SafeEnumerable";
import { TrendCalculators } from "igniteui-angular-core";
import { AxisRange } from "./AxisRange";
import { FinancialSeriesView } from "./FinancialSeriesView";
import { ScalerParams } from "./ScalerParams";
import { ISortingAxis_$type } from "./ISortingAxis";
import { FinancialBucketCalculator } from "./FinancialBucketCalculator";
import { Polyline } from "igniteui-angular-core";
import { Polygon } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { isNaN_ } from "igniteui-angular-core";
import { Flattener } from "igniteui-angular-core";
/**
 * @hidden
 */
var BollingerBandsOverlay = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BollingerBandsOverlay, _super);
    function BollingerBandsOverlay() {
        var _this = _super.call(this) || this;
        _this.aa0 = 0;
        _this.aa1 = 0;
        _this.aay = new List$1(Number_$type, 0);
        _this.aaz = new List$1(Number_$type, 0);
        _this.ab = BollingerBandsOverlay.$;
        _this.w2 = new CategoryFrame(4);
        _this.w4 = new CategoryFrame(4);
        _this.w0 = new CategoryFrame(4);
        return _this;
    }
    BollingerBandsOverlay.prototype.bi = function () {
        return new BollingerBandsOverlay();
    };
    BollingerBandsOverlay.prototype.q5 = function (a, b, c, d) {
        if (this.yh().contains(b)) {
            this.aas = false;
        }
        _super.prototype.q5.call(this, a, b, c, d);
        switch (b) {
            case BollingerBandsOverlay.$$p[1]:
                this.aas = false;
                this.rn(false);
                break;
            case BollingerBandsOverlay.$$p[0]:
                this.rn(false);
                break;
        }
    };
    BollingerBandsOverlay.prototype.aat = function () {
        this.aay.clear();
        this.aaz.clear();
        var a = getEnumerator(TrendCalculators.e(new SafeEnumerable(this.yg), this.aa3));
        var b = getEnumerator(TrendCalculators.f(new SafeEnumerable(this.yg), this.aa3));
        this.aa1 = 1.7976931348623157E+308;
        this.aa0 = -1.7976931348623157E+308;
        var c = true;
        var d = true;
        var e = this.aa2;
        while (c || d) {
            if (a.moveNext()) {
                this.aay.add(a.current);
            }
            else {
                c = false;
            }
            if (b.moveNext()) {
                this.aaz.add(b.current);
            }
            else {
                d = false;
            }
            if (c && d) {
                this.aa1 = Math.min(this.aa1, a.current - b.current * e);
                this.aa0 = Math.max(this.aa0, a.current + b.current * e);
            }
        }
        return true;
    };
    BollingerBandsOverlay.prototype.ax = function (a) {
        if (this.aay == null || this.aaz == null || a == null || this.dl == null || a != this.xh) {
            return null;
        }
        var b = new AxisRange(this.aa1, this.aa0);
        return b;
    };
    BollingerBandsOverlay.prototype.z4 = function (a, b) {
        _super.prototype.z4.call(this, a, b);
        var c = b.by;
        var d = b.bx;
        var e = this.getEffectiveViewport1(b);
        var f = this.wy;
        var g = this.xh;
        var h = new ScalerParams(0, c, d, f.ch, e);
        var i = new ScalerParams(0, c, d, g.ch, e);
        a.f.clear();
        a.n.clear();
        a.q.clear();
        var j = typeCast(ISortingAxis_$type, this.wy);
        if (j != null && j.i$b.count != this.dl.count) {
            return;
        }
        var k = 0;
        k = this.yp(c, d, e);
        if (!this.aas) {
            this.aas = this.aat();
            if (this.xh != null) {
                this.xh.c2(true);
            }
        }
        i.b = this.iw();
        var l = this.wy.en(2, h) - this.wy.en(1, h);
        for (var m = b.b0.c; m <= b.b0.d; ++m) {
            var n = void 0;
            if (j == null) {
                n = b.b0.getBucket(m);
            }
            else {
                var o = j.i$b._inner[m];
                var p = j.i$d(o);
                var q = this.aay._inner[m];
                var r = this.aaz._inner[m];
                var s = q;
                var t = r;
                var u = p;
                var v = 1;
                while (m < b.b0.d) {
                    o = j.i$b._inner[m + 1];
                    u = j.i$d(o);
                    if (u - p > l) {
                        break;
                    }
                    m++;
                    s += this.aay._inner[m];
                    t += this.aaz._inner[m];
                    v++;
                }
                s /= v;
                t /= v;
                var w = s - t * this.aa2;
                var x = s;
                var y = s + t * this.aa2;
                var z = NaN;
                if (!isNaN_(p)) {
                    z = this.wy.ek(p, h);
                }
                n = [z, w, x, y];
            }
            var aa = Math.max(1, l);
            if (!isNaN_(n[0]) && m * aa >= this.aau) {
                if (this.wy != null && typeCast(ISortingAxis_$type, this.wy) !== null) {
                    n[0] = n[0] + k;
                }
                else {
                    n[0] = f.ek(n[0], h) + k;
                }
                n[1] = g.ek(n[1], i);
                n[2] = g.ek(n[2], i);
                n[3] = g.ek(n[3], i);
                a.f.add(n);
            }
        }
    };
    BollingerBandsOverlay.prototype.o3 = function (a, b) {
        _super.prototype.o3.call(this, a, b);
        var c = b;
        if (c != null) {
            c.b7();
        }
    };
    BollingerBandsOverlay.prototype.i3 = function (a, b, c) {
        if (this.aay == null) {
            return _super.prototype.i3.call(this, a, b, c);
        }
        var d = this.getEffectiveViewport1(this.ct);
        var e = this.yp(this.ct.by, this.ct.bx, d);
        var f = new ScalerParams(0, this.ct.by, this.ct.bx, this.wy.ch, d);
        f.c = this.c9.v0;
        return this.i5(this.aay, a, this.wy, f, e, runOn(this, this.yo), b, c);
    };
    BollingerBandsOverlay.prototype.j8 = function (a, b) {
        if (this.aay == null) {
            return _super.prototype.j8.call(this, a, b);
        }
        return this.j9(a, b, this.wy, runOn(this, this.yo), this.aay);
    };
    BollingerBandsOverlay.prototype.j6 = function (a, b) {
        if (this.aay == null) {
            return _super.prototype.j6.call(this, a, b);
        }
        return this.j7(a, b, this.wy, runOn(this, this.yo), this.aay);
    };
    BollingerBandsOverlay.prototype.hc = function (a, b, c) {
        var d = _super.prototype.hc.call(this, a, b, c);
        if (!this.x7(this.yh())) {
            d = false;
        }
        return d;
    };
    BollingerBandsOverlay.prototype.z6 = function (a, b) {
        _super.prototype.z6.call(this, a, b);
        var c = typeCast(BollingerBandsOverlayView.$, b);
        var d = this.getEffectiveViewport1(b);
        if (b.b3(a)) {
            c.b7();
            var e_1 = a.f.count;
            var f = function (g) { return a.f._inner[g][0]; };
            var g = function (h) { return a.f._inner[e_1 - 1 - h][0]; };
            var h = function (i) { return a.f._inner[i][1]; };
            var i = function (j) { return a.f._inner[j][2]; };
            var j = function (k) { return a.f._inner[e_1 - 1 - k][3]; };
            c.b8(e_1, f, g, h, i, j);
            b.b5(a);
        }
        this.w7.a2(this, this.x6(), runOn(this, this.wq), this.yv(b), this.yx(b));
        var k = false;
        var l = this.w7.b;
        var m = a.f;
        var n = this.dl.count;
        if (l != null) {
            k = true;
        }
        if (k) {
            var o = new ScalerParams(0, b.by, b.bx, this.wy.ch, d);
            this.z3(m, -1, n, this.wy, o, b.q, -1);
        }
        var p = c.cf;
        var q = c.cg;
        var r = c.ch;
        var s = c.ca;
        this.w7.ad(q, true, false, true, true);
        this.w7.an = null;
        this.w7.am = null;
        this.w7.ad(p, true, false, false, true);
        this.w7.ad(r, true, false, false, true);
        this.w7.ad(s, false, true, false, false);
        s._opacity = this.w7.j * this.h7;
    };
    BollingerBandsOverlay.prototype.cq = function () {
        return new BollingerBandsOverlayView(this);
    };
    BollingerBandsOverlay.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.aax = typeCast(BollingerBandsOverlayView.$, a);
    };
    BollingerBandsOverlay.prototype.yy = function () {
        return 3;
    };
    BollingerBandsOverlay.prototype.iw = function () {
        return Series.iy(this.aay, null);
    };
    BollingerBandsOverlay.$t = markType(BollingerBandsOverlay, 'BollingerBandsOverlay', FinancialOverlay.$);
    BollingerBandsOverlay.$$p = markDep(DependencyProperty, PropertyMetadata, BollingerBandsOverlay, 'raisePropertyChanged', ['Multiplier:aa2:aa6', [1, 2], 'Period:aa3:aa7', [1, 14]]);
    return BollingerBandsOverlay;
}(FinancialOverlay));
export { BollingerBandsOverlay };
/**
 * @hidden
 */
var BollingerBandsBucketCalculator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BollingerBandsBucketCalculator, _super);
    function BollingerBandsBucketCalculator(a) {
        var _this = _super.call(this, a) || this;
        _this.f = null;
        _this.f = a;
        return _this;
    }
    BollingerBandsBucketCalculator.prototype.getBucket = function (a) {
        var b = a * this.b;
        var c = Math.min(b + this.b - 1, this.a.b2.dl.count - 1);
        if (b <= c) {
            var d = this.f.b6.aa2;
            var e = 0;
            var f = 0;
            var g = 0;
            for (var h = b; h <= c; ++h) {
                if (!isNaN_(this.f.b6.aay._inner[h]) && !isNaN_(this.f.b6.aaz._inner[h])) {
                    e += this.f.b6.aay._inner[h];
                    f += this.f.b6.aaz._inner[h];
                    ++g;
                }
            }
            if (g > 0) {
                e = e / g;
                f = f / g;
                return [0.5 * (b + c), e - f * d, e, e + f * d];
            }
        }
        return [NaN, NaN, NaN, NaN];
    };
    BollingerBandsBucketCalculator.$t = markType(BollingerBandsBucketCalculator, 'BollingerBandsBucketCalculator', FinancialBucketCalculator.$);
    return BollingerBandsBucketCalculator;
}(FinancialBucketCalculator));
export { BollingerBandsBucketCalculator };
/**
 * @hidden
 */
var BollingerBandsOverlayView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BollingerBandsOverlayView, _super);
    function BollingerBandsOverlayView(a) {
        var _this = _super.call(this, a) || this;
        _this.b6 = null;
        _this.cb = new Polygon();
        _this.ci = new Polyline();
        _this.cj = new Polyline();
        _this.ck = new Polyline();
        _this.b9 = new Polygon();
        _this.cc = new Polyline();
        _this.cd = new Polyline();
        _this.ce = new Polyline();
        _this.b6 = a;
        return _this;
    }
    BollingerBandsOverlayView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
        }
    };
    BollingerBandsOverlayView.prototype.b1 = function () {
        return new BollingerBandsBucketCalculator(this);
    };
    BollingerBandsOverlayView.prototype.b8 = function (a, b, c, d, e, f) {
        var e_2, _a, e_3, _b, e_4, _c;
        try {
            for (var _d = __values(fromEnum(Flattener.d(a, b, d, this.e.ib))), _e = _d.next(); !_e.done; _e = _d.next()) {
                var g = _e.value;
                this.cb.an.add({ $type: Point_$type, x: b(g), y: d(g) });
                this.ci.an.add({ $type: Point_$type, x: b(g), y: d(g) });
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return))
                    _a.call(_d);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
        try {
            for (var _f = __values(fromEnum(Flattener.d(a, b, e, this.e.ib))), _g = _f.next(); !_g.done; _g = _f.next()) {
                var h = _g.value;
                this.cj.an.add({ $type: Point_$type, x: b(h), y: e(h) });
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_g && !_g.done && (_b = _f.return))
                    _b.call(_f);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
        try {
            for (var _h = __values(fromEnum(Flattener.d(a, c, f, this.e.ib))), _j = _h.next(); !_j.done; _j = _h.next()) {
                var i = _j.value;
                this.cb.an.add({ $type: Point_$type, x: c(i), y: f(i) });
                this.ck.an.add({ $type: Point_$type, x: c(i), y: f(i) });
            }
        }
        catch (e_4_1) {
            e_4 = { error: e_4_1 };
        }
        finally {
            try {
                if (_j && !_j.done && (_c = _h.return))
                    _c.call(_h);
            }
            finally {
                if (e_4)
                    throw e_4.error;
            }
        }
        this.as();
    };
    BollingerBandsOverlayView.prototype.b7 = function () {
        this.cb.an.clear();
        this.ci.an.clear();
        this.cj.an.clear();
        this.ck.an.clear();
        this.as();
    };
    Object.defineProperty(BollingerBandsOverlayView.prototype, "ca", {
        get: function () {
            return this.cb;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BollingerBandsOverlayView.prototype, "cf", {
        get: function () {
            return this.ci;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BollingerBandsOverlayView.prototype, "cg", {
        get: function () {
            return this.cj;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BollingerBandsOverlayView.prototype, "ch", {
        get: function () {
            return this.ck;
        },
        enumerable: false,
        configurable: true
    });
    BollingerBandsOverlayView.prototype.be = function () {
        _super.prototype.be.call(this);
        this.b9.an = this.cb.an;
        this.cc.an = this.ci.an;
        this.cd.an = this.cj.an;
        this.ce.an = this.ck.an;
        var a = this.br();
        this.b9._fill = a;
        this.b9._opacity = 1;
        this.cc._stroke = a;
        this.cc.ad = this.e.ic + 3;
        this.cd._stroke = a;
        this.cd.ad = this.e.ic + 3;
        this.ce._stroke = a;
        this.ce.ad = this.e.ic + 3;
    };
    BollingerBandsOverlayView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (a.d) {
            if (b) {
                a.u(this.b9);
                a.v(this.cc);
                a.v(this.cd);
                a.v(this.ce);
            }
            else {
                a.u(this.cb);
                a.v(this.ci);
                a.v(this.cj);
                a.v(this.ck);
            }
        }
    };
    BollingerBandsOverlayView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        this.e.bb.exportPolygonData(a, this.cb, "FillShape", ["Fill"]);
        this.e.bb.exportPolylineData(a, this.ci, "BottomLine", ["Lower", "Main"]);
        this.e.bb.exportPolylineData(a, this.cj, "CentralLine", ["Central"]);
        this.e.bb.exportPolylineData(a, this.ck, "TopLine", ["Upper"]);
    };
    BollingerBandsOverlayView.$t = markType(BollingerBandsOverlayView, 'BollingerBandsOverlayView', FinancialSeriesView.$);
    return BollingerBandsOverlayView;
}(FinancialSeriesView));
export { BollingerBandsOverlayView };
