// igniteui-angular-charts/esm5/lib/CategoryBucketCalculator_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { Base, typeCast, markType, Array_$type, Number_$type, Point_$type, enumGetBox, runOn, delegateCombine, fromEnum, markDep } from "igniteui-angular-core";
import { IBucketizer_$type } from "./IBucketizer";
import { ArgumentNullException } from "igniteui-angular-core";
import { Rect } from "igniteui-angular-core";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { ISortingAxis_$type } from "./ISortingAxis";
import { ScalerParams } from "./ScalerParams";
import { Series } from "./Series";
import { truncate, intDivide, isNaN_, isInfinity } from "igniteui-angular-core";
import { FramePreparer } from "./FramePreparer";
import { IIsCategoryBased_$type } from "./IIsCategoryBased";
import { List$1 } from "igniteui-angular-core";
import { ISupportsMarkers_$type } from "./ISupportsMarkers";
import { IProvidesViewport_$type } from "./IProvidesViewport";
import { ISupportsErrorBars_$type } from "./ISupportsErrorBars";
import { CategoryFrame } from "./CategoryFrame";
import { PreparationParams } from "./PreparationParams";
import { IDateTimeAxis_$type } from "./IDateTimeAxis";
import { NumericAxisBase } from "./NumericAxisBase";
import { IHasCategoryTrendline_$type } from "./IHasCategoryTrendline";
import { IHasSingleValueCategory_$type } from "./IHasSingleValueCategory";
import { DefaultCategoryTrendlineHost } from "./DefaultCategoryTrendlineHost";
import { DefaultSingleValueProvider } from "./DefaultSingleValueProvider";
import { SingleValuesHolder } from "./SingleValuesHolder";
import { MarkerSeries } from "./MarkerSeries";
import { IHasCategoryModePreference_$type } from "./IHasCategoryModePreference";
import { ICategorySeriesInteractionManager_$type } from "./ICategorySeriesInteractionManager";
import { DVContainer } from "igniteui-angular-core";
import { DependencyProperty } from "igniteui-angular-core";
import { CategoryTransitionInMode_$type } from "./CategoryTransitionInMode";
import { HighlightingInfo } from "igniteui-angular-core";
import { CategoryTransitionSourceFramePreparer } from "./CategoryTransitionSourceFramePreparer";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { Marker } from "./Marker";
import { MarkerSeriesView } from "./MarkerSeriesView";
import { CategorySeriesRenderManager } from "./CategorySeriesRenderManager";
import { Pool$1 } from "igniteui-angular-core";
import { SeriesLayerPropertyOverlay } from "./SeriesLayerPropertyOverlay";
import { DoubleCollection } from "igniteui-angular-core";
import { Clipper } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { round10N } from "igniteui-angular-core";
/**
 * @hidden
 */
var CategoryBucketCalculator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryBucketCalculator, _super);
    function CategoryBucketCalculator(a) {
        var _this = _super.call(this) || this;
        _this.c = null;
        _this.e = 0;
        _this.f = 0;
        _this.d = 0;
        if (a == null) {
            throw new ArgumentNullException(0, "view");
        }
        _this.c = a;
        _this.e = -1;
        _this.d = 0;
        return _this;
    }
    CategoryBucketCalculator.prototype.getBucket = function (a) {
        return null;
    };
    CategoryBucketCalculator.prototype.getBucketWithoutUnknowns = function (a) {
        return this.getBucket(a);
    };
    CategoryBucketCalculator.prototype.getErrorBucket = function (a, b) {
        return NaN;
    };
    CategoryBucketCalculator.prototype.g = function (a) {
        var b = this.c.by;
        var c = this.c.bx;
        var d = this.c.cy;
        var e = typeCast(CategoryAxisBase.$, d.fetchXAxis());
        var f = d.dl;
        if (b.isEmpty || c.isEmpty || e == null || f == null || f.count == 0) {
            this.d = 0;
            return;
        }
        var g = d.getEffectiveViewport1(this.c);
        var h = d.wd(this.c);
        var i = typeCast(ISortingAxis_$type, e);
        if (i == null || i.i$b == null) {
            var j = new ScalerParams(0, b, c, e.ch, g);
            var k = Math.floor(e.en(h.left, j));
            var l = Math.ceil(e.en(h.right, j));
            if (e.ch) {
                l = Math.ceil(e.en(h.left, j));
                k = Math.floor(e.en(h.right, j));
            }
            l = Math.min(l, e.lh - 1);
            var m = Math.floor((l - k + 1) * a / h.width);
            this.d = truncate(Math.max(1, m));
            this.e = truncate(Math.max(0, Math.floor(k / this.d) - 1));
            this.f = truncate(Math.ceil(l / this.d));
        }
        else {
            this.e = i.i$e(b, c, g, h);
            this.f = i.i$g(b, c, g, h);
            this.d = 1;
        }
    };
    CategoryBucketCalculator.prototype.getBucketInfo = function (a, b, c, d) {
        a = this.e;
        b = this.f;
        c = this.d;
        d = this.c.cy.ib;
        return {
            p0: a,
            p1: b,
            p2: c,
            p3: d
        };
    };
    CategoryBucketCalculator.prototype.cacheValues = function () {
    };
    CategoryBucketCalculator.prototype.unCacheValues = function () {
    };
    CategoryBucketCalculator.$t = markType(CategoryBucketCalculator, 'CategoryBucketCalculator', Base.$, [IBucketizer_$type]);
    return CategoryBucketCalculator;
}(Base));
export { CategoryBucketCalculator };
/**
 * @hidden
 */
var CategoryFramePreparerBase = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryFramePreparerBase, _super);
    function CategoryFramePreparerBase(a) {
        var _rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            _rest[_i - 1] = arguments[_i];
        }
        var _this = this;
        a = (a == void 0) ? 0 : a;
        switch (a) {
            case 0:
                {
                    var c = _rest[0];
                    var d = [c, typeCast(ISupportsMarkers_$type, c), typeCast(IProvidesViewport_$type, c), typeCast(ISupportsErrorBars_$type, c), typeCast(IBucketizer_$type, c)];
                    {
                        var f = d[0];
                        var g = d[1];
                        var h = d[2];
                        var i = d[3];
                        var j = d[4];
                        _this = _super.call(this, 1, g, h, i) || this;
                        _this.g = null;
                        _this.h = null;
                        _this.m = false;
                        _this.o = new List$1(Array_$type, 0);
                        _this.p = new List$1(Number_$type, 0);
                        _this.n = new List$1(Number_$type, 0);
                        _this.h = f;
                        _this.g = j;
                    }
                }
                break;
            case 1:
                {
                    var c = _rest[0];
                    var d = _rest[1];
                    var e = _rest[2];
                    var f = _rest[3];
                    var g = _rest[4];
                    _this = _super.call(this, 1, d, e, f) || this;
                    _this.g = null;
                    _this.h = null;
                    _this.m = false;
                    _this.o = new List$1(Array_$type, 0);
                    _this.p = new List$1(Number_$type, 0);
                    _this.n = new List$1(Number_$type, 0);
                    _this.h = c;
                    _this.g = g;
                }
                break;
        }
        return _this;
    }
    CategoryFramePreparerBase.prototype.i = function (a) {
        var b = new PreparationParams();
        b.f = this.h.useHighMarkerFidelity;
        b.b = this.h.scaler;
        b.c = this.h.yScaler;
        b.d = typeCast(ISortingAxis_$type, b.b);
        b.a = typeCast(CategoryFrame.$, a);
        if (b.a == null || b.b == null || b.c == null) {
            return null;
        }
        var c;
        var d;
        var e;
        var f;
        var g = this.g.getBucketInfo(c, d, e, f);
        c = g.p0;
        d = g.p1;
        e = g.p2;
        f = g.p3;
        b.k = c;
        b.l = d;
        b.j = e;
        b.i = f;
        if (b.l < b.k) {
            return null;
        }
        var h;
        var i;
        var j;
        var k = this.a.getViewInfo(i, h);
        i = k.p0;
        h = k.p1;
        j = this.a.getEffectiveViewport();
        var l = this.a.getMainContentViewport();
        b.p = h;
        b.o = i;
        b.n = j;
        b.m = l;
        if (Rect.l_op_Equality(b.p, Rect.empty) || Rect.l_op_Equality(b.o, Rect.empty)) {
            return null;
        }
        if (this.h != null && (typeCast(Series.$, this.h) !== null && this.h.e9) && this.g != null) {
            b.e = true;
        }
        return b;
    };
    CategoryFramePreparerBase.prototype.r = function (a, b, c, d) {
        return a.getSeriesOffset(b, c, d, this.h.currentMode2Index, this.h.currentCategoryMode);
    };
    CategoryFramePreparerBase.prototype.s = function (a) {
        return this.r(a.b, a.p, a.o, a.n);
    };
    CategoryFramePreparerBase.prototype.d = function (a, b) {
        var c = this.i(a);
        if (c == null || this.g == null) {
            return;
        }
        c.a.f.clear();
        c.a.k.clear();
        c.a.n.clear();
        c.a.q.clear();
        c.a.i.clear();
        c.a.h.clear();
        c.a.g.clear();
        if (c.a.m != null) {
            c.a.m.clear();
        }
        var d = this.c.shouldDisplayMarkers;
        var e = false;
        var f = this.s(c);
        var g = this.j(c);
        if (c.d != null && c.d.i$b != null && c.d.i$b.count != g.e) {
            return;
        }
        if (c.d != null && c.d.i$a) {
            c.d.i$h();
        }
        this.g.cacheValues();
        if (this.m && c.d != null && c.d.i$b.count > 0) {
            var h = c.d.i$b._inner[c.k];
            this.z(g, h, true, c.e);
            while ((isNaN_(g.c) || isNaN_(g.d)) && c.k > 0) {
                c.k = c.k - 1;
                h = c.d.i$b._inner[c.k];
                this.z(g, h, true, c.e);
            }
            var i = c.d.i$b._inner[c.l];
            this.z(g, i, true, c.e);
            while ((isNaN_(g.c) || isNaN_(g.d)) && c.l < g.e - 1) {
                c.l = c.l + 1;
                i = c.d.i$b._inner[c.l];
                this.z(g, i, true, c.e);
            }
        }
        if (this.m && c.d == null) {
            var j = c.k;
            var k = this.g.getBucket(j);
            while ((isNaN_(k[1]) || isNaN_(k[2])) && c.k > 0) {
                c.k = c.k - 1;
                j = c.k;
                k = this.g.getBucket(j);
            }
            var l = c.l;
            k = this.g.getBucket(l);
            while ((isNaN_(k[1]) || isNaN_(k[2])) && c.l < (intDivide(g.e, c.j))) {
                c.l = c.l + 1;
                l = c.l;
                k = this.g.getBucket(l);
            }
        }
        this.w(c, g, f);
        var m = this.t(c, g, f, d, e);
        this.c.updateMarkerCount(m);
        this.v(typeCast(CategoryFrame.$, a), b);
        this.g.unCacheValues();
        return;
    };
    CategoryFramePreparerBase.prototype.v = function (a, b) {
    };
    Object.defineProperty(CategoryFramePreparerBase.prototype, "q", {
        get: function () {
            return this.o;
        },
        enumerable: false,
        configurable: true
    });
    CategoryFramePreparerBase.prototype.t = function (a, b, c, d, e) {
        var _this = this;
        var f = 0;
        var g = false;
        var h = null;
        var i = false;
        var j = a.d != null;
        var k = a.p;
        var l = a.o;
        var m = typeCast(NumericAxisBase.$, a.c) !== null && a.c.lk;
        var n = a.f;
        var o = this.h.mayContainUnknowns;
        var p = this.o;
        var q = this.n;
        var r = this.p;
        p.clear();
        q.clear();
        r.clear();
        var s = this.g;
        var t = this.h.provideCollisionDetector();
        var u = 0;
        var v = a.n;
        var w = new ScalerParams(0, k, l, a.b.j$b, v);
        var x = ((function () {
            var $ret = new ScalerParams(0, k, l, a.c.j$b, v);
            $ret.b = typeCast(Series.$, _this.h) !== null ? _this.h.iw() : NaN;
            return $ret;
        })());
        if (j) {
            u = a.b.j$d(2, w) - a.b.j$d(1, w);
            i = this.k();
        }
        var y = a.k;
        var z = a.l;
        var aa = a.a.f;
        var ab = typeCast(IDateTimeAxis_$type, a.b);
        if (ab != null) {
            ab.StartTimeCursor();
        }
        var _loop_1 = function (ac) {
            var ad = void 0;
            if (!j) {
                if (!o) {
                    ad = s.getBucketWithoutUnknowns(ac);
                }
                else {
                    ad = s.getBucket(ac);
                }
            }
            else {
                ad = ((function () { var ae = _this.f(a, ac, b, u, g, h, c); ac = ae.p1; g = ae.p4; h = ae.p5; return ae.ret; })());
            }
            var ae = !m || (m && ad[1] > 0);
            var af = ad[0];
            if (!j || !isNaN_(ad[0])) {
                aa.add(ad);
                if (j) {
                    if (a.a.m == null) {
                        a.a.m = new List$1(Number_$type, 0);
                    }
                    a.a.m.add(ac * a.j);
                }
                if (g && i) {
                    if (h != null) {
                        aa.add(h);
                        if (j) {
                            if (a.a.m == null) {
                                a.a.m = new List$1(Number_$type, 0);
                            }
                            a.a.m.add(ac * a.j);
                        }
                    }
                }
                if (d && ae) {
                    var ag = ac * a.j;
                    var ah = ag;
                    if (j && a.d.i$b != null && ag >= 0 && ag < a.d.i$b.count) {
                        ag = a.d.i$b._inner[ag];
                    }
                    var ai = new Array(ad.length);
                    ai[0] = ad[0];
                    ai[1] = ad[1];
                    ai[2] = ad[2];
                    if (n && ag < b.e) {
                        ai = new Array(ad.length);
                        ai[0] = af;
                        this_1.z(b, ag, true, a.e);
                        ai[1] = b.c;
                        ai[2] = b.d;
                    }
                    p.add(ai);
                    q.add(a.a.f.count - 1);
                    r.add(ag);
                }
            }
            out_ac_1 = ac;
        };
        var this_1 = this, out_ac_1;
        for (var ac = y; ac <= z; ++ac) {
            _loop_1(ac);
            ac = out_ac_1;
        }
        if (ab != null) {
            ab.ResetTimeCursor();
        }
        this.y(a, c, j, w, x, e, d);
        if (d) {
            var aj = p.count;
            var ak = void 0;
            for (var al = 0; al < aj; al++) {
                ak = p._inner[al];
                if (this.l(a.a, ak, t, Math.min(r._inner[al], b.e - 1), f, q._inner[al])) {
                    ++f;
                }
            }
        }
        return f;
    };
    CategoryFramePreparerBase.prototype.k = function () {
        if (typeCast(Series.$, this.h) !== null) {
            return this.h.et;
        }
        return false;
    };
    CategoryFramePreparerBase.prototype.f = function (a, b, c, d, e, f, g) {
        var h;
        f = null;
        e = false;
        var i = typeCast(CategorySeries.$, this.h);
        var j = a.o;
        var k = a.p;
        var l = a.n;
        var m = new ScalerParams(0, k, j, a.b.j$b, l);
        var n = new ScalerParams(0, k, j, a.c.j$b, l);
        var o = a.e;
        if (a.d == null || a.d.i$b == null || a.d.i$b.count == 0) {
            return {
                ret: [NaN, NaN, NaN],
                p1: b,
                p4: e,
                p5: f
            };
        }
        if (i != null && i.dl != null && i.dl.count < a.d.i$b.count) {
            return {
                ret: [NaN, NaN, NaN],
                p1: b,
                p4: e,
                p5: f
            };
        }
        var p = a.d.i$b._inner[b];
        var q = a.d.i$d(p);
        var r = q;
        this.z(c, p, false, o);
        while (b < a.l) {
            p = a.d.i$b._inner[b + 1];
            r = a.d.i$d(p);
            this.z(c, p, true, o);
            if (r - q >= d || isNaN_(c.c) || isNaN_(c.d)) {
                if (e) {
                    var s = a.d.i$b._inner[b];
                    this.z(c, s, true, o);
                    f = [a.b.j$c(q, m), c.c, c.d];
                }
                break;
            }
            if (!e && this.k()) {
                var t = a.d.i$b._inner[b];
                this.z(c, t, true, o);
                var u = [a.b.j$c(q, m), c.c, c.d];
                if (!isNaN_(u[0])) {
                    if (!isNaN_(u[1]) && !isNaN_(u[2])) {
                        if (a.d != null) {
                            u[0] = u[0] + g;
                        }
                        a.a.f.add(u);
                        if (a.d != null) {
                            if (a.a.m == null) {
                                a.a.m = new List$1(Number_$type, 0);
                            }
                            a.a.m.add(b);
                        }
                        e = true;
                    }
                    else {
                        break;
                    }
                }
            }
            b++;
            this.u(c, p, o);
        }
        var v = NaN;
        if (!isNaN_(q)) {
            v = a.b.j$c(q, m);
        }
        h = this.e(v, c);
        return {
            ret: h,
            p1: b,
            p4: e,
            p5: f
        };
    };
    CategoryFramePreparerBase.prototype.y = function (a, b, c, d, e, f, g) {
        var h = a.j;
        var i = a.a.f.count;
        var j = a.a.f;
        var k = this.o;
        var l = k.count;
        if (f) {
            var m = a.c.j$c(a.c.l8, e);
            a.c.j$f(a.a.k, 0, a.a.k.count, e);
            var n = a.a.k;
            var o = n.count;
            for (var p = 0; p < n.count; p++) {
                n._inner[p] = Math.abs(n._inner[p] - m);
            }
        }
    };
    CategoryFramePreparerBase.prototype.w = function (a, b, c) {
    };
    CategoryFramePreparerBase.$t = markType(CategoryFramePreparerBase, 'CategoryFramePreparerBase', FramePreparer.$);
    return CategoryFramePreparerBase;
}(FramePreparer));
export { CategoryFramePreparerBase };
/**
 * @hidden
 */
var CategoryFramePreparer = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryFramePreparer, _super);
    function CategoryFramePreparer(a) {
        var _rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            _rest[_i - 1] = arguments[_i];
        }
        var _this = this;
        a = (a == void 0) ? 0 : a;
        switch (a) {
            case 0:
                {
                    var c = _rest[0];
                    var d = [c, typeCast(ISupportsMarkers_$type, c), typeCast(IProvidesViewport_$type, c), typeCast(ISupportsErrorBars_$type, c), typeCast(IBucketizer_$type, c)];
                    {
                        var f = d[0];
                        var g = d[1];
                        var h = d[2];
                        var i = d[3];
                        var j = d[4];
                        _this = _super.call(this, 1, f, g, h, i, j) || this;
                        _this.aa = new DefaultCategoryTrendlineHost();
                        _this.ab = new DefaultSingleValueProvider();
                        _this.ac = null;
                        _this.ad = null;
                        _this.ac = _this.aa;
                        if (typeCast(IHasCategoryTrendline_$type, f) !== null) {
                            _this.ac = typeCast(IHasCategoryTrendline_$type, f);
                        }
                        _this.ad = _this.ab;
                        if (typeCast(IHasSingleValueCategory_$type, f) !== null) {
                            _this.ad = typeCast(IHasSingleValueCategory_$type, f);
                        }
                    }
                }
                break;
            case 1:
                {
                    var c = _rest[0];
                    var d = _rest[1];
                    var e = _rest[2];
                    var f = _rest[3];
                    var g = _rest[4];
                    _this = _super.call(this, 1, c, d, e, f, g) || this;
                    _this.aa = new DefaultCategoryTrendlineHost();
                    _this.ab = new DefaultSingleValueProvider();
                    _this.ac = null;
                    _this.ad = null;
                    _this.ac = _this.aa;
                    if (typeCast(IHasCategoryTrendline_$type, c) !== null) {
                        _this.ac = typeCast(IHasCategoryTrendline_$type, c);
                    }
                    _this.ad = _this.ab;
                    if (typeCast(IHasSingleValueCategory_$type, c) !== null) {
                        _this.ad = typeCast(IHasSingleValueCategory_$type, c);
                    }
                }
                break;
        }
        return _this;
    }
    CategoryFramePreparer.prototype.l = function (a, b, c, d, e, f) {
        var g = b[0];
        var h = b[1];
        var i = new Rect(0, g - 5, h - 5, 11, 11);
        if (!isNaN_(g) && !isNaN_(h) && !isInfinity(g) && !isInfinity(h) && c.tryAdd(i)) {
            a.n.add({ $type: Point_$type, x: g, y: h });
            this.c.updateMarkerTemplate(e, d, f);
            return true;
        }
        return false;
    };
    CategoryFramePreparer.prototype.j = function (a) {
        var b = new SingleValuesHolder();
        var c = this.ad.valueColumn;
        b.f = c;
        return b;
    };
    CategoryFramePreparer.prototype.x = function (a, b, c, d, e, f) {
        if (d) {
            b[0] = b[0] + c;
        }
        else {
            b[0] = a.b.j$c(b[0], e) + c;
        }
        b[1] = a.c.j$c(b[1], f);
        if (a.j > 1 || d) {
            b[2] = a.c.j$c(b[2], f);
        }
        else {
            b[2] = b[1];
        }
    };
    CategoryFramePreparer.prototype.y = function (a, b, c, d, e, f, g) {
        _super.prototype.y.call(this, a, b, c, d, e, f, g);
        var h = a.j;
        var i = a.a.f.count;
        var j = a.a.f;
        var k = this.q;
        var l = k.count;
        var m = new List$1(Number_$type, 0);
        if (!c) {
            m.add(0);
        }
        var n = new List$1(Number_$type, 0);
        n.add(1);
        if (c || h > 1) {
            n.add(2);
        }
        a.b.j$e(j, m, 0, i, d);
        a.c.j$e(j, n, 0, i, e);
        for (var o = 0; o < i; o++) {
            j._inner[o][0] = j._inner[o][0] + b;
        }
        if (h == 1 && !c) {
            for (var p = 0; p < i; p++) {
                j._inner[p][2] = j._inner[p][1];
            }
        }
        if (g) {
            a.b.j$e(k, m, 0, l, d);
            a.c.j$e(k, n, 0, l, e);
            for (var q = 0; q < l; q++) {
                k._inner[q][0] = k._inner[q][0] + b;
            }
            if (h == 1 && !c) {
                for (var r = 0; r < l; r++) {
                    k._inner[r][2] = k._inner[r][1];
                }
            }
        }
    };
    CategoryFramePreparer.prototype.w = function (a, b, c) {
        if (this.ac.trendLineType == 0 || this.ac.trendlinePreparer == null || this.ac.trendLinePeriod < 1) {
            return;
        }
        this.ac.trendlinePreparer.prepareLine(this.ac, a, b, c);
    };
    CategoryFramePreparer.prototype.z = function (a, b, c, d) {
        var e = a;
        var f = e.f;
        var g = f.item(b);
        var h = g;
        if (d) {
            var i = this.g.getBucket(b);
            g = i[1];
            h = i[1];
        }
        if (c) {
            e.c = g;
            e.d = h;
        }
        else {
            e.a = g;
            e.b = h;
        }
    };
    CategoryFramePreparer.prototype.u = function (a, b, c) {
        var d = a;
        var e = d.f;
        if (b < e.count) {
            var f = void 0;
            if (c) {
                var g = this.g.getBucket(b);
                f = g[1];
            }
            else {
                f = e.item(b);
            }
            d.a = Math.min(d.a, f);
            d.b = Math.max(d.b, f);
        }
    };
    CategoryFramePreparer.prototype.e = function (a, b) {
        var c = b;
        return [a, c.a, c.b];
    };
    CategoryFramePreparer.$t = markType(CategoryFramePreparer, 'CategoryFramePreparer', CategoryFramePreparerBase.$);
    return CategoryFramePreparer;
}(CategoryFramePreparerBase));
export { CategoryFramePreparer };
/**
 * @hidden
 */
var CategorySeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategorySeries, _super);
    function CategorySeries() {
        var _this = _super.call(this) || this;
        _this.yq = null;
        _this.yy = null;
        _this.y7 = false;
        _this.y6 = false;
        _this.assigningCategoryStyle = null;
        _this.assigningCategoryMarkerStyle = null;
        _this.y4 = null;
        _this.y5 = null;
        _this.yn = new CategoryFrame(3);
        _this.yp = new CategoryFrame(3);
        _this.yl = new CategoryFrame(3);
        _this.yo = new CategoryFrame(3);
        _this.yt = null;
        _this.yx = null;
        _this.yt = new CategorySeriesRenderManager();
        _this.yx = new CategoryTransitionSourceFramePreparer();
        _this.y7 = false;
        _this.d0 = false;
        _this.y6 = false;
        _this.ab = CategorySeries.$;
        return _this;
    }
    CategorySeries.prototype.cq = function () {
        var a = new CategorySeriesView(this);
        return a;
    };
    Object.defineProperty(CategorySeries.prototype, "yz", {
        get: function () {
            var _this = this;
            if (this.yy == null) {
                this.yy = DVContainer.instance.createInstance(ICategorySeriesInteractionManager_$type, function () { return _this.dp.n; });
            }
            return this.yy;
        },
        enumerable: false,
        configurable: true
    });
    CategorySeries.prototype.d9 = function (a, b, c, d) {
        var _this = this;
        if (!this.hasMarkers) {
            return false;
        }
        var e = this.getEffectiveViewport1(this.ct);
        var f = this.yk();
        var g = new ScalerParams(1, this.c9.vl, this.wl, f.ch);
        g.c = e;
        var h = this.y0();
        var i = new ScalerParams(1, this.c9.vl, this.wl, h.ch);
        i.c = e;
        var j = this.c9.vl;
        var k = this.wl;
        var l = this.currentMode2Index;
        var m = 0;
        if (f.ce && f.cc) {
            m = f.getSeriesOffset(j, k, e, l, f.categoryMode);
        }
        a.clear();
        this.wy.cj(function (n) {
            if (n._visibility == 0) {
                var o = new SeriesPointOfInterest();
                var p = n.n;
                var q = n.o;
                if (p < _this.wl.left || p > _this.wl.right || q < _this.wl.top || q > _this.wl.bottom) {
                    return;
                }
                if (_this.isVertical) {
                    if (f.cb && !f.ce && f.categoryMode != 0) {
                        o.e = n.ak;
                    }
                    else {
                        o.e = f.en(q, g);
                    }
                    o.d = h.en(p, i);
                }
                else {
                    if (f.cb && !f.ce && f.categoryMode != 0) {
                        o.d = n.ak;
                    }
                    else {
                        o.d = f.en(p, g);
                    }
                    o.e = h.en(q, i);
                }
                o.b = n.content;
                if (b) {
                    if (_this.isVertical) {
                        if (c) {
                            o.f = _this.zq(o.d, d);
                        }
                        else {
                            var r = _this.jb(n);
                            if (!isNaN_(r)) {
                                o.f = _this.zq(r, d);
                            }
                            else {
                                o.f = _this.zq(o.d, d);
                            }
                        }
                    }
                    else {
                        if (c) {
                            o.f = _this.zq(o.e, d);
                        }
                        else {
                            var s = _this.jb(n);
                            if (!isNaN_(s)) {
                                o.f = _this.zq(s, d);
                            }
                            else {
                                o.f = _this.zq(o.e, d);
                            }
                        }
                    }
                }
                else {
                    var t = _this.at(true);
                    if (t != null) {
                        var u = _this.kq(n.ak, false);
                        if (u != null) {
                            o.f = t.i(u);
                        }
                        else {
                            o.f = null;
                        }
                    }
                    else {
                        o.f = f.ge(n.ak);
                    }
                }
                o.c = n.ak;
                o.g = _this.va;
                if (_this.g7) {
                    o.g = _this.vh(n.ak);
                }
                a.add(o);
            }
        });
        return true;
    };
    CategorySeries.prototype.zq = function (a, b) {
        var c;
        if (b >= 0) {
            c = round10N(a, b);
        }
        else {
            c = a;
        }
        var d = this.at(false);
        if (d != null) {
            var e = d;
            return e.i(c);
        }
        else {
            return c.toString();
        }
    };
    CategorySeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.yu = a;
    };
    CategorySeries.prototype.preferredCategoryMode = function (a) {
        return 0;
    };
    Object.defineProperty(CategorySeries.prototype, "categoryAxis", {
        get: function () {
            return this.yk();
        },
        enumerable: false,
        configurable: true
    });
    CategorySeries.prototype.yk = function () {
        return this.fetchXAxis();
    };
    CategorySeries.prototype.y0 = function () {
        return this.fetchYAxis();
    };
    CategorySeries.prototype.get_ei = function () {
        return true;
    };
    CategorySeries.prototype.get_f3 = function () {
        return this.y0() != null && this.y0().ch;
    };
    CategorySeries.prototype.get_ew = function () {
        return true;
    };
    CategorySeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.yu.cw.g(this.ib);
    };
    CategorySeries.prototype.zd = function () {
        return (this.assigningCategoryStyle != null && this.y7) || (this.bv != 1) || this.d0 || (this.cn != 1) || (this.cm != 1);
    };
    CategorySeries.prototype.ze = function () {
        return (this.assigningCategoryMarkerStyle != null && this.y6) || (this.bv != 1) || this.d0 || (this.cn != 1) || (this.cm != 1);
    };
    CategorySeries.prototype.z3 = function (a) {
        if (this.assigningCategoryStyle != null && this.y7) {
            this.assigningCategoryStyle(this, a);
        }
    };
    CategorySeries.prototype.z2 = function (a) {
        if (this.assigningCategoryMarkerStyle != null && this.y6) {
            this.assigningCategoryMarkerStyle(this, a);
        }
    };
    CategorySeries.prototype.ye = function (a, b) {
        return null;
    };
    CategorySeries.prototype.o3 = function (a, b) {
        _super.prototype.o3.call(this, a, b);
        if (a) {
            this.zx(b);
        }
    };
    CategorySeries.prototype.bp = function () {
        var a = _super.prototype.bp.call(this);
        return a;
    };
    CategorySeries.prototype.p6 = function () {
        _super.prototype.p6.call(this);
        var a = this.fetchXAxis();
        if (a != null) {
            a.h9(false);
        }
        var b = this.fetchYAxis();
        if (b != null) {
            b.h9(false);
        }
    };
    CategorySeries.prototype.sl = function (a, b) {
        _super.prototype.sl.call(this, a, b);
        this.yu.cw.g(this.ib);
        this.rn(false);
    };
    CategorySeries.prototype.sk = function (a, b) {
        if (this.gz && this.jo < 0.05 && (this.c9 == null || !this.c9.fj())) {
            this.g0 = true;
            this.g3 = false;
            this.gi = false;
        }
        _super.prototype.sk.call(this, a, b);
        this.yu.cw.g(this.ib);
        this.rn(false);
    };
    CategorySeries.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
        switch (b) {
            case "SeriesViewer":
                var e = this.fetchYAxis();
                var f = this.fetchXAxis();
                if (c != null && d == null) {
                    this.pa(f);
                    this.pa(e);
                }
                if (c == null && d != null) {
                    this.rd(f);
                    this.rd(e);
                }
                this.yu.cw.g(this.ib);
                this.rn(false);
                break;
            case "SyncLink":
                if (this.c1 != null && this.c9 != null) {
                    this.yu.cw.g(this.ib);
                }
                this.rn(false);
                break;
            case "TransitionProgress":
                this.yp.c(this.jo, this.yn, this.yl);
                if (this.d3(this.ct)) {
                    return;
                }
                if (this.jo == 1) {
                    this.yl.w();
                    this.z4(this.yl, this.yu);
                    if (this.gz) {
                        this.gz = false;
                        this.o4();
                        this.da.r = truncate(this.i7());
                        this.da.b = this.db;
                    }
                    if (this.g2) {
                        this.g2 = false;
                        this.o4();
                        this.da.r = truncate(this.i7());
                        this.da.b = this.db;
                        this.qw();
                    }
                    this.qk();
                }
                else {
                    this.z4(this.yp, this.yu);
                }
                if (this.c9 != null) {
                    this.c9.qa(this);
                }
                break;
            case "ErrorBarSettings":
                this.qh();
                break;
            case CategorySeries.$$p[4]:
                this.rn(false);
                break;
            case CategorySeries.$$p[1]:
                this.y7 = this.y9();
                this.rn(false);
                break;
            case CategorySeries.$$p[0]:
                this.y6 = this.y8();
                this.rn(false);
                break;
        }
    };
    CategorySeries.prototype.y9 = function () {
        return this.zb;
    };
    CategorySeries.prototype.y8 = function () {
        return this.za;
    };
    CategorySeries.prototype.zk = function (a) {
        return this.ir(a, this.fetchXAxis());
    };
    CategorySeries.prototype.get_fq = function () {
        return typeCast(ISortingAxis_$type, this.yk()) !== null;
    };
    CategorySeries.prototype.az = function () {
        return this.yk();
    };
    CategorySeries.prototype.j3 = function (a) {
        return this.yz.b$d(this, a);
    };
    CategorySeries.prototype.kl = function (a) {
        return this.yz.b$f(this, a);
    };
    CategorySeries.prototype.j1 = function (a) {
        var b = truncate(Math.round(this.ip(a)));
        return b;
    };
    CategorySeries.prototype.ip = function (a) {
        return this.yz.b$c(this, a);
    };
    CategorySeries.prototype.df = function (a, b, c) {
        if (this.fx(this.bv) && !c) {
            var d = this.y3(a, b, this.y5);
            this.y5 = d;
            return d;
        }
        else {
            var e = this.y1(a, b, this.y5);
            e.d = true;
            this.y5 = e;
            return e;
        }
    };
    CategorySeries.prototype.get_fc = function () {
        return true;
    };
    CategorySeries.prototype.o1 = function () {
        _super.prototype.o1.call(this);
        this.y4 = null;
        this.y5 = null;
    };
    CategorySeries.prototype.de = function (a, b, c) {
        if (this.eb && this.fx(this.bv) && !c) {
            var d = this.y2(a, b, this.y4);
            this.y4 = d;
            return d;
        }
        else {
            var e = this.y1(a, b, this.y4);
            this.y4 = e;
            return e;
        }
    };
    CategorySeries.prototype.zj = function () {
        return this.yu.cz.a;
    };
    CategorySeries.prototype.y3 = function (a, b, c) {
        if (!this.w8()) {
            return null;
        }
        var d = this.ym();
        var e = this.zj();
        return this.yz.b$b(this, e, this.zn(this.ct), this.zm(this.ct), d.f, a, b, c);
    };
    CategorySeries.prototype.y2 = function (a, b, c) {
        var d = this.zn(this.ct);
        var e = this.zo(this.ct);
        var f = this.zm(this.ct);
        var g = this.ym();
        return this.yz.b$a(this, g.f, g.m, d, e, f, a, b, c);
    };
    CategorySeries.prototype.y1 = function (a, b, c) {
        var d = new HighlightingInfo();
        d.i = this;
        d.g = 0;
        d.f = this.dl != null ? this.dl.count - 1 : 0;
        d.c = true;
        if (c != null && c.g == d.g && c.f == d.f) {
            return c;
        }
        return d;
    };
    CategorySeries.prototype.ym = function () {
        var a = this.yl;
        if (this.d1()) {
            if (this.da.g) {
                this.da.t();
            }
            a = this.yp;
        }
        return a;
    };
    CategorySeries.prototype.zz = function (a, b, c, d, e, f, g) {
        var h = this.bv != 1;
        var i = this.d0;
        var j = this.cn != 1;
        var k = this.cm != 1;
        this.yt.a3(this.yt.d, d.cn, a, b, c, d, e, f, true, g);
        this.yt.aa(this.h9);
        var l = this.yt.ax;
        var m = null;
        var n = this.dl;
        var o = null;
        var p = null;
        var q = null;
        var r = null;
        var s = null;
        if (i && this.c9 != null && !l.e) {
            var t = this.vi(l.z);
            var u = this.vj(l.z);
            this.yt.ae = t;
            this.yt.af = u;
            l.aa = t;
            l.ab = u;
        }
        if (h && this.c9 != null) {
            m = this.c9.dj.getHighlightingInfo(this, n, l.z, l.y, true);
            if (this.c9.dj.isCrossContextHighlight(this)) {
                o = this.c9.dj.getHighlightingInfo(this, n, l.z, l.y, true);
            }
            l.c = m;
        }
        if (j && this.c9 != null) {
            p = this.c9.dk.getHighlightingInfo(this, n, l.z, l.y, true);
            if (p == null && this.bl != null) {
                p = this.c9.dk.getHighlightingInfo(this.bl, this.bl.dl, l.z, l.y, true);
            }
            if (this.c9.dk.isCrossContextHighlight(this)) {
                r = this.c9.dk.getHighlightingInfo(this, n, l.z, l.y, true);
            }
            l.d = p;
        }
        if (k && this.c9 != null) {
            q = this.c9.di.getHighlightingInfo(this, n, l.z, l.y, true);
            if (this.c9.di.isCrossContextHighlight(this)) {
                s = this.c9.di.getHighlightingInfo(this, n, l.z, l.y, true);
            }
            l.b = q;
        }
        l.h = f;
        this.z2(l);
        this.yt.y(this, m, p, q, f, h, j, k, l.v, this.c9.dj.totalHighWaterMark, l.x, this.c9.dk.totalHighWaterMark, l.t, this.c9.di.totalHighWaterMark);
    };
    CategorySeries.prototype.z0 = function (a, b, c, d, e, f, g) {
        var h = this.bv != 1;
        var i = this.cn != 1;
        var j = this.cm != 1;
        var k = this.d0;
        this.yt.a3(this.yt.b, d.cn, a, b, c, d, e, f, false, g);
        this.yt.ab(this.h9);
        var l = this.yt.ay;
        var m = null;
        var n = null;
        var o = null;
        var p = this.dl;
        if (k && this.c9 != null && !l.e) {
            var q = this.vh(l.z);
            var r = this.vk(l.z);
            this.yt.ag = q;
            this.yt.ah = r;
            l.aa = q;
            l.ab = r;
        }
        if (h && this.c9 != null) {
            m = this.c9.dj.getHighlightingInfo(this, p, l.z, l.y, false);
            l.c = m;
        }
        if (i && this.c9 != null) {
            n = this.c9.dk.getHighlightingInfo(this, p, l.z, l.y, false);
            if (n == null && this.bl != null) {
                n = this.c9.dk.getHighlightingInfo(this.bl, this.bl.dl, l.z, l.y, false);
            }
            l.d = n;
        }
        if (j && this.c9 != null) {
            o = this.c9.di.getHighlightingInfo(this, p, l.z, l.y, false);
            if (o == null && this.bl != null) {
                o = this.c9.di.getHighlightingInfo(this.bl, this.bl.dl, l.z, l.y, false);
            }
            l.b = o;
        }
        l.h = f;
        this.z3(l);
        this.yt.z(this, m, n, o, f, h, i, j, b < 0);
    };
    CategorySeries.prototype.hc = function (a, b, c) {
        var d = _super.prototype.hc.call(this, a, b, c);
        d = d && this.zi(this.fetchXAxis());
        d = d && this.zi(this.fetchYAxis());
        var e = c;
        if (!c.o() || b.isEmpty || a.isEmpty || this.dl == null) {
            d = false;
        }
        if (!d) {
            e.cw.d = 0;
        }
        return d;
    };
    CategorySeries.prototype.zi = function (a) {
        if (a == null || a.bi == null) {
            return false;
        }
        var b = typeCast(CategoryAxisBase.$, a);
        if (b != null) {
            if (b.itemsSource == null) {
                return false;
            }
            if (b.lg < 1) {
                return false;
            }
        }
        else {
            var c = typeCast(NumericAxisBase.$, a);
            if (c != null) {
                return c.ly != c.lx;
            }
        }
        return true;
    };
    CategorySeries.prototype.gl = function () {
        return this.zc;
    };
    CategorySeries.prototype.gm = function () {
        return true;
    };
    CategorySeries.prototype.yv = function () {
        return 12;
    };
    CategorySeries.prototype.o4 = function () {
        _super.prototype.o4.call(this);
        this.yn.w();
        this.yl.w();
        this.yp.w();
    };
    CategorySeries.prototype.ro = function (a) {
        _super.prototype.ro.call(this, a);
        this.yu.cw.g(this.ib);
        if (this.d3(this.ct)) {
            this.yl.z();
            this.yp.z();
            this.yn.z();
            return;
        }
        if (this.gp) {
            if (this.d1()) {
                if (this.da.g) {
                    this.da.t();
                }
                this.z4(this.yp, this.yu);
            }
            else {
                this.z4(this.yl, this.yu);
            }
            return;
        }
        if (this.gd(a)) {
            var b = this.yn;
            if (this.d1()) {
                if (this.da.g) {
                    this.da.t();
                }
                this.yn = this.yp;
                this.yp = b;
            }
            else {
                this.yn = this.yl;
                this.yl = b;
            }
            var c = false;
            this.z1(this.yl, this.yu);
            if (this.g1) {
                this.da.w();
                this.da.r = truncate(this.i8());
                this.da.b = this.dc != null ? this.dc : this.db;
                if (this.g2) {
                    var d = this.yp;
                    this.yp = this.yn;
                    this.yn = d;
                    c = true;
                }
                this.yx.s(this, this.yn, this.yl, this.isVertical, this.fetchXAxis(), this.fetchYAxis(), this.yw, this.yv(), this.c3, this.c2(), runOn(this, this.zl), this.yu.bx);
                if (this.g2) {
                    var e = this.yp;
                    this.yp = this.yn;
                    this.yn = e;
                }
                this.g2 = false;
            }
            if (this.g4) {
                this.da.w();
                this.da.r = truncate(this.i9());
                this.da.b = this.dd != null ? this.dd : this.db;
                if (this.gz) {
                    var f = this.yp;
                    this.yp = this.yn;
                    this.yn = f;
                    c = true;
                }
                this.yx.s(this, this.yn, this.yl, this.isVertical, this.fetchXAxis(), this.fetchYAxis(), this.yw, this.yv(), this.c3, this.c2(), runOn(this, this.zl), this.yu.bx);
                var g = this.yl;
                this.yl = this.yn;
                this.yn = g;
                if (this.gz) {
                    var h = this.yp;
                    this.yp = this.yn;
                    this.yn = h;
                }
                this.gz = false;
            }
            this.d7 = false;
            if (!c) {
                this.oz();
            }
            this.r1();
            if (this.g1) {
                this.g0 = false;
                this.gz = true;
            }
            if (this.g4) {
                this.g3 = false;
                this.g2 = true;
            }
        }
        else {
            this.z1(this.yl, this.yu);
            this.z4(this.yl, this.yu);
        }
    };
    CategorySeries.prototype.c2 = function () {
        return 3;
    };
    CategorySeries.prototype.c5 = function () {
        var a = this.c2();
        switch (a) {
            case 0: return 0;
            case 1: return 1;
            case 4: return 4;
            case 2: return 2;
            case 3: return 3;
        }
        return 3;
    };
    CategorySeries.prototype.zl = function () {
        if (this.isVertical) {
            return this.wl.left;
        }
        return this.wl.bottom;
    };
    CategorySeries.prototype.zf = function () {
        return false;
    };
    CategorySeries.prototype.hr = function () {
        var a = _super.prototype.hr.call(this);
        if (this.zf()) {
            var b = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "DashArray";
                $ret.p = true;
                $ret.x = ((function () {
                    var $ret = new DoubleCollection();
                    $ret.add(5);
                    $ret.add(5);
                    return $ret;
                })());
                $ret.q = true;
                return $ret;
            })());
            a.add(b);
        }
        if (!this.zf()) {
            var c = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Brush";
                $ret.p = true;
                $ret.y = "ActualBrush";
                $ret.q = true;
                return $ret;
            })());
            c.valueResolving = delegateCombine(c.valueResolving, runOn(this, this.rv));
            a.add(c);
        }
        if (!this.zf()) {
            var d = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Outline";
                $ret.p = true;
                $ret.y = "ActualOutline";
                $ret.q = true;
                return $ret;
            })());
            d.valueResolving = delegateCombine(d.valueResolving, runOn(this, this.rv));
            a.add(d);
        }
        var e = ((function () {
            var $ret = new SeriesLayerPropertyOverlay();
            $ret.p = true;
            $ret.ad = "IsTransitionInEnabled";
            $ret.x = true;
            $ret.q = true;
            return $ret;
        })());
        a.add(e);
        var f = ((function () {
            var $ret = new SeriesLayerPropertyOverlay();
            $ret.p = true;
            $ret.ad = "TransitionInMode";
            $ret.x = enumGetBox(CategoryTransitionInMode_$type, 20);
            $ret.q = true;
            return $ret;
        })());
        a.add(f);
        return a;
    };
    CategorySeries.prototype.kt = function () {
        if (this.d1()) {
            return this.bl.yp;
        }
        else {
            return this.bl.yl;
        }
    };
    CategorySeries.prototype.zp = function () {
        var e_1, _a;
        if (this.bl != null) {
            var a = this.bl;
            return a.zp();
        }
        var b = 0;
        var c = this.fetchXAxis();
        try {
            for (var _b = __values(fromEnum(this.c9.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var d = _c.value;
                if (d == this) {
                    return b;
                }
                var e = typeCast(CategorySeries.$, d);
                if (e != null) {
                    var f = typeCast(CategoryAxisBase.$, e.fetchXAxis());
                    if (f == c && e.preferredCategoryMode(f) == 2) {
                        b++;
                    }
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        return -1;
    };
    Object.defineProperty(CategorySeries.prototype, "GetErrorBarsXAxis", {
        get: function () {
            return this.fetchXAxis();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeries.prototype, "GetErrorBarsYAxis", {
        get: function () {
            return this.fetchYAxis();
        },
        enumerable: false,
        configurable: true
    });
    CategorySeries.prototype.zy = function (a, b) {
        this.yt.a1(this, this.ze(), runOn(this, this.ye), this.zm(a), this.zn(a));
        var c = this.yt.d;
        var d = c != null;
        var e = this.yk();
        if (e == null) {
            return;
        }
        var f = this.getEffectiveViewport1(a);
        var g = new ScalerParams(0, a.by, a.bx, e.ch, f);
        var h = this.dl == null ? 0 : this.dl.count;
        var i = a;
        var j = b.f;
        var k = i.cw.e;
        var l = i.cw.f;
        var m = i.cw.d;
        var n = k * m;
        for (var o = 0; o < a.cz.count; o++) {
            var p = a.cz.item(o);
            var q = p.content;
            if (d) {
                var r = p.al;
                this.zz(j, r, h, e, g, a.q, p.ak);
            }
            this.yt.ac(p, q);
        }
    };
    CategorySeries.prototype.z4 = function (a, b) {
        this.qt();
        this.wc = a.ab;
        b.aw();
    };
    CategorySeries.prototype.z1 = function (a, b) {
        a.v();
    };
    CategorySeries.prototype.zm = function (a) {
        return a.cw.d;
    };
    CategorySeries.prototype.zn = function (a) {
        return a.cw.e;
    };
    CategorySeries.prototype.zo = function (a) {
        return a.cw.f;
    };
    CategorySeries.prototype.zx = function (a) {
        var b = a;
        b.cz.count = 0;
    };
    CategorySeries.prototype.rp = function (a, b) {
        _super.prototype.rp.call(this, a, b);
        if (!this.gy) {
            this.ct.a4(b);
            return;
        }
        var c = typeCast(CategorySeriesView.$, this.cs);
        c.cw.g(this.ib);
        this.ct.a4(b);
        if (this.d3(this.cs)) {
            return;
        }
        this.z5();
        this.gy = false;
    };
    CategorySeries.prototype.z5 = function () {
        var a = typeCast(CategorySeriesView.$, this.cs);
        if (!this.gq) {
            this.z1(this.yo, a);
        }
        this.gq = false;
        this.z4(this.yo, a);
    };
    CategorySeries.prototype.yr = function (a) {
        var b = typeCast(CategorySeriesView.$, a);
        if (b != null && b.q) {
            if (b.p) {
                return new CategoryFramePreparer(1, typeCast(IIsCategoryBased_$type, this), typeCast(ISupportsMarkers_$type, b), b, this, b.cw);
            }
            else {
                return new CategoryFramePreparer(1, typeCast(IIsCategoryBased_$type, this), typeCast(ISupportsMarkers_$type, b), this.c9.c8.o, this, b.cw);
            }
        }
        else {
            return this.yq;
        }
    };
    CategorySeries.prototype.av = function (a) {
        if (this.fetchXAxis() == a) {
            return this.fetchYAxis();
        }
        if (this.fetchYAxis() == a) {
            return this.fetchXAxis();
        }
        return null;
    };
    CategorySeries.prototype.ai = function () {
        return [this.fetchXAxis(), this.fetchYAxis()];
    };
    CategorySeries.yf = function (a, b, c, d, e) {
        var f = null;
        if (b > -1 && !d.isEmpty && !c.isEmpty) {
            var g = a._inner[0][0] < c.left - 2000 ? c.left - 10 : NaN;
            var h = c.bottom + 10;
            var i = a._inner[b][0] > c.right + 2000 ? c.right + 10 : NaN;
            var j = c.top - 10;
            f = new Clipper(1, g, h, i, j, e);
        }
        return f;
    };
    CategorySeries.$t = markType(CategorySeries, 'CategorySeries', MarkerSeries.$, [IHasCategoryModePreference_$type, ISupportsErrorBars_$type]);
    CategorySeries.$$p = markDep(DependencyProperty, PropertyMetadata, CategorySeries, 'raisePropertyChanged', ['IsCustomCategoryMarkerStyleAllowed:za:z6', [0, false], 'IsCustomCategoryStyleAllowed:zb:z7', [0, false], 'IsTransitionInEnabled:zc:z8', [0, false], 'TransitionInMode:yw:z9', [CategoryTransitionInMode_$type, enumGetBox(CategoryTransitionInMode_$type, 0)], 'UseHighMarkerFidelity:zh:aaa', [0, false]]);
    return CategorySeries;
}(MarkerSeries));
export { CategorySeries };
/**
 * @hidden
 */
var CategorySeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategorySeriesView, _super);
    function CategorySeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.cy = null;
        _this.c1 = 0;
        _this.cw = null;
        _this.cz = null;
        _this.c1 = -1;
        _this.cy = a;
        _this.cw = _this.cx();
        _this.cz = ((function () {
            var $ret = new Pool$1(Marker.$);
            $ret.create = runOn(_this, _this.b3);
            $ret.activate = runOn(_this, _this.co);
            $ret.disactivate = runOn(_this, _this.cq);
            $ret.destroy = runOn(_this, _this.cp);
            return $ret;
        })());
        return _this;
    }
    CategorySeriesView.prototype.c0 = function (a) {
        if (this.c1 != a.u) {
            return true;
        }
        return false;
    };
    CategorySeriesView.prototype.c3 = function (a) {
        this.c1 = a.u;
    };
    CategorySeriesView.prototype.updateMarkerTemplate = function (a, b, c) {
        if (!this.b4.xa) {
            var d = this.cz.item(a);
            var e = d.content;
            e.item = this.e.dl.item(b);
            d.ak = b;
            d.al = c;
        }
    };
    CategorySeriesView.prototype.cx = function () {
        return new CategoryBucketCalculator(this);
    };
    Object.defineProperty(CategorySeriesView.prototype, "shouldDisplayMarkers", {
        get: function () {
            return this.cy.w8();
        },
        enumerable: false,
        configurable: true
    });
    CategorySeriesView.prototype.updateMarkerCount = function (a) {
        this.cz.count = a;
    };
    CategorySeriesView.prototype.cj = function (a) {
        this.cz.c(a);
    };
    CategorySeriesView.prototype.c2 = function () {
    };
    CategorySeriesView.$t = markType(CategorySeriesView, 'CategorySeriesView', MarkerSeriesView.$, [ISupportsMarkers_$type]);
    return CategorySeriesView;
}(MarkerSeriesView));
export { CategorySeriesView };
