// igniteui-angular-charts/esm5/lib/ChaikinVolatilityIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ChaikinVolatilityIndicatorStrategy } from "./ChaikinVolatilityIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var ChaikinVolatilityIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ChaikinVolatilityIndicator, _super);
    function ChaikinVolatilityIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChaikinVolatilityIndicator.prototype.bi = function () {
        return new ChaikinVolatilityIndicator();
    };
    ChaikinVolatilityIndicator.prototype.get_abn = function () {
        return new ChaikinVolatilityIndicatorStrategy();
    };
    Object.defineProperty(ChaikinVolatilityIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    ChaikinVolatilityIndicator.prototype.get_ab0 = function () {
        return ChaikinVolatilityIndicator.$;
    };
    Object.defineProperty(ChaikinVolatilityIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ChaikinVolatilityIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(ChaikinVolatilityIndicator.ab9));
        },
        set: function (a) {
            this.h(ChaikinVolatilityIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    ChaikinVolatilityIndicator.prototype.abt = function () {
        return this.ab8;
    };
    ChaikinVolatilityIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    ChaikinVolatilityIndicator.$t = markType(ChaikinVolatilityIndicator, 'ChaikinVolatilityIndicator', StrategyBasedIndicator.$);
    ChaikinVolatilityIndicator.ab9 = StrategyBasedIndicator.ab5(10, ChaikinVolatilityIndicator.$);
    return ChaikinVolatilityIndicator;
}(StrategyBasedIndicator));
export { ChaikinVolatilityIndicator };
