// igniteui-angular-charts/esm5/lib/ChaikinOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { ChaikinOscillatorIndicatorStrategy } from "./ChaikinOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var ChaikinOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ChaikinOscillatorIndicator, _super);
    function ChaikinOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ChaikinOscillatorIndicator.prototype.bi = function () {
        return new ChaikinOscillatorIndicator();
    };
    ChaikinOscillatorIndicator.prototype.get_abn = function () {
        return new ChaikinOscillatorIndicatorStrategy();
    };
    Object.defineProperty(ChaikinOscillatorIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    ChaikinOscillatorIndicator.prototype.get_ab0 = function () {
        return ChaikinOscillatorIndicator.$;
    };
    Object.defineProperty(ChaikinOscillatorIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ChaikinOscillatorIndicator.prototype, "shortPeriod", {
        get: function () {
            return typeGetValue(this.c(ChaikinOscillatorIndicator.ab9));
        },
        set: function (a) {
            this.h(ChaikinOscillatorIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    ChaikinOscillatorIndicator.prototype.abv = function () {
        return this.shortPeriod;
    };
    Object.defineProperty(ChaikinOscillatorIndicator.prototype, "longPeriod", {
        get: function () {
            return typeGetValue(this.c(ChaikinOscillatorIndicator.ab8));
        },
        set: function (a) {
            this.h(ChaikinOscillatorIndicator.ab8, a);
        },
        enumerable: false,
        configurable: true
    });
    ChaikinOscillatorIndicator.prototype.abs = function () {
        return this.longPeriod;
    };
    ChaikinOscillatorIndicator.$t = markType(ChaikinOscillatorIndicator, 'ChaikinOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    ChaikinOscillatorIndicator.ab9 = StrategyBasedIndicator.ab7(3, ChaikinOscillatorIndicator.$);
    ChaikinOscillatorIndicator.ab8 = StrategyBasedIndicator.ab3(10, ChaikinOscillatorIndicator.$);
    return ChaikinOscillatorIndicator;
}(StrategyBasedIndicator));
export { ChaikinOscillatorIndicator };
