// igniteui-angular-charts/esm5/lib/CommodityChannelIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { CommodityChannelIndexIndicatorStrategy } from "./CommodityChannelIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var CommodityChannelIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CommodityChannelIndexIndicator, _super);
    function CommodityChannelIndexIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CommodityChannelIndexIndicator.prototype.bi = function () {
        return new CommodityChannelIndexIndicator();
    };
    CommodityChannelIndexIndicator.prototype.get_abn = function () {
        return new CommodityChannelIndexIndicatorStrategy();
    };
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    CommodityChannelIndexIndicator.prototype.get_ab0 = function () {
        return CommodityChannelIndexIndicator.$;
    };
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CommodityChannelIndexIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(CommodityChannelIndexIndicator.ab9));
        },
        set: function (a) {
            this.h(CommodityChannelIndexIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    CommodityChannelIndexIndicator.prototype.abt = function () {
        return this.ab8;
    };
    CommodityChannelIndexIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    CommodityChannelIndexIndicator.$t = markType(CommodityChannelIndexIndicator, 'CommodityChannelIndexIndicator', StrategyBasedIndicator.$);
    CommodityChannelIndexIndicator.ab9 = StrategyBasedIndicator.ab5(20, CommodityChannelIndexIndicator.$);
    return CommodityChannelIndexIndicator;
}(StrategyBasedIndicator));
export { CommodityChannelIndexIndicator };
