// igniteui-angular-charts/esm5/lib/CrosshairLayer_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { AnnotationLayer } from "./AnnotationLayer";
import { Brush } from "igniteui-angular-core";
import { DependencyProperty } from "igniteui-angular-core";
import { Series } from "./Series";
import { Thickness } from "igniteui-angular-core";
import { Visibility_$type } from "igniteui-angular-core";
import { CrosshairLayerFrame } from "./CrosshairLayerFrame";
import { List$1 } from "igniteui-angular-core";
import { Axis } from "./Axis";
import { Dictionary$2 } from "igniteui-angular-core";
import { Pool$1 } from "igniteui-angular-core";
import { AxisAnnotation } from "./AxisAnnotation";
import { Base, Number_$type, enumGetBox, fromEnum, Point_$type, markType, markDep, runOn } from "igniteui-angular-core";
import { AxisAnnotationFrameManager } from "./AxisAnnotationFrameManager";
import { Tuple$2 } from "igniteui-angular-core";
import { DoubleCollection } from "igniteui-angular-core";
import { DeviceUtils } from "igniteui-angular-core";
import { Path } from "igniteui-angular-core";
import { PathGeometry } from "igniteui-angular-core";
import { PathFigure } from "igniteui-angular-core";
import { LineSegment } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { isNaN_, truncate } from "igniteui-angular-core";
import { AnnotationLayerView } from "./AnnotationLayerView";
import { Defaults } from "./Defaults";
import { TranslateTransform } from "igniteui-angular-core";
import { LineGeometry } from "igniteui-angular-core";
import { BrushUtil } from "igniteui-angular-core";
import { Color } from "igniteui-angular-core";
/**
 * @hidden
 */
var CrosshairLayer = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CrosshairLayer, _super);
    function CrosshairLayer() {
        var _this = _super.call(this) || this;
        _this.xAxisAnnotationFormatLabel = null;
        _this.yAxisAnnotationFormatLabel = null;
        _this.acc = null;
        _this.zz = null;
        _this.zy = null;
        var a = new CrosshairLayerFrame();
        var b = new CrosshairLayerFrame();
        var c = new CrosshairLayerFrame();
        var d = _this.kb / 1000;
        _this.w4 = a;
        _this.w3 = b;
        _this.w5 = c;
        _this.ab = CrosshairLayer.$;
        return _this;
    }
    CrosshairLayer.prototype.bi = function () {
        return new CrosshairLayer();
    };
    CrosshairLayer.prototype.get_ep = function () {
        return true;
    };
    CrosshairLayer.prototype.get_en = function () {
        return true;
    };
    CrosshairLayer.prototype.cq = function () {
        return new CrosshairLayerView(this);
    };
    CrosshairLayer.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.zr = a;
    };
    CrosshairLayer.prototype.gs = function (a) {
        if (this.zs == null) {
            return true;
        }
        if (this.zs == a) {
            return true;
        }
        return false;
    };
    CrosshairLayer.prototype.zt = function () {
        return this.xAxisAnnotationFormatLabel != null;
    };
    CrosshairLayer.prototype.zu = function () {
        return this.yAxisAnnotationFormatLabel != null;
    };
    Object.defineProperty(CrosshairLayer.prototype, "ace", {
        get: function () {
            return new Thickness(1, this.z2, this.z4, this.z3, this.z1);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CrosshairLayer.prototype, "acf", {
        get: function () {
            return new Thickness(1, this.z8, this.aaa, this.z9, this.z7);
        },
        enumerable: false,
        configurable: true
    });
    CrosshairLayer.prototype.yh = function (a, b) {
        if (this.zs == null || this.zs == a) {
            if (!this.a0.c) {
                this.rn(b);
            }
        }
    };
    CrosshairLayer.prototype.yj = function (a, b) {
        var e_1, _a;
        var _this = this;
        _super.prototype.yj.call(this, a, b);
        var c = a;
        var d = this.zx;
        var e = this.zw;
        c.af.clear();
        c.ag.clear();
        c.ae.clear();
        c.ah.clear();
        AxisAnnotationFrameManager.b(c);
        if (this.xa(this.zs)) {
            this.aa9(this.zs, c, d, e);
        }
        else {
            try {
                for (var _b = __values(fromEnum(this.c9.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var f = _c.value;
                    if (this.xa(f)) {
                        this.aa9(f, c, d, e);
                    }
                    if (f.isStacked) {
                        f.pz(function (g) {
                            if (_this.xa(g)) {
                                _this.aa9(g, c, d, e);
                            }
                            if (g.ee) {
                                g.px(function (h) {
                                    if (_this.xa(h)) {
                                        _this.aa9(h, c, d, e);
                                    }
                                    return true;
                                });
                            }
                            return true;
                        });
                    }
                    if (f.ee) {
                        f.px(function (g) {
                            if (_this.xa(g)) {
                                _this.aa9(g, c, d, e);
                            }
                            return true;
                        });
                    }
                }
            }
            catch (e_1_1) {
                e_1 = { error: e_1_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_1)
                        throw e_1.error;
                }
            }
        }
    };
    CrosshairLayer.prototype.xa = function (a) {
        if (!_super.prototype.xa.call(this, a)) {
            return false;
        }
        if (this.aai != null) {
            if (a.e9) {
                var b = a.na();
                if (!Base.equalsStatic(this.aai, b)) {
                    return false;
                }
            }
            else {
                if (!Base.equalsStatic(this.aai, a.name)) {
                    return false;
                }
            }
        }
        var c = this.yc(a);
        if (!c.c) {
            return false;
        }
        return true;
    };
    CrosshairLayer.prototype.ql = function () {
        _super.prototype.ql.call(this);
        if (!this.a0.c && !this.d1()) {
            return;
        }
        if (this.d1()) {
            this.aba(this.w5, false);
        }
        else {
            this.aba(this.w3, false);
        }
    };
    CrosshairLayer.prototype.wu = function () {
        return 1;
    };
    CrosshairLayer.prototype.xh = function () {
        return 0.1;
    };
    CrosshairLayer.prototype.xk = function () {
        return -0.4;
    };
    CrosshairLayer.prototype.xj = function () {
        return 0.5;
    };
    CrosshairLayer.prototype.zg = function () {
        return this.wr == 3 ? this.acd() : null;
    };
    CrosshairLayer.prototype.zh = function () {
        return this.ws == 3 ? this.acd() : null;
    };
    CrosshairLayer.prototype.acd = function () {
        if (this.acc == null) {
            this.acc = new DoubleCollection();
            this.acc.add(DeviceUtils.g(5));
            this.acc.add(DeviceUtils.g(5));
        }
        return this.acc;
    };
    CrosshairLayer.prototype.aa9 = function (a, b, c, d) {
        var e = this.yc(a);
        if (!e.c) {
            return;
        }
        var f = e.d;
        f = a.v6(f, c, d);
        var g = false;
        var h = this.u3;
        if (h == null) {
            h = this.u9;
        }
        if (h == null) {
            h = a.u3;
            g = true;
        }
        var i = false;
        var j = this.ab6;
        if (j == null) {
            if (this.u9 != null) {
                j = this.u9;
            }
            else {
                j = a.u3;
                i = true;
            }
        }
        var k = false;
        var l = this.ab9;
        if (l == null) {
            if (this.u9 != null) {
                l = this.u9;
            }
            else {
                l = a.u3;
                k = true;
            }
        }
        var m = h;
        var n = h;
        var o = j;
        var p = l;
        if (this.ab4 != null) {
            m = this.ab4;
        }
        else if (g) {
            switch (this.wr) {
                case 0:
                case 1:
                    m = this.zr.ce(m, this.xe);
                    break;
                case 2:
                    m = this.zr.cg(m, this.xe);
                    break;
                case 4:
                    m = this.zr.cf(m, this.xe);
                    break;
            }
        }
        if (this.ab5 != null) {
            n = this.ab5;
        }
        else if (g) {
            switch (this.ws) {
                case 0:
                case 1:
                    n = this.zr.ce(n, this.xg);
                    break;
                case 2:
                    n = this.zr.cg(n, this.xg);
                    break;
                case 4:
                    n = this.zr.cf(n, this.xg);
                    break;
            }
        }
        if (i) {
            switch (this.ws) {
                case 1:
                    o = this.zr.ce(o, this.xg);
                    break;
                case 2:
                    o = this.zr.cg(o, this.xg);
                    break;
            }
        }
        if (k) {
            switch (this.wr) {
                case 1:
                    p = this.zr.ce(p, this.xe);
                    break;
                case 2:
                    p = this.zr.cg(p, this.xe);
                    break;
            }
        }
        b.ae.add(m);
        b.ah.add(n);
        b.af.add(f.y);
        b.ag.add(f.x);
        if (this.zv && !isNaN_(f.y)) {
            AxisAnnotationFrameManager.c(b, f, a, this.c9, this.getEffectiveViewport(), h, o, this.z0, this.ab8, this.ab7, this.ace, this.z5, p, this.z6, this.acb, this.aca, this.acf, this.aab, this.ach == 0, this.acg == 0);
        }
    };
    CrosshairLayer.prototype.aba = function (a, b) {
        var c = false;
        if (this.zy == null) {
            c = true;
            this.zz = new List$1(Axis.$, 0);
            this.zy = new Dictionary$2(Axis.$, Pool$1.$.specialize(AxisAnnotation.$), 0);
        }
        AxisAnnotationFrameManager.d(this, a, c, this.zy, this.zz, b, this.aac, this.aad, this.xAxisAnnotationFormatLabel, this.yAxisAnnotationFormatLabel);
    };
    CrosshairLayer.prototype.aa8 = function (a, b, c, d, e) {
        var f = a.item(b);
        b++;
        if (e) {
            f._stroke = d.ah._inner[c];
            f._fill = d.ah._inner[c];
        }
        else {
            f._stroke = d.ae._inner[c];
            f._fill = d.ae._inner[c];
        }
        f.ad = this.ic;
        f.ai = e ? this.zf : this.ze;
        f.am = this.vz;
        f.al = this.vy;
        f.ak = this.vy;
        f.aj = this.vx;
        f.ac = this.jf;
        return new Tuple$2(Path.$, Number_$type, f, b);
    };
    CrosshairLayer.prototype.ym = function (a, b) {
        _super.prototype.ym.call(this, a, b);
        var c = this.zr.b8;
        var d = a;
        var e = d.af.count;
        var f = 0;
        var g = b.bx.left;
        var h = b.bx.right;
        var i = b.bx.top;
        var j = b.bx.bottom;
        var k = this.ach == 0;
        var l = this.acg == 0;
        for (var m = 0; m < e; m++) {
            var n = Math.floor(d.af._inner[m]);
            var o = Math.floor(d.ag._inner[m]);
            var p = null;
            if (k) {
                if (isNaN_(n)) {
                    continue;
                }
                var q = this.aa8(c, f, m, d, true);
                p = q.c;
                f = q.d;
                var r = new PathGeometry();
                var s = new PathFigure();
                s._startPoint = { $type: Point_$type, x: o, y: i };
                var t = new LineSegment(1);
                t.c = { $type: Point_$type, x: o, y: j };
                s._segments.add(t);
                r.c.add(s);
                p.an = r;
            }
            if (l) {
                if (isNaN_(n)) {
                    continue;
                }
                var u = this.aa8(c, f, m, d, false);
                p = u.c;
                f = u.d;
                var v = new PathGeometry();
                var w = new PathFigure();
                w._startPoint = { $type: Point_$type, x: g, y: n };
                var x = new LineSegment(1);
                x.c = { $type: Point_$type, x: h, y: n };
                w._segments.add(x);
                v.c.add(w);
                p.an = v;
            }
        }
        c.count = f;
        if (this.d1()) {
            this.aba(this.w5, true);
        }
        else {
            this.aba(this.w3, false);
        }
    };
    CrosshairLayer.prototype.f8 = function () {
        return true;
    };
    CrosshairLayer.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
        switch (b) {
            case "SeriesViewer":
                if (d == null) {
                    if (this.zy != null && this.zy.count > 0) {
                        var e = void 0;
                        if (this.d1()) {
                            e = this.w5;
                        }
                        else {
                            e = this.w3;
                        }
                        AxisAnnotationFrameManager.b(e);
                        AxisAnnotationFrameManager.d(this, e, true, this.zy, this.zz, true, this.aac, this.aad, this.xAxisAnnotationFormatLabel, this.yAxisAnnotationFormatLabel);
                    }
                }
                break;
            case CrosshairLayer.$$p[1]:
                this.rn(true);
                break;
            case CrosshairLayer.$$p[8]:
                this.rn(true);
                break;
            case CrosshairLayer.$$p[4]:
            case CrosshairLayer.$$p[5]:
                this.rn(true);
                break;
            case CrosshairLayer.$$p[2]:
            case CrosshairLayer.$$p[10]:
            case CrosshairLayer.$$p[9]:
            case CrosshairLayer.$$p[12]:
            case CrosshairLayer.$$p[14]:
            case CrosshairLayer.$$p[16]:
            case CrosshairLayer.$$p[15]:
            case CrosshairLayer.$$p[13]:
            case CrosshairLayer.$$p[17]:
            case CrosshairLayer.$$p[18]:
            case CrosshairLayer.$$p[20]:
            case CrosshairLayer.$$p[22]:
            case CrosshairLayer.$$p[24]:
            case CrosshairLayer.$$p[26]:
            case CrosshairLayer.$$p[25]:
            case CrosshairLayer.$$p[23]:
            case CrosshairLayer.$$p[27]:
            case CrosshairLayer.$$p[28]:
            case CrosshairLayer.$$p[0]:
            case CrosshairLayer.$$p[7]:
            case CrosshairLayer.$$p[6]:
                this.rn(true);
                break;
        }
    };
    CrosshairLayer.$t = markType(CrosshairLayer, 'CrosshairLayer', AnnotationLayer.$);
    CrosshairLayer.$$p = markDep(DependencyProperty, PropertyMetadata, CrosshairLayer, 'raisePropertyChanged', ['HorizontalLineStroke:ab4:abb', [Brush.$, null], 'HorizontalLineVisibility:acg:abc', [Visibility_$type, enumGetBox(Visibility_$type, 0)], 'IsAxisAnnotationEnabled:zv:abd', [0, false], 'SkipUnknownValues:zw:abe', [0, true], 'TargetSeriesName:aai:abf', [2, null], 'TargetSeries:zs:abg', [Series.$, null], 'UseInterpolation:zx:abh', [0, true], 'VerticalLineStroke:ab5:abi', [Brush.$, null], 'VerticalLineVisibility:ach:abj', [Visibility_$type, enumGetBox(Visibility_$type, 0)], 'XAxisAnnotationBackgroundCornerRadius:z0:abk', [1, DeviceUtils.g(3)], 'XAxisAnnotationBackground:ab6:abl', [Brush.$, null], 'XAxisAnnotationInterpolatedValuePrecision:aac:abm', [1, -1], 'XAxisAnnotationOutline:ab7:abn', [Brush.$, null], 'XAxisAnnotationPaddingBottom:z1:abo', [1, NaN], 'XAxisAnnotationPaddingLeft:z2:abp', [1, NaN], 'XAxisAnnotationPaddingRight:z3:abq', [1, NaN], 'XAxisAnnotationPaddingTop:z4:abr', [1, NaN], 'XAxisAnnotationStrokeThickness:z5:abs', [1, NaN], 'XAxisAnnotationTextColor:ab8:abt', [Brush.$, null], 'YAxisAnnotationBackgroundCornerRadius:z6:abu', [1, DeviceUtils.g(3)], 'YAxisAnnotationBackground:ab9:abv', [Brush.$, null], 'YAxisAnnotationInterpolatedValuePrecision:aad:abw', [1, -1], 'YAxisAnnotationOutline:aca:abx', [Brush.$, null], 'YAxisAnnotationPaddingBottom:z7:aby', [1, NaN], 'YAxisAnnotationPaddingLeft:z8:abz', [1, NaN], 'YAxisAnnotationPaddingRight:z9:ab0', [1, NaN], 'YAxisAnnotationPaddingTop:aaa:ab1', [1, NaN], 'YAxisAnnotationStrokeThickness:aab:ab2', [1, NaN], 'YAxisAnnotationTextColor:acb:ab3', [Brush.$, null]]);
    return CrosshairLayer;
}(AnnotationLayer));
export { CrosshairLayer };
/**
 * @hidden
 */
var CrosshairLayerView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CrosshairLayerView, _super);
    function CrosshairLayerView(a) {
        var _this = _super.call(this, a) || this;
        _this.b7 = null;
        _this.b8 = null;
        _this.b9 = null;
        _this.b7 = a;
        _this.b9 = new List$1(Path.$, 0);
        return _this;
    }
    CrosshairLayerView.prototype.au = function () {
        var _this = this;
        _super.prototype.au.call(this);
        this.b8 = ((function () {
            var $ret = new Pool$1(Path.$);
            $ret.create = runOn(_this, _this.ch);
            $ret.activate = runOn(_this, _this.ca);
            $ret.disactivate = runOn(_this, _this.cc);
            $ret.destroy = runOn(_this, _this.cb);
            return $ret;
        })());
        this.b7.jn = Defaults.j;
    };
    CrosshairLayerView.prototype.ch = function () {
        var a = new Path();
        a.j = new TranslateTransform();
        this.b9.add(a);
        return a;
    };
    CrosshairLayerView.prototype.ca = function (a) {
        a._visibility = 0;
    };
    CrosshairLayerView.prototype.cc = function (a) {
        a._visibility = 1;
    };
    CrosshairLayerView.prototype.cb = function (a) {
        this.b9.remove(a);
    };
    CrosshairLayerView.prototype.cd = function (a, b, c, d) {
        if (a.an == null || a.an.c.x != d || a.an.d.x != c) {
            var e = ((function () {
                var $ret = new LineGeometry();
                $ret.d = { $type: Point_$type, x: c, y: b };
                $ret.c = { $type: Point_$type, x: d, y: b };
                return $ret;
            })());
            a.an = e;
        }
    };
    CrosshairLayerView.prototype.ce = function (a, b) {
        if (a == null) {
            return a;
        }
        return BrushUtil.m(a, b);
    };
    CrosshairLayerView.prototype.cg = function (a, b) {
        if (a == null) {
            return a;
        }
        return BrushUtil.o(a, b);
    };
    CrosshairLayerView.prototype.cf = function (a, b) {
        if (a == null) {
            return a;
        }
        var c = new Brush();
        c.color = Color.u(truncate((a.color.l * b)), a.color.o, a.color.n, a.color.m);
        return c;
    };
    CrosshairLayerView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (a.d && !b) {
            for (var c = 0; c < this.b9.count; c++) {
                var d = this.b9._inner[c];
                if (d._visibility == 0) {
                    a.t(d);
                }
            }
        }
    };
    CrosshairLayerView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        for (var b = 0; b < this.b9.count; b++) {
            this.e.bb.exportPathData(a, this.b9._inner[b], "crosshairLine", ["Main", "Crosshair"]);
        }
    };
    CrosshairLayerView.$t = markType(CrosshairLayerView, 'CrosshairLayerView', AnnotationLayerView.$);
    return CrosshairLayerView;
}(AnnotationLayerView));
export { CrosshairLayerView };
