// igniteui-angular-charts/esm5/lib/DetrendedPriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { DetrendedPriceOscillatorIndicatorStrategy } from "./DetrendedPriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var DetrendedPriceOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DetrendedPriceOscillatorIndicator, _super);
    function DetrendedPriceOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DetrendedPriceOscillatorIndicator.prototype.bi = function () {
        return new DetrendedPriceOscillatorIndicator();
    };
    DetrendedPriceOscillatorIndicator.prototype.get_abn = function () {
        return new DetrendedPriceOscillatorIndicatorStrategy();
    };
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    DetrendedPriceOscillatorIndicator.prototype.get_ab0 = function () {
        return DetrendedPriceOscillatorIndicator.$;
    };
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DetrendedPriceOscillatorIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(DetrendedPriceOscillatorIndicator.ab9));
        },
        set: function (a) {
            this.h(DetrendedPriceOscillatorIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    DetrendedPriceOscillatorIndicator.prototype.abt = function () {
        return this.ab8;
    };
    DetrendedPriceOscillatorIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    DetrendedPriceOscillatorIndicator.$t = markType(DetrendedPriceOscillatorIndicator, 'DetrendedPriceOscillatorIndicator', StrategyBasedIndicator.$);
    DetrendedPriceOscillatorIndicator.ab9 = StrategyBasedIndicator.ab5(20, DetrendedPriceOscillatorIndicator.$);
    return DetrendedPriceOscillatorIndicator;
}(StrategyBasedIndicator));
export { DetrendedPriceOscillatorIndicator };
