// igniteui-angular-charts/esm5/lib/FastStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { FastStochasticOscillatorIndicatorStrategy } from "./FastStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
import { Brush } from "igniteui-angular-core";
/**
 * @hidden
 */
var FastStochasticOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FastStochasticOscillatorIndicator, _super);
    function FastStochasticOscillatorIndicator() {
        var _this = _super.call(this) || this;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 10;
        _this.trendLinePeriod = 3;
        return _this;
    }
    FastStochasticOscillatorIndicator.prototype.bi = function () {
        return new FastStochasticOscillatorIndicator();
    };
    FastStochasticOscillatorIndicator.prototype.get_abn = function () {
        return new FastStochasticOscillatorIndicatorStrategy();
    };
    Object.defineProperty(FastStochasticOscillatorIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    FastStochasticOscillatorIndicator.prototype.get_ab0 = function () {
        return FastStochasticOscillatorIndicator.$;
    };
    Object.defineProperty(FastStochasticOscillatorIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FastStochasticOscillatorIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(FastStochasticOscillatorIndicator.ab9));
        },
        set: function (a) {
            this.h(FastStochasticOscillatorIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    FastStochasticOscillatorIndicator.prototype.abt = function () {
        return this.ab8;
    };
    FastStochasticOscillatorIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    FastStochasticOscillatorIndicator.$t = markType(FastStochasticOscillatorIndicator, 'FastStochasticOscillatorIndicator', StrategyBasedIndicator.$);
    FastStochasticOscillatorIndicator.ab9 = StrategyBasedIndicator.ab5(14, FastStochasticOscillatorIndicator.$);
    return FastStochasticOscillatorIndicator;
}(StrategyBasedIndicator));
export { FastStochasticOscillatorIndicator };
