// igniteui-angular-charts/esm5/lib/FinancialPriceSeriesProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Base, markType } from "igniteui-angular-core";
import { IFinancialPriceSeriesProxy_$type } from "./IFinancialPriceSeriesProxy";
/**
 * @hidden
 */
var FinancialPriceSeriesProxy = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialPriceSeriesProxy, _super);
    function FinancialPriceSeriesProxy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FinancialPriceSeriesProxy.prototype.getCloseMemberPath = function (a) {
        return a.y2;
    };
    FinancialPriceSeriesProxy.prototype.getHighMemberPath = function (a) {
        return a.zf;
    };
    FinancialPriceSeriesProxy.prototype.getLowMemberPath = function (a) {
        return a.zk;
    };
    FinancialPriceSeriesProxy.prototype.getOpenMemberPath = function (a) {
        return a.zo;
    };
    FinancialPriceSeriesProxy.prototype.getVolumeMemberPath = function (a) {
        return a.zs;
    };
    FinancialPriceSeriesProxy.prototype.matchesType = function (a) {
        if (!a.e4) {
            return false;
        }
        if (a.e5 || a.e6) {
            return false;
        }
        return true;
    };
    FinancialPriceSeriesProxy.prototype.setCategoryMode = function (a, b) {
        a.w6 = b;
    };
    FinancialPriceSeriesProxy.prototype.setCloseMemberPath = function (a, b) {
        a.y2 = b;
    };
    FinancialPriceSeriesProxy.prototype.setDisplayType = function (a, b) {
        a.aau = b;
    };
    FinancialPriceSeriesProxy.prototype.setHighMemberPath = function (a, b) {
        a.zf = b;
    };
    FinancialPriceSeriesProxy.prototype.setLowMemberPath = function (a, b) {
        a.zk = b;
    };
    FinancialPriceSeriesProxy.prototype.setOpenMemberPath = function (a, b) {
        a.zo = b;
    };
    FinancialPriceSeriesProxy.prototype.setVolumeMemberPath = function (a, b) {
        a.zs = b;
    };
    FinancialPriceSeriesProxy.prototype.setXAxis = function (a, b) {
        a.wy = b;
    };
    FinancialPriceSeriesProxy.prototype.setYAxis = function (a, b) {
        a.xh = b;
    };
    FinancialPriceSeriesProxy.prototype.setTrendLineType = function (a, b) {
        if (a.e7) {
            a.trendLineType = b;
        }
    };
    FinancialPriceSeriesProxy.prototype.getTrendLineType = function (a) {
        if (a.e7) {
            return a.trendLineType;
        }
        return 0;
    };
    FinancialPriceSeriesProxy.$t = markType(FinancialPriceSeriesProxy, 'FinancialPriceSeriesProxy', Base.$, [IFinancialPriceSeriesProxy_$type]);
    return FinancialPriceSeriesProxy;
}(Base));
export { FinancialPriceSeriesProxy };
