// igniteui-angular-charts/esm5/lib/FinancialChart_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __read, __spreadArray, __values } from "tslib";
import { Brush } from "igniteui-angular-core";
import { XYChart } from "./XYChart";
import { IndicatorPane } from "./IndicatorPane";
import { Base, BaseError, EventArgs, runOn, delegateRemove, delegateCombine, fromEnum, typeCast, EnumUtil, Number_$type, String_$type, enumGetBox, markType, TypeRegistrar, fromEn } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { DataSeriesType_$type } from "igniteui-angular-core";
import { IDataSeriesAdapterRule_$type } from "igniteui-angular-core";
import { SimpleFinancialPriceSeriesRule } from "igniteui-angular-core";
import { SubCollectionsRule } from "igniteui-angular-core";
import { XamDataChart } from "./XamDataChart";
import { FinancialChartVolumeType_$type } from "./FinancialChartVolumeType";
import { FinancialChartXAxisMode_$type } from "./FinancialChartXAxisMode";
import { FinancialChartYAxisMode_$type } from "./FinancialChartYAxisMode";
import { FinancialChartType_$type } from "./FinancialChartType";
import { NumericYAxis } from "./NumericYAxis";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { FinancialChartZoomSliderType_$type } from "./FinancialChartZoomSliderType";
import { BrushCollection } from "igniteui-angular-core";
import { FinancialChartRangeSelectorOptionCollection } from "./FinancialChartRangeSelectorOptionCollection";
import { ObservableCollection$1 } from "igniteui-angular-core";
import { TemplatedHtmlControl } from "./TemplatedHtmlControl";
import { FinancialChartContext } from "./FinancialChartContext";
import { FinancialChartToolbarContext } from "./FinancialChartToolbarContext";
import { FinancialChartViewModel } from "./FinancialChartViewModel";
import { FinancialChartToolbarViewModel } from "./FinancialChartToolbarViewModel";
import { AxisDefaults } from "./AxisDefaults";
import { Defaults } from "./Defaults";
import { BrushUtil } from "igniteui-angular-core";
import { Series } from "./Series";
import { Rect } from "igniteui-angular-core";
import { Axis } from "./Axis";
import { AxisLabelSettings } from "./AxisLabelSettings";
import { MarkerSeries } from "./MarkerSeries";
import { IZoomableCategoryAxis_$type } from "./IZoomableCategoryAxis";
import { SyncSettings } from "./SyncSettings";
import { SyncManager } from "./SyncManager";
import { FinancialOverlayType_$type } from "./FinancialOverlayType";
import { ArrayExtension } from "igniteui-angular-core";
import { FinancialIndicatorType_$type } from "./FinancialIndicatorType";
import { IndicatorDisplayType_$type } from "./IndicatorDisplayType";
import { ITimeRangeSelectorAxis_$type } from "./ITimeRangeSelectorAxis";
import { FinancialChartVisualData } from "./FinancialChartVisualData";
import { FinancialChartCustomIndicatorArgs } from "./FinancialChartCustomIndicatorArgs";
import { ScalerParams } from "./ScalerParams";
import { stringReplace, stringJoin, stringIsNullOrEmpty } from "igniteui-angular-core";
import { timeSpanFromDays } from "igniteui-angular-core";
import { dateMaxValue, dateMinValue } from "igniteui-angular-core";
import { isNaN_, isInfinity } from "igniteui-angular-core";
import { FinancialChartTypePicker } from "./FinancialChartTypePicker";
import { FinancialChartIndicatorMenu } from "./FinancialChartIndicatorMenu";
import { TrendLineType_$type } from "igniteui-angular-core";
import { FinancialChartIndicatorMenuContext } from "./FinancialChartIndicatorMenuContext";
import { FinancialChartTypePickerContext } from "./FinancialChartTypePickerContext";
import { FinancialChartRangeSelectorContext } from "./FinancialChartRangeSelectorContext";
import { FinancialChartToolbarVisualData } from "./FinancialChartToolbarVisualData";
import { FinancialChartRangeSelectorViewModel } from "./FinancialChartRangeSelectorViewModel";
import { FinancialChartRangeSelectorView } from "./FinancialChartRangeSelectorView";
/**
 * @hidden
 */
var FinancialChart = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialChart, _super);
    function FinancialChart() {
        var _this = _super.call(this) || this;
        _this.ah9 = null;
        _this.ak7 = null;
        _this.ak5 = null;
        _this.ak2 = null;
        _this.ak4 = null;
        _this.ak6 = null;
        _this.ak3 = null;
        _this.ahf = null;
        _this.alp = null;
        _this.alq = null;
        _this.alm = null;
        _this.alr = null;
        _this.als = null;
        _this.aln = null;
        _this.alo = null;
        _this.aix = null;
        _this.aj7 = 40;
        _this.ah3 = null;
        _this.ah1 = null;
        _this.ah5 = null;
        _this.aia = null;
        _this.al0 = null;
        _this.ait = null;
        _this.ajj = false;
        _this.aji = false;
        _this.ajq = false;
        _this.ai5 = false;
        _this.ald = 10;
        _this.ake = NaN;
        _this.akg = NaN;
        _this.akf = NaN;
        _this.akh = 0;
        _this.ahl = 0;
        _this.aho = 0;
        _this.aj9 = 3;
        _this.akc = NaN;
        _this.aka = NaN;
        _this.akb = NaN;
        _this.akd = NaN;
        _this.ahq = 0;
        _this.ai0 = true;
        _this.ahh = 0;
        _this.ai1 = false;
        _this.ahv = null;
        _this.ahx = null;
        _this.ajs = null;
        _this.ai3 = true;
        _this.ai2 = false;
        _this.ai4 = false;
        _this.aiu = null;
        _this.ahs = 1;
        _this.agj = null;
        _this.agk = null;
        _this.agl = null;
        _this.agm = null;
        _this.ago = null;
        _this.agn = null;
        _this.agi = null;
        _this.agh = null;
        _this.aj8 = 1;
        _this.aj6 = 1;
        _this.aj5 = 2;
        _this.aj4 = 1;
        _this.ah7 = null;
        _this.ak9 = -2147483648;
        _this.aj3 = 2;
        _this.alc = 3;
        _this.ala = 10;
        _this.ak8 = 30;
        _this.alb = 9;
        _this.ahc = null;
        _this.ahd = null;
        _this.ag8 = null;
        _this.applyCustomIndicators = null;
        _this.apu = null;
        _this.aki = 1;
        _this.aiz = false;
        _this.ahz = null;
        _this.aj0 = dateMinValue();
        _this.ajz = dateMaxValue();
        _this.ajx = null;
        _this.ue();
        _this.ajx = new ObservableCollection$1(Base.$, 0);
        if (_this.xAxisBreaks != null) {
            var a = _this.xAxisBreaks;
            a.collectionChanged = delegateCombine(a.collectionChanged, function (a, b) { return _this.apj(); });
        }
        _this.ao4();
        return _this;
    }
    FinancialChart.prototype.vp = function (a, b, c) {
        _super.prototype.vp.call(this, a, b, c);
        switch (a) {
            case "ToolbarTemplate":
            case "RangeSelectorTemplate":
                this.ahg = new FinancialChartToolbar(this, this.alx);
                break;
            case "Toolbar":
                var d = b;
                if (d != null) {
                    d.chartTypeChanged = delegateRemove(d.chartTypeChanged, runOn(this, this.aor));
                    d.trendLineTypeChanged = delegateRemove(d.trendLineTypeChanged, runOn(this, this.aou));
                    d.volumeTypeChanged = delegateRemove(d.volumeTypeChanged, runOn(this, this.aov));
                    d.l(null);
                }
                if (this.ahg != null) {
                    this.ahg.ac = this.ahn;
                    this.ahg.ai = this.ef;
                    this.ahg.y = this.ahi;
                    this.ahg.al = this.ajg;
                    if (this.ak4 != null) {
                        this.ahg.l(this.ak4);
                    }
                    var e = this.ahg;
                    e.chartTypeChanged = delegateCombine(e.chartTypeChanged, runOn(this, this.aor));
                    ;
                    var f = this.ahg;
                    f.trendLineTypeChanged = delegateCombine(f.trendLineTypeChanged, runOn(this, this.aou));
                    var g = this.ahg;
                    g.volumeTypeChanged = delegateCombine(g.volumeTypeChanged, runOn(this, this.aov));
                }
                break;
            case "ChartType":
                if (this.ahg != null) {
                    this.ahg.y = this.ahi;
                }
                break;
            case "VolumeType":
                if (this.ahg != null) {
                    this.ahg.ac = this.ahn;
                }
                break;
            case "TrendLineType":
                if (this.ahg != null) {
                    this.ahg.ai = this.ef;
                }
                break;
            case "IsToolbarVisible":
                if (this.ahg == null) {
                    break;
                }
                this.ahg.al = this.ajg;
                this.setGridRows();
                break;
        }
    };
    FinancialChart.prototype.aov = function (a, b) {
        this.ahn = this.ahg.ac;
    };
    FinancialChart.prototype.aou = function (a, b) {
        this.ef = this.ahg.ai;
    };
    FinancialChart.prototype.aor = function (a, b) {
        this.ahi = this.ahg.y;
    };
    FinancialChart.prototype.ao4 = function () {
        this.ah9 = new IndicatorPane();
        this.ah0 = TypeRegistrar.create("FinancialLegend");
    };
    FinancialChart.prototype.provideGrid = function (a) {
        this.ak7 = a;
        this.app();
        this.setGridRows();
        if (this.ahg != null) {
            this.ahg.bg(a);
        }
    };
    FinancialChart.prototype.onDocumentClick = function (a) {
        if (this.ahg != null) {
            this.ahg.bl(a);
        }
    };
    FinancialChart.prototype.getContext = function () {
        var a = this.ahg.v();
        var b = new FinancialChartContext();
        b.vm = new FinancialChartViewModel();
        b.vm.toolbar = a;
        b.container = this.lu;
        b.toolbarContainer = this.ak4;
        b.legendContainer = this.ak3;
        b.indicatorsContainer = this.ak2;
        b.volumeContainer = this.ak5;
        b.zoomContainer = this.ak6;
        return b;
    };
    FinancialChart.prototype.provideContainer = function (a) {
        var b = a;
        if (b != null) {
            var c = "financialChartMainGrid";
            var d = b.createElement("div");
            d.setStyleProperty("width", "100%");
            d.setStyleProperty("height", "100%");
            if (this.ahg == null) {
                this.ahg = new FinancialChartToolbar(this, this.alx);
            }
            this.ahg.w.isIndicatorMenuVisible = this.ai7 && this.ai8;
            this.ak4 = b.getSubRenderer(d);
            this.ahg.l(this.ak4);
            var e = b.createElement("div");
            e.setStyleProperty("width", "100%");
            e.setStyleProperty("height", "100%");
            var f = b.createElement("div");
            f.setStyleProperty("width", "100%");
            f.setStyleProperty("height", "100%");
            this.ak3 = b.getSubRenderer(f);
            var g = b.createElement("div");
            g.setStyleProperty("width", "100%");
            g.setStyleProperty("height", "100%");
            this.ak2 = b.getSubRenderer(g);
            var h = b.createElement("div");
            h.setStyleProperty("width", "100%");
            h.setStyleProperty("height", "100%");
            this.ak5 = b.getSubRenderer(h);
            var i = b.createElement("div");
            i.setStyleProperty("width", "100%");
            i.setStyleProperty("height", "100%");
            this.ak6 = b.getSubRenderer(i);
            a = b.getSubRenderer(e);
        }
        else {
            this.ak4 = null;
            if (this.ahg != null) {
                this.ahg.l(null);
            }
            if (this.aiy != null) {
                this.aiy.provideContainer(null);
                this.aiy = null;
            }
            if (this.ah0 != null) {
                this.ah0.provideContainer(null);
            }
        }
        _super.prototype.provideContainer.call(this, a);
        if (b != null) {
            if (this.volumeChart != null) {
                this.volumeChart.c8.b4(this.ak5);
            }
            if (this.ah9 != null) {
                this.ah9.h(this.ak2);
            }
            if (this.zoomSliderChart != null) {
                this.zoomSliderChart.c8.b4(this.ak6);
                if (a != null) {
                    this.aob();
                }
                else {
                    this.aiy = null;
                }
            }
            if (this.ah0 != null) {
                this.ah0.provideContainer(this.ak3);
            }
            this.setGridRows();
        }
    };
    FinancialChart.prototype.apf = function (a, b, c) {
        if (a != null) {
            a.provideContainer(null);
        }
        if (b != null) {
            b.c8.b4(c);
        }
        this.setGridRows();
    };
    FinancialChart.prototype.apt = function (a, b) {
        this.apf(a, b, this.ak6);
        if (b == null) {
            return;
        }
        this.aob();
    };
    FinancialChart.prototype.aob = function () {
        var _this = this;
        if (this.ak6 == null) {
            return;
        }
        if (this.aiy != null) {
            this.aiy.provideContainer(null);
        }
        this.aiy = TypeRegistrar.create("ZoomSlider");
        this.aiy.provideContainer(this.ak6);
        var a = this.aiy;
        a.windowRectChanged = delegateCombine(a.windowRectChanged, function (a, b) {
            if (_this.dataChart != null) {
                _this.dataChart.v1 = new Rect(0, _this.aiy.windowRect.left, _this.dataChart.vl.top, _this.aiy.windowRect.width, _this.dataChart.vl.height);
            }
        });
        this.aoc();
    };
    FinancialChart.prototype.apr = function (a, b) {
        this.apf(a, b, this.ak5);
    };
    FinancialChart.prototype.setGridRows = function () {
        if (this.ak7 == null) {
            return;
        }
        var a = new Array(6);
        a[0] = (this.ajg ? this.ako : 0) + "px";
        a[1] = this.ajc ? "min-content" : "0px";
        a[2] = "auto";
        a[3] = this.indicatorCharts == null || this.indicatorCharts.count == 0 ? "0px" : this.indicatorCharts.count * 150 + "px";
        a[4] = this.volumeChart == null ? "0px" : "100px";
        a[5] = this.zoomSliderChart == null ? "0px" : "100px";
        if (this.aiy != null) {
            if (a[5] == "0px") {
                this.aiy.hide();
            }
            else {
                this.aiy.show();
            }
        }
        this.ak7.setStyleProperty("-ms-grid-rows", stringReplace(stringJoin.apply(void 0, __spreadArray([" "], __read(a))), "auto", "1fr"));
        this.ak7.setStyleProperty("grid-template-rows", stringJoin.apply(void 0, __spreadArray([" "], __read(a))));
        if (this.lu == null) {
            return;
        }
        if (this.dataChart != null) {
            this.dataChart.notifyContainerResized();
        }
        if (this.ah9 != null) {
            this.ah9.f();
        }
        if (this.volumeChart != null) {
            this.volumeChart.notifyContainerResized();
        }
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.notifyContainerResized();
        }
        if (this.aiy != null) {
            this.aiy.notifySizeChanged();
        }
    };
    FinancialChart.prototype.ux = function () {
        var e_1, _a;
        _super.prototype.ux.call(this);
        this.zv = 1;
        this.afa = 0;
        this.abs = 0;
        this.af4 = null;
        this.afx = AxisDefaults.g;
        this.abc = AxisDefaults.e;
        this.abb = AxisDefaults.f;
        this.afy = AxisDefaults.j;
        this.afu = null;
        this.apv = AxisDefaults.h;
        this.ak1 = AxisDefaults.b;
        this.gw = true;
        this.go = true;
        this.gp = false;
        this.bf = 0;
        this.jm = Defaults.j;
        this.xw = Defaults.ae;
        this.xr = Defaults.af;
        this.xs = Defaults.ag;
        this.j6 = NaN;
        this.jy = 40;
        this.g6 = true;
        this.he = false;
        if (this.dataChart != null) {
            var a = this.dataChart.c8.j.e();
            this.agr = a.d;
            this.ags = a.e;
            this.agq = a.d;
            this.apv = this.dataChart.z6;
        }
        var b = new BrushCollection();
        b.add(((function () {
            var $ret = new Brush();
            $ret.fill = "#8BDC5C";
            return $ret;
        })()));
        b.add(((function () {
            var $ret = new Brush();
            $ret.fill = "#8961A9";
            return $ret;
        })()));
        b.add(((function () {
            var $ret = new Brush();
            $ret.fill = "#6DB1FF";
            return $ret;
        })()));
        b.add(((function () {
            var $ret = new Brush();
            $ret.fill = "#82E9D9";
            return $ret;
        })()));
        b.add(((function () {
            var $ret = new Brush();
            $ret.fill = "#EA3C63";
            return $ret;
        })()));
        b.add(((function () {
            var $ret = new Brush();
            $ret.fill = "#735656";
            return $ret;
        })()));
        b.add(((function () {
            var $ret = new Brush();
            $ret.fill = "#F8CE4F";
            return $ret;
        })()));
        b.add(((function () {
            var $ret = new Brush();
            $ret.fill = "#A8A8B7";
            return $ret;
        })()));
        b.add(((function () {
            var $ret = new Brush();
            $ret.fill = "#E051A9";
            return $ret;
        })()));
        b.add(((function () {
            var $ret = new Brush();
            $ret.fill = "#FF903B";
            return $ret;
        })()));
        var c = new BrushCollection();
        try {
            for (var _b = __values(fromEnum(b)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var d = _c.value;
                c.add(BrushUtil.n(d, 0.3));
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        this.agt = c;
        this.agu = b;
    };
    FinancialChart.prototype.apb = function () {
        if (this.ah9 != null) {
            this.ah9.b = null;
            this.ah9.b = this.indicatorCharts;
        }
        this.setGridRows();
    };
    FinancialChart.prototype.apc = function () {
        this.setGridRows();
    };
    FinancialChart.prototype.apl = function () {
        this.apb();
    };
    FinancialChart.prototype.aoa = function () {
        if (this.aiy == null || this.zoomSliderChart == null || this.dataChart == null || this.dataChart.series.count == 0) {
            return;
        }
        var a = this.lu.rootWrapper.width();
        var b = this.dataChart.v0.left + this.dataChart.v0.width;
        var c = this.dataChart.v0.left;
        var d = a - b;
        this.aiy.startInset = c - this.aiy.trackStartInset;
        this.aiy.endInset = d - this.aiy.trackEndInset;
        if (this.aiy.endInset < 0) {
            var e = this.aiy.endInset;
            this.aiy.endInset = 0;
            this.dataChart.h0 += (e * -1);
        }
        if (this.aiy.startInset < 0) {
            var f = this.aiy.startInset;
            this.aiy.startInset = 0;
            this.dataChart.hv += (f * -1);
        }
        this.zoomSliderChart.hv = c;
        this.zoomSliderChart.h0 = d;
        this.zoomSliderChart.hi = this.aiy.barExtent;
    };
    FinancialChart.prototype.ao1 = function () {
        this.aoa();
    };
    FinancialChart.prototype.uu = function (a) {
        _super.prototype.uu.call(this, a);
        if (a != null) {
            a.hh = 0;
        }
    };
    FinancialChart.prototype.apw = function (a) {
        return a.ve();
    };
    FinancialChart.prototype.apz = function () {
        return this.ak3 == null ? null : this.apy(this.ak3.rootWrapper);
    };
    FinancialChart.prototype.apx = function (a) {
        if (a == null || a.d == null) {
            return null;
        }
        return this.apy(a.d.rootWrapper);
    };
    FinancialChart.prototype.apy = function (a) {
        if (a == null) {
            return null;
        }
        var b = a.getOffset();
        return new Rect(0, b.left, b.top, a.width(), a.height());
    };
    FinancialChart.prototype.akj = function () {
        if (this.ak7 == null) {
            return NaN;
        }
        var a = this.ak7.getOffset();
        return this.ak7.width() + (a == null ? 0 : a.left);
    };
    FinancialChart.prototype.u6 = function (source_, a, b) {
        _super.prototype.u6.call(this, source_, a, b);
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.n2(source_, a, b);
        }
    };
    FinancialChart.prototype.u7 = function (source_, a, b) {
        _super.prototype.u7.call(this, source_, a, b);
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.n3(source_, a, b);
        }
    };
    FinancialChart.prototype.u5 = function (source_) {
        _super.prototype.u5.call(this, source_);
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.nv(source_);
        }
    };
    FinancialChart.prototype.u9 = function (source_, a, b, c) {
        _super.prototype.u9.call(this, source_, a, b, c);
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.n4(source_, a, b, c);
        }
    };
    FinancialChart.prototype.u0 = function (a, b) {
        _super.prototype.u0.call(this, a, b);
        if (this.ajd(a)) {
            a.toolTip = null;
        }
    };
    Object.defineProperty(FinancialChart.prototype, "ahg", {
        get: function () {
            return this.ahf;
        },
        set: function (a) {
            var b = this.ahg;
            if (a != b) {
                this.ahf = a;
                this.vl("Toolbar", b, this.ahg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "alw", {
        get: function () {
            return this.alp;
        },
        set: function (a) {
            var b = this.alw;
            if (a != b) {
                this.alp = a;
                this.vl("RangeSelectorTemplate", b, this.alw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "alx", {
        get: function () {
            return this.alq;
        },
        set: function (a) {
            var b = this.alx;
            if (a != b) {
                this.alq = a;
                this.vl("ToolbarTemplate", b, this.alx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "alt", {
        get: function () {
            return this.alm;
        },
        set: function (a) {
            var b = this.alt;
            if (a != b) {
                this.alm = a;
                this.vl("ChartTypePickerTemplate", b, this.alt);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "aly", {
        get: function () {
            return this.alr;
        },
        set: function (a) {
            var b = this.aly;
            if (a != b) {
                this.alr = a;
                this.vl("TrendLineTypePickerTemplate", b, this.aly);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "alz", {
        get: function () {
            return this.als;
        },
        set: function (a) {
            var b = this.alz;
            if (a != b) {
                this.als = a;
                this.vl("VolumeTypePickerTemplate", b, this.alz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "alu", {
        get: function () {
            return this.aln;
        },
        set: function (a) {
            var b = this.alu;
            if (a != b) {
                this.aln = a;
                this.vl("IndicatorMenuTemplate", b, this.alu);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "alv", {
        get: function () {
            return this.alo;
        },
        set: function (a) {
            var b = this.alv;
            if (a != b) {
                this.alo = a;
                this.vl("OverlayPickerTemplate", b, this.alv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "aiy", {
        get: function () {
            return this.aix;
        },
        set: function (a) {
            var b = this.aiy;
            if (a != b) {
                this.aix = a;
                this.vl("ZoomSlider", b, this.aiy);
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.get_jy = function () {
        return _super.prototype.get_jy.call(this);
    };
    FinancialChart.prototype.set_jy = function (a) {
        _super.prototype.set_jy.call(this, a);
    };
    FinancialChart.prototype.get_g6 = function () {
        return _super.prototype.get_g6.call(this);
    };
    FinancialChart.prototype.set_g6 = function (a) {
        _super.prototype.set_g6.call(this, a);
    };
    Object.defineProperty(FinancialChart.prototype, "ako", {
        get: function () {
            return this.aj7;
        },
        set: function (a) {
            var b = this.ako;
            if (a != b) {
                this.aj7 = a;
                this.vl("ToolbarHeight", b, this.ako);
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.zk = function () {
        return this.xAxis;
    };
    FinancialChart.prototype.zl = function () {
        return this.yAxis;
    };
    FinancialChart.prototype.onAttachedToUI = function () {
        _super.prototype.onAttachedToUI.call(this);
        if (this.aiy != null) {
            this.aiy.onAttachedToUI();
        }
    };
    FinancialChart.prototype.onDetachedFromUI = function () {
        _super.prototype.onDetachedFromUI.call(this);
        if (this.aiy != null) {
            this.aiy.onDetachedFromUI();
        }
    };
    FinancialChart.prototype.bi = function (a) {
        return 20;
    };
    FinancialChart.prototype.get_hy = function () {
        return ((function () {
            var $ret = new List$1(DataSeriesType_$type, 0);
            $ret.add(20);
            return $ret;
        })());
    };
    Object.defineProperty(FinancialChart.prototype, "hy", {
        get: function () {
            return this.get_hy();
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.get_hx = function () {
        return ((function () {
            var $ret = new List$1(IDataSeriesAdapterRule_$type, 0);
            $ret.add(new SimpleFinancialPriceSeriesRule());
            $ret.add(new SubCollectionsRule());
            return $ret;
        })());
    };
    Object.defineProperty(FinancialChart.prototype, "hx", {
        get: function () {
            return this.get_hx();
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.ag2 = function (a) {
        return this.ag3(a);
    };
    FinancialChart.prototype.aig = function (a) {
        return this.aih(a);
    };
    FinancialChart.prototype.ag1 = function (a) {
        var b = this.ag2(0);
        b.name = a;
        return b;
    };
    FinancialChart.prototype.t9 = function () {
        this.xAxis = this.ag1("xAxis");
        this.yAxis = this.aig(0);
        this.yAxis.name = "yAxis";
        this.volumeAxis = this.aig(0);
        this.volumeAxis.name = "volumeAxis";
        this.volumeXAxis = this.ag1("volumeXAxis");
        this.aju = new List$1(NumericYAxis.$, 0);
        this.ajw = new List$1(CategoryAxisBase.$, 0);
        this.indicatorAxis = this.aig(0);
        this.indicatorAxis.name = "indicatorAxis";
        this.indicatorXAxis = this.ag1("indicatorXAxis_0");
        this.aju.add(this.indicatorAxis);
        this.ajw.add(this.indicatorXAxis);
        if (this.xAxis.az == null) {
            this.xAxis.az = new AxisLabelSettings();
        }
        if (this.yAxis.az == null) {
            this.yAxis.az = new AxisLabelSettings();
        }
        if (this.volumeAxis.az == null) {
            this.volumeAxis.az = new AxisLabelSettings();
        }
        if (this.volumeXAxis.az == null) {
            this.volumeAxis.az = new AxisLabelSettings();
        }
        if (this.indicatorAxis.az == null) {
            this.indicatorAxis.az = new AxisLabelSettings();
        }
        if (this.indicatorXAxis.az == null) {
            this.indicatorXAxis.az = new AxisLabelSettings();
        }
        this.zoomSliderAxis = this.aig(0);
        this.zoomSliderAxis.name = "zoomSliderAxis";
        this.zoomSliderXAxis = this.ag1("zoomSliderXAxis");
        if (this.zoomSliderAxis.az == null) {
            this.zoomSliderAxis.az = new AxisLabelSettings();
        }
        if (this.zoomSliderXAxis.az == null) {
            this.zoomSliderXAxis.az = new AxisLabelSettings();
        }
    };
    FinancialChart.prototype.t3 = function (a, b) {
        var c = typeCast(MarkerSeries.$, a);
        if (c != null) {
            c.w2 = this.c6(a, b);
        }
    };
    FinancialChart.prototype.c5 = function (a) {
        if (a != null) {
            return 1;
        }
        return _super.prototype.c5.call(this, a);
    };
    FinancialChart.prototype.we = function (a) {
        if (a.suggestedSecondaryAxis == 2) {
            this.yAxis.lj = true;
            this.yAxis.ma = 10;
        }
    };
    FinancialChart.prototype.wh = function () {
        this.yAxis.lj = this.ajn;
        this.yAxis.ma = this.all;
    };
    FinancialChart.prototype.ahk = function (a) {
        var b;
        if (this.ahi == 0) {
            if (a == null || a.index > 0 || stringIsNullOrEmpty(a.getMemberPathFor(12))) {
                b = 4;
            }
            else {
                b = 2;
            }
        }
        else {
            b = this.ahi;
        }
        b = this.ahj(b);
        return b;
    };
    FinancialChart.prototype.ahj = function (a) {
        if ((a == 1 || a == 2)) {
            if (TypeRegistrar.isRegistered("FinancialPriceSeries")) {
                return a;
            }
            a = 4;
        }
        if (a == 4) {
            if (TypeRegistrar.isRegistered("LineSeries")) {
                return a;
            }
            a = 3;
        }
        if (a == 3) {
            if (TypeRegistrar.isRegistered("ColumnSeries")) {
                return a;
            }
        }
        return a;
    };
    FinancialChart.prototype.aht = function (a) {
        if ((a == 2 || a == 3)) {
            if (TypeRegistrar.isRegistered("FinancialPriceSeries")) {
                return a;
            }
            a = 5;
        }
        if (a == 5) {
            if (TypeRegistrar.isRegistered("LineSeries")) {
                return a;
            }
            a = 6;
        }
        if (a == 6) {
            if (TypeRegistrar.isRegistered("AreaSeries")) {
                return a;
            }
            a = 4;
        }
        if (a == 4) {
            if (TypeRegistrar.isRegistered("ColumnSeries")) {
                return a;
            }
        }
        return a;
    };
    FinancialChart.prototype.ahm = function (a) {
        if (a == 2) {
            if (TypeRegistrar.isRegistered("LineSeries")) {
                return a;
            }
            a = 3;
        }
        if (a == 3) {
            if (TypeRegistrar.isRegistered("AreaSeries")) {
                return a;
            }
            a = 1;
        }
        if (a == 1) {
            if (TypeRegistrar.isRegistered("ColumnSeries")) {
                return a;
            }
        }
        return a;
    };
    FinancialChart.prototype.al8 = function (a) {
        var b = a.getMemberPathFor(14);
        if (stringIsNullOrEmpty(b)) {
            b = a.getMemberPathFor(11);
        }
        return b;
    };
    FinancialChart.prototype.ag3 = function (axisType_) {
        if ((this.externalObject)) {
            return (this.externalObject.createXAxis(axisType_).i);
        }
        else {
            var a = void 0;
            switch (axisType_) {
                case 0:
                    if (TypeRegistrar.isRegistered("OrdinalTimeXAxis")) {
                        a = TypeRegistrar.create("OrdinalTimeXAxis");
                    }
                    else if (TypeRegistrar.isRegistered("TimeXAxis")) {
                        a = TypeRegistrar.create("TimeXAxis");
                    }
                    else {
                        throw new BaseError(1, "no valid axes loaded");
                    }
                    return a;
                case 1:
                    if (TypeRegistrar.isRegistered("TimeXAxis")) {
                        a = TypeRegistrar.create("TimeXAxis");
                    }
                    else if (TypeRegistrar.isRegistered("OrdinalTimeXAxis")) {
                        a = TypeRegistrar.create("OrdinalTimeXAxis");
                    }
                    else {
                        throw new BaseError(1, "no valid axes loaded");
                    }
                    return a;
            }
            return null;
        }
    };
    FinancialChart.prototype.aih = function (axisType_) {
        if ((this.externalObject)) {
            return (this.externalObject.createYAxis(axisType_).i);
        }
        else {
            var a = void 0;
            switch (axisType_) {
                case 0:
                    if (TypeRegistrar.isRegistered("NumericYAxis")) {
                        a = TypeRegistrar.create("NumericYAxis");
                    }
                    else if (TypeRegistrar.isRegistered("PercentChangeYAxis")) {
                        a = TypeRegistrar.create("PercentChangeYAxis");
                    }
                    else {
                        throw new BaseError(1, "no valid axes loaded");
                    }
                    return a;
                case 1:
                    if (TypeRegistrar.isRegistered("PercentChangeYAxis")) {
                        a = TypeRegistrar.create("PercentChangeYAxis");
                    }
                    else if (TypeRegistrar.isRegistered("NumericYAxis")) {
                        a = TypeRegistrar.create("NumericYAxis");
                    }
                    else {
                        throw new BaseError(1, "no valid axes loaded");
                    }
                    return a;
            }
            return null;
        }
    };
    FinancialChart.prototype.ain = function () {
        return this.di(1);
    };
    FinancialChart.prototype.aio = function () {
        return this.di(0);
    };
    FinancialChart.prototype.aim = function () {
        return this.di(2);
    };
    FinancialChart.prototype.aip = function () {
        return this.di(20);
    };
    Object.defineProperty(FinancialChart.prototype, "ai6", {
        get: function () {
            return (TypeRegistrar.isRegistered("HorizontalAnchoredCategorySeriesProxy"));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ah4", {
        get: function () {
            if (TypeRegistrar.isRegistered("HorizontalAnchoredCategorySeriesProxy")) {
                return TypeRegistrar.create("HorizontalAnchoredCategorySeriesProxy");
            }
            else {
                throw new BaseError(1, "horizontal series module is not loaded, but is required.");
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ai9", {
        get: function () {
            return (TypeRegistrar.isRegistered("FinancialPriceSeriesProxy"));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ah2", {
        get: function () {
            if (TypeRegistrar.isRegistered("FinancialPriceSeriesProxy")) {
                return TypeRegistrar.create("FinancialPriceSeriesProxy");
            }
            else {
                throw new BaseError(1, "financial series module is not loaded, but is required.");
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ai7", {
        get: function () {
            return (TypeRegistrar.isRegistered("IndicatorProxy"));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ah6", {
        get: function () {
            if (TypeRegistrar.isRegistered("IndicatorProxy")) {
                return TypeRegistrar.create("IndicatorProxy");
            }
            else {
                throw new BaseError(1, "financial series module is not loaded, but is required.");
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ai8", {
        get: function () {
            return (TypeRegistrar.isRegistered("OverlayProxy"));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "aib", {
        get: function () {
            if (TypeRegistrar.isRegistered("OverlayProxy")) {
                return TypeRegistrar.create("OverlayProxy");
            }
            else {
                throw new BaseError(1, "overlay series module is not loaded, but is required.");
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.dg = function (a, b) {
        var c = this.al8(a);
        switch (this.ahk(a)) {
            case 3:
                var d = this.ain();
                this.ah4.setValueMemberPath(d, c);
                this.ah4.setXAxis(d, this.xAxis);
                this.ah4.setYAxis(d, this.yAxis);
                this.ah4.setCategoryMode(d, 1);
                return d;
            case 4:
                var e = this.aio();
                this.ah4.setValueMemberPath(e, c);
                this.ah4.setXAxis(e, this.xAxis);
                this.ah4.setYAxis(e, this.yAxis);
                this.ah4.setCategoryMode(e, 1);
                return e;
            default:
            case 1:
            case 2:
                var f = this.aip();
                this.ah2.setOpenMemberPath(f, a.getMemberPathFor(11));
                this.ah2.setHighMemberPath(f, a.getMemberPathFor(12));
                this.ah2.setLowMemberPath(f, a.getMemberPathFor(13));
                this.ah2.setCloseMemberPath(f, a.getMemberPathFor(14));
                this.ah2.setVolumeMemberPath(f, a.getMemberPathFor(15));
                this.ah2.setXAxis(f, this.xAxis);
                this.ah2.setYAxis(f, this.yAxis);
                this.ah2.setDisplayType(f, this.ahi == 1 ? 1 : 0);
                this.ah2.setCategoryMode(f, 1);
                return f;
        }
    };
    FinancialChart.prototype.wg = function () {
        _super.prototype.wg.call(this);
        this.aoz(this.xAxis);
        this.apn(false);
        if (this.ahn != 0 && this.volumeChart != null) {
            this.ao6();
        }
        this.aoz(this.volumeXAxis);
        this.aoz(this.indicatorXAxis);
        this.apk();
        this.apm();
        if (this.ahu != 0) {
            this.ao7();
            this.aoz(this.zoomSliderXAxis);
        }
    };
    FinancialChart.prototype.aew = function (a, b) {
        var e_2, _a;
        _super.prototype.aew.call(this, a, b);
        a(this.volumeXAxis);
        if (this.ajw != null) {
            try {
                for (var _b = __values(fromEnum(this.ajw)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var c = _c.value;
                    a(c);
                }
            }
            catch (e_2_1) {
                e_2 = { error: e_2_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_2)
                        throw e_2.error;
                }
            }
        }
        var d = true;
        switch (b) {
            case "XAxisMajorStroke":
            case "XAxisMajorStrokeThickness":
                d = false;
                break;
        }
        if (d) {
            a(this.zoomSliderXAxis);
        }
    };
    FinancialChart.prototype.ae1 = function (a, b) {
        var e_3, _a;
        _super.prototype.ae1.call(this, a, b);
        if (b) {
            a(this.volumeAxis);
            if (this.aju != null) {
                try {
                    for (var _b = __values(fromEnum(this.aju)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var c = _c.value;
                        a(c);
                    }
                }
                catch (e_3_1) {
                    e_3 = { error: e_3_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_3)
                            throw e_3.error;
                    }
                }
            }
            a(this.zoomSliderAxis);
        }
    };
    FinancialChart.prototype.ao3 = function (a) {
        var b = typeCast(IZoomableCategoryAxis_$type, this.xAxis);
        if (b != null) {
            a(b);
        }
    };
    FinancialChart.prototype.aez = function (a) {
        _super.prototype.aez.call(this, a);
        this.aeq(this.volumeXAxis, a);
        this.aeq(this.indicatorXAxis, a);
        this.aeq(this.zoomSliderXAxis, a);
    };
    FinancialChart.prototype.ae4 = function (a) {
        _super.prototype.ae4.call(this, a);
        this.aeq(this.volumeAxis, a);
        this.aeq(this.indicatorAxis, a);
        this.aeq(this.zoomSliderAxis, a);
    };
    FinancialChart.prototype.ae2 = function (a) {
        _super.prototype.ae2.call(this, a);
        this.aep(this.volumeAxis, a);
        this.aep(this.indicatorAxis, a);
        this.aep(this.zoomSliderAxis, a);
    };
    FinancialChart.prototype.aex = function (a) {
        _super.prototype.aex.call(this, a);
        this.aep(this.volumeXAxis, a);
        this.aep(this.indicatorXAxis, a);
        this.aep(this.zoomSliderXAxis, a);
    };
    FinancialChart.prototype.ut = function (a) {
        var _this = this;
        _super.prototype.ut.call(this, a);
        if (a == null) {
            return;
        }
        if (a == this.zoomSliderChart) {
            a.ww = a.wz = false;
            a.legend = null;
            if (this.dataChart != null) {
                a.ag = this.dataChart.ag;
                a.aj = this.dataChart.aj;
            }
        }
        else {
            SyncManager.d(a, ((function () {
                var $ret = new SyncSettings();
                $ret.syncChannel = _this.am3();
                $ret.synchronizeHorizontally = true;
                $ret.synchronizeVertically = false;
                return $ret;
            })()));
            SyncManager.c(a);
        }
        a.cw = 2;
    };
    FinancialChart.prototype.am3 = function () {
        if (this.al0 == null) {
            FinancialChart.alk++;
            this.al0 = FinancialChart.alk.toString();
        }
        return this.al0;
    };
    FinancialChart.prototype.aol = function (a, b) {
        this.an7();
        this.ao1();
    };
    FinancialChart.prototype.an7 = function () {
        var e_4, _a;
        if (this.yAxis.b8()) {
            return;
        }
        var a = this.yAxis.aw.bd;
        if (this.ahn != 0) {
            a = Math.max(a, this.volumeAxis.aw.bd);
        }
        try {
            for (var _b = __values(fromEnum(this.aju)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var b = _c.value;
                a = Math.max(a, b.aw.bd);
            }
        }
        catch (e_4_1) {
            e_4 = { error: e_4_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_4)
                    throw e_4.error;
            }
        }
        this.an8(a);
    };
    FinancialChart.prototype.an8 = function (a) {
        var e_5, _a;
        if (this.yAxis.b8()) {
            return;
        }
        this.yAxis.et = this.volumeAxis.et = a;
        try {
            for (var _b = __values(fromEnum(this.aju)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var b = _c.value;
                b.et = a;
            }
        }
        catch (e_5_1) {
            e_5 = { error: e_5_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_5)
                    throw e_5.error;
            }
        }
        this.zoomSliderAxis.et = this.yAxis.et;
    };
    FinancialChart.prototype.aod = function (a) {
        if (this.indicatorCharts == null) {
            return;
        }
        for (var b = 0; b < this.indicatorCharts.count; b++) {
            var c = this.indicatorCharts._inner[b];
            for (var d = 0; d < c.series.count; d++) {
                var e = c.series._inner[d];
                if (!(this.ajb(e))) {
                    continue;
                }
                a(c.series._inner[d], b);
            }
        }
    };
    FinancialChart.prototype.aog = function (a) {
        if (this.volumeChart == null) {
            return;
        }
        for (var b = 0; b < this.volumeChart.series.count; b++) {
            var c = this.volumeChart.series._inner[b];
            if (c.eq) {
                continue;
            }
            a(this.volumeChart.series._inner[b], b);
        }
    };
    FinancialChart.prototype.aoe = function (a) {
        if (this.dataChart == null) {
            return;
        }
        for (var b = 0; b < this.dataChart.series.count; b++) {
            var c = this.dataChart.series._inner[b];
            if (!this.ajd(c)) {
                continue;
            }
            a(c, b);
        }
    };
    FinancialChart.prototype.aof = function (a, b) {
        if (this.dataChart == null) {
            return;
        }
        for (var c = 0; c < this.dataChart.series.count; c++) {
            var d = this.dataChart.series._inner[c];
            if (this.ajd(d)) {
                continue;
            }
            var e = this.ale(d, c);
            a(d, e);
        }
        if (b) {
            this.aoh(a);
        }
    };
    FinancialChart.prototype.aoh = function (a) {
        if (this.zoomSliderChart == null) {
            return;
        }
        for (var b = 0; b < this.zoomSliderChart.series.count; b++) {
            var c = this.zoomSliderChart.series._inner[b];
            if (c.eq) {
                continue;
            }
            a(this.zoomSliderChart.series._inner[b], b);
        }
    };
    FinancialChart.prototype.ale = function (a, b) {
        if (a.c7 != this.dataChart) {
            return b;
        }
        var c = this.bj == null || this.bj.dataSeries == null ? 0 : this.bj.dataSeries.count;
        var d = this.ahy == null ? 0 : this.ahy.count;
        return b - (c * d);
    };
    FinancialChart.prototype.uj = function (a) {
        _super.prototype.uj.call(this, a);
        this.aog(a);
        this.aod(a);
        this.aoh(a);
    };
    Object.defineProperty(FinancialChart.prototype, "volumeChart", {
        get: function () {
            return this.ait;
        },
        set: function (a) {
            var b = this.volumeChart;
            if (a != b) {
                this.ait = a;
                this.vl("VolumeChart", b, this.volumeChart);
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.ao6 = function () {
        if (this.dataChart == null) {
            return;
        }
        this.volumeChart.series.clear();
        var a = new List$1(Series.$, 0);
        var b = -1;
        for (var c = 0; c < this.dataChart.series.count; c++) {
            var d = this.dataChart.series._inner[c];
            if (this.ajd(d) || d.eq) {
                continue;
            }
            b++;
            var e = void 0;
            var f = this.ahn;
            f = this.ahm(f);
            switch (f) {
                case 2:
                    e = this.aio();
                    break;
                case 3:
                    e = this.aim();
                    break;
                default:
                case 1:
                    e = this.ain();
                    break;
            }
            this.ah4.setCategoryMode(e, 1);
            e.name = "volumeSeries_" + b;
            e.itemsSource = d.itemsSource;
            this.ah4.setXAxis(e, this.volumeXAxis);
            this.ah4.setYAxis(e, this.volumeAxis);
            this.ah4.setValueMemberPath(e, this.bj != null && this.bj.dataSeries != null && this.bj.dataSeries.count > b ? this.bj.dataSeries._inner[b].getMemberPathFor(15) : null);
            e.title = d.title + " Volume";
            e.u9 = ArrayExtension.getModulus$1(Brush.$, this.agv, b);
            e.vp = ArrayExtension.getModulus$1(Brush.$, this.agw, b);
            e.jn = this.akp;
            a.add(e);
        }
        var g = this.hu("volume", false);
        if (g != null) {
            a.o(g);
        }
        for (var h = 0; h < a.count; h++) {
            var i = a._inner[h];
            this.volumeChart.series.add(i);
            this.uy(i, h);
        }
    };
    FinancialChart.prototype.aja = function (a) {
        if (!this.ai6) {
            return false;
        }
        return this.ah4.matchesType(a);
    };
    FinancialChart.prototype.aje = function (a) {
        if (!this.ai9) {
            return false;
        }
        return this.ah2.matchesType(a);
    };
    FinancialChart.prototype.ajb = function (a) {
        if (!this.ai7) {
            return false;
        }
        return this.ah6.matchesType(a);
    };
    FinancialChart.prototype.ajd = function (a) {
        if (!this.ai8) {
            return false;
        }
        return this.aib.matchesType(a);
    };
    FinancialChart.prototype.ao5 = function () {
        if (this.dataChart == null) {
            return;
        }
        var a = 0;
        var b = this.ahw == null ? 0 : this.ahw.count;
        var c = this.customIndicatorNames == null ? 0 : this.customIndicatorNames.count;
        var d = b + c;
        for (var e = 0; e < d; e++) {
            var f = new List$1(Series.$, 0);
            var g = this.indicatorCharts._inner[e];
            var h = e >= b;
            var i = h ? this.customIndicatorNames._inner[e - b] : EnumUtil.getName(FinancialIndicatorType_$type, this.ahw._inner[e]);
            var j = -1;
            for (var k = 0; k < this.dataChart.series.count; k++) {
                var l = this.dataChart.series._inner[k];
                if (this.ajd(l) || l.eq) {
                    continue;
                }
                j++;
                var m = void 0;
                if (h) {
                    var n = this.aiq("Custom");
                    this.ah6.handleCustomIndicatorEvent(n, runOn(this, this.aoj));
                    m = n;
                }
                else {
                    m = this.ais(this.ahw._inner[e]);
                }
                if (m == null) {
                    continue;
                }
                m.name = "indicator_" + a++;
                this.ah6.setXAxis(m, this.ajw._inner[e]);
                this.ah6.setYAxis(m, this.aju._inner[e]);
                m.itemsSource = l.itemsSource;
                this.ah6.setOpenMemberPath(m, this.bj.dataSeries._inner[j].getMemberPathFor(11));
                this.ah6.setHighMemberPath(m, this.bj.dataSeries._inner[j].getMemberPathFor(12));
                this.ah6.setLowMemberPath(m, this.bj.dataSeries._inner[j].getMemberPathFor(13));
                this.ah6.setCloseMemberPath(m, this.bj.dataSeries._inner[j].getMemberPathFor(14));
                this.ah6.setVolumeMemberPath(m, this.bj.dataSeries._inner[j].getMemberPathFor(15));
                m.title = l.title + " " + i;
                m.u9 = ArrayExtension.getModulus$1(Brush.$, this.agp, e);
                m.jn = this.akl;
                if (this.ajb(m)) {
                    this.ah6.setNegativeBrush(m, ArrayExtension.getModulus$1(Brush.$, this.agq, e));
                    this.ah6.setDisplayType(m, ArrayExtension.getModulus$1(IndicatorDisplayType_$type, this.ah8, e));
                }
                if (this.alg != -2147483648 && this.ajb(m)) {
                    this.ah6.setPeriodOverride(m, this.alg);
                }
                if (this.ajb(m)) {
                    this.ah6.setSmoothingPeriod(m, this.alj);
                }
                if (this.ajb(m)) {
                    this.ah6.setSignalPeriod(m, this.ali);
                }
                if (this.ajb(m)) {
                    this.ah6.setShortPeriod(m, this.alh);
                    this.ah6.setLongPeriod(m, this.alf);
                }
                this.uy(m, e);
                f.add(m);
            }
            var o = this.hu(i, false);
            for (var p = 0; p < f.count; p++) {
                var q = f._inner[p];
                g.series.add(q);
            }
            if (o != null) {
                for (var r = 0; r < o.count; r++) {
                    var s = o._inner[r];
                    g.series.add(s);
                    this.uy(s, r);
                }
            }
        }
    };
    FinancialChart.prototype.aiq = function (indicatorName_) {
        return (this.externalObject.createIndicator(indicatorName_ + 'Indicator').i);
    };
    FinancialChart.prototype.ais = function (a) {
        return this.aiq(EnumUtil.getName(FinancialIndicatorType_$type, a));
    };
    FinancialChart.prototype.apn = function (a) {
        if (this.ahg == null) {
            return;
        }
        var b = this.xAxis;
        if (b.actualMinimumValue > b.actualMaximumValue) {
            return;
        }
        this.ahg.as = b.actualMinimumValue;
        this.ahg.ar = b.actualMaximumValue;
        if (a) {
            this.ahg.au = b.visibleMinimum;
            this.ahg.at = b.visibleMaximum;
        }
        this.ahg.bn();
    };
    FinancialChart.prototype.lv = function () {
        var e_6, _a, e_7, _b;
        var _this = this;
        var a = ((function () {
            var $ret = new FinancialChartVisualData();
            $ret.mainChart = _super.prototype.lv.call(_this);
            $ret.volumeChart = _this.volumeChart == null ? null : _this.volumeChart.x8();
            $ret.toolbar = _this.ahg == null ? null : _this.ahg.aj();
            $ret.zoomChart = _this.zoomSliderChart == null ? null : _this.zoomSliderChart.x8();
            return $ret;
        })());
        if (this.indicatorCharts != null) {
            a.indicatorCharts = new List$1(Base.$, 0);
            try {
                for (var _c = __values(fromEnum(this.indicatorCharts)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var b = _d.value;
                    a.indicatorCharts.add1(b.x8());
                }
            }
            catch (e_6_1) {
                e_6 = { error: e_6_1 };
            }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return))
                        _a.call(_c);
                }
                finally {
                    if (e_6)
                        throw e_6.error;
                }
            }
        }
        if (this.dataChart != null) {
            a.mainChartLeft = this.apw(this.dataChart).x;
            a.mainChartTop = this.apw(this.dataChart).y;
        }
        if (this.volumeChart != null) {
            a.volumeChartLeft = this.apw(this.volumeChart).x;
            a.volumeChartTop = this.apw(this.volumeChart).y;
        }
        if (this.zoomSliderChart != null) {
            a.zoomChartLeft = this.apw(this.zoomSliderChart).x;
            a.zoomChartTop = this.apw(this.zoomSliderChart).y;
        }
        if (this.ahg != null) {
            var c = this.apx(this.ahg);
            a.toolbarLeft = c.left;
            a.toolbarTop = c.top;
            a.toolbarWidth = c.width;
            a.toolbarHeight = c.height;
        }
        if (this.indicatorCharts != null) {
            a.indicatorChartLefts = new List$1(Number_$type, 0);
            a.indicatorChartTops = new List$1(Number_$type, 0);
            try {
                for (var _e = __values(fromEnum(this.indicatorCharts)), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var d = _f.value;
                    a.indicatorChartLefts.add(this.apw(d).x);
                    a.indicatorChartTops.add(this.apw(d).y);
                }
            }
            catch (e_7_1) {
                e_7 = { error: e_7_1 };
            }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return))
                        _b.call(_e);
                }
                finally {
                    if (e_7)
                        throw e_7.error;
                }
            }
        }
        var e = this.apz();
        a.legendLeft = e.left;
        a.legendTop = e.top;
        a.legendWidth = e.width;
        a.legendHeight = e.height;
        a.legend = this.ah0 == null ? null : this.ah0.exportVisualData();
        return a;
    };
    FinancialChart.prototype.al5 = function () {
        var a = this.lv();
        a.scaleByViewport();
        return a.serialize();
    };
    FinancialChart.prototype.air = function (overlayName_) {
        return (this.externalObject.createOverlay(overlayName_ + 'Overlay').i);
    };
    FinancialChart.prototype.apm = function () {
        var e_8, _a, e_9, _b;
        if (this.dataChart == null) {
            return;
        }
        var a = new List$1(Series.$, 0);
        try {
            for (var _c = __values(fromEnum(this.dataChart.series)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var b = _d.value;
                if (this.ajd(b)) {
                    a.add(b);
                }
            }
        }
        catch (e_8_1) {
            e_8 = { error: e_8_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_8)
                    throw e_8.error;
            }
        }
        try {
            for (var _e = __values(fromEnum(a)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var c = _f.value;
                this.dataChart.series.remove(c);
            }
        }
        catch (e_9_1) {
            e_9 = { error: e_9_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_9)
                    throw e_9.error;
            }
        }
        if (this.ahy == null) {
            return;
        }
        var d = new List$1(Series.$, 0);
        var e = -1;
        for (var f = 0; f < this.dataChart.series.count; f++) {
            var g = this.dataChart.series._inner[f];
            if (this.ajd(g) || g.eq) {
                continue;
            }
            e++;
            for (var h = 0; h < this.ahy.count; h++) {
                var i = this.ahy._inner[h];
                var j = void 0;
                switch (i) {
                    case 1:
                        j = this.air(EnumUtil.getName(FinancialOverlayType_$type, i));
                        break;
                    default:
                    case 0:
                        j = this.air(EnumUtil.getName(FinancialOverlayType_$type, i));
                        break;
                }
                j.name = g.name + "_" + EnumUtil.getName(FinancialOverlayType_$type, i);
                this.aib.setXAxis(j, this.xAxis);
                this.aib.setYAxis(j, this.yAxis);
                j.itemsSource = g.itemsSource;
                var k = e * this.ahy.count + h;
                j.u9 = ArrayExtension.getModulus$1(Brush.$, this.agt, k);
                j.vp = ArrayExtension.getModulus$1(Brush.$, this.agu, k);
                j.jn = this.akn;
                j.title = g.title + " " + EnumUtil.getName(FinancialOverlayType_$type, i);
                var l = g;
                if (l == null) {
                    continue;
                }
                var m = this.bj.dataSeries._inner[e];
                this.aib.setOpenMemberPath(j, m.getMemberPathFor(11));
                this.aib.setHighMemberPath(j, m.getMemberPathFor(12));
                this.aib.setLowMemberPath(j, m.getMemberPathFor(13));
                this.aib.setCloseMemberPath(j, m.getMemberPathFor(14));
                this.aib.setVolumeMemberPath(j, m.getMemberPathFor(15));
                d.add(j);
            }
        }
        var n = new Array(this.dataChart.series.count);
        this.dataChart.series.copyTo(n, 0);
        this.dataChart.skipDataCleanup = true;
        this.dataChart.series.clear();
        this.dataChart.skipDataCleanup = false;
        for (var o = 0; o < n.length; o++) {
            n[o].index = -1;
        }
        d.o(n);
        for (var p = 0; p < d.count; p++) {
            var q = d._inner[p];
            this.dataChart.series.add(q);
            this.uy(q, p);
        }
    };
    FinancialChart.prototype.apk = function () {
        var e_10, _a;
        if (this.ajj) {
            return;
        }
        this.ajj = true;
        if (this.indicatorCharts != null) {
            var a = this.indicatorCharts;
            this.indicatorCharts = null;
            try {
                for (var _b = __values(fromEnum(a)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var b = _c.value;
                    this.ao0(b);
                }
            }
            catch (e_10_1) {
                e_10 = { error: e_10_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_10)
                        throw e_10.error;
                }
            }
        }
        if ((this.ahw == null || this.ahw.count == 0) && (this.customIndicatorNames == null || this.customIndicatorNames.count == 0)) {
            this.indicatorCharts = null;
            if (this.aju != null) {
                this.aju.clear();
            }
            if (this.ajw != null) {
                this.ajw.clear();
            }
            this.apb();
            this.apl();
            this.ajj = false;
            return;
        }
        if (this.indicatorCharts == null) {
            this.indicatorCharts = new List$1(XamDataChart.$, 0);
        }
        this.indicatorCharts.clear();
        this.aju.clear();
        this.ajw.clear();
        var c = this.ahw == null ? 0 : this.ahw.count;
        var d = this.customIndicatorNames == null ? 0 : this.customIndicatorNames.count;
        var e = c + d;
        for (var f = 0; f < e; f++) {
            var g = new XamDataChart();
            this.indicatorCharts.add(g);
            this.aoy(g, null);
            var h = f >= c;
            if (h) {
                g.ls = this.customIndicatorNames._inner[f - c];
            }
            else {
                g.ls = this.al7(this.ahw._inner[f]);
            }
            if (this.indicatorAxis != null) {
                var i = void 0;
                var j = void 0;
                if (f == 0) {
                    i = this.indicatorAxis;
                    j = this.indicatorXAxis;
                }
                else {
                    i = new NumericYAxis();
                    if (!this.indicatorXAxis.ce) {
                        j = this.ag2(0);
                    }
                    else {
                        j = this.ag2(1);
                    }
                }
                j.name = "indicatorXAxis_" + f;
                i.name = "indicatorAxis_" + f;
                g.axes.add(i);
                g.axes.add(j);
                this.aju.add(i);
                this.ajw.add(j);
                if (this.gh) {
                    this.aet(i);
                }
                if (this.gh) {
                    this.aer(j);
                }
            }
        }
        this.apb();
        this.ao5();
        if (this.ahg != null) {
            this.ahg.ae = this.ahw;
        }
        this.an9();
        this.apl();
        this.ajj = false;
    };
    FinancialChart.prototype.uq = function () {
        var e_11, _a, e_12, _b;
        _super.prototype.uq.call(this);
        this.aer(this.xAxis);
        this.aer(this.volumeXAxis);
        if (this.ajw != null) {
            try {
                for (var _c = __values(fromEnum(this.ajw)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var a = _d.value;
                    this.aer(a);
                }
            }
            catch (e_11_1) {
                e_11 = { error: e_11_1 };
            }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return))
                        _a.call(_c);
                }
                finally {
                    if (e_11)
                        throw e_11.error;
                }
            }
        }
        this.aet(this.yAxis);
        this.aet(this.volumeAxis);
        if (this.aju != null) {
            try {
                for (var _e = __values(fromEnum(this.aju)), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var b = _f.value;
                    this.aet(b);
                }
            }
            catch (e_12_1) {
                e_12 = { error: e_12_1 };
            }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return))
                        _b.call(_e);
                }
                finally {
                    if (e_12)
                        throw e_12.error;
                }
            }
        }
        this.aer(this.zoomSliderXAxis);
        this.aet(this.zoomSliderAxis);
    };
    FinancialChart.prototype.get_hd = function () {
        return !this.ajh;
    };
    Object.defineProperty(FinancialChart.prototype, "ajp", {
        get: function () {
            return (this.externalObject._dynamicContent);
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.aps = function (a, b) {
        this.ape(a, b, this.ak6);
        this.an9();
        if (this.zoomSliderChart != null) {
            this.zoomSliderChart.axes.add(this.zoomSliderAxis);
            this.zoomSliderChart.axes.add(this.zoomSliderXAxis);
        }
        this.ao7();
        this.zoomSliderXAxis.itemsSource = this.xAxis.itemsSource;
        this.apc();
        this.apt(a, b);
    };
    FinancialChart.prototype.ao7 = function () {
        if (this.dataChart == null || this.zoomSliderChart == null) {
            return;
        }
        this.zoomSliderChart.series.clear();
        var a = new List$1(Series.$, 0);
        for (var b = 0; b < this.dataChart.series.count; b++) {
            var c = this.dataChart.series._inner[b];
            if (this.ajd(c) || c.eq) {
                continue;
            }
            var d = void 0;
            var e = this.ahu;
            e = this.aht(e);
            var f = this.ale(c, b);
            if (e == 1) {
                switch (this.ahk(this.bj != null && this.bj.dataSeries != null && this.bj.dataSeries.count > f ? this.bj.dataSeries._inner[f] : null)) {
                    case 1:
                        e = 2;
                        break;
                    case 2:
                        e = 3;
                        break;
                    case 3:
                        e = 4;
                        break;
                    case 4:
                        e = 5;
                        break;
                }
            }
            switch (e) {
                case 2:
                    d = this.aip();
                    this.ah2.setDisplayType(d, 1);
                    break;
                case 3:
                    d = this.aip();
                    this.ah2.setDisplayType(d, 0);
                    break;
                case 5:
                    d = this.aio();
                    break;
                case 6:
                    d = this.aim();
                    break;
                default:
                case 4:
                    d = this.ain();
                    break;
            }
            var g = this.bj != null && this.bj.dataSeries != null && this.bj.dataSeries.count > f ? this.bj.dataSeries._inner[f] : null;
            if (d.ew && !d.e4) {
                this.ah4.setCategoryMode(d, 1);
                this.ah4.setXAxis(d, this.zoomSliderXAxis);
                this.ah4.setYAxis(d, this.zoomSliderAxis);
                this.ah4.setValueMemberPath(d, g == null ? null : this.al8(g));
            }
            else if (d.e4) {
                this.ah2.setCategoryMode(d, 1);
                this.ah2.setXAxis(d, this.zoomSliderXAxis);
                this.ah2.setYAxis(d, this.zoomSliderAxis);
                if (g != null) {
                    this.ah2.setOpenMemberPath(d, g.getMemberPathFor(11));
                    this.ah2.setHighMemberPath(d, g.getMemberPathFor(12));
                    this.ah2.setLowMemberPath(d, g.getMemberPathFor(13));
                    this.ah2.setCloseMemberPath(d, g.getMemberPathFor(14));
                    this.ah2.setVolumeMemberPath(d, g.getMemberPathFor(15));
                }
            }
            d.name = "zoomSliderSeries_" + f;
            d.itemsSource = c.itemsSource;
            d.title = c.title;
            d.jn = this.kc;
            a.add(d);
        }
        for (var h = 0; h < a.count; h++) {
            var i = a._inner[h];
            this.zoomSliderChart.series.add(i);
            this.uy(i, h);
        }
    };
    FinancialChart.prototype.ve = function (a, b, c) {
        var e_13, _a;
        _super.prototype.ve.call(this, a, b, c);
        if (a == null || !b) {
            return;
        }
        var d = true;
        var e = true;
        switch (c) {
            case "Brushes":
            case "Outlines":
                d = e = false;
                break;
        }
        if (this.indicatorCharts != null && d) {
            try {
                for (var _b = __values(fromEnum(this.indicatorCharts)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var f = _c.value;
                    a(f);
                }
            }
            catch (e_13_1) {
                e_13 = { error: e_13_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_13)
                        throw e_13.error;
                }
            }
        }
        if (this.volumeChart != null && e) {
            a(this.volumeChart);
        }
        if (this.zoomSliderChart != null) {
            a(this.zoomSliderChart);
        }
    };
    FinancialChart.prototype.aoj = function (a, b) {
        var c = b;
        var d = a;
        var e = -1;
        if (this.indicatorCharts != null) {
            e = this.indicatorCharts.indexOf(d.c7);
        }
        e -= this.ahw == null ? 0 : this.ahw.count;
        if (this.applyCustomIndicators != null) {
            this.applyCustomIndicators(this, new FinancialChartCustomIndicatorArgs(c, d, e));
        }
    };
    FinancialChart.prototype.apg = function (a, b, c) {
        if (a != null) {
            if (b != null) {
                a.axes.remove(b);
            }
            a.axes.add(c);
        }
        if (this.gh) {
            this.aer(c);
        }
    };
    Object.defineProperty(FinancialChart.prototype, "ajo", {
        get: function () {
            return TypeRegistrar.isRegistered("IgxZoomSliderComponent");
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.aok = function (a, b) {
        this.apk();
    };
    FinancialChart.prototype.ajf = function (a, b) {
        if (b <= (0)) {
            return false;
        }
        switch (a) {
            case 5: return true;
            case 4:
            case 3: return b >= timeSpanFromDays(365);
            case 2: return b >= timeSpanFromDays(180);
            case 1: return b >= timeSpanFromDays(90);
            case 0: return b >= timeSpanFromDays(30);
        }
        return false;
    };
    FinancialChart.prototype.apa = function () {
        var e_14, _a;
        var a = this.ahe != null ? this.ahe : new FinancialChartRangeSelectorOptionCollection(1, [5, 4, 3, 2, 1, 0]);
        var b = new FinancialChartRangeSelectorOptionCollection(0);
        var c = this.xAxis;
        var d = +(c.actualMaximumValue) - +(c.actualMinimumValue);
        try {
            for (var _b = __values(fromEnum(a)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var e = _c.value;
                if (this.ajf(e, d)) {
                    b.add(e);
                }
            }
        }
        catch (e_14_1) {
            e_14 = { error: e_14_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_14)
                    throw e_14.error;
            }
        }
        this.ahd = b;
        if (this.ahg != null) {
            this.ahg.u = this.ahd;
        }
    };
    FinancialChart.prototype.aoo = function (a, b) {
        this.apa();
    };
    FinancialChart.prototype.aot = function (a, b) {
        this.ahy = this.ahg.ag;
    };
    FinancialChart.prototype.aos = function (a, b) {
        this.ahw = this.ahg.ae;
    };
    FinancialChart.prototype.aom = function (a, b) {
        this.apk();
    };
    FinancialChart.prototype.aon = function (a, b) {
        this.apm();
    };
    FinancialChart.prototype.aox = function (a, b) {
        var c = a;
        if (c.bg == null) {
            return;
        }
        if (this.ajh) {
            this.ao9(c.bg);
        }
    };
    FinancialChart.prototype.ao0 = function (a) {
        SyncManager.e(a);
        a.axes.clear();
        a.series.clear();
        a.gridAreaRectChanged = delegateRemove(a.gridAreaRectChanged, runOn(this, this.aol));
        a.provideContainer(null);
        if (this.volumeChart == null && this.indicatorCharts == null) {
            this.yAxis.et = this.volumeAxis.et = this.indicatorAxis.et = 0;
            this.xAxis.az.visibility = this.agf;
        }
    };
    FinancialChart.prototype.aoy = function (a, b) {
        var c = a != this.zoomSliderChart;
        if (c) {
            if (this.dataChart != null) {
                a.v1 = new Rect(0, this.dataChart.vl.x, a.v1.y, this.dataChart.vl.width, a.v1.height);
            }
            var d = ((function () {
                var $ret = new SyncSettings();
                $ret.syncChannel = "one";
                $ret.synchronizeHorizontally = true;
                $ret.synchronizeVertically = false;
                return $ret;
            })());
            SyncManager.d(a, d);
            SyncManager.c(a);
        }
        a.gridAreaRectChanged = delegateCombine(a.gridAreaRectChanged, runOn(this, this.aol));
        this.us(a, b);
    };
    FinancialChart.prototype.an9 = function () {
        if (this.indicatorCharts != null || this.volumeChart != null) {
            this.aep(this.xAxis, function (a) { return a.visibility = 1; });
        }
        var _loop_1 = function (a) {
            var b = this_1.ajw._inner[a];
            var c = a < this_1.ajw.count - 1 || this_1.volumeChart != null;
            this_1.aep(b, function (d) { return d.visibility = c ? 1 : 0; });
        };
        var this_1 = this;
        for (var a = 0; a < this.ajw.count; a++) {
            _loop_1(a);
        }
    };
    FinancialChart.prototype.ape = function (a, b, c) {
        if (a != null) {
            this.ao0(a);
        }
        if (b != null) {
            this.aoy(b, c);
        }
    };
    FinancialChart.prototype.apq = function (a, b) {
        this.ape(a, b, this.ak5);
        this.an9();
        if (this.volumeChart != null) {
            if (this.volumeAxis != null) {
                this.volumeChart.axes.add(this.volumeAxis);
                this.volumeChart.axes.add(this.volumeXAxis);
            }
            this.ao6();
            this.volumeXAxis.itemsSource = this.xAxis.itemsSource;
        }
        this.apr(a, b);
    };
    FinancialChart.prototype.aop = function (a, b) {
        this.apo();
    };
    FinancialChart.prototype.apo = function () {
        if (this.ahg == null || +(this.ahg.au) == +(dateMaxValue()) || +(this.ahg.at) == +(dateMinValue()) || this.dataChart == null) {
            return;
        }
        var a = this.xAxis;
        var b = a.getWindowRectForSelection(this.ahg.au, this.ahg.at);
        if (b.width == 0 || isNaN_(b.width) || isInfinity(b.width)) {
            return;
        }
        this.aji = true;
        this.dataChart.v1 = new Rect(0, b.left, this.dataChart.v1.top, b.width, this.dataChart.v1.height);
        if (this.aiy != null) {
            this.aiy.windowRect = this.dataChart.vl;
        }
        if (this.ajh) {
            this.ao9(this.dataChart);
        }
        this.aji = false;
    };
    FinancialChart.prototype.aow = function (a, b) {
        this.apa();
        this.apn(!this.aji);
    };
    FinancialChart.prototype.aoi = function (a, b) {
        var _this = this;
        if (!this.aji) {
            if (this.ajh) {
                this.ao9(a);
            }
            var c = false;
            var d = this.vt(function () { return _this.apn(true); }, c);
            c = d.p1;
            if (this.aiy != null) {
                this.aiy.windowRect = this.dataChart.vl;
            }
        }
    };
    FinancialChart.prototype.api = function (a, b, c, d) {
        var e_15, _a;
        if (a == null || b == null) {
            return;
        }
        if (c != -1) {
            this.aji = true;
            a.v1 = new Rect(0, d, a.vl.top, c, a.vl.height);
            this.aji = false;
        }
        var e = 1.7976931348623157E+308;
        var f = -1.7976931348623157E+308;
        var g = new Rect(0, 0, 0, 1, 1);
        var h = new ScalerParams(0, g, g, this.yAxis.ch, a.vp);
        try {
            for (var _b = __values(fromEnum(a.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var i = _c.value;
                h.b = i.iw();
                var j = i.ay(a.vl);
                var k = j == null || j.c == 1.7976931348623157E+308 || isNaN_(j.c);
                if (k) {
                    continue;
                }
                var l = b.ek(j.c, h);
                var m = b.ek(j.b, h);
                f = Math.max(f, l);
                e = Math.min(e, m);
            }
        }
        catch (e_15_1) {
            e_15 = { error: e_15_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_15)
                    throw e_15.error;
            }
        }
        var n = a.vl.width;
        if (c != -1) {
            n = c;
        }
        var o = a.vl.left;
        if (d != -1) {
            o = d;
        }
        if (e == 1.7976931348623157E+308 || f == -1.7976931348623157E+308 || isNaN_(e) || isNaN_(f)) {
            return;
        }
        this.aji = true;
        var p = new Rect(0, o, e, n, f - e);
        a.v1 = p;
        this.aji = false;
    };
    FinancialChart.prototype.aph = function (a) {
        if (a == null) {
            return;
        }
        if (a == this.dataChart) {
            this.api(this.dataChart, this.yAxis, -1, -1);
            if (this.volumeChart != null) {
                this.api(this.volumeChart, this.volumeAxis, this.dataChart.vl.width, this.dataChart.vl.left);
            }
        }
        else if (a == this.volumeChart) {
            this.api(this.volumeChart, this.volumeAxis, -1, -1);
            this.api(this.dataChart, this.yAxis, this.volumeChart.vl.width, this.volumeChart.vl.left);
        }
    };
    FinancialChart.prototype.ao9 = function (a) {
        var _this = this;
        this.vs(function () { return _this.aph(a); });
    };
    FinancialChart.prototype.aoq = function (a, b) {
        if (this.ajh) {
            this.ao9(a);
        }
    };
    FinancialChart.prototype.app = function () {
        if (!this.ajp) {
            return;
        }
        this.aoc();
        if (this.ahu == 0) {
            this.wl();
            this.zoomSliderChart = null;
            this.ak6.rootWrapper.removeChildren();
        }
        else if (this.zoomSliderChart == null) {
            this.zoomSliderChart = ((function () {
                var $ret = new XamDataChart();
                $ret.name = "ZoomSliderChart";
                return $ret;
            })());
            this.apc();
        }
        else {
            this.ao7();
        }
    };
    FinancialChart.prototype.aoc = function () {
        var container_ = this.ak6.rootWrapper.getNativeElement();
        var containerParent_ = container_.parentElement;
        if (this.aiy != null && containerParent_ != null && !this.ajq) {
            this.ajq = true;
            if (TypeRegistrar.isRegistered("IgxZoomSliderComponent")) {
                var z_ = this.aiy;
                var f_ = this.externalObject;
                var sliderComponent_ = TypeRegistrar.get("IgxZoomSliderComponent");
                var sliderInstance_ = TypeRegistrar.create("IgxZoomSliderComponent");
                sliderInstance_._implementation = z_;
                sliderComponent_._staticStyling(container_, sliderInstance_, f_);
            }
        }
    };
    FinancialChart.prototype.apj = function () {
        var _this = this;
        if (this.ahp != 1) {
            return;
        }
        this.aev(function (a) {
            var e_16, _a;
            var b;
            b = a.breaks;
            b.clear();
            if (_this.xAxisBreaks == null) {
                return;
            }
            try {
                for (var _b = __values(fromEnum(_this.xAxisBreaks)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var c = _c.value;
                    b.add(c);
                }
            }
            catch (e_16_1) {
                e_16 = { error: e_16_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_16)
                        throw e_16.error;
                }
            }
        });
    };
    FinancialChart.prototype.wm = function (a, b, c) {
        _super.prototype.wm.call(this, a, b, c);
        if (this.bj.dataSeries.count > 0) {
            if (this.ahp == 1 && stringIsNullOrEmpty(this.nk)) {
                this.ci.setXMemberPath(a, c.getMemberPathFor(17));
            }
            if (stringIsNullOrEmpty(this.nm)) {
                this.ci.setYMemberPath(a, this.al8(c));
            }
            if (stringIsNullOrEmpty(this.nc)) {
                this.ci.setLabelMemberPath(a, this.al8(c));
            }
            if (stringIsNullOrEmpty(this.m9)) {
                this.ci.setContentMemberPath(a, this.al8(c));
            }
        }
    };
    FinancialChart.prototype.al7 = function (a) {
        switch (a) {
            case 0: return "Absolute Volume Oscillator";
            case 1: return "Accumulation/Distribution Indicator";
            case 2: return "Average Directional Index";
            case 3: return "Average True Range Indicator";
            case 4: return "Bollinger Bandwidth";
            case 5: return "Chaikin Oscillator";
            case 6: return "Chaikin Volatility Indicator";
            case 7: return "Commodity Channel Index";
            case 8: return "Detrended Price Oscillator";
            case 9: return "Ease of Movement Indicator";
            case 10: return "Fast Stochastic Oscillator";
            case 11: return "Force Index Indicator";
            case 12: return "Full Stochastic Oscillator";
            case 13: return "Market Facilitation Index";
            case 14: return "Mass Index Indicator";
            case 15: return "Median Price Indicator";
            case 16: return "Money Flow Index";
            case 17: return "MACD Indicator";
            case 18: return "Negative Volume Index";
            case 19: return "On Balance Volume Indicator";
            case 20: return "Percentage Price Oscillator";
            case 21: return "Percentage Volume Oscillator";
            case 22: return "Positive Volume Index";
            case 23: return "Price Volume Trend";
            case 24: return "ROC - Rate of Change Indicator";
            case 25: return "RSI - Relative Strength Index";
            case 26: return "Slow Stochastic Oscillator";
            case 27: return "Standard Deviation Indicator";
            case 28: return "Stoch RSI";
            case 29: return "TRIX Indicator";
            case 30: return "Typical Price Indicator";
            case 31: return "Ultimate Oscillator";
            case 32: return "Weighted Close Indicator";
            case 33: return "Williams %R Indicator";
            default: return "Unknown";
        }
    };
    FinancialChart.prototype.aer = function (a) {
        var e_17, _a;
        var _this = this;
        _super.prototype.aer.call(this, a);
        var b = a;
        this.aoz(b);
        if (a.bg != null) {
            try {
                for (var _b = __values(fromEnum(a.bg.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var c = _c.value;
                    if (this.aja(c)) {
                        this.ah4.setXAxis(c, b);
                    }
                    else if (this.aje(c)) {
                        this.ah2.setXAxis(c, b);
                    }
                    else if (this.ajb(c)) {
                        this.ah6.setXAxis(c, b);
                    }
                    else if (this.ajd(c)) {
                        this.aib.setXAxis(c, b);
                    }
                }
            }
            catch (e_17_1) {
                e_17 = { error: e_17_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_17)
                        throw e_17.error;
                }
            }
        }
        this.ao3(function (d) {
            d.zoomMaximumCategoryRange = _this.akq;
            d.zoomMaximumItemSpan = _this.akr;
            d.zoomToCategoryRange = _this.aks;
            d.zoomToCategoryStart = _this.akt;
            d.zoomToItemSpan = _this.aku;
        });
        var d = a == this.zoomSliderXAxis;
        if (d) {
            a.jx = this.apv;
            a.es = this.ak1;
            this.aep(a, function (e) { return e.visibility = 1; });
        }
        var e = a;
        e.minimumValue = this.aj2;
        e.maximumValue = this.aj1;
        a.b1 = this.ajk;
    };
    FinancialChart.prototype.aet = function (a) {
        var e_18, _a;
        _super.prototype.aet.call(this, a);
        if (a.bg != null && a == this.yAxis) {
            try {
                for (var _b = __values(fromEnum(a.bg.series)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var b = _c.value;
                    if (this.aja(b)) {
                        this.ah4.setYAxis(b, this.yAxis);
                    }
                    else if (this.aje(b)) {
                        this.ah2.setYAxis(b, this.yAxis);
                    }
                    else if (this.ajb(b)) {
                        this.ah6.setYAxis(b, this.yAxis);
                    }
                    else if (this.ajd(b)) {
                        this.aib.setYAxis(b, this.yAxis);
                    }
                }
            }
            catch (e_18_1) {
                e_18 = { error: e_18_1 };
            }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return))
                        _a.call(_b);
                }
                finally {
                    if (e_18)
                        throw e_18.error;
                }
            }
        }
        a.lb = this.ajl;
        a.l6 = this.akz;
        a.l5 = this.aky;
        if (a == this.yAxis) {
            a.l2 = this.akx;
        }
        var c = a == this.zoomSliderAxis;
        if (c) {
            this.aep(a, function (d) { return d.visibility = 1; });
        }
        a.b1 = this.ajm;
    };
    FinancialChart.prototype.aoz = function (a) {
        a.itemsSource = this.dataChart != null && this.dataChart.series.count > 0 ? this.dataChart.series._inner[0].itemsSource : null;
        a.f1 = XYChart.ace(a, this.bj.dataSeries);
        if (this.bj.dataSeries.count > 0) {
            var b = this.bj.dataSeries._inner[0].getMemberPathFor(17);
            a.hz(b);
        }
    };
    FinancialChart.prototype.uy = function (a, b) {
        if (a == null) {
            return;
        }
        var c = this.zoomSliderChart == a.c7;
        var d = this.dataChart != null && this.dataChart.series.contains(a);
        if (d && !a.eq && !this.ajd(a)) {
            a.l1 = "Financial Series";
        }
        _super.prototype.uy.call(this, a, b);
        if ((d || c) && a.fk) {
            a.rx(this.agr != null ? ArrayExtension.getModulus$1(Brush.$, this.agr, b) : null, this.ags != null ? ArrayExtension.getModulus$1(Brush.$, this.ags, b) : null);
        }
        var e = this.ajb(a);
        if (this.ajd(a)) {
            a.u9 = this.agt != null ? ArrayExtension.getModulus$1(Brush.$, this.agt, b) : null;
            a.vp = this.agu != null ? ArrayExtension.getModulus$1(Brush.$, this.agu, b) : null;
            a.jn = this.akn;
        }
        else if (e) {
            a.u9 = this.agp != null ? ArrayExtension.getModulus$1(Brush.$, this.agp, b) : null;
            a.jn = this.akl;
        }
        else if (!a.eq && (d || c)) {
            a.u9 = this.ab != null ? ArrayExtension.getModulus$1(Brush.$, this.ab, b) : null;
            a.vp = this.af != null ? ArrayExtension.getModulus$1(Brush.$, this.af, b) : null;
        }
        var f = this.volumeChart != null && this.volumeChart == a.c7;
        if (f) {
            a.jn = this.akp;
        }
        a.legend = c || f || e || !a.isUsableInLegend ? null : this.legend != null ? this.legend : this.ajc ? this.ah0 : null;
    };
    FinancialChart.prototype.uz = function (a, b) {
        _super.prototype.uz.call(this, a, b);
        if (a == null) {
            return;
        }
    };
    FinancialChart.prototype.vl = function (a, b, c) {
        var e_19, _a, e_20, _b;
        var _this = this;
        _super.prototype.vl.call(this, a, b, c);
        switch (a) {
            case "YAxisIsLogarithmic":
                this.yAxis.lj = this.ajn;
                if (this.volumeAxis != null) {
                    this.volumeAxis.lj = this.ajn;
                }
                if (this.zoomSliderAxis != null) {
                    this.zoomSliderAxis.lj = this.ajn;
                }
                break;
            case "YAxisLogarithmBase":
                this.yAxis.ma = this.all;
                if (this.volumeAxis != null) {
                    this.volumeAxis.ma = this.all;
                }
                if (this.zoomSliderAxis != null) {
                    this.zoomSliderAxis.ma = this.all;
                }
                break;
            case "VolumeType":
                var d = EnumUtil.getEnumValue(FinancialChartVolumeType_$type, b);
                if (this.ahn == 0) {
                    this.volumeChart = null;
                }
                else if (this.volumeChart == null) {
                    this.volumeChart = ((function () {
                        var $ret = new XamDataChart();
                        $ret.name = "VolumeChart";
                        return $ret;
                    })());
                }
                else {
                    this.ao6();
                }
                break;
            case "VolumeChart":
                this.apq(typeCast(XamDataChart.$, b), typeCast(XamDataChart.$, c));
                break;
            case "DataChart":
                if (typeCast(XamDataChart.$, b) !== null) {
                    var e = b;
                    e.gridAreaRectChanged = delegateRemove(e.gridAreaRectChanged, runOn(this, this.aol));
                    e.actualWindowRectChanged = delegateRemove(e.actualWindowRectChanged, runOn(this, this.aoi));
                    e.seriesDataUpdated = delegateRemove(e.seriesDataUpdated, runOn(this, this.aoq));
                }
                if (this.dataChart != null) {
                    var f = this.dataChart;
                    f.gridAreaRectChanged = delegateCombine(f.gridAreaRectChanged, runOn(this, this.aol));
                    var g = this.dataChart;
                    g.actualWindowRectChanged = delegateCombine(g.actualWindowRectChanged, runOn(this, this.aoi));
                    var h = this.dataChart;
                    h.seriesDataUpdated = delegateCombine(h.seriesDataUpdated, runOn(this, this.aoq));
                }
                this.app();
                break;
            case "XAxisMode":
                this.xAxis = this.ag2(this.ahp);
                this.xAxis.name = "xAxis";
                this.volumeXAxis = this.ag2(this.ahp);
                this.volumeXAxis.name = "volumeXAxis";
                this.indicatorXAxis = this.ag2(this.ahp);
                this.indicatorXAxis.name = "indicatorXAxis_0";
                this.zoomSliderXAxis = this.ag2(this.ahp);
                this.zoomSliderXAxis.name = "zoomSliderXAxis";
                this.aep(this.zoomSliderXAxis, function (i) { return i.visibility = 1; });
                this.an9();
                if (this.ahp == 1) {
                    this.apj();
                }
                break;
            case "YAxisMode":
                this.yAxis = this.aig(this.ahr);
                this.yAxis.name = "yAxis";
                this.zoomSliderAxis = this.aig(this.ahr);
                this.zoomSliderAxis.name = "zoomSliderAxis";
                this.volumeAxis = this.aig(this.ahr);
                this.volumeAxis.name = "volumeAxis";
                break;
            case "Toolbar":
                var i = typeCast(FinancialChartToolbar.$, b);
                if (i != null) {
                    i.rangeChanged = delegateRemove(i.rangeChanged, runOn(this, this.aop));
                    i.indicatorTypesChanged = delegateRemove(i.indicatorTypesChanged, runOn(this, this.aos));
                    i.overlayTypesChanged = delegateRemove(i.overlayTypesChanged, runOn(this, this.aot));
                }
                if (this.ahg != null) {
                    this.ahg.ae = this.ahw;
                    this.ahg.ag = this.ahy;
                    this.ahg.u = this.ahd;
                    this.apn(true);
                    var j = this.ahg;
                    j.rangeChanged = delegateCombine(j.rangeChanged, runOn(this, this.aop));
                    var k = this.ahg;
                    k.indicatorTypesChanged = delegateCombine(k.indicatorTypesChanged, runOn(this, this.aos));
                    var l = this.ahg;
                    l.overlayTypesChanged = delegateCombine(l.overlayTypesChanged, runOn(this, this.aot));
                }
                break;
            case "XAxis":
                var m = typeCast(CategoryAxisBase.$, b);
                if (m != null) {
                    if (this.dataChart != null) {
                        this.dataChart.axes.remove(m);
                    }
                }
                var n = typeCast(ITimeRangeSelectorAxis_$type, b);
                if (n != null) {
                    n.visibleRangeChanged = delegateRemove(n.visibleRangeChanged, runOn(this, this.aow));
                }
                if (this.dataChart != null) {
                    this.dataChart.axes.add(this.xAxis);
                }
                if (this.gh) {
                    this.aer(this.xAxis);
                    this.ug();
                }
                this.apn(false);
                this.apo();
                this.apa();
                var o = this.xAxis;
                o.visibleRangeChanged = delegateCombine(o.visibleRangeChanged, runOn(this, this.aow));
                break;
            case "XAxisZoomMaximumCategoryRange":
                this.ao3(function (p) { return p.zoomMaximumCategoryRange = _this.akq; });
                break;
            case "XAxisZoomMaximumItemSpan":
                this.ao3(function (p) { return p.zoomMaximumItemSpan = _this.akr; });
                break;
            case "XAxisZoomToCategoryRange":
                this.ao3(function (p) { return p.zoomToCategoryRange = _this.aks; });
                break;
            case "XAxisZoomToCategoryStart":
                this.ao3(function (p) { return p.zoomToCategoryStart = _this.akt; });
                break;
            case "XAxisZoomToItemSpan":
                this.ao3(function (p) { return p.zoomToItemSpan = _this.aku; });
                break;
            case "ChartType":
                this.ud();
                break;
            case "VolumeXAxis":
                this.apg(this.volumeChart, typeCast(Axis.$, b), this.volumeXAxis);
                break;
            case "ZoomSliderXAxis":
                this.apg(this.zoomSliderChart, typeCast(Axis.$, b), this.zoomSliderXAxis);
                break;
            case "IndicatorXAxis":
                this.apk();
                break;
            case "YAxis":
                var p = typeCast(NumericYAxis.$, b);
                if (p != null) {
                    if (this.dataChart != null) {
                        this.dataChart.axes.remove(p);
                    }
                }
                if (this.dataChart != null) {
                    this.dataChart.axes.add(this.yAxis);
                }
                if (this.gh) {
                    this.aet(this.yAxis);
                    this.aet(this.yAxis);
                }
                var q = this.yAxis;
                q.rangeChanged = delegateCombine(q.rangeChanged, runOn(this, this.aox));
                break;
            case "VolumeAxis":
                var r = typeCast(Axis.$, b);
                if (this.volumeChart != null && r != null) {
                    this.volumeChart.axes.remove(r);
                }
                if (this.volumeChart != null && this.volumeAxis != null) {
                    this.volumeChart.axes.add(this.volumeAxis);
                    this.ao6();
                }
                if (this.gh) {
                    this.aet(this.volumeAxis);
                }
                var s = this.volumeAxis;
                s.rangeChanged = delegateCombine(s.rangeChanged, runOn(this, this.aox));
                break;
            case "IndicatorAxis":
                if (this.indicatorCharts != null) {
                    for (var t = 0; t < this.indicatorCharts.count; t++) {
                        var u = this.indicatorCharts._inner[t];
                        var v = new List$1(Axis.$, 0);
                        try {
                            for (var _c = (e_19 = void 0, __values(fromEnum(u.axes))), _d = _c.next(); !_d.done; _d = _c.next()) {
                                var w = _d.value;
                                if (typeCast(NumericYAxis.$, w) !== null) {
                                    v.add(w);
                                }
                            }
                        }
                        catch (e_19_1) {
                            e_19 = { error: e_19_1 };
                        }
                        finally {
                            try {
                                if (_d && !_d.done && (_a = _c.return))
                                    _a.call(_c);
                            }
                            finally {
                                if (e_19)
                                    throw e_19.error;
                            }
                        }
                        try {
                            for (var _e = (e_20 = void 0, __values(fromEnum(v))), _f = _e.next(); !_f.done; _f = _e.next()) {
                                var x = _f.value;
                                u.axes.remove(x);
                            }
                        }
                        catch (e_20_1) {
                            e_20 = { error: e_20_1 };
                        }
                        finally {
                            try {
                                if (_f && !_f.done && (_b = _e.return))
                                    _b.call(_e);
                            }
                            finally {
                                if (e_20)
                                    throw e_20.error;
                            }
                        }
                        var y = null;
                        if (t == 0) {
                            y = this.indicatorAxis;
                        }
                        else if (this.ahr == 1) {
                            y = this.aig(this.ahr);
                        }
                        else if (typeCast(NumericYAxis.$, this.indicatorAxis) !== null) {
                            y = new NumericYAxis();
                        }
                        if (y != null) {
                            u.axes.add(y);
                        }
                        if (this.gh) {
                            this.aet(y);
                        }
                    }
                }
                break;
            case "IsWindowSyncedToVisibleRange":
                if (this.dataChart != null) {
                    this.dataChart.wz = this.he && this.hd;
                }
                if (this.volumeChart != null) {
                    this.volumeChart.wz = this.he && this.hd;
                }
                if (this.ajh) {
                    this.aph(this.dataChart);
                }
                else {
                    this.volumeChart.x6 = 1;
                }
                break;
            case "IndicatorTypes":
                var z = typeCast(ObservableCollection$1.$.specialize(FinancialIndicatorType_$type), b);
                if (z != null) {
                    z.collectionChanged = delegateRemove(z.collectionChanged, runOn(this, this.aom));
                }
                var aa = typeCast(ObservableCollection$1.$.specialize(FinancialIndicatorType_$type), c);
                if (aa != null) {
                    aa.collectionChanged = delegateCombine(aa.collectionChanged, runOn(this, this.aom));
                }
                this.apk();
                break;
            case "OverlayTypes":
                var ab = typeCast(ObservableCollection$1.$.specialize(FinancialOverlayType_$type), b);
                if (ab != null) {
                    ab.collectionChanged = delegateRemove(ab.collectionChanged, runOn(this, this.aon));
                }
                var ac = typeCast(ObservableCollection$1.$.specialize(FinancialOverlayType_$type), c);
                if (ac != null) {
                    ac.collectionChanged = delegateCombine(ac.collectionChanged, runOn(this, this.aon));
                }
                this.apm();
                break;
            case "YAxisAbbreviateLargeNumbers":
                this.ae0(function (ad) { return ad.lb = _this.ajl; });
                break;
            case "YAxisMinimumValue":
                this.ae1(function (ad) { return ad.l6 = _this.akz; }, false);
                break;
            case "YAxisMaximumValue":
                this.ae1(function (ad) { return ad.l5 = _this.aky; }, false);
                break;
            case "YAxisInterval":
                this.ae1(function (ad) { return ad.l2 = _this.akx; }, false);
                break;
            case "Thickness":
            case "ZoomSliderType":
                this.app();
                break;
            case "ZoomSliderChart":
                if (!this.ajo) {
                    return;
                }
                this.aps(typeCast(XamDataChart.$, b), typeCast(XamDataChart.$, c));
                break;
            case "NegativeBrushes":
            case "NegativeOutlines":
                this.aof(function (ad, ae) {
                    if (ad.fk) {
                        ad.rx(_this.agr != null ? ArrayExtension.getModulus$1(Brush.$, _this.agr, ae) : null, _this.ags != null ? ArrayExtension.getModulus$1(Brush.$, _this.ags, ae) : null);
                    }
                }, true);
                break;
            case "OverlayBrushes":
                this.aoe(function (ad, ae) { return ad.u9 = _this.agt != null ? ArrayExtension.getModulus$1(Brush.$, _this.agt, ae) : null; });
                break;
            case "OverlayOutlines":
                this.aoe(function (ad, ae) { return ad.vp = _this.agu != null ? ArrayExtension.getModulus$1(Brush.$, _this.agu, ae) : null; });
                break;
            case "OverlayThickness":
                this.aoe(function (ad, ae) { return ad.jn = _this.akn; });
                break;
            case "OverlayMultiplier":
                this.aoe(function (ad, ae) { return _this.aib.setMultiplier(ad, _this.akm); });
                break;
            case "VolumeBrushes":
                this.aog(function (ad, ae) { return ad.u9 = ArrayExtension.getModulus$1(Brush.$, _this.agv, ae); });
                break;
            case "VolumeOutlines":
                this.aog(function (ad, ae) { return ad.vp = ArrayExtension.getModulus$1(Brush.$, _this.agw, ae); });
                break;
            case "VolumeThickness":
                this.aog(function (ad, ae) { return ad.jn = _this.akp; });
                break;
            case "IndicatorNegativeBrushes":
                this.aod(function (ad, ae) { return _this.ah6.setNegativeBrush(ad, ArrayExtension.getModulus$1(Brush.$, _this.agq, ae)); });
                break;
            case "IndicatorBrushes":
                this.aod(function (ad, ae) { return ad.u9 = ArrayExtension.getModulus$1(Brush.$, _this.agp, ae); });
                break;
            case "IndicatorThickness":
                this.aod(function (ad, ae) { return ad.jn = _this.akl; });
                break;
            case "IndicatorDisplayTypes":
                this.aod(function (ad, ae) { return _this.ah6.setDisplayType(ad, ArrayExtension.getModulus$1(IndicatorDisplayType_$type, _this.ah8, ae)); });
                break;
            case "IndicatorPeriod":
                this.aod(function (ad, ae) {
                    if (_this.alg == -2147483648) {
                        return;
                    }
                    _this.ah6.setPeriodOverride(ad, _this.alg);
                });
                break;
            case "IndicatorMultiplier":
                this.aod(function (ad, ae) { return _this.ah6.setMultiplier(ad, _this.akk); });
                break;
            case "IndicatorSmoothingPeriod":
                this.aod(function (ad, ae) { return _this.ah6.setSmoothingPeriod(ad, _this.alj); });
                break;
            case "IndicatorShortPeriod":
                this.aod(function (ad, ae) { return _this.ah6.setShortPeriod(ad, _this.alh); });
                break;
            case "IndicatorLongPeriod":
                this.aod(function (ad, ae) { return _this.ah6.setLongPeriod(ad, _this.alf); });
                break;
            case "IndicatorSignalPeriod":
                this.aod(function (ad, ae) { return _this.ah6.setSignalPeriod(ad, _this.ali); });
                break;
            case "RangeSelectorOptions":
                var ad = typeCast(FinancialChartRangeSelectorOptionCollection.$, b);
                if (ad != null) {
                    ad.collectionChanged = delegateRemove(ad.collectionChanged, runOn(this, this.aoo));
                }
                if (this.ahe != null) {
                    var ae = this.ahe;
                    ae.collectionChanged = delegateCombine(ae.collectionChanged, runOn(this, this.aoo));
                }
                this.apa();
                break;
            case "CustomIndicatorNames":
                var af = typeCast(ObservableCollection$1.$.specialize(String_$type), b);
                if (af != null) {
                    af.collectionChanged = delegateRemove(af.collectionChanged, runOn(this, this.aok));
                }
                if (this.customIndicatorNames != null) {
                    var ag = this.customIndicatorNames;
                    ag.collectionChanged = delegateCombine(ag.collectionChanged, runOn(this, this.aok));
                }
                this.apk();
                break;
            case "Legend":
                if (this.zoomSliderChart != null) {
                    this.zoomSliderChart.legend = null;
                }
                break;
            case "Brushes":
                this.aof(function (ah, ai) { return ah.u9 = _this.ab != null ? ArrayExtension.getModulus$1(Brush.$, _this.ab, ai) : null; }, true);
                break;
            case "Outlines":
                this.aof(function (ah, ai) { return ah.vp = _this.af != null ? ArrayExtension.getModulus$1(Brush.$, _this.af, ai) : null; }, true);
                break;
            case "XAxisLabelVisibility":
                if (this.agf == 0 && this.zoomSliderXAxis != null) {
                    this.aep(this.zoomSliderXAxis, function (ah) { return ah.visibility = 1; });
                }
                this.an9();
                break;
            case "YAxisLabelVisibility":
                if (this.agg != 0 || this.zoomSliderAxis == null) {
                    break;
                }
                this.aep(this.zoomSliderAxis, function (ah) { return ah.visibility = 1; });
                break;
            case "ZoomSliderXAxisMajorStroke":
                this.zoomSliderXAxis.jx = this.apv;
                break;
            case "ZoomSliderXAxisMajorStrokeThickness":
                this.zoomSliderXAxis.es = this.ak1;
                break;
            case "FinancialChartLegend":
            case "IsLegendVisible":
                var ah_1 = this.ajc && this.legend == null;
                {
                    var ai = function (aj, ak) {
                        if (aj.isUsableInLegend) {
                            aj.legend = ah_1 ? _this.ah0 : null;
                        }
                    };
                    this.aof(ai, false);
                    this.aoe(ai);
                }
                break;
            case "ZoomSliderAxis":
                var aj = typeCast(Axis.$, b);
                if (aj != null && this.zoomSliderChart != null) {
                    this.zoomSliderChart.axes.remove(aj);
                }
                if (this.zoomSliderAxis != null) {
                    if (this.zoomSliderChart != null) {
                        this.zoomSliderChart.axes.add(this.zoomSliderAxis);
                        this.ao7();
                        this.zoomSliderXAxis.itemsSource = this.xAxis.itemsSource;
                    }
                    if (this.gh) {
                        this.aet(this.zoomSliderAxis);
                    }
                    this.aep(this.zoomSliderAxis, function (ak) { return ak.visibility = 1; });
                }
                break;
            case "XAxisMinimumValue":
                this.aev(function (ak) {
                    var al = typeCast(ITimeRangeSelectorAxis_$type, ak);
                    if (al != null) {
                        al.minimumValue = _this.aj2;
                    }
                });
                break;
            case "XAxisMaximumValue":
                this.aev(function (ak) {
                    var al = typeCast(ITimeRangeSelectorAxis_$type, ak);
                    if (al != null) {
                        al.maximumValue = _this.aj1;
                    }
                });
                break;
            case FinancialChart.ang:
                this.aew(function (ak) { return ak.b1 = _this.ajk; }, a);
                break;
            case FinancialChart.anr:
                this.ae0(function (ak) { return ak.b1 = _this.ajm; });
                break;
        }
    };
    FinancialChart.prototype.get_cf = function () {
        return 1;
    };
    Object.defineProperty(FinancialChart.prototype, "cf", {
        get: function () {
            return this.get_cf();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ajn", {
        get: function () {
            return this.ai5;
        },
        set: function (a) {
            var b = this.ajn;
            if (a != b) {
                this.ai5 = a;
                this.vl("YAxisIsLogarithmic", b, this.ajn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "all", {
        get: function () {
            return this.ald;
        },
        set: function (a) {
            var b = this.all;
            if (a != b) {
                this.ald = a;
                this.vl("YAxisLogarithmBase", b, this.all);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akx", {
        get: function () {
            return this.ake;
        },
        set: function (a) {
            var b = this.akx;
            if (a != b) {
                this.ake = a;
                this.vl("YAxisInterval", b, this.akx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akz", {
        get: function () {
            return this.akg;
        },
        set: function (a) {
            var b = this.akz;
            if (a != b) {
                this.akg = a;
                this.vl("YAxisMinimumValue", b, this.akz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "aky", {
        get: function () {
            return this.akf;
        },
        set: function (a) {
            var b = this.aky;
            if (a != b) {
                this.akf = a;
                this.vl("YAxisMaximumValue", b, this.aky);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ak0", {
        get: function () {
            return this.akh;
        },
        set: function (a) {
            var b = this.ak0;
            if (a != b) {
                this.akh = a;
                this.vl("YAxisMinorInterval", b, this.ak0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akw", {
        get: function () {
            return this.yAxis.ly;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akv", {
        get: function () {
            return this.yAxis.lx;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ahn", {
        get: function () {
            return this.ahl;
        },
        set: function (a) {
            var b = this.ahn;
            if (a != b) {
                this.ahl = a;
                this.vl("VolumeType", enumGetBox(FinancialChartVolumeType_$type, b), enumGetBox(FinancialChartVolumeType_$type, this.ahn));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ahp", {
        get: function () {
            return this.aho;
        },
        set: function (a) {
            var b = this.ahp;
            if (a != b) {
                this.aho = a;
                this.vl("XAxisMode", enumGetBox(FinancialChartXAxisMode_$type, b), enumGetBox(FinancialChartXAxisMode_$type, this.ahp));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akq", {
        get: function () {
            return this.aj9;
        },
        set: function (a) {
            var b = this.akq;
            if (a != b) {
                this.aj9 = a;
                this.vl("XAxisZoomMaximumCategoryRange", b, this.akq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akt", {
        get: function () {
            return this.akc;
        },
        set: function (a) {
            var b = this.akt;
            if (a != b) {
                this.akc = a;
                this.vl("XAxisZoomToCategoryStart", b, this.akt);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akr", {
        get: function () {
            return this.aka;
        },
        set: function (a) {
            var b = this.akr;
            if (a != b) {
                this.aka = a;
                this.vl("XAxisZoomMaximumItemSpan", b, this.akr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "aks", {
        get: function () {
            return this.akb;
        },
        set: function (a) {
            var b = this.aks;
            if (a != b) {
                this.akb = a;
                this.vl("XAxisZoomToCategoryRange", b, this.aks);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "aku", {
        get: function () {
            return this.akd;
        },
        set: function (a) {
            var b = this.aku;
            if (a != b) {
                this.akd = a;
                this.vl("XAxisZoomToItemSpan", b, this.aku);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ahr", {
        get: function () {
            return this.ahq;
        },
        set: function (a) {
            var b = this.ahr;
            if (a != b) {
                this.ahq = a;
                this.vl("YAxisMode", enumGetBox(FinancialChartYAxisMode_$type, b), enumGetBox(FinancialChartYAxisMode_$type, this.ahr));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ajg", {
        get: function () {
            return this.ai0;
        },
        set: function (a) {
            var b = this.ajg;
            if (a != b) {
                this.ai0 = a;
                this.vl("IsToolbarVisible", b, this.ajg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ahi", {
        get: function () {
            return this.ahh;
        },
        set: function (a) {
            var b = this.ahi;
            if (a != b) {
                this.ahh = a;
                this.vl("ChartType", enumGetBox(FinancialChartType_$type, b), enumGetBox(FinancialChartType_$type, this.ahi));
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.get_afa = function () {
        return _super.prototype.get_afa.call(this);
    };
    FinancialChart.prototype.set_afa = function (a) {
        _super.prototype.set_afa.call(this, a);
    };
    Object.defineProperty(FinancialChart.prototype, "ajh", {
        get: function () {
            return this.ai1;
        },
        set: function (a) {
            var b = this.ajh;
            if (a != b) {
                this.ai1 = a;
                this.vl("IsWindowSyncedToVisibleRange", b, this.ajh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ahw", {
        get: function () {
            return this.ahv;
        },
        set: function (a) {
            var b = this.ahw;
            if (a != b) {
                this.ahv = a;
                this.vl("IndicatorTypes", b, this.ahw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ahy", {
        get: function () {
            return this.ahx;
        },
        set: function (a) {
            var b = this.ahy;
            if (a != b) {
                this.ahx = a;
                this.vl("OverlayTypes", b, this.ahy);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "indicatorCharts", {
        get: function () {
            return this.ajs;
        },
        set: function (a) {
            var b = this.indicatorCharts;
            if (a != b) {
                this.ajs = a;
                this.vl("IndicatorCharts", b, this.indicatorCharts);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ajl", {
        get: function () {
            return this.ai3;
        },
        set: function (a) {
            var b = this.ajl;
            if (a != b) {
                this.ai3 = a;
                this.vl("YAxisAbbreviateLargeNumbers", b, this.ajl);
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.get_j6 = function () {
        return _super.prototype.get_j6.call(this);
    };
    FinancialChart.prototype.set_j6 = function (a) {
        _super.prototype.set_j6.call(this, a);
    };
    Object.defineProperty(FinancialChart.prototype, "ajk", {
        get: function () {
            return this.ai2;
        },
        set: function (a) {
            var b = this.ajk;
            if (a != b) {
                this.ai2 = a;
                this.vl(FinancialChart.ang, b, this.ajk);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ajm", {
        get: function () {
            return this.ai4;
        },
        set: function (a) {
            var b = this.ajm;
            if (a != b) {
                this.ai4 = a;
                this.vl(FinancialChart.anr, b, this.ajm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "zoomSliderChart", {
        get: function () {
            return this.aiu;
        },
        set: function (a) {
            var b = this.zoomSliderChart;
            if (a != b) {
                this.aiu = a;
                this.vl("ZoomSliderChart", b, this.zoomSliderChart);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "zoomSliderAxis", {
        get: function () {
            return this.aif;
        },
        set: function (a) {
            var b = this.zoomSliderAxis;
            if (a != b) {
                this.aif = a;
                this.vl("ZoomSliderAxis", b, this.zoomSliderAxis);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "zoomSliderXAxis", {
        get: function () {
            return this.ag0;
        },
        set: function (a) {
            var b = this.zoomSliderXAxis;
            if (a != b) {
                this.ag0 = a;
                this.vl("ZoomSliderXAxis", b, this.zoomSliderXAxis);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ahu", {
        get: function () {
            return this.ahs;
        },
        set: function (a) {
            var b = this.ahu;
            if (a != b) {
                this.ahs = a;
                this.vl("ZoomSliderType", enumGetBox(FinancialChartZoomSliderType_$type, b), enumGetBox(FinancialChartZoomSliderType_$type, this.ahu));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "agr", {
        get: function () {
            return this.agj;
        },
        set: function (a) {
            var b = this.agr;
            if (a != b) {
                this.agj = a;
                this.vl("NegativeBrushes", b, this.agr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ags", {
        get: function () {
            return this.agk;
        },
        set: function (a) {
            var b = this.ags;
            if (a != b) {
                this.agk = a;
                this.vl("NegativeOutlines", b, this.ags);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "agt", {
        get: function () {
            return this.agl;
        },
        set: function (a) {
            var b = this.agt;
            if (a != b) {
                this.agl = a;
                this.vl("OverlayBrushes", b, this.agt);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "agu", {
        get: function () {
            return this.agm;
        },
        set: function (a) {
            var b = this.agu;
            if (a != b) {
                this.agm = a;
                this.vl("OverlayOutlines", b, this.agu);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "agw", {
        get: function () {
            return this.ago;
        },
        set: function (a) {
            var b = this.agw;
            if (a != b) {
                this.ago = a;
                this.vl("VolumeOutlines", b, this.agw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "agv", {
        get: function () {
            return this.agn;
        },
        set: function (a) {
            var b = this.agv;
            if (a != b) {
                this.agn = a;
                this.vl("VolumeBrushes", b, this.agv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "agq", {
        get: function () {
            return this.agi;
        },
        set: function (a) {
            var b = this.agq;
            if (a != b) {
                this.agi = a;
                this.vl("IndicatorNegativeBrushes", b, this.agq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "agp", {
        get: function () {
            return this.agh;
        },
        set: function (a) {
            var b = this.agp;
            if (a != b) {
                this.agh = a;
                this.vl("IndicatorBrushes", b, this.agp);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akp", {
        get: function () {
            return this.aj8;
        },
        set: function (a) {
            var b = this.akp;
            if (a != b) {
                this.aj8 = a;
                this.vl("VolumeThickness", b, this.akp);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akn", {
        get: function () {
            return this.aj6;
        },
        set: function (a) {
            var b = this.akn;
            if (a != b) {
                this.aj6 = a;
                this.vl("OverlayThickness", b, this.akn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akm", {
        get: function () {
            return this.aj5;
        },
        set: function (a) {
            var b = this.akm;
            if (a != b) {
                this.aj5 = a;
                this.vl("OverlayMultiplier", b, this.akm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akl", {
        get: function () {
            return this.aj4;
        },
        set: function (a) {
            var b = this.akl;
            if (a != b) {
                this.aj4 = a;
                this.vl("IndicatorThickness", b, this.akl);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ah8", {
        get: function () {
            return this.ah7;
        },
        set: function (a) {
            var b = this.ah8;
            if (a != b) {
                this.ah7 = a;
                this.vl("IndicatorDisplayTypes", b, this.ah8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "alg", {
        get: function () {
            return this.ak9;
        },
        set: function (a) {
            var b = this.alg;
            if (a != b) {
                this.ak9 = a;
                this.vl("IndicatorPeriod", b, this.alg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "akk", {
        get: function () {
            return this.aj3;
        },
        set: function (a) {
            var b = this.akk;
            if (a != b) {
                this.aj3 = a;
                this.vl("IndicatorMultiplier", b, this.akk);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "alj", {
        get: function () {
            return this.alc;
        },
        set: function (a) {
            var b = this.alj;
            if (a != b) {
                this.alc = a;
                this.vl("IndicatorSmoothingPeriod", b, this.alj);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "alh", {
        get: function () {
            return this.ala;
        },
        set: function (a) {
            var b = this.alh;
            if (a != b) {
                this.ala = a;
                this.vl("IndicatorShortPeriod", b, this.alh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "alf", {
        get: function () {
            return this.ak8;
        },
        set: function (a) {
            var b = this.alf;
            if (a != b) {
                this.ak8 = a;
                this.vl("IndicatorLongPeriod", b, this.alf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ali", {
        get: function () {
            return this.alb;
        },
        set: function (a) {
            var b = this.ali;
            if (a != b) {
                this.alb = a;
                this.vl("IndicatorSignalPeriod", b, this.ali);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ahe", {
        get: function () {
            return this.ahc;
        },
        set: function (a) {
            var b = this.ahe;
            if (a != b) {
                this.ahc = a;
                this.vl("RangeSelectorOptions", b, this.ahe);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "customIndicatorNames", {
        get: function () {
            return this.ag8;
        },
        set: function (a) {
            var b = this.ag8;
            if (a != b) {
                this.ag8 = a;
                this.vl("CustomIndicatorNames", b, this.customIndicatorNames);
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.prototype.get_he = function () {
        return _super.prototype.get_he.call(this);
    };
    FinancialChart.prototype.set_he = function (a) {
        _super.prototype.set_he.call(this, a);
    };
    Object.defineProperty(FinancialChart.prototype, "apv", {
        get: function () {
            return this.apu;
        },
        set: function (a) {
            var b = this.apv;
            if (a != b) {
                this.apu = a;
                this.vl("ZoomSliderXAxisMajorStroke", b, this.apv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ak1", {
        get: function () {
            return this.aki;
        },
        set: function (a) {
            var b = this.ak1;
            if (a != b) {
                this.aki = a;
                this.vl("ZoomSliderXAxisMajorStrokeThickness", b, this.ak1);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ajc", {
        get: function () {
            return this.aiz;
        },
        set: function (a) {
            var b = this.ajc;
            if (a != b) {
                this.aiz = a;
                this.vl("IsLegendVisible", b, this.ajc);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ah0", {
        get: function () {
            return this.ahz;
        },
        set: function (a) {
            var b = this.ah0;
            if (a != b) {
                this.ahz = a;
                this.vl("FinancialChartLegend", b, this.ah0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "xAxis", {
        get: function () {
            return this.agz;
        },
        set: function (a) {
            var b = this.xAxis;
            if (a != b) {
                this.agz = a;
                this.vl("XAxis", b, this.xAxis);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "yAxis", {
        get: function () {
            return this.aie;
        },
        set: function (a) {
            var b = this.yAxis;
            if (a != b) {
                this.aie = a;
                this.vl("YAxis", b, this.yAxis);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "volumeAxis", {
        get: function () {
            return this.aid;
        },
        set: function (a) {
            var b = this.volumeAxis;
            if (a != b) {
                this.aid = a;
                this.vl("VolumeAxis", b, this.volumeAxis);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "volumeXAxis", {
        get: function () {
            return this.agy;
        },
        set: function (a) {
            var b = this.volumeXAxis;
            if (a != b) {
                this.agy = a;
                this.vl("VolumeXAxis", b, this.volumeXAxis);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "indicatorAxis", {
        get: function () {
            return this.aic;
        },
        set: function (a) {
            var b = this.indicatorAxis;
            if (a != b) {
                this.aic = a;
                this.vl("IndicatorAxis", b, this.indicatorAxis);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "indicatorXAxis", {
        get: function () {
            return this.agx;
        },
        set: function (a) {
            var b = this.indicatorXAxis;
            if (a != b) {
                this.agx = a;
                this.vl("IndicatorXAxis", b, this.indicatorXAxis);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "aju", {
        get: function () {
            return this.ajr;
        },
        set: function (a) {
            var b = this.aju;
            if (a != b) {
                this.ajr = a;
                this.vl("IndicatorAxes", b, this.aju);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "ajw", {
        get: function () {
            return this.ajt;
        },
        set: function (a) {
            var b = this.ajw;
            if (a != b) {
                this.ajt = a;
                this.vl("IndicatorXAxes", b, this.ajw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "aj2", {
        get: function () {
            return this.aj0;
        },
        set: function (a) {
            var b = this.aj2;
            if (+a != +b) {
                this.aj0 = a;
                this.vl("XAxisMinimumValue", b, this.aj2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "aj1", {
        get: function () {
            return this.ajz;
        },
        set: function (a) {
            var b = this.aj1;
            if (+a != +b) {
                this.ajz = a;
                this.vl("XAxisMaximumValue", b, this.aj1);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChart.prototype, "xAxisBreaks", {
        get: function () {
            return this.ajx;
        },
        enumerable: false,
        configurable: true
    });
    FinancialChart.$t = markType(FinancialChart, 'FinancialChart', XYChart.$);
    FinancialChart.alk = 0;
    FinancialChart.ang = "XAxisEnhancedIntervalPreferMoreCategoryLabels";
    FinancialChart.anr = "YAxisEnhancedIntervalPreferMoreCategoryLabels";
    return FinancialChart;
}(XYChart));
export { FinancialChart };
/**
 * @hidden
 */
var FinancialChartToolbar = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialChartToolbar, _super);
    function FinancialChartToolbar(a, b) {
        var _this = _super.call(this, "financialChartToolbar", b, "") || this;
        _this.o = null;
        _this.r = null;
        _this.z = null;
        _this.p = null;
        _this.w = null;
        _this.aq = dateMaxValue();
        _this.ap = dateMinValue();
        _this.ao = dateMaxValue();
        _this.an = dateMinValue();
        _this.rangeChanged = null;
        _this.chartTypeChanged = null;
        _this.trendLineTypeChanged = null;
        _this.volumeTypeChanged = null;
        _this.indicatorTypesChanged = null;
        _this.overlayTypesChanged = null;
        _this.x = 0;
        _this.ah = 0;
        _this.ab = 0;
        _this.ad = null;
        _this.af = null;
        _this.t = null;
        _this.ak = false;
        _this.o = a;
        _this.w = new FinancialChartToolbarViewModel();
        _this.s = new FinancialChartRangeSelector(a != null ? a.alw : null);
        _this.aa = new FinancialChartTypePicker(a != null ? a.alt : null);
        _this.q = new FinancialChartIndicatorMenu(1, a != null ? a.alu : null);
        return _this;
    }
    FinancialChartToolbar.prototype.bg = function (a) {
        var b = a.querySelectorAll(".financialChartToolbarArea");
        if (b != null && b.length > 0) {
            this.g = b[0];
            if (!this.al) {
                this.g.hide();
            }
        }
        var c = a.querySelectorAll(".financialChartIndicatorMenu");
        if (c != null && c.length > 0) {
            this.q.g = c[0];
        }
        var d = a.querySelectorAll(".financialChartRangeSelector");
        if (d != null && d.length > 0) {
            this.s.g = d[0];
        }
        var e = a.querySelectorAll(".financialChartTypePicker");
        if (e != null && e.length > 0) {
            this.aa.g = e[0];
        }
    };
    FinancialChartToolbar.prototype.bl = function (a) {
        this.q.u(a);
        this.aa.u(a);
    };
    Object.defineProperty(FinancialChartToolbar.prototype, "s", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            var b = this.s;
            if (a != b) {
                this.r = a;
                this.bm("RangeSelector", b, this.s);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartToolbar.prototype, "aa", {
        get: function () {
            return this.z;
        },
        set: function (a) {
            var b = this.aa;
            if (a != b) {
                this.z = a;
                this.bm("ChartTypePicker", b, this.aa);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartToolbar.prototype, "q", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            var b = this.q;
            if (a != b) {
                this.p = a;
                this.bm("IndicatorMenu", b, this.q);
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChartToolbar.prototype.bm = function (a, b, c) {
        switch (a) {
            case "RangeStart":
                if (this.s != null) {
                    this.s.t = this.as;
                }
                break;
            case "RangeEnd":
                if (this.s != null) {
                    this.s.s = this.ar;
                }
                break;
            case "SelectionStart":
                if (this.s != null) {
                    this.s.v = this.au;
                }
                break;
            case "SelectionEnd":
                if (this.s != null) {
                    this.s.u = this.at;
                }
                break;
            case "RangeSelector":
                var d = b;
                if (d != null) {
                    d.rangeChanged = delegateRemove(d.rangeChanged, runOn(this, this.bj));
                }
                if (this.s != null) {
                    this.s.t = this.as;
                    this.s.s = this.ar;
                    this.s.o = this.u;
                    this.s.v = this.au;
                    this.s.u = this.at;
                    var e = this.s;
                    e.rangeChanged = delegateCombine(e.rangeChanged, runOn(this, this.bj));
                }
                break;
            case "ChartTypePicker":
                var f = b;
                if (f != null) {
                    f.selectionChanged = delegateRemove(f.selectionChanged, runOn(this, this.bi));
                }
                if (this.aa != null) {
                    this.aa.w = this.y;
                    var g = this.aa;
                    g.selectionChanged = delegateCombine(g.selectionChanged, runOn(this, this.bi));
                }
                break;
            case "ChartType":
                if (this.aa != null) {
                    this.aa.w = this.y;
                }
                break;
            case "RangeSelectorOptions":
                if (this.s != null) {
                    this.s.o = this.u;
                }
                break;
            case "IsToolbarVisible":
                if (this.g == null) {
                    break;
                }
                if (this.al) {
                    this.g.show();
                }
                else {
                    this.g.hide();
                }
                break;
            case "IndicatorMenu":
                var h = typeCast(FinancialChartIndicatorMenu.$, b);
                if (h != null) {
                    h.selectionChanged = delegateRemove(h.selectionChanged, runOn(this, this.bh));
                }
                if (this.q != null) {
                    var i = this.q;
                    i.selectionChanged = delegateCombine(i.selectionChanged, runOn(this, this.bh));
                }
                break;
            case "IndicatorTypes":
                if (this.indicatorTypesChanged != null) {
                    this.indicatorTypesChanged(this, EventArgs.empty);
                }
                break;
            case "OverlayTypes":
                if (this.overlayTypesChanged != null) {
                    this.overlayTypesChanged(this, EventArgs.empty);
                }
                break;
            case "TrendLineType":
                if (this.trendLineTypeChanged != null) {
                    this.trendLineTypeChanged(this, EventArgs.empty);
                }
                break;
            case "VolumeType":
                if (this.volumeTypeChanged != null) {
                    this.volumeTypeChanged(this, EventArgs.empty);
                }
                break;
        }
    };
    FinancialChartToolbar.prototype.bh = function (a, b) {
        this.ag = this.q.z;
        this.ae = this.q.y;
        this.ai = this.q.aa;
        this.ac = this.q.x;
    };
    FinancialChartToolbar.prototype.bi = function (a, b) {
        this.y = this.aa.w;
        if (this.chartTypeChanged != null) {
            this.chartTypeChanged(this, b);
        }
    };
    FinancialChartToolbar.prototype.bj = function (a, b) {
        this.au = this.s.v;
        this.at = this.s.u;
        if (this.rangeChanged != null) {
            this.rangeChanged(this, b);
        }
    };
    FinancialChartToolbar.prototype.v = function () {
        var _this = this;
        var a = new FinancialChartToolbarContext();
        a.container = this.d;
        a.vm = this.w;
        a.vm.financialChartIndicatorMenu = ((function () {
            var $ret = new FinancialChartIndicatorMenuContext();
            $ret.vm = _this.q.p;
            return $ret;
        })());
        a.vm.financialChartTypePicker = ((function () {
            var $ret = new FinancialChartTypePickerContext();
            $ret.vm = _this.aa.p;
            return $ret;
        })());
        a.vm.financialChartRangeSelector = ((function () {
            var $ret = new FinancialChartRangeSelectorContext();
            $ret.vm = _this.s.q;
            return $ret;
        })());
        return a;
    };
    FinancialChartToolbar.prototype.l = function (a) {
        _super.prototype.l.call(this, a);
        if (a == null) {
            return;
        }
        this.s.l(a);
        this.aa.l(a);
        this.q.l(a);
        this.bk();
    };
    FinancialChartToolbar.prototype.bk = function () {
    };
    Object.defineProperty(FinancialChartToolbar.prototype, "au", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            var b = this.au;
            if (+a != +b) {
                this.aq = a;
                this.bm("SelectionStart", b, this.au);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartToolbar.prototype, "at", {
        get: function () {
            return this.ap;
        },
        set: function (a) {
            var b = this.at;
            if (+a != +b) {
                this.ap = a;
                this.bm("SelectionEnd", b, this.at);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartToolbar.prototype, "as", {
        get: function () {
            return this.ao;
        },
        set: function (a) {
            var b = this.as;
            if (+a != +b) {
                this.ao = a;
                this.bm("RangeStart", b, this.as);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartToolbar.prototype, "ar", {
        get: function () {
            return this.an;
        },
        set: function (a) {
            var b = this.ar;
            if (+a != +b) {
                this.an = a;
                this.bm("RangeEnd", b, this.ar);
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChartToolbar.prototype.bn = function () {
        if (this.s != null) {
            this.s.ab();
        }
    };
    Object.defineProperty(FinancialChartToolbar.prototype, "y", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            var b = this.y;
            if (a != b) {
                this.x = a;
                this.bm("ChartType", enumGetBox(FinancialChartType_$type, b), enumGetBox(FinancialChartType_$type, this.y));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartToolbar.prototype, "ai", {
        get: function () {
            return this.ah;
        },
        set: function (a) {
            var b = this.ai;
            if (a != b) {
                this.ah = a;
                this.bm("TrendLineType", enumGetBox(TrendLineType_$type, b), enumGetBox(TrendLineType_$type, this.ai));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartToolbar.prototype, "ac", {
        get: function () {
            return this.ab;
        },
        set: function (a) {
            var b = this.ac;
            if (a != b) {
                this.ab = a;
                this.bm("VolumeType", enumGetBox(FinancialChartVolumeType_$type, b), enumGetBox(FinancialChartVolumeType_$type, this.ac));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartToolbar.prototype, "ae", {
        get: function () {
            return this.ad;
        },
        set: function (a) {
            var b = this.ae;
            if (!FinancialChartToolbar.am(b, a)) {
                this.ad = a;
                this.bm("IndicatorTypes", b, this.ae);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartToolbar.prototype, "ag", {
        get: function () {
            return this.af;
        },
        set: function (a) {
            var b = this.ag;
            if (!FinancialChartToolbar.am(b, a)) {
                this.af = a;
                this.bm("OverlayTypes", b, this.ag);
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChartToolbar.am = function (a, b) {
        var e_21, _a;
        if (a == b) {
            return true;
        }
        if ((a == null && b != null) || (a != null && b == null)) {
            return false;
        }
        if (a.count != b.count) {
            return false;
        }
        try {
            for (var _b = __values(fromEn(a)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                if (!b.contains(c)) {
                    return false;
                }
            }
        }
        catch (e_21_1) {
            e_21 = { error: e_21_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_21)
                    throw e_21.error;
            }
        }
        return true;
    };
    FinancialChartToolbar.prototype.aj = function () {
        var _this = this;
        return ((function () {
            var $ret = new FinancialChartToolbarVisualData();
            $ret.rangeSelector = _this.s == null ? null : _this.s.r();
            return $ret;
        })());
    };
    Object.defineProperty(FinancialChartToolbar.prototype, "u", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            var b = this.u;
            if (a != b) {
                this.t = a;
                this.bm("RangeSelectorOptions", b, this.u);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartToolbar.prototype, "al", {
        get: function () {
            return this.ak;
        },
        set: function (a) {
            var b = this.al;
            if (a != b) {
                this.ak = a;
                this.bm("IsToolbarVisible", b, this.al);
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChartToolbar.$t = markType(FinancialChartToolbar, 'FinancialChartToolbar', TemplatedHtmlControl.$);
    return FinancialChartToolbar;
}(TemplatedHtmlControl));
export { FinancialChartToolbar };
/**
 * @hidden
 */
var FinancialChartRangeSelector = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialChartRangeSelector, _super);
    function FinancialChartRangeSelector(a) {
        var _this = _super.call(this, "financialChartRangeSelector", a, "") || this;
        _this.q = null;
        _this.rangeChanged = null;
        _this.p = null;
        _this.p = new FinancialChartRangeSelectorView();
        _this.q = new FinancialChartRangeSelectorViewModel();
        _this.q.onRangeChanged = runOn(_this, _this.z);
        _this.p.c = _this.q;
        return _this;
    }
    Object.defineProperty(FinancialChartRangeSelector.prototype, "v", {
        get: function () {
            return this.q.selectionStart;
        },
        set: function (a) {
            this.q.selectionStart = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartRangeSelector.prototype, "u", {
        get: function () {
            return this.q.selectionEnd;
        },
        set: function (a) {
            this.q.selectionEnd = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartRangeSelector.prototype, "t", {
        get: function () {
            return this.q.rangeStart;
        },
        set: function (a) {
            this.q.rangeStart = a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FinancialChartRangeSelector.prototype, "s", {
        get: function () {
            return this.q.rangeEnd;
        },
        set: function (a) {
            this.q.rangeEnd = a;
        },
        enumerable: false,
        configurable: true
    });
    FinancialChartRangeSelector.prototype.ab = function () {
        if (this.q.onPropertiesChanged()) {
            this.p.g(this.q);
        }
    };
    FinancialChartRangeSelector.prototype.l = function (a) {
        _super.prototype.l.call(this, a);
        this.p.f(this.d, this.q);
    };
    FinancialChartRangeSelector.prototype.z = function () {
        if (this.rangeChanged != null) {
            this.rangeChanged(this, EventArgs.empty);
        }
    };
    FinancialChartRangeSelector.prototype.r = function () {
        return this.p.d(this.g);
    };
    Object.defineProperty(FinancialChartRangeSelector.prototype, "o", {
        get: function () {
            return this.q.options;
        },
        set: function (a) {
            var b = !FinancialChartToolbar.am(this.q.options, a);
            if (b) {
                this.q.options = a;
                this.aa();
            }
        },
        enumerable: false,
        configurable: true
    });
    FinancialChartRangeSelector.prototype.aa = function () {
        var a = false;
        if (this.q != null) {
            a = this.q.p();
        }
        if (this.p != null) {
            this.p.i(this.o);
            if (a) {
                if (this.q.notifyChanged != null) {
                    this.q.notifyChanged();
                }
            }
        }
    };
    FinancialChartRangeSelector.$t = markType(FinancialChartRangeSelector, 'FinancialChartRangeSelector', TemplatedHtmlControl.$);
    FinancialChartRangeSelector.x = 0;
    return FinancialChartRangeSelector;
}(TemplatedHtmlControl));
export { FinancialChartRangeSelector };
