// igniteui-angular-charts/esm5/lib/FinancialOverlay.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { FinancialSeries } from "./FinancialSeries";
import { DependencyProperty } from "igniteui-angular-core";
import { ISortingAxis_$type } from "./ISortingAxis";
import { typeCast, markType, markDep } from "igniteui-angular-core";
import { Rect } from "igniteui-angular-core";
import { ScalerParams } from "./ScalerParams";
import { AxisRange } from "./AxisRange";
import { PropertyMetadata } from "igniteui-angular-core";
import { isNaN_ } from "igniteui-angular-core";
/**
 * @hidden
 */
var FinancialOverlay = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FinancialOverlay, _super);
    function FinancialOverlay() {
        var _this = _super.call(this) || this;
        _this.aas = false;
        _this.aas = false;
        return _this;
    }
    FinancialOverlay.prototype.get_e6 = function () {
        return true;
    };
    FinancialOverlay.prototype.o9 = function (a, b, c, d) {
        _super.prototype.o9.call(this, a, b, c, d);
        if (this.wy != null && typeCast(ISortingAxis_$type, this.wy) !== null) {
            this.wy.i$i();
        }
        this.xf.b0.e(this.ib);
        this.aat();
        if (this.xh != null) {
            this.xh.c1();
        }
        this.rn(true);
    };
    FinancialOverlay.prototype.get_hb = function () {
        return true;
    };
    FinancialOverlay.prototype.q5 = function (a, b, c, d) {
        switch (b) {
            case "FastItemsSource":
                this.aas = false;
                break;
            case FinancialOverlay.$$p[0]:
                this.aas = false;
                this.rn(false);
                this.qh();
                break;
            case "XAxis":
                this.aas = false;
                break;
        }
        _super.prototype.q5.call(this, a, b, c, d);
    };
    FinancialOverlay.prototype.b0 = function () {
        if (this.b1 == 0) {
            return 1;
        }
        else {
            return _super.prototype.b0.call(this);
        }
    };
    FinancialOverlay.prototype.gu = function (a, b) {
        if (this.gt(a, b)) {
            return true;
        }
        return false;
    };
    FinancialOverlay.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rn(false);
    };
    FinancialOverlay.prototype.gb = function (a) {
        var b = this.dl != null ? this.dl.indexOf(a) : -1;
        var c = this.ct.by;
        c = c.copy();
        var d = this.ct.bx;
        var e = new Rect(0, 0, 0, 1, 1);
        var f = this.wf(this.ct);
        var g = new ScalerParams(0, e, e, this.wy.ch, f);
        var h = new ScalerParams(0, e, e, this.xh.ch, f);
        h.b = this.iw();
        if (b >= 0 && Rect.l_op_Inequality(c, null) && Rect.l_op_Inequality(d, null)) {
            if (this.wy != null) {
                var i = this.wy.ek(b, g);
                if (i < c.left + 0.1 * c.width) {
                    i = i + 0.4 * c.width;
                }
                if (i > c.right - 0.1 * c.width) {
                    i = i - 0.4 * c.width;
                }
                c.x = i - 0.5 * c.width;
            }
            if (this.xh != null && this.xt != null && this.xs != null && b < this.xt.count && b < this.xs.count) {
                var j = this.xh.ek(this.xt.item(b), h);
                var k = this.xh.ek(this.xs.item(b), h);
                if (!isNaN_(j) && !isNaN_(k)) {
                    var l = Math.abs(j - k);
                    if (c.height < l) {
                        c.height = l;
                        c.y = Math.min(j, k);
                    }
                    else {
                        if (j < c.top + 0.1 * c.height) {
                            j = j + 0.4 * c.height;
                        }
                        if (j > c.bottom - 0.1 * c.height) {
                            j = j - 0.4 * c.height;
                        }
                        c.y = j - 0.5 * c.height;
                    }
                }
            }
            this.c1.au(this.c9, c, true);
        }
        return b >= 0;
    };
    FinancialOverlay.prototype.ax = function (a) {
        if (a != null && a == this.xh && this.xt != null && this.xs != null) {
            return new AxisRange(this.xt.minimum, this.xs.maximum);
        }
        return null;
    };
    FinancialOverlay.$t = markType(FinancialOverlay, 'FinancialOverlay', FinancialSeries.$);
    FinancialOverlay.$$p = markDep(DependencyProperty, PropertyMetadata, FinancialOverlay, 'raisePropertyChanged', ['IgnoreFirst:aau:aaw', [1, 0]]);
    return FinancialOverlay;
}(FinancialSeries));
export { FinancialOverlay };
