// igniteui-angular-charts/esm5/lib/ForceIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ForceIndexIndicatorStrategy } from "./ForceIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var ForceIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ForceIndexIndicator, _super);
    function ForceIndexIndicator() {
        return _super.call(this) || this;
    }
    ForceIndexIndicator.prototype.bi = function () {
        return new ForceIndexIndicator();
    };
    ForceIndexIndicator.prototype.get_aau = function () {
        return 1;
    };
    ForceIndexIndicator.prototype.get_abn = function () {
        return new ForceIndexIndicatorStrategy();
    };
    Object.defineProperty(ForceIndexIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    ForceIndexIndicator.prototype.get_ab0 = function () {
        return ForceIndexIndicator.$;
    };
    Object.defineProperty(ForceIndexIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ForceIndexIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(ForceIndexIndicator.ab9));
        },
        set: function (a) {
            this.h(ForceIndexIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    ForceIndexIndicator.prototype.abt = function () {
        return this.ab8;
    };
    ForceIndexIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    ForceIndexIndicator.$t = markType(ForceIndexIndicator, 'ForceIndexIndicator', StrategyBasedIndicator.$);
    ForceIndexIndicator.ab9 = StrategyBasedIndicator.ab5(0, ForceIndexIndicator.$);
    return ForceIndexIndicator;
}(StrategyBasedIndicator));
export { ForceIndexIndicator };
