// igniteui-angular-charts/esm5/lib/FullStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { FullStochasticOscillatorIndicatorStrategy } from "./FullStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
import { Brush } from "igniteui-angular-core";
/**
 * @hidden
 */
var FullStochasticOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(FullStochasticOscillatorIndicator, _super);
    function FullStochasticOscillatorIndicator() {
        var _this = _super.call(this) || this;
        _this.trendLineBrush = ((function () {
            var $ret = new Brush();
            $ret.fill = "#565656";
            return $ret;
        })());
        _this.trendLineType = 10;
        _this.trendLinePeriod = 3;
        return _this;
    }
    FullStochasticOscillatorIndicator.prototype.bi = function () {
        return new FullStochasticOscillatorIndicator();
    };
    FullStochasticOscillatorIndicator.prototype.get_abn = function () {
        return new FullStochasticOscillatorIndicatorStrategy();
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.get_ab0 = function () {
        return FullStochasticOscillatorIndicator.$;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.acd));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.acd, a);
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.abt = function () {
        return this.ab8;
    };
    FullStochasticOscillatorIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "smoothingPeriod", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.ace));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.ace, a);
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.abv = function () {
        return this.smoothingPeriod;
    };
    Object.defineProperty(FullStochasticOscillatorIndicator.prototype, "aca", {
        get: function () {
            return typeGetValue(this.c(FullStochasticOscillatorIndicator.acf));
        },
        set: function (a) {
            this.h(FullStochasticOscillatorIndicator.acf, a);
        },
        enumerable: false,
        configurable: true
    });
    FullStochasticOscillatorIndicator.prototype.abs = function () {
        return this.aca;
    };
    FullStochasticOscillatorIndicator.prototype.aa3 = function () {
        return this.aca;
    };
    FullStochasticOscillatorIndicator.$t = markType(FullStochasticOscillatorIndicator, 'FullStochasticOscillatorIndicator', StrategyBasedIndicator.$);
    FullStochasticOscillatorIndicator.acd = StrategyBasedIndicator.ab5(14, FullStochasticOscillatorIndicator.$);
    FullStochasticOscillatorIndicator.ace = StrategyBasedIndicator.ab6(3, FullStochasticOscillatorIndicator.$, "SmoothingPeriod");
    FullStochasticOscillatorIndicator.acf = StrategyBasedIndicator.ab6(3, FullStochasticOscillatorIndicator.$, "TriggerPeriod");
    return FullStochasticOscillatorIndicator;
}(StrategyBasedIndicator));
export { FullStochasticOscillatorIndicator };
