// igniteui-angular-charts/esm5/lib/HorizontalRangeCategorySeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { RangeCategorySeries } from "./RangeCategorySeries";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { DependencyProperty } from "igniteui-angular-core";
import { NumericYAxis } from "./NumericYAxis";
import { Axis } from "./Axis";
import { Point_$type, runOn, typeCast, markType, markDep } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { RangeValueList } from "./RangeValueList";
import { ScalerParams } from "./ScalerParams";
import { CategoryXAxis } from "./CategoryXAxis";
import { DataLegendSeriesValueInfo } from "igniteui-angular-core";
import { DataLegendSeriesContext } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
/**
 * @hidden
 */
var HorizontalRangeCategorySeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(HorizontalRangeCategorySeries, _super);
    function HorizontalRangeCategorySeries() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HorizontalRangeCategorySeries.prototype.fetchXAxis = function () {
        return this.xAxis;
    };
    HorizontalRangeCategorySeries.prototype.fetchYAxis = function () {
        return this.yAxis;
    };
    HorizontalRangeCategorySeries.prototype.o0 = function () {
        _super.prototype.o0.call(this);
        this.xAxis = null;
        this.yAxis = null;
    };
    HorizontalRangeCategorySeries.prototype.aj = function (a) {
        var b = this.am(this.yk(), this.yl.f, this.zm(this.ct), this.zn(this.ct), this.wb(a), true);
        if (b == null) {
            return null;
        }
        var c = b[0];
        var d = b[1];
        var e = new Array(1);
        var f = new Array(2);
        f[0] = { $type: Point_$type, x: c[0], y: c[2] };
        f[1] = { $type: Point_$type, x: d[0], y: d[2] };
        e[0] = f;
        return e;
    };
    HorizontalRangeCategorySeries.prototype.ak = function (a) {
        var b = this.am(this.yk(), this.yl.f, this.zm(this.ct), this.zn(this.ct), this.wb(a), true);
        if (b == null) {
            return null;
        }
        var c = b[0];
        var d = b[1];
        var e = new Array(1);
        var f = new Array(2);
        f[0] = { $type: Point_$type, x: c[0], y: c[1] };
        f[1] = { $type: Point_$type, x: d[0], y: d[1] };
        e[0] = f;
        return e;
    };
    HorizontalRangeCategorySeries.prototype.get_fs = function () {
        return true;
    };
    HorizontalRangeCategorySeries.prototype.b0 = function () {
        if (this.b1 == 0) {
            return 1;
        }
        else {
            return _super.prototype.b0.call(this);
        }
    };
    HorizontalRangeCategorySeries.prototype.getOffsetValue = function () {
        return this.RangeFramePreparer.r(this.xAxis, this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct));
    };
    HorizontalRangeCategorySeries.prototype.getCategoryWidth = function () {
        return this.xAxis.getCategorySize(this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct));
    };
    HorizontalRangeCategorySeries.prototype.j6 = function (a, b) {
        return this.j7(a, b, this.xAxis, runOn(this, this.zk), new RangeValueList(this.highColumn, this.lowColumn));
    };
    HorizontalRangeCategorySeries.prototype.j8 = function (a, b) {
        return this.j9(a, b, this.xAxis, runOn(this, this.zk), new RangeValueList(this.highColumn, this.lowColumn));
    };
    HorizontalRangeCategorySeries.prototype.im = function (a, b, c, d, e) {
        if (c == null) {
            return Number.POSITIVE_INFINITY;
        }
        var f = this.xAxis.lg;
        return this.io(a, b, this.xAxis, d, e, f, runOn(this, this.zk));
    };
    HorizontalRangeCategorySeries.prototype.i3 = function (a, b, c) {
        if (this.c9 == null) {
            return NaN;
        }
        var d = this.getEffectiveViewport1(this.ct);
        var e = new ScalerParams(0, this.c9.vl, this.ct.bx, this.xAxis.ch, d);
        var f = this.RangeFramePreparer.r(this.xAxis, this.c9.vl, this.ct.bx, d);
        return this.i5(new RangeValueList(this.highColumn, this.lowColumn), a, this.xAxis, e, f, runOn(this, this.zk), b, c);
    };
    HorizontalRangeCategorySeries.prototype.i1 = function (a, b, c) {
        if (this.c9 == null) {
            return NaN;
        }
        var d = this.getEffectiveViewport1(this.ct);
        var e = new ScalerParams(0, this.c9.vl, this.ct.bx, this.xAxis.ch, d);
        var f = this.RangeFramePreparer.r(this.xAxis, this.c9.vl, this.ct.bx, d);
        return this.i5(this.lowColumn, a, this.xAxis, e, f, runOn(this, this.zk), b, c);
    };
    HorizontalRangeCategorySeries.prototype.iz = function (a, b, c) {
        if (this.c9 == null) {
            return NaN;
        }
        var d = this.getEffectiveViewport1(this.ct);
        var e = new ScalerParams(0, this.c9.vl, this.ct.bx, this.xAxis.ch, d);
        var f = this.RangeFramePreparer.r(this.xAxis, this.c9.vl, this.ct.bx, d);
        return this.i5(this.highColumn, a, this.xAxis, e, f, runOn(this, this.zk), b, c);
    };
    HorizontalRangeCategorySeries.prototype.v2 = function (a, b, c) {
        var _this = this;
        return this.v8(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct)), this.yAxis, this.xAxis, runOn(this, this.iz), function (d, e) { return _this.j9(d, e, _this.xAxis, runOn(_this, _this.zk), _this.highColumn); }, function (d, e) { return _this.j7(d, e, _this.xAxis, runOn(_this, _this.zk), _this.highColumn); });
    };
    HorizontalRangeCategorySeries.prototype.v4 = function (a, b, c) {
        var _this = this;
        return this.v8(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct)), this.yAxis, this.xAxis, runOn(this, this.i1), function (d, e) { return _this.j9(d, e, _this.xAxis, runOn(_this, _this.zk), _this.lowColumn); }, function (d, e) { return _this.j7(d, e, _this.xAxis, runOn(_this, _this.zk), _this.lowColumn); });
    };
    HorizontalRangeCategorySeries.prototype.v6 = function (a, b, c) {
        return this.v8(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct)), this.yAxis, this.xAxis, null, null, null);
    };
    HorizontalRangeCategorySeries.prototype.zg = function () {
        return this.yAxis != null && this.yAxis.c1();
    };
    HorizontalRangeCategorySeries.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
        switch (b) {
            case HorizontalRangeCategorySeries.$$p[0]:
            case HorizontalRangeCategorySeries.$$p[1]:
            case HorizontalRangeCategorySeries.$$p[2]:
            case HorizontalRangeCategorySeries.$$p[3]:
                this.qb();
                break;
            case HorizontalRangeCategorySeries.$$p[4]:
                this.pa(typeCast(Axis.$, c));
                this.rd(typeCast(Axis.$, d));
                this.yu.cw.g(this.ib);
                this.rn(false);
                this.qh();
                break;
            case HorizontalRangeCategorySeries.$$p[5]:
                this.pa(typeCast(Axis.$, c));
                this.rd(typeCast(Axis.$, d));
                this.yu.cw.g(this.ib);
                this.zg();
                this.rn(false);
                this.qh();
                break;
        }
    };
    HorizontalRangeCategorySeries.prototype.aa2 = function (a) {
        if (typeCast(NumericYAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    HorizontalRangeCategorySeries.prototype.aa1 = function (a) {
        if (typeCast(CategoryXAxis.$, a) !== null || a.ce) {
            return true;
        }
        return false;
    };
    HorizontalRangeCategorySeries.prototype.getDataLegendSeriesContextAt = function (a, b) {
        var c = { $type: Point_$type, x: a, y: b };
        var d = this.i1(c, false, false);
        var e = this.iz(c, false, false);
        var f = new DataLegendSeriesValueInfo();
        f.a = 1;
        f.l = e > d ? d : e;
        f.q = "Low";
        f.r = this.aaq != null ? this.aaq : "Low";
        f.s = this.aa7 != null ? this.aa7 : "L:";
        f.t = this.aa9 != null ? this.aa9 : "";
        var g = new DataLegendSeriesValueInfo();
        g.a = 2;
        g.l = e > d ? e : d;
        g.q = "High";
        g.r = this.aan != null ? this.aan : "High";
        g.s = this.aa3 != null ? this.aa3 : "H:";
        g.t = this.aa5 != null ? this.aa5 : "";
        var h = new DataLegendSeriesValueInfo();
        h.a = 7;
        h.l = Math.abs(g.l - f.l);
        h.q = "Range";
        h.r = "Range";
        h.s = "R:";
        h.k = true;
        var i = new DataLegendSeriesValueInfo();
        i.a = 6;
        i.l = f.l + (h.l / 2);
        i.q = "Average";
        i.r = "Average";
        i.s = "A:";
        i.k = true;
        var j = new DataLegendSeriesContext();
        j.e = ((function () {
            var $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
            $ret.add(g);
            $ret.add(f);
            $ret.add(h);
            $ret.add(i);
            return $ret;
        })());
        j.h = this.ma();
        return j;
    };
    Object.defineProperty(HorizontalRangeCategorySeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zh;
        },
        enumerable: false,
        configurable: true
    });
    HorizontalRangeCategorySeries.$t = markType(HorizontalRangeCategorySeries, 'HorizontalRangeCategorySeries', RangeCategorySeries.$);
    HorizontalRangeCategorySeries.$$p = markDep(DependencyProperty, PropertyMetadata, HorizontalRangeCategorySeries, 'raisePropertyChanged', ['HighMemberAsLegendLabel:aa3:abd', [2, null], 'HighMemberAsLegendUnit:aa5:abe', [2, null], 'LowMemberAsLegendLabel:aa7:abf', [2, null], 'LowMemberAsLegendUnit:aa9:abg', [2, null], 'XAxis::abh', [CategoryAxisBase.$, null], 'YAxis::abi', [NumericYAxis.$, null]]);
    return HorizontalRangeCategorySeries;
}(RangeCategorySeries));
export { HorizontalRangeCategorySeries };
