// igniteui-angular-charts/esm5/lib/igx-market-facilitation-index-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxMarketFacilitationIndexIndicatorComponent } from './igx-market-facilitation-index-indicator-component';
import * as i0 from "@angular/core";
var IgxMarketFacilitationIndexIndicatorModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxMarketFacilitationIndexIndicatorModule() {
    }
    IgxMarketFacilitationIndexIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMarketFacilitationIndexIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxMarketFacilitationIndexIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMarketFacilitationIndexIndicatorModule, declarations: [IgxMarketFacilitationIndexIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxMarketFacilitationIndexIndicatorComponent, IgxIndicatorProxyModule] });
    IgxMarketFacilitationIndexIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMarketFacilitationIndexIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxMarketFacilitationIndexIndicatorModule;
}());
export { IgxMarketFacilitationIndexIndicatorModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMarketFacilitationIndexIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxMarketFacilitationIndexIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxMarketFacilitationIndexIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
