// igniteui-angular-charts/esm5/lib/igx-data-chart-financial-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryXAxisDynamicModule } from './igx-category-x-axis-dynamic-module';
import { IgxNumericYAxisDynamicModule } from './igx-numeric-y-axis-dynamic-module';
import { IgxDataChartCoreDynamicModule } from './igx-data-chart-core-dynamic-module';
import { IgxDataChartFinancialCoreModule } from './igx-data-chart-financial-core-module';
import * as i0 from "@angular/core";
var IgxDataChartFinancialCoreDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartFinancialCoreDynamicModule() {
    }
    IgxDataChartFinancialCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartFinancialCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialCoreDynamicModule, imports: [CommonModule, IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartFinancialCoreModule], exports: [IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartFinancialCoreModule] });
    IgxDataChartFinancialCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialCoreDynamicModule, imports: [[CommonModule, IgxCategoryXAxisDynamicModule,
                IgxNumericYAxisDynamicModule,
                IgxDataChartCoreDynamicModule,
                IgxDataChartFinancialCoreModule
            ], IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartFinancialCoreModule] });
    return IgxDataChartFinancialCoreDynamicModule;
}());
export { IgxDataChartFinancialCoreDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxCategoryXAxisDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartFinancialCoreModule
                    ],
                    imports: [CommonModule, IgxCategoryXAxisDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartFinancialCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
