// igniteui-angular-charts/esm5/lib/igx-stacked-spline-area-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedSplineAreaSeriesModule } from './igx-stacked-spline-area-series-module';
import { IgxStackedSplineAreaSeriesComponent } from './igx-stacked-spline-area-series-component';
import { StackedSplineAreaSeries } from './StackedSplineAreaSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxStackedSplineAreaSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStackedSplineAreaSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxStackedSplineAreaSeriesComponent", IgxStackedSplineAreaSeriesComponent);
        TypeRegistrar.register("StackedSplineAreaSeries", StackedSplineAreaSeries.$type);
    }
    IgxStackedSplineAreaSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineAreaSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedSplineAreaSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineAreaSeriesDynamicModule, imports: [CommonModule, IgxStackedSplineAreaSeriesModule], exports: [IgxStackedSplineAreaSeriesModule] });
    IgxStackedSplineAreaSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineAreaSeriesDynamicModule, imports: [[CommonModule, IgxStackedSplineAreaSeriesModule], IgxStackedSplineAreaSeriesModule] });
    return IgxStackedSplineAreaSeriesDynamicModule;
}());
export { IgxStackedSplineAreaSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineAreaSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStackedSplineAreaSeriesModule],
                    exports: [IgxStackedSplineAreaSeriesModule],
                    entryComponents: [IgxStackedSplineAreaSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
