// igniteui-angular-charts/esm5/lib/igx-funnel-chart-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFunnelChartCoreDynamicModule } from './igx-funnel-chart-core-dynamic-module';
import { IgxDataChartInteractivityDynamicModule } from './igx-data-chart-interactivity-dynamic-module';
import { IgxFunnelChartModule } from './igx-funnel-chart-module';
import * as i0 from "@angular/core";
var IgxFunnelChartDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxFunnelChartDynamicModule() {
    }
    IgxFunnelChartDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFunnelChartDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartDynamicModule, imports: [CommonModule, IgxFunnelChartCoreDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxFunnelChartModule], exports: [IgxFunnelChartCoreDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxFunnelChartModule] });
    IgxFunnelChartDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartDynamicModule, imports: [[CommonModule, IgxFunnelChartCoreDynamicModule,
                IgxDataChartInteractivityDynamicModule,
                IgxFunnelChartModule
            ], IgxFunnelChartCoreDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxFunnelChartModule] });
    return IgxFunnelChartDynamicModule;
}());
export { IgxFunnelChartDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxFunnelChartCoreDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxFunnelChartModule
                    ],
                    imports: [CommonModule, IgxFunnelChartCoreDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxFunnelChartModule
                    ],
                    entryComponents: []
                }]
        }] });
