// igniteui-angular-charts/esm5/lib/igx-stoch-rsi-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStochRSIIndicatorModule } from './igx-stoch-rsi-indicator-module';
import { IgxStochRSIIndicatorComponent } from './igx-stoch-rsi-indicator-component';
import { StochRSIIndicator } from './StochRSIIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxStochRSIIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStochRSIIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxStochRSIIndicatorComponent", IgxStochRSIIndicatorComponent);
        TypeRegistrar.register("StochRSIIndicator", StochRSIIndicator.$type);
    }
    IgxStochRSIIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStochRSIIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStochRSIIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStochRSIIndicatorDynamicModule, imports: [CommonModule, IgxStochRSIIndicatorModule], exports: [IgxStochRSIIndicatorModule] });
    IgxStochRSIIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStochRSIIndicatorDynamicModule, imports: [[CommonModule, IgxStochRSIIndicatorModule], IgxStochRSIIndicatorModule] });
    return IgxStochRSIIndicatorDynamicModule;
}());
export { IgxStochRSIIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStochRSIIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStochRSIIndicatorModule],
                    exports: [IgxStochRSIIndicatorModule],
                    entryComponents: [IgxStochRSIIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
