// igniteui-angular-charts/esm5/lib/igx-force-index-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxForceIndexIndicatorComponent } from './igx-force-index-indicator-component';
import * as i0 from "@angular/core";
var IgxForceIndexIndicatorModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxForceIndexIndicatorModule() {
    }
    IgxForceIndexIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxForceIndexIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxForceIndexIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxForceIndexIndicatorModule, declarations: [IgxForceIndexIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxForceIndexIndicatorComponent, IgxIndicatorProxyModule] });
    IgxForceIndexIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxForceIndexIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxForceIndexIndicatorModule;
}());
export { IgxForceIndexIndicatorModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxForceIndexIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxForceIndexIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxForceIndexIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
