// igniteui-angular-charts/esm5/lib/igx-crosshair-layer-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCrosshairLayerModule } from './igx-crosshair-layer-module';
import { IgxCrosshairLayerComponent } from './igx-crosshair-layer-component';
import { CrosshairLayer } from './CrosshairLayer';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxCrosshairLayerDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCrosshairLayerDynamicModule() {
        TypeRegistrar.registerCons("IgxCrosshairLayerComponent", IgxCrosshairLayerComponent);
        TypeRegistrar.register("CrosshairLayer", CrosshairLayer.$type);
    }
    IgxCrosshairLayerDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCrosshairLayerDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCrosshairLayerDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCrosshairLayerDynamicModule, imports: [CommonModule, IgxCrosshairLayerModule], exports: [IgxCrosshairLayerModule] });
    IgxCrosshairLayerDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCrosshairLayerDynamicModule, imports: [[CommonModule, IgxCrosshairLayerModule], IgxCrosshairLayerModule] });
    return IgxCrosshairLayerDynamicModule;
}());
export { IgxCrosshairLayerDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCrosshairLayerDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCrosshairLayerModule],
                    exports: [IgxCrosshairLayerModule],
                    entryComponents: [IgxCrosshairLayerComponent]
                }]
        }], ctorParameters: function () { return []; } });
