// igniteui-angular-charts/esm5/lib/igx-time-axis-interval-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTimeAxisIntervalModule } from './igx-time-axis-interval-module';
import { IgxTimeAxisIntervalComponent } from './igx-time-axis-interval-component';
import { TimeAxisInterval } from './TimeAxisInterval';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxTimeAxisIntervalCollection } from './igx-time-axis-interval-collection';
import { TimeAxisIntervalCollection } from './TimeAxisIntervalCollection';
import * as i0 from "@angular/core";
var IgxTimeAxisIntervalDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxTimeAxisIntervalDynamicModule() {
        TypeRegistrar.registerCons("IgxTimeAxisIntervalComponent", IgxTimeAxisIntervalComponent);
        TypeRegistrar.register("TimeAxisInterval", TimeAxisInterval.$type);
        TypeRegistrar.registerCons("IgxTimeAxisIntervalCollection", IgxTimeAxisIntervalCollection);
        TypeRegistrar.register("TimeAxisIntervalCollection", TimeAxisIntervalCollection.$type);
    }
    IgxTimeAxisIntervalDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisIntervalDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTimeAxisIntervalDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisIntervalDynamicModule, imports: [CommonModule, IgxTimeAxisIntervalModule], exports: [IgxTimeAxisIntervalModule] });
    IgxTimeAxisIntervalDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisIntervalDynamicModule, imports: [[CommonModule, IgxTimeAxisIntervalModule], IgxTimeAxisIntervalModule] });
    return IgxTimeAxisIntervalDynamicModule;
}());
export { IgxTimeAxisIntervalDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisIntervalDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxTimeAxisIntervalModule],
                    exports: [IgxTimeAxisIntervalModule],
                    entryComponents: [IgxTimeAxisIntervalComponent]
                }]
        }], ctorParameters: function () { return []; } });
