// igniteui-angular-charts/esm5/lib/igx-area-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { UnknownValuePlotting_$type } from "igniteui-angular-core";
import { IgxHorizontalAnchoredCategorySeriesComponent } from "./igx-horizontal-anchored-category-series-component";
import { IgxAnchoredCategorySeriesComponent } from "./igx-anchored-category-series-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { AreaSeries } from "./AreaSeries";
import { ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent area series.
 * Displays trend over time or ordered categories.
 * Useful when there are many data points and the order is important.
*
* The `AreaSeries` class represents a IgxDataChartComponent area series.Displays trend over time or ordered categories.
*
* Useful when there are many data points and the order is important.
*
* ```html
* <igx-data-chart #chart
*     [dataSource]="data">
*     <igx-category-x-axis
*         label="label"
*         #xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
*         #yAxis>
*     </igx-numeric-y-axis>
*     <igx-area-series
*         [xAxis]="xAxis"
*         [yAxis]="yAxis"
*         valueMemberPath="value"  >
*     </igx-area-series>
* </igx-data-chart>
* ```
*
* ```ts
* let series: IgxAreaSeriesComponent = new IgxAreaSeriesComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.valueMemberPath = "Value";
* this.chart.series.add(series);
* ```
*/
var IgxAreaSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxAreaSeriesComponent, _super);
    function IgxAreaSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxAreaSeriesComponent.prototype.createImplementation = function () {
        return new AreaSeries();
    };
    Object.defineProperty(IgxAreaSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxAreaSeriesComponent.prototype, "isAreaOrLine", {
        /**
         * Gets whether the current series shows an area or line shape.
        *
        * The `IsAreaOrLine` is used to check if the current series shows an area or line shape.
        *
        * ```ts
        * let series: IgxAreaSeriesComponent = new IgxAreaSeriesComponent();
        * let x= series.isAreaOrLine;
        * ```
        */
        get: function () {
            return this.i.et;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxAreaSeriesComponent.prototype, "isArea", {
        /**
         * Gets whether the current series shows an area shape.
        */
        get: function () {
            return this.i.es;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxAreaSeriesComponent.prototype, "unknownValuePlotting", {
        /**
         * Determines how unknown values will be plotted on the chart.
         * Null and Double.NaN are two examples of unknown values.
        *
        * The `unknownValuePlotting` property is used to determines how unknown values will be plotted on the chart.
        *
        * Null and Double.NaN are two examples of unknown values.
        *
        * ```html
        * <igx-data-chart #chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-area-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        * 		unknownValuePlotting="DontPlot"  >
        *     </igx-area-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * series.unknownValuePlotting = UnknownValuePlotting.LinearInterpolate;
        * ```
        */
        get: function () {
            return this.i.unknownValuePlotting;
        },
        set: function (v) {
            this.i.unknownValuePlotting = ensureEnum(UnknownValuePlotting_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    IgxAreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxAreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxAreaSeriesComponent, selector: "igx-area-series", inputs: { unknownValuePlotting: "unknownValuePlotting" }, providers: [{ provide: IgxHorizontalAnchoredCategorySeriesComponent, useExisting: forwardRef(function () { return IgxAreaSeriesComponent; }) }, { provide: IgxAnchoredCategorySeriesComponent, useExisting: forwardRef(function () { return IgxAreaSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxAreaSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxAreaSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxAreaSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxAreaSeriesComponent;
}(IgxHorizontalAnchoredCategorySeriesComponent));
export { IgxAreaSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-area-series',
                    template: "",
                    providers: [{ provide: IgxHorizontalAnchoredCategorySeriesComponent, useExisting: forwardRef(function () { return IgxAreaSeriesComponent; }) }, { provide: IgxAnchoredCategorySeriesComponent, useExisting: forwardRef(function () { return IgxAreaSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxAreaSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxAreaSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxAreaSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { unknownValuePlotting: [{
                type: Input
            }] } });
