// igniteui-angular-charts/esm5/lib/igx-hierarchical-ring-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxRingSeriesBaseComponent } from "./igx-ring-series-base-component";
import { HierarchicalRingSeries } from "./HierarchicalRingSeries";
import * as i0 from "@angular/core";
/**
 * Represents a
 * IgxDoughnutChartComponent hierarchical series.
*/
var IgxHierarchicalRingSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxHierarchicalRingSeriesComponent, _super);
    function IgxHierarchicalRingSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxHierarchicalRingSeriesComponent.prototype.createImplementation = function () {
        return new HierarchicalRingSeries();
    };
    Object.defineProperty(IgxHierarchicalRingSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxHierarchicalRingSeriesComponent.prototype, "childrenMemberPath", {
        /**
         * Gets or sets the property name that contains the children items collection.
        */
        get: function () {
            return this.i.et;
        },
        set: function (v) {
            this.i.et = v;
        },
        enumerable: false,
        configurable: true
    });
    IgxHierarchicalRingSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHierarchicalRingSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxHierarchicalRingSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxHierarchicalRingSeriesComponent, selector: "igx-hierarchical-ring-series", inputs: { childrenMemberPath: "childrenMemberPath" }, providers: [{ provide: IgxRingSeriesBaseComponent, useExisting: forwardRef(function () { return IgxHierarchicalRingSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxHierarchicalRingSeriesComponent;
}(IgxRingSeriesBaseComponent));
export { IgxHierarchicalRingSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHierarchicalRingSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-hierarchical-ring-series',
                    template: "",
                    providers: [{ provide: IgxRingSeriesBaseComponent, useExisting: forwardRef(function () { return IgxHierarchicalRingSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { childrenMemberPath: [{
                type: Input
            }] } });
