// igniteui-angular-charts/esm5/lib/igx-data-pie-chart-toolbar-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataChartToolbarIconsModule } from './igx-data-chart-toolbar-icons-module';
import { DataChartToolbarFactoryManager } from './DataChartToolbarFactoryManager';
import { DataPieChartToolbarFactoryManager } from './DataPieChartToolbarFactoryManager';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxDataPieChartToolbarModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataPieChartToolbarModule() {
        TypeRegistrar.register('DataChartToolbarFactoryManager', DataChartToolbarFactoryManager.$type);
        DataChartToolbarFactoryManager.register();
        TypeRegistrar.register('DataPieChartToolbarFactoryManager', DataPieChartToolbarFactoryManager.$type);
        DataPieChartToolbarFactoryManager.register();
    }
    IgxDataPieChartToolbarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartToolbarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataPieChartToolbarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartToolbarModule, imports: [CommonModule, IgxDataChartToolbarIconsModule], exports: [IgxDataChartToolbarIconsModule] });
    IgxDataPieChartToolbarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartToolbarModule, imports: [[CommonModule, IgxDataChartToolbarIconsModule
            ], IgxDataChartToolbarIconsModule] });
    return IgxDataPieChartToolbarModule;
}());
export { IgxDataPieChartToolbarModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartToolbarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataChartToolbarIconsModule
                    ],
                    imports: [CommonModule, IgxDataChartToolbarIconsModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
