// igniteui-angular-charts/esm5/lib/igx-value-brush-scale-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxValueBrushScaleComponent } from './igx-value-brush-scale-component';
import * as i0 from "@angular/core";
var IgxValueBrushScaleModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxValueBrushScaleModule() {
    }
    IgxValueBrushScaleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueBrushScaleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxValueBrushScaleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueBrushScaleModule, declarations: [IgxValueBrushScaleComponent], imports: [CommonModule], exports: [IgxValueBrushScaleComponent] });
    IgxValueBrushScaleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueBrushScaleModule, imports: [[CommonModule]] });
    return IgxValueBrushScaleModule;
}());
export { IgxValueBrushScaleModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueBrushScaleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxValueBrushScaleComponent],
                    imports: [CommonModule],
                    exports: [IgxValueBrushScaleComponent],
                    entryComponents: []
                }]
        }] });
