// igniteui-angular-charts/esm5/lib/igx-stacked-100-bar-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStackedBarSeriesComponent } from "./igx-stacked-bar-series-component";
import { IgxVerticalStackedSeriesBaseComponent } from "./igx-vertical-stacked-series-base-component";
import { IgxStackedSeriesBaseComponent } from "./igx-stacked-series-base-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { Stacked100BarSeries } from "./Stacked100BarSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent stacked100 bar series.
*/
var IgxStacked100BarSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxStacked100BarSeriesComponent, _super);
    function IgxStacked100BarSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxStacked100BarSeriesComponent.prototype.createImplementation = function () {
        return new Stacked100BarSeries();
    };
    Object.defineProperty(IgxStacked100BarSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxStacked100BarSeriesComponent.prototype, "isPercentBased", {
        get: function () {
            return this.i.aan;
        },
        enumerable: false,
        configurable: true
    });
    IgxStacked100BarSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStacked100BarSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStacked100BarSeriesComponent, selector: "igx-stacked-100-bar-series", providers: [{ provide: IgxStackedBarSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }, { provide: IgxVerticalStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStacked100BarSeriesComponent;
}(IgxStackedBarSeriesComponent));
export { IgxStacked100BarSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-stacked-100-bar-series',
                    template: "",
                    providers: [{ provide: IgxStackedBarSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }, { provide: IgxVerticalStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100BarSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
