// igniteui-angular-charts/esm5/lib/igx-slow-stochastic-oscillator-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSlowStochasticOscillatorIndicatorModule } from './igx-slow-stochastic-oscillator-indicator-module';
import { IgxSlowStochasticOscillatorIndicatorComponent } from './igx-slow-stochastic-oscillator-indicator-component';
import { SlowStochasticOscillatorIndicator } from './SlowStochasticOscillatorIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxSlowStochasticOscillatorIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxSlowStochasticOscillatorIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxSlowStochasticOscillatorIndicatorComponent", IgxSlowStochasticOscillatorIndicatorComponent);
        TypeRegistrar.register("SlowStochasticOscillatorIndicator", SlowStochasticOscillatorIndicator.$type);
    }
    IgxSlowStochasticOscillatorIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSlowStochasticOscillatorIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSlowStochasticOscillatorIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSlowStochasticOscillatorIndicatorDynamicModule, imports: [CommonModule, IgxSlowStochasticOscillatorIndicatorModule], exports: [IgxSlowStochasticOscillatorIndicatorModule] });
    IgxSlowStochasticOscillatorIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSlowStochasticOscillatorIndicatorDynamicModule, imports: [[CommonModule, IgxSlowStochasticOscillatorIndicatorModule], IgxSlowStochasticOscillatorIndicatorModule] });
    return IgxSlowStochasticOscillatorIndicatorDynamicModule;
}());
export { IgxSlowStochasticOscillatorIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSlowStochasticOscillatorIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxSlowStochasticOscillatorIndicatorModule],
                    exports: [IgxSlowStochasticOscillatorIndicatorModule],
                    entryComponents: [IgxSlowStochasticOscillatorIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
