// igniteui-angular-charts/esm5/lib/igx-percentage-volume-oscillator-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPercentageVolumeOscillatorIndicatorModule } from './igx-percentage-volume-oscillator-indicator-module';
import { IgxPercentageVolumeOscillatorIndicatorComponent } from './igx-percentage-volume-oscillator-indicator-component';
import { PercentageVolumeOscillatorIndicator } from './PercentageVolumeOscillatorIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxPercentageVolumeOscillatorIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPercentageVolumeOscillatorIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxPercentageVolumeOscillatorIndicatorComponent", IgxPercentageVolumeOscillatorIndicatorComponent);
        TypeRegistrar.register("PercentageVolumeOscillatorIndicator", PercentageVolumeOscillatorIndicator.$type);
    }
    IgxPercentageVolumeOscillatorIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentageVolumeOscillatorIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPercentageVolumeOscillatorIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentageVolumeOscillatorIndicatorDynamicModule, imports: [CommonModule, IgxPercentageVolumeOscillatorIndicatorModule], exports: [IgxPercentageVolumeOscillatorIndicatorModule] });
    IgxPercentageVolumeOscillatorIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentageVolumeOscillatorIndicatorDynamicModule, imports: [[CommonModule, IgxPercentageVolumeOscillatorIndicatorModule], IgxPercentageVolumeOscillatorIndicatorModule] });
    return IgxPercentageVolumeOscillatorIndicatorDynamicModule;
}());
export { IgxPercentageVolumeOscillatorIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentageVolumeOscillatorIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPercentageVolumeOscillatorIndicatorModule],
                    exports: [IgxPercentageVolumeOscillatorIndicatorModule],
                    entryComponents: [IgxPercentageVolumeOscillatorIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
