// igniteui-angular-charts/esm5/lib/igx-scatter-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterSeriesComponent } from './igx-scatter-series-component';
import * as i0 from "@angular/core";
var IgxScatterSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxScatterSeriesModule() {
    }
    IgxScatterSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesModule, declarations: [IgxScatterSeriesComponent], imports: [CommonModule], exports: [IgxScatterSeriesComponent] });
    IgxScatterSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesModule, imports: [[CommonModule]] });
    return IgxScatterSeriesModule;
}());
export { IgxScatterSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxScatterSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxScatterSeriesComponent],
                    entryComponents: []
                }]
        }] });
