// igniteui-angular-charts/esm5/lib/igx-median-price-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxMedianPriceIndicatorModule } from './igx-median-price-indicator-module';
import { IgxMedianPriceIndicatorComponent } from './igx-median-price-indicator-component';
import { MedianPriceIndicator } from './MedianPriceIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxMedianPriceIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxMedianPriceIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxMedianPriceIndicatorComponent", IgxMedianPriceIndicatorComponent);
        TypeRegistrar.register("MedianPriceIndicator", MedianPriceIndicator.$type);
    }
    IgxMedianPriceIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMedianPriceIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxMedianPriceIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMedianPriceIndicatorDynamicModule, imports: [CommonModule, IgxMedianPriceIndicatorModule], exports: [IgxMedianPriceIndicatorModule] });
    IgxMedianPriceIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMedianPriceIndicatorDynamicModule, imports: [[CommonModule, IgxMedianPriceIndicatorModule], IgxMedianPriceIndicatorModule] });
    return IgxMedianPriceIndicatorDynamicModule;
}());
export { IgxMedianPriceIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMedianPriceIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxMedianPriceIndicatorModule],
                    exports: [IgxMedianPriceIndicatorModule],
                    entryComponents: [IgxMedianPriceIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
