// igniteui-angular-charts/esm5/lib/igx-category-item-highlight-layer-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryItemHighlightLayerModule } from './igx-category-item-highlight-layer-module';
import { IgxCategoryItemHighlightLayerComponent } from './igx-category-item-highlight-layer-component';
import { CategoryItemHighlightLayer } from './CategoryItemHighlightLayer';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxCategoryItemHighlightLayerDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCategoryItemHighlightLayerDynamicModule() {
        TypeRegistrar.registerCons("IgxCategoryItemHighlightLayerComponent", IgxCategoryItemHighlightLayerComponent);
        TypeRegistrar.register("CategoryItemHighlightLayer", CategoryItemHighlightLayer.$type);
    }
    IgxCategoryItemHighlightLayerDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryItemHighlightLayerDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryItemHighlightLayerDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryItemHighlightLayerDynamicModule, imports: [CommonModule, IgxCategoryItemHighlightLayerModule], exports: [IgxCategoryItemHighlightLayerModule] });
    IgxCategoryItemHighlightLayerDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryItemHighlightLayerDynamicModule, imports: [[CommonModule, IgxCategoryItemHighlightLayerModule], IgxCategoryItemHighlightLayerModule] });
    return IgxCategoryItemHighlightLayerDynamicModule;
}());
export { IgxCategoryItemHighlightLayerDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryItemHighlightLayerDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCategoryItemHighlightLayerModule],
                    exports: [IgxCategoryItemHighlightLayerModule],
                    entryComponents: [IgxCategoryItemHighlightLayerComponent]
                }]
        }], ctorParameters: function () { return []; } });
