// igniteui-angular-charts/esm5/lib/igx-number-abbreviator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NumberAbbreviator } from './NumberAbbreviator';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxNumberAbbreviatorModule } from './igx-number-abbreviator-module';
import * as i0 from "@angular/core";
var IgxNumberAbbreviatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxNumberAbbreviatorDynamicModule() {
        TypeRegistrar.registerCons('NumberAbbreviator', NumberAbbreviator);
        TypeRegistrar.register('NumberAbbreviator', NumberAbbreviator.$type);
    }
    IgxNumberAbbreviatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumberAbbreviatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxNumberAbbreviatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumberAbbreviatorDynamicModule, imports: [CommonModule, IgxNumberAbbreviatorModule], exports: [IgxNumberAbbreviatorModule] });
    IgxNumberAbbreviatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumberAbbreviatorDynamicModule, imports: [[CommonModule, IgxNumberAbbreviatorModule
            ], IgxNumberAbbreviatorModule] });
    return IgxNumberAbbreviatorDynamicModule;
}());
export { IgxNumberAbbreviatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumberAbbreviatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumberAbbreviatorModule
                    ],
                    imports: [CommonModule, IgxNumberAbbreviatorModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
