// igniteui-angular-charts/esm5/lib/igx-numeric-angle-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericAngleAxisModule } from './igx-numeric-angle-axis-module';
import { IgxNumericAngleAxisComponent } from './igx-numeric-angle-axis-component';
import { NumericAngleAxis } from './NumericAngleAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxNumericAngleAxisDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxNumericAngleAxisDynamicModule() {
        TypeRegistrar.registerCons("IgxNumericAngleAxisComponent", IgxNumericAngleAxisComponent);
        TypeRegistrar.register("NumericAngleAxis", NumericAngleAxis.$type);
    }
    IgxNumericAngleAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAngleAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxNumericAngleAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAngleAxisDynamicModule, imports: [CommonModule, IgxNumericAngleAxisModule], exports: [IgxNumericAngleAxisModule] });
    IgxNumericAngleAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAngleAxisDynamicModule, imports: [[CommonModule, IgxNumericAngleAxisModule], IgxNumericAngleAxisModule] });
    return IgxNumericAngleAxisDynamicModule;
}());
export { IgxNumericAngleAxisDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAngleAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxNumericAngleAxisModule],
                    exports: [IgxNumericAngleAxisModule],
                    entryComponents: [IgxNumericAngleAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
