// igniteui-angular-charts/esm5/lib/igx-bollinger-band-width-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxBollingerBandWidthIndicatorModule } from './igx-bollinger-band-width-indicator-module';
import { IgxBollingerBandWidthIndicatorComponent } from './igx-bollinger-band-width-indicator-component';
import { BollingerBandWidthIndicator } from './BollingerBandWidthIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxBollingerBandWidthIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxBollingerBandWidthIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxBollingerBandWidthIndicatorComponent", IgxBollingerBandWidthIndicatorComponent);
        TypeRegistrar.register("BollingerBandWidthIndicator", BollingerBandWidthIndicator.$type);
    }
    IgxBollingerBandWidthIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBollingerBandWidthIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxBollingerBandWidthIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBollingerBandWidthIndicatorDynamicModule, imports: [CommonModule, IgxBollingerBandWidthIndicatorModule], exports: [IgxBollingerBandWidthIndicatorModule] });
    IgxBollingerBandWidthIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBollingerBandWidthIndicatorDynamicModule, imports: [[CommonModule, IgxBollingerBandWidthIndicatorModule], IgxBollingerBandWidthIndicatorModule] });
    return IgxBollingerBandWidthIndicatorDynamicModule;
}());
export { IgxBollingerBandWidthIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBollingerBandWidthIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxBollingerBandWidthIndicatorModule],
                    exports: [IgxBollingerBandWidthIndicatorModule],
                    entryComponents: [IgxBollingerBandWidthIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
