// igniteui-angular-charts/esm5/lib/igx-zoom-slider-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxZoomSliderComponent } from './igx-zoom-slider-component';
import { ZoomSlider } from './ZoomSlider';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxZoomSliderModule } from './igx-zoom-slider-module';
import * as i0 from "@angular/core";
var IgxZoomSliderDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxZoomSliderDynamicModule() {
        TypeRegistrar.registerCons('ZoomSlider', ZoomSlider);
        TypeRegistrar.registerCons('IgxZoomSliderComponent', IgxZoomSliderComponent);
        TypeRegistrar.register('ZoomSlider', ZoomSlider.$type);
    }
    IgxZoomSliderDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxZoomSliderDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxZoomSliderDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxZoomSliderDynamicModule, imports: [CommonModule, IgxZoomSliderModule], exports: [IgxZoomSliderModule] });
    IgxZoomSliderDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxZoomSliderDynamicModule, imports: [[CommonModule, IgxZoomSliderModule
            ], IgxZoomSliderModule] });
    return IgxZoomSliderDynamicModule;
}());
export { IgxZoomSliderDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxZoomSliderDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxZoomSliderModule
                    ],
                    imports: [CommonModule, IgxZoomSliderModule
                    ],
                    entryComponents: [IgxZoomSliderComponent]
                }]
        }], ctorParameters: function () { return []; } });
