// igniteui-angular-charts/esm5/lib/igx-linear-contour-value-resolver-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxContourValueResolverComponent } from "./igx-contour-value-resolver-component";
import { LinearContourValueResolver } from "./LinearContourValueResolver";
import * as i0 from "@angular/core";
/**
 * ContourValueResolver class which resolves contour lines to be distributed at linear intervals from the data minimum to the data maximum.
*/
var IgxLinearContourValueResolverComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxLinearContourValueResolverComponent, _super);
    function IgxLinearContourValueResolverComponent() {
        return _super.call(this) || this;
    }
    IgxLinearContourValueResolverComponent.prototype.createImplementation = function () {
        return new LinearContourValueResolver();
    };
    Object.defineProperty(IgxLinearContourValueResolverComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxLinearContourValueResolverComponent.prototype, "valueCount", {
        /**
         * Gets or sets how many contour values to return.
        */
        get: function () {
            return this.i.valueCount;
        },
        set: function (v) {
            this.i.valueCount = +v;
        },
        enumerable: false,
        configurable: true
    });
    IgxLinearContourValueResolverComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxLinearContourValueResolverComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxLinearContourValueResolverComponent, selector: "igx-linear-contour-value-resolver", inputs: { valueCount: "valueCount" }, providers: [{ provide: IgxContourValueResolverComponent, useExisting: forwardRef(function () { return IgxLinearContourValueResolverComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxLinearContourValueResolverComponent;
}(IgxContourValueResolverComponent));
export { IgxLinearContourValueResolverComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-linear-contour-value-resolver',
                    template: "",
                    providers: [{ provide: IgxContourValueResolverComponent, useExisting: forwardRef(function () { return IgxLinearContourValueResolverComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { valueCount: [{
                type: Input
            }] } });
