// igniteui-angular-charts/esm5/lib/igx-data-chart-financial-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFinancialPriceSeriesDynamicModule } from './igx-financial-price-series-dynamic-module';
import { IgxDataChartFinancialCoreDynamicModule } from './igx-data-chart-financial-core-dynamic-module';
import { IgxDataChartFinancialModule } from './igx-data-chart-financial-module';
import * as i0 from "@angular/core";
var IgxDataChartFinancialDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartFinancialDynamicModule() {
    }
    IgxDataChartFinancialDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartFinancialDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialDynamicModule, imports: [CommonModule, IgxFinancialPriceSeriesDynamicModule,
            IgxDataChartFinancialCoreDynamicModule,
            IgxDataChartFinancialModule], exports: [IgxFinancialPriceSeriesDynamicModule,
            IgxDataChartFinancialCoreDynamicModule,
            IgxDataChartFinancialModule] });
    IgxDataChartFinancialDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialDynamicModule, imports: [[CommonModule, IgxFinancialPriceSeriesDynamicModule,
                IgxDataChartFinancialCoreDynamicModule,
                IgxDataChartFinancialModule
            ], IgxFinancialPriceSeriesDynamicModule,
            IgxDataChartFinancialCoreDynamicModule,
            IgxDataChartFinancialModule] });
    return IgxDataChartFinancialDynamicModule;
}());
export { IgxDataChartFinancialDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxFinancialPriceSeriesDynamicModule,
                        IgxDataChartFinancialCoreDynamicModule,
                        IgxDataChartFinancialModule
                    ],
                    imports: [CommonModule, IgxFinancialPriceSeriesDynamicModule,
                        IgxDataChartFinancialCoreDynamicModule,
                        IgxDataChartFinancialModule
                    ],
                    entryComponents: []
                }]
        }] });
