// igniteui-angular-charts/esm5/lib/igx-waterfall-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxHorizontalAnchoredCategorySeriesProxyModule } from './igx-horizontal-anchored-category-series-proxy-module';
import { IgxWaterfallSeriesComponent } from './igx-waterfall-series-component';
import * as i0 from "@angular/core";
var IgxWaterfallSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxWaterfallSeriesModule() {
    }
    IgxWaterfallSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWaterfallSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxWaterfallSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWaterfallSeriesModule, declarations: [IgxWaterfallSeriesComponent], imports: [CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule], exports: [IgxWaterfallSeriesComponent, IgxHorizontalAnchoredCategorySeriesProxyModule] });
    IgxWaterfallSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWaterfallSeriesModule, imports: [[CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule], IgxHorizontalAnchoredCategorySeriesProxyModule] });
    return IgxWaterfallSeriesModule;
}());
export { IgxWaterfallSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWaterfallSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxWaterfallSeriesComponent],
                    imports: [CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule],
                    exports: [IgxWaterfallSeriesComponent, IgxHorizontalAnchoredCategorySeriesProxyModule],
                    entryComponents: []
                }]
        }] });
