// igniteui-angular-charts/esm5/lib/igx-scatter-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxScatterBaseComponent } from "./igx-scatter-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { ScatterSeries } from "./ScatterSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent scatter series
*/
var IgxScatterSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxScatterSeriesComponent, _super);
    function IgxScatterSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxScatterSeriesComponent.prototype.createImplementation = function () {
        return new ScatterSeries();
    };
    Object.defineProperty(IgxScatterSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxScatterSeriesComponent.prototype, "hasOnlyMarkers", {
        /**
         * Gets whether the series has only marker as visuals
        */
        get: function () {
            return this.i.eg;
        },
        enumerable: false,
        configurable: true
    });
    IgxScatterSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxScatterSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxScatterSeriesComponent, selector: "igx-scatter-series", providers: [{ provide: IgxScatterBaseComponent, useExisting: forwardRef(function () { return IgxScatterSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxScatterSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxScatterSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxScatterSeriesComponent;
}(IgxScatterBaseComponent));
export { IgxScatterSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-scatter-series',
                    template: "",
                    providers: [{ provide: IgxScatterBaseComponent, useExisting: forwardRef(function () { return IgxScatterSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxScatterSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxScatterSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
