// igniteui-angular-charts/esm5/lib/igx-financial-chart-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxLineSeriesDynamicModule } from './igx-line-series-dynamic-module';
import { IgxAreaSeriesDynamicModule } from './igx-area-series-dynamic-module';
import { IgxColumnSeriesDynamicModule } from './igx-column-series-dynamic-module';
import { IgxNumericYAxisDynamicModule } from './igx-numeric-y-axis-dynamic-module';
import { IgxTimeXAxisDynamicModule } from './igx-time-x-axis-dynamic-module';
import { IgxTimeAxisBreakDynamicModule } from './igx-time-axis-break-dynamic-module';
import { IgxTimeAxisIntervalDynamicModule } from './igx-time-axis-interval-dynamic-module';
import { IgxTimeAxisLabelFormatDynamicModule } from './igx-time-axis-label-format-dynamic-module';
import { IgxCategoryXAxisDynamicModule } from './igx-category-x-axis-dynamic-module';
import { IgxOrdinalTimeXAxisDynamicModule } from './igx-ordinal-time-x-axis-dynamic-module';
import { IgxPercentChangeYAxisDynamicModule } from './igx-percent-change-y-axis-dynamic-module';
import { IgxFinancialPriceSeriesDynamicModule } from './igx-financial-price-series-dynamic-module';
import { IgxDataChartCategoryTrendLineDynamicModule } from './igx-data-chart-category-trend-line-dynamic-module';
import { IgxIndicatorsDynamicModule } from './igx-indicators-dynamic-module';
import { IgxOverlaysDynamicModule } from './igx-overlays-dynamic-module';
import { IgxZoomSliderDynamicModule } from './igx-zoom-slider-dynamic-module';
import { IgxCrosshairLayerDynamicModule } from './igx-crosshair-layer-dynamic-module';
import { IgxCalloutLayerDynamicModule } from './igx-callout-layer-dynamic-module';
import { IgxItemToolTipLayerDynamicModule } from './igx-item-tool-tip-layer-dynamic-module';
import { IgxCategoryToolTipLayerDynamicModule } from './igx-category-tool-tip-layer-dynamic-module';
import { IgxCategoryHighlightLayerDynamicModule } from './igx-category-highlight-layer-dynamic-module';
import { IgxCategoryItemHighlightLayerDynamicModule } from './igx-category-item-highlight-layer-dynamic-module';
import { IgxDataToolTipLayerDynamicModule } from './igx-data-tool-tip-layer-dynamic-module';
import { IgxFinalValueLayerDynamicModule } from './igx-final-value-layer-dynamic-module';
import { IgxNumberAbbreviatorDynamicModule } from './igx-number-abbreviator-dynamic-module';
import { IgxDataChartInteractivityDynamicModule } from './igx-data-chart-interactivity-dynamic-module';
import { IgxDataChartAnnotationDynamicModule } from './igx-data-chart-annotation-dynamic-module';
import { IgxFinancialLegendDynamicModule } from './igx-financial-legend-dynamic-module';
import { IgxFinancialChartCoreDynamicModule } from './igx-financial-chart-core-dynamic-module';
import { IgxFinancialChartModule } from './igx-financial-chart-module';
import * as i0 from "@angular/core";
var IgxFinancialChartDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxFinancialChartDynamicModule() {
    }
    IgxFinancialChartDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFinancialChartDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartDynamicModule, imports: [CommonModule, IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxTimeXAxisDynamicModule,
            IgxTimeAxisBreakDynamicModule,
            IgxTimeAxisIntervalDynamicModule,
            IgxTimeAxisLabelFormatDynamicModule,
            IgxCategoryXAxisDynamicModule,
            IgxOrdinalTimeXAxisDynamicModule,
            IgxPercentChangeYAxisDynamicModule,
            IgxFinancialPriceSeriesDynamicModule,
            IgxDataChartCategoryTrendLineDynamicModule,
            IgxIndicatorsDynamicModule,
            IgxOverlaysDynamicModule,
            IgxZoomSliderDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxNumberAbbreviatorDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxDataChartAnnotationDynamicModule,
            IgxFinancialLegendDynamicModule,
            IgxFinancialChartCoreDynamicModule,
            IgxFinancialChartModule], exports: [IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxTimeXAxisDynamicModule,
            IgxTimeAxisBreakDynamicModule,
            IgxTimeAxisIntervalDynamicModule,
            IgxTimeAxisLabelFormatDynamicModule,
            IgxCategoryXAxisDynamicModule,
            IgxOrdinalTimeXAxisDynamicModule,
            IgxPercentChangeYAxisDynamicModule,
            IgxFinancialPriceSeriesDynamicModule,
            IgxDataChartCategoryTrendLineDynamicModule,
            IgxIndicatorsDynamicModule,
            IgxOverlaysDynamicModule,
            IgxZoomSliderDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxNumberAbbreviatorDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxDataChartAnnotationDynamicModule,
            IgxFinancialLegendDynamicModule,
            IgxFinancialChartCoreDynamicModule,
            IgxFinancialChartModule] });
    IgxFinancialChartDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartDynamicModule, imports: [[CommonModule, IgxLineSeriesDynamicModule,
                IgxAreaSeriesDynamicModule,
                IgxColumnSeriesDynamicModule,
                IgxNumericYAxisDynamicModule,
                IgxTimeXAxisDynamicModule,
                IgxTimeAxisBreakDynamicModule,
                IgxTimeAxisIntervalDynamicModule,
                IgxTimeAxisLabelFormatDynamicModule,
                IgxCategoryXAxisDynamicModule,
                IgxOrdinalTimeXAxisDynamicModule,
                IgxPercentChangeYAxisDynamicModule,
                IgxFinancialPriceSeriesDynamicModule,
                IgxDataChartCategoryTrendLineDynamicModule,
                IgxIndicatorsDynamicModule,
                IgxOverlaysDynamicModule,
                IgxZoomSliderDynamicModule,
                IgxCrosshairLayerDynamicModule,
                IgxCalloutLayerDynamicModule,
                IgxItemToolTipLayerDynamicModule,
                IgxCategoryToolTipLayerDynamicModule,
                IgxCategoryHighlightLayerDynamicModule,
                IgxCategoryItemHighlightLayerDynamicModule,
                IgxDataToolTipLayerDynamicModule,
                IgxFinalValueLayerDynamicModule,
                IgxNumberAbbreviatorDynamicModule,
                IgxDataChartInteractivityDynamicModule,
                IgxDataChartAnnotationDynamicModule,
                IgxFinancialLegendDynamicModule,
                IgxFinancialChartCoreDynamicModule,
                IgxFinancialChartModule
            ], IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxTimeXAxisDynamicModule,
            IgxTimeAxisBreakDynamicModule,
            IgxTimeAxisIntervalDynamicModule,
            IgxTimeAxisLabelFormatDynamicModule,
            IgxCategoryXAxisDynamicModule,
            IgxOrdinalTimeXAxisDynamicModule,
            IgxPercentChangeYAxisDynamicModule,
            IgxFinancialPriceSeriesDynamicModule,
            IgxDataChartCategoryTrendLineDynamicModule,
            IgxIndicatorsDynamicModule,
            IgxOverlaysDynamicModule,
            IgxZoomSliderDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxNumberAbbreviatorDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxDataChartAnnotationDynamicModule,
            IgxFinancialLegendDynamicModule,
            IgxFinancialChartCoreDynamicModule,
            IgxFinancialChartModule] });
    return IgxFinancialChartDynamicModule;
}());
export { IgxFinancialChartDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxLineSeriesDynamicModule,
                        IgxAreaSeriesDynamicModule,
                        IgxColumnSeriesDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxTimeXAxisDynamicModule,
                        IgxTimeAxisBreakDynamicModule,
                        IgxTimeAxisIntervalDynamicModule,
                        IgxTimeAxisLabelFormatDynamicModule,
                        IgxCategoryXAxisDynamicModule,
                        IgxOrdinalTimeXAxisDynamicModule,
                        IgxPercentChangeYAxisDynamicModule,
                        IgxFinancialPriceSeriesDynamicModule,
                        IgxDataChartCategoryTrendLineDynamicModule,
                        IgxIndicatorsDynamicModule,
                        IgxOverlaysDynamicModule,
                        IgxZoomSliderDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxFinalValueLayerDynamicModule,
                        IgxNumberAbbreviatorDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxDataChartAnnotationDynamicModule,
                        IgxFinancialLegendDynamicModule,
                        IgxFinancialChartCoreDynamicModule,
                        IgxFinancialChartModule
                    ],
                    imports: [CommonModule, IgxLineSeriesDynamicModule,
                        IgxAreaSeriesDynamicModule,
                        IgxColumnSeriesDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxTimeXAxisDynamicModule,
                        IgxTimeAxisBreakDynamicModule,
                        IgxTimeAxisIntervalDynamicModule,
                        IgxTimeAxisLabelFormatDynamicModule,
                        IgxCategoryXAxisDynamicModule,
                        IgxOrdinalTimeXAxisDynamicModule,
                        IgxPercentChangeYAxisDynamicModule,
                        IgxFinancialPriceSeriesDynamicModule,
                        IgxDataChartCategoryTrendLineDynamicModule,
                        IgxIndicatorsDynamicModule,
                        IgxOverlaysDynamicModule,
                        IgxZoomSliderDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxFinalValueLayerDynamicModule,
                        IgxNumberAbbreviatorDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxDataChartAnnotationDynamicModule,
                        IgxFinancialLegendDynamicModule,
                        IgxFinancialChartCoreDynamicModule,
                        IgxFinancialChartModule
                    ],
                    entryComponents: []
                }]
        }] });
