// igniteui-angular-charts/esm5/lib/igx-polar-scatter-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxPolarBaseComponent } from "./igx-polar-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { PolarScatterSeries } from "./PolarScatterSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent polar scatter series.
*/
var IgxPolarScatterSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxPolarScatterSeriesComponent, _super);
    function IgxPolarScatterSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxPolarScatterSeriesComponent.prototype.createImplementation = function () {
        return new PolarScatterSeries();
    };
    Object.defineProperty(IgxPolarScatterSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxPolarScatterSeriesComponent.prototype, "hasOnlyMarkers", {
        /**
         * Gets whether the series has only marker as visuals
        */
        get: function () {
            return this.i.eg;
        },
        enumerable: false,
        configurable: true
    });
    IgxPolarScatterSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarScatterSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxPolarScatterSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxPolarScatterSeriesComponent, selector: "igx-polar-scatter-series", providers: [{ provide: IgxPolarBaseComponent, useExisting: forwardRef(function () { return IgxPolarScatterSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxPolarScatterSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxPolarScatterSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxPolarScatterSeriesComponent;
}(IgxPolarBaseComponent));
export { IgxPolarScatterSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarScatterSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-polar-scatter-series',
                    template: "",
                    providers: [{ provide: IgxPolarBaseComponent, useExisting: forwardRef(function () { return IgxPolarScatterSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxPolarScatterSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxPolarScatterSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
