// igniteui-angular-charts/esm5/lib/igx-positive-volume-index-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPositiveVolumeIndexIndicatorModule } from './igx-positive-volume-index-indicator-module';
import { IgxPositiveVolumeIndexIndicatorComponent } from './igx-positive-volume-index-indicator-component';
import { PositiveVolumeIndexIndicator } from './PositiveVolumeIndexIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxPositiveVolumeIndexIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPositiveVolumeIndexIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxPositiveVolumeIndexIndicatorComponent", IgxPositiveVolumeIndexIndicatorComponent);
        TypeRegistrar.register("PositiveVolumeIndexIndicator", PositiveVolumeIndexIndicator.$type);
    }
    IgxPositiveVolumeIndexIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPositiveVolumeIndexIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPositiveVolumeIndexIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPositiveVolumeIndexIndicatorDynamicModule, imports: [CommonModule, IgxPositiveVolumeIndexIndicatorModule], exports: [IgxPositiveVolumeIndexIndicatorModule] });
    IgxPositiveVolumeIndexIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPositiveVolumeIndexIndicatorDynamicModule, imports: [[CommonModule, IgxPositiveVolumeIndexIndicatorModule], IgxPositiveVolumeIndexIndicatorModule] });
    return IgxPositiveVolumeIndexIndicatorDynamicModule;
}());
export { IgxPositiveVolumeIndexIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPositiveVolumeIndexIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPositiveVolumeIndexIndicatorModule],
                    exports: [IgxPositiveVolumeIndexIndicatorModule],
                    entryComponents: [IgxPositiveVolumeIndexIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
