// igniteui-angular-charts/esm5/lib/igx-financial-price-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFinancialPriceSeriesModule } from './igx-financial-price-series-module';
import { IgxFinancialPriceSeriesComponent } from './igx-financial-price-series-component';
import { FinancialPriceSeries } from './FinancialPriceSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxFinancialPriceSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxFinancialPriceSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxFinancialPriceSeriesComponent", IgxFinancialPriceSeriesComponent);
        TypeRegistrar.register("FinancialPriceSeries", FinancialPriceSeries.$type);
    }
    IgxFinancialPriceSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFinancialPriceSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesDynamicModule, imports: [CommonModule, IgxFinancialPriceSeriesModule], exports: [IgxFinancialPriceSeriesModule] });
    IgxFinancialPriceSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesDynamicModule, imports: [[CommonModule, IgxFinancialPriceSeriesModule], IgxFinancialPriceSeriesModule] });
    return IgxFinancialPriceSeriesDynamicModule;
}());
export { IgxFinancialPriceSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxFinancialPriceSeriesModule],
                    exports: [IgxFinancialPriceSeriesModule],
                    entryComponents: [IgxFinancialPriceSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
