// igniteui-angular-charts/esm5/lib/igx-chart-summary-description-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxChartSummaryDescriptionModule } from './igx-chart-summary-description-module';
import { IgxChartSummaryDescription } from './igx-chart-summary-description';
import { ChartSummaryDescription } from './ChartSummaryDescription';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxChartSummaryDescriptionDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxChartSummaryDescriptionDynamicModule() {
        TypeRegistrar.registerCons("IgxChartSummaryDescription", IgxChartSummaryDescription);
        TypeRegistrar.register("ChartSummaryDescription", ChartSummaryDescription.$type);
    }
    IgxChartSummaryDescriptionDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSummaryDescriptionDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxChartSummaryDescriptionDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSummaryDescriptionDynamicModule, imports: [CommonModule, IgxChartSummaryDescriptionModule], exports: [IgxChartSummaryDescriptionModule] });
    IgxChartSummaryDescriptionDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSummaryDescriptionDynamicModule, imports: [[CommonModule, IgxChartSummaryDescriptionModule], IgxChartSummaryDescriptionModule] });
    return IgxChartSummaryDescriptionDynamicModule;
}());
export { IgxChartSummaryDescriptionDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSummaryDescriptionDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxChartSummaryDescriptionModule],
                    exports: [IgxChartSummaryDescriptionModule],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
