// igniteui-angular-charts/esm5/lib/igx-range-category-series-component.js
import { __extends } from "tslib";
import { Component, Input } from '@angular/core';
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import * as i0 from "@angular/core";
/**
 * Represents the base class for all IgxDataChartComponent ranged category/value series.
*
* The `RangeCategorySeries` represents the base class for all IgxDataChartComponent ranged category/value series.
*
* ```html
* <igx-data-chart  #chart
*     [dataSource]="data">
*     <igx-category-x-axis
*     label="label"
*     #xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
*     minimumValue="0"
*     #yAxis>
*     </igx-numeric-y-axis>
*     <igx-range-column-series
*     [xAxis]="xAxis"
*     [yAxis]="yAxis"
*     highMemberPath="high"
*     lowMemberPath="low">
* </igx-range-column-series>
* </igx-data-chart>
* ```
*
* ```ts
* let series = new IgxRangeColumnSeriesComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* this.chart.series.add(series);
* ```
*/
var IgxRangeCategorySeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxRangeCategorySeriesComponent, _super);
    function IgxRangeCategorySeriesComponent() {
        return _super.call(this) || this;
    }
    Object.defineProperty(IgxRangeCategorySeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRangeCategorySeriesComponent.prototype, "lowMemberPath", {
        /**
         * Gets or sets the value mapping property for the current series object.
        *
        * The `LowMemberPath` is used for low mapping property of the current series object.
        *
        * ```html
        * <igx-data-chart  #chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *     label="label"
        *     #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *     minimumValue="0"
        *     #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-range-column-series
        *     [xAxis]="xAxis"
        *     [yAxis]="yAxis"
        *     highMemberPath="high"
        *     lowMemberPath="low">
        * </igx-range-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * let series = new IgxRangeColumnSeriesComponent();
        * series.xAxis = this.xAxis;
        * series.yAxis = this.yAxis;
        * series.highMemberPath = "high";
        * series.lowMemberPath = "low";
        * this.chart.series.add(series);
        * ```
        */
        get: function () {
            return this.i.aaq;
        },
        set: function (v) {
            this.i.aaq = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRangeCategorySeriesComponent.prototype, "highMemberPath", {
        /**
         * Gets or sets the value mapping property for the current series object.
        *
        * Use `HighMemberPath` propert for high mapping of the current series object.
        *
        * ```html
        * <igx-data-chart  #chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *     label="label"
        *     #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *     minimumValue="0"
        *     #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-range-column-series
        *     [xAxis]="xAxis"
        *     [yAxis]="yAxis"
        *     highMemberPath="high"
        *     lowMemberPath="low">
        * </igx-range-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * series.highMemberPath = "high";
        * ```
        */
        get: function () {
            return this.i.aan;
        },
        set: function (v) {
            this.i.aan = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRangeCategorySeriesComponent.prototype, "highlightedLowMemberPath", {
        /**
         * Gets or sets the highlighted low value mapping property for the current series object.
        */
        get: function () {
            return this.i.aal;
        },
        set: function (v) {
            this.i.aal = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRangeCategorySeriesComponent.prototype, "highlightedHighMemberPath", {
        /**
         * Gets or sets the highlighted High value mapping property for the current series object.
        */
        get: function () {
            return this.i.aaj;
        },
        set: function (v) {
            this.i.aaj = v;
        },
        enumerable: false,
        configurable: true
    });
    IgxRangeCategorySeriesComponent.prototype.getItemValue = function (item, memberPathName) {
        var iv = this.i.ko(item, memberPathName);
        return (iv);
    };
    /**
     * Gets the value of a requested member path from the series.
    
    * @param memberPathName  * The property name of a valid member path for the series
    */
    IgxRangeCategorySeriesComponent.prototype.getMemberPathValue = function (memberPathName) {
        var iv = this.i.mb(memberPathName);
        return (iv);
    };
    /**
     * Scrolls the series to display the item for the specified data item.
     * The series is scrolled by the minimum amount required to place the specified data item within
     * the central 80% of the visible axis.
    
    * @param item  * The data item (item) to scroll to.
    *
    * `ScrollIntoView` method Scrolls the series to display the item for the specified data item.
    */
    IgxRangeCategorySeriesComponent.prototype.scrollIntoView = function (item) {
        var iv = this.i.gb(item);
        return (iv);
    };
    IgxRangeCategorySeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeCategorySeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxRangeCategorySeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxRangeCategorySeriesComponent, selector: "ng-component", inputs: { lowMemberPath: "lowMemberPath", highMemberPath: "highMemberPath", highlightedLowMemberPath: "highlightedLowMemberPath", highlightedHighMemberPath: "highlightedHighMemberPath" }, usesInheritance: true, ngImport: i0, template: "", isInline: true });
    return IgxRangeCategorySeriesComponent;
}(IgxCategorySeriesComponent));
export { IgxRangeCategorySeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeCategorySeriesComponent, decorators: [{
            type: Component,
            args: [{
                    template: "",
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { lowMemberPath: [{
                type: Input
            }], highMemberPath: [{
                type: Input
            }], highlightedLowMemberPath: [{
                type: Input
            }], highlightedHighMemberPath: [{
                type: Input
            }] } });
