// igniteui-angular-charts/esm5/lib/igx-market-facilitation-index-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxMarketFacilitationIndexIndicatorModule } from './igx-market-facilitation-index-indicator-module';
import { IgxMarketFacilitationIndexIndicatorComponent } from './igx-market-facilitation-index-indicator-component';
import { MarketFacilitationIndexIndicator } from './MarketFacilitationIndexIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxMarketFacilitationIndexIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxMarketFacilitationIndexIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxMarketFacilitationIndexIndicatorComponent", IgxMarketFacilitationIndexIndicatorComponent);
        TypeRegistrar.register("MarketFacilitationIndexIndicator", MarketFacilitationIndexIndicator.$type);
    }
    IgxMarketFacilitationIndexIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMarketFacilitationIndexIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxMarketFacilitationIndexIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMarketFacilitationIndexIndicatorDynamicModule, imports: [CommonModule, IgxMarketFacilitationIndexIndicatorModule], exports: [IgxMarketFacilitationIndexIndicatorModule] });
    IgxMarketFacilitationIndexIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMarketFacilitationIndexIndicatorDynamicModule, imports: [[CommonModule, IgxMarketFacilitationIndexIndicatorModule], IgxMarketFacilitationIndexIndicatorModule] });
    return IgxMarketFacilitationIndexIndicatorDynamicModule;
}());
export { IgxMarketFacilitationIndexIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMarketFacilitationIndexIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxMarketFacilitationIndexIndicatorModule],
                    exports: [IgxMarketFacilitationIndexIndicatorModule],
                    entryComponents: [IgxMarketFacilitationIndexIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
