// igniteui-angular-charts/esm5/lib/igx-commodity-channel-index-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxCommodityChannelIndexIndicatorComponent } from './igx-commodity-channel-index-indicator-component';
import * as i0 from "@angular/core";
var IgxCommodityChannelIndexIndicatorModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCommodityChannelIndexIndicatorModule() {
    }
    IgxCommodityChannelIndexIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCommodityChannelIndexIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCommodityChannelIndexIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCommodityChannelIndexIndicatorModule, declarations: [IgxCommodityChannelIndexIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxCommodityChannelIndexIndicatorComponent, IgxIndicatorProxyModule] });
    IgxCommodityChannelIndexIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCommodityChannelIndexIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxCommodityChannelIndexIndicatorModule;
}());
export { IgxCommodityChannelIndexIndicatorModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCommodityChannelIndexIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxCommodityChannelIndexIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxCommodityChannelIndexIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
