// igniteui-angular-charts/esm5/lib/igx-step-area-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxHorizontalAnchoredCategorySeriesComponent } from "./igx-horizontal-anchored-category-series-component";
import { IgxAnchoredCategorySeriesComponent } from "./igx-anchored-category-series-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { StepAreaSeries } from "./StepAreaSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent step area series.
*
* ```html
* <igx-data-chart
*     [dataSource]="data">
*     <igx-category-x-axis
*         label="label"
*         #xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
*         #yAxis>
*     </igx-numeric-y-axis>
*     <igx-step-area-series
*         [xAxis]="xAxis"
*         [yAxis]="yAxis"
*         valueMemberPath="value">
*     </igx-step-area-series>
* </igx-data-chart>
* ```
*
* ```ts
*    let series = new IgxStepAreaSeriesComponent();
*    series.xAxis = this.xAxis;
*    series.yAxis = this.yAxis;
*    series.valueMemberPath="value";
*    this.chart.series.add(series);
* ```
*/
var IgxStepAreaSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxStepAreaSeriesComponent, _super);
    function IgxStepAreaSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxStepAreaSeriesComponent.prototype.createImplementation = function () {
        return new StepAreaSeries();
    };
    Object.defineProperty(IgxStepAreaSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxStepAreaSeriesComponent.prototype, "isAreaOrLine", {
        /**
         * Gets whether the current series shows an area or line shape.
        */
        get: function () {
            return this.i.et;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxStepAreaSeriesComponent.prototype, "isArea", {
        /**
         * Gets whether the current series shows an area shape.
        */
        get: function () {
            return this.i.es;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxStepAreaSeriesComponent.prototype, "isStep", {
        /**
         * Gets whether the current series shows step shapes.
        */
        get: function () {
            return this.i.fz;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxStepAreaSeriesComponent.prototype, "isMarkerlessDisplayPreferred", {
        /**
         * Overridden by derived series classes to indicate when marker-less display is preferred or not.
        */
        get: function () {
            return this.i.fj;
        },
        enumerable: false,
        configurable: true
    });
    IgxStepAreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStepAreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStepAreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStepAreaSeriesComponent, selector: "igx-step-area-series", providers: [{ provide: IgxHorizontalAnchoredCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStepAreaSeriesComponent; }) }, { provide: IgxAnchoredCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStepAreaSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStepAreaSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxStepAreaSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxStepAreaSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStepAreaSeriesComponent;
}(IgxHorizontalAnchoredCategorySeriesComponent));
export { IgxStepAreaSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStepAreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-step-area-series',
                    template: "",
                    providers: [{ provide: IgxHorizontalAnchoredCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStepAreaSeriesComponent; }) }, { provide: IgxAnchoredCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStepAreaSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStepAreaSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxStepAreaSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxStepAreaSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
