// igniteui-angular-charts/esm5/lib/igx-custom-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCustomIndicatorModule } from './igx-custom-indicator-module';
import { IgxCustomIndicatorComponent } from './igx-custom-indicator-component';
import { CustomIndicator } from './CustomIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxCustomIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCustomIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxCustomIndicatorComponent", IgxCustomIndicatorComponent);
        TypeRegistrar.register("CustomIndicator", CustomIndicator.$type);
    }
    IgxCustomIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCustomIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomIndicatorDynamicModule, imports: [CommonModule, IgxCustomIndicatorModule], exports: [IgxCustomIndicatorModule] });
    IgxCustomIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomIndicatorDynamicModule, imports: [[CommonModule, IgxCustomIndicatorModule], IgxCustomIndicatorModule] });
    return IgxCustomIndicatorDynamicModule;
}());
export { IgxCustomIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCustomIndicatorModule],
                    exports: [IgxCustomIndicatorModule],
                    entryComponents: [IgxCustomIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
