// igniteui-angular-charts/esm5/lib/igx-size-scale-component.js
import { EventEmitter, Output, Component, Input, ChangeDetectionStrategy } from '@angular/core';
import { delegateCombine } from "igniteui-angular-core";
import { IgxPropertyUpdatedEventArgs } from "igniteui-angular-core";
import { SizeScale } from "./SizeScale";
import { TypeRegistrar } from "igniteui-angular-core";
import { NamePatcher, ensureBool, toSpinal, initializePropertiesFromCss } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a scale that is used determine an object's size.
*/
var IgxSizeScaleComponent = /** @class */ /*@__PURE__*/ (function () {
    function IgxSizeScaleComponent() {
        this.__p = null;
        this._hasUserValues = new Set();
        this._stylingContainer = null;
        this._stylingParent = null;
        this._inStyling = false;
        this._propertyUpdated = null;
        this._zoneRunner = null;
        if (this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
        }
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgxSizeScaleComponent.prototype.createImplementation = function () {
        return new SizeScale();
    };
    IgxSizeScaleComponent.prototype.ngOnInit = function () {
    };
    Object.defineProperty(IgxSizeScaleComponent.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        } /**
                                     * @hidden
                                     */,
        enumerable: false,
        configurable: true
    });
    IgxSizeScaleComponent._createFromInternal = function (internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        var name = internal.$type.name;
        var externalName = "Igx" + name + "Component";
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    };
    IgxSizeScaleComponent.prototype.onImplementationCreated = function () {
    };
    Object.defineProperty(IgxSizeScaleComponent.prototype, "globalMinimum", {
        /**
         * Gets or sets global minimum value that will be mapped to MinumumValue of this scale
        */
        get: function () {
            return this.i.globalMinimum;
        },
        set: function (v) {
            this.i.globalMinimum = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSizeScaleComponent.prototype, "globalMaximum", {
        /**
         * Gets or sets global maximum value that will be mapped to MaxumumValue of this scale
        */
        get: function () {
            return this.i.globalMaximum;
        },
        set: function (v) {
            this.i.globalMaximum = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSizeScaleComponent.prototype, "minimumValue", {
        /**
         * Gets or sets the minimum size (in pixels) for this scale.
        */
        get: function () {
            return this.i.minimumValue;
        },
        set: function (v) {
            this.i.minimumValue = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSizeScaleComponent.prototype, "maximumValue", {
        /**
         * Gets or sets the maximum size (in pixels) for this scale.
        */
        get: function () {
            return this.i.maximumValue;
        },
        set: function (v) {
            this.i.maximumValue = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSizeScaleComponent.prototype, "isLogarithmic", {
        /**
         * Gets or sets whether the scale is logarithmic.
        */
        get: function () {
            return this.i.isLogarithmic;
        },
        set: function (v) {
            this.i.isLogarithmic = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSizeScaleComponent.prototype, "logarithmBase", {
        /**
         * Gets or sets the logarithm base for this scale.
        */
        get: function () {
            return this.i.logarithmBase;
        },
        set: function (v) {
            this.i.logarithmBase = +v;
        },
        enumerable: false,
        configurable: true
    });
    IgxSizeScaleComponent.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    Object.defineProperty(IgxSizeScaleComponent.prototype, "hasUserValues", {
        get: function () {
            return this._hasUserValues;
        },
        enumerable: false,
        configurable: true
    });
    IgxSizeScaleComponent.prototype.__m = function (propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    };
    IgxSizeScaleComponent.prototype._styling = function (container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        var genericPrefix = "";
        var typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("SizeScaleComponent");
        var additionalPrefixes = [];
        var prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        var b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            var basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            var parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            var parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    };
    IgxSizeScaleComponent.prototype.getCurrentGlobalMinimum = function () {
        var iv = this.i.getCurrentGlobalMinimum();
        return (iv);
    };
    IgxSizeScaleComponent.prototype.getCurrentGlobalMaximum = function () {
        var iv = this.i.getCurrentGlobalMaximum();
        return (iv);
    };
    Object.defineProperty(IgxSizeScaleComponent.prototype, "propertyUpdated", {
        /**
         * Event raised when a property (including "effective" and non-dependency property) value changes.
        */
        get: function () {
            var _this = this;
            if (this._propertyUpdated == null) {
                this._propertyUpdated = new EventEmitter();
                this.i.propertyUpdated = delegateCombine(this.i.propertyUpdated, function (o, e) {
                    _this._runInZone(function () {
                        var outerArgs = new IgxPropertyUpdatedEventArgs();
                        outerArgs._provideImplementation(e);
                        if (_this.beforePropertyUpdated) {
                            _this.beforePropertyUpdated(_this, outerArgs);
                        }
                        _this._propertyUpdated.emit({
                            sender: _this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._propertyUpdated;
        },
        enumerable: false,
        configurable: true
    });
    IgxSizeScaleComponent.prototype._runInZone = function (act) {
        if (this._zoneRunner != null) {
            this._zoneRunner(act);
        }
        else {
            act();
        }
    };
    IgxSizeScaleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxSizeScaleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxSizeScaleComponent, selector: "igx-size-scale", inputs: { globalMinimum: "globalMinimum", globalMaximum: "globalMaximum", minimumValue: "minimumValue", maximumValue: "maximumValue", isLogarithmic: "isLogarithmic", logarithmBase: "logarithmBase" }, outputs: { propertyUpdated: "propertyUpdated" }, providers: [], ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxSizeScaleComponent;
}());
export { IgxSizeScaleComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-size-scale',
                    template: "",
                    providers: [],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { globalMinimum: [{
                type: Input
            }], globalMaximum: [{
                type: Input
            }], minimumValue: [{
                type: Input
            }], maximumValue: [{
                type: Input
            }], isLogarithmic: [{
                type: Input
            }], logarithmBase: [{
                type: Input
            }], propertyUpdated: [{
                type: Output
            }] } });
