// igniteui-angular-charts/esm5/lib/igx-data-legend-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataLegendComponent } from './igx-data-legend-component';
import { XamDataLegend } from './XamDataLegend';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDataLegendModule } from './igx-data-legend-module';
import * as i0 from "@angular/core";
var IgxDataLegendDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataLegendDynamicModule() {
        TypeRegistrar.registerCons('XamDataLegend', XamDataLegend);
        TypeRegistrar.registerCons('IgxDataLegendComponent', IgxDataLegendComponent);
        TypeRegistrar.register('XamDataLegend', XamDataLegend.$type);
    }
    IgxDataLegendDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataLegendDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataLegendDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataLegendDynamicModule, imports: [CommonModule, IgxDataLegendModule], exports: [IgxDataLegendModule] });
    IgxDataLegendDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataLegendDynamicModule, imports: [[CommonModule, IgxDataLegendModule
            ], IgxDataLegendModule] });
    return IgxDataLegendDynamicModule;
}());
export { IgxDataLegendDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataLegendDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataLegendModule
                    ],
                    imports: [CommonModule, IgxDataLegendModule
                    ],
                    entryComponents: [IgxDataLegendComponent]
                }]
        }], ctorParameters: function () { return []; } });
