// igniteui-angular-charts/esm5/lib/igx-stacked-100-spline-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStacked100SplineSeriesModule } from './igx-stacked-100-spline-series-module';
import { IgxStacked100SplineSeriesComponent } from './igx-stacked-100-spline-series-component';
import { Stacked100SplineSeries } from './Stacked100SplineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxStacked100SplineSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStacked100SplineSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxStacked100SplineSeriesComponent", IgxStacked100SplineSeriesComponent);
        TypeRegistrar.register("Stacked100SplineSeries", Stacked100SplineSeries.$type);
    }
    IgxStacked100SplineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStacked100SplineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesDynamicModule, imports: [CommonModule, IgxStacked100SplineSeriesModule], exports: [IgxStacked100SplineSeriesModule] });
    IgxStacked100SplineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesDynamicModule, imports: [[CommonModule, IgxStacked100SplineSeriesModule], IgxStacked100SplineSeriesModule] });
    return IgxStacked100SplineSeriesDynamicModule;
}());
export { IgxStacked100SplineSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStacked100SplineSeriesModule],
                    exports: [IgxStacked100SplineSeriesModule],
                    entryComponents: [IgxStacked100SplineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
