// igniteui-angular-charts/esm5/lib/igx-typical-price-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxTypicalPriceIndicatorComponent } from './igx-typical-price-indicator-component';
import * as i0 from "@angular/core";
var IgxTypicalPriceIndicatorModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxTypicalPriceIndicatorModule() {
    }
    IgxTypicalPriceIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTypicalPriceIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorModule, declarations: [IgxTypicalPriceIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxTypicalPriceIndicatorComponent, IgxIndicatorProxyModule] });
    IgxTypicalPriceIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxTypicalPriceIndicatorModule;
}());
export { IgxTypicalPriceIndicatorModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxTypicalPriceIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxTypicalPriceIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
