// igniteui-angular-charts/esm5/lib/igx-radial-line-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { UnknownValuePlotting_$type } from "igniteui-angular-core";
import { IgxAnchoredRadialSeriesComponent } from "./igx-anchored-radial-series-component";
import { IgxRadialBaseComponent } from "./igx-radial-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { RadialLineSeries } from "./RadialLineSeries";
import { ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent radial line series.
*/
var IgxRadialLineSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxRadialLineSeriesComponent, _super);
    function IgxRadialLineSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxRadialLineSeriesComponent.prototype.createImplementation = function () {
        return new RadialLineSeries();
    };
    Object.defineProperty(IgxRadialLineSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialLineSeriesComponent.prototype, "isLineOnly", {
        /**
         * Gets whether the current series shows only line shapes.
        */
        get: function () {
            return this.i.fi;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialLineSeriesComponent.prototype, "unknownValuePlotting", {
        /**
         * Determines how unknown values will be plotted on the chart.
         * Null and Double.NaN are two examples of unknown values.
        */
        get: function () {
            return this.i.ade;
        },
        set: function (v) {
            this.i.ade = ensureEnum(UnknownValuePlotting_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    /**
     * When overridden in a derived class, is invoked whenever application code or internal processes
     * call ApplyTemplate.
    
    */
    IgxRadialLineSeriesComponent.prototype.onApplyTemplate = function () {
        this.i.ac();
    };
    IgxRadialLineSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialLineSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxRadialLineSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxRadialLineSeriesComponent, selector: "igx-radial-line-series", inputs: { unknownValuePlotting: "unknownValuePlotting" }, providers: [{ provide: IgxAnchoredRadialSeriesComponent, useExisting: forwardRef(function () { return IgxRadialLineSeriesComponent; }) }, { provide: IgxRadialBaseComponent, useExisting: forwardRef(function () { return IgxRadialLineSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxRadialLineSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxRadialLineSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxRadialLineSeriesComponent;
}(IgxAnchoredRadialSeriesComponent));
export { IgxRadialLineSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialLineSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-radial-line-series',
                    template: "",
                    providers: [{ provide: IgxAnchoredRadialSeriesComponent, useExisting: forwardRef(function () { return IgxRadialLineSeriesComponent; }) }, { provide: IgxRadialBaseComponent, useExisting: forwardRef(function () { return IgxRadialLineSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxRadialLineSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxRadialLineSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { unknownValuePlotting: [{
                type: Input
            }] } });
