// igniteui-angular-charts/esm5/lib/igx-series-component.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { EventEmitter, Output, Component, Input } from '@angular/core';
import { Visibility_$type } from "igniteui-angular-core";
import { TransitionInSpeedType_$type } from './TransitionInSpeedType';
import { SeriesHitTestMode_$type } from './SeriesHitTestMode';
import { fromDoubleCollection, toDoubleCollection, brushToString, stringToBrush, toPoint, fromPoint, fromRect, stringToColor, colorToString, ensureEnum, ensureBool, toSpinal, initializePropertiesFromCss, NamePatcher, arrayFindByName } from "igniteui-angular-core";
import { IgxRenderRequestedEventArgs } from './igx-render-requested-event-args';
import { delegateCombine, TypeRegistrar } from "igniteui-angular-core";
import { SeriesVisibleRangeMode_$type } from './SeriesVisibleRangeMode';
import { SeriesHighlightingMode_$type } from './SeriesHighlightingMode';
import { SeriesOutlineMode_$type } from './SeriesOutlineMode';
import { LegendItemBadgeMode_$type } from "igniteui-angular-core";
import { LegendItemBadgeShape_$type } from "igniteui-angular-core";
import { PenLineJoin_$type } from "igniteui-angular-core";
import { PenLineCap_$type } from "igniteui-angular-core";
import { IgxSeriesLayerCollection } from './igx-series-layer-collection';
import { SeriesLayerCollection as SeriesLayerCollection_internal } from './SeriesLayerCollection';
import { SeriesLayer } from './SeriesLayer';
import { SyncableObservableCollection$1 } from "igniteui-angular-core";
import { TransitionOutSpeedType_$type } from './TransitionOutSpeedType';
import { SeriesHighlightedValuesDisplayMode_$type } from './SeriesHighlightedValuesDisplayMode';
//import { DataLegendSeriesContext, DataLegendSeriesContext_$type } from 'igniteui-core/DataLegendSeriesContext';
//import { DataLegendSeriesInfo, DataLegendSeriesInfo_$type } from 'igniteui-core/DataLegendSeriesInfo';
import { IgxSeriesLayerPropertyOverlayCollection } from './igx-series-layer-property-overlay-collection';
import { SeriesLayerPropertyOverlayCollection as SeriesLayerPropertyOverlayCollection_internal } from './SeriesLayerPropertyOverlayCollection';
import { SeriesLayerPropertyOverlay } from './SeriesLayerPropertyOverlay';
import { SeriesSelectionMode_$type } from './SeriesSelectionMode';
import { IgxTransitionOutCompletedEventArgs } from './igx-transition-out-completed-event-args';
import * as i0 from "@angular/core";
export var IgxSeriesComponent_PROVIDERS = [];
/**
 * Represents the base class for all IgxDataChartComponent series.
*/
var IgxSeriesComponent = /** @class */ /*@__PURE__*/ (function () {
    function IgxSeriesComponent() {
        //@Input()
        //set opacity(value: number) {
        //    this.i.opacity = +value;
        //}
        //get opacity(): number {
        //    return this.i.opacity;
        //}
        //@Input()
        //set visibility(v: Visibility) {
        //    this.i.visibility = ensureEnum<Visibility>(Visibility_$type, v);
        //}
        //get visibility(): Visibility {
        //    return this.i.visibility;
        //}
        this._chartLevelData = null;
        this._dataSource = null;
        this._chartLevelHighlightedData = null;
        this._highlightedDataSource = null;
        this._tooltipTemplate = null;
        this._tooltipContent = null;
        this._tooltipContainerTemplate = null;
        this._renderer = null;
        this._highlightedValuesExtraPropertyOverlays = null;
        this._layers = null;
        this._actualLayers = null;
        this.__p = null;
        this._hasUserValues = new Set();
        this._stylingContainer = null;
        this._stylingParent = null;
        this._inStyling = false;
        this._transitionOutCompleted = null;
        this._renderRequested = null;
        this._zoneRunner = null;
        if (this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
        }
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    Object.defineProperty(IgxSeriesComponent.prototype, "i", {
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesComponent.prototype.onImplementationCreated = function () {
    };
    IgxSeriesComponent.prototype.createImplementation = function () {
        return null;
    };
    Object.defineProperty(IgxSeriesComponent.prototype, "seriesInternal", {
        get: function () {
            return this.i;
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesComponent.prototype.ngOnInit = function () {
    };
    Object.defineProperty(IgxSeriesComponent.prototype, "name", {
        get: function () {
            return this.i.name;
        },
        set: function (value) {
            this.i.name = value;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "legend", {
        /**
         * Gets or sets the legend used for the current series.
        */
        get: function () {
            if (this.i.legend != null)
                return this.i.legend.externalObject;
        },
        set: function (v) {
            if (v != undefined && v != null)
                this.i.legend = v.i;
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesComponent.prototype.provideData = function (data) {
        this._chartLevelData = data;
        this.updateDataSource();
    };
    IgxSeriesComponent.prototype.onUpdateDataSource = function (dataSource) {
        return dataSource;
    };
    IgxSeriesComponent.prototype.updateDataSource = function () {
        if (this._dataSource == null) {
            this.i.itemsSource = this.onUpdateDataSource(this._chartLevelData);
        }
        else {
            this.i.itemsSource = this.onUpdateDataSource(this._dataSource);
        }
    };
    Object.defineProperty(IgxSeriesComponent.prototype, "dataSource", {
        get: function () {
            if (this._dataSource != null) {
                return this._dataSource;
            }
            return this.i.itemsSource;
        },
        set: function (value) {
            this._dataSource = value;
            this.updateDataSource();
            //console.log("setting axis data source: " + value)
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesComponent.prototype.provideHighlightedData = function (data) {
        this._chartLevelHighlightedData = data;
        this.updatehighlightedDataSource();
    };
    IgxSeriesComponent.prototype.onUpdatehighlightedDataSource = function (highlightedDataSource) {
        return highlightedDataSource;
    };
    IgxSeriesComponent.prototype.updatehighlightedDataSource = function () {
        if (this._highlightedDataSource == null) {
            this.i.highlightedItemsSource = this.onUpdatehighlightedDataSource(this._chartLevelHighlightedData);
        }
        else {
            this.i.highlightedItemsSource = this.onUpdatehighlightedDataSource(this._highlightedDataSource);
        }
    };
    Object.defineProperty(IgxSeriesComponent.prototype, "highlightedDataSource", {
        get: function () {
            if (this._highlightedDataSource != null) {
                return this._highlightedDataSource;
            }
            return this.i.highlightedItemsSource;
        },
        set: function (value) {
            this._highlightedDataSource = value;
            this.updatehighlightedDataSource();
            //console.log("setting axis data source: " + value)
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesComponent.prototype.bindAxes = function (axes) {
    };
    Object.defineProperty(IgxSeriesComponent.prototype, "tooltipTemplate", {
        get: function () {
            return this._tooltipTemplate;
        },
        set: function (value) {
            this._tooltipTemplate = value;
            if (value == null) {
                if (this._tooltipContent !== null) {
                    this._tooltipContent.destroy();
                    this._tooltipContent = null;
                }
            }
            if (this._tooltipContent != null) {
                this._tooltipContent.instance.template = this._tooltipTemplate;
            }
            else {
                if (this.owner != null) {
                    this.owner._ensureTooltipCreated(this);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "tooltipContainerTemplate", {
        get: function () {
            return this._tooltipContainerTemplate;
        },
        set: function (value) {
            this._tooltipContainerTemplate = value;
            if (this._tooltipContent != null) {
                this._tooltipContent.instance.containerTemplate = this._tooltipContainerTemplate;
            }
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesComponent.prototype._ensureTooltipCreated = function (createTooltip, createWrapper) {
        if (this._tooltipTemplate == null) {
            this.i.toolTip = null;
        }
        if (this._tooltipContent == null && this._tooltipTemplate != null) {
            var tooltip = createTooltip();
            if (tooltip == null) {
                return;
            }
            this._tooltipContent = tooltip;
            tooltip.instance.template = this._tooltipTemplate;
            this.i.toolTip = createWrapper(tooltip.location.nativeElement);
        }
    };
    IgxSeriesComponent.prototype._ensureTooltipDestroyed = function () {
        if (this._tooltipContent !== null) {
            this._tooltipContent.destroy();
            this._tooltipContent = null;
        }
    };
    IgxSeriesComponent.prototype._provideRenderer = function (renderer) {
        this._renderer = renderer;
        this._zoneRunner = function (act) {
            if (renderer !== null) {
                renderer._ngZone.run(act);
            }
            else {
                act();
            }
        };
    };
    IgxSeriesComponent._createFromInternal = function (internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        var name = internal.$type.name;
        var externalName = "Igx" + name + "Component";
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    };
    Object.defineProperty(IgxSeriesComponent.prototype, "dataLegendKey", {
        /**
     * Gets unique key used to identify the series in data legend
    */
        get: function () {
            return this.i.dataLegendKey;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "dataLegendGroup", {
        /**
         * Gets or sets a name used for grouping multiple series in the Data Legend
         * If not set, series will be grouped by series family, e.g. Category, Range, Financial, Scatter
        */
        get: function () {
            return this.i.l1;
        },
        set: function (v) {
            this.i.l1 = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "highlightedValuesDataLegendGroup", {
        /**
         * Gets or sets a name used for grouping highlighted series in the Data Legend
         * If not set, series will default to parent series' DataLegendGroup, e.g. Category, Range, Financial, Scatter
        */
        get: function () {
            return this.i.mh;
        },
        set: function (v) {
            this.i.mh = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isAnnotationLayer", {
        /**
         * Gets whether the series is an annotation layer.
        *
        * ```ts
        * let isAnnotationLayer: boolean = this.series.isAnnotationLayer;
        * ```
        */
        get: function () {
            return this.i.eq;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isAnnotationCrosshairLayer", {
        /**
         * Gets whether the series is an crosshair annotation layer.
        */
        get: function () {
            return this.i.en;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isAnnotationCalloutLayer", {
        /**
         * Gets whether the series is an callout annotation layer.
        */
        get: function () {
            return this.i.em;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isAnnotationHoverLayer", {
        /**
         * Gets whether the series is an annotation layer displayed only when hovering over the chart.
         * Crosshair Layer, Category Highlight Layer, Category Tooltip Layer, Item Tooltip Layer
        */
        get: function () {
            return this.i.ep;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isAnnotationFinalValue", {
        /**
         * Gets whether the series is final value annotation layer.
        */
        get: function () {
            return this.i.eo;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isAnnotationValueLayer", {
        /**
         * Gets whether the series is a value annotation layer.
        */
        get: function () {
            return this.i.er;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isVertical", {
        /**
         * Gets whether the current series is oriented vertically.
        *
        * ```ts
        * let isVertical: boolean = this.series.isVertical;
        * ```
        */
        get: function () {
            return this.i.isVertical;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isFragment", {
        /**
         * Gets whether the current series is a stack fragment.
        *
        * ```ts
        * let isFragment: boolean = this.series.isFragment;
        * ```
        */
        get: function () {
            return this.i.e9;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isAreaOrLine", {
        /**
         * Gets whether the current series shows an area or line shape.
        *
        * ```ts
        * let isAreaorLine: boolean = this.series.isAreaorLine;
        * ```
        */
        get: function () {
            return this.i.et;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isArea", {
        /**
         * Gets whether the current series shows an area shape.
        */
        get: function () {
            return this.i.es;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "hasValueAxis", {
        /**
         * Gets whether the current series shows an area shape.
        */
        get: function () {
            return this.i.ei;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isValueAxisInverted", {
        /**
         * Gets whether the current series shows an area shape.
        */
        get: function () {
            return this.i.f3;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isLineOnly", {
        /**
         * Gets whether the current series shows only line shapes.
        */
        get: function () {
            return this.i.fi;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isLineContour", {
        /**
         * Gets whether the current series shows a line contour shape.
        */
        get: function () {
            return this.i.fh;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isSpline", {
        /**
         * Gets whether the current series shows a spline shape.
        */
        get: function () {
            return this.i.fy;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isColumn", {
        /**
         * Gets whether the current series shows a column shape.
        */
        get: function () {
            return this.i.ex;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isBar", {
        /**
         * Gets whether the current series shows a horizontal column shape.
        */
        get: function () {
            return this.i.ev;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isWaterfall", {
        /**
         * Gets whether the current series shows a waterfall column shape.
        */
        get: function () {
            return this.i.f4;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isPolyline", {
        /**
         * Gets whether the current series shows a polyline shape.
        */
        get: function () {
            return this.i.fp;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isPolygon", {
        /**
         * Gets whether the current series shows a polygon shape.
        */
        get: function () {
            return this.i.fo;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isPixel", {
        /**
         * Gets whether the current series shows pixels.
        */
        get: function () {
            return this.i.fm;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isPie", {
        /**
         * Gets whether the current series shows a pie shape.
        */
        get: function () {
            return this.i.fl;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isTile", {
        /**
         * Gets whether the current series shows a tile imagery.
        */
        get: function () {
            return this.i.f1;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isStep", {
        /**
         * Gets whether the current series shows step shapes.
        */
        get: function () {
            return this.i.fz;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isShapeControl", {
        /**
         * Gets whether the current series shows interactive shape controls.
        */
        get: function () {
            return this.i.fw;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isPolar", {
        /**
         * Gets whether the current series is a polar type series.
        *
        * ```ts
        * let isPolar: boolean = this.series.isPolar;
        * ```
        */
        get: function () {
            return this.i.fn;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isScatter", {
        /**
         * Gets whether the current series is a scatter type series.
        *
        * ```ts
        * let isScatter: boolean = this.series.isScatter;
        * ```
        */
        get: function () {
            return this.i.fu;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isRadial", {
        /**
         * Gets whether the current series is a radial type series.
        *
        * ```ts
        * let isRadial: boolean = this.series.isRadial;
        * ```
        */
        get: function () {
            return this.i.fr;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isStacked", {
        /**
         * Gets whether the current series is a stacked type series.
        *
        * ```ts
        * let isStacked: boolean = this.series.isStacked;
        * ```
        */
        get: function () {
            return this.i.isStacked;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isFinancial", {
        /**
         * Gets whether the current series is a financial type series.
        *
        * ```ts
        * let isFinancial: boolean = this.series.isFinancial;
        * ```
        */
        get: function () {
            return this.i.e4;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isRange", {
        /**
         * Gets whether the current series is a range type series.
        *
        * ```ts
        * let isRange: boolean = this.series.isRange;
        * ```
        */
        get: function () {
            return this.i.fs;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isCategory", {
        /**
         * Gets whether the current series is a category type series.
        *
        * ```ts
        * let isCategory: boolean = this.series.isCategory;
        * ```
        */
        get: function () {
            return this.i.ew;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isGeographic", {
        /**
         * Gets whether the current series is a geographic series.
        *
        * ```ts
        * let isGeographic: boolean = this.series.isGeographic;
        * ```
        */
        get: function () {
            return this.i.fa;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isShape", {
        /**
         * Gets whether the current series is a shape type series.
        */
        get: function () {
            return this.i.fv;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isValueOverlay", {
        /**
         * Gets whether the series is a value overlay.
        */
        get: function () {
            return this.i.isValueOverlay;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isHighlightOverlay", {
        /**
         * Gets whether the series is a highlight overlay.
        */
        get: function () {
            return this.i.fd;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isSummarizationSupported", {
        /**
         * Gets whether the series supports summarization in Data Legend and Data Tooltip.
        */
        get: function () {
            return this.i.isSummarizationSupported;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isLayer", {
        get: function () {
            return this.i.isLayer;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "hasOnlyMarkers", {
        /**
         * Gets whether the series has only marker as visuals
        */
        get: function () {
            return this.i.eg;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isFinancialSeries", {
        /**
         * Gets whether the series has financial OHLC/candlestick visuals
        */
        get: function () {
            return this.i.e7;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isFinancialOverlay", {
        /**
         * Gets whether the series is financial overlay
        */
        get: function () {
            return this.i.e6;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isFinancialIndicator", {
        /**
         * Gets whether the series is financial indicator
        */
        get: function () {
            return this.i.e5;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isFinancialWaterfall", {
        /**
         * Gets whether the series is financial waterfall
        */
        get: function () {
            return this.i.e8;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isDefaultCrosshairBehaviorDisabled", {
        /**
         * Gets whether the default crosshair behavior should be disabled if this series is present.
        *
        * `IsDefaultCrosshairBehaviorDisabled` defaults to true for annotation layers. When not using annotaion layers, you can show crosshairs by setting the chart's `CrosshairVisibility` to true.
        *
        * ```ts
        * let isDefaultCrosshairBehaviorDisabled: boolean = this.series.isDefaultCrosshairBehaviorDisabled;
        * ```
        */
        get: function () {
            return this.i.ez;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isDefaultTooltipBehaviorDisabled", {
        /**
         * Gets whether the default tooltip behavior should be disabled if this series is present.
        *
        * ```ts
        *  let isDefaultTooltipBehaviorDisabled: boolean = this.series.isDefaultTooltipBehaviorDisabled;
        * ```
        */
        get: function () {
            return this.i.e0;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "hasMarkers", {
        /**
         * Returns whether the current series supports visual markers.
        *
        * ```ts
        * let hasMarkers: boolean = this.series.hasMarkers;
        * ```
        */
        get: function () {
            return this.i.hasMarkers;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "hasVisibleMarkers", {
        /**
         * Returns whether the current series has a visible marker style.
        */
        get: function () {
            return this.i.ej;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "highlightedValuesExtraPropertyOverlays", {
        /**
         * Gets the current extra layers for the series.
        */
        get: function () {
            if (this._highlightedValuesExtraPropertyOverlays === null) {
                var coll = new IgxSeriesLayerPropertyOverlayCollection();
                var innerColl = this.i.cc;
                if (!innerColl) {
                    innerColl = new SeriesLayerPropertyOverlayCollection_internal();
                }
                this._highlightedValuesExtraPropertyOverlays = coll._fromInner(innerColl);
            }
            return this._highlightedValuesExtraPropertyOverlays;
        },
        set: function (v) {
            if (this._highlightedValuesExtraPropertyOverlays !== null) {
                this._highlightedValuesExtraPropertyOverlays._setSyncTarget(null);
                this._highlightedValuesExtraPropertyOverlays = null;
            }
            var coll = new IgxSeriesLayerPropertyOverlayCollection();
            this._highlightedValuesExtraPropertyOverlays = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(SeriesLayerPropertyOverlay.$type);
            var innerColl = this.i.cc;
            if (!innerColl) {
                innerColl = new SeriesLayerPropertyOverlayCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._highlightedValuesExtraPropertyOverlays._setSyncTarget(syncColl);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isIndexed", {
        /**
         * Gets if the series uses an index, which can be used to auto-assign it a color.
        *
        * ```ts
        * let isIndexed: boolean = this.series.isIndexed;
        * ```
        */
        get: function () {
            return this.i.fe;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isUsableInLegend", {
        /**
         * Gets if the series should appear in any legends.
        *
        * ```ts
        * let isUsableInLegend: boolean = this.series.isUsableInLegend;
        * ```
        */
        get: function () {
            return this.i.isUsableInLegend;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "legendItemVisibility", {
        /**
         * Gets or sets whether or not the current series will have a legend item displayed in a legend.
        *
        * The `LegendItemVisibilty` can be used to exclude only certain `Series` from the `Legend`.
        */
        get: function () {
            return this.i.legendItemVisibility;
        },
        set: function (v) {
            this.i.legendItemVisibility = ensureEnum(Visibility_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "legendItemBadgeTemplate", {
        /**
         * Gets or sets the LegendItemBadgeTemplate property.
         * The legend item badge is created according to the LegendItemBadgeTemplate on-demand by
         * the series object itself.
        */
        get: function () {
            return this.i.legendItemBadgeTemplate;
        },
        set: function (v) {
            this.i.legendItemBadgeTemplate = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualLegendItemBadgeTemplate", {
        /**
         * Gets the actual legend item badge template used by the series.
        */
        get: function () {
            return this.i.sq;
        },
        set: function (v) {
            this.i.sq = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualLegendItemBadgeOutline", {
        /**
         * Gets the effective legend badge outline for the current series.
        */
        get: function () {
            return brushToString(this.i.u6);
        },
        set: function (v) {
            this.i.u6 = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualLegendItemBadgeBrush", {
        /**
         * Gets the effective legend badge fill for the current series.
        */
        get: function () {
            return brushToString(this.i.u5);
        },
        set: function (v) {
            this.i.u5 = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "legendItemBadgeShape", {
        /**
         * Gets or sets the type of legend badge representing the current series in a legend.
         * This property will be ignored when the LegendItemTemplate/LegendItemBadgeTemplate property is set on the series
        */
        get: function () {
            return this.i.legendItemBadgeShape;
        },
        set: function (v) {
            this.i.legendItemBadgeShape = ensureEnum(LegendItemBadgeShape_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "legendItemBadgeMode", {
        /**
         * Gets or sets the mode of legend badge representing the current series in a legend.
         * This property will be ignored when the LegendItemTemplate/LegendItemBadgeTemplate property is set on the series
        */
        get: function () {
            return this.i.bc;
        },
        set: function (v) {
            this.i.bc = ensureEnum(LegendItemBadgeMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "legendItemTemplate", {
        /**
         * Gets or sets the LegendItemTemplate property.
         * The legend item control content is created according to the LegendItemTemplate on-demand by
         * the series object itself.
        */
        get: function () {
            return this.i.legendItemTemplate;
        },
        set: function (v) {
            this.i.legendItemTemplate = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "discreteLegendItemTemplate", {
        /**
         * Gets or sets the DiscreteLegendItemTemplate property.
         * The legend item control content is created according to the DiscreteLegendItemTemplate on-demand by
         * the series object itself.
        */
        get: function () {
            return this.i.sr;
        },
        set: function (v) {
            this.i.sr = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "transitionEasingFunction", {
        /**
         * Gets or sets the easing function used to morph the current series.
        *
        * The `TransitioninDuration` and `TransitionEasingFunction` can be used to play animation when data is added or removed from a `Series`. To play an initial animation see `TransitionInDuration`.
        *
        * ```ts
        * this.series.transitionEasingFunction =  EasingFunctions.cubicEase;
        * ```
        */
        get: function () {
            return this.i.db;
        },
        set: function (v) {
            this.i.db = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "transitionInEasingFunction", {
        /**
         * Gets or sets the EasingFunction used to morph the current series during the initial transition.
        *
        * The `TransitionInEasingFunction` can be used with `IsTransitionInEnabled` to cofigure the animation when a new datasource is loaded.
        *
        * ```ts
        * this.series.transitionInEasingFunction =  EasingFunctions.cubicEase;
        * ```
        */
        get: function () {
            return this.i.dc;
        },
        set: function (v) {
            this.i.dc = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "transitionOutEasingFunction", {
        /**
         * Gets or sets the EasingFunction used to morph the current series during the initial transition.
        */
        get: function () {
            return this.i.dd;
        },
        set: function (v) {
            this.i.dd = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "transitionDuration", {
        /**
         * Gets or sets the duration of the current series' morph.
        *
        * The `TransitionDuration` can be used to play animation when data is added or removed from a `Series`. To play an initial animation see `TransitionInDuration`.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     animateSeriesWhenAxisRangeChanges=true>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         transitionInMode="accordionFromBottom"
        *         isTransitionInEnabled=true
        *         transitionInSpeedType="indexScaled"
        *         transitionDuration=500
        *         transitionInDuration=500>
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.transitionDuration = 500;
        * ```
        */
        get: function () {
            return this.i.kb;
        },
        set: function (v) {
            this.i.kb = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualResolution", {
        get: function () {
            return this.i.ib;
        },
        set: function (v) {
            this.i.ib = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "resolution", {
        /**
         * Gets or sets the current series object's rendering resolution.
        *
        * Setting the `Resolution` on a Series to a higher value can help with performance, but it will lower the graphical fidelity of the line. As such, it can be increased up until the fidelity is unacceptable.
        *
        * ```html
        *  <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         resolution=1.5>
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.resolution = 1.5;
        * ```
        */
        get: function () {
            return this.i.jh;
        },
        set: function (v) {
            this.i.jh = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "visibleRangeMarginTop", {
        /**
         * Gets or sets the top margin to use when getting a visible axis range for the series.
        */
        get: function () {
            return this.i.js;
        },
        set: function (v) {
            this.i.js = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "visibleRangeMarginBottom", {
        /**
         * Gets or sets the bottom margin to use when getting a visible axis range for the series.
        */
        get: function () {
            return this.i.jp;
        },
        set: function (v) {
            this.i.jp = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "visibleRangeMarginLeft", {
        /**
         * Gets or sets the left margin to use when getting a visible axis range for the series.
        */
        get: function () {
            return this.i.jq;
        },
        set: function (v) {
            this.i.jq = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "visibleRangeMarginRight", {
        /**
         * Gets or sets the right margin to use when getting a visible axis range for the series.
        */
        get: function () {
            return this.i.jr;
        },
        set: function (v) {
            this.i.jr = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "title", {
        /**
         * Gets or sets the Title property.
         * The legend item control is created according to the Title on-demand by
         * the series object itself.
        *
        * The `Series` `Title` may be used in tooltips and legends.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         showDefaultTooltip="true"
        *         title="InStock Items">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.title = "InStock Items";
        * ```
        */
        get: function () {
            return this.i.title;
        },
        set: function (v) {
            this.i.title = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "highlightedTitleSuffix", {
        /**
         * Gets or sets the HighlightedTitleSuffix property.
        */
        get: function () {
            return this.i.mf;
        },
        set: function (v) {
            this.i.mf = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "highlightedLegendItemVisibility", {
        /**
         * Gets or sets whether the highlighted values layer should have a legend item.
        */
        get: function () {
            return this.i.wp;
        },
        set: function (v) {
            this.i.wp = ensureEnum(Visibility_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "brush", {
        /**
         * Gets or sets the brush to use for the series.
        *
        * The `Brush` along with the `Outline` and `Negative Brush` can be used to affect the visuals of the `Series`.
        *
        * ```html
        * <igx-data-chart #chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-area-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         brush="red">
        *     </igx-area-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.brush = "red";
        * ```
        */
        get: function () {
            return brushToString(this.i.u9);
        },
        set: function (v) {
            this.i.u9 = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualBrush", {
        /**
         * Gets the effective brush for the current series object.
        *
        * ```ts
        * let actualBrush: string = series.actualBrush;
        * ```
        */
        get: function () {
            return brushToString(this.i.u3);
        },
        set: function (v) {
            this.i.u3 = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "selectionBrush", {
        /**
         * Gets or sets the selection brush to use for the series.
        */
        get: function () {
            return brushToString(this.i.vt);
        },
        set: function (v) {
            this.i.vt = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "focusBrush", {
        /**
         * Gets or sets the Focus brush to use for the series.
        */
        get: function () {
            return brushToString(this.i.ve);
        },
        set: function (v) {
            this.i.ve = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualSelectionBrush", {
        /**
         * Gets the effective emphasis brush for the current series object.
        */
        get: function () {
            return brushToString(this.i.u8);
        },
        set: function (v) {
            this.i.u8 = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualFocusBrush", {
        /**
         * Gets the effective emphasis brush for the current series object.
        */
        get: function () {
            return brushToString(this.i.u4);
        },
        set: function (v) {
            this.i.u4 = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "safeActualBrush", {
        /**
         * Gets the effective brush for the current series object with opacity removed so its contrasty for use as a font color.
        */
        get: function () {
            return brushToString(this.i.safeActualBrush);
        },
        set: function (v) {
            this.i.safeActualBrush = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "outline", {
        /**
         * Gets or sets the brush to use for the outline of the series.
         * Some series types, such as LineSeries, do not display outlines.  Therefore, this property does not affect some charts.
        *
        * The `Outline` along with the `Brush` and `NegativeBrush` can be used to affect the visuals of the `Series`.
        *
        * ```html
        * <igx-data-chart #chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-area-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         outline="red">
        *     </igx-area-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.brush = "red";
        * ```
        */
        get: function () {
            return brushToString(this.i.vp);
        },
        set: function (v) {
            this.i.vp = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualOutline", {
        /**
         * Gets the effective outline for the current series object.
        *
        * ```ts
        * let actualOutline: string = series.ActualOutline;
        * ```
        */
        get: function () {
            return brushToString(this.i.u7);
        },
        set: function (v) {
            this.i.u7 = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "lineJoin", {
        /**
         * Gets or sets the brush that specifies current series object's line join style.
        */
        get: function () {
            return this.i.vz;
        },
        set: function (v) {
            this.i.vz = ensureEnum(PenLineJoin_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "highlightedValuesDisplayMode", {
        /**
         * Gets or sets whether and how to display highlighted values for the series. Note, this is distinct from the highlighting feature that indicates what is closest or under the mouse.
        */
        get: function () {
            return this.i.bt;
        },
        set: function (v) {
            this.i.bt = ensureEnum(SeriesHighlightedValuesDisplayMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "shouldAnimateOnDataSourceSwap", {
        /**
         * Gets or sets whether and how to display highlighted values for the series. Note, this is distinct from the highlighting feature that indicates what is closest or under the mouse.
        */
        get: function () {
            return this.i.ge;
        },
        set: function (v) {
            this.i.ge = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "thickness", {
        /**
         * Gets or sets the width of the current series object's line thickness.
        *
        * Depending on the `Series` type, this can be the main brush used, or just the outline. For example, when using a `LineSeries` it will affect the thickness of the lines drawn, whereas when using a `ColumnSeries` it will affect the outer border thickness of the columns.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         thickness=5>
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.thickness=5;
        * ```
        */
        get: function () {
            return this.i.jn;
        },
        set: function (v) {
            this.i.jn = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualThickness", {
        /**
         * Gets the effective Thickness for the current series object.
        */
        get: function () {
            return this.i.ic;
        },
        set: function (v) {
            this.i.ic = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "dashArray", {
        /**
         * Gets or sets a collection of Double values that indicate the pattern of dashes and gaps that
         * is used to outline the current series object.
        */
        get: function () {
            return fromDoubleCollection(this.i.vw);
        },
        set: function (v) {
            this.i.vw = toDoubleCollection(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualHighlightingMode", {
        /**
         * Gets actual highlighting mode
        */
        get: function () {
            return this.i.bv;
        },
        set: function (v) {
            this.i.bv = ensureEnum(SeriesHighlightingMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualSelectionMode", {
        /**
         * Gets actual Selection mode
        */
        get: function () {
            return this.i.cn;
        },
        set: function (v) {
            this.i.cn = ensureEnum(SeriesSelectionMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualFocusMode", {
        /**
         * Gets actual Selection mode
        */
        get: function () {
            return this.i.cm;
        },
        set: function (v) {
            this.i.cm = ensureEnum(SeriesSelectionMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualHighlightedValuesFadeOpacity", {
        /**
         * Gets actual HighlightedValues fade opacity
        */
        get: function () {
            return this.i.h8;
        },
        set: function (v) {
            this.i.h8 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "highlightedValuesFadeOpacity", {
        /**
         * Gets or sets the target opacity to fade to for fade style HighlightedValues.
        */
        get: function () {
            return this.i.jc;
        },
        set: function (v) {
            this.i.jc = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualHighlightingFadeOpacity", {
        /**
         * Gets actual highlighting fade opacity
        */
        get: function () {
            return this.i.h9;
        },
        set: function (v) {
            this.i.h9 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "highlightingFadeOpacity", {
        /**
         * Gets or sets the target opacity to fade to for fade style highlighting.
        */
        get: function () {
            return this.i.jd;
        },
        set: function (v) {
            this.i.jd = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isActualLegendFinancial", {
        /**
         * Gets or sets whether the ActualLegend is FinancialLegend or normal Legend.
        */
        get: function () {
            return this.i.el;
        },
        set: function (v) {
            this.i.el = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isComponentHighlightingModeIgnored", {
        /**
         * Gets or sets whether the component level highlight mode is ignored.
        */
        get: function () {
            return this.i.ey;
        },
        set: function (v) {
            this.i.ey = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isHighlightingEnabled", {
        /**
         * Gets or sets whether highlighting should be enabled for this series, if this type of series supports highlighting.
        *
        * ```html
        * <igx-data-chart #chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         isHighlightEnabled="true">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.isHighlightingEnabled = true;
        * ```
        */
        get: function () {
            return this.i.fb;
        },
        set: function (v) {
            this.i.fb = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "useItemWiseColors", {
        /**
         * Gets or sets whether the series should use individual palette colors for each item.
        */
        get: function () {
            return this.i.g7;
        },
        set: function (v) {
            this.i.g7 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isColoredItemwise", {
        get: function () {
            return this.i.isColoredItemwise;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "shouldShiftOpacityForSafeActualBrush", {
        /**
         * Gets or sets whether the opacity should be automatically shifted for the safe actual brush.
        */
        get: function () {
            return this.i.gj;
        },
        set: function (v) {
            this.i.gj = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "shouldRemoveHighlightedDataOnLayerHidden", {
        /**
         * Gets or sets whether the opacity should be automatically shifted for the safe actual brush.
        */
        get: function () {
            return this.i.gh;
        },
        set: function (v) {
            this.i.gh = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "shouldHideAutoCallouts", {
        /**
         * Gets or sets whether this series should suppress it's auto callouts
        */
        get: function () {
            return this.i.gg;
        },
        set: function (v) {
            this.i.gg = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isDropShadowEnabled", {
        /**
         * Gets or sets whether drop shadow should be enabled for this series.
        *
        * `IsDropShadowEnabled` is used to decide whether drop shadow should be enabled for this series.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         isDropShadowEnabled=true>
        *     </igx-line-series>
        *     </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.isDropShadowEnabled = true;
        * ```
        */
        get: function () {
            return this.i.e2;
        },
        set: function (v) {
            this.i.e2 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "shadowBlur", {
        /**
         * Gets or sets the shadow blur.
         * This property is ignored when
         * Series.UseSingleShadow is set to true.
        *
        * `ShadowBlur` can be set in conjection with `IsDropShadowEnabled` to further define the drop shadow effect.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         isDropShadowEnabled="true"
        *         shadowBlur="7"
        *         shadowColor="blue">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.shadowBlur = 7;
        * ```
        */
        get: function () {
            return this.i.jk;
        },
        set: function (v) {
            this.i.jk = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "shadowColor", {
        /**
         * Gets or sets the drop shadow color.
        *
        * `ShadowColor` can be set in conjection with `IsDropShadowEnabled` to further define the drop shadow effect.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         isDropShadowEnabled="true"
        *         shadowBlur="7"
        *         shadowColor="blue">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.shadowColor = "blue";
        * ```
        */
        get: function () {
            return colorToString(this.i.vv);
        },
        set: function (v) {
            this.i.vv = stringToColor(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "useSingleShadow", {
        /**
         * Gets or sets whether drop shadow is applied to the whole series visual or to each of the individual shapes forming the series.
         * When this property is set to true, no
         * Series.ShadowBlur is applied.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         isDropShadowEnabled=true
        *         markerType="circle"
        *         useSingleShadow=true>
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.useSingleShadow = true;
        * ```
        */
        get: function () {
            return this.i.ha;
        },
        set: function (v) {
            this.i.ha = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "shadowOffsetX", {
        /**
         * Gets or sets the drop shadow x-offset.
        *
        * `ShadowOffsetX` can be set in conjection with `IsDropShadowEnabled` to further define the drop shadow effect.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         isDropShadowEnabled="true"
        *         shadowOffsetX="10"
        *         shadowOffsetY="10">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.shadowOffsetX = 10;
        * ```
        */
        get: function () {
            return this.i.jl;
        },
        set: function (v) {
            this.i.jl = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "shadowOffsetY", {
        /**
         * Gets or sets the drop shadow y-offset.
        *
        * `ShadowOffsetY` can be set in conjection with `IsDropShadowEnabled` to further define the drop shadow effect.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         isDropShadowEnabled="true"
        *         shadowOffsetX="10"
        *         shadowOffsetY="10">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.shadowOffsetY = 10;
        * ```
        */
        get: function () {
            return this.i.jm;
        },
        set: function (v) {
            this.i.jm = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "areaFillOpacity", {
        /**
         * Gets or sets the opacity applied to the area fill visual.
         * This property only applies to series that have area visual.
        *
        * ```html
        * <igx-data-chart #chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-area-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         areaFillOpacity=.5>
        *     </igx-area-series>
        * </igx-data-chart>
        * ```
        */
        get: function () {
            return this.i.ii;
        },
        set: function (v) {
            this.i.ii = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualAreaFillOpacity", {
        /**
         * Gets the actual opacity applied to the area fill visual.
        *
        * ```ts
        * let actualFillOpacity: number = series.actualAreaFillOpacity;
        * ```
        */
        get: function () {
            return this.i.h7;
        },
        set: function (v) {
            this.i.h7 = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "markerFillOpacity", {
        /**
         * Gets or sets the opacity applied to the fill of the markers.
         * This property only applies to series that have area visual.
        */
        get: function () {
            return this.i.je;
        },
        set: function (v) {
            this.i.je = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualMarkerFillOpacity", {
        /**
         * Gets the actual opacity applied to the fill of the markers.
        */
        get: function () {
            return this.i.ia;
        },
        set: function (v) {
            this.i.ia = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isDefaultToolTipSelected", {
        /**
         * Gets if the default tooltip has been selected for display.
        *
        * To use the default tooltip set `ShowDefaultTooltip` to true.
        *
        * ```ts
        * let isDefaultToolTipSelected: boolean = this.series.isDefaultToolTipSelected;
        * ```
        */
        get: function () {
            return this.i.e1;
        },
        set: function (v) {
            this.i.e1 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "showDefaultTooltip", {
        /**
         * Gets or sets whether default tooltip will be shown.
        *
        * The default tooltips display all the information relevant to the particular series item (series title, data values, axis values etc.) and are styled to match the series' style. When using default tooltips, you should also set the series `Title`.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         showDefaultTooltip="true"
        *         title="InStock Items">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.showDefaultTooltip ="true";
        * ```
        */
        get: function () {
            return this.i.gn;
        },
        set: function (v) {
            this.i.gn = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "visibleRangeMode", {
        /**
         * Gets or sets the visible range mode to use.
        */
        get: function () {
            return this.i.cx;
        },
        set: function (v) {
            this.i.cx = ensureEnum(SeriesVisibleRangeMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "outlineMode", {
        /**
         * Gets or sets the outline mode to use for the series.
        */
        get: function () {
            return this.i.cg;
        },
        set: function (v) {
            this.i.cg = ensureEnum(SeriesOutlineMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "transitionInDuration", {
        /**
         * Gets or sets the duration of the current series' transition in morph.
        *
        * The `TransitionInDuration` can be used with `IsTransitionInEnabled` to cofigure the animation when a new datasource is loaded.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         transitionInMode="accordionFromBottom"
        *         isTransitionInEnabled=true
        *         transitionInSpeedType="indexScaled"
        *         transitionDuration=500
        *         transitionInDuration=500>
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.transitionDuration = 500;
        * ```
        */
        get: function () {
            return this.i.kc;
        },
        set: function (v) {
            this.i.kc = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "transitionOutDuration", {
        /**
         * Gets or sets the duration of the current series' transition out morph.
        */
        get: function () {
            return this.i.kd;
        },
        set: function (v) {
            this.i.kd = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "transitionInSpeedType", {
        /**
         * Gets or sets the duration of the current series' transition in morph.
        *
        * The `TransitionInSpeedType` can be used with `IsTransitionInEnabled` to cofigure the animation when a new datasource is loaded.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         transitionInMode="accordionFromBottom"
        *         isTransitionInEnabled=true
        *         transitionInSpeedType="indexScaled"
        *         transitionInDuration=500>
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.transitionInSpeedType = TransitionInSpeedType.IndexScaled;
        * ```
        */
        get: function () {
            return this.i.c3;
        },
        set: function (v) {
            this.i.c3 = ensureEnum(TransitionInSpeedType_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "transitionOutSpeedType", {
        /**
         * Gets or sets the duration of the current series' transition out morph.
        */
        get: function () {
            return this.i.c6;
        },
        set: function (v) {
            this.i.c6 = ensureEnum(TransitionOutSpeedType_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "lineCap", {
        /**
         * Gets or sets the style of the starting point of any lines or polylines representing this series.
         * Not every series type has a line at which it would be appropriate to display a start cap, so this property does not affect every series type.  LineSeries, for example, is affected by StartCap, but ColumnSeries is not.
        */
        get: function () {
            return this.i.vy;
        },
        set: function (v) {
            this.i.vy = ensureEnum(PenLineCap_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "autoCalloutLabelFormat", {
        /**
         * Gets or sets the AutoCalloutLabel format string to use for the AutoCalloutLabel.
        */
        get: function () {
            return this.i.ls;
        },
        set: function (v) {
            this.i.ls = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "autoCalloutLabelFormatSpecifiers", {
        /**
         * Gets or sets the format specifiers to use with the AutoCalloutLabelFormat string.
        */
        get: function () {
            return this.i.ag;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.ag = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "autoCalloutValueLabelFormat", {
        /**
         * Gets or sets the AutoCalloutLabel format string to use for the AutoCalloutLabel.
        */
        get: function () {
            return this.i.lv;
        },
        set: function (v) {
            this.i.lv = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "autoCalloutValueLabelFormatSpecifiers", {
        /**
         * Gets or sets the format specifiers to use with the AutoCalloutValueLabelFormat string.
        */
        get: function () {
            return this.i.ah;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.ah = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "mouseOverEnabled", {
        /**
         * Gets or sets the whether the chart reacts to mouse move events.
        */
        get: function () {
            return this.i.f6;
        },
        set: function (v) {
            this.i.f6 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "coercionMethods", {
        /**
         * Gets or sets the coercion methods to use when loading data from data sources.
         * Should be specified before setting any member paths, if being used. Setting it later
         * will not cause data to be reimported into the chart.
        *
        * ```html
        *  <igx-data-chart
        *       [dataSource]="data"
        *      >
        *       <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *       </igx-category-x-axis>
        *       <igx-numeric-y-axis
        *         #yAxis>
        *       </igx-numeric-y-axis>
        *       <igx-line-series
        *       [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       valueMemberPath="value">
        *      </igx-line-series>
        *        <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         [coercionMethods]= cMethods
        *         valueMemberPath="value!ValueCoercion">
        *        </igx-line-series>
        *     </igx-data-chart>
        * ```
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        * cMethods: any = {
        *     ValueCoercion: function(value:number): number {return value+2;}
        *  }
        * ```
        */
        get: function () {
            return this.i.coercionMethods;
        },
        set: function (v) {
            this.i.coercionMethods = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "expectFunctions", {
        /**
         * Gets or sets whether the series should expect that properties in its items source may be functions that need to be
         * evaluated to yield their value. This defaults to false, as there will be a subtle performance impact to using this feature.
         * This value must be set before assigning an item's source to take effect.
        *
        * `expectFunctions` is used to decide whether the series should expect that its data source members need to be called as functions to get their values.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *     label="label"
        *     #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *     #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *     [xAxis]="xAxis"
        *     [yAxis]="yAxis"
        *     expectFunctions=true
        *     valueMemberPath="MyFunction">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.expectFunctions=true;
        * ```
        */
        get: function () {
            return this.i.expectFunctions;
        },
        set: function (v) {
            this.i.expectFunctions = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "hitTestMode", {
        /**
         * Method of hit testing to use when pointing to items in this series.
        *
        * The `HitTestMode` can be set to either:
        *
        * -   `Auto`: automatically decide the appropriate hit test mode for the series.
        * -   `ColorEncoded`: use a color encoded off screen buffer for hit testing. Should always be O(1) time for determining a hit series. This increases frame render time and memory usage, however. Consider using this if hit testing time is degrading performance.
        * -   `Computational`: use a computational based approach to determine whether the series has been hit. This uses loose bounding boxes, in some cases, and can range in time complexity between O(1) and O(log n) to find a hit. This decreases frame render time compared to color encoded
        *
        * ```ts
        * this.series.hitTestMode = SeriesHitTestMode.Computational;
        * ```
        */
        get: function () {
            return this.i.b1;
        },
        set: function (v) {
            this.i.b1 = ensureEnum(SeriesHitTestMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualHitTestMode", {
        /**
         * Resolved method of hit testing to use when pointing to items in the chart.
        *
        * ```ts
        * let hitTestMode: SeriesHitTestMode = this.series.actualHitTestMode;
        * ```
        */
        get: function () {
            return this.i.bz;
        },
        set: function (v) {
            this.i.bz = ensureEnum(SeriesHitTestMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isNegativeColorSupported", {
        /**
         * Overridden by derived series classes to indicate when negative colors are supported or not.
        */
        get: function () {
            return this.i.fk;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "isMarkerlessDisplayPreferred", {
        /**
         * Overridden by derived series classes to indicate when markerless display is preferred or not.
        */
        get: function () {
            return this.i.fj;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "finalValue", {
        /**
         * The final value of this series.
        */
        get: function () {
            return this.i.ik;
        },
        set: function (v) {
            this.i.ik = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "percentChange", {
        /**
         * The percent change from the beginning to the end of this series.
        */
        get: function () {
            return this.i.jg;
        },
        set: function (v) {
            this.i.jg = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualOpenLabel", {
        /**
         * Gets the text that is displayed for the Open label in the tooltip.
        */
        get: function () {
            return this.i.li;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualCloseLabel", {
        /**
         * Gets the text that is displayed for the Close label in the tooltip.
        */
        get: function () {
            return this.i.k4;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualHighLabel", {
        /**
         * Gets the text that is displayed for the High label in the tooltip.
        */
        get: function () {
            return this.i.k7;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualLowLabel", {
        /**
         * Gets the text that is displayed for the Low label in the tooltip.
        */
        get: function () {
            return this.i.lg;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualVolumeLabel", {
        /**
         * Gets the text that is displayed for the Volume label in the tooltip.
        */
        get: function () {
            return this.i.lq;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualValueLabel", {
        /**
         * Gets the text that is displayed for the Value label in the tooltip.
        */
        get: function () {
            return this.i.lp;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualRadiusLabel", {
        /**
         * Gets the text that is displayed for the Value label in the tooltip.
        */
        get: function () {
            return this.i.lk;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "layers", {
        /**
         * Gets the current extra layers for the series.
        */
        get: function () {
            if (this._layers === null) {
                var coll = new IgxSeriesLayerCollection();
                var innerColl = this.i.b9;
                if (!innerColl) {
                    innerColl = new SeriesLayerCollection_internal();
                }
                this._layers = coll._fromInner(innerColl);
            }
            return this._layers;
        },
        set: function (v) {
            if (this._layers !== null) {
                this._layers._setSyncTarget(null);
                this._layers = null;
            }
            var coll = new IgxSeriesLayerCollection();
            this._layers = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(SeriesLayer.$type);
            var innerColl = this.i.b9;
            if (!innerColl) {
                innerColl = new SeriesLayerCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._layers._setSyncTarget(syncColl);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "actualLayers", {
        /**
         * Gets the actual extra layers for the series.
        */
        get: function () {
            if (this._actualLayers === null) {
                var coll = new IgxSeriesLayerCollection();
                var innerColl = this.i.actualLayers;
                if (!innerColl) {
                    innerColl = new SeriesLayerCollection_internal();
                }
                this._actualLayers = coll._fromInner(innerColl);
            }
            return this._actualLayers;
        },
        set: function (v) {
            if (this._actualLayers !== null) {
                this._actualLayers._setSyncTarget(null);
                this._actualLayers = null;
            }
            var coll = new IgxSeriesLayerCollection();
            this._actualLayers = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(SeriesLayer.$type);
            var innerColl = this.i.actualLayers;
            if (!innerColl) {
                innerColl = new SeriesLayerCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._actualLayers._setSyncTarget(syncColl);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "opacity", {
        get: function () {
            return this.i.opacity;
        },
        set: function (v) {
            this.i.opacity = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "visibility", {
        get: function () {
            return this.i.visibility;
        },
        set: function (v) {
            this.i.visibility = ensureEnum(Visibility_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesComponent.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.highlightedValuesExtraPropertyOverlays != null && this.highlightedValuesExtraPropertyOverlays.findByName && this.highlightedValuesExtraPropertyOverlays.findByName(name)) {
            return this.highlightedValuesExtraPropertyOverlays.findByName(name);
        }
        if (this.legend && this.legend.name && this.legend.name == name) {
            return this.legend;
        }
        if (this.autoCalloutLabelFormatSpecifiers != null && arrayFindByName(this.autoCalloutLabelFormatSpecifiers, name)) {
            return arrayFindByName(this.autoCalloutLabelFormatSpecifiers, name);
        }
        if (this.autoCalloutValueLabelFormatSpecifiers != null && arrayFindByName(this.autoCalloutValueLabelFormatSpecifiers, name)) {
            return arrayFindByName(this.autoCalloutValueLabelFormatSpecifiers, name);
        }
        if (this.layers != null && this.layers.findByName && this.layers.findByName(name)) {
            return this.layers.findByName(name);
        }
        if (this.actualLayers != null && this.actualLayers.findByName && this.actualLayers.findByName(name)) {
            return this.actualLayers.findByName(name);
        }
        return null;
    };
    Object.defineProperty(IgxSeriesComponent.prototype, "hasUserValues", {
        get: function () {
            return this._hasUserValues;
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesComponent.prototype.__m = function (propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    };
    IgxSeriesComponent.prototype._styling = function (container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        var genericPrefix = "";
        var typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("SeriesComponent");
        var additionalPrefixes = [];
        var prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        var b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            var basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            var parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            var parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    };
    /**
 * Gets the brush for tooltip labels.

*/
    IgxSeriesComponent.prototype.resolveTooltipBrush = function () {
        var iv = this.i.nc();
        return (iv);
    };
    IgxSeriesComponent.prototype.getItemValue = function (item, memberPathName) {
        var iv = this.i.ko(item, memberPathName);
        return (iv);
    };
    /**
     * Gets the value of a requested member path from the series.
    
    * @param memberPathName  * The property name of a valid member path for the series
    */
    IgxSeriesComponent.prototype.getMemberPathValue = function (memberPathName) {
        var iv = this.i.mb(memberPathName);
        return (iv);
    };
    /**
     * Gets the precise item index, if possible, based on the closeness to the previous or next whole integer. If the series cannot provide this information, GetExactItemIndex will return the same integer value as GetItemIndex.
    
    * @param world  * The world position for which to return the index.
    *
    * ```ts
    * let index: number = this.series.getExactItemIndex({x:.5,y:.5});
    * ```
    */
    IgxSeriesComponent.prototype.getExactItemIndex = function (world) {
        var iv = this.i.ip(toPoint(world));
        return (iv);
    };
    /**
     * Gets the item index associated with the specified world position
    
    *
    * ```ts
    * let index: number = this.series.getItemIndex({x:.5,y:.5});
    * ```
    */
    IgxSeriesComponent.prototype.getItemIndex = function (world) {
        var iv = this.i.j1(toPoint(world));
        return (iv);
    };
    /**
     * Gets the item that is the best match for the specified world coordinates.
    
    * @param world  * The world coordinates to use.
    *
    * ```ts
    * let item: number = this.series.getItem({x:.5,y:.5});
    * ```
    */
    IgxSeriesComponent.prototype.getItem = function (world) {
        var iv = this.i.kl(toPoint(world));
        return (iv);
    };
    IgxSeriesComponent.prototype.getPreviousOrExactIndex = function (world, skipUnknowns) {
        var iv = this.i.j8(toPoint(world), skipUnknowns);
        return (iv);
    };
    IgxSeriesComponent.prototype.getNextOrExactIndex = function (world, skipUnknowns) {
        var iv = this.i.j6(toPoint(world), skipUnknowns);
        return (iv);
    };
    IgxSeriesComponent.prototype.getSeriesValue = function (world, useInterpolation, skipUnknowns) {
        var iv = this.i.i3(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    };
    IgxSeriesComponent.prototype.getSeriesValuePosition = function (world, useInterpolation, skipUnknowns) {
        var iv = this.i.v6(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    };
    /**
     * If possible, will return the best available value bounding box within the series that has the best value match for the world position provided.
    
    * @param world  * The world coordinate for which to get a value bounding box for
    *
    * ```ts
    * let box: IgxRect = this.series.getSeriesValueBoundingBox({x:.8,y:.8});
    * ```
    */
    IgxSeriesComponent.prototype.getSeriesValueBoundingBox = function (world) {
        var iv = this.i.wh(toPoint(world));
        return fromRect(iv);
    };
    /**
     * If possible, will return the best available value marker bounding box within the series that has the best value match for the world position provided.
    
    * @param world  * The world coordinates for which to get a value marker bounding box for
    *
    * ```ts
    * let box: IgxRect = this.series.getSeriesValueMarkerBoundingBox({x:.8,y:.8});
    * ```
    */
    IgxSeriesComponent.prototype.getSeriesValueMarkerBoundingBox = function (world) {
        var iv = this.i.wj(toPoint(world));
        return fromRect(iv);
    };
    IgxSeriesComponent.prototype.getSeriesHighValue = function (world, useInterpolation, skipUnknowns) {
        var iv = this.i.iz(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    };
    IgxSeriesComponent.prototype.getSeriesHighValuePosition = function (world, useInterpolation, skipUnknowns) {
        var iv = this.i.v2(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    };
    IgxSeriesComponent.prototype.getSeriesLowValue = function (world, useInterpolation, skipUnknowns) {
        var iv = this.i.i1(toPoint(world), useInterpolation, skipUnknowns);
        return (iv);
    };
    IgxSeriesComponent.prototype.getSeriesLowValuePosition = function (world, useInterpolation, skipUnknowns) {
        var iv = this.i.v4(toPoint(world), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    };
    IgxSeriesComponent.prototype.getSeriesValuePositionFromSeriesPixel = function (mouse, useInterpolation, skipUnknowns) {
        var iv = this.i.v7(toPoint(mouse), useInterpolation, skipUnknowns);
        return fromPoint(iv);
    };
    IgxSeriesComponent.prototype.getSeriesValueFromSeriesPixel = function (mouse, useInterpolation, skipUnknowns) {
        var iv = this.i.i4(toPoint(mouse), useInterpolation, skipUnknowns);
        return (iv);
    };
    /**
     * For a category plotted series, returns the current width of the items within the categories. This only returns a value if the items have some form of width (e.g. columns, bars, etc.) otherwise 0 is returned.
    
    *
    * ```ts
    * let span: number = this.series.getItemSpan();
    * ```
    */
    IgxSeriesComponent.prototype.getItemSpan = function () {
        var iv = this.i.iu();
        return (iv);
    };
    /**
     * Gets the data values of each axis associated with the specified value type. For example, if you specify
     * mode as Maximum then it will return the maximum numerical value for each axis. For non-numerical axes it
     * will return NaN.
    
    * @param mode  * The type of value desired from the series axes.
    */
    IgxSeriesComponent.prototype.getSeriesValueType = function (mode) {
        var iv = this.i.ap((mode == null ? null : mode));
        return (iv);
    };
    /**
     * Gets position of series value for specified layer value mode
    
    */
    IgxSeriesComponent.prototype.getSeriesValueTypePosition = function (mode) {
        var iv = this.i.v9((mode == null ? null : mode));
        return fromPoint(iv);
    };
    IgxSeriesComponent.prototype.getSeriesValueTypePositionFromValue = function (values) {
        var iv = this.i.wa(values);
        return fromPoint(iv);
    };
    /**
     * Hides any internal tooltips presented by the series, if any.
    
    */
    IgxSeriesComponent.prototype.hideToolTips = function () {
        this.i.p2();
    };
    /**
     * Hides any internal tooltips presented by the series, if any.
    
    */
    IgxSeriesComponent.prototype.hideToolTipsImmediate = function () {
        this.i.p3();
    };
    /**
     * Converts a point from coordinates within the series plot area to a world position within axis space.
    
    * @param seriesPoint  * The pixel location within the plot area of the series.
    *
    * ```ts
    * this.series.toWorldPosition({x:500,y:500});
    * ```
    */
    IgxSeriesComponent.prototype.toWorldPosition = function (seriesPoint) {
        var iv = this.i.wb(toPoint(seriesPoint));
        return fromPoint(iv);
    };
    /**
     * Converts a point from world coordinates to coordinates within the viewport of the series.
    
    * @param world  * The world position from which to convert.
    *
    * ```ts
    * let pt: IgxPoint = this.series.fromWorldPosition({x:.5,y:.5});
    * ```
    */
    IgxSeriesComponent.prototype.fromWorldPosition = function (world) {
        var iv = this.i.v1(toPoint(world));
        return fromPoint(iv);
    };
    /**
     * Requests that the provided item should be brought into view if possible.
    
    * @param item  * The item to attempt to bring into view.
    *
    * ```ts
    * this.series.scrollIntoView(dataItem);
    * ```
    */
    IgxSeriesComponent.prototype.scrollIntoView = function (item) {
        var iv = this.i.gb(item);
        return (iv);
    };
    /**
     * Renders the series.
    
    * @param animate  * True if the change should be animated.
    *
    * ```ts
    * this.series.renderSeries(true);
    * ```
    */
    IgxSeriesComponent.prototype.renderSeries = function (animate) {
        this.i.rn(animate);
    };
    IgxSeriesComponent.prototype.getMainContentViewport = function () {
        var iv = this.i.getMainContentViewport();
        return fromRect(iv);
    };
    /**
     * Gets the effective viewport, adjusted to account for margins and other factors.
    
    */
    IgxSeriesComponent.prototype.getEffectiveViewport = function () {
        var iv = this.i.getEffectiveViewport();
        return fromRect(iv);
    };
    /**
     * Removes all alternate views of this series.
    
    */
    IgxSeriesComponent.prototype.removeAllAlternateViews = function () {
        this.i.rg();
    };
    /**
     * Removes an alternate view of this series.
    
    * @param viewIdentifier  * The string identifier for the view to remove.
    */
    IgxSeriesComponent.prototype.removeAlternateView = function (viewIdentifier) {
        this.i.ri(viewIdentifier);
    };
    /**
     * Called to notify about changes to indexed-based properties, e.g. Brushes, Outlines, MarkerBrushes, MarkerOutlines and refresh series
    
    */
    IgxSeriesComponent.prototype.notifyIndexedPropertiesChanged = function () {
        this.i.qc();
    };
    /**
     * Notifies the series that a visual property has changed, requiring a visual update.
    
    */
    IgxSeriesComponent.prototype.notifyVisualPropertiesChanged = function () {
        this.i.qj();
    };
    /**
     * Replays the assigned transition in animation, if any.
    
    *
    * ```ts
    * this.series.replayTransitionIn();
    * ```
    */
    IgxSeriesComponent.prototype.replayTransitionIn = function () {
        this.i.rr();
    };
    /**
     * Plays the assigned transition out animation, if any.
    
    */
    IgxSeriesComponent.prototype.playTransitionOut = function () {
        this.i.q3();
    };
    /**
     * Plays the assigned transition out animation, if any.
    
    */
    IgxSeriesComponent.prototype.playTransitionIn = function () {
        this.i.q2();
    };
    /**
     * Plays the assigned transition out animation, if any.
    
    */
    IgxSeriesComponent.prototype.playTransitionOutAndRemove = function () {
        this.i.q4();
    };
    /**
     * Call to null out the axes of the series;
    
    */
    IgxSeriesComponent.prototype.removeAxes = function () {
        this.i.rj();
    };
    /**
     * Simulates a pointer hover over the series surface.
    
    * @param point  * The pointer position relative to the series viewport over which to hover.
    *
    * When tooltips are enabled, `SimulateHover` can be used to automatically display the tooltip over a particular series item.
    *
    * ```ts
    * this.series.simulateHover({x:250,y:250});
    * ```
    */
    IgxSeriesComponent.prototype.simulateHover = function (point) {
        this.i.r0(toPoint(point));
    };
    /**
     * Called when this series' Style is updated.
    
    */
    IgxSeriesComponent.prototype.styleUpdated = function () {
        this.i.r6();
    };
    IgxSeriesComponent.prototype.setNegativeColors = function (negativeBrush, negativeOutline) {
        this.i.rx(stringToBrush(negativeBrush), stringToBrush(negativeOutline));
    };
    /**
     * Moves the cursor point to the provided world position. Some series may react to the current cursor position.
    
    * @param point  * The cursor point, in world coordinates.
    *
    * `MoveCursorPoint` can be used on annotation layers to programitically show the annotation at the given point.
    *
    * ```ts
    * this.crosshair.moveCursorPoint({x:.5,y:.5});
    * ```
    */
    IgxSeriesComponent.prototype.moveCursorPoint = function (point) {
        this.i.p8(toPoint(point));
    };
    Object.defineProperty(IgxSeriesComponent.prototype, "transitionOutCompleted", {
        /**
         * Fired when a transition out has completed.
        */
        get: function () {
            var _this = this;
            if (this._transitionOutCompleted == null) {
                this._transitionOutCompleted = new EventEmitter();
                this.i.transitionOutCompleted = delegateCombine(this.i.transitionOutCompleted, function (o, e) {
                    _this._runInZone(function () {
                        var outerArgs = new IgxTransitionOutCompletedEventArgs();
                        outerArgs._provideImplementation(e);
                        if (_this.beforeTransitionOutCompleted) {
                            _this.beforeTransitionOutCompleted(_this, outerArgs);
                        }
                        _this._transitionOutCompleted.emit({
                            sender: _this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._transitionOutCompleted;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesComponent.prototype, "renderRequested", {
        /**
         * This event is raised every time a render has been requested from a series.
        */
        get: function () {
            var _this = this;
            if (this._renderRequested == null) {
                this._renderRequested = new EventEmitter();
                this.i.renderRequested = delegateCombine(this.i.renderRequested, function (o, e) {
                    var outerArgs = new IgxRenderRequestedEventArgs();
                    outerArgs._provideImplementation(e);
                    if (_this.beforeRenderRequested) {
                        _this.beforeRenderRequested(_this, outerArgs);
                    }
                    _this._renderRequested.emit({
                        sender: _this,
                        args: outerArgs
                    });
                });
            }
            return this._renderRequested;
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesComponent.prototype._runInZone = function (act) {
        if (this._zoneRunner != null) {
            this._zoneRunner(act);
        }
        else {
            act();
        }
    };
    IgxSeriesComponent.fac = i0.ngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSeriesComponent, deps: [], target: i0.FactoryTarget.Component });
    IgxSeriesComponent.cmp = i0.ngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxSeriesComponent, selector: "ng-component", inputs: { name: "name", legend: "legend", dataSource: "dataSource", highlightedDataSource: "highlightedDataSource", tooltipTemplate: "tooltipTemplate", tooltipContainerTemplate: "tooltipContainerTemplate", dataLegendGroup: "dataLegendGroup", highlightedValuesDataLegendGroup: "highlightedValuesDataLegendGroup", highlightedValuesExtraPropertyOverlays: "highlightedValuesExtraPropertyOverlays", legendItemVisibility: "legendItemVisibility", legendItemBadgeTemplate: "legendItemBadgeTemplate", actualLegendItemBadgeTemplate: "actualLegendItemBadgeTemplate", actualLegendItemBadgeOutline: "actualLegendItemBadgeOutline", actualLegendItemBadgeBrush: "actualLegendItemBadgeBrush", legendItemBadgeShape: "legendItemBadgeShape", legendItemBadgeMode: "legendItemBadgeMode", legendItemTemplate: "legendItemTemplate", discreteLegendItemTemplate: "discreteLegendItemTemplate", transitionEasingFunction: "transitionEasingFunction", transitionInEasingFunction: "transitionInEasingFunction", transitionOutEasingFunction: "transitionOutEasingFunction", transitionDuration: "transitionDuration", actualResolution: "actualResolution", resolution: "resolution", visibleRangeMarginTop: "visibleRangeMarginTop", visibleRangeMarginBottom: "visibleRangeMarginBottom", visibleRangeMarginLeft: "visibleRangeMarginLeft", visibleRangeMarginRight: "visibleRangeMarginRight", title: "title", highlightedTitleSuffix: "highlightedTitleSuffix", highlightedLegendItemVisibility: "highlightedLegendItemVisibility", brush: "brush", actualBrush: "actualBrush", selectionBrush: "selectionBrush", focusBrush: "focusBrush", actualSelectionBrush: "actualSelectionBrush", actualFocusBrush: "actualFocusBrush", safeActualBrush: "safeActualBrush", outline: "outline", actualOutline: "actualOutline", lineJoin: "lineJoin", highlightedValuesDisplayMode: "highlightedValuesDisplayMode", shouldAnimateOnDataSourceSwap: "shouldAnimateOnDataSourceSwap", thickness: "thickness", actualThickness: "actualThickness", dashArray: "dashArray", actualHighlightingMode: "actualHighlightingMode", actualSelectionMode: "actualSelectionMode", actualFocusMode: "actualFocusMode", actualHighlightedValuesFadeOpacity: "actualHighlightedValuesFadeOpacity", highlightedValuesFadeOpacity: "highlightedValuesFadeOpacity", actualHighlightingFadeOpacity: "actualHighlightingFadeOpacity", highlightingFadeOpacity: "highlightingFadeOpacity", isActualLegendFinancial: "isActualLegendFinancial", isComponentHighlightingModeIgnored: "isComponentHighlightingModeIgnored", isHighlightingEnabled: "isHighlightingEnabled", useItemWiseColors: "useItemWiseColors", shouldShiftOpacityForSafeActualBrush: "shouldShiftOpacityForSafeActualBrush", shouldRemoveHighlightedDataOnLayerHidden: "shouldRemoveHighlightedDataOnLayerHidden", shouldHideAutoCallouts: "shouldHideAutoCallouts", isDropShadowEnabled: "isDropShadowEnabled", shadowBlur: "shadowBlur", shadowColor: "shadowColor", useSingleShadow: "useSingleShadow", shadowOffsetX: "shadowOffsetX", shadowOffsetY: "shadowOffsetY", areaFillOpacity: "areaFillOpacity", actualAreaFillOpacity: "actualAreaFillOpacity", markerFillOpacity: "markerFillOpacity", actualMarkerFillOpacity: "actualMarkerFillOpacity", isDefaultToolTipSelected: "isDefaultToolTipSelected", showDefaultTooltip: "showDefaultTooltip", visibleRangeMode: "visibleRangeMode", outlineMode: "outlineMode", transitionInDuration: "transitionInDuration", transitionOutDuration: "transitionOutDuration", transitionInSpeedType: "transitionInSpeedType", transitionOutSpeedType: "transitionOutSpeedType", lineCap: "lineCap", autoCalloutLabelFormat: "autoCalloutLabelFormat", autoCalloutLabelFormatSpecifiers: "autoCalloutLabelFormatSpecifiers", autoCalloutValueLabelFormat: "autoCalloutValueLabelFormat", autoCalloutValueLabelFormatSpecifiers: "autoCalloutValueLabelFormatSpecifiers", mouseOverEnabled: "mouseOverEnabled", coercionMethods: "coercionMethods", expectFunctions: "expectFunctions", hitTestMode: "hitTestMode", actualHitTestMode: "actualHitTestMode", finalValue: "finalValue", percentChange: "percentChange", layers: "layers", actualLayers: "actualLayers", opacity: "opacity", visibility: "visibility" }, outputs: { transitionOutCompleted: "transitionOutCompleted", renderRequested: "renderRequested" }, ngImport: i0, template: "", isInline: true });
    return IgxSeriesComponent;
}());
export { IgxSeriesComponent };
i0.ngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    template: "",
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { name: [{
                type: Input
            }], legend: [{
                type: Input
            }], dataSource: [{
                type: Input
            }], highlightedDataSource: [{
                type: Input
            }], tooltipTemplate: [{
                type: Input
            }], tooltipContainerTemplate: [{
                type: Input
            }], dataLegendGroup: [{
                type: Input
            }], highlightedValuesDataLegendGroup: [{
                type: Input
            }], highlightedValuesExtraPropertyOverlays: [{
                type: Input
            }], legendItemVisibility: [{
                type: Input
            }], legendItemBadgeTemplate: [{
                type: Input
            }], actualLegendItemBadgeTemplate: [{
                type: Input
            }], actualLegendItemBadgeOutline: [{
                type: Input
            }], actualLegendItemBadgeBrush: [{
                type: Input
            }], legendItemBadgeShape: [{
                type: Input
            }], legendItemBadgeMode: [{
                type: Input
            }], legendItemTemplate: [{
                type: Input
            }], discreteLegendItemTemplate: [{
                type: Input
            }], transitionEasingFunction: [{
                type: Input
            }], transitionInEasingFunction: [{
                type: Input
            }], transitionOutEasingFunction: [{
                type: Input
            }], transitionDuration: [{
                type: Input
            }], actualResolution: [{
                type: Input
            }], resolution: [{
                type: Input
            }], visibleRangeMarginTop: [{
                type: Input
            }], visibleRangeMarginBottom: [{
                type: Input
            }], visibleRangeMarginLeft: [{
                type: Input
            }], visibleRangeMarginRight: [{
                type: Input
            }], title: [{
                type: Input
            }], highlightedTitleSuffix: [{
                type: Input
            }], highlightedLegendItemVisibility: [{
                type: Input
            }], brush: [{
                type: Input
            }], actualBrush: [{
                type: Input
            }], selectionBrush: [{
                type: Input
            }], focusBrush: [{
                type: Input
            }], actualSelectionBrush: [{
                type: Input
            }], actualFocusBrush: [{
                type: Input
            }], safeActualBrush: [{
                type: Input
            }], outline: [{
                type: Input
            }], actualOutline: [{
                type: Input
            }], lineJoin: [{
                type: Input
            }], highlightedValuesDisplayMode: [{
                type: Input
            }], shouldAnimateOnDataSourceSwap: [{
                type: Input
            }], thickness: [{
                type: Input
            }], actualThickness: [{
                type: Input
            }], dashArray: [{
                type: Input
            }], actualHighlightingMode: [{
                type: Input
            }], actualSelectionMode: [{
                type: Input
            }], actualFocusMode: [{
                type: Input
            }], actualHighlightedValuesFadeOpacity: [{
                type: Input
            }], highlightedValuesFadeOpacity: [{
                type: Input
            }], actualHighlightingFadeOpacity: [{
                type: Input
            }], highlightingFadeOpacity: [{
                type: Input
            }], isActualLegendFinancial: [{
                type: Input
            }], isComponentHighlightingModeIgnored: [{
                type: Input
            }], isHighlightingEnabled: [{
                type: Input
            }], useItemWiseColors: [{
                type: Input
            }], shouldShiftOpacityForSafeActualBrush: [{
                type: Input
            }], shouldRemoveHighlightedDataOnLayerHidden: [{
                type: Input
            }], shouldHideAutoCallouts: [{
                type: Input
            }], isDropShadowEnabled: [{
                type: Input
            }], shadowBlur: [{
                type: Input
            }], shadowColor: [{
                type: Input
            }], useSingleShadow: [{
                type: Input
            }], shadowOffsetX: [{
                type: Input
            }], shadowOffsetY: [{
                type: Input
            }], areaFillOpacity: [{
                type: Input
            }], actualAreaFillOpacity: [{
                type: Input
            }], markerFillOpacity: [{
                type: Input
            }], actualMarkerFillOpacity: [{
                type: Input
            }], isDefaultToolTipSelected: [{
                type: Input
            }], showDefaultTooltip: [{
                type: Input
            }], visibleRangeMode: [{
                type: Input
            }], outlineMode: [{
                type: Input
            }], transitionInDuration: [{
                type: Input
            }], transitionOutDuration: [{
                type: Input
            }], transitionInSpeedType: [{
                type: Input
            }], transitionOutSpeedType: [{
                type: Input
            }], lineCap: [{
                type: Input
            }], autoCalloutLabelFormat: [{
                type: Input
            }], autoCalloutLabelFormatSpecifiers: [{
                type: Input
            }], autoCalloutValueLabelFormat: [{
                type: Input
            }], autoCalloutValueLabelFormatSpecifiers: [{
                type: Input
            }], mouseOverEnabled: [{
                type: Input
            }], coercionMethods: [{
                type: Input
            }], expectFunctions: [{
                type: Input
            }], hitTestMode: [{
                type: Input
            }], actualHitTestMode: [{
                type: Input
            }], finalValue: [{
                type: Input
            }], percentChange: [{
                type: Input
            }], layers: [{
                type: Input
            }], actualLayers: [{
                type: Input
            }], opacity: [{
                type: Input
            }], visibility: [{
                type: Input
            }], transitionOutCompleted: [{
                type: Output
            }], renderRequested: [{
                type: Output
            }] } });
