// igniteui-angular-charts/esm5/lib/igx-financial-price-series-proxy-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FinancialPriceSeriesProxy } from './FinancialPriceSeriesProxy';
import { IgxFinancialEventArgs } from './igx-financial-event-args';
import { FinancialEventArgs as FinancialEventArgs_internal } from './FinancialEventArgs';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxFinancialPriceSeriesProxyModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxFinancialPriceSeriesProxyModule() {
        TypeRegistrar.register('FinancialPriceSeriesProxy', FinancialPriceSeriesProxy.$type);
        TypeRegistrar.register('FinancialEventArgs', FinancialEventArgs_internal.$type);
        TypeRegistrar.registerCons('igx-financial-event-args', IgxFinancialEventArgs);
    }
    IgxFinancialPriceSeriesProxyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesProxyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFinancialPriceSeriesProxyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesProxyModule, imports: [CommonModule] });
    IgxFinancialPriceSeriesProxyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesProxyModule, imports: [[CommonModule,
            ]] });
    return IgxFinancialPriceSeriesProxyModule;
}());
export { IgxFinancialPriceSeriesProxyModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesProxyModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
