// igniteui-angular-charts/esm5/lib/igx-funnel-chart-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumberFormatSpecifierDynamicModule } from "igniteui-angular-core";
import { IgxDateTimeFormatSpecifierDynamicModule } from "igniteui-angular-core";
import { IgxFunnelChartComponent } from './igx-funnel-chart-component';
import { FastItemObjectColumn } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxFunnelChartCoreModule } from './igx-funnel-chart-core-module';
import * as i0 from "@angular/core";
var IgxFunnelChartCoreDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxFunnelChartCoreDynamicModule() {
        TypeRegistrar.registerCons('IgxFunnelChartComponent', IgxFunnelChartComponent);
        TypeRegistrar.register('FastItemObjectColumn', FastItemObjectColumn.$type);
    }
    IgxFunnelChartCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFunnelChartCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartCoreDynamicModule, imports: [CommonModule, IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxFunnelChartCoreModule], exports: [IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxFunnelChartCoreModule] });
    IgxFunnelChartCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartCoreDynamicModule, imports: [[CommonModule, IgxNumberFormatSpecifierDynamicModule,
                IgxDateTimeFormatSpecifierDynamicModule,
                IgxFunnelChartCoreModule
            ], IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxFunnelChartCoreModule] });
    return IgxFunnelChartCoreDynamicModule;
}());
export { IgxFunnelChartCoreDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumberFormatSpecifierDynamicModule,
                        IgxDateTimeFormatSpecifierDynamicModule,
                        IgxFunnelChartCoreModule
                    ],
                    imports: [CommonModule, IgxNumberFormatSpecifierDynamicModule,
                        IgxDateTimeFormatSpecifierDynamicModule,
                        IgxFunnelChartCoreModule
                    ],
                    entryComponents: [IgxFunnelChartComponent]
                }]
        }], ctorParameters: function () { return []; } });
