// igniteui-angular-charts/esm5/lib/igx-stacked-spline-area-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxHorizontalStackedSeriesBaseComponent } from "./igx-horizontal-stacked-series-base-component";
import { IgxStackedSeriesBaseComponent } from "./igx-stacked-series-base-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { StackedSplineAreaSeries } from "./StackedSplineAreaSeries";
import { ensureBool } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent stacked spline area series.
*/
var IgxStackedSplineAreaSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxStackedSplineAreaSeriesComponent, _super);
    function IgxStackedSplineAreaSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxStackedSplineAreaSeriesComponent.prototype.createImplementation = function () {
        return new StackedSplineAreaSeries();
    };
    Object.defineProperty(IgxStackedSplineAreaSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxStackedSplineAreaSeriesComponent.prototype, "isSplineShapePartOfRange", {
        /**
         * Gets or sets whether to include the spline shape in the axis range requested of the axis.
        */
        get: function () {
            return this.i.abl;
        },
        set: function (v) {
            this.i.abl = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxStackedSplineAreaSeriesComponent.prototype, "isArea", {
        /**
         * Gets whether the current series shows an area shape.
        */
        get: function () {
            return this.i.es;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxStackedSplineAreaSeriesComponent.prototype, "isSpline", {
        /**
         * Gets whether the current series shows a spline shape.
        */
        get: function () {
            return this.i.fy;
        },
        enumerable: false,
        configurable: true
    });
    IgxStackedSplineAreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineAreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStackedSplineAreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStackedSplineAreaSeriesComponent, selector: "igx-stacked-spline-area-series", inputs: { isSplineShapePartOfRange: "isSplineShapePartOfRange" }, providers: [{ provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStackedSplineAreaSeriesComponent; }) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStackedSplineAreaSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStackedSplineAreaSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxStackedSplineAreaSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxStackedSplineAreaSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStackedSplineAreaSeriesComponent;
}(IgxHorizontalStackedSeriesBaseComponent));
export { IgxStackedSplineAreaSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineAreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-stacked-spline-area-series',
                    template: "",
                    providers: [{ provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStackedSplineAreaSeriesComponent; }) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStackedSplineAreaSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStackedSplineAreaSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxStackedSplineAreaSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxStackedSplineAreaSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { isSplineShapePartOfRange: [{
                type: Input
            }] } });
