// igniteui-angular-charts/esm5/lib/igx-numeric-angle-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericAngleAxisComponent } from './igx-numeric-angle-axis-component';
import * as i0 from "@angular/core";
var IgxNumericAngleAxisModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxNumericAngleAxisModule() {
    }
    IgxNumericAngleAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAngleAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxNumericAngleAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAngleAxisModule, declarations: [IgxNumericAngleAxisComponent], imports: [CommonModule], exports: [IgxNumericAngleAxisComponent] });
    IgxNumericAngleAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAngleAxisModule, imports: [[CommonModule]] });
    return IgxNumericAngleAxisModule;
}());
export { IgxNumericAngleAxisModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAngleAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxNumericAngleAxisComponent],
                    imports: [CommonModule],
                    exports: [IgxNumericAngleAxisComponent],
                    entryComponents: []
                }]
        }] });
