// igniteui-angular-charts/esm5/lib/igx-proportional-category-angle-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxProportionalCategoryAngleAxisComponent } from './igx-proportional-category-angle-axis-component';
import * as i0 from "@angular/core";
var IgxProportionalCategoryAngleAxisModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxProportionalCategoryAngleAxisModule() {
    }
    IgxProportionalCategoryAngleAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxProportionalCategoryAngleAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxProportionalCategoryAngleAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxProportionalCategoryAngleAxisModule, declarations: [IgxProportionalCategoryAngleAxisComponent], imports: [CommonModule], exports: [IgxProportionalCategoryAngleAxisComponent] });
    IgxProportionalCategoryAngleAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxProportionalCategoryAngleAxisModule, imports: [[CommonModule]] });
    return IgxProportionalCategoryAngleAxisModule;
}());
export { IgxProportionalCategoryAngleAxisModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxProportionalCategoryAngleAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxProportionalCategoryAngleAxisComponent],
                    imports: [CommonModule],
                    exports: [IgxProportionalCategoryAngleAxisComponent],
                    entryComponents: []
                }]
        }] });
