// igniteui-angular-charts/esm5/lib/igx-radial-area-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxRadialAreaSeriesModule } from './igx-radial-area-series-module';
import { IgxRadialAreaSeriesComponent } from './igx-radial-area-series-component';
import { RadialAreaSeries } from './RadialAreaSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxRadialAreaSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxRadialAreaSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxRadialAreaSeriesComponent", IgxRadialAreaSeriesComponent);
        TypeRegistrar.register("RadialAreaSeries", RadialAreaSeries.$type);
    }
    IgxRadialAreaSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialAreaSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxRadialAreaSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialAreaSeriesDynamicModule, imports: [CommonModule, IgxRadialAreaSeriesModule], exports: [IgxRadialAreaSeriesModule] });
    IgxRadialAreaSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialAreaSeriesDynamicModule, imports: [[CommonModule, IgxRadialAreaSeriesModule], IgxRadialAreaSeriesModule] });
    return IgxRadialAreaSeriesDynamicModule;
}());
export { IgxRadialAreaSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialAreaSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxRadialAreaSeriesModule],
                    exports: [IgxRadialAreaSeriesModule],
                    entryComponents: [IgxRadialAreaSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
