// igniteui-angular-charts/esm5/lib/igx-time-axis-label-format-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTimeAxisLabelFormatModule } from './igx-time-axis-label-format-module';
import { IgxTimeAxisLabelFormatComponent } from './igx-time-axis-label-format-component';
import { TimeAxisLabelFormat } from './TimeAxisLabelFormat';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxTimeAxisLabelFormatCollection } from './igx-time-axis-label-format-collection';
import { TimeAxisLabelFormatCollection } from './TimeAxisLabelFormatCollection';
import * as i0 from "@angular/core";
var IgxTimeAxisLabelFormatDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxTimeAxisLabelFormatDynamicModule() {
        TypeRegistrar.registerCons("IgxTimeAxisLabelFormatComponent", IgxTimeAxisLabelFormatComponent);
        TypeRegistrar.register("TimeAxisLabelFormat", TimeAxisLabelFormat.$type);
        TypeRegistrar.registerCons("IgxTimeAxisLabelFormatCollection", IgxTimeAxisLabelFormatCollection);
        TypeRegistrar.register("TimeAxisLabelFormatCollection", TimeAxisLabelFormatCollection.$type);
    }
    IgxTimeAxisLabelFormatDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisLabelFormatDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTimeAxisLabelFormatDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisLabelFormatDynamicModule, imports: [CommonModule, IgxTimeAxisLabelFormatModule], exports: [IgxTimeAxisLabelFormatModule] });
    IgxTimeAxisLabelFormatDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisLabelFormatDynamicModule, imports: [[CommonModule, IgxTimeAxisLabelFormatModule], IgxTimeAxisLabelFormatModule] });
    return IgxTimeAxisLabelFormatDynamicModule;
}());
export { IgxTimeAxisLabelFormatDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisLabelFormatDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxTimeAxisLabelFormatModule],
                    exports: [IgxTimeAxisLabelFormatModule],
                    entryComponents: [IgxTimeAxisLabelFormatComponent]
                }]
        }], ctorParameters: function () { return []; } });
