// igniteui-angular-charts/esm5/lib/igx-itemwise-strategy-based-indicator-component.js
import { __extends } from "tslib";
import { Component } from '@angular/core';
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import * as i0 from "@angular/core";
/**
 * Financial indicator base class for doing calculating an itemwize indicator.
 * An itemwise indicator's individual values don't depend on any other values in the
 * collection so can always be independently recalculated. Also, because of this, the
 * base class handles all the looping logic and inheritors only need to provide the
 * actual calculation.
*/
var IgxItemwiseStrategyBasedIndicatorComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxItemwiseStrategyBasedIndicatorComponent, _super);
    function IgxItemwiseStrategyBasedIndicatorComponent() {
        return _super.call(this) || this;
    }
    Object.defineProperty(IgxItemwiseStrategyBasedIndicatorComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * This is a marker method to distinguish this from a public api perspective on some platforms compared to the base strategy based indicators.
    
    */
    IgxItemwiseStrategyBasedIndicatorComponent.prototype.resolveIsItemwise = function () {
        var iv = this.i.aca();
        return (iv);
    };
    IgxItemwiseStrategyBasedIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemwiseStrategyBasedIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxItemwiseStrategyBasedIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxItemwiseStrategyBasedIndicatorComponent, selector: "ng-component", usesInheritance: true, ngImport: i0, template: "", isInline: true });
    return IgxItemwiseStrategyBasedIndicatorComponent;
}(IgxStrategyBasedIndicatorComponent));
export { IgxItemwiseStrategyBasedIndicatorComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemwiseStrategyBasedIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    template: "",
                }]
        }], ctorParameters: function () { return []; } });
