// igniteui-angular-charts/esm5/lib/igx-numeric-radius-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericRadiusAxisComponent } from './igx-numeric-radius-axis-component';
import * as i0 from "@angular/core";
var IgxNumericRadiusAxisModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxNumericRadiusAxisModule() {
    }
    IgxNumericRadiusAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericRadiusAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxNumericRadiusAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericRadiusAxisModule, declarations: [IgxNumericRadiusAxisComponent], imports: [CommonModule], exports: [IgxNumericRadiusAxisComponent] });
    IgxNumericRadiusAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericRadiusAxisModule, imports: [[CommonModule]] });
    return IgxNumericRadiusAxisModule;
}());
export { IgxNumericRadiusAxisModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericRadiusAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxNumericRadiusAxisComponent],
                    imports: [CommonModule],
                    exports: [IgxNumericRadiusAxisComponent],
                    entryComponents: []
                }]
        }] });
