// igniteui-angular-charts/esm5/lib/igx-data-chart-vertical-category-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericXAxisModule } from './igx-numeric-x-axis-module';
import { IgxCategoryYAxisModule } from './igx-category-y-axis-module';
import { IgxDataChartCoreModule } from './igx-data-chart-core-module';
import * as i0 from "@angular/core";
var IgxDataChartVerticalCategoryCoreModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartVerticalCategoryCoreModule() {
    }
    IgxDataChartVerticalCategoryCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartVerticalCategoryCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryCoreModule, imports: [CommonModule, IgxNumericXAxisModule,
            IgxCategoryYAxisModule,
            IgxDataChartCoreModule], exports: [IgxNumericXAxisModule,
            IgxCategoryYAxisModule,
            IgxDataChartCoreModule] });
    IgxDataChartVerticalCategoryCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryCoreModule, imports: [[CommonModule, IgxNumericXAxisModule,
                IgxCategoryYAxisModule,
                IgxDataChartCoreModule
            ], IgxNumericXAxisModule,
            IgxCategoryYAxisModule,
            IgxDataChartCoreModule] });
    return IgxDataChartVerticalCategoryCoreModule;
}());
export { IgxDataChartVerticalCategoryCoreModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumericXAxisModule,
                        IgxCategoryYAxisModule,
                        IgxDataChartCoreModule
                    ],
                    imports: [CommonModule, IgxNumericXAxisModule,
                        IgxCategoryYAxisModule,
                        IgxDataChartCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
