// igniteui-angular-charts/esm5/lib/igx-data-chart-radial-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxRadialAreaSeriesModule } from './igx-radial-area-series-module';
import { IgxRadialColumnSeriesModule } from './igx-radial-column-series-module';
import { IgxRadialLineSeriesModule } from './igx-radial-line-series-module';
import { IgxRadialPieSeriesModule } from './igx-radial-pie-series-module';
import { IgxDataChartRadialCoreModule } from './igx-data-chart-radial-core-module';
import * as i0 from "@angular/core";
var IgxDataChartRadialModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartRadialModule() {
    }
    IgxDataChartRadialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartRadialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialModule, imports: [CommonModule, IgxRadialAreaSeriesModule,
            IgxRadialColumnSeriesModule,
            IgxRadialLineSeriesModule,
            IgxRadialPieSeriesModule,
            IgxDataChartRadialCoreModule], exports: [IgxRadialAreaSeriesModule,
            IgxRadialColumnSeriesModule,
            IgxRadialLineSeriesModule,
            IgxRadialPieSeriesModule,
            IgxDataChartRadialCoreModule] });
    IgxDataChartRadialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialModule, imports: [[CommonModule, IgxRadialAreaSeriesModule,
                IgxRadialColumnSeriesModule,
                IgxRadialLineSeriesModule,
                IgxRadialPieSeriesModule,
                IgxDataChartRadialCoreModule
            ], IgxRadialAreaSeriesModule,
            IgxRadialColumnSeriesModule,
            IgxRadialLineSeriesModule,
            IgxRadialPieSeriesModule,
            IgxDataChartRadialCoreModule] });
    return IgxDataChartRadialModule;
}());
export { IgxDataChartRadialModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxRadialAreaSeriesModule,
                        IgxRadialColumnSeriesModule,
                        IgxRadialLineSeriesModule,
                        IgxRadialPieSeriesModule,
                        IgxDataChartRadialCoreModule
                    ],
                    imports: [CommonModule, IgxRadialAreaSeriesModule,
                        IgxRadialColumnSeriesModule,
                        IgxRadialLineSeriesModule,
                        IgxRadialPieSeriesModule,
                        IgxDataChartRadialCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
