// igniteui-angular-charts/esm5/lib/igx-category-chart-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryChartComponent } from './igx-category-chart-component';
import { IgxCategoryXAxisDynamicModule } from './igx-category-x-axis-dynamic-module';
import { IgxNumericYAxisDynamicModule } from './igx-numeric-y-axis-dynamic-module';
import * as i0 from "@angular/core";
var IgxCategoryChartCoreModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCategoryChartCoreModule() {
    }
    IgxCategoryChartCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryChartCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartCoreModule, declarations: [IgxCategoryChartComponent], imports: [CommonModule, IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule], exports: [IgxCategoryChartComponent,
            IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule] });
    IgxCategoryChartCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartCoreModule, imports: [[CommonModule, IgxCategoryXAxisDynamicModule,
                IgxNumericYAxisDynamicModule
            ], IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule] });
    return IgxCategoryChartCoreModule;
}());
export { IgxCategoryChartCoreModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxCategoryChartComponent],
                    exports: [IgxCategoryChartComponent,
                        IgxCategoryXAxisDynamicModule,
                        IgxNumericYAxisDynamicModule
                    ],
                    imports: [CommonModule, IgxCategoryXAxisDynamicModule,
                        IgxNumericYAxisDynamicModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
