// igniteui-angular-charts/esm5/lib/igx-data-chart-financial-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFinancialPriceSeriesModule } from './igx-financial-price-series-module';
import { IgxDataChartFinancialCoreModule } from './igx-data-chart-financial-core-module';
import * as i0 from "@angular/core";
var IgxDataChartFinancialModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartFinancialModule() {
    }
    IgxDataChartFinancialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartFinancialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialModule, imports: [CommonModule, IgxFinancialPriceSeriesModule,
            IgxDataChartFinancialCoreModule], exports: [IgxFinancialPriceSeriesModule,
            IgxDataChartFinancialCoreModule] });
    IgxDataChartFinancialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialModule, imports: [[CommonModule, IgxFinancialPriceSeriesModule,
                IgxDataChartFinancialCoreModule
            ], IgxFinancialPriceSeriesModule,
            IgxDataChartFinancialCoreModule] });
    return IgxDataChartFinancialModule;
}());
export { IgxDataChartFinancialModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxFinancialPriceSeriesModule,
                        IgxDataChartFinancialCoreModule
                    ],
                    imports: [CommonModule, IgxFinancialPriceSeriesModule,
                        IgxDataChartFinancialCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
