// igniteui-angular-charts/esm5/lib/igx-value-layer-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxValueLayerModule } from './igx-value-layer-module';
import { IgxValueLayerComponent } from './igx-value-layer-component';
import { ValueLayer } from './ValueLayer';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxValueLayerDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxValueLayerDynamicModule() {
        TypeRegistrar.registerCons("IgxValueLayerComponent", IgxValueLayerComponent);
        TypeRegistrar.register("ValueLayer", ValueLayer.$type);
    }
    IgxValueLayerDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueLayerDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxValueLayerDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueLayerDynamicModule, imports: [CommonModule, IgxValueLayerModule], exports: [IgxValueLayerModule] });
    IgxValueLayerDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueLayerDynamicModule, imports: [[CommonModule, IgxValueLayerModule], IgxValueLayerModule] });
    return IgxValueLayerDynamicModule;
}());
export { IgxValueLayerDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueLayerDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxValueLayerModule],
                    exports: [IgxValueLayerModule],
                    entryComponents: [IgxValueLayerComponent]
                }]
        }], ctorParameters: function () { return []; } });
