// igniteui-angular-charts/esm5/lib/igx-annotation-layer-proxy-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AnnotationLayerProxy } from './AnnotationLayerProxy';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxAnnotationLayerProxyModule } from './igx-annotation-layer-proxy-module';
import * as i0 from "@angular/core";
var IgxAnnotationLayerProxyDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxAnnotationLayerProxyDynamicModule() {
        TypeRegistrar.registerCons('AnnotationLayerProxy', AnnotationLayerProxy);
        TypeRegistrar.register('AnnotationLayerProxy', AnnotationLayerProxy.$type);
    }
    IgxAnnotationLayerProxyDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAnnotationLayerProxyDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAnnotationLayerProxyDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAnnotationLayerProxyDynamicModule, imports: [CommonModule, IgxAnnotationLayerProxyModule], exports: [IgxAnnotationLayerProxyModule] });
    IgxAnnotationLayerProxyDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAnnotationLayerProxyDynamicModule, imports: [[CommonModule, IgxAnnotationLayerProxyModule
            ], IgxAnnotationLayerProxyModule] });
    return IgxAnnotationLayerProxyDynamicModule;
}());
export { IgxAnnotationLayerProxyDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAnnotationLayerProxyDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxAnnotationLayerProxyModule
                    ],
                    imports: [CommonModule, IgxAnnotationLayerProxyModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
