// igniteui-angular-charts/esm5/lib/igx-range-column-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxHorizontalRangeCategorySeriesProxyModule } from './igx-horizontal-range-category-series-proxy-module';
import { IgxRangeColumnSeriesComponent } from './igx-range-column-series-component';
import * as i0 from "@angular/core";
var IgxRangeColumnSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxRangeColumnSeriesModule() {
    }
    IgxRangeColumnSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeColumnSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxRangeColumnSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeColumnSeriesModule, declarations: [IgxRangeColumnSeriesComponent], imports: [CommonModule, IgxHorizontalRangeCategorySeriesProxyModule], exports: [IgxRangeColumnSeriesComponent, IgxHorizontalRangeCategorySeriesProxyModule] });
    IgxRangeColumnSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeColumnSeriesModule, imports: [[CommonModule, IgxHorizontalRangeCategorySeriesProxyModule], IgxHorizontalRangeCategorySeriesProxyModule] });
    return IgxRangeColumnSeriesModule;
}());
export { IgxRangeColumnSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeColumnSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxRangeColumnSeriesComponent],
                    imports: [CommonModule, IgxHorizontalRangeCategorySeriesProxyModule],
                    exports: [IgxRangeColumnSeriesComponent, IgxHorizontalRangeCategorySeriesProxyModule],
                    entryComponents: []
                }]
        }] });
