// igniteui-angular-charts/esm5/lib/igx-range-column-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxRangeColumnSeriesModule } from './igx-range-column-series-module';
import { IgxRangeColumnSeriesComponent } from './igx-range-column-series-component';
import { RangeColumnSeries } from './RangeColumnSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxRangeColumnSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxRangeColumnSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxRangeColumnSeriesComponent", IgxRangeColumnSeriesComponent);
        TypeRegistrar.register("RangeColumnSeries", RangeColumnSeries.$type);
    }
    IgxRangeColumnSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeColumnSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxRangeColumnSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeColumnSeriesDynamicModule, imports: [CommonModule, IgxRangeColumnSeriesModule], exports: [IgxRangeColumnSeriesModule] });
    IgxRangeColumnSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeColumnSeriesDynamicModule, imports: [[CommonModule, IgxRangeColumnSeriesModule], IgxRangeColumnSeriesModule] });
    return IgxRangeColumnSeriesDynamicModule;
}());
export { IgxRangeColumnSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeColumnSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxRangeColumnSeriesModule],
                    exports: [IgxRangeColumnSeriesModule],
                    entryComponents: [IgxRangeColumnSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
