// igniteui-angular-charts/esm5/lib/igx-data-chart-interactivity-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { InteractivityFactoryManager } from './InteractivityFactoryManager';
import { IgxDataChartDefaultTooltipsComponent } from './igx-data-chart-default-tooltips-component';
import { IgxDVInteractivityModule } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxDataChartInteractivityModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartInteractivityModule() {
        TypeRegistrar.register('InteractivityFactoryManager', InteractivityFactoryManager.$type);
        InteractivityFactoryManager.register();
        TypeRegistrar.registerCons('IgxDataChartDefaultTooltipsComponent', IgxDataChartDefaultTooltipsComponent);
    }
    IgxDataChartInteractivityModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartInteractivityModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartInteractivityModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartInteractivityModule, declarations: [IgxDataChartDefaultTooltipsComponent], imports: [CommonModule, IgxDVInteractivityModule], exports: [IgxDataChartDefaultTooltipsComponent,
            IgxDVInteractivityModule] });
    IgxDataChartInteractivityModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartInteractivityModule, imports: [[CommonModule, IgxDVInteractivityModule
            ], IgxDVInteractivityModule] });
    return IgxDataChartInteractivityModule;
}());
export { IgxDataChartInteractivityModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartInteractivityModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxDataChartDefaultTooltipsComponent],
                    exports: [IgxDataChartDefaultTooltipsComponent,
                        IgxDVInteractivityModule
                    ],
                    imports: [CommonModule, IgxDVInteractivityModule
                    ],
                    entryComponents: [IgxDataChartDefaultTooltipsComponent]
                }]
        }], ctorParameters: function () { return []; } });
