// igniteui-angular-charts/esm5/lib/igx-numeric-y-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericYAxisModule } from './igx-numeric-y-axis-module';
import { IgxNumericYAxisComponent } from './igx-numeric-y-axis-component';
import { NumericYAxis } from './NumericYAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxNumericYAxisDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxNumericYAxisDynamicModule() {
        TypeRegistrar.registerCons("IgxNumericYAxisComponent", IgxNumericYAxisComponent);
        TypeRegistrar.register("NumericYAxis", NumericYAxis.$type);
    }
    IgxNumericYAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericYAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxNumericYAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericYAxisDynamicModule, imports: [CommonModule, IgxNumericYAxisModule], exports: [IgxNumericYAxisModule] });
    IgxNumericYAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericYAxisDynamicModule, imports: [[CommonModule, IgxNumericYAxisModule], IgxNumericYAxisModule] });
    return IgxNumericYAxisDynamicModule;
}());
export { IgxNumericYAxisDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericYAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxNumericYAxisModule],
                    exports: [IgxNumericYAxisModule],
                    entryComponents: [IgxNumericYAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
