// igniteui-angular-charts/esm5/lib/igx-range-area-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxHorizontalRangeCategorySeriesComponent } from "./igx-horizontal-range-category-series-component";
import { IgxRangeCategorySeriesComponent } from "./igx-range-category-series-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { RangeAreaSeries } from "./RangeAreaSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent range area series.
*
* The `RangeAreaSeries` class represents the range area series of the IgxDataChartComponent.
*
* ```html
* <igx-data-chart
* [dataSource]="data">
* <igx-category-x-axis
* 	label="label"
* 	#xAxis>
* </igx-category-x-axis>
* <igx-numeric-y-axis
* 	#yAxis>
* </igx-numeric-y-axis>
*     <igx-range-area-series
*       [xAxis]="xAxis"
*       [yAxis]="yAxis"
*       highMemberPath="high"
*       lowMemberPath="low">
*   </igx-range-area-series>
*     </igx-data-chart>
* </igx-data-chart>
* ```
*
* ```ts
* let series: IgxRangeAreaSeriesComponent = new IgxRangeAreaSeriesComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.highMemberPath = "High";
* series.lowMemberPath = "Low";
* this.chart.series.add(series);
* ```
*/
var IgxRangeAreaSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxRangeAreaSeriesComponent, _super);
    function IgxRangeAreaSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxRangeAreaSeriesComponent.prototype.createImplementation = function () {
        return new RangeAreaSeries();
    };
    Object.defineProperty(IgxRangeAreaSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRangeAreaSeriesComponent.prototype, "isAreaOrLine", {
        /**
         * Gets whether the current series shows an area or line shape.
        *
        * Use the `IsAreaOrLine` property to check if the current series shows an area or line shape.
        *
        * ```ts
        * var x= series.isAreaOrLine;
        * ```
        */
        get: function () {
            return this.i.et;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRangeAreaSeriesComponent.prototype, "isArea", {
        /**
         * Gets whether the current series shows an area shape.
        */
        get: function () {
            return this.i.es;
        },
        enumerable: false,
        configurable: true
    });
    IgxRangeAreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeAreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxRangeAreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxRangeAreaSeriesComponent, selector: "igx-range-area-series", providers: [{ provide: IgxHorizontalRangeCategorySeriesComponent, useExisting: forwardRef(function () { return IgxRangeAreaSeriesComponent; }) }, { provide: IgxRangeCategorySeriesComponent, useExisting: forwardRef(function () { return IgxRangeAreaSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxRangeAreaSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxRangeAreaSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxRangeAreaSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxRangeAreaSeriesComponent;
}(IgxHorizontalRangeCategorySeriesComponent));
export { IgxRangeAreaSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeAreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-range-area-series',
                    template: "",
                    providers: [{ provide: IgxHorizontalRangeCategorySeriesComponent, useExisting: forwardRef(function () { return IgxRangeAreaSeriesComponent; }) }, { provide: IgxRangeCategorySeriesComponent, useExisting: forwardRef(function () { return IgxRangeAreaSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxRangeAreaSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxRangeAreaSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxRangeAreaSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
