// igniteui-angular-charts/esm5/lib/igx-data-chart-visual-data-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VisualDataFactoryManager } from './VisualDataFactoryManager';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxDataChartVisualDataModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartVisualDataModule() {
        TypeRegistrar.register('VisualDataFactoryManager', VisualDataFactoryManager.$type);
        VisualDataFactoryManager.register();
    }
    IgxDataChartVisualDataModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVisualDataModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartVisualDataModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVisualDataModule, imports: [CommonModule] });
    IgxDataChartVisualDataModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVisualDataModule, imports: [[CommonModule,
            ]] });
    return IgxDataChartVisualDataModule;
}());
export { IgxDataChartVisualDataModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVisualDataModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
