// igniteui-angular-charts/esm5/lib/igx-scatter-polygon-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterPolygonSeriesModule } from './igx-scatter-polygon-series-module';
import { IgxScatterPolygonSeriesComponent } from './igx-scatter-polygon-series-component';
import { ScatterPolygonSeries } from './ScatterPolygonSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxScatterPolygonSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxScatterPolygonSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxScatterPolygonSeriesComponent", IgxScatterPolygonSeriesComponent);
        TypeRegistrar.register("ScatterPolygonSeries", ScatterPolygonSeries.$type);
    }
    IgxScatterPolygonSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolygonSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterPolygonSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolygonSeriesDynamicModule, imports: [CommonModule, IgxScatterPolygonSeriesModule], exports: [IgxScatterPolygonSeriesModule] });
    IgxScatterPolygonSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolygonSeriesDynamicModule, imports: [[CommonModule, IgxScatterPolygonSeriesModule], IgxScatterPolygonSeriesModule] });
    return IgxScatterPolygonSeriesDynamicModule;
}());
export { IgxScatterPolygonSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolygonSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxScatterPolygonSeriesModule],
                    exports: [IgxScatterPolygonSeriesModule],
                    entryComponents: [IgxScatterPolygonSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
