// igniteui-angular-charts/esm5/lib/igx-price-volume-trend-indicator-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import { IgxFinancialIndicatorComponent } from "./igx-financial-indicator-component";
import { IgxFinancialSeriesComponent } from "./igx-financial-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { PriceVolumeTrendIndicator } from "./PriceVolumeTrendIndicator";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent Price Volume Trend Indicator series.
 * Default required members: Volume, Close
*
* You can use the `PriceVolumeTrendIndicator` to measure money flow by adding or subtracting a portion of the daily volume.
*
* ```html
* <igx-data-chart
*     [dataSource]="data" >
*     <igx-category-x-axis
* 		label="label"
* 		#xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
* 		#yAxis>
*     </igx-numeric-y-axis>
*     <igx-price-volume-trend-indicator
* 		[xAxis]="xAxis"
* 		[yAxis]="yAxis"
* 		openMemberPath="open"
* 		highMemberPath="high"
* 		lowMemberPath="low"
* 		closeMemberPath="close">
*     </igx-price-volume-trend-indicator>
* </igx-data-chart>
* ```
*
* ```ts
* let series = new IgxPriceVolumeTrendIndicatorComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.openMemberPath = "open";
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* series.closeMemberPath = "close";
* this.chart.series.add(series);
* ```
*/
var IgxPriceVolumeTrendIndicatorComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxPriceVolumeTrendIndicatorComponent, _super);
    function IgxPriceVolumeTrendIndicatorComponent() {
        return _super.call(this) || this;
    }
    IgxPriceVolumeTrendIndicatorComponent.prototype.createImplementation = function () {
        return new PriceVolumeTrendIndicator();
    };
    Object.defineProperty(IgxPriceVolumeTrendIndicatorComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgxPriceVolumeTrendIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceVolumeTrendIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxPriceVolumeTrendIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxPriceVolumeTrendIndicatorComponent, selector: "igx-price-volume-trend-indicator", providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxPriceVolumeTrendIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxPriceVolumeTrendIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxPriceVolumeTrendIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxPriceVolumeTrendIndicatorComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxPriceVolumeTrendIndicatorComponent;
}(IgxStrategyBasedIndicatorComponent));
export { IgxPriceVolumeTrendIndicatorComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceVolumeTrendIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-price-volume-trend-indicator',
                    template: "",
                    providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxPriceVolumeTrendIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxPriceVolumeTrendIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxPriceVolumeTrendIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxPriceVolumeTrendIndicatorComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
