// igniteui-angular-charts/esm5/lib/igx-average-true-range-indicator-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import { IgxFinancialIndicatorComponent } from "./igx-financial-indicator-component";
import { IgxFinancialSeriesComponent } from "./igx-financial-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { AverageTrueRangeIndicator } from "./AverageTrueRangeIndicator";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent Average True Range indicator series.
 * Default required members: High, Low, Close
*
* You can use the `AverageTrueRangeIndicator` to measures a security's degree of price movement or volatility within a given period of time.
*
* ```html
* <igx-data-chart
* 	#chart
*     [dataSource]="data" >
*     <igx-category-x-axis
* 		label="label"
* 		#xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
* 		#yAxis>
*     </igx-numeric-y-axis>
*     <igx-average-true-range-indicator
* 		[xAxis]="xAxis"
* 		[yAxis]="yAxis"
* 		openMemberPath="open"
* 		highMemberPath="high"
* 		lowMemberPath="low"
* 		closeMemberPath="close">
*     </igx-average-true-range-indicator>
* </igx-data-chart>
* ```
*
* ```ts
* let series = new IgxAverageTrueRangeIndicatorComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.openMemberPath = "open";
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* series.closeMemberPath = "close";
* this.chart.series.add(series);
* ```
*/
var IgxAverageTrueRangeIndicatorComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxAverageTrueRangeIndicatorComponent, _super);
    function IgxAverageTrueRangeIndicatorComponent() {
        return _super.call(this) || this;
    }
    IgxAverageTrueRangeIndicatorComponent.prototype.createImplementation = function () {
        return new AverageTrueRangeIndicator();
    };
    Object.defineProperty(IgxAverageTrueRangeIndicatorComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxAverageTrueRangeIndicatorComponent.prototype, "period", {
        /**
         * Gets or sets the moving average period for the current AverageTrueRangeSeries object.
         * The typical, and initial, value for AverageTrueRange periods is 14.
        *
        * You can use the `Period` to set the moving average.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data" >
        *     <igx-category-x-axis
        * 		label="label"
        * 		#xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        * 		#yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-average-true-range-indicator
        * 		#series
        * 		[xAxis]="xAxis"
        * 		[yAxis]="yAxis"
        * 		openMemberPath="open"
        * 		highMemberPath="high"
        * 		lowMemberPath="low"
        * 		closeMemberPath="close"
        * 		period="14">
        *     </igx-average-true-range-indicator>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.period = 14;
        * ```
        */
        get: function () {
            return this.i.ab8;
        },
        set: function (v) {
            this.i.ab8 = +v;
        },
        enumerable: false,
        configurable: true
    });
    IgxAverageTrueRangeIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAverageTrueRangeIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxAverageTrueRangeIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxAverageTrueRangeIndicatorComponent, selector: "igx-average-true-range-indicator", inputs: { period: "period" }, providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxAverageTrueRangeIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxAverageTrueRangeIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxAverageTrueRangeIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxAverageTrueRangeIndicatorComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxAverageTrueRangeIndicatorComponent;
}(IgxStrategyBasedIndicatorComponent));
export { IgxAverageTrueRangeIndicatorComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAverageTrueRangeIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-average-true-range-indicator',
                    template: "",
                    providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxAverageTrueRangeIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxAverageTrueRangeIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxAverageTrueRangeIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxAverageTrueRangeIndicatorComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { period: [{
                type: Input
            }] } });
