// igniteui-angular-charts/esm5/lib/igx-treemap-node-style-mapping-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { TreemapNodeStyleMappingTargetType_$type } from "./TreemapNodeStyleMappingTargetType";
import { TreemapValueMappingMode_$type } from "./TreemapValueMappingMode";
import { IgxTreemapNodeStyleComponent } from "./igx-treemap-node-style-component";
import { TreemapNodeStyleMapping } from "./TreemapNodeStyleMapping";
import { ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a style to apply to a set of nodes in the treemap.
*/
var IgxTreemapNodeStyleMappingComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxTreemapNodeStyleMappingComponent, _super);
    function IgxTreemapNodeStyleMappingComponent() {
        return _super.call(this) || this;
    }
    IgxTreemapNodeStyleMappingComponent.prototype.createImplementation = function () {
        return new TreemapNodeStyleMapping();
    };
    Object.defineProperty(IgxTreemapNodeStyleMappingComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxTreemapNodeStyleMappingComponent.prototype, "targetType", {
        /**
         * Gets or sets the unique name of the mapping.
        */
        get: function () {
            return this.i.bo;
        },
        set: function (v) {
            this.i.bo = ensureEnum(TreemapNodeStyleMappingTargetType_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxTreemapNodeStyleMappingComponent.prototype, "name", {
        /**
         * Gets or sets the unique name of the mapping.
        */
        get: function () {
            return this.i.b1;
        },
        set: function (v) {
            this.i.b1 = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxTreemapNodeStyleMappingComponent.prototype, "minimumValue", {
        /**
         * Gets or sets the minimum value to which the mapping applies.
        */
        get: function () {
            return this.i.bu;
        },
        set: function (v) {
            this.i.bu = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxTreemapNodeStyleMappingComponent.prototype, "maximumValue", {
        /**
         * Gets or sets the maximum value to which the mapping applies.
        */
        get: function () {
            return this.i.bt;
        },
        set: function (v) {
            this.i.bt = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxTreemapNodeStyleMappingComponent.prototype, "value", {
        /**
         * Gets or sets the value to which the mapping applies.
        */
        get: function () {
            return this.i.bw;
        },
        set: function (v) {
            this.i.bw = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxTreemapNodeStyleMappingComponent.prototype, "mappingMode", {
        /**
         * Gets or sets the mapping mode to use to evaluate whether the mapping applies.
        */
        get: function () {
            return this.i.bq;
        },
        set: function (v) {
            this.i.bq = ensureEnum(TreemapValueMappingMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    IgxTreemapNodeStyleMappingComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxTreemapNodeStyleMappingComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxTreemapNodeStyleMappingComponent, selector: "igx-treemap-node-style-mapping", inputs: { targetType: "targetType", name: "name", minimumValue: "minimumValue", maximumValue: "maximumValue", value: "value", mappingMode: "mappingMode" }, providers: [{ provide: IgxTreemapNodeStyleComponent, useExisting: forwardRef(function () { return IgxTreemapNodeStyleMappingComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxTreemapNodeStyleMappingComponent;
}(IgxTreemapNodeStyleComponent));
export { IgxTreemapNodeStyleMappingComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-treemap-node-style-mapping',
                    template: "",
                    providers: [{ provide: IgxTreemapNodeStyleComponent, useExisting: forwardRef(function () { return IgxTreemapNodeStyleMappingComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { targetType: [{
                type: Input
            }], name: [{
                type: Input
            }], minimumValue: [{
                type: Input
            }], maximumValue: [{
                type: Input
            }], value: [{
                type: Input
            }], mappingMode: [{
                type: Input
            }] } });
