// igniteui-angular-charts/esm5/lib/igx-money-flow-index-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxMoneyFlowIndexIndicatorModule } from './igx-money-flow-index-indicator-module';
import { IgxMoneyFlowIndexIndicatorComponent } from './igx-money-flow-index-indicator-component';
import { MoneyFlowIndexIndicator } from './MoneyFlowIndexIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxMoneyFlowIndexIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxMoneyFlowIndexIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxMoneyFlowIndexIndicatorComponent", IgxMoneyFlowIndexIndicatorComponent);
        TypeRegistrar.register("MoneyFlowIndexIndicator", MoneyFlowIndexIndicator.$type);
    }
    IgxMoneyFlowIndexIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMoneyFlowIndexIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxMoneyFlowIndexIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMoneyFlowIndexIndicatorDynamicModule, imports: [CommonModule, IgxMoneyFlowIndexIndicatorModule], exports: [IgxMoneyFlowIndexIndicatorModule] });
    IgxMoneyFlowIndexIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMoneyFlowIndexIndicatorDynamicModule, imports: [[CommonModule, IgxMoneyFlowIndexIndicatorModule], IgxMoneyFlowIndexIndicatorModule] });
    return IgxMoneyFlowIndexIndicatorDynamicModule;
}());
export { IgxMoneyFlowIndexIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMoneyFlowIndexIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxMoneyFlowIndexIndicatorModule],
                    exports: [IgxMoneyFlowIndexIndicatorModule],
                    entryComponents: [IgxMoneyFlowIndexIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
