// igniteui-angular-charts/esm5/lib/igx-category-chart-toolbar-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataChartToolbarIconsModule } from './igx-data-chart-toolbar-icons-module';
import { DataChartToolbarFactoryManager } from './DataChartToolbarFactoryManager';
import { CategoryChartToolbarFactoryManager } from './CategoryChartToolbarFactoryManager';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxCategoryChartToolbarModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCategoryChartToolbarModule() {
        TypeRegistrar.register('DataChartToolbarFactoryManager', DataChartToolbarFactoryManager.$type);
        DataChartToolbarFactoryManager.register();
        TypeRegistrar.register('CategoryChartToolbarFactoryManager', CategoryChartToolbarFactoryManager.$type);
        CategoryChartToolbarFactoryManager.register();
    }
    IgxCategoryChartToolbarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartToolbarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryChartToolbarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartToolbarModule, imports: [CommonModule, IgxDataChartToolbarIconsModule], exports: [IgxDataChartToolbarIconsModule] });
    IgxCategoryChartToolbarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartToolbarModule, imports: [[CommonModule, IgxDataChartToolbarIconsModule
            ], IgxDataChartToolbarIconsModule] });
    return IgxCategoryChartToolbarModule;
}());
export { IgxCategoryChartToolbarModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartToolbarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataChartToolbarIconsModule
                    ],
                    imports: [CommonModule, IgxDataChartToolbarIconsModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
