// igniteui-angular-charts/esm5/lib/igx-data-chart-polar-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPolarAreaSeriesDynamicModule } from './igx-polar-area-series-dynamic-module';
import { IgxPolarLineSeriesDynamicModule } from './igx-polar-line-series-dynamic-module';
import { IgxPolarScatterSeriesDynamicModule } from './igx-polar-scatter-series-dynamic-module';
import { IgxPolarSplineSeriesDynamicModule } from './igx-polar-spline-series-dynamic-module';
import { IgxPolarSplineAreaSeriesDynamicModule } from './igx-polar-spline-area-series-dynamic-module';
import { IgxDataChartPolarCoreDynamicModule } from './igx-data-chart-polar-core-dynamic-module';
import { IgxDataChartPolarModule } from './igx-data-chart-polar-module';
import * as i0 from "@angular/core";
var IgxDataChartPolarDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartPolarDynamicModule() {
    }
    IgxDataChartPolarDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartPolarDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarDynamicModule, imports: [CommonModule, IgxPolarAreaSeriesDynamicModule,
            IgxPolarLineSeriesDynamicModule,
            IgxPolarScatterSeriesDynamicModule,
            IgxPolarSplineSeriesDynamicModule,
            IgxPolarSplineAreaSeriesDynamicModule,
            IgxDataChartPolarCoreDynamicModule,
            IgxDataChartPolarModule], exports: [IgxPolarAreaSeriesDynamicModule,
            IgxPolarLineSeriesDynamicModule,
            IgxPolarScatterSeriesDynamicModule,
            IgxPolarSplineSeriesDynamicModule,
            IgxPolarSplineAreaSeriesDynamicModule,
            IgxDataChartPolarCoreDynamicModule,
            IgxDataChartPolarModule] });
    IgxDataChartPolarDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarDynamicModule, imports: [[CommonModule, IgxPolarAreaSeriesDynamicModule,
                IgxPolarLineSeriesDynamicModule,
                IgxPolarScatterSeriesDynamicModule,
                IgxPolarSplineSeriesDynamicModule,
                IgxPolarSplineAreaSeriesDynamicModule,
                IgxDataChartPolarCoreDynamicModule,
                IgxDataChartPolarModule
            ], IgxPolarAreaSeriesDynamicModule,
            IgxPolarLineSeriesDynamicModule,
            IgxPolarScatterSeriesDynamicModule,
            IgxPolarSplineSeriesDynamicModule,
            IgxPolarSplineAreaSeriesDynamicModule,
            IgxDataChartPolarCoreDynamicModule,
            IgxDataChartPolarModule] });
    return IgxDataChartPolarDynamicModule;
}());
export { IgxDataChartPolarDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxPolarAreaSeriesDynamicModule,
                        IgxPolarLineSeriesDynamicModule,
                        IgxPolarScatterSeriesDynamicModule,
                        IgxPolarSplineSeriesDynamicModule,
                        IgxPolarSplineAreaSeriesDynamicModule,
                        IgxDataChartPolarCoreDynamicModule,
                        IgxDataChartPolarModule
                    ],
                    imports: [CommonModule, IgxPolarAreaSeriesDynamicModule,
                        IgxPolarLineSeriesDynamicModule,
                        IgxPolarScatterSeriesDynamicModule,
                        IgxPolarSplineSeriesDynamicModule,
                        IgxPolarSplineAreaSeriesDynamicModule,
                        IgxDataChartPolarCoreDynamicModule,
                        IgxDataChartPolarModule
                    ],
                    entryComponents: []
                }]
        }] });
