// igniteui-angular-charts/esm5/lib/igx-horizontal-range-category-series-proxy-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HorizontalRangeCategorySeriesProxy } from './HorizontalRangeCategorySeriesProxy';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxHorizontalRangeCategorySeriesProxyModule } from './igx-horizontal-range-category-series-proxy-module';
import * as i0 from "@angular/core";
var IgxHorizontalRangeCategorySeriesProxyDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxHorizontalRangeCategorySeriesProxyDynamicModule() {
        TypeRegistrar.registerCons('HorizontalRangeCategorySeriesProxy', HorizontalRangeCategorySeriesProxy);
        TypeRegistrar.register('HorizontalRangeCategorySeriesProxy', HorizontalRangeCategorySeriesProxy.$type);
    }
    IgxHorizontalRangeCategorySeriesProxyDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHorizontalRangeCategorySeriesProxyDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxHorizontalRangeCategorySeriesProxyDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHorizontalRangeCategorySeriesProxyDynamicModule, imports: [CommonModule, IgxHorizontalRangeCategorySeriesProxyModule], exports: [IgxHorizontalRangeCategorySeriesProxyModule] });
    IgxHorizontalRangeCategorySeriesProxyDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHorizontalRangeCategorySeriesProxyDynamicModule, imports: [[CommonModule, IgxHorizontalRangeCategorySeriesProxyModule
            ], IgxHorizontalRangeCategorySeriesProxyModule] });
    return IgxHorizontalRangeCategorySeriesProxyDynamicModule;
}());
export { IgxHorizontalRangeCategorySeriesProxyDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHorizontalRangeCategorySeriesProxyDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxHorizontalRangeCategorySeriesProxyModule
                    ],
                    imports: [CommonModule, IgxHorizontalRangeCategorySeriesProxyModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
