// igniteui-angular-charts/esm5/lib/igx-stacked-fragment-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedFragmentSeriesComponent } from './igx-stacked-fragment-series-component';
import * as i0 from "@angular/core";
var IgxStackedFragmentSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStackedFragmentSeriesModule() {
    }
    IgxStackedFragmentSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedFragmentSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedFragmentSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedFragmentSeriesModule, declarations: [IgxStackedFragmentSeriesComponent], imports: [CommonModule], exports: [IgxStackedFragmentSeriesComponent] });
    IgxStackedFragmentSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedFragmentSeriesModule, imports: [[CommonModule]] });
    return IgxStackedFragmentSeriesModule;
}());
export { IgxStackedFragmentSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedFragmentSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStackedFragmentSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxStackedFragmentSeriesComponent],
                    entryComponents: []
                }]
        }] });
