// igniteui-angular-charts/esm5/lib/igx-stacked-100-line-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStacked100LineSeriesModule } from './igx-stacked-100-line-series-module';
import { IgxStacked100LineSeriesComponent } from './igx-stacked-100-line-series-component';
import { Stacked100LineSeries } from './Stacked100LineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxStacked100LineSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStacked100LineSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxStacked100LineSeriesComponent", IgxStacked100LineSeriesComponent);
        TypeRegistrar.register("Stacked100LineSeries", Stacked100LineSeries.$type);
    }
    IgxStacked100LineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100LineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStacked100LineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100LineSeriesDynamicModule, imports: [CommonModule, IgxStacked100LineSeriesModule], exports: [IgxStacked100LineSeriesModule] });
    IgxStacked100LineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100LineSeriesDynamicModule, imports: [[CommonModule, IgxStacked100LineSeriesModule], IgxStacked100LineSeriesModule] });
    return IgxStacked100LineSeriesDynamicModule;
}());
export { IgxStacked100LineSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100LineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStacked100LineSeriesModule],
                    exports: [IgxStacked100LineSeriesModule],
                    entryComponents: [IgxStacked100LineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
