// igniteui-angular-charts/esm5/lib/igx-average-true-range-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAverageTrueRangeIndicatorModule } from './igx-average-true-range-indicator-module';
import { IgxAverageTrueRangeIndicatorComponent } from './igx-average-true-range-indicator-component';
import { AverageTrueRangeIndicator } from './AverageTrueRangeIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxAverageTrueRangeIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxAverageTrueRangeIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxAverageTrueRangeIndicatorComponent", IgxAverageTrueRangeIndicatorComponent);
        TypeRegistrar.register("AverageTrueRangeIndicator", AverageTrueRangeIndicator.$type);
    }
    IgxAverageTrueRangeIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAverageTrueRangeIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAverageTrueRangeIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAverageTrueRangeIndicatorDynamicModule, imports: [CommonModule, IgxAverageTrueRangeIndicatorModule], exports: [IgxAverageTrueRangeIndicatorModule] });
    IgxAverageTrueRangeIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAverageTrueRangeIndicatorDynamicModule, imports: [[CommonModule, IgxAverageTrueRangeIndicatorModule], IgxAverageTrueRangeIndicatorModule] });
    return IgxAverageTrueRangeIndicatorDynamicModule;
}());
export { IgxAverageTrueRangeIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAverageTrueRangeIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxAverageTrueRangeIndicatorModule],
                    exports: [IgxAverageTrueRangeIndicatorModule],
                    entryComponents: [IgxAverageTrueRangeIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
