// igniteui-angular-charts/esm5/lib/igx-scatter-line-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterLineSeriesModule } from './igx-scatter-line-series-module';
import { IgxScatterLineSeriesComponent } from './igx-scatter-line-series-component';
import { ScatterLineSeries } from './ScatterLineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxScatterLineSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxScatterLineSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxScatterLineSeriesComponent", IgxScatterLineSeriesComponent);
        TypeRegistrar.register("ScatterLineSeries", ScatterLineSeries.$type);
    }
    IgxScatterLineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterLineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterLineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterLineSeriesDynamicModule, imports: [CommonModule, IgxScatterLineSeriesModule], exports: [IgxScatterLineSeriesModule] });
    IgxScatterLineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterLineSeriesDynamicModule, imports: [[CommonModule, IgxScatterLineSeriesModule], IgxScatterLineSeriesModule] });
    return IgxScatterLineSeriesDynamicModule;
}());
export { IgxScatterLineSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterLineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxScatterLineSeriesModule],
                    exports: [IgxScatterLineSeriesModule],
                    entryComponents: [IgxScatterLineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
