// igniteui-angular-charts/esm5/lib/igx-value-overlay-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxValueOverlayModule } from './igx-value-overlay-module';
import { IgxValueOverlayComponent } from './igx-value-overlay-component';
import { ValueOverlay } from './ValueOverlay';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxValueOverlayDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxValueOverlayDynamicModule() {
        TypeRegistrar.registerCons("IgxValueOverlayComponent", IgxValueOverlayComponent);
        TypeRegistrar.register("ValueOverlay", ValueOverlay.$type);
    }
    IgxValueOverlayDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueOverlayDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxValueOverlayDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueOverlayDynamicModule, imports: [CommonModule, IgxValueOverlayModule], exports: [IgxValueOverlayModule] });
    IgxValueOverlayDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueOverlayDynamicModule, imports: [[CommonModule, IgxValueOverlayModule], IgxValueOverlayModule] });
    return IgxValueOverlayDynamicModule;
}());
export { IgxValueOverlayDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueOverlayDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxValueOverlayModule],
                    exports: [IgxValueOverlayModule],
                    entryComponents: [IgxValueOverlayComponent]
                }]
        }], ctorParameters: function () { return []; } });
