// igniteui-angular-charts/esm5/lib/igx-on-balance-volume-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxOnBalanceVolumeIndicatorModule } from './igx-on-balance-volume-indicator-module';
import { IgxOnBalanceVolumeIndicatorComponent } from './igx-on-balance-volume-indicator-component';
import { OnBalanceVolumeIndicator } from './OnBalanceVolumeIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxOnBalanceVolumeIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxOnBalanceVolumeIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxOnBalanceVolumeIndicatorComponent", IgxOnBalanceVolumeIndicatorComponent);
        TypeRegistrar.register("OnBalanceVolumeIndicator", OnBalanceVolumeIndicator.$type);
    }
    IgxOnBalanceVolumeIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOnBalanceVolumeIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxOnBalanceVolumeIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOnBalanceVolumeIndicatorDynamicModule, imports: [CommonModule, IgxOnBalanceVolumeIndicatorModule], exports: [IgxOnBalanceVolumeIndicatorModule] });
    IgxOnBalanceVolumeIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOnBalanceVolumeIndicatorDynamicModule, imports: [[CommonModule, IgxOnBalanceVolumeIndicatorModule], IgxOnBalanceVolumeIndicatorModule] });
    return IgxOnBalanceVolumeIndicatorDynamicModule;
}());
export { IgxOnBalanceVolumeIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOnBalanceVolumeIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxOnBalanceVolumeIndicatorModule],
                    exports: [IgxOnBalanceVolumeIndicatorModule],
                    entryComponents: [IgxOnBalanceVolumeIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
