// igniteui-angular-charts/esm5/lib/igx-spline-area-fragment-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSplineAreaFragmentModule } from './igx-spline-area-fragment-module';
import { IgxSplineAreaFragmentComponent } from './igx-spline-area-fragment-component';
import { SplineAreaFragment } from './SplineAreaFragment';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxSplineAreaFragmentDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxSplineAreaFragmentDynamicModule() {
        TypeRegistrar.registerCons("IgxSplineAreaFragmentComponent", IgxSplineAreaFragmentComponent);
        TypeRegistrar.register("SplineAreaFragment", SplineAreaFragment.$type);
    }
    IgxSplineAreaFragmentDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineAreaFragmentDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSplineAreaFragmentDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineAreaFragmentDynamicModule, imports: [CommonModule, IgxSplineAreaFragmentModule], exports: [IgxSplineAreaFragmentModule] });
    IgxSplineAreaFragmentDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineAreaFragmentDynamicModule, imports: [[CommonModule, IgxSplineAreaFragmentModule], IgxSplineAreaFragmentModule] });
    return IgxSplineAreaFragmentDynamicModule;
}());
export { IgxSplineAreaFragmentDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineAreaFragmentDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxSplineAreaFragmentModule],
                    exports: [IgxSplineAreaFragmentModule],
                    entryComponents: [IgxSplineAreaFragmentComponent]
                }]
        }], ctorParameters: function () { return []; } });
