// igniteui-angular-charts/esm5/lib/igx-linear-contour-value-resolver-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxLinearContourValueResolverModule } from './igx-linear-contour-value-resolver-module';
import { IgxLinearContourValueResolverComponent } from './igx-linear-contour-value-resolver-component';
import { LinearContourValueResolver } from './LinearContourValueResolver';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxLinearContourValueResolverDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxLinearContourValueResolverDynamicModule() {
        TypeRegistrar.registerCons("IgxLinearContourValueResolverComponent", IgxLinearContourValueResolverComponent);
        TypeRegistrar.register("LinearContourValueResolver", LinearContourValueResolver.$type);
    }
    IgxLinearContourValueResolverDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxLinearContourValueResolverDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverDynamicModule, imports: [CommonModule, IgxLinearContourValueResolverModule], exports: [IgxLinearContourValueResolverModule] });
    IgxLinearContourValueResolverDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverDynamicModule, imports: [[CommonModule, IgxLinearContourValueResolverModule], IgxLinearContourValueResolverModule] });
    return IgxLinearContourValueResolverDynamicModule;
}());
export { IgxLinearContourValueResolverDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxLinearContourValueResolverModule],
                    exports: [IgxLinearContourValueResolverModule],
                    entryComponents: [IgxLinearContourValueResolverComponent]
                }]
        }], ctorParameters: function () { return []; } });
