// igniteui-angular-charts/esm5/lib/igx-data-chart-stacked-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedFragmentSeriesModule } from './igx-stacked-fragment-series-module';
import { IgxStacked100AreaSeriesModule } from './igx-stacked-100-area-series-module';
import { IgxStacked100BarSeriesModule } from './igx-stacked-100-bar-series-module';
import { IgxStacked100ColumnSeriesModule } from './igx-stacked-100-column-series-module';
import { IgxStacked100LineSeriesModule } from './igx-stacked-100-line-series-module';
import { IgxStacked100SplineSeriesModule } from './igx-stacked-100-spline-series-module';
import { IgxStacked100SplineAreaSeriesModule } from './igx-stacked-100-spline-area-series-module';
import { IgxStackedAreaSeriesModule } from './igx-stacked-area-series-module';
import { IgxStackedBarSeriesModule } from './igx-stacked-bar-series-module';
import { IgxStackedColumnSeriesModule } from './igx-stacked-column-series-module';
import { IgxStackedLineSeriesModule } from './igx-stacked-line-series-module';
import { IgxStackedSplineSeriesModule } from './igx-stacked-spline-series-module';
import { IgxStackedSplineAreaSeriesModule } from './igx-stacked-spline-area-series-module';
import { IgxAreaFragmentDynamicModule } from './igx-area-fragment-dynamic-module';
import { IgxBarFragmentDynamicModule } from './igx-bar-fragment-dynamic-module';
import { IgxColumnFragmentDynamicModule } from './igx-column-fragment-dynamic-module';
import { IgxLineFragmentDynamicModule } from './igx-line-fragment-dynamic-module';
import { IgxSplineFragmentDynamicModule } from './igx-spline-fragment-dynamic-module';
import { IgxSplineAreaFragmentDynamicModule } from './igx-spline-area-fragment-dynamic-module';
import { IgxDataChartCategoryCoreModule } from './igx-data-chart-category-core-module';
import { IgxDataChartVerticalCategoryCoreModule } from './igx-data-chart-vertical-category-core-module';
import * as i0 from "@angular/core";
var IgxDataChartStackedModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartStackedModule() {
    }
    IgxDataChartStackedModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartStackedModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartStackedModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartStackedModule, imports: [CommonModule, IgxStackedFragmentSeriesModule,
            IgxStacked100AreaSeriesModule,
            IgxStacked100BarSeriesModule,
            IgxStacked100ColumnSeriesModule,
            IgxStacked100LineSeriesModule,
            IgxStacked100SplineSeriesModule,
            IgxStacked100SplineAreaSeriesModule,
            IgxStackedAreaSeriesModule,
            IgxStackedBarSeriesModule,
            IgxStackedColumnSeriesModule,
            IgxStackedLineSeriesModule,
            IgxStackedSplineSeriesModule,
            IgxStackedSplineAreaSeriesModule,
            IgxAreaFragmentDynamicModule,
            IgxBarFragmentDynamicModule,
            IgxColumnFragmentDynamicModule,
            IgxLineFragmentDynamicModule,
            IgxSplineFragmentDynamicModule,
            IgxSplineAreaFragmentDynamicModule,
            IgxDataChartCategoryCoreModule,
            IgxDataChartVerticalCategoryCoreModule], exports: [IgxStackedFragmentSeriesModule,
            IgxStacked100AreaSeriesModule,
            IgxStacked100BarSeriesModule,
            IgxStacked100ColumnSeriesModule,
            IgxStacked100LineSeriesModule,
            IgxStacked100SplineSeriesModule,
            IgxStacked100SplineAreaSeriesModule,
            IgxStackedAreaSeriesModule,
            IgxStackedBarSeriesModule,
            IgxStackedColumnSeriesModule,
            IgxStackedLineSeriesModule,
            IgxStackedSplineSeriesModule,
            IgxStackedSplineAreaSeriesModule,
            IgxAreaFragmentDynamicModule,
            IgxBarFragmentDynamicModule,
            IgxColumnFragmentDynamicModule,
            IgxLineFragmentDynamicModule,
            IgxSplineFragmentDynamicModule,
            IgxSplineAreaFragmentDynamicModule,
            IgxDataChartCategoryCoreModule,
            IgxDataChartVerticalCategoryCoreModule] });
    IgxDataChartStackedModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartStackedModule, imports: [[CommonModule, IgxStackedFragmentSeriesModule,
                IgxStacked100AreaSeriesModule,
                IgxStacked100BarSeriesModule,
                IgxStacked100ColumnSeriesModule,
                IgxStacked100LineSeriesModule,
                IgxStacked100SplineSeriesModule,
                IgxStacked100SplineAreaSeriesModule,
                IgxStackedAreaSeriesModule,
                IgxStackedBarSeriesModule,
                IgxStackedColumnSeriesModule,
                IgxStackedLineSeriesModule,
                IgxStackedSplineSeriesModule,
                IgxStackedSplineAreaSeriesModule,
                IgxAreaFragmentDynamicModule,
                IgxBarFragmentDynamicModule,
                IgxColumnFragmentDynamicModule,
                IgxLineFragmentDynamicModule,
                IgxSplineFragmentDynamicModule,
                IgxSplineAreaFragmentDynamicModule,
                IgxDataChartCategoryCoreModule,
                IgxDataChartVerticalCategoryCoreModule
            ], IgxStackedFragmentSeriesModule,
            IgxStacked100AreaSeriesModule,
            IgxStacked100BarSeriesModule,
            IgxStacked100ColumnSeriesModule,
            IgxStacked100LineSeriesModule,
            IgxStacked100SplineSeriesModule,
            IgxStacked100SplineAreaSeriesModule,
            IgxStackedAreaSeriesModule,
            IgxStackedBarSeriesModule,
            IgxStackedColumnSeriesModule,
            IgxStackedLineSeriesModule,
            IgxStackedSplineSeriesModule,
            IgxStackedSplineAreaSeriesModule,
            IgxAreaFragmentDynamicModule,
            IgxBarFragmentDynamicModule,
            IgxColumnFragmentDynamicModule,
            IgxLineFragmentDynamicModule,
            IgxSplineFragmentDynamicModule,
            IgxSplineAreaFragmentDynamicModule,
            IgxDataChartCategoryCoreModule,
            IgxDataChartVerticalCategoryCoreModule] });
    return IgxDataChartStackedModule;
}());
export { IgxDataChartStackedModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartStackedModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxStackedFragmentSeriesModule,
                        IgxStacked100AreaSeriesModule,
                        IgxStacked100BarSeriesModule,
                        IgxStacked100ColumnSeriesModule,
                        IgxStacked100LineSeriesModule,
                        IgxStacked100SplineSeriesModule,
                        IgxStacked100SplineAreaSeriesModule,
                        IgxStackedAreaSeriesModule,
                        IgxStackedBarSeriesModule,
                        IgxStackedColumnSeriesModule,
                        IgxStackedLineSeriesModule,
                        IgxStackedSplineSeriesModule,
                        IgxStackedSplineAreaSeriesModule,
                        IgxAreaFragmentDynamicModule,
                        IgxBarFragmentDynamicModule,
                        IgxColumnFragmentDynamicModule,
                        IgxLineFragmentDynamicModule,
                        IgxSplineFragmentDynamicModule,
                        IgxSplineAreaFragmentDynamicModule,
                        IgxDataChartCategoryCoreModule,
                        IgxDataChartVerticalCategoryCoreModule
                    ],
                    imports: [CommonModule, IgxStackedFragmentSeriesModule,
                        IgxStacked100AreaSeriesModule,
                        IgxStacked100BarSeriesModule,
                        IgxStacked100ColumnSeriesModule,
                        IgxStacked100LineSeriesModule,
                        IgxStacked100SplineSeriesModule,
                        IgxStacked100SplineAreaSeriesModule,
                        IgxStackedAreaSeriesModule,
                        IgxStackedBarSeriesModule,
                        IgxStackedColumnSeriesModule,
                        IgxStackedLineSeriesModule,
                        IgxStackedSplineSeriesModule,
                        IgxStackedSplineAreaSeriesModule,
                        IgxAreaFragmentDynamicModule,
                        IgxBarFragmentDynamicModule,
                        IgxColumnFragmentDynamicModule,
                        IgxLineFragmentDynamicModule,
                        IgxSplineFragmentDynamicModule,
                        IgxSplineAreaFragmentDynamicModule,
                        IgxDataChartCategoryCoreModule,
                        IgxDataChartVerticalCategoryCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
