// igniteui-angular-charts/esm5/lib/igx-market-facilitation-index-indicator-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxItemwiseStrategyBasedIndicatorComponent } from "./igx-itemwise-strategy-based-indicator-component";
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import { IgxFinancialIndicatorComponent } from "./igx-financial-indicator-component";
import { IgxFinancialSeriesComponent } from "./igx-financial-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { MarketFacilitationIndexIndicator } from "./MarketFacilitationIndexIndicator";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent Market Facilitation Index indicator series.
 * The Market Facilitation Index (MFI) indicator is represented by
 * the difference of the high and the low for a bar divided by the volume.
 * Default required members: High, Low, Volume
*
* The Market Facilitation Index (MFI) indicator is represented by the difference of the high and the low for a bar divided by the volume.
*
* Example:
*
* ```html
* <igx-data-chart
*     [dataSource]="data">
* 	<igx-category-x-axis
* 		label="label"
* 		#xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
* 		#yAxis>
*     </igx-numeric-y-axis>
* 	<igx-market-facilitation-index-indicator
*         #series
*         [xAxis]="xAxis"
*         [yAxis]="yAxis"
*         highMemberPath="high"
* 		lowMemberPath="low"
* 		volumeMemberPath="volume">
*     </igx-market-facilitation-index-indicator>
* </igx-data-chart>
* ```
*/
var IgxMarketFacilitationIndexIndicatorComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxMarketFacilitationIndexIndicatorComponent, _super);
    function IgxMarketFacilitationIndexIndicatorComponent() {
        return _super.call(this) || this;
    }
    IgxMarketFacilitationIndexIndicatorComponent.prototype.createImplementation = function () {
        return new MarketFacilitationIndexIndicator();
    };
    Object.defineProperty(IgxMarketFacilitationIndexIndicatorComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxMarketFacilitationIndexIndicatorComponent.prototype, "defaultDisplayType", {
        /**
         * Gets default display type for the current Financial Indicator
        */
        get: function () {
            return this.i.aau;
        },
        enumerable: false,
        configurable: true
    });
    IgxMarketFacilitationIndexIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMarketFacilitationIndexIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxMarketFacilitationIndexIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxMarketFacilitationIndexIndicatorComponent, selector: "igx-market-facilitation-index-indicator", providers: [{ provide: IgxItemwiseStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxMarketFacilitationIndexIndicatorComponent; }) }, { provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxMarketFacilitationIndexIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxMarketFacilitationIndexIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxMarketFacilitationIndexIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxMarketFacilitationIndexIndicatorComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxMarketFacilitationIndexIndicatorComponent;
}(IgxItemwiseStrategyBasedIndicatorComponent));
export { IgxMarketFacilitationIndexIndicatorComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMarketFacilitationIndexIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-market-facilitation-index-indicator',
                    template: "",
                    providers: [{ provide: IgxItemwiseStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxMarketFacilitationIndexIndicatorComponent; }) }, { provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxMarketFacilitationIndexIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxMarketFacilitationIndexIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxMarketFacilitationIndexIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxMarketFacilitationIndexIndicatorComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
