// igniteui-angular-charts/esm5/lib/igx-stacked-line-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedLineSeriesComponent } from './igx-stacked-line-series-component';
import * as i0 from "@angular/core";
var IgxStackedLineSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStackedLineSeriesModule() {
    }
    IgxStackedLineSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedLineSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedLineSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedLineSeriesModule, declarations: [IgxStackedLineSeriesComponent], imports: [CommonModule], exports: [IgxStackedLineSeriesComponent] });
    IgxStackedLineSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedLineSeriesModule, imports: [[CommonModule]] });
    return IgxStackedLineSeriesModule;
}());
export { IgxStackedLineSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedLineSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStackedLineSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxStackedLineSeriesComponent],
                    entryComponents: []
                }]
        }] });
