// igniteui-angular-charts/esm5/lib/igx-data-chart-polar-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericAngleAxisModule } from './igx-numeric-angle-axis-module';
import { IgxNumericRadiusAxisModule } from './igx-numeric-radius-axis-module';
import { IgxDataChartCoreModule } from './igx-data-chart-core-module';
import * as i0 from "@angular/core";
var IgxDataChartPolarCoreModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartPolarCoreModule() {
    }
    IgxDataChartPolarCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartPolarCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarCoreModule, imports: [CommonModule, IgxNumericAngleAxisModule,
            IgxNumericRadiusAxisModule,
            IgxDataChartCoreModule], exports: [IgxNumericAngleAxisModule,
            IgxNumericRadiusAxisModule,
            IgxDataChartCoreModule] });
    IgxDataChartPolarCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarCoreModule, imports: [[CommonModule, IgxNumericAngleAxisModule,
                IgxNumericRadiusAxisModule,
                IgxDataChartCoreModule
            ], IgxNumericAngleAxisModule,
            IgxNumericRadiusAxisModule,
            IgxDataChartCoreModule] });
    return IgxDataChartPolarCoreModule;
}());
export { IgxDataChartPolarCoreModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumericAngleAxisModule,
                        IgxNumericRadiusAxisModule,
                        IgxDataChartCoreModule
                    ],
                    imports: [CommonModule, IgxNumericAngleAxisModule,
                        IgxNumericRadiusAxisModule,
                        IgxDataChartCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
