// igniteui-angular-charts/esm5/lib/igx-fast-stochastic-oscillator-indicator-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import { IgxFinancialIndicatorComponent } from "./igx-financial-indicator-component";
import { IgxFinancialSeriesComponent } from "./igx-financial-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { FastStochasticOscillatorIndicator } from "./FastStochasticOscillatorIndicator";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent Fast Stochastic Oscillator indicator series.
 * Default required members: High, Low, Close
*/
var IgxFastStochasticOscillatorIndicatorComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxFastStochasticOscillatorIndicatorComponent, _super);
    function IgxFastStochasticOscillatorIndicatorComponent() {
        return _super.call(this) || this;
    }
    IgxFastStochasticOscillatorIndicatorComponent.prototype.createImplementation = function () {
        return new FastStochasticOscillatorIndicator();
    };
    Object.defineProperty(IgxFastStochasticOscillatorIndicatorComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxFastStochasticOscillatorIndicatorComponent.prototype, "period", {
        /**
         * Gets or sets the moving average period for the current FastStochasticOscillatorIndicator object.
         * The typical, and initial, value for FastStochasticOscillatorIndicator periods is 14.
        */
        get: function () {
            return this.i.ab8;
        },
        set: function (v) {
            this.i.ab8 = +v;
        },
        enumerable: false,
        configurable: true
    });
    IgxFastStochasticOscillatorIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFastStochasticOscillatorIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxFastStochasticOscillatorIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxFastStochasticOscillatorIndicatorComponent, selector: "igx-fast-stochastic-oscillator-indicator", inputs: { period: "period" }, providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxFastStochasticOscillatorIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxFastStochasticOscillatorIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxFastStochasticOscillatorIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxFastStochasticOscillatorIndicatorComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxFastStochasticOscillatorIndicatorComponent;
}(IgxStrategyBasedIndicatorComponent));
export { IgxFastStochasticOscillatorIndicatorComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFastStochasticOscillatorIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-fast-stochastic-oscillator-indicator',
                    template: "",
                    providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxFastStochasticOscillatorIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxFastStochasticOscillatorIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxFastStochasticOscillatorIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxFastStochasticOscillatorIndicatorComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { period: [{
                type: Input
            }] } });
