// igniteui-angular-charts/esm5/lib/igx-callout-layer-component.js
import { __extends } from "tslib";
import { EventEmitter, Output, Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { delegateCombine } from "igniteui-angular-core";
import { IgxSeriesComponent } from "./igx-series-component";
import { IgxCalloutPlacementPositionsCollection } from "./igx-callout-placement-positions-collection";
import { HighlightedValueLabelMode_$type } from "igniteui-angular-core";
import { AutoCalloutVisibilityMode_$type } from "./AutoCalloutVisibilityMode";
import { CalloutCollisionMode_$type } from "igniteui-angular-core";
import { IgxCalloutStyleUpdatingEventArgs } from "./igx-callout-style-updating-event-args";
import { IgxCalloutRenderStyleUpdatingEventArgs } from "./igx-callout-render-style-updating-event-args";
import { IgxCalloutContentUpdatingEventArgs } from "./igx-callout-content-updating-event-args";
import { IgxCalloutLabelUpdatingEventArgs } from "./igx-callout-label-updating-event-args";
import { IgxCalloutSeriesSelectingEventArgs } from "./igx-callout-series-selecting-event-args";
import { IgxAnnotationLayerComponent } from "./igx-annotation-layer-component";
import { CalloutLayer } from "./CalloutLayer";
import { ensureBool, ensureEnum, brushToString, stringToBrush, arrayFindByName } from "igniteui-angular-core";
import { CalloutPlacementPositionsCollection as CalloutPlacementPositionsCollection_internal } from "./CalloutPlacementPositionsCollection";
import { CalloutPlacementPositions_$type } from "./CalloutPlacementPositions";
import { SyncableObservableCollection$1 } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents an annotation layer that displays crosshair lines that cross through the closest value of the target series under the cursor.
*/
var IgxCalloutLayerComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxCalloutLayerComponent, _super);
    function IgxCalloutLayerComponent() {
        var _this = _super.call(this) || this;
        _this._allowedPositions = null;
        _this._calloutStyleUpdating = null;
        _this._calloutRenderStyleUpdating = null;
        _this._calloutContentUpdating = null;
        _this._calloutLabelUpdating = null;
        _this._calloutSeriesSelecting = null;
        return _this;
    }
    IgxCalloutLayerComponent.prototype.createImplementation = function () {
        return new CalloutLayer();
    };
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "isAnnotationCalloutLayer", {
        /**
         * Gets whether the series is an callout annotation layer.
        */
        get: function () {
            return this.i.em;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "labelMemberPath", {
        /**
         * Gets or sets the label mapping property for the callouts.
        */
        get: function () {
            return this.i.acr;
        },
        set: function (v) {
            this.i.acr = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "xMemberPath", {
        /**
         * Gets or sets the label mapping property for the callouts.
        */
        get: function () {
            return this.i.ac5;
        },
        set: function (v) {
            this.i.ac5 = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "yMemberPath", {
        /**
         * Gets or sets the label mapping property for the callouts.
        */
        get: function () {
            return this.i.ac8;
        },
        set: function (v) {
            this.i.ac8 = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "keyMemberPath", {
        /**
         * Gets or sets the key mapping property for the callouts.
        */
        get: function () {
            return this.i.acl;
        },
        set: function (v) {
            this.i.acl = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "isCalloutOffsettingEnabled", {
        /**
         * Gets or sets whether to allow the callouts to be variable distances from the target points, for suppporting collision modes.
        */
        get: function () {
            return this.i.aaw;
        },
        set: function (v) {
            this.i.aaw = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "labelFormat", {
        /**
         * Gets or sets the label format string to use for the label.
        */
        get: function () {
            return this.i.aco;
        },
        set: function (v) {
            this.i.aco = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "labelFormatSpecifiers", {
        /**
         * Gets or sets the format specifiers to use with the LabelFormat string.
        */
        get: function () {
            return this.i.zq;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.zq = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "contentMemberPath", {
        /**
         * Gets or sets the content mapping property for the callouts.
        */
        get: function () {
            return this.i.acd;
        },
        set: function (v) {
            this.i.acd = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "targetSeriesName", {
        /**
         * Gets or sets the name of the series series to target this annotation to. If null, this annotation targets all series simultaneously.
        */
        get: function () {
            return this.i.act;
        },
        set: function (v) {
            this.i.act = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "targetSeries", {
        /**
         * Gets or sets the series to target this annotation to. If null, this annotation targets all series simultaneously.
        */
        get: function () {
            var r = this.i.z7;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgxSeriesComponent._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            if (v != null && this._stylingContainer && v._styling)
                v._styling(this._stylingContainer, this, this);
            v == null ? this.i.z7 = null : this.i.z7 = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "collisionChannel", {
        /**
         * Gets or sets a channel name to identify other callout layers within the chart to share collision information with.
        */
        get: function () {
            return this.i.aca;
        },
        set: function (v) {
            this.i.aca = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "allowedPositions", {
        /**
         * Gets or sets the allowed positions that the callout layer is allowed to use.
        */
        get: function () {
            if (this._allowedPositions === null) {
                var coll = new IgxCalloutPlacementPositionsCollection();
                var innerColl = this.i.z2;
                if (!innerColl) {
                    innerColl = new CalloutPlacementPositionsCollection_internal();
                }
                this._allowedPositions = coll._fromInner(innerColl);
                this.i.z2 = innerColl;
            }
            return this._allowedPositions;
        },
        set: function (v) {
            if (this._allowedPositions !== null) {
                this._allowedPositions._setSyncTarget(null);
                this._allowedPositions = null;
            }
            var coll = new IgxCalloutPlacementPositionsCollection();
            this._allowedPositions = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(CalloutPlacementPositions_$type);
            var innerColl = this.i.z2;
            if (!innerColl) {
                innerColl = new CalloutPlacementPositionsCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._allowedPositions._setSyncTarget(syncColl);
            this.i.z2 = innerColl;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "isAutoCalloutBehaviorEnabled", {
        /**
         * Gets or sets how to select the final value to annotate.
        */
        get: function () {
            return this.i.aav;
        },
        set: function (v) {
            this.i.aav = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "useInterpolatedValueForAutoCalloutLabels", {
        /**
         * Gets or sets whether to use values, rather than labels for auto callout labels.
        */
        get: function () {
            return this.i.aa0;
        },
        set: function (v) {
            this.i.aa0 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "useSeriesColorForOutline", {
        /**
         * Gets or sets whether to the series color as the outline of the callout.
        */
        get: function () {
            return this.i.aa3;
        },
        set: function (v) {
            this.i.aa3 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "useAutoContrastingLabelColors", {
        /**
         * Gets or sets whether to the label color is automatically decided between a light and dark option for contrast.
        */
        get: function () {
            return this.i.aaz;
        },
        set: function (v) {
            this.i.aaz = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "useItemColorForFill", {
        /**
         * Gets or sets whether to the series color as the outline of the callout.
        */
        get: function () {
            return this.i.aa1;
        },
        set: function (v) {
            this.i.aa1 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "useItemColorForOutline", {
        /**
         * Gets or sets whether to the series item color as the outline of the callout.
        */
        get: function () {
            return this.i.aa2;
        },
        set: function (v) {
            this.i.aa2 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "useValueForAutoCalloutLabels", {
        /**
         * Gets or sets whether to use values, rather than labels for auto callout labels.
        */
        get: function () {
            return this.i.aa4;
        },
        set: function (v) {
            this.i.aa4 = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "highlightedValueLabelMode", {
        /**
         * Gets or sets whether to use values, rather than labels for auto callout labels.
        */
        get: function () {
            return this.i.z6;
        },
        set: function (v) {
            this.i.z6 = ensureEnum(HighlightedValueLabelMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "autoCalloutVisibilityMode", {
        /**
         * Gets or sets the rules for when the callouts should be visible.
        */
        get: function () {
            return this.i.zv;
        },
        set: function (v) {
            this.i.zv = ensureEnum(AutoCalloutVisibilityMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutCornerRadius", {
        /**
         * Gets or sets the corner radius to use for the callout borders.
        */
        get: function () {
            return this.i.abe;
        },
        set: function (v) {
            this.i.abe = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "isCustomCalloutStyleEnabled", {
        /**
         * Gets or sets whether to allow for custom callout styles.
        */
        get: function () {
            return this.i.aay;
        },
        set: function (v) {
            this.i.aay = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "isCustomCalloutRenderStyleEnabled", {
        /**
         * Gets or sets whether to allow for custom callout styles.
        */
        get: function () {
            return this.i.aax;
        },
        set: function (v) {
            this.i.aax = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "autoCalloutLabelPrecision", {
        /**
         * Gets or sets the precision to use when values are used for the auto callout labels.
        */
        get: function () {
            return this.i.abn;
        },
        set: function (v) {
            this.i.abn = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutTextColor", {
        /**
         * Gets or sets the color to use for the callout text. Leave unset for an automatic value.
        */
        get: function () {
            return brushToString(this.i.aey);
        },
        set: function (v) {
            this.i.aey = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutLightTextColor", {
        /**
         * Gets or sets the light color to use for the callout text. Leave unset for an automatic value.
        */
        get: function () {
            return brushToString(this.i.aew);
        },
        set: function (v) {
            this.i.aew = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutDarkTextColor", {
        /**
         * Gets or sets the dark color to use for the callout text. Leave unset for an automatic value.
        */
        get: function () {
            return brushToString(this.i.aeu);
        },
        set: function (v) {
            this.i.aeu = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutBackground", {
        /**
         * Gets or sets the brush that specifies how the backgrounds for the callouts of the layer are painted. Leave unset for an automatic value.
        */
        get: function () {
            return brushToString(this.i.aet);
        },
        set: function (v) {
            this.i.aet = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutInterpolatedValuePrecision", {
        /**
         * Gets or sets the precision to use displaying values for interpolated crosshair positions.
        */
        get: function () {
            return this.i.abo;
        },
        set: function (v) {
            this.i.abo = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutOutline", {
        /**
         * Gets or sets the brush that specifies how the outlines for the callouts of the layer are painted.
        */
        get: function () {
            return brushToString(this.i.aex);
        },
        set: function (v) {
            this.i.aex = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutLeaderBrush", {
        /**
         * Gets or sets the brush that specifies how the leader lines for the callouts of the layer are painted.
        */
        get: function () {
            return brushToString(this.i.aev);
        },
        set: function (v) {
            this.i.aev = stringToBrush(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutPaddingLeft", {
        /**
         * Gets or sets the left padding to use withing the callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.abg;
        },
        set: function (v) {
            this.i.abg = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutPaddingTop", {
        /**
         * Gets or sets the top padding to use within the callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.abi;
        },
        set: function (v) {
            this.i.abi = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutPaddingRight", {
        /**
         * Gets or sets the right padding to use within the callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.abh;
        },
        set: function (v) {
            this.i.abh = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutPaddingBottom", {
        /**
         * Gets or sets the bottom padding to use within the callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get: function () {
            return this.i.abf;
        },
        set: function (v) {
            this.i.abf = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutCollisionMode", {
        /**
         * Gets or sets the strategy to use for avoiding collisions between the callouts in this layer. Leave unset for an automatic value.
        */
        get: function () {
            return this.i.zx;
        },
        set: function (v) {
            this.i.zx = ensureEnum(CalloutCollisionMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutPositionPadding", {
        /**
         * Gets or sets the padding to add to the callout positioning. Leave unset for an automatic value.
        */
        get: function () {
            return this.i.abj;
        },
        set: function (v) {
            this.i.abj = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutStrokeThickness", {
        /**
         * Gets or sets the stroke thickness for the callout backing. Leave unset for an automatic value.
        */
        get: function () {
            return this.i.abk;
        },
        set: function (v) {
            this.i.abk = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "textStyle", {
        /**
         * Gets or sets the stroke thickness for the axis annotation backing. Leave unset for an automatic value.
        */
        get: function () {
            return this.i.acw;
        },
        set: function (v) {
            this.i.acw = v;
        },
        enumerable: false,
        configurable: true
    });
    IgxCalloutLayerComponent.prototype.findByName = function (name) {
        var baseResult = _super.prototype.findByName.call(this, name);
        if (baseResult) {
            return baseResult;
        }
        if (this.labelFormatSpecifiers != null && arrayFindByName(this.labelFormatSpecifiers, name)) {
            return arrayFindByName(this.labelFormatSpecifiers, name);
        }
        if (this.targetSeries && this.targetSeries.name && this.targetSeries.name == name) {
            return this.targetSeries;
        }
        if (this.allowedPositions != null && this.allowedPositions.findByName && this.allowedPositions.findByName(name)) {
            return this.allowedPositions.findByName(name);
        }
        return null;
    };
    IgxCalloutLayerComponent.prototype._styling = function (container, component, parent) {
        _super.prototype._styling.call(this, container, component, parent);
        this._inStyling = true;
        if (this.targetSeries && this.targetSeries._styling) {
            this.targetSeries._styling(container, component, this);
        }
        this._inStyling = false;
    };
    IgxCalloutLayerComponent.prototype.getItemValue = function (item, memberPathName) {
        var iv = this.i.ko(item, memberPathName);
        return (iv);
    };
    /**
     * Gets the value of a requested member path from the series.
    
    * @param memberPathName  * The property name of a valid member path for the series
    */
    IgxCalloutLayerComponent.prototype.getMemberPathValue = function (memberPathName) {
        var iv = this.i.mb(memberPathName);
        return (iv);
    };
    /**
     * Invalidates content of callout layer
    
    */
    IgxCalloutLayerComponent.prototype.invalidateCalloutContent = function () {
        this.i.adi();
    };
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutStyleUpdating", {
        /**
         * Event raised when updating style of callout layer
        */
        get: function () {
            var _this = this;
            if (this._calloutStyleUpdating == null) {
                this._calloutStyleUpdating = new EventEmitter();
                this.i.calloutStyleUpdating = delegateCombine(this.i.calloutStyleUpdating, function (o, e) {
                    _this._runInZone(function () {
                        var outerArgs = new IgxCalloutStyleUpdatingEventArgs();
                        outerArgs._provideImplementation(e);
                        if (_this.beforeCalloutStyleUpdating) {
                            _this.beforeCalloutStyleUpdating(_this, outerArgs);
                        }
                        _this._calloutStyleUpdating.emit({
                            sender: _this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._calloutStyleUpdating;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutRenderStyleUpdating", {
        /**
         * Event raised when updating style of callout layer
        */
        get: function () {
            var _this = this;
            if (this._calloutRenderStyleUpdating == null) {
                this._calloutRenderStyleUpdating = new EventEmitter();
                this.i.calloutRenderStyleUpdating = delegateCombine(this.i.calloutRenderStyleUpdating, function (o, e) {
                    _this._runInZone(function () {
                        var outerArgs = new IgxCalloutRenderStyleUpdatingEventArgs();
                        outerArgs._provideImplementation(e);
                        if (_this.beforeCalloutRenderStyleUpdating) {
                            _this.beforeCalloutRenderStyleUpdating(_this, outerArgs);
                        }
                        _this._calloutRenderStyleUpdating.emit({
                            sender: _this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._calloutRenderStyleUpdating;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutContentUpdating", {
        /**
         * Raised to allow you to decide on the content for an automatically created callout.
        */
        get: function () {
            var _this = this;
            if (this._calloutContentUpdating == null) {
                this._calloutContentUpdating = new EventEmitter();
                this.i.calloutContentUpdating = delegateCombine(this.i.calloutContentUpdating, function (o, e) {
                    _this._runInZone(function () {
                        var outerArgs = new IgxCalloutContentUpdatingEventArgs();
                        outerArgs._provideImplementation(e);
                        if (_this.beforeCalloutContentUpdating) {
                            _this.beforeCalloutContentUpdating(_this, outerArgs);
                        }
                        _this._calloutContentUpdating.emit({
                            sender: _this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._calloutContentUpdating;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutLabelUpdating", {
        /**
         * Raised to allow you to decide on the label for an automatically created callout.
        */
        get: function () {
            var _this = this;
            if (this._calloutLabelUpdating == null) {
                this._calloutLabelUpdating = new EventEmitter();
                this.i.calloutLabelUpdating = delegateCombine(this.i.calloutLabelUpdating, function (o, e) {
                    _this._runInZone(function () {
                        var outerArgs = new IgxCalloutLabelUpdatingEventArgs();
                        outerArgs._provideImplementation(e);
                        if (_this.beforeCalloutLabelUpdating) {
                            _this.beforeCalloutLabelUpdating(_this, outerArgs);
                        }
                        _this._calloutLabelUpdating.emit({
                            sender: _this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._calloutLabelUpdating;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxCalloutLayerComponent.prototype, "calloutSeriesSelecting", {
        /**
         * Raised to allow you to select a target series for a data bound callout.
        */
        get: function () {
            var _this = this;
            if (this._calloutSeriesSelecting == null) {
                this._calloutSeriesSelecting = new EventEmitter();
                this.i.calloutSeriesSelecting = delegateCombine(this.i.calloutSeriesSelecting, function (o, e) {
                    _this._runInZone(function () {
                        var outerArgs = new IgxCalloutSeriesSelectingEventArgs();
                        outerArgs._provideImplementation(e);
                        if (_this.beforeCalloutSeriesSelecting) {
                            _this.beforeCalloutSeriesSelecting(_this, outerArgs);
                        }
                        _this._calloutSeriesSelecting.emit({
                            sender: _this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._calloutSeriesSelecting;
        },
        enumerable: false,
        configurable: true
    });
    IgxCalloutLayerComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCalloutLayerComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxCalloutLayerComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxCalloutLayerComponent, selector: "igx-callout-layer", inputs: { labelMemberPath: "labelMemberPath", xMemberPath: "xMemberPath", yMemberPath: "yMemberPath", keyMemberPath: "keyMemberPath", isCalloutOffsettingEnabled: "isCalloutOffsettingEnabled", labelFormat: "labelFormat", labelFormatSpecifiers: "labelFormatSpecifiers", contentMemberPath: "contentMemberPath", targetSeriesName: "targetSeriesName", targetSeries: "targetSeries", collisionChannel: "collisionChannel", allowedPositions: "allowedPositions", isAutoCalloutBehaviorEnabled: "isAutoCalloutBehaviorEnabled", useInterpolatedValueForAutoCalloutLabels: "useInterpolatedValueForAutoCalloutLabels", useSeriesColorForOutline: "useSeriesColorForOutline", useAutoContrastingLabelColors: "useAutoContrastingLabelColors", useItemColorForFill: "useItemColorForFill", useItemColorForOutline: "useItemColorForOutline", useValueForAutoCalloutLabels: "useValueForAutoCalloutLabels", highlightedValueLabelMode: "highlightedValueLabelMode", autoCalloutVisibilityMode: "autoCalloutVisibilityMode", calloutCornerRadius: "calloutCornerRadius", isCustomCalloutStyleEnabled: "isCustomCalloutStyleEnabled", isCustomCalloutRenderStyleEnabled: "isCustomCalloutRenderStyleEnabled", autoCalloutLabelPrecision: "autoCalloutLabelPrecision", calloutTextColor: "calloutTextColor", calloutLightTextColor: "calloutLightTextColor", calloutDarkTextColor: "calloutDarkTextColor", calloutBackground: "calloutBackground", calloutInterpolatedValuePrecision: "calloutInterpolatedValuePrecision", calloutOutline: "calloutOutline", calloutLeaderBrush: "calloutLeaderBrush", calloutPaddingLeft: "calloutPaddingLeft", calloutPaddingTop: "calloutPaddingTop", calloutPaddingRight: "calloutPaddingRight", calloutPaddingBottom: "calloutPaddingBottom", calloutCollisionMode: "calloutCollisionMode", calloutPositionPadding: "calloutPositionPadding", calloutStrokeThickness: "calloutStrokeThickness", textStyle: "textStyle" }, outputs: { calloutStyleUpdating: "calloutStyleUpdating", calloutRenderStyleUpdating: "calloutRenderStyleUpdating", calloutContentUpdating: "calloutContentUpdating", calloutLabelUpdating: "calloutLabelUpdating", calloutSeriesSelecting: "calloutSeriesSelecting" }, providers: [{ provide: IgxAnnotationLayerComponent, useExisting: forwardRef(function () { return IgxCalloutLayerComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxCalloutLayerComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxCalloutLayerComponent;
}(IgxAnnotationLayerComponent));
export { IgxCalloutLayerComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCalloutLayerComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-callout-layer',
                    template: "",
                    providers: [{ provide: IgxAnnotationLayerComponent, useExisting: forwardRef(function () { return IgxCalloutLayerComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxCalloutLayerComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { labelMemberPath: [{
                type: Input
            }], xMemberPath: [{
                type: Input
            }], yMemberPath: [{
                type: Input
            }], keyMemberPath: [{
                type: Input
            }], isCalloutOffsettingEnabled: [{
                type: Input
            }], labelFormat: [{
                type: Input
            }], labelFormatSpecifiers: [{
                type: Input
            }], contentMemberPath: [{
                type: Input
            }], targetSeriesName: [{
                type: Input
            }], targetSeries: [{
                type: Input
            }], collisionChannel: [{
                type: Input
            }], allowedPositions: [{
                type: Input
            }], isAutoCalloutBehaviorEnabled: [{
                type: Input
            }], useInterpolatedValueForAutoCalloutLabels: [{
                type: Input
            }], useSeriesColorForOutline: [{
                type: Input
            }], useAutoContrastingLabelColors: [{
                type: Input
            }], useItemColorForFill: [{
                type: Input
            }], useItemColorForOutline: [{
                type: Input
            }], useValueForAutoCalloutLabels: [{
                type: Input
            }], highlightedValueLabelMode: [{
                type: Input
            }], autoCalloutVisibilityMode: [{
                type: Input
            }], calloutCornerRadius: [{
                type: Input
            }], isCustomCalloutStyleEnabled: [{
                type: Input
            }], isCustomCalloutRenderStyleEnabled: [{
                type: Input
            }], autoCalloutLabelPrecision: [{
                type: Input
            }], calloutTextColor: [{
                type: Input
            }], calloutLightTextColor: [{
                type: Input
            }], calloutDarkTextColor: [{
                type: Input
            }], calloutBackground: [{
                type: Input
            }], calloutInterpolatedValuePrecision: [{
                type: Input
            }], calloutOutline: [{
                type: Input
            }], calloutLeaderBrush: [{
                type: Input
            }], calloutPaddingLeft: [{
                type: Input
            }], calloutPaddingTop: [{
                type: Input
            }], calloutPaddingRight: [{
                type: Input
            }], calloutPaddingBottom: [{
                type: Input
            }], calloutCollisionMode: [{
                type: Input
            }], calloutPositionPadding: [{
                type: Input
            }], calloutStrokeThickness: [{
                type: Input
            }], textStyle: [{
                type: Input
            }], calloutStyleUpdating: [{
                type: Output
            }], calloutRenderStyleUpdating: [{
                type: Output
            }], calloutContentUpdating: [{
                type: Output
            }], calloutLabelUpdating: [{
                type: Output
            }], calloutSeriesSelecting: [{
                type: Output
            }] } });
