// igniteui-angular-charts/esm5/lib/igx-pie-chart-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPieChartCoreModule } from './igx-pie-chart-core-module';
import { IgxDataChartInteractivityModule } from './igx-data-chart-interactivity-module';
import * as i0 from "@angular/core";
var IgxPieChartModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPieChartModule() {
    }
    IgxPieChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPieChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartModule, imports: [CommonModule, IgxPieChartCoreModule,
            IgxDataChartInteractivityModule], exports: [IgxPieChartCoreModule,
            IgxDataChartInteractivityModule] });
    IgxPieChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartModule, imports: [[CommonModule, IgxPieChartCoreModule,
                IgxDataChartInteractivityModule
            ], IgxPieChartCoreModule,
            IgxDataChartInteractivityModule] });
    return IgxPieChartModule;
}());
export { IgxPieChartModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxPieChartCoreModule,
                        IgxDataChartInteractivityModule
                    ],
                    imports: [CommonModule, IgxPieChartCoreModule,
                        IgxDataChartInteractivityModule
                    ],
                    entryComponents: []
                }]
        }] });
