// igniteui-angular-charts/esm5/lib/igx-bubble-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxBubbleSeriesModule } from './igx-bubble-series-module';
import { IgxBubbleSeriesComponent } from './igx-bubble-series-component';
import { BubbleSeries } from './BubbleSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxBubbleSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxBubbleSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxBubbleSeriesComponent", IgxBubbleSeriesComponent);
        TypeRegistrar.register("BubbleSeries", BubbleSeries.$type);
    }
    IgxBubbleSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBubbleSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxBubbleSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBubbleSeriesDynamicModule, imports: [CommonModule, IgxBubbleSeriesModule], exports: [IgxBubbleSeriesModule] });
    IgxBubbleSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBubbleSeriesDynamicModule, imports: [[CommonModule, IgxBubbleSeriesModule], IgxBubbleSeriesModule] });
    return IgxBubbleSeriesDynamicModule;
}());
export { IgxBubbleSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBubbleSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxBubbleSeriesModule],
                    exports: [IgxBubbleSeriesModule],
                    entryComponents: [IgxBubbleSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
