// igniteui-angular-charts/esm5/lib/igx-area-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAreaSeriesModule } from './igx-area-series-module';
import { IgxAreaSeriesComponent } from './igx-area-series-component';
import { AreaSeries } from './AreaSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxAreaSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxAreaSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxAreaSeriesComponent", IgxAreaSeriesComponent);
        TypeRegistrar.register("AreaSeries", AreaSeries.$type);
    }
    IgxAreaSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAreaSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaSeriesDynamicModule, imports: [CommonModule, IgxAreaSeriesModule], exports: [IgxAreaSeriesModule] });
    IgxAreaSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaSeriesDynamicModule, imports: [[CommonModule, IgxAreaSeriesModule], IgxAreaSeriesModule] });
    return IgxAreaSeriesDynamicModule;
}());
export { IgxAreaSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxAreaSeriesModule],
                    exports: [IgxAreaSeriesModule],
                    entryComponents: [IgxAreaSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
