// igniteui-angular-charts/esm5/lib/igx-absolute-volume-oscillator-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAbsoluteVolumeOscillatorIndicatorModule } from './igx-absolute-volume-oscillator-indicator-module';
import { IgxAbsoluteVolumeOscillatorIndicatorComponent } from './igx-absolute-volume-oscillator-indicator-component';
import { AbsoluteVolumeOscillatorIndicator } from './AbsoluteVolumeOscillatorIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxAbsoluteVolumeOscillatorIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxAbsoluteVolumeOscillatorIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxAbsoluteVolumeOscillatorIndicatorComponent", IgxAbsoluteVolumeOscillatorIndicatorComponent);
        TypeRegistrar.register("AbsoluteVolumeOscillatorIndicator", AbsoluteVolumeOscillatorIndicator.$type);
    }
    IgxAbsoluteVolumeOscillatorIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAbsoluteVolumeOscillatorIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAbsoluteVolumeOscillatorIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAbsoluteVolumeOscillatorIndicatorDynamicModule, imports: [CommonModule, IgxAbsoluteVolumeOscillatorIndicatorModule], exports: [IgxAbsoluteVolumeOscillatorIndicatorModule] });
    IgxAbsoluteVolumeOscillatorIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAbsoluteVolumeOscillatorIndicatorDynamicModule, imports: [[CommonModule, IgxAbsoluteVolumeOscillatorIndicatorModule], IgxAbsoluteVolumeOscillatorIndicatorModule] });
    return IgxAbsoluteVolumeOscillatorIndicatorDynamicModule;
}());
export { IgxAbsoluteVolumeOscillatorIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAbsoluteVolumeOscillatorIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxAbsoluteVolumeOscillatorIndicatorModule],
                    exports: [IgxAbsoluteVolumeOscillatorIndicatorModule],
                    entryComponents: [IgxAbsoluteVolumeOscillatorIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
