// igniteui-angular-charts/esm5/lib/igx-doughnut-chart-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxRingSeriesModule } from './igx-ring-series-module';
import { IgxDoughnutChartComponent } from './igx-doughnut-chart-component';
import * as i0 from "@angular/core";
var IgxDoughnutChartCoreModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDoughnutChartCoreModule() {
    }
    IgxDoughnutChartCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDoughnutChartCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartCoreModule, declarations: [IgxDoughnutChartComponent], imports: [CommonModule, IgxRingSeriesModule], exports: [IgxRingSeriesModule,
            IgxDoughnutChartComponent] });
    IgxDoughnutChartCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartCoreModule, imports: [[CommonModule, IgxRingSeriesModule
            ], IgxRingSeriesModule] });
    return IgxDoughnutChartCoreModule;
}());
export { IgxDoughnutChartCoreModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxDoughnutChartComponent],
                    exports: [IgxRingSeriesModule,
                        IgxDoughnutChartComponent
                    ],
                    imports: [CommonModule, IgxRingSeriesModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
