// igniteui-angular-charts/esm5/lib/igx-scale-legend-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScaleLegendComponent } from './igx-scale-legend-component';
import { LegendTemplates } from './LegendTemplates';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxScaleLegendModule } from './igx-scale-legend-module';
import * as i0 from "@angular/core";
var IgxScaleLegendDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxScaleLegendDynamicModule() {
        TypeRegistrar.registerCons('IgxScaleLegendComponent', IgxScaleLegendComponent);
        TypeRegistrar.registerCons('LegendTemplates', LegendTemplates);
        TypeRegistrar.register('LegendTemplates', LegendTemplates.$type);
    }
    IgxScaleLegendDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScaleLegendDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendDynamicModule, imports: [CommonModule, IgxScaleLegendModule], exports: [IgxScaleLegendModule] });
    IgxScaleLegendDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendDynamicModule, imports: [[CommonModule, IgxScaleLegendModule
            ], IgxScaleLegendModule] });
    return IgxScaleLegendDynamicModule;
}());
export { IgxScaleLegendDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxScaleLegendModule
                    ],
                    imports: [CommonModule, IgxScaleLegendModule
                    ],
                    entryComponents: [IgxScaleLegendComponent]
                }]
        }], ctorParameters: function () { return []; } });
