// igniteui-angular-charts/esm5/lib/igx-ring-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxRingSeriesModule } from './igx-ring-series-module';
import { IgxRingSeriesComponent } from './igx-ring-series-component';
import { RingSeries } from './RingSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxRingSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxRingSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxRingSeriesComponent", IgxRingSeriesComponent);
        TypeRegistrar.register("RingSeries", RingSeries.$type);
    }
    IgxRingSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRingSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxRingSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRingSeriesDynamicModule, imports: [CommonModule, IgxRingSeriesModule], exports: [IgxRingSeriesModule] });
    IgxRingSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRingSeriesDynamicModule, imports: [[CommonModule, IgxRingSeriesModule], IgxRingSeriesModule] });
    return IgxRingSeriesDynamicModule;
}());
export { IgxRingSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRingSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxRingSeriesModule],
                    exports: [IgxRingSeriesModule],
                    entryComponents: [IgxRingSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
