// igniteui-angular-charts/esm5/lib/igx-bar-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxBarSeriesComponent } from './igx-bar-series-component';
import * as i0 from "@angular/core";
var IgxBarSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxBarSeriesModule() {
    }
    IgxBarSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxBarSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarSeriesModule, declarations: [IgxBarSeriesComponent], imports: [CommonModule], exports: [IgxBarSeriesComponent] });
    IgxBarSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarSeriesModule, imports: [[CommonModule]] });
    return IgxBarSeriesModule;
}());
export { IgxBarSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxBarSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxBarSeriesComponent],
                    entryComponents: []
                }]
        }] });
