// igniteui-angular-charts/esm5/lib/igx-stoch-rsi-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxStochRSIIndicatorComponent } from './igx-stoch-rsi-indicator-component';
import * as i0 from "@angular/core";
var IgxStochRSIIndicatorModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStochRSIIndicatorModule() {
    }
    IgxStochRSIIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStochRSIIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStochRSIIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStochRSIIndicatorModule, declarations: [IgxStochRSIIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxStochRSIIndicatorComponent, IgxIndicatorProxyModule] });
    IgxStochRSIIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStochRSIIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxStochRSIIndicatorModule;
}());
export { IgxStochRSIIndicatorModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStochRSIIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStochRSIIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxStochRSIIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
