// igniteui-angular-charts/esm5/lib/igx-data-chart-vertical-category-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericXAxisDynamicModule } from './igx-numeric-x-axis-dynamic-module';
import { IgxCategoryYAxisDynamicModule } from './igx-category-y-axis-dynamic-module';
import { IgxDataChartCoreDynamicModule } from './igx-data-chart-core-dynamic-module';
import { IgxDataChartVerticalCategoryCoreModule } from './igx-data-chart-vertical-category-core-module';
import * as i0 from "@angular/core";
var IgxDataChartVerticalCategoryCoreDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartVerticalCategoryCoreDynamicModule() {
    }
    IgxDataChartVerticalCategoryCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartVerticalCategoryCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryCoreDynamicModule, imports: [CommonModule, IgxNumericXAxisDynamicModule,
            IgxCategoryYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartVerticalCategoryCoreModule], exports: [IgxNumericXAxisDynamicModule,
            IgxCategoryYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartVerticalCategoryCoreModule] });
    IgxDataChartVerticalCategoryCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryCoreDynamicModule, imports: [[CommonModule, IgxNumericXAxisDynamicModule,
                IgxCategoryYAxisDynamicModule,
                IgxDataChartCoreDynamicModule,
                IgxDataChartVerticalCategoryCoreModule
            ], IgxNumericXAxisDynamicModule,
            IgxCategoryYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartVerticalCategoryCoreModule] });
    return IgxDataChartVerticalCategoryCoreDynamicModule;
}());
export { IgxDataChartVerticalCategoryCoreDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumericXAxisDynamicModule,
                        IgxCategoryYAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartVerticalCategoryCoreModule
                    ],
                    imports: [CommonModule, IgxNumericXAxisDynamicModule,
                        IgxCategoryYAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartVerticalCategoryCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
