// igniteui-angular-charts/esm5/lib/igx-stacked-column-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedColumnSeriesComponent } from './igx-stacked-column-series-component';
import * as i0 from "@angular/core";
var IgxStackedColumnSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStackedColumnSeriesModule() {
    }
    IgxStackedColumnSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedColumnSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedColumnSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedColumnSeriesModule, declarations: [IgxStackedColumnSeriesComponent], imports: [CommonModule], exports: [IgxStackedColumnSeriesComponent] });
    IgxStackedColumnSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedColumnSeriesModule, imports: [[CommonModule]] });
    return IgxStackedColumnSeriesModule;
}());
export { IgxStackedColumnSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedColumnSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStackedColumnSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxStackedColumnSeriesComponent],
                    entryComponents: []
                }]
        }] });
