// igniteui-angular-charts/esm5/lib/igx-category-y-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryYAxisModule } from './igx-category-y-axis-module';
import { IgxCategoryYAxisComponent } from './igx-category-y-axis-component';
import { CategoryYAxis } from './CategoryYAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxCategoryYAxisDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCategoryYAxisDynamicModule() {
        TypeRegistrar.registerCons("IgxCategoryYAxisComponent", IgxCategoryYAxisComponent);
        TypeRegistrar.register("CategoryYAxis", CategoryYAxis.$type);
    }
    IgxCategoryYAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryYAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryYAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryYAxisDynamicModule, imports: [CommonModule, IgxCategoryYAxisModule], exports: [IgxCategoryYAxisModule] });
    IgxCategoryYAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryYAxisDynamicModule, imports: [[CommonModule, IgxCategoryYAxisModule], IgxCategoryYAxisModule] });
    return IgxCategoryYAxisDynamicModule;
}());
export { IgxCategoryYAxisDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryYAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCategoryYAxisModule],
                    exports: [IgxCategoryYAxisModule],
                    entryComponents: [IgxCategoryYAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
