// igniteui-angular-charts/esm5/lib/igx-stacked-100-spline-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStacked100SplineSeriesComponent } from './igx-stacked-100-spline-series-component';
import * as i0 from "@angular/core";
var IgxStacked100SplineSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStacked100SplineSeriesModule() {
    }
    IgxStacked100SplineSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStacked100SplineSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesModule, declarations: [IgxStacked100SplineSeriesComponent], imports: [CommonModule], exports: [IgxStacked100SplineSeriesComponent] });
    IgxStacked100SplineSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesModule, imports: [[CommonModule]] });
    return IgxStacked100SplineSeriesModule;
}());
export { IgxStacked100SplineSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStacked100SplineSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxStacked100SplineSeriesComponent],
                    entryComponents: []
                }]
        }] });
