// igniteui-angular-charts/esm5/lib/igx-scatter-contour-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterContourSeriesComponent } from './igx-scatter-contour-series-component';
import * as i0 from "@angular/core";
var IgxScatterContourSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxScatterContourSeriesModule() {
    }
    IgxScatterContourSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterContourSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterContourSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterContourSeriesModule, declarations: [IgxScatterContourSeriesComponent], imports: [CommonModule], exports: [IgxScatterContourSeriesComponent] });
    IgxScatterContourSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterContourSeriesModule, imports: [[CommonModule]] });
    return IgxScatterContourSeriesModule;
}());
export { IgxScatterContourSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterContourSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxScatterContourSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxScatterContourSeriesComponent],
                    entryComponents: []
                }]
        }] });
