// igniteui-angular-charts/esm5/lib/igx-moving-average-convergence-divergence-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxMovingAverageConvergenceDivergenceIndicatorModule } from './igx-moving-average-convergence-divergence-indicator-module';
import { IgxMovingAverageConvergenceDivergenceIndicatorComponent } from './igx-moving-average-convergence-divergence-indicator-component';
import { MovingAverageConvergenceDivergenceIndicator } from './MovingAverageConvergenceDivergenceIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxMovingAverageConvergenceDivergenceIndicatorComponent", IgxMovingAverageConvergenceDivergenceIndicatorComponent);
        TypeRegistrar.register("MovingAverageConvergenceDivergenceIndicator", MovingAverageConvergenceDivergenceIndicator.$type);
    }
    IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule, imports: [CommonModule, IgxMovingAverageConvergenceDivergenceIndicatorModule], exports: [IgxMovingAverageConvergenceDivergenceIndicatorModule] });
    IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule, imports: [[CommonModule, IgxMovingAverageConvergenceDivergenceIndicatorModule], IgxMovingAverageConvergenceDivergenceIndicatorModule] });
    return IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule;
}());
export { IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxMovingAverageConvergenceDivergenceIndicatorModule],
                    exports: [IgxMovingAverageConvergenceDivergenceIndicatorModule],
                    entryComponents: [IgxMovingAverageConvergenceDivergenceIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
