// igniteui-angular-charts/esm5/lib/igx-chart-sort-description-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxChartSortDescriptionModule } from './igx-chart-sort-description-module';
import { IgxChartSortDescription } from './igx-chart-sort-description';
import { ChartSortDescription } from './ChartSortDescription';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxChartSortDescriptionDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxChartSortDescriptionDynamicModule() {
        TypeRegistrar.registerCons("IgxChartSortDescription", IgxChartSortDescription);
        TypeRegistrar.register("ChartSortDescription", ChartSortDescription.$type);
    }
    IgxChartSortDescriptionDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSortDescriptionDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxChartSortDescriptionDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSortDescriptionDynamicModule, imports: [CommonModule, IgxChartSortDescriptionModule], exports: [IgxChartSortDescriptionModule] });
    IgxChartSortDescriptionDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSortDescriptionDynamicModule, imports: [[CommonModule, IgxChartSortDescriptionModule], IgxChartSortDescriptionModule] });
    return IgxChartSortDescriptionDynamicModule;
}());
export { IgxChartSortDescriptionDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSortDescriptionDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxChartSortDescriptionModule],
                    exports: [IgxChartSortDescriptionModule],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
