// igniteui-angular-charts/esm5/lib/igx-williams-percent-r-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxWilliamsPercentRIndicatorModule } from './igx-williams-percent-r-indicator-module';
import { IgxWilliamsPercentRIndicatorComponent } from './igx-williams-percent-r-indicator-component';
import { WilliamsPercentRIndicator } from './WilliamsPercentRIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxWilliamsPercentRIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxWilliamsPercentRIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxWilliamsPercentRIndicatorComponent", IgxWilliamsPercentRIndicatorComponent);
        TypeRegistrar.register("WilliamsPercentRIndicator", WilliamsPercentRIndicator.$type);
    }
    IgxWilliamsPercentRIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWilliamsPercentRIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxWilliamsPercentRIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWilliamsPercentRIndicatorDynamicModule, imports: [CommonModule, IgxWilliamsPercentRIndicatorModule], exports: [IgxWilliamsPercentRIndicatorModule] });
    IgxWilliamsPercentRIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWilliamsPercentRIndicatorDynamicModule, imports: [[CommonModule, IgxWilliamsPercentRIndicatorModule], IgxWilliamsPercentRIndicatorModule] });
    return IgxWilliamsPercentRIndicatorDynamicModule;
}());
export { IgxWilliamsPercentRIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWilliamsPercentRIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxWilliamsPercentRIndicatorModule],
                    exports: [IgxWilliamsPercentRIndicatorModule],
                    entryComponents: [IgxWilliamsPercentRIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
