// igniteui-angular-charts/esm5/lib/igx-value-brush-scale-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxValueBrushScaleModule } from './igx-value-brush-scale-module';
import { IgxValueBrushScaleComponent } from './igx-value-brush-scale-component';
import { ValueBrushScale } from './ValueBrushScale';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxValueBrushScaleDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxValueBrushScaleDynamicModule() {
        TypeRegistrar.registerCons("IgxValueBrushScaleComponent", IgxValueBrushScaleComponent);
        TypeRegistrar.register("ValueBrushScale", ValueBrushScale.$type);
    }
    IgxValueBrushScaleDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueBrushScaleDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxValueBrushScaleDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueBrushScaleDynamicModule, imports: [CommonModule, IgxValueBrushScaleModule], exports: [IgxValueBrushScaleModule] });
    IgxValueBrushScaleDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueBrushScaleDynamicModule, imports: [[CommonModule, IgxValueBrushScaleModule], IgxValueBrushScaleModule] });
    return IgxValueBrushScaleDynamicModule;
}());
export { IgxValueBrushScaleDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueBrushScaleDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxValueBrushScaleModule],
                    exports: [IgxValueBrushScaleModule],
                    entryComponents: [IgxValueBrushScaleComponent]
                }]
        }], ctorParameters: function () { return []; } });
