// igniteui-angular-charts/esm5/lib/igx-data-chart-shape-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericXAxisModule } from './igx-numeric-x-axis-module';
import { IgxNumericYAxisModule } from './igx-numeric-y-axis-module';
import { IgxDataChartCoreModule } from './igx-data-chart-core-module';
import * as i0 from "@angular/core";
var IgxDataChartShapeCoreModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartShapeCoreModule() {
    }
    IgxDataChartShapeCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartShapeCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartShapeCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartShapeCoreModule, imports: [CommonModule, IgxNumericXAxisModule,
            IgxNumericYAxisModule,
            IgxDataChartCoreModule], exports: [IgxNumericXAxisModule,
            IgxNumericYAxisModule,
            IgxDataChartCoreModule] });
    IgxDataChartShapeCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartShapeCoreModule, imports: [[CommonModule, IgxNumericXAxisModule,
                IgxNumericYAxisModule,
                IgxDataChartCoreModule
            ], IgxNumericXAxisModule,
            IgxNumericYAxisModule,
            IgxDataChartCoreModule] });
    return IgxDataChartShapeCoreModule;
}());
export { IgxDataChartShapeCoreModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartShapeCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumericXAxisModule,
                        IgxNumericYAxisModule,
                        IgxDataChartCoreModule
                    ],
                    imports: [CommonModule, IgxNumericXAxisModule,
                        IgxNumericYAxisModule,
                        IgxDataChartCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
