// igniteui-angular-charts/esm5/lib/igx-category-chart-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryChartCoreModule } from './igx-category-chart-core-module';
import { IgxLineSeriesDynamicModule } from './igx-line-series-dynamic-module';
import { IgxAreaSeriesDynamicModule } from './igx-area-series-dynamic-module';
import { IgxColumnSeriesDynamicModule } from './igx-column-series-dynamic-module';
import { IgxSplineSeriesDynamicModule } from './igx-spline-series-dynamic-module';
import { IgxSplineAreaSeriesDynamicModule } from './igx-spline-area-series-dynamic-module';
import { IgxStepLineSeriesDynamicModule } from './igx-step-line-series-dynamic-module';
import { IgxStepAreaSeriesDynamicModule } from './igx-step-area-series-dynamic-module';
import { IgxPointSeriesDynamicModule } from './igx-point-series-dynamic-module';
import { IgxWaterfallSeriesDynamicModule } from './igx-waterfall-series-dynamic-module';
import { IgxCrosshairLayerDynamicModule } from './igx-crosshair-layer-dynamic-module';
import { IgxCalloutLayerDynamicModule } from './igx-callout-layer-dynamic-module';
import { IgxItemToolTipLayerDynamicModule } from './igx-item-tool-tip-layer-dynamic-module';
import { IgxCategoryToolTipLayerDynamicModule } from './igx-category-tool-tip-layer-dynamic-module';
import { IgxCategoryHighlightLayerDynamicModule } from './igx-category-highlight-layer-dynamic-module';
import { IgxCategoryItemHighlightLayerDynamicModule } from './igx-category-item-highlight-layer-dynamic-module';
import { IgxDataToolTipLayerDynamicModule } from './igx-data-tool-tip-layer-dynamic-module';
import { IgxFinalValueLayerDynamicModule } from './igx-final-value-layer-dynamic-module';
import { IgxNumberAbbreviatorDynamicModule } from './igx-number-abbreviator-dynamic-module';
import { IgxDataChartInteractivityDynamicModule } from './igx-data-chart-interactivity-dynamic-module';
import { IgxDataChartCategoryTrendLineDynamicModule } from './igx-data-chart-category-trend-line-dynamic-module';
import { IgxDataChartAnnotationDynamicModule } from './igx-data-chart-annotation-dynamic-module';
import * as i0 from "@angular/core";
var IgxCategoryChartModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCategoryChartModule() {
    }
    IgxCategoryChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartModule, imports: [CommonModule, IgxCategoryChartCoreModule,
            IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxSplineSeriesDynamicModule,
            IgxSplineAreaSeriesDynamicModule,
            IgxStepLineSeriesDynamicModule,
            IgxStepAreaSeriesDynamicModule,
            IgxPointSeriesDynamicModule,
            IgxWaterfallSeriesDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxNumberAbbreviatorDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxDataChartCategoryTrendLineDynamicModule,
            IgxDataChartAnnotationDynamicModule], exports: [IgxCategoryChartCoreModule,
            IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxSplineSeriesDynamicModule,
            IgxSplineAreaSeriesDynamicModule,
            IgxStepLineSeriesDynamicModule,
            IgxStepAreaSeriesDynamicModule,
            IgxPointSeriesDynamicModule,
            IgxWaterfallSeriesDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxNumberAbbreviatorDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxDataChartCategoryTrendLineDynamicModule,
            IgxDataChartAnnotationDynamicModule] });
    IgxCategoryChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartModule, imports: [[CommonModule, IgxCategoryChartCoreModule,
                IgxLineSeriesDynamicModule,
                IgxAreaSeriesDynamicModule,
                IgxColumnSeriesDynamicModule,
                IgxSplineSeriesDynamicModule,
                IgxSplineAreaSeriesDynamicModule,
                IgxStepLineSeriesDynamicModule,
                IgxStepAreaSeriesDynamicModule,
                IgxPointSeriesDynamicModule,
                IgxWaterfallSeriesDynamicModule,
                IgxCrosshairLayerDynamicModule,
                IgxCalloutLayerDynamicModule,
                IgxItemToolTipLayerDynamicModule,
                IgxCategoryToolTipLayerDynamicModule,
                IgxCategoryHighlightLayerDynamicModule,
                IgxCategoryItemHighlightLayerDynamicModule,
                IgxDataToolTipLayerDynamicModule,
                IgxFinalValueLayerDynamicModule,
                IgxNumberAbbreviatorDynamicModule,
                IgxDataChartInteractivityDynamicModule,
                IgxDataChartCategoryTrendLineDynamicModule,
                IgxDataChartAnnotationDynamicModule
            ], IgxCategoryChartCoreModule,
            IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxSplineSeriesDynamicModule,
            IgxSplineAreaSeriesDynamicModule,
            IgxStepLineSeriesDynamicModule,
            IgxStepAreaSeriesDynamicModule,
            IgxPointSeriesDynamicModule,
            IgxWaterfallSeriesDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxNumberAbbreviatorDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxDataChartCategoryTrendLineDynamicModule,
            IgxDataChartAnnotationDynamicModule] });
    return IgxCategoryChartModule;
}());
export { IgxCategoryChartModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxCategoryChartCoreModule,
                        IgxLineSeriesDynamicModule,
                        IgxAreaSeriesDynamicModule,
                        IgxColumnSeriesDynamicModule,
                        IgxSplineSeriesDynamicModule,
                        IgxSplineAreaSeriesDynamicModule,
                        IgxStepLineSeriesDynamicModule,
                        IgxStepAreaSeriesDynamicModule,
                        IgxPointSeriesDynamicModule,
                        IgxWaterfallSeriesDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxFinalValueLayerDynamicModule,
                        IgxNumberAbbreviatorDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxDataChartCategoryTrendLineDynamicModule,
                        IgxDataChartAnnotationDynamicModule
                    ],
                    imports: [CommonModule, IgxCategoryChartCoreModule,
                        IgxLineSeriesDynamicModule,
                        IgxAreaSeriesDynamicModule,
                        IgxColumnSeriesDynamicModule,
                        IgxSplineSeriesDynamicModule,
                        IgxSplineAreaSeriesDynamicModule,
                        IgxStepLineSeriesDynamicModule,
                        IgxStepAreaSeriesDynamicModule,
                        IgxPointSeriesDynamicModule,
                        IgxWaterfallSeriesDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxFinalValueLayerDynamicModule,
                        IgxNumberAbbreviatorDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxDataChartCategoryTrendLineDynamicModule,
                        IgxDataChartAnnotationDynamicModule
                    ],
                    entryComponents: []
                }]
        }] });
