// igniteui-angular-charts/esm5/lib/igx-radial-base-component.js
import { __extends } from "tslib";
import { EventEmitter, Output, Component, Input } from '@angular/core';
import { delegateCombine } from "igniteui-angular-core";
import { RadialLabelMode_$type } from "igniteui-angular-core";
import { CategoryTransitionInMode_$type } from "./CategoryTransitionInMode";
import { IgxCategoryAngleAxisComponent } from "./igx-category-angle-axis-component";
import { IgxNumericRadiusAxisComponent } from "./igx-numeric-radius-axis-component";
import { IgxAssigningRadialStyleEventArgs } from "./igx-assigning-radial-style-event-args";
import { IgxAssigningRadialMarkerStyleEventArgs } from "./igx-assigning-radial-marker-style-event-args";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { ensureEnum, ensureBool, arrayFindByName, toPoint } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents the base class for all IgxDataChartComponent radial series
*/
var IgxRadialBaseComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxRadialBaseComponent, _super);
    function IgxRadialBaseComponent() {
        var _this = _super.call(this) || this;
        _this._assigningRadialStyle = null;
        _this._assigningRadialMarkerStyle = null;
        return _this;
    }
    Object.defineProperty(IgxRadialBaseComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "autoCalloutLabelPrecision", {
        /**
         * Gets or sets the format string to use for the auto callout labels for the others category.
        */
        get: function () {
            return this.i.zv;
        },
        set: function (v) {
            this.i.zv = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "autoCalloutPercentagePrecision", {
        /**
         * Gets or sets the format string to use for the auto callout labels for the others category.
        */
        get: function () {
            return this.i.zw;
        },
        set: function (v) {
            this.i.zw = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "autoCalloutLabelValueSeparator", {
        /**
         * Gets or sets the labelling mode to use for the auto callouts in preference to their mode.
        */
        get: function () {
            return this.i.z4;
        },
        set: function (v) {
            this.i.z4 = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "legendRadialLabelMode", {
        /**
         * Gets or sets the labelling mode to use for the auto callouts in preference to their mode.
        */
        get: function () {
            return this.i.y6;
        },
        set: function (v) {
            this.i.y6 = ensureEnum(RadialLabelMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "autoCalloutRadialLabelMode", {
        /**
         * Gets or sets the labelling mode to use for the auto callouts in preference to their mode.
        */
        get: function () {
            return this.i.y5;
        },
        set: function (v) {
            this.i.y5 = ensureEnum(RadialLabelMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "autoCalloutOthersLabelFormat", {
        /**
         * Gets or sets the format string to use for the auto callout labels for the others category.
        */
        get: function () {
            return this.i.z6;
        },
        set: function (v) {
            this.i.z6 = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "autoCalloutOthersLabelFormatSpecifiers", {
        /**
         * Gets or sets the format specifiers to use with the auto callouts for the others category.
        */
        get: function () {
            return this.i.ye;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.ye = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "proportionalRadialLabelFormat", {
        /**
         * Gets or sets the format string to use for the auto callout labels for the others category.
        */
        get: function () {
            return this.i.aas;
        },
        set: function (v) {
            this.i.aas = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "proportionalRadialLabelFormatSpecifiers", {
        /**
         * Gets or sets the format specifiers to use with the auto callouts for the others category.
        */
        get: function () {
            return this.i.yj;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.yj = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "legendProportionalRadialLabelFormat", {
        /**
         * Gets or sets the format string to use for the auto callout labels for the others category.
        */
        get: function () {
            return this.i.aai;
        },
        set: function (v) {
            this.i.aai = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "legendProportionalRadialLabelFormatSpecifiers", {
        /**
         * Gets or sets the format specifiers to use with the auto callouts for the others category.
        */
        get: function () {
            return this.i.yg;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.yg = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "othersProportionalRadialLabelFormat", {
        /**
         * Gets or sets the format string to use for the auto callout labels for the others category.
        */
        get: function () {
            return this.i.aap;
        },
        set: function (v) {
            this.i.aap = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "othersProportionalRadialLabelFormatSpecifiers", {
        /**
         * Gets or sets the format specifiers to use with the auto callouts for the others category.
        */
        get: function () {
            return this.i.yi;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.yi = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "othersLegendProportionalRadialLabelFormat", {
        /**
         * Gets or sets the format string to use for the auto callout labels for the others category.
        */
        get: function () {
            return this.i.aam;
        },
        set: function (v) {
            this.i.aam = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "othersLegendProportionalRadialLabelFormatSpecifiers", {
        /**
         * Gets or sets the format specifiers to use with the auto callouts for the others category.
        */
        get: function () {
            return this.i.yh;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.yh = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "isRadial", {
        /**
         * Checks if this series is a radial series
        */
        get: function () {
            return this.i.fr;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "isCustomRadialStyleAllowed", {
        /**
         * Gets or sets whether this Radial series should allow custom style overrides of its individual visuals.
        */
        get: function () {
            return this.i.zj;
        },
        set: function (v) {
            this.i.zj = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "isCustomRadialMarkerStyleAllowed", {
        /**
         * Gets or sets whether this Radial series should allow custom style overrides of its individual marker visuals.
        */
        get: function () {
            return this.i.zi;
        },
        set: function (v) {
            this.i.zi = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "transitionInMode", {
        /**
         * Gets or sets the method by which to animate the data into the chart when the chart data source is swapped.
         * Note: Transitions are not currently supported for stacked series.
        */
        get: function () {
            return this.i.yu;
        },
        set: function (v) {
            this.i.yu = ensureEnum(CategoryTransitionInMode_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "isTransitionInEnabled", {
        /**
         * Gets or sets whether the series should transition into the plot area when a new data source is assigned.
         * Note: Transitions are not currently supported for stacked series.
        */
        get: function () {
            return this.i.zk;
        },
        set: function (v) {
            this.i.zk = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "angleAxis", {
        /**
         * Gets the effective angle axis for the current series object.
        */
        get: function () {
            var r = this.i.yr;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgxCategoryAngleAxisComponent._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            if (v != null && this._stylingContainer && v._styling)
                v._styling(this._stylingContainer, this, this);
            v == null ? this.i.yr = null : this.i.yr = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "valueAxis", {
        /**
         * Gets the effective value axis for the current series object.
        */
        get: function () {
            var r = this.i.yw;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgxNumericRadiusAxisComponent._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            if (v != null && this._stylingContainer && v._styling)
                v._styling(this._stylingContainer, this, this);
            v == null ? this.i.yw = null : this.i.yw = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "clipSeriesToBounds", {
        /**
         * Gets or sets whether to clip the series to the bounds.
         * Setting this property to true will decrease chart performance.
        */
        get: function () {
            return this.i.zh;
        },
        set: function (v) {
            this.i.zh = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    IgxRadialBaseComponent.prototype.findByName = function (name) {
        var baseResult = _super.prototype.findByName.call(this, name);
        if (baseResult) {
            return baseResult;
        }
        if (this.autoCalloutOthersLabelFormatSpecifiers != null && arrayFindByName(this.autoCalloutOthersLabelFormatSpecifiers, name)) {
            return arrayFindByName(this.autoCalloutOthersLabelFormatSpecifiers, name);
        }
        if (this.proportionalRadialLabelFormatSpecifiers != null && arrayFindByName(this.proportionalRadialLabelFormatSpecifiers, name)) {
            return arrayFindByName(this.proportionalRadialLabelFormatSpecifiers, name);
        }
        if (this.legendProportionalRadialLabelFormatSpecifiers != null && arrayFindByName(this.legendProportionalRadialLabelFormatSpecifiers, name)) {
            return arrayFindByName(this.legendProportionalRadialLabelFormatSpecifiers, name);
        }
        if (this.othersProportionalRadialLabelFormatSpecifiers != null && arrayFindByName(this.othersProportionalRadialLabelFormatSpecifiers, name)) {
            return arrayFindByName(this.othersProportionalRadialLabelFormatSpecifiers, name);
        }
        if (this.othersLegendProportionalRadialLabelFormatSpecifiers != null && arrayFindByName(this.othersLegendProportionalRadialLabelFormatSpecifiers, name)) {
            return arrayFindByName(this.othersLegendProportionalRadialLabelFormatSpecifiers, name);
        }
        if (this.angleAxis && this.angleAxis.name && this.angleAxis.name == name) {
            return this.angleAxis;
        }
        if (this.valueAxis && this.valueAxis.name && this.valueAxis.name == name) {
            return this.valueAxis;
        }
        return null;
    };
    IgxRadialBaseComponent.prototype._styling = function (container, component, parent) {
        _super.prototype._styling.call(this, container, component, parent);
        this._inStyling = true;
        if (this.angleAxis && this.angleAxis._styling) {
            this.angleAxis._styling(container, component, this);
        }
        if (this.valueAxis && this.valueAxis._styling) {
            this.valueAxis._styling(container, component, this);
        }
        this._inStyling = false;
    };
    /**
     * Determine if object can be used as ValueAxis
    
    * @param axis  * The object to check
    */
    IgxRadialBaseComponent.prototype.canUseAsValueAxis = function (axis) {
        var iv = this.i.zg(axis);
        return (iv);
    };
    /**
     * Determine if object can be used as AngleAxis
    
    * @param axis  * The object to check
    */
    IgxRadialBaseComponent.prototype.canUseAsAngleAxis = function (axis) {
        var iv = this.i.zf(axis);
        return (iv);
    };
    /**
     * Gets the item that is the best match for the specified world coordinates.
    
    * @param world  * The world coordinates to use.
    */
    IgxRadialBaseComponent.prototype.getItem = function (world) {
        var iv = this.i.kl(toPoint(world));
        return (iv);
    };
    /**
     * Gets the angle to the provided world position from the center of the radial series.
    
    * @param world  * The world position to get the angle for
    */
    IgxRadialBaseComponent.prototype.getAngleFromWorld = function (world) {
        var iv = this.i.zq(toPoint(world));
        return (iv);
    };
    /**
     * Gets the precise item index, if possible, based on the closeness to the previous or next whole integer. If the series cannot provide this information, GetExactItemIndex will return the same integer value as GetItemIndex.
    
    * @param world  * The world position for which to return the index.
    */
    IgxRadialBaseComponent.prototype.getExactItemIndex = function (world) {
        var iv = this.i.ip(toPoint(world));
        return (iv);
    };
    /**
     * Get the index of the item near the provided world coordinates.
    
    * @param world  * The world coordinates for which to getch the item index.
    */
    IgxRadialBaseComponent.prototype.getItemIndex = function (world) {
        var iv = this.i.j1(toPoint(world));
        return (iv);
    };
    /**
     * Scrolls the specified item into the view.
    
    * @param item  * The item to scroll into view.
    */
    IgxRadialBaseComponent.prototype.scrollIntoView = function (item) {
        var iv = this.i.gb(item);
        return (iv);
    };
    /**
     * Returns the offset value for this series if grouped on a category axis.
    
    */
    IgxRadialBaseComponent.prototype.getOffsetValue = function () {
        var iv = this.i.getOffsetValue();
        return (iv);
    };
    /**
     * Returns the width of the category grouping this series is in.
    
    */
    IgxRadialBaseComponent.prototype.getCategoryWidth = function () {
        var iv = this.i.getCategoryWidth();
        return (iv);
    };
    Object.defineProperty(IgxRadialBaseComponent.prototype, "assigningRadialStyle", {
        /**
         * Event raised when Assigning Category Style
        */
        get: function () {
            var _this = this;
            if (this._assigningRadialStyle == null) {
                this._assigningRadialStyle = new EventEmitter();
                this.i.assigningRadialStyle = delegateCombine(this.i.assigningRadialStyle, function (o, e) {
                    var outerArgs = new IgxAssigningRadialStyleEventArgs();
                    outerArgs._provideImplementation(e);
                    if (_this.beforeAssigningRadialStyle) {
                        _this.beforeAssigningRadialStyle(_this, outerArgs);
                    }
                    _this._assigningRadialStyle.emit({
                        sender: _this,
                        args: outerArgs
                    });
                });
            }
            return this._assigningRadialStyle;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxRadialBaseComponent.prototype, "assigningRadialMarkerStyle", {
        /**
         * Event raised when Assigning Radial Marker Style
        */
        get: function () {
            var _this = this;
            if (this._assigningRadialMarkerStyle == null) {
                this._assigningRadialMarkerStyle = new EventEmitter();
                this.i.assigningRadialMarkerStyle = delegateCombine(this.i.assigningRadialMarkerStyle, function (o, e) {
                    var outerArgs = new IgxAssigningRadialMarkerStyleEventArgs();
                    outerArgs._provideImplementation(e);
                    if (_this.beforeAssigningRadialMarkerStyle) {
                        _this.beforeAssigningRadialMarkerStyle(_this, outerArgs);
                    }
                    _this._assigningRadialMarkerStyle.emit({
                        sender: _this,
                        args: outerArgs
                    });
                });
            }
            return this._assigningRadialMarkerStyle;
        },
        enumerable: false,
        configurable: true
    });
    IgxRadialBaseComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialBaseComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxRadialBaseComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxRadialBaseComponent, selector: "ng-component", inputs: { autoCalloutLabelPrecision: "autoCalloutLabelPrecision", autoCalloutPercentagePrecision: "autoCalloutPercentagePrecision", autoCalloutLabelValueSeparator: "autoCalloutLabelValueSeparator", legendRadialLabelMode: "legendRadialLabelMode", autoCalloutRadialLabelMode: "autoCalloutRadialLabelMode", autoCalloutOthersLabelFormat: "autoCalloutOthersLabelFormat", autoCalloutOthersLabelFormatSpecifiers: "autoCalloutOthersLabelFormatSpecifiers", proportionalRadialLabelFormat: "proportionalRadialLabelFormat", proportionalRadialLabelFormatSpecifiers: "proportionalRadialLabelFormatSpecifiers", legendProportionalRadialLabelFormat: "legendProportionalRadialLabelFormat", legendProportionalRadialLabelFormatSpecifiers: "legendProportionalRadialLabelFormatSpecifiers", othersProportionalRadialLabelFormat: "othersProportionalRadialLabelFormat", othersProportionalRadialLabelFormatSpecifiers: "othersProportionalRadialLabelFormatSpecifiers", othersLegendProportionalRadialLabelFormat: "othersLegendProportionalRadialLabelFormat", othersLegendProportionalRadialLabelFormatSpecifiers: "othersLegendProportionalRadialLabelFormatSpecifiers", isCustomRadialStyleAllowed: "isCustomRadialStyleAllowed", isCustomRadialMarkerStyleAllowed: "isCustomRadialMarkerStyleAllowed", transitionInMode: "transitionInMode", isTransitionInEnabled: "isTransitionInEnabled", angleAxis: "angleAxis", valueAxis: "valueAxis", clipSeriesToBounds: "clipSeriesToBounds" }, outputs: { assigningRadialStyle: "assigningRadialStyle", assigningRadialMarkerStyle: "assigningRadialMarkerStyle" }, usesInheritance: true, ngImport: i0, template: "", isInline: true });
    return IgxRadialBaseComponent;
}(IgxMarkerSeriesComponent));
export { IgxRadialBaseComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialBaseComponent, decorators: [{
            type: Component,
            args: [{
                    template: "",
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { autoCalloutLabelPrecision: [{
                type: Input
            }], autoCalloutPercentagePrecision: [{
                type: Input
            }], autoCalloutLabelValueSeparator: [{
                type: Input
            }], legendRadialLabelMode: [{
                type: Input
            }], autoCalloutRadialLabelMode: [{
                type: Input
            }], autoCalloutOthersLabelFormat: [{
                type: Input
            }], autoCalloutOthersLabelFormatSpecifiers: [{
                type: Input
            }], proportionalRadialLabelFormat: [{
                type: Input
            }], proportionalRadialLabelFormatSpecifiers: [{
                type: Input
            }], legendProportionalRadialLabelFormat: [{
                type: Input
            }], legendProportionalRadialLabelFormatSpecifiers: [{
                type: Input
            }], othersProportionalRadialLabelFormat: [{
                type: Input
            }], othersProportionalRadialLabelFormatSpecifiers: [{
                type: Input
            }], othersLegendProportionalRadialLabelFormat: [{
                type: Input
            }], othersLegendProportionalRadialLabelFormatSpecifiers: [{
                type: Input
            }], isCustomRadialStyleAllowed: [{
                type: Input
            }], isCustomRadialMarkerStyleAllowed: [{
                type: Input
            }], transitionInMode: [{
                type: Input
            }], isTransitionInEnabled: [{
                type: Input
            }], angleAxis: [{
                type: Input
            }], valueAxis: [{
                type: Input
            }], clipSeriesToBounds: [{
                type: Input
            }], assigningRadialStyle: [{
                type: Output
            }], assigningRadialMarkerStyle: [{
                type: Output
            }] } });
