// igniteui-angular-charts/esm5/lib/igx-waterfall-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxWaterfallSeriesModule } from './igx-waterfall-series-module';
import { IgxWaterfallSeriesComponent } from './igx-waterfall-series-component';
import { WaterfallSeries } from './WaterfallSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxWaterfallSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxWaterfallSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxWaterfallSeriesComponent", IgxWaterfallSeriesComponent);
        TypeRegistrar.register("WaterfallSeries", WaterfallSeries.$type);
    }
    IgxWaterfallSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWaterfallSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxWaterfallSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWaterfallSeriesDynamicModule, imports: [CommonModule, IgxWaterfallSeriesModule], exports: [IgxWaterfallSeriesModule] });
    IgxWaterfallSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWaterfallSeriesDynamicModule, imports: [[CommonModule, IgxWaterfallSeriesModule], IgxWaterfallSeriesModule] });
    return IgxWaterfallSeriesDynamicModule;
}());
export { IgxWaterfallSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWaterfallSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxWaterfallSeriesModule],
                    exports: [IgxWaterfallSeriesModule],
                    entryComponents: [IgxWaterfallSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
