// igniteui-angular-charts/esm5/lib/igx-trix-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTRIXIndicatorModule } from './igx-trix-indicator-module';
import { IgxTRIXIndicatorComponent } from './igx-trix-indicator-component';
import { TRIXIndicator } from './TRIXIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxTRIXIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxTRIXIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxTRIXIndicatorComponent", IgxTRIXIndicatorComponent);
        TypeRegistrar.register("TRIXIndicator", TRIXIndicator.$type);
    }
    IgxTRIXIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTRIXIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTRIXIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTRIXIndicatorDynamicModule, imports: [CommonModule, IgxTRIXIndicatorModule], exports: [IgxTRIXIndicatorModule] });
    IgxTRIXIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTRIXIndicatorDynamicModule, imports: [[CommonModule, IgxTRIXIndicatorModule], IgxTRIXIndicatorModule] });
    return IgxTRIXIndicatorDynamicModule;
}());
export { IgxTRIXIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTRIXIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxTRIXIndicatorModule],
                    exports: [IgxTRIXIndicatorModule],
                    entryComponents: [IgxTRIXIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
