// igniteui-angular-charts/esm5/lib/igx-absolute-volume-oscillator-indicator-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import { IgxFinancialIndicatorComponent } from "./igx-financial-indicator-component";
import { IgxFinancialSeriesComponent } from "./igx-financial-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { AbsoluteVolumeOscillatorIndicator } from "./AbsoluteVolumeOscillatorIndicator";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent Absolute Volume Oscillator indicator series.
 * Default required members: Volume
*
* You can use the `AbsoluteVolumeOscillatorIndicator` to identify whether volume trands are increasing or decreasing.
*
* ```html
* <igx-data-chart
* 	#chart
*     [dataSource]="data" >
*     <igx-category-x-axis
* 		label="label"
* 		#xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
* 		#yAxis>
*     </igx-numeric-y-axis>
*     <igx-absolute-volume-oscillator-indicator
* 		[xAxis]="xAxis"
* 		[yAxis]="yAxis"
* 		openMemberPath="open"
* 		highMemberPath="high"
* 		lowMemberPath="low"
* 		closeMemberPath="close">
*     </igx-absolute-volume-oscillator-indicator>
* </igx-data-chart>
* ```
*/
var IgxAbsoluteVolumeOscillatorIndicatorComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxAbsoluteVolumeOscillatorIndicatorComponent, _super);
    function IgxAbsoluteVolumeOscillatorIndicatorComponent() {
        return _super.call(this) || this;
    }
    IgxAbsoluteVolumeOscillatorIndicatorComponent.prototype.createImplementation = function () {
        return new AbsoluteVolumeOscillatorIndicator();
    };
    Object.defineProperty(IgxAbsoluteVolumeOscillatorIndicatorComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxAbsoluteVolumeOscillatorIndicatorComponent.prototype, "shortPeriod", {
        /**
         * Gets or sets the short moving average period for the current AbsoluteVolumeOscillatorIndicator object.
         * The typical, and initial, value for short AVO periods is 10.
        *
        * You can use the `ShortPeriod` to set the short moving average period.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data" >
        *     <igx-category-x-axis
        * 		label="label"
        * 		#xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        * 		#yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-absolute-volume-oscillator-indicator
        * 		#series
        * 		[xAxis]="xAxis"
        * 		[yAxis]="yAxis"
        * 		openMemberPath="open"
        * 		highMemberPath="high"
        * 		lowMemberPath="low"
        * 		closeMemberPath="close"
        * 		shortPeriod="10">
        *     </igx-absolute-volume-oscillator-indicator>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.shortPeriod = 10;
        * ```
        */
        get: function () {
            return this.i.shortPeriod;
        },
        set: function (v) {
            this.i.shortPeriod = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxAbsoluteVolumeOscillatorIndicatorComponent.prototype, "longPeriod", {
        /**
         * Gets or sets the short moving average period for the current AbsoluteVolumeOscillatorIndicator object.
         * The typical, and initial, value for long AVO periods is 30.
        *
        * You can use the `LongPeriod` to set the long moving average period.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data" >
        *     <igx-category-x-axis
        * 		label="label"
        * 		#xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        * 		#yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-absolute-volume-oscillator-indicator
        * 		#series
        * 		[xAxis]="xAxis"
        * 		[yAxis]="yAxis"
        * 		openMemberPath="open"
        * 		highMemberPath="high"
        * 		lowMemberPath="low"
        * 		closeMemberPath="close"
        * 		longPeriod="30">
        *     </igx-absolute-volume-oscillator-indicator>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.series.longPeriod = 30;
        * ```
        */
        get: function () {
            return this.i.longPeriod;
        },
        set: function (v) {
            this.i.longPeriod = +v;
        },
        enumerable: false,
        configurable: true
    });
    IgxAbsoluteVolumeOscillatorIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAbsoluteVolumeOscillatorIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxAbsoluteVolumeOscillatorIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxAbsoluteVolumeOscillatorIndicatorComponent, selector: "igx-absolute-volume-oscillator-indicator", inputs: { shortPeriod: "shortPeriod", longPeriod: "longPeriod" }, providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxAbsoluteVolumeOscillatorIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxAbsoluteVolumeOscillatorIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxAbsoluteVolumeOscillatorIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxAbsoluteVolumeOscillatorIndicatorComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxAbsoluteVolumeOscillatorIndicatorComponent;
}(IgxStrategyBasedIndicatorComponent));
export { IgxAbsoluteVolumeOscillatorIndicatorComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAbsoluteVolumeOscillatorIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-absolute-volume-oscillator-indicator',
                    template: "",
                    providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxAbsoluteVolumeOscillatorIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxAbsoluteVolumeOscillatorIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxAbsoluteVolumeOscillatorIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxAbsoluteVolumeOscillatorIndicatorComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { shortPeriod: [{
                type: Input
            }], longPeriod: [{
                type: Input
            }] } });
