// igniteui-angular-charts/esm5/lib/igx-category-x-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryXAxisComponent } from './igx-category-x-axis-component';
import * as i0 from "@angular/core";
var IgxCategoryXAxisModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCategoryXAxisModule() {
    }
    IgxCategoryXAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryXAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryXAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryXAxisModule, declarations: [IgxCategoryXAxisComponent], imports: [CommonModule], exports: [IgxCategoryXAxisComponent] });
    IgxCategoryXAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryXAxisModule, imports: [[CommonModule]] });
    return IgxCategoryXAxisModule;
}());
export { IgxCategoryXAxisModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryXAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxCategoryXAxisComponent],
                    imports: [CommonModule],
                    exports: [IgxCategoryXAxisComponent],
                    entryComponents: []
                }]
        }] });
