// igniteui-angular-charts/esm5/lib/igx-negative-volume-index-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNegativeVolumeIndexIndicatorModule } from './igx-negative-volume-index-indicator-module';
import { IgxNegativeVolumeIndexIndicatorComponent } from './igx-negative-volume-index-indicator-component';
import { NegativeVolumeIndexIndicator } from './NegativeVolumeIndexIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxNegativeVolumeIndexIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxNegativeVolumeIndexIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxNegativeVolumeIndexIndicatorComponent", IgxNegativeVolumeIndexIndicatorComponent);
        TypeRegistrar.register("NegativeVolumeIndexIndicator", NegativeVolumeIndexIndicator.$type);
    }
    IgxNegativeVolumeIndexIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNegativeVolumeIndexIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxNegativeVolumeIndexIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNegativeVolumeIndexIndicatorDynamicModule, imports: [CommonModule, IgxNegativeVolumeIndexIndicatorModule], exports: [IgxNegativeVolumeIndexIndicatorModule] });
    IgxNegativeVolumeIndexIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNegativeVolumeIndexIndicatorDynamicModule, imports: [[CommonModule, IgxNegativeVolumeIndexIndicatorModule], IgxNegativeVolumeIndexIndicatorModule] });
    return IgxNegativeVolumeIndexIndicatorDynamicModule;
}());
export { IgxNegativeVolumeIndexIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNegativeVolumeIndexIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxNegativeVolumeIndexIndicatorModule],
                    exports: [IgxNegativeVolumeIndexIndicatorModule],
                    entryComponents: [IgxNegativeVolumeIndexIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
