// igniteui-angular-charts/esm5/lib/igx-percent-change-y-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPercentChangeYAxisModule } from './igx-percent-change-y-axis-module';
import { IgxPercentChangeYAxisComponent } from './igx-percent-change-y-axis-component';
import { PercentChangeYAxis } from './PercentChangeYAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxPercentChangeYAxisDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPercentChangeYAxisDynamicModule() {
        TypeRegistrar.registerCons("IgxPercentChangeYAxisComponent", IgxPercentChangeYAxisComponent);
        TypeRegistrar.register("PercentChangeYAxis", PercentChangeYAxis.$type);
    }
    IgxPercentChangeYAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentChangeYAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPercentChangeYAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentChangeYAxisDynamicModule, imports: [CommonModule, IgxPercentChangeYAxisModule], exports: [IgxPercentChangeYAxisModule] });
    IgxPercentChangeYAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentChangeYAxisDynamicModule, imports: [[CommonModule, IgxPercentChangeYAxisModule], IgxPercentChangeYAxisModule] });
    return IgxPercentChangeYAxisDynamicModule;
}());
export { IgxPercentChangeYAxisDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentChangeYAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPercentChangeYAxisModule],
                    exports: [IgxPercentChangeYAxisModule],
                    entryComponents: [IgxPercentChangeYAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
