// igniteui-angular-charts/esm5/lib/igx-point-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPointSeriesModule } from './igx-point-series-module';
import { IgxPointSeriesComponent } from './igx-point-series-component';
import { PointSeries } from './PointSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxPointSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPointSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxPointSeriesComponent", IgxPointSeriesComponent);
        TypeRegistrar.register("PointSeries", PointSeries.$type);
    }
    IgxPointSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPointSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPointSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPointSeriesDynamicModule, imports: [CommonModule, IgxPointSeriesModule], exports: [IgxPointSeriesModule] });
    IgxPointSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPointSeriesDynamicModule, imports: [[CommonModule, IgxPointSeriesModule], IgxPointSeriesModule] });
    return IgxPointSeriesDynamicModule;
}());
export { IgxPointSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPointSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPointSeriesModule],
                    exports: [IgxPointSeriesModule],
                    entryComponents: [IgxPointSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
