// igniteui-angular-charts/esm5/lib/igx-indicator-proxy-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IndicatorProxy } from './IndicatorProxy';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import * as i0 from "@angular/core";
var IgxIndicatorProxyDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxIndicatorProxyDynamicModule() {
        TypeRegistrar.registerCons('IndicatorProxy', IndicatorProxy);
        TypeRegistrar.register('IndicatorProxy', IndicatorProxy.$type);
    }
    IgxIndicatorProxyDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorProxyDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxIndicatorProxyDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorProxyDynamicModule, imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxIndicatorProxyModule] });
    IgxIndicatorProxyDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorProxyDynamicModule, imports: [[CommonModule, IgxIndicatorProxyModule
            ], IgxIndicatorProxyModule] });
    return IgxIndicatorProxyDynamicModule;
}());
export { IgxIndicatorProxyDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorProxyDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxIndicatorProxyModule
                    ],
                    imports: [CommonModule, IgxIndicatorProxyModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
