// igniteui-angular-charts/esm5/lib/igx-data-chart-financial-indicators-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorsDynamicModule } from './igx-indicators-dynamic-module';
import { IgxDataChartFinancialCoreDynamicModule } from './igx-data-chart-financial-core-dynamic-module';
import { IgxDataChartFinancialIndicatorsModule } from './igx-data-chart-financial-indicators-module';
import * as i0 from "@angular/core";
var IgxDataChartFinancialIndicatorsDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartFinancialIndicatorsDynamicModule() {
    }
    IgxDataChartFinancialIndicatorsDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialIndicatorsDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartFinancialIndicatorsDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialIndicatorsDynamicModule, imports: [CommonModule, IgxIndicatorsDynamicModule,
            IgxDataChartFinancialCoreDynamicModule,
            IgxDataChartFinancialIndicatorsModule], exports: [IgxIndicatorsDynamicModule,
            IgxDataChartFinancialCoreDynamicModule,
            IgxDataChartFinancialIndicatorsModule] });
    IgxDataChartFinancialIndicatorsDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialIndicatorsDynamicModule, imports: [[CommonModule, IgxIndicatorsDynamicModule,
                IgxDataChartFinancialCoreDynamicModule,
                IgxDataChartFinancialIndicatorsModule
            ], IgxIndicatorsDynamicModule,
            IgxDataChartFinancialCoreDynamicModule,
            IgxDataChartFinancialIndicatorsModule] });
    return IgxDataChartFinancialIndicatorsDynamicModule;
}());
export { IgxDataChartFinancialIndicatorsDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialIndicatorsDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxIndicatorsDynamicModule,
                        IgxDataChartFinancialCoreDynamicModule,
                        IgxDataChartFinancialIndicatorsModule
                    ],
                    imports: [CommonModule, IgxIndicatorsDynamicModule,
                        IgxDataChartFinancialCoreDynamicModule,
                        IgxDataChartFinancialIndicatorsModule
                    ],
                    entryComponents: []
                }]
        }] });
