// igniteui-angular-charts/esm5/lib/igx-scatter-polygon-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterPolygonSeriesComponent } from './igx-scatter-polygon-series-component';
import * as i0 from "@angular/core";
var IgxScatterPolygonSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxScatterPolygonSeriesModule() {
    }
    IgxScatterPolygonSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolygonSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterPolygonSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolygonSeriesModule, declarations: [IgxScatterPolygonSeriesComponent], imports: [CommonModule], exports: [IgxScatterPolygonSeriesComponent] });
    IgxScatterPolygonSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolygonSeriesModule, imports: [[CommonModule]] });
    return IgxScatterPolygonSeriesModule;
}());
export { IgxScatterPolygonSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolygonSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxScatterPolygonSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxScatterPolygonSeriesComponent],
                    entryComponents: []
                }]
        }] });
