// igniteui-angular-charts/esm5/lib/igx-stacked-100-spline-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStackedSplineSeriesComponent } from "./igx-stacked-spline-series-component";
import { IgxHorizontalStackedSeriesBaseComponent } from "./igx-horizontal-stacked-series-base-component";
import { IgxStackedSeriesBaseComponent } from "./igx-stacked-series-base-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { Stacked100SplineSeries } from "./Stacked100SplineSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent stacked100 spline series.
*/
var IgxStacked100SplineSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxStacked100SplineSeriesComponent, _super);
    function IgxStacked100SplineSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxStacked100SplineSeriesComponent.prototype.createImplementation = function () {
        return new Stacked100SplineSeries();
    };
    Object.defineProperty(IgxStacked100SplineSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxStacked100SplineSeriesComponent.prototype, "isPercentBased", {
        get: function () {
            return this.i.aan;
        },
        enumerable: false,
        configurable: true
    });
    IgxStacked100SplineSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStacked100SplineSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStacked100SplineSeriesComponent, selector: "igx-stacked-100-spline-series", providers: [{ provide: IgxStackedSplineSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStacked100SplineSeriesComponent;
}(IgxStackedSplineSeriesComponent));
export { IgxStacked100SplineSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-stacked-100-spline-series',
                    template: "",
                    providers: [{ provide: IgxStackedSplineSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100SplineSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
