// igniteui-angular-charts/esm5/lib/igx-scatter-area-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterAreaSeriesModule } from './igx-scatter-area-series-module';
import { IgxScatterAreaSeriesComponent } from './igx-scatter-area-series-component';
import { ScatterAreaSeries } from './ScatterAreaSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxScatterAreaSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxScatterAreaSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxScatterAreaSeriesComponent", IgxScatterAreaSeriesComponent);
        TypeRegistrar.register("ScatterAreaSeries", ScatterAreaSeries.$type);
    }
    IgxScatterAreaSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterAreaSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterAreaSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterAreaSeriesDynamicModule, imports: [CommonModule, IgxScatterAreaSeriesModule], exports: [IgxScatterAreaSeriesModule] });
    IgxScatterAreaSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterAreaSeriesDynamicModule, imports: [[CommonModule, IgxScatterAreaSeriesModule], IgxScatterAreaSeriesModule] });
    return IgxScatterAreaSeriesDynamicModule;
}());
export { IgxScatterAreaSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterAreaSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxScatterAreaSeriesModule],
                    exports: [IgxScatterAreaSeriesModule],
                    entryComponents: [IgxScatterAreaSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
