// igniteui-angular-charts/esm5/lib/igx-stacked-column-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedColumnSeriesModule } from './igx-stacked-column-series-module';
import { IgxStackedColumnSeriesComponent } from './igx-stacked-column-series-component';
import { StackedColumnSeries } from './StackedColumnSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxStackedColumnSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStackedColumnSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxStackedColumnSeriesComponent", IgxStackedColumnSeriesComponent);
        TypeRegistrar.register("StackedColumnSeries", StackedColumnSeries.$type);
    }
    IgxStackedColumnSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedColumnSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedColumnSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedColumnSeriesDynamicModule, imports: [CommonModule, IgxStackedColumnSeriesModule], exports: [IgxStackedColumnSeriesModule] });
    IgxStackedColumnSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedColumnSeriesDynamicModule, imports: [[CommonModule, IgxStackedColumnSeriesModule], IgxStackedColumnSeriesModule] });
    return IgxStackedColumnSeriesDynamicModule;
}());
export { IgxStackedColumnSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedColumnSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStackedColumnSeriesModule],
                    exports: [IgxStackedColumnSeriesModule],
                    entryComponents: [IgxStackedColumnSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
