// igniteui-angular-charts/esm5/lib/igx-polar-spline-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxPolarLineSeriesBaseComponent } from "./igx-polar-line-series-base-component";
import { IgxPolarBaseComponent } from "./igx-polar-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { PolarSplineSeries } from "./PolarSplineSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent polar spline series.
*/
var IgxPolarSplineSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxPolarSplineSeriesComponent, _super);
    function IgxPolarSplineSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxPolarSplineSeriesComponent.prototype.createImplementation = function () {
        return new PolarSplineSeries();
    };
    Object.defineProperty(IgxPolarSplineSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxPolarSplineSeriesComponent.prototype, "isSpline", {
        /**
         * Gets whether the current series shows a spline shape.
        */
        get: function () {
            return this.i.fy;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxPolarSplineSeriesComponent.prototype, "stiffness", {
        /**
         * Gets or sets the Stiffness property.
        */
        get: function () {
            return this.i.acn;
        },
        set: function (v) {
            this.i.acn = +v;
        },
        enumerable: false,
        configurable: true
    });
    IgxPolarSplineSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxPolarSplineSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxPolarSplineSeriesComponent, selector: "igx-polar-spline-series", inputs: { stiffness: "stiffness" }, providers: [{ provide: IgxPolarLineSeriesBaseComponent, useExisting: forwardRef(function () { return IgxPolarSplineSeriesComponent; }) }, { provide: IgxPolarBaseComponent, useExisting: forwardRef(function () { return IgxPolarSplineSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxPolarSplineSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxPolarSplineSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxPolarSplineSeriesComponent;
}(IgxPolarLineSeriesBaseComponent));
export { IgxPolarSplineSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-polar-spline-series',
                    template: "",
                    providers: [{ provide: IgxPolarLineSeriesBaseComponent, useExisting: forwardRef(function () { return IgxPolarSplineSeriesComponent; }) }, { provide: IgxPolarBaseComponent, useExisting: forwardRef(function () { return IgxPolarSplineSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxPolarSplineSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxPolarSplineSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { stiffness: [{
                type: Input
            }] } });
