// igniteui-angular-charts/esm5/lib/igx-ease-of-movement-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxEaseOfMovementIndicatorModule } from './igx-ease-of-movement-indicator-module';
import { IgxEaseOfMovementIndicatorComponent } from './igx-ease-of-movement-indicator-component';
import { EaseOfMovementIndicator } from './EaseOfMovementIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxEaseOfMovementIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxEaseOfMovementIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxEaseOfMovementIndicatorComponent", IgxEaseOfMovementIndicatorComponent);
        TypeRegistrar.register("EaseOfMovementIndicator", EaseOfMovementIndicator.$type);
    }
    IgxEaseOfMovementIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxEaseOfMovementIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxEaseOfMovementIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxEaseOfMovementIndicatorDynamicModule, imports: [CommonModule, IgxEaseOfMovementIndicatorModule], exports: [IgxEaseOfMovementIndicatorModule] });
    IgxEaseOfMovementIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxEaseOfMovementIndicatorDynamicModule, imports: [[CommonModule, IgxEaseOfMovementIndicatorModule], IgxEaseOfMovementIndicatorModule] });
    return IgxEaseOfMovementIndicatorDynamicModule;
}());
export { IgxEaseOfMovementIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxEaseOfMovementIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxEaseOfMovementIndicatorModule],
                    exports: [IgxEaseOfMovementIndicatorModule],
                    entryComponents: [IgxEaseOfMovementIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
