// igniteui-angular-charts/esm5/lib/igx-pie-chart-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumberFormatSpecifierModule } from "igniteui-angular-core";
import { IgxDateTimeFormatSpecifierModule } from "igniteui-angular-core";
import { IgxPieChartComponent } from './igx-pie-chart-component';
import { FastItemObjectColumn } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxPieChartCoreModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPieChartCoreModule() {
        TypeRegistrar.register('FastItemObjectColumn', FastItemObjectColumn.$type);
    }
    IgxPieChartCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPieChartCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartCoreModule, declarations: [IgxPieChartComponent], imports: [CommonModule, IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule], exports: [IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule,
            IgxPieChartComponent] });
    IgxPieChartCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartCoreModule, imports: [[CommonModule, IgxNumberFormatSpecifierModule,
                IgxDateTimeFormatSpecifierModule
            ], IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule] });
    return IgxPieChartCoreModule;
}());
export { IgxPieChartCoreModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxPieChartComponent],
                    exports: [IgxNumberFormatSpecifierModule,
                        IgxDateTimeFormatSpecifierModule,
                        IgxPieChartComponent
                    ],
                    imports: [CommonModule, IgxNumberFormatSpecifierModule,
                        IgxDateTimeFormatSpecifierModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
