// igniteui-angular-charts/esm5/lib/igx-data-pie-chart-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataPieChartCoreDynamicModule } from './igx-data-pie-chart-core-dynamic-module';
import { IgxRadialPieSeriesDynamicModule } from './igx-radial-pie-series-dynamic-module';
import { IgxCrosshairLayerDynamicModule } from './igx-crosshair-layer-dynamic-module';
import { IgxCalloutLayerDynamicModule } from './igx-callout-layer-dynamic-module';
import { IgxItemToolTipLayerDynamicModule } from './igx-item-tool-tip-layer-dynamic-module';
import { IgxCategoryToolTipLayerDynamicModule } from './igx-category-tool-tip-layer-dynamic-module';
import { IgxCategoryHighlightLayerDynamicModule } from './igx-category-highlight-layer-dynamic-module';
import { IgxCategoryItemHighlightLayerDynamicModule } from './igx-category-item-highlight-layer-dynamic-module';
import { IgxDataToolTipLayerDynamicModule } from './igx-data-tool-tip-layer-dynamic-module';
import { IgxNumberAbbreviatorDynamicModule } from './igx-number-abbreviator-dynamic-module';
import { IgxDataChartInteractivityDynamicModule } from './igx-data-chart-interactivity-dynamic-module';
import { IgxDataChartAnnotationDynamicModule } from './igx-data-chart-annotation-dynamic-module';
import { IgxDataPieChartModule } from './igx-data-pie-chart-module';
import * as i0 from "@angular/core";
var IgxDataPieChartDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataPieChartDynamicModule() {
    }
    IgxDataPieChartDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataPieChartDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartDynamicModule, imports: [CommonModule, IgxDataPieChartCoreDynamicModule,
            IgxRadialPieSeriesDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxNumberAbbreviatorDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxDataChartAnnotationDynamicModule,
            IgxDataPieChartModule], exports: [IgxDataPieChartCoreDynamicModule,
            IgxRadialPieSeriesDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxNumberAbbreviatorDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxDataChartAnnotationDynamicModule,
            IgxDataPieChartModule] });
    IgxDataPieChartDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartDynamicModule, imports: [[CommonModule, IgxDataPieChartCoreDynamicModule,
                IgxRadialPieSeriesDynamicModule,
                IgxCrosshairLayerDynamicModule,
                IgxCalloutLayerDynamicModule,
                IgxItemToolTipLayerDynamicModule,
                IgxCategoryToolTipLayerDynamicModule,
                IgxCategoryHighlightLayerDynamicModule,
                IgxCategoryItemHighlightLayerDynamicModule,
                IgxDataToolTipLayerDynamicModule,
                IgxNumberAbbreviatorDynamicModule,
                IgxDataChartInteractivityDynamicModule,
                IgxDataChartAnnotationDynamicModule,
                IgxDataPieChartModule
            ], IgxDataPieChartCoreDynamicModule,
            IgxRadialPieSeriesDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxNumberAbbreviatorDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxDataChartAnnotationDynamicModule,
            IgxDataPieChartModule] });
    return IgxDataPieChartDynamicModule;
}());
export { IgxDataPieChartDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataPieChartCoreDynamicModule,
                        IgxRadialPieSeriesDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxNumberAbbreviatorDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxDataChartAnnotationDynamicModule,
                        IgxDataPieChartModule
                    ],
                    imports: [CommonModule, IgxDataPieChartCoreDynamicModule,
                        IgxRadialPieSeriesDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxNumberAbbreviatorDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxDataChartAnnotationDynamicModule,
                        IgxDataPieChartModule
                    ],
                    entryComponents: []
                }]
        }] });
