// igniteui-angular-charts/esm5/lib/igx-typical-price-indicator-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import { IgxFinancialIndicatorComponent } from "./igx-financial-indicator-component";
import { IgxFinancialSeriesComponent } from "./igx-financial-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { TypicalPriceIndicator } from "./TypicalPriceIndicator";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent Typical Price indicator series.
 * The typical price indicator is represented as a arithmetic average
 * of the High, Low, and Close for a day.
 * Default required members: High, Low, Close
*
* The `TypicalPriceIndicator` class represents a IgxDataChartComponent Typical Price indicator series.
*
* ```html
* <igx-data-chart
*     [dataSource]="data">
* 	<igx-category-x-axis
* 		label="label"
* 		#xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
* 		#yAxis>
*     </igx-numeric-y-axis>
* 	<igx-typical-price-indicator
*         #series
*         [xAxis]="xAxis"
*         [yAxis]="yAxis"
*         highMemberPath="high"
* 		lowMemberPath="low"
* 		volumeMemberPath="volume">
*     </igx-typical-price-indicator>
* </igx-data-chart>
* ```
*
* ```ts
* let series = new IgxTypicalPriceIndicatorComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.openMemberPath = "open";
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* series.closeMemberPath = "close";
* this.chart.series.add(series);
* ```
*/
var IgxTypicalPriceIndicatorComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxTypicalPriceIndicatorComponent, _super);
    function IgxTypicalPriceIndicatorComponent() {
        return _super.call(this) || this;
    }
    IgxTypicalPriceIndicatorComponent.prototype.createImplementation = function () {
        return new TypicalPriceIndicator();
    };
    Object.defineProperty(IgxTypicalPriceIndicatorComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgxTypicalPriceIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxTypicalPriceIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxTypicalPriceIndicatorComponent, selector: "igx-typical-price-indicator", providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxTypicalPriceIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxTypicalPriceIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxTypicalPriceIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxTypicalPriceIndicatorComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxTypicalPriceIndicatorComponent;
}(IgxStrategyBasedIndicatorComponent));
export { IgxTypicalPriceIndicatorComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-typical-price-indicator',
                    template: "",
                    providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(function () { return IgxTypicalPriceIndicatorComponent; }) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(function () { return IgxTypicalPriceIndicatorComponent; }) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(function () { return IgxTypicalPriceIndicatorComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxTypicalPriceIndicatorComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
