// igniteui-angular-charts/esm5/lib/igx-data-pie-chart-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataPieChartComponent } from './igx-data-pie-chart-component';
import { IgxProportionalCategoryAngleAxisDynamicModule } from './igx-proportional-category-angle-axis-dynamic-module';
import { IgxNumericRadiusAxisDynamicModule } from './igx-numeric-radius-axis-dynamic-module';
import { FastItemObjectColumn } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxDataPieChartCoreModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataPieChartCoreModule() {
        TypeRegistrar.register('FastItemObjectColumn', FastItemObjectColumn.$type);
    }
    IgxDataPieChartCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataPieChartCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartCoreModule, declarations: [IgxDataPieChartComponent], imports: [CommonModule, IgxProportionalCategoryAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule], exports: [IgxDataPieChartComponent,
            IgxProportionalCategoryAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule] });
    IgxDataPieChartCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartCoreModule, imports: [[CommonModule, IgxProportionalCategoryAngleAxisDynamicModule,
                IgxNumericRadiusAxisDynamicModule
            ], IgxProportionalCategoryAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule] });
    return IgxDataPieChartCoreModule;
}());
export { IgxDataPieChartCoreModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxDataPieChartComponent],
                    exports: [IgxDataPieChartComponent,
                        IgxProportionalCategoryAngleAxisDynamicModule,
                        IgxNumericRadiusAxisDynamicModule
                    ],
                    imports: [CommonModule, IgxProportionalCategoryAngleAxisDynamicModule,
                        IgxNumericRadiusAxisDynamicModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
