// igniteui-angular-charts/esm5/lib/igx-area-fragment-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAreaFragmentModule } from './igx-area-fragment-module';
import { IgxAreaFragmentComponent } from './igx-area-fragment-component';
import { AreaFragment } from './AreaFragment';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxAreaFragmentDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxAreaFragmentDynamicModule() {
        TypeRegistrar.registerCons("IgxAreaFragmentComponent", IgxAreaFragmentComponent);
        TypeRegistrar.register("AreaFragment", AreaFragment.$type);
    }
    IgxAreaFragmentDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaFragmentDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAreaFragmentDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaFragmentDynamicModule, imports: [CommonModule, IgxAreaFragmentModule], exports: [IgxAreaFragmentModule] });
    IgxAreaFragmentDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaFragmentDynamicModule, imports: [[CommonModule, IgxAreaFragmentModule], IgxAreaFragmentModule] });
    return IgxAreaFragmentDynamicModule;
}());
export { IgxAreaFragmentDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaFragmentDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxAreaFragmentModule],
                    exports: [IgxAreaFragmentModule],
                    entryComponents: [IgxAreaFragmentComponent]
                }]
        }], ctorParameters: function () { return []; } });
