// igniteui-angular-charts/esm5/lib/igx-stacked-spline-area-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedSplineAreaSeriesComponent } from './igx-stacked-spline-area-series-component';
import * as i0 from "@angular/core";
var IgxStackedSplineAreaSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStackedSplineAreaSeriesModule() {
    }
    IgxStackedSplineAreaSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineAreaSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedSplineAreaSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineAreaSeriesModule, declarations: [IgxStackedSplineAreaSeriesComponent], imports: [CommonModule], exports: [IgxStackedSplineAreaSeriesComponent] });
    IgxStackedSplineAreaSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineAreaSeriesModule, imports: [[CommonModule]] });
    return IgxStackedSplineAreaSeriesModule;
}());
export { IgxStackedSplineAreaSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineAreaSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStackedSplineAreaSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxStackedSplineAreaSeriesComponent],
                    entryComponents: []
                }]
        }] });
