// igniteui-angular-charts/esm5/lib/igx-detrended-price-oscillator-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDetrendedPriceOscillatorIndicatorModule } from './igx-detrended-price-oscillator-indicator-module';
import { IgxDetrendedPriceOscillatorIndicatorComponent } from './igx-detrended-price-oscillator-indicator-component';
import { DetrendedPriceOscillatorIndicator } from './DetrendedPriceOscillatorIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxDetrendedPriceOscillatorIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDetrendedPriceOscillatorIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxDetrendedPriceOscillatorIndicatorComponent", IgxDetrendedPriceOscillatorIndicatorComponent);
        TypeRegistrar.register("DetrendedPriceOscillatorIndicator", DetrendedPriceOscillatorIndicator.$type);
    }
    IgxDetrendedPriceOscillatorIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDetrendedPriceOscillatorIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDetrendedPriceOscillatorIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDetrendedPriceOscillatorIndicatorDynamicModule, imports: [CommonModule, IgxDetrendedPriceOscillatorIndicatorModule], exports: [IgxDetrendedPriceOscillatorIndicatorModule] });
    IgxDetrendedPriceOscillatorIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDetrendedPriceOscillatorIndicatorDynamicModule, imports: [[CommonModule, IgxDetrendedPriceOscillatorIndicatorModule], IgxDetrendedPriceOscillatorIndicatorModule] });
    return IgxDetrendedPriceOscillatorIndicatorDynamicModule;
}());
export { IgxDetrendedPriceOscillatorIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDetrendedPriceOscillatorIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxDetrendedPriceOscillatorIndicatorModule],
                    exports: [IgxDetrendedPriceOscillatorIndicatorModule],
                    entryComponents: [IgxDetrendedPriceOscillatorIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
