// igniteui-angular-charts/esm5/lib/igx-sparkline-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSparklineCoreDynamicModule } from './igx-sparkline-core-dynamic-module';
import { IgxDataChartInteractivityDynamicModule } from './igx-data-chart-interactivity-dynamic-module';
import { IgxSparklineModule } from './igx-sparkline-module';
import * as i0 from "@angular/core";
var IgxSparklineDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxSparklineDynamicModule() {
    }
    IgxSparklineDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSparklineDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineDynamicModule, imports: [CommonModule, IgxSparklineCoreDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxSparklineModule], exports: [IgxSparklineCoreDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxSparklineModule] });
    IgxSparklineDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineDynamicModule, imports: [[CommonModule, IgxSparklineCoreDynamicModule,
                IgxDataChartInteractivityDynamicModule,
                IgxSparklineModule
            ], IgxSparklineCoreDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxSparklineModule] });
    return IgxSparklineDynamicModule;
}());
export { IgxSparklineDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxSparklineCoreDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxSparklineModule
                    ],
                    imports: [CommonModule, IgxSparklineCoreDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxSparklineModule
                    ],
                    entryComponents: []
                }]
        }] });
