// igniteui-angular-charts/esm5/lib/igx-numeric-x-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericXAxisComponent } from './igx-numeric-x-axis-component';
import * as i0 from "@angular/core";
var IgxNumericXAxisModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxNumericXAxisModule() {
    }
    IgxNumericXAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericXAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxNumericXAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericXAxisModule, declarations: [IgxNumericXAxisComponent], imports: [CommonModule], exports: [IgxNumericXAxisComponent] });
    IgxNumericXAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericXAxisModule, imports: [[CommonModule]] });
    return IgxNumericXAxisModule;
}());
export { IgxNumericXAxisModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericXAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxNumericXAxisComponent],
                    imports: [CommonModule],
                    exports: [IgxNumericXAxisComponent],
                    entryComponents: []
                }]
        }] });
