// igniteui-angular-charts/esm5/lib/igx-price-channel-overlay-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxOverlayProxyModule } from './igx-overlay-proxy-module';
import { IgxPriceChannelOverlayComponent } from './igx-price-channel-overlay-component';
import * as i0 from "@angular/core";
var IgxPriceChannelOverlayModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPriceChannelOverlayModule() {
    }
    IgxPriceChannelOverlayModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceChannelOverlayModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPriceChannelOverlayModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceChannelOverlayModule, declarations: [IgxPriceChannelOverlayComponent], imports: [CommonModule, IgxOverlayProxyModule], exports: [IgxPriceChannelOverlayComponent, IgxOverlayProxyModule] });
    IgxPriceChannelOverlayModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceChannelOverlayModule, imports: [[CommonModule, IgxOverlayProxyModule], IgxOverlayProxyModule] });
    return IgxPriceChannelOverlayModule;
}());
export { IgxPriceChannelOverlayModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceChannelOverlayModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxPriceChannelOverlayComponent],
                    imports: [CommonModule, IgxOverlayProxyModule],
                    exports: [IgxPriceChannelOverlayComponent, IgxOverlayProxyModule],
                    entryComponents: []
                }]
        }] });
