// igniteui-angular-charts/esm5/lib/igx-stacked-spline-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedSplineSeriesModule } from './igx-stacked-spline-series-module';
import { IgxStackedSplineSeriesComponent } from './igx-stacked-spline-series-component';
import { StackedSplineSeries } from './StackedSplineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxStackedSplineSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStackedSplineSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxStackedSplineSeriesComponent", IgxStackedSplineSeriesComponent);
        TypeRegistrar.register("StackedSplineSeries", StackedSplineSeries.$type);
    }
    IgxStackedSplineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedSplineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineSeriesDynamicModule, imports: [CommonModule, IgxStackedSplineSeriesModule], exports: [IgxStackedSplineSeriesModule] });
    IgxStackedSplineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineSeriesDynamicModule, imports: [[CommonModule, IgxStackedSplineSeriesModule], IgxStackedSplineSeriesModule] });
    return IgxStackedSplineSeriesDynamicModule;
}());
export { IgxStackedSplineSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedSplineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStackedSplineSeriesModule],
                    exports: [IgxStackedSplineSeriesModule],
                    entryComponents: [IgxStackedSplineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
