// igniteui-angular-charts/esm5/lib/igx-data-chart-annotation-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AnnotationFactoryManager } from './AnnotationFactoryManager';
import { IgxCrosshairLayerModule } from './igx-crosshair-layer-module';
import { IgxCategoryToolTipLayerModule } from './igx-category-tool-tip-layer-module';
import { IgxItemToolTipLayerModule } from './igx-item-tool-tip-layer-module';
import { IgxCategoryItemHighlightLayerModule } from './igx-category-item-highlight-layer-module';
import { IgxCategoryHighlightLayerModule } from './igx-category-highlight-layer-module';
import { IgxFinalValueLayerModule } from './igx-final-value-layer-module';
import { IgxCalloutLayerModule } from './igx-callout-layer-module';
import { IgxDataToolTipLayerModule } from './igx-data-tool-tip-layer-module';
import { IgxValueLayerModule } from './igx-value-layer-module';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxDataChartAnnotationModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataChartAnnotationModule() {
        TypeRegistrar.register('AnnotationFactoryManager', AnnotationFactoryManager.$type);
    }
    IgxDataChartAnnotationModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartAnnotationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartAnnotationModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartAnnotationModule, imports: [CommonModule, IgxCrosshairLayerModule,
            IgxCategoryToolTipLayerModule,
            IgxItemToolTipLayerModule,
            IgxCategoryItemHighlightLayerModule,
            IgxCategoryHighlightLayerModule,
            IgxFinalValueLayerModule,
            IgxCalloutLayerModule,
            IgxDataToolTipLayerModule,
            IgxValueLayerModule], exports: [IgxCrosshairLayerModule,
            IgxCategoryToolTipLayerModule,
            IgxItemToolTipLayerModule,
            IgxCategoryItemHighlightLayerModule,
            IgxCategoryHighlightLayerModule,
            IgxFinalValueLayerModule,
            IgxCalloutLayerModule,
            IgxDataToolTipLayerModule,
            IgxValueLayerModule] });
    IgxDataChartAnnotationModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartAnnotationModule, imports: [[CommonModule, IgxCrosshairLayerModule,
                IgxCategoryToolTipLayerModule,
                IgxItemToolTipLayerModule,
                IgxCategoryItemHighlightLayerModule,
                IgxCategoryHighlightLayerModule,
                IgxFinalValueLayerModule,
                IgxCalloutLayerModule,
                IgxDataToolTipLayerModule,
                IgxValueLayerModule
            ], IgxCrosshairLayerModule,
            IgxCategoryToolTipLayerModule,
            IgxItemToolTipLayerModule,
            IgxCategoryItemHighlightLayerModule,
            IgxCategoryHighlightLayerModule,
            IgxFinalValueLayerModule,
            IgxCalloutLayerModule,
            IgxDataToolTipLayerModule,
            IgxValueLayerModule] });
    return IgxDataChartAnnotationModule;
}());
export { IgxDataChartAnnotationModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartAnnotationModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxCrosshairLayerModule,
                        IgxCategoryToolTipLayerModule,
                        IgxItemToolTipLayerModule,
                        IgxCategoryItemHighlightLayerModule,
                        IgxCategoryHighlightLayerModule,
                        IgxFinalValueLayerModule,
                        IgxCalloutLayerModule,
                        IgxDataToolTipLayerModule,
                        IgxValueLayerModule
                    ],
                    imports: [CommonModule, IgxCrosshairLayerModule,
                        IgxCategoryToolTipLayerModule,
                        IgxItemToolTipLayerModule,
                        IgxCategoryItemHighlightLayerModule,
                        IgxCategoryHighlightLayerModule,
                        IgxFinalValueLayerModule,
                        IgxCalloutLayerModule,
                        IgxDataToolTipLayerModule,
                        IgxValueLayerModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
