// igniteui-angular-charts/esm5/lib/igx-category-highlight-layer-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryHighlightLayerModule } from './igx-category-highlight-layer-module';
import { IgxCategoryHighlightLayerComponent } from './igx-category-highlight-layer-component';
import { CategoryHighlightLayer } from './CategoryHighlightLayer';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxCategoryHighlightLayerDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCategoryHighlightLayerDynamicModule() {
        TypeRegistrar.registerCons("IgxCategoryHighlightLayerComponent", IgxCategoryHighlightLayerComponent);
        TypeRegistrar.register("CategoryHighlightLayer", CategoryHighlightLayer.$type);
    }
    IgxCategoryHighlightLayerDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryHighlightLayerDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryHighlightLayerDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryHighlightLayerDynamicModule, imports: [CommonModule, IgxCategoryHighlightLayerModule], exports: [IgxCategoryHighlightLayerModule] });
    IgxCategoryHighlightLayerDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryHighlightLayerDynamicModule, imports: [[CommonModule, IgxCategoryHighlightLayerModule], IgxCategoryHighlightLayerModule] });
    return IgxCategoryHighlightLayerDynamicModule;
}());
export { IgxCategoryHighlightLayerDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryHighlightLayerDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCategoryHighlightLayerModule],
                    exports: [IgxCategoryHighlightLayerModule],
                    entryComponents: [IgxCategoryHighlightLayerComponent]
                }]
        }], ctorParameters: function () { return []; } });
