// igniteui-angular-charts/esm5/lib/igx-stacked-100-area-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStackedAreaSeriesComponent } from "./igx-stacked-area-series-component";
import { IgxHorizontalStackedSeriesBaseComponent } from "./igx-horizontal-stacked-series-base-component";
import { IgxStackedSeriesBaseComponent } from "./igx-stacked-series-base-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { Stacked100AreaSeries } from "./Stacked100AreaSeries";
import * as i0 from "@angular/core";
/**
 * Represents a stacked area series, where values are presented as percentages of the total.
*/
var IgxStacked100AreaSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxStacked100AreaSeriesComponent, _super);
    function IgxStacked100AreaSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxStacked100AreaSeriesComponent.prototype.createImplementation = function () {
        return new Stacked100AreaSeries();
    };
    Object.defineProperty(IgxStacked100AreaSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxStacked100AreaSeriesComponent.prototype, "isPercentBased", {
        get: function () {
            return this.i.aan;
        },
        enumerable: false,
        configurable: true
    });
    IgxStacked100AreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100AreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStacked100AreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStacked100AreaSeriesComponent, selector: "igx-stacked-100-area-series", providers: [{ provide: IgxStackedAreaSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStacked100AreaSeriesComponent;
}(IgxStackedAreaSeriesComponent));
export { IgxStacked100AreaSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100AreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-stacked-100-area-series',
                    template: "",
                    providers: [{ provide: IgxStackedAreaSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxStacked100AreaSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
