// igniteui-angular-charts/esm5/lib/igx-standard-deviation-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStandardDeviationIndicatorModule } from './igx-standard-deviation-indicator-module';
import { IgxStandardDeviationIndicatorComponent } from './igx-standard-deviation-indicator-component';
import { StandardDeviationIndicator } from './StandardDeviationIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxStandardDeviationIndicatorDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStandardDeviationIndicatorDynamicModule() {
        TypeRegistrar.registerCons("IgxStandardDeviationIndicatorComponent", IgxStandardDeviationIndicatorComponent);
        TypeRegistrar.register("StandardDeviationIndicator", StandardDeviationIndicator.$type);
    }
    IgxStandardDeviationIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStandardDeviationIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStandardDeviationIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStandardDeviationIndicatorDynamicModule, imports: [CommonModule, IgxStandardDeviationIndicatorModule], exports: [IgxStandardDeviationIndicatorModule] });
    IgxStandardDeviationIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStandardDeviationIndicatorDynamicModule, imports: [[CommonModule, IgxStandardDeviationIndicatorModule], IgxStandardDeviationIndicatorModule] });
    return IgxStandardDeviationIndicatorDynamicModule;
}());
export { IgxStandardDeviationIndicatorDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStandardDeviationIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStandardDeviationIndicatorModule],
                    exports: [IgxStandardDeviationIndicatorModule],
                    entryComponents: [IgxStandardDeviationIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
