// igniteui-angular-charts/esm5/lib/igx-polar-spline-area-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPolarSplineAreaSeriesModule } from './igx-polar-spline-area-series-module';
import { IgxPolarSplineAreaSeriesComponent } from './igx-polar-spline-area-series-component';
import { PolarSplineAreaSeries } from './PolarSplineAreaSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxPolarSplineAreaSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPolarSplineAreaSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxPolarSplineAreaSeriesComponent", IgxPolarSplineAreaSeriesComponent);
        TypeRegistrar.register("PolarSplineAreaSeries", PolarSplineAreaSeries.$type);
    }
    IgxPolarSplineAreaSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineAreaSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPolarSplineAreaSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineAreaSeriesDynamicModule, imports: [CommonModule, IgxPolarSplineAreaSeriesModule], exports: [IgxPolarSplineAreaSeriesModule] });
    IgxPolarSplineAreaSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineAreaSeriesDynamicModule, imports: [[CommonModule, IgxPolarSplineAreaSeriesModule], IgxPolarSplineAreaSeriesModule] });
    return IgxPolarSplineAreaSeriesDynamicModule;
}());
export { IgxPolarSplineAreaSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineAreaSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPolarSplineAreaSeriesModule],
                    exports: [IgxPolarSplineAreaSeriesModule],
                    entryComponents: [IgxPolarSplineAreaSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
