// igniteui-angular-charts/esm5/lib/igx-area-fragment-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxHorizontalAnchoredCategorySeriesProxyModule } from './igx-horizontal-anchored-category-series-proxy-module';
import { IgxAreaFragmentComponent } from './igx-area-fragment-component';
import * as i0 from "@angular/core";
var IgxAreaFragmentModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxAreaFragmentModule() {
    }
    IgxAreaFragmentModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaFragmentModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAreaFragmentModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaFragmentModule, declarations: [IgxAreaFragmentComponent], imports: [CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule], exports: [IgxAreaFragmentComponent, IgxHorizontalAnchoredCategorySeriesProxyModule] });
    IgxAreaFragmentModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaFragmentModule, imports: [[CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule], IgxHorizontalAnchoredCategorySeriesProxyModule] });
    return IgxAreaFragmentModule;
}());
export { IgxAreaFragmentModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaFragmentModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxAreaFragmentComponent],
                    imports: [CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule],
                    exports: [IgxAreaFragmentComponent, IgxHorizontalAnchoredCategorySeriesProxyModule],
                    entryComponents: []
                }]
        }] });
