// igniteui-angular-charts/esm5/lib/igx-category-tool-tip-layer-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryToolTipLayerModule } from './igx-category-tool-tip-layer-module';
import { IgxCategoryToolTipLayerComponent } from './igx-category-tool-tip-layer-component';
import { CategoryToolTipLayer } from './CategoryToolTipLayer';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxCategoryToolTipLayerDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCategoryToolTipLayerDynamicModule() {
        TypeRegistrar.registerCons("IgxCategoryToolTipLayerComponent", IgxCategoryToolTipLayerComponent);
        TypeRegistrar.register("CategoryToolTipLayer", CategoryToolTipLayer.$type);
    }
    IgxCategoryToolTipLayerDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryToolTipLayerDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryToolTipLayerDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryToolTipLayerDynamicModule, imports: [CommonModule, IgxCategoryToolTipLayerModule], exports: [IgxCategoryToolTipLayerModule] });
    IgxCategoryToolTipLayerDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryToolTipLayerDynamicModule, imports: [[CommonModule, IgxCategoryToolTipLayerModule], IgxCategoryToolTipLayerModule] });
    return IgxCategoryToolTipLayerDynamicModule;
}());
export { IgxCategoryToolTipLayerDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryToolTipLayerDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCategoryToolTipLayerModule],
                    exports: [IgxCategoryToolTipLayerModule],
                    entryComponents: [IgxCategoryToolTipLayerComponent]
                }]
        }], ctorParameters: function () { return []; } });
