// igniteui-angular-charts/esm5/lib/igx-item-legend-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxItemLegendComponent } from './igx-item-legend-component';
import { LegendTemplates } from './LegendTemplates';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxItemLegendModule } from './igx-item-legend-module';
import * as i0 from "@angular/core";
var IgxItemLegendDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxItemLegendDynamicModule() {
        TypeRegistrar.registerCons('IgxItemLegendComponent', IgxItemLegendComponent);
        TypeRegistrar.registerCons('LegendTemplates', LegendTemplates);
        TypeRegistrar.register('LegendTemplates', LegendTemplates.$type);
    }
    IgxItemLegendDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemLegendDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxItemLegendDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemLegendDynamicModule, imports: [CommonModule, IgxItemLegendModule], exports: [IgxItemLegendModule] });
    IgxItemLegendDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemLegendDynamicModule, imports: [[CommonModule, IgxItemLegendModule
            ], IgxItemLegendModule] });
    return IgxItemLegendDynamicModule;
}());
export { IgxItemLegendDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemLegendDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxItemLegendModule
                    ],
                    imports: [CommonModule, IgxItemLegendModule
                    ],
                    entryComponents: [IgxItemLegendComponent]
                }]
        }], ctorParameters: function () { return []; } });
