// igniteui-angular-charts/esm5/lib/igx-percentage-volume-oscillator-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxPercentageVolumeOscillatorIndicatorComponent } from './igx-percentage-volume-oscillator-indicator-component';
import * as i0 from "@angular/core";
var IgxPercentageVolumeOscillatorIndicatorModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPercentageVolumeOscillatorIndicatorModule() {
    }
    IgxPercentageVolumeOscillatorIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentageVolumeOscillatorIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPercentageVolumeOscillatorIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentageVolumeOscillatorIndicatorModule, declarations: [IgxPercentageVolumeOscillatorIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxPercentageVolumeOscillatorIndicatorComponent, IgxIndicatorProxyModule] });
    IgxPercentageVolumeOscillatorIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentageVolumeOscillatorIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxPercentageVolumeOscillatorIndicatorModule;
}());
export { IgxPercentageVolumeOscillatorIndicatorModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentageVolumeOscillatorIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxPercentageVolumeOscillatorIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxPercentageVolumeOscillatorIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
