// igniteui-angular-charts/esm5/lib/igx-step-area-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStepAreaSeriesModule } from './igx-step-area-series-module';
import { IgxStepAreaSeriesComponent } from './igx-step-area-series-component';
import { StepAreaSeries } from './StepAreaSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxStepAreaSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxStepAreaSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxStepAreaSeriesComponent", IgxStepAreaSeriesComponent);
        TypeRegistrar.register("StepAreaSeries", StepAreaSeries.$type);
    }
    IgxStepAreaSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStepAreaSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStepAreaSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStepAreaSeriesDynamicModule, imports: [CommonModule, IgxStepAreaSeriesModule], exports: [IgxStepAreaSeriesModule] });
    IgxStepAreaSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStepAreaSeriesDynamicModule, imports: [[CommonModule, IgxStepAreaSeriesModule], IgxStepAreaSeriesModule] });
    return IgxStepAreaSeriesDynamicModule;
}());
export { IgxStepAreaSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStepAreaSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStepAreaSeriesModule],
                    exports: [IgxStepAreaSeriesModule],
                    entryComponents: [IgxStepAreaSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
