// igniteui-angular-charts/esm5/lib/igx-series-layer.js
import { IgxSeriesLayerPropertyOverlayCollection } from "./igx-series-layer-property-overlay-collection";
import { SeriesLayer as SeriesLayer_internal } from "./SeriesLayer";
import { SeriesLayerPropertyOverlayCollection as SeriesLayerPropertyOverlayCollection_internal } from "./SeriesLayerPropertyOverlayCollection";
import { SeriesLayerPropertyOverlay } from "./SeriesLayerPropertyOverlay";
import { SyncableObservableCollection$1 } from "igniteui-angular-core";
/**
 * Represents an axis annotation
*/
var IgxSeriesLayer = /** @class */ /*@__PURE__*/ (function () {
    function IgxSeriesLayer() {
        this._propertyOverlays = null;
        this._zoneRunner = null;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgxSeriesLayer.prototype.createImplementation = function () {
        return new SeriesLayer_internal();
    };
    Object.defineProperty(IgxSeriesLayer.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesLayer.prototype.onImplementationCreated = function () {
    };
    IgxSeriesLayer.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    };
    Object.defineProperty(IgxSeriesLayer.prototype, "transitionOutIsInProgress", {
        get: function () {
            return this.i.t;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesLayer.prototype, "zIndex", {
        /**
         * Gets or sets the ZIndex to use for layering the series layers.
        */
        get: function () {
            return this.i.ac;
        },
        set: function (v) {
            this.i.ac = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxSeriesLayer.prototype, "propertyOverlays", {
        /**
         * Gets the current extra layers for the series.
        */
        get: function () {
            if (this._propertyOverlays === null) {
                var coll = new IgxSeriesLayerPropertyOverlayCollection();
                var innerColl = this.i.m;
                if (!innerColl) {
                    innerColl = new SeriesLayerPropertyOverlayCollection_internal();
                }
                this._propertyOverlays = coll._fromInner(innerColl);
            }
            return this._propertyOverlays;
        },
        set: function (v) {
            if (this._propertyOverlays !== null) {
                this._propertyOverlays._setSyncTarget(null);
                this._propertyOverlays = null;
            }
            var coll = new IgxSeriesLayerPropertyOverlayCollection();
            this._propertyOverlays = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(SeriesLayerPropertyOverlay.$type);
            var innerColl = this.i.m;
            if (!innerColl) {
                innerColl = new SeriesLayerPropertyOverlayCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._propertyOverlays._setSyncTarget(syncColl);
        },
        enumerable: false,
        configurable: true
    });
    IgxSeriesLayer.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.propertyOverlays != null && this.propertyOverlays.findByName && this.propertyOverlays.findByName(name)) {
            return this.propertyOverlays.findByName(name);
        }
        return null;
    };
    IgxSeriesLayer.prototype.playTransitionOutAndRemove = function () {
        this.i.au();
    };
    IgxSeriesLayer.prototype.playTransitionIn = function () {
        this.i.at();
    };
    IgxSeriesLayer.prototype._runInZone = function (act) {
        if (this._zoneRunner != null) {
            this._zoneRunner(act);
        }
        else {
            act();
        }
    };
    return IgxSeriesLayer;
}());
export { IgxSeriesLayer };
