// igniteui-angular-charts/esm5/lib/igx-ordinal-time-x-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxOrdinalTimeXAxisComponent } from './igx-ordinal-time-x-axis-component';
import * as i0 from "@angular/core";
var IgxOrdinalTimeXAxisModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxOrdinalTimeXAxisModule() {
    }
    IgxOrdinalTimeXAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOrdinalTimeXAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxOrdinalTimeXAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOrdinalTimeXAxisModule, declarations: [IgxOrdinalTimeXAxisComponent], imports: [CommonModule], exports: [IgxOrdinalTimeXAxisComponent] });
    IgxOrdinalTimeXAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOrdinalTimeXAxisModule, imports: [[CommonModule]] });
    return IgxOrdinalTimeXAxisModule;
}());
export { IgxOrdinalTimeXAxisModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOrdinalTimeXAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxOrdinalTimeXAxisComponent],
                    imports: [CommonModule],
                    exports: [IgxOrdinalTimeXAxisComponent],
                    entryComponents: []
                }]
        }] });
