// igniteui-angular-charts/esm5/lib/igx-category-tool-tip-layer-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAnnotationLayerProxyModule } from './igx-annotation-layer-proxy-module';
import { IgxCategoryToolTipLayerComponent } from './igx-category-tool-tip-layer-component';
import * as i0 from "@angular/core";
var IgxCategoryToolTipLayerModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCategoryToolTipLayerModule() {
    }
    IgxCategoryToolTipLayerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryToolTipLayerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryToolTipLayerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryToolTipLayerModule, declarations: [IgxCategoryToolTipLayerComponent], imports: [CommonModule, IgxAnnotationLayerProxyModule], exports: [IgxCategoryToolTipLayerComponent, IgxAnnotationLayerProxyModule] });
    IgxCategoryToolTipLayerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryToolTipLayerModule, imports: [[CommonModule, IgxAnnotationLayerProxyModule], IgxAnnotationLayerProxyModule] });
    return IgxCategoryToolTipLayerModule;
}());
export { IgxCategoryToolTipLayerModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryToolTipLayerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxCategoryToolTipLayerComponent],
                    imports: [CommonModule, IgxAnnotationLayerProxyModule],
                    exports: [IgxCategoryToolTipLayerComponent, IgxAnnotationLayerProxyModule],
                    entryComponents: []
                }]
        }] });
