// igniteui-angular-charts/esm5/lib/igx-value-overlay-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxValueOverlayComponent } from './igx-value-overlay-component';
import * as i0 from "@angular/core";
var IgxValueOverlayModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxValueOverlayModule() {
    }
    IgxValueOverlayModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueOverlayModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxValueOverlayModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueOverlayModule, declarations: [IgxValueOverlayComponent], imports: [CommonModule], exports: [IgxValueOverlayComponent] });
    IgxValueOverlayModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueOverlayModule, imports: [[CommonModule]] });
    return IgxValueOverlayModule;
}());
export { IgxValueOverlayModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueOverlayModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxValueOverlayComponent],
                    imports: [CommonModule],
                    exports: [IgxValueOverlayComponent],
                    entryComponents: []
                }]
        }] });
