// igniteui-angular-charts/esm5/lib/igx-series-matcher-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSeriesMatcherModule } from './igx-series-matcher-module';
import { IgxSeriesMatcher } from './igx-series-matcher';
import { SeriesMatcher } from './SeriesMatcher';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxSeriesMatcherDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxSeriesMatcherDynamicModule() {
        TypeRegistrar.registerCons("IgxSeriesMatcher", IgxSeriesMatcher);
        TypeRegistrar.register("SeriesMatcher", SeriesMatcher.$type);
    }
    IgxSeriesMatcherDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSeriesMatcherDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSeriesMatcherDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSeriesMatcherDynamicModule, imports: [CommonModule, IgxSeriesMatcherModule], exports: [IgxSeriesMatcherModule] });
    IgxSeriesMatcherDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSeriesMatcherDynamicModule, imports: [[CommonModule, IgxSeriesMatcherModule], IgxSeriesMatcherModule] });
    return IgxSeriesMatcherDynamicModule;
}());
export { IgxSeriesMatcherDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSeriesMatcherDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxSeriesMatcherModule],
                    exports: [IgxSeriesMatcherModule],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
