// igniteui-angular-charts/esm5/lib/igx-category-chart-toolbar-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataChartToolbarIconsDynamicModule } from './igx-data-chart-toolbar-icons-dynamic-module';
import { DataChartToolbarFactoryManager } from './DataChartToolbarFactoryManager';
import { CategoryChartToolbarFactoryManager } from './CategoryChartToolbarFactoryManager';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxCategoryChartToolbarModule } from './igx-category-chart-toolbar-module';
import * as i0 from "@angular/core";
var IgxCategoryChartToolbarDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCategoryChartToolbarDynamicModule() {
        TypeRegistrar.register('DataChartToolbarFactoryManager', DataChartToolbarFactoryManager.$type);
        DataChartToolbarFactoryManager.register();
        TypeRegistrar.register('CategoryChartToolbarFactoryManager', CategoryChartToolbarFactoryManager.$type);
        CategoryChartToolbarFactoryManager.register();
    }
    IgxCategoryChartToolbarDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartToolbarDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryChartToolbarDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartToolbarDynamicModule, imports: [CommonModule, IgxDataChartToolbarIconsDynamicModule,
            IgxCategoryChartToolbarModule], exports: [IgxDataChartToolbarIconsDynamicModule,
            IgxCategoryChartToolbarModule] });
    IgxCategoryChartToolbarDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartToolbarDynamicModule, imports: [[CommonModule, IgxDataChartToolbarIconsDynamicModule,
                IgxCategoryChartToolbarModule
            ], IgxDataChartToolbarIconsDynamicModule,
            IgxCategoryChartToolbarModule] });
    return IgxCategoryChartToolbarDynamicModule;
}());
export { IgxCategoryChartToolbarDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartToolbarDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataChartToolbarIconsDynamicModule,
                        IgxCategoryChartToolbarModule
                    ],
                    imports: [CommonModule, IgxDataChartToolbarIconsDynamicModule,
                        IgxCategoryChartToolbarModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
