// igniteui-angular-charts/esm5/lib/igx-polar-area-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPolarAreaSeriesComponent } from './igx-polar-area-series-component';
import * as i0 from "@angular/core";
var IgxPolarAreaSeriesModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPolarAreaSeriesModule() {
    }
    IgxPolarAreaSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarAreaSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPolarAreaSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarAreaSeriesModule, declarations: [IgxPolarAreaSeriesComponent], imports: [CommonModule], exports: [IgxPolarAreaSeriesComponent] });
    IgxPolarAreaSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarAreaSeriesModule, imports: [[CommonModule]] });
    return IgxPolarAreaSeriesModule;
}());
export { IgxPolarAreaSeriesModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarAreaSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxPolarAreaSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxPolarAreaSeriesComponent],
                    entryComponents: []
                }]
        }] });
