// igniteui-angular-charts/esm5/lib/igx-percent-change-y-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPercentChangeYAxisComponent } from './igx-percent-change-y-axis-component';
import * as i0 from "@angular/core";
var IgxPercentChangeYAxisModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPercentChangeYAxisModule() {
    }
    IgxPercentChangeYAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentChangeYAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPercentChangeYAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentChangeYAxisModule, declarations: [IgxPercentChangeYAxisComponent], imports: [CommonModule], exports: [IgxPercentChangeYAxisComponent] });
    IgxPercentChangeYAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentChangeYAxisModule, imports: [[CommonModule]] });
    return IgxPercentChangeYAxisModule;
}());
export { IgxPercentChangeYAxisModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentChangeYAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxPercentChangeYAxisComponent],
                    imports: [CommonModule],
                    exports: [IgxPercentChangeYAxisComponent],
                    entryComponents: []
                }]
        }] });
