// igniteui-angular-charts/esm5/lib/igx-data-tool-tip-layer-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataLegendDynamicModule } from "./igx-data-legend-dynamic-module";
import { IgxDataToolTipLayerComponent } from './igx-data-tool-tip-layer-component';
import * as i0 from "@angular/core";
var IgxDataToolTipLayerModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataToolTipLayerModule() {
    }
    IgxDataToolTipLayerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataToolTipLayerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataToolTipLayerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataToolTipLayerModule, declarations: [IgxDataToolTipLayerComponent], imports: [CommonModule, IgxDataLegendDynamicModule], exports: [IgxDataToolTipLayerComponent, IgxDataLegendDynamicModule] });
    IgxDataToolTipLayerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataToolTipLayerModule, imports: [[CommonModule, IgxDataLegendDynamicModule], IgxDataLegendDynamicModule] });
    return IgxDataToolTipLayerModule;
}());
export { IgxDataToolTipLayerModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataToolTipLayerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxDataToolTipLayerComponent],
                    imports: [CommonModule, IgxDataLegendDynamicModule],
                    exports: [IgxDataToolTipLayerComponent, IgxDataLegendDynamicModule],
                    entryComponents: []
                }]
        }] });
