// igniteui-angular-charts/esm5/lib/igx-scale-legend-component.js
import { __extends } from "tslib";
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxLegendBaseComponent } from "./igx-legend-base-component";
import { ScaleLegend } from "./ScaleLegend";
import * as i0 from "@angular/core";
/**
 * Represents a legend that indicates the size and the color scale for a collection of series.
*/
var IgxScaleLegendComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxScaleLegendComponent, _super);
    function IgxScaleLegendComponent() {
        return _super.call(this) || this;
    }
    IgxScaleLegendComponent.prototype.createImplementation = function () {
        return new ScaleLegend();
    };
    Object.defineProperty(IgxScaleLegendComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgxScaleLegendComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxScaleLegendComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxScaleLegendComponent, selector: "igx-scale-legend", providers: [{ provide: IgxLegendBaseComponent, useExisting: forwardRef(function () { return IgxScaleLegendComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxScaleLegendComponent;
}(IgxLegendBaseComponent));
export { IgxScaleLegendComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-scale-legend',
                    template: "",
                    providers: [{ provide: IgxLegendBaseComponent, useExisting: forwardRef(function () { return IgxScaleLegendComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
