// igniteui-angular-charts/esm5/lib/igx-custom-palette-brush-scale-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCustomPaletteBrushScaleModule } from './igx-custom-palette-brush-scale-module';
import { IgxCustomPaletteBrushScaleComponent } from './igx-custom-palette-brush-scale-component';
import { CustomPaletteBrushScale } from './CustomPaletteBrushScale';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxCustomPaletteBrushScaleDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCustomPaletteBrushScaleDynamicModule() {
        TypeRegistrar.registerCons("IgxCustomPaletteBrushScaleComponent", IgxCustomPaletteBrushScaleComponent);
        TypeRegistrar.register("CustomPaletteBrushScale", CustomPaletteBrushScale.$type);
    }
    IgxCustomPaletteBrushScaleDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteBrushScaleDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCustomPaletteBrushScaleDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteBrushScaleDynamicModule, imports: [CommonModule, IgxCustomPaletteBrushScaleModule], exports: [IgxCustomPaletteBrushScaleModule] });
    IgxCustomPaletteBrushScaleDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteBrushScaleDynamicModule, imports: [[CommonModule, IgxCustomPaletteBrushScaleModule], IgxCustomPaletteBrushScaleModule] });
    return IgxCustomPaletteBrushScaleDynamicModule;
}());
export { IgxCustomPaletteBrushScaleDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteBrushScaleDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCustomPaletteBrushScaleModule],
                    exports: [IgxCustomPaletteBrushScaleModule],
                    entryComponents: [IgxCustomPaletteBrushScaleComponent]
                }]
        }], ctorParameters: function () { return []; } });
