// igniteui-angular-charts/esm5/lib/igx-custom-palette-brush-scale-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCustomPaletteBrushScaleComponent } from './igx-custom-palette-brush-scale-component';
import * as i0 from "@angular/core";
var IgxCustomPaletteBrushScaleModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxCustomPaletteBrushScaleModule() {
    }
    IgxCustomPaletteBrushScaleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteBrushScaleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCustomPaletteBrushScaleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteBrushScaleModule, declarations: [IgxCustomPaletteBrushScaleComponent], imports: [CommonModule], exports: [IgxCustomPaletteBrushScaleComponent] });
    IgxCustomPaletteBrushScaleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteBrushScaleModule, imports: [[CommonModule]] });
    return IgxCustomPaletteBrushScaleModule;
}());
export { IgxCustomPaletteBrushScaleModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteBrushScaleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxCustomPaletteBrushScaleComponent],
                    imports: [CommonModule],
                    exports: [IgxCustomPaletteBrushScaleComponent],
                    entryComponents: []
                }]
        }] });
