// igniteui-angular-charts/esm5/lib/igx-sparkline-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSparklineComponent } from './igx-sparkline-component';
import { FastItemObjectColumn } from "igniteui-angular-core";
import { IgxNumberFormatSpecifierModule } from "igniteui-angular-core";
import { IgxDateTimeFormatSpecifierModule } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxSparklineCoreModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxSparklineCoreModule() {
        TypeRegistrar.register('FastItemObjectColumn', FastItemObjectColumn.$type);
    }
    IgxSparklineCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSparklineCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineCoreModule, declarations: [IgxSparklineComponent], imports: [CommonModule, IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule], exports: [IgxSparklineComponent,
            IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule] });
    IgxSparklineCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineCoreModule, imports: [[CommonModule, IgxNumberFormatSpecifierModule,
                IgxDateTimeFormatSpecifierModule
            ], IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule] });
    return IgxSparklineCoreModule;
}());
export { IgxSparklineCoreModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxSparklineComponent],
                    exports: [IgxSparklineComponent,
                        IgxNumberFormatSpecifierModule,
                        IgxDateTimeFormatSpecifierModule
                    ],
                    imports: [CommonModule, IgxNumberFormatSpecifierModule,
                        IgxDateTimeFormatSpecifierModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
