// igniteui-angular-charts/esm5/lib/igx-radial-pie-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxRadialPieSeriesModule } from './igx-radial-pie-series-module';
import { IgxRadialPieSeriesComponent } from './igx-radial-pie-series-component';
import { RadialPieSeries } from './RadialPieSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxRadialPieSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxRadialPieSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxRadialPieSeriesComponent", IgxRadialPieSeriesComponent);
        TypeRegistrar.register("RadialPieSeries", RadialPieSeries.$type);
    }
    IgxRadialPieSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialPieSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxRadialPieSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialPieSeriesDynamicModule, imports: [CommonModule, IgxRadialPieSeriesModule], exports: [IgxRadialPieSeriesModule] });
    IgxRadialPieSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialPieSeriesDynamicModule, imports: [[CommonModule, IgxRadialPieSeriesModule], IgxRadialPieSeriesModule] });
    return IgxRadialPieSeriesDynamicModule;
}());
export { IgxRadialPieSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialPieSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxRadialPieSeriesModule],
                    exports: [IgxRadialPieSeriesModule],
                    entryComponents: [IgxRadialPieSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
