// igniteui-angular-charts/esm5/lib/igx-polar-line-series-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { UnknownValuePlotting_$type } from "igniteui-angular-core";
import { IgxPolarLineSeriesBaseComponent } from "./igx-polar-line-series-base-component";
import { IgxPolarBaseComponent } from "./igx-polar-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { PolarLineSeries } from "./PolarLineSeries";
import { ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent polar line series.
*/
var IgxPolarLineSeriesComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxPolarLineSeriesComponent, _super);
    function IgxPolarLineSeriesComponent() {
        return _super.call(this) || this;
    }
    IgxPolarLineSeriesComponent.prototype.createImplementation = function () {
        return new PolarLineSeries();
    };
    Object.defineProperty(IgxPolarLineSeriesComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxPolarLineSeriesComponent.prototype, "isLineOnly", {
        /**
         * Gets whether the current series shows only line shapes.
        */
        get: function () {
            return this.i.fi;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxPolarLineSeriesComponent.prototype, "unknownValuePlotting", {
        /**
         * Determines how unknown values will be plotted on the chart.
         * Null and Double.NaN are two examples of unknown values.
        */
        get: function () {
            return this.i.acn;
        },
        set: function (v) {
            this.i.acn = ensureEnum(UnknownValuePlotting_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    IgxPolarLineSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxPolarLineSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxPolarLineSeriesComponent, selector: "igx-polar-line-series", inputs: { unknownValuePlotting: "unknownValuePlotting" }, providers: [{ provide: IgxPolarLineSeriesBaseComponent, useExisting: forwardRef(function () { return IgxPolarLineSeriesComponent; }) }, { provide: IgxPolarBaseComponent, useExisting: forwardRef(function () { return IgxPolarLineSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxPolarLineSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxPolarLineSeriesComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxPolarLineSeriesComponent;
}(IgxPolarLineSeriesBaseComponent));
export { IgxPolarLineSeriesComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-polar-line-series',
                    template: "",
                    providers: [{ provide: IgxPolarLineSeriesBaseComponent, useExisting: forwardRef(function () { return IgxPolarLineSeriesComponent; }) }, { provide: IgxPolarBaseComponent, useExisting: forwardRef(function () { return IgxPolarLineSeriesComponent; }) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(function () { return IgxPolarLineSeriesComponent; }) }, { provide: IgxSeriesComponent, useExisting: forwardRef(function () { return IgxPolarLineSeriesComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { unknownValuePlotting: [{
                type: Input
            }] } });
