// igniteui-angular-charts/esm5/lib/igx-value-layer-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAnnotationLayerProxyModule } from './igx-annotation-layer-proxy-module';
import { IgxValueLayerComponent } from './igx-value-layer-component';
import * as i0 from "@angular/core";
var IgxValueLayerModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxValueLayerModule() {
    }
    IgxValueLayerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueLayerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxValueLayerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueLayerModule, declarations: [IgxValueLayerComponent], imports: [CommonModule, IgxAnnotationLayerProxyModule], exports: [IgxValueLayerComponent, IgxAnnotationLayerProxyModule] });
    IgxValueLayerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueLayerModule, imports: [[CommonModule, IgxAnnotationLayerProxyModule], IgxAnnotationLayerProxyModule] });
    return IgxValueLayerModule;
}());
export { IgxValueLayerModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxValueLayerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxValueLayerComponent],
                    imports: [CommonModule, IgxAnnotationLayerProxyModule],
                    exports: [IgxValueLayerComponent, IgxAnnotationLayerProxyModule],
                    entryComponents: []
                }]
        }] });
