// igniteui-angular-charts/esm5/lib/igx-polar-line-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPolarLineSeriesModule } from './igx-polar-line-series-module';
import { IgxPolarLineSeriesComponent } from './igx-polar-line-series-component';
import { PolarLineSeries } from './PolarLineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxPolarLineSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxPolarLineSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxPolarLineSeriesComponent", IgxPolarLineSeriesComponent);
        TypeRegistrar.register("PolarLineSeries", PolarLineSeries.$type);
    }
    IgxPolarLineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPolarLineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesDynamicModule, imports: [CommonModule, IgxPolarLineSeriesModule], exports: [IgxPolarLineSeriesModule] });
    IgxPolarLineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesDynamicModule, imports: [[CommonModule, IgxPolarLineSeriesModule], IgxPolarLineSeriesModule] });
    return IgxPolarLineSeriesDynamicModule;
}());
export { IgxPolarLineSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPolarLineSeriesModule],
                    exports: [IgxPolarLineSeriesModule],
                    entryComponents: [IgxPolarLineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
