// igniteui-angular-charts/esm5/lib/igx-numeric-radius-axis-component.js
import { __extends } from "tslib";
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxNumericAxisBaseComponent } from "./igx-numeric-axis-base-component";
import { IgxAxisComponent } from "./igx-axis-component";
import { NumericRadiusAxis } from "./NumericRadiusAxis";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent radius axis for polar and radial series.
*
* ```html
* <igx-data-chart
*     [dataSource]="dataSource"   >
*     <igx-category-angle-axis name="angleAxis" label="Department" />
*     <igx-numeric-radius-axis name="radiusAxis" />
*  </igx-data-chart>
* ```
*
* ```ts
* this.series1.angleAxisName = "AngleAxis";
* this.series1.valueAxisName = "numericYAxis";
* ```
*/
var IgxNumericRadiusAxisComponent = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgxNumericRadiusAxisComponent, _super);
    function IgxNumericRadiusAxisComponent() {
        return _super.call(this) || this;
    }
    IgxNumericRadiusAxisComponent.prototype.createImplementation = function () {
        return new NumericRadiusAxis();
    };
    Object.defineProperty(IgxNumericRadiusAxisComponent.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxNumericRadiusAxisComponent.prototype, "isRadial", {
        /**
         * Checks if the axis is of radial axis type
        *
        * ```ts
        * let isRadial: boolean = this.series.isRadial;
        * ```
        */
        get: function () {
            return this.i.cm;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxNumericRadiusAxisComponent.prototype, "radiusExtentScale", {
        /**
         * Defines the percentage of the maximum radius extent to use as the maximum radius. Should be
         * a value between 0.0 and 1.0.
        */
        get: function () {
            return this.i.nz;
        },
        set: function (v) {
            this.i.nz = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgxNumericRadiusAxisComponent.prototype, "innerRadiusExtentScale", {
        /**
         * Defines the percentage of the maximum radius extent to leave blank at the center of the chart. Should be
         * a value between 0.0 and 1.0.
        */
        get: function () {
            return this.i.ny;
        },
        set: function (v) {
            this.i.ny = +v;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Returns a world coordinates radius length (0 - 0.5) from a raw axis value.
    
    * @param unscaledValue  * The raw axis value.
    */
    IgxNumericRadiusAxisComponent.prototype.getScaledValue = function (unscaledValue) {
        var iv = this.i.nv(unscaledValue);
        return (iv);
    };
    /**
     * Returns a raw axis value from the world coordinates radius length provided.
    
    * @param scaledValue  * The scaled world coordinates radius length.
    *
    * Class containing several properties which are used as parameters passed to scaling operations in a `SeriesViewer`.
    */
    IgxNumericRadiusAxisComponent.prototype.getUnscaledValue = function (scaledValue) {
        var iv = this.i.nx(scaledValue);
        return (iv);
    };
    IgxNumericRadiusAxisComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericRadiusAxisComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxNumericRadiusAxisComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxNumericRadiusAxisComponent, selector: "igx-numeric-radius-axis", inputs: { radiusExtentScale: "radiusExtentScale", innerRadiusExtentScale: "innerRadiusExtentScale" }, providers: [{ provide: IgxNumericAxisBaseComponent, useExisting: forwardRef(function () { return IgxNumericRadiusAxisComponent; }) }, { provide: IgxAxisComponent, useExisting: forwardRef(function () { return IgxNumericRadiusAxisComponent; }) }], usesInheritance: true, ngImport: i0, template: "", isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxNumericRadiusAxisComponent;
}(IgxNumericAxisBaseComponent));
export { IgxNumericRadiusAxisComponent };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericRadiusAxisComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-numeric-radius-axis',
                    template: "",
                    providers: [{ provide: IgxNumericAxisBaseComponent, useExisting: forwardRef(function () { return IgxNumericRadiusAxisComponent; }) }, { provide: IgxAxisComponent, useExisting: forwardRef(function () { return IgxNumericRadiusAxisComponent; }) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { radiusExtentScale: [{
                type: Input
            }], innerRadiusExtentScale: [{
                type: Input
            }] } });
