// igniteui-angular-charts/esm5/lib/igx-high-density-scatter-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxHighDensityScatterSeriesModule } from './igx-high-density-scatter-series-module';
import { IgxHighDensityScatterSeriesComponent } from './igx-high-density-scatter-series-component';
import { HighDensityScatterSeries } from './HighDensityScatterSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxHighDensityScatterSeriesDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxHighDensityScatterSeriesDynamicModule() {
        TypeRegistrar.registerCons("IgxHighDensityScatterSeriesComponent", IgxHighDensityScatterSeriesComponent);
        TypeRegistrar.register("HighDensityScatterSeries", HighDensityScatterSeries.$type);
    }
    IgxHighDensityScatterSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHighDensityScatterSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxHighDensityScatterSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHighDensityScatterSeriesDynamicModule, imports: [CommonModule, IgxHighDensityScatterSeriesModule], exports: [IgxHighDensityScatterSeriesModule] });
    IgxHighDensityScatterSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHighDensityScatterSeriesDynamicModule, imports: [[CommonModule, IgxHighDensityScatterSeriesModule], IgxHighDensityScatterSeriesModule] });
    return IgxHighDensityScatterSeriesDynamicModule;
}());
export { IgxHighDensityScatterSeriesDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHighDensityScatterSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxHighDensityScatterSeriesModule],
                    exports: [IgxHighDensityScatterSeriesModule],
                    entryComponents: [IgxHighDensityScatterSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
