// igniteui-angular-charts/esm5/lib/igx-zoom-slider-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxZoomSliderComponent } from './igx-zoom-slider-component';
import { ZoomSlider } from './ZoomSlider';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
var IgxZoomSliderModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxZoomSliderModule() {
        TypeRegistrar.registerCons('IgxZoomSliderComponent', IgxZoomSliderComponent);
        TypeRegistrar.register('ZoomSlider', ZoomSlider.$type);
    }
    IgxZoomSliderModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxZoomSliderModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxZoomSliderModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxZoomSliderModule, declarations: [IgxZoomSliderComponent], imports: [CommonModule], exports: [IgxZoomSliderComponent] });
    IgxZoomSliderModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxZoomSliderModule, imports: [[CommonModule,
            ]] });
    return IgxZoomSliderModule;
}());
export { IgxZoomSliderModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxZoomSliderModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxZoomSliderComponent],
                    exports: [IgxZoomSliderComponent
                    ],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
