// igniteui-angular-charts/esm5/lib/igx-linear-contour-value-resolver-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxLinearContourValueResolverComponent } from './igx-linear-contour-value-resolver-component';
import * as i0 from "@angular/core";
var IgxLinearContourValueResolverModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxLinearContourValueResolverModule() {
    }
    IgxLinearContourValueResolverModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxLinearContourValueResolverModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverModule, declarations: [IgxLinearContourValueResolverComponent], imports: [CommonModule], exports: [IgxLinearContourValueResolverComponent] });
    IgxLinearContourValueResolverModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverModule, imports: [[CommonModule]] });
    return IgxLinearContourValueResolverModule;
}());
export { IgxLinearContourValueResolverModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxLinearContourValueResolverComponent],
                    imports: [CommonModule],
                    exports: [IgxLinearContourValueResolverComponent],
                    entryComponents: []
                }]
        }] });
