// igniteui-angular-charts/esm5/lib/MoneyFlowIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { MoneyFlowIndexIndicatorStrategy } from "./MoneyFlowIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var MoneyFlowIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(MoneyFlowIndexIndicator, _super);
    function MoneyFlowIndexIndicator() {
        return _super.call(this) || this;
    }
    MoneyFlowIndexIndicator.prototype.bi = function () {
        return new MoneyFlowIndexIndicator();
    };
    MoneyFlowIndexIndicator.prototype.get_aau = function () {
        return 1;
    };
    MoneyFlowIndexIndicator.prototype.get_abn = function () {
        return new MoneyFlowIndexIndicatorStrategy();
    };
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    MoneyFlowIndexIndicator.prototype.get_ab0 = function () {
        return MoneyFlowIndexIndicator.$;
    };
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(MoneyFlowIndexIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(MoneyFlowIndexIndicator.ab9));
        },
        set: function (a) {
            this.h(MoneyFlowIndexIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    MoneyFlowIndexIndicator.prototype.abt = function () {
        return this.ab8;
    };
    MoneyFlowIndexIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    MoneyFlowIndexIndicator.$t = markType(MoneyFlowIndexIndicator, 'MoneyFlowIndexIndicator', StrategyBasedIndicator.$);
    MoneyFlowIndexIndicator.ab9 = StrategyBasedIndicator.ab5(14, MoneyFlowIndexIndicator.$);
    return MoneyFlowIndexIndicator;
}(StrategyBasedIndicator));
export { MoneyFlowIndexIndicator };
