// igniteui-angular-charts/esm5/lib/NumericXAxis_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StraightNumericAxisBase } from "./StraightNumericAxisBase";
import { IScaler_$type } from "./IScaler";
import { markType, TypeRegistrar } from "igniteui-angular-core";
import { HorizontalAxisLabelPanel } from "./HorizontalAxisLabelPanel";
import { ScalerParams } from "./ScalerParams";
import { Rect } from "igniteui-angular-core";
import { RangeInfo } from "./RangeInfo";
import { HorizontalLogarithmicScaler } from "./HorizontalLogarithmicScaler";
import { FastHorizontalLinearScaler } from "./FastHorizontalLinearScaler";
import { isNaN_ } from "igniteui-angular-core";
import { StraightNumericAxisBaseView } from "./StraightNumericAxisBaseView";
import { AxisDefaults } from "./AxisDefaults";
/**
 * @hidden
 */
var NumericXAxis = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NumericXAxis, _super);
    function NumericXAxis() {
        var _this = _super.call(this) || this;
        _this.ab = NumericXAxis.$;
        _this.k1 = _this.kz();
        return _this;
    }
    NumericXAxis.prototype.a4 = function () {
        return new NumericXAxisView(this);
    };
    NumericXAxis.prototype.ht = function (a) {
        _super.prototype.ht.call(this, a);
        this.n1 = a;
    };
    NumericXAxis.prototype.av = function () {
        if (this.c6) {
            return (TypeRegistrar.create('HorizontalSmartAxisLabelPanel'));
        }
        else {
            return new HorizontalAxisLabelPanel();
        }
    };
    NumericXAxis.prototype.kz = function () {
        var _this = this;
        var a = _super.prototype.kz.call(this);
        a.b.c = function (b) {
            if ((_this.az == null || _this.az.visibility == 0) && _this.aj != null) {
                _this.aw.bg = b;
                if (_this.az != null && (_this.az.location == 4 || _this.az.location == 5)) {
                    _this.bi.nr();
                }
            }
        };
        a.k = function (b, c, d) { return _this.io(c, d, b.z, b.b, false); };
        a.l = function (b, c, d, e) { return _this.ip(c, d, e, b.z); };
        a.m = function (b, c) {
            var d = b.y;
            var e = new ScalerParams(0, b.aa, b.z, _this.ci, d);
            return _this.ek(c, e);
        };
        a.q = function (b, c) { return true; };
        a.j = function (b) { return _this.hr(b.t, b.n, b.z, b.a, true); };
        a.e = function (b) {
            b.n = b.z.bottom;
            if (_this.aj != null && _this.aj.bi != null) {
                var c = b.n;
                var d = b.r;
                var e = _this.ig(c, d, b.aa, b.z, b.y, b.z.top, b.z.bottom);
                c = e.p0;
                d = e.p1;
                b.n = c;
                b.r = d;
            }
        };
        a.p = function (b, c, d) {
            var e = Math.round(c);
            return e >= Math.floor(b.z.left) && e <= Math.ceil(b.z.right);
        };
        return a;
    };
    NumericXAxis.prototype.ek = function (a, b) {
        if (this.nk != null) {
            return this.nk.p(a, b);
        }
        return this.nj.p(a, b);
    };
    NumericXAxis.prototype.en = function (a, b) {
        if (this.nk != null) {
            return this.nk.q(a, b);
        }
        return this.nj.q(a, b);
    };
    NumericXAxis.prototype.ho = function (a, b, c, d) {
        if (this.nk != null) {
            this.nk.v(a, b, c, d);
            return;
        }
        this.nj.v(a, b, c, d);
    };
    NumericXAxis.prototype.hn = function (a, b, c, d, e) {
        if (this.nk != null) {
            this.nk.u(a, b, c, d, e);
            return;
        }
        this.nj.u(a, b, c, d, e);
    };
    NumericXAxis.prototype.hp = function (a, b, c, d) {
        if (this.nk != null) {
            this.nk.j(a, b, c, d);
            return;
        }
        this.nj.j(a, b, c, d);
    };
    NumericXAxis.prototype.n2 = function (a, b) {
        if (this.bi == null || isNaN_(a) || isNaN_(b)) {
            return;
        }
        var c = this.np(a, this);
        var d = this.np(b, this);
        if (isNaN_(c) || isNaN_(d)) {
            return;
        }
        this.bi.v1 = new Rect(0, c, this.bi.v1.y, d - c, this.bi.v1.height);
    };
    NumericXAxis.prototype.k2 = function (a, b, c, d) {
        var e = d;
        var f = _super.prototype.k2.call(this, a, b, c, d);
        var g = new ScalerParams(0, b, a, this.ch, c);
        var h = NaN;
        var i = NaN;
        if (!this.ch && b.right == 1) {
            i = this.lx;
        }
        else if (this.ch && b.left == 0) {
            h = this.lx;
        }
        if (isNaN_(h)) {
            h = this.en(e.left, g);
        }
        if (isNaN_(i)) {
            i = this.en(e.right, g);
        }
        var j = Math.min(h, i);
        var k = Math.max(h, i);
        this.l1 = j;
        this.l0 = k;
        f.k.add(((function () {
            var $ret = new RangeInfo();
            $ret.d = j;
            $ret.c = k;
            $ret.b = e.width;
            return $ret;
        })()));
        return f;
    };
    NumericXAxis.prototype.ia = function (a) {
        _super.prototype.ia.call(this, a);
        var b = this.bi != null ? this.bi.vl : Rect.empty;
        var c = !b.isEmpty ? this.ke : Rect.empty;
        var d = this.j9();
        var e = this.j8();
        var f = this.k1.c(a, c, b, d, e);
        if (f != null) {
            this.lw = f.f.m;
            this.lz = f.f.l;
        }
    };
    NumericXAxis.prototype.iq = function (a, b) {
        _super.prototype.iq.call(this, a, b);
        if (b.height != a.height) {
            this.c1();
        }
    };
    NumericXAxis.prototype.get_a2 = function () {
        return 0;
    };
    Object.defineProperty(NumericXAxis.prototype, "a2", {
        get: function () {
            return this.get_a2();
        },
        enumerable: false,
        configurable: true
    });
    NumericXAxis.prototype.k4 = function () {
        if (this.lj) {
            return new HorizontalLogarithmicScaler();
        }
        switch (this.ni) {
            case 0: return new FastHorizontalLinearScaler();
            case 1: return new HorizontalLogarithmicScaler();
        }
        return null;
    };
    NumericXAxis.prototype.j$d = function (a, b) {
        return this.en.apply(this, arguments);
    };
    NumericXAxis.prototype.j$c = function (a, b) {
        return this.ek.apply(this, arguments);
    };
    NumericXAxis.prototype.j$e = function (a, b, c, d, e) {
        this.hn.apply(this, arguments);
    };
    NumericXAxis.prototype.j$f = function (a, b, c, d) {
        this.ho.apply(this, arguments);
    };
    NumericXAxis.prototype.j$g = function (a, b, c, d) {
        this.hp.apply(this, arguments);
    };
    Object.defineProperty(NumericXAxis.prototype, "j$b", {
        get: function () {
            return this.ch;
        },
        enumerable: false,
        configurable: true
    });
    NumericXAxis.prototype.j$a = function () {
        return this.ag.apply(this, arguments);
    };
    NumericXAxis.$t = markType(NumericXAxis, 'NumericXAxis', StraightNumericAxisBase.$, [IScaler_$type]);
    return NumericXAxis;
}(StraightNumericAxisBase));
export { NumericXAxis };
/**
 * @hidden
 */
var NumericXAxisView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NumericXAxisView, _super);
    function NumericXAxisView(a) {
        var _this = _super.call(this, a) || this;
        _this.bo = null;
        _this.bo = a;
        return _this;
    }
    NumericXAxisView.prototype.al = function () {
        _super.prototype.al.call(this);
        this.a.jx = AxisDefaults.h;
    };
    NumericXAxisView.$t = markType(NumericXAxisView, 'NumericXAxisView', StraightNumericAxisBaseView.$);
    return NumericXAxisView;
}(StraightNumericAxisBaseView));
export { NumericXAxisView };
