// igniteui-angular-charts/esm5/lib/NumericYAxis_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StraightNumericAxisBase } from "./StraightNumericAxisBase";
import { IScaler_$type } from "./IScaler";
import { markType } from "igniteui-angular-core";
import { VerticalAxisLabelPanel } from "./VerticalAxisLabelPanel";
import { Rect } from "igniteui-angular-core";
import { ScalerParams } from "./ScalerParams";
import { RangeInfo } from "./RangeInfo";
import { VerticalLogarithmicScaler } from "./VerticalLogarithmicScaler";
import { FastVerticalLinearScaler } from "./FastVerticalLinearScaler";
import { isNaN_ } from "igniteui-angular-core";
import { StraightNumericAxisBaseView } from "./StraightNumericAxisBaseView";
import { AxisDefaults } from "./AxisDefaults";
/**
 * @hidden
 */
var NumericYAxis = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NumericYAxis, _super);
    function NumericYAxis() {
        var _this = _super.call(this) || this;
        _this.ab = NumericYAxis.$;
        _this.k1 = _this.kz();
        return _this;
    }
    NumericYAxis.prototype.a4 = function () {
        return new NumericYAxisView(this);
    };
    NumericYAxis.prototype.av = function () {
        return new VerticalAxisLabelPanel();
    };
    NumericYAxis.prototype.get_cp = function () {
        return true;
    };
    NumericYAxis.prototype.ej = function () {
        return this.j8().height;
    };
    NumericYAxis.prototype.ek = function (a, b) {
        if (this.nk != null) {
            return this.nk.p(a, b);
        }
        return this.nj.p(a, b);
    };
    NumericYAxis.prototype.ho = function (a, b, c, d) {
        if (this.nk != null) {
            this.nk.v(a, b, c, d);
            return;
        }
        this.nj.v(a, b, c, d);
    };
    NumericYAxis.prototype.hn = function (a, b, c, d, e) {
        if (this.nk != null) {
            this.nk.u(a, b, c, d, e);
            return;
        }
        this.nj.u(a, b, c, d, e);
    };
    NumericYAxis.prototype.en = function (a, b) {
        if (this.nk != null) {
            return this.nk.q(a, b);
        }
        return this.nj.q(a, b);
    };
    NumericYAxis.prototype.hp = function (a, b, c, d) {
        if (this.nk != null) {
            this.nk.j(a, b, c, d);
            return;
        }
        this.nj.j(a, b, c, d);
    };
    NumericYAxis.prototype.n1 = function (a, b) {
        if (this.bi == null || isNaN_(a) || isNaN_(b)) {
            return;
        }
        var c = this.np(a, this);
        var d = this.np(b, this);
        if (isNaN_(c) || isNaN_(d)) {
            return;
        }
        this.bi.v1 = new Rect(0, this.bi.v1.x, d, this.bi.v1.width, c - d);
    };
    NumericYAxis.prototype.kz = function () {
        var _this = this;
        var a = _super.prototype.kz.call(this);
        a.b.c = function (b) {
            if (_this.az == null || _this.az.visibility == 0) {
                _this.aw.bg = b;
                if (_this.az != null && (_this.az.location == 7 || _this.az.location == 6)) {
                    _this.bi.nr();
                }
            }
        };
        a.k = function (b, c, d) { return _this.hr(c, d, b.z, b.b, false); };
        a.l = function (b, c, d, e) { return _this.hs(c, d, e, b.z); };
        a.m = function (b, c) {
            var d = b.y;
            var e = new ScalerParams(0, b.aa, b.z, _this.ci, d);
            return _this.ek(c, e);
        };
        a.q = function (b, c) { return true; };
        a.j = function (b) { return _this.io(b.t, b.n, b.z, b.a, true); };
        a.e = function (b) {
            b.n = _this.az != null && (_this.az.location == 7 || _this.az.location == 3) ? b.z.right : b.z.left;
            var c = b.y;
            var d = new ScalerParams(0, b.aa, b.z, _this.ci, c);
            if (_this.aj != null && _this.aj.bi != null) {
                var e = b.n;
                var f = b.r;
                var g = _this.ig(e, f, b.aa, b.z, b.y, b.z.left, b.z.right);
                e = g.p0;
                f = g.p1;
                b.n = e;
                b.r = f;
            }
        };
        a.p = function (b, c, d) {
            var e = Math.round(c);
            return e >= Math.floor(b.z.top) && e <= Math.ceil(b.z.bottom);
        };
        return a;
    };
    NumericYAxis.prototype.k2 = function (a, b, c, d) {
        var e = d;
        var f = _super.prototype.k2.call(this, a, b, c, d);
        var g = new ScalerParams(0, b, a, this.ch, c);
        var h = NaN;
        var i = NaN;
        if (!this.ch && b.top == 0) {
            i = this.lx;
        }
        else if (this.ch && b.bottom == 1) {
            h = this.lx;
        }
        if (isNaN_(h)) {
            h = this.en(e.bottom, g);
        }
        if (isNaN_(i)) {
            i = this.en(e.top, g);
        }
        var j = Math.min(h, i);
        var k = Math.max(h, i);
        this.l1 = j;
        this.l0 = k;
        var l = ((function () {
            var $ret = new RangeInfo();
            $ret.d = j;
            $ret.c = k;
            $ret.b = e.height;
            return $ret;
        })());
        f.k.add(l);
        return f;
    };
    NumericYAxis.prototype.ia = function (a) {
        _super.prototype.ia.call(this, a);
        var b = this.bi != null ? this.bi.vl : Rect.empty;
        var c = this.ke;
        var d = this.j9();
        var e = this.j8();
        var f = this.k1.c(a, c, b, d, e);
        if (f != null) {
            this.lw = f.f.m;
            this.lz = f.f.l;
        }
    };
    NumericYAxis.prototype.iq = function (a, b) {
        _super.prototype.iq.call(this, a, b);
        if (b.height != a.height) {
            this.c1();
        }
    };
    NumericYAxis.prototype.get_a2 = function () {
        return 1;
    };
    Object.defineProperty(NumericYAxis.prototype, "a2", {
        get: function () {
            return this.get_a2();
        },
        enumerable: false,
        configurable: true
    });
    NumericYAxis.prototype.k4 = function () {
        if (this.lj) {
            return new VerticalLogarithmicScaler();
        }
        switch (this.ni) {
            case 0: return new FastVerticalLinearScaler();
            case 1: return new VerticalLogarithmicScaler();
        }
        return null;
    };
    NumericYAxis.prototype.j$d = function (a, b) {
        return this.en.apply(this, arguments);
    };
    NumericYAxis.prototype.j$c = function (a, b) {
        return this.ek.apply(this, arguments);
    };
    NumericYAxis.prototype.j$e = function (a, b, c, d, e) {
        this.hn.apply(this, arguments);
    };
    NumericYAxis.prototype.j$f = function (a, b, c, d) {
        this.ho.apply(this, arguments);
    };
    NumericYAxis.prototype.j$g = function (a, b, c, d) {
        this.hp.apply(this, arguments);
    };
    Object.defineProperty(NumericYAxis.prototype, "j$b", {
        get: function () {
            return this.ch;
        },
        enumerable: false,
        configurable: true
    });
    NumericYAxis.prototype.j$a = function () {
        return this.ag.apply(this, arguments);
    };
    NumericYAxis.$t = markType(NumericYAxis, 'NumericYAxis', StraightNumericAxisBase.$, [IScaler_$type]);
    return NumericYAxis;
}(StraightNumericAxisBase));
export { NumericYAxis };
/**
 * @hidden
 */
var NumericYAxisView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NumericYAxisView, _super);
    function NumericYAxisView(a) {
        var _this = _super.call(this, a) || this;
        _this.bo = null;
        _this.bo = a;
        return _this;
    }
    NumericYAxisView.prototype.al = function () {
        _super.prototype.al.call(this);
        this.a.jx = AxisDefaults.h;
    };
    NumericYAxisView.$t = markType(NumericYAxisView, 'NumericYAxisView', StraightNumericAxisBaseView.$);
    return NumericYAxisView;
}(StraightNumericAxisBaseView));
export { NumericYAxisView };
