// igniteui-angular-charts/esm5/lib/PercentagePriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentagePriceOscillatorIndicatorStrategy } from "./PercentagePriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var PercentagePriceOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PercentagePriceOscillatorIndicator, _super);
    function PercentagePriceOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PercentagePriceOscillatorIndicator.prototype.bi = function () {
        return new PercentagePriceOscillatorIndicator();
    };
    PercentagePriceOscillatorIndicator.prototype.get_abn = function () {
        return new PercentagePriceOscillatorIndicatorStrategy();
    };
    Object.defineProperty(PercentagePriceOscillatorIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    PercentagePriceOscillatorIndicator.prototype.get_ab0 = function () {
        return PercentagePriceOscillatorIndicator.$;
    };
    Object.defineProperty(PercentagePriceOscillatorIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PercentagePriceOscillatorIndicator.prototype, "shortPeriod", {
        get: function () {
            return typeGetValue(this.c(PercentagePriceOscillatorIndicator.ab9));
        },
        set: function (a) {
            this.h(PercentagePriceOscillatorIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    PercentagePriceOscillatorIndicator.prototype.abv = function () {
        return this.shortPeriod;
    };
    Object.defineProperty(PercentagePriceOscillatorIndicator.prototype, "longPeriod", {
        get: function () {
            return typeGetValue(this.c(PercentagePriceOscillatorIndicator.ab8));
        },
        set: function (a) {
            this.h(PercentagePriceOscillatorIndicator.ab8, a);
        },
        enumerable: false,
        configurable: true
    });
    PercentagePriceOscillatorIndicator.prototype.abs = function () {
        return this.longPeriod;
    };
    PercentagePriceOscillatorIndicator.$t = markType(PercentagePriceOscillatorIndicator, 'PercentagePriceOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    PercentagePriceOscillatorIndicator.ab9 = StrategyBasedIndicator.ab7(10, PercentagePriceOscillatorIndicator.$);
    PercentagePriceOscillatorIndicator.ab8 = StrategyBasedIndicator.ab3(30, PercentagePriceOscillatorIndicator.$);
    return PercentagePriceOscillatorIndicator;
}(StrategyBasedIndicator));
export { PercentagePriceOscillatorIndicator };
