// igniteui-angular-charts/esm5/lib/PercentageVolumeOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentageVolumeOscillatorIndicatorStrategy } from "./PercentageVolumeOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var PercentageVolumeOscillatorIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PercentageVolumeOscillatorIndicator, _super);
    function PercentageVolumeOscillatorIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PercentageVolumeOscillatorIndicator.prototype.bi = function () {
        return new PercentageVolumeOscillatorIndicator();
    };
    PercentageVolumeOscillatorIndicator.prototype.get_abn = function () {
        return new PercentageVolumeOscillatorIndicatorStrategy();
    };
    Object.defineProperty(PercentageVolumeOscillatorIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    PercentageVolumeOscillatorIndicator.prototype.get_ab0 = function () {
        return PercentageVolumeOscillatorIndicator.$;
    };
    Object.defineProperty(PercentageVolumeOscillatorIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PercentageVolumeOscillatorIndicator.prototype, "shortPeriod", {
        get: function () {
            return typeGetValue(this.c(PercentageVolumeOscillatorIndicator.ab9));
        },
        set: function (a) {
            this.h(PercentageVolumeOscillatorIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    PercentageVolumeOscillatorIndicator.prototype.abv = function () {
        return this.shortPeriod;
    };
    Object.defineProperty(PercentageVolumeOscillatorIndicator.prototype, "longPeriod", {
        get: function () {
            return typeGetValue(this.c(PercentageVolumeOscillatorIndicator.ab8));
        },
        set: function (a) {
            this.h(PercentageVolumeOscillatorIndicator.ab8, a);
        },
        enumerable: false,
        configurable: true
    });
    PercentageVolumeOscillatorIndicator.prototype.abs = function () {
        return this.longPeriod;
    };
    PercentageVolumeOscillatorIndicator.$t = markType(PercentageVolumeOscillatorIndicator, 'PercentageVolumeOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    PercentageVolumeOscillatorIndicator.ab9 = StrategyBasedIndicator.ab7(10, PercentageVolumeOscillatorIndicator.$);
    PercentageVolumeOscillatorIndicator.ab8 = StrategyBasedIndicator.ab3(30, PercentageVolumeOscillatorIndicator.$);
    return PercentageVolumeOscillatorIndicator;
}(StrategyBasedIndicator));
export { PercentageVolumeOscillatorIndicator };
