// igniteui-angular-charts/esm5/lib/PointSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { HorizontalAnchoredCategorySeries } from "./HorizontalAnchoredCategorySeries";
import { markType } from "igniteui-angular-core";
import { AnchoredCategorySeriesView } from "./AnchoredCategorySeriesView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-angular-core";
/**
 * @hidden
 */
var PointSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PointSeries, _super);
    function PointSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PointSeries.$;
        return _this;
    }
    PointSeries.prototype.bi = function () {
        return new PointSeries();
    };
    PointSeries.prototype.get_eg = function () {
        return true;
    };
    PointSeries.prototype.cq = function () {
        return new PointSeriesView(this);
    };
    PointSeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.acd = a;
    };
    PointSeries.prototype.jj = function () {
        return 0.7;
    };
    PointSeries.prototype.wt = function () {
        if (this.ws == 0) {
            return 2;
        }
        return _super.prototype.wt.call(this);
    };
    PointSeries.prototype.gu = function (a, b) {
        var c = this.v6(this.wb(a), true, false);
        if (this.w9(a, b)) {
            return true;
        }
        return false;
    };
    PointSeries.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
    };
    Object.defineProperty(PointSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zh;
        },
        enumerable: false,
        configurable: true
    });
    PointSeries.$t = markType(PointSeries, 'PointSeries', HorizontalAnchoredCategorySeries.$);
    return PointSeries;
}(HorizontalAnchoredCategorySeries));
export { PointSeries };
/**
 * @hidden
 */
var PointSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PointSeriesView, _super);
    function PointSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.df = null;
        _this.df = a;
        return _this;
    }
    PointSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.b4.w2 = Defaults.e;
        }
    };
    PointSeriesView.prototype.af = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.vv = a;
        this.e.jk = 3;
        this.e.jl = 2;
        this.e.jm = 2;
        this.e.ha = false;
    };
    PointSeriesView.$t = markType(PointSeriesView, 'PointSeriesView', AnchoredCategorySeriesView.$);
    return PointSeriesView;
}(AnchoredCategorySeriesView));
export { PointSeriesView };
