// igniteui-angular-charts/esm5/lib/PolarLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { UnknownValuePlotting_$type } from "igniteui-angular-core";
import { DependencyProperty } from "igniteui-angular-core";
import { runOn, enumGetBox, markType, markDep } from "igniteui-angular-core";
import { Path } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
var PolarLineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarLineSeries, _super);
    function PolarLineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PolarLineSeries.$;
        return _this;
    }
    PolarLineSeries.prototype.bi = function () {
        return new PolarLineSeries();
    };
    PolarLineSeries.prototype.get_fi = function () {
        return true;
    };
    PolarLineSeries.prototype.cq = function () {
        return new PolarLineSeriesView(this);
    };
    PolarLineSeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.acm = a;
    };
    PolarLineSeries.prototype.ack = function (a) {
        var b = a;
        b.da();
    };
    PolarLineSeries.prototype.acl = function (a, b) {
        var c = b;
        c.db(a.m, this.ib);
        this.yx.az(this, this.zp(), runOn(this, this.yf));
        var d = false;
        var e = this.yx.b;
        if (e != null) {
            d = true;
        }
        if (d) {
            this.aa4(-1, this.zb.count, b.q);
        }
        this.yx.ad(b.dd, true, false, true, false);
    };
    PolarLineSeries.prototype.acc = function () {
        return this.acn;
    };
    PolarLineSeries.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
        switch (b) {
            case PolarLineSeries.$$p[0]:
                this.rn(false);
                this.qh();
                break;
        }
    };
    PolarLineSeries.$t = markType(PolarLineSeries, 'PolarLineSeries', PolarLineSeriesBase.$);
    PolarLineSeries.$$p = markDep(DependencyProperty, PropertyMetadata, PolarLineSeries, 'raisePropertyChanged', ['UnknownValuePlotting:acn:acp', [UnknownValuePlotting_$type, enumGetBox(UnknownValuePlotting_$type, 1)]]);
    return PolarLineSeries;
}(PolarLineSeriesBase));
export { PolarLineSeries };
/**
 * @hidden
 */
var PolarLineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarLineSeriesView, _super);
    function PolarLineSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c9 = null;
        _this.dd = new Path();
        _this.dc = new Path();
        _this.c9 = a;
        return _this;
    }
    PolarLineSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.b4.w2 = Defaults.e;
        }
    };
    PolarLineSeriesView.prototype.da = function () {
        this.dd.an = null;
        this.dc.an = null;
        this.as();
    };
    PolarLineSeriesView.prototype.db = function (a, b) {
        this.c9.ab9.d(this.dd, a, b);
        this.as();
    };
    PolarLineSeriesView.prototype.bd = function () {
        _super.prototype.bd.call(this);
    };
    PolarLineSeriesView.prototype.be = function () {
        _super.prototype.be.call(this);
        var a = this.br();
        this.dc.an = this.dd.an;
        this.dc._stroke = a;
        this.dc.ad = this.e.ic + 3;
    };
    PolarLineSeriesView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (a.d) {
            if (b) {
                a.t(this.dc);
            }
            else {
                a.t(this.dd);
            }
        }
    };
    PolarLineSeriesView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        this.e.bb.exportPathData(a, this.dd, "polarShape", ["Main"]);
    };
    PolarLineSeriesView.$t = markType(PolarLineSeriesView, 'PolarLineSeriesView', PolarLineSeriesBaseView.$);
    return PolarLineSeriesView;
}(PolarLineSeriesBaseView));
export { PolarLineSeriesView };
