// igniteui-angular-charts/esm5/lib/PolarAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { UnknownValuePlotting_$type } from "igniteui-angular-core";
import { DependencyProperty } from "igniteui-angular-core";
import { runOn, enumGetBox, markType, markDep, Point_$type } from "igniteui-angular-core";
import { Path } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-angular-core";
/**
 * @hidden
 */
var PolarAreaSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarAreaSeries, _super);
    function PolarAreaSeries() {
        var _this = _super.call(this) || this;
        _this.ab = PolarAreaSeries.$;
        return _this;
    }
    PolarAreaSeries.prototype.bi = function () {
        return new PolarAreaSeries();
    };
    PolarAreaSeries.prototype.get_es = function () {
        return true;
    };
    PolarAreaSeries.prototype.cq = function () {
        return new PolarAreaSeriesView(this);
    };
    PolarAreaSeries.prototype.zq = function () {
        return false;
    };
    PolarAreaSeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.acm = a;
    };
    PolarAreaSeries.prototype.ack = function (a) {
        var b = a;
        b.da();
    };
    PolarAreaSeries.prototype.acl = function (a, b) {
        if (a.m.count > 0) {
            var c = a.m._inner[0];
            var d = a.m._inner[a.m.count - 1];
            var e = c.x == d.x && c.y == d.y;
            if (!e) {
                a.m.add(a.m._inner[0]);
            }
        }
        var f = b;
        f.db(a.m, this.ib);
        this.yx.az(this, this.zp(), runOn(this, this.yf));
        var g = false;
        var h = this.yx.b;
        if (h != null) {
            g = true;
        }
        if (g) {
            this.aa4(-1, this.zb.count, b.q);
        }
        var i = f.dh;
        var j = f.dg;
        this.yx.ad(i, true, false, true, true);
        this.yx.ad(j, false, true, false, false);
        j._opacity = this.yx.j * this.h7;
    };
    PolarAreaSeries.prototype.get_ach = function () {
        return true;
    };
    PolarAreaSeries.prototype.acc = function () {
        return this.acn;
    };
    PolarAreaSeries.$t = markType(PolarAreaSeries, 'PolarAreaSeries', PolarLineSeriesBase.$);
    PolarAreaSeries.$$p = markDep(DependencyProperty, PropertyMetadata, PolarAreaSeries, 'raisePropertyChanged', ['UnknownValuePlotting:acn:acp', [UnknownValuePlotting_$type, enumGetBox(UnknownValuePlotting_$type, 1)]]);
    return PolarAreaSeries;
}(PolarLineSeriesBase));
export { PolarAreaSeries };
/**
 * @hidden
 */
var PolarAreaSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarAreaSeriesView, _super);
    function PolarAreaSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c9 = null;
        _this.df = new Path();
        _this.de = new Path();
        _this.dd = new Path();
        _this.dc = new Path();
        _this.c9 = a;
        return _this;
    }
    PolarAreaSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c9.w2 = Defaults.e;
        }
    };
    Object.defineProperty(PolarAreaSeriesView.prototype, "dg", {
        get: function () {
            return this.de;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarAreaSeriesView.prototype, "dh", {
        get: function () {
            return this.df;
        },
        enumerable: false,
        configurable: true
    });
    PolarAreaSeriesView.prototype.da = function () {
        this.de.an = null;
        this.df.an = null;
        this.dc.an = null;
        this.dd.an = null;
        this.as();
    };
    PolarAreaSeriesView.prototype.db = function (a, b) {
        var c = new List$1(Point_$type, 1, a);
        this.c9.ab9.d(this.df, c, b);
        this.c9.ab9.d(this.de, c, b);
        this.as();
    };
    PolarAreaSeriesView.prototype.bd = function () {
        _super.prototype.bd.call(this);
    };
    PolarAreaSeriesView.prototype.be = function () {
        _super.prototype.be.call(this);
        var a = this.br();
        this.dd.an = this.df.an;
        this.dc.an = this.de.an;
        this.dd._stroke = a;
        this.dd.ad = this.e.ic + 3;
        this.dc._fill = a;
        this.dc._opacity = 1;
    };
    PolarAreaSeriesView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (a.d) {
            if (b) {
                a.t(this.dc);
                a.t(this.dd);
            }
            else {
                a.t(this.de);
                a.t(this.df);
            }
        }
    };
    PolarAreaSeriesView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        this.e.bb.exportPathData(a, this.df, "polarShape", ["Main"]);
        this.e.bb.exportPathData(a, this.de, "fillShape", ["Fill"]);
    };
    PolarAreaSeriesView.prototype.af = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.vv = a;
        this.e.jk = 5;
        this.e.jl = 5;
        this.e.jm = 5;
    };
    PolarAreaSeriesView.$t = markType(PolarAreaSeriesView, 'PolarAreaSeriesView', PolarLineSeriesBaseView.$);
    return PolarAreaSeriesView;
}(PolarLineSeriesBaseView));
export { PolarAreaSeriesView };
