// igniteui-angular-charts/esm5/lib/RadialLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnchoredRadialSeries } from "./AnchoredRadialSeries";
import { UnknownValuePlotting_$type } from "igniteui-angular-core";
import { DependencyProperty } from "igniteui-angular-core";
import { Array_$type, runOn, enumGetBox, markType, markDep } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { Path } from "igniteui-angular-core";
import { ScalerParams } from "./ScalerParams";
import { PropertyMetadata } from "igniteui-angular-core";
import { AnchoredRadialSeriesView } from "./AnchoredRadialSeriesView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-angular-core";
/**
 * @hidden
 */
var RadialLineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RadialLineSeries, _super);
    function RadialLineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = RadialLineSeries.$;
        return _this;
    }
    RadialLineSeries.prototype.bi = function () {
        return new RadialLineSeries();
    };
    RadialLineSeries.prototype.get_fi = function () {
        return true;
    };
    RadialLineSeries.prototype.cq = function () {
        return new RadialLineSeriesView(this);
    };
    RadialLineSeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.add = a;
    };
    RadialLineSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
    };
    RadialLineSeries.prototype.preferredCategoryMode = function (a) {
        return 0;
    };
    RadialLineSeries.prototype.zp = function () {
        return true;
    };
    RadialLineSeries.prototype.o3 = function (a, b) {
        _super.prototype.o3.call(this, a, b);
        var c = b;
        c.c6();
    };
    RadialLineSeries.prototype.aa2 = function (a, b) {
        _super.prototype.aa2.call(this, a, b);
        var c = b.by;
        var d = b.bx;
        var e = b.getEffectiveViewport();
        var f = a.f;
        var g = f.count;
        var h = new List$1(Array_$type, 2, g);
        var i = 0;
        for (var j = 0; j < g; j++) {
            var k = new Array(4);
            h.add(k);
            k[0] = this.yz.d(a.f._inner[j][0], a.f._inner[j][1], c, d, e);
            k[1] = this.yz.e(a.f._inner[j][0], a.f._inner[j][1], c, d, e);
            k[2] = this.yz.d(a.f._inner[j][0], a.f._inner[j][2], c, d, e);
            var l = a.f._inner[j][2];
            k[3] = this.yz.e(a.f._inner[j][0], l, c, d, e);
            if (l > i) {
                i = l;
            }
        }
        var m = this.getEffectiveViewport1(b);
        this.y7.a2(this, this.zo(), runOn(this, this.yf), this.zx(b), this.zy(b));
        var n = false;
        var o = this.y7.b;
        if (o != null) {
            n = true;
        }
        if (n) {
            var p = new ScalerParams(0, b.by, b.bx, this.yr.ch, m);
            this.act(f, -1, this.abx.count, this.yr, p, b.q);
        }
        var q = b;
        var r = q.df;
        var s = q.dg;
        var t = q.de;
        this.y7.ad(r, true, false, true, false);
        this.y7.ad(s, true, false, true, false);
        this.y7.ad(t, false, true, false, false);
        t._opacity = 0.75 * this.y7.j;
        q.c7(g, h, false, this.ade, this.abq(h, g - 1, b), q.cx.b, this.ib);
    };
    RadialLineSeries.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
        switch (b) {
            case RadialLineSeries.$$p[0]:
                this.rn(false);
                this.qh();
                break;
        }
    };
    RadialLineSeries.prototype.get_ab3 = function () {
        return true;
    };
    RadialLineSeries.$t = markType(RadialLineSeries, 'RadialLineSeries', AnchoredRadialSeries.$);
    RadialLineSeries.$$p = markDep(DependencyProperty, PropertyMetadata, RadialLineSeries, 'raisePropertyChanged', ['UnknownValuePlotting:ade:adg', [UnknownValuePlotting_$type, enumGetBox(UnknownValuePlotting_$type, 1)]]);
    return RadialLineSeries;
}(AnchoredRadialSeries));
export { RadialLineSeries };
/**
 * @hidden
 */
var RadialLineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RadialLineSeriesView, _super);
    function RadialLineSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.c5 = null;
        _this.dc = new Path();
        _this.db = new Path();
        _this.dd = new Path();
        _this.c9 = new Path();
        _this.c8 = new Path();
        _this.da = new Path();
        _this.c5 = a;
        return _this;
    }
    RadialLineSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.b4.w2 = Defaults.e;
        }
    };
    Object.defineProperty(RadialLineSeriesView.prototype, "df", {
        get: function () {
            return this.dc;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialLineSeriesView.prototype, "dg", {
        get: function () {
            return this.dd;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialLineSeriesView.prototype, "de", {
        get: function () {
            return this.db;
        },
        enumerable: false,
        configurable: true
    });
    RadialLineSeriesView.prototype.c6 = function () {
        this.dc.an = null;
        this.db.an = null;
        this.dd.an = null;
        this.c9.an = null;
        this.c8.an = null;
        this.da.an = null;
        this.as();
    };
    RadialLineSeriesView.prototype.c7 = function (a, b, c, d, e, f, g) {
        this.c1.abs.m(this.dc, this.db, this.dd, a, b, c, d, e, f, g);
        this.as();
    };
    RadialLineSeriesView.prototype.bd = function () {
        _super.prototype.bd.call(this);
    };
    RadialLineSeriesView.prototype.be = function () {
        _super.prototype.be.call(this);
        var a = this.br();
        this.c8.an = this.db.an;
        this.c9.an = this.dc.an;
        this.da.an = this.dd.an;
        this.c9._stroke = a;
        this.c9.ad = this.e.ic + 3;
        this.da._stroke = a;
        this.da.ad = this.e.ic + 3;
        this.c8._fill = a;
        this.c8._opacity = 1;
    };
    RadialLineSeriesView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (a.d) {
            if (b) {
                a.t(this.c8);
                a.t(this.c9);
                a.t(this.da);
            }
            else {
                a.t(this.db);
                a.t(this.dc);
                a.t(this.dd);
            }
        }
    };
    RadialLineSeriesView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        this.e.bb.exportPathData(a, this.dc, "lowerShape", ["Lower", "Main"]);
        this.e.bb.exportPathData(a, this.dd, "upperShape", ["Upper"]);
        this.e.bb.exportPathData(a, this.db, "translucentShape", ["Translucent"]);
    };
    RadialLineSeriesView.prototype.af = function () {
        var a = new Color();
        a.colorString = "rgba(95,95,95,0.5)";
        this.e.vv = a;
        this.e.jk = 3;
        this.e.jl = 1;
        this.e.jm = 4;
        this.e.ha = false;
    };
    RadialLineSeriesView.$t = markType(RadialLineSeriesView, 'RadialLineSeriesView', AnchoredRadialSeriesView.$);
    return RadialLineSeriesView;
}(AnchoredRadialSeriesView));
export { RadialLineSeriesView };
