// igniteui-angular-charts/esm5/lib/RateOfChangeAndMomentumIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RateOfChangeAndMomentumIndicatorStrategy } from "./RateOfChangeAndMomentumIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var RateOfChangeAndMomentumIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RateOfChangeAndMomentumIndicator, _super);
    function RateOfChangeAndMomentumIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RateOfChangeAndMomentumIndicator.prototype.bi = function () {
        return new RateOfChangeAndMomentumIndicator();
    };
    RateOfChangeAndMomentumIndicator.prototype.get_abn = function () {
        return new RateOfChangeAndMomentumIndicatorStrategy();
    };
    Object.defineProperty(RateOfChangeAndMomentumIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    RateOfChangeAndMomentumIndicator.prototype.get_ab0 = function () {
        return RateOfChangeAndMomentumIndicator.$;
    };
    Object.defineProperty(RateOfChangeAndMomentumIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RateOfChangeAndMomentumIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(RateOfChangeAndMomentumIndicator.ab9));
        },
        set: function (a) {
            this.h(RateOfChangeAndMomentumIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    RateOfChangeAndMomentumIndicator.prototype.abt = function () {
        return this.ab8;
    };
    RateOfChangeAndMomentumIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    RateOfChangeAndMomentumIndicator.$t = markType(RateOfChangeAndMomentumIndicator, 'RateOfChangeAndMomentumIndicator', StrategyBasedIndicator.$);
    RateOfChangeAndMomentumIndicator.ab9 = StrategyBasedIndicator.ab5(14, RateOfChangeAndMomentumIndicator.$);
    return RateOfChangeAndMomentumIndicator;
}(StrategyBasedIndicator));
export { RateOfChangeAndMomentumIndicator };
