// igniteui-angular-charts/esm5/lib/RelativeStrengthIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RelativeStrengthIndexIndicatorStrategy } from "./RelativeStrengthIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var RelativeStrengthIndexIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RelativeStrengthIndexIndicator, _super);
    function RelativeStrengthIndexIndicator() {
        return _super.call(this) || this;
    }
    RelativeStrengthIndexIndicator.prototype.bi = function () {
        return new RelativeStrengthIndexIndicator();
    };
    RelativeStrengthIndexIndicator.prototype.get_aau = function () {
        return 1;
    };
    RelativeStrengthIndexIndicator.prototype.get_abn = function () {
        return new RelativeStrengthIndexIndicatorStrategy();
    };
    Object.defineProperty(RelativeStrengthIndexIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    RelativeStrengthIndexIndicator.prototype.get_ab0 = function () {
        return RelativeStrengthIndexIndicator.$;
    };
    Object.defineProperty(RelativeStrengthIndexIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RelativeStrengthIndexIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(RelativeStrengthIndexIndicator.ab9));
        },
        set: function (a) {
            this.h(RelativeStrengthIndexIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    RelativeStrengthIndexIndicator.prototype.abt = function () {
        return this.ab8;
    };
    RelativeStrengthIndexIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    RelativeStrengthIndexIndicator.$t = markType(RelativeStrengthIndexIndicator, 'RelativeStrengthIndexIndicator', StrategyBasedIndicator.$);
    RelativeStrengthIndexIndicator.ab9 = StrategyBasedIndicator.ab5(14, RelativeStrengthIndexIndicator.$);
    return RelativeStrengthIndexIndicator;
}(StrategyBasedIndicator));
export { RelativeStrengthIndexIndicator };
