// igniteui-angular-charts/esm5/lib/ScatterAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ScatterTriangulationSeries } from "./ScatterTriangulationSeries";
import { ColorScale } from "./ColorScale";
import { runOn, delegateRemove, delegateCombine, typeCast, Point_$type, markType, markDep } from "igniteui-angular-core";
import { DependencyProperty } from "igniteui-angular-core";
import { FastItemIntColumn } from "igniteui-angular-core";
import { IFastItemsSource_$type } from "igniteui-angular-core";
import { Rect } from "igniteui-angular-core";
import { SeriesView } from "./SeriesView";
import { TriangleRasterizer } from "./TriangleRasterizer";
import { ScalerParams } from "./ScalerParams";
import { ObservableColorCollection } from "igniteui-angular-core";
import { Color } from "igniteui-angular-core";
import { ColorUtil } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { CustomPaletteColorScale } from "./CustomPaletteColorScale";
import { DataLegendSeriesValueInfo } from "igniteui-angular-core";
import { DataLegendSeriesContext } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { truncate, intDivide } from "igniteui-angular-core";
import { boxArray$1 } from "igniteui-angular-core";
import { RenderingContext } from "igniteui-angular-core";
import { CanvasViewRenderer } from "igniteui-angular-core";
import { Rectangle } from "igniteui-angular-core";
import { stringIsNullOrEmpty } from "igniteui-angular-core";
/**
 * @hidden
 */
var ScatterAreaSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterAreaSeries, _super);
    function ScatterAreaSeries() {
        var _this = _super.call(this) || this;
        _this.yl = null;
        _this.yq = null;
        _this.yj = null;
        _this.yk = null;
        FastItemIntColumn.m();
        _this.ab = ScatterAreaSeries.$;
        return _this;
    }
    ScatterAreaSeries.prototype.bi = function () {
        return new ScatterAreaSeries();
    };
    ScatterAreaSeries.prototype.get_es = function () {
        return true;
    };
    Object.defineProperty(ScatterAreaSeries.prototype, "yn", {
        get: function () {
            return this.yl;
        },
        set: function (a) {
            var b = this.yl != a;
            if (b) {
                var c = this.yl;
                if (this.yl != null) {
                    var d = this.yl;
                    d.propertyChanged = delegateRemove(d.propertyChanged, runOn(this, this.y4));
                }
                this.yl = a;
                if (this.yl != null) {
                    var e = this.yl;
                    e.propertyChanged = delegateCombine(e.propertyChanged, runOn(this, this.y4));
                }
                this.raisePropertyChanged("ColorScale", c, this.yl);
            }
        },
        enumerable: false,
        configurable: true
    });
    ScatterAreaSeries.prototype.y4 = function (a, b) {
        this.y6();
        this.rn(false);
    };
    ScatterAreaSeries.prototype.ko = function (a, b) {
        var c = this.nb(b);
        if (c == ScatterAreaSeries.$$p[3]) {
            return this.kp(a, b, this.yz);
        }
        return _super.prototype.ko.call(this, a, b);
    };
    ScatterAreaSeries.prototype.mb = function (a) {
        var b = this.nb(a);
        if (b == ScatterAreaSeries.$$p[3]) {
            return this.yz;
        }
        return _super.prototype.mb.call(this, a);
    };
    Object.defineProperty(ScatterAreaSeries.prototype, "yr", {
        get: function () {
            return this.yq;
        },
        set: function (a) {
            var b = this.yr != a;
            if (b) {
                var c = this.yr;
                this.yq = a;
                this.raisePropertyChanged("ColorColumn", c, this.yr);
            }
        },
        enumerable: false,
        configurable: true
    });
    ScatterAreaSeries.prototype.get_g5 = function () {
        return true;
    };
    ScatterAreaSeries.prototype.set_g5 = function (a) {
        _super.prototype.set_g5.call(this, a);
    };
    ScatterAreaSeries.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
        switch (b) {
            case ScatterAreaSeries.$$p[1]:
            case ScatterAreaSeries.$$p[2]:
                this.qb();
                break;
            case "FastItemsSource":
                var e = typeCast(IFastItemsSource_$type, c);
                if (e != null) {
                    e.deregisterColumn(this.yr);
                    this.yr = null;
                }
                var f = typeCast(IFastItemsSource_$type, d);
                if (f != null) {
                    this.yr = this.di(this.yz);
                }
                this.rn(this.ge && c != null);
                break;
            case ScatterAreaSeries.$$p[3]:
                if (this.dl != null) {
                    this.dl.deregisterColumn(this.yr);
                    this.yr = this.di(this.yz);
                }
                this.rn(false);
                this.qh();
                break;
            case "ColorScale":
                this.y6();
                this.rn(false);
                this.qh();
                break;
        }
    };
    ScatterAreaSeries.prototype.ro = function (a) {
        _super.prototype.ro.call(this, a);
        this.y5(this.ct);
    };
    ScatterAreaSeries.prototype.rm = function (a, b, c, d, e) {
        _super.prototype.rm.call(this, a, b, c, d, e);
        var f = this.a9.a$j.item(d);
        f.a3(c);
        if (this.d3(f)) {
            return;
        }
        this.y5(f);
    };
    ScatterAreaSeries.prototype.y5 = function (a) {
        if (this.d3(a)) {
            return;
        }
        var b;
        var c;
        var d = a.getViewInfo(c, b);
        c = d.p0;
        b = d.p1;
        var e = a;
        var f = null;
        var g = null;
        var h = this.wq;
        var i = this.wr;
        var j = this.getEffectiveViewport1(a);
        var k = new ScalerParams(1, b, c, h.ch);
        k.c = j;
        var l = new ScalerParams(1, b, c, i.ch);
        l.c = j;
        g = this.yp(a, k, l);
        if (g != null) {
            e.cb(g);
        }
        else {
            e.ce(f);
        }
    };
    ScatterAreaSeries.prototype.y2 = function (a) {
        var b;
        var c;
        var d = a.getViewInfo(b, c);
        b = d.p0;
        c = d.p1;
        var e = truncate(Math.round(b.width)) * truncate(Math.round(b.height));
        var f = a.b3;
        if (f == null || f.length != e) {
            f = new Array(e);
            for (var g = 0; g < e; g++) {
                f[g] = 0;
            }
        }
        else {
            for (var h = 0; h < e; h++) {
                f[h] = 0;
            }
        }
        a.b3 = f;
    };
    ScatterAreaSeries.prototype.kl = function (a) {
        var b = truncate(Math.round(this.wl.width));
        var c = truncate(Math.round(this.wl.height));
        var d = this.ct.b3;
        if (d == null || d.length != (b * c) || this.w0 == null || this.w0.count < 1 || this.w1 == null || this.w1.count < 1 || this.w2 == null || this.w2.count < 1) {
            return null;
        }
        var e = this.c9.v1;
        var f = (a.x - e.left) / e.width;
        var g = (a.y - e.top) / e.height;
        var h = truncate(Math.round(b * f));
        var i = truncate(Math.round(c * g));
        var j = (b * i) + h;
        if (j < 0 || j > d.length - 1) {
            return null;
        }
        var k = d[j] - 1;
        if (k < 0 || k > this.w0.count) {
            return null;
        }
        var l = this.yj.length;
        var m = this.w0.item(k);
        var n = this.w1.item(k);
        var o = this.w2.item(k);
        if ((m >= l) || (n >= l) || (o >= l)) {
            return null;
        }
        var p = (this.yj[m] - this.wl.left) / this.wl.width;
        var q = (this.yk[m] - this.wl.top) / this.wl.height;
        var r = (this.yj[n] - this.wl.left) / this.wl.width;
        var s = (this.yk[n] - this.wl.top) / this.wl.height;
        var t = (this.yj[o] - this.wl.left) / this.wl.width;
        var u = (this.yk[o] - this.wl.top) / this.wl.height;
        var v = Math.pow(p - f, 2) + Math.pow(q - g, 2);
        var w = Math.pow(r - f, 2) + Math.pow(s - g, 2);
        var x = Math.pow(t - f, 2) + Math.pow(u - g, 2);
        if (v < w && v < x) {
            return this.dl.item(m);
        }
        if (w < v && w < x) {
            return this.dl.item(n);
        }
        if (x < v && x < w) {
            return this.dl.item(o);
        }
        return this.dl.item(m);
    };
    ScatterAreaSeries.prototype.yp = function (a, b, c) {
        var d;
        var e;
        var f = a.getViewInfo(e, d);
        e = f.p0;
        d = f.p1;
        e = new Rect(0, 0, 0, e.width, e.height);
        b = new ScalerParams(0, b.e, e, b.a, b.c);
        c = new ScalerParams(0, c.e, e, c.a, c.c);
        c.b = this.iw();
        var g = a;
        var h = this.wq;
        var i = this.wr;
        var j = this.w0.count;
        var k;
        var xArr_ = this.w3.asArray();
        k = (xArr_.slice(0));
        var l;
        var yArr_ = this.w4.asArray();
        l = (yArr_.slice(0));
        var m = this.yr.asArray();
        var n = this.w0.asArray();
        var o = this.w1.asArray();
        var p = this.w2.asArray();
        this.yj = k;
        this.yk = l;
        h.ho(boxArray$1(k), 0, k.length, b);
        i.ho(boxArray$1(l), 0, l.length, c);
        var q = this.ym;
        var r = g.b5(b.d, q, this.yr);
        var s = e.top;
        var t = e.left;
        var u = e.right;
        var v = e.bottom;
        var w;
        var x;
        var y;
        var z;
        this.y2(g);
        var aa = g.b3;
        var ab = k.length;
        for (var ac = 0; ac < j; ac++) {
            var ad = n[ac];
            var ae = o[ac];
            var af = p[ac];
            if ((ad >= ab) || (ae >= ab) || (af >= ab) || (ad < 0) || (ae < 0) || (af < 0)) {
                continue;
            }
            var ag = k[ad];
            var ah = l[ad];
            var ai = { $type: Point_$type, x: ag, y: ah };
            var aj = k[ae];
            var ak = l[ae];
            var al = { $type: Point_$type, x: aj, y: ak };
            var am = k[af];
            var an = l[af];
            var ao = { $type: Point_$type, x: am, y: an };
            w = al.y < ao.y ? al.y : ao.y;
            w = ai.y < w ? ai.y : w;
            x = al.y > ao.y ? al.y : ao.y;
            x = ai.y > x ? ai.y : x;
            y = al.x < ao.x ? al.x : ao.x;
            y = ai.x < y ? ai.x : y;
            z = al.x > ao.x ? al.x : ao.x;
            z = ai.x > z ? ai.x : z;
            if (w < v && x > s && y < u && z > t) {
                var ap = m[ad];
                var aq = m[ae];
                var ar = m[af];
                r.p(ac, aa, ai, al, ao, ap, aq, ar);
            }
        }
        return r;
    };
    ScatterAreaSeries.prototype.ys = function () {
        var a = new Color();
        var b = this.u3;
        if (b == null) {
            a = ColorUtil.t("#000000");
        }
        else {
            a = ColorUtil.s(b);
        }
        var c = new ObservableColorCollection();
        c.add(ColorUtil.x(a, -0.5));
        c.add(a);
        return c;
    };
    ScatterAreaSeries.prototype.y6 = function () {
        var _this = this;
        var a = this.yn;
        if (a == null) {
            a = ((function () {
                var $ret = new CustomPaletteColorScale();
                $ret.palette = _this.ys();
                $ret.interpolationMode = 2;
                return $ret;
            })());
        }
        else if (typeCast(CustomPaletteColorScale.$, a) !== null) {
            var b = a;
            if (b.palette == null || b.palette.count == 0) {
                b.palette = this.ys();
            }
            a = b;
        }
        var c = new ObservableColorCollection();
        if (typeCast(CustomPaletteColorScale.$, a) !== null) {
            c = a.palette;
        }
        else {
            c = this.ys();
        }
        this.ym = a;
    };
    ScatterAreaSeries.prototype.sb = function () {
        _super.prototype.sb.call(this);
        this.y6();
    };
    ScatterAreaSeries.prototype.o3 = function (a, b) {
        _super.prototype.o3.call(this, a, b);
        var c = b;
        c.ca();
    };
    ScatterAreaSeries.prototype.cq = function () {
        return new ScatterAreaSeriesView(this);
    };
    ScatterAreaSeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.yo = typeCast(ScatterAreaSeriesView.$, a);
    };
    ScatterAreaSeries.prototype.y3 = function (a) {
        if (this.yo != null) {
            this.yo.b9(a);
        }
    };
    ScatterAreaSeries.prototype.getDataLegendSeriesContextAt = function (a, b) {
        var c = { $type: Point_$type, x: a, y: b };
        var d = this.kl(c);
        var e = d == null ? NaN : this.iv(d, "XMemberPath");
        var f = d == null ? NaN : this.iv(d, "YMemberPath");
        var g = d == null ? NaN : this.iv(d, "ColorMemberPath");
        var h = new DataLegendSeriesValueInfo();
        h.l = e;
        h.a = 13;
        h.q = "X";
        h.r = this.xs != null ? this.xs : "X";
        h.s = this.xo != null ? this.xo : "X:";
        h.t = this.xq != null ? this.xq : "";
        var i = new DataLegendSeriesValueInfo();
        i.l = f;
        i.a = 14;
        i.q = "Y";
        i.r = this.x0 != null ? this.x0 : "Y";
        i.s = this.xw != null ? this.xw : "Y:";
        i.t = this.xy != null ? this.xy : "";
        var j = new DataLegendSeriesValueInfo();
        j.l = g;
        j.a = 0;
        j.q = "Value";
        j.r = this.yz != null ? this.yz : "Value";
        j.s = this.yv != null ? this.yv : "V:";
        j.t = this.yx != null ? this.yx : "";
        var k = new DataLegendSeriesContext();
        k.e = ((function () {
            var $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
            $ret.add(h);
            $ret.add(i);
            $ret.add(j);
            return $ret;
        })());
        k.h = this.ma();
        return k;
    };
    ScatterAreaSeries.$t = markType(ScatterAreaSeries, 'ScatterAreaSeries', ScatterTriangulationSeries.$);
    ScatterAreaSeries.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterAreaSeries, 'raisePropertyChanged', ['ActualColorScale:ym:y7', [ColorScale.$, null], 'ColorMemberAsLegendLabel:yv:y8', [2, null], 'ColorMemberAsLegendUnit:yx:y9', [2, null], 'ColorMemberPath:yz:za', [2, null]]);
    return ScatterAreaSeries;
}(ScatterTriangulationSeries));
export { ScatterAreaSeries };
/**
 * @hidden
 */
var ScatterAreaSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterAreaSeriesView, _super);
    function ScatterAreaSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.b4 = null;
        _this.b0 = null;
        _this.b8 = null;
        _this.b7 = null;
        _this.b6 = null;
        _this.b4 = a;
        return _this;
    }
    Object.defineProperty(ScatterAreaSeriesView.prototype, "b3", {
        get: function () {
            return this.b0;
        },
        set: function (a) {
            this.b0 = a;
        },
        enumerable: false,
        configurable: true
    });
    ScatterAreaSeriesView.prototype.cd = function (a, b) {
    };
    ScatterAreaSeriesView.prototype.b9 = function (a) {
    };
    ScatterAreaSeriesView.prototype.b1 = function () {
        return this.b2(true);
    };
    ScatterAreaSeriesView.prototype.b2 = function (a) {
        this.cc();
        var b = truncate(Math.round(this.bx.left));
        var c = truncate(Math.round(this.bx.left));
        var d = truncate(Math.round(this.bx.width));
        var e = truncate(Math.round(this.bx.height));
        var f = d.toString();
        var g = e.toString();
        if (this.b7.getAttribute("width") != f) {
            this.b7.setAttribute("width", d.toString());
        }
        if (this.b7.getAttribute("height") != g) {
            this.b7.setAttribute("height", e.toString());
        }
        var h = this.b6.h();
        if (a) {
            h.clearRect(0, 0, d, e);
        }
        this.b8 = h.getImageData(0, 0, d, e);
        var imageDataArray_ = this.b8.data;
        return imageDataArray_;
    };
    ScatterAreaSeriesView.prototype.cc = function () {
        if (this.b7 == null) {
            this.b7 = this.e.jx.createElement("canvas");
            var a = this.e.jx.get2DCanvasContext(this.b7);
            this.b6 = new RenderingContext(new CanvasViewRenderer(), a);
        }
    };
    ScatterAreaSeriesView.prototype.b5 = function (a, b, c) {
        var d = truncate(Math.round(this.bx.left));
        var e = truncate(Math.round(this.bx.left));
        var f = truncate(Math.round(this.bx.width));
        var g = truncate(Math.round(this.bx.height));
        return new TriangleRasterizer(this.b1(), b, c.minimum, c.maximum, c, f, g);
    };
    ScatterAreaSeriesView.prototype.cb = function (a) {
        this.ce(null);
    };
    ScatterAreaSeriesView.prototype.ca = function () {
        if (this.b6 != null) {
            this.b6.l(this.bx.left, this.bx.top, this.bx.width, this.bx.height);
        }
        this.as();
    };
    ScatterAreaSeriesView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (this.b7 == null) {
            return;
        }
        if (b) {
            var c = new Rectangle();
            c.n = this.bx.left;
            c.o = this.bx.top;
            c.width = this.bx.width;
            c.height = this.bx.height;
            c._fill = this.br();
            a.w(c);
        }
        else {
            var d = truncate(Math.round(this.bx.left));
            var e = truncate(Math.round(this.bx.top));
            var f = truncate(Math.round(this.bx.width));
            var g = truncate(Math.round(this.bx.height));
            a.n(this.b7.getNativeElement(), 1, d, e, f, g);
        }
    };
    ScatterAreaSeriesView.prototype.ce = function (a) {
        var b = truncate(Math.round(this.bx.left));
        var c = truncate(Math.round(this.bx.left));
        var d = truncate(Math.round(this.bx.width));
        var e = truncate(Math.round(this.bx.height));
        this.cc();
        this.b6.h().putImageData(this.b8, 0, 0);
        this.as();
    };
    ScatterAreaSeriesView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        var b = truncate(Math.round(this.bx.width));
        var c = truncate(Math.round(this.bx.height));
        var d = this.b2(false);
        if (d == null || d.length == 0) {
            this.e.bb.exportSeriesPixelsData(a, null, b, c);
            return;
        }
        var e = new Array(intDivide(d.length, 4));
        var f;
        for (var g = 0; g < intDivide(d.length, 4); g++) {
            f = g * 4;
            e[g] = d[f + 3] << 24 | d[f] << 16 | d[f + 1] << 8 | d[f + 2];
        }
        this.e.bb.exportSeriesPixelsData(a, e, b, c);
    };
    ScatterAreaSeriesView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span";
        var b = this.e.a9.a$u(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.b4.title + "</span><br/><span>" + "(${item." + this.b4.xs + "}, ${item." + this.b4.x0 + "})</span>";
        if (!stringIsNullOrEmpty(this.b4.yz)) {
            a += "<span>, Value: ${item." + this.b4.yz + "}</span>";
        }
        a += "</div>";
        return a;
    };
    ScatterAreaSeriesView.$t = markType(ScatterAreaSeriesView, 'ScatterAreaSeriesView', SeriesView.$);
    return ScatterAreaSeriesView;
}(SeriesView));
export { ScatterAreaSeriesView };
