// igniteui-angular-charts/esm5/lib/ScatterPolygonSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { ShapeSeriesBase } from "./ShapeSeriesBase";
import { ShapeAxisInfoCache } from "./ShapeAxisInfoCache";
import { StyleSelector } from "./StyleSelector";
import { DependencyProperty } from "igniteui-angular-core";
import { Style } from "igniteui-angular-core";
import { MarkerType_$type } from "./MarkerType";
import { DataTemplate } from "igniteui-angular-core";
import { Brush } from "igniteui-angular-core";
import { CollisionAvoidanceType_$type } from "./CollisionAvoidanceType";
import { MarkerOutlineMode_$type } from "./MarkerOutlineMode";
import { MarkerFillMode_$type } from "./MarkerFillMode";
import { Rect } from "igniteui-angular-core";
import { ScalerParams } from "./ScalerParams";
import { Base, runOn, enumGetBox, Number_$type, typeCast, fromEnum, markType, markDep, PointUtil } from "igniteui-angular-core";
import { Dictionary$2 } from "igniteui-angular-core";
import { OwnedPoint } from "./OwnedPoint";
import { ShapeSeriesViewBase } from "./ShapeSeriesViewBase";
import { Marker } from "./Marker";
import { MarkerSeries } from "./MarkerSeries";
import { BrushUtil } from "igniteui-angular-core";
import { DataContext } from "igniteui-angular-core";
import { HashPool$2 } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { RectUtil } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { DeviceUtils } from "igniteui-angular-core";
import { isNaN_, isInfinity } from "igniteui-angular-core";
import { NumericMarkerManager } from "./NumericMarkerManager";
import { DataTemplateMeasureInfo } from "igniteui-angular-core";
import { Size } from "igniteui-angular-core";
import { Defaults } from "./Defaults";
import { DataTemplatePassInfo } from "igniteui-angular-core";
import { DataTemplateRenderInfo } from "igniteui-angular-core";
import { MarkerVisualData } from "./MarkerVisualData";
import { stringIsNullOrEmpty } from "igniteui-angular-core";
/**
 * @hidden
 */
var ScatterPolygonSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolygonSeries, _super);
    function ScatterPolygonSeries() {
        var _this = _super.call(this) || this;
        _this.zv = null;
        _this.zs = new ScalerParams(1, ScatterPolygonSeries.aa1, ScatterPolygonSeries.aa1, false);
        _this.zt = new ScalerParams(1, ScatterPolygonSeries.aa1, ScatterPolygonSeries.aa1, false);
        _this.aah = null;
        _this.zn = 1;
        _this.zj = 1;
        _this.ab = ScatterPolygonSeries.$;
        return _this;
    }
    ScatterPolygonSeries.prototype.bi = function () {
        return new ScatterPolygonSeries();
    };
    ScatterPolygonSeries.prototype.get_fo = function () {
        return true;
    };
    ScatterPolygonSeries.prototype.get_fj = function () {
        return true;
    };
    ScatterPolygonSeries.prototype.xy = function () {
        return this.zx();
    };
    ScatterPolygonSeries.prototype.y0 = function (a) {
        var _this = this;
        var b = a;
        b.cy(this.aa3);
        b.cz(this.zw);
        _super.prototype.y0.call(this, a);
        if (this.d3(a)) {
            return;
        }
        var c, d;
        var e = a.getViewInfo(c, d);
        c = e.p0;
        d = e.p1;
        var f = this.getEffectiveViewport1(a);
        this.zs = new ScalerParams(1, d, c, this.w2.ch);
        this.zs.c = f;
        this.zt = new ScalerParams(1, d, c, this.w4.ch);
        this.zt.c = f;
        if (c.width < 1 || c.height < 1) {
            return;
        }
        this.zv = ((function () {
            var $ret = new ShapeAxisInfoCache();
            $ret.a = _this.w2;
            $ret.b = _this.w4;
            $ret.e = _this.w2.ch;
            $ret.f = _this.w4.ch;
            $ret.d = _this.dl;
            $ret.c = _this.xh;
            return $ret;
        })());
        if (this.zx()) {
            var g = new Dictionary$2(Base.$, OwnedPoint.$, 0);
            b.dg.l(g, 400, d, c, this.ib);
            b.dg.j(g, false);
            if (this.xw()) {
                this.aae(a);
            }
        }
    };
    ScatterPolygonSeries.prototype.aae = function (a) {
        var _this = this;
        a.dk = true;
        this.w9.ay(this, this.xw(), runOn(this, this.ww));
        var b = this.w9.d;
        var c = b != null;
        var d = this.getEffectiveViewport1(a);
        var e = this.dl == null ? 0 : this.dl.count;
        a.dw(function (f) {
            if (f._visibility == 0) {
                var g = f.ak;
                var h = f.content;
                if (h == null) {
                    h = new DataContext();
                    f.content = h;
                }
                if (c) {
                    _this.yv(g, e, a.q);
                }
                _this.w9.ac(f, h);
            }
        });
    };
    ScatterPolygonSeries.prototype.zx = function () {
        return this.aah != null && ((this.zr != 1 && this.zr != 0) || this.aai != null);
    };
    ScatterPolygonSeries.prototype.cq = function () {
        return new ScatterPolygonSeriesView(this);
    };
    ScatterPolygonSeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.zu = a;
    };
    ScatterPolygonSeries.prototype.get_hasMarkers = function () {
        return true;
    };
    ScatterPolygonSeries.prototype.get_ej = function () {
        return this.zr != 13 && this.zr != 1;
    };
    ScatterPolygonSeries.prototype.vf = function () {
        return this.aax;
    };
    ScatterPolygonSeries.prototype.vg = function () {
        return this.aay;
    };
    ScatterPolygonSeries.prototype.st = function () {
        return this.aag;
    };
    ScatterPolygonSeries.prototype.l8 = function () {
        return this.l9(this, this.aah);
    };
    ScatterPolygonSeries.prototype.il = function () {
        return this.zz;
    };
    Object.defineProperty(ScatterPolygonSeries.prototype, "zo", {
        get: function () {
            return this.zn;
        },
        set: function (a) {
            this.zn = a;
        },
        enumerable: false,
        configurable: true
    });
    ScatterPolygonSeries.prototype.zq = function () {
        if (this.zp == 0) {
            return 1;
        }
        return this.zp;
    };
    ScatterPolygonSeries.prototype.sf = function (a) {
        _super.prototype.sf.call(this, a);
        this.zp = a;
    };
    Object.defineProperty(ScatterPolygonSeries.prototype, "zk", {
        get: function () {
            return this.zj;
        },
        set: function (a) {
            this.zj = a;
        },
        enumerable: false,
        configurable: true
    });
    ScatterPolygonSeries.prototype.zm = function () {
        if (this.zl == 0) {
            return 2;
        }
        return this.zl;
    };
    ScatterPolygonSeries.prototype.se = function (a) {
        _super.prototype.se.call(this, a);
        this.zl = a;
    };
    ScatterPolygonSeries.prototype.sg = function (a) {
        _super.prototype.sg.call(this, a);
        this.zz = a;
    };
    ScatterPolygonSeries.prototype.sb = function () {
        _super.prototype.sb.call(this);
        if (this.index < 0) {
            return;
        }
        if (this.aai != null) {
            this.zu.dv();
            this.zu.ds();
        }
        else {
            var a = MarkerSeries.w3(this, this.zr);
            var b = MarkerSeries.xg(a);
            if (b == null) {
                this.aag = MarkerSeries.xv;
            }
            else {
                this.zu.dp(b);
            }
        }
        if (this.aaz != null && this.zr != 13) {
            this.r2("ActualMarkerBrush");
            var c = this.aax;
            this.zu.dt();
            this.zu.dq();
            var d = this.aax;
            var e = this.ff("ActualMarkerBrush", this.ht.f > 0 ? this.ht.d() : null, c, d);
            this.r4("ActualMarkerBrush", e);
        }
        else {
            if (this.zr == 13) {
                this.aax = this.ct.bt();
            }
            else {
                if (this.zk != 2) {
                    if (this.c9 != null) {
                        var f = this.c9.uv(this.index);
                        if (!isNaN_(this.ia) && this.ia != 1) {
                            f = BrushUtil.q(f, this.ia);
                        }
                        this.aax = f;
                    }
                    else {
                        this.aax = null;
                    }
                }
            }
        }
        if (this.aa0 != null) {
            this.zu.du();
            this.zu.dr();
        }
        else {
            if (this.zr == 13) {
                this.aay = this.ct.bt();
            }
            else {
                if (this.zp == 2) {
                    this.aay = this.aax;
                }
                else {
                    this.aay = this.c9 == null ? null : this.c9.uw(this.index);
                }
            }
        }
        if (this.zk == 2 && this.aaz == null && this.zr != 13) {
            if (this.aay != null) {
                var g = this.aay;
                if (!isNaN_(this.ia) && this.ia != 1) {
                    g = BrushUtil.q(g, this.ia);
                }
                this.aax = g;
            }
            else {
                this.aax = null;
            }
        }
    };
    ScatterPolygonSeries.prototype.qo = function () {
        _super.prototype.qo.call(this);
        this.zo = this.zq();
        this.zk = this.zm();
    };
    ScatterPolygonSeries.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
        switch (b) {
            case ScatterPolygonSeries.$$p[3]:
            case ScatterPolygonSeries.$$p[11]:
            case ScatterPolygonSeries.$$p[7]:
            case ScatterPolygonSeries.$$p[9]:
            case ScatterPolygonSeries.$$p[10]:
                this.rb(b);
                this.qz();
                break;
            case ScatterPolygonSeries.$$p[6]:
                this.zo = this.zq();
                this.rb(null);
                this.qz();
                break;
            case ScatterPolygonSeries.$$p[5]:
                this.zk = this.zm();
                this.rb(null);
                this.qz();
                break;
            case ScatterPolygonSeries.$$p[2]:
                this.aah = d;
                if (c == MarkerSeries.xv || d == MarkerSeries.xv || (c == null || d != null)) {
                    this.zu.dx();
                    this.rn(false);
                }
                this.ct.bl();
                if (this.toolTip != null && this.fg) {
                    this.a9.a$as(this, this.toolTip);
                }
                this.rn(false);
                this.qh();
                break;
            case ScatterPolygonSeries.$$p[12]:
                this.zu.cy(this.aa3);
                this.rn(false);
                this.qh();
                break;
            case ScatterPolygonSeries.$$p[13]:
                this.zu.cz(this.zw);
                this.rn(false);
                this.qh();
                break;
            case ScatterPolygonSeries.$$p[4]:
                this.rn(false);
                this.qh();
                break;
            case ScatterPolygonSeries.$$p[0]:
            case ScatterPolygonSeries.$$p[1]:
                if (this.zx()) {
                    this.rn(false);
                }
                break;
        }
    };
    ScatterPolygonSeries.prototype.a1 = function (a, b) {
        var c = this.zu.b4(a);
        var d = null;
        if (c != null) {
            d = c.content;
        }
        if (d != null) {
            return d;
        }
        return _super.prototype.a1.call(this, a, b);
    };
    ScatterPolygonSeries.prototype.aaf = function (a, b) {
        var e_1, _a, e_2, _b;
        var c = new List$1(Base.$, 0);
        try {
            for (var _c = __values(fromEnum(b.c)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var d = _d.value;
                if (!a.containsKey(d)) {
                    c.add1(d);
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        try {
            for (var _e = __values(fromEnum(c)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var e = _f.value;
                b.h(e);
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
    };
    ScatterPolygonSeries.prototype.d9 = function (a, b, c, d) {
        var _this = this;
        if (!this.hasMarkers) {
            return false;
        }
        var e = this.w2;
        var f = new ScalerParams(1, this.c9.vl, this.wl, e.ch);
        f.c = this.getEffectiveViewport1(this.ct);
        var g = this.w4;
        var h = new ScalerParams(1, this.c9.vl, this.wl, g.ch);
        h.c = this.getEffectiveViewport1(this.ct);
        a.clear();
        this.zu.dw(function (i) {
            if (i._visibility == 0) {
                var j = new SeriesPointOfInterest();
                var k = i.n;
                var l = i.o;
                if (k < _this.wl.left || k > _this.wl.right || l < _this.wl.top || l > _this.wl.bottom) {
                    return;
                }
                j.d = e.en(k, f);
                j.e = g.en(l, h);
                j.b = i.content;
                j.g = _this.va;
                j.c = i.ak;
                if (_this.g7) {
                    j.g = _this.vh(i.ak);
                }
                a.add(j);
            }
        });
        return true;
    };
    ScatterPolygonSeries.prototype.zh = function (a, b, c, d, e) {
        var _this = this;
        if (c == null || c.length != this.zv.d.count) {
            c = new Array(this.zv.d.count);
            for (var f = 0; f < this.zv.d.count; f++) {
                c[f] = PointUtil.create();
            }
        }
        var g = ((function () {
            var $ret = new ScalerParams(1, d, e, _this.w2.ch);
            $ret.c = _this.getEffectiveViewport1(a);
            return $ret;
        })());
        var h = ((function () {
            var $ret = new ScalerParams(1, d, e, _this.w4.ch);
            $ret.c = _this.getEffectiveViewport1(a);
            return $ret;
        })());
        var i = this.wd(a);
        var j = this.zv.a.en(i.left, g);
        var k = this.zv.a.en(i.right, g);
        var l = this.zv.b.en(i.bottom, h);
        var m = this.zv.b.en(i.top, h);
        if (this.zv.e) {
            var n = j;
            j = k;
            k = n;
        }
        if (this.zv.f) {
            var o = l;
            l = m;
            m = o;
        }
        var p = this.zv;
        var q = p.a;
        var r = p.b;
        var s;
        var t;
        var u = g.d.left;
        var v = g.d.right;
        var w = h.d.top;
        var x = h.d.bottom;
        for (var y = 0; y < this.zv.d.count; y++) {
            s = 0;
            t = 0;
            var z = null;
            var aa = a.cc.tryGetValue(y, z);
            z = aa.p1;
            if (z == null) {
                c[y].x = NaN;
                c[y].y = NaN;
            }
            else {
                var ab = z.c;
                var ac = RectUtil.l(ab);
                s = ac.x;
                t = ac.y;
                if (s >= u && s <= v && t >= w && t <= x) {
                    c[y].x = s;
                    c[y].y = t;
                }
                else {
                    c[y].x = NaN;
                    c[y].y = NaN;
                }
            }
        }
        return c;
    };
    ScatterPolygonSeries.prototype.zy = function (a, b) {
        var e_3, _a;
        if (b == null) {
            b = new List$1(Number_$type, 2, a.e);
        }
        else {
            b.clear();
        }
        var c = 0;
        var d = this.dl;
        try {
            for (var _b = __values(fromEnum(a.c)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var e = _c.value;
                b.add(d.indexOf(e));
                c++;
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
        return b;
    };
    ScatterPolygonSeries.$t = markType(ScatterPolygonSeries, 'ScatterPolygonSeries', ShapeSeriesBase.$);
    ScatterPolygonSeries.aa1 = new Rect(0, 0, 0, 1, 1);
    ScatterPolygonSeries.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterPolygonSeries, 'raisePropertyChanged', ['ActualMarkerBrush:aax:aaj', [Brush.$], 'ActualMarkerOutline:aay:aak', [Brush.$], 'ActualMarkerTemplate:aag:aal', [DataTemplate.$, null], 'MarkerBrush:aaz:aam', [Brush.$], 'MarkerCollisionAvoidance:zi:aan', [CollisionAvoidanceType_$type, enumGetBox(CollisionAvoidanceType_$type, 0)], 'MarkerFillMode:zl:aao', [MarkerFillMode_$type, enumGetBox(MarkerFillMode_$type, 0)], 'MarkerOutlineMode:zp:aap', [MarkerOutlineMode_$type, enumGetBox(MarkerOutlineMode_$type, 0)], 'MarkerOutline:aa0:aaq', [Brush.$], 'MarkerStyle:aa2:aar', [Style.$], 'MarkerTemplate:aai:aas', [DataTemplate.$, null], 'MarkerThickness:zz:aat', [1, DeviceUtils.g(2)], 'MarkerType:zr:aau', [MarkerType_$type, enumGetBox(MarkerType_$type, 1)], 'ShapeStyle:aa3:aav', [Style.$, null], 'ShapeStyleSelector:zw:aaw', [StyleSelector.$, null]]);
    return ScatterPolygonSeries;
}(ShapeSeriesBase));
export { ScatterPolygonSeries };
/**
 * @hidden
 */
var ScatterPolygonSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolygonSeriesView, _super);
    function ScatterPolygonSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.dk = false;
        _this.db = null;
        _this.dm = null;
        _this.dh = null;
        _this.df = null;
        _this.dj = false;
        _this.di = null;
        _this.dn = null;
        _this.dc = new Marker();
        _this.d4 = null;
        _this.dc = new Marker();
        _this.dc.content = new DataContext();
        _this.dh = a;
        _this.di = new HashPool$2(Base.$, Marker.$);
        _this.dn = new List$1(Marker.$, 0);
        _this.dy(_this.di);
        _this.df = new NumericMarkerManager(1, function (b) { return _this.di.item(b); }, function (b) { return _this.dh.zv.d.item(b); }, runOn(_this, _this.d2), runOn(_this, _this.da), runOn(_this, _this.dl), function () { return _this.dh.zi; });
        _this.df.q = runOn(_this, _this.d5);
        return _this;
    }
    ScatterPolygonSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.dh.aa0 = Defaults.am;
            this.dh.jn = Defaults.z;
            this.dh.zz = Defaults.y;
            this.dh.zr = Defaults.f;
            this.dh.zl = Defaults.b;
        }
    };
    Object.defineProperty(ScatterPolygonSeriesView.prototype, "dg", {
        get: function () {
            return this.df;
        },
        enumerable: false,
        configurable: true
    });
    ScatterPolygonSeriesView.prototype.d2 = function (a) {
        this.dh.aaf(a, this.di);
    };
    ScatterPolygonSeriesView.prototype.da = function () {
        this.db = this.dh.zh(this, this.di, this.db, this.by, this.bx);
        return this.db;
    };
    ScatterPolygonSeriesView.prototype.dl = function () {
        this.dm = this.dh.zy(this.di, this.dm);
        return this.dm;
    };
    ScatterPolygonSeriesView.prototype.c5 = function (a, b) {
        return this.c6(a, b, true);
    };
    ScatterPolygonSeriesView.prototype.cm = function (a, b, c) {
        _super.prototype.cm.call(this, a, b, c);
        ShapeSeriesViewBase.cl(this, a, b);
    };
    ScatterPolygonSeriesView.prototype.ck = function (a, b) {
        var c = a;
        if (c == null) {
            return;
        }
        c.an = b;
    };
    ScatterPolygonSeriesView.prototype.dw = function (a) {
        this.di.f(a);
    };
    ScatterPolygonSeriesView.prototype.dy = function (a) {
        a.create = runOn(this, this.de);
        a.destroy = runOn(this, this.d0);
        a.activate = runOn(this, this.dz);
        a.disactivate = runOn(this, this.d1);
    };
    ScatterPolygonSeriesView.prototype.de = function () {
        var _this = this;
        var a = new Marker();
        if (!this.dj) {
            a.content = ((function () {
                var $ret = new DataContext();
                $ret.series = _this.e;
                return $ret;
            })());
        }
        a.ah = this.e.aag;
        this.dn.add(a);
        return a;
    };
    ScatterPolygonSeriesView.prototype.b4 = function (a) {
        var b = this.dd(a, 0);
        if (b == null) {
            b = this.dd(a, 3);
        }
        return b;
    };
    ScatterPolygonSeriesView.prototype.dd = function (a, b) {
        var c;
        var d;
        var e;
        var f;
        for (var g = this.dn.count - 1; g >= 0; g--) {
            var h = this.dn._inner[g];
            if (h._visibility == 1 || h._opacity == 0) {
                continue;
            }
            c = (h.m / 2) + b;
            d = (h.l / 2) + b;
            e = h.ai;
            f = h.aj;
            if ((h.n + e) - c <= a.x && (h.n + e) + c >= a.x && (h.o + f) - d <= a.y && (h.o + f) + d >= a.y) {
                return h;
            }
        }
        return null;
    };
    ScatterPolygonSeriesView.prototype.d0 = function (a) {
        a.content = null;
        this.dn.remove(a);
    };
    ScatterPolygonSeriesView.prototype.dz = function (a) {
        a._visibility = 0;
    };
    ScatterPolygonSeriesView.prototype.d1 = function (a) {
        if (a.content != null) {
            (typeCast(DataContext.$, a.content)).item = null;
        }
        a._visibility = 1;
    };
    ScatterPolygonSeriesView.prototype.d3 = function (a) {
        this.dj = a;
    };
    ScatterPolygonSeriesView.prototype.bj = function (a, b) {
        _super.prototype.bj.call(this, a, b);
        if (!this.dk) {
            var c = a;
            var d = c.content;
            if (d != null) {
                if (this.dh.aax != null) {
                    d.actualItemBrush = this.dh.aax;
                }
                else {
                    d.actualItemBrush = this.b6.u3;
                }
                if (this.dh.aax != null) {
                    d.outline = this.dh.aay;
                }
                else {
                    d.outline = this.b6.u7;
                }
                d.thickness = this.dh.zz;
            }
        }
    };
    ScatterPolygonSeriesView.prototype.bk = function (a, b) {
        _super.prototype.bk.call(this, a, b);
        var c = a;
        this.dc._visibility = c._visibility;
        this.dc.ah = c.ah;
        this.dc.width = c.width;
        this.dc.height = c.height;
        this.dc.m = c.m;
        this.dc.l = c.l;
        this.dc.n = c.n;
        this.dc.o = c.o;
        var d = this.bs(b);
        var e = this.dc.content;
        var f = c.content;
        e.item = f.item;
        e.series = f.series;
        e.thickness = f.thickness;
        if (e != null) {
            e.actualItemBrush = d;
            e.outline = d;
            e.thickness = this.dh.zz + 3;
        }
    };
    ScatterPolygonSeriesView.prototype.a7 = function (a, b) {
        _super.prototype.a7.call(this, a, b);
        var c = new DataTemplatePassInfo();
        c.isHitTestRender = b;
        c.context = a.h();
        c.viewportTop = this.bx.top;
        c.viewportLeft = this.bx.left;
        c.viewportWidth = this.bx.width;
        c.viewportHeight = this.bx.height;
        c.passID = "Markers";
        var d = new DataTemplateRenderInfo();
        d.passInfo = c;
        d.isHitTestRender = b;
        var e = new DataTemplateMeasureInfo();
        e.passInfo = c;
        var f = false;
        var g = a.h();
        e.context = g;
        d.context = g;
        var h = 0;
        var i = 0;
        if (this.dh.aag != null && this.dh.aag.passStarting != null) {
            this.dh.aag.passStarting(c);
        }
        for (var j = 0; j < this.dn.count; j++) {
            var k = this.dn._inner[j];
            if (k._visibility == 1) {
                continue;
            }
            this.bi(k, j, b);
            if (b) {
                k = this.dc;
            }
            if (!f) {
                e.width = k.width;
                e.height = k.height;
                e.renderOffsetX = 0;
                e.renderOffsetY = 0;
                e.renderContext = a;
                var l = k.ah;
                if (l.measure != null) {
                    e.data = k.content;
                    l.measure(e);
                    f = e.isConstant;
                    if (f) {
                        h = e.width;
                        i = e.height;
                    }
                }
                d.availableWidth = e.width;
                d.availableHeight = e.height;
                d.renderOffsetX = e.renderOffsetX;
                d.renderOffsetY = e.renderOffsetY;
                d.renderContext = a;
            }
            else {
                d.availableWidth = h;
                d.availableHeight = i;
            }
            if (!isNaN_(k.width) && !isInfinity(k.width)) {
                d.availableWidth = k.width;
            }
            if (!isNaN_(k.height) && !isInfinity(k.height)) {
                d.availableHeight = k.height;
            }
            a.q(d, k);
            k.m = d.availableWidth;
            k.l = d.availableHeight;
            k.ai = d.renderOffsetX;
            k.aj = d.renderOffsetY;
        }
        if (this.dh.aag != null && this.dh.aag.passCompleted != null) {
            this.dh.aag.passCompleted(c);
        }
    };
    ScatterPolygonSeriesView.prototype.dv = function () {
        this.dh.aag = null;
    };
    ScatterPolygonSeriesView.prototype.ds = function () {
        this.dh.aag = null;
        this.dh.aag = this.dh.aai;
    };
    ScatterPolygonSeriesView.prototype.dt = function () {
        this.dh.aax = null;
    };
    ScatterPolygonSeriesView.prototype.dq = function () {
        this.dh.aax = null;
        if (!isNaN_(this.e.ia) && this.e.ia != 1) {
            var a = BrushUtil.q(this.dh.aaz, this.e.ia);
            this.dh.aax = a;
        }
        else {
            this.dh.aax = this.dh.aaz;
        }
    };
    ScatterPolygonSeriesView.prototype.du = function () {
        this.dh.aay = null;
    };
    ScatterPolygonSeriesView.prototype.dr = function () {
        this.dh.aay = null;
        this.dh.aay = this.dh.aa0;
    };
    ScatterPolygonSeriesView.prototype.dp = function (a) {
        switch (a) {
            case "CircleMarkerTemplate":
                this.dh.aag = this.dh.c9.rn;
                break;
            case "TriangleMarkerTemplate":
                this.dh.aag = this.dh.c9.rw;
                break;
            case "PyramidMarkerTemplate":
                this.dh.aag = this.dh.c9.rt;
                break;
            case "SquareMarkerTemplate":
                this.dh.aag = this.dh.c9.ru;
                break;
            case "DiamondMarkerTemplate":
                this.dh.aag = this.dh.c9.ro;
                break;
            case "PentagonMarkerTemplate":
                this.dh.aag = this.dh.c9.rr;
                break;
            case "HexagonMarkerTemplate":
                this.dh.aag = this.dh.c9.rp;
                break;
            case "TetragramMarkerTemplate":
                this.dh.aag = this.dh.c9.rr;
                break;
            case "PentagramMarkerTemplate":
                this.dh.aag = this.dh.c9.rs;
                break;
            case "HexagramMarkerTemplate":
                this.dh.aag = this.dh.c9.rq;
                break;
        }
    };
    ScatterPolygonSeriesView.prototype.dx = function () {
        var e_4, _a;
        try {
            for (var _b = __values(fromEnum(this.dn)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var a = _c.value;
                a.ah = this.dh.aag;
            }
        }
        catch (e_4_1) {
            e_4 = { error: e_4_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_4)
                    throw e_4.error;
            }
        }
        this.as();
    };
    ScatterPolygonSeriesView.prototype.d5 = function (a) {
        if (this.d4 == null) {
            this.d4 = new DataTemplateMeasureInfo();
            this.d4.context = this.k.h();
        }
        this.d4.width = a.width;
        this.d4.height = a.height;
        this.d4.data = a.content;
        var b = a.ah;
        if (b.measure != null) {
            b.measure(this.d4);
        }
        return new Size(1, this.d4.width, this.d4.height);
    };
    ScatterPolygonSeriesView.prototype.an = function (a) {
        var e_5, _a;
        var b = a;
        _super.prototype.an.call(this, b);
        try {
            for (var _b = __values(fromEnum(this.di.c)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                var d = this.di.item(c);
                if (d._visibility != 1) {
                    var e = typeCast(ScatterPolygonSeries.$, this.e);
                    var f = this.e.bb.getMarkerVisualData(e, d, e.aag);
                    b.markerShapes.add(typeCast(MarkerVisualData.$, f));
                }
            }
        }
        catch (e_5_1) {
            e_5 = { error: e_5_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_5)
                    throw e_5.error;
            }
        }
    };
    ScatterPolygonSeriesView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span";
        var b = this.e.a9.a$u(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.dh.title + "</span><br/>";
        a += "</div>";
        return a;
    };
    ScatterPolygonSeriesView.$t = markType(ScatterPolygonSeriesView, 'ScatterPolygonSeriesView', ShapeSeriesViewBase.$);
    return ScatterPolygonSeriesView;
}(ShapeSeriesViewBase));
export { ScatterPolygonSeriesView };
