// igniteui-angular-charts/esm5/lib/ScatterBase_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { MarkerSeries } from "./MarkerSeries";
import { ISupportsErrorBars_$type } from "./ISupportsErrorBars";
import { ITrendLineSeries_$type } from "./ITrendLineSeries";
import { ScatterAxisInfoCache } from "./ScatterAxisInfoCache";
import { NumericXAxis } from "./NumericXAxis";
import { DependencyProperty } from "igniteui-angular-core";
import { NumericYAxis } from "./NumericYAxis";
import { Series } from "./Series";
import { TrendLineType_$type } from "igniteui-angular-core";
import { Brush } from "igniteui-angular-core";
import { PenLineCap_$type } from "igniteui-angular-core";
import { DoubleCollection } from "igniteui-angular-core";
import { CollisionAvoidanceType_$type } from "./CollisionAvoidanceType";
import { ScatterItemSearchMode_$type } from "./ScatterItemSearchMode";
import { ScatterErrorBarSettings } from "./ScatterErrorBarSettings";
import { HighlightingInfo } from "igniteui-angular-core";
import { ScatterFrame } from "./ScatterFrame";
import { Base, runOn, delegateCombine, enumGetBox, Number_$type, typeCast, fromEnum, Point_$type, markType, markDep, PointUtil } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { ScalerParams } from "./ScalerParams";
import { Rect } from "igniteui-angular-core";
import { Marker } from "./Marker";
import { MarkerSeriesView } from "./MarkerSeriesView";
import { SeriesLayerPropertyOverlay } from "./SeriesLayerPropertyOverlay";
import { IFastItemsSource_$type } from "igniteui-angular-core";
import { Clipper } from "igniteui-angular-core";
import { OwnedPoint } from "./OwnedPoint";
import { Dictionary$2 } from "igniteui-angular-core";
import { AxisRange } from "./AxisRange";
import { GeometryUtil } from "igniteui-angular-core";
import { HashPool$2 } from "igniteui-angular-core";
import { DataContext } from "igniteui-angular-core";
import { TrendResolutionParams } from "./TrendResolutionParams";
import { ScatterTrendLineManager } from "./ScatterTrendLineManager";
import { Thickness } from "igniteui-angular-core";
import { Size } from "igniteui-angular-core";
import { AssigningScatterStyleEventArgs } from "./AssigningScatterStyleEventArgs";
import { AssigningScatterMarkerStyleEventArgs } from "./AssigningScatterMarkerStyleEventArgs";
import { SeriesRenderManager } from "./SeriesRenderManager";
import { ErrorBarsHelper } from "./ErrorBarsHelper";
import { PathGeometry } from "igniteui-angular-core";
import { DataLegendSeriesValueInfo } from "igniteui-angular-core";
import { DataLegendSeriesContext } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { truncate, isNaN_, isInfinity } from "igniteui-angular-core";
import { boxArray$1, arrayShallowClone } from "igniteui-angular-core";
import { Path } from "igniteui-angular-core";
import { DataTemplateMeasureInfo } from "igniteui-angular-core";
import { NumericMarkerManager } from "./NumericMarkerManager";
import { stringIsNullOrEmpty } from "igniteui-angular-core";
/**
 * @hidden
 */
var ScatterBase = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterBase, _super);
    function ScatterBase() {
        var _this = _super.call(this) || this;
        _this.yo = null;
        _this.y7 = null;
        _this.y8 = null;
        _this.acc = null;
        _this.acd = null;
        _this.yx = 1;
        _this.ye = null;
        _this.yf = null;
        _this.yu = null;
        _this.yw = null;
        _this.ys = null;
        _this.zd = false;
        _this.zc = false;
        _this.y5 = null;
        _this.y6 = null;
        _this.assigningScatterStyle = null;
        _this.assigningScatterMarkerStyle = null;
        _this.y1 = null;
        _this.yv = null;
        _this.yr = null;
        _this.zb = true;
        _this.zq = 0;
        _this.zr = 0;
        _this.zs = 0;
        _this.zt = 0;
        _this.yy = _this.yz();
        _this.y1 = new ScatterSeriesRenderManager();
        _this.yv = new ScatterFrame();
        _this.acd = Rect.empty;
        _this.acc = Rect.empty;
        return _this;
    }
    ScatterBase.prototype.cq = function () {
        return new ScatterBaseView(this);
    };
    ScatterBase.prototype.o0 = function () {
        _super.prototype.o0.call(this);
        this.xAxis = null;
        this.yAxis = null;
    };
    ScatterBase.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.yp = a;
    };
    ScatterBase.prototype.get_fu = function () {
        return true;
    };
    ScatterBase.prototype.ai = function () {
        return [this.xAxis, this.yAxis];
    };
    ScatterBase.prototype.ko = function (a, b) {
        var c = this.nb(b);
        if (c == ScatterBase.$$p[12]) {
            return this.kp(a, b, this.aal);
        }
        if (c == ScatterBase.$$p[16]) {
            return this.kp(a, b, this.aat);
        }
        if (c == ScatterBase.$$p[1]) {
            return this.kp(a, b, this.z5);
        }
        if (c == ScatterBase.$$p[2]) {
            return this.kp(a, b, this.z7);
        }
        return null;
    };
    ScatterBase.prototype.mb = function (a) {
        var b = this.nb(a);
        if (b == ScatterBase.$$p[12]) {
            return this.aal;
        }
        if (b == ScatterBase.$$p[16]) {
            return this.aat;
        }
        if (b == ScatterBase.$$p[1]) {
            return this.z5;
        }
        if (b == ScatterBase.$$p[2]) {
            return this.z7;
        }
        return _super.prototype.mb.call(this, a);
    };
    ScatterBase.prototype.d9 = function (a, b, c, d) {
        var _this = this;
        if (!this.hasMarkers) {
            return false;
        }
        var e = this.xAxis;
        var f = new ScalerParams(1, this.c9.vl, this.wl, e.ch);
        f.c = this.getEffectiveViewport1(this.ct);
        var g = this.yAxis;
        var h = new ScalerParams(1, this.c9.vl, this.wl, g.ch);
        h.c = this.getEffectiveViewport1(this.ct);
        a.clear();
        this.wy.cj(function (i) {
            if (i._visibility == 0) {
                var j = new SeriesPointOfInterest();
                var k = i.n;
                var l = i.o;
                if (k < _this.wl.left || k > _this.wl.right || l < _this.wl.top || l > _this.wl.bottom) {
                    return;
                }
                j.d = e.en(k, f);
                j.e = g.en(l, h);
                j.b = i.content;
                j.g = _this.va;
                j.c = i.ak;
                if (_this.g7) {
                    j.g = _this.vh(i.ak);
                }
                a.add(j);
            }
        });
        return true;
    };
    ScatterBase.prototype.b0 = function () {
        if (this.b1 == 0) {
            if (this.z1 <= 2000) {
                return 1;
            }
            else {
                return _super.prototype.b0.call(this);
            }
        }
        else {
            return _super.prototype.b0.call(this);
        }
    };
    ScatterBase.prototype.wj = function (a) {
        if (!this.hasMarkers || !this.w8()) {
            return Rect.empty;
        }
        var b = this.wy.b1(this.v1(a));
        if (b != null) {
            return this.wy.cu(b);
        }
        return Rect.empty;
    };
    ScatterBase.prototype.gu = function (a, b) {
        if (this.w9(a, b)) {
            return true;
        }
        return false;
    };
    Object.defineProperty(ScatterBase.prototype, "y9", {
        get: function () {
            return this.y7;
        },
        set: function (a) {
            if (this.y7 != a) {
                var b = this.y9;
                this.y7 = a;
                this.raisePropertyChanged("XColumn", b, this.y9);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBase.prototype, "za", {
        get: function () {
            return this.y8;
        },
        set: function (a) {
            if (this.y8 != a) {
                var b = this.za;
                this.y8 = a;
                this.raisePropertyChanged("YColumn", b, this.za);
            }
        },
        enumerable: false,
        configurable: true
    });
    ScatterBase.prototype.f9 = function () {
        return this.highlightedItemsSource != null || (this.z5 != null && this.z7 != null);
    };
    ScatterBase.prototype.zo = function () {
        return false;
    };
    ScatterBase.prototype.hr = function () {
        var a = _super.prototype.hr.call(this);
        {
            var b = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "YMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedYMemberPath";
                $ret.q = true;
                return $ret;
            })());
            b.valueResolving = delegateCombine(b.valueResolving, runOn(this, this.aa0));
            a.add(b);
        }
        {
            var c = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.ad = "XMemberPath";
                $ret.p = true;
                $ret.y = "HighlightedXMemberPath";
                $ret.q = true;
                return $ret;
            })());
            c.valueResolving = delegateCombine(c.valueResolving, runOn(this, this.aaz));
            a.add(c);
        }
        if (this.zo()) {
            var d = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "DashArray";
                $ret.p = true;
                $ret.x = ((function () {
                    var $ret = new DoubleCollection();
                    $ret.add(5);
                    $ret.add(5);
                    return $ret;
                })());
                $ret.q = true;
                return $ret;
            })());
            a.add(d);
        }
        if (!this.zo()) {
            var e = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Brush";
                $ret.p = true;
                $ret.y = "ActualBrush";
                $ret.q = true;
                return $ret;
            })());
            e.valueResolving = delegateCombine(e.valueResolving, runOn(this, this.rv));
            a.add(e);
        }
        if (!this.zo()) {
            var f = ((function () {
                var $ret = new SeriesLayerPropertyOverlay();
                $ret.r = true;
                $ret.ad = "Outline";
                $ret.p = true;
                $ret.y = "ActualOutline";
                $ret.q = true;
                return $ret;
            })());
            f.valueResolving = delegateCombine(f.valueResolving, runOn(this, this.rv));
            a.add(f);
        }
        return a;
    };
    ScatterBase.prototype.aaz = function (a, b) {
        if (this.z5 != null) {
            b.value = this.z5;
            return;
        }
        b.value = this.aal;
    };
    ScatterBase.prototype.aa0 = function (a, b) {
        if (this.z7 != null) {
            b.value = this.z7;
            return;
        }
        b.value = this.aat;
    };
    Object.defineProperty(ScatterBase.prototype, "ab9", {
        get: function () {
            return this.c(ScatterBase.abk);
        },
        set: function (a) {
            this.h(ScatterBase.abk, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBase.prototype, "GetErrorBarsXAxis", {
        get: function () {
            return this.xAxis;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterBase.prototype, "GetErrorBarsYAxis", {
        get: function () {
            return this.yAxis;
        },
        enumerable: false,
        configurable: true
    });
    ScatterBase.prototype.p6 = function () {
        _super.prototype.p6.call(this);
        if (this.xAxis != null) {
            this.xAxis.h9(false);
        }
        if (this.yAxis != null) {
            this.yAxis.h9(false);
        }
    };
    ScatterBase.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
        if (this.yp.c2.g$b(a, b, c, d)) {
            this.rn(false);
            this.qh();
        }
        switch (b) {
            case ScatterBase.$$p[10]:
            case ScatterBase.$$p[14]:
            case ScatterBase.$$p[11]:
            case ScatterBase.$$p[15]:
                this.qb();
                break;
            case "FastItemsSource":
                if (typeCast(IFastItemsSource_$type, c) != null) {
                    c.deregisterColumn(this.y9);
                    c.deregisterColumn(this.za);
                    this.y9 = null;
                    this.za = null;
                }
                if (typeCast(IFastItemsSource_$type, d) != null) {
                    this.y9 = this.di(this.aal);
                    this.za = this.di(this.aat);
                }
                if ((this.yAxis != null && !this.yAxis.c1()) || (this.xAxis != null && !this.xAxis.c1())) {
                    this.rn(this.ge && c != null);
                }
                break;
            case ScatterBase.$$p[9]:
                if (c != null) {
                    c.b0(this);
                }
                if (d != null) {
                    d.ct(this);
                }
                if ((this.xAxis != null && !this.xAxis.c1()) || (d == null && c != null)) {
                    this.rn(false);
                }
                this.qh();
                break;
            case ScatterBase.$$p[13]:
                if (c != null) {
                    c.b0(this);
                }
                if (d != null) {
                    d.ct(this);
                }
                if ((this.yAxis != null && !this.yAxis.c1()) || (d == null && c != null)) {
                    this.rn(false);
                }
                this.qh();
                break;
            case ScatterBase.$$p[12]:
                if (this.dl != null) {
                    this.dl.deregisterColumn(this.y9);
                    this.y9 = this.di(this.aal);
                }
                break;
            case "XColumn":
                this.yp.c2.g$i();
                if (this.xAxis != null && !this.xAxis.c1()) {
                    this.rn(this.ge && this.d5);
                }
                break;
            case ScatterBase.$$p[16]:
                if (this.dl != null) {
                    this.dl.deregisterColumn(this.za);
                    this.za = this.di(this.aat);
                }
                break;
            case "YColumn":
                this.yp.c2.g$i();
                if (this.yAxis != null && !this.yAxis.c1()) {
                    this.rn(this.ge && this.d5);
                }
                break;
            case ScatterBase.$$p[7]:
                this.rn(false);
                this.qh();
                break;
            case ScatterBase.$$p[8]:
                this.rn(false);
                this.qh();
                break;
            case "TransitionProgress":
                this.yw.c(this.jo, this.yu, this.ys);
                this.aav();
                try {
                    if (this.d3(this.ct)) {
                        return;
                    }
                    if ((Math.round(this.jo * 100000) / 100000) == 1) {
                        this.abh(this.ys, this.yp);
                    }
                    else {
                        this.abh(this.yw, this.yp);
                    }
                    if (this.c9 != null) {
                        this.c9.qa(this);
                    }
                }
                finally {
                    this.abi();
                }
                break;
            case "TrendLineBrush":
                this.rb(b);
                break;
            case ScatterBase.$$p[0]:
                if (this.yq != null) {
                    this.yq.r = this;
                }
                this.rn(false);
                this.qh();
                break;
            case "TrendLineType":
                this.qh();
                break;
            case ScatterBase.$$p[4]:
                this.zd = this.zl();
                this.rn(false);
                break;
            case ScatterBase.$$p[3]:
                this.zc = this.zk();
                this.rn(false);
                break;
            case ScatterBase.$$p[5]:
                this.yy = this.yz();
                break;
        }
    };
    ScatterBase.prototype.zl = function () {
        return this.zh;
    };
    ScatterBase.prototype.zk = function () {
        return this.zg;
    };
    ScatterBase.prototype.zf = function (a) {
        if (typeCast(NumericYAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    ScatterBase.prototype.ze = function (a) {
        if (typeCast(NumericXAxis.$, a) !== null) {
            return true;
        }
        return false;
    };
    ScatterBase.prototype.abi = function () {
        this.acc = Rect.empty;
        this.acd = Rect.empty;
    };
    ScatterBase.prototype.aav = function () {
        var a = this.getViewInfo(this.acc, this.acd);
        this.acc = a.p0;
        this.acd = a.p1;
    };
    ScatterBase.prototype.zj = function (a, b) {
        if (b != 3) {
            return true;
        }
        if (a == null) {
            return true;
        }
        if (this.aal == a || this.aat == a) {
            return true;
        }
        return false;
    };
    ScatterBase.prototype.o9 = function (a, b, c, d) {
        var e = false;
        if (!this.zj(d, a)) {
            return;
        }
        this.yp.c2.g$e(a, b, c, d);
        if (this.xAxis != null && !this.xAxis.c1()) {
            e = true;
        }
        if (this.yAxis != null && !this.yAxis.c1()) {
            e = true;
        }
        if (a == 0 || a == 2 || a == 3 || a == 4) {
            this.zb = true;
        }
        else if (a == 1) {
            for (var f = b; f < b + c; f++) {
                this.zq += this.y9.item(f);
                this.zr += this.za.item(f);
            }
            this.zs = this.zq / this.y9.count;
            this.zt = this.zr / this.za.count;
        }
        if (e) {
            this.rn(true);
        }
    };
    ScatterBase.prototype.aa9 = function (a) {
        this.aba(a, null);
    };
    ScatterBase.prototype.aba = function (a, b) {
        var e_1, _a, e_2, _b;
        var c = this.y9 != null ? this.y9.count : 0;
        var d = this.za != null ? this.za.count : 0;
        var e = Math.min(c, d);
        if (e <= this.z1) {
            a.m.clear();
            var f = new List$1(OwnedPoint.$, 0);
            try {
                for (var _c = __values(fromEnum(a.u.values)), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var g = _d.value;
                    f.add(g);
                }
            }
            catch (e_1_1) {
                e_1 = { error: e_1_1 };
            }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return))
                        _a.call(_c);
                }
                finally {
                    if (e_1)
                        throw e_1.error;
                }
            }
            var h_1 = this.dl;
            f.aa(function (i, j) {
                var k = h_1.indexOf(i.a);
                var l = h_1.indexOf(j.a);
                if (k < l) {
                    return -1;
                }
                if (k > l) {
                    return 1;
                }
                return 0;
            });
            if (b != null) {
                b.i = a.m;
            }
            try {
                for (var _e = __values(fromEnum(f)), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var i = _f.value;
                    if (h_1.indexOf(i.a) >= 0) {
                        if (b != null) {
                            b.j(i.c);
                        }
                        else {
                            a.m.add({ $type: Point_$type, x: i.c.x, y: i.c.y });
                        }
                    }
                }
            }
            catch (e_2_1) {
                e_2 = { error: e_2_1 };
            }
            finally {
                try {
                    if (_f && !_f.done && (_b = _e.return))
                        _b.call(_e);
                }
                finally {
                    if (e_2)
                        throw e_2.error;
                }
            }
        }
    };
    ScatterBase.prototype.ax = function (a) {
        if (a != null && a == this.xAxis && this.y9 != null) {
            return new AxisRange(this.y9.minimum, this.y9.maximum);
        }
        if (a != null && a == this.yAxis && this.za != null) {
            return new AxisRange(this.za.minimum, this.za.maximum);
        }
        return null;
    };
    ScatterBase.prototype.kl = function (a) {
        var b = this.j1(a);
        if (this.dl != null && b >= 0 && b < this.dl.count) {
            return this.dl.item(b);
        }
        return null;
    };
    Object.defineProperty(ScatterBase.prototype, "yy", {
        get: function () {
            return this.yx;
        },
        set: function (a) {
            var b = this.yx;
            this.yx = a;
            if (b != this.yx) {
                this.raisePropertyChanged("ActualItemSearchMode", enumGetBox(ScatterItemSearchMode_$type, b), enumGetBox(ScatterItemSearchMode_$type, this.yx));
            }
        },
        enumerable: false,
        configurable: true
    });
    ScatterBase.prototype.yz = function () {
        if (this.y0 != 0) {
            return this.y0;
        }
        return this.zi ? 4 : 1;
    };
    ScatterBase.prototype.j1 = function (a) {
        var b = truncate(Math.round(this.ip(a)));
        return b;
    };
    ScatterBase.prototype.ip = function (a) {
        switch (this.yy) {
            case 1: return this.zx(a);
            case 3: return this.zy(a);
            case 2: return this.zv(a);
            case 4: return this.zw(a);
            case 5: return this.zu(a);
        }
        return -1;
    };
    ScatterBase.prototype.zw = function (a) {
        return this.zz(a, false);
    };
    ScatterBase.prototype.zv = function (a) {
        var b = this.v1(a);
        var c = b.x;
        var d = b.y;
        if (this.dl == null || this.xAxis == null || this.yAxis == null || this.y9 == null || this.za == null) {
            return -1;
        }
        if (this.dl.count > this.z0) {
            return -1;
        }
        this.aay();
        var e = this.ye;
        var f = this.yf;
        if (e == null || f == null) {
            return -1;
        }
        var g = 1.7976931348623157E+308;
        var h = -1;
        for (var i = 0; i < this.y9.count; i++) {
            var j = e[i];
            var k = f[i];
            var l = (c - j) * (c - j) + (d - k) * (d - k);
            if (l < g) {
                g = l;
                h = i;
            }
        }
        return h;
    };
    ScatterBase.prototype.aay = function () {
        var _this = this;
        var a = this.ct.by;
        var b = this.ct.bx;
        var c = this.getEffectiveViewport();
        if (this.yo == null) {
            return;
        }
        var d = new ScalerParams(0, a, b, this.yo.d, c);
        var e = ((function () {
            var $ret = new ScalerParams(0, a, b, _this.yo.e, c);
            $ret.b = _this.iw();
            return $ret;
        })());
        if (this.ye == null || this.yf == null) {
            var f = arrayShallowClone(this.y9.asArray());
            var g = arrayShallowClone(this.za.asArray());
            this.xAxis.ho(boxArray$1(f), 0, f.length, d);
            this.yAxis.ho(boxArray$1(g), 0, g.length, e);
            this.ye = f;
            this.yf = g;
        }
    };
    ScatterBase.prototype.get_zi = function () {
        return false;
    };
    Object.defineProperty(ScatterBase.prototype, "zi", {
        get: function () {
            return this.get_zi();
        },
        enumerable: false,
        configurable: true
    });
    ScatterBase.prototype.zz = function (a, b) {
        if (!this.zi) {
            return -1;
        }
        var c = { $type: Point_$type, x: 0, y: 0 };
        var d = true;
        var e = this.yt();
        if (e == null) {
            return -1;
        }
        if (e.m == null) {
            return -1;
        }
        if (this.dl == null || this.dl.count > this.z0) {
            return -1;
        }
        this.aay();
        var f = this.ye;
        var g = this.yf;
        if (f == null || g == null) {
            return -1;
        }
        var h = this.v1(a);
        var i = 1.7976931348623157E+308;
        var j = -1;
        for (var k = 0; k < this.ye.length; k++) {
            var l = { $type: Point_$type, x: this.ye[k], y: this.yf[k] };
            if (k > 0) {
                if (!b || e.j.containsKey(this.dl.item(k)) || e.j.containsKey(this.dl.item(k - 1))) {
                    var m = GeometryUtil.g(h, c, l);
                    if (m < i) {
                        i = m;
                        j = k;
                    }
                }
            }
            c = l;
        }
        if (j >= 0) {
            var n = { $type: Point_$type, x: this.ye[j], y: this.yf[j] };
            var o = { $type: Point_$type, x: this.ye[j - 1], y: this.yf[j - 1] };
            if (!b || (e.j.containsKey(this.dl.item(j)) && e.j.containsKey(this.dl.item(j - 1)))) {
                var p = GeometryUtil.l(h, o, n);
                return (j - 1) + p;
            }
            else if (e.j.containsKey(this.dl.item(j))) {
                return j;
            }
            else {
                return j - 1;
            }
        }
        return -1;
    };
    ScatterBase.prototype.zy = function (a) {
        return this.zz(a, true);
    };
    ScatterBase.prototype.zu = function (a) {
        var _this = this;
        var b = this.v1(a);
        var c = b.x;
        var d = b.y;
        if (this.yp.c3.e > this.z0) {
            return -1;
        }
        var e = -1;
        this.wy.ck(function (f) {
            var g = _this.yp.cu(f);
            var h = g.left;
            var i = g.top;
            var j = g.width;
            var k = g.height;
            if (_this.w5) {
                var l = h + j / 2;
                var m = i + k / 2;
                var n = (j / 2) * (j / 2);
                var o = (c - l) * (c - l) + (d - m) * (d - m);
                if (o < n) {
                    if (e == -1) {
                        e = f.ak;
                    }
                }
            }
            else {
                if (c > h && c < h + j && d > i && d < i + k) {
                    if (e == -1) {
                        e = f.ak;
                    }
                }
            }
        });
        return e;
    };
    ScatterBase.prototype.zx = function (a) {
        var _this = this;
        var b = null;
        var c = 1.7976931348623157E+308;
        var d = this.v1(a);
        var e = d.x;
        var f = d.y;
        if (this.yp.c3.e > this.z0) {
            return -1;
        }
        this.yp.da(function (g) {
            if (g._visibility == 0) {
                var h = _this.yp.cu(g);
                var i = h.left + h.width / 2;
                var j = h.top + h.height / 2;
                var k = (e - i) * (e - i) + (f - j) * (f - j);
                if (k < c) {
                    c = k;
                    b = g;
                }
            }
        });
        if (b != null) {
            return b.ak;
        }
        return -1;
    };
    ScatterBase.prototype.v6 = function (a, b, c) {
        var _this = this;
        if (this.xAxis == null || this.yAxis == null || this.y9 == null || this.za == null) {
            return { $type: Point_$type, x: NaN, y: NaN };
        }
        var d = this.ct.by;
        var e = this.ct.bx;
        var f = this.getEffectiveViewport();
        var g = new ScalerParams(0, d, e, this.xAxis.ch, f);
        var h = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yAxis.ch, f);
            $ret.b = _this.iw();
            return $ret;
        })());
        if (b) {
            var i = this.ip(a);
            if (i == -1) {
                return { $type: Point_$type, x: NaN, y: NaN };
            }
            var j = truncate(Math.floor(i));
            var k = truncate(Math.ceil(i));
            if (j < 0) {
                j = 0;
            }
            if (k < 0) {
                k = 0;
            }
            if (k > this.dl.count - 1) {
                k = this.dl.count - 1;
            }
            if (j > this.dl.count - 1) {
                j = this.dl.count - 1;
            }
            if (k == j) {
                var l = this.xAxis.ek(this.y9.item(j), g);
                var m = this.yAxis.ek(this.za.item(j), h);
                return { $type: Point_$type, x: l, y: m };
            }
            var n = i - Math.floor(i);
            var o = this.xAxis.ek(this.y9.item(j), g);
            var p = this.yAxis.ek(this.za.item(j), h);
            var q = this.xAxis.ek(this.y9.item(k), g);
            var r = this.yAxis.ek(this.za.item(k), h);
            var s = q - o;
            var t = r - p;
            return { $type: Point_$type, x: o + s * n, y: p + t * n };
        }
        else {
            var u = this.j1(a);
            if (u == -1) {
                return { $type: Point_$type, x: NaN, y: NaN };
            }
            var v = this.xAxis.ek(this.y9.item(u), g);
            var w = this.yAxis.ek(this.za.item(u), h);
            return { $type: Point_$type, x: v, y: w };
        }
    };
    ScatterBase.prototype.gb = function (a) {
        var _this = this;
        var b = this.c9 != null ? this.c9.vl : Rect.empty;
        b = b.copy();
        var c = this.c9 != null ? this.c9.v0 : Rect.empty;
        var d = new Rect(0, 0, 0, 1, 1);
        var e = this.wf(this.ct);
        var f = new ScalerParams(0, d, d, this.xAxis.ch, e);
        var g = ((function () {
            var $ret = new ScalerParams(0, d, d, _this.yAxis.ch, e);
            $ret.b = _this.iw();
            return $ret;
        })());
        var h = !b.isEmpty && !c.isEmpty && this.dl != null ? this.dl.indexOf(a) : -1;
        var i = this.xAxis != null && this.y9 != null && h < this.y9.count ? this.xAxis.ek(this.y9.item(h), f) : NaN;
        var j = this.yAxis != null && this.za != null && h < this.za.count ? this.yAxis.ek(this.za.item(h), g) : NaN;
        if (!isNaN_(i)) {
            if (i < b.left + 0.1 * b.width) {
                i = i + 0.4 * b.width;
                b.x = i - 0.5 * b.width;
            }
            if (i > b.right - 0.1 * b.width) {
                i = i - 0.4 * b.width;
                b.x = i - 0.5 * b.width;
            }
        }
        if (!isNaN_(j)) {
            if (j < b.top + 0.1 * b.height) {
                j = j + 0.4 * b.height;
                b.y = j - 0.5 * b.height;
            }
            if (j > b.bottom - 0.1 * b.height) {
                j = j - 0.4 * b.height;
                b.y = j - 0.5 * b.height;
            }
        }
        if (this.c1 != null) {
            this.c1.au(this.c9, b, true);
        }
        return h >= 0;
    };
    ScatterBase.prototype.sk = function (a, b) {
        this.rn(false);
    };
    ScatterBase.prototype.sl = function (a, b) {
        this.rn(false);
    };
    ScatterBase.prototype.yt = function () {
        var a = this.ys;
        if (this.d1()) {
            if (this.da.g) {
                this.da.t();
            }
            a = this.yw;
        }
        return a;
    };
    ScatterBase.prototype.aaw = function (a, b, c, d, e) {
        var _this = this;
        b.g = new Dictionary$2(Base.$, OwnedPoint.$, 1, c);
        var f = this.getEffectiveViewport1(a);
        var g = this.dl;
        var h;
        var i;
        var j = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yo.d, f);
            $ret.c = _this.getEffectiveViewport1(a);
            return $ret;
        })());
        var k = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yo.e, f);
            $ret.c = _this.getEffectiveViewport1(a);
            $ret.b = _this.iw();
            return $ret;
        })());
        var l = false;
        var m = 0;
        if (this.c9 != null) {
            if (this.c9.fz()) {
                var n = this.c9.mt(d, j);
                l = n.c;
                m = n.d;
            }
        }
        var _loop_1 = function (o) {
            h = this_1.y9.item(o);
            i = this_1.za.item(o);
            var p = { $type: Point_$type, x: this_1.yo.a.ek(h, j), y: this_1.yo.b.ek(i, k) };
            if (l) {
                if (p.x < e.left || p.x > e.right) {
                    if (p.x + m >= e.left && p.x + m <= e.right) {
                        p = { $type: Point_$type, x: p.x + m, y: i };
                    }
                }
            }
            if (!isInfinity(p.x) && !isInfinity(p.y)) {
                var q_1 = g.item(o);
                if (!b.g.containsKey(q_1)) {
                    var r_1 = { $type: Point_$type, x: h, y: i };
                    var s_1 = { $type: Point_$type, x: p.x, y: p.y };
                    b.g.addItem(q_1, ((function () {
                        var $ret = new OwnedPoint();
                        $ret.a = q_1;
                        $ret.b = r_1;
                        $ret.c = s_1;
                        return $ret;
                    })()));
                }
            }
        };
        var this_1 = this;
        for (var o = 0; o < c; o++) {
            _loop_1(o);
        }
    };
    ScatterBase.prototype.aa6 = function (a, b) {
        var _this = this;
        this.ye = null;
        this.yf = null;
        a.j.clear();
        a.n.clear();
        a.h.clear();
        a.l.clear();
        a.i.clear();
        a.k.clear();
        var c = Math.min(this.y9 != null ? this.y9.count : 0, this.za != null ? this.za.count : 0);
        var d = b.by;
        var e = b.bx;
        var f = this.getEffectiveViewport1(b);
        var g = new ScalerParams(0, d, e, this.xAxis.ch, f);
        g.c = this.getEffectiveViewport1(b);
        var h = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yAxis.ch, f);
            $ret.b = _this.iw();
            return $ret;
        })());
        h.c = this.getEffectiveViewport1(b);
        if (c < 1) {
            return;
        }
        this.yo = ((function () {
            var $ret = new ScatterAxisInfoCache();
            $ret.a = _this.xAxis;
            $ret.b = _this.yAxis;
            $ret.d = _this.xAxis.ch;
            $ret.e = _this.yAxis.ch;
            $ret.c = _this.dl;
            return $ret;
        })());
        var i = b;
        if (c <= this.z1) {
            this.aaw(i, a, c, d, e);
        }
        if (this.w8()) {
            b.c0.l(a.j, this.z1, d, e, this.ib);
        }
        var j = ((function () {
            var $ret = new Clipper(0, e, false);
            $ret.i = a.n;
            return $ret;
        })());
        var k = this.wd(b);
        var l = this.xAxis.en(k.left, g);
        var m = this.xAxis.en(k.right, g);
        b.c2.h(a.n, this.trendLineType, this.y9, this.za, this.trendLinePeriod, function (n) { return _this.xAxis.ek(n, g); }, function (n) { return _this.yAxis.ek(n, h); }, ((function () {
            var $ret = new TrendResolutionParams();
            $ret.d = _this.ib;
            $ret.i = e;
            $ret.h = f;
            $ret.j = d;
            return $ret;
        })()), j, l, m);
        this.aa4(a, b);
    };
    ScatterBase.prototype.o3 = function (a, b) {
        _super.prototype.o3.call(this, a, b);
        var c = b;
        c.c9(a);
    };
    ScatterBase.prototype.get_fc = function () {
        return true;
    };
    ScatterBase.prototype.y3 = function (a, b, c) {
        return null;
    };
    ScatterBase.prototype.y4 = function (a, b, c) {
        if (this.dl == null) {
            return null;
        }
        var d = this.yt();
        var e = this.yp.b1(this.v1(b));
        var f = -1;
        if (e == null) {
            if (a != null) {
                f = this.dl.indexOf(a);
            }
            else {
                return null;
            }
        }
        else {
            f = e.ak;
        }
        if (f == -1) {
            return null;
        }
        var g = new HighlightingInfo();
        g.i = this;
        g.g = f;
        g.f = f;
        g.c = false;
        g.d = true;
        if (c != null && c.g == g.g && c.f == g.f) {
            return c;
        }
        return g;
    };
    ScatterBase.prototype.wn = function (a, b) {
        if (this.w0 != 1 && this.w0 != 13) {
            var c = this.wy.cv();
            var d = (c.width / 2) + (this.xb / 2);
            var e = (c.height / 2) + (this.xb / 2);
            var f = d;
            var g = d;
            var h = e;
            var i = e;
            if (a == 2 || a == 4) {
                f = 0;
            }
            if (a == 2 || a == 3) {
                g = 0;
            }
            if (b == 2 || b == 3) {
                h = 0;
            }
            if (b == 2 || b == 4) {
                i = 0;
            }
            return new Thickness(1, f, h, g, i);
        }
        return new Thickness(1, 0, 0, 0, 0);
    };
    ScatterBase.prototype.y2 = function (a, b, c) {
        var d = new HighlightingInfo();
        d.i = this;
        d.g = 0;
        d.f = this.dl.count - 1;
        d.c = true;
        if (c != null && c.g == d.g && c.f == d.f) {
            return c;
        }
        return d;
    };
    ScatterBase.prototype.de = function (a, b, c) {
        if (this.eb && this.fx(this.bv) && !c) {
            var d = this.y3(a, b, this.y5);
            this.y5 = d;
            return d;
        }
        else {
            var e = this.y2(a, b, this.y5);
            this.y5 = e;
            return e;
        }
    };
    ScatterBase.prototype.df = function (a, b, c) {
        if (this.fx(this.bv) && !c) {
            var d = this.y4(a, b, this.y6);
            this.y6 = d;
            return d;
        }
        else {
            var e = this.y2(a, b, this.y6);
            e.d = true;
            this.y6 = e;
            return e;
        }
    };
    ScatterBase.prototype.zn = function () {
        return (this.assigningScatterStyle != null && this.zd) || this.d0 || (this.bv != 1) || (this.cn != 1) || (this.cm != 1);
    };
    ScatterBase.prototype.zm = function () {
        return (this.assigningScatterMarkerStyle != null && this.zc) || this.d0 || (this.bv != 1) || (this.cn != 1) || (this.cm != 1);
    };
    ScatterBase.prototype.abh = function (a, b) {
        var _this = this;
        this.qt();
        var c = b.bx;
        this.yo = ((function () {
            var $ret = new ScatterAxisInfoCache();
            $ret.a = _this.xAxis;
            $ret.b = _this.yAxis;
            $ret.d = _this.xAxis.ch;
            $ret.e = _this.yAxis.ch;
            return $ret;
        })());
        if (this.w8()) {
            b.c0.j(a.j, this.xa);
            if (this.zm()) {
                this.aax(b, a);
            }
        }
        b.cr();
        var d = ((function () {
            var $ret = new Clipper(1, NaN, c.bottom, NaN, c.top, false);
            $ret.i = b.c2.g$j.an;
            return $ret;
        })());
        b.c2.g$h(a.n, d);
        this.abe(a, b);
    };
    ScatterBase.prototype.yh = function (a, b) {
        return this.yi(a, b);
    };
    ScatterBase.prototype.yi = function (a, b) {
        if (b < a || a < 0 || a > this.y9.count || b < 0 || b > this.y9.count) {
            return null;
        }
        var c = new Array((b - a) + 1);
        for (var d = a; d <= b; d++) {
            var e = d;
            c[d - a] = this.dl.item(d);
        }
        return c;
    };
    ScatterBase.prototype.aax = function (a, b) {
        var _this = this;
        a.b7 = true;
        this.y1.ay(this, this.zm(), runOn(this, this.yh));
        var c = this.y1.d;
        var d = c != null;
        var e = this.getEffectiveViewport1(a);
        var f = this.dl == null ? 0 : this.dl.count;
        this.yp.cj(function (g) {
            if (g._visibility == 0) {
                var h = g.ak;
                var i = g.content;
                if (i == null) {
                    i = new DataContext();
                    g.content = i;
                }
                if (d) {
                    _this.aa1(h, f, a.q);
                }
                _this.y1.ac(g, i);
            }
        });
    };
    ScatterBase.prototype.aa2 = function (a, b, c) {
        var d = this.bv != 1;
        var e = this.cn != 1;
        var f = this.cm != 1;
        this.y1.a0(this.y1.b, a, b, false);
        this.y1.ab(this.h9);
        var g = this.y1.ax;
        var h = null;
        var i = null;
        var j = null;
        var k = this.dl;
        if (d && this.c9 != null) {
            h = this.c9.dj.getHighlightingInfo(this, k, g.z, g.y, false);
            g.c = h;
        }
        if (e && this.c9 != null) {
            h = this.c9.dk.getHighlightingInfo(this, k, g.z, g.y, false);
            g.d = h;
        }
        if (f && this.c9 != null) {
            h = this.c9.di.getHighlightingInfo(this, k, g.z, g.y, false);
            g.b = h;
        }
        g.h = c;
        this.abc(g);
        this.y1.z(this, h, i, j, c, d, e, f, a < 0);
    };
    ScatterBase.prototype.aa1 = function (a, b, c) {
        var d = this.bv != 1;
        var e = this.cn != 1;
        var f = this.cm != 1;
        this.y1.a0(this.y1.d, a, b, true);
        this.y1.aa(this.h9);
        var g = this.y1.aw;
        var h = null;
        var i = null;
        var j = null;
        var k = this.dl;
        var l = null;
        var m = null;
        var n = null;
        if (d && this.c9 != null) {
            h = this.c9.dj.getHighlightingInfo(this, k, g.z, g.y, true);
            if (this.c9.dj.isCrossContextHighlight(this)) {
                l = this.c9.dj.getHighlightingInfo(this, k, g.z, g.y, true);
            }
            g.c = h;
        }
        if (e && this.c9 != null) {
            i = this.c9.dk.getHighlightingInfo(this, k, g.z, g.y, true);
            if (i == null && this.bl != null) {
                i = this.c9.dk.getHighlightingInfo(this.bl, this.bl.dl, g.z, g.y, true);
            }
            if (this.c9.dk.isCrossContextHighlight(this)) {
                m = this.c9.dk.getHighlightingInfo(this, k, g.z, g.y, true);
            }
            g.d = i;
        }
        if (f && this.c9 != null) {
            j = this.c9.di.getHighlightingInfo(this, k, g.z, g.y, true);
            if (this.c9.di.isCrossContextHighlight(this)) {
                n = this.c9.di.getHighlightingInfo(this, k, g.z, g.y, true);
            }
            g.d = i;
        }
        g.h = c;
        this.abb(g);
        this.y1.y(this, h, i, j, c, d, e, f, g.v, this.c9.dj.totalHighWaterMark, g.x, this.c9.dk.totalHighWaterMark, g.t, this.c9.di.totalHighWaterMark);
    };
    ScatterBase.prototype.abc = function (a) {
        if (this.assigningScatterStyle != null && this.zd) {
            this.assigningScatterStyle(this, a);
        }
    };
    ScatterBase.prototype.abb = function (a) {
        if (this.assigningScatterMarkerStyle != null && this.zc) {
            this.assigningScatterMarkerStyle(this, a);
        }
    };
    ScatterBase.prototype.aa4 = function (a, b) {
        var e_3, _a;
        var c = new ErrorBarsHelper(this, b);
        if (this.yq == null) {
            return;
        }
        var d = this.yq.u;
        var e = this.yq.v;
        var f = 0;
        var g = 0;
        var h = 0;
        var i = 0;
        try {
            for (var _b = __values(fromEnum(a.j.keys)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var j = _c.value;
                var k = a.j.item(j);
                if (d != null) {
                    switch (d.getCalculatorType()) {
                        case 1:
                            var l = this.aa7(k, c, d, h, i, true);
                            h = l.p3;
                            i = l.p4;
                            break;
                        case 0:
                            var m = this.aa5(k, c, d, h, i, true);
                            h = m.p3;
                            i = m.p4;
                            break;
                        case 2:
                            var n = this.aa3(k, c, d, j, h, i, true);
                            h = n.p4;
                            i = n.p5;
                            break;
                        case 3:
                            var o = this.aa8(c, d, f, h, i, true);
                            f = o.p2;
                            h = o.p3;
                            i = o.p4;
                            break;
                        case 4:
                            var p = this.aa5(k, c, d, h, i, true);
                            h = p.p3;
                            i = p.p4;
                            break;
                    }
                    ;
                    var q = new OwnedPoint();
                    q.c = c.l(d, this.yo.a, k.c, f);
                    q.a = k.a;
                    a.h.addItem(j, q);
                    a.i.addItem(j, [h, i]);
                }
                if (e != null) {
                    switch (e.getCalculatorType()) {
                        case 1:
                            var r = this.aa7(k, c, e, h, i, false);
                            h = r.p3;
                            i = r.p4;
                            break;
                        case 0:
                            var s = this.aa5(k, c, e, h, i, false);
                            h = s.p3;
                            i = s.p4;
                            break;
                        case 2:
                            var t = this.aa3(k, c, e, j, h, i, false);
                            h = t.p4;
                            i = t.p5;
                            break;
                        case 3:
                            var u = this.aa8(c, e, g, h, i, false);
                            g = u.p2;
                            h = u.p3;
                            i = u.p4;
                            break;
                        case 4:
                            var v = this.aa5(k, c, e, h, i, false);
                            h = v.p3;
                            i = v.p4;
                            break;
                    }
                    ;
                    var w = new OwnedPoint();
                    w.c = c.m(e, this.yo.b, k.c, g);
                    w.a = k.a;
                    a.l.addItem(j, w);
                    a.k.addItem(j, [h, i]);
                }
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
    };
    ScatterBase.prototype.aa7 = function (a, b, c, d, e, f) {
        if (f) {
            var g = void 0;
            var h = void 0, i = void 0;
            i = this.yo.a;
            if (this.yq.s == 0) {
                g = a.c.x;
                h = this.yo.a;
            }
            else {
                g = a.c.y;
                h = this.yo.b;
            }
            var j = b.f(g, c, h, i, d, e);
            d = j.p4;
            e = j.p5;
        }
        else {
            var k = void 0;
            var l = void 0, m = void 0;
            m = this.yo.b;
            if (this.yq.t == 0) {
                k = a.c.x;
                l = this.yo.a;
            }
            else {
                k = a.c.y;
                l = this.yo.b;
            }
            var n = b.f(k, c, l, m, d, e);
            d = n.p4;
            e = n.p5;
        }
        return {
            p3: d,
            p4: e
        };
    };
    ScatterBase.prototype.aa5 = function (a, b, c, d, e, f) {
        if (f) {
            var g = b.j(a.c.x, c, this.yo.a, d, e);
            d = g.p3;
            e = g.p4;
        }
        else {
            var h = b.j(a.c.y, c, this.yo.b, d, e);
            d = h.p3;
            e = h.p4;
        }
        return {
            p3: d,
            p4: e
        };
    };
    ScatterBase.prototype.aa3 = function (a, b, c, d, e, f, g) {
        var h = c.getItemColumn();
        var i = this.dl.indexOf(d);
        if (h != null && i < h.count) {
            var j = h.item(i);
            if (g) {
                var k = b.h(a.c.x, j, this.yo.a, e, f);
                e = k.p3;
                f = k.p4;
            }
            else {
                var l = b.h(a.c.y, j, this.yo.b, e, f);
                e = l.p3;
                f = l.p4;
            }
        }
        else {
            e = NaN;
            f = NaN;
        }
        return {
            p4: e,
            p5: f
        };
    };
    ScatterBase.prototype.aa8 = function (a, b, c, d, e, f) {
        var g = a.i(b, c);
        c = g.p1;
        if (f) {
            var h = a.k(c, b, this.yo.a, d, e);
            d = h.p3;
            e = h.p4;
        }
        else {
            var i = a.k(c, b, this.yo.b, d, e);
            d = i.p3;
            e = i.p4;
        }
        return {
            p2: c,
            p3: d,
            p4: e
        };
    };
    ScatterBase.prototype.abe = function (a, b) {
        if (!b.o() || this.yq == null) {
            b.db();
            return;
        }
        this.abf(a, b);
        this.abg(a, b);
    };
    ScatterBase.prototype.abf = function (a, b) {
        var e_4, _a;
        b.c7();
        var c = new ErrorBarsHelper(this, b);
        var d = new PathGeometry();
        var e = this.yq.u;
        try {
            for (var _b = __values(fromEnum(a.j.keys)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var f = _c.value;
                if (e != null && a.i.containsKey(f)) {
                    var g = a.i.item(f)[0];
                    var h = a.i.item(f)[1];
                    if (!isNaN_(g) && !isNaN_(h)) {
                        var i = a.h.item(f).c;
                        if (this.yq.o == 1 || this.yq.o == 2) {
                            c.d(d, i, g, true);
                        }
                        if (this.yq.o == 1 || this.yq.o == 3) {
                            c.d(d, i, h, false);
                        }
                    }
                }
            }
        }
        catch (e_4_1) {
            e_4 = { error: e_4_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_4)
                    throw e_4.error;
            }
        }
        b.dc(d);
    };
    ScatterBase.prototype.abg = function (a, b) {
        var e_5, _a;
        b.c8();
        var c = new ErrorBarsHelper(this, b);
        var d = new PathGeometry();
        var e = this.yq.v;
        try {
            for (var _b = __values(fromEnum(a.j.keys)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var f = _c.value;
                if (e != null && a.k.containsKey(f)) {
                    var g = a.k.item(f)[0];
                    var h = a.k.item(f)[1];
                    if (!isNaN_(g) && !isNaN_(h)) {
                        var i = a.l.item(f).c;
                        if (this.yq.p == 1 || this.yq.p == 2) {
                            c.e(d, i, g, true);
                        }
                        if (this.yq.p == 1 || this.yq.p == 3) {
                            c.e(d, i, h, false);
                        }
                    }
                }
            }
        }
        catch (e_5_1) {
            e_5 = { error: e_5_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_5)
                    throw e_5.error;
            }
        }
    };
    ScatterBase.prototype.hc = function (a, b, c) {
        var d = true;
        if (!_super.prototype.hc.call(this, a, b, c) || b.isEmpty || a.isEmpty || this.xAxis == null || this.yAxis == null || this.xAxis.bi == null || this.yAxis.bi == null || this.y9 == null || this.za == null || this.y9.count == 0 || this.za.count == 0 || this.dl == null || this.dl.count != this.y9.count || this.dl.count != this.za.count || this.xAxis.bi == null || this.yAxis.bi == null || this.xAxis.ly == this.xAxis.lx || this.yAxis.ly == this.yAxis.lx) {
            d = false;
        }
        return d;
    };
    ScatterBase.prototype.getViewInfo = function (a, b) {
        if (!this.acc.isEmpty && !this.acd.isEmpty) {
            a = this.acc;
            b = this.acd;
            return {
                p0: a,
                p1: b
            };
        }
        a = this.ct.bx;
        b = this.ct.by;
        return {
            p0: a,
            p1: b
        };
    };
    ScatterBase.prototype.ro = function (a) {
        _super.prototype.ro.call(this, a);
        this.aav();
        try {
            if (this.d3(this.ct)) {
                return;
            }
            if (this.dl != null && this.dl.count > this.z1) {
                a = false;
            }
            if (this.gd(a) && !this.gp) {
                var b = this.yu;
                if (this.d1()) {
                    if (this.da.g) {
                        this.da.t();
                    }
                    this.yu = this.yw;
                    this.yw = b;
                }
                else {
                    this.yu = this.ys;
                    this.ys = b;
                }
                this.aa6(this.ys, this.yp);
                this.r1();
            }
            else {
                if (!this.gp) {
                    this.aa6(this.ys, this.yp);
                }
                this.abh(this.ys, this.yp);
            }
        }
        finally {
            this.abi();
        }
    };
    ScatterBase.prototype.sb = function () {
        _super.prototype.sb.call(this);
        if (this.index < 0) {
            return;
        }
        this.yp.df();
    };
    ScatterBase.prototype.rp = function (a, b) {
        _super.prototype.rp.call(this, a, b);
        if (!this.gy) {
            this.ct.a4(b);
            return;
        }
        this.ct.a4(b);
        if (this.d3(this.cs)) {
            return;
        }
        var c = typeCast(ScatterBaseView.$, this.cs);
        if (!this.gq) {
            this.yv = new ScatterFrame();
            this.aa6(this.yv, c);
        }
        this.gq = false;
        this.abh(this.yv, c);
        this.gy = false;
    };
    ScatterBase.prototype.rm = function (a, b, c, d, e) {
        _super.prototype.rm.call(this, a, b, c, d, e);
        var f = this.a9.a$j.item(d);
        var g = f;
        f.a3(c);
        if (this.d3(g)) {
            return;
        }
        if (this.yr == null) {
            this.yr = new ScatterFrame();
        }
        this.aa6(this.yr, g);
        this.abh(this.yr, g);
    };
    ScatterBase.prototype.abd = function (a, b) {
        var e_6, _a, e_7, _b;
        var c = new List$1(Base.$, 0);
        try {
            for (var _c = __values(fromEnum(b.c)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var d = _d.value;
                if (!a.containsKey(d)) {
                    c.add1(d);
                }
            }
        }
        catch (e_6_1) {
            e_6 = { error: e_6_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_6)
                    throw e_6.error;
            }
        }
        try {
            for (var _e = __values(fromEnum(c)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var e = _f.value;
                b.h(e);
            }
        }
        catch (e_7_1) {
            e_7 = { error: e_7_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_7)
                    throw e_7.error;
            }
        }
    };
    ScatterBase.prototype.yg = function (a, b, c, d, e) {
        var _this = this;
        if (c == null || c.length != this.yo.c.count) {
            c = new Array(this.yo.c.count);
            for (var f = 0; f < this.yo.c.count; f++) {
                c[f] = PointUtil.create();
            }
        }
        var g = this.getEffectiveViewport1(a);
        var h = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.xAxis.ch, g);
            $ret.c = _this.getEffectiveViewport1(a);
            return $ret;
        })());
        var i = ((function () {
            var $ret = new ScalerParams(0, d, e, _this.yAxis.ch, g);
            $ret.c = _this.getEffectiveViewport1(a);
            $ret.b = _this.iw();
            return $ret;
        })());
        var j = this.wd(a);
        var k = this.yo.a.en(j.left, h);
        var l = this.yo.a.en(j.right, h);
        var m = this.yo.b.en(j.bottom, i);
        var n = this.yo.b.en(j.top, i);
        if (this.yo.d) {
            var o = k;
            k = l;
            l = o;
        }
        if (this.yo.e) {
            var p = m;
            m = n;
            n = p;
        }
        var q = this.yo;
        var r = q.a;
        var s = q.b;
        var t;
        var u;
        var v = this.y9;
        var w = this.za;
        var x = false;
        var y = 0;
        var z = 0;
        if (this.c9 != null) {
            if (this.c9.fz()) {
                var aa = this.c9.mt(d, h);
                x = aa.c;
                y = aa.d;
                z = r.en(0 + y, h);
                var ab = r.en(0, h);
                z -= ab;
            }
        }
        for (var ac = 0; ac < this.yo.c.count; ac++) {
            t = v.item(ac);
            u = w.item(ac);
            var ad = 1E-10;
            if (x) {
                if (t >= k - ad && t <= l + ad && u >= m - ad && u <= n + ad) {
                    c[ac].x = r.ek(t, h);
                    c[ac].y = s.ek(u, i);
                }
                else {
                    if ((t + z) >= k && (t + z) <= l && u >= m && u <= n) {
                        c[ac].x = r.ek(t, h) + y;
                        c[ac].y = s.ek(u, i);
                    }
                    else {
                        c[ac].x = NaN;
                        c[ac].y = NaN;
                    }
                }
            }
            else {
                if (t >= k - ad && t <= l + ad && u >= m - ad && u <= n + ad) {
                    c[ac].x = r.ek(t, h);
                    c[ac].y = s.ek(u, i);
                }
                else {
                    c[ac].x = NaN;
                    c[ac].y = NaN;
                }
            }
        }
        return c;
    };
    ScatterBase.prototype.zp = function (a, b) {
        var e_8, _a;
        if (b == null) {
            b = new List$1(Number_$type, 2, a.e);
        }
        else {
            b.clear();
        }
        var c = 0;
        var d = this.dl;
        try {
            for (var _b = __values(fromEnum(a.c)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var e = _c.value;
                b.add(d.indexOf(e));
            }
        }
        catch (e_8_1) {
            e_8 = { error: e_8_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_8)
                    throw e_8.error;
            }
        }
        return b;
    };
    ScatterBase.prototype.pv = function (a) {
        _super.prototype.pv.call(this, a);
        this.bb.exportTrendlineData(this, a, this.yp.c2.g$j);
    };
    ScatterBase.prototype.iw = function () {
        return Series.ix(this.za, null);
    };
    ScatterBase.prototype.getDataLegendSeriesContextAt = function (a, b) {
        var c = { $type: Point_$type, x: a, y: b };
        var d = this.kl(c);
        var e = d == null ? NaN : this.iv(d, "XMemberPath");
        var f = d == null ? NaN : this.iv(d, "YMemberPath");
        var g = new DataLegendSeriesValueInfo();
        g.l = e;
        g.a = 13;
        g.q = "X";
        g.r = this.aal != null ? this.aal : "X";
        g.s = this.aah != null ? this.aah : "X:";
        g.t = this.aaj != null ? this.aaj : "";
        var h = new DataLegendSeriesValueInfo();
        h.l = f;
        h.a = 14;
        h.q = "Y";
        h.r = this.aat != null ? this.aat : "Y";
        h.s = this.aap != null ? this.aap : "Y:";
        h.t = this.aar != null ? this.aar : "";
        var i = new DataLegendSeriesContext();
        i.e = ((function () {
            var $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
            $ret.add(g);
            $ret.add(h);
            return $ret;
        })());
        i.h = this.ma();
        return i;
    };
    ScatterBase.prototype.abj = function () {
        if (this.zb) {
            this.zq = 0;
            this.zr = 0;
            for (var a = 0; a < this.y9.count; a++) {
                this.zq += this.y9.item(a);
                this.zr += this.za.item(a);
            }
            this.zs = this.zq / this.y9.count;
            this.zt = this.zr / this.za.count;
            this.zb = false;
        }
    };
    ScatterBase.prototype.i6 = function (a, b) {
        if (b == null || !b.cj) {
            return NaN;
        }
        if (b != this.xAxis && b != this.yAxis) {
            return NaN;
        }
        var c = this.ap(a);
        if (b.cp) {
            return c[1];
        }
        else {
            return c[0];
        }
    };
    ScatterBase.prototype.ap = function (a) {
        var b = NaN;
        var c = NaN;
        switch (a) {
            case 5:
            case 2:
                b = this.y9.maximum;
                c = this.za.maximum;
                break;
            case 4:
            case 1:
                b = this.y9.minimum;
                c = this.za.minimum;
                break;
            case 6:
            case 3:
                if (this.zb) {
                    this.abj();
                }
                b = this.zs;
                c = this.zt;
                break;
        }
        return [b, c];
    };
    ScatterBase.prototype.wa = function (a) {
        var b = this.getEffectiveViewport1(this.ct);
        var c = new ScalerParams(0, this.ct.by, this.ct.bx, this.xAxis.ch, b);
        c.c = b;
        var d = isNaN_(a[0]) ? NaN : this.xAxis.ek(a[0], c);
        c = new ScalerParams(0, this.ct.by, this.ct.bx, this.yAxis.ch, b);
        c.b = this.iw();
        c.c = b;
        var e = isNaN_(a[1]) ? NaN : this.yAxis.ek(a[1], c);
        return { $type: Point_$type, x: d, y: e };
    };
    ScatterBase.prototype.v9 = function (a) {
        var b = this.ap(a);
        return this.wa(b);
    };
    ScatterBase.$t = markType(ScatterBase, 'ScatterBase', MarkerSeries.$, [ISupportsErrorBars_$type, ITrendLineSeries_$type]);
    ScatterBase.abk = DependencyProperty.i("ActualTrendLineBrush", Brush.$, ScatterBase.$, new PropertyMetadata(2, null, function (a, b) { return a.raisePropertyChanged("ActualTrendLineBrush", b.oldValue, b.newValue); }));
    ScatterBase.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterBase, 'raisePropertyChanged', ['ErrorBarSettings:yq:abl', [ScatterErrorBarSettings.$, null], 'HighlightedXMemberPath:z5:abm', [2, null], 'HighlightedYMemberPath:z7:abn', [2, null], 'IsCustomScatterMarkerStyleAllowed:zg:abo', [0, false], 'IsCustomScatterStyleAllowed:zh:abp', [0, false], 'ItemSearchMode:y0:abq', [ScatterItemSearchMode_$type, enumGetBox(ScatterItemSearchMode_$type, 0)], 'ItemSearchThreshold:z0:abr', [1, 10000], 'MarkerCollisionAvoidance:yl:abs', [CollisionAvoidanceType_$type, enumGetBox(CollisionAvoidanceType_$type, 0)], 'MaximumMarkers:z1:abt', [1, 400], 'XAxis::ab1', [NumericXAxis.$, null], 'XMemberAsLegendLabel:aah:ab2', [2, null], 'XMemberAsLegendUnit:aaj:ab3', [2, null], 'XMemberPath:aal:ab4', [2, null], 'YAxis::ab5', [NumericYAxis.$, null], 'YMemberAsLegendLabel:aap:ab6', [2, null], 'YMemberAsLegendUnit:aar:ab7', [2, null], 'YMemberPath:aat:ab8', [2, null], 'TrendLineType::abz', [TrendLineType_$type, enumGetBox(TrendLineType_$type, 0)], 'TrendLineThickness::aby', [1, 1.5], 'TrendLinePeriod::abx', [1, 7], 'TrendLineBrush::abu', [Brush.$, null], 'TrendLineZIndex:z2:ab0', [1, 1001], 'TrendLineDashArray:aca:abv', [DoubleCollection.$], 'TrendLineDashCap:acb:abw', [PenLineCap_$type, enumGetBox(PenLineCap_$type, 0)]]);
    return ScatterBase;
}(MarkerSeries));
export { ScatterBase };
/**
 * @hidden
 */
var ScatterSeriesRenderManager = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterSeriesRenderManager, _super);
    function ScatterSeriesRenderManager() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.aw = null;
        _this.ax = null;
        return _this;
    }
    ScatterSeriesRenderManager.prototype.c = function () {
        this.aw = new AssigningScatterMarkerStyleEventArgs();
        return this.aw;
    };
    ScatterSeriesRenderManager.prototype.a = function () {
        this.ax = new AssigningScatterStyleEventArgs();
        return this.ax;
    };
    ScatterSeriesRenderManager.prototype.ay = function (a, b, c) {
        this.w(a, b, c);
    };
    ScatterSeriesRenderManager.prototype.az = function (a, b, c) {
        this.x(a, b, c);
    };
    ScatterSeriesRenderManager.prototype.a0 = function (a, b, c, d) {
        if (b == -1) {
            a.e = false;
            a.z = 0;
            a.y = c - 1;
        }
        else {
            if (d) {
                a.z = b;
                a.y = b;
            }
            else {
                a.e = false;
                a.z = b;
                a.y = b;
            }
        }
    };
    ScatterSeriesRenderManager.$t = markType(ScatterSeriesRenderManager, 'ScatterSeriesRenderManager', SeriesRenderManager.$);
    return ScatterSeriesRenderManager;
}(SeriesRenderManager));
export { ScatterSeriesRenderManager };
/**
 * @hidden
 */
var ScatterBaseView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterBaseView, _super);
    function ScatterBaseView(a) {
        var _this = _super.call(this, a) || this;
        _this.c0 = null;
        _this.cy = null;
        _this.c1 = null;
        _this.c6 = null;
        _this.dh = null;
        _this.di = null;
        _this.c2 = null;
        _this.dg = null;
        _this.c3 = null;
        _this.c1 = a;
        _this.c3 = new HashPool$2(Base.$, Marker.$);
        _this.cm(_this.c3);
        _this.c2 = new ScatterTrendLineManager();
        return _this;
    }
    ScatterBaseView.prototype.au = function () {
        _super.prototype.au.call(this);
        this.c0 = this.cz();
        this.dh = new Path();
        this.di = new Path();
        if (!this.q) {
            this.c1.z1 = 2000;
        }
    };
    ScatterBaseView.prototype.dj = function (a) {
        if (this.dg == null) {
            this.dg = new DataTemplateMeasureInfo();
            this.dg.context = this.k.h();
        }
        this.dg.width = a.width;
        this.dg.height = a.height;
        this.dg.data = a.content;
        var b = a.ah;
        if (b.measure != null) {
            b.measure(this.dg);
        }
        return new Size(1, this.dg.width, this.dg.height);
    };
    ScatterBaseView.prototype.cz = function () {
        var _this = this;
        var a = new NumericMarkerManager(1, function (b) { return _this.c3.item(b); }, function (b) { return _this.c1.yo.c.item(b); }, runOn(this, this.de), runOn(this, this.cw), runOn(this, this.c4), function () { return _this.c1.yl; });
        a.q = runOn(this, this.dj);
        return a;
    };
    ScatterBaseView.prototype.de = function (a) {
        this.c1.abd(a, this.c3);
    };
    ScatterBaseView.prototype.c5 = function () {
        return this.b8;
    };
    ScatterBaseView.prototype.cx = function () {
        return this.cy;
    };
    ScatterBaseView.prototype.cw = function () {
        this.cy = this.c1.yg(this, this.c3, this.cy, this.by, this.bx);
        return this.cy;
    };
    ScatterBaseView.prototype.c4 = function () {
        this.c6 = this.c1.zp(this.c3, this.c6);
        return this.c6;
    };
    ScatterBaseView.prototype.c7 = function () {
    };
    ScatterBaseView.prototype.dc = function (a) {
    };
    ScatterBaseView.prototype.c8 = function () {
    };
    ScatterBaseView.prototype.dd = function (a) {
    };
    ScatterBaseView.prototype.df = function () {
        if (this.c1.trendLineBrush != null) {
            this.c1.ab9 = this.c1.trendLineBrush;
        }
        else {
            this.c1.ab9 = this.c1.u3;
        }
    };
    ScatterBaseView.prototype.cj = function (a) {
        this.c3.f(a);
    };
    ScatterBaseView.prototype.da = function (a) {
        this.c3.g(a);
    };
    ScatterBaseView.prototype.db = function () {
    };
    ScatterBaseView.prototype.a7 = function (a, b) {
        _super.prototype.a7.call(this, a, b);
        if (a.d) {
            if (this.c1.trendLineType != 0 && !b) {
                var c = this.c2.g$j;
                c.ad = this.c1.trendLineThickness;
                c._stroke = this.c1.ab9;
                c.ai = this.c1.aca;
                c.aj = this.c1.acb;
                a.v(c);
            }
        }
    };
    ScatterBaseView.prototype.c9 = function (a) {
        if (a) {
            this.db();
            this.c3.clear();
        }
        this.c2.g$d();
        this.as();
    };
    ScatterBaseView.prototype.getDefaultTooltipTemplate = function () {
        var a = "<div class='ui-chart-default-tooltip-content'><span";
        var b = this.e.a9.a$u(this.e);
        if (!stringIsNullOrEmpty(b)) {
            a += " style='color:" + b + "'";
        }
        a += ">" + this.c1.title + "</span><br/><span>" + "x: </span><span class='ui-priority-primary'>${item." + this.c1.aal + "}</span><br/><span>" + "y: </span><span class='ui-priority-primary'>${item." + this.c1.aat + "}</span></div>";
        return a;
    };
    ScatterBaseView.$t = markType(ScatterBaseView, 'ScatterBaseView', MarkerSeriesView.$);
    return ScatterBaseView;
}(MarkerSeriesView));
export { ScatterBaseView };
