// igniteui-angular-charts/esm5/lib/StackedLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { HorizontalStackedSeriesBase } from "./HorizontalStackedSeriesBase";
import { markType } from "igniteui-angular-core";
import { AxisRange } from "./AxisRange";
import { StackedSeriesView } from "./StackedSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
var StackedLineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StackedLineSeries, _super);
    function StackedLineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = StackedLineSeries.$;
        return _this;
    }
    StackedLineSeries.prototype.bi = function () {
        return new StackedLineSeries();
    };
    StackedLineSeries.prototype.get_fi = function () {
        return true;
    };
    StackedLineSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rn(false);
    };
    StackedLineSeries.prototype.cq = function () {
        return new StackedLineSeriesView(this);
    };
    StackedLineSeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.abk = a;
    };
    StackedLineSeries.prototype.ax = function (a) {
        if (this.itemsSource == null) {
            return null;
        }
        if (a == this.xAxis) {
            return new AxisRange(0, this.dl.count - 1);
        }
        if (a == this.yAxis) {
            this.aa3();
            return new AxisRange(this.aau, this.aat);
        }
        return null;
    };
    StackedLineSeries.prototype.preferredCategoryMode = function (a) {
        return 0;
    };
    Object.defineProperty(StackedLineSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zh;
        },
        enumerable: false,
        configurable: true
    });
    StackedLineSeries.$t = markType(StackedLineSeries, 'StackedLineSeries', HorizontalStackedSeriesBase.$);
    return StackedLineSeries;
}(HorizontalStackedSeriesBase));
export { StackedLineSeries };
/**
 * @hidden
 */
var StackedLineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StackedLineSeriesView, _super);
    function StackedLineSeriesView(a) {
        return _super.call(this, a) || this;
    }
    StackedLineSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c5.w2 = Defaults.e;
        }
    };
    StackedLineSeriesView.$t = markType(StackedLineSeriesView, 'StackedLineSeriesView', StackedSeriesView.$);
    return StackedLineSeriesView;
}(StackedSeriesView));
export { StackedLineSeriesView };
