// igniteui-angular-charts/esm5/lib/Stacked100LineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StackedLineSeries } from "./StackedLineSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-angular-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedLineSeriesView } from "./StackedLineSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
var Stacked100LineSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100LineSeries, _super);
    function Stacked100LineSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100LineSeries.$;
        return _this;
    }
    Stacked100LineSeries.prototype.cq = function () {
        return new Stacked100LineSeriesView(this);
    };
    Stacked100LineSeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.abl = a;
    };
    Stacked100LineSeries.prototype.get_aan = function () {
        return true;
    };
    Stacked100LineSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rn(false);
    };
    Stacked100LineSeries.prototype.aa3 = function () {
        _super.prototype.aa3.call(this);
        Stacked100DataPreparer.a(this);
    };
    Object.defineProperty(Stacked100LineSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zh;
        },
        enumerable: false,
        configurable: true
    });
    Stacked100LineSeries.$t = markType(Stacked100LineSeries, 'Stacked100LineSeries', StackedLineSeries.$, [IStacked100Series_$type]);
    return Stacked100LineSeries;
}(StackedLineSeries));
export { Stacked100LineSeries };
/**
 * @hidden
 */
var Stacked100LineSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100LineSeriesView, _super);
    function Stacked100LineSeriesView(a) {
        return _super.call(this, a) || this;
    }
    Stacked100LineSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c5.w2 = Defaults.e;
        }
    };
    Stacked100LineSeriesView.$t = markType(Stacked100LineSeriesView, 'Stacked100LineSeriesView', StackedLineSeriesView.$);
    return Stacked100LineSeriesView;
}(StackedLineSeriesView));
export { Stacked100LineSeriesView };
