// igniteui-angular-charts/esm5/lib/Stacked100SplineAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StackedSplineAreaSeries } from "./StackedSplineAreaSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-angular-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedSplineAreaSeriesView } from "./StackedSplineAreaSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
var Stacked100SplineAreaSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100SplineAreaSeries, _super);
    function Stacked100SplineAreaSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100SplineAreaSeries.$;
        return _this;
    }
    Stacked100SplineAreaSeries.prototype.cq = function () {
        return new Stacked100SplineAreaSeriesView(this);
    };
    Stacked100SplineAreaSeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.abo = a;
    };
    Stacked100SplineAreaSeries.prototype.get_aan = function () {
        return true;
    };
    Stacked100SplineAreaSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rn(false);
    };
    Stacked100SplineAreaSeries.prototype.aa3 = function () {
        _super.prototype.aa3.call(this);
        Stacked100DataPreparer.a(this);
    };
    Object.defineProperty(Stacked100SplineAreaSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zh;
        },
        enumerable: false,
        configurable: true
    });
    Stacked100SplineAreaSeries.$t = markType(Stacked100SplineAreaSeries, 'Stacked100SplineAreaSeries', StackedSplineAreaSeries.$, [IStacked100Series_$type]);
    return Stacked100SplineAreaSeries;
}(StackedSplineAreaSeries));
export { Stacked100SplineAreaSeries };
/**
 * @hidden
 */
var Stacked100SplineAreaSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100SplineAreaSeriesView, _super);
    function Stacked100SplineAreaSeriesView(a) {
        return _super.call(this, a) || this;
    }
    Stacked100SplineAreaSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c5.w2 = Defaults.e;
        }
    };
    Stacked100SplineAreaSeriesView.$t = markType(Stacked100SplineAreaSeriesView, 'Stacked100SplineAreaSeriesView', StackedSplineAreaSeriesView.$);
    return Stacked100SplineAreaSeriesView;
}(StackedSplineAreaSeriesView));
export { Stacked100SplineAreaSeriesView };
