// igniteui-angular-charts/esm5/lib/Stacked100ColumnSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StackedColumnSeries } from "./StackedColumnSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { Defaults } from "./Defaults";
import { markType } from "igniteui-angular-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { isNaN_ } from "igniteui-angular-core";
import { StackedColumnSeriesView } from "./StackedColumnSeriesView";
import { Stacked100ColumnBucketCalculator } from "./Stacked100ColumnBucketCalculator";
/**
 * @hidden
 */
var Stacked100ColumnSeries = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100ColumnSeries, _super);
    function Stacked100ColumnSeries() {
        var _this = _super.call(this) || this;
        _this.ab = Stacked100ColumnSeries.$;
        return _this;
    }
    Stacked100ColumnSeries.prototype.get_aan = function () {
        return true;
    };
    Stacked100ColumnSeries.prototype.ji = function () {
        if (!isNaN_(this.jh)) {
            return _super.prototype.ji.call(this);
        }
        return Defaults.h;
    };
    Stacked100ColumnSeries.prototype.ac = function () {
        _super.prototype.ac.call(this);
        this.rn(false);
    };
    Stacked100ColumnSeries.prototype.cq = function () {
        return new Stacked100ColumnSeriesView(this);
    };
    Stacked100ColumnSeries.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.abr = a;
    };
    Stacked100ColumnSeries.prototype.aae = function () {
        return this.abr;
    };
    Stacked100ColumnSeries.prototype.aa3 = function () {
        _super.prototype.aa3.call(this);
        Stacked100DataPreparer.a(this);
    };
    Object.defineProperty(Stacked100ColumnSeries.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.zh;
        },
        enumerable: false,
        configurable: true
    });
    Stacked100ColumnSeries.$t = markType(Stacked100ColumnSeries, 'Stacked100ColumnSeries', StackedColumnSeries.$, [IStacked100Series_$type]);
    return Stacked100ColumnSeries;
}(StackedColumnSeries));
export { Stacked100ColumnSeries };
/**
 * @hidden
 */
var Stacked100ColumnSeriesView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Stacked100ColumnSeriesView, _super);
    function Stacked100ColumnSeriesView(a) {
        var _this = _super.call(this, a) || this;
        _this.da = null;
        _this.da = a;
        return _this;
    }
    Stacked100ColumnSeriesView.prototype.cx = function () {
        return new Stacked100ColumnBucketCalculator(this);
    };
    Stacked100ColumnSeriesView.prototype.au = function () {
        _super.prototype.au.call(this);
        if (!this.q) {
            this.c5.w2 = Defaults.c;
            this.c5.jn = Defaults.i;
        }
    };
    Stacked100ColumnSeriesView.$t = markType(Stacked100ColumnSeriesView, 'Stacked100ColumnSeriesView', StackedColumnSeriesView.$);
    return Stacked100ColumnSeriesView;
}(StackedColumnSeriesView));
export { Stacked100ColumnSeriesView };
