// igniteui-angular-charts/esm5/lib/StandardDeviationIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { StandardDeviationIndicatorStrategy } from "./StandardDeviationIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var StandardDeviationIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StandardDeviationIndicator, _super);
    function StandardDeviationIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StandardDeviationIndicator.prototype.bi = function () {
        return new StandardDeviationIndicator();
    };
    StandardDeviationIndicator.prototype.get_abn = function () {
        return new StandardDeviationIndicatorStrategy();
    };
    Object.defineProperty(StandardDeviationIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    StandardDeviationIndicator.prototype.get_ab0 = function () {
        return StandardDeviationIndicator.$;
    };
    Object.defineProperty(StandardDeviationIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StandardDeviationIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(StandardDeviationIndicator.ab9));
        },
        set: function (a) {
            this.h(StandardDeviationIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    StandardDeviationIndicator.prototype.abt = function () {
        return this.ab8;
    };
    StandardDeviationIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    StandardDeviationIndicator.$t = markType(StandardDeviationIndicator, 'StandardDeviationIndicator', StrategyBasedIndicator.$);
    StandardDeviationIndicator.ab9 = StrategyBasedIndicator.ab5(20, StandardDeviationIndicator.$);
    return StandardDeviationIndicator;
}(StrategyBasedIndicator));
export { StandardDeviationIndicator };
