// igniteui-angular-charts/esm5/lib/StochRSIIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { StochRSIIndicatorStrategy } from "./StochRSIIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var StochRSIIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StochRSIIndicator, _super);
    function StochRSIIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StochRSIIndicator.prototype.bi = function () {
        return new StochRSIIndicator();
    };
    StochRSIIndicator.prototype.get_abn = function () {
        return new StochRSIIndicatorStrategy();
    };
    Object.defineProperty(StochRSIIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    StochRSIIndicator.prototype.get_ab0 = function () {
        return StochRSIIndicator.$;
    };
    Object.defineProperty(StochRSIIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StochRSIIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(StochRSIIndicator.ab9));
        },
        set: function (a) {
            this.h(StochRSIIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    StochRSIIndicator.prototype.abt = function () {
        return this.ab8;
    };
    StochRSIIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    StochRSIIndicator.$t = markType(StochRSIIndicator, 'StochRSIIndicator', StrategyBasedIndicator.$);
    StochRSIIndicator.ab9 = StrategyBasedIndicator.ab5(14, StochRSIIndicator.$);
    return StochRSIIndicator;
}(StrategyBasedIndicator));
export { StochRSIIndicator };
