// igniteui-angular-charts/esm5/lib/StrategyBasedIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { FinancialIndicator } from "./FinancialIndicator";
import { String_$type, Number_$type, typeCast, fromEnum, markType } from "igniteui-angular-core";
import { DependencyProperty } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { AxisRange } from "./AxisRange";
import { isNaN_ } from "igniteui-angular-core";
/**
 * @hidden
 */
var StrategyBasedIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StrategyBasedIndicator, _super);
    function StrategyBasedIndicator() {
        var _this = _super.call(this) || this;
        _this.ab1();
        return _this;
    }
    StrategyBasedIndicator.prototype.ab1 = function () {
        this.abm = this.abn;
        this.ab = this.ab0;
    };
    StrategyBasedIndicator.prototype.abt = function () {
        return -2147483648;
    };
    StrategyBasedIndicator.prototype.ab2 = function (a) {
    };
    StrategyBasedIndicator.prototype.abv = function () {
        return -2147483648;
    };
    StrategyBasedIndicator.prototype.abs = function () {
        return -2147483648;
    };
    StrategyBasedIndicator.ab6 = function (a, b, c) {
        var d = DependencyProperty.i(c, Number_$type, b, new PropertyMetadata(2, a, function (e, f) { return (typeCast(StrategyBasedIndicator.$, e)).raisePropertyChanged(c, f.oldValue, f.newValue); }));
        StrategyBasedIndicator.abq.add(c);
        return d;
    };
    StrategyBasedIndicator.ab5 = function (a, b) {
        return StrategyBasedIndicator.ab6(a, b, "Period");
    };
    StrategyBasedIndicator.ab3 = function (a, b) {
        return StrategyBasedIndicator.ab6(a, b, "LongPeriod");
    };
    StrategyBasedIndicator.ab7 = function (a, b) {
        return StrategyBasedIndicator.ab6(a, b, "ShortPeriod");
    };
    StrategyBasedIndicator.prototype.aaz = function (a, b) {
        var c = this.xb(a, b);
        var d = this.xc(c);
        return this.abm.b(c, d);
    };
    StrategyBasedIndicator.prototype.aaw = function (a, b) {
        var c = this.xb(a, b);
        if (b == 0) {
            return false;
        }
        if (!this.x7(this.aaz(a, b))) {
            return false;
        }
        var d = this.xc(c);
        if (this.aas != null) {
            c.minimumValue = this.aas.c;
            c.maximumValue = this.aas.b;
        }
        var e = this.abm.a(c, d);
        for (var f = 0; f < this.aa1 && f < c.indicatorColumn.count; f++) {
            c.indicatorColumn.item(f, NaN);
        }
        if (this.xh != null && this.abp(c)) {
            this.xh.c1();
        }
        return e;
    };
    StrategyBasedIndicator.prototype.abp = function (a) {
        var e_1, _a;
        if (!isNaN_(a.minimumValue) && !isNaN_(a.maximumValue) && a.specifiesRange) {
            var b = this.aas;
            this.aas = new AxisRange(a.minimumValue, a.maximumValue);
            return this.abo(b, this.aas);
        }
        var c = 1.7976931348623157E+308;
        var d = -1.7976931348623157E+308;
        try {
            for (var _b = __values(fromEnum(a.indicatorColumn)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var e = _c.value;
                if (!isNaN_(e)) {
                    c = Math.min(c, e);
                    d = Math.max(d, e);
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        var f = this.aas;
        this.aas = new AxisRange(c, d);
        return this.abo(f, this.aas);
    };
    StrategyBasedIndicator.prototype.abo = function (a, b) {
        if (a == null || b == null) {
            return true;
        }
        if (a.c != b.c) {
            return true;
        }
        if (a.b != b.b) {
            return true;
        }
        return false;
    };
    StrategyBasedIndicator.prototype.abu = function (a) {
        if (a == -2147483648) {
            return 0;
        }
        if (a > this.indicatorColumn.count && this.indicatorColumn.count > 0) {
            a = this.indicatorColumn.count - 1;
        }
        if (a < 1) {
            return 1;
        }
        return a;
    };
    StrategyBasedIndicator.prototype.xb = function (a, b) {
        var c = _super.prototype.xb.call(this, a, b);
        c.indicatorColumn = this.indicatorColumn;
        c.period = this.abu(this.abt());
        c.shortPeriod = this.abu(this.abv());
        c.longPeriod = this.abu(this.abs());
        c.multiplier = this.abr();
        return c;
    };
    StrategyBasedIndicator.prototype.abr = function () {
        return 1;
    };
    StrategyBasedIndicator.ab4 = function (a, b) {
        var c = DependencyProperty.i("Multiplier", Number_$type, b, new PropertyMetadata(2, a, function (d, e) { return (typeCast(StrategyBasedIndicator.$, d)).raisePropertyChanged("Multiplier", e.oldValue, e.newValue); }));
        StrategyBasedIndicator.abq.add("Multiplier");
        return c;
    };
    StrategyBasedIndicator.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
        if (StrategyBasedIndicator.abq.contains(b)) {
            if (this.xh != null && !this.xh.c1()) {
                this.xf.b0.e(this.ib);
                this.aaw(0, this.indicatorColumn.count);
                if (this.aat.ca) {
                    this.aat.b8.g$i();
                }
                this.rn(false);
            }
        }
    };
    StrategyBasedIndicator.$t = markType(StrategyBasedIndicator, 'StrategyBasedIndicator', FinancialIndicator.$);
    StrategyBasedIndicator.abq = new List$1(String_$type, 0);
    return StrategyBasedIndicator;
}(FinancialIndicator));
export { StrategyBasedIndicator };
