// igniteui-angular-charts/esm5/lib/ValueLayer_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { AnnotationLayer } from "./AnnotationLayer";
import { ValueLayerValueMode_$type } from "./ValueLayerValueMode";
import { DependencyProperty } from "igniteui-angular-core";
import { Brush } from "igniteui-angular-core";
import { Series } from "./Series";
import { Axis } from "./Axis";
import { Base, enumGetBox, Delegate_$type, Number_$type, fromEnum, Point_$type, markType, markDep, runOn } from "igniteui-angular-core";
import { Thickness } from "igniteui-angular-core";
import { Visibility_$type } from "igniteui-angular-core";
import { ValueLayerFrame } from "./ValueLayerFrame";
import { List$1 } from "igniteui-angular-core";
import { Dictionary$2 } from "igniteui-angular-core";
import { Pool$1 } from "igniteui-angular-core";
import { AxisAnnotation } from "./AxisAnnotation";
import { AxisAnnotationFrameManager } from "./AxisAnnotationFrameManager";
import { DoubleCollection } from "igniteui-angular-core";
import { DeviceUtils } from "igniteui-angular-core";
import { Tuple$2 } from "igniteui-angular-core";
import { Path } from "igniteui-angular-core";
import { PathGeometry } from "igniteui-angular-core";
import { PathFigure } from "igniteui-angular-core";
import { LineSegment } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { isNaN_, truncate } from "igniteui-angular-core";
import { AnnotationLayerView } from "./AnnotationLayerView";
import { Defaults } from "./Defaults";
import { TranslateTransform } from "igniteui-angular-core";
import { LineGeometry } from "igniteui-angular-core";
import { BrushUtil } from "igniteui-angular-core";
import { Color } from "igniteui-angular-core";
/**
 * @hidden
 */
var ValueLayer = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ValueLayer, _super);
    function ValueLayer() {
        var _this = _super.call(this) || this;
        _this.acq = null;
        _this.z1 = null;
        _this.z0 = null;
        var a = new ValueLayerFrame();
        var b = new ValueLayerFrame();
        var c = new ValueLayerFrame();
        var d = _this.kb / 1000;
        _this.w4 = a;
        _this.w3 = b;
        _this.w5 = c;
        _this.ab = ValueLayer.$;
        return _this;
    }
    ValueLayer.prototype.bi = function () {
        return new ValueLayer();
    };
    ValueLayer.prototype.get_er = function () {
        return true;
    };
    ValueLayer.prototype.cq = function () {
        return new ValueLayerView(this);
    };
    ValueLayer.prototype.qx = function (a) {
        _super.prototype.qx.call(this, a);
        this.zv = a;
    };
    ValueLayer.prototype.gs = function (a) {
        if (this.zs == null) {
            return true;
        }
        if (this.zs == a) {
            return true;
        }
        return false;
    };
    Object.defineProperty(ValueLayer.prototype, "acs", {
        get: function () {
            return new Thickness(1, this.z4, this.z6, this.z5, this.z3);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueLayer.prototype, "act", {
        get: function () {
            return new Thickness(1, this.aaa, this.aac, this.aab, this.z9);
        },
        enumerable: false,
        configurable: true
    });
    ValueLayer.prototype.yh = function (a, b) {
        if (this.zs == null || this.zs == a) {
            if (!this.a0.c) {
                this.rn(b);
            }
        }
    };
    ValueLayer.prototype.yj = function (a, b) {
        var e_1, _a, e_2, _b;
        var _this = this;
        _super.prototype.yj.call(this, a, b);
        var c = a;
        var d = this.zz;
        var e = this.zy;
        c.af.clear();
        c.ag.clear();
        c.ae.clear();
        c.ah.clear();
        AxisAnnotationFrameManager.b(c);
        if (this.xa(this.zs)) {
            this.abh(this.zs, c, d, e);
        }
        else {
            var f = this.zt();
            if (f == 5 || f == 4 || f == 6) {
                var g = null;
                var h = null;
                var i = [NaN, NaN];
                var j = 0;
                try {
                    for (var _c = __values(fromEnum(this.c9.series)), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var k = _d.value;
                        if (!this.xa(k)) {
                            continue;
                        }
                        var l = k.ai();
                        for (var n = 0; n < l.length; n++) {
                            var m = l[n];
                            if (!this.zx(m)) {
                                continue;
                            }
                            var o = m.cp ? 1 : 0;
                            var p = i[o];
                            var q = null;
                            if (f == 5) {
                                var r = k.i6(2, m);
                                if (!isNaN_(r) && (r > p || isNaN_(p))) {
                                    q = k;
                                }
                                p = !isNaN_(p) ? Math.max(r, p) : r;
                            }
                            else if (f == 4) {
                                var s = k.i6(1, m);
                                if (!isNaN_(s) && (s < p || isNaN_(p))) {
                                    q = k;
                                }
                                p = !isNaN_(p) ? Math.min(s, p) : s;
                            }
                            else if (f == 6) {
                                var t = k.i6(3, m);
                                q = k;
                                p = !isNaN_(p) ? p + t : t;
                            }
                            if (m.cp) {
                                h = q != null ? q : h;
                            }
                            else {
                                g = q != null ? q : g;
                            }
                            i[o] = p;
                        }
                        j++;
                    }
                }
                catch (e_1_1) {
                    e_1 = { error: e_1_1 };
                }
                finally {
                    try {
                        if (_d && !_d.done && (_a = _c.return))
                            _a.call(_c);
                    }
                    finally {
                        if (e_1)
                            throw e_1.error;
                    }
                }
                if (f == 6) {
                    i[0] = !isNaN_(i[0]) ? i[0] / j : NaN;
                    i[1] = !isNaN_(i[1]) ? i[1] / j : NaN;
                }
                if (g != null && !isNaN_(i[0])) {
                    var u = g.wa(i);
                    this.abi(g, { $type: Point_$type, x: u.x, y: NaN }, c, d, e);
                }
                if (h != null && !isNaN_(i[1])) {
                    var v = h.wa(i);
                    this.abi(h, { $type: Point_$type, x: NaN, y: v.y }, c, d, e);
                }
            }
            else {
                try {
                    for (var _e = __values(fromEnum(this.c9.series)), _f = _e.next(); !_f.done; _f = _e.next()) {
                        var w = _f.value;
                        if (this.xa(w)) {
                            this.abh(w, c, d, e);
                        }
                        if (w.isStacked) {
                            w.pz(function (x) {
                                if (_this.xa(x)) {
                                    _this.abh(x, c, d, e);
                                }
                                if (x.ee) {
                                    x.px(function (y) {
                                        if (_this.xa(y)) {
                                            _this.abh(y, c, d, e);
                                        }
                                        return true;
                                    });
                                }
                                return true;
                            });
                        }
                        if (w.ee) {
                            w.px(function (x) {
                                if (_this.xa(x)) {
                                    _this.abh(x, c, d, e);
                                }
                                return true;
                            });
                        }
                    }
                }
                catch (e_2_1) {
                    e_2 = { error: e_2_1 };
                }
                finally {
                    try {
                        if (_f && !_f.done && (_b = _e.return))
                            _b.call(_e);
                    }
                    finally {
                        if (e_2)
                            throw e_2.error;
                    }
                }
            }
        }
    };
    ValueLayer.prototype.xa = function (a) {
        if (!_super.prototype.xa.call(this, a)) {
            return false;
        }
        if (this.aan != null) {
            if (a.e9) {
                var b = a.na();
                if (!Base.equalsStatic(this.aan, b)) {
                    return false;
                }
            }
            else {
                if (!Base.equalsStatic(this.aan, a.name)) {
                    return false;
                }
            }
        }
        return true;
    };
    ValueLayer.prototype.zx = function (a) {
        if (!a.cj) {
            return false;
        }
        if (this.zp != null && this.zp != a) {
            return false;
        }
        if (this.aak != null && !Base.equalsStatic(this.aak, a.name)) {
            return false;
        }
        return true;
    };
    ValueLayer.prototype.ql = function () {
        _super.prototype.ql.call(this);
        if (!this.a0.c && !this.d1()) {
            return;
        }
        if (this.d1()) {
            this.abj(this.w5, false);
        }
        else {
            this.abj(this.w3, false);
        }
    };
    ValueLayer.prototype.zt = function () {
        if (this.zu == 0) {
            return 2;
        }
        return this.zu;
    };
    ValueLayer.prototype.wu = function () {
        return 3;
    };
    ValueLayer.prototype.xh = function () {
        return 0.3;
    };
    ValueLayer.prototype.xk = function () {
        return -0.4;
    };
    ValueLayer.prototype.xj = function () {
        return 0.5;
    };
    ValueLayer.prototype.zg = function () {
        return this.wr == 3 ? this.acr() : null;
    };
    ValueLayer.prototype.zh = function () {
        return this.ws == 3 ? this.acr() : null;
    };
    ValueLayer.prototype.acr = function () {
        if (this.acq == null) {
            this.acq = new DoubleCollection();
            this.acq.add(DeviceUtils.g(5));
            this.acq.add(DeviceUtils.g(5));
        }
        return this.acq;
    };
    ValueLayer.prototype.abi = function (a, b, c, d, e) {
        var f = false;
        var g = this.u3;
        if (g == null) {
            g = this.u9;
        }
        if (g == null) {
            g = a.u3;
            f = true;
        }
        var h = false;
        var i = this.ack;
        if (i == null) {
            if (this.u9 != null) {
                i = this.u9;
            }
            else {
                i = a.u3;
                h = true;
            }
        }
        var j = false;
        var k = this.acn;
        if (k == null) {
            if (this.u9 != null) {
                k = this.u9;
            }
            else {
                k = a.u3;
                j = true;
            }
        }
        var l = g;
        var m = g;
        var n = i;
        var o = k;
        if (this.aci != null) {
            l = this.aci;
        }
        else if (f) {
            switch (this.wr) {
                case 1:
                    l = this.zv.ce(l, this.xe);
                    break;
                case 2:
                    l = this.zv.cg(l, this.xe);
                    break;
                case 4:
                    l = this.zv.cf(l, this.xe);
                    break;
            }
        }
        if (this.acj != null) {
            m = this.acj;
        }
        else if (f) {
            switch (this.ws) {
                case 1:
                    m = this.zv.ce(m, this.xg);
                    break;
                case 2:
                    m = this.zv.cg(m, this.xg);
                    break;
                case 4:
                    m = this.zv.cf(m, this.xg);
                    break;
            }
        }
        if (h) {
            switch (this.ws) {
                case 1:
                    n = this.zv.ce(n, this.xg);
                    break;
                case 2:
                    n = this.zv.cg(n, this.xg);
                    break;
            }
        }
        if (j) {
            switch (this.wr) {
                case 1:
                    o = this.zv.ce(o, this.xe);
                    break;
                case 2:
                    o = this.zv.cg(o, this.xe);
                    break;
            }
        }
        c.ae.add(l);
        c.ah.add(m);
        c.af.add(b.y);
        c.ag.add(b.x);
        if (this.zw) {
            AxisAnnotationFrameManager.c(c, b, a, this.c9, this.getEffectiveViewport(), g, n, this.z2, this.acm, this.acl, this.acs, this.z7, o, this.z8, this.acp, this.aco, this.act, this.aad, this.acv == 0, this.acu == 0);
        }
    };
    ValueLayer.prototype.abh = function (a, b, c, d) {
        var e = { $type: Point_$type, x: NaN, y: NaN };
        var f = this.zt();
        switch (f) {
            case 5:
                e = a.v9(2);
                break;
            case 4:
                e = a.v9(1);
                break;
            case 6:
                e = a.v9(3);
                break;
            default:
                e = a.v9(f);
                break;
        }
        var i = a.ai();
        for (var h = 0; h < i.length; h++) {
            var g = i[h];
            if (!this.zx(g)) {
                continue;
            }
            if (g.cp) {
                this.abi(a, { $type: Point_$type, x: NaN, y: e.y }, b, c, d);
            }
            else {
                this.abi(a, { $type: Point_$type, x: e.x, y: NaN }, b, c, d);
            }
        }
    };
    ValueLayer.prototype.abj = function (a, b) {
        var c = false;
        if (this.z0 == null) {
            c = true;
            this.z1 = new List$1(Axis.$, 0);
            this.z0 = new Dictionary$2(Axis.$, Pool$1.$.specialize(AxisAnnotation.$), 0);
        }
        AxisAnnotationFrameManager.d(this, a, c, this.z0, this.z1, b, this.aae, this.aaf, this.zq, this.zr);
    };
    ValueLayer.prototype.abg = function (a, b, c, d, e) {
        var f = a.item(b);
        b++;
        if (e) {
            f._stroke = d.ah._inner[c];
            f._fill = d.ah._inner[c];
        }
        else {
            f._stroke = d.ae._inner[c];
            f._fill = d.ae._inner[c];
        }
        f.ad = this.ic;
        f.ai = e ? this.zf : this.ze;
        f.am = this.vz;
        f.al = this.vy;
        f.ak = this.vy;
        f.aj = this.vx;
        f.ac = this.jf;
        return new Tuple$2(Path.$, Number_$type, f, b);
    };
    ValueLayer.prototype.ym = function (a, b) {
        _super.prototype.ym.call(this, a, b);
        var c = this.zv.b8;
        var d = a;
        var e = d.af.count;
        var f = 0;
        var g = b.bx.left;
        var h = b.bx.right;
        var i = b.bx.top;
        var j = b.bx.bottom;
        var k = this.acv == 0;
        var l = this.acu == 0;
        for (var m = 0; m < e; m++) {
            var n = Math.floor(d.af._inner[m]);
            var o = Math.floor(d.ag._inner[m]);
            var p = null;
            if (k) {
                if (!isNaN_(o)) {
                    var q = this.abg(c, f, m, d, true);
                    p = q.c;
                    f = q.d;
                    var r = new PathGeometry();
                    var s = new PathFigure();
                    s._startPoint = { $type: Point_$type, x: o, y: i };
                    var t = new LineSegment(1);
                    t.c = { $type: Point_$type, x: o, y: j };
                    s._segments.add(t);
                    r.c.add(s);
                    p.an = r;
                }
            }
            if (l) {
                if (!isNaN_(n)) {
                    var u = this.abg(c, f, m, d, false);
                    p = u.c;
                    f = u.d;
                    var v = new PathGeometry();
                    var w = new PathFigure();
                    w._startPoint = { $type: Point_$type, x: g, y: n };
                    var x = new LineSegment(1);
                    x.c = { $type: Point_$type, x: h, y: n };
                    w._segments.add(x);
                    v.c.add(w);
                    p.an = v;
                }
            }
        }
        c.count = f;
        if (this.d1()) {
            this.abj(this.w5, true);
        }
        else {
            this.abj(this.w3, false);
        }
    };
    ValueLayer.prototype.f8 = function () {
        return true;
    };
    ValueLayer.prototype.q5 = function (a, b, c, d) {
        _super.prototype.q5.call(this, a, b, c, d);
        switch (b) {
            case "SeriesViewer":
                if (d == null) {
                    if (this.z0 != null && this.z0.count > 0) {
                        var e = void 0;
                        if (this.d1()) {
                            e = this.w5;
                        }
                        else {
                            e = this.w3;
                        }
                        AxisAnnotationFrameManager.b(e);
                        AxisAnnotationFrameManager.d(this, e, true, this.z0, this.z1, true, this.aae, this.aaf, this.zq, this.zr);
                    }
                }
                break;
            case ValueLayer.$$p[1]:
                this.rn(true);
                break;
            case ValueLayer.$$p[11]:
                this.rn(true);
                break;
            case ValueLayer.$$p[6]:
            case ValueLayer.$$p[7]:
            case "Visibility":
                this.rn(true);
                break;
            case ValueLayer.$$p[2]:
            case ValueLayer.$$p[13]:
            case ValueLayer.$$p[12]:
            case ValueLayer.$$p[16]:
            case ValueLayer.$$p[18]:
            case ValueLayer.$$p[20]:
            case ValueLayer.$$p[19]:
            case ValueLayer.$$p[17]:
            case ValueLayer.$$p[21]:
            case ValueLayer.$$p[22]:
            case ValueLayer.$$p[24]:
            case ValueLayer.$$p[27]:
            case ValueLayer.$$p[29]:
            case ValueLayer.$$p[31]:
            case ValueLayer.$$p[30]:
            case ValueLayer.$$p[28]:
            case ValueLayer.$$p[32]:
            case ValueLayer.$$p[33]:
            case ValueLayer.$$p[0]:
            case ValueLayer.$$p[10]:
            case ValueLayer.$$p[8]:
                this.rn(true);
                break;
        }
    };
    ValueLayer.prototype.o3 = function (a, b) {
        _super.prototype.o3.call(this, a, b);
        if (a) {
            AxisAnnotationFrameManager.b(this.w3);
            this.abj(this.w3, false);
        }
    };
    ValueLayer.$t = markType(ValueLayer, 'ValueLayer', AnnotationLayer.$);
    ValueLayer.$$p = markDep(DependencyProperty, PropertyMetadata, ValueLayer, 'raisePropertyChanged', ['HorizontalLineStroke:aci:abk', [Brush.$, null], 'HorizontalLineVisibility:acu:abl', [Visibility_$type, enumGetBox(Visibility_$type, 0)], 'IsAxisAnnotationEnabled:zw:abm', [0, false], 'SkipUnknownValues:zy:abn', [0, true], 'TargetAxisName:aak:abo', [2, null], 'TargetAxis:zp:abp', [Axis.$, null], 'TargetSeriesName:aan:abq', [2, null], 'TargetSeries:zs:abr', [Series.$, null], 'UseInterpolation:zz:abs', [0, true], 'ValueMode:zu:abt', [ValueLayerValueMode_$type, enumGetBox(ValueLayerValueMode_$type, 2)], 'VerticalLineStroke:acj:abu', [Brush.$, null], 'VerticalLineVisibility:acv:abv', [Visibility_$type, enumGetBox(Visibility_$type, 0)], 'XAxisAnnotationBackgroundCornerRadius:z2:abw', [1, DeviceUtils.g(3)], 'XAxisAnnotationBackground:ack:abx', [Brush.$, null], 'XAxisAnnotationFormatLabel:zq:aby', [Delegate_$type], 'XAxisAnnotationInterpolatedValuePrecision:aae:abz', [1, -1], 'XAxisAnnotationOutline:acl:ab0', [Brush.$, null], 'XAxisAnnotationPaddingBottom:z3:ab1', [1, NaN], 'XAxisAnnotationPaddingLeft:z4:ab2', [1, NaN], 'XAxisAnnotationPaddingRight:z5:ab3', [1, NaN], 'XAxisAnnotationPaddingTop:z6:ab4', [1, NaN], 'XAxisAnnotationStrokeThickness:z7:ab5', [1, NaN], 'XAxisAnnotationTextColor:acm:ab6', [Brush.$, null], 'YAxisAnnotationBackgroundCornerRadius:z8:ab7', [1, DeviceUtils.g(3)], 'YAxisAnnotationBackground:acn:ab8', [Brush.$, null], 'YAxisAnnotationFormatLabel:zr:ab9', [Delegate_$type], 'YAxisAnnotationInterpolatedValuePrecision:aaf:aca', [1, -1], 'YAxisAnnotationOutline:aco:acb', [Brush.$, null], 'YAxisAnnotationPaddingBottom:z9:acc', [1, NaN], 'YAxisAnnotationPaddingLeft:aaa:acd', [1, NaN], 'YAxisAnnotationPaddingRight:aab:ace', [1, NaN], 'YAxisAnnotationPaddingTop:aac:acf', [1, NaN], 'YAxisAnnotationStrokeThickness:aad:acg', [1, NaN], 'YAxisAnnotationTextColor:acp:ach', [Brush.$, null]]);
    return ValueLayer;
}(AnnotationLayer));
export { ValueLayer };
/**
 * @hidden
 */
var ValueLayerView = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ValueLayerView, _super);
    function ValueLayerView(a) {
        var _this = _super.call(this, a) || this;
        _this.b7 = null;
        _this.b8 = null;
        _this.b9 = null;
        _this.b7 = a;
        _this.b9 = new List$1(Path.$, 0);
        return _this;
    }
    ValueLayerView.prototype.au = function () {
        var _this = this;
        _super.prototype.au.call(this);
        this.b8 = ((function () {
            var $ret = new Pool$1(Path.$);
            $ret.create = runOn(_this, _this.ch);
            $ret.activate = runOn(_this, _this.ca);
            $ret.disactivate = runOn(_this, _this.cc);
            $ret.destroy = runOn(_this, _this.cb);
            return $ret;
        })());
        this.b7.jn = Defaults.j;
    };
    ValueLayerView.prototype.ch = function () {
        var a = new Path();
        a.j = new TranslateTransform();
        this.b9.add(a);
        return a;
    };
    ValueLayerView.prototype.ca = function (a) {
        a._visibility = 0;
    };
    ValueLayerView.prototype.cc = function (a) {
        a._visibility = 1;
    };
    ValueLayerView.prototype.cb = function (a) {
        this.b9.remove(a);
    };
    ValueLayerView.prototype.cd = function (a, b, c, d) {
        if (a.an == null || a.an.c.x != d || a.an.d.x != c) {
            var e = ((function () {
                var $ret = new LineGeometry();
                $ret.d = { $type: Point_$type, x: c, y: b };
                $ret.c = { $type: Point_$type, x: d, y: b };
                return $ret;
            })());
            a.an = e;
        }
    };
    ValueLayerView.prototype.ce = function (a, b) {
        if (a == null) {
            return a;
        }
        return BrushUtil.m(a, b);
    };
    ValueLayerView.prototype.cg = function (a, b) {
        if (a == null) {
            return a;
        }
        return BrushUtil.o(a, b);
    };
    ValueLayerView.prototype.cf = function (a, b) {
        if (a == null) {
            return a;
        }
        var c = new Brush();
        c.color = Color.u(truncate((a.color.l * b)), a.color.o, a.color.n, a.color.m);
        return c;
    };
    ValueLayerView.prototype.a8 = function (a, b) {
        _super.prototype.a8.call(this, a, b);
        if (a.d && !b) {
            for (var c = 0; c < this.b9.count; c++) {
                var d = this.b9._inner[c];
                if (d._visibility == 0) {
                    a.t(d);
                }
            }
        }
    };
    ValueLayerView.prototype.an = function (a) {
        _super.prototype.an.call(this, a);
        for (var b = 0; b < this.b9.count; b++) {
            this.e.bb.exportPathData(a, this.b9._inner[b], "crosshairLine", ["Main", "Crosshair"]);
        }
    };
    ValueLayerView.$t = markType(ValueLayerView, 'ValueLayerView', AnnotationLayerView.$);
    return ValueLayerView;
}(AnnotationLayerView));
export { ValueLayerView };
