// igniteui-angular-charts/esm5/lib/WilliamsPercentRIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { WilliamsPercentRIndicatorStrategy } from "./WilliamsPercentRIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
var WilliamsPercentRIndicator = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WilliamsPercentRIndicator, _super);
    function WilliamsPercentRIndicator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    WilliamsPercentRIndicator.prototype.bi = function () {
        return new WilliamsPercentRIndicator();
    };
    WilliamsPercentRIndicator.prototype.get_abn = function () {
        return new WilliamsPercentRIndicatorStrategy();
    };
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "abn", {
        get: function () {
            return this.get_abn();
        },
        enumerable: false,
        configurable: true
    });
    WilliamsPercentRIndicator.prototype.get_ab0 = function () {
        return WilliamsPercentRIndicator.$;
    };
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "ab0", {
        get: function () {
            return this.get_ab0();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WilliamsPercentRIndicator.prototype, "ab8", {
        get: function () {
            return typeGetValue(this.c(WilliamsPercentRIndicator.ab9));
        },
        set: function (a) {
            this.h(WilliamsPercentRIndicator.ab9, a);
        },
        enumerable: false,
        configurable: true
    });
    WilliamsPercentRIndicator.prototype.abt = function () {
        return this.ab8;
    };
    WilliamsPercentRIndicator.prototype.ab2 = function (a) {
        this.ab8 = a;
    };
    WilliamsPercentRIndicator.$t = markType(WilliamsPercentRIndicator, 'WilliamsPercentRIndicator', StrategyBasedIndicator.$);
    WilliamsPercentRIndicator.ab9 = StrategyBasedIndicator.ab5(14, WilliamsPercentRIndicator.$);
    return WilliamsPercentRIndicator;
}(StrategyBasedIndicator));
export { WilliamsPercentRIndicator };
