// igniteui-angular-charts/esm5/lib/XYChart.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { DomainChart } from "./DomainChart";
import { fromEnum, enumGetBox, markType } from "igniteui-angular-core";
import { HorizontalAlignment_$type } from "igniteui-angular-core";
import { VerticalAlignment_$type } from "igniteui-angular-core";
import { Visibility_$type } from "igniteui-angular-core";
import { YAxisLabelLocation_$type } from "./YAxisLabelLocation";
import { XAxisLabelLocation_$type } from "./XAxisLabelLocation";
import { AxisLabelSettings } from "./AxisLabelSettings";
import { AxisDefaults } from "./AxisDefaults";
import { TitleSettings } from "./TitleSettings";
import { ScalerParams } from "./ScalerParams";
import { DeviceUtils } from "igniteui-angular-core";
/**
 * @hidden
 */
var XYChart = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(XYChart, _super);
    function XYChart() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.zm = null;
        _this.zn = null;
        _this.z3 = 0;
        _this.z5 = 5;
        _this.z4 = 0;
        _this.z2 = 5;
        _this.aak = 5;
        _this.aam = 0;
        _this.aal = 5;
        _this.aaj = 0;
        _this.afd = null;
        _this.afk = null;
        _this.afb = null;
        _this.afc = null;
        _this.aae = 0;
        _this.aav = 0;
        _this.aad = NaN;
        _this.aau = NaN;
        _this.aag = NaN;
        _this.aax = NaN;
        _this.aaf = NaN;
        _this.aaw = NaN;
        _this.aac = NaN;
        _this.aat = NaN;
        _this.afj = null;
        _this.afq = null;
        _this.ab5 = null;
        _this.ab9 = null;
        _this.ab7 = null;
        _this.acb = null;
        _this.ab0 = null;
        _this.ab1 = null;
        _this.afe = null;
        _this.afl = null;
        _this.z6 = DeviceUtils.g(1);
        _this.aan = DeviceUtils.g(1);
        _this.z7 = DeviceUtils.g(1);
        _this.aao = DeviceUtils.g(1);
        _this.afg = null;
        _this.afn = null;
        _this.afh = null;
        _this.afo = null;
        _this.z8 = DeviceUtils.g(1);
        _this.aap = DeviceUtils.g(1);
        _this.z9 = DeviceUtils.g(8);
        _this.aaq = 0;
        _this.afi = null;
        _this.afp = null;
        _this.aaa = DeviceUtils.g(1);
        _this.aar = DeviceUtils.g(1);
        _this.ab6 = null;
        _this.aca = null;
        _this.aff = null;
        _this.afm = null;
        _this.z1 = 0;
        _this.aai = 0;
        _this.z0 = NaN;
        _this.aah = NaN;
        _this.aab = 0;
        _this.aas = -90;
        _this.zw = false;
        _this.zx = false;
        _this.ae6 = 1;
        _this.af9 = 1;
        _this.ae5 = 1;
        _this.ae7 = 2;
        _this.af7 = 0;
        _this.af8 = 1;
        _this.agd = 0;
        _this.age = 0;
        _this.zu = 4;
        _this.zs = 4;
        _this.ab4 = null;
        _this.zg = null;
        _this.ab8 = null;
        _this.zh = null;
        return _this;
    }
    XYChart.prototype.provideContainer = function (a) {
        if (a != null) {
            var b = this.zk();
            var c = this.zl();
            if (b != null) {
                b.provideRenderer(a);
            }
            if (c != null) {
                c.provideRenderer(a);
            }
        }
        _super.prototype.provideContainer.call(this, a);
    };
    XYChart.prototype.ux = function () {
        _super.prototype.ux.call(this);
        var a = this.zk();
        var b = this.zl();
        if (a != null && a.az != null) {
            this.aft = a.az.textColor;
        }
        if (b != null && b.az != null) {
            this.af0 = b.az.textColor;
        }
        this.afy = AxisDefaults.j;
        this.af5 = AxisDefaults.j;
        this.afu = AxisDefaults.h;
        this.af1 = AxisDefaults.h;
        this.afx = AxisDefaults.g;
        this.af4 = AxisDefaults.g;
    };
    XYChart.prototype.vo = function (a, b, c) {
        var _this = this;
        _super.prototype.vo.call(this, a, b, c);
        switch (a) {
            case "XAxisLabelLeftMargin":
                this.aex(function (d) { return d.leftMargin = _this.aa5; });
                break;
            case "XAxisLabelTopMargin":
                this.aex(function (d) { return d.topMargin = _this.aa7; });
                break;
            case "XAxisLabelRightMargin":
                this.aex(function (d) { return d.rightMargin = _this.aa6; });
                break;
            case "XAxisLabelBottomMargin":
                this.aex(function (d) { return d.bottomMargin = _this.aa4; });
                break;
            case "YAxisLabelLeftMargin":
                this.ae2(function (d) { return d.leftMargin = _this.abm; });
                break;
            case "YAxisLabelTopMargin":
                this.ae2(function (d) { return d.topMargin = _this.abo; });
                break;
            case "YAxisLabelRightMargin":
                this.ae2(function (d) { return d.rightMargin = _this.abn; });
                break;
            case "YAxisLabelBottomMargin":
                this.ae2(function (d) { return d.bottomMargin = _this.abl; });
                break;
            case "ActualItemsSource":
            case "ShouldDisplayMockData":
            case "ShouldUseSkeletonStyleForMockData":
                this.afr = this.yg(this.aft);
                this.afs = this.yg(this.af0);
                break;
            case "XAxisLabelTextColor":
                this.afr = this.yg(this.aft);
                break;
            case "YAxisLabelTextColor":
                this.afs = this.yg(this.af0);
                break;
            case "ActualXAxisLabelTextColor":
                this.aex(function (d) { return d.textColor = _this.afr; });
                break;
            case "ActualYAxisLabelTextColor":
                this.ae2(function (d) { return d.textColor = _this.afs; });
                break;
            case "XAxisTitleMargin":
                this.aez(function (d) { return d.margin = _this.abg; });
                break;
            case "YAxisTitleMargin":
                this.ae4(function (d) { return d.margin = _this.abx; });
                break;
            case "XAxisTitleLeftMargin":
                this.aez(function (d) { return d.leftMargin = _this.abf; });
                break;
            case "YAxisTitleLeftMargin":
                this.ae4(function (d) { return d.leftMargin = _this.abw; });
                break;
            case "XAxisTitleTopMargin":
                this.aez(function (d) { return d.topMargin = _this.abi; });
                break;
            case "YAxisTitleTopMargin":
                this.ae4(function (d) { return d.topMargin = _this.abz; });
                break;
            case "XAxisTitleRightMargin":
                this.aez(function (d) { return d.rightMargin = _this.abh; });
                break;
            case "YAxisTitleRightMargin":
                this.ae4(function (d) { return d.rightMargin = _this.aby; });
                break;
            case "XAxisTitleBottomMargin":
                this.aez(function (d) { return d.bottomMargin = _this.abe; });
                break;
            case "YAxisTitleBottomMargin":
                this.ae4(function (d) { return d.bottomMargin = _this.abv; });
                break;
            case "XAxisTitleTextColor":
                this.aez(function (d) { return d.textColor = _this.afz; });
                break;
            case "YAxisTitleTextColor":
                this.ae4(function (d) { return d.textColor = _this.af6; });
                break;
            case "XAxisFormatLabel":
                this.aev(function (d) { return d.au = _this.zo; });
                break;
            case "YAxisFormatLabel":
                this.ae0(function (d) { return d.au = _this.zp; });
                break;
        }
    };
    Object.defineProperty(XYChart.prototype, "zo", {
        get: function () {
            return this.zm;
        },
        set: function (a) {
            var b = this.zo;
            if (a != b) {
                this.zm = a;
                this.vl("XAxisFormatLabel", b, this.zo);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "zp", {
        get: function () {
            return this.zn;
        },
        set: function (a) {
            var b = this.zp;
            if (a != b) {
                this.zn = a;
                this.vl("YAxisFormatLabel", b, this.zp);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aa5", {
        get: function () {
            return this.z3;
        },
        set: function (a) {
            var b = this.aa5;
            if (a != b) {
                this.z3 = a;
                this.vl("XAxisLabelLeftMargin", b, this.aa5);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aa7", {
        get: function () {
            return this.z5;
        },
        set: function (a) {
            var b = this.aa7;
            if (a != b) {
                this.z5 = a;
                this.vl("XAxisLabelTopMargin", b, this.aa7);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aa6", {
        get: function () {
            return this.z4;
        },
        set: function (a) {
            var b = this.aa6;
            if (a != b) {
                this.z4 = a;
                this.vl("XAxisLabelRightMargin", b, this.aa6);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aa4", {
        get: function () {
            return this.z2;
        },
        set: function (a) {
            var b = this.aa4;
            if (a != b) {
                this.z2 = a;
                this.vl("XAxisLabelBottomMargin", b, this.aa4);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abm", {
        get: function () {
            return this.aak;
        },
        set: function (a) {
            var b = this.abm;
            if (a != b) {
                this.aak = a;
                this.vl("YAxisLabelLeftMargin", b, this.abm);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abo", {
        get: function () {
            return this.aam;
        },
        set: function (a) {
            var b = this.abo;
            if (a != b) {
                this.aam = a;
                this.vl("YAxisLabelTopMargin", b, this.abo);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abn", {
        get: function () {
            return this.aal;
        },
        set: function (a) {
            var b = this.abn;
            if (a != b) {
                this.aal = a;
                this.vl("YAxisLabelRightMargin", b, this.abn);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abl", {
        get: function () {
            return this.aaj;
        },
        set: function (a) {
            var b = this.abl;
            if (a != b) {
                this.aaj = a;
                this.vl("YAxisLabelBottomMargin", b, this.abl);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aft", {
        get: function () {
            return this.afd;
        },
        set: function (a) {
            var b = this.aft;
            if (a != b) {
                this.afd = a;
                this.vl("XAxisLabelTextColor", b, this.aft);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "af0", {
        get: function () {
            return this.afk;
        },
        set: function (a) {
            var b = this.af0;
            if (a != b) {
                this.afk = a;
                this.vl("YAxisLabelTextColor", b, this.af0);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "afr", {
        get: function () {
            return this.afb;
        },
        set: function (a) {
            var b = this.afb;
            if (a != b) {
                this.afb = a;
                this.vl("ActualXAxisLabelTextColor", b, this.afr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "afs", {
        get: function () {
            return this.afc;
        },
        set: function (a) {
            var b = this.afc;
            if (a != b) {
                this.afc = a;
                this.vl("ActualYAxisLabelTextColor", b, this.afs);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abg", {
        get: function () {
            return this.aae;
        },
        set: function (a) {
            var b = this.abg;
            if (a != b) {
                this.aae = a;
                this.vl("XAxisTitleMargin", b, this.abg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abx", {
        get: function () {
            return this.aav;
        },
        set: function (a) {
            var b = this.abx;
            if (a != b) {
                this.aav = a;
                this.vl("YAxisTitleMargin", b, this.abx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abf", {
        get: function () {
            return this.aad;
        },
        set: function (a) {
            var b = this.abf;
            if (a != b) {
                this.aad = a;
                this.vl("XAxisTitleLeftMargin", b, this.abf);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abw", {
        get: function () {
            return this.aau;
        },
        set: function (a) {
            var b = this.abw;
            if (a != b) {
                this.aau = a;
                this.vl("YAxisTitleLeftMargin", b, this.abw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abi", {
        get: function () {
            return this.aag;
        },
        set: function (a) {
            var b = this.abi;
            if (a != b) {
                this.aag = a;
                this.vl("XAxisTitleTopMargin", b, this.abi);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abz", {
        get: function () {
            return this.aax;
        },
        set: function (a) {
            var b = this.abz;
            if (a != b) {
                this.aax = a;
                this.vl("YAxisTitleTopMargin", b, this.abz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abh", {
        get: function () {
            return this.aaf;
        },
        set: function (a) {
            var b = this.abh;
            if (a != b) {
                this.aaf = a;
                this.vl("XAxisTitleRightMargin", b, this.abh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aby", {
        get: function () {
            return this.aaw;
        },
        set: function (a) {
            var b = this.aby;
            if (a != b) {
                this.aaw = a;
                this.vl("YAxisTitleRightMargin", b, this.aby);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abe", {
        get: function () {
            return this.aac;
        },
        set: function (a) {
            var b = this.abe;
            if (a != b) {
                this.aac = a;
                this.vl("XAxisTitleBottomMargin", b, this.abe);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abv", {
        get: function () {
            return this.aat;
        },
        set: function (a) {
            var b = this.abv;
            if (a != b) {
                this.aat = a;
                this.vl("YAxisTitleBottomMargin", b, this.abv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "afz", {
        get: function () {
            return this.afj;
        },
        set: function (a) {
            var b = this.afz;
            if (a != b) {
                this.afj = a;
                this.vl("XAxisTitleTextColor", b, this.afz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "af6", {
        get: function () {
            return this.afq;
        },
        set: function (a) {
            var b = this.af6;
            if (a != b) {
                this.afq = a;
                this.vl("YAxisTitleTextColor", b, this.af6);
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.vp = function (a, b, c) {
        var _this = this;
        _super.prototype.vp.call(this, a, b, c);
        switch (a) {
            case "XAxisLabelTextStyle":
                this.aex(function (d) { return d.textStyle = _this.act; });
                break;
            case "YAxisLabelTextStyle":
                this.ae2(function (d) { return d.textStyle = _this.ady; });
                break;
            case "XAxisTitleTextStyle":
                this.aez(function (d) { return d.textStyle = _this.adh; });
                break;
            case "YAxisTitleTextStyle":
                this.ae4(function (d) { return d.textStyle = _this.aem; });
                break;
        }
    };
    Object.defineProperty(XYChart.prototype, "act", {
        get: function () {
            return this.ab5;
        },
        set: function (a) {
            var b = this.act;
            if (a != b) {
                this.ab5 = a;
                this.vl("XAxisLabelTextStyle", b, this.act);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ady", {
        get: function () {
            return this.ab9;
        },
        set: function (a) {
            var b = this.ady;
            if (a != b) {
                this.ab9 = a;
                this.vl("YAxisLabelTextStyle", b, this.ady);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "adh", {
        get: function () {
            return this.ab7;
        },
        set: function (a) {
            var b = this.adh;
            if (a != b) {
                this.ab7 = a;
                this.vl("XAxisTitleTextStyle", b, this.adh);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aem", {
        get: function () {
            return this.acb;
        },
        set: function (a) {
            var b = this.aem;
            if (a != b) {
                this.acb = a;
                this.vl("YAxisTitleTextStyle", b, this.aem);
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.aes = function (a) {
        var _this = this;
        this.aep(a, function (b) { return b.leftMargin = _this.aa5; });
        this.aep(a, function (b) { return b.topMargin = _this.aa7; });
        this.aep(a, function (b) { return b.rightMargin = _this.aa6; });
        this.aep(a, function (b) { return b.bottomMargin = _this.aa4; });
        this.aep(a, function (b) { return b.textColor = _this.afr; });
        this.aeq(a, function (b) { return b.textColor = _this.afz; });
        this.aeq(a, function (b) { return b.leftMargin = _this.abf; });
        this.aeq(a, function (b) { return b.topMargin = _this.abi; });
        this.aeq(a, function (b) { return b.rightMargin = _this.abh; });
        this.aeq(a, function (b) { return b.bottomMargin = _this.abe; });
        this.aep(a, function (b) { return b.textStyle = _this.act; });
        this.aeq(a, function (b) { return b.textStyle = _this.adh; });
    };
    XYChart.prototype.aeu = function (a) {
        var _this = this;
        this.aep(a, function (b) { return b.leftMargin = _this.abm; });
        this.aep(a, function (b) { return b.topMargin = _this.abo; });
        this.aep(a, function (b) { return b.rightMargin = _this.abn; });
        this.aep(a, function (b) { return b.bottomMargin = _this.abl; });
        this.aep(a, function (b) { return b.textColor = _this.afs; });
        this.aeq(a, function (b) { return b.textColor = _this.af6; });
        this.aeq(a, function (b) { return b.leftMargin = _this.abw; });
        this.aeq(a, function (b) { return b.topMargin = _this.abz; });
        this.aeq(a, function (b) { return b.rightMargin = _this.aby; });
        this.aeq(a, function (b) { return b.bottomMargin = _this.abv; });
        this.aep(a, function (b) { return b.textStyle = _this.ady; });
        this.aeq(a, function (b) { return b.textStyle = _this.aem; });
    };
    XYChart.prototype.wf = function (a) {
        var b = this.bi(a);
        var c = this.dg(a, b);
        if (c == null) {
            return;
        }
        c.itemsSource = this.h2(a);
        if (a.highlightedData != null) {
            c.highlightedItemsSource = this.h3(a);
        }
        c.name = this.qk(a);
        c.title = this.ql(a, b);
        this.dataChart.series.add(c);
    };
    XYChart.prototype.aev = function (a) {
        this.aew(a, null);
    };
    XYChart.prototype.aew = function (a, b) {
        a(this.zk());
    };
    XYChart.prototype.ae1 = function (a, b) {
        a(this.zl());
    };
    XYChart.prototype.ae0 = function (a) {
        this.ae1(a, true);
    };
    XYChart.prototype.uh = function (a, b) {
        this.aev(a);
        this.ae1(a, b);
    };
    XYChart.prototype.aez = function (a) {
        this.aeq(this.zk(), a);
    };
    XYChart.prototype.ae4 = function (a) {
        this.aeq(this.zl(), a);
    };
    XYChart.prototype.ae2 = function (a) {
        this.aep(this.zl(), a);
    };
    XYChart.prototype.aex = function (a) {
        this.aep(this.zk(), a);
    };
    XYChart.prototype.aay = function (a) {
        var b = this.zk();
        return b.ek(a, new ScalerParams(1, this.yp, this.yo, b.ch));
    };
    XYChart.prototype.aa0 = function (a) {
        var b = this.zk();
        return b.en(a, new ScalerParams(1, this.yp, this.yo, b.ch));
    };
    XYChart.prototype.aaz = function (a) {
        var b = this.zl();
        return b.ek(a, new ScalerParams(1, this.yp, this.yo, b.ch));
    };
    XYChart.prototype.aa1 = function (a) {
        var b = this.zl();
        return b.en(a, new ScalerParams(1, this.yp, this.yo, b.ch));
    };
    XYChart.prototype.wc = function (a) {
        _super.prototype.wc.call(this, a);
        if (a == null) {
            return;
        }
        if (a.axes != null) {
            var b = this.zk();
            var c = this.zl();
            if (a.axes.contains(b)) {
                a.axes.remove(b);
            }
            if (a.axes.contains(c)) {
                a.axes.remove(c);
            }
        }
    };
    XYChart.ace = function (a, b) {
        var e_1, _a;
        try {
            for (var _b = __values(fromEnum(b)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                var d = c.findMatchingHint(16);
                if (d != null) {
                    var e = d.path;
                    return e;
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
        return null;
    };
    XYChart.prototype.us = function (a, b) {
        _super.prototype.us.call(this, a, b);
        if (a == this.dataChart) {
            this.dataChart.axes.add(this.zk());
            this.dataChart.axes.add(this.zl());
        }
    };
    XYChart.prototype.aer = function (a) {
        var _this = this;
        if (a != null && this.lu != null && a.fm == null) {
            a.provideRenderer(this.lu);
        }
        a.jx = this.afu;
        a.es = this.aa8;
        a.jy = this.afv;
        a.ev = this.aa9;
        a.j0 = this.afx;
        a.fh = this.aba;
        a.jz = this.afw;
        a.fi = this.abb;
        a.j1 = this.afy;
        a.fj = this.abc;
        a.c4 = true;
        a.cu = this.hg;
        a.cw = this.hg;
        var b = a == this.zk();
        if (b) {
            a.g7 = this.ac8;
        }
        a.f1 = this.ab2;
        a.gh = this.ack;
        a.ah = this.zi;
        this.aep(a, function (c) { return c.angle = _this.aa3; });
        this.aep(a, function (c) {
            if (c.extent != _this.aa2) {
                c.extent = _this.aa2;
            }
        });
        this.aep(a, function (c) { return c.horizontalAlignment = _this.ae8; });
        this.aep(a, function (c) { return c.verticalAlignment = _this.aga; });
        this.aep(a, function (c) { return c.visibility = _this.agf; });
        this.aep(a, function (c) { return c.location = _this.zq(_this.zt); });
        this.aeq(a, function (c) { return c.horizontalAlignment = _this.ae9; });
        this.aeq(a, function (c) { return c.angle = _this.abd; });
        this.aes(a);
    };
    XYChart.prototype.aet = function (a) {
        var _this = this;
        if (a == null) {
            return;
        }
        if (this.lu != null && a.fm == null) {
            a.provideRenderer(this.lu);
        }
        a.cu = this.hg;
        a.cw = this.hg;
        this.aep(a, function (b) { return b.angle = _this.abk; });
        this.aep(a, function (b) {
            if (b.extent != _this.abj) {
                b.extent = _this.abj;
            }
        });
        this.aep(a, function (b) { return b.horizontalAlignment = _this.afa; });
        this.aep(a, function (b) { return b.verticalAlignment = _this.agb; });
        this.aep(a, function (b) { return b.visibility = _this.agg; });
        this.aep(a, function (b) { return b.location = _this.zr(_this.zv); });
        this.aeq(a, function (b) { return b.verticalAlignment = _this.agc; });
        this.aeq(a, function (b) { return b.angle = _this.abu; });
        a.jx = this.af1;
        a.es = this.abp;
        a.jy = this.af2;
        a.ev = this.abq;
        a.j0 = this.af4;
        a.fh = this.abr;
        a.jz = this.af3;
        a.fi = this.abs;
        a.j1 = this.af5;
        a.fj = this.abt;
        var b = a == this.zl();
        if (b) {
            a.g7 = this.aed;
        }
        a.gh = this.adp;
        a.ah = this.zj;
        this.aeu(a);
    };
    XYChart.prototype.up = function () {
        _super.prototype.up.call(this);
        this.aer(this.zk());
        this.aet(this.zl());
    };
    XYChart.prototype.aep = function (a, b) {
        if (a == null) {
            return;
        }
        if (a.az == null) {
            a.az = new AxisLabelSettings();
            a.az.angle = a.d6();
            a.az.bottomMargin = a.d7();
            a.az.leftMargin = a.d8();
            a.az.rightMargin = a.d9();
            a.az.topMargin = a.ea();
            a.az.horizontalAlignment = a.js();
            a.az.location = a.a0();
            a.az.verticalAlignment = a.kl();
        }
        b(a.az);
    };
    XYChart.prototype.aeq = function (a, b) {
        if (a == null) {
            return;
        }
        if (a.bf == null) {
            a.bf = new TitleSettings();
            a.bf.bottomMargin = a.ec();
            a.bf.leftMargin = a.ed();
            a.bf.rightMargin = a.ee();
            a.bf.topMargin = a.ef();
            a.bf.location = a.a1();
            a.bf.angle = a.eb();
            a.bf.horizontalAlignment = a.jt();
            a.bf.verticalAlignment = a.km();
        }
        b(a.bf);
    };
    XYChart.prototype.vl = function (a, b, c) {
        var _this = this;
        _super.prototype.vl.call(this, a, b, c);
        switch (a) {
            case "XAxisLabelAngle":
                this.aex(function (d) { return d.angle = _this.aa3; });
                break;
            case "YAxisLabelAngle":
                this.ae2(function (d) { return d.angle = _this.abk; });
                break;
            case "XAxisLabelHorizontalAlignment":
                this.aex(function (d) { return d.horizontalAlignment = _this.ae8; });
                break;
            case "YAxisLabelHorizontalAlignment":
                this.ae2(function (d) { return d.horizontalAlignment = _this.afa; });
                break;
            case "XAxisLabelVerticalAlignment":
                this.aex(function (d) { return d.verticalAlignment = _this.aga; });
                break;
            case "YAxisLabelVerticalAlignment":
                this.ae2(function (d) { return d.verticalAlignment = _this.agb; });
                break;
            case "XAxisLabelVisibility":
                this.aex(function (d) { return d.visibility = _this.agf; });
                break;
            case "YAxisLabelVisibility":
                this.ae2(function (d) { return d.visibility = _this.agg; });
                break;
            case "XAxisLabel":
                this.aev(function (d) { return d.f1 = _this.ab2; });
                break;
            case "YAxisLabel":
                this.ae0(function (d) { return d.f1 = _this.ab3; });
                break;
            case "XAxisLabelFormat":
                this.aev(function (d) { return d.gh = _this.ack; });
                break;
            case "XAxisLabelFormatSpecifiers":
                this.aev(function (d) { return d.ah = _this.zi; });
                break;
            case "YAxisLabelFormat":
                this.ae0(function (d) { return d.gh = _this.adp; });
                break;
            case "YAxisLabelFormatSpecifiers":
                this.ae0(function (d) { return d.ah = _this.zj; });
                break;
            case "XAxisTitleAngle":
                this.aez(function (d) { return d.angle = _this.abd; });
                break;
            case "YAxisTitleAngle":
                this.ae4(function (d) { return d.angle = _this.abu; });
                break;
            case "XAxisTitleAlignment":
                this.aez(function (d) { return d.horizontalAlignment = _this.ae9; });
                break;
            case "YAxisTitleAlignment":
                this.ae4(function (d) { return d.verticalAlignment = _this.agc; });
                break;
            case "XAxisTitle":
                this.aev(function (d) { return d.g7 = _this.ac8; });
                break;
            case "YAxisTitle":
                this.ae1(function (d) { return d.g7 = _this.aed; }, false);
                break;
            case "XAxisMajorStroke":
                this.aew(function (d) { return d.jx = _this.afu; }, a);
                break;
            case "YAxisMajorStroke":
                this.ae0(function (d) { return d.jx = _this.af1; });
                break;
            case "XAxisMajorStrokeThickness":
                this.aew(function (d) { return d.es = _this.aa8; }, a);
                break;
            case "YAxisMajorStrokeThickness":
                this.ae0(function (d) { return d.es = _this.abp; });
                break;
            case "XAxisMinorStroke":
                this.aev(function (d) { return d.jy = _this.afv; });
                break;
            case "YAxisMinorStroke":
                this.ae0(function (d) { return d.jy = _this.af2; });
                break;
            case "XAxisMinorStrokeThickness":
                this.aev(function (d) { return d.ev = _this.aa9; });
                break;
            case "YAxisMinorStrokeThickness":
                this.ae0(function (d) { return d.ev = _this.abq; });
                break;
            case "XAxisTickLength":
                this.aev(function (d) { return d.fi = _this.abb; });
                break;
            case "YAxisTickLength":
                this.ae0(function (d) { return d.fi = _this.abs; });
                break;
            case "XAxisTickStroke":
                this.aev(function (d) { return d.j1 = _this.afy; });
                break;
            case "YAxisTickStroke":
                this.ae0(function (d) { return d.j1 = _this.af5; });
                break;
            case "XAxisTickStrokeThickness":
                this.aev(function (d) { return d.fj = _this.abc; });
                break;
            case "YAxisTickStrokeThickness":
                this.ae0(function (d) { return d.fj = _this.abt; });
                break;
            case "XAxisStroke":
                this.aev(function (d) { return d.j0 = _this.afx; });
                break;
            case "YAxisStroke":
                this.ae0(function (d) { return d.j0 = _this.af4; });
                break;
            case "XAxisStrokeThickness":
                this.aev(function (d) { return d.fh = _this.aba; });
                break;
            case "YAxisStrokeThickness":
                this.ae0(function (d) { return d.fh = _this.abr; });
                break;
            case "XAxisExtent":
                this.aex(function (d) { return d.extent = _this.aa2; });
                break;
            case "YAxisExtent":
                this.ae2(function (d) { return d.extent = _this.abj; });
                break;
            case "XAxisStrip":
                this.aev(function (d) { return d.jz = _this.afw; });
                break;
            case "YAxisStrip":
                this.ae0(function (d) { return d.jz = _this.af3; });
                break;
            case "XAxisInverted":
                this.aev(function (d) { return d.ch = _this.zy; });
                break;
            case "YAxisInverted":
                this.ae0(function (d) { return d.ch = _this.zz; });
                break;
            case "YAxisLabelLocation":
                this.ae3();
                break;
            case "XAxisLabelLocation":
                this.aey();
                break;
        }
    };
    XYChart.prototype.ae3 = function () {
        var a = this.zr(this.zv);
        if (a == 2 || a == 7) {
            this.afa = 2;
        }
        else {
            this.afa = 0;
        }
        this.ae2(function (b) { return b.location = a; });
    };
    XYChart.prototype.aey = function () {
        var a = this.zq(this.zt);
        if (a == 1 || a == 5) {
            this.aga = 0;
        }
        else {
            this.aga = 2;
        }
        this.aex(function (b) { return b.location = a; });
    };
    XYChart.prototype.zr = function (a) {
        if (a == 0) {
            return 2;
        }
        else if (a == 1) {
            return 3;
        }
        else if (a == 3) {
            return 7;
        }
        else if (a == 2) {
            return 6;
        }
        else {
            if (this.cf == 1) {
                return 3;
            }
            else {
                return 2;
            }
        }
    };
    XYChart.prototype.zq = function (a) {
        if (a == 1) {
            return 1;
        }
        else if (a == 0) {
            return 0;
        }
        else if (a == 3) {
            return 5;
        }
        else if (a == 2) {
            return 4;
        }
        else {
            return 1;
        }
    };
    Object.defineProperty(XYChart.prototype, "ab2", {
        get: function () {
            return this.ab0;
        },
        set: function (a) {
            var b = this.ab2;
            if (a != b) {
                this.ab0 = a;
                this.vl("XAxisLabel", b, this.ab2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ab3", {
        get: function () {
            return this.ab1;
        },
        set: function (a) {
            var b = this.ab3;
            if (a != b) {
                this.ab1 = a;
                this.vl("YAxisLabel", b, this.ab3);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "afu", {
        get: function () {
            return this.afe;
        },
        set: function (a) {
            var b = this.afu;
            if (a != b) {
                this.afe = a;
                this.vl("XAxisMajorStroke", b, this.afu);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "af1", {
        get: function () {
            return this.afl;
        },
        set: function (a) {
            var b = this.af1;
            if (a != b) {
                this.afl = a;
                this.vl("YAxisMajorStroke", b, this.af1);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aa8", {
        get: function () {
            return this.z6;
        },
        set: function (a) {
            var b = this.aa8;
            if (a != b) {
                this.z6 = a;
                this.vl("XAxisMajorStrokeThickness", b, this.aa8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abp", {
        get: function () {
            return this.aan;
        },
        set: function (a) {
            var b = this.abp;
            if (a != b) {
                this.aan = a;
                this.vl("YAxisMajorStrokeThickness", b, this.abp);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aa9", {
        get: function () {
            return this.z7;
        },
        set: function (a) {
            var b = this.aa9;
            if (a != b) {
                this.z7 = a;
                this.vl("XAxisMinorStrokeThickness", b, this.aa9);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abq", {
        get: function () {
            return this.aao;
        },
        set: function (a) {
            var b = this.abq;
            if (a != b) {
                this.aao = a;
                this.vl("YAxisMinorStrokeThickness", b, this.abq);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "afw", {
        get: function () {
            return this.afg;
        },
        set: function (a) {
            var b = this.afw;
            if (a != b) {
                this.afg = a;
                this.vl("XAxisStrip", b, this.afw);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "af3", {
        get: function () {
            return this.afn;
        },
        set: function (a) {
            var b = this.af3;
            if (a != b) {
                this.afn = a;
                this.vl("YAxisStrip", b, this.af3);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "afx", {
        get: function () {
            return this.afh;
        },
        set: function (a) {
            var b = this.afx;
            if (a != b) {
                this.afh = a;
                this.vl("XAxisStroke", b, this.afx);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "af4", {
        get: function () {
            return this.afo;
        },
        set: function (a) {
            var b = this.af4;
            if (a != b) {
                this.afo = a;
                this.vl("YAxisStroke", b, this.af4);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aba", {
        get: function () {
            return this.z8;
        },
        set: function (a) {
            var b = this.aba;
            if (a != b) {
                this.z8 = a;
                this.vl("XAxisStrokeThickness", b, this.aba);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abr", {
        get: function () {
            return this.aap;
        },
        set: function (a) {
            var b = this.abr;
            if (a != b) {
                this.aap = a;
                this.vl("YAxisStrokeThickness", b, this.abr);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abb", {
        get: function () {
            return this.z9;
        },
        set: function (a) {
            var b = this.abb;
            if (a != b) {
                this.z9 = a;
                this.vl("XAxisTickLength", b, this.abb);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abs", {
        get: function () {
            return this.aaq;
        },
        set: function (a) {
            var b = this.abs;
            if (a != b) {
                this.aaq = a;
                this.vl("YAxisTickLength", b, this.abs);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "afy", {
        get: function () {
            return this.afi;
        },
        set: function (a) {
            var b = this.afy;
            if (a != b) {
                this.afi = a;
                this.vl("XAxisTickStroke", b, this.afy);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "af5", {
        get: function () {
            return this.afp;
        },
        set: function (a) {
            var b = this.af5;
            if (a != b) {
                this.afp = a;
                this.vl("YAxisTickStroke", b, this.af5);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abc", {
        get: function () {
            return this.aaa;
        },
        set: function (a) {
            var b = this.abc;
            if (a != b) {
                this.aaa = a;
                this.vl("XAxisTickStrokeThickness", b, this.abc);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abt", {
        get: function () {
            return this.aar;
        },
        set: function (a) {
            var b = this.abt;
            if (a != b) {
                this.aar = a;
                this.vl("YAxisTickStrokeThickness", b, this.abt);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ac8", {
        get: function () {
            return this.ab6;
        },
        set: function (a) {
            var b = this.ac8;
            if (a != b) {
                this.ab6 = a;
                this.vl("XAxisTitle", b, this.ac8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aed", {
        get: function () {
            return this.aca;
        },
        set: function (a) {
            var b = this.aed;
            if (a != b) {
                this.aca = a;
                this.vl("YAxisTitle", b, this.aed);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "afv", {
        get: function () {
            return this.aff;
        },
        set: function (a) {
            var b = this.afv;
            if (a != b) {
                this.aff = a;
                this.vl("XAxisMinorStroke", b, this.afv);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "af2", {
        get: function () {
            return this.afm;
        },
        set: function (a) {
            var b = this.af2;
            if (a != b) {
                this.afm = a;
                this.vl("YAxisMinorStroke", b, this.af2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aa3", {
        get: function () {
            return this.z1;
        },
        set: function (a) {
            var b = this.aa3;
            if (a != b) {
                this.z1 = a;
                this.vl("XAxisLabelAngle", b, this.aa3);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abk", {
        get: function () {
            return this.aai;
        },
        set: function (a) {
            var b = this.abk;
            if (a != b) {
                this.aai = a;
                this.vl("YAxisLabelAngle", b, this.abk);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aa2", {
        get: function () {
            return this.z0;
        },
        set: function (a) {
            var b = this.aa2;
            if (a != b) {
                this.z0 = a;
                this.vl("XAxisExtent", b, this.aa2);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abj", {
        get: function () {
            return this.aah;
        },
        set: function (a) {
            var b = this.abj;
            if (a != b) {
                this.aah = a;
                this.vl("YAxisExtent", b, this.abj);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abd", {
        get: function () {
            return this.aab;
        },
        set: function (a) {
            var b = this.abd;
            if (a != b) {
                this.aab = a;
                this.vl("XAxisTitleAngle", b, this.abd);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "abu", {
        get: function () {
            return this.aas;
        },
        set: function (a) {
            var b = this.abu;
            if (a != b) {
                this.aas = a;
                this.vl("YAxisTitleAngle", b, this.abu);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "zy", {
        get: function () {
            return this.zw;
        },
        set: function (a) {
            var b = this.zy;
            if (a != b) {
                this.zw = a;
                this.vl("XAxisInverted", b, this.zy);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "zz", {
        get: function () {
            return this.zx;
        },
        set: function (a) {
            var b = this.zz;
            if (a != b) {
                this.zx = a;
                this.vl("YAxisInverted", b, this.zz);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ae9", {
        get: function () {
            return this.ae6;
        },
        set: function (a) {
            var b = this.ae9;
            if (a != b) {
                this.ae6 = a;
                this.vl("XAxisTitleAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.ae9));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agc", {
        get: function () {
            return this.af9;
        },
        set: function (a) {
            var b = this.agc;
            if (a != b) {
                this.af9 = a;
                this.vl("YAxisTitleAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.agc));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ae8", {
        get: function () {
            return this.ae5;
        },
        set: function (a) {
            var b = this.ae8;
            if (a != b) {
                this.ae5 = a;
                this.vl("XAxisLabelHorizontalAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.ae8));
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.get_afa = function () {
        return this.ae7;
    };
    XYChart.prototype.set_afa = function (a) {
        var b = this.afa;
        if (a != b) {
            this.ae7 = a;
            this.vl("YAxisLabelHorizontalAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.afa));
        }
    };
    Object.defineProperty(XYChart.prototype, "afa", {
        get: function () {
            return this.get_afa();
        },
        set: function (a) {
            this.set_afa(a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "aga", {
        get: function () {
            return this.af7;
        },
        set: function (a) {
            var b = this.aga;
            if (a != b) {
                this.af7 = a;
                this.vl("XAxisLabelVerticalAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.aga));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agb", {
        get: function () {
            return this.af8;
        },
        set: function (a) {
            var b = this.agb;
            if (a != b) {
                this.af8 = a;
                this.vl("YAxisLabelVerticalAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.agb));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agf", {
        get: function () {
            return this.agd;
        },
        set: function (a) {
            var b = this.agf;
            if (a != b) {
                this.agd = a;
                this.vl("XAxisLabelVisibility", enumGetBox(Visibility_$type, b), enumGetBox(Visibility_$type, this.agf));
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "agg", {
        get: function () {
            return this.age;
        },
        set: function (a) {
            var b = this.agg;
            if (a != b) {
                this.age = a;
                this.vl("YAxisLabelVisibility", enumGetBox(Visibility_$type, b), enumGetBox(Visibility_$type, this.agg));
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.get_zv = function () {
        return this.zu;
    };
    XYChart.prototype.set_zv = function (a) {
        var b = this.zv;
        if (a != b) {
            this.zu = a;
            this.vl("YAxisLabelLocation", enumGetBox(YAxisLabelLocation_$type, b), enumGetBox(YAxisLabelLocation_$type, this.zv));
        }
    };
    Object.defineProperty(XYChart.prototype, "zv", {
        get: function () {
            return this.get_zv();
        },
        set: function (a) {
            this.set_zv(a);
        },
        enumerable: false,
        configurable: true
    });
    XYChart.prototype.get_zt = function () {
        return this.zs;
    };
    XYChart.prototype.set_zt = function (a) {
        var b = this.zt;
        if (a != b) {
            this.zs = a;
            this.vl("XAxisLabelLocation", enumGetBox(XAxisLabelLocation_$type, b), enumGetBox(XAxisLabelLocation_$type, this.zt));
        }
    };
    Object.defineProperty(XYChart.prototype, "zt", {
        get: function () {
            return this.get_zt();
        },
        set: function (a) {
            this.set_zt(a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "ack", {
        get: function () {
            return this.ab4;
        },
        set: function (a) {
            var b = this.ab4;
            if (a != b) {
                this.ab4 = a;
                this.vl("XAxisLabelFormat", b, this.ab4);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "zi", {
        get: function () {
            return this.zg;
        },
        set: function (a) {
            var b = this.zg;
            if (a != b) {
                this.zg = a;
                this.vl("XAxisLabelFormatSpecifiers", b, this.zg);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "adp", {
        get: function () {
            return this.ab8;
        },
        set: function (a) {
            var b = this.ab8;
            if (a != b) {
                this.ab8 = a;
                this.vl("YAxisLabelFormat", b, this.ab8);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XYChart.prototype, "zj", {
        get: function () {
            return this.zh;
        },
        set: function (a) {
            var b = this.zh;
            if (a != b) {
                this.zh = a;
                this.vl("YAxisLabelFormatSpecifiers", b, this.zh);
            }
        },
        enumerable: false,
        configurable: true
    });
    XYChart.$t = markType(XYChart, 'XYChart', DomainChart.$);
    return XYChart;
}(DomainChart));
export { XYChart };
