// igniteui-angular-charts/esm2015/lib/AnnotationLayerProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, typeCast, delegateCombine, delegateRemove, markType } from "igniteui-angular-core";
import { IAnnotationLayerProxy_$type } from "./IAnnotationLayerProxy";
import { CrosshairLayer } from "./CrosshairLayer";
import { ValueLayer } from "./ValueLayer";
import { CalloutLayer } from "./CalloutLayer";
import { FinalValueLayer } from "./FinalValueLayer";
import { DataToolTipLayer } from "./DataToolTipLayer";
import { DoubleCollection } from "igniteui-angular-core";
/**
 * @hidden
 */
export let AnnotationLayerProxy = /*@__PURE__*/ (() => {
    class AnnotationLayerProxy extends Base {
        constructor() {
            super(...arguments);
            this._container = null;
        }
        get container() {
            return this._container;
        }
        set container(a) {
            this._container = a;
        }
        matchesType(a) {
            if (!a.eq) {
                return false;
            }
            return true;
        }
        createCrosshairLayer() {
            let a = new CrosshairLayer();
            let b = this.container.getExternal(a, null, null);
            return a;
        }
        createValueLayer() {
            let a = new ValueLayer();
            let b = this.container.getExternal(a, null, null);
            return a;
        }
        createCalloutLayer() {
            let a = new CalloutLayer();
            let b = this.container.getExternal(a, null, null);
            return a;
        }
        createFinalValueLayer() {
            let a = new FinalValueLayer();
            let b = this.container.getExternal(a, null, null);
            return a;
        }
        setCrosshairHorizontalLineVisibility(a, b) {
            a.acg = b;
        }
        setCrosshairVerticalLineVisibility(a, b) {
            a.ach = b;
        }
        setCrosshairHorizontalLineStroke(a, b) {
            a.ab4 = b;
        }
        setCrosshairVerticalLineStroke(a, b) {
            a.ab5 = b;
        }
        setCrosshairXAxisForeground(a, b) {
            a.ab8 = b;
        }
        setCrosshairYAxisForeground(a, b) {
            a.acb = b;
        }
        setCrosshairXAxisBackground(a, b) {
            a.ab6 = b;
        }
        setCrosshairYAxisBackground(a, b) {
            a.ab9 = b;
        }
        setCrosshairYAxisPrecision(a, b) {
            a.aad = b;
        }
        setCrosshairXAxisPrecision(a, b) {
            a.aac = b;
        }
        setCrosshairOnAxesEnabled(a, b) {
            a.zv = b;
        }
        setCrosshairUseInterpolation(a, b) {
            a.zx = b;
        }
        setCrosshairThickness(a, b) {
            a.jn = b;
        }
        isCrosshairLayer(a) {
            return typeCast(CrosshairLayer.$, a) !== null;
        }
        setFinalValuePrecision(a, b) {
            a.z2 = b;
        }
        setFinalValueBackground(a, b) {
            a.aaw = b;
        }
        setFinalValueForeground(a, b) {
            a.aay = b;
        }
        setContentMemberPath(a, b) {
            a.acd = b;
        }
        setCalloutsAllowedPositions(a, b) {
            a.z2 = b;
        }
        setLabelMemberPath(a, b) {
            a.acr = b;
        }
        setXMemberPath(a, b) {
            a.ac5 = b;
        }
        setYMemberPath(a, b) {
            a.ac8 = b;
        }
        setTargetSeries(a, b) {
            a.z7 = b;
        }
        setCollisionChannel(a, b) {
            a.aca = b;
        }
        setIsAutoCalloutBehaviorEnabled(a, b) {
            a.aav = b;
        }
        setGreedyCollisionMode(a, b) {
            a.zx = b ? 2 : 0;
        }
        setCollisionMode(a, b) {
            a.zx = b;
        }
        setUseItemColorForFill(a, b) {
            a.aa1 = b;
        }
        setUseItemColorForOutline(a, b) {
            a.aa2 = b;
        }
        setCalloutStrokeThickness(a, b) {
            a.abk = b;
        }
        setUseAutoContrastingLabelColors(a, b) {
            a.aaz = b;
        }
        setCalloutBackground(a, b) {
            a.aet = b;
        }
        setAutoCalloutLabelPrecision(a, b) {
            a.abn = b;
        }
        setCalloutLightColor(a, b) {
            a.aew = b;
        }
        setCalloutDarkColor(a, b) {
            a.aeu = b;
        }
        setUseDedicatedLanes(a, b) {
            a.zv = b ? 2 : 0;
        }
        isCalloutLayer(a) {
            if (typeCast(CalloutLayer.$, a) !== null) {
                return true;
            }
            return false;
        }
        isDataToolTipLayer(a) {
            if (typeCast(DataToolTipLayer.$, a) !== null) {
                return true;
            }
            return false;
        }
        isFinalValueLayer(a) {
            if (typeCast(FinalValueLayer.$, a) !== null) {
                return true;
            }
            return false;
        }
        addCalloutStyleUpdating(a, b) {
            let c = a;
            c.calloutStyleUpdating = delegateCombine(c.calloutStyleUpdating, b);
        }
        removeCalloutStyleUpdating(a, b) {
            let c = a;
            c.calloutStyleUpdating = delegateRemove(c.calloutStyleUpdating, b);
        }
        addCalloutRenderStyleUpdating(a, b) {
            let c = a;
            c.calloutRenderStyleUpdating = delegateCombine(c.calloutRenderStyleUpdating, b);
        }
        removeCalloutRenderStyleUpdating(a, b) {
            let c = a;
            c.calloutRenderStyleUpdating = delegateRemove(c.calloutRenderStyleUpdating, b);
        }
        setCalloutStyleUpdatingEventEnabled(a, b) {
            a.aay = b;
        }
        setUseValueForAutoCalloutLabels(a, b) {
            a.aa4 = b;
        }
        setCalloutRenderStyleUpdatingEventEnabled(a, b) {
            a.aax = b;
        }
        updateDataToolTipPropertyByName(a, b, c) {
            a.afc(b, c);
        }
        getIsAutoCalloutBehaviorEnabled(a) {
            return a.aav;
        }
        setValueMode(a, b) {
            a.zu = b;
        }
        setValueAxisAnnotationEnabled(a, b) {
            a.zw = b;
        }
        getValueMode(a) {
            return a.zu;
        }
        isValueLayer(a) {
            return typeCast(ValueLayer.$, a) !== null;
        }
        setBrush(a, b) {
            a.u9 = b;
        }
        setDashArray(a, b) {
            let c = new DoubleCollection();
            for (let d = 0; d < b.length; d++) {
                c.add(b[d]);
            }
            a.vw = c;
        }
    }
    AnnotationLayerProxy.$t = markType(AnnotationLayerProxy, 'AnnotationLayerProxy', Base.$, [IAnnotationLayerProxy_$type]);
    return AnnotationLayerProxy;
})();
