// igniteui-angular-charts/esm2015/lib/AverageDirectionalIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { AverageDirectionalIndexIndicatorStrategy } from "./AverageDirectionalIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let AverageDirectionalIndexIndicator = /*@__PURE__*/ (() => {
    class AverageDirectionalIndexIndicator extends StrategyBasedIndicator {
        bi() {
            return new AverageDirectionalIndexIndicator();
        }
        get_abn() {
            return new AverageDirectionalIndexIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return AverageDirectionalIndexIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        get ab8() {
            return typeGetValue(this.c(AverageDirectionalIndexIndicator.ab9));
        }
        set ab8(a) {
            this.h(AverageDirectionalIndexIndicator.ab9, a);
        }
        abt() {
            return this.ab8;
        }
        ab2(a) {
            this.ab8 = a;
        }
    }
    AverageDirectionalIndexIndicator.$t = markType(AverageDirectionalIndexIndicator, 'AverageDirectionalIndexIndicator', StrategyBasedIndicator.$);
    AverageDirectionalIndexIndicator.ab9 = StrategyBasedIndicator.ab5(14, AverageDirectionalIndexIndicator.$);
    return AverageDirectionalIndexIndicator;
})();
