// igniteui-angular-charts/esm2015/lib/BollingerBandWidthIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { BollingerBandWidthIndicatorStrategy } from "./BollingerBandWidthIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let BollingerBandWidthIndicator = /*@__PURE__*/ (() => {
    class BollingerBandWidthIndicator extends StrategyBasedIndicator {
        bi() {
            return new BollingerBandWidthIndicator();
        }
        get_abn() {
            return new BollingerBandWidthIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return BollingerBandWidthIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        get ab9() {
            return typeGetValue(this.c(BollingerBandWidthIndicator.acc));
        }
        set ab9(a) {
            this.h(BollingerBandWidthIndicator.acc, a);
        }
        abt() {
            return this.ab9;
        }
        get ab8() {
            return this.c(BollingerBandWidthIndicator.acb);
        }
        set ab8(a) {
            this.h(BollingerBandWidthIndicator.acb, a);
        }
        abr() {
            return this.ab8;
        }
    }
    BollingerBandWidthIndicator.$t = markType(BollingerBandWidthIndicator, 'BollingerBandWidthIndicator', StrategyBasedIndicator.$);
    BollingerBandWidthIndicator.acc = StrategyBasedIndicator.ab5(20, BollingerBandWidthIndicator.$);
    BollingerBandWidthIndicator.acb = StrategyBasedIndicator.ab4(2, BollingerBandWidthIndicator.$);
    return BollingerBandWidthIndicator;
})();
