// igniteui-angular-charts/esm2015/lib/CategoryHighlightLayer_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayer } from "./AnnotationLayer";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { DependencyProperty } from "igniteui-angular-core";
import { CategoryHighlightLayerFrame } from "./CategoryHighlightLayerFrame";
import { Base, Number_$type, typeCast, fromEnum, Point_$type, markType, markDep, runOn } from "igniteui-angular-core";
import { XamDataChart } from "./XamDataChart";
import { CategoryHighlightFrameRect } from "./CategoryHighlightFrameRect";
import { List$1 } from "igniteui-angular-core";
import { Pool$1 } from "igniteui-angular-core";
import { Rectangle } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { isNaN_, truncate } from "igniteui-angular-core";
import { AnnotationLayerView } from "./AnnotationLayerView";
import { BrushUtil } from "igniteui-angular-core";
import { Color } from "igniteui-angular-core";
/**
 * @hidden
 */
export let CategoryHighlightLayer = /*@__PURE__*/ (() => {
    class CategoryHighlightLayer extends AnnotationLayer {
        constructor() {
            super();
            let a = new CategoryHighlightLayerFrame();
            let b = new CategoryHighlightLayerFrame();
            let c = new CategoryHighlightLayerFrame();
            let d = this.kb / 1000;
            this.w4 = a;
            this.w3 = b;
            this.w5 = c;
            this.ab = CategoryHighlightLayer.$;
        }
        bi() {
            return new CategoryHighlightLayer();
        }
        get_ep() {
            return true;
        }
        cq() {
            return new CategoryHighlightLayerView(this);
        }
        qx(a) {
            super.qx(a);
            this.zq = a;
        }
        get zt() {
            return this.c(CategoryHighlightLayer.z0);
        }
        set zt(a) {
            this.h(CategoryHighlightLayer.z0, a);
        }
        yg(a, b) {
            if (this.zp == null || this.zp == a) {
                if (!this.a0.c) {
                    this.rn(b);
                }
            }
        }
        zr(a) {
            return a != null && a.co();
        }
        yj(a, b) {
            super.yj(a, b);
            let c = a;
            c.g.clear();
            let d = this.zs;
            if (isNaN_(this.zm.x) && isNaN_(this.zm.y)) {
                return;
            }
            if (this.zp != null) {
                if (this.zr(this.zp)) {
                    this.zy(this.zp, c, d);
                }
            }
            else {
                if (typeCast(XamDataChart.$, this.c9) !== null) {
                    let e = this.c9;
                    for (let f of fromEnum(e.axes)) {
                        if (typeCast(CategoryAxisBase.$, f) !== null && this.zr(f)) {
                            this.zy(f, c, d);
                        }
                    }
                }
            }
        }
        zy(a, b, c) {
            if (this.ct == null || this.ct.by.isEmpty || this.ct.bx.isEmpty) {
                return;
            }
            let d = this.zm.x;
            if (a.cp && isNaN_(d)) {
                d = 0;
            }
            let e = this.wd(this.ct);
            d = e.left + e.width * (d - this.ct.by.left) / this.ct.by.width;
            let f = this.zm.y;
            if (!a.cp && isNaN_(f)) {
                f = 0;
            }
            f = e.top + e.height * (f - this.ct.by.top) / this.ct.by.height;
            let g = a.mg({ $type: Point_$type, x: d, y: f }, c, this.zt);
            this.zx(g, b, a);
        }
        zx(a, b, c) {
            let d = new CategoryHighlightFrameRect();
            d.b = a.left;
            d.e = a.top;
            d.c = a.right;
            d.a = a.bottom;
            d.g = this.zc();
            d.h = this.zd();
            if (d.g == null) {
                d.g = this.zq.cf(c.jw);
            }
            if (d.h == null) {
                d.h = this.zq.cf(c.jw);
            }
            d.d = this.ic;
            d.f = Base.getHashCodeStatic(c);
            if (!a.isEmpty) {
                b.g.addItem(d.f, d);
            }
        }
        ym(a, b) {
            super.ym(a, b);
            let c = this.zq.b8;
            let d = a;
            let e = 0;
            for (let f of fromEnum(d.g.values)) {
                let g = c.item(e);
                this.zz(f, g);
                e++;
            }
            c.count = e;
        }
        f8() {
            return true;
        }
        zz(a, b) {
            if (isNaN_(a.b) || isNaN_(a.e) || isNaN_(a.a) || isNaN_(a.c) || isNaN_(a.d)) {
                b._visibility = 1;
                return;
            }
            this.zq.ca(b, a.b, a.e, a.c - a.b, a.a - a.e);
            b._fill = a.g;
            b._stroke = a.h;
            b.ad = a.d;
            if (this.vw != null) {
                b.ai = this.vw;
            }
            else {
                b.ai = null;
            }
            b.aj = this.vx;
            b.am = this.vz;
            b.al = this.vy;
            b.ak = this.vy;
            b.ac = this.jf;
        }
        q5(a, b, c, d) {
            super.q5(a, b, c, d);
            switch (b) {
                case CategoryHighlightLayer.$$p[0]:
                    this.rn(true);
                    break;
                case CategoryHighlightLayer.$$p[1]:
                    this.rn(true);
                    break;
            }
        }
    }
    CategoryHighlightLayer.$t = markType(CategoryHighlightLayer, 'CategoryHighlightLayer', AnnotationLayer.$);
    CategoryHighlightLayer.z0 = DependencyProperty.i("BAndHighlightWidth", Number_$type, CategoryHighlightLayer.$, new PropertyMetadata(2, 10, (a, b) => a.raisePropertyChanged("BAndHighlightWidth", b.oldValue, b.newValue)));
    CategoryHighlightLayer.$$p = markDep(DependencyProperty, PropertyMetadata, CategoryHighlightLayer, 'raisePropertyChanged', ['TargetAxis:zp:z1', [CategoryAxisBase.$, null], 'UseInterpolation:zs:z2', [0, false]]);
    return CategoryHighlightLayer;
})();
/**
 * @hidden
 */
export let CategoryHighlightLayerView = /*@__PURE__*/ (() => {
    class CategoryHighlightLayerView extends AnnotationLayerView {
        constructor(a) {
            super(a);
            this.b7 = null;
            this.b8 = null;
            this.b9 = null;
            this.b7 = a;
            this.b9 = new List$1(Rectangle.$, 0);
        }
        au() {
            super.au();
            this.b8 = ((() => {
                let $ret = new Pool$1(Rectangle.$);
                $ret.create = runOn(this, this.cg);
                $ret.activate = runOn(this, this.cb);
                $ret.disactivate = runOn(this, this.cd);
                $ret.destroy = runOn(this, this.cc);
                return $ret;
            })());
        }
        cg() {
            let a = new Rectangle();
            a.ab = false;
            this.b9.add(a);
            return a;
        }
        cb(a) {
            a._visibility = 0;
        }
        cd(a) {
            a._visibility = 1;
        }
        cc(a) {
            this.b9.remove(a);
        }
        ce(a) {
            if (a == null) {
                return a;
            }
            return BrushUtil.m(a, 0.1);
        }
        cf(a) {
            if (a != null) {
                let b = BrushUtil.m(a, 0.3);
                if (b.isGradient) {
                    let c = b;
                    if (c.gradientStops != null) {
                        for (let d = 0; d < c.gradientStops.length; d++) {
                            let e = c.gradientStops[d];
                            e.color = Color.u(truncate(Math.round((e.color.l) * 0.2)), e.color.o, e.color.n, e.color.m);
                            e.color = e.color;
                        }
                    }
                }
                else {
                    b.color = Color.u(truncate(Math.round((b.color.l) * 0.2)), b.color.o, b.color.n, b.color.m);
                    b.color = b.color;
                }
                return b;
            }
            else {
                return null;
            }
        }
        ca(a, b, c, d, e) {
            a._visibility = 0;
            a.n = b;
            a.o = c;
            a.width = d;
            a.height = e;
        }
        a8(a, b) {
            super.a8(a, b);
            if (a.d && !b) {
                for (let c = 0; c < this.b9.count; c++) {
                    let d = this.b9._inner[c];
                    if (d._visibility == 0) {
                        a.w(d);
                    }
                }
            }
        }
        an(a) {
            super.an(a);
            for (let b = 0; b < this.b9.count; b++) {
                this.e.bb.exportRectangleData(a, this.b9._inner[b], "cihRect", ["Main", "CategoryItem"]);
            }
        }
    }
    CategoryHighlightLayerView.$t = markType(CategoryHighlightLayerView, 'CategoryHighlightLayerView', AnnotationLayerView.$);
    return CategoryHighlightLayerView;
})();
