// igniteui-angular-charts/esm2015/lib/CategoryXAxis_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategoryAxisBase } from "./CategoryAxisBase";
import { IZoomableCategoryAxis_$type } from "./IZoomableCategoryAxis";
import { Base, Number_$type, typeCast, Array_$type, markType, markDep, TypeRegistrar } from "igniteui-angular-core";
import { DependencyProperty } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { HorizontalAxisLabelPanel } from "./HorizontalAxisLabelPanel";
import { Rect } from "igniteui-angular-core";
import { ScalerParams } from "./ScalerParams";
import { NumericScaler } from "./NumericScaler";
import { LinearCategorySnapper } from "./LinearCategorySnapper";
import { NumericYAxis } from "./NumericYAxis";
import { LabelPosition } from "./LabelPosition";
import { AxisRangeChangedEventArgs } from "./AxisRangeChangedEventArgs";
import { MathUtil } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { truncate, isNaN_, isInfinity } from "igniteui-angular-core";
import { CategoryAxisBaseView } from "./CategoryAxisBaseView";
import { AxisDefaults } from "./AxisDefaults";
/**
 * @hidden
 */
export let CategoryXAxis = /*@__PURE__*/ (() => {
    class CategoryXAxis extends CategoryAxisBase {
        constructor() {
            super();
            this.mw = 1;
            this.mv = 1;
            this.mz = null;
            this.mm = 0;
            this.mn = 0;
            this.kz = new List$1(Number_$type, 0);
            this.ab = CategoryXAxis.$;
        }
        a4() {
            return new CategoryXAxisView(this);
        }
        ht(a) {
            super.ht(a);
            this.mj = a;
        }
        get my() {
            return this.mw;
        }
        set my(a) {
            this.mw = a;
        }
        get mx() {
            return this.mv;
        }
        set mx(a) {
            this.mv = a;
        }
        av() {
            if (this.c6) {
                return (TypeRegistrar.create('HorizontalSmartAxisLabelPanel'));
            }
            else {
                return new HorizontalAxisLabelPanel();
            }
        }
        getCategorySize(a, b, c) {
            let d = this.kc(b, c);
            return d.width / (this.lg * a.width);
        }
        k5(a, b, c) {
            let d = this.lc(a, b, c);
            let e = 0;
            if (!isNaN_(this.lb)) {
                e = Math.min(this.lb, 1);
            }
            else {
                e = 0;
            }
            ;
            let f = 1 - 0.5 * d;
            let g = this.li == 0 ? 1 : this.li;
            let h = this.getCategorySize(a, b, c) * f / (g - (g - 1) * e);
            return h;
        }
        getGroupCenter(a, b, c, d) {
            let e = 0.5;
            if (this.li > 1) {
                let f = this.lc(b, c, d);
                let g = 0;
                if (!isNaN_(this.lb)) {
                    g = Math.min(this.lb, 1);
                }
                let h = 1 - 0.5 * f;
                let i = h / (this.li - (this.li - 1) * g);
                let j = (h - i) / (this.li - 1);
                e = 0.25 * f + 0.5 * i + a * j;
            }
            return this.getCategorySize(b, c, d) * e;
        }
        d5() {
            return this.mt;
        }
        m9(a) {
            let b = this.bi != null ? this.bi.vl : Rect.empty;
            b = b.copy();
            let c = this.ke;
            let d = this.kb();
            let e = new Rect(0, 0, 0, 1, 1);
            let f = new ScalerParams(0, e, e, this.ch, d);
            let g = !b.isEmpty && !c.isEmpty && this.ks != null ? this.ks.indexOf(a) : -1;
            let h = g > -1 ? this.ek(g, f) : NaN;
            if (!isNaN_(h) && this.bi.fc) {
                if (!isNaN_(h)) {
                    if (h < b.left + 0.1 * b.width) {
                        h = h + 0.4 * b.width;
                        b.x = h - 0.5 * b.width;
                    }
                    if (h > b.right - 0.1 * b.width) {
                        h = h - 0.4 * b.width;
                        b.x = h - 0.5 * b.width;
                    }
                }
                this.bi.rk(b, false, true);
            }
        }
        ek(a, b) {
            if (!b.c.isEmpty) {
                let c = this.ms(a, NumericScaler.ac, b.c, this.categoryMode);
                c = (c - b.d.left) / b.d.width;
                c = b.d.left + b.d.width * (c - b.e.left) / b.e.width;
                return c;
            }
            return this.ms(a, b.e, b.d, this.categoryMode);
        }
        mi(a) {
            if (typeCast(List$1.$.specialize(Array_$type), a) !== null) {
                return a.asArray();
            }
            return null;
        }
        hn(a, b, c, d, e) {
            let f;
            let g = e.e;
            let h = e.d;
            let i = e.c;
            let j = e.a;
            let k = !i.isEmpty;
            let l = this.categoryMode == 0 ? this.lg - 1 : this.lg;
            if (l < 0) {
                l = 0;
            }
            let m = i.left;
            let n = i.width;
            let o = g.left;
            let p = g.width;
            let q = h.left;
            let r = h.width;
            let s = 0;
            let t = 1;
            let u = this.mi(a);
            let v = false;
            if (u != null) {
                v = true;
            }
            let w;
            for (let x = 0; x < b.count; x++) {
                w = b.item(x);
                for (let y = c; y < d; y++) {
                    if (v) {
                        f = u[y];
                    }
                    else {
                        f = a.item(y);
                    }
                    if (k) {
                        let z = l >= 1 ? (f[w]) / (l) : l == 0 ? 0.5 : NaN;
                        if (j) {
                            z = 1 - z;
                        }
                        z = m + n * (z - s) / t;
                        z = (z - q) / r;
                        let aa = q + r * (z - o) / p;
                        f[w] = aa;
                    }
                    else {
                        let ab = l >= 1 ? (f[w]) / (l) : l == 0 ? 0.5 : NaN;
                        if (j) {
                            ab = 1 - ab;
                        }
                        ab = q + r * (ab - o) / p;
                        f[w] = ab;
                    }
                }
            }
        }
        el(a, b, c) {
            if (!b.c.isEmpty) {
                let d = this.ms(a, NumericScaler.ac, b.c, c);
                d = (d - b.d.left) / b.d.width;
                d = b.d.left + b.d.width * (d - b.e.left) / b.e.width;
                return d;
            }
            return this.ms(a, b.e, b.d, c);
        }
        ms(a, b, c, d) {
            let e = d == 0 ? this.lg - 1 : this.lg;
            if (e < 0) {
                e = 0;
            }
            let f = e >= 1 ? (a) / (e) : e == 0 ? 0.5 : NaN;
            if (this.ci) {
                f = 1 - f;
            }
            return c.left + c.width * (f - b.left) / b.width;
        }
        en(a, b) {
            if (!b.c.isEmpty) {
                let c = b.e.left + b.e.width * (a - b.d.left) / b.d.width;
                c = (c * b.d.width) + b.d.left;
                c = this.getUnscaledValue3(c, NumericScaler.ac, b.c, this.categoryMode);
                return c;
            }
            return this.getUnscaledValue3(a, b.e, b.d, this.categoryMode);
        }
        eo(a, b, c) {
            if (!b.c.isEmpty) {
                let d = b.e.left + b.e.width * (a - b.d.left) / b.d.width;
                d = (d * b.d.width) + b.d.left;
                d = this.getUnscaledValue3(d, NumericScaler.ac, b.c, c);
                return d;
            }
            return this.getUnscaledValue3(a, b.e, b.d, c);
        }
        getUnscaledValue3(a, b, c, d) {
            let e = b.left + (a - c.left) * b.width / c.width;
            if (this.ci) {
                e = 1 - e;
            }
            let f = d == 0 ? this.lg - 1 : this.lg;
            if (f < 0) {
                f = 0;
            }
            return e * f;
        }
        na(a, b) {
            if (isNaN_(a) || isNaN_(b) || this.ks == null || this.ks.count == 0) {
                return;
            }
            let c = a / this.ks.count;
            let d = (b + 1) / this.ks.count;
            this.bi.v1 = new Rect(0, c, this.bi.v1.y, d - c, this.bi.v1.height);
        }
        ia(a) {
            super.ia(a);
            let b = this.bi != null ? this.bi.vl : Rect.empty;
            let c = this.ke;
            let d = this.j9();
            let e = this.j8();
            let f = new ScalerParams(0, b, c, this.ch, d);
            let g = this.a5.a4();
            let h = this.a5.a7();
            let i = this.a5.a5();
            let j = this.a5.a6();
            let k = this.a5.e();
            let l = this.a5.f();
            let m = this.a5.g();
            let n = this.ks;
            let o = this.mk();
            this.im();
            this.hj(g);
            this.hj(h);
            this.hj(i);
            this.hj(j);
            this.dc.clear();
            this.dd.clear();
            this.kz.clear();
            this.a5.az(this, b, c);
            if (b.isEmpty || c.isEmpty) {
                this.br.count = 0;
                this.bq.count = 0;
                this.bp.count = 0;
            }
            if (this.br.count == 0 && this.bq.count == 0 && this.a5.l()) {
                this.a5.ad();
            }
            if (this.az != null) {
                this.az.registerAxis(this);
            }
            if (this.itemsSource == null || o == null || n.count == 0) {
                this.br.count = 0;
                this.bq.count = 0;
                this.bp.count = 0;
                this.a5.ad();
                return;
            }
            if (!b.isEmpty && !c.isEmpty) {
                let p = this.en(e.left, f);
                let q = this.en(e.right, f);
                let r = 0;
                let s = false;
                if (this.c4) {
                    let t = Math.abs(q - p);
                    r = Math.ceil(t) + 1;
                    if (this.ch) {
                        if (Math.abs(q - Math.floor(q)) < Math.abs(p - Math.ceil(p))) {
                            s = true;
                        }
                    }
                    else {
                        if (Math.abs(q - Math.ceil(q)) < Math.abs(p - Math.floor(p))) {
                            s = true;
                        }
                    }
                }
                if (this.ch) {
                    p = Math.ceil(p);
                    q = Math.floor(q);
                }
                else {
                    p = Math.floor(p);
                    q = Math.ceil(q);
                }
                if (this.c4) {
                    if (Math.abs(q - p) < r) {
                        if (this.ch) {
                            if (s) {
                                q--;
                            }
                            else {
                                p++;
                            }
                        }
                        else {
                            if (s) {
                                q++;
                            }
                            else {
                                p--;
                            }
                        }
                    }
                }
                let u = c.bottom;
                let v = u - c.top;
                if (this.aj != null) {
                    let w = typeCast(NumericYAxis.$, this.aj);
                    if (w != null) {
                        let x = this.ig(u, v, b, c, d, c.top, c.bottom);
                        u = x.p0;
                        v = x.p1;
                    }
                }
                this.hr(g, u, c, k, true);
                this.a5.au(v);
                let y = Math.min(p, q);
                let z = Math.max(p, q);
                let aa = e.width;
                if (this.c4) {
                    let ab = this.ek(y, f);
                    let ac = this.ek(z, f);
                    if (ab != ac) {
                        aa = Math.abs(ac - ab);
                    }
                }
                let ad = new LinearCategorySnapper(1, y, z, aa, this.mt, this.categoryMode, this.ez());
                let ae = ad.g;
                ae = this.d2(ae, this.ml(), b, c, d);
                if (ad.g != ae) {
                    ad = new LinearCategorySnapper(1, y, z, aa, ae, this.categoryMode, this.ez());
                }
                let af = Math.floor((y - 0) / ad.g);
                let ag = Math.ceil((z - 0) / ad.g);
                this.mo = ad.g;
                if (!isNaN_(af) && !isNaN_(ag)) {
                    let ah = truncate(af);
                    let ai = truncate(ag);
                    let aj = this.mr(0 + ah * ad.g, f);
                    let ak = this.ek(ad.g, f);
                    let al = Math.abs(ak - this.ek(0, f));
                    this.a5.av(ak);
                    let am = truncate(Math.ceil(e.right));
                    let an = truncate(Math.floor(e.left));
                    let ao = this.mu;
                    let ap = ao > 0 && ao < ad.g;
                    let aq = c.right;
                    let ar = this.kz;
                    let as = ad.g;
                    let at = truncate(as);
                    let au = this.li;
                    let av = this.ch;
                    let aw = this.categoryMode;
                    this.mz = null;
                    for (let ax = ah; ax <= ai; ++ax) {
                        let ay = this.mr(0 + (ax + 1) * ad.g, f);
                        if (aj <= aq) {
                            if (ax % 2 == 0) {
                                this.ip(h, aj, ay, c);
                            }
                            this.io(i, aj, c, l, false);
                            ar.add(aj);
                        }
                        if (this.categoryMode != 0 && this.li != 0 && this.cy) {
                            if (ap) {
                                let az = al * this.mu;
                                if (this.ch) {
                                    let a0 = aj;
                                    while (a0 - az > ay) {
                                        a0 -= az;
                                        if (Math.abs(a0 - ay) < 0.0001) {
                                            break;
                                        }
                                        this.io(j, a0, c, m, false);
                                    }
                                }
                                else {
                                    let a1 = aj;
                                    while (a1 + az < ay) {
                                        a1 += az;
                                        if (Math.abs(a1 - ay) < 0.0001) {
                                            break;
                                        }
                                        this.io(j, a1, c, m, false);
                                    }
                                }
                                this.mp = this.mu;
                            }
                            else {
                                for (let a2 = 0; a2 < truncate(ad.g); a2++) {
                                    for (let a3 = 0; a3 < this.li; a3++) {
                                        let a4 = this.getGroupCenter(a3, b, c, d);
                                        if (this.ch) {
                                            a4 = -a4;
                                        }
                                        let a5 = this.ek(a2 + ax * ad.g, f) + a4;
                                        this.io(j, a5, c, m, false);
                                    }
                                }
                                this.mp = 0.5;
                            }
                        }
                        let a6 = this.mq(ax * as, aj, aw == 0, f);
                        let a7 = truncate(Math.round(a6));
                        if (a7 >= an && a7 <= am) {
                            let a8 = 0;
                            if (as >= 1) {
                                a8 = ax * truncate(Math.floor(as));
                            }
                            else {
                                if ((ax * as) * 2 % 2 == 0) {
                                    a8 = truncate(Math.floor(ax * as));
                                }
                                else {
                                    a8 = -1;
                                }
                            }
                            if (o != null && a8 < n.count && a8 >= 0) {
                                let a9 = o.getItem(a8);
                                let ba = this.f0(a9);
                                this.mz = a9;
                                if (!isNaN_(a6) && !isInfinity(a6) && ba != null) {
                                    if ((typeof ba === 'string') && Base.equalsStatic(ba, "")) {
                                    }
                                    else {
                                        this.dc.add1(ba);
                                        this.dd.add(new LabelPosition(a6));
                                    }
                                }
                            }
                        }
                        aj = ay;
                    }
                }
                this.mz = null;
                if ((this.az == null || this.az.visibility == 0) && this.aj != null) {
                    if (this.az != null && (this.az.location == 4 || this.az.location == 5)) {
                        this.bi.nr();
                    }
                }
                this.a5.a0(this.dc, this.dd);
                this.l6();
            }
        }
        c3() {
            if (this.cq && this.c4) {
                this.bc = this.bb();
            }
            if (this.ks == null) {
                return false;
            }
            let a = false;
            let b = this.ks.count;
            if (b != this.mx) {
                let c = new AxisRangeChangedEventArgs(0, 1, 1, this.mx, b);
                this.mx = b;
                this.h4(c);
                a = true;
            }
            this.nc();
            this.nb();
            return a;
        }
        getWindowZoomFromCategories(a) {
            let b = MathUtil.a(a) && a > 0;
            let c = this.ks.count;
            if (c > 0 && b) {
                let d = this.bi.v0.width;
                let e = isNaN_(d) || isInfinity(d);
                if (e) {
                    return NaN;
                }
                let f = new Rect(0, 0, 0, 1, 1);
                let g = this.bi.v0;
                let h = this.getCategorySize(f, g, g);
                let i = isNaN_(h) || isInfinity(h);
                if (i) {
                    return NaN;
                }
                let j = h * a / d;
                if (j >= 0 && j <= 1) {
                    return j;
                }
            }
            return NaN;
        }
        getWindowZoomFromItemSpan(a) {
            let b = MathUtil.a(a) && a > 0;
            let c = this.ks.count;
            if (c > 0 && b) {
                let d = this.bi.v0.width;
                let e = isNaN_(d) || isInfinity(d);
                if (e) {
                    return NaN;
                }
                let f = new Rect(0, 0, 0, 1, 1);
                let g = this.bi.v0;
                let h = this.k5(f, g, g);
                let i = a / h;
                let j = Math.min(1, 1 / i);
                if (j > 0 && j <= 1) {
                    return j;
                }
            }
            return NaN;
        }
        nc() {
            if (this.ks == null || this.bi == null) {
                return;
            }
            let a = !isNaN_(this.bi.il) && !isNaN_(this.bi.io);
            if (a) {
                return;
            }
            if (!isNaN_(this.zoomMaximumItemSpan) && !isInfinity(this.zoomMaximumItemSpan)) {
                let b = this.getWindowZoomFromItemSpan(this.zoomMaximumItemSpan);
                if (!isNaN_(b) && !isInfinity(b)) {
                    this.bi.hf = b;
                }
            }
            else if (!isNaN_(this.zoomMaximumCategoryRange) && !isInfinity(this.zoomMaximumCategoryRange)) {
                let c = this.getWindowZoomFromCategories(this.zoomMaximumCategoryRange);
                if (!isNaN_(c) && !isInfinity(c)) {
                    this.bi.hf = c;
                }
            }
        }
        nb() {
            if (this.ks == null || this.bi == null) {
                return;
            }
            let a = NaN;
            let b = NaN;
            let c = this.bi.hl();
            let d = this.bi.hm();
            let e = false;
            if (MathUtil.a(this.zoomToItemSpan) && this.zoomToItemSpan > 0) {
                e = true;
                c = this.getWindowZoomFromItemSpan(this.zoomToItemSpan);
                c = Math.max(c, this.bi.hf);
            }
            else if (MathUtil.a(this.zoomToCategoryRange) && this.zoomToCategoryRange > 0 && this.zoomToCategoryRange <= this.lh) {
                e = true;
                c = this.getWindowZoomFromCategories(this.zoomToCategoryRange);
                c = Math.max(c, this.bi.hf);
            }
            if (MathUtil.a(this.zoomToCategoryStart) && this.zoomToCategoryStart >= 0 && this.zoomToCategoryStart < this.lh) {
                e = true;
                a = this.getWindowZoomFromCategories(this.zoomToCategoryStart);
                let f = c + a - 1;
                let g = c - f;
                if (f > 0 && g > this.bi.hf) {
                    c = g;
                    c = Math.max(c, this.bi.hf);
                }
            }
            if (e) {
                this.bi.ri(a, b, c, d);
            }
        }
        hx(a, b, c, d) {
            super.hx(a, b, c, d);
            switch (b) {
                case CategoryXAxis.$$p[0]:
                case CategoryXAxis.$$p[1]:
                    this.nc();
                    break;
                case CategoryXAxis.$$p[2]:
                case CategoryXAxis.$$p[3]:
                case CategoryXAxis.$$p[4]:
                    this.nb();
                    break;
            }
        }
        get mt() {
            return this.c(CategoryXAxis.nd);
        }
        set mt(a) {
            this.h(CategoryXAxis.nd, a);
        }
        ml() {
            return !isNaN_(this.mt);
        }
        get mo() {
            return this.mm;
        }
        set mo(a) {
            if (this.mo != a) {
                let b = this.mm;
                this.mm = a;
                this.h3("ActualInterval", b, this.mo);
            }
        }
        ew() {
            return this.mo;
        }
        get mu() {
            return this.c(CategoryXAxis.ne);
        }
        set mu(a) {
            this.h(CategoryXAxis.ne, a);
        }
        get mp() {
            return this.mn;
        }
        set mp(a) {
            if (this.mp != a) {
                let b = this.mn;
                this.mn = a;
                this.h3("ActualMinorInterval", b, this.mp);
            }
        }
        kv(a) {
            if (a == null) {
                return false;
            }
            let b = this.be();
            if (b == null) {
                return false;
            }
            return b.synchronizeHorizontally;
        }
        get_a2() {
            return 0;
        }
        get a2() {
            return this.get_a2();
        }
        mk() {
            return this.ks;
        }
        mr(a, b) {
            return this.ek(a, b);
        }
        mq(a, b, c, d) {
            if (!c) {
                let e = this.mr(a + 1, d);
                return (b + e) / 2;
            }
            return b;
        }
        get isVertical() {
            return this.cp;
        }
        j$d(a, b) {
            return this.en.apply(this, arguments);
        }
        j$c(a, b) {
            return this.ek.apply(this, arguments);
        }
        j$e(a, b, c, d, e) {
            this.hn.apply(this, arguments);
        }
        j$f(a, b, c, d) {
            this.ho.apply(this, arguments);
        }
        j$g(a, b, c, d) {
            this.hp.apply(this, arguments);
        }
        get j$b() {
            return this.ch;
        }
        j$a() {
            return this.ag.apply(this, arguments);
        }
    }
    CategoryXAxis.$t = markType(CategoryXAxis, 'CategoryXAxis', CategoryAxisBase.$, [IZoomableCategoryAxis_$type]);
    CategoryXAxis.nd = DependencyProperty.i("Interval", Number_$type, CategoryXAxis.$, new PropertyMetadata(2, NaN, (a, b) => {
        a.h3("Interval", b.oldValue, b.newValue);
        a.h9(false);
    }));
    CategoryXAxis.ne = DependencyProperty.i("MinorInterval", Number_$type, CategoryXAxis.$, new PropertyMetadata(2, NaN, (a, b) => {
        (typeCast(CategoryXAxis.$, a)).h3("MinorInterval", b.oldValue, b.newValue);
        (typeCast(CategoryXAxis.$, a)).h9(false);
    }));
    CategoryXAxis.$$p = markDep(DependencyProperty, PropertyMetadata, CategoryXAxis, 'h3', ['ZoomMaximumCategoryRange::nf', [1, NaN], 'ZoomMaximumItemSpan::ng', [1, NaN], 'ZoomToCategoryRange::nh', [1, NaN], 'ZoomToCategoryStart::ni', [1, NaN], 'ZoomToItemSpan::nj', [1, NaN]]);
    return CategoryXAxis;
})();
/**
 * @hidden
 */
export let CategoryXAxisView = /*@__PURE__*/ (() => {
    class CategoryXAxisView extends CategoryAxisBaseView {
        constructor(a) {
            super(a);
            this.bm = null;
            this.bm = a;
        }
        al() {
            super.al();
            this.a.fi = AxisDefaults.f;
        }
    }
    CategoryXAxisView.$t = markType(CategoryXAxisView, 'CategoryXAxisView', CategoryAxisBaseView.$);
    return CategoryXAxisView;
})();
