// igniteui-angular-charts/esm2015/lib/ChaikinOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { ChaikinOscillatorIndicatorStrategy } from "./ChaikinOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let ChaikinOscillatorIndicator = /*@__PURE__*/ (() => {
    class ChaikinOscillatorIndicator extends StrategyBasedIndicator {
        bi() {
            return new ChaikinOscillatorIndicator();
        }
        get_abn() {
            return new ChaikinOscillatorIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return ChaikinOscillatorIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        get shortPeriod() {
            return typeGetValue(this.c(ChaikinOscillatorIndicator.ab9));
        }
        set shortPeriod(a) {
            this.h(ChaikinOscillatorIndicator.ab9, a);
        }
        abv() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(ChaikinOscillatorIndicator.ab8));
        }
        set longPeriod(a) {
            this.h(ChaikinOscillatorIndicator.ab8, a);
        }
        abs() {
            return this.longPeriod;
        }
    }
    ChaikinOscillatorIndicator.$t = markType(ChaikinOscillatorIndicator, 'ChaikinOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    ChaikinOscillatorIndicator.ab9 = StrategyBasedIndicator.ab7(3, ChaikinOscillatorIndicator.$);
    ChaikinOscillatorIndicator.ab8 = StrategyBasedIndicator.ab3(10, ChaikinOscillatorIndicator.$);
    return ChaikinOscillatorIndicator;
})();
