// igniteui-angular-charts/esm2015/lib/CustomIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { CustomIndicatorStrategy } from "./CustomIndicatorStrategy";
import { String_$type, fromEnum, markType } from "igniteui-angular-core";
import { FinancialEventArgs } from "./FinancialEventArgs";
import { List$1 } from "igniteui-angular-core";
/**
 * @hidden
 */
export let CustomIndicator = /*@__PURE__*/ (() => {
    class CustomIndicator extends StrategyBasedIndicator {
        constructor() {
            super(...arguments);
            this.ab9 = null;
            this.ab8 = null;
        }
        bi() {
            return new CustomIndicator();
        }
        get_abn() {
            return new CustomIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return CustomIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        aaz(a, b) {
            let c = this.xb(a, b);
            let d = this.xc(c);
            let e = new List$1(String_$type, 0);
            if (this.basedOnColumns != null) {
                let f = new FinancialEventArgs(c.calculateFrom, c.calculateCount, c, d, this);
                this.basedOnColumns(this, f);
                if (f.basedOn != null) {
                    for (let g of fromEnum(f.basedOn)) {
                        e.add(g);
                    }
                }
            }
            return e;
        }
        get indicator() {
            return this.ab9;
        }
        set indicator(a) {
            this.ab9 = a;
            this.aa7();
        }
        get basedOnColumns() {
            return this.ab8;
        }
        set basedOnColumns(a) {
            this.ab8 = a;
        }
        aaw(a, b) {
            super.aaw(a, b);
            if (this.indicator != null) {
                let c = this.xb(a, b);
                if (b == 0) {
                    return false;
                }
                if (!this.x7(this.aaz(a, b))) {
                    return false;
                }
                this.indicator(this, new FinancialEventArgs(a, b, c, this.xc(c), this));
                if (this.abp(c) && this.xh != null) {
                    this.xh.c1();
                }
                return true;
            }
            return false;
        }
    }
    CustomIndicator.$t = markType(CustomIndicator, 'CustomIndicator', StrategyBasedIndicator.$);
    return CustomIndicator;
})();
