// igniteui-angular-charts/esm2015/lib/DetrendedPriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { DetrendedPriceOscillatorIndicatorStrategy } from "./DetrendedPriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let DetrendedPriceOscillatorIndicator = /*@__PURE__*/ (() => {
    class DetrendedPriceOscillatorIndicator extends StrategyBasedIndicator {
        bi() {
            return new DetrendedPriceOscillatorIndicator();
        }
        get_abn() {
            return new DetrendedPriceOscillatorIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return DetrendedPriceOscillatorIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        get ab8() {
            return typeGetValue(this.c(DetrendedPriceOscillatorIndicator.ab9));
        }
        set ab8(a) {
            this.h(DetrendedPriceOscillatorIndicator.ab9, a);
        }
        abt() {
            return this.ab8;
        }
        ab2(a) {
            this.ab8 = a;
        }
    }
    DetrendedPriceOscillatorIndicator.$t = markType(DetrendedPriceOscillatorIndicator, 'DetrendedPriceOscillatorIndicator', StrategyBasedIndicator.$);
    DetrendedPriceOscillatorIndicator.ab9 = StrategyBasedIndicator.ab5(20, DetrendedPriceOscillatorIndicator.$);
    return DetrendedPriceOscillatorIndicator;
})();
