// igniteui-angular-charts/esm2015/lib/FastStochasticOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { FastStochasticOscillatorIndicatorStrategy } from "./FastStochasticOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
import { Brush } from "igniteui-angular-core";
/**
 * @hidden
 */
export let FastStochasticOscillatorIndicator = /*@__PURE__*/ (() => {
    class FastStochasticOscillatorIndicator extends StrategyBasedIndicator {
        constructor() {
            super();
            this.trendLineBrush = ((() => {
                let $ret = new Brush();
                $ret.fill = "#565656";
                return $ret;
            })());
            this.trendLineType = 10;
            this.trendLinePeriod = 3;
        }
        bi() {
            return new FastStochasticOscillatorIndicator();
        }
        get_abn() {
            return new FastStochasticOscillatorIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return FastStochasticOscillatorIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        get ab8() {
            return typeGetValue(this.c(FastStochasticOscillatorIndicator.ab9));
        }
        set ab8(a) {
            this.h(FastStochasticOscillatorIndicator.ab9, a);
        }
        abt() {
            return this.ab8;
        }
        ab2(a) {
            this.ab8 = a;
        }
    }
    FastStochasticOscillatorIndicator.$t = markType(FastStochasticOscillatorIndicator, 'FastStochasticOscillatorIndicator', StrategyBasedIndicator.$);
    FastStochasticOscillatorIndicator.ab9 = StrategyBasedIndicator.ab5(14, FastStochasticOscillatorIndicator.$);
    return FastStochasticOscillatorIndicator;
})();
