// igniteui-angular-charts/esm2015/lib/FinancialPriceSeriesProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-angular-core";
import { IFinancialPriceSeriesProxy_$type } from "./IFinancialPriceSeriesProxy";
/**
 * @hidden
 */
export let FinancialPriceSeriesProxy = /*@__PURE__*/ (() => {
    class FinancialPriceSeriesProxy extends Base {
        getCloseMemberPath(a) {
            return a.y2;
        }
        getHighMemberPath(a) {
            return a.zf;
        }
        getLowMemberPath(a) {
            return a.zk;
        }
        getOpenMemberPath(a) {
            return a.zo;
        }
        getVolumeMemberPath(a) {
            return a.zs;
        }
        matchesType(a) {
            if (!a.e4) {
                return false;
            }
            if (a.e5 || a.e6) {
                return false;
            }
            return true;
        }
        setCategoryMode(a, b) {
            a.w6 = b;
        }
        setCloseMemberPath(a, b) {
            a.y2 = b;
        }
        setDisplayType(a, b) {
            a.aau = b;
        }
        setHighMemberPath(a, b) {
            a.zf = b;
        }
        setLowMemberPath(a, b) {
            a.zk = b;
        }
        setOpenMemberPath(a, b) {
            a.zo = b;
        }
        setVolumeMemberPath(a, b) {
            a.zs = b;
        }
        setXAxis(a, b) {
            a.wy = b;
        }
        setYAxis(a, b) {
            a.xh = b;
        }
        setTrendLineType(a, b) {
            if (a.e7) {
                a.trendLineType = b;
            }
        }
        getTrendLineType(a) {
            if (a.e7) {
                return a.trendLineType;
            }
            return 0;
        }
    }
    FinancialPriceSeriesProxy.$t = markType(FinancialPriceSeriesProxy, 'FinancialPriceSeriesProxy', Base.$, [IFinancialPriceSeriesProxy_$type]);
    return FinancialPriceSeriesProxy;
})();
