// igniteui-angular-charts/esm2015/lib/FinancialOverlay.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FinancialSeries } from "./FinancialSeries";
import { DependencyProperty } from "igniteui-angular-core";
import { ISortingAxis_$type } from "./ISortingAxis";
import { typeCast, markType, markDep } from "igniteui-angular-core";
import { Rect } from "igniteui-angular-core";
import { ScalerParams } from "./ScalerParams";
import { AxisRange } from "./AxisRange";
import { PropertyMetadata } from "igniteui-angular-core";
import { isNaN_ } from "igniteui-angular-core";
/**
 * @hidden
 */
export let FinancialOverlay = /*@__PURE__*/ (() => {
    class FinancialOverlay extends FinancialSeries {
        constructor() {
            super();
            this.aas = false;
            this.aas = false;
        }
        get_e6() {
            return true;
        }
        o9(a, b, c, d) {
            super.o9(a, b, c, d);
            if (this.wy != null && typeCast(ISortingAxis_$type, this.wy) !== null) {
                this.wy.i$i();
            }
            this.xf.b0.e(this.ib);
            this.aat();
            if (this.xh != null) {
                this.xh.c1();
            }
            this.rn(true);
        }
        get_hb() {
            return true;
        }
        q5(a, b, c, d) {
            switch (b) {
                case "FastItemsSource":
                    this.aas = false;
                    break;
                case FinancialOverlay.$$p[0]:
                    this.aas = false;
                    this.rn(false);
                    this.qh();
                    break;
                case "XAxis":
                    this.aas = false;
                    break;
            }
            super.q5(a, b, c, d);
        }
        b0() {
            if (this.b1 == 0) {
                return 1;
            }
            else {
                return super.b0();
            }
        }
        gu(a, b) {
            if (this.gt(a, b)) {
                return true;
            }
            return false;
        }
        ac() {
            super.ac();
            this.rn(false);
        }
        gb(a) {
            let b = this.dl != null ? this.dl.indexOf(a) : -1;
            let c = this.ct.by;
            c = c.copy();
            let d = this.ct.bx;
            let e = new Rect(0, 0, 0, 1, 1);
            let f = this.wf(this.ct);
            let g = new ScalerParams(0, e, e, this.wy.ch, f);
            let h = new ScalerParams(0, e, e, this.xh.ch, f);
            h.b = this.iw();
            if (b >= 0 && Rect.l_op_Inequality(c, null) && Rect.l_op_Inequality(d, null)) {
                if (this.wy != null) {
                    let i = this.wy.ek(b, g);
                    if (i < c.left + 0.1 * c.width) {
                        i = i + 0.4 * c.width;
                    }
                    if (i > c.right - 0.1 * c.width) {
                        i = i - 0.4 * c.width;
                    }
                    c.x = i - 0.5 * c.width;
                }
                if (this.xh != null && this.xt != null && this.xs != null && b < this.xt.count && b < this.xs.count) {
                    let j = this.xh.ek(this.xt.item(b), h);
                    let k = this.xh.ek(this.xs.item(b), h);
                    if (!isNaN_(j) && !isNaN_(k)) {
                        let l = Math.abs(j - k);
                        if (c.height < l) {
                            c.height = l;
                            c.y = Math.min(j, k);
                        }
                        else {
                            if (j < c.top + 0.1 * c.height) {
                                j = j + 0.4 * c.height;
                            }
                            if (j > c.bottom - 0.1 * c.height) {
                                j = j - 0.4 * c.height;
                            }
                            c.y = j - 0.5 * c.height;
                        }
                    }
                }
                this.c1.au(this.c9, c, true);
            }
            return b >= 0;
        }
        ax(a) {
            if (a != null && a == this.xh && this.xt != null && this.xs != null) {
                return new AxisRange(this.xt.minimum, this.xs.maximum);
            }
            return null;
        }
    }
    FinancialOverlay.$t = markType(FinancialOverlay, 'FinancialOverlay', FinancialSeries.$);
    FinancialOverlay.$$p = markDep(DependencyProperty, PropertyMetadata, FinancialOverlay, 'raisePropertyChanged', ['IgnoreFirst:aau:aaw', [1, 0]]);
    return FinancialOverlay;
})();
