// igniteui-angular-charts/esm2015/lib/HorizontalRangeCategorySeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RangeCategorySeries } from "./RangeCategorySeries";
import { CategoryAxisBase } from "./CategoryAxisBase";
import { DependencyProperty } from "igniteui-angular-core";
import { NumericYAxis } from "./NumericYAxis";
import { Axis } from "./Axis";
import { Point_$type, runOn, typeCast, markType, markDep } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { RangeValueList } from "./RangeValueList";
import { ScalerParams } from "./ScalerParams";
import { CategoryXAxis } from "./CategoryXAxis";
import { DataLegendSeriesValueInfo } from "igniteui-angular-core";
import { DataLegendSeriesContext } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
/**
 * @hidden
 */
export let HorizontalRangeCategorySeries = /*@__PURE__*/ (() => {
    class HorizontalRangeCategorySeries extends RangeCategorySeries {
        fetchXAxis() {
            return this.xAxis;
        }
        fetchYAxis() {
            return this.yAxis;
        }
        o0() {
            super.o0();
            this.xAxis = null;
            this.yAxis = null;
        }
        aj(a) {
            let b = this.am(this.yk(), this.yl.f, this.zm(this.ct), this.zn(this.ct), this.wb(a), true);
            if (b == null) {
                return null;
            }
            let c = b[0];
            let d = b[1];
            let e = new Array(1);
            let f = new Array(2);
            f[0] = { $type: Point_$type, x: c[0], y: c[2] };
            f[1] = { $type: Point_$type, x: d[0], y: d[2] };
            e[0] = f;
            return e;
        }
        ak(a) {
            let b = this.am(this.yk(), this.yl.f, this.zm(this.ct), this.zn(this.ct), this.wb(a), true);
            if (b == null) {
                return null;
            }
            let c = b[0];
            let d = b[1];
            let e = new Array(1);
            let f = new Array(2);
            f[0] = { $type: Point_$type, x: c[0], y: c[1] };
            f[1] = { $type: Point_$type, x: d[0], y: d[1] };
            e[0] = f;
            return e;
        }
        get_fs() {
            return true;
        }
        b0() {
            if (this.b1 == 0) {
                return 1;
            }
            else {
                return super.b0();
            }
        }
        getOffsetValue() {
            return this.RangeFramePreparer.r(this.xAxis, this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct));
        }
        getCategoryWidth() {
            return this.xAxis.getCategorySize(this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct));
        }
        j6(a, b) {
            return this.j7(a, b, this.xAxis, runOn(this, this.zk), new RangeValueList(this.highColumn, this.lowColumn));
        }
        j8(a, b) {
            return this.j9(a, b, this.xAxis, runOn(this, this.zk), new RangeValueList(this.highColumn, this.lowColumn));
        }
        im(a, b, c, d, e) {
            if (c == null) {
                return Number.POSITIVE_INFINITY;
            }
            let f = this.xAxis.lg;
            return this.io(a, b, this.xAxis, d, e, f, runOn(this, this.zk));
        }
        i3(a, b, c) {
            if (this.c9 == null) {
                return NaN;
            }
            let d = this.getEffectiveViewport1(this.ct);
            let e = new ScalerParams(0, this.c9.vl, this.ct.bx, this.xAxis.ch, d);
            let f = this.RangeFramePreparer.r(this.xAxis, this.c9.vl, this.ct.bx, d);
            return this.i5(new RangeValueList(this.highColumn, this.lowColumn), a, this.xAxis, e, f, runOn(this, this.zk), b, c);
        }
        i1(a, b, c) {
            if (this.c9 == null) {
                return NaN;
            }
            let d = this.getEffectiveViewport1(this.ct);
            let e = new ScalerParams(0, this.c9.vl, this.ct.bx, this.xAxis.ch, d);
            let f = this.RangeFramePreparer.r(this.xAxis, this.c9.vl, this.ct.bx, d);
            return this.i5(this.lowColumn, a, this.xAxis, e, f, runOn(this, this.zk), b, c);
        }
        iz(a, b, c) {
            if (this.c9 == null) {
                return NaN;
            }
            let d = this.getEffectiveViewport1(this.ct);
            let e = new ScalerParams(0, this.c9.vl, this.ct.bx, this.xAxis.ch, d);
            let f = this.RangeFramePreparer.r(this.xAxis, this.c9.vl, this.ct.bx, d);
            return this.i5(this.highColumn, a, this.xAxis, e, f, runOn(this, this.zk), b, c);
        }
        v2(a, b, c) {
            return this.v8(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct)), this.yAxis, this.xAxis, runOn(this, this.iz), (d, e) => this.j9(d, e, this.xAxis, runOn(this, this.zk), this.highColumn), (d, e) => this.j7(d, e, this.xAxis, runOn(this, this.zk), this.highColumn));
        }
        v4(a, b, c) {
            return this.v8(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct)), this.yAxis, this.xAxis, runOn(this, this.i1), (d, e) => this.j9(d, e, this.xAxis, runOn(this, this.zk), this.lowColumn), (d, e) => this.j7(d, e, this.xAxis, runOn(this, this.zk), this.lowColumn));
        }
        v6(a, b, c) {
            return this.v8(a, b, c, this.RangeFramePreparer.r(this.xAxis, this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct)), this.yAxis, this.xAxis, null, null, null);
        }
        zg() {
            return this.yAxis != null && this.yAxis.c1();
        }
        q5(a, b, c, d) {
            super.q5(a, b, c, d);
            switch (b) {
                case HorizontalRangeCategorySeries.$$p[0]:
                case HorizontalRangeCategorySeries.$$p[1]:
                case HorizontalRangeCategorySeries.$$p[2]:
                case HorizontalRangeCategorySeries.$$p[3]:
                    this.qb();
                    break;
                case HorizontalRangeCategorySeries.$$p[4]:
                    this.pa(typeCast(Axis.$, c));
                    this.rd(typeCast(Axis.$, d));
                    this.yu.cw.g(this.ib);
                    this.rn(false);
                    this.qh();
                    break;
                case HorizontalRangeCategorySeries.$$p[5]:
                    this.pa(typeCast(Axis.$, c));
                    this.rd(typeCast(Axis.$, d));
                    this.yu.cw.g(this.ib);
                    this.zg();
                    this.rn(false);
                    this.qh();
                    break;
            }
        }
        aa2(a) {
            if (typeCast(NumericYAxis.$, a) !== null) {
                return true;
            }
            return false;
        }
        aa1(a) {
            if (typeCast(CategoryXAxis.$, a) !== null || a.ce) {
                return true;
            }
            return false;
        }
        getDataLegendSeriesContextAt(a, b) {
            let c = { $type: Point_$type, x: a, y: b };
            let d = this.i1(c, false, false);
            let e = this.iz(c, false, false);
            let f = new DataLegendSeriesValueInfo();
            f.a = 1;
            f.l = e > d ? d : e;
            f.q = "Low";
            f.r = this.aaq != null ? this.aaq : "Low";
            f.s = this.aa7 != null ? this.aa7 : "L:";
            f.t = this.aa9 != null ? this.aa9 : "";
            let g = new DataLegendSeriesValueInfo();
            g.a = 2;
            g.l = e > d ? e : d;
            g.q = "High";
            g.r = this.aan != null ? this.aan : "High";
            g.s = this.aa3 != null ? this.aa3 : "H:";
            g.t = this.aa5 != null ? this.aa5 : "";
            let h = new DataLegendSeriesValueInfo();
            h.a = 7;
            h.l = Math.abs(g.l - f.l);
            h.q = "Range";
            h.r = "Range";
            h.s = "R:";
            h.k = true;
            let i = new DataLegendSeriesValueInfo();
            i.a = 6;
            i.l = f.l + (h.l / 2);
            i.q = "Average";
            i.r = "Average";
            i.s = "A:";
            i.k = true;
            let j = new DataLegendSeriesContext();
            j.e = ((() => {
                let $ret = new List$1(DataLegendSeriesValueInfo.$, 0);
                $ret.add(g);
                $ret.add(f);
                $ret.add(h);
                $ret.add(i);
                return $ret;
            })());
            j.h = this.ma();
            return j;
        }
        get useHighMarkerFidelity() {
            return this.zh;
        }
    }
    HorizontalRangeCategorySeries.$t = markType(HorizontalRangeCategorySeries, 'HorizontalRangeCategorySeries', RangeCategorySeries.$);
    HorizontalRangeCategorySeries.$$p = markDep(DependencyProperty, PropertyMetadata, HorizontalRangeCategorySeries, 'raisePropertyChanged', ['HighMemberAsLegendLabel:aa3:abd', [2, null], 'HighMemberAsLegendUnit:aa5:abe', [2, null], 'LowMemberAsLegendLabel:aa7:abf', [2, null], 'LowMemberAsLegendUnit:aa9:abg', [2, null], 'XAxis::abh', [CategoryAxisBase.$, null], 'YAxis::abi', [NumericYAxis.$, null]]);
    return HorizontalRangeCategorySeries;
})();
