// igniteui-angular-charts/esm2015/lib/HorizontalAnchoredCategorySeriesProxy.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, typeCast, markType } from "igniteui-angular-core";
import { IHorizontalAnchoredCategorySeriesProxy_$type } from "./IHorizontalAnchoredCategorySeriesProxy";
import { HorizontalAnchoredCategorySeries } from "./HorizontalAnchoredCategorySeries";
/**
 * @hidden
 */
export let HorizontalAnchoredCategorySeriesProxy = /*@__PURE__*/ (() => {
    class HorizontalAnchoredCategorySeriesProxy extends Base {
        matchesType(a) {
            return typeCast(HorizontalAnchoredCategorySeries.$, a) !== null;
        }
        setCategoryMode(a, b) {
            a.abx = b;
        }
        setValueMemberPath(a, b) {
            a.aa4 = b;
        }
        setXAxis(a, b) {
            a.xAxis = b;
        }
        setYAxis(a, b) {
            a.yAxis = b;
        }
        setTrendLineType(a, b) {
            a.trendLineType = b;
        }
        setMarkerType(a, b) {
            a.w2 = b;
        }
        getMarkerType(a) {
            return a.w0;
        }
        getTrendLineType(a) {
            return a.trendLineType;
        }
    }
    HorizontalAnchoredCategorySeriesProxy.$t = markType(HorizontalAnchoredCategorySeriesProxy, 'HorizontalAnchoredCategorySeriesProxy', Base.$, [IHorizontalAnchoredCategorySeriesProxy_$type]);
    return HorizontalAnchoredCategorySeriesProxy;
})();
