// igniteui-angular-charts/esm2015/lib/igx-data-chart-financial-overlays-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxOverlaysModule } from './igx-overlays-module';
import { IgxDataChartFinancialCoreModule } from './igx-data-chart-financial-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartFinancialOverlaysModule = /*@__PURE__*/ (() => {
    class IgxDataChartFinancialOverlaysModule {
    }
    IgxDataChartFinancialOverlaysModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialOverlaysModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartFinancialOverlaysModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialOverlaysModule, imports: [CommonModule, IgxOverlaysModule,
            IgxDataChartFinancialCoreModule], exports: [IgxOverlaysModule,
            IgxDataChartFinancialCoreModule] });
    IgxDataChartFinancialOverlaysModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialOverlaysModule, imports: [[CommonModule, IgxOverlaysModule,
                IgxDataChartFinancialCoreModule
            ], IgxOverlaysModule,
            IgxDataChartFinancialCoreModule] });
    return IgxDataChartFinancialOverlaysModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialOverlaysModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxOverlaysModule,
                        IgxDataChartFinancialCoreModule
                    ],
                    imports: [CommonModule, IgxOverlaysModule,
                        IgxDataChartFinancialCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
