// igniteui-angular-charts/esm2015/lib/igx-stacked-100-bar-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStacked100BarSeriesComponent } from './igx-stacked-100-bar-series-component';
import * as i0 from "@angular/core";
export let IgxStacked100BarSeriesModule = /*@__PURE__*/ (() => {
    class IgxStacked100BarSeriesModule {
    }
    IgxStacked100BarSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStacked100BarSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesModule, declarations: [IgxStacked100BarSeriesComponent], imports: [CommonModule], exports: [IgxStacked100BarSeriesComponent] });
    IgxStacked100BarSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesModule, imports: [[CommonModule]] });
    return IgxStacked100BarSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStacked100BarSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxStacked100BarSeriesComponent],
                    entryComponents: []
                }]
        }] });
