// igniteui-angular-charts/esm2015/lib/igx-chaikin-volatility-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxChaikinVolatilityIndicatorModule } from './igx-chaikin-volatility-indicator-module';
import { IgxChaikinVolatilityIndicatorComponent } from './igx-chaikin-volatility-indicator-component';
import { ChaikinVolatilityIndicator } from './ChaikinVolatilityIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxChaikinVolatilityIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxChaikinVolatilityIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxChaikinVolatilityIndicatorComponent", IgxChaikinVolatilityIndicatorComponent);
            TypeRegistrar.register("ChaikinVolatilityIndicator", ChaikinVolatilityIndicator.$type);
        }
    }
    IgxChaikinVolatilityIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinVolatilityIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxChaikinVolatilityIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinVolatilityIndicatorDynamicModule, imports: [CommonModule, IgxChaikinVolatilityIndicatorModule], exports: [IgxChaikinVolatilityIndicatorModule] });
    IgxChaikinVolatilityIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinVolatilityIndicatorDynamicModule, imports: [[CommonModule, IgxChaikinVolatilityIndicatorModule], IgxChaikinVolatilityIndicatorModule] });
    return IgxChaikinVolatilityIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinVolatilityIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxChaikinVolatilityIndicatorModule],
                    exports: [IgxChaikinVolatilityIndicatorModule],
                    entryComponents: [IgxChaikinVolatilityIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
