// igniteui-angular-charts/esm2015/lib/igx-scatter-series-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxScatterBaseComponent } from "./igx-scatter-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { ScatterSeries } from "./ScatterSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent scatter series
*/
export let IgxScatterSeriesComponent = /*@__PURE__*/ (() => {
    class IgxScatterSeriesComponent extends IgxScatterBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new ScatterSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the series has only marker as visuals
        */
        get hasOnlyMarkers() {
            return this.i.eg;
        }
    }
    IgxScatterSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxScatterSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxScatterSeriesComponent, selector: "igx-scatter-series", providers: [{ provide: IgxScatterBaseComponent, useExisting: forwardRef(() => IgxScatterSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxScatterSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxScatterSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxScatterSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-scatter-series',
                    template: ``,
                    providers: [{ provide: IgxScatterBaseComponent, useExisting: forwardRef(() => IgxScatterSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxScatterSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxScatterSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
