// igniteui-angular-charts/esm2015/lib/igx-numeric-axis-base-component.js
import { EventEmitter, Output, Component, Input } from '@angular/core';
import { delegateCombine } from "igniteui-angular-core";
import { AxisRangeBufferMode_$type } from "./AxisRangeBufferMode";
import { IgxAxisComponent } from "./igx-axis-component";
import { ensureBool, ensureEnum, arrayFindByName } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents the base class for all IgxDataChartComponent numeric axes.
*/
export let IgxNumericAxisBaseComponent = /*@__PURE__*/ (() => {
    class IgxNumericAxisBaseComponent extends IgxAxisComponent {
        constructor() {
            super();
            this._actualMinimumValueChange = null;
            this._actualMaximumValueChange = null;
            this._actualIntervalChange = null;
            this._actualMinorIntervalChange = null;
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets or sets if the current axis is of numeric axis type
        */
        get isNumeric() {
            return this.i.cj;
        }
        /**
         * Gets or sets how the numeric axis will adjust its range buffer to less closely fix the data from the series.
        */
        get autoRangeBufferMode() {
            return this.i.kv;
        }
        set autoRangeBufferMode(v) {
            this.i.kv = ensureEnum(AxisRangeBufferMode_$type, v);
        }
        /**
         * Gets or sets the MinimumValue property.
        *
        * Used to set the minimum value of a numeric axis. To get the minimum value in effect on a numeric axis use `ActualMinimumValue`.
        *
        * ```ts
        * this.yAxis.minimumValue = 0;
        * ```
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         minimumValue=0
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        */
        get minimumValue() {
            return this.i.l6;
        }
        set minimumValue(v) {
            this.i.l6 = +v;
        }
        /**
         * Gets the effective minimum value for the current numeric axis object.
        *
        * ```ts
        * let effectiveMinValue: number = yAxis.actualMinimumValue;
        * ```
        */
        get actualMinimumValue() {
            return this.i.ly;
        }
        set actualMinimumValue(v) {
            this.i.ly = +v;
        }
        /**
         * Gets the effective minimum value for the current numeric axis object.
        *
        * The `actualVisibleMinimumValue` is used to get the current visible maximum value for the a numeric axis. Note, at times the `actualVisibleMinimumValue` may differ from the `actualMinimumValue`, for example when the chart is zoomed.
        *
        * ```ts
        * let effectiveVisibleMinValue: number = yAxis.actualVisibleMinimumValue;
        * ```
        */
        get actualVisibleMinimumValue() {
            return this.i.l1;
        }
        set actualVisibleMinimumValue(v) {
            this.i.l1 = +v;
        }
        /**
         * Gets or sets the MaximumValue property.
        *
        * Used to set the maximum value of a numeric axis. To get the maximum value in effect on a numeric axis use `ActualMaximumValue`.
        *
        * ```ts
        * this.yAxis.maximumValue = 80;
        * ```
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         maximumValue=80
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        */
        get maximumValue() {
            return this.i.l5;
        }
        set maximumValue(v) {
            this.i.l5 = +v;
        }
        /**
         * Gets the effective maximum value for the current numeric axis object.
        *
        * ```ts
        * let effectiveMaxValue: number = yAxis.actualMaximumValue;
        * ```
        */
        get actualMaximumValue() {
            return this.i.lx;
        }
        set actualMaximumValue(v) {
            this.i.lx = +v;
        }
        /**
         * Gets the effective visible maximum value for the current numeric axis object.
        *
        * The `ActualVisibleMaximumValue` is used to get the current visible maximum value for the a numeric axis. Note, at times the `ActualVisibleMaximumValue` may differ from the `ActualMaximumValue`, for example when the chart is zoomed.
        *
        * ```ts
        * let effectiveVisibleMaxValue: number = yAxis.actualVisibleMaximumValue;
        * ```
        */
        get actualVisibleMaximumValue() {
            return this.i.l0;
        }
        set actualVisibleMaximumValue(v) {
            this.i.l0 = +v;
        }
        /**
         * Gets or sets the Interval property.
        *
        * The `Interval` is used to get or set the numeric axis major interval which specifies how frequent major gridlines and axis labels are rendered on an axis.
        *
        * ```ts
        * this.yAxis.interval = 20;
        * ```
        *
        * ```html
        * <igx-data-chart
        *       [dataSource]="data">
        *       <igx-category-x-axis
        *         label="label"
        *         #xAxis
        *         >
        *       </igx-category-x-axis>
        *       <igx-numeric-y-axis
        *         interval="20"
        *         #yAxis>
        *       </igx-numeric-y-axis>
        *
        *       <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *       </igx-column-series>
        * </igx-data-chart>
        * ```
        */
        get interval() {
            return this.i.l2;
        }
        set interval(v) {
            this.i.l2 = +v;
        }
        /**
         * Gets the effective value for the current Interval.
        *
        * ```ts
        * let effectiveInterval: number = yAxis.actualInterval;
        * ```
        */
        get actualInterval() {
            return this.i.lw;
        }
        set actualInterval(v) {
            this.i.lw = +v;
        }
        /**
         * Gets or sets the MinorInterval property.
        *
        * The `MinorInterval` is used to get or set the numeric axis minor interval which specifies how frequent minor gridlines are rendered on an axis. Please note, in order for minor gridlines to be displayed the following may also need to be set: `MinorStroke`, `MinorStrokeThickness`.
        *
        * ```ts
        * this.yAxis.minorInterval = 2.5;
        * this.yAxis.minorStroke = 'blue';
        * this.yAxis.minorStrokeThickness = 1;
        * ```
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data" >
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         minorInterval= 2.5
        *         minorStroke="blue"
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        */
        get minorInterval() {
            return this.i.l7;
        }
        set minorInterval(v) {
            this.i.l7 = +v;
        }
        /**
         * Gets the effective value for the current MinorInterval.
        *
        * ```ts
        * let effectiveMinorInterval: number = yAxis.actualMinorInterval;
        * ```
        */
        get actualMinorInterval() {
            return this.i.lz;
        }
        set actualMinorInterval(v) {
            this.i.lz = +v;
        }
        /**
         * Gets or sets the ReferenceValue property.
        *
        * Used to get or set a reference value on a numeric axis which when used with an area, bar, or column series will result in the series shapes being drawn above or below the specified value.
        *
        * ```ts
        * this.yAxis.referenceValue = 30;
        * ```
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         referenceValue=30
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        */
        get referenceValue() {
            return this.i.l8;
        }
        set referenceValue(v) {
            this.i.l8 = +v;
        }
        /**
         * Gets or sets the IsLogarithmic property.
        *
        * `IsLogarithmic` is used to get or set whether the numeric axis scaler is in linear or logarithmic mode. For logarithmic scaling, this can be used in conjunction with `LogarithmBase` to further define the logarithmic scale.
        *
        * ```ts
        * this.yAxis.isLogarithmic= true;
        * this.yAxis.logarithmBase = 2;
        * ```
        *
        * ```html
        * <igx-data-chart
        *       [dataSource]="data">
        *       <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *       </igx-category-x-axis>
        *       <igx-numeric-y-axis
        *         isLogarithmic=true
        *         logarithmBase=2
        *         #yAxis>
        *       </igx-numeric-y-axis>
        *       <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *       </igx-line-series>
        * </igx-data-chart>
        * ```
        */
        get isLogarithmic() {
            return this.i.lj;
        }
        set isLogarithmic(v) {
            this.i.lj = ensureBool(v);
        }
        /**
         * Determines if the axis has a valid logarithmic scale.
        *
        * ```ts
        * let effectiveLogarithmic: boolean = this.yAxis.actualIsLogarithmic
        * ```
        */
        get actualIsLogarithmic() {
            return this.i.lc;
        }
        set actualIsLogarithmic(v) {
            this.i.lc = ensureBool(v);
        }
        /**
         * Gets or sets whether the axis should favor emitting a label at the end of the scale.
        */
        get favorLabellingScaleEnd() {
            return this.i.le;
        }
        set favorLabellingScaleEnd(v) {
            this.i.le = ensureBool(v);
        }
        /**
         * Gets or sets the LogarithmBase property.
        *
        * `LogarithmBase` is used to get or set logarithm base when a numeric axis is in Logarithmic scale mode. For logarithmic scaling, this is used in conjunction with `IsLogarithmic` or `ScaleMode` to set logarithmic scaling.
        *
        * ```ts
        * this.yAxis.scaleMode = NumericScaleMode.Logarithmic;
        * this.yAxis.logarithmBase = 2;
        * ```
        *
        * ```html
        * <igx-data-chart
        *       [dataSource]="data">
        *       <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *       </igx-category-x-axis>
        *       <igx-numeric-y-axis
        *         scaleMode="Logarithmic"
        *         logarithmBase=2
        *         #yAxis>
        *       </igx-numeric-y-axis>
        *       <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *       </igx-line-series>
        * </igx-data-chart>
        * ```
        */
        get logarithmBase() {
            return this.i.ma;
        }
        set logarithmBase(v) {
            this.i.ma = +v;
        }
        /**
         * Sets or gets a function which takes an object that produces a formatted label for displaying in the chart.
        */
        get formatAbbreviatedLabel() {
            return this.i.kq;
        }
        set formatAbbreviatedLabel(v) {
            this.i.kq = v;
        }
        /**
         * Gets or sets the label format string to use for the label.
        */
        get abbreviatedLabelFormat() {
            return this.i.mc;
        }
        set abbreviatedLabelFormat(v) {
            this.i.mc = v;
        }
        /**
         * Gets or sets the format specifiers to use with the AbbreviatedLabelFormat string.
        */
        get abbreviatedLabelFormatSpecifiers() {
            return this.i.ko;
        }
        set abbreviatedLabelFormatSpecifiers(v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                const re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.ko = v;
        }
        get isFormattingAbbreviatedLargeNumber() {
            return this.i.li;
        }
        set isFormattingAbbreviatedLargeNumber(v) {
            this.i.li = ensureBool(v);
        }
        /**
         * Determines if the axis has a user-defined minimum.
        *
        *  `HasUserMaximum` is used to determine if a numeric axis has a user definied maximum value.
        *
        * ```ts
        * let hasSetMaximum: boolean = this.yAxis.hasUserMaximum;
        * ```
        */
        get hasUserMinimum() {
            return this.i.lh;
        }
        /**
         * Determines if the axis has a user-defined maximum.
        *
        *  `HasUserMinimum` is used to determine if a numeric axis has a user definied minium value.
        *
        * ```ts
        * let hasSetMinimum: boolean = this.yAxis.hasUserMinimum;
        * ```
        */
        get hasUserMaximum() {
            return this.i.lg;
        }
        /**
         * A boolean indicating whether or not to abbreviate large numbers.
        */
        get abbreviateLargeNumbers() {
            return this.i.lb;
        }
        set abbreviateLargeNumbers(v) {
            this.i.lb = ensureBool(v);
        }
        findByName(name) {
            var baseResult = super.findByName(name);
            if (baseResult) {
                return baseResult;
            }
            if (this.abbreviatedLabelFormatSpecifiers != null && arrayFindByName(this.abbreviatedLabelFormatSpecifiers, name)) {
                return arrayFindByName(this.abbreviatedLabelFormatSpecifiers, name);
            }
            return null;
        }
        getFullRange() {
            let iv = this.i.ag();
            return (iv);
        }
        /**
         * Unscales a value from screen space into axis space.
        
        * @param unscaledValue  * The scaled value in screen coordinates to unscale into axis space.
        *
        * Unscales a value from screen space into axis space.
        *
        * ```ts
        * let dataValue: number = this.yAxis.unscaleValue(pixel);
        * ```
        */
        unscaleValue(unscaledValue) {
            let iv = this.i.l9(unscaledValue);
            return (iv);
        }
        get actualMinimumValueChange() {
            if (this._actualMinimumValueChange == null) {
                this._actualMinimumValueChange = new EventEmitter();
                this.i.propertyChanged = delegateCombine(this.i.propertyChanged, (o, e) => {
                    let iv = e;
                    let ext = this.actualMinimumValue;
                    if (this.beforeActualMinimumValueChange) {
                        this.beforeActualMinimumValueChange(this, ext);
                    }
                    this._actualMinimumValueChange.emit(ext);
                });
            }
            return this._actualMinimumValueChange;
        }
        get actualMaximumValueChange() {
            if (this._actualMaximumValueChange == null) {
                this._actualMaximumValueChange = new EventEmitter();
                this.i.propertyChanged = delegateCombine(this.i.propertyChanged, (o, e) => {
                    let iv = e;
                    let ext = this.actualMaximumValue;
                    if (this.beforeActualMaximumValueChange) {
                        this.beforeActualMaximumValueChange(this, ext);
                    }
                    this._actualMaximumValueChange.emit(ext);
                });
            }
            return this._actualMaximumValueChange;
        }
        get actualIntervalChange() {
            if (this._actualIntervalChange == null) {
                this._actualIntervalChange = new EventEmitter();
                this.i.propertyChanged = delegateCombine(this.i.propertyChanged, (o, e) => {
                    let iv = e;
                    let ext = this.actualInterval;
                    if (this.beforeActualIntervalChange) {
                        this.beforeActualIntervalChange(this, ext);
                    }
                    this._actualIntervalChange.emit(ext);
                });
            }
            return this._actualIntervalChange;
        }
        get actualMinorIntervalChange() {
            if (this._actualMinorIntervalChange == null) {
                this._actualMinorIntervalChange = new EventEmitter();
                this.i.propertyChanged = delegateCombine(this.i.propertyChanged, (o, e) => {
                    let iv = e;
                    let ext = this.actualMinorInterval;
                    if (this.beforeActualMinorIntervalChange) {
                        this.beforeActualMinorIntervalChange(this, ext);
                    }
                    this._actualMinorIntervalChange.emit(ext);
                });
            }
            return this._actualMinorIntervalChange;
        }
    }
    IgxNumericAxisBaseComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAxisBaseComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxNumericAxisBaseComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxNumericAxisBaseComponent, selector: "ng-component", inputs: { autoRangeBufferMode: "autoRangeBufferMode", minimumValue: "minimumValue", actualMinimumValue: "actualMinimumValue", actualVisibleMinimumValue: "actualVisibleMinimumValue", maximumValue: "maximumValue", actualMaximumValue: "actualMaximumValue", actualVisibleMaximumValue: "actualVisibleMaximumValue", interval: "interval", actualInterval: "actualInterval", minorInterval: "minorInterval", actualMinorInterval: "actualMinorInterval", referenceValue: "referenceValue", isLogarithmic: "isLogarithmic", actualIsLogarithmic: "actualIsLogarithmic", favorLabellingScaleEnd: "favorLabellingScaleEnd", logarithmBase: "logarithmBase", formatAbbreviatedLabel: "formatAbbreviatedLabel", abbreviatedLabelFormat: "abbreviatedLabelFormat", abbreviatedLabelFormatSpecifiers: "abbreviatedLabelFormatSpecifiers", isFormattingAbbreviatedLargeNumber: "isFormattingAbbreviatedLargeNumber", abbreviateLargeNumbers: "abbreviateLargeNumbers" }, outputs: { actualMinimumValueChange: "actualMinimumValueChange", actualMaximumValueChange: "actualMaximumValueChange", actualIntervalChange: "actualIntervalChange", actualMinorIntervalChange: "actualMinorIntervalChange" }, usesInheritance: true, ngImport: i0, template: ``, isInline: true });
    return IgxNumericAxisBaseComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAxisBaseComponent, decorators: [{
            type: Component,
            args: [{
                    template: ``,
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { autoRangeBufferMode: [{
                type: Input
            }], minimumValue: [{
                type: Input
            }], actualMinimumValue: [{
                type: Input
            }], actualVisibleMinimumValue: [{
                type: Input
            }], maximumValue: [{
                type: Input
            }], actualMaximumValue: [{
                type: Input
            }], actualVisibleMaximumValue: [{
                type: Input
            }], interval: [{
                type: Input
            }], actualInterval: [{
                type: Input
            }], minorInterval: [{
                type: Input
            }], actualMinorInterval: [{
                type: Input
            }], referenceValue: [{
                type: Input
            }], isLogarithmic: [{
                type: Input
            }], actualIsLogarithmic: [{
                type: Input
            }], favorLabellingScaleEnd: [{
                type: Input
            }], logarithmBase: [{
                type: Input
            }], formatAbbreviatedLabel: [{
                type: Input
            }], abbreviatedLabelFormat: [{
                type: Input
            }], abbreviatedLabelFormatSpecifiers: [{
                type: Input
            }], isFormattingAbbreviatedLargeNumber: [{
                type: Input
            }], abbreviateLargeNumbers: [{
                type: Input
            }], actualMinimumValueChange: [{
                type: Output
            }], actualMaximumValueChange: [{
                type: Output
            }], actualIntervalChange: [{
                type: Output
            }], actualMinorIntervalChange: [{
                type: Output
            }] } });
