// igniteui-angular-charts/esm2015/lib/igx-step-line-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStepLineSeriesModule } from './igx-step-line-series-module';
import { IgxStepLineSeriesComponent } from './igx-step-line-series-component';
import { StepLineSeries } from './StepLineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxStepLineSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxStepLineSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxStepLineSeriesComponent", IgxStepLineSeriesComponent);
            TypeRegistrar.register("StepLineSeries", StepLineSeries.$type);
        }
    }
    IgxStepLineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStepLineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStepLineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStepLineSeriesDynamicModule, imports: [CommonModule, IgxStepLineSeriesModule], exports: [IgxStepLineSeriesModule] });
    IgxStepLineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStepLineSeriesDynamicModule, imports: [[CommonModule, IgxStepLineSeriesModule], IgxStepLineSeriesModule] });
    return IgxStepLineSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStepLineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStepLineSeriesModule],
                    exports: [IgxStepLineSeriesModule],
                    entryComponents: [IgxStepLineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
