// igniteui-angular-charts/esm2015/lib/igx-fast-stochastic-oscillator-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFastStochasticOscillatorIndicatorModule } from './igx-fast-stochastic-oscillator-indicator-module';
import { IgxFastStochasticOscillatorIndicatorComponent } from './igx-fast-stochastic-oscillator-indicator-component';
import { FastStochasticOscillatorIndicator } from './FastStochasticOscillatorIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxFastStochasticOscillatorIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxFastStochasticOscillatorIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxFastStochasticOscillatorIndicatorComponent", IgxFastStochasticOscillatorIndicatorComponent);
            TypeRegistrar.register("FastStochasticOscillatorIndicator", FastStochasticOscillatorIndicator.$type);
        }
    }
    IgxFastStochasticOscillatorIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFastStochasticOscillatorIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFastStochasticOscillatorIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFastStochasticOscillatorIndicatorDynamicModule, imports: [CommonModule, IgxFastStochasticOscillatorIndicatorModule], exports: [IgxFastStochasticOscillatorIndicatorModule] });
    IgxFastStochasticOscillatorIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFastStochasticOscillatorIndicatorDynamicModule, imports: [[CommonModule, IgxFastStochasticOscillatorIndicatorModule], IgxFastStochasticOscillatorIndicatorModule] });
    return IgxFastStochasticOscillatorIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFastStochasticOscillatorIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxFastStochasticOscillatorIndicatorModule],
                    exports: [IgxFastStochasticOscillatorIndicatorModule],
                    entryComponents: [IgxFastStochasticOscillatorIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
