// igniteui-angular-charts/esm2015/lib/igx-polar-area-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPolarAreaSeriesModule } from './igx-polar-area-series-module';
import { IgxPolarAreaSeriesComponent } from './igx-polar-area-series-component';
import { PolarAreaSeries } from './PolarAreaSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxPolarAreaSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxPolarAreaSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxPolarAreaSeriesComponent", IgxPolarAreaSeriesComponent);
            TypeRegistrar.register("PolarAreaSeries", PolarAreaSeries.$type);
        }
    }
    IgxPolarAreaSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarAreaSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPolarAreaSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarAreaSeriesDynamicModule, imports: [CommonModule, IgxPolarAreaSeriesModule], exports: [IgxPolarAreaSeriesModule] });
    IgxPolarAreaSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarAreaSeriesDynamicModule, imports: [[CommonModule, IgxPolarAreaSeriesModule], IgxPolarAreaSeriesModule] });
    return IgxPolarAreaSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarAreaSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPolarAreaSeriesModule],
                    exports: [IgxPolarAreaSeriesModule],
                    entryComponents: [IgxPolarAreaSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
