// igniteui-angular-charts/esm2015/lib/igx-polar-spline-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPolarSplineSeriesModule } from './igx-polar-spline-series-module';
import { IgxPolarSplineSeriesComponent } from './igx-polar-spline-series-component';
import { PolarSplineSeries } from './PolarSplineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxPolarSplineSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxPolarSplineSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxPolarSplineSeriesComponent", IgxPolarSplineSeriesComponent);
            TypeRegistrar.register("PolarSplineSeries", PolarSplineSeries.$type);
        }
    }
    IgxPolarSplineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPolarSplineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesDynamicModule, imports: [CommonModule, IgxPolarSplineSeriesModule], exports: [IgxPolarSplineSeriesModule] });
    IgxPolarSplineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesDynamicModule, imports: [[CommonModule, IgxPolarSplineSeriesModule], IgxPolarSplineSeriesModule] });
    return IgxPolarSplineSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPolarSplineSeriesModule],
                    exports: [IgxPolarSplineSeriesModule],
                    entryComponents: [IgxPolarSplineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
