// igniteui-angular-charts/esm2015/lib/igx-funnel-slice-event-args.js
import { FunnelSliceEventArgs as FunnelSliceEventArgs_internal } from "./FunnelSliceEventArgs";
import { fromRect, toRect, fromPoint, toPoint } from "igniteui-angular-core";
/**
 * Information about the clicked slice.
*/
export class IgxFunnelSliceEventArgs {
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    createImplementation() {
        return new FunnelSliceEventArgs_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    /**
     * The index of the slice.
    */
    get index() {
        return this.i.index;
    }
    set index(v) {
        this.i.index = +v;
    }
    /**
     * The bounds of the slice.
    */
    get bounds() {
        return fromRect(this.i.bounds);
    }
    set bounds(v) {
        this.i.bounds = toRect(v);
    }
    /**
     * The data item for the clicked slice.
    */
    get item() {
        return this.i.item;
    }
    set item(v) {
        this.i.item = v;
    }
    /**
     * The position of the mouse relative to the funnel chart.
    */
    get position() {
        return fromPoint(this.i.position);
    }
    set position(v) {
        this.i.position = toPoint(v);
    }
}
