// igniteui-angular-charts/esm2015/lib/igx-hierarchical-ring-series-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxRingSeriesBaseComponent } from "./igx-ring-series-base-component";
import { HierarchicalRingSeries } from "./HierarchicalRingSeries";
import * as i0 from "@angular/core";
/**
 * Represents a
 * IgxDoughnutChartComponent hierarchical series.
*/
export let IgxHierarchicalRingSeriesComponent = /*@__PURE__*/ (() => {
    class IgxHierarchicalRingSeriesComponent extends IgxRingSeriesBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new HierarchicalRingSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets or sets the property name that contains the children items collection.
        */
        get childrenMemberPath() {
            return this.i.et;
        }
        set childrenMemberPath(v) {
            this.i.et = v;
        }
    }
    IgxHierarchicalRingSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHierarchicalRingSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxHierarchicalRingSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxHierarchicalRingSeriesComponent, selector: "igx-hierarchical-ring-series", inputs: { childrenMemberPath: "childrenMemberPath" }, providers: [{ provide: IgxRingSeriesBaseComponent, useExisting: forwardRef(() => IgxHierarchicalRingSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxHierarchicalRingSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHierarchicalRingSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-hierarchical-ring-series',
                    template: ``,
                    providers: [{ provide: IgxRingSeriesBaseComponent, useExisting: forwardRef(() => IgxHierarchicalRingSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { childrenMemberPath: [{
                type: Input
            }] } });
