// igniteui-angular-charts/esm2015/lib/igx-item-tool-tip-layer-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxItemToolTipLayerModule } from './igx-item-tool-tip-layer-module';
import { IgxItemToolTipLayerComponent } from './igx-item-tool-tip-layer-component';
import { ItemToolTipLayer } from './ItemToolTipLayer';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxItemToolTipLayerDynamicModule = /*@__PURE__*/ (() => {
    class IgxItemToolTipLayerDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxItemToolTipLayerComponent", IgxItemToolTipLayerComponent);
            TypeRegistrar.register("ItemToolTipLayer", ItemToolTipLayer.$type);
        }
    }
    IgxItemToolTipLayerDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemToolTipLayerDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxItemToolTipLayerDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemToolTipLayerDynamicModule, imports: [CommonModule, IgxItemToolTipLayerModule], exports: [IgxItemToolTipLayerModule] });
    IgxItemToolTipLayerDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemToolTipLayerDynamicModule, imports: [[CommonModule, IgxItemToolTipLayerModule], IgxItemToolTipLayerModule] });
    return IgxItemToolTipLayerDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemToolTipLayerDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxItemToolTipLayerModule],
                    exports: [IgxItemToolTipLayerModule],
                    entryComponents: [IgxItemToolTipLayerComponent]
                }]
        }], ctorParameters: function () { return []; } });
