// igniteui-angular-charts/esm2015/lib/igx-spline-fragment-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSplineFragmentModule } from './igx-spline-fragment-module';
import { IgxSplineFragmentComponent } from './igx-spline-fragment-component';
import { SplineFragment } from './SplineFragment';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxSplineFragmentDynamicModule = /*@__PURE__*/ (() => {
    class IgxSplineFragmentDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxSplineFragmentComponent", IgxSplineFragmentComponent);
            TypeRegistrar.register("SplineFragment", SplineFragment.$type);
        }
    }
    IgxSplineFragmentDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineFragmentDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSplineFragmentDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineFragmentDynamicModule, imports: [CommonModule, IgxSplineFragmentModule], exports: [IgxSplineFragmentModule] });
    IgxSplineFragmentDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineFragmentDynamicModule, imports: [[CommonModule, IgxSplineFragmentModule], IgxSplineFragmentModule] });
    return IgxSplineFragmentDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineFragmentDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxSplineFragmentModule],
                    exports: [IgxSplineFragmentModule],
                    entryComponents: [IgxSplineFragmentComponent]
                }]
        }], ctorParameters: function () { return []; } });
