// igniteui-angular-charts/esm2015/lib/igx-category-angle-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryAngleAxisModule } from './igx-category-angle-axis-module';
import { IgxCategoryAngleAxisComponent } from './igx-category-angle-axis-component';
import { CategoryAngleAxis } from './CategoryAngleAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxCategoryAngleAxisDynamicModule = /*@__PURE__*/ (() => {
    class IgxCategoryAngleAxisDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxCategoryAngleAxisComponent", IgxCategoryAngleAxisComponent);
            TypeRegistrar.register("CategoryAngleAxis", CategoryAngleAxis.$type);
        }
    }
    IgxCategoryAngleAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryAngleAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryAngleAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryAngleAxisDynamicModule, imports: [CommonModule, IgxCategoryAngleAxisModule], exports: [IgxCategoryAngleAxisModule] });
    IgxCategoryAngleAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryAngleAxisDynamicModule, imports: [[CommonModule, IgxCategoryAngleAxisModule], IgxCategoryAngleAxisModule] });
    return IgxCategoryAngleAxisDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryAngleAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCategoryAngleAxisModule],
                    exports: [IgxCategoryAngleAxisModule],
                    entryComponents: [IgxCategoryAngleAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
