// igniteui-angular-charts/esm2015/lib/igx-range-area-series-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxHorizontalRangeCategorySeriesComponent } from "./igx-horizontal-range-category-series-component";
import { IgxRangeCategorySeriesComponent } from "./igx-range-category-series-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { RangeAreaSeries } from "./RangeAreaSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent range area series.
*
* The `RangeAreaSeries` class represents the range area series of the IgxDataChartComponent.
*
* ```html
* <igx-data-chart
* [dataSource]="data">
* <igx-category-x-axis
* 	label="label"
* 	#xAxis>
* </igx-category-x-axis>
* <igx-numeric-y-axis
* 	#yAxis>
* </igx-numeric-y-axis>
*     <igx-range-area-series
*       [xAxis]="xAxis"
*       [yAxis]="yAxis"
*       highMemberPath="high"
*       lowMemberPath="low">
*   </igx-range-area-series>
*     </igx-data-chart>
* </igx-data-chart>
* ```
*
* ```ts
* let series: IgxRangeAreaSeriesComponent = new IgxRangeAreaSeriesComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.highMemberPath = "High";
* series.lowMemberPath = "Low";
* this.chart.series.add(series);
* ```
*/
export let IgxRangeAreaSeriesComponent = /*@__PURE__*/ (() => {
    class IgxRangeAreaSeriesComponent extends IgxHorizontalRangeCategorySeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new RangeAreaSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows an area or line shape.
        *
        * Use the `IsAreaOrLine` property to check if the current series shows an area or line shape.
        *
        * ```ts
        * var x= series.isAreaOrLine;
        * ```
        */
        get isAreaOrLine() {
            return this.i.et;
        }
        /**
         * Gets whether the current series shows an area shape.
        */
        get isArea() {
            return this.i.es;
        }
    }
    IgxRangeAreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeAreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxRangeAreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxRangeAreaSeriesComponent, selector: "igx-range-area-series", providers: [{ provide: IgxHorizontalRangeCategorySeriesComponent, useExisting: forwardRef(() => IgxRangeAreaSeriesComponent) }, { provide: IgxRangeCategorySeriesComponent, useExisting: forwardRef(() => IgxRangeAreaSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxRangeAreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxRangeAreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxRangeAreaSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxRangeAreaSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeAreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-range-area-series',
                    template: ``,
                    providers: [{ provide: IgxHorizontalRangeCategorySeriesComponent, useExisting: forwardRef(() => IgxRangeAreaSeriesComponent) }, { provide: IgxRangeCategorySeriesComponent, useExisting: forwardRef(() => IgxRangeAreaSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxRangeAreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxRangeAreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxRangeAreaSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
