// igniteui-angular-charts/esm2015/lib/igx-treemap-node-style-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTreemapNodeStyleModule } from './igx-treemap-node-style-module';
import { IgxTreemapNodeStyleComponent } from './igx-treemap-node-style-component';
import { TreemapNodeStyle } from './TreemapNodeStyle';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxTreemapNodeStyleDynamicModule = /*@__PURE__*/ (() => {
    class IgxTreemapNodeStyleDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxTreemapNodeStyleComponent", IgxTreemapNodeStyleComponent);
            TypeRegistrar.register("TreemapNodeStyle", TreemapNodeStyle.$type);
        }
    }
    IgxTreemapNodeStyleDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTreemapNodeStyleDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleDynamicModule, imports: [CommonModule, IgxTreemapNodeStyleModule], exports: [IgxTreemapNodeStyleModule] });
    IgxTreemapNodeStyleDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleDynamicModule, imports: [[CommonModule, IgxTreemapNodeStyleModule], IgxTreemapNodeStyleModule] });
    return IgxTreemapNodeStyleDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxTreemapNodeStyleModule],
                    exports: [IgxTreemapNodeStyleModule],
                    entryComponents: [IgxTreemapNodeStyleComponent]
                }]
        }], ctorParameters: function () { return []; } });
