// igniteui-angular-charts/esm2015/lib/igx-chaikin-oscillator-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxChaikinOscillatorIndicatorModule } from './igx-chaikin-oscillator-indicator-module';
import { IgxChaikinOscillatorIndicatorComponent } from './igx-chaikin-oscillator-indicator-component';
import { ChaikinOscillatorIndicator } from './ChaikinOscillatorIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxChaikinOscillatorIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxChaikinOscillatorIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxChaikinOscillatorIndicatorComponent", IgxChaikinOscillatorIndicatorComponent);
            TypeRegistrar.register("ChaikinOscillatorIndicator", ChaikinOscillatorIndicator.$type);
        }
    }
    IgxChaikinOscillatorIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinOscillatorIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxChaikinOscillatorIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinOscillatorIndicatorDynamicModule, imports: [CommonModule, IgxChaikinOscillatorIndicatorModule], exports: [IgxChaikinOscillatorIndicatorModule] });
    IgxChaikinOscillatorIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinOscillatorIndicatorDynamicModule, imports: [[CommonModule, IgxChaikinOscillatorIndicatorModule], IgxChaikinOscillatorIndicatorModule] });
    return IgxChaikinOscillatorIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinOscillatorIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxChaikinOscillatorIndicatorModule],
                    exports: [IgxChaikinOscillatorIndicatorModule],
                    entryComponents: [IgxChaikinOscillatorIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
