// igniteui-angular-charts/esm2015/lib/igx-category-chart-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryChartComponent } from './igx-category-chart-component';
import { IgxCategoryXAxisDynamicModule } from './igx-category-x-axis-dynamic-module';
import { IgxNumericYAxisDynamicModule } from './igx-numeric-y-axis-dynamic-module';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxCategoryChartCoreModule } from './igx-category-chart-core-module';
import * as i0 from "@angular/core";
export let IgxCategoryChartCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxCategoryChartCoreDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('IgxCategoryChartComponent', IgxCategoryChartComponent);
        }
    }
    IgxCategoryChartCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryChartCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartCoreDynamicModule, imports: [CommonModule, IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxCategoryChartCoreModule], exports: [IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxCategoryChartCoreModule] });
    IgxCategoryChartCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartCoreDynamicModule, imports: [[CommonModule, IgxCategoryXAxisDynamicModule,
                IgxNumericYAxisDynamicModule,
                IgxCategoryChartCoreModule
            ], IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxCategoryChartCoreModule] });
    return IgxCategoryChartCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryChartCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxCategoryXAxisDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxCategoryChartCoreModule
                    ],
                    imports: [CommonModule, IgxCategoryXAxisDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxCategoryChartCoreModule
                    ],
                    entryComponents: [IgxCategoryChartComponent]
                }]
        }], ctorParameters: function () { return []; } });
