// igniteui-angular-charts/esm2015/lib/igx-size-scale-component.js
import { EventEmitter, Output, Component, Input, ChangeDetectionStrategy } from '@angular/core';
import { delegateCombine } from "igniteui-angular-core";
import { IgxPropertyUpdatedEventArgs } from "igniteui-angular-core";
import { SizeScale } from "./SizeScale";
import { TypeRegistrar } from "igniteui-angular-core";
import { NamePatcher, ensureBool, toSpinal, initializePropertiesFromCss } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a scale that is used determine an object's size.
*/
export let IgxSizeScaleComponent = /*@__PURE__*/ (() => {
    class IgxSizeScaleComponent {
        constructor() {
            this.__p = null;
            this._hasUserValues = new Set();
            this._stylingContainer = null;
            this._stylingParent = null;
            this._inStyling = false;
            this._propertyUpdated = null;
            this._zoneRunner = null;
            if (this._styling) {
                NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
            }
            this._implementation = this.createImplementation();
            this._implementation.externalObject = this;
            this.onImplementationCreated();
            if (this._initializeAdapters) {
                this._initializeAdapters();
            }
        }
        createImplementation() {
            return new SizeScale();
        }
        ngOnInit() {
        }
        /**
         * @hidden
         */
        get i() {
            return this._implementation;
        } /**
                                     * @hidden
                                     */
        static _createFromInternal(internal) {
            if (!internal) {
                return null;
            }
            if (!internal.$type) {
                return null;
            }
            let name = internal.$type.name;
            let externalName = "Igx" + name + "Component";
            if (!TypeRegistrar.isRegistered(externalName)) {
                return null;
            }
            return TypeRegistrar.create(externalName);
        }
        onImplementationCreated() {
        }
        /**
         * Gets or sets global minimum value that will be mapped to MinumumValue of this scale
        */
        get globalMinimum() {
            return this.i.globalMinimum;
        }
        set globalMinimum(v) {
            this.i.globalMinimum = +v;
        }
        /**
         * Gets or sets global maximum value that will be mapped to MaxumumValue of this scale
        */
        get globalMaximum() {
            return this.i.globalMaximum;
        }
        set globalMaximum(v) {
            this.i.globalMaximum = +v;
        }
        /**
         * Gets or sets the minimum size (in pixels) for this scale.
        */
        get minimumValue() {
            return this.i.minimumValue;
        }
        set minimumValue(v) {
            this.i.minimumValue = +v;
        }
        /**
         * Gets or sets the maximum size (in pixels) for this scale.
        */
        get maximumValue() {
            return this.i.maximumValue;
        }
        set maximumValue(v) {
            this.i.maximumValue = +v;
        }
        /**
         * Gets or sets whether the scale is logarithmic.
        */
        get isLogarithmic() {
            return this.i.isLogarithmic;
        }
        set isLogarithmic(v) {
            this.i.isLogarithmic = ensureBool(v);
        }
        /**
         * Gets or sets the logarithm base for this scale.
        */
        get logarithmBase() {
            return this.i.logarithmBase;
        }
        set logarithmBase(v) {
            this.i.logarithmBase = +v;
        }
        findByName(name) {
            if (this.findEphemera) {
                if (name && name.indexOf("@@e:") == 0) {
                    return this.findEphemera(name);
                }
            }
            return null;
        }
        get hasUserValues() {
            return this._hasUserValues;
        }
        __m(propertyName) {
            if (!this._inStyling) {
                this._hasUserValues.add(propertyName);
            }
        }
        _styling(container, component, parent) {
            if (this._inStyling) {
                return;
            }
            this._inStyling = true;
            this._stylingContainer = container;
            this._stylingParent = component;
            let genericPrefix = "";
            let typeName = this.i.$type.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            genericPrefix = toSpinal("SizeScaleComponent");
            let additionalPrefixes = [];
            let prefix = toSpinal(typeName);
            additionalPrefixes.push(prefix + "-");
            let b = this.i.$type.baseType;
            while (b && b.name != "Object" &&
                b.name != "Base" &&
                b.name != "Control" &&
                b.Name != "DependencyObject" &&
                b.Name != "FrameworkElement") {
                typeName = b.name;
                if (typeName.indexOf("Xam") === 0) {
                    typeName = typeName.substring(3);
                }
                let basePrefix = toSpinal(typeName);
                additionalPrefixes.push(basePrefix + "-");
                b = b.baseType;
            }
            if (parent) {
                let parentTypeName = parent.i.$type.name;
                if (parentTypeName.indexOf("Xam") === 0) {
                    parentTypeName = parentTypeName.substring(3);
                }
                let parentPrefix = toSpinal(parentTypeName);
                additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
                additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
            }
            initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
            if (this._otherStyling) {
                this._otherStyling(container, component, parent);
            }
            this._inStyling = false;
        }
        getCurrentGlobalMinimum() {
            let iv = this.i.getCurrentGlobalMinimum();
            return (iv);
        }
        getCurrentGlobalMaximum() {
            let iv = this.i.getCurrentGlobalMaximum();
            return (iv);
        }
        /**
         * Event raised when a property (including "effective" and non-dependency property) value changes.
        */
        get propertyUpdated() {
            if (this._propertyUpdated == null) {
                this._propertyUpdated = new EventEmitter();
                this.i.propertyUpdated = delegateCombine(this.i.propertyUpdated, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxPropertyUpdatedEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforePropertyUpdated) {
                            this.beforePropertyUpdated(this, outerArgs);
                        }
                        this._propertyUpdated.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._propertyUpdated;
        }
        _runInZone(act) {
            if (this._zoneRunner != null) {
                this._zoneRunner(act);
            }
            else {
                act();
            }
        }
    }
    IgxSizeScaleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxSizeScaleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxSizeScaleComponent, selector: "igx-size-scale", inputs: { globalMinimum: "globalMinimum", globalMaximum: "globalMaximum", minimumValue: "minimumValue", maximumValue: "maximumValue", isLogarithmic: "isLogarithmic", logarithmBase: "logarithmBase" }, outputs: { propertyUpdated: "propertyUpdated" }, providers: [], ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxSizeScaleComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-size-scale',
                    template: ``,
                    providers: [],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { globalMinimum: [{
                type: Input
            }], globalMaximum: [{
                type: Input
            }], minimumValue: [{
                type: Input
            }], maximumValue: [{
                type: Input
            }], isLogarithmic: [{
                type: Input
            }], logarithmBase: [{
                type: Input
            }], propertyUpdated: [{
                type: Output
            }] } });
