// igniteui-angular-charts/esm2015/lib/igx-full-stochastic-oscillator-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFullStochasticOscillatorIndicatorModule } from './igx-full-stochastic-oscillator-indicator-module';
import { IgxFullStochasticOscillatorIndicatorComponent } from './igx-full-stochastic-oscillator-indicator-component';
import { FullStochasticOscillatorIndicator } from './FullStochasticOscillatorIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxFullStochasticOscillatorIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxFullStochasticOscillatorIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxFullStochasticOscillatorIndicatorComponent", IgxFullStochasticOscillatorIndicatorComponent);
            TypeRegistrar.register("FullStochasticOscillatorIndicator", FullStochasticOscillatorIndicator.$type);
        }
    }
    IgxFullStochasticOscillatorIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFullStochasticOscillatorIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFullStochasticOscillatorIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFullStochasticOscillatorIndicatorDynamicModule, imports: [CommonModule, IgxFullStochasticOscillatorIndicatorModule], exports: [IgxFullStochasticOscillatorIndicatorModule] });
    IgxFullStochasticOscillatorIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFullStochasticOscillatorIndicatorDynamicModule, imports: [[CommonModule, IgxFullStochasticOscillatorIndicatorModule], IgxFullStochasticOscillatorIndicatorModule] });
    return IgxFullStochasticOscillatorIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFullStochasticOscillatorIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxFullStochasticOscillatorIndicatorModule],
                    exports: [IgxFullStochasticOscillatorIndicatorModule],
                    entryComponents: [IgxFullStochasticOscillatorIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
