// igniteui-angular-charts/esm2015/lib/igx-data-chart-radial-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryAngleAxisDynamicModule } from './igx-category-angle-axis-dynamic-module';
import { IgxNumericRadiusAxisDynamicModule } from './igx-numeric-radius-axis-dynamic-module';
import { IgxDataChartCoreDynamicModule } from './igx-data-chart-core-dynamic-module';
import { IgxDataChartRadialCoreModule } from './igx-data-chart-radial-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartRadialCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartRadialCoreDynamicModule {
    }
    IgxDataChartRadialCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartRadialCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialCoreDynamicModule, imports: [CommonModule, IgxCategoryAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartRadialCoreModule], exports: [IgxCategoryAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartRadialCoreModule] });
    IgxDataChartRadialCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialCoreDynamicModule, imports: [[CommonModule, IgxCategoryAngleAxisDynamicModule,
                IgxNumericRadiusAxisDynamicModule,
                IgxDataChartCoreDynamicModule,
                IgxDataChartRadialCoreModule
            ], IgxCategoryAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartRadialCoreModule] });
    return IgxDataChartRadialCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxCategoryAngleAxisDynamicModule,
                        IgxNumericRadiusAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartRadialCoreModule
                    ],
                    imports: [CommonModule, IgxCategoryAngleAxisDynamicModule,
                        IgxNumericRadiusAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartRadialCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
