// igniteui-angular-charts/esm2015/lib/igx-data-pie-chart-toolbar-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataChartToolbarIconsDynamicModule } from './igx-data-chart-toolbar-icons-dynamic-module';
import { DataChartToolbarFactoryManager } from './DataChartToolbarFactoryManager';
import { DataPieChartToolbarFactoryManager } from './DataPieChartToolbarFactoryManager';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDataPieChartToolbarModule } from './igx-data-pie-chart-toolbar-module';
import * as i0 from "@angular/core";
export let IgxDataPieChartToolbarDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataPieChartToolbarDynamicModule {
        constructor() {
            TypeRegistrar.register('DataChartToolbarFactoryManager', DataChartToolbarFactoryManager.$type);
            DataChartToolbarFactoryManager.register();
            TypeRegistrar.register('DataPieChartToolbarFactoryManager', DataPieChartToolbarFactoryManager.$type);
            DataPieChartToolbarFactoryManager.register();
        }
    }
    IgxDataPieChartToolbarDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartToolbarDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataPieChartToolbarDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartToolbarDynamicModule, imports: [CommonModule, IgxDataChartToolbarIconsDynamicModule,
            IgxDataPieChartToolbarModule], exports: [IgxDataChartToolbarIconsDynamicModule,
            IgxDataPieChartToolbarModule] });
    IgxDataPieChartToolbarDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartToolbarDynamicModule, imports: [[CommonModule, IgxDataChartToolbarIconsDynamicModule,
                IgxDataPieChartToolbarModule
            ], IgxDataChartToolbarIconsDynamicModule,
            IgxDataPieChartToolbarModule] });
    return IgxDataPieChartToolbarDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartToolbarDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataChartToolbarIconsDynamicModule,
                        IgxDataPieChartToolbarModule
                    ],
                    imports: [CommonModule, IgxDataChartToolbarIconsDynamicModule,
                        IgxDataPieChartToolbarModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
