// igniteui-angular-charts/esm2015/lib/igx-weighted-close-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxWeightedCloseIndicatorModule } from './igx-weighted-close-indicator-module';
import { IgxWeightedCloseIndicatorComponent } from './igx-weighted-close-indicator-component';
import { WeightedCloseIndicator } from './WeightedCloseIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxWeightedCloseIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxWeightedCloseIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxWeightedCloseIndicatorComponent", IgxWeightedCloseIndicatorComponent);
            TypeRegistrar.register("WeightedCloseIndicator", WeightedCloseIndicator.$type);
        }
    }
    IgxWeightedCloseIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWeightedCloseIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxWeightedCloseIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWeightedCloseIndicatorDynamicModule, imports: [CommonModule, IgxWeightedCloseIndicatorModule], exports: [IgxWeightedCloseIndicatorModule] });
    IgxWeightedCloseIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWeightedCloseIndicatorDynamicModule, imports: [[CommonModule, IgxWeightedCloseIndicatorModule], IgxWeightedCloseIndicatorModule] });
    return IgxWeightedCloseIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxWeightedCloseIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxWeightedCloseIndicatorModule],
                    exports: [IgxWeightedCloseIndicatorModule],
                    entryComponents: [IgxWeightedCloseIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
