// igniteui-angular-charts/esm2015/lib/igx-brush-scale-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxBrushScaleModule } from './igx-brush-scale-module';
import { IgxBrushScaleComponent } from './igx-brush-scale-component';
import { BrushScale } from './BrushScale';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxBrushScaleDynamicModule = /*@__PURE__*/ (() => {
    class IgxBrushScaleDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxBrushScaleComponent", IgxBrushScaleComponent);
            TypeRegistrar.register("BrushScale", BrushScale.$type);
        }
    }
    IgxBrushScaleDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBrushScaleDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxBrushScaleDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBrushScaleDynamicModule, imports: [CommonModule, IgxBrushScaleModule], exports: [IgxBrushScaleModule] });
    IgxBrushScaleDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBrushScaleDynamicModule, imports: [[CommonModule, IgxBrushScaleModule], IgxBrushScaleModule] });
    return IgxBrushScaleDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBrushScaleDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxBrushScaleModule],
                    exports: [IgxBrushScaleModule],
                    entryComponents: [IgxBrushScaleComponent]
                }]
        }], ctorParameters: function () { return []; } });
