// igniteui-angular-charts/esm2015/lib/igx-category-date-time-x-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryDateTimeXAxisModule } from './igx-category-date-time-x-axis-module';
import { IgxCategoryDateTimeXAxisComponent } from './igx-category-date-time-x-axis-component';
import { CategoryDateTimeXAxis } from './CategoryDateTimeXAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxCategoryDateTimeXAxisDynamicModule = /*@__PURE__*/ (() => {
    class IgxCategoryDateTimeXAxisDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxCategoryDateTimeXAxisComponent", IgxCategoryDateTimeXAxisComponent);
            TypeRegistrar.register("CategoryDateTimeXAxis", CategoryDateTimeXAxis.$type);
        }
    }
    IgxCategoryDateTimeXAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryDateTimeXAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryDateTimeXAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryDateTimeXAxisDynamicModule, imports: [CommonModule, IgxCategoryDateTimeXAxisModule], exports: [IgxCategoryDateTimeXAxisModule] });
    IgxCategoryDateTimeXAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryDateTimeXAxisDynamicModule, imports: [[CommonModule, IgxCategoryDateTimeXAxisModule], IgxCategoryDateTimeXAxisModule] });
    return IgxCategoryDateTimeXAxisDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryDateTimeXAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCategoryDateTimeXAxisModule],
                    exports: [IgxCategoryDateTimeXAxisModule],
                    entryComponents: [IgxCategoryDateTimeXAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
