// igniteui-angular-charts/esm2015/lib/igx-doughnut-chart-interactivity-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { InteractivityFactoryManager } from './InteractivityFactoryManager';
import { IgxDoughnutChartDefaultTooltipsComponent } from './igx-doughnut-chart-default-tooltips-component';
import { IgxDVInteractivityModule } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxDoughnutChartInteractivityModule = /*@__PURE__*/ (() => {
    class IgxDoughnutChartInteractivityModule {
        constructor() {
            TypeRegistrar.register('InteractivityFactoryManager', InteractivityFactoryManager.$type);
            InteractivityFactoryManager.register();
            IgxDoughnutChartDefaultTooltipsComponent.register();
        }
    }
    IgxDoughnutChartInteractivityModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartInteractivityModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDoughnutChartInteractivityModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartInteractivityModule, declarations: [IgxDoughnutChartDefaultTooltipsComponent], imports: [CommonModule, IgxDVInteractivityModule], exports: [IgxDoughnutChartDefaultTooltipsComponent,
            IgxDVInteractivityModule] });
    IgxDoughnutChartInteractivityModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartInteractivityModule, imports: [[CommonModule, IgxDVInteractivityModule
            ], IgxDVInteractivityModule] });
    return IgxDoughnutChartInteractivityModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartInteractivityModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxDoughnutChartDefaultTooltipsComponent],
                    exports: [IgxDoughnutChartDefaultTooltipsComponent,
                        IgxDVInteractivityModule
                    ],
                    imports: [CommonModule, IgxDVInteractivityModule
                    ],
                    entryComponents: [IgxDoughnutChartDefaultTooltipsComponent]
                }]
        }], ctorParameters: function () { return []; } });
