// igniteui-angular-charts/esm2015/lib/igx-column-fragment-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxColumnFragmentModule } from './igx-column-fragment-module';
import { IgxColumnFragmentComponent } from './igx-column-fragment-component';
import { ColumnFragment } from './ColumnFragment';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxColumnFragmentDynamicModule = /*@__PURE__*/ (() => {
    class IgxColumnFragmentDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxColumnFragmentComponent", IgxColumnFragmentComponent);
            TypeRegistrar.register("ColumnFragment", ColumnFragment.$type);
        }
    }
    IgxColumnFragmentDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnFragmentDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxColumnFragmentDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnFragmentDynamicModule, imports: [CommonModule, IgxColumnFragmentModule], exports: [IgxColumnFragmentModule] });
    IgxColumnFragmentDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnFragmentDynamicModule, imports: [[CommonModule, IgxColumnFragmentModule], IgxColumnFragmentModule] });
    return IgxColumnFragmentDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnFragmentDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxColumnFragmentModule],
                    exports: [IgxColumnFragmentModule],
                    entryComponents: [IgxColumnFragmentComponent]
                }]
        }], ctorParameters: function () { return []; } });
