// igniteui-angular-charts/esm2015/lib/igx-polar-area-series-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { UnknownValuePlotting_$type } from "igniteui-angular-core";
import { IgxPolarLineSeriesBaseComponent } from "./igx-polar-line-series-base-component";
import { IgxPolarBaseComponent } from "./igx-polar-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { PolarAreaSeries } from "./PolarAreaSeries";
import { ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent polar area series.
*/
export let IgxPolarAreaSeriesComponent = /*@__PURE__*/ (() => {
    class IgxPolarAreaSeriesComponent extends IgxPolarLineSeriesBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new PolarAreaSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows an area shape.
        */
        get isArea() {
            return this.i.es;
        }
        /**
         * Determines how unknown values will be plotted on the chart.
         * Null and Double.NaN are two examples of unknown values.
        */
        get unknownValuePlotting() {
            return this.i.acn;
        }
        set unknownValuePlotting(v) {
            this.i.acn = ensureEnum(UnknownValuePlotting_$type, v);
        }
    }
    IgxPolarAreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarAreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxPolarAreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxPolarAreaSeriesComponent, selector: "igx-polar-area-series", inputs: { unknownValuePlotting: "unknownValuePlotting" }, providers: [{ provide: IgxPolarLineSeriesBaseComponent, useExisting: forwardRef(() => IgxPolarAreaSeriesComponent) }, { provide: IgxPolarBaseComponent, useExisting: forwardRef(() => IgxPolarAreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxPolarAreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxPolarAreaSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxPolarAreaSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarAreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-polar-area-series',
                    template: ``,
                    providers: [{ provide: IgxPolarLineSeriesBaseComponent, useExisting: forwardRef(() => IgxPolarAreaSeriesComponent) }, { provide: IgxPolarBaseComponent, useExisting: forwardRef(() => IgxPolarAreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxPolarAreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxPolarAreaSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { unknownValuePlotting: [{
                type: Input
            }] } });
