// igniteui-angular-charts/esm2015/lib/igx-line-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxLineSeriesModule } from './igx-line-series-module';
import { IgxLineSeriesComponent } from './igx-line-series-component';
import { LineSeries } from './LineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxLineSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxLineSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxLineSeriesComponent", IgxLineSeriesComponent);
            TypeRegistrar.register("LineSeries", LineSeries.$type);
        }
    }
    IgxLineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxLineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineSeriesDynamicModule, imports: [CommonModule, IgxLineSeriesModule], exports: [IgxLineSeriesModule] });
    IgxLineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineSeriesDynamicModule, imports: [[CommonModule, IgxLineSeriesModule], IgxLineSeriesModule] });
    return IgxLineSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxLineSeriesModule],
                    exports: [IgxLineSeriesModule],
                    entryComponents: [IgxLineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
