// igniteui-angular-charts/esm2015/lib/igx-stacked-100-line-series-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStackedLineSeriesComponent } from "./igx-stacked-line-series-component";
import { IgxHorizontalStackedSeriesBaseComponent } from "./igx-horizontal-stacked-series-base-component";
import { IgxStackedSeriesBaseComponent } from "./igx-stacked-series-base-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { Stacked100LineSeries } from "./Stacked100LineSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent stacked100 line series.
*/
export let IgxStacked100LineSeriesComponent = /*@__PURE__*/ (() => {
    class IgxStacked100LineSeriesComponent extends IgxStackedLineSeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new Stacked100LineSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        get isPercentBased() {
            return this.i.aan;
        }
    }
    IgxStacked100LineSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100LineSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStacked100LineSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStacked100LineSeriesComponent, selector: "igx-stacked-100-line-series", providers: [{ provide: IgxStackedLineSeriesComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStacked100LineSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100LineSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-stacked-100-line-series',
                    template: ``,
                    providers: [{ provide: IgxStackedLineSeriesComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100LineSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
