// igniteui-angular-charts/esm2015/lib/igx-numeric-y-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericYAxisComponent } from './igx-numeric-y-axis-component';
import * as i0 from "@angular/core";
export let IgxNumericYAxisModule = /*@__PURE__*/ (() => {
    class IgxNumericYAxisModule {
    }
    IgxNumericYAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericYAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxNumericYAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericYAxisModule, declarations: [IgxNumericYAxisComponent], imports: [CommonModule], exports: [IgxNumericYAxisComponent] });
    IgxNumericYAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericYAxisModule, imports: [[CommonModule]] });
    return IgxNumericYAxisModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericYAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxNumericYAxisComponent],
                    imports: [CommonModule],
                    exports: [IgxNumericYAxisComponent],
                    entryComponents: []
                }]
        }] });
