// igniteui-angular-charts/esm2015/lib/igx-scatter-polyline-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterPolylineSeriesModule } from './igx-scatter-polyline-series-module';
import { IgxScatterPolylineSeriesComponent } from './igx-scatter-polyline-series-component';
import { ScatterPolylineSeries } from './ScatterPolylineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxScatterPolylineSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxScatterPolylineSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxScatterPolylineSeriesComponent", IgxScatterPolylineSeriesComponent);
            TypeRegistrar.register("ScatterPolylineSeries", ScatterPolylineSeries.$type);
        }
    }
    IgxScatterPolylineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolylineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterPolylineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolylineSeriesDynamicModule, imports: [CommonModule, IgxScatterPolylineSeriesModule], exports: [IgxScatterPolylineSeriesModule] });
    IgxScatterPolylineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolylineSeriesDynamicModule, imports: [[CommonModule, IgxScatterPolylineSeriesModule], IgxScatterPolylineSeriesModule] });
    return IgxScatterPolylineSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterPolylineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxScatterPolylineSeriesModule],
                    exports: [IgxScatterPolylineSeriesModule],
                    entryComponents: [IgxScatterPolylineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
