// igniteui-angular-charts/esm2015/lib/igx-category-highlight-layer-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAnnotationLayerProxyModule } from './igx-annotation-layer-proxy-module';
import { IgxCategoryHighlightLayerComponent } from './igx-category-highlight-layer-component';
import * as i0 from "@angular/core";
export let IgxCategoryHighlightLayerModule = /*@__PURE__*/ (() => {
    class IgxCategoryHighlightLayerModule {
    }
    IgxCategoryHighlightLayerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryHighlightLayerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryHighlightLayerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryHighlightLayerModule, declarations: [IgxCategoryHighlightLayerComponent], imports: [CommonModule, IgxAnnotationLayerProxyModule], exports: [IgxCategoryHighlightLayerComponent, IgxAnnotationLayerProxyModule] });
    IgxCategoryHighlightLayerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryHighlightLayerModule, imports: [[CommonModule, IgxAnnotationLayerProxyModule], IgxAnnotationLayerProxyModule] });
    return IgxCategoryHighlightLayerModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryHighlightLayerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxCategoryHighlightLayerComponent],
                    imports: [CommonModule, IgxAnnotationLayerProxyModule],
                    exports: [IgxCategoryHighlightLayerComponent, IgxAnnotationLayerProxyModule],
                    entryComponents: []
                }]
        }] });
