// igniteui-angular-charts/esm2015/lib/igx-sparkline-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSparklineComponent } from './igx-sparkline-component';
import { FastItemObjectColumn } from "igniteui-angular-core";
import { IgxNumberFormatSpecifierDynamicModule } from "igniteui-angular-core";
import { IgxDateTimeFormatSpecifierDynamicModule } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxSparklineCoreModule } from './igx-sparkline-core-module';
import * as i0 from "@angular/core";
export let IgxSparklineCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxSparklineCoreDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('IgxSparklineComponent', IgxSparklineComponent);
            TypeRegistrar.register('FastItemObjectColumn', FastItemObjectColumn.$type);
        }
    }
    IgxSparklineCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSparklineCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineCoreDynamicModule, imports: [CommonModule, IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxSparklineCoreModule], exports: [IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxSparklineCoreModule] });
    IgxSparklineCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineCoreDynamicModule, imports: [[CommonModule, IgxNumberFormatSpecifierDynamicModule,
                IgxDateTimeFormatSpecifierDynamicModule,
                IgxSparklineCoreModule
            ], IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxSparklineCoreModule] });
    return IgxSparklineCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSparklineCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumberFormatSpecifierDynamicModule,
                        IgxDateTimeFormatSpecifierDynamicModule,
                        IgxSparklineCoreModule
                    ],
                    imports: [CommonModule, IgxNumberFormatSpecifierDynamicModule,
                        IgxDateTimeFormatSpecifierDynamicModule,
                        IgxSparklineCoreModule
                    ],
                    entryComponents: [IgxSparklineComponent]
                }]
        }], ctorParameters: function () { return []; } });
