// igniteui-angular-charts/esm2015/lib/igx-horizontal-anchored-category-series-proxy-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HorizontalAnchoredCategorySeriesProxy } from './HorizontalAnchoredCategorySeriesProxy';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxHorizontalAnchoredCategorySeriesProxyModule = /*@__PURE__*/ (() => {
    class IgxHorizontalAnchoredCategorySeriesProxyModule {
        constructor() {
            TypeRegistrar.register('HorizontalAnchoredCategorySeriesProxy', HorizontalAnchoredCategorySeriesProxy.$type);
        }
    }
    IgxHorizontalAnchoredCategorySeriesProxyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHorizontalAnchoredCategorySeriesProxyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxHorizontalAnchoredCategorySeriesProxyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHorizontalAnchoredCategorySeriesProxyModule, imports: [CommonModule] });
    IgxHorizontalAnchoredCategorySeriesProxyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHorizontalAnchoredCategorySeriesProxyModule, imports: [[CommonModule,
            ]] });
    return IgxHorizontalAnchoredCategorySeriesProxyModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHorizontalAnchoredCategorySeriesProxyModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
