// igniteui-angular-charts/esm2015/lib/igx-final-value-layer-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFinalValueLayerModule } from './igx-final-value-layer-module';
import { IgxFinalValueLayerComponent } from './igx-final-value-layer-component';
import { FinalValueLayer } from './FinalValueLayer';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxFinalValueLayerDynamicModule = /*@__PURE__*/ (() => {
    class IgxFinalValueLayerDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxFinalValueLayerComponent", IgxFinalValueLayerComponent);
            TypeRegistrar.register("FinalValueLayer", FinalValueLayer.$type);
        }
    }
    IgxFinalValueLayerDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinalValueLayerDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFinalValueLayerDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinalValueLayerDynamicModule, imports: [CommonModule, IgxFinalValueLayerModule], exports: [IgxFinalValueLayerModule] });
    IgxFinalValueLayerDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinalValueLayerDynamicModule, imports: [[CommonModule, IgxFinalValueLayerModule], IgxFinalValueLayerModule] });
    return IgxFinalValueLayerDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinalValueLayerDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxFinalValueLayerModule],
                    exports: [IgxFinalValueLayerModule],
                    entryComponents: [IgxFinalValueLayerComponent]
                }]
        }], ctorParameters: function () { return []; } });
