// igniteui-angular-charts/esm2015/lib/igx-straight-numeric-axis-base-component.js
import { Component, Input } from '@angular/core';
import { NumericScaleMode_$type } from "./NumericScaleMode";
import { IgxNumericAxisBaseComponent } from "./igx-numeric-axis-base-component";
import { ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Defines a set of basic methods and properties used to create a StraightNumeric axis.
*/
export let IgxStraightNumericAxisBaseComponent = /*@__PURE__*/ (() => {
    class IgxStraightNumericAxisBaseComponent extends IgxNumericAxisBaseComponent {
        constructor() {
            super();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets or sets the axis scale mode.
        *
        * `ScaleMode` can be used on numeric axes to allow scaling data values using built-in scalers. The available scalers are linear or logarithmic.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis
        *         scaleMode="logarithmic">
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.yAxis.scaleMode = NumericScaleMode.Logarithmic;
        * ```
        */
        get scaleMode() {
            return this.i.ni;
        }
        set scaleMode(v) {
            this.i.ni = ensureEnum(NumericScaleMode_$type, v);
        }
    }
    IgxStraightNumericAxisBaseComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStraightNumericAxisBaseComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStraightNumericAxisBaseComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStraightNumericAxisBaseComponent, selector: "ng-component", inputs: { scaleMode: "scaleMode" }, usesInheritance: true, ngImport: i0, template: ``, isInline: true });
    return IgxStraightNumericAxisBaseComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStraightNumericAxisBaseComponent, decorators: [{
            type: Component,
            args: [{
                    template: ``,
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { scaleMode: [{
                type: Input
            }] } });
