// igniteui-angular-charts/esm2015/lib/igx-treemap-node-style-mapping-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTreemapNodeStyleMappingModule } from './igx-treemap-node-style-mapping-module';
import { IgxTreemapNodeStyleMappingComponent } from './igx-treemap-node-style-mapping-component';
import { TreemapNodeStyleMapping } from './TreemapNodeStyleMapping';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxTreemapNodeStyleMappingDynamicModule = /*@__PURE__*/ (() => {
    class IgxTreemapNodeStyleMappingDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxTreemapNodeStyleMappingComponent", IgxTreemapNodeStyleMappingComponent);
            TypeRegistrar.register("TreemapNodeStyleMapping", TreemapNodeStyleMapping.$type);
        }
    }
    IgxTreemapNodeStyleMappingDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTreemapNodeStyleMappingDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingDynamicModule, imports: [CommonModule, IgxTreemapNodeStyleMappingModule], exports: [IgxTreemapNodeStyleMappingModule] });
    IgxTreemapNodeStyleMappingDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingDynamicModule, imports: [[CommonModule, IgxTreemapNodeStyleMappingModule], IgxTreemapNodeStyleMappingModule] });
    return IgxTreemapNodeStyleMappingDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxTreemapNodeStyleMappingModule],
                    exports: [IgxTreemapNodeStyleMappingModule],
                    entryComponents: [IgxTreemapNodeStyleMappingComponent]
                }]
        }], ctorParameters: function () { return []; } });
