// igniteui-angular-charts/esm2015/lib/igx-data-chart-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumberFormatSpecifierDynamicModule } from "igniteui-angular-core";
import { IgxDateTimeFormatSpecifierDynamicModule } from "igniteui-angular-core";
import { IgxChartSelectionDynamicModule } from './igx-chart-selection-dynamic-module';
import { IgxSeriesMatcherDynamicModule } from './igx-series-matcher-dynamic-module';
import { IgxDataChartComponent } from './igx-data-chart-component';
import { XamDataChart } from './XamDataChart';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDataChartCoreModule } from './igx-data-chart-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartCoreDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('IgxDataChartComponent', IgxDataChartComponent);
            TypeRegistrar.registerCons('XamDataChart', XamDataChart);
            TypeRegistrar.register('XamDataChart', XamDataChart.$type);
        }
    }
    IgxDataChartCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCoreDynamicModule, imports: [CommonModule, IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxChartSelectionDynamicModule,
            IgxSeriesMatcherDynamicModule,
            IgxDataChartCoreModule], exports: [IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxChartSelectionDynamicModule,
            IgxSeriesMatcherDynamicModule,
            IgxDataChartCoreModule] });
    IgxDataChartCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCoreDynamicModule, imports: [[CommonModule, IgxNumberFormatSpecifierDynamicModule,
                IgxDateTimeFormatSpecifierDynamicModule,
                IgxChartSelectionDynamicModule,
                IgxSeriesMatcherDynamicModule,
                IgxDataChartCoreModule
            ], IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxChartSelectionDynamicModule,
            IgxSeriesMatcherDynamicModule,
            IgxDataChartCoreModule] });
    return IgxDataChartCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumberFormatSpecifierDynamicModule,
                        IgxDateTimeFormatSpecifierDynamicModule,
                        IgxChartSelectionDynamicModule,
                        IgxSeriesMatcherDynamicModule,
                        IgxDataChartCoreModule
                    ],
                    imports: [CommonModule, IgxNumberFormatSpecifierDynamicModule,
                        IgxDateTimeFormatSpecifierDynamicModule,
                        IgxChartSelectionDynamicModule,
                        IgxSeriesMatcherDynamicModule,
                        IgxDataChartCoreModule
                    ],
                    entryComponents: [IgxDataChartComponent]
                }]
        }], ctorParameters: function () { return []; } });
