// igniteui-angular-charts/esm2015/lib/igx-accumulation-distribution-indicator-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import { IgxFinancialIndicatorComponent } from "./igx-financial-indicator-component";
import { IgxFinancialSeriesComponent } from "./igx-financial-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { AccumulationDistributionIndicator } from "./AccumulationDistributionIndicator";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent Accumulation/Distribution indicator series.
 * Default required members: Close, Low, High, Volume
*
* You can use the `AccumulationDistributionIndicator` to evaluate the suppy and demand of a stock, security, or index.
*
* ```html
* <igx-data-chart
* 	#chart
*     [dataSource]="data" >
*     <igx-category-x-axis
* 		label="label"
* 		#xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
* 		#yAxis>
*     </igx-numeric-y-axis>
*     <igx-accumulation-distribution-indicator
* 		[xAxis]="xAxis"
* 		[yAxis]="yAxis"
* 		openMemberPath="open"
* 		highMemberPath="high"
* 		lowMemberPath="low"
* 		closeMemberPath="close">
*     </igx-accumulation-distribution-indicator>
* </igx-data-chart>
* ```
*
* ```ts
* let series = new IgxAccumulationDistributionIndicatorComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.openMemberPath = "open";
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* series.closeMemberPath = "close";
* this.chart.series.add(series);
* ```
*/
export let IgxAccumulationDistributionIndicatorComponent = /*@__PURE__*/ (() => {
    class IgxAccumulationDistributionIndicatorComponent extends IgxStrategyBasedIndicatorComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new AccumulationDistributionIndicator();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
    }
    IgxAccumulationDistributionIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAccumulationDistributionIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxAccumulationDistributionIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxAccumulationDistributionIndicatorComponent, selector: "igx-accumulation-distribution-indicator", providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(() => IgxAccumulationDistributionIndicatorComponent) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(() => IgxAccumulationDistributionIndicatorComponent) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(() => IgxAccumulationDistributionIndicatorComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxAccumulationDistributionIndicatorComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxAccumulationDistributionIndicatorComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAccumulationDistributionIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-accumulation-distribution-indicator',
                    template: ``,
                    providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(() => IgxAccumulationDistributionIndicatorComponent) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(() => IgxAccumulationDistributionIndicatorComponent) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(() => IgxAccumulationDistributionIndicatorComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxAccumulationDistributionIndicatorComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
