// igniteui-angular-charts/esm2015/lib/igx-range-area-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxHorizontalRangeCategorySeriesProxyModule } from './igx-horizontal-range-category-series-proxy-module';
import { IgxRangeAreaSeriesComponent } from './igx-range-area-series-component';
import * as i0 from "@angular/core";
export let IgxRangeAreaSeriesModule = /*@__PURE__*/ (() => {
    class IgxRangeAreaSeriesModule {
    }
    IgxRangeAreaSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeAreaSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxRangeAreaSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeAreaSeriesModule, declarations: [IgxRangeAreaSeriesComponent], imports: [CommonModule, IgxHorizontalRangeCategorySeriesProxyModule], exports: [IgxRangeAreaSeriesComponent, IgxHorizontalRangeCategorySeriesProxyModule] });
    IgxRangeAreaSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeAreaSeriesModule, imports: [[CommonModule, IgxHorizontalRangeCategorySeriesProxyModule], IgxHorizontalRangeCategorySeriesProxyModule] });
    return IgxRangeAreaSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeAreaSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxRangeAreaSeriesComponent],
                    imports: [CommonModule, IgxHorizontalRangeCategorySeriesProxyModule],
                    exports: [IgxRangeAreaSeriesComponent, IgxHorizontalRangeCategorySeriesProxyModule],
                    entryComponents: []
                }]
        }] });
