// igniteui-angular-charts/esm2015/lib/igx-legend-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxLegendComponent } from './igx-legend-component';
import { LegendTemplates } from './LegendTemplates';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxLegendModule } from './igx-legend-module';
import * as i0 from "@angular/core";
export let IgxLegendDynamicModule = /*@__PURE__*/ (() => {
    class IgxLegendDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('IgxLegendComponent', IgxLegendComponent);
            TypeRegistrar.registerCons('LegendTemplates', LegendTemplates);
            TypeRegistrar.register('LegendTemplates', LegendTemplates.$type);
        }
    }
    IgxLegendDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLegendDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxLegendDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLegendDynamicModule, imports: [CommonModule, IgxLegendModule], exports: [IgxLegendModule] });
    IgxLegendDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLegendDynamicModule, imports: [[CommonModule, IgxLegendModule
            ], IgxLegendModule] });
    return IgxLegendDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLegendDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxLegendModule
                    ],
                    imports: [CommonModule, IgxLegendModule
                    ],
                    entryComponents: [IgxLegendComponent]
                }]
        }], ctorParameters: function () { return []; } });
