// igniteui-angular-charts/esm2015/lib/igx-stacked-100-bar-series-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStackedBarSeriesComponent } from "./igx-stacked-bar-series-component";
import { IgxVerticalStackedSeriesBaseComponent } from "./igx-vertical-stacked-series-base-component";
import { IgxStackedSeriesBaseComponent } from "./igx-stacked-series-base-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { Stacked100BarSeries } from "./Stacked100BarSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent stacked100 bar series.
*/
export let IgxStacked100BarSeriesComponent = /*@__PURE__*/ (() => {
    class IgxStacked100BarSeriesComponent extends IgxStackedBarSeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new Stacked100BarSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        get isPercentBased() {
            return this.i.aan;
        }
    }
    IgxStacked100BarSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStacked100BarSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStacked100BarSeriesComponent, selector: "igx-stacked-100-bar-series", providers: [{ provide: IgxStackedBarSeriesComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }, { provide: IgxVerticalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStacked100BarSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-stacked-100-bar-series',
                    template: ``,
                    providers: [{ provide: IgxStackedBarSeriesComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }, { provide: IgxVerticalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100BarSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
