// igniteui-angular-charts/esm2015/lib/igx-line-fragment-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxLineFragmentModule } from './igx-line-fragment-module';
import { IgxLineFragmentComponent } from './igx-line-fragment-component';
import { LineFragment } from './LineFragment';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxLineFragmentDynamicModule = /*@__PURE__*/ (() => {
    class IgxLineFragmentDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxLineFragmentComponent", IgxLineFragmentComponent);
            TypeRegistrar.register("LineFragment", LineFragment.$type);
        }
    }
    IgxLineFragmentDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineFragmentDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxLineFragmentDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineFragmentDynamicModule, imports: [CommonModule, IgxLineFragmentModule], exports: [IgxLineFragmentModule] });
    IgxLineFragmentDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineFragmentDynamicModule, imports: [[CommonModule, IgxLineFragmentModule], IgxLineFragmentModule] });
    return IgxLineFragmentDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineFragmentDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxLineFragmentModule],
                    exports: [IgxLineFragmentModule],
                    entryComponents: [IgxLineFragmentComponent]
                }]
        }], ctorParameters: function () { return []; } });
