// igniteui-angular-charts/esm2015/lib/igx-doughnut-chart-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxRingSeriesDynamicModule } from './igx-ring-series-dynamic-module';
import { IgxDoughnutChartComponent } from './igx-doughnut-chart-component';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDoughnutChartCoreModule } from './igx-doughnut-chart-core-module';
import * as i0 from "@angular/core";
export let IgxDoughnutChartCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxDoughnutChartCoreDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('IgxDoughnutChartComponent', IgxDoughnutChartComponent);
        }
    }
    IgxDoughnutChartCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDoughnutChartCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartCoreDynamicModule, imports: [CommonModule, IgxRingSeriesDynamicModule,
            IgxDoughnutChartCoreModule], exports: [IgxRingSeriesDynamicModule,
            IgxDoughnutChartCoreModule] });
    IgxDoughnutChartCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartCoreDynamicModule, imports: [[CommonModule, IgxRingSeriesDynamicModule,
                IgxDoughnutChartCoreModule
            ], IgxRingSeriesDynamicModule,
            IgxDoughnutChartCoreModule] });
    return IgxDoughnutChartCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxRingSeriesDynamicModule,
                        IgxDoughnutChartCoreModule
                    ],
                    imports: [CommonModule, IgxRingSeriesDynamicModule,
                        IgxDoughnutChartCoreModule
                    ],
                    entryComponents: [IgxDoughnutChartComponent]
                }]
        }], ctorParameters: function () { return []; } });
