// igniteui-angular-charts/esm2015/lib/igx-spline-area-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSplineAreaSeriesModule } from './igx-spline-area-series-module';
import { IgxSplineAreaSeriesComponent } from './igx-spline-area-series-component';
import { SplineAreaSeries } from './SplineAreaSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxSplineAreaSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxSplineAreaSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxSplineAreaSeriesComponent", IgxSplineAreaSeriesComponent);
            TypeRegistrar.register("SplineAreaSeries", SplineAreaSeries.$type);
        }
    }
    IgxSplineAreaSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineAreaSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSplineAreaSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineAreaSeriesDynamicModule, imports: [CommonModule, IgxSplineAreaSeriesModule], exports: [IgxSplineAreaSeriesModule] });
    IgxSplineAreaSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineAreaSeriesDynamicModule, imports: [[CommonModule, IgxSplineAreaSeriesModule], IgxSplineAreaSeriesModule] });
    return IgxSplineAreaSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineAreaSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxSplineAreaSeriesModule],
                    exports: [IgxSplineAreaSeriesModule],
                    entryComponents: [IgxSplineAreaSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
