// igniteui-angular-charts/esm2015/lib/igx-overlays-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPriceChannelOverlayDynamicModule } from './igx-price-channel-overlay-dynamic-module';
import { IgxBollingerBandsOverlayDynamicModule } from './igx-bollinger-bands-overlay-dynamic-module';
import { IgxOverlaysModule } from './igx-overlays-module';
import * as i0 from "@angular/core";
export let IgxOverlaysDynamicModule = /*@__PURE__*/ (() => {
    class IgxOverlaysDynamicModule {
    }
    IgxOverlaysDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOverlaysDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxOverlaysDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOverlaysDynamicModule, imports: [CommonModule, IgxPriceChannelOverlayDynamicModule,
            IgxBollingerBandsOverlayDynamicModule,
            IgxOverlaysModule], exports: [IgxPriceChannelOverlayDynamicModule,
            IgxBollingerBandsOverlayDynamicModule,
            IgxOverlaysModule] });
    IgxOverlaysDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOverlaysDynamicModule, imports: [[CommonModule, IgxPriceChannelOverlayDynamicModule,
                IgxBollingerBandsOverlayDynamicModule,
                IgxOverlaysModule
            ], IgxPriceChannelOverlayDynamicModule,
            IgxBollingerBandsOverlayDynamicModule,
            IgxOverlaysModule] });
    return IgxOverlaysDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOverlaysDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxPriceChannelOverlayDynamicModule,
                        IgxBollingerBandsOverlayDynamicModule,
                        IgxOverlaysModule
                    ],
                    imports: [CommonModule, IgxPriceChannelOverlayDynamicModule,
                        IgxBollingerBandsOverlayDynamicModule,
                        IgxOverlaysModule
                    ],
                    entryComponents: []
                }]
        }] });
