// igniteui-angular-charts/esm2015/lib/igx-doughnut-chart-interactivity-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { InteractivityFactoryManager } from './InteractivityFactoryManager';
import { IgxDoughnutChartDefaultTooltipsComponent } from './igx-doughnut-chart-default-tooltips-component';
import { IgxDVInteractivityDynamicModule } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDoughnutChartInteractivityModule } from './igx-doughnut-chart-interactivity-module';
import * as i0 from "@angular/core";
export let IgxDoughnutChartInteractivityDynamicModule = /*@__PURE__*/ (() => {
    class IgxDoughnutChartInteractivityDynamicModule {
        constructor() {
            TypeRegistrar.register('InteractivityFactoryManager', InteractivityFactoryManager.$type);
            InteractivityFactoryManager.register();
            IgxDoughnutChartDefaultTooltipsComponent.register();
        }
    }
    IgxDoughnutChartInteractivityDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartInteractivityDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDoughnutChartInteractivityDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartInteractivityDynamicModule, imports: [CommonModule, IgxDVInteractivityDynamicModule,
            IgxDoughnutChartInteractivityModule], exports: [IgxDVInteractivityDynamicModule,
            IgxDoughnutChartInteractivityModule] });
    IgxDoughnutChartInteractivityDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartInteractivityDynamicModule, imports: [[CommonModule, IgxDVInteractivityDynamicModule,
                IgxDoughnutChartInteractivityModule
            ], IgxDVInteractivityDynamicModule,
            IgxDoughnutChartInteractivityModule] });
    return IgxDoughnutChartInteractivityDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartInteractivityDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDVInteractivityDynamicModule,
                        IgxDoughnutChartInteractivityModule
                    ],
                    imports: [CommonModule, IgxDVInteractivityDynamicModule,
                        IgxDoughnutChartInteractivityModule
                    ],
                    entryComponents: [IgxDoughnutChartDefaultTooltipsComponent]
                }]
        }], ctorParameters: function () { return []; } });
