// igniteui-angular-charts/esm2015/lib/igx-annotation-layer-proxy-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AnnotationLayerProxy } from './AnnotationLayerProxy';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxAnnotationLayerProxyModule = /*@__PURE__*/ (() => {
    class IgxAnnotationLayerProxyModule {
        constructor() {
            TypeRegistrar.register('AnnotationLayerProxy', AnnotationLayerProxy.$type);
        }
    }
    IgxAnnotationLayerProxyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAnnotationLayerProxyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAnnotationLayerProxyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAnnotationLayerProxyModule, imports: [CommonModule] });
    IgxAnnotationLayerProxyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAnnotationLayerProxyModule, imports: [[CommonModule,
            ]] });
    return IgxAnnotationLayerProxyModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAnnotationLayerProxyModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
