// igniteui-angular-charts/esm2015/lib/igx-data-legend-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataLegendComponent } from './igx-data-legend-component';
import { XamDataLegend } from './XamDataLegend';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxDataLegendModule = /*@__PURE__*/ (() => {
    class IgxDataLegendModule {
        constructor() {
            TypeRegistrar.registerCons('IgxDataLegendComponent', IgxDataLegendComponent);
            TypeRegistrar.register('XamDataLegend', XamDataLegend.$type);
        }
    }
    IgxDataLegendModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataLegendModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataLegendModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataLegendModule, declarations: [IgxDataLegendComponent], imports: [CommonModule], exports: [IgxDataLegendComponent] });
    IgxDataLegendModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataLegendModule, imports: [[CommonModule,
            ]] });
    return IgxDataLegendModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataLegendModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxDataLegendComponent],
                    exports: [IgxDataLegendComponent
                    ],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
