// igniteui-angular-charts/esm2015/lib/igx-scatter-spline-series-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxScatterBaseComponent } from "./igx-scatter-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { ScatterSplineSeries } from "./ScatterSplineSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent scatter spline series
*/
export let IgxScatterSplineSeriesComponent = /*@__PURE__*/ (() => {
    class IgxScatterSplineSeriesComponent extends IgxScatterBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new ScatterSplineSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows a spline shape.
        */
        get isSpline() {
            return this.i.fy;
        }
        /**
         * Gets or sets the Stiffness property.
        */
        get stiffness() {
            return this.i.acf;
        }
        set stiffness(v) {
            this.i.acf = +v;
        }
    }
    IgxScatterSplineSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSplineSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxScatterSplineSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxScatterSplineSeriesComponent, selector: "igx-scatter-spline-series", inputs: { stiffness: "stiffness" }, providers: [{ provide: IgxScatterBaseComponent, useExisting: forwardRef(() => IgxScatterSplineSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxScatterSplineSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxScatterSplineSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxScatterSplineSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSplineSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-scatter-spline-series',
                    template: ``,
                    providers: [{ provide: IgxScatterBaseComponent, useExisting: forwardRef(() => IgxScatterSplineSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxScatterSplineSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxScatterSplineSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { stiffness: [{
                type: Input
            }] } });
