// igniteui-angular-charts/esm2015/lib/igx-data-chart-category-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryXAxisDynamicModule } from './igx-category-x-axis-dynamic-module';
import { IgxNumericYAxisDynamicModule } from './igx-numeric-y-axis-dynamic-module';
import { IgxDataChartCoreDynamicModule } from './igx-data-chart-core-dynamic-module';
import { IgxDataChartCategoryCoreModule } from './igx-data-chart-category-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartCategoryCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartCategoryCoreDynamicModule {
    }
    IgxDataChartCategoryCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartCategoryCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryCoreDynamicModule, imports: [CommonModule, IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartCategoryCoreModule], exports: [IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartCategoryCoreModule] });
    IgxDataChartCategoryCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryCoreDynamicModule, imports: [[CommonModule, IgxCategoryXAxisDynamicModule,
                IgxNumericYAxisDynamicModule,
                IgxDataChartCoreDynamicModule,
                IgxDataChartCategoryCoreModule
            ], IgxCategoryXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartCategoryCoreModule] });
    return IgxDataChartCategoryCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxCategoryXAxisDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartCategoryCoreModule
                    ],
                    imports: [CommonModule, IgxCategoryXAxisDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartCategoryCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
