// igniteui-angular-charts/esm2015/lib/igx-chaikin-volatility-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxChaikinVolatilityIndicatorComponent } from './igx-chaikin-volatility-indicator-component';
import * as i0 from "@angular/core";
export let IgxChaikinVolatilityIndicatorModule = /*@__PURE__*/ (() => {
    class IgxChaikinVolatilityIndicatorModule {
    }
    IgxChaikinVolatilityIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinVolatilityIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxChaikinVolatilityIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinVolatilityIndicatorModule, declarations: [IgxChaikinVolatilityIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxChaikinVolatilityIndicatorComponent, IgxIndicatorProxyModule] });
    IgxChaikinVolatilityIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinVolatilityIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxChaikinVolatilityIndicatorModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChaikinVolatilityIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxChaikinVolatilityIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxChaikinVolatilityIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
