// igniteui-angular-charts/esm2015/lib/igx-treemap-node-style-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTreemapNodeStyleComponent } from './igx-treemap-node-style-component';
import * as i0 from "@angular/core";
export let IgxTreemapNodeStyleModule = /*@__PURE__*/ (() => {
    class IgxTreemapNodeStyleModule {
    }
    IgxTreemapNodeStyleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTreemapNodeStyleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleModule, declarations: [IgxTreemapNodeStyleComponent], imports: [CommonModule], exports: [IgxTreemapNodeStyleComponent] });
    IgxTreemapNodeStyleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleModule, imports: [[CommonModule]] });
    return IgxTreemapNodeStyleModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxTreemapNodeStyleComponent],
                    imports: [CommonModule],
                    exports: [IgxTreemapNodeStyleComponent],
                    entryComponents: []
                }]
        }] });
