// igniteui-angular-charts/esm2015/lib/igx-data-chart-radial-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryAngleAxisModule } from './igx-category-angle-axis-module';
import { IgxNumericRadiusAxisModule } from './igx-numeric-radius-axis-module';
import { IgxDataChartCoreModule } from './igx-data-chart-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartRadialCoreModule = /*@__PURE__*/ (() => {
    class IgxDataChartRadialCoreModule {
    }
    IgxDataChartRadialCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartRadialCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialCoreModule, imports: [CommonModule, IgxCategoryAngleAxisModule,
            IgxNumericRadiusAxisModule,
            IgxDataChartCoreModule], exports: [IgxCategoryAngleAxisModule,
            IgxNumericRadiusAxisModule,
            IgxDataChartCoreModule] });
    IgxDataChartRadialCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialCoreModule, imports: [[CommonModule, IgxCategoryAngleAxisModule,
                IgxNumericRadiusAxisModule,
                IgxDataChartCoreModule
            ], IgxCategoryAngleAxisModule,
            IgxNumericRadiusAxisModule,
            IgxDataChartCoreModule] });
    return IgxDataChartRadialCoreModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartRadialCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxCategoryAngleAxisModule,
                        IgxNumericRadiusAxisModule,
                        IgxDataChartCoreModule
                    ],
                    imports: [CommonModule, IgxCategoryAngleAxisModule,
                        IgxNumericRadiusAxisModule,
                        IgxDataChartCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
