// igniteui-angular-charts/esm2015/lib/igx-data-chart-financial-overlays-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxOverlaysDynamicModule } from './igx-overlays-dynamic-module';
import { IgxDataChartFinancialCoreDynamicModule } from './igx-data-chart-financial-core-dynamic-module';
import { IgxDataChartFinancialOverlaysModule } from './igx-data-chart-financial-overlays-module';
import * as i0 from "@angular/core";
export let IgxDataChartFinancialOverlaysDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartFinancialOverlaysDynamicModule {
    }
    IgxDataChartFinancialOverlaysDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialOverlaysDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartFinancialOverlaysDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialOverlaysDynamicModule, imports: [CommonModule, IgxOverlaysDynamicModule,
            IgxDataChartFinancialCoreDynamicModule,
            IgxDataChartFinancialOverlaysModule], exports: [IgxOverlaysDynamicModule,
            IgxDataChartFinancialCoreDynamicModule,
            IgxDataChartFinancialOverlaysModule] });
    IgxDataChartFinancialOverlaysDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialOverlaysDynamicModule, imports: [[CommonModule, IgxOverlaysDynamicModule,
                IgxDataChartFinancialCoreDynamicModule,
                IgxDataChartFinancialOverlaysModule
            ], IgxOverlaysDynamicModule,
            IgxDataChartFinancialCoreDynamicModule,
            IgxDataChartFinancialOverlaysModule] });
    return IgxDataChartFinancialOverlaysDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartFinancialOverlaysDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxOverlaysDynamicModule,
                        IgxDataChartFinancialCoreDynamicModule,
                        IgxDataChartFinancialOverlaysModule
                    ],
                    imports: [CommonModule, IgxOverlaysDynamicModule,
                        IgxDataChartFinancialCoreDynamicModule,
                        IgxDataChartFinancialOverlaysModule
                    ],
                    entryComponents: []
                }]
        }] });
