// igniteui-angular-charts/esm2015/lib/igx-financial-chart-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFinancialChartComponent } from './igx-financial-chart-component';
import { IgxFinancialChartDefaultTemplatesComponent } from './igx-financial-chart-default-templates-component';
import { IgxTemplateContentDynamicModule } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxFinancialChartCoreModule } from './igx-financial-chart-core-module';
import * as i0 from "@angular/core";
export let IgxFinancialChartCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxFinancialChartCoreDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('IgxFinancialChartComponent', IgxFinancialChartComponent);
            TypeRegistrar.registerCons('IgxFinancialChartDefaultTemplatesComponent', IgxFinancialChartDefaultTemplatesComponent);
        }
    }
    IgxFinancialChartCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFinancialChartCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartCoreDynamicModule, imports: [CommonModule, IgxTemplateContentDynamicModule,
            IgxFinancialChartCoreModule], exports: [IgxTemplateContentDynamicModule,
            IgxFinancialChartCoreModule] });
    IgxFinancialChartCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartCoreDynamicModule, imports: [[CommonModule, IgxTemplateContentDynamicModule,
                IgxFinancialChartCoreModule
            ], IgxTemplateContentDynamicModule,
            IgxFinancialChartCoreModule] });
    return IgxFinancialChartCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxTemplateContentDynamicModule,
                        IgxFinancialChartCoreModule
                    ],
                    imports: [CommonModule, IgxTemplateContentDynamicModule,
                        IgxFinancialChartCoreModule
                    ],
                    entryComponents: [IgxFinancialChartComponent,
                        IgxFinancialChartDefaultTemplatesComponent]
                }]
        }], ctorParameters: function () { return []; } });
