// igniteui-angular-charts/esm2015/lib/igx-custom-palette-brush-scale-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { BrushSelectionMode_$type } from "./BrushSelectionMode";
import { IgxBrushScaleComponent } from "./igx-brush-scale-component";
import { CustomPaletteBrushScale } from "./CustomPaletteBrushScale";
import { ensureEnum, brushToString } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a brush scale that uses index-based brush selection mode.
*/
export let IgxCustomPaletteBrushScaleComponent = /*@__PURE__*/ (() => {
    class IgxCustomPaletteBrushScaleComponent extends IgxBrushScaleComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new CustomPaletteBrushScale();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets or sets the brush selection mode.
        */
        get brushSelectionMode() {
            return this.i.brushSelectionMode;
        }
        set brushSelectionMode(v) {
            this.i.brushSelectionMode = ensureEnum(BrushSelectionMode_$type, v);
        }
        /**
         * Checks if this scale is ready for useage with a series
        */
        get isReady() {
            return this.i.isReady;
        }
        getBrush1(index, total) {
            let iv = this.i.getBrush1(index, total);
            return brushToString(iv);
        }
    }
    IgxCustomPaletteBrushScaleComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteBrushScaleComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxCustomPaletteBrushScaleComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxCustomPaletteBrushScaleComponent, selector: "igx-custom-palette-brush-scale", inputs: { brushSelectionMode: "brushSelectionMode" }, providers: [{ provide: IgxBrushScaleComponent, useExisting: forwardRef(() => IgxCustomPaletteBrushScaleComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxCustomPaletteBrushScaleComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteBrushScaleComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-custom-palette-brush-scale',
                    template: ``,
                    providers: [{ provide: IgxBrushScaleComponent, useExisting: forwardRef(() => IgxCustomPaletteBrushScaleComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { brushSelectionMode: [{
                type: Input
            }] } });
