// igniteui-angular-charts/esm2015/lib/igx-time-x-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTimeAxisBreakModule } from './igx-time-axis-break-module';
import { IgxTimeAxisIntervalModule } from './igx-time-axis-interval-module';
import { IgxTimeAxisLabelFormatModule } from './igx-time-axis-label-format-module';
import { IgxTimeXAxisComponent } from './igx-time-x-axis-component';
import * as i0 from "@angular/core";
export let IgxTimeXAxisModule = /*@__PURE__*/ (() => {
    class IgxTimeXAxisModule {
    }
    IgxTimeXAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeXAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTimeXAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeXAxisModule, declarations: [IgxTimeXAxisComponent], imports: [CommonModule, IgxTimeAxisBreakModule, IgxTimeAxisIntervalModule, IgxTimeAxisLabelFormatModule], exports: [IgxTimeXAxisComponent, IgxTimeAxisBreakModule, IgxTimeAxisIntervalModule, IgxTimeAxisLabelFormatModule] });
    IgxTimeXAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeXAxisModule, imports: [[CommonModule, IgxTimeAxisBreakModule, IgxTimeAxisIntervalModule, IgxTimeAxisLabelFormatModule], IgxTimeAxisBreakModule, IgxTimeAxisIntervalModule, IgxTimeAxisLabelFormatModule] });
    return IgxTimeXAxisModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeXAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxTimeXAxisComponent],
                    imports: [CommonModule, IgxTimeAxisBreakModule, IgxTimeAxisIntervalModule, IgxTimeAxisLabelFormatModule],
                    exports: [IgxTimeXAxisComponent, IgxTimeAxisBreakModule, IgxTimeAxisIntervalModule, IgxTimeAxisLabelFormatModule],
                    entryComponents: []
                }]
        }] });
