// igniteui-angular-charts/esm2015/lib/igx-scatter-area-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCustomPaletteColorScaleModule } from './igx-custom-palette-color-scale-module';
import { IgxScatterAreaSeriesComponent } from './igx-scatter-area-series-component';
import * as i0 from "@angular/core";
export let IgxScatterAreaSeriesModule = /*@__PURE__*/ (() => {
    class IgxScatterAreaSeriesModule {
    }
    IgxScatterAreaSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterAreaSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterAreaSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterAreaSeriesModule, declarations: [IgxScatterAreaSeriesComponent], imports: [CommonModule, IgxCustomPaletteColorScaleModule], exports: [IgxScatterAreaSeriesComponent, IgxCustomPaletteColorScaleModule] });
    IgxScatterAreaSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterAreaSeriesModule, imports: [[CommonModule, IgxCustomPaletteColorScaleModule], IgxCustomPaletteColorScaleModule] });
    return IgxScatterAreaSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterAreaSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxScatterAreaSeriesComponent],
                    imports: [CommonModule, IgxCustomPaletteColorScaleModule],
                    exports: [IgxScatterAreaSeriesComponent, IgxCustomPaletteColorScaleModule],
                    entryComponents: []
                }]
        }] });
