// igniteui-angular-charts/esm2015/lib/igx-category-x-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryXAxisModule } from './igx-category-x-axis-module';
import { IgxCategoryXAxisComponent } from './igx-category-x-axis-component';
import { CategoryXAxis } from './CategoryXAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxCategoryXAxisDynamicModule = /*@__PURE__*/ (() => {
    class IgxCategoryXAxisDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxCategoryXAxisComponent", IgxCategoryXAxisComponent);
            TypeRegistrar.register("CategoryXAxis", CategoryXAxis.$type);
        }
    }
    IgxCategoryXAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryXAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryXAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryXAxisDynamicModule, imports: [CommonModule, IgxCategoryXAxisModule], exports: [IgxCategoryXAxisModule] });
    IgxCategoryXAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryXAxisDynamicModule, imports: [[CommonModule, IgxCategoryXAxisModule], IgxCategoryXAxisModule] });
    return IgxCategoryXAxisDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryXAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCategoryXAxisModule],
                    exports: [IgxCategoryXAxisModule],
                    entryComponents: [IgxCategoryXAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
