// igniteui-angular-charts/esm2015/lib/igx-price-volume-trend-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPriceVolumeTrendIndicatorModule } from './igx-price-volume-trend-indicator-module';
import { IgxPriceVolumeTrendIndicatorComponent } from './igx-price-volume-trend-indicator-component';
import { PriceVolumeTrendIndicator } from './PriceVolumeTrendIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxPriceVolumeTrendIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxPriceVolumeTrendIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxPriceVolumeTrendIndicatorComponent", IgxPriceVolumeTrendIndicatorComponent);
            TypeRegistrar.register("PriceVolumeTrendIndicator", PriceVolumeTrendIndicator.$type);
        }
    }
    IgxPriceVolumeTrendIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceVolumeTrendIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPriceVolumeTrendIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceVolumeTrendIndicatorDynamicModule, imports: [CommonModule, IgxPriceVolumeTrendIndicatorModule], exports: [IgxPriceVolumeTrendIndicatorModule] });
    IgxPriceVolumeTrendIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceVolumeTrendIndicatorDynamicModule, imports: [[CommonModule, IgxPriceVolumeTrendIndicatorModule], IgxPriceVolumeTrendIndicatorModule] });
    return IgxPriceVolumeTrendIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceVolumeTrendIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPriceVolumeTrendIndicatorModule],
                    exports: [IgxPriceVolumeTrendIndicatorModule],
                    entryComponents: [IgxPriceVolumeTrendIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
