// igniteui-angular-charts/esm2015/lib/igx-price-channel-overlay-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPriceChannelOverlayModule } from './igx-price-channel-overlay-module';
import { IgxPriceChannelOverlayComponent } from './igx-price-channel-overlay-component';
import { PriceChannelOverlay } from './PriceChannelOverlay';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxPriceChannelOverlayDynamicModule = /*@__PURE__*/ (() => {
    class IgxPriceChannelOverlayDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxPriceChannelOverlayComponent", IgxPriceChannelOverlayComponent);
            TypeRegistrar.register("PriceChannelOverlay", PriceChannelOverlay.$type);
        }
    }
    IgxPriceChannelOverlayDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceChannelOverlayDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPriceChannelOverlayDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceChannelOverlayDynamicModule, imports: [CommonModule, IgxPriceChannelOverlayModule], exports: [IgxPriceChannelOverlayModule] });
    IgxPriceChannelOverlayDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceChannelOverlayDynamicModule, imports: [[CommonModule, IgxPriceChannelOverlayModule], IgxPriceChannelOverlayModule] });
    return IgxPriceChannelOverlayDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPriceChannelOverlayDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPriceChannelOverlayModule],
                    exports: [IgxPriceChannelOverlayModule],
                    entryComponents: [IgxPriceChannelOverlayComponent]
                }]
        }], ctorParameters: function () { return []; } });
