// igniteui-angular-charts/esm2015/lib/igx-custom-palette-color-scale-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCustomPaletteColorScaleComponent } from './igx-custom-palette-color-scale-component';
import * as i0 from "@angular/core";
export let IgxCustomPaletteColorScaleModule = /*@__PURE__*/ (() => {
    class IgxCustomPaletteColorScaleModule {
    }
    IgxCustomPaletteColorScaleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteColorScaleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCustomPaletteColorScaleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteColorScaleModule, declarations: [IgxCustomPaletteColorScaleComponent], imports: [CommonModule], exports: [IgxCustomPaletteColorScaleComponent] });
    IgxCustomPaletteColorScaleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteColorScaleModule, imports: [[CommonModule]] });
    return IgxCustomPaletteColorScaleModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteColorScaleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxCustomPaletteColorScaleComponent],
                    imports: [CommonModule],
                    exports: [IgxCustomPaletteColorScaleComponent],
                    entryComponents: []
                }]
        }] });
