// igniteui-angular-charts/esm2015/lib/igx-radial-line-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxRadialLineSeriesModule } from './igx-radial-line-series-module';
import { IgxRadialLineSeriesComponent } from './igx-radial-line-series-component';
import { RadialLineSeries } from './RadialLineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxRadialLineSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxRadialLineSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxRadialLineSeriesComponent", IgxRadialLineSeriesComponent);
            TypeRegistrar.register("RadialLineSeries", RadialLineSeries.$type);
        }
    }
    IgxRadialLineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialLineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxRadialLineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialLineSeriesDynamicModule, imports: [CommonModule, IgxRadialLineSeriesModule], exports: [IgxRadialLineSeriesModule] });
    IgxRadialLineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialLineSeriesDynamicModule, imports: [[CommonModule, IgxRadialLineSeriesModule], IgxRadialLineSeriesModule] });
    return IgxRadialLineSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialLineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxRadialLineSeriesModule],
                    exports: [IgxRadialLineSeriesModule],
                    entryComponents: [IgxRadialLineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
