// igniteui-angular-charts/esm2015/lib/igx-spline-fragment-base-component.js
import { Component } from '@angular/core';
import { IgxSplineSeriesBaseComponent } from "./igx-spline-series-base-component";
import { toPoint, fromPoint } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents one part in a StackedSplineSeries or StackedSplineAreaSeries.
*/
export let IgxSplineFragmentBaseComponent = /*@__PURE__*/ (() => {
    class IgxSplineFragmentBaseComponent extends IgxSplineSeriesBaseComponent {
        constructor() {
            super();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series is a stack fragment.
        */
        get isFragment() {
            return this.i.e9;
        }
        getSeriesValuePosition(world, useInterpolation, skipUnknowns) {
            let iv = this.i.v6(toPoint(world), useInterpolation, skipUnknowns);
            return fromPoint(iv);
        }
    }
    IgxSplineFragmentBaseComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineFragmentBaseComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxSplineFragmentBaseComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxSplineFragmentBaseComponent, selector: "ng-component", usesInheritance: true, ngImport: i0, template: ``, isInline: true });
    return IgxSplineFragmentBaseComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineFragmentBaseComponent, decorators: [{
            type: Component,
            args: [{
                    template: ``,
                }]
        }], ctorParameters: function () { return []; } });
