// igniteui-angular-charts/esm2015/lib/igx-linear-contour-value-resolver-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxContourValueResolverComponent } from "./igx-contour-value-resolver-component";
import { LinearContourValueResolver } from "./LinearContourValueResolver";
import * as i0 from "@angular/core";
/**
 * ContourValueResolver class which resolves contour lines to be distributed at linear intervals from the data minimum to the data maximum.
*/
export let IgxLinearContourValueResolverComponent = /*@__PURE__*/ (() => {
    class IgxLinearContourValueResolverComponent extends IgxContourValueResolverComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new LinearContourValueResolver();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets or sets how many contour values to return.
        */
        get valueCount() {
            return this.i.valueCount;
        }
        set valueCount(v) {
            this.i.valueCount = +v;
        }
    }
    IgxLinearContourValueResolverComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxLinearContourValueResolverComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxLinearContourValueResolverComponent, selector: "igx-linear-contour-value-resolver", inputs: { valueCount: "valueCount" }, providers: [{ provide: IgxContourValueResolverComponent, useExisting: forwardRef(() => IgxLinearContourValueResolverComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxLinearContourValueResolverComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLinearContourValueResolverComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-linear-contour-value-resolver',
                    template: ``,
                    providers: [{ provide: IgxContourValueResolverComponent, useExisting: forwardRef(() => IgxLinearContourValueResolverComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { valueCount: [{
                type: Input
            }] } });
