// igniteui-angular-charts/esm2015/lib/igx-stacked-fragment-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedFragmentSeriesModule } from './igx-stacked-fragment-series-module';
import { IgxStackedFragmentSeriesComponent } from './igx-stacked-fragment-series-component';
import { StackedFragmentSeries } from './StackedFragmentSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxStackedFragmentSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxStackedFragmentSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxStackedFragmentSeriesComponent", IgxStackedFragmentSeriesComponent);
            TypeRegistrar.register("StackedFragmentSeries", StackedFragmentSeries.$type);
        }
    }
    IgxStackedFragmentSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedFragmentSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedFragmentSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedFragmentSeriesDynamicModule, imports: [CommonModule, IgxStackedFragmentSeriesModule], exports: [IgxStackedFragmentSeriesModule] });
    IgxStackedFragmentSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedFragmentSeriesDynamicModule, imports: [[CommonModule, IgxStackedFragmentSeriesModule], IgxStackedFragmentSeriesModule] });
    return IgxStackedFragmentSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedFragmentSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStackedFragmentSeriesModule],
                    exports: [IgxStackedFragmentSeriesModule],
                    entryComponents: [IgxStackedFragmentSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
