// igniteui-angular-charts/esm2015/lib/igx-radial-area-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxRadialAreaSeriesComponent } from './igx-radial-area-series-component';
import * as i0 from "@angular/core";
export let IgxRadialAreaSeriesModule = /*@__PURE__*/ (() => {
    class IgxRadialAreaSeriesModule {
    }
    IgxRadialAreaSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialAreaSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxRadialAreaSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialAreaSeriesModule, declarations: [IgxRadialAreaSeriesComponent], imports: [CommonModule], exports: [IgxRadialAreaSeriesComponent] });
    IgxRadialAreaSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialAreaSeriesModule, imports: [[CommonModule]] });
    return IgxRadialAreaSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialAreaSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxRadialAreaSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxRadialAreaSeriesComponent],
                    entryComponents: []
                }]
        }] });
