// igniteui-angular-charts/esm2015/lib/igx-proportional-category-angle-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxProportionalCategoryAngleAxisModule } from './igx-proportional-category-angle-axis-module';
import { IgxProportionalCategoryAngleAxisComponent } from './igx-proportional-category-angle-axis-component';
import { ProportionalCategoryAngleAxis } from './ProportionalCategoryAngleAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxProportionalCategoryAngleAxisDynamicModule = /*@__PURE__*/ (() => {
    class IgxProportionalCategoryAngleAxisDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxProportionalCategoryAngleAxisComponent", IgxProportionalCategoryAngleAxisComponent);
            TypeRegistrar.register("ProportionalCategoryAngleAxis", ProportionalCategoryAngleAxis.$type);
        }
    }
    IgxProportionalCategoryAngleAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxProportionalCategoryAngleAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxProportionalCategoryAngleAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxProportionalCategoryAngleAxisDynamicModule, imports: [CommonModule, IgxProportionalCategoryAngleAxisModule], exports: [IgxProportionalCategoryAngleAxisModule] });
    IgxProportionalCategoryAngleAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxProportionalCategoryAngleAxisDynamicModule, imports: [[CommonModule, IgxProportionalCategoryAngleAxisModule], IgxProportionalCategoryAngleAxisModule] });
    return IgxProportionalCategoryAngleAxisDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxProportionalCategoryAngleAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxProportionalCategoryAngleAxisModule],
                    exports: [IgxProportionalCategoryAngleAxisModule],
                    entryComponents: [IgxProportionalCategoryAngleAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
