// igniteui-angular-charts/esm2015/lib/igx-crosshair-layer-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxSeriesComponent } from "./igx-series-component";
import { Visibility_$type } from "igniteui-angular-core";
import { IgxAnnotationLayerComponent } from "./igx-annotation-layer-component";
import { CrosshairLayer } from "./CrosshairLayer";
import { ensureBool, brushToString, stringToBrush, ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents an annotation layer that displays crosshair lines that cross through the closest value of the target series under the cursor.
*/
export let IgxCrosshairLayerComponent = /*@__PURE__*/ (() => {
    class IgxCrosshairLayerComponent extends IgxAnnotationLayerComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new CrosshairLayer();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the series is an annotation layer displayed only when hovering over the chart.
        */
        get isAnnotationHoverLayer() {
            return this.i.ep;
        }
        /**
         * Gets whether the series is an crosshair annotation layer.
        */
        get isAnnotationCrosshairLayer() {
            return this.i.en;
        }
        /**
         * Gets or sets the color to use for the horizontal line. Leave null for an automatic value.
        */
        get horizontalLineStroke() {
            return brushToString(this.i.ab4);
        }
        set horizontalLineStroke(v) {
            this.i.ab4 = stringToBrush(v);
        }
        /**
         * Gets or sets the color to use for vertical line. Leave null for an automatic value.
        */
        get verticalLineStroke() {
            return brushToString(this.i.ab5);
        }
        set verticalLineStroke(v) {
            this.i.ab5 = stringToBrush(v);
        }
        /**
         * Gets or sets the name of the series to target this annotation to. If null, this annotation targets all series simultaneously.
        */
        get targetSeriesName() {
            return this.i.aai;
        }
        set targetSeriesName(v) {
            this.i.aai = v;
        }
        /**
         * Gets or sets the series to target this annotation to. If null, this annotation targets all series simultaneously.
        */
        get targetSeries() {
            const r = this.i.zs;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                let e = IgxSeriesComponent._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        }
        set targetSeries(v) {
            if (v != null && this._stylingContainer && v._styling)
                v._styling(this._stylingContainer, this, this);
            v == null ? this.i.zs = null : this.i.zs = v.i;
        }
        /**
         * Gets or sets whether to use value interpolation when drawing a line through the best value for the cursor position.
        */
        get useInterpolation() {
            return this.i.zx;
        }
        set useInterpolation(v) {
            this.i.zx = ensureBool(v);
        }
        /**
         * Gets or sets whether to draw annotations over the axes where the crosshair meets with them.
        */
        get isAxisAnnotationEnabled() {
            return this.i.zv;
        }
        set isAxisAnnotationEnabled(v) {
            this.i.zv = ensureBool(v);
        }
        /**
         * Gets or sets the color to use for the x axis annotation text. Leave unset for an automatic value.
        */
        get xAxisAnnotationTextColor() {
            return brushToString(this.i.ab8);
        }
        set xAxisAnnotationTextColor(v) {
            this.i.ab8 = stringToBrush(v);
        }
        /**
         * Gets or sets the color to use for the x axis annotation backing. Leave unset for an automatic value.
        */
        get xAxisAnnotationBackground() {
            return brushToString(this.i.ab6);
        }
        set xAxisAnnotationBackground(v) {
            this.i.ab6 = stringToBrush(v);
        }
        /**
         * Gets or sets the corner radius to use for the x axis annotation backing. Leave unset for an automatic value.
        */
        get xAxisAnnotationBackgroundCornerRadius() {
            return this.i.z0;
        }
        set xAxisAnnotationBackgroundCornerRadius(v) {
            this.i.z0 = +v;
        }
        /**
         * Gets or sets the precision to use displaying values for interpolated crosshair positions.
        */
        get xAxisAnnotationInterpolatedValuePrecision() {
            return this.i.aac;
        }
        set xAxisAnnotationInterpolatedValuePrecision(v) {
            this.i.aac = +v;
        }
        /**
         * Gets or sets the color to use for the x axis annotation outline. Leave unset for an automatic value.
        */
        get xAxisAnnotationOutline() {
            return brushToString(this.i.ab7);
        }
        set xAxisAnnotationOutline(v) {
            this.i.ab7 = stringToBrush(v);
        }
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get xAxisAnnotationPaddingLeft() {
            return this.i.z2;
        }
        set xAxisAnnotationPaddingLeft(v) {
            this.i.z2 = +v;
        }
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get xAxisAnnotationPaddingTop() {
            return this.i.z4;
        }
        set xAxisAnnotationPaddingTop(v) {
            this.i.z4 = +v;
        }
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get xAxisAnnotationPaddingRight() {
            return this.i.z3;
        }
        set xAxisAnnotationPaddingRight(v) {
            this.i.z3 = +v;
        }
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get xAxisAnnotationPaddingBottom() {
            return this.i.z1;
        }
        set xAxisAnnotationPaddingBottom(v) {
            this.i.z1 = +v;
        }
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get yAxisAnnotationPaddingLeft() {
            return this.i.z8;
        }
        set yAxisAnnotationPaddingLeft(v) {
            this.i.z8 = +v;
        }
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get yAxisAnnotationPaddingTop() {
            return this.i.aaa;
        }
        set yAxisAnnotationPaddingTop(v) {
            this.i.aaa = +v;
        }
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get yAxisAnnotationPaddingRight() {
            return this.i.z9;
        }
        set yAxisAnnotationPaddingRight(v) {
            this.i.z9 = +v;
        }
        /**
         * Gets or sets the padding to use withing the axis annotation callout. Leaving this NaN will use an automatic value related to the axis label margins.
        */
        get yAxisAnnotationPaddingBottom() {
            return this.i.z7;
        }
        set yAxisAnnotationPaddingBottom(v) {
            this.i.z7 = +v;
        }
        /**
         * Gets or sets the stroke thickness for the x axis annotation backing. Leave unset for an automatic value.
        */
        get xAxisAnnotationStrokeThickness() {
            return this.i.z5;
        }
        set xAxisAnnotationStrokeThickness(v) {
            this.i.z5 = +v;
        }
        /**
         * Gets or sets the color to use for the y axis annotation text. Leave unset for an automatic value.
        */
        get yAxisAnnotationTextColor() {
            return brushToString(this.i.acb);
        }
        set yAxisAnnotationTextColor(v) {
            this.i.acb = stringToBrush(v);
        }
        /**
         * Gets or sets the color to use for the y axis annotation backing. Leave unset for an automatic value.
        */
        get yAxisAnnotationBackground() {
            return brushToString(this.i.ab9);
        }
        set yAxisAnnotationBackground(v) {
            this.i.ab9 = stringToBrush(v);
        }
        /**
         * Gets or sets the corner radius to use for the y axis annotation backing. Leave unset for an automatic value.
        */
        get yAxisAnnotationBackgroundCornerRadius() {
            return this.i.z6;
        }
        set yAxisAnnotationBackgroundCornerRadius(v) {
            this.i.z6 = +v;
        }
        /**
         * Gets or sets the precision to use displaying values for interpolated crosshair positions.
        */
        get yAxisAnnotationInterpolatedValuePrecision() {
            return this.i.aad;
        }
        set yAxisAnnotationInterpolatedValuePrecision(v) {
            this.i.aad = +v;
        }
        /**
         * Gets or sets the color to use for the y axis annotation outline. Leave unset for an automatic value.
        */
        get yAxisAnnotationOutline() {
            return brushToString(this.i.aca);
        }
        set yAxisAnnotationOutline(v) {
            this.i.aca = stringToBrush(v);
        }
        /**
         * Gets or sets the stroke thickness for the y axis annotation backing. Leave unset for an automatic value.
        */
        get yAxisAnnotationStrokeThickness() {
            return this.i.aab;
        }
        set yAxisAnnotationStrokeThickness(v) {
            this.i.aab = +v;
        }
        /**
         * Gets or sets whether to display the vertical line.
        */
        get verticalLineVisibility() {
            return this.i.ach;
        }
        set verticalLineVisibility(v) {
            this.i.ach = ensureEnum(Visibility_$type, v);
        }
        /**
         * Gets or sets whether to display the horizontal line.
        */
        get horizontalLineVisibility() {
            return this.i.acg;
        }
        set horizontalLineVisibility(v) {
            this.i.acg = ensureEnum(Visibility_$type, v);
        }
        /**
         * Gets or sets whether to skip unknown values when searching for series values.
        */
        get skipUnknownValues() {
            return this.i.zw;
        }
        set skipUnknownValues(v) {
            this.i.zw = ensureBool(v);
        }
        findByName(name) {
            var baseResult = super.findByName(name);
            if (baseResult) {
                return baseResult;
            }
            if (this.targetSeries && this.targetSeries.name && this.targetSeries.name == name) {
                return this.targetSeries;
            }
            return null;
        }
        _styling(container, component, parent) {
            super._styling(container, component, parent);
            this._inStyling = true;
            if (this.targetSeries && this.targetSeries._styling) {
                this.targetSeries._styling(container, component, this);
            }
            this._inStyling = false;
        }
    }
    IgxCrosshairLayerComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCrosshairLayerComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxCrosshairLayerComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxCrosshairLayerComponent, selector: "igx-crosshair-layer", inputs: { horizontalLineStroke: "horizontalLineStroke", verticalLineStroke: "verticalLineStroke", targetSeriesName: "targetSeriesName", targetSeries: "targetSeries", useInterpolation: "useInterpolation", isAxisAnnotationEnabled: "isAxisAnnotationEnabled", xAxisAnnotationTextColor: "xAxisAnnotationTextColor", xAxisAnnotationBackground: "xAxisAnnotationBackground", xAxisAnnotationBackgroundCornerRadius: "xAxisAnnotationBackgroundCornerRadius", xAxisAnnotationInterpolatedValuePrecision: "xAxisAnnotationInterpolatedValuePrecision", xAxisAnnotationOutline: "xAxisAnnotationOutline", xAxisAnnotationPaddingLeft: "xAxisAnnotationPaddingLeft", xAxisAnnotationPaddingTop: "xAxisAnnotationPaddingTop", xAxisAnnotationPaddingRight: "xAxisAnnotationPaddingRight", xAxisAnnotationPaddingBottom: "xAxisAnnotationPaddingBottom", yAxisAnnotationPaddingLeft: "yAxisAnnotationPaddingLeft", yAxisAnnotationPaddingTop: "yAxisAnnotationPaddingTop", yAxisAnnotationPaddingRight: "yAxisAnnotationPaddingRight", yAxisAnnotationPaddingBottom: "yAxisAnnotationPaddingBottom", xAxisAnnotationStrokeThickness: "xAxisAnnotationStrokeThickness", yAxisAnnotationTextColor: "yAxisAnnotationTextColor", yAxisAnnotationBackground: "yAxisAnnotationBackground", yAxisAnnotationBackgroundCornerRadius: "yAxisAnnotationBackgroundCornerRadius", yAxisAnnotationInterpolatedValuePrecision: "yAxisAnnotationInterpolatedValuePrecision", yAxisAnnotationOutline: "yAxisAnnotationOutline", yAxisAnnotationStrokeThickness: "yAxisAnnotationStrokeThickness", verticalLineVisibility: "verticalLineVisibility", horizontalLineVisibility: "horizontalLineVisibility", skipUnknownValues: "skipUnknownValues" }, providers: [{ provide: IgxAnnotationLayerComponent, useExisting: forwardRef(() => IgxCrosshairLayerComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxCrosshairLayerComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxCrosshairLayerComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCrosshairLayerComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-crosshair-layer',
                    template: ``,
                    providers: [{ provide: IgxAnnotationLayerComponent, useExisting: forwardRef(() => IgxCrosshairLayerComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxCrosshairLayerComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { horizontalLineStroke: [{
                type: Input
            }], verticalLineStroke: [{
                type: Input
            }], targetSeriesName: [{
                type: Input
            }], targetSeries: [{
                type: Input
            }], useInterpolation: [{
                type: Input
            }], isAxisAnnotationEnabled: [{
                type: Input
            }], xAxisAnnotationTextColor: [{
                type: Input
            }], xAxisAnnotationBackground: [{
                type: Input
            }], xAxisAnnotationBackgroundCornerRadius: [{
                type: Input
            }], xAxisAnnotationInterpolatedValuePrecision: [{
                type: Input
            }], xAxisAnnotationOutline: [{
                type: Input
            }], xAxisAnnotationPaddingLeft: [{
                type: Input
            }], xAxisAnnotationPaddingTop: [{
                type: Input
            }], xAxisAnnotationPaddingRight: [{
                type: Input
            }], xAxisAnnotationPaddingBottom: [{
                type: Input
            }], yAxisAnnotationPaddingLeft: [{
                type: Input
            }], yAxisAnnotationPaddingTop: [{
                type: Input
            }], yAxisAnnotationPaddingRight: [{
                type: Input
            }], yAxisAnnotationPaddingBottom: [{
                type: Input
            }], xAxisAnnotationStrokeThickness: [{
                type: Input
            }], yAxisAnnotationTextColor: [{
                type: Input
            }], yAxisAnnotationBackground: [{
                type: Input
            }], yAxisAnnotationBackgroundCornerRadius: [{
                type: Input
            }], yAxisAnnotationInterpolatedValuePrecision: [{
                type: Input
            }], yAxisAnnotationOutline: [{
                type: Input
            }], yAxisAnnotationStrokeThickness: [{
                type: Input
            }], verticalLineVisibility: [{
                type: Input
            }], horizontalLineVisibility: [{
                type: Input
            }], skipUnknownValues: [{
                type: Input
            }] } });
