// igniteui-angular-charts/esm2015/lib/igx-area-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxHorizontalAnchoredCategorySeriesProxyModule } from './igx-horizontal-anchored-category-series-proxy-module';
import { IgxAreaSeriesComponent } from './igx-area-series-component';
import * as i0 from "@angular/core";
export let IgxAreaSeriesModule = /*@__PURE__*/ (() => {
    class IgxAreaSeriesModule {
    }
    IgxAreaSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAreaSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaSeriesModule, declarations: [IgxAreaSeriesComponent], imports: [CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule], exports: [IgxAreaSeriesComponent, IgxHorizontalAnchoredCategorySeriesProxyModule] });
    IgxAreaSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaSeriesModule, imports: [[CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule], IgxHorizontalAnchoredCategorySeriesProxyModule] });
    return IgxAreaSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxAreaSeriesComponent],
                    imports: [CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule],
                    exports: [IgxAreaSeriesComponent, IgxHorizontalAnchoredCategorySeriesProxyModule],
                    entryComponents: []
                }]
        }] });
