// igniteui-angular-charts/esm2015/lib/igx-financial-chart-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFinancialChartComponent } from './igx-financial-chart-component';
import { IgxFinancialChartDefaultTemplatesComponent } from './igx-financial-chart-default-templates-component';
import { IgxTemplateContentModule } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxFinancialChartCoreModule = /*@__PURE__*/ (() => {
    class IgxFinancialChartCoreModule {
        constructor() {
        }
    }
    IgxFinancialChartCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFinancialChartCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartCoreModule, declarations: [IgxFinancialChartComponent,
            IgxFinancialChartDefaultTemplatesComponent], imports: [CommonModule, IgxTemplateContentModule], exports: [IgxFinancialChartComponent,
            IgxFinancialChartDefaultTemplatesComponent,
            IgxTemplateContentModule] });
    IgxFinancialChartCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartCoreModule, imports: [[CommonModule, IgxTemplateContentModule
            ], IgxTemplateContentModule] });
    return IgxFinancialChartCoreModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxFinancialChartComponent,
                        IgxFinancialChartDefaultTemplatesComponent],
                    exports: [IgxFinancialChartComponent,
                        IgxFinancialChartDefaultTemplatesComponent,
                        IgxTemplateContentModule
                    ],
                    imports: [CommonModule, IgxTemplateContentModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
