// igniteui-angular-charts/esm2015/lib/igx-stacked-100-area-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStacked100AreaSeriesModule } from './igx-stacked-100-area-series-module';
import { IgxStacked100AreaSeriesComponent } from './igx-stacked-100-area-series-component';
import { Stacked100AreaSeries } from './Stacked100AreaSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxStacked100AreaSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxStacked100AreaSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxStacked100AreaSeriesComponent", IgxStacked100AreaSeriesComponent);
            TypeRegistrar.register("Stacked100AreaSeries", Stacked100AreaSeries.$type);
        }
    }
    IgxStacked100AreaSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100AreaSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStacked100AreaSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100AreaSeriesDynamicModule, imports: [CommonModule, IgxStacked100AreaSeriesModule], exports: [IgxStacked100AreaSeriesModule] });
    IgxStacked100AreaSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100AreaSeriesDynamicModule, imports: [[CommonModule, IgxStacked100AreaSeriesModule], IgxStacked100AreaSeriesModule] });
    return IgxStacked100AreaSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100AreaSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStacked100AreaSeriesModule],
                    exports: [IgxStacked100AreaSeriesModule],
                    entryComponents: [IgxStacked100AreaSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
