// igniteui-angular-charts/esm2015/lib/igx-rate-of-change-and-momentum-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxRateOfChangeAndMomentumIndicatorModule } from './igx-rate-of-change-and-momentum-indicator-module';
import { IgxRateOfChangeAndMomentumIndicatorComponent } from './igx-rate-of-change-and-momentum-indicator-component';
import { RateOfChangeAndMomentumIndicator } from './RateOfChangeAndMomentumIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxRateOfChangeAndMomentumIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxRateOfChangeAndMomentumIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxRateOfChangeAndMomentumIndicatorComponent", IgxRateOfChangeAndMomentumIndicatorComponent);
            TypeRegistrar.register("RateOfChangeAndMomentumIndicator", RateOfChangeAndMomentumIndicator.$type);
        }
    }
    IgxRateOfChangeAndMomentumIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRateOfChangeAndMomentumIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxRateOfChangeAndMomentumIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRateOfChangeAndMomentumIndicatorDynamicModule, imports: [CommonModule, IgxRateOfChangeAndMomentumIndicatorModule], exports: [IgxRateOfChangeAndMomentumIndicatorModule] });
    IgxRateOfChangeAndMomentumIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRateOfChangeAndMomentumIndicatorDynamicModule, imports: [[CommonModule, IgxRateOfChangeAndMomentumIndicatorModule], IgxRateOfChangeAndMomentumIndicatorModule] });
    return IgxRateOfChangeAndMomentumIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRateOfChangeAndMomentumIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxRateOfChangeAndMomentumIndicatorModule],
                    exports: [IgxRateOfChangeAndMomentumIndicatorModule],
                    entryComponents: [IgxRateOfChangeAndMomentumIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
