// igniteui-angular-charts/esm2015/lib/igx-indicators-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAbsoluteVolumeOscillatorIndicatorDynamicModule } from './igx-absolute-volume-oscillator-indicator-dynamic-module';
import { IgxAccumulationDistributionIndicatorDynamicModule } from './igx-accumulation-distribution-indicator-dynamic-module';
import { IgxAverageDirectionalIndexIndicatorDynamicModule } from './igx-average-directional-index-indicator-dynamic-module';
import { IgxAverageTrueRangeIndicatorDynamicModule } from './igx-average-true-range-indicator-dynamic-module';
import { IgxBollingerBandWidthIndicatorDynamicModule } from './igx-bollinger-band-width-indicator-dynamic-module';
import { IgxChaikinOscillatorIndicatorDynamicModule } from './igx-chaikin-oscillator-indicator-dynamic-module';
import { IgxChaikinVolatilityIndicatorDynamicModule } from './igx-chaikin-volatility-indicator-dynamic-module';
import { IgxCommodityChannelIndexIndicatorDynamicModule } from './igx-commodity-channel-index-indicator-dynamic-module';
import { IgxDetrendedPriceOscillatorIndicatorDynamicModule } from './igx-detrended-price-oscillator-indicator-dynamic-module';
import { IgxEaseOfMovementIndicatorDynamicModule } from './igx-ease-of-movement-indicator-dynamic-module';
import { IgxFastStochasticOscillatorIndicatorDynamicModule } from './igx-fast-stochastic-oscillator-indicator-dynamic-module';
import { IgxForceIndexIndicatorDynamicModule } from './igx-force-index-indicator-dynamic-module';
import { IgxFullStochasticOscillatorIndicatorDynamicModule } from './igx-full-stochastic-oscillator-indicator-dynamic-module';
import { IgxMarketFacilitationIndexIndicatorDynamicModule } from './igx-market-facilitation-index-indicator-dynamic-module';
import { IgxMassIndexIndicatorDynamicModule } from './igx-mass-index-indicator-dynamic-module';
import { IgxMedianPriceIndicatorDynamicModule } from './igx-median-price-indicator-dynamic-module';
import { IgxMoneyFlowIndexIndicatorDynamicModule } from './igx-money-flow-index-indicator-dynamic-module';
import { IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule } from './igx-moving-average-convergence-divergence-indicator-dynamic-module';
import { IgxNegativeVolumeIndexIndicatorDynamicModule } from './igx-negative-volume-index-indicator-dynamic-module';
import { IgxOnBalanceVolumeIndicatorDynamicModule } from './igx-on-balance-volume-indicator-dynamic-module';
import { IgxPercentagePriceOscillatorIndicatorDynamicModule } from './igx-percentage-price-oscillator-indicator-dynamic-module';
import { IgxPercentageVolumeOscillatorIndicatorDynamicModule } from './igx-percentage-volume-oscillator-indicator-dynamic-module';
import { IgxPositiveVolumeIndexIndicatorDynamicModule } from './igx-positive-volume-index-indicator-dynamic-module';
import { IgxPriceVolumeTrendIndicatorDynamicModule } from './igx-price-volume-trend-indicator-dynamic-module';
import { IgxRateOfChangeAndMomentumIndicatorDynamicModule } from './igx-rate-of-change-and-momentum-indicator-dynamic-module';
import { IgxRelativeStrengthIndexIndicatorDynamicModule } from './igx-relative-strength-index-indicator-dynamic-module';
import { IgxSlowStochasticOscillatorIndicatorDynamicModule } from './igx-slow-stochastic-oscillator-indicator-dynamic-module';
import { IgxStandardDeviationIndicatorDynamicModule } from './igx-standard-deviation-indicator-dynamic-module';
import { IgxStochRSIIndicatorDynamicModule } from './igx-stoch-rsi-indicator-dynamic-module';
import { IgxTRIXIndicatorDynamicModule } from './igx-trix-indicator-dynamic-module';
import { IgxTypicalPriceIndicatorDynamicModule } from './igx-typical-price-indicator-dynamic-module';
import { IgxUltimateOscillatorIndicatorDynamicModule } from './igx-ultimate-oscillator-indicator-dynamic-module';
import { IgxWeightedCloseIndicatorDynamicModule } from './igx-weighted-close-indicator-dynamic-module';
import { IgxWilliamsPercentRIndicatorDynamicModule } from './igx-williams-percent-r-indicator-dynamic-module';
import { IgxCustomIndicatorDynamicModule } from './igx-custom-indicator-dynamic-module';
import { IgxIndicatorsModule } from './igx-indicators-module';
import * as i0 from "@angular/core";
export let IgxIndicatorsDynamicModule = /*@__PURE__*/ (() => {
    class IgxIndicatorsDynamicModule {
    }
    IgxIndicatorsDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorsDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxIndicatorsDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorsDynamicModule, imports: [CommonModule, IgxAbsoluteVolumeOscillatorIndicatorDynamicModule,
            IgxAccumulationDistributionIndicatorDynamicModule,
            IgxAverageDirectionalIndexIndicatorDynamicModule,
            IgxAverageTrueRangeIndicatorDynamicModule,
            IgxBollingerBandWidthIndicatorDynamicModule,
            IgxChaikinOscillatorIndicatorDynamicModule,
            IgxChaikinVolatilityIndicatorDynamicModule,
            IgxCommodityChannelIndexIndicatorDynamicModule,
            IgxDetrendedPriceOscillatorIndicatorDynamicModule,
            IgxEaseOfMovementIndicatorDynamicModule,
            IgxFastStochasticOscillatorIndicatorDynamicModule,
            IgxForceIndexIndicatorDynamicModule,
            IgxFullStochasticOscillatorIndicatorDynamicModule,
            IgxMarketFacilitationIndexIndicatorDynamicModule,
            IgxMassIndexIndicatorDynamicModule,
            IgxMedianPriceIndicatorDynamicModule,
            IgxMoneyFlowIndexIndicatorDynamicModule,
            IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule,
            IgxNegativeVolumeIndexIndicatorDynamicModule,
            IgxOnBalanceVolumeIndicatorDynamicModule,
            IgxPercentagePriceOscillatorIndicatorDynamicModule,
            IgxPercentageVolumeOscillatorIndicatorDynamicModule,
            IgxPositiveVolumeIndexIndicatorDynamicModule,
            IgxPriceVolumeTrendIndicatorDynamicModule,
            IgxRateOfChangeAndMomentumIndicatorDynamicModule,
            IgxRelativeStrengthIndexIndicatorDynamicModule,
            IgxSlowStochasticOscillatorIndicatorDynamicModule,
            IgxStandardDeviationIndicatorDynamicModule,
            IgxStochRSIIndicatorDynamicModule,
            IgxTRIXIndicatorDynamicModule,
            IgxTypicalPriceIndicatorDynamicModule,
            IgxUltimateOscillatorIndicatorDynamicModule,
            IgxWeightedCloseIndicatorDynamicModule,
            IgxWilliamsPercentRIndicatorDynamicModule,
            IgxCustomIndicatorDynamicModule,
            IgxIndicatorsModule], exports: [IgxAbsoluteVolumeOscillatorIndicatorDynamicModule,
            IgxAccumulationDistributionIndicatorDynamicModule,
            IgxAverageDirectionalIndexIndicatorDynamicModule,
            IgxAverageTrueRangeIndicatorDynamicModule,
            IgxBollingerBandWidthIndicatorDynamicModule,
            IgxChaikinOscillatorIndicatorDynamicModule,
            IgxChaikinVolatilityIndicatorDynamicModule,
            IgxCommodityChannelIndexIndicatorDynamicModule,
            IgxDetrendedPriceOscillatorIndicatorDynamicModule,
            IgxEaseOfMovementIndicatorDynamicModule,
            IgxFastStochasticOscillatorIndicatorDynamicModule,
            IgxForceIndexIndicatorDynamicModule,
            IgxFullStochasticOscillatorIndicatorDynamicModule,
            IgxMarketFacilitationIndexIndicatorDynamicModule,
            IgxMassIndexIndicatorDynamicModule,
            IgxMedianPriceIndicatorDynamicModule,
            IgxMoneyFlowIndexIndicatorDynamicModule,
            IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule,
            IgxNegativeVolumeIndexIndicatorDynamicModule,
            IgxOnBalanceVolumeIndicatorDynamicModule,
            IgxPercentagePriceOscillatorIndicatorDynamicModule,
            IgxPercentageVolumeOscillatorIndicatorDynamicModule,
            IgxPositiveVolumeIndexIndicatorDynamicModule,
            IgxPriceVolumeTrendIndicatorDynamicModule,
            IgxRateOfChangeAndMomentumIndicatorDynamicModule,
            IgxRelativeStrengthIndexIndicatorDynamicModule,
            IgxSlowStochasticOscillatorIndicatorDynamicModule,
            IgxStandardDeviationIndicatorDynamicModule,
            IgxStochRSIIndicatorDynamicModule,
            IgxTRIXIndicatorDynamicModule,
            IgxTypicalPriceIndicatorDynamicModule,
            IgxUltimateOscillatorIndicatorDynamicModule,
            IgxWeightedCloseIndicatorDynamicModule,
            IgxWilliamsPercentRIndicatorDynamicModule,
            IgxCustomIndicatorDynamicModule,
            IgxIndicatorsModule] });
    IgxIndicatorsDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorsDynamicModule, imports: [[CommonModule, IgxAbsoluteVolumeOscillatorIndicatorDynamicModule,
                IgxAccumulationDistributionIndicatorDynamicModule,
                IgxAverageDirectionalIndexIndicatorDynamicModule,
                IgxAverageTrueRangeIndicatorDynamicModule,
                IgxBollingerBandWidthIndicatorDynamicModule,
                IgxChaikinOscillatorIndicatorDynamicModule,
                IgxChaikinVolatilityIndicatorDynamicModule,
                IgxCommodityChannelIndexIndicatorDynamicModule,
                IgxDetrendedPriceOscillatorIndicatorDynamicModule,
                IgxEaseOfMovementIndicatorDynamicModule,
                IgxFastStochasticOscillatorIndicatorDynamicModule,
                IgxForceIndexIndicatorDynamicModule,
                IgxFullStochasticOscillatorIndicatorDynamicModule,
                IgxMarketFacilitationIndexIndicatorDynamicModule,
                IgxMassIndexIndicatorDynamicModule,
                IgxMedianPriceIndicatorDynamicModule,
                IgxMoneyFlowIndexIndicatorDynamicModule,
                IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule,
                IgxNegativeVolumeIndexIndicatorDynamicModule,
                IgxOnBalanceVolumeIndicatorDynamicModule,
                IgxPercentagePriceOscillatorIndicatorDynamicModule,
                IgxPercentageVolumeOscillatorIndicatorDynamicModule,
                IgxPositiveVolumeIndexIndicatorDynamicModule,
                IgxPriceVolumeTrendIndicatorDynamicModule,
                IgxRateOfChangeAndMomentumIndicatorDynamicModule,
                IgxRelativeStrengthIndexIndicatorDynamicModule,
                IgxSlowStochasticOscillatorIndicatorDynamicModule,
                IgxStandardDeviationIndicatorDynamicModule,
                IgxStochRSIIndicatorDynamicModule,
                IgxTRIXIndicatorDynamicModule,
                IgxTypicalPriceIndicatorDynamicModule,
                IgxUltimateOscillatorIndicatorDynamicModule,
                IgxWeightedCloseIndicatorDynamicModule,
                IgxWilliamsPercentRIndicatorDynamicModule,
                IgxCustomIndicatorDynamicModule,
                IgxIndicatorsModule
            ], IgxAbsoluteVolumeOscillatorIndicatorDynamicModule,
            IgxAccumulationDistributionIndicatorDynamicModule,
            IgxAverageDirectionalIndexIndicatorDynamicModule,
            IgxAverageTrueRangeIndicatorDynamicModule,
            IgxBollingerBandWidthIndicatorDynamicModule,
            IgxChaikinOscillatorIndicatorDynamicModule,
            IgxChaikinVolatilityIndicatorDynamicModule,
            IgxCommodityChannelIndexIndicatorDynamicModule,
            IgxDetrendedPriceOscillatorIndicatorDynamicModule,
            IgxEaseOfMovementIndicatorDynamicModule,
            IgxFastStochasticOscillatorIndicatorDynamicModule,
            IgxForceIndexIndicatorDynamicModule,
            IgxFullStochasticOscillatorIndicatorDynamicModule,
            IgxMarketFacilitationIndexIndicatorDynamicModule,
            IgxMassIndexIndicatorDynamicModule,
            IgxMedianPriceIndicatorDynamicModule,
            IgxMoneyFlowIndexIndicatorDynamicModule,
            IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule,
            IgxNegativeVolumeIndexIndicatorDynamicModule,
            IgxOnBalanceVolumeIndicatorDynamicModule,
            IgxPercentagePriceOscillatorIndicatorDynamicModule,
            IgxPercentageVolumeOscillatorIndicatorDynamicModule,
            IgxPositiveVolumeIndexIndicatorDynamicModule,
            IgxPriceVolumeTrendIndicatorDynamicModule,
            IgxRateOfChangeAndMomentumIndicatorDynamicModule,
            IgxRelativeStrengthIndexIndicatorDynamicModule,
            IgxSlowStochasticOscillatorIndicatorDynamicModule,
            IgxStandardDeviationIndicatorDynamicModule,
            IgxStochRSIIndicatorDynamicModule,
            IgxTRIXIndicatorDynamicModule,
            IgxTypicalPriceIndicatorDynamicModule,
            IgxUltimateOscillatorIndicatorDynamicModule,
            IgxWeightedCloseIndicatorDynamicModule,
            IgxWilliamsPercentRIndicatorDynamicModule,
            IgxCustomIndicatorDynamicModule,
            IgxIndicatorsModule] });
    return IgxIndicatorsDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorsDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxAbsoluteVolumeOscillatorIndicatorDynamicModule,
                        IgxAccumulationDistributionIndicatorDynamicModule,
                        IgxAverageDirectionalIndexIndicatorDynamicModule,
                        IgxAverageTrueRangeIndicatorDynamicModule,
                        IgxBollingerBandWidthIndicatorDynamicModule,
                        IgxChaikinOscillatorIndicatorDynamicModule,
                        IgxChaikinVolatilityIndicatorDynamicModule,
                        IgxCommodityChannelIndexIndicatorDynamicModule,
                        IgxDetrendedPriceOscillatorIndicatorDynamicModule,
                        IgxEaseOfMovementIndicatorDynamicModule,
                        IgxFastStochasticOscillatorIndicatorDynamicModule,
                        IgxForceIndexIndicatorDynamicModule,
                        IgxFullStochasticOscillatorIndicatorDynamicModule,
                        IgxMarketFacilitationIndexIndicatorDynamicModule,
                        IgxMassIndexIndicatorDynamicModule,
                        IgxMedianPriceIndicatorDynamicModule,
                        IgxMoneyFlowIndexIndicatorDynamicModule,
                        IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule,
                        IgxNegativeVolumeIndexIndicatorDynamicModule,
                        IgxOnBalanceVolumeIndicatorDynamicModule,
                        IgxPercentagePriceOscillatorIndicatorDynamicModule,
                        IgxPercentageVolumeOscillatorIndicatorDynamicModule,
                        IgxPositiveVolumeIndexIndicatorDynamicModule,
                        IgxPriceVolumeTrendIndicatorDynamicModule,
                        IgxRateOfChangeAndMomentumIndicatorDynamicModule,
                        IgxRelativeStrengthIndexIndicatorDynamicModule,
                        IgxSlowStochasticOscillatorIndicatorDynamicModule,
                        IgxStandardDeviationIndicatorDynamicModule,
                        IgxStochRSIIndicatorDynamicModule,
                        IgxTRIXIndicatorDynamicModule,
                        IgxTypicalPriceIndicatorDynamicModule,
                        IgxUltimateOscillatorIndicatorDynamicModule,
                        IgxWeightedCloseIndicatorDynamicModule,
                        IgxWilliamsPercentRIndicatorDynamicModule,
                        IgxCustomIndicatorDynamicModule,
                        IgxIndicatorsModule
                    ],
                    imports: [CommonModule, IgxAbsoluteVolumeOscillatorIndicatorDynamicModule,
                        IgxAccumulationDistributionIndicatorDynamicModule,
                        IgxAverageDirectionalIndexIndicatorDynamicModule,
                        IgxAverageTrueRangeIndicatorDynamicModule,
                        IgxBollingerBandWidthIndicatorDynamicModule,
                        IgxChaikinOscillatorIndicatorDynamicModule,
                        IgxChaikinVolatilityIndicatorDynamicModule,
                        IgxCommodityChannelIndexIndicatorDynamicModule,
                        IgxDetrendedPriceOscillatorIndicatorDynamicModule,
                        IgxEaseOfMovementIndicatorDynamicModule,
                        IgxFastStochasticOscillatorIndicatorDynamicModule,
                        IgxForceIndexIndicatorDynamicModule,
                        IgxFullStochasticOscillatorIndicatorDynamicModule,
                        IgxMarketFacilitationIndexIndicatorDynamicModule,
                        IgxMassIndexIndicatorDynamicModule,
                        IgxMedianPriceIndicatorDynamicModule,
                        IgxMoneyFlowIndexIndicatorDynamicModule,
                        IgxMovingAverageConvergenceDivergenceIndicatorDynamicModule,
                        IgxNegativeVolumeIndexIndicatorDynamicModule,
                        IgxOnBalanceVolumeIndicatorDynamicModule,
                        IgxPercentagePriceOscillatorIndicatorDynamicModule,
                        IgxPercentageVolumeOscillatorIndicatorDynamicModule,
                        IgxPositiveVolumeIndexIndicatorDynamicModule,
                        IgxPriceVolumeTrendIndicatorDynamicModule,
                        IgxRateOfChangeAndMomentumIndicatorDynamicModule,
                        IgxRelativeStrengthIndexIndicatorDynamicModule,
                        IgxSlowStochasticOscillatorIndicatorDynamicModule,
                        IgxStandardDeviationIndicatorDynamicModule,
                        IgxStochRSIIndicatorDynamicModule,
                        IgxTRIXIndicatorDynamicModule,
                        IgxTypicalPriceIndicatorDynamicModule,
                        IgxUltimateOscillatorIndicatorDynamicModule,
                        IgxWeightedCloseIndicatorDynamicModule,
                        IgxWilliamsPercentRIndicatorDynamicModule,
                        IgxCustomIndicatorDynamicModule,
                        IgxIndicatorsModule
                    ],
                    entryComponents: []
                }]
        }] });
