// igniteui-angular-charts/esm2015/lib/igx-treemap-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTreemapComponent } from './igx-treemap-component';
import { IgxTreemapNodeStyleDynamicModule } from './igx-treemap-node-style-dynamic-module';
import { IgxTreemapNodeStyleMappingModule } from './igx-treemap-node-style-mapping-module';
import { Treemap } from './Treemap';
import { FastItemObjectColumn } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxTreemapModule = /*@__PURE__*/ (() => {
    class IgxTreemapModule {
        constructor() {
            TypeRegistrar.registerCons('IgxTreemapComponent', IgxTreemapComponent);
            TypeRegistrar.register('Treemap', Treemap.$type);
            TypeRegistrar.register('FastItemObjectColumn', FastItemObjectColumn.$type);
        }
    }
    IgxTreemapModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTreemapModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapModule, declarations: [IgxTreemapComponent], imports: [CommonModule, IgxTreemapNodeStyleDynamicModule,
            IgxTreemapNodeStyleMappingModule], exports: [IgxTreemapComponent,
            IgxTreemapNodeStyleDynamicModule,
            IgxTreemapNodeStyleMappingModule] });
    IgxTreemapModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapModule, imports: [[CommonModule, IgxTreemapNodeStyleDynamicModule,
                IgxTreemapNodeStyleMappingModule
            ], IgxTreemapNodeStyleDynamicModule,
            IgxTreemapNodeStyleMappingModule] });
    return IgxTreemapModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxTreemapComponent],
                    exports: [IgxTreemapComponent,
                        IgxTreemapNodeStyleDynamicModule,
                        IgxTreemapNodeStyleMappingModule
                    ],
                    imports: [CommonModule, IgxTreemapNodeStyleDynamicModule,
                        IgxTreemapNodeStyleMappingModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
