// igniteui-angular-charts/esm2015/lib/igx-data-chart-scatter-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericXAxisDynamicModule } from './igx-numeric-x-axis-dynamic-module';
import { IgxNumericYAxisDynamicModule } from './igx-numeric-y-axis-dynamic-module';
import { IgxDataChartCoreDynamicModule } from './igx-data-chart-core-dynamic-module';
import { IgxDataChartScatterCoreModule } from './igx-data-chart-scatter-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartScatterCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartScatterCoreDynamicModule {
    }
    IgxDataChartScatterCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartScatterCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterCoreDynamicModule, imports: [CommonModule, IgxNumericXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartScatterCoreModule], exports: [IgxNumericXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartScatterCoreModule] });
    IgxDataChartScatterCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterCoreDynamicModule, imports: [[CommonModule, IgxNumericXAxisDynamicModule,
                IgxNumericYAxisDynamicModule,
                IgxDataChartCoreDynamicModule,
                IgxDataChartScatterCoreModule
            ], IgxNumericXAxisDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartScatterCoreModule] });
    return IgxDataChartScatterCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumericXAxisDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartScatterCoreModule
                    ],
                    imports: [CommonModule, IgxNumericXAxisDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartScatterCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
