// igniteui-angular-charts/esm2015/lib/igx-radial-pie-series-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxAnchoredRadialSeriesComponent } from "./igx-anchored-radial-series-component";
import { IgxRadialBaseComponent } from "./igx-radial-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { RadialPieSeries } from "./RadialPieSeries";
import { toPoint, fromRect } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent radial pie series.
 * Compare values across categories by using radial pie slices.
*/
export let IgxRadialPieSeriesComponent = /*@__PURE__*/ (() => {
    class IgxRadialPieSeriesComponent extends IgxAnchoredRadialSeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new RadialPieSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows a pie shape.
        */
        get isPie() {
            return this.i.fl;
        }
        /**
         * Overridden by derived series classes to indicate when marker-less display is preferred or not.
        */
        get isMarkerlessDisplayPreferred() {
            return this.i.fj;
        }
        /**
         * Gets or sets the x-radius of the ellipse that is used to round the corners of the column.
        */
        get radiusX() {
            return this.i.adh;
        }
        set radiusX(v) {
            this.i.adh = +v;
        }
        /**
         * Gets or sets the y-radius of the ellipse that is used to round the corners of the column.
        */
        get radiusY() {
            return this.i.adi;
        }
        set radiusY(v) {
            this.i.adi = +v;
        }
        /**
         * Gets or sets the property name that contains the legend labels.
        */
        get legendLabelMemberPath() {
            return this.i.adl;
        }
        set legendLabelMemberPath(v) {
            this.i.adl = v;
        }
        /**
         * For a category plotted series, returns the current width of the items within the categories. This only returns a value if the items have some form of width (e.g. columns, bars, etc.) otherwise 0 is returned.
        
        */
        getItemSpan() {
            let iv = this.i.iu();
            return (iv);
        }
        getSeriesValueBoundingBox(world) {
            let iv = this.i.wh(toPoint(world));
            return fromRect(iv);
        }
    }
    IgxRadialPieSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialPieSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxRadialPieSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxRadialPieSeriesComponent, selector: "igx-radial-pie-series", inputs: { radiusX: "radiusX", radiusY: "radiusY", legendLabelMemberPath: "legendLabelMemberPath" }, providers: [{ provide: IgxAnchoredRadialSeriesComponent, useExisting: forwardRef(() => IgxRadialPieSeriesComponent) }, { provide: IgxRadialBaseComponent, useExisting: forwardRef(() => IgxRadialPieSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxRadialPieSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxRadialPieSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxRadialPieSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialPieSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-radial-pie-series',
                    template: ``,
                    providers: [{ provide: IgxAnchoredRadialSeriesComponent, useExisting: forwardRef(() => IgxRadialPieSeriesComponent) }, { provide: IgxRadialBaseComponent, useExisting: forwardRef(() => IgxRadialPieSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxRadialPieSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxRadialPieSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { radiusX: [{
                type: Input
            }], radiusY: [{
                type: Input
            }], legendLabelMemberPath: [{
                type: Input
            }] } });
