// igniteui-angular-charts/esm2015/lib/igx-median-price-indicator-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxItemwiseStrategyBasedIndicatorComponent } from "./igx-itemwise-strategy-based-indicator-component";
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import { IgxFinancialIndicatorComponent } from "./igx-financial-indicator-component";
import { IgxFinancialSeriesComponent } from "./igx-financial-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { MedianPriceIndicator } from "./MedianPriceIndicator";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent Median Price indicator series.
 * Default required members: High, Low
*
* Represents a Ignite UIDataChart Median Price indicator series.
*
* Example:
*
* ```html
* <igx-data-chart
*     [dataSource]="data">
* 	<igx-category-x-axis
* 		label="label"
* 		#xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
* 		#yAxis>
*     </igx-numeric-y-axis>
* 	<igx-median-price-indicator
*         #series
*         [xAxis]="xAxis"
*         [yAxis]="yAxis"
*         highMemberPath="high"
* 		lowMemberPath="low"
* 		volumeMemberPath="volume">
*     </igx-median-price-indicator>
* </igx-data-chart>
* ```
*
* ```ts
* let series = new IgxMedianPriceIndicatorComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* series.volumeMemberPath= "volume";
* this.chart.series.add(series);
* ```
*/
export let IgxMedianPriceIndicatorComponent = /*@__PURE__*/ (() => {
    class IgxMedianPriceIndicatorComponent extends IgxItemwiseStrategyBasedIndicatorComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new MedianPriceIndicator();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
    }
    IgxMedianPriceIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMedianPriceIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxMedianPriceIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxMedianPriceIndicatorComponent, selector: "igx-median-price-indicator", providers: [{ provide: IgxItemwiseStrategyBasedIndicatorComponent, useExisting: forwardRef(() => IgxMedianPriceIndicatorComponent) }, { provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(() => IgxMedianPriceIndicatorComponent) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(() => IgxMedianPriceIndicatorComponent) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(() => IgxMedianPriceIndicatorComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxMedianPriceIndicatorComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxMedianPriceIndicatorComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMedianPriceIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-median-price-indicator',
                    template: ``,
                    providers: [{ provide: IgxItemwiseStrategyBasedIndicatorComponent, useExisting: forwardRef(() => IgxMedianPriceIndicatorComponent) }, { provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(() => IgxMedianPriceIndicatorComponent) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(() => IgxMedianPriceIndicatorComponent) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(() => IgxMedianPriceIndicatorComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxMedianPriceIndicatorComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
