// igniteui-angular-charts/esm2015/lib/igx-column-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxHorizontalAnchoredCategorySeriesProxyModule } from './igx-horizontal-anchored-category-series-proxy-module';
import { IgxColumnSeriesComponent } from './igx-column-series-component';
import * as i0 from "@angular/core";
export let IgxColumnSeriesModule = /*@__PURE__*/ (() => {
    class IgxColumnSeriesModule {
    }
    IgxColumnSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxColumnSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnSeriesModule, declarations: [IgxColumnSeriesComponent], imports: [CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule], exports: [IgxColumnSeriesComponent, IgxHorizontalAnchoredCategorySeriesProxyModule] });
    IgxColumnSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnSeriesModule, imports: [[CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule], IgxHorizontalAnchoredCategorySeriesProxyModule] });
    return IgxColumnSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxColumnSeriesComponent],
                    imports: [CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule],
                    exports: [IgxColumnSeriesComponent, IgxHorizontalAnchoredCategorySeriesProxyModule],
                    entryComponents: []
                }]
        }] });
