// igniteui-angular-charts/esm2015/lib/igx-time-x-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTimeXAxisModule } from './igx-time-x-axis-module';
import { IgxTimeXAxisComponent } from './igx-time-x-axis-component';
import { TimeXAxis } from './TimeXAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxTimeXAxisDynamicModule = /*@__PURE__*/ (() => {
    class IgxTimeXAxisDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxTimeXAxisComponent", IgxTimeXAxisComponent);
            TypeRegistrar.register("TimeXAxis", TimeXAxis.$type);
        }
    }
    IgxTimeXAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeXAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTimeXAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeXAxisDynamicModule, imports: [CommonModule, IgxTimeXAxisModule], exports: [IgxTimeXAxisModule] });
    IgxTimeXAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeXAxisDynamicModule, imports: [[CommonModule, IgxTimeXAxisModule], IgxTimeXAxisModule] });
    return IgxTimeXAxisDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeXAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxTimeXAxisModule],
                    exports: [IgxTimeXAxisModule],
                    entryComponents: [IgxTimeXAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
