// igniteui-angular-charts/esm2015/lib/igx-doughnut-chart-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDoughnutChartInteractivityDynamicModule } from './igx-doughnut-chart-interactivity-dynamic-module';
import { IgxDoughnutChartCoreDynamicModule } from './igx-doughnut-chart-core-dynamic-module';
import { IgxDoughnutChartModule } from './igx-doughnut-chart-module';
import * as i0 from "@angular/core";
export let IgxDoughnutChartDynamicModule = /*@__PURE__*/ (() => {
    class IgxDoughnutChartDynamicModule {
    }
    IgxDoughnutChartDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDoughnutChartDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartDynamicModule, imports: [CommonModule, IgxDoughnutChartInteractivityDynamicModule,
            IgxDoughnutChartCoreDynamicModule,
            IgxDoughnutChartModule], exports: [IgxDoughnutChartInteractivityDynamicModule,
            IgxDoughnutChartCoreDynamicModule,
            IgxDoughnutChartModule] });
    IgxDoughnutChartDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartDynamicModule, imports: [[CommonModule, IgxDoughnutChartInteractivityDynamicModule,
                IgxDoughnutChartCoreDynamicModule,
                IgxDoughnutChartModule
            ], IgxDoughnutChartInteractivityDynamicModule,
            IgxDoughnutChartCoreDynamicModule,
            IgxDoughnutChartModule] });
    return IgxDoughnutChartDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDoughnutChartInteractivityDynamicModule,
                        IgxDoughnutChartCoreDynamicModule,
                        IgxDoughnutChartModule
                    ],
                    imports: [CommonModule, IgxDoughnutChartInteractivityDynamicModule,
                        IgxDoughnutChartCoreDynamicModule,
                        IgxDoughnutChartModule
                    ],
                    entryComponents: []
                }]
        }] });
