// igniteui-angular-charts/esm2015/lib/igx-stacked-line-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedLineSeriesModule } from './igx-stacked-line-series-module';
import { IgxStackedLineSeriesComponent } from './igx-stacked-line-series-component';
import { StackedLineSeries } from './StackedLineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxStackedLineSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxStackedLineSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxStackedLineSeriesComponent", IgxStackedLineSeriesComponent);
            TypeRegistrar.register("StackedLineSeries", StackedLineSeries.$type);
        }
    }
    IgxStackedLineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedLineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedLineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedLineSeriesDynamicModule, imports: [CommonModule, IgxStackedLineSeriesModule], exports: [IgxStackedLineSeriesModule] });
    IgxStackedLineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedLineSeriesDynamicModule, imports: [[CommonModule, IgxStackedLineSeriesModule], IgxStackedLineSeriesModule] });
    return IgxStackedLineSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedLineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStackedLineSeriesModule],
                    exports: [IgxStackedLineSeriesModule],
                    entryComponents: [IgxStackedLineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
