// igniteui-angular-charts/esm2015/lib/igx-scale-legend-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScaleLegendComponent } from './igx-scale-legend-component';
import { LegendTemplates } from './LegendTemplates';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxScaleLegendModule = /*@__PURE__*/ (() => {
    class IgxScaleLegendModule {
        constructor() {
            TypeRegistrar.register('LegendTemplates', LegendTemplates.$type);
        }
    }
    IgxScaleLegendModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScaleLegendModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendModule, declarations: [IgxScaleLegendComponent], imports: [CommonModule], exports: [IgxScaleLegendComponent] });
    IgxScaleLegendModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendModule, imports: [[CommonModule,
            ]] });
    return IgxScaleLegendModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxScaleLegendComponent],
                    exports: [IgxScaleLegendComponent
                    ],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
