// igniteui-angular-charts/esm2015/lib/igx-ordinal-time-x-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxOrdinalTimeXAxisModule } from './igx-ordinal-time-x-axis-module';
import { IgxOrdinalTimeXAxisComponent } from './igx-ordinal-time-x-axis-component';
import { OrdinalTimeXAxis } from './OrdinalTimeXAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxOrdinalTimeXAxisDynamicModule = /*@__PURE__*/ (() => {
    class IgxOrdinalTimeXAxisDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxOrdinalTimeXAxisComponent", IgxOrdinalTimeXAxisComponent);
            TypeRegistrar.register("OrdinalTimeXAxis", OrdinalTimeXAxis.$type);
        }
    }
    IgxOrdinalTimeXAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOrdinalTimeXAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxOrdinalTimeXAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOrdinalTimeXAxisDynamicModule, imports: [CommonModule, IgxOrdinalTimeXAxisModule], exports: [IgxOrdinalTimeXAxisModule] });
    IgxOrdinalTimeXAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOrdinalTimeXAxisDynamicModule, imports: [[CommonModule, IgxOrdinalTimeXAxisModule], IgxOrdinalTimeXAxisModule] });
    return IgxOrdinalTimeXAxisDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOrdinalTimeXAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxOrdinalTimeXAxisModule],
                    exports: [IgxOrdinalTimeXAxisModule],
                    entryComponents: [IgxOrdinalTimeXAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
