// igniteui-angular-charts/esm2015/lib/igx-size-scale-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSizeScaleModule } from './igx-size-scale-module';
import { IgxSizeScaleComponent } from './igx-size-scale-component';
import { SizeScale } from './SizeScale';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxSizeScaleDynamicModule = /*@__PURE__*/ (() => {
    class IgxSizeScaleDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxSizeScaleComponent", IgxSizeScaleComponent);
            TypeRegistrar.register("SizeScale", SizeScale.$type);
        }
    }
    IgxSizeScaleDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSizeScaleDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleDynamicModule, imports: [CommonModule, IgxSizeScaleModule], exports: [IgxSizeScaleModule] });
    IgxSizeScaleDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleDynamicModule, imports: [[CommonModule, IgxSizeScaleModule], IgxSizeScaleModule] });
    return IgxSizeScaleDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxSizeScaleModule],
                    exports: [IgxSizeScaleModule],
                    entryComponents: [IgxSizeScaleComponent]
                }]
        }], ctorParameters: function () { return []; } });
