// igniteui-angular-charts/esm2015/lib/igx-radial-column-series-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxAnchoredRadialSeriesComponent } from "./igx-anchored-radial-series-component";
import { IgxRadialBaseComponent } from "./igx-radial-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { RadialColumnSeries } from "./RadialColumnSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent radial column series.
 * Compare values across categories by using radial rectangles.
*/
export let IgxRadialColumnSeriesComponent = /*@__PURE__*/ (() => {
    class IgxRadialColumnSeriesComponent extends IgxAnchoredRadialSeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new RadialColumnSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows a column shape.
        */
        get isColumn() {
            return this.i.ex;
        }
        /**
         * Overridden by derived series classes to indicate when marker-less display is preferred or not.
        */
        get isMarkerlessDisplayPreferred() {
            return this.i.fj;
        }
        /**
         * Gets or sets the x-radius of the ellipse that is used to round the corners of the column.
        */
        get radiusX() {
            return this.i.ade;
        }
        set radiusX(v) {
            this.i.ade = +v;
        }
        /**
         * Gets or sets the y-radius of the ellipse that is used to round the corners of the column.
        */
        get radiusY() {
            return this.i.adf;
        }
        set radiusY(v) {
            this.i.adf = +v;
        }
        /**
         * When overridden in a derived class, is invoked whenever application code or internal processes
         * call ApplyTemplate.
        
        */
        onApplyTemplate() {
            this.i.ac();
        }
        /**
         * For a category plotted series, returns the current width of the items within the categories. This only returns a value if the items have some form of width (e.g. columns, bars, etc.) otherwise 0 is returned.
        
        */
        getItemSpan() {
            let iv = this.i.iu();
            return (iv);
        }
    }
    IgxRadialColumnSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialColumnSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxRadialColumnSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxRadialColumnSeriesComponent, selector: "igx-radial-column-series", inputs: { radiusX: "radiusX", radiusY: "radiusY" }, providers: [{ provide: IgxAnchoredRadialSeriesComponent, useExisting: forwardRef(() => IgxRadialColumnSeriesComponent) }, { provide: IgxRadialBaseComponent, useExisting: forwardRef(() => IgxRadialColumnSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxRadialColumnSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxRadialColumnSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxRadialColumnSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialColumnSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-radial-column-series',
                    template: ``,
                    providers: [{ provide: IgxAnchoredRadialSeriesComponent, useExisting: forwardRef(() => IgxRadialColumnSeriesComponent) }, { provide: IgxRadialBaseComponent, useExisting: forwardRef(() => IgxRadialColumnSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxRadialColumnSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxRadialColumnSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { radiusX: [{
                type: Input
            }], radiusY: [{
                type: Input
            }] } });
