// igniteui-angular-charts/esm2015/lib/igx-percentage-price-oscillator-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPercentagePriceOscillatorIndicatorModule } from './igx-percentage-price-oscillator-indicator-module';
import { IgxPercentagePriceOscillatorIndicatorComponent } from './igx-percentage-price-oscillator-indicator-component';
import { PercentagePriceOscillatorIndicator } from './PercentagePriceOscillatorIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxPercentagePriceOscillatorIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxPercentagePriceOscillatorIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxPercentagePriceOscillatorIndicatorComponent", IgxPercentagePriceOscillatorIndicatorComponent);
            TypeRegistrar.register("PercentagePriceOscillatorIndicator", PercentagePriceOscillatorIndicator.$type);
        }
    }
    IgxPercentagePriceOscillatorIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentagePriceOscillatorIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPercentagePriceOscillatorIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentagePriceOscillatorIndicatorDynamicModule, imports: [CommonModule, IgxPercentagePriceOscillatorIndicatorModule], exports: [IgxPercentagePriceOscillatorIndicatorModule] });
    IgxPercentagePriceOscillatorIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentagePriceOscillatorIndicatorDynamicModule, imports: [[CommonModule, IgxPercentagePriceOscillatorIndicatorModule], IgxPercentagePriceOscillatorIndicatorModule] });
    return IgxPercentagePriceOscillatorIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPercentagePriceOscillatorIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxPercentagePriceOscillatorIndicatorModule],
                    exports: [IgxPercentagePriceOscillatorIndicatorModule],
                    entryComponents: [IgxPercentagePriceOscillatorIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
