// igniteui-angular-charts/esm2015/lib/igx-category-angle-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryAngleAxisComponent } from './igx-category-angle-axis-component';
import * as i0 from "@angular/core";
export let IgxCategoryAngleAxisModule = /*@__PURE__*/ (() => {
    class IgxCategoryAngleAxisModule {
    }
    IgxCategoryAngleAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryAngleAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryAngleAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryAngleAxisModule, declarations: [IgxCategoryAngleAxisComponent], imports: [CommonModule], exports: [IgxCategoryAngleAxisComponent] });
    IgxCategoryAngleAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryAngleAxisModule, imports: [[CommonModule]] });
    return IgxCategoryAngleAxisModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryAngleAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxCategoryAngleAxisComponent],
                    imports: [CommonModule],
                    exports: [IgxCategoryAngleAxisComponent],
                    entryComponents: []
                }]
        }] });
