// igniteui-angular-charts/esm2015/lib/igx-data-chart-polar-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericAngleAxisDynamicModule } from './igx-numeric-angle-axis-dynamic-module';
import { IgxNumericRadiusAxisDynamicModule } from './igx-numeric-radius-axis-dynamic-module';
import { IgxDataChartCoreDynamicModule } from './igx-data-chart-core-dynamic-module';
import { IgxDataChartPolarCoreModule } from './igx-data-chart-polar-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartPolarCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartPolarCoreDynamicModule {
    }
    IgxDataChartPolarCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartPolarCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarCoreDynamicModule, imports: [CommonModule, IgxNumericAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartPolarCoreModule], exports: [IgxNumericAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartPolarCoreModule] });
    IgxDataChartPolarCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarCoreDynamicModule, imports: [[CommonModule, IgxNumericAngleAxisDynamicModule,
                IgxNumericRadiusAxisDynamicModule,
                IgxDataChartCoreDynamicModule,
                IgxDataChartPolarCoreModule
            ], IgxNumericAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartPolarCoreModule] });
    return IgxDataChartPolarCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartPolarCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumericAngleAxisDynamicModule,
                        IgxNumericRadiusAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartPolarCoreModule
                    ],
                    imports: [CommonModule, IgxNumericAngleAxisDynamicModule,
                        IgxNumericRadiusAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartPolarCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
