// igniteui-angular-charts/esm2015/lib/igx-data-chart-vertical-category-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxBarSeriesModule } from './igx-bar-series-module';
import { IgxDataChartVerticalCategoryCoreModule } from './igx-data-chart-vertical-category-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartVerticalCategoryModule = /*@__PURE__*/ (() => {
    class IgxDataChartVerticalCategoryModule {
    }
    IgxDataChartVerticalCategoryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartVerticalCategoryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryModule, imports: [CommonModule, IgxBarSeriesModule,
            IgxDataChartVerticalCategoryCoreModule], exports: [IgxBarSeriesModule,
            IgxDataChartVerticalCategoryCoreModule] });
    IgxDataChartVerticalCategoryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryModule, imports: [[CommonModule, IgxBarSeriesModule,
                IgxDataChartVerticalCategoryCoreModule
            ], IgxBarSeriesModule,
            IgxDataChartVerticalCategoryCoreModule] });
    return IgxDataChartVerticalCategoryModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxBarSeriesModule,
                        IgxDataChartVerticalCategoryCoreModule
                    ],
                    imports: [CommonModule, IgxBarSeriesModule,
                        IgxDataChartVerticalCategoryCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
