// igniteui-angular-charts/esm2015/lib/igx-ultimate-oscillator-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxUltimateOscillatorIndicatorComponent } from './igx-ultimate-oscillator-indicator-component';
import * as i0 from "@angular/core";
export let IgxUltimateOscillatorIndicatorModule = /*@__PURE__*/ (() => {
    class IgxUltimateOscillatorIndicatorModule {
    }
    IgxUltimateOscillatorIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxUltimateOscillatorIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxUltimateOscillatorIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxUltimateOscillatorIndicatorModule, declarations: [IgxUltimateOscillatorIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxUltimateOscillatorIndicatorComponent, IgxIndicatorProxyModule] });
    IgxUltimateOscillatorIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxUltimateOscillatorIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxUltimateOscillatorIndicatorModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxUltimateOscillatorIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxUltimateOscillatorIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxUltimateOscillatorIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
