// igniteui-angular-charts/esm2015/lib/igx-stacked-area-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedAreaSeriesComponent } from './igx-stacked-area-series-component';
import * as i0 from "@angular/core";
export let IgxStackedAreaSeriesModule = /*@__PURE__*/ (() => {
    class IgxStackedAreaSeriesModule {
    }
    IgxStackedAreaSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedAreaSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedAreaSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedAreaSeriesModule, declarations: [IgxStackedAreaSeriesComponent], imports: [CommonModule], exports: [IgxStackedAreaSeriesComponent] });
    IgxStackedAreaSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedAreaSeriesModule, imports: [[CommonModule]] });
    return IgxStackedAreaSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedAreaSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStackedAreaSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxStackedAreaSeriesComponent],
                    entryComponents: []
                }]
        }] });
