// igniteui-angular-charts/esm2015/lib/igx-data-chart-extended-axes-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryDateTimeXAxisModule } from './igx-category-date-time-x-axis-module';
import { IgxTimeAxisBreakModule } from './igx-time-axis-break-module';
import { IgxTimeAxisIntervalModule } from './igx-time-axis-interval-module';
import { IgxTimeAxisLabelFormatModule } from './igx-time-axis-label-format-module';
import { IgxTimeXAxisModule } from './igx-time-x-axis-module';
import { IgxOrdinalTimeXAxisModule } from './igx-ordinal-time-x-axis-module';
import { IgxDataChartCoreModule } from './igx-data-chart-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartExtendedAxesModule = /*@__PURE__*/ (() => {
    class IgxDataChartExtendedAxesModule {
    }
    IgxDataChartExtendedAxesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartExtendedAxesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartExtendedAxesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartExtendedAxesModule, imports: [CommonModule, IgxCategoryDateTimeXAxisModule,
            IgxTimeAxisBreakModule,
            IgxTimeAxisIntervalModule,
            IgxTimeAxisLabelFormatModule,
            IgxTimeXAxisModule,
            IgxOrdinalTimeXAxisModule,
            IgxDataChartCoreModule], exports: [IgxCategoryDateTimeXAxisModule,
            IgxTimeAxisBreakModule,
            IgxTimeAxisIntervalModule,
            IgxTimeAxisLabelFormatModule,
            IgxTimeXAxisModule,
            IgxOrdinalTimeXAxisModule,
            IgxDataChartCoreModule] });
    IgxDataChartExtendedAxesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartExtendedAxesModule, imports: [[CommonModule, IgxCategoryDateTimeXAxisModule,
                IgxTimeAxisBreakModule,
                IgxTimeAxisIntervalModule,
                IgxTimeAxisLabelFormatModule,
                IgxTimeXAxisModule,
                IgxOrdinalTimeXAxisModule,
                IgxDataChartCoreModule
            ], IgxCategoryDateTimeXAxisModule,
            IgxTimeAxisBreakModule,
            IgxTimeAxisIntervalModule,
            IgxTimeAxisLabelFormatModule,
            IgxTimeXAxisModule,
            IgxOrdinalTimeXAxisModule,
            IgxDataChartCoreModule] });
    return IgxDataChartExtendedAxesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartExtendedAxesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxCategoryDateTimeXAxisModule,
                        IgxTimeAxisBreakModule,
                        IgxTimeAxisIntervalModule,
                        IgxTimeAxisLabelFormatModule,
                        IgxTimeXAxisModule,
                        IgxOrdinalTimeXAxisModule,
                        IgxDataChartCoreModule
                    ],
                    imports: [CommonModule, IgxCategoryDateTimeXAxisModule,
                        IgxTimeAxisBreakModule,
                        IgxTimeAxisIntervalModule,
                        IgxTimeAxisLabelFormatModule,
                        IgxTimeXAxisModule,
                        IgxOrdinalTimeXAxisModule,
                        IgxDataChartCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
