// igniteui-angular-charts/esm2015/lib/igx-polar-line-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPolarLineSeriesComponent } from './igx-polar-line-series-component';
import * as i0 from "@angular/core";
export let IgxPolarLineSeriesModule = /*@__PURE__*/ (() => {
    class IgxPolarLineSeriesModule {
    }
    IgxPolarLineSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPolarLineSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesModule, declarations: [IgxPolarLineSeriesComponent], imports: [CommonModule], exports: [IgxPolarLineSeriesComponent] });
    IgxPolarLineSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesModule, imports: [[CommonModule]] });
    return IgxPolarLineSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxPolarLineSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxPolarLineSeriesComponent],
                    entryComponents: []
                }]
        }] });
