// igniteui-angular-charts/esm2015/lib/igx-data-chart-stacked-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedFragmentSeriesDynamicModule } from './igx-stacked-fragment-series-dynamic-module';
import { IgxStacked100AreaSeriesDynamicModule } from './igx-stacked-100-area-series-dynamic-module';
import { IgxStacked100BarSeriesDynamicModule } from './igx-stacked-100-bar-series-dynamic-module';
import { IgxStacked100ColumnSeriesDynamicModule } from './igx-stacked-100-column-series-dynamic-module';
import { IgxStacked100LineSeriesDynamicModule } from './igx-stacked-100-line-series-dynamic-module';
import { IgxStacked100SplineSeriesDynamicModule } from './igx-stacked-100-spline-series-dynamic-module';
import { IgxStacked100SplineAreaSeriesDynamicModule } from './igx-stacked-100-spline-area-series-dynamic-module';
import { IgxStackedAreaSeriesDynamicModule } from './igx-stacked-area-series-dynamic-module';
import { IgxStackedBarSeriesDynamicModule } from './igx-stacked-bar-series-dynamic-module';
import { IgxStackedColumnSeriesDynamicModule } from './igx-stacked-column-series-dynamic-module';
import { IgxStackedLineSeriesDynamicModule } from './igx-stacked-line-series-dynamic-module';
import { IgxStackedSplineSeriesDynamicModule } from './igx-stacked-spline-series-dynamic-module';
import { IgxStackedSplineAreaSeriesDynamicModule } from './igx-stacked-spline-area-series-dynamic-module';
import { IgxAreaFragmentDynamicModule } from './igx-area-fragment-dynamic-module';
import { IgxBarFragmentDynamicModule } from './igx-bar-fragment-dynamic-module';
import { IgxColumnFragmentDynamicModule } from './igx-column-fragment-dynamic-module';
import { IgxLineFragmentDynamicModule } from './igx-line-fragment-dynamic-module';
import { IgxSplineFragmentDynamicModule } from './igx-spline-fragment-dynamic-module';
import { IgxSplineAreaFragmentDynamicModule } from './igx-spline-area-fragment-dynamic-module';
import { IgxDataChartCategoryCoreDynamicModule } from './igx-data-chart-category-core-dynamic-module';
import { IgxDataChartVerticalCategoryCoreDynamicModule } from './igx-data-chart-vertical-category-core-dynamic-module';
import { IgxDataChartStackedModule } from './igx-data-chart-stacked-module';
import * as i0 from "@angular/core";
export let IgxDataChartStackedDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartStackedDynamicModule {
    }
    IgxDataChartStackedDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartStackedDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartStackedDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartStackedDynamicModule, imports: [CommonModule, IgxStackedFragmentSeriesDynamicModule,
            IgxStacked100AreaSeriesDynamicModule,
            IgxStacked100BarSeriesDynamicModule,
            IgxStacked100ColumnSeriesDynamicModule,
            IgxStacked100LineSeriesDynamicModule,
            IgxStacked100SplineSeriesDynamicModule,
            IgxStacked100SplineAreaSeriesDynamicModule,
            IgxStackedAreaSeriesDynamicModule,
            IgxStackedBarSeriesDynamicModule,
            IgxStackedColumnSeriesDynamicModule,
            IgxStackedLineSeriesDynamicModule,
            IgxStackedSplineSeriesDynamicModule,
            IgxStackedSplineAreaSeriesDynamicModule,
            IgxAreaFragmentDynamicModule,
            IgxBarFragmentDynamicModule,
            IgxColumnFragmentDynamicModule,
            IgxLineFragmentDynamicModule,
            IgxSplineFragmentDynamicModule,
            IgxSplineAreaFragmentDynamicModule,
            IgxDataChartCategoryCoreDynamicModule,
            IgxDataChartVerticalCategoryCoreDynamicModule,
            IgxDataChartStackedModule], exports: [IgxStackedFragmentSeriesDynamicModule,
            IgxStacked100AreaSeriesDynamicModule,
            IgxStacked100BarSeriesDynamicModule,
            IgxStacked100ColumnSeriesDynamicModule,
            IgxStacked100LineSeriesDynamicModule,
            IgxStacked100SplineSeriesDynamicModule,
            IgxStacked100SplineAreaSeriesDynamicModule,
            IgxStackedAreaSeriesDynamicModule,
            IgxStackedBarSeriesDynamicModule,
            IgxStackedColumnSeriesDynamicModule,
            IgxStackedLineSeriesDynamicModule,
            IgxStackedSplineSeriesDynamicModule,
            IgxStackedSplineAreaSeriesDynamicModule,
            IgxAreaFragmentDynamicModule,
            IgxBarFragmentDynamicModule,
            IgxColumnFragmentDynamicModule,
            IgxLineFragmentDynamicModule,
            IgxSplineFragmentDynamicModule,
            IgxSplineAreaFragmentDynamicModule,
            IgxDataChartCategoryCoreDynamicModule,
            IgxDataChartVerticalCategoryCoreDynamicModule,
            IgxDataChartStackedModule] });
    IgxDataChartStackedDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartStackedDynamicModule, imports: [[CommonModule, IgxStackedFragmentSeriesDynamicModule,
                IgxStacked100AreaSeriesDynamicModule,
                IgxStacked100BarSeriesDynamicModule,
                IgxStacked100ColumnSeriesDynamicModule,
                IgxStacked100LineSeriesDynamicModule,
                IgxStacked100SplineSeriesDynamicModule,
                IgxStacked100SplineAreaSeriesDynamicModule,
                IgxStackedAreaSeriesDynamicModule,
                IgxStackedBarSeriesDynamicModule,
                IgxStackedColumnSeriesDynamicModule,
                IgxStackedLineSeriesDynamicModule,
                IgxStackedSplineSeriesDynamicModule,
                IgxStackedSplineAreaSeriesDynamicModule,
                IgxAreaFragmentDynamicModule,
                IgxBarFragmentDynamicModule,
                IgxColumnFragmentDynamicModule,
                IgxLineFragmentDynamicModule,
                IgxSplineFragmentDynamicModule,
                IgxSplineAreaFragmentDynamicModule,
                IgxDataChartCategoryCoreDynamicModule,
                IgxDataChartVerticalCategoryCoreDynamicModule,
                IgxDataChartStackedModule
            ], IgxStackedFragmentSeriesDynamicModule,
            IgxStacked100AreaSeriesDynamicModule,
            IgxStacked100BarSeriesDynamicModule,
            IgxStacked100ColumnSeriesDynamicModule,
            IgxStacked100LineSeriesDynamicModule,
            IgxStacked100SplineSeriesDynamicModule,
            IgxStacked100SplineAreaSeriesDynamicModule,
            IgxStackedAreaSeriesDynamicModule,
            IgxStackedBarSeriesDynamicModule,
            IgxStackedColumnSeriesDynamicModule,
            IgxStackedLineSeriesDynamicModule,
            IgxStackedSplineSeriesDynamicModule,
            IgxStackedSplineAreaSeriesDynamicModule,
            IgxAreaFragmentDynamicModule,
            IgxBarFragmentDynamicModule,
            IgxColumnFragmentDynamicModule,
            IgxLineFragmentDynamicModule,
            IgxSplineFragmentDynamicModule,
            IgxSplineAreaFragmentDynamicModule,
            IgxDataChartCategoryCoreDynamicModule,
            IgxDataChartVerticalCategoryCoreDynamicModule,
            IgxDataChartStackedModule] });
    return IgxDataChartStackedDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartStackedDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxStackedFragmentSeriesDynamicModule,
                        IgxStacked100AreaSeriesDynamicModule,
                        IgxStacked100BarSeriesDynamicModule,
                        IgxStacked100ColumnSeriesDynamicModule,
                        IgxStacked100LineSeriesDynamicModule,
                        IgxStacked100SplineSeriesDynamicModule,
                        IgxStacked100SplineAreaSeriesDynamicModule,
                        IgxStackedAreaSeriesDynamicModule,
                        IgxStackedBarSeriesDynamicModule,
                        IgxStackedColumnSeriesDynamicModule,
                        IgxStackedLineSeriesDynamicModule,
                        IgxStackedSplineSeriesDynamicModule,
                        IgxStackedSplineAreaSeriesDynamicModule,
                        IgxAreaFragmentDynamicModule,
                        IgxBarFragmentDynamicModule,
                        IgxColumnFragmentDynamicModule,
                        IgxLineFragmentDynamicModule,
                        IgxSplineFragmentDynamicModule,
                        IgxSplineAreaFragmentDynamicModule,
                        IgxDataChartCategoryCoreDynamicModule,
                        IgxDataChartVerticalCategoryCoreDynamicModule,
                        IgxDataChartStackedModule
                    ],
                    imports: [CommonModule, IgxStackedFragmentSeriesDynamicModule,
                        IgxStacked100AreaSeriesDynamicModule,
                        IgxStacked100BarSeriesDynamicModule,
                        IgxStacked100ColumnSeriesDynamicModule,
                        IgxStacked100LineSeriesDynamicModule,
                        IgxStacked100SplineSeriesDynamicModule,
                        IgxStacked100SplineAreaSeriesDynamicModule,
                        IgxStackedAreaSeriesDynamicModule,
                        IgxStackedBarSeriesDynamicModule,
                        IgxStackedColumnSeriesDynamicModule,
                        IgxStackedLineSeriesDynamicModule,
                        IgxStackedSplineSeriesDynamicModule,
                        IgxStackedSplineAreaSeriesDynamicModule,
                        IgxAreaFragmentDynamicModule,
                        IgxBarFragmentDynamicModule,
                        IgxColumnFragmentDynamicModule,
                        IgxLineFragmentDynamicModule,
                        IgxSplineFragmentDynamicModule,
                        IgxSplineAreaFragmentDynamicModule,
                        IgxDataChartCategoryCoreDynamicModule,
                        IgxDataChartVerticalCategoryCoreDynamicModule,
                        IgxDataChartStackedModule
                    ],
                    entryComponents: []
                }]
        }] });
