// igniteui-angular-charts/esm2015/lib/igx-item-legend-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxItemLegendComponent } from './igx-item-legend-component';
import { LegendTemplates } from './LegendTemplates';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxItemLegendModule = /*@__PURE__*/ (() => {
    class IgxItemLegendModule {
        constructor() {
            TypeRegistrar.register('LegendTemplates', LegendTemplates.$type);
        }
    }
    IgxItemLegendModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemLegendModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxItemLegendModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemLegendModule, declarations: [IgxItemLegendComponent], imports: [CommonModule], exports: [IgxItemLegendComponent] });
    IgxItemLegendModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemLegendModule, imports: [[CommonModule,
            ]] });
    return IgxItemLegendModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemLegendModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxItemLegendComponent],
                    exports: [IgxItemLegendComponent
                    ],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
