// igniteui-angular-charts/esm2015/lib/igx-rate-of-change-and-momentum-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxRateOfChangeAndMomentumIndicatorComponent } from './igx-rate-of-change-and-momentum-indicator-component';
import * as i0 from "@angular/core";
export let IgxRateOfChangeAndMomentumIndicatorModule = /*@__PURE__*/ (() => {
    class IgxRateOfChangeAndMomentumIndicatorModule {
    }
    IgxRateOfChangeAndMomentumIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRateOfChangeAndMomentumIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxRateOfChangeAndMomentumIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRateOfChangeAndMomentumIndicatorModule, declarations: [IgxRateOfChangeAndMomentumIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxRateOfChangeAndMomentumIndicatorComponent, IgxIndicatorProxyModule] });
    IgxRateOfChangeAndMomentumIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRateOfChangeAndMomentumIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxRateOfChangeAndMomentumIndicatorModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRateOfChangeAndMomentumIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxRateOfChangeAndMomentumIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxRateOfChangeAndMomentumIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
