// igniteui-angular-charts/esm2015/lib/igx-stacked-100-area-series-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStackedAreaSeriesComponent } from "./igx-stacked-area-series-component";
import { IgxHorizontalStackedSeriesBaseComponent } from "./igx-horizontal-stacked-series-base-component";
import { IgxStackedSeriesBaseComponent } from "./igx-stacked-series-base-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { Stacked100AreaSeries } from "./Stacked100AreaSeries";
import * as i0 from "@angular/core";
/**
 * Represents a stacked area series, where values are presented as percentages of the total.
*/
export let IgxStacked100AreaSeriesComponent = /*@__PURE__*/ (() => {
    class IgxStacked100AreaSeriesComponent extends IgxStackedAreaSeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new Stacked100AreaSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        get isPercentBased() {
            return this.i.aan;
        }
    }
    IgxStacked100AreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100AreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStacked100AreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStacked100AreaSeriesComponent, selector: "igx-stacked-100-area-series", providers: [{ provide: IgxStackedAreaSeriesComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStacked100AreaSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100AreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-stacked-100-area-series',
                    template: ``,
                    providers: [{ provide: IgxStackedAreaSeriesComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100AreaSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
