// igniteui-angular-charts/esm2015/lib/igx-data-chart-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumberFormatSpecifierModule } from "igniteui-angular-core";
import { IgxDateTimeFormatSpecifierModule } from "igniteui-angular-core";
import { IgxChartSelectionModule } from './igx-chart-selection-module';
import { IgxSeriesMatcherModule } from './igx-series-matcher-module';
import { IgxDataChartComponent } from './igx-data-chart-component';
import { XamDataChart } from './XamDataChart';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxDataChartCoreModule = /*@__PURE__*/ (() => {
    class IgxDataChartCoreModule {
        constructor() {
            TypeRegistrar.register('XamDataChart', XamDataChart.$type);
        }
    }
    IgxDataChartCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCoreModule, declarations: [IgxDataChartComponent], imports: [CommonModule, IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule,
            IgxChartSelectionModule,
            IgxSeriesMatcherModule], exports: [IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule,
            IgxChartSelectionModule,
            IgxSeriesMatcherModule,
            IgxDataChartComponent] });
    IgxDataChartCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCoreModule, imports: [[CommonModule, IgxNumberFormatSpecifierModule,
                IgxDateTimeFormatSpecifierModule,
                IgxChartSelectionModule,
                IgxSeriesMatcherModule
            ], IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule,
            IgxChartSelectionModule,
            IgxSeriesMatcherModule] });
    return IgxDataChartCoreModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxDataChartComponent],
                    exports: [IgxNumberFormatSpecifierModule,
                        IgxDateTimeFormatSpecifierModule,
                        IgxChartSelectionModule,
                        IgxSeriesMatcherModule,
                        IgxDataChartComponent
                    ],
                    imports: [CommonModule, IgxNumberFormatSpecifierModule,
                        IgxDateTimeFormatSpecifierModule,
                        IgxChartSelectionModule,
                        IgxSeriesMatcherModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
