// igniteui-angular-charts/esm2015/lib/igx-financial-legend-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFinancialLegendComponent } from './igx-financial-legend-component';
import { LegendTemplates } from './LegendTemplates';
import { FinancialLegend } from './FinancialLegend';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxFinancialLegendModule = /*@__PURE__*/ (() => {
    class IgxFinancialLegendModule {
        constructor() {
            TypeRegistrar.register('LegendTemplates', LegendTemplates.$type);
            TypeRegistrar.register('FinancialLegend', FinancialLegend.$type);
        }
    }
    IgxFinancialLegendModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialLegendModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFinancialLegendModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialLegendModule, declarations: [IgxFinancialLegendComponent], imports: [CommonModule], exports: [IgxFinancialLegendComponent] });
    IgxFinancialLegendModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialLegendModule, imports: [[CommonModule,
            ]] });
    return IgxFinancialLegendModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialLegendModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxFinancialLegendComponent],
                    exports: [IgxFinancialLegendComponent
                    ],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
