// igniteui-angular-charts/esm2015/lib/igx-treemap-node-style-mapping-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { TreemapNodeStyleMappingTargetType_$type } from "./TreemapNodeStyleMappingTargetType";
import { TreemapValueMappingMode_$type } from "./TreemapValueMappingMode";
import { IgxTreemapNodeStyleComponent } from "./igx-treemap-node-style-component";
import { TreemapNodeStyleMapping } from "./TreemapNodeStyleMapping";
import { ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a style to apply to a set of nodes in the treemap.
*/
export let IgxTreemapNodeStyleMappingComponent = /*@__PURE__*/ (() => {
    class IgxTreemapNodeStyleMappingComponent extends IgxTreemapNodeStyleComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new TreemapNodeStyleMapping();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets or sets the unique name of the mapping.
        */
        get targetType() {
            return this.i.bo;
        }
        set targetType(v) {
            this.i.bo = ensureEnum(TreemapNodeStyleMappingTargetType_$type, v);
        }
        /**
         * Gets or sets the unique name of the mapping.
        */
        get name() {
            return this.i.b1;
        }
        set name(v) {
            this.i.b1 = v;
        }
        /**
         * Gets or sets the minimum value to which the mapping applies.
        */
        get minimumValue() {
            return this.i.bu;
        }
        set minimumValue(v) {
            this.i.bu = +v;
        }
        /**
         * Gets or sets the maximum value to which the mapping applies.
        */
        get maximumValue() {
            return this.i.bt;
        }
        set maximumValue(v) {
            this.i.bt = +v;
        }
        /**
         * Gets or sets the value to which the mapping applies.
        */
        get value() {
            return this.i.bw;
        }
        set value(v) {
            this.i.bw = v;
        }
        /**
         * Gets or sets the mapping mode to use to evaluate whether the mapping applies.
        */
        get mappingMode() {
            return this.i.bq;
        }
        set mappingMode(v) {
            this.i.bq = ensureEnum(TreemapValueMappingMode_$type, v);
        }
    }
    IgxTreemapNodeStyleMappingComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxTreemapNodeStyleMappingComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxTreemapNodeStyleMappingComponent, selector: "igx-treemap-node-style-mapping", inputs: { targetType: "targetType", name: "name", minimumValue: "minimumValue", maximumValue: "maximumValue", value: "value", mappingMode: "mappingMode" }, providers: [{ provide: IgxTreemapNodeStyleComponent, useExisting: forwardRef(() => IgxTreemapNodeStyleMappingComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxTreemapNodeStyleMappingComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-treemap-node-style-mapping',
                    template: ``,
                    providers: [{ provide: IgxTreemapNodeStyleComponent, useExisting: forwardRef(() => IgxTreemapNodeStyleMappingComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { targetType: [{
                type: Input
            }], name: [{
                type: Input
            }], minimumValue: [{
                type: Input
            }], maximumValue: [{
                type: Input
            }], value: [{
                type: Input
            }], mappingMode: [{
                type: Input
            }] } });
