// igniteui-angular-charts/esm2015/lib/igx-stacked-bar-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStackedBarSeriesComponent } from './igx-stacked-bar-series-component';
import * as i0 from "@angular/core";
export let IgxStackedBarSeriesModule = /*@__PURE__*/ (() => {
    class IgxStackedBarSeriesModule {
    }
    IgxStackedBarSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedBarSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStackedBarSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedBarSeriesModule, declarations: [IgxStackedBarSeriesComponent], imports: [CommonModule], exports: [IgxStackedBarSeriesComponent] });
    IgxStackedBarSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedBarSeriesModule, imports: [[CommonModule]] });
    return IgxStackedBarSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStackedBarSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStackedBarSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxStackedBarSeriesComponent],
                    entryComponents: []
                }]
        }] });
