// igniteui-angular-charts/esm2015/lib/igx-numeric-radius-axis-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumericRadiusAxisModule } from './igx-numeric-radius-axis-module';
import { IgxNumericRadiusAxisComponent } from './igx-numeric-radius-axis-component';
import { NumericRadiusAxis } from './NumericRadiusAxis';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxNumericRadiusAxisDynamicModule = /*@__PURE__*/ (() => {
    class IgxNumericRadiusAxisDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxNumericRadiusAxisComponent", IgxNumericRadiusAxisComponent);
            TypeRegistrar.register("NumericRadiusAxis", NumericRadiusAxis.$type);
        }
    }
    IgxNumericRadiusAxisDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericRadiusAxisDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxNumericRadiusAxisDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericRadiusAxisDynamicModule, imports: [CommonModule, IgxNumericRadiusAxisModule], exports: [IgxNumericRadiusAxisModule] });
    IgxNumericRadiusAxisDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericRadiusAxisDynamicModule, imports: [[CommonModule, IgxNumericRadiusAxisModule], IgxNumericRadiusAxisModule] });
    return IgxNumericRadiusAxisDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericRadiusAxisDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxNumericRadiusAxisModule],
                    exports: [IgxNumericRadiusAxisModule],
                    entryComponents: [IgxNumericRadiusAxisComponent]
                }]
        }], ctorParameters: function () { return []; } });
