// igniteui-angular-charts/esm2015/lib/igx-money-flow-index-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxMoneyFlowIndexIndicatorComponent } from './igx-money-flow-index-indicator-component';
import * as i0 from "@angular/core";
export let IgxMoneyFlowIndexIndicatorModule = /*@__PURE__*/ (() => {
    class IgxMoneyFlowIndexIndicatorModule {
    }
    IgxMoneyFlowIndexIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMoneyFlowIndexIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxMoneyFlowIndexIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMoneyFlowIndexIndicatorModule, declarations: [IgxMoneyFlowIndexIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxMoneyFlowIndexIndicatorComponent, IgxIndicatorProxyModule] });
    IgxMoneyFlowIndexIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMoneyFlowIndexIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxMoneyFlowIndexIndicatorModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMoneyFlowIndexIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxMoneyFlowIndexIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxMoneyFlowIndexIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
