// igniteui-angular-charts/esm2015/lib/igx-polar-scatter-series-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxPolarBaseComponent } from "./igx-polar-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { PolarScatterSeries } from "./PolarScatterSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent polar scatter series.
*/
export let IgxPolarScatterSeriesComponent = /*@__PURE__*/ (() => {
    class IgxPolarScatterSeriesComponent extends IgxPolarBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new PolarScatterSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the series has only marker as visuals
        */
        get hasOnlyMarkers() {
            return this.i.eg;
        }
    }
    IgxPolarScatterSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarScatterSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxPolarScatterSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxPolarScatterSeriesComponent, selector: "igx-polar-scatter-series", providers: [{ provide: IgxPolarBaseComponent, useExisting: forwardRef(() => IgxPolarScatterSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxPolarScatterSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxPolarScatterSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxPolarScatterSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarScatterSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-polar-scatter-series',
                    template: ``,
                    providers: [{ provide: IgxPolarBaseComponent, useExisting: forwardRef(() => IgxPolarScatterSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxPolarScatterSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxPolarScatterSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
