// igniteui-angular-charts/esm2015/lib/igx-scatter-contour-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterContourSeriesModule } from './igx-scatter-contour-series-module';
import { IgxScatterContourSeriesComponent } from './igx-scatter-contour-series-component';
import { ScatterContourSeries } from './ScatterContourSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxScatterContourSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxScatterContourSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxScatterContourSeriesComponent", IgxScatterContourSeriesComponent);
            TypeRegistrar.register("ScatterContourSeries", ScatterContourSeries.$type);
        }
    }
    IgxScatterContourSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterContourSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterContourSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterContourSeriesDynamicModule, imports: [CommonModule, IgxScatterContourSeriesModule], exports: [IgxScatterContourSeriesModule] });
    IgxScatterContourSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterContourSeriesDynamicModule, imports: [[CommonModule, IgxScatterContourSeriesModule], IgxScatterContourSeriesModule] });
    return IgxScatterContourSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterContourSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxScatterContourSeriesModule],
                    exports: [IgxScatterContourSeriesModule],
                    entryComponents: [IgxScatterContourSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
