// igniteui-angular-charts/esm2015/lib/igx-data-chart-interactivity-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { InteractivityFactoryManager } from './InteractivityFactoryManager';
import { IgxDataChartDefaultTooltipsComponent } from './igx-data-chart-default-tooltips-component';
import { IgxDVInteractivityDynamicModule } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDataChartInteractivityModule } from './igx-data-chart-interactivity-module';
import * as i0 from "@angular/core";
export let IgxDataChartInteractivityDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartInteractivityDynamicModule {
        constructor() {
            TypeRegistrar.register('InteractivityFactoryManager', InteractivityFactoryManager.$type);
            InteractivityFactoryManager.register();
            TypeRegistrar.registerCons('IgxDataChartDefaultTooltipsComponent', IgxDataChartDefaultTooltipsComponent);
        }
    }
    IgxDataChartInteractivityDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartInteractivityDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartInteractivityDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartInteractivityDynamicModule, imports: [CommonModule, IgxDVInteractivityDynamicModule,
            IgxDataChartInteractivityModule], exports: [IgxDVInteractivityDynamicModule,
            IgxDataChartInteractivityModule] });
    IgxDataChartInteractivityDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartInteractivityDynamicModule, imports: [[CommonModule, IgxDVInteractivityDynamicModule,
                IgxDataChartInteractivityModule
            ], IgxDVInteractivityDynamicModule,
            IgxDataChartInteractivityModule] });
    return IgxDataChartInteractivityDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartInteractivityDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDVInteractivityDynamicModule,
                        IgxDataChartInteractivityModule
                    ],
                    imports: [CommonModule, IgxDVInteractivityDynamicModule,
                        IgxDataChartInteractivityModule
                    ],
                    entryComponents: [IgxDataChartDefaultTooltipsComponent]
                }]
        }], ctorParameters: function () { return []; } });
