// igniteui-angular-charts/esm2015/lib/igx-accumulation-distribution-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAccumulationDistributionIndicatorModule } from './igx-accumulation-distribution-indicator-module';
import { IgxAccumulationDistributionIndicatorComponent } from './igx-accumulation-distribution-indicator-component';
import { AccumulationDistributionIndicator } from './AccumulationDistributionIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxAccumulationDistributionIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxAccumulationDistributionIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxAccumulationDistributionIndicatorComponent", IgxAccumulationDistributionIndicatorComponent);
            TypeRegistrar.register("AccumulationDistributionIndicator", AccumulationDistributionIndicator.$type);
        }
    }
    IgxAccumulationDistributionIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAccumulationDistributionIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAccumulationDistributionIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAccumulationDistributionIndicatorDynamicModule, imports: [CommonModule, IgxAccumulationDistributionIndicatorModule], exports: [IgxAccumulationDistributionIndicatorModule] });
    IgxAccumulationDistributionIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAccumulationDistributionIndicatorDynamicModule, imports: [[CommonModule, IgxAccumulationDistributionIndicatorModule], IgxAccumulationDistributionIndicatorModule] });
    return IgxAccumulationDistributionIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAccumulationDistributionIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxAccumulationDistributionIndicatorModule],
                    exports: [IgxAccumulationDistributionIndicatorModule],
                    entryComponents: [IgxAccumulationDistributionIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
