// igniteui-angular-charts/esm2015/lib/igx-stacked-100-spline-area-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStacked100SplineAreaSeriesModule } from './igx-stacked-100-spline-area-series-module';
import { IgxStacked100SplineAreaSeriesComponent } from './igx-stacked-100-spline-area-series-component';
import { Stacked100SplineAreaSeries } from './Stacked100SplineAreaSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxStacked100SplineAreaSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxStacked100SplineAreaSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxStacked100SplineAreaSeriesComponent", IgxStacked100SplineAreaSeriesComponent);
            TypeRegistrar.register("Stacked100SplineAreaSeries", Stacked100SplineAreaSeries.$type);
        }
    }
    IgxStacked100SplineAreaSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineAreaSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStacked100SplineAreaSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineAreaSeriesDynamicModule, imports: [CommonModule, IgxStacked100SplineAreaSeriesModule], exports: [IgxStacked100SplineAreaSeriesModule] });
    IgxStacked100SplineAreaSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineAreaSeriesDynamicModule, imports: [[CommonModule, IgxStacked100SplineAreaSeriesModule], IgxStacked100SplineAreaSeriesModule] });
    return IgxStacked100SplineAreaSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineAreaSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStacked100SplineAreaSeriesModule],
                    exports: [IgxStacked100SplineAreaSeriesModule],
                    entryComponents: [IgxStacked100SplineAreaSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
