// igniteui-angular-charts/esm2015/lib/igx-stacked-100-column-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStacked100ColumnSeriesModule } from './igx-stacked-100-column-series-module';
import { IgxStacked100ColumnSeriesComponent } from './igx-stacked-100-column-series-component';
import { Stacked100ColumnSeries } from './Stacked100ColumnSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxStacked100ColumnSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxStacked100ColumnSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxStacked100ColumnSeriesComponent", IgxStacked100ColumnSeriesComponent);
            TypeRegistrar.register("Stacked100ColumnSeries", Stacked100ColumnSeries.$type);
        }
    }
    IgxStacked100ColumnSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100ColumnSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStacked100ColumnSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100ColumnSeriesDynamicModule, imports: [CommonModule, IgxStacked100ColumnSeriesModule], exports: [IgxStacked100ColumnSeriesModule] });
    IgxStacked100ColumnSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100ColumnSeriesDynamicModule, imports: [[CommonModule, IgxStacked100ColumnSeriesModule], IgxStacked100ColumnSeriesModule] });
    return IgxStacked100ColumnSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100ColumnSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStacked100ColumnSeriesModule],
                    exports: [IgxStacked100ColumnSeriesModule],
                    entryComponents: [IgxStacked100ColumnSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
