// igniteui-angular-charts/esm2015/lib/igx-doughnut-chart-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDoughnutChartInteractivityModule } from './igx-doughnut-chart-interactivity-module';
import { IgxDoughnutChartCoreModule } from './igx-doughnut-chart-core-module';
import * as i0 from "@angular/core";
export let IgxDoughnutChartModule = /*@__PURE__*/ (() => {
    class IgxDoughnutChartModule {
    }
    IgxDoughnutChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDoughnutChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartModule, imports: [CommonModule, IgxDoughnutChartInteractivityModule,
            IgxDoughnutChartCoreModule], exports: [IgxDoughnutChartInteractivityModule,
            IgxDoughnutChartCoreModule] });
    IgxDoughnutChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartModule, imports: [[CommonModule, IgxDoughnutChartInteractivityModule,
                IgxDoughnutChartCoreModule
            ], IgxDoughnutChartInteractivityModule,
            IgxDoughnutChartCoreModule] });
    return IgxDoughnutChartModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDoughnutChartModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDoughnutChartInteractivityModule,
                        IgxDoughnutChartCoreModule
                    ],
                    imports: [CommonModule, IgxDoughnutChartInteractivityModule,
                        IgxDoughnutChartCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
