// igniteui-angular-charts/esm2015/lib/igx-radial-line-series-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { UnknownValuePlotting_$type } from "igniteui-angular-core";
import { IgxAnchoredRadialSeriesComponent } from "./igx-anchored-radial-series-component";
import { IgxRadialBaseComponent } from "./igx-radial-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { RadialLineSeries } from "./RadialLineSeries";
import { ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent radial line series.
*/
export let IgxRadialLineSeriesComponent = /*@__PURE__*/ (() => {
    class IgxRadialLineSeriesComponent extends IgxAnchoredRadialSeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new RadialLineSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows only line shapes.
        */
        get isLineOnly() {
            return this.i.fi;
        }
        /**
         * Determines how unknown values will be plotted on the chart.
         * Null and Double.NaN are two examples of unknown values.
        */
        get unknownValuePlotting() {
            return this.i.ade;
        }
        set unknownValuePlotting(v) {
            this.i.ade = ensureEnum(UnknownValuePlotting_$type, v);
        }
        /**
         * When overridden in a derived class, is invoked whenever application code or internal processes
         * call ApplyTemplate.
        
        */
        onApplyTemplate() {
            this.i.ac();
        }
    }
    IgxRadialLineSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialLineSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxRadialLineSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxRadialLineSeriesComponent, selector: "igx-radial-line-series", inputs: { unknownValuePlotting: "unknownValuePlotting" }, providers: [{ provide: IgxAnchoredRadialSeriesComponent, useExisting: forwardRef(() => IgxRadialLineSeriesComponent) }, { provide: IgxRadialBaseComponent, useExisting: forwardRef(() => IgxRadialLineSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxRadialLineSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxRadialLineSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxRadialLineSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialLineSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-radial-line-series',
                    template: ``,
                    providers: [{ provide: IgxAnchoredRadialSeriesComponent, useExisting: forwardRef(() => IgxRadialLineSeriesComponent) }, { provide: IgxRadialBaseComponent, useExisting: forwardRef(() => IgxRadialLineSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxRadialLineSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxRadialLineSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { unknownValuePlotting: [{
                type: Input
            }] } });
