// igniteui-angular-charts/esm2015/lib/igx-slow-stochastic-oscillator-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxSlowStochasticOscillatorIndicatorComponent } from './igx-slow-stochastic-oscillator-indicator-component';
import * as i0 from "@angular/core";
export let IgxSlowStochasticOscillatorIndicatorModule = /*@__PURE__*/ (() => {
    class IgxSlowStochasticOscillatorIndicatorModule {
    }
    IgxSlowStochasticOscillatorIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSlowStochasticOscillatorIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSlowStochasticOscillatorIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSlowStochasticOscillatorIndicatorModule, declarations: [IgxSlowStochasticOscillatorIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxSlowStochasticOscillatorIndicatorComponent, IgxIndicatorProxyModule] });
    IgxSlowStochasticOscillatorIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSlowStochasticOscillatorIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxSlowStochasticOscillatorIndicatorModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSlowStochasticOscillatorIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxSlowStochasticOscillatorIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxSlowStochasticOscillatorIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
