// igniteui-angular-charts/esm2015/lib/igx-funnel-chart-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFunnelChartCoreModule } from './igx-funnel-chart-core-module';
import { IgxDataChartInteractivityModule } from './igx-data-chart-interactivity-module';
import * as i0 from "@angular/core";
export let IgxFunnelChartModule = /*@__PURE__*/ (() => {
    class IgxFunnelChartModule {
    }
    IgxFunnelChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFunnelChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartModule, imports: [CommonModule, IgxFunnelChartCoreModule,
            IgxDataChartInteractivityModule], exports: [IgxFunnelChartCoreModule,
            IgxDataChartInteractivityModule] });
    IgxFunnelChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartModule, imports: [[CommonModule, IgxFunnelChartCoreModule,
                IgxDataChartInteractivityModule
            ], IgxFunnelChartCoreModule,
            IgxDataChartInteractivityModule] });
    return IgxFunnelChartModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxFunnelChartCoreModule,
                        IgxDataChartInteractivityModule
                    ],
                    imports: [CommonModule, IgxFunnelChartCoreModule,
                        IgxDataChartInteractivityModule
                    ],
                    entryComponents: []
                }]
        }] });
