// igniteui-angular-charts/esm2015/lib/igx-financial-price-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxFinancialPriceSeriesProxyModule } from './igx-financial-price-series-proxy-module';
import { IgxFinancialPriceSeriesComponent } from './igx-financial-price-series-component';
import * as i0 from "@angular/core";
export let IgxFinancialPriceSeriesModule = /*@__PURE__*/ (() => {
    class IgxFinancialPriceSeriesModule {
    }
    IgxFinancialPriceSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFinancialPriceSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesModule, declarations: [IgxFinancialPriceSeriesComponent], imports: [CommonModule, IgxFinancialPriceSeriesProxyModule], exports: [IgxFinancialPriceSeriesComponent, IgxFinancialPriceSeriesProxyModule] });
    IgxFinancialPriceSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesModule, imports: [[CommonModule, IgxFinancialPriceSeriesProxyModule], IgxFinancialPriceSeriesProxyModule] });
    return IgxFinancialPriceSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxFinancialPriceSeriesComponent],
                    imports: [CommonModule, IgxFinancialPriceSeriesProxyModule],
                    exports: [IgxFinancialPriceSeriesComponent, IgxFinancialPriceSeriesProxyModule],
                    entryComponents: []
                }]
        }] });
