// igniteui-angular-charts/esm2015/lib/igx-polar-spline-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPolarSplineSeriesComponent } from './igx-polar-spline-series-component';
import * as i0 from "@angular/core";
export let IgxPolarSplineSeriesModule = /*@__PURE__*/ (() => {
    class IgxPolarSplineSeriesModule {
    }
    IgxPolarSplineSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPolarSplineSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesModule, declarations: [IgxPolarSplineSeriesComponent], imports: [CommonModule], exports: [IgxPolarSplineSeriesComponent] });
    IgxPolarSplineSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesModule, imports: [[CommonModule]] });
    return IgxPolarSplineSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxPolarSplineSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxPolarSplineSeriesComponent],
                    entryComponents: []
                }]
        }] });
