// igniteui-angular-charts/esm2015/lib/igx-bubble-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSizeScaleModule } from './igx-size-scale-module';
import { IgxBrushScaleModule } from './igx-brush-scale-module';
import { IgxValueBrushScaleModule } from './igx-value-brush-scale-module';
import { IgxCustomPaletteBrushScaleModule } from './igx-custom-palette-brush-scale-module';
import { IgxBubbleSeriesComponent } from './igx-bubble-series-component';
import * as i0 from "@angular/core";
export let IgxBubbleSeriesModule = /*@__PURE__*/ (() => {
    class IgxBubbleSeriesModule {
    }
    IgxBubbleSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBubbleSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxBubbleSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBubbleSeriesModule, declarations: [IgxBubbleSeriesComponent], imports: [CommonModule, IgxSizeScaleModule, IgxBrushScaleModule, IgxValueBrushScaleModule, IgxCustomPaletteBrushScaleModule], exports: [IgxBubbleSeriesComponent, IgxSizeScaleModule, IgxBrushScaleModule, IgxValueBrushScaleModule, IgxCustomPaletteBrushScaleModule] });
    IgxBubbleSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBubbleSeriesModule, imports: [[CommonModule, IgxSizeScaleModule, IgxBrushScaleModule, IgxValueBrushScaleModule, IgxCustomPaletteBrushScaleModule], IgxSizeScaleModule, IgxBrushScaleModule, IgxValueBrushScaleModule, IgxCustomPaletteBrushScaleModule] });
    return IgxBubbleSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBubbleSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxBubbleSeriesComponent],
                    imports: [CommonModule, IgxSizeScaleModule, IgxBrushScaleModule, IgxValueBrushScaleModule, IgxCustomPaletteBrushScaleModule],
                    exports: [IgxBubbleSeriesComponent, IgxSizeScaleModule, IgxBrushScaleModule, IgxValueBrushScaleModule, IgxCustomPaletteBrushScaleModule],
                    entryComponents: []
                }]
        }] });
