// igniteui-angular-charts/esm2015/lib/igx-line-fragment-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxHorizontalAnchoredCategorySeriesProxyModule } from './igx-horizontal-anchored-category-series-proxy-module';
import { IgxLineFragmentComponent } from './igx-line-fragment-component';
import * as i0 from "@angular/core";
export let IgxLineFragmentModule = /*@__PURE__*/ (() => {
    class IgxLineFragmentModule {
    }
    IgxLineFragmentModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineFragmentModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxLineFragmentModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineFragmentModule, declarations: [IgxLineFragmentComponent], imports: [CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule], exports: [IgxLineFragmentComponent, IgxHorizontalAnchoredCategorySeriesProxyModule] });
    IgxLineFragmentModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineFragmentModule, imports: [[CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule], IgxHorizontalAnchoredCategorySeriesProxyModule] });
    return IgxLineFragmentModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineFragmentModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxLineFragmentComponent],
                    imports: [CommonModule, IgxHorizontalAnchoredCategorySeriesProxyModule],
                    exports: [IgxLineFragmentComponent, IgxHorizontalAnchoredCategorySeriesProxyModule],
                    entryComponents: []
                }]
        }] });
