// igniteui-angular-charts/esm2015/lib/igx-series-viewer-component.js
import { EventEmitter, Output, Component, Input } from '@angular/core';
import { delegateCombine } from "igniteui-angular-core";
import { MarkerAutomaticBehavior_$type } from "./MarkerAutomaticBehavior";
import { SeriesHighlightingMode_$type } from "./SeriesHighlightingMode";
import { SeriesSelectionMode_$type } from "./SeriesSelectionMode";
import { SeriesSelectionBehavior_$type } from "./SeriesSelectionBehavior";
import { SeriesViewerScrollbarMode_$type } from "./SeriesViewerScrollbarMode";
import { SeriesViewerHorizontalScrollbarPosition_$type } from "./SeriesViewerHorizontalScrollbarPosition";
import { SeriesViewerVerticalScrollbarPosition_$type } from "./SeriesViewerVerticalScrollbarPosition";
import { SeriesHighlightingBehavior_$type } from "./SeriesHighlightingBehavior";
import { LegendHighlightingMode_$type } from "./LegendHighlightingMode";
import { LegendItemBadgeShape_$type } from "igniteui-angular-core";
import { LegendItemBadgeMode_$type } from "igniteui-angular-core";
import { ChartHitTestMode_$type } from "./ChartHitTestMode";
import { WindowResponse_$type } from "./WindowResponse";
import { Visibility_$type } from "igniteui-angular-core";
import { ZoomCoercionMode_$type } from "./ZoomCoercionMode";
import { InteractionState_$type } from "igniteui-angular-core";
import { ModifierKeys_$type } from "igniteui-angular-core";
import { SeriesHighlightedValuesDisplayMode_$type } from "./SeriesHighlightedValuesDisplayMode";
import { IgxChartSelectedItemCollection } from "./igx-chart-selected-item-collection";
import { HorizontalAlignment_$type } from "igniteui-angular-core";
import { IgxPlotAreaMouseButtonEventArgs } from "./igx-plot-area-mouse-button-event-args";
import { IgxPlotAreaMouseEventArgs } from "./igx-plot-area-mouse-event-args";
import { IgxChartCursorEventArgs } from "./igx-chart-cursor-event-args";
import { IgxDataChartMouseButtonEventArgs } from "./igx-data-chart-mouse-button-event-args";
import { IgxChartMouseEventArgs } from "./igx-chart-mouse-event-args";
import { IgxSeriesViewerManipulationEventArgs } from "./igx-series-viewer-manipulation-event-args";
import { IgxRectChangedEventArgs } from "igniteui-angular-core";
import { IgxSeriesViewerSelectedSeriesItemsChangedEventArgs } from "./igx-series-viewer-selected-series-items-changed-event-args";
import { IgxRefreshCompletedEventArgs } from "./igx-refresh-completed-event-args";
import { IgxImageCapturedEventArgs } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import { NamePatcher, toBrushCollection, fromBrushCollection, ensureEnum, fromPoint, toPoint, ensureBool, brushToString, stringToBrush, fromRect, toRect, toSpinal, initializePropertiesFromCss } from "igniteui-angular-core";
import { ChartSelectedItemCollection as ChartSelectedItemCollection_internal } from "./ChartSelectedItemCollection";
import { ChartSelection } from "./ChartSelection";
import { SyncableObservableCollection$1 } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a control for viewing series objects.
*/
export let IgxSeriesViewerComponent = /*@__PURE__*/ (() => {
    class IgxSeriesViewerComponent {
        constructor() {
            this._selectedSeriesItems = null;
            this._focusedSeriesItems = null;
            this.__p = null;
            this._hasUserValues = new Set();
            this._stylingContainer = null;
            this._stylingParent = null;
            this._inStyling = false;
            this._plotAreaMouseLeftButtonDown = null;
            this._plotAreaMouseLeftButtonUp = null;
            this._plotAreaClicked = null;
            this._plotAreaMouseEnter = null;
            this._plotAreaMouseLeave = null;
            this._plotAreaMouseOver = null;
            this._seriesCursorMouseMove = null;
            this._seriesMouseLeftButtonDown = null;
            this._seriesMouseLeftButtonUp = null;
            this._seriesClick = null;
            this._seriesMouseMove = null;
            this._seriesMouseEnter = null;
            this._seriesMouseLeave = null;
            this._viewerManipulationStarting = null;
            this._viewerManipulationEnding = null;
            this._windowRectChanged = null;
            this._sizeChanged = null;
            this._actualWindowRectChanged = null;
            this._gridAreaRectChanged = null;
            this._selectedSeriesItemsChanged = null;
            this._focusedSeriesItemsChanged = null;
            this._refreshCompleted = null;
            this._imageCaptured = null;
            this._zoneRunner = null;
            if (this._styling) {
                NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
            }
            this._implementation = this.createImplementation();
            this._implementation.externalObject = this;
            this.onImplementationCreated();
            if (this._initializeAdapters) {
                this._initializeAdapters();
            }
        }
        createImplementation() {
            return null;
        }
        ngOnInit() {
        }
        /**
         * @hidden
         */
        get i() {
            return this._implementation;
        } /**
                                     * @hidden
                                     */
        static _createFromInternal(internal) {
            if (!internal) {
                return null;
            }
            if (!internal.$type) {
                return null;
            }
            let name = internal.$type.name;
            let externalName = "Igx" + name + "Component";
            if (!TypeRegistrar.isRegistered(externalName)) {
                return null;
            }
            return TypeRegistrar.create(externalName);
        }
        onImplementationCreated() {
        }
        /**
         * Gets or sets the palette of brushes used for fill of plotted series.
         * These brushes are automatically assigned to series based on the index of series.
        */
        get brushes() {
            return fromBrushCollection(this.i.ag);
        }
        set brushes(v) {
            this.i.ag = toBrushCollection(v);
        }
        /**
         * Gets or sets the palette of brushes used for outlines of plotted series.
         * These outlines are automatically assigned to series based on the index of series.
        */
        get outlines() {
            return fromBrushCollection(this.i.aj);
        }
        set outlines(v) {
            this.i.aj = toBrushCollection(v);
        }
        /**
         * Gets or sets the palette of brushes used for fill of marker series.
         * These brushes are automatically assigned to markers based on the index of series.
        */
        get markerBrushes() {
            return fromBrushCollection(this.i.ah);
        }
        set markerBrushes(v) {
            this.i.ah = toBrushCollection(v);
        }
        /**
         * Gets or sets the palette of brushes used for outlines of plotted markers.
         * These outlines are automatically assigned to markers based on the index of series.
        */
        get markerOutlines() {
            return fromBrushCollection(this.i.ai);
        }
        set markerOutlines(v) {
            this.i.ai = toBrushCollection(v);
        }
        /**
         * Gets or sets the marker mode used for assigning a marker type to series when the marker type is automatic.
        */
        get markerAutomaticBehavior() {
            return this.i.bg;
        }
        set markerAutomaticBehavior(v) {
            this.i.bg = ensureEnum(MarkerAutomaticBehavior_$type, v);
        }
        /**
         * Gets or sets the cross hair point (in world coordinates)
         * Either or both of the crosshair point's X and Y may be set to double.NaN, in which
         * case the relevant crosshair line is hidden.
        *
        * The `CrosshairPoint` property is used to gets/sets the cross hair point (in world coordinates).
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     crosshairVisibility="visible"
        *     crosshairPoint={x:.8,y:.2}>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.crosshairPoint = {x:.8,y:.2};
        * ```
        */
        get crosshairPoint() {
            return fromPoint(this.i.vc);
        }
        set crosshairPoint(v) {
            this.i.vc = toPoint(v);
        }
        /**
         * Gets or sets whether the series animations should be allowed when a range change has been detected on an axis.
        *
        * `AnimateSeriesWhenAxisRangeChanges` is a Boolean property to override the default behavior in which series do not animate if an axis range changes.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     animateSeriesWhenAxisRangeChanges=true>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         transitionInMode="accordionFromBottom"
        *         isTransitionInEnabled=true
        *         transitionInSpeedType="indexScaled"
        *         transitionDuration=1000>
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.animateSeriesWhenAxisRangeChanges = true;
        * ```
        */
        get animateSeriesWhenAxisRangeChanges() {
            return this.i.eh;
        }
        set animateSeriesWhenAxisRangeChanges(v) {
            this.i.eh = ensureBool(v);
        }
        /**
         * Gets or sets whether to sync the vertical aspect of the window with the contained series content. This is only supported for a subset of the available series.
        */
        get isWindowSyncedToVisibleRange() {
            return this.i.ff;
        }
        set isWindowSyncedToVisibleRange(v) {
            this.i.ff = ensureBool(v);
        }
        /**
         * Gets or sets whether calling SimulateHover should shift the crosshair point.
        */
        get shouldSimulateHoverMoveCrosshairPoint() {
            return this.i.f1;
        }
        set shouldSimulateHoverMoveCrosshairPoint(v) {
            this.i.f1 = ensureBool(v);
        }
        /**
         * Gets or sets the highlighting mode to use for the series in the component, when supported. This takes precedence over the series level IsHighlightingEnabled.
        */
        get highlightingMode() {
            return this.i.by;
        }
        set highlightingMode(v) {
            this.i.by = ensureEnum(SeriesHighlightingMode_$type, v);
        }
        /**
         * Gets or sets the selection mode to use for the series in the component, when supported.
        */
        get selectionMode() {
            return this.i.cb;
        }
        set selectionMode(v) {
            this.i.cb = ensureEnum(SeriesSelectionMode_$type, v);
        }
        /**
         * Gets or sets the selection behavior to use for the series in the component, when supported.
        */
        get selectionBehavior() {
            return this.i.b6;
        }
        set selectionBehavior(v) {
            this.i.b6 = ensureEnum(SeriesSelectionBehavior_$type, v);
        }
        /**
         * Gets or sets the focus mode to use for the series in the component, when supported.
        */
        get focusMode() {
            return this.i.b9;
        }
        set focusMode(v) {
            this.i.b9 = ensureEnum(SeriesSelectionMode_$type, v);
        }
        /**
         * Gets or sets the horizontal scrollbar mode to use for the series viewer.
        */
        get horizontalViewScrollbarMode() {
            return this.i.ch;
        }
        set horizontalViewScrollbarMode(v) {
            this.i.ch = ensureEnum(SeriesViewerScrollbarMode_$type, v);
        }
        /**
         * Gets or sets the vertical scrollbar mode to use for the series viewer.
        */
        get verticalViewScrollbarMode() {
            return this.i.ci;
        }
        set verticalViewScrollbarMode(v) {
            this.i.ci = ensureEnum(SeriesViewerScrollbarMode_$type, v);
        }
        /**
         * Gets or sets the position to use for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarPosition() {
            return this.i.ce;
        }
        set horizontalViewScrollbarPosition(v) {
            this.i.ce = ensureEnum(SeriesViewerHorizontalScrollbarPosition_$type, v);
        }
        /**
         * Gets or sets the position to use for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarPosition() {
            return this.i.cl;
        }
        set verticalViewScrollbarPosition(v) {
            this.i.cl = ensureEnum(SeriesViewerVerticalScrollbarPosition_$type, v);
        }
        /**
         * Gets or sets the fill to use for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarFill() {
            return brushToString(this.i.uz);
        }
        set horizontalViewScrollbarFill(v) {
            this.i.uz = stringToBrush(v);
        }
        /**
         * Gets or sets the outline to use for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarOutline() {
            return brushToString(this.i.u0);
        }
        set horizontalViewScrollbarOutline(v) {
            this.i.u0 = stringToBrush(v);
        }
        /**
         * Gets or sets the stroke thickness to use for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarStrokeThickness() {
            return this.i.hr;
        }
        set horizontalViewScrollbarStrokeThickness(v) {
            this.i.hr = +v;
        }
        /**
         * Gets or sets the max opacity to use for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarMaxOpacity() {
            return this.i.hq;
        }
        set horizontalViewScrollbarMaxOpacity(v) {
            this.i.hq = +v;
        }
        /**
         * Gets or sets the corner radius to use for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarCornerRadius() {
            return this.i.hn;
        }
        set horizontalViewScrollbarCornerRadius(v) {
            this.i.hn = +v;
        }
        /**
         * Gets or sets the height to use for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarHeight() {
            return this.i.ho;
        }
        set horizontalViewScrollbarHeight(v) {
            this.i.ho = +v;
        }
        /**
         * Gets or sets the inset distance to use for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarInset() {
            return this.i.hp;
        }
        set horizontalViewScrollbarInset(v) {
            this.i.hp = +v;
        }
        /**
         * Gets or sets the track start inset to use for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarTrackStartInset() {
            return this.i.ht;
        }
        set horizontalViewScrollbarTrackStartInset(v) {
            this.i.ht = +v;
        }
        /**
         * Gets or sets the track end inset to use for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarTrackEndInset() {
            return this.i.hs;
        }
        set horizontalViewScrollbarTrackEndInset(v) {
            this.i.hs = +v;
        }
        /**
         * Gets or sets whether to use automatic track insets for the horizontal scrollbar in the component, when enabled.
        */
        get horizontalViewScrollbarShouldAddAutoTrackInsets() {
            return this.i.ev;
        }
        set horizontalViewScrollbarShouldAddAutoTrackInsets(v) {
            this.i.ev = ensureBool(v);
        }
        /**
         * Gets or sets the fill to use for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarFill() {
            return brushToString(this.i.u9);
        }
        set verticalViewScrollbarFill(v) {
            this.i.u9 = stringToBrush(v);
        }
        /**
         * Gets or sets the outline to use for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarOutline() {
            return brushToString(this.i.va);
        }
        set verticalViewScrollbarOutline(v) {
            this.i.va = stringToBrush(v);
        }
        /**
         * Gets or sets the stroke thickness to use for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarStrokeThickness() {
            return this.i.id;
        }
        set verticalViewScrollbarStrokeThickness(v) {
            this.i.id = +v;
        }
        /**
         * Gets or sets the max opacity to use for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarMaxOpacity() {
            return this.i.ic;
        }
        set verticalViewScrollbarMaxOpacity(v) {
            this.i.ic = +v;
        }
        /**
         * Gets or sets the corner radius to use for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarCornerRadius() {
            return this.i.ia;
        }
        set verticalViewScrollbarCornerRadius(v) {
            this.i.ia = +v;
        }
        /**
         * Gets or sets the width to use for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarWidth() {
            return this.i.ih;
        }
        set verticalViewScrollbarWidth(v) {
            this.i.ih = +v;
        }
        /**
         * Gets or sets the inset distance to use for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarInset() {
            return this.i.ib;
        }
        set verticalViewScrollbarInset(v) {
            this.i.ib = +v;
        }
        /**
         * Gets or sets the track start inset to use for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarTrackStartInset() {
            return this.i.ig;
        }
        set verticalViewScrollbarTrackStartInset(v) {
            this.i.ig = +v;
        }
        /**
         * Gets or sets the track end inset to use for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarTrackEndInset() {
            return this.i.ie;
        }
        set verticalViewScrollbarTrackEndInset(v) {
            this.i.ie = +v;
        }
        /**
         * Gets or sets whether to use automatic track insets for the vertical scrollbar in the component, when enabled.
        */
        get verticalViewScrollbarShouldAddAutoTrackInsets() {
            return this.i.gd;
        }
        set verticalViewScrollbarShouldAddAutoTrackInsets(v) {
            this.i.gd = ensureBool(v);
        }
        /**
         * Gets or sets the highlighting Behavior to use for the series in the component, when supported. This takes precedence over the series level IsHighlightingEnabled.
        */
        get highlightingBehavior() {
            return this.i.bw;
        }
        set highlightingBehavior(v) {
            this.i.bw = ensureEnum(SeriesHighlightingBehavior_$type, v);
        }
        /**
         * Gets or sets the highlighting mode to use for the legend linked to the component, when supported.
        */
        get legendHighlightingMode() {
            return this.i.bd;
        }
        set legendHighlightingMode(v) {
            this.i.bd = ensureEnum(LegendHighlightingMode_$type, v);
        }
        /**
         * Gets or sets type of legend badges representing all series displayed in a legend linked to this component
         * This property will be ignored when the LegendItemTemplate or LegendItemBadgeTemplate property is set on a series
        */
        get legendItemBadgeShape() {
            return this.i.bf;
        }
        set legendItemBadgeShape(v) {
            this.i.bf = ensureEnum(LegendItemBadgeShape_$type, v);
        }
        /**
         * Gets or sets the mode of legend badges representing all series in this chart.
         * This property will be ignored when the LegendItemTemplate or LegendItemBadgeTemplate property is set on a series
        */
        get legendItemBadgeMode() {
            return this.i.be;
        }
        set legendItemBadgeMode(v) {
            this.i.be = ensureEnum(LegendItemBadgeMode_$type, v);
        }
        /**
         * Gets the viewport rectangle associated with the SeriesViewer, the physical dimensions of the plot area.
        *
        * ```ts
        * let viewPortRect: IgxRect = this.chart.viewportRect;
        * ```
        */
        get viewportRect() {
            return fromRect(this.i.v0);
        }
        /**
         * Gets the EffectiveViewport rectangle, representing the effective viewport area after margins have been subtracted.
        *
        *  ``ts
        * let viewport: IgxRect = this.chart.effectiveViewport;
        *
        * ```
        *
        * ```
        */
        get effectiveViewport() {
            return fromRect(this.i.vp);
        }
        set effectiveViewport(v) {
            this.i.vp = toRect(v);
        }
        /**
         * Represents the subset area within the viewport into which content should be mapped.
         * This is the projection of the current EffectiveViewport into the viewport at current zoom level.
        *
        * ```ts
        * let contentViewport: IgxRect = this.chart.contentViewport;
        * ```
        */
        get contentViewport() {
            return fromRect(this.i.vo);
        }
        /**
         * A rectangle representing the portion of the SeriesViewer currently in view.
         * A rectangle at X=0, Y=0 with a Height and Width of 1 implies the entire plotting area is in view.  A Height and Width of .5 would imply that the view is halfway zoomed in.
        *
        * You can set the `WindowRect` to zoom in on an area of interest in the chart.
        *
        * ```ts
        * this.chart.windowRect = {left:0, top:1, width: .5, height: .5};
        * ```
        */
        get windowRect() {
            return fromRect(this.i.v1);
        }
        set windowRect(v) {
            this.i.v1 = toRect(v);
        }
        /**
         * Gets or sets whether the series viewer should use cached tiles during zooms rather than the default live content.
        */
        get useTiledZooming() {
            return this.i.gc;
        }
        set useTiledZooming(v) {
            this.i.gc = ensureBool(v);
        }
        /**
         * Gets or sets whether the series viewer should prefer selecting higher resolution tiles over lower resolution tiles when performing tile zooming. Setting this to true will lower performance but increase quality.
        */
        get preferHigherResolutionTiles() {
            return this.i.fv;
        }
        set preferHigherResolutionTiles(v) {
            this.i.fv = ensureBool(v);
        }
        /**
         * Gets or sets the maximum number of zoom tiles that the series viewer should cache while in tiled zooming mode.
        */
        get zoomTileCacheSize() {
            return this.i.i1;
        }
        set zoomTileCacheSize(v) {
            this.i.i1 = +v;
        }
        /**
         * Gets or sets the duration the highlighting/de-highlighting effects.
        *
        * The `HighlightingTransitionDuration` property is used to set the time for highlighting transition in milliseconds .
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        * 	HighlightingTransitionDuration=500 >
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value" >
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * chart.HighlightingTransitionDuration=500;
        * ```
        */
        get highlightingTransitionDuration() {
            return this.i.iy;
        }
        set highlightingTransitionDuration(v) {
            this.i.iy = +v;
        }
        /**
         * Gets or sets the duration the selection/de-Selection effects.
        */
        get selectionTransitionDuration() {
            return this.i.i0;
        }
        set selectionTransitionDuration(v) {
            this.i.i0 = +v;
        }
        /**
         * Gets or sets the duration the Focus/de-Focus effects.
        */
        get focusTransitionDuration() {
            return this.i.iw;
        }
        set focusTransitionDuration(v) {
            this.i.iw = +v;
        }
        /**
         * Gets or sets the duration of the scrollbar effects.
        */
        get scrollbarsAnimationDuration() {
            return this.i.iz;
        }
        set scrollbarsAnimationDuration(v) {
            this.i.iz = +v;
        }
        /**
         * Gets or sets the whether the series viewer can allow the page to pan if a control pan is not possible in the requested direction.
        */
        get isPagePanningAllowed() {
            return this.i.fa;
        }
        set isPagePanningAllowed(v) {
            this.i.fa = ensureBool(v);
        }
        /**
         * Determines the method of hit testing used when mousing over elements in the chart.
        *
        * ```ts
        * this.chart.contentHitTestMode = ChartHitTestMode.Computational;
        * ```
        */
        get contentHitTestMode() {
            return this.i.av;
        }
        set contentHitTestMode(v) {
            this.i.av = ensureEnum(ChartHitTestMode_$type, v);
        }
        /**
         * Resolved method of hit testing used when the pointer is over elements in the chart.
        *
        * ```ts
        * let actualContentHitTestMode: ChartHitTestMode  =this.chart.actualContentHitTestMode;
        * ```
        */
        get actualContentHitTestMode() {
            return this.i.au;
        }
        set actualContentHitTestMode(v) {
            this.i.au = ensureEnum(ChartHitTestMode_$type, v);
        }
        /**
         * Gets or sets whether all surface interactions with the plot area should be disabled.
        *
        * The `IsSurfaceInteractionDisabled` property is used to enable/disable interactions with the plot surface.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     isSurfaceInteractionDisabled=true>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * chart.isSurfaceInteractionDisabled=true;
        * ```
        */
        get isSurfaceInteractionDisabled() {
            return this.i.fb;
        }
        set isSurfaceInteractionDisabled(v) {
            this.i.fb = ensureBool(v);
        }
        /**
         * Indicates if this SeriesViewer is a map.
        */
        get isMap() {
            return this.i.e9;
        }
        /**
         * The response to user panning and zooming: whether to update the view immediately while the user action is happening, or to defer the update to after the user action is complete.  The user action will be an action such as a mouse drag which causes panning and/or zooming to occur.
        *
        * The `WindowResponse` property is used to set the response to user panning and zooming: whether to update the view immediately while the user action is happening, or to defer the update to after the user action is complete.
        * The user action will be an action such as a mouse drag which causes panning and/or zooming to occur.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     isHorizontalZoomEnabled=true
        *     windowResponse="deferred">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.windowResponse="deferred";
        * ```
        */
        get windowResponse() {
            return this.i.cp;
        }
        set windowResponse(v) {
            this.i.cp = ensureEnum(WindowResponse_$type, v);
        }
        /**
         * Sets or gets the minimum width that the window rect is allowed to reach before being clamped.
         * Decrease this value if you want to allow for further zooming into the viewer.
         * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
         * This property accepts values between 0.0 (max zooming) and 1.0 (no zooming)
        *
        * The `WindowRectMinWidth` property is used to set or get the minimum width that the window rect is allowed to reach before being clamped.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     isHorizontalZoomEnabled=true
        *     windowRectMinWidth=.2>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.WindowRectMinWidth=".2";
        * ```
        */
        get windowRectMinWidth() {
            return this.i.il;
        }
        set windowRectMinWidth(v) {
            this.i.il = +v;
        }
        /**
         * Sets or gets the minimum height that the window rect is allowed to reach before being clamped.
         * Decrease this value if you want to allow for further zooming into the viewer.
         * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
         * This property accepts values between 0.0 (max zooming) and 1.0 (no zooming)
        */
        get windowRectMinHeight() {
            return this.i.ik;
        }
        set windowRectMinHeight(v) {
            this.i.ik = +v;
        }
        /**
         * Gets the actual minimum width that the window rect is allowed to reach before being clamped.
        */
        get actualWindowRectMinWidth() {
            return this.i.hf;
        }
        set actualWindowRectMinWidth(v) {
            this.i.hf = +v;
        }
        /**
         * Gets the actual minimum height that the window rect is allowed to reach before being clamped.
        */
        get actualWindowRectMinHeight() {
            return this.i.he;
        }
        set actualWindowRectMinHeight(v) {
            this.i.he = +v;
        }
        /**
         * Sets or gets minimum pixel width that the window is allowed to reach before being clamped.
         * Decrease this value if you want to allow for further zooming into the viewer.
         * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
         * This property is overridden by the WindowRectMinWidth property
        */
        get windowSizeMinWidth() {
            return this.i.io;
        }
        set windowSizeMinWidth(v) {
            this.i.io = +v;
        }
        /**
         * Sets or gets minimum pixel height that the window is allowed to reach before being clamped.
         * Decrease this value if you want to allow for further zooming into the viewer.
         * If this value is lowered too much it can cause graphical corruption due to floating point arithmetic inaccuracy.
         * This property is overridden by the WindowRectMinHeight property
        */
        get windowSizeMinHeight() {
            return this.i.im;
        }
        set windowSizeMinHeight(v) {
            this.i.im = +v;
        }
        /**
         * Gets or sets the current SeriesViewer's crosshair visibility override.
        *
        * The `CrosshairVisibility` property is used to gets or sets the current Chart's crosshair visibility override. Note: setting this property does not affect the mobile browser version of the chart.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     crosshairVisibility="visible">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.crosshairVisibility = Visibility.Visible;
        * ```
        */
        get crosshairVisibility() {
            return this.i.v5;
        }
        set crosshairVisibility(v) {
            this.i.v5 = ensureEnum(Visibility_$type, v);
        }
        /**
         * Gets or sets the current SeriesViewer's crosshair horizontal brush override.
        */
        get horizontalCrosshairBrush() {
            return brushToString(this.i.uy);
        }
        set horizontalCrosshairBrush(v) {
            this.i.uy = stringToBrush(v);
        }
        /**
         * Gets or sets the current SeriesViewer's crosshair vertical brush override.
        */
        get verticalCrosshairBrush() {
            return brushToString(this.i.u8);
        }
        set verticalCrosshairBrush(v) {
            this.i.u8 = stringToBrush(v);
        }
        /**
         * Gets or sets zoom is constrained to within the axes. Setting this to false is a preview feature at the present time.
        */
        get zoomCoercionMode() {
            return this.i.ct;
        }
        set zoomCoercionMode(v) {
            this.i.ct = ensureEnum(ZoomCoercionMode_$type, v);
        }
        /**
         * Gets or sets the brush used as the background for the current SeriesViewer object's plot area.
        *
        * `PlotAreaBackground` property set the background brush to the current Chart object's plot area.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     plotAreaBackground="#F0F8FF">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.plotAreaBackground="#F0F8FF";
        * ```
        */
        get plotAreaBackground() {
            return brushToString(this.i.u1);
        }
        set plotAreaBackground(v) {
            this.i.u1 = stringToBrush(v);
        }
        /**
         * Gets or sets the DefaultInteraction property.
         * The default interaction state defines the SeriesViewer's response to mouse events.
        *
        * The default interaction state defines the chart's response to mouse events.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     isHorizontalZoomEnabled=true
        *     defaultInteraction="dragPan">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.defaultInteraction = InteractionState.DragPan;
        * ```
        */
        get defaultInteraction() {
            return this.i.cw;
        }
        set defaultInteraction(v) {
            this.i.cw = ensureEnum(InteractionState_$type, v);
        }
        /**
         * Gets or sets the RightButtonDefaultInteraction property.
         * The default interaction state defines the SeriesViewer's response to right button mouse events.
        */
        get rightButtonDefaultInteraction() {
            return this.i.cy;
        }
        set rightButtonDefaultInteraction(v) {
            this.i.cy = ensureEnum(InteractionState_$type, v);
        }
        /**
         * Gets or sets the current SeriesViewer's DragModifier property.
        *
        * `DragModifier` is useful to enable zooming when zooming is not the default interaction.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     isHorizontalZoomEnabled=true
        *     defaultInteraction="dragPan"
        *     dragModifier="alt">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.dragModifier = ModifierKeys.Alt;
        * ```
        */
        get dragModifier() {
            return this.i.uj;
        }
        set dragModifier(v) {
            this.i.uj = ensureEnum(ModifierKeys_$type, v);
        }
        /**
         * Gets or sets the current SeriesViewer's PanModifier property.
        *
        * `PanModifier` is useful to enable panning when panning is not the default interaction.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     isHorizontalZoomEnabled=true
        *     defaultInteraction="dragZoom"
        *     panModifier="alt">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.panModifier = ModifierKeys.Alt;
        * ```
        */
        get panModifier() {
            return this.i.uk;
        }
        set panModifier(v) {
            this.i.uk = ensureEnum(ModifierKeys_$type, v);
        }
        /**
         * Gets or sets the preview rectangle.
         * The preview rectangle may be set to Rect.Empty, in which case the visible preview
         * strokePath is hidden.
        *
        * `PreviewRect` can be used to highlight an area of importance.
        *
        * ```ts
        * this.chart.previewRect = {left:0,top:0,height:.5,width:.5};
        * ```
        */
        get previewRect() {
            return fromRect(this.i.vv);
        }
        set previewRect(v) {
            this.i.vv = toRect(v);
        }
        /**
         * Gets or sets whether and how to display highlighted values for the series by default. Note, this is distinct from the highlighting feature that indicates what is closest or under the mouse.
        */
        get highlightedValuesDisplayMode() {
            return this.i.bt;
        }
        set highlightedValuesDisplayMode(v) {
            this.i.bt = ensureEnum(SeriesHighlightedValuesDisplayMode_$type, v);
        }
        /**
         * Gets the currently selected data items.  Adding or removing data items from this collection will
         * select or deselect the visuals associated with those items.
        */
        get selectedSeriesItems() {
            if (this._selectedSeriesItems === null) {
                let coll = new IgxChartSelectedItemCollection();
                let innerColl = this.i.selectedSeriesItems;
                if (!innerColl) {
                    innerColl = new ChartSelectedItemCollection_internal();
                }
                this._selectedSeriesItems = coll._fromInner(innerColl);
                this.i.selectedSeriesItems = innerColl;
            }
            return this._selectedSeriesItems;
        }
        set selectedSeriesItems(v) {
            if (this._selectedSeriesItems !== null) {
                this._selectedSeriesItems._setSyncTarget(null);
                this._selectedSeriesItems = null;
            }
            let coll = new IgxChartSelectedItemCollection();
            this._selectedSeriesItems = coll._fromOuter(v);
            let syncColl = new SyncableObservableCollection$1(ChartSelection.$type);
            let innerColl = this.i.selectedSeriesItems;
            if (!innerColl) {
                innerColl = new ChartSelectedItemCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._selectedSeriesItems._setSyncTarget(syncColl);
            this.i.selectedSeriesItems = innerColl;
        }
        /**
         * Gets the currently focused data items.  Adding or removing data items from this collection will
         * focus or blur the visuals associated with those items.
        */
        get focusedSeriesItems() {
            if (this._focusedSeriesItems === null) {
                let coll = new IgxChartSelectedItemCollection();
                let innerColl = this.i.focusedSeriesItems;
                if (!innerColl) {
                    innerColl = new ChartSelectedItemCollection_internal();
                }
                this._focusedSeriesItems = coll._fromInner(innerColl);
                this.i.focusedSeriesItems = innerColl;
            }
            return this._focusedSeriesItems;
        }
        set focusedSeriesItems(v) {
            if (this._focusedSeriesItems !== null) {
                this._focusedSeriesItems._setSyncTarget(null);
                this._focusedSeriesItems = null;
            }
            let coll = new IgxChartSelectedItemCollection();
            this._focusedSeriesItems = coll._fromOuter(v);
            let syncColl = new SyncableObservableCollection$1(ChartSelection.$type);
            let innerColl = this.i.focusedSeriesItems;
            if (!innerColl) {
                innerColl = new ChartSelectedItemCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._focusedSeriesItems._setSyncTarget(syncColl);
            this.i.focusedSeriesItems = innerColl;
        }
        /**
         * A number between 0 and 1 determining the position of the horizontal scroll.
         * This property is effectively a shortcut to the X position of the WindowRect property.
        *
        * ```ts
        * this.chart.windowPositionHorizontal = .2;
        * ```
        */
        get windowPositionHorizontal() {
            return this.i.ii;
        }
        set windowPositionHorizontal(v) {
            this.i.ii = +v;
        }
        /**
         * A number between 0 and 1 determining the position of the vertical scroll.
         * This property is effectively a shortcut to the Y position of the WindowRect property.
        *
        * ```ts
        * this.chart.windowPositionVertical = .2;
        * ```
        */
        get windowPositionVertical() {
            return this.i.ij;
        }
        set windowPositionVertical(v) {
            this.i.ij = +v;
        }
        /**
         * Gets or sets the selection brush to use for the series.
        */
        get selectionBrush() {
            return brushToString(this.i.u5);
        }
        set selectionBrush(v) {
            this.i.u5 = stringToBrush(v);
        }
        /**
         * Gets or sets the Focus brush to use for the series.
        */
        get focusBrush() {
            return brushToString(this.i.uo);
        }
        set focusBrush(v) {
            this.i.uo = stringToBrush(v);
        }
        /**
         * Text to display above the plot area.
        *
        * The `Title` property is used to set the title to display for the component.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.title="Items Sold";
        * ```
        */
        get chartTitle() {
            return this.i.ls;
        }
        set chartTitle(v) {
            this.i.ls = v;
        }
        /**
         * Gets or sets the TitleHorizontalAlignment property.
        *
        * The `subtitleHorizontalAlignment` property is used to set horizontal alignment for the subtitle.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     subtitle="date of sale: 1/1/2018"
        *     titleHorizontalAlignment="left"
        *     subtitleHorizontalAlignment="left">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.titleHorizontalAlignment="left";
        * ```
        */
        get titleHorizontalAlignment() {
            return this.i.ui;
        }
        set titleHorizontalAlignment(v) {
            this.i.ui = ensureEnum(HorizontalAlignment_$type, v);
        }
        /**
         * Gets or sets the SubtitleHorizontalAlignment property.
        *
        * The `subtitleBottomMargin` property is used to set the horizontal alignment for the subtitle.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     subtitle="Date of Sale: 1/1/2018"
        *     titleHorizontalAlignment="left"
        *     subtitleHorizontalAlignment="left">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.subtitleHorizontalAlignment="left";
        * ```
        */
        get subtitleHorizontalAlignment() {
            return this.i.uh;
        }
        set subtitleHorizontalAlignment(v) {
            this.i.uh = ensureEnum(HorizontalAlignment_$type, v);
        }
        /**
         * Font settings for the title.
        */
        get titleTextStyle() {
            return this.i.lz;
        }
        set titleTextStyle(v) {
            this.i.lz = v;
        }
        /**
         * Font settings for the subtitle.
        */
        get subtitleTextStyle() {
            return this.i.ln;
        }
        set subtitleTextStyle(v) {
            this.i.ln = v;
        }
        /**
         * Gets or sets the Color used for the Title Text.
        *
        * The `titleTextColor` property is used to set the title brush.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     subtitle="date of sale: 1/1/2018"
        *     titleTextColor="blue">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.titleTextColor="blue" ;
        * ```
        */
        get titleTextColor() {
            return brushToString(this.i.u7);
        }
        set titleTextColor(v) {
            this.i.u7 = stringToBrush(v);
        }
        /**
         * Gets or sets the Color used for the Subtitle Text.
        *
        * The `subtitleTextColor` property is used to set the color for the subtitle.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     subtitle="Date of Sale: 1/1/2018"
        *     subtitleTextColor="blue">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.subtitleTextColor="blue";
        * ```
        */
        get subtitleTextColor() {
            return brushToString(this.i.u6);
        }
        set subtitleTextColor(v) {
            this.i.u6 = stringToBrush(v);
        }
        /**
         * Margin applied above the title.
        *
        * The `titleTopMargin` property is used to set the top margin for the title.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     titleLeftMarging=10
        *     titleTopMargin=10
        *     titleRightMargin=10
        *     titleBottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.titleTopMargin=10;
        * ```
        */
        get titleTopMargin() {
            return this.i.h8;
        }
        set titleTopMargin(v) {
            this.i.h8 = +v;
        }
        /**
         * Margin applied to the left of the title.
        *
        * The `titleLeftMarging` property is used to set the left margin for the title.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     titleLeftMarging=10
        *     titleTopMargin=10
        *     titleRightMargin=10
        *     titleBottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.titleLeftMarging=10;
        * ```
        */
        get titleLeftMargin() {
            return this.i.h6;
        }
        set titleLeftMargin(v) {
            this.i.h6 = +v;
        }
        /**
         * Margin applied to the right of the title.
        *
        * The `titleRightMargin` property is used to set the right margin for the title.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     titleLeftMarging=10
        *     titleTopMargin=10
        *     titleRightMargin=10
        *     titleBottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.titleRightMargin=10;
        * ```
        */
        get titleRightMargin() {
            return this.i.h7;
        }
        set titleRightMargin(v) {
            this.i.h7 = +v;
        }
        /**
         * Margin applied below the title.
        *
        * The `titleBottomMargin` property is used to set the bottom margin for the title.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     titleLeftMarging=10
        *     titleTopMargin=10
        *     titleRightMargin=10
        *     titleBottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.titleBottomMargin=20;
        * ```
        */
        get titleBottomMargin() {
            return this.i.h5;
        }
        set titleBottomMargin(v) {
            this.i.h5 = +v;
        }
        /**
         * Margin applied above the subtitle.
        *
        * The `subtitleTopMargin` property is used to set the top margin for the subtitle.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     subtitle="date of sale: 1/1/2018"
        *     subtitleLeftMarging=10
        *     subtitleTopMargin=10
        *     subtitleRightMargin=10
        *     subtitleBottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.subtitleTopMargin =10;
        * ```
        */
        get subtitleTopMargin() {
            return this.i.h4;
        }
        set subtitleTopMargin(v) {
            this.i.h4 = +v;
        }
        /**
         * Margin applied to the left of the subtitle.
        *
        * The `subtitleLeftMarging` property is used to set the left margin for the subtitle.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     subtitle="date of sale: 1/1/2018"
        *     subtitleLeftMarging=10
        *     subtitleTopMargin=10
        *     subtitleRightMargin=10
        *     subtitleBottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.subtitleLeftMarging=10;
        * ```
        */
        get subtitleLeftMargin() {
            return this.i.h2;
        }
        set subtitleLeftMargin(v) {
            this.i.h2 = +v;
        }
        /**
         * Margin applied to the right of the subtitle.
        *
        * The `subTitleRightMargin` property is used to set the right margin for the subtitle.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     subtitle="date of sale: 1/1/2018"
        *     subtitleLeftMarging=10
        *     subtitleTopMargin=10
        *     subTitleRightMargin=10
        *     subtitleBottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.subtitleLeftMarging=10;
        * ```
        */
        get subtitleRightMargin() {
            return this.i.h3;
        }
        set subtitleRightMargin(v) {
            this.i.h3 = +v;
        }
        /**
         * Margin applied below the subtitle.
        *
        * The `subtitleBottomMargin` property is used to set the bottom margin for the subtitle.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     subtitle="date of sale: 1/1/2018"
        *     subtitleLeftMarging=10
        *     subtitleTopMargin=10
        *     subtitleRightMargin=10
        *     subtitleBottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.subtitleBottomMargin = 20;
        * ```
        */
        get subtitleBottomMargin() {
            return this.i.h1;
        }
        set subtitleBottomMargin(v) {
            this.i.h1 = +v;
        }
        /**
         * Gets or sets the Subtitle property.
         * The Title property defines the Subtitle of the chart
        *
        * The `Subtitle` property is used to display subtitle for the component.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     title="Items Sold"
        *     subtitle="Date of Sale: 1/1/2018">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.subtitle="Date of Sale: 1/1/2018";
        * ```
        */
        get subtitle() {
            return this.i.lg;
        }
        set subtitle(v) {
            this.i.lg = v;
        }
        /**
         * Sets or gets the top margin to use around the chart content.
        *
        * The `topMargin` property is used to set the top margin.
        *
        * ```html
        *  <igx-data-chart
        *     [dataSource]="data"
        *     leftMargin=20
        *     rightMargin=20
        *     topMargin=30
        *     bottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.topMargin=30;
        * ```
        */
        get topMargin() {
            return this.i.h9;
        }
        set topMargin(v) {
            this.i.h9 = +v;
        }
        /**
         * Sets or gets the left margin to use around the chart content.
        *
        * The `LeftMargin` property is used to set the left margin.
        *
        * ```html
        *  <igx-data-chart
        *     [dataSource]="data"
        *     leftMargin=20
        *     rightMargin=20
        *     topMargin=30
        *     bottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        *  this.chart.leftMargin = 20;
        * ```
        */
        get leftMargin() {
            return this.i.hv;
        }
        set leftMargin(v) {
            this.i.hv = +v;
        }
        /**
         * Sets or gets the right margin to use around the chart content.
        *
        * The `RightMargin` property is used to set the left margin around the chart content in the canvas.
        *
        * ```html
        *  <igx-data-chart
        *     [dataSource]="data"
        *     leftMargin=20
        *     rightMargin=20
        *     topMargin=30
        *     bottomMargin=20>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.rightMargin=20;
        * ```
        */
        get rightMargin() {
            return this.i.h0;
        }
        set rightMargin(v) {
            this.i.h0 = +v;
        }
        /**
         * Sets or gets the bottom margin to use around the chart content.
        *
        * The `bottomMargin` property is used to gets/set the bottom margin around the chart content in the canvas.
        *
        * ```ts
        * this.chart.bottomMargin = 50;
        * ```
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     bottomMargin=50>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        */
        get bottomMargin() {
            return this.i.hi;
        }
        set bottomMargin(v) {
            this.i.hi = +v;
        }
        /**
         * Sets or gets the automatic width to add when automatically adding margins to the chart.
        *
        * The `autoMarginWidth` propertry is used find the automatic width to add when automatically adding margins to the chart.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     autoMarginWidth=50>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * this.chart.autoMarginWidth = 50;
        * ```
        */
        get autoMarginWidth() {
            return this.i.hh;
        }
        set autoMarginWidth(v) {
            this.i.hh = +v;
        }
        /**
         * Sets or gets the automatic height to add when automatically adding margins to the chart.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     autoMarginHeight=50>
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-column-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        *  this.chart.autoMarginHeight = 50;
        * ```
        */
        get autoMarginHeight() {
            return this.i.hg;
        }
        set autoMarginHeight(v) {
            this.i.hg = +v;
        }
        /**
         * Gets or sets the IsAntiAliasingEnabledDuringInteraction property.
        */
        get isAntiAliasingEnabledDuringInteraction() {
            return this.i.e0;
        }
        set isAntiAliasingEnabledDuringInteraction(v) {
            this.i.e0 = ensureBool(v);
        }
        /**
         * Gets or sets the scaling value used to affect the pixel density of the control.
         * A higher scaling ratio will produce crisper visuals at the expense of memory.  Lower values will cause the control
         * to appear blurry.
        */
        get pixelScalingRatio() {
            return this.i.hw;
        }
        set pixelScalingRatio(v) {
            this.i.hw = +v;
        }
        /**
         * Method of pixel scaling to use during end user interaction with the chart.
        */
        get interactionPixelScalingRatio() {
            return this.i.hu;
        }
        set interactionPixelScalingRatio(v) {
            this.i.hu = +v;
        }
        /**
         * Gets the resolved pixel scaling ratio.
        *
        * A higher scaling ratio will produce crisper visuals at the expense of memory. Lower values will cause the control to appear blurry
        *
        * ```ts
        * let actualPixelScalingRatio: number = this.chart.actualPixelScalingRatio;
        * ```
        */
        get actualPixelScalingRatio() {
            return this.i.hb;
        }
        set actualPixelScalingRatio(v) {
            this.i.hb = +v;
        }
        /**
         * Resolved method of pixel scaling to use during end user interaction with the chart.
        *
        * ```ts
        * let actualInteractionPixelScalingRatio: number = this.chart.actualInteractionPixelScalingRatio;
        * ```
        */
        get actualInteractionPixelScalingRatio() {
            return this.i.ha;
        }
        set actualInteractionPixelScalingRatio(v) {
            this.i.ha = +v;
        }
        /**
         * Gets or sets a whether the chart should pan its content when zooming in passed max zoom level.
        */
        get shouldPanOnMaximumZoom() {
            return this.i.f0;
        }
        set shouldPanOnMaximumZoom(v) {
            this.i.f0 = ensureBool(v);
        }
        /**
         * Gets the actual value of the window rectangle, which represents the current zoom level.
        *
        * ```ts
        * let actualWindowRect: IgxRect = this.chart.actualWindowRect;
        * ``
        * ```
        */
        get actualWindowRect() {
            return fromRect(this.i.vl);
        }
        set actualWindowRect(v) {
            this.i.vl = toRect(v);
        }
        /**
         * A number between 0 and 1 determining the position of the horizontal scroll.
         * This property is effectively a shortcut to the Left of the ActualWindowRect property.
        *
        * ```ts
        * let actualWindowPositionHorizontal: number = this.chart.actualWindowPositionHorizontal;
        * ```
        */
        get actualWindowPositionHorizontal() {
            return this.i.hc;
        }
        set actualWindowPositionHorizontal(v) {
            this.i.hc = +v;
        }
        /**
         * A number between 0 and 1 determining the position of the vertical scroll.
         * This property is effectively a shortcut to the Top of the ActualWindowRect property.
        *
        * ```ts
        * let actualWindowPositionVertical: number = this.chart.actualWindowPositionVertical;
        * ``
        * ```
        */
        get actualWindowPositionVertical() {
            return this.i.hd;
        }
        set actualWindowPositionVertical(v) {
            this.i.hd = +v;
        }
        get isDetached() {
            return this.i.isDetached;
        }
        findByName(name) {
            if (this.findEphemera) {
                if (name && name.indexOf("@@e:") == 0) {
                    return this.findEphemera(name);
                }
            }
            if (this.series != null && this.series.findByName && this.series.findByName(name)) {
                return this.series.findByName(name);
            }
            if (this.legend && this.legend.name && this.legend.name == name) {
                return this.legend;
            }
            if (this.selectedSeriesItems != null && this.selectedSeriesItems.findByName && this.selectedSeriesItems.findByName(name)) {
                return this.selectedSeriesItems.findByName(name);
            }
            if (this.focusedSeriesItems != null && this.focusedSeriesItems.findByName && this.focusedSeriesItems.findByName(name)) {
                return this.focusedSeriesItems.findByName(name);
            }
            return null;
        }
        get hasUserValues() {
            return this._hasUserValues;
        }
        __m(propertyName) {
            if (!this._inStyling) {
                this._hasUserValues.add(propertyName);
            }
        }
        _styling(container, component, parent) {
            if (this._inStyling) {
                return;
            }
            this._inStyling = true;
            this._stylingContainer = container;
            this._stylingParent = component;
            let genericPrefix = "";
            let typeName = this.i.$type.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            genericPrefix = toSpinal("SeriesViewerComponent");
            let additionalPrefixes = [];
            let prefix = toSpinal(typeName);
            additionalPrefixes.push(prefix + "-");
            let b = this.i.$type.baseType;
            while (b && b.name != "Object" &&
                b.name != "Base" &&
                b.name != "Control" &&
                b.Name != "DependencyObject" &&
                b.Name != "FrameworkElement") {
                typeName = b.name;
                if (typeName.indexOf("Xam") === 0) {
                    typeName = typeName.substring(3);
                }
                let basePrefix = toSpinal(typeName);
                additionalPrefixes.push(basePrefix + "-");
                b = b.baseType;
            }
            if (parent) {
                let parentTypeName = parent.i.$type.name;
                if (parentTypeName.indexOf("Xam") === 0) {
                    parentTypeName = parentTypeName.substring(3);
                }
                let parentPrefix = toSpinal(parentTypeName);
                additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
                additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
            }
            initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
            if (this._otherStyling) {
                this._otherStyling(container, component, parent);
            }
            this._inStyling = false;
        }
        getDesiredToolbarActions() {
            let iv = this.i.getDesiredToolbarActions();
            return (iv);
        }
        addCommandAvailabilityListener(listener) {
            this.i.addCommandAvailabilityListener(listener);
        }
        removeCommandAvailabilityListener(listener) {
            this.i.removeCommandAvailabilityListener(listener);
        }
        addCommandStateChangedListener(listener) {
            this.i.addCommandStateChangedListener(listener);
        }
        removeCommandStateChangedListener(listener) {
            this.i.removeCommandStateChangedListener(listener);
        }
        /**
         * Hides the active main tooltip, if displayed.
        
        *
        * ```ts
        * this.chart.hideToolTip();
        * ```
        */
        hideToolTip() {
            this.i.nl();
        }
        simulateHover(point) {
            this.i.q2(toPoint(point));
        }
        queueForAnimationIdle(action, version) {
            this.i.queueForAnimationIdle(action, version);
        }
        getAnimationIdleVersionNumber() {
            let iv = this.i.getAnimationIdleVersionNumber();
            return (iv);
        }
        isAnimationActive() {
            let iv = this.i.isAnimationActive();
            return (iv);
        }
        /**
         * Manually starts a tiled zoom if one isn't already running.
        
        */
        startTiledZoomingIfNecessary() {
            this.i.q4();
        }
        /**
         * Manually ends a tiled zoom if one is running.
        
        */
        endTiledZoomingIfRunning() {
            this.i.m9();
        }
        /**
         * Clears the tile zoom tile cache so that new tiles will be generated. Only applies if the viewer is using a tile based zoom.
        
        */
        clearTileZoomCache() {
            this.i.m3();
        }
        /**
         * Performs zoom in action on the chart by specified percentage of current window rect
        
        */
        zoomIn(percentage) {
            this.i.rl(percentage);
        }
        /**
         * Performs zoom out action on the chart by specified percentage of current window rect
        
        */
        zoomOut(percentage) {
            this.i.rm(percentage);
        }
        /**
         * Resets the zoom level to default.
        
        */
        resetZoom() {
            this.i.qu();
        }
        cancelManipulation() {
            this.i.my();
        }
        /**
         * Gets actual window scale for horizontal dimension of the control
        
        *
        * ```ts
        * let actualWindowScaleHorizontal: number = chart.getActualWindowScaleHorizontal();
        * ```
        */
        getActualWindowScaleHorizontal() {
            let iv = this.i.hl();
            return (iv);
        }
        /**
         * Gets actual window scale for vertical dimension of the control
        
        *
        * ```ts
        * let actualWindowScaleVertical: number = chart.getActualWindowScaleVertical();
        * ```
        */
        getActualWindowScaleVertical() {
            let iv = this.i.hm();
            return (iv);
        }
        /**
         * Notification from the containing object that the container has been resized.
        
        */
        notifyContainerResized() {
            this.i.notifyContainerResized();
        }
        /**
         * Attaches the given series to the chart.
        
        * @param s  * The series to attach to the chart.
        */
        attachSeries(s) {
            this.i.mw((s == null ? null : s.i));
        }
        notifySetItem(source_, index, oldItem, newItem) {
            this.i.n4(source_, index, oldItem, newItem);
        }
        /**
         * Used to manually notify the SeriesViewer that the data source has reset or cleared its items.  Invoking this method is only necessary if that datasource is not observable.
        
        * @param source_  * The data source which was cleared or reset.
        *
        * ```ts
        * this.chart.notifyClearItems(this.data);
        * ```
        */
        notifyClearItems(source_) {
            this.i.nv(source_);
        }
        notifyInsertItem(source_, index, newItem) {
            this.i.n2(source_, index, newItem);
        }
        notifyRemoveItem(source_, index, oldItem) {
            this.i.n3(source_, index, oldItem);
        }
        /**
         * Release any resources held by the control.
        
        */
        destroy() {
            this.i.destroy();
        }
        getCurrentActualWindowRect() {
            let iv = this.i.vr();
            return fromRect(iv);
        }
        renderToImage(width, height) {
            let iv = this.i.i4(width, height);
            return (iv);
        }
        /**
         * Use to force the SeriesViewer to finish any deferred work before printing or evaluating its visual.
         * This should only be called if the visual of the SeriesViewer needs to be synchronously saved or evaluated.
         * Calling this method too often will hinder the performance of the SeriesViewer.
        
        *
        * ```ts
        * this.chart.flush();
        * ```
        */
        flush() {
            this.i.flush();
        }
        captureImage(settings) {
            this.i.mz((settings == null ? null : settings.i));
        }
        /**
         * Occurs when the left mouse button is pressed while the mouse pointer is over the plot area.
        */
        get plotAreaMouseLeftButtonDown() {
            if (this._plotAreaMouseLeftButtonDown == null) {
                this._plotAreaMouseLeftButtonDown = new EventEmitter();
                this.i.plotAreaMouseLeftButtonDown = delegateCombine(this.i.plotAreaMouseLeftButtonDown, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxPlotAreaMouseButtonEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforePlotAreaMouseLeftButtonDown) {
                            this.beforePlotAreaMouseLeftButtonDown(this, outerArgs);
                        }
                        this._plotAreaMouseLeftButtonDown.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._plotAreaMouseLeftButtonDown;
        }
        /**
         * Occurs when the left mouse button is pressed while the mouse pointer is over the plot area.
        */
        get plotAreaMouseLeftButtonUp() {
            if (this._plotAreaMouseLeftButtonUp == null) {
                this._plotAreaMouseLeftButtonUp = new EventEmitter();
                this.i.plotAreaMouseLeftButtonUp = delegateCombine(this.i.plotAreaMouseLeftButtonUp, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxPlotAreaMouseButtonEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforePlotAreaMouseLeftButtonUp) {
                            this.beforePlotAreaMouseLeftButtonUp(this, outerArgs);
                        }
                        this._plotAreaMouseLeftButtonUp.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._plotAreaMouseLeftButtonUp;
        }
        /**
         * Occurs when the left mouse button is clicked while the mouse pointer is over the plot area.
        */
        get plotAreaClicked() {
            if (this._plotAreaClicked == null) {
                this._plotAreaClicked = new EventEmitter();
                this.i.plotAreaClicked = delegateCombine(this.i.plotAreaClicked, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxPlotAreaMouseButtonEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforePlotAreaClicked) {
                            this.beforePlotAreaClicked(this, outerArgs);
                        }
                        this._plotAreaClicked.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._plotAreaClicked;
        }
        /**
         * Occurs when the pointer enters the plot area.
        */
        get plotAreaMouseEnter() {
            if (this._plotAreaMouseEnter == null) {
                this._plotAreaMouseEnter = new EventEmitter();
                this.i.plotAreaMouseEnter = delegateCombine(this.i.plotAreaMouseEnter, (o, e) => {
                    let outerArgs = new IgxPlotAreaMouseEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforePlotAreaMouseEnter) {
                        this.beforePlotAreaMouseEnter(this, outerArgs);
                    }
                    this._plotAreaMouseEnter.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._plotAreaMouseEnter;
        }
        /**
         * Occurs when the pointer exits the plot area.
        */
        get plotAreaMouseLeave() {
            if (this._plotAreaMouseLeave == null) {
                this._plotAreaMouseLeave = new EventEmitter();
                this.i.plotAreaMouseLeave = delegateCombine(this.i.plotAreaMouseLeave, (o, e) => {
                    let outerArgs = new IgxPlotAreaMouseEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforePlotAreaMouseLeave) {
                        this.beforePlotAreaMouseLeave(this, outerArgs);
                    }
                    this._plotAreaMouseLeave.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._plotAreaMouseLeave;
        }
        /**
         * Occurs when the pointer if over the plot area.
        */
        get plotAreaMouseOver() {
            if (this._plotAreaMouseOver == null) {
                this._plotAreaMouseOver = new EventEmitter();
                this.i.plotAreaMouseOver = delegateCombine(this.i.plotAreaMouseOver, (o, e) => {
                    let outerArgs = new IgxPlotAreaMouseEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforePlotAreaMouseOver) {
                        this.beforePlotAreaMouseOver(this, outerArgs);
                    }
                    this._plotAreaMouseOver.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._plotAreaMouseOver;
        }
        /**
         * Occurs when the cursors are moved over a series in this SeriesViewer.
        */
        get seriesCursorMouseMove() {
            if (this._seriesCursorMouseMove == null) {
                this._seriesCursorMouseMove = new EventEmitter();
                this.i.seriesCursorMouseMove = delegateCombine(this.i.seriesCursorMouseMove, (o, e) => {
                    let outerArgs = new IgxChartCursorEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeSeriesCursorMouseMove) {
                        this.beforeSeriesCursorMouseMove(this, outerArgs);
                    }
                    this._seriesCursorMouseMove.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._seriesCursorMouseMove;
        }
        /**
         * Occurs when the left mouse button is pressed while the mouse pointer is over a Series.
        *
        * The `SeriesMouseLeftButtonDown` event occurs when the left mouse button is pressed while the mouse pointer is over an element of this chart.
        *
        * &lt;-- position: content member-->
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     (seriesMouseLeftButtonDown)="onSeriesMouseLeftButtonDown($event)">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         thickness=5>
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * public onSeriesMouseLeftButtonDown(event: {sender: any, args: DataChartMouseButtonEventArgs})
        * {
        *     var item = event.args.item;
        * }
        * ```
        */
        get seriesMouseLeftButtonDown() {
            if (this._seriesMouseLeftButtonDown == null) {
                this._seriesMouseLeftButtonDown = new EventEmitter();
                this.i.seriesMouseLeftButtonDown = delegateCombine(this.i.seriesMouseLeftButtonDown, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxDataChartMouseButtonEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforeSeriesMouseLeftButtonDown) {
                            this.beforeSeriesMouseLeftButtonDown(this, outerArgs);
                        }
                        this._seriesMouseLeftButtonDown.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._seriesMouseLeftButtonDown;
        }
        /**
         * Occurs when the left mouse button is released while the mouse pointer is over a Series.
        *
        * The `SeriesMouseLeftButtonUp` event occurs when the left mouse button is released while the mouse pointer is over an element of this chart.
        *
        * &lt;-- position: content member-->
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     (seriesMouseLeftButtonUp)="onSeriesMouseLeftButtonUp($event)">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         thickness=5>
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        * public onSeriesMouseLeftButtonUp(event: {sender: any, args: DataChartMouseButtonEventArgs})
        * {
        *     var item = event.args.item;
        * }
        * ```
        *
        * ```ts
        * public onSeriesMouseLeftButtonUp =(event: any, args: DataChartMouseButtonEventArgs )
        * {
        *
        * }
        * ```
        */
        get seriesMouseLeftButtonUp() {
            if (this._seriesMouseLeftButtonUp == null) {
                this._seriesMouseLeftButtonUp = new EventEmitter();
                this.i.seriesMouseLeftButtonUp = delegateCombine(this.i.seriesMouseLeftButtonUp, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxDataChartMouseButtonEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforeSeriesMouseLeftButtonUp) {
                            this.beforeSeriesMouseLeftButtonUp(this, outerArgs);
                        }
                        this._seriesMouseLeftButtonUp.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._seriesMouseLeftButtonUp;
        }
        /**
         * Occurs when the left mouse button is released while the mouse pointer is over a Series.
        */
        get seriesClick() {
            if (this._seriesClick == null) {
                this._seriesClick = new EventEmitter();
                this.i.seriesClick = delegateCombine(this.i.seriesClick, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxDataChartMouseButtonEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforeSeriesClick) {
                            this.beforeSeriesClick(this, outerArgs);
                        }
                        this._seriesClick.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._seriesClick;
        }
        /**
         * Occurs when the mouse pointer moves while over a Series.
        *
        * The `SeriesMouseMove` event occurs when the left mouse pointer moves while over an element of this chart.
        *
        * &lt;-- position: content member-->
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     (seriesMouseMove)="onSeriesMouseMove($event)">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         thickness=5>
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        * public onSeriesMouseMove(event: {sender: any, args: ChartMouseEventArgs})
        * {
        *     var item = event.args.item;
        * }
        * ```
        *
        * ```ts
        * public onSeriesMouseMove =( event: any, args: ChartMouseEventArgs) => {
        *                  }
        * ```
        */
        get seriesMouseMove() {
            if (this._seriesMouseMove == null) {
                this._seriesMouseMove = new EventEmitter();
                this.i.seriesMouseMove = delegateCombine(this.i.seriesMouseMove, (o, e) => {
                    let outerArgs = new IgxChartMouseEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeSeriesMouseMove) {
                        this.beforeSeriesMouseMove(this, outerArgs);
                    }
                    this._seriesMouseMove.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._seriesMouseMove;
        }
        /**
         * Occurs when the mouse pointer enters a Series.
        *
        * The `SeriesMouseEnter` event occurs when the left mouse pointer enters an element of this chart.
        *
        * &lt;-- position: content member-->
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     (seriesMouseEnter)="onSeriesMouseEnter($event)">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         thickness=5>
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * public onSeriesMouseEnter(event: {sender: any, args: ChartMouseEventArgs})
        * {
        *     var item = event.args.item;
        * }
        * ```
        */
        get seriesMouseEnter() {
            if (this._seriesMouseEnter == null) {
                this._seriesMouseEnter = new EventEmitter();
                this.i.seriesMouseEnter = delegateCombine(this.i.seriesMouseEnter, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxChartMouseEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforeSeriesMouseEnter) {
                            this.beforeSeriesMouseEnter(this, outerArgs);
                        }
                        this._seriesMouseEnter.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._seriesMouseEnter;
        }
        /**
         * Occurs when the mouse pointer leaves a Series.
        *
        * The `SeriesMouseLeave` event occurs when the left mouse pointer leaves an element of this chart.
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     (seriesMouseLeave)="onSeriesMouseLeave($event)">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value"
        *         thickness=5>
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * public onSeriesMouseLeave(event: {sender: any, args: ChartMouseEventArgs})
        * {
        *     var item = event.args.item;
        * }
        * ```
        */
        get seriesMouseLeave() {
            if (this._seriesMouseLeave == null) {
                this._seriesMouseLeave = new EventEmitter();
                this.i.seriesMouseLeave = delegateCombine(this.i.seriesMouseLeave, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxChartMouseEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforeSeriesMouseLeave) {
                            this.beforeSeriesMouseLeave(this, outerArgs);
                        }
                        this._seriesMouseLeave.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._seriesMouseLeave;
        }
        /**
         * Raised when an ongoing manipulation of the series viewer is started.
        */
        get viewerManipulationStarting() {
            if (this._viewerManipulationStarting == null) {
                this._viewerManipulationStarting = new EventEmitter();
                this.i.viewerManipulationStarting = delegateCombine(this.i.viewerManipulationStarting, (o, e) => {
                    let outerArgs = new IgxSeriesViewerManipulationEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeViewerManipulationStarting) {
                        this.beforeViewerManipulationStarting(this, outerArgs);
                    }
                    this._viewerManipulationStarting.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._viewerManipulationStarting;
        }
        /**
         * Raised when an ongoing manipulation of the series viewer is ended.
        */
        get viewerManipulationEnding() {
            if (this._viewerManipulationEnding == null) {
                this._viewerManipulationEnding = new EventEmitter();
                this.i.viewerManipulationEnding = delegateCombine(this.i.viewerManipulationEnding, (o, e) => {
                    let outerArgs = new IgxSeriesViewerManipulationEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeViewerManipulationEnding) {
                        this.beforeViewerManipulationEnding(this, outerArgs);
                    }
                    this._viewerManipulationEnding.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._viewerManipulationEnding;
        }
        /**
         * Occurs just after the current SeriesViewer's window rectangle is changed.
        *
        * &lt;-- position: content member-->
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     isHorizontalZoomEnabled=true
        *     (windowRectChanged)="onWindowRectChanged($event)">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        * public onWindowRectChanged(event: {sender: any, args: RectChangedEventArgs})
        * {
        *     let newRect: IgxRect = event.args.newRect;
        * }
        * ```
        *
        * ```ts
        * public onWindowRectChanged =( s:any, e: RectChangedEventArgs) => {
        *
        *        }
        * ```
        */
        get windowRectChanged() {
            if (this._windowRectChanged == null) {
                this._windowRectChanged = new EventEmitter();
                this.i.windowRectChanged = delegateCombine(this.i.windowRectChanged, (o, e) => {
                    let outerArgs = new IgxRectChangedEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeWindowRectChanged) {
                        this.beforeWindowRectChanged(this, outerArgs);
                    }
                    this._windowRectChanged.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._windowRectChanged;
        }
        /**
         * Occurs after the size of the series viewer changes.
        *
        * &lt;-- position: content member-->
        *
        * ```html
        * <igx-data-chart
        *     [dataSource]="data"
        *     (sizeChanged)="onSizeChanged($event)">
        *     <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *         #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-line-series
        *         [xAxis]="xAxis"
        *         [yAxis]="yAxis"
        *         valueMemberPath="value">
        *     </igx-line-series>
        * </igx-data-chart>
        * ```
        *
        * &lt;-- position: content member-->
        *
        * ```ts
        * public onSizeChanged(event: {sender: any, args: RectChangedEventArgs})
        * {
        *     let newRect: IgxRect = event.args.newRect;
        * }
        * ```
        *
        * ```ts
        * public onSizeChanged =(event: any, args: RectChangedEventArgs}){
        *
        * }
        * ```
        */
        get sizeChanged() {
            if (this._sizeChanged == null) {
                this._sizeChanged = new EventEmitter();
                this.i.sizeChanged = delegateCombine(this.i.sizeChanged, (o, e) => {
                    let outerArgs = new IgxRectChangedEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeSizeChanged) {
                        this.beforeSizeChanged(this, outerArgs);
                    }
                    this._sizeChanged.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._sizeChanged;
        }
        /**
         * Raised when the actual window rectangle of the SeriesViewer changes.
        */
        get actualWindowRectChanged() {
            if (this._actualWindowRectChanged == null) {
                this._actualWindowRectChanged = new EventEmitter();
                this.i.actualWindowRectChanged = delegateCombine(this.i.actualWindowRectChanged, (o, e) => {
                    let outerArgs = new IgxRectChangedEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeActualWindowRectChanged) {
                        this.beforeActualWindowRectChanged(this, outerArgs);
                    }
                    this._actualWindowRectChanged.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._actualWindowRectChanged;
        }
        /**
         * Occurs just after the current SeriesViewer's grid area rectangle is changed.
         * The grid area may change as the result of the SeriesViewer being resized, or
         * of an axis being added or changing size, possibly in another SeriesViewer.
        */
        get gridAreaRectChanged() {
            if (this._gridAreaRectChanged == null) {
                this._gridAreaRectChanged = new EventEmitter();
                this.i.gridAreaRectChanged = delegateCombine(this.i.gridAreaRectChanged, (o, e) => {
                    let outerArgs = new IgxRectChangedEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeGridAreaRectChanged) {
                        this.beforeGridAreaRectChanged(this, outerArgs);
                    }
                    this._gridAreaRectChanged.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._gridAreaRectChanged;
        }
        /**
         * Raised when the SelectedItems collection has changed.
        */
        get selectedSeriesItemsChanged() {
            if (this._selectedSeriesItemsChanged == null) {
                this._selectedSeriesItemsChanged = new EventEmitter();
                this.i.selectedSeriesItemsChanged = delegateCombine(this.i.selectedSeriesItemsChanged, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxSeriesViewerSelectedSeriesItemsChangedEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforeSelectedSeriesItemsChanged) {
                            this.beforeSelectedSeriesItemsChanged(this, outerArgs);
                        }
                        this._selectedSeriesItemsChanged.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._selectedSeriesItemsChanged;
        }
        /**
         * Raised when the SelectedItems collection has changed.
        */
        get focusedSeriesItemsChanged() {
            if (this._focusedSeriesItemsChanged == null) {
                this._focusedSeriesItemsChanged = new EventEmitter();
                this.i.focusedSeriesItemsChanged = delegateCombine(this.i.focusedSeriesItemsChanged, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxSeriesViewerSelectedSeriesItemsChangedEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforeFocusedSeriesItemsChanged) {
                            this.beforeFocusedSeriesItemsChanged(this, outerArgs);
                        }
                        this._focusedSeriesItemsChanged.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._focusedSeriesItemsChanged;
        }
        /**
         * Raised when the SeriesViewer's processing for an update has completed.
        */
        get refreshCompleted() {
            if (this._refreshCompleted == null) {
                this._refreshCompleted = new EventEmitter();
                this.i.refreshCompleted = delegateCombine(this.i.refreshCompleted, (o, e) => {
                    let outerArgs = new IgxRefreshCompletedEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeRefreshCompleted) {
                        this.beforeRefreshCompleted(this, outerArgs);
                    }
                    this._refreshCompleted.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._refreshCompleted;
        }
        /**
         * Event raised when an image of the component is saved
        */
        get imageCaptured() {
            if (this._imageCaptured == null) {
                this._imageCaptured = new EventEmitter();
                this.i.imageCaptured = delegateCombine(this.i.imageCaptured, (o, e) => {
                    let outerArgs = new IgxImageCapturedEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeImageCaptured) {
                        this.beforeImageCaptured(this, outerArgs);
                    }
                    this._imageCaptured.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._imageCaptured;
        }
        _runInZone(act) {
            if (this._zoneRunner != null) {
                this._zoneRunner(act);
            }
            else {
                act();
            }
        }
    }
    IgxSeriesViewerComponent.fac = i0.ngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSeriesViewerComponent, deps: [], target: i0.FactoryTarget.Component });
    IgxSeriesViewerComponent.cmp = i0.ngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxSeriesViewerComponent, selector: "ng-component", inputs: { brushes: "brushes", outlines: "outlines", markerBrushes: "markerBrushes", markerOutlines: "markerOutlines", markerAutomaticBehavior: "markerAutomaticBehavior", crosshairPoint: "crosshairPoint", animateSeriesWhenAxisRangeChanges: "animateSeriesWhenAxisRangeChanges", isWindowSyncedToVisibleRange: "isWindowSyncedToVisibleRange", shouldSimulateHoverMoveCrosshairPoint: "shouldSimulateHoverMoveCrosshairPoint", highlightingMode: "highlightingMode", selectionMode: "selectionMode", selectionBehavior: "selectionBehavior", focusMode: "focusMode", horizontalViewScrollbarMode: "horizontalViewScrollbarMode", verticalViewScrollbarMode: "verticalViewScrollbarMode", horizontalViewScrollbarPosition: "horizontalViewScrollbarPosition", verticalViewScrollbarPosition: "verticalViewScrollbarPosition", horizontalViewScrollbarFill: "horizontalViewScrollbarFill", horizontalViewScrollbarOutline: "horizontalViewScrollbarOutline", horizontalViewScrollbarStrokeThickness: "horizontalViewScrollbarStrokeThickness", horizontalViewScrollbarMaxOpacity: "horizontalViewScrollbarMaxOpacity", horizontalViewScrollbarCornerRadius: "horizontalViewScrollbarCornerRadius", horizontalViewScrollbarHeight: "horizontalViewScrollbarHeight", horizontalViewScrollbarInset: "horizontalViewScrollbarInset", horizontalViewScrollbarTrackStartInset: "horizontalViewScrollbarTrackStartInset", horizontalViewScrollbarTrackEndInset: "horizontalViewScrollbarTrackEndInset", horizontalViewScrollbarShouldAddAutoTrackInsets: "horizontalViewScrollbarShouldAddAutoTrackInsets", verticalViewScrollbarFill: "verticalViewScrollbarFill", verticalViewScrollbarOutline: "verticalViewScrollbarOutline", verticalViewScrollbarStrokeThickness: "verticalViewScrollbarStrokeThickness", verticalViewScrollbarMaxOpacity: "verticalViewScrollbarMaxOpacity", verticalViewScrollbarCornerRadius: "verticalViewScrollbarCornerRadius", verticalViewScrollbarWidth: "verticalViewScrollbarWidth", verticalViewScrollbarInset: "verticalViewScrollbarInset", verticalViewScrollbarTrackStartInset: "verticalViewScrollbarTrackStartInset", verticalViewScrollbarTrackEndInset: "verticalViewScrollbarTrackEndInset", verticalViewScrollbarShouldAddAutoTrackInsets: "verticalViewScrollbarShouldAddAutoTrackInsets", highlightingBehavior: "highlightingBehavior", legendHighlightingMode: "legendHighlightingMode", legendItemBadgeShape: "legendItemBadgeShape", legendItemBadgeMode: "legendItemBadgeMode", effectiveViewport: "effectiveViewport", windowRect: "windowRect", useTiledZooming: "useTiledZooming", preferHigherResolutionTiles: "preferHigherResolutionTiles", zoomTileCacheSize: "zoomTileCacheSize", highlightingTransitionDuration: "highlightingTransitionDuration", selectionTransitionDuration: "selectionTransitionDuration", focusTransitionDuration: "focusTransitionDuration", scrollbarsAnimationDuration: "scrollbarsAnimationDuration", isPagePanningAllowed: "isPagePanningAllowed", contentHitTestMode: "contentHitTestMode", actualContentHitTestMode: "actualContentHitTestMode", isSurfaceInteractionDisabled: "isSurfaceInteractionDisabled", windowResponse: "windowResponse", windowRectMinWidth: "windowRectMinWidth", windowRectMinHeight: "windowRectMinHeight", actualWindowRectMinWidth: "actualWindowRectMinWidth", actualWindowRectMinHeight: "actualWindowRectMinHeight", windowSizeMinWidth: "windowSizeMinWidth", windowSizeMinHeight: "windowSizeMinHeight", crosshairVisibility: "crosshairVisibility", horizontalCrosshairBrush: "horizontalCrosshairBrush", verticalCrosshairBrush: "verticalCrosshairBrush", zoomCoercionMode: "zoomCoercionMode", plotAreaBackground: "plotAreaBackground", defaultInteraction: "defaultInteraction", rightButtonDefaultInteraction: "rightButtonDefaultInteraction", dragModifier: "dragModifier", panModifier: "panModifier", previewRect: "previewRect", highlightedValuesDisplayMode: "highlightedValuesDisplayMode", selectedSeriesItems: "selectedSeriesItems", focusedSeriesItems: "focusedSeriesItems", windowPositionHorizontal: "windowPositionHorizontal", windowPositionVertical: "windowPositionVertical", selectionBrush: "selectionBrush", focusBrush: "focusBrush", chartTitle: "chartTitle", titleHorizontalAlignment: "titleHorizontalAlignment", subtitleHorizontalAlignment: "subtitleHorizontalAlignment", titleTextStyle: "titleTextStyle", subtitleTextStyle: "subtitleTextStyle", titleTextColor: "titleTextColor", subtitleTextColor: "subtitleTextColor", titleTopMargin: "titleTopMargin", titleLeftMargin: "titleLeftMargin", titleRightMargin: "titleRightMargin", titleBottomMargin: "titleBottomMargin", subtitleTopMargin: "subtitleTopMargin", subtitleLeftMargin: "subtitleLeftMargin", subtitleRightMargin: "subtitleRightMargin", subtitleBottomMargin: "subtitleBottomMargin", subtitle: "subtitle", topMargin: "topMargin", leftMargin: "leftMargin", rightMargin: "rightMargin", bottomMargin: "bottomMargin", autoMarginWidth: "autoMarginWidth", autoMarginHeight: "autoMarginHeight", isAntiAliasingEnabledDuringInteraction: "isAntiAliasingEnabledDuringInteraction", pixelScalingRatio: "pixelScalingRatio", interactionPixelScalingRatio: "interactionPixelScalingRatio", actualPixelScalingRatio: "actualPixelScalingRatio", actualInteractionPixelScalingRatio: "actualInteractionPixelScalingRatio", shouldPanOnMaximumZoom: "shouldPanOnMaximumZoom", actualWindowRect: "actualWindowRect", actualWindowPositionHorizontal: "actualWindowPositionHorizontal", actualWindowPositionVertical: "actualWindowPositionVertical" }, outputs: { plotAreaMouseLeftButtonDown: "plotAreaMouseLeftButtonDown", plotAreaMouseLeftButtonUp: "plotAreaMouseLeftButtonUp", plotAreaClicked: "plotAreaClicked", plotAreaMouseEnter: "plotAreaMouseEnter", plotAreaMouseLeave: "plotAreaMouseLeave", plotAreaMouseOver: "plotAreaMouseOver", seriesCursorMouseMove: "seriesCursorMouseMove", seriesMouseLeftButtonDown: "seriesMouseLeftButtonDown", seriesMouseLeftButtonUp: "seriesMouseLeftButtonUp", seriesClick: "seriesClick", seriesMouseMove: "seriesMouseMove", seriesMouseEnter: "seriesMouseEnter", seriesMouseLeave: "seriesMouseLeave", viewerManipulationStarting: "viewerManipulationStarting", viewerManipulationEnding: "viewerManipulationEnding", windowRectChanged: "windowRectChanged", sizeChanged: "sizeChanged", actualWindowRectChanged: "actualWindowRectChanged", gridAreaRectChanged: "gridAreaRectChanged", selectedSeriesItemsChanged: "selectedSeriesItemsChanged", focusedSeriesItemsChanged: "focusedSeriesItemsChanged", refreshCompleted: "refreshCompleted", imageCaptured: "imageCaptured" }, ngImport: i0, template: ``, isInline: true });
    return IgxSeriesViewerComponent;
})();
i0.ngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSeriesViewerComponent, decorators: [{
            type: Component,
            args: [{
                    template: ``,
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { brushes: [{
                type: Input
            }], outlines: [{
                type: Input
            }], markerBrushes: [{
                type: Input
            }], markerOutlines: [{
                type: Input
            }], markerAutomaticBehavior: [{
                type: Input
            }], crosshairPoint: [{
                type: Input
            }], animateSeriesWhenAxisRangeChanges: [{
                type: Input
            }], isWindowSyncedToVisibleRange: [{
                type: Input
            }], shouldSimulateHoverMoveCrosshairPoint: [{
                type: Input
            }], highlightingMode: [{
                type: Input
            }], selectionMode: [{
                type: Input
            }], selectionBehavior: [{
                type: Input
            }], focusMode: [{
                type: Input
            }], horizontalViewScrollbarMode: [{
                type: Input
            }], verticalViewScrollbarMode: [{
                type: Input
            }], horizontalViewScrollbarPosition: [{
                type: Input
            }], verticalViewScrollbarPosition: [{
                type: Input
            }], horizontalViewScrollbarFill: [{
                type: Input
            }], horizontalViewScrollbarOutline: [{
                type: Input
            }], horizontalViewScrollbarStrokeThickness: [{
                type: Input
            }], horizontalViewScrollbarMaxOpacity: [{
                type: Input
            }], horizontalViewScrollbarCornerRadius: [{
                type: Input
            }], horizontalViewScrollbarHeight: [{
                type: Input
            }], horizontalViewScrollbarInset: [{
                type: Input
            }], horizontalViewScrollbarTrackStartInset: [{
                type: Input
            }], horizontalViewScrollbarTrackEndInset: [{
                type: Input
            }], horizontalViewScrollbarShouldAddAutoTrackInsets: [{
                type: Input
            }], verticalViewScrollbarFill: [{
                type: Input
            }], verticalViewScrollbarOutline: [{
                type: Input
            }], verticalViewScrollbarStrokeThickness: [{
                type: Input
            }], verticalViewScrollbarMaxOpacity: [{
                type: Input
            }], verticalViewScrollbarCornerRadius: [{
                type: Input
            }], verticalViewScrollbarWidth: [{
                type: Input
            }], verticalViewScrollbarInset: [{
                type: Input
            }], verticalViewScrollbarTrackStartInset: [{
                type: Input
            }], verticalViewScrollbarTrackEndInset: [{
                type: Input
            }], verticalViewScrollbarShouldAddAutoTrackInsets: [{
                type: Input
            }], highlightingBehavior: [{
                type: Input
            }], legendHighlightingMode: [{
                type: Input
            }], legendItemBadgeShape: [{
                type: Input
            }], legendItemBadgeMode: [{
                type: Input
            }], effectiveViewport: [{
                type: Input
            }], windowRect: [{
                type: Input
            }], useTiledZooming: [{
                type: Input
            }], preferHigherResolutionTiles: [{
                type: Input
            }], zoomTileCacheSize: [{
                type: Input
            }], highlightingTransitionDuration: [{
                type: Input
            }], selectionTransitionDuration: [{
                type: Input
            }], focusTransitionDuration: [{
                type: Input
            }], scrollbarsAnimationDuration: [{
                type: Input
            }], isPagePanningAllowed: [{
                type: Input
            }], contentHitTestMode: [{
                type: Input
            }], actualContentHitTestMode: [{
                type: Input
            }], isSurfaceInteractionDisabled: [{
                type: Input
            }], windowResponse: [{
                type: Input
            }], windowRectMinWidth: [{
                type: Input
            }], windowRectMinHeight: [{
                type: Input
            }], actualWindowRectMinWidth: [{
                type: Input
            }], actualWindowRectMinHeight: [{
                type: Input
            }], windowSizeMinWidth: [{
                type: Input
            }], windowSizeMinHeight: [{
                type: Input
            }], crosshairVisibility: [{
                type: Input
            }], horizontalCrosshairBrush: [{
                type: Input
            }], verticalCrosshairBrush: [{
                type: Input
            }], zoomCoercionMode: [{
                type: Input
            }], plotAreaBackground: [{
                type: Input
            }], defaultInteraction: [{
                type: Input
            }], rightButtonDefaultInteraction: [{
                type: Input
            }], dragModifier: [{
                type: Input
            }], panModifier: [{
                type: Input
            }], previewRect: [{
                type: Input
            }], highlightedValuesDisplayMode: [{
                type: Input
            }], selectedSeriesItems: [{
                type: Input
            }], focusedSeriesItems: [{
                type: Input
            }], windowPositionHorizontal: [{
                type: Input
            }], windowPositionVertical: [{
                type: Input
            }], selectionBrush: [{
                type: Input
            }], focusBrush: [{
                type: Input
            }], chartTitle: [{
                type: Input
            }], titleHorizontalAlignment: [{
                type: Input
            }], subtitleHorizontalAlignment: [{
                type: Input
            }], titleTextStyle: [{
                type: Input
            }], subtitleTextStyle: [{
                type: Input
            }], titleTextColor: [{
                type: Input
            }], subtitleTextColor: [{
                type: Input
            }], titleTopMargin: [{
                type: Input
            }], titleLeftMargin: [{
                type: Input
            }], titleRightMargin: [{
                type: Input
            }], titleBottomMargin: [{
                type: Input
            }], subtitleTopMargin: [{
                type: Input
            }], subtitleLeftMargin: [{
                type: Input
            }], subtitleRightMargin: [{
                type: Input
            }], subtitleBottomMargin: [{
                type: Input
            }], subtitle: [{
                type: Input
            }], topMargin: [{
                type: Input
            }], leftMargin: [{
                type: Input
            }], rightMargin: [{
                type: Input
            }], bottomMargin: [{
                type: Input
            }], autoMarginWidth: [{
                type: Input
            }], autoMarginHeight: [{
                type: Input
            }], isAntiAliasingEnabledDuringInteraction: [{
                type: Input
            }], pixelScalingRatio: [{
                type: Input
            }], interactionPixelScalingRatio: [{
                type: Input
            }], actualPixelScalingRatio: [{
                type: Input
            }], actualInteractionPixelScalingRatio: [{
                type: Input
            }], shouldPanOnMaximumZoom: [{
                type: Input
            }], actualWindowRect: [{
                type: Input
            }], actualWindowPositionHorizontal: [{
                type: Input
            }], actualWindowPositionVertical: [{
                type: Input
            }], plotAreaMouseLeftButtonDown: [{
                type: Output
            }], plotAreaMouseLeftButtonUp: [{
                type: Output
            }], plotAreaClicked: [{
                type: Output
            }], plotAreaMouseEnter: [{
                type: Output
            }], plotAreaMouseLeave: [{
                type: Output
            }], plotAreaMouseOver: [{
                type: Output
            }], seriesCursorMouseMove: [{
                type: Output
            }], seriesMouseLeftButtonDown: [{
                type: Output
            }], seriesMouseLeftButtonUp: [{
                type: Output
            }], seriesClick: [{
                type: Output
            }], seriesMouseMove: [{
                type: Output
            }], seriesMouseEnter: [{
                type: Output
            }], seriesMouseLeave: [{
                type: Output
            }], viewerManipulationStarting: [{
                type: Output
            }], viewerManipulationEnding: [{
                type: Output
            }], windowRectChanged: [{
                type: Output
            }], sizeChanged: [{
                type: Output
            }], actualWindowRectChanged: [{
                type: Output
            }], gridAreaRectChanged: [{
                type: Output
            }], selectedSeriesItemsChanged: [{
                type: Output
            }], focusedSeriesItemsChanged: [{
                type: Output
            }], refreshCompleted: [{
                type: Output
            }], imageCaptured: [{
                type: Output
            }] } });
