// igniteui-angular-charts/esm2015/lib/igx-bar-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxBarSeriesModule } from './igx-bar-series-module';
import { IgxBarSeriesComponent } from './igx-bar-series-component';
import { BarSeries } from './BarSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxBarSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxBarSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxBarSeriesComponent", IgxBarSeriesComponent);
            TypeRegistrar.register("BarSeries", BarSeries.$type);
        }
    }
    IgxBarSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxBarSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarSeriesDynamicModule, imports: [CommonModule, IgxBarSeriesModule], exports: [IgxBarSeriesModule] });
    IgxBarSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarSeriesDynamicModule, imports: [[CommonModule, IgxBarSeriesModule], IgxBarSeriesModule] });
    return IgxBarSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxBarSeriesModule],
                    exports: [IgxBarSeriesModule],
                    entryComponents: [IgxBarSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
