// igniteui-angular-charts/esm2015/lib/igx-range-category-series-component.js
import { Component, Input } from '@angular/core';
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import * as i0 from "@angular/core";
/**
 * Represents the base class for all IgxDataChartComponent ranged category/value series.
*
* The `RangeCategorySeries` represents the base class for all IgxDataChartComponent ranged category/value series.
*
* ```html
* <igx-data-chart  #chart
*     [dataSource]="data">
*     <igx-category-x-axis
*     label="label"
*     #xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
*     minimumValue="0"
*     #yAxis>
*     </igx-numeric-y-axis>
*     <igx-range-column-series
*     [xAxis]="xAxis"
*     [yAxis]="yAxis"
*     highMemberPath="high"
*     lowMemberPath="low">
* </igx-range-column-series>
* </igx-data-chart>
* ```
*
* ```ts
* let series = new IgxRangeColumnSeriesComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* this.chart.series.add(series);
* ```
*/
export let IgxRangeCategorySeriesComponent = /*@__PURE__*/ (() => {
    class IgxRangeCategorySeriesComponent extends IgxCategorySeriesComponent {
        constructor() {
            super();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets or sets the value mapping property for the current series object.
        *
        * The `LowMemberPath` is used for low mapping property of the current series object.
        *
        * ```html
        * <igx-data-chart  #chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *     label="label"
        *     #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *     minimumValue="0"
        *     #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-range-column-series
        *     [xAxis]="xAxis"
        *     [yAxis]="yAxis"
        *     highMemberPath="high"
        *     lowMemberPath="low">
        * </igx-range-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * let series = new IgxRangeColumnSeriesComponent();
        * series.xAxis = this.xAxis;
        * series.yAxis = this.yAxis;
        * series.highMemberPath = "high";
        * series.lowMemberPath = "low";
        * this.chart.series.add(series);
        * ```
        */
        get lowMemberPath() {
            return this.i.aaq;
        }
        set lowMemberPath(v) {
            this.i.aaq = v;
        }
        /**
         * Gets or sets the value mapping property for the current series object.
        *
        * Use `HighMemberPath` propert for high mapping of the current series object.
        *
        * ```html
        * <igx-data-chart  #chart
        *     [dataSource]="data">
        *     <igx-category-x-axis
        *     label="label"
        *     #xAxis>
        *     </igx-category-x-axis>
        *     <igx-numeric-y-axis
        *     minimumValue="0"
        *     #yAxis>
        *     </igx-numeric-y-axis>
        *     <igx-range-column-series
        *     [xAxis]="xAxis"
        *     [yAxis]="yAxis"
        *     highMemberPath="high"
        *     lowMemberPath="low">
        * </igx-range-column-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * series.highMemberPath = "high";
        * ```
        */
        get highMemberPath() {
            return this.i.aan;
        }
        set highMemberPath(v) {
            this.i.aan = v;
        }
        /**
         * Gets or sets the highlighted low value mapping property for the current series object.
        */
        get highlightedLowMemberPath() {
            return this.i.aal;
        }
        set highlightedLowMemberPath(v) {
            this.i.aal = v;
        }
        /**
         * Gets or sets the highlighted High value mapping property for the current series object.
        */
        get highlightedHighMemberPath() {
            return this.i.aaj;
        }
        set highlightedHighMemberPath(v) {
            this.i.aaj = v;
        }
        getItemValue(item, memberPathName) {
            let iv = this.i.ko(item, memberPathName);
            return (iv);
        }
        /**
         * Gets the value of a requested member path from the series.
        
        * @param memberPathName  * The property name of a valid member path for the series
        */
        getMemberPathValue(memberPathName) {
            let iv = this.i.mb(memberPathName);
            return (iv);
        }
        /**
         * Scrolls the series to display the item for the specified data item.
         * The series is scrolled by the minimum amount required to place the specified data item within
         * the central 80% of the visible axis.
        
        * @param item  * The data item (item) to scroll to.
        *
        * `ScrollIntoView` method Scrolls the series to display the item for the specified data item.
        */
        scrollIntoView(item) {
            let iv = this.i.gb(item);
            return (iv);
        }
    }
    IgxRangeCategorySeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeCategorySeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxRangeCategorySeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxRangeCategorySeriesComponent, selector: "ng-component", inputs: { lowMemberPath: "lowMemberPath", highMemberPath: "highMemberPath", highlightedLowMemberPath: "highlightedLowMemberPath", highlightedHighMemberPath: "highlightedHighMemberPath" }, usesInheritance: true, ngImport: i0, template: ``, isInline: true });
    return IgxRangeCategorySeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRangeCategorySeriesComponent, decorators: [{
            type: Component,
            args: [{
                    template: ``,
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { lowMemberPath: [{
                type: Input
            }], highMemberPath: [{
                type: Input
            }], highlightedLowMemberPath: [{
                type: Input
            }], highlightedHighMemberPath: [{
                type: Input
            }] } });
