// igniteui-angular-charts/esm2015/lib/igx-data-chart-visual-data-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { VisualDataFactoryManager } from './VisualDataFactoryManager';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDataChartVisualDataModule } from './igx-data-chart-visual-data-module';
import * as i0 from "@angular/core";
export let IgxDataChartVisualDataDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartVisualDataDynamicModule {
        constructor() {
            TypeRegistrar.register('VisualDataFactoryManager', VisualDataFactoryManager.$type);
            VisualDataFactoryManager.register();
        }
    }
    IgxDataChartVisualDataDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVisualDataDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartVisualDataDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVisualDataDynamicModule, imports: [CommonModule, IgxDataChartVisualDataModule], exports: [IgxDataChartVisualDataModule] });
    IgxDataChartVisualDataDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVisualDataDynamicModule, imports: [[CommonModule, IgxDataChartVisualDataModule
            ], IgxDataChartVisualDataModule] });
    return IgxDataChartVisualDataDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVisualDataDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataChartVisualDataModule
                    ],
                    imports: [CommonModule, IgxDataChartVisualDataModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
