// igniteui-angular-charts/esm2015/lib/igx-polar-spline-area-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPolarSplineAreaSeriesComponent } from './igx-polar-spline-area-series-component';
import * as i0 from "@angular/core";
export let IgxPolarSplineAreaSeriesModule = /*@__PURE__*/ (() => {
    class IgxPolarSplineAreaSeriesModule {
    }
    IgxPolarSplineAreaSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineAreaSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPolarSplineAreaSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineAreaSeriesModule, declarations: [IgxPolarSplineAreaSeriesComponent], imports: [CommonModule], exports: [IgxPolarSplineAreaSeriesComponent] });
    IgxPolarSplineAreaSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineAreaSeriesModule, imports: [[CommonModule]] });
    return IgxPolarSplineAreaSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineAreaSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxPolarSplineAreaSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxPolarSplineAreaSeriesComponent],
                    entryComponents: []
                }]
        }] });
