// igniteui-angular-charts/esm2015/lib/igx-on-balance-volume-indicator-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import { IgxFinancialIndicatorComponent } from "./igx-financial-indicator-component";
import { IgxFinancialSeriesComponent } from "./igx-financial-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { OnBalanceVolumeIndicator } from "./OnBalanceVolumeIndicator";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent On Balance Bolume indicator series.
 * Default required members: Close, Volume
*
* You can use the `OnBalanceVolumeIndicator` to calculate a running total of sales volume for a stock.
*
* ```html
* <igx-data-chart
* 	#chart
*     [dataSource]="data" >
*     <igx-category-x-axis
* 		label="label"
* 		#xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
* 		#yAxis>
*     </igx-numeric-y-axis>
*     <igx-on-balance-volume-indicator
* 		[xAxis]="xAxis"
* 		[yAxis]="yAxis"
* 		openMemberPath="open"
* 		highMemberPath="high"
* 		lowMemberPath="low"
* 		closeMemberPath="close">
*     </igx-on-balance-volume-indicator>
* </igx-data-chart>
* ```
*
* ```ts
* let series = new IgxOnBalanceVolumeIndicatorComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.openMemberPath = "open";
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* series.closeMemberPath = "close";
* this.chart.series.add(series);
* ```
*/
export let IgxOnBalanceVolumeIndicatorComponent = /*@__PURE__*/ (() => {
    class IgxOnBalanceVolumeIndicatorComponent extends IgxStrategyBasedIndicatorComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new OnBalanceVolumeIndicator();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
    }
    IgxOnBalanceVolumeIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOnBalanceVolumeIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxOnBalanceVolumeIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxOnBalanceVolumeIndicatorComponent, selector: "igx-on-balance-volume-indicator", providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(() => IgxOnBalanceVolumeIndicatorComponent) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(() => IgxOnBalanceVolumeIndicatorComponent) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(() => IgxOnBalanceVolumeIndicatorComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxOnBalanceVolumeIndicatorComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxOnBalanceVolumeIndicatorComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOnBalanceVolumeIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-on-balance-volume-indicator',
                    template: ``,
                    providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(() => IgxOnBalanceVolumeIndicatorComponent) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(() => IgxOnBalanceVolumeIndicatorComponent) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(() => IgxOnBalanceVolumeIndicatorComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxOnBalanceVolumeIndicatorComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
