// igniteui-angular-charts/esm2015/lib/igx-column-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxColumnSeriesModule } from './igx-column-series-module';
import { IgxColumnSeriesComponent } from './igx-column-series-component';
import { ColumnSeries } from './ColumnSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxColumnSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxColumnSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxColumnSeriesComponent", IgxColumnSeriesComponent);
            TypeRegistrar.register("ColumnSeries", ColumnSeries.$type);
        }
    }
    IgxColumnSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxColumnSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnSeriesDynamicModule, imports: [CommonModule, IgxColumnSeriesModule], exports: [IgxColumnSeriesModule] });
    IgxColumnSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnSeriesDynamicModule, imports: [[CommonModule, IgxColumnSeriesModule], IgxColumnSeriesModule] });
    return IgxColumnSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxColumnSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxColumnSeriesModule],
                    exports: [IgxColumnSeriesModule],
                    entryComponents: [IgxColumnSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
