// igniteui-angular-charts/esm2015/lib/igx-on-balance-volume-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxOnBalanceVolumeIndicatorComponent } from './igx-on-balance-volume-indicator-component';
import * as i0 from "@angular/core";
export let IgxOnBalanceVolumeIndicatorModule = /*@__PURE__*/ (() => {
    class IgxOnBalanceVolumeIndicatorModule {
    }
    IgxOnBalanceVolumeIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOnBalanceVolumeIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxOnBalanceVolumeIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOnBalanceVolumeIndicatorModule, declarations: [IgxOnBalanceVolumeIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxOnBalanceVolumeIndicatorComponent, IgxIndicatorProxyModule] });
    IgxOnBalanceVolumeIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOnBalanceVolumeIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxOnBalanceVolumeIndicatorModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOnBalanceVolumeIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxOnBalanceVolumeIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxOnBalanceVolumeIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
