// igniteui-angular-charts/esm2015/lib/igx-polar-spline-area-series-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxPolarLineSeriesBaseComponent } from "./igx-polar-line-series-base-component";
import { IgxPolarBaseComponent } from "./igx-polar-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { PolarSplineAreaSeries } from "./PolarSplineAreaSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent polar spline area series.
*/
export let IgxPolarSplineAreaSeriesComponent = /*@__PURE__*/ (() => {
    class IgxPolarSplineAreaSeriesComponent extends IgxPolarLineSeriesBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new PolarSplineAreaSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows an area shape.
        */
        get isArea() {
            return this.i.es;
        }
        /**
         * Gets whether the current series shows a spline shape.
        */
        get isSpline() {
            return this.i.fy;
        }
        /**
         * Gets or sets the Stiffness property.
        */
        get stiffness() {
            return this.i.acn;
        }
        set stiffness(v) {
            this.i.acn = +v;
        }
    }
    IgxPolarSplineAreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineAreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxPolarSplineAreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxPolarSplineAreaSeriesComponent, selector: "igx-polar-spline-area-series", inputs: { stiffness: "stiffness" }, providers: [{ provide: IgxPolarLineSeriesBaseComponent, useExisting: forwardRef(() => IgxPolarSplineAreaSeriesComponent) }, { provide: IgxPolarBaseComponent, useExisting: forwardRef(() => IgxPolarSplineAreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxPolarSplineAreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxPolarSplineAreaSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxPolarSplineAreaSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineAreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-polar-spline-area-series',
                    template: ``,
                    providers: [{ provide: IgxPolarLineSeriesBaseComponent, useExisting: forwardRef(() => IgxPolarSplineAreaSeriesComponent) }, { provide: IgxPolarBaseComponent, useExisting: forwardRef(() => IgxPolarSplineAreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxPolarSplineAreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxPolarSplineAreaSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { stiffness: [{
                type: Input
            }] } });
