// igniteui-angular-charts/esm2015/lib/igx-data-chart-scatter-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSizeScaleDynamicModule } from './igx-size-scale-dynamic-module';
import { IgxValueBrushScaleDynamicModule } from './igx-value-brush-scale-dynamic-module';
import { IgxCustomPaletteColorScaleDynamicModule } from './igx-custom-palette-color-scale-dynamic-module';
import { IgxBubbleSeriesDynamicModule } from './igx-bubble-series-dynamic-module';
import { IgxScatterSeriesDynamicModule } from './igx-scatter-series-dynamic-module';
import { IgxScatterLineSeriesDynamicModule } from './igx-scatter-line-series-dynamic-module';
import { IgxScatterSplineSeriesDynamicModule } from './igx-scatter-spline-series-dynamic-module';
import { IgxHighDensityScatterSeriesDynamicModule } from './igx-high-density-scatter-series-dynamic-module';
import { IgxDataChartScatterCoreDynamicModule } from './igx-data-chart-scatter-core-dynamic-module';
import { IgxDataChartScatterModule } from './igx-data-chart-scatter-module';
import * as i0 from "@angular/core";
export let IgxDataChartScatterDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartScatterDynamicModule {
    }
    IgxDataChartScatterDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartScatterDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterDynamicModule, imports: [CommonModule, IgxSizeScaleDynamicModule,
            IgxValueBrushScaleDynamicModule,
            IgxCustomPaletteColorScaleDynamicModule,
            IgxBubbleSeriesDynamicModule,
            IgxScatterSeriesDynamicModule,
            IgxScatterLineSeriesDynamicModule,
            IgxScatterSplineSeriesDynamicModule,
            IgxHighDensityScatterSeriesDynamicModule,
            IgxDataChartScatterCoreDynamicModule,
            IgxDataChartScatterModule], exports: [IgxSizeScaleDynamicModule,
            IgxValueBrushScaleDynamicModule,
            IgxCustomPaletteColorScaleDynamicModule,
            IgxBubbleSeriesDynamicModule,
            IgxScatterSeriesDynamicModule,
            IgxScatterLineSeriesDynamicModule,
            IgxScatterSplineSeriesDynamicModule,
            IgxHighDensityScatterSeriesDynamicModule,
            IgxDataChartScatterCoreDynamicModule,
            IgxDataChartScatterModule] });
    IgxDataChartScatterDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterDynamicModule, imports: [[CommonModule, IgxSizeScaleDynamicModule,
                IgxValueBrushScaleDynamicModule,
                IgxCustomPaletteColorScaleDynamicModule,
                IgxBubbleSeriesDynamicModule,
                IgxScatterSeriesDynamicModule,
                IgxScatterLineSeriesDynamicModule,
                IgxScatterSplineSeriesDynamicModule,
                IgxHighDensityScatterSeriesDynamicModule,
                IgxDataChartScatterCoreDynamicModule,
                IgxDataChartScatterModule
            ], IgxSizeScaleDynamicModule,
            IgxValueBrushScaleDynamicModule,
            IgxCustomPaletteColorScaleDynamicModule,
            IgxBubbleSeriesDynamicModule,
            IgxScatterSeriesDynamicModule,
            IgxScatterLineSeriesDynamicModule,
            IgxScatterSplineSeriesDynamicModule,
            IgxHighDensityScatterSeriesDynamicModule,
            IgxDataChartScatterCoreDynamicModule,
            IgxDataChartScatterModule] });
    return IgxDataChartScatterDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxSizeScaleDynamicModule,
                        IgxValueBrushScaleDynamicModule,
                        IgxCustomPaletteColorScaleDynamicModule,
                        IgxBubbleSeriesDynamicModule,
                        IgxScatterSeriesDynamicModule,
                        IgxScatterLineSeriesDynamicModule,
                        IgxScatterSplineSeriesDynamicModule,
                        IgxHighDensityScatterSeriesDynamicModule,
                        IgxDataChartScatterCoreDynamicModule,
                        IgxDataChartScatterModule
                    ],
                    imports: [CommonModule, IgxSizeScaleDynamicModule,
                        IgxValueBrushScaleDynamicModule,
                        IgxCustomPaletteColorScaleDynamicModule,
                        IgxBubbleSeriesDynamicModule,
                        IgxScatterSeriesDynamicModule,
                        IgxScatterLineSeriesDynamicModule,
                        IgxScatterSplineSeriesDynamicModule,
                        IgxHighDensityScatterSeriesDynamicModule,
                        IgxDataChartScatterCoreDynamicModule,
                        IgxDataChartScatterModule
                    ],
                    entryComponents: []
                }]
        }] });
