// igniteui-angular-charts/esm2015/lib/igx-data-tool-tip-layer-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataToolTipLayerModule } from './igx-data-tool-tip-layer-module';
import { IgxDataToolTipLayerComponent } from './igx-data-tool-tip-layer-component';
import { DataToolTipLayer } from './DataToolTipLayer';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxDataToolTipLayerDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataToolTipLayerDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxDataToolTipLayerComponent", IgxDataToolTipLayerComponent);
            TypeRegistrar.register("DataToolTipLayer", DataToolTipLayer.$type);
        }
    }
    IgxDataToolTipLayerDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataToolTipLayerDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataToolTipLayerDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataToolTipLayerDynamicModule, imports: [CommonModule, IgxDataToolTipLayerModule], exports: [IgxDataToolTipLayerModule] });
    IgxDataToolTipLayerDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataToolTipLayerDynamicModule, imports: [[CommonModule, IgxDataToolTipLayerModule], IgxDataToolTipLayerModule] });
    return IgxDataToolTipLayerDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataToolTipLayerDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxDataToolTipLayerModule],
                    exports: [IgxDataToolTipLayerModule],
                    entryComponents: [IgxDataToolTipLayerComponent]
                }]
        }], ctorParameters: function () { return []; } });
