// igniteui-angular-charts/esm2015/lib/igx-data-chart-extended-axes-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryDateTimeXAxisDynamicModule } from './igx-category-date-time-x-axis-dynamic-module';
import { IgxTimeAxisBreakDynamicModule } from './igx-time-axis-break-dynamic-module';
import { IgxTimeAxisIntervalDynamicModule } from './igx-time-axis-interval-dynamic-module';
import { IgxTimeAxisLabelFormatDynamicModule } from './igx-time-axis-label-format-dynamic-module';
import { IgxTimeXAxisDynamicModule } from './igx-time-x-axis-dynamic-module';
import { IgxOrdinalTimeXAxisDynamicModule } from './igx-ordinal-time-x-axis-dynamic-module';
import { IgxDataChartCoreDynamicModule } from './igx-data-chart-core-dynamic-module';
import { IgxDataChartExtendedAxesModule } from './igx-data-chart-extended-axes-module';
import * as i0 from "@angular/core";
export let IgxDataChartExtendedAxesDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartExtendedAxesDynamicModule {
    }
    IgxDataChartExtendedAxesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartExtendedAxesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartExtendedAxesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartExtendedAxesDynamicModule, imports: [CommonModule, IgxCategoryDateTimeXAxisDynamicModule,
            IgxTimeAxisBreakDynamicModule,
            IgxTimeAxisIntervalDynamicModule,
            IgxTimeAxisLabelFormatDynamicModule,
            IgxTimeXAxisDynamicModule,
            IgxOrdinalTimeXAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartExtendedAxesModule], exports: [IgxCategoryDateTimeXAxisDynamicModule,
            IgxTimeAxisBreakDynamicModule,
            IgxTimeAxisIntervalDynamicModule,
            IgxTimeAxisLabelFormatDynamicModule,
            IgxTimeXAxisDynamicModule,
            IgxOrdinalTimeXAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartExtendedAxesModule] });
    IgxDataChartExtendedAxesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartExtendedAxesDynamicModule, imports: [[CommonModule, IgxCategoryDateTimeXAxisDynamicModule,
                IgxTimeAxisBreakDynamicModule,
                IgxTimeAxisIntervalDynamicModule,
                IgxTimeAxisLabelFormatDynamicModule,
                IgxTimeXAxisDynamicModule,
                IgxOrdinalTimeXAxisDynamicModule,
                IgxDataChartCoreDynamicModule,
                IgxDataChartExtendedAxesModule
            ], IgxCategoryDateTimeXAxisDynamicModule,
            IgxTimeAxisBreakDynamicModule,
            IgxTimeAxisIntervalDynamicModule,
            IgxTimeAxisLabelFormatDynamicModule,
            IgxTimeXAxisDynamicModule,
            IgxOrdinalTimeXAxisDynamicModule,
            IgxDataChartCoreDynamicModule,
            IgxDataChartExtendedAxesModule] });
    return IgxDataChartExtendedAxesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartExtendedAxesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxCategoryDateTimeXAxisDynamicModule,
                        IgxTimeAxisBreakDynamicModule,
                        IgxTimeAxisIntervalDynamicModule,
                        IgxTimeAxisLabelFormatDynamicModule,
                        IgxTimeXAxisDynamicModule,
                        IgxOrdinalTimeXAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartExtendedAxesModule
                    ],
                    imports: [CommonModule, IgxCategoryDateTimeXAxisDynamicModule,
                        IgxTimeAxisBreakDynamicModule,
                        IgxTimeAxisIntervalDynamicModule,
                        IgxTimeAxisLabelFormatDynamicModule,
                        IgxTimeXAxisDynamicModule,
                        IgxOrdinalTimeXAxisDynamicModule,
                        IgxDataChartCoreDynamicModule,
                        IgxDataChartExtendedAxesModule
                    ],
                    entryComponents: []
                }]
        }] });
