// igniteui-angular-charts/esm2015/lib/igx-data-chart-toolbar-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataChartToolbarIconsModule } from './igx-data-chart-toolbar-icons-module';
import { IgxDataChartAnnotationModule } from './igx-data-chart-annotation-module';
import { IgxAnnotationLayerProxyModule } from './igx-annotation-layer-proxy-module';
import { IgxDataChartCategoryTrendLineModule } from './igx-data-chart-category-trend-line-module';
import { IgxDataChartInteractivityModule } from './igx-data-chart-interactivity-module';
import { IgxCalloutLayerDynamicModule } from './igx-callout-layer-dynamic-module';
import { IgxValueLayerDynamicModule } from './igx-value-layer-dynamic-module';
import { IgxCrosshairLayerDynamicModule } from './igx-crosshair-layer-dynamic-module';
import { DataChartToolbarFactoryManager } from './DataChartToolbarFactoryManager';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxDataChartToolbarModule = /*@__PURE__*/ (() => {
    class IgxDataChartToolbarModule {
        constructor() {
            TypeRegistrar.register('DataChartToolbarFactoryManager', DataChartToolbarFactoryManager.$type);
            DataChartToolbarFactoryManager.register();
        }
    }
    IgxDataChartToolbarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartToolbarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartToolbarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartToolbarModule, imports: [CommonModule, IgxDataChartToolbarIconsModule,
            IgxDataChartAnnotationModule,
            IgxAnnotationLayerProxyModule,
            IgxDataChartCategoryTrendLineModule,
            IgxDataChartInteractivityModule,
            IgxCalloutLayerDynamicModule,
            IgxValueLayerDynamicModule,
            IgxCrosshairLayerDynamicModule], exports: [IgxDataChartToolbarIconsModule,
            IgxDataChartAnnotationModule,
            IgxAnnotationLayerProxyModule,
            IgxDataChartCategoryTrendLineModule,
            IgxDataChartInteractivityModule,
            IgxCalloutLayerDynamicModule,
            IgxValueLayerDynamicModule,
            IgxCrosshairLayerDynamicModule] });
    IgxDataChartToolbarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartToolbarModule, imports: [[CommonModule, IgxDataChartToolbarIconsModule,
                IgxDataChartAnnotationModule,
                IgxAnnotationLayerProxyModule,
                IgxDataChartCategoryTrendLineModule,
                IgxDataChartInteractivityModule,
                IgxCalloutLayerDynamicModule,
                IgxValueLayerDynamicModule,
                IgxCrosshairLayerDynamicModule
            ], IgxDataChartToolbarIconsModule,
            IgxDataChartAnnotationModule,
            IgxAnnotationLayerProxyModule,
            IgxDataChartCategoryTrendLineModule,
            IgxDataChartInteractivityModule,
            IgxCalloutLayerDynamicModule,
            IgxValueLayerDynamicModule,
            IgxCrosshairLayerDynamicModule] });
    return IgxDataChartToolbarModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartToolbarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataChartToolbarIconsModule,
                        IgxDataChartAnnotationModule,
                        IgxAnnotationLayerProxyModule,
                        IgxDataChartCategoryTrendLineModule,
                        IgxDataChartInteractivityModule,
                        IgxCalloutLayerDynamicModule,
                        IgxValueLayerDynamicModule,
                        IgxCrosshairLayerDynamicModule
                    ],
                    imports: [CommonModule, IgxDataChartToolbarIconsModule,
                        IgxDataChartAnnotationModule,
                        IgxAnnotationLayerProxyModule,
                        IgxDataChartCategoryTrendLineModule,
                        IgxDataChartInteractivityModule,
                        IgxCalloutLayerDynamicModule,
                        IgxValueLayerDynamicModule,
                        IgxCrosshairLayerDynamicModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
