// igniteui-angular-charts/esm2015/lib/igx-data-chart-category-trend-line-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CategoryTrendLineManagerFactory } from './CategoryTrendLineManagerFactory';
import { CategoryTrendLineManager } from './CategoryTrendLineManager';
import { SortingTrendLineManager } from './SortingTrendLineManager';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDataChartCategoryTrendLineModule } from './igx-data-chart-category-trend-line-module';
import * as i0 from "@angular/core";
export let IgxDataChartCategoryTrendLineDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartCategoryTrendLineDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('CategoryTrendLineManagerFactory', CategoryTrendLineManagerFactory);
            TypeRegistrar.registerCons('CategoryTrendLineManager', CategoryTrendLineManager);
            TypeRegistrar.registerCons('SortingTrendLineManager', SortingTrendLineManager);
            TypeRegistrar.register('CategoryTrendLineManagerFactory', CategoryTrendLineManagerFactory.$type);
            TypeRegistrar.register('CategoryTrendLineManager', CategoryTrendLineManager.$type);
            TypeRegistrar.register('SortingTrendLineManager', SortingTrendLineManager.$type);
        }
    }
    IgxDataChartCategoryTrendLineDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryTrendLineDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartCategoryTrendLineDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryTrendLineDynamicModule, imports: [CommonModule, IgxDataChartCategoryTrendLineModule], exports: [IgxDataChartCategoryTrendLineModule] });
    IgxDataChartCategoryTrendLineDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryTrendLineDynamicModule, imports: [[CommonModule, IgxDataChartCategoryTrendLineModule
            ], IgxDataChartCategoryTrendLineModule] });
    return IgxDataChartCategoryTrendLineDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryTrendLineDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataChartCategoryTrendLineModule
                    ],
                    imports: [CommonModule, IgxDataChartCategoryTrendLineModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
