// igniteui-angular-charts/esm2015/lib/igx-accumulation-distribution-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxAccumulationDistributionIndicatorComponent } from './igx-accumulation-distribution-indicator-component';
import * as i0 from "@angular/core";
export let IgxAccumulationDistributionIndicatorModule = /*@__PURE__*/ (() => {
    class IgxAccumulationDistributionIndicatorModule {
    }
    IgxAccumulationDistributionIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAccumulationDistributionIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAccumulationDistributionIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAccumulationDistributionIndicatorModule, declarations: [IgxAccumulationDistributionIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxAccumulationDistributionIndicatorComponent, IgxIndicatorProxyModule] });
    IgxAccumulationDistributionIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAccumulationDistributionIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxAccumulationDistributionIndicatorModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAccumulationDistributionIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxAccumulationDistributionIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxAccumulationDistributionIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
