// igniteui-angular-charts/esm2015/lib/igx-indicators-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAbsoluteVolumeOscillatorIndicatorModule } from './igx-absolute-volume-oscillator-indicator-module';
import { IgxAccumulationDistributionIndicatorModule } from './igx-accumulation-distribution-indicator-module';
import { IgxAverageDirectionalIndexIndicatorModule } from './igx-average-directional-index-indicator-module';
import { IgxAverageTrueRangeIndicatorModule } from './igx-average-true-range-indicator-module';
import { IgxBollingerBandWidthIndicatorModule } from './igx-bollinger-band-width-indicator-module';
import { IgxChaikinOscillatorIndicatorModule } from './igx-chaikin-oscillator-indicator-module';
import { IgxChaikinVolatilityIndicatorModule } from './igx-chaikin-volatility-indicator-module';
import { IgxCommodityChannelIndexIndicatorModule } from './igx-commodity-channel-index-indicator-module';
import { IgxDetrendedPriceOscillatorIndicatorModule } from './igx-detrended-price-oscillator-indicator-module';
import { IgxEaseOfMovementIndicatorModule } from './igx-ease-of-movement-indicator-module';
import { IgxFastStochasticOscillatorIndicatorModule } from './igx-fast-stochastic-oscillator-indicator-module';
import { IgxForceIndexIndicatorModule } from './igx-force-index-indicator-module';
import { IgxFullStochasticOscillatorIndicatorModule } from './igx-full-stochastic-oscillator-indicator-module';
import { IgxMarketFacilitationIndexIndicatorModule } from './igx-market-facilitation-index-indicator-module';
import { IgxMassIndexIndicatorModule } from './igx-mass-index-indicator-module';
import { IgxMedianPriceIndicatorModule } from './igx-median-price-indicator-module';
import { IgxMoneyFlowIndexIndicatorModule } from './igx-money-flow-index-indicator-module';
import { IgxMovingAverageConvergenceDivergenceIndicatorModule } from './igx-moving-average-convergence-divergence-indicator-module';
import { IgxNegativeVolumeIndexIndicatorModule } from './igx-negative-volume-index-indicator-module';
import { IgxOnBalanceVolumeIndicatorModule } from './igx-on-balance-volume-indicator-module';
import { IgxPercentagePriceOscillatorIndicatorModule } from './igx-percentage-price-oscillator-indicator-module';
import { IgxPercentageVolumeOscillatorIndicatorModule } from './igx-percentage-volume-oscillator-indicator-module';
import { IgxPositiveVolumeIndexIndicatorModule } from './igx-positive-volume-index-indicator-module';
import { IgxPriceVolumeTrendIndicatorModule } from './igx-price-volume-trend-indicator-module';
import { IgxRateOfChangeAndMomentumIndicatorModule } from './igx-rate-of-change-and-momentum-indicator-module';
import { IgxRelativeStrengthIndexIndicatorModule } from './igx-relative-strength-index-indicator-module';
import { IgxSlowStochasticOscillatorIndicatorModule } from './igx-slow-stochastic-oscillator-indicator-module';
import { IgxStandardDeviationIndicatorModule } from './igx-standard-deviation-indicator-module';
import { IgxStochRSIIndicatorModule } from './igx-stoch-rsi-indicator-module';
import { IgxTRIXIndicatorModule } from './igx-trix-indicator-module';
import { IgxTypicalPriceIndicatorModule } from './igx-typical-price-indicator-module';
import { IgxUltimateOscillatorIndicatorModule } from './igx-ultimate-oscillator-indicator-module';
import { IgxWeightedCloseIndicatorModule } from './igx-weighted-close-indicator-module';
import { IgxWilliamsPercentRIndicatorModule } from './igx-williams-percent-r-indicator-module';
import { IgxCustomIndicatorModule } from './igx-custom-indicator-module';
import * as i0 from "@angular/core";
export let IgxIndicatorsModule = /*@__PURE__*/ (() => {
    class IgxIndicatorsModule {
    }
    IgxIndicatorsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxIndicatorsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorsModule, imports: [CommonModule, IgxAbsoluteVolumeOscillatorIndicatorModule,
            IgxAccumulationDistributionIndicatorModule,
            IgxAverageDirectionalIndexIndicatorModule,
            IgxAverageTrueRangeIndicatorModule,
            IgxBollingerBandWidthIndicatorModule,
            IgxChaikinOscillatorIndicatorModule,
            IgxChaikinVolatilityIndicatorModule,
            IgxCommodityChannelIndexIndicatorModule,
            IgxDetrendedPriceOscillatorIndicatorModule,
            IgxEaseOfMovementIndicatorModule,
            IgxFastStochasticOscillatorIndicatorModule,
            IgxForceIndexIndicatorModule,
            IgxFullStochasticOscillatorIndicatorModule,
            IgxMarketFacilitationIndexIndicatorModule,
            IgxMassIndexIndicatorModule,
            IgxMedianPriceIndicatorModule,
            IgxMoneyFlowIndexIndicatorModule,
            IgxMovingAverageConvergenceDivergenceIndicatorModule,
            IgxNegativeVolumeIndexIndicatorModule,
            IgxOnBalanceVolumeIndicatorModule,
            IgxPercentagePriceOscillatorIndicatorModule,
            IgxPercentageVolumeOscillatorIndicatorModule,
            IgxPositiveVolumeIndexIndicatorModule,
            IgxPriceVolumeTrendIndicatorModule,
            IgxRateOfChangeAndMomentumIndicatorModule,
            IgxRelativeStrengthIndexIndicatorModule,
            IgxSlowStochasticOscillatorIndicatorModule,
            IgxStandardDeviationIndicatorModule,
            IgxStochRSIIndicatorModule,
            IgxTRIXIndicatorModule,
            IgxTypicalPriceIndicatorModule,
            IgxUltimateOscillatorIndicatorModule,
            IgxWeightedCloseIndicatorModule,
            IgxWilliamsPercentRIndicatorModule,
            IgxCustomIndicatorModule], exports: [IgxAbsoluteVolumeOscillatorIndicatorModule,
            IgxAccumulationDistributionIndicatorModule,
            IgxAverageDirectionalIndexIndicatorModule,
            IgxAverageTrueRangeIndicatorModule,
            IgxBollingerBandWidthIndicatorModule,
            IgxChaikinOscillatorIndicatorModule,
            IgxChaikinVolatilityIndicatorModule,
            IgxCommodityChannelIndexIndicatorModule,
            IgxDetrendedPriceOscillatorIndicatorModule,
            IgxEaseOfMovementIndicatorModule,
            IgxFastStochasticOscillatorIndicatorModule,
            IgxForceIndexIndicatorModule,
            IgxFullStochasticOscillatorIndicatorModule,
            IgxMarketFacilitationIndexIndicatorModule,
            IgxMassIndexIndicatorModule,
            IgxMedianPriceIndicatorModule,
            IgxMoneyFlowIndexIndicatorModule,
            IgxMovingAverageConvergenceDivergenceIndicatorModule,
            IgxNegativeVolumeIndexIndicatorModule,
            IgxOnBalanceVolumeIndicatorModule,
            IgxPercentagePriceOscillatorIndicatorModule,
            IgxPercentageVolumeOscillatorIndicatorModule,
            IgxPositiveVolumeIndexIndicatorModule,
            IgxPriceVolumeTrendIndicatorModule,
            IgxRateOfChangeAndMomentumIndicatorModule,
            IgxRelativeStrengthIndexIndicatorModule,
            IgxSlowStochasticOscillatorIndicatorModule,
            IgxStandardDeviationIndicatorModule,
            IgxStochRSIIndicatorModule,
            IgxTRIXIndicatorModule,
            IgxTypicalPriceIndicatorModule,
            IgxUltimateOscillatorIndicatorModule,
            IgxWeightedCloseIndicatorModule,
            IgxWilliamsPercentRIndicatorModule,
            IgxCustomIndicatorModule] });
    IgxIndicatorsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorsModule, imports: [[CommonModule, IgxAbsoluteVolumeOscillatorIndicatorModule,
                IgxAccumulationDistributionIndicatorModule,
                IgxAverageDirectionalIndexIndicatorModule,
                IgxAverageTrueRangeIndicatorModule,
                IgxBollingerBandWidthIndicatorModule,
                IgxChaikinOscillatorIndicatorModule,
                IgxChaikinVolatilityIndicatorModule,
                IgxCommodityChannelIndexIndicatorModule,
                IgxDetrendedPriceOscillatorIndicatorModule,
                IgxEaseOfMovementIndicatorModule,
                IgxFastStochasticOscillatorIndicatorModule,
                IgxForceIndexIndicatorModule,
                IgxFullStochasticOscillatorIndicatorModule,
                IgxMarketFacilitationIndexIndicatorModule,
                IgxMassIndexIndicatorModule,
                IgxMedianPriceIndicatorModule,
                IgxMoneyFlowIndexIndicatorModule,
                IgxMovingAverageConvergenceDivergenceIndicatorModule,
                IgxNegativeVolumeIndexIndicatorModule,
                IgxOnBalanceVolumeIndicatorModule,
                IgxPercentagePriceOscillatorIndicatorModule,
                IgxPercentageVolumeOscillatorIndicatorModule,
                IgxPositiveVolumeIndexIndicatorModule,
                IgxPriceVolumeTrendIndicatorModule,
                IgxRateOfChangeAndMomentumIndicatorModule,
                IgxRelativeStrengthIndexIndicatorModule,
                IgxSlowStochasticOscillatorIndicatorModule,
                IgxStandardDeviationIndicatorModule,
                IgxStochRSIIndicatorModule,
                IgxTRIXIndicatorModule,
                IgxTypicalPriceIndicatorModule,
                IgxUltimateOscillatorIndicatorModule,
                IgxWeightedCloseIndicatorModule,
                IgxWilliamsPercentRIndicatorModule,
                IgxCustomIndicatorModule
            ], IgxAbsoluteVolumeOscillatorIndicatorModule,
            IgxAccumulationDistributionIndicatorModule,
            IgxAverageDirectionalIndexIndicatorModule,
            IgxAverageTrueRangeIndicatorModule,
            IgxBollingerBandWidthIndicatorModule,
            IgxChaikinOscillatorIndicatorModule,
            IgxChaikinVolatilityIndicatorModule,
            IgxCommodityChannelIndexIndicatorModule,
            IgxDetrendedPriceOscillatorIndicatorModule,
            IgxEaseOfMovementIndicatorModule,
            IgxFastStochasticOscillatorIndicatorModule,
            IgxForceIndexIndicatorModule,
            IgxFullStochasticOscillatorIndicatorModule,
            IgxMarketFacilitationIndexIndicatorModule,
            IgxMassIndexIndicatorModule,
            IgxMedianPriceIndicatorModule,
            IgxMoneyFlowIndexIndicatorModule,
            IgxMovingAverageConvergenceDivergenceIndicatorModule,
            IgxNegativeVolumeIndexIndicatorModule,
            IgxOnBalanceVolumeIndicatorModule,
            IgxPercentagePriceOscillatorIndicatorModule,
            IgxPercentageVolumeOscillatorIndicatorModule,
            IgxPositiveVolumeIndexIndicatorModule,
            IgxPriceVolumeTrendIndicatorModule,
            IgxRateOfChangeAndMomentumIndicatorModule,
            IgxRelativeStrengthIndexIndicatorModule,
            IgxSlowStochasticOscillatorIndicatorModule,
            IgxStandardDeviationIndicatorModule,
            IgxStochRSIIndicatorModule,
            IgxTRIXIndicatorModule,
            IgxTypicalPriceIndicatorModule,
            IgxUltimateOscillatorIndicatorModule,
            IgxWeightedCloseIndicatorModule,
            IgxWilliamsPercentRIndicatorModule,
            IgxCustomIndicatorModule] });
    return IgxIndicatorsModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxIndicatorsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxAbsoluteVolumeOscillatorIndicatorModule,
                        IgxAccumulationDistributionIndicatorModule,
                        IgxAverageDirectionalIndexIndicatorModule,
                        IgxAverageTrueRangeIndicatorModule,
                        IgxBollingerBandWidthIndicatorModule,
                        IgxChaikinOscillatorIndicatorModule,
                        IgxChaikinVolatilityIndicatorModule,
                        IgxCommodityChannelIndexIndicatorModule,
                        IgxDetrendedPriceOscillatorIndicatorModule,
                        IgxEaseOfMovementIndicatorModule,
                        IgxFastStochasticOscillatorIndicatorModule,
                        IgxForceIndexIndicatorModule,
                        IgxFullStochasticOscillatorIndicatorModule,
                        IgxMarketFacilitationIndexIndicatorModule,
                        IgxMassIndexIndicatorModule,
                        IgxMedianPriceIndicatorModule,
                        IgxMoneyFlowIndexIndicatorModule,
                        IgxMovingAverageConvergenceDivergenceIndicatorModule,
                        IgxNegativeVolumeIndexIndicatorModule,
                        IgxOnBalanceVolumeIndicatorModule,
                        IgxPercentagePriceOscillatorIndicatorModule,
                        IgxPercentageVolumeOscillatorIndicatorModule,
                        IgxPositiveVolumeIndexIndicatorModule,
                        IgxPriceVolumeTrendIndicatorModule,
                        IgxRateOfChangeAndMomentumIndicatorModule,
                        IgxRelativeStrengthIndexIndicatorModule,
                        IgxSlowStochasticOscillatorIndicatorModule,
                        IgxStandardDeviationIndicatorModule,
                        IgxStochRSIIndicatorModule,
                        IgxTRIXIndicatorModule,
                        IgxTypicalPriceIndicatorModule,
                        IgxUltimateOscillatorIndicatorModule,
                        IgxWeightedCloseIndicatorModule,
                        IgxWilliamsPercentRIndicatorModule,
                        IgxCustomIndicatorModule
                    ],
                    imports: [CommonModule, IgxAbsoluteVolumeOscillatorIndicatorModule,
                        IgxAccumulationDistributionIndicatorModule,
                        IgxAverageDirectionalIndexIndicatorModule,
                        IgxAverageTrueRangeIndicatorModule,
                        IgxBollingerBandWidthIndicatorModule,
                        IgxChaikinOscillatorIndicatorModule,
                        IgxChaikinVolatilityIndicatorModule,
                        IgxCommodityChannelIndexIndicatorModule,
                        IgxDetrendedPriceOscillatorIndicatorModule,
                        IgxEaseOfMovementIndicatorModule,
                        IgxFastStochasticOscillatorIndicatorModule,
                        IgxForceIndexIndicatorModule,
                        IgxFullStochasticOscillatorIndicatorModule,
                        IgxMarketFacilitationIndexIndicatorModule,
                        IgxMassIndexIndicatorModule,
                        IgxMedianPriceIndicatorModule,
                        IgxMoneyFlowIndexIndicatorModule,
                        IgxMovingAverageConvergenceDivergenceIndicatorModule,
                        IgxNegativeVolumeIndexIndicatorModule,
                        IgxOnBalanceVolumeIndicatorModule,
                        IgxPercentagePriceOscillatorIndicatorModule,
                        IgxPercentageVolumeOscillatorIndicatorModule,
                        IgxPositiveVolumeIndexIndicatorModule,
                        IgxPriceVolumeTrendIndicatorModule,
                        IgxRateOfChangeAndMomentumIndicatorModule,
                        IgxRelativeStrengthIndexIndicatorModule,
                        IgxSlowStochasticOscillatorIndicatorModule,
                        IgxStandardDeviationIndicatorModule,
                        IgxStochRSIIndicatorModule,
                        IgxTRIXIndicatorModule,
                        IgxTypicalPriceIndicatorModule,
                        IgxUltimateOscillatorIndicatorModule,
                        IgxWeightedCloseIndicatorModule,
                        IgxWilliamsPercentRIndicatorModule,
                        IgxCustomIndicatorModule
                    ],
                    entryComponents: []
                }]
        }] });
