// igniteui-angular-charts/esm2015/lib/igx-polar-spline-series-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { IgxPolarLineSeriesBaseComponent } from "./igx-polar-line-series-base-component";
import { IgxPolarBaseComponent } from "./igx-polar-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { PolarSplineSeries } from "./PolarSplineSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent polar spline series.
*/
export let IgxPolarSplineSeriesComponent = /*@__PURE__*/ (() => {
    class IgxPolarSplineSeriesComponent extends IgxPolarLineSeriesBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new PolarSplineSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows a spline shape.
        */
        get isSpline() {
            return this.i.fy;
        }
        /**
         * Gets or sets the Stiffness property.
        */
        get stiffness() {
            return this.i.acn;
        }
        set stiffness(v) {
            this.i.acn = +v;
        }
    }
    IgxPolarSplineSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxPolarSplineSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxPolarSplineSeriesComponent, selector: "igx-polar-spline-series", inputs: { stiffness: "stiffness" }, providers: [{ provide: IgxPolarLineSeriesBaseComponent, useExisting: forwardRef(() => IgxPolarSplineSeriesComponent) }, { provide: IgxPolarBaseComponent, useExisting: forwardRef(() => IgxPolarSplineSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxPolarSplineSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxPolarSplineSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxPolarSplineSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarSplineSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-polar-spline-series',
                    template: ``,
                    providers: [{ provide: IgxPolarLineSeriesBaseComponent, useExisting: forwardRef(() => IgxPolarSplineSeriesComponent) }, { provide: IgxPolarBaseComponent, useExisting: forwardRef(() => IgxPolarSplineSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxPolarSplineSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxPolarSplineSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { stiffness: [{
                type: Input
            }] } });
