// igniteui-angular-charts/esm2015/lib/igx-bollinger-bands-overlay-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxBollingerBandsOverlayModule } from './igx-bollinger-bands-overlay-module';
import { IgxBollingerBandsOverlayComponent } from './igx-bollinger-bands-overlay-component';
import { BollingerBandsOverlay } from './BollingerBandsOverlay';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxBollingerBandsOverlayDynamicModule = /*@__PURE__*/ (() => {
    class IgxBollingerBandsOverlayDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxBollingerBandsOverlayComponent", IgxBollingerBandsOverlayComponent);
            TypeRegistrar.register("BollingerBandsOverlay", BollingerBandsOverlay.$type);
        }
    }
    IgxBollingerBandsOverlayDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBollingerBandsOverlayDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxBollingerBandsOverlayDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBollingerBandsOverlayDynamicModule, imports: [CommonModule, IgxBollingerBandsOverlayModule], exports: [IgxBollingerBandsOverlayModule] });
    IgxBollingerBandsOverlayDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBollingerBandsOverlayDynamicModule, imports: [[CommonModule, IgxBollingerBandsOverlayModule], IgxBollingerBandsOverlayModule] });
    return IgxBollingerBandsOverlayDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBollingerBandsOverlayDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxBollingerBandsOverlayModule],
                    exports: [IgxBollingerBandsOverlayModule],
                    entryComponents: [IgxBollingerBandsOverlayComponent]
                }]
        }], ctorParameters: function () { return []; } });
