// igniteui-angular-charts/esm2015/lib/igx-treemap-node-style-mapping-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTreemapNodeStyleMappingComponent } from './igx-treemap-node-style-mapping-component';
import * as i0 from "@angular/core";
export let IgxTreemapNodeStyleMappingModule = /*@__PURE__*/ (() => {
    class IgxTreemapNodeStyleMappingModule {
    }
    IgxTreemapNodeStyleMappingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTreemapNodeStyleMappingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingModule, declarations: [IgxTreemapNodeStyleMappingComponent], imports: [CommonModule], exports: [IgxTreemapNodeStyleMappingComponent] });
    IgxTreemapNodeStyleMappingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingModule, imports: [[CommonModule]] });
    return IgxTreemapNodeStyleMappingModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapNodeStyleMappingModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxTreemapNodeStyleMappingComponent],
                    imports: [CommonModule],
                    exports: [IgxTreemapNodeStyleMappingComponent],
                    entryComponents: []
                }]
        }] });
