// igniteui-angular-charts/esm2015/lib/igx-numeric-y-axis-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStraightNumericAxisBaseComponent } from "./igx-straight-numeric-axis-base-component";
import { IgxNumericAxisBaseComponent } from "./igx-numeric-axis-base-component";
import { IgxAxisComponent } from "./igx-axis-component";
import { NumericYAxis } from "./NumericYAxis";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent numeric Y axis.
*
* The `NumericYAxis` treats the data as continuously varying numerical data items. Labels on this axis are placed along the Y-axis. Location of labels varies according to the value in a data column that is mapped using the `YMemberPath` property for Scatter Series or `ValueMemberPath` property for Category Series.
*
* ```html
* <igx-data-chart
*     [dataSource]="data" >
*     <igx-category-x-axis
*         label="label"
*         #xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
*         #yAxis>
*     </igx-numeric-y-axis>
*     <igx-column-series
*         [xAxis]="xAxis"
*         [yAxis]="yAxis"
*         valueMemberPath="value">
*     </igx-column-series>
* </igx-data-chart>
* ```
*/
export let IgxNumericYAxisComponent = /*@__PURE__*/ (() => {
    class IgxNumericYAxisComponent extends IgxStraightNumericAxisBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new NumericYAxis();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets if the current axis is of vertical axis
        */
        get isVertical() {
            return this.i.cp;
        }
        scrollRangeIntoView(minimum, maximum) {
            this.i.n1(minimum, maximum);
        }
    }
    IgxNumericYAxisComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericYAxisComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxNumericYAxisComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxNumericYAxisComponent, selector: "igx-numeric-y-axis", providers: [{ provide: IgxStraightNumericAxisBaseComponent, useExisting: forwardRef(() => IgxNumericYAxisComponent) }, { provide: IgxNumericAxisBaseComponent, useExisting: forwardRef(() => IgxNumericYAxisComponent) }, { provide: IgxAxisComponent, useExisting: forwardRef(() => IgxNumericYAxisComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxNumericYAxisComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericYAxisComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-numeric-y-axis',
                    template: ``,
                    providers: [{ provide: IgxStraightNumericAxisBaseComponent, useExisting: forwardRef(() => IgxNumericYAxisComponent) }, { provide: IgxNumericAxisBaseComponent, useExisting: forwardRef(() => IgxNumericYAxisComponent) }, { provide: IgxAxisComponent, useExisting: forwardRef(() => IgxNumericYAxisComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
