// igniteui-angular-charts/esm2015/lib/igx-item-tool-tip-layer-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAnnotationLayerProxyModule } from './igx-annotation-layer-proxy-module';
import { IgxItemToolTipLayerComponent } from './igx-item-tool-tip-layer-component';
import * as i0 from "@angular/core";
export let IgxItemToolTipLayerModule = /*@__PURE__*/ (() => {
    class IgxItemToolTipLayerModule {
    }
    IgxItemToolTipLayerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemToolTipLayerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxItemToolTipLayerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemToolTipLayerModule, declarations: [IgxItemToolTipLayerComponent], imports: [CommonModule, IgxAnnotationLayerProxyModule], exports: [IgxItemToolTipLayerComponent, IgxAnnotationLayerProxyModule] });
    IgxItemToolTipLayerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemToolTipLayerModule, imports: [[CommonModule, IgxAnnotationLayerProxyModule], IgxAnnotationLayerProxyModule] });
    return IgxItemToolTipLayerModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxItemToolTipLayerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxItemToolTipLayerComponent],
                    imports: [CommonModule, IgxAnnotationLayerProxyModule],
                    exports: [IgxItemToolTipLayerComponent, IgxAnnotationLayerProxyModule],
                    entryComponents: []
                }]
        }] });
