// igniteui-angular-charts/esm2015/lib/igx-chart-selection-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxChartSelectionModule } from './igx-chart-selection-module';
import { IgxChartSelection } from './igx-chart-selection';
import { ChartSelection } from './ChartSelection';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxChartSelectionDynamicModule = /*@__PURE__*/ (() => {
    class IgxChartSelectionDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxChartSelection", IgxChartSelection);
            TypeRegistrar.register("ChartSelection", ChartSelection.$type);
        }
    }
    IgxChartSelectionDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSelectionDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxChartSelectionDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSelectionDynamicModule, imports: [CommonModule, IgxChartSelectionModule], exports: [IgxChartSelectionModule] });
    IgxChartSelectionDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSelectionDynamicModule, imports: [[CommonModule, IgxChartSelectionModule], IgxChartSelectionModule] });
    return IgxChartSelectionDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChartSelectionDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxChartSelectionModule],
                    exports: [IgxChartSelectionModule],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
