// igniteui-angular-charts/esm2015/lib/igx-data-chart-toolbar-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataChartToolbarIconsDynamicModule } from './igx-data-chart-toolbar-icons-dynamic-module';
import { IgxDataChartAnnotationDynamicModule } from './igx-data-chart-annotation-dynamic-module';
import { IgxAnnotationLayerProxyDynamicModule } from './igx-annotation-layer-proxy-dynamic-module';
import { IgxDataChartCategoryTrendLineDynamicModule } from './igx-data-chart-category-trend-line-dynamic-module';
import { IgxDataChartInteractivityDynamicModule } from './igx-data-chart-interactivity-dynamic-module';
import { IgxCalloutLayerDynamicModule } from './igx-callout-layer-dynamic-module';
import { IgxValueLayerDynamicModule } from './igx-value-layer-dynamic-module';
import { IgxCrosshairLayerDynamicModule } from './igx-crosshair-layer-dynamic-module';
import { DataChartToolbarFactoryManager } from './DataChartToolbarFactoryManager';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDataChartToolbarModule } from './igx-data-chart-toolbar-module';
import * as i0 from "@angular/core";
export let IgxDataChartToolbarDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartToolbarDynamicModule {
        constructor() {
            TypeRegistrar.register('DataChartToolbarFactoryManager', DataChartToolbarFactoryManager.$type);
            DataChartToolbarFactoryManager.register();
        }
    }
    IgxDataChartToolbarDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartToolbarDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartToolbarDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartToolbarDynamicModule, imports: [CommonModule, IgxDataChartToolbarIconsDynamicModule,
            IgxDataChartAnnotationDynamicModule,
            IgxAnnotationLayerProxyDynamicModule,
            IgxDataChartCategoryTrendLineDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxValueLayerDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxDataChartToolbarModule], exports: [IgxDataChartToolbarIconsDynamicModule,
            IgxDataChartAnnotationDynamicModule,
            IgxAnnotationLayerProxyDynamicModule,
            IgxDataChartCategoryTrendLineDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxValueLayerDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxDataChartToolbarModule] });
    IgxDataChartToolbarDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartToolbarDynamicModule, imports: [[CommonModule, IgxDataChartToolbarIconsDynamicModule,
                IgxDataChartAnnotationDynamicModule,
                IgxAnnotationLayerProxyDynamicModule,
                IgxDataChartCategoryTrendLineDynamicModule,
                IgxDataChartInteractivityDynamicModule,
                IgxCalloutLayerDynamicModule,
                IgxValueLayerDynamicModule,
                IgxCrosshairLayerDynamicModule,
                IgxDataChartToolbarModule
            ], IgxDataChartToolbarIconsDynamicModule,
            IgxDataChartAnnotationDynamicModule,
            IgxAnnotationLayerProxyDynamicModule,
            IgxDataChartCategoryTrendLineDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxValueLayerDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxDataChartToolbarModule] });
    return IgxDataChartToolbarDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartToolbarDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataChartToolbarIconsDynamicModule,
                        IgxDataChartAnnotationDynamicModule,
                        IgxAnnotationLayerProxyDynamicModule,
                        IgxDataChartCategoryTrendLineDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxValueLayerDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxDataChartToolbarModule
                    ],
                    imports: [CommonModule, IgxDataChartToolbarIconsDynamicModule,
                        IgxDataChartAnnotationDynamicModule,
                        IgxAnnotationLayerProxyDynamicModule,
                        IgxDataChartCategoryTrendLineDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxValueLayerDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxDataChartToolbarModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
