// igniteui-angular-charts/esm2015/lib/igx-data-chart-category-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxLineSeriesDynamicModule } from './igx-line-series-dynamic-module';
import { IgxAreaSeriesDynamicModule } from './igx-area-series-dynamic-module';
import { IgxColumnSeriesDynamicModule } from './igx-column-series-dynamic-module';
import { IgxSplineSeriesDynamicModule } from './igx-spline-series-dynamic-module';
import { IgxSplineAreaSeriesDynamicModule } from './igx-spline-area-series-dynamic-module';
import { IgxStepLineSeriesDynamicModule } from './igx-step-line-series-dynamic-module';
import { IgxStepAreaSeriesDynamicModule } from './igx-step-area-series-dynamic-module';
import { IgxPointSeriesDynamicModule } from './igx-point-series-dynamic-module';
import { IgxWaterfallSeriesDynamicModule } from './igx-waterfall-series-dynamic-module';
import { IgxRangeAreaSeriesDynamicModule } from './igx-range-area-series-dynamic-module';
import { IgxRangeColumnSeriesDynamicModule } from './igx-range-column-series-dynamic-module';
import { IgxDataChartCategoryCoreDynamicModule } from './igx-data-chart-category-core-dynamic-module';
import { IgxDataChartCategoryModule } from './igx-data-chart-category-module';
import * as i0 from "@angular/core";
export let IgxDataChartCategoryDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartCategoryDynamicModule {
    }
    IgxDataChartCategoryDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartCategoryDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryDynamicModule, imports: [CommonModule, IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxSplineSeriesDynamicModule,
            IgxSplineAreaSeriesDynamicModule,
            IgxStepLineSeriesDynamicModule,
            IgxStepAreaSeriesDynamicModule,
            IgxPointSeriesDynamicModule,
            IgxWaterfallSeriesDynamicModule,
            IgxRangeAreaSeriesDynamicModule,
            IgxRangeColumnSeriesDynamicModule,
            IgxDataChartCategoryCoreDynamicModule,
            IgxDataChartCategoryModule], exports: [IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxSplineSeriesDynamicModule,
            IgxSplineAreaSeriesDynamicModule,
            IgxStepLineSeriesDynamicModule,
            IgxStepAreaSeriesDynamicModule,
            IgxPointSeriesDynamicModule,
            IgxWaterfallSeriesDynamicModule,
            IgxRangeAreaSeriesDynamicModule,
            IgxRangeColumnSeriesDynamicModule,
            IgxDataChartCategoryCoreDynamicModule,
            IgxDataChartCategoryModule] });
    IgxDataChartCategoryDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryDynamicModule, imports: [[CommonModule, IgxLineSeriesDynamicModule,
                IgxAreaSeriesDynamicModule,
                IgxColumnSeriesDynamicModule,
                IgxSplineSeriesDynamicModule,
                IgxSplineAreaSeriesDynamicModule,
                IgxStepLineSeriesDynamicModule,
                IgxStepAreaSeriesDynamicModule,
                IgxPointSeriesDynamicModule,
                IgxWaterfallSeriesDynamicModule,
                IgxRangeAreaSeriesDynamicModule,
                IgxRangeColumnSeriesDynamicModule,
                IgxDataChartCategoryCoreDynamicModule,
                IgxDataChartCategoryModule
            ], IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxSplineSeriesDynamicModule,
            IgxSplineAreaSeriesDynamicModule,
            IgxStepLineSeriesDynamicModule,
            IgxStepAreaSeriesDynamicModule,
            IgxPointSeriesDynamicModule,
            IgxWaterfallSeriesDynamicModule,
            IgxRangeAreaSeriesDynamicModule,
            IgxRangeColumnSeriesDynamicModule,
            IgxDataChartCategoryCoreDynamicModule,
            IgxDataChartCategoryModule] });
    return IgxDataChartCategoryDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartCategoryDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxLineSeriesDynamicModule,
                        IgxAreaSeriesDynamicModule,
                        IgxColumnSeriesDynamicModule,
                        IgxSplineSeriesDynamicModule,
                        IgxSplineAreaSeriesDynamicModule,
                        IgxStepLineSeriesDynamicModule,
                        IgxStepAreaSeriesDynamicModule,
                        IgxPointSeriesDynamicModule,
                        IgxWaterfallSeriesDynamicModule,
                        IgxRangeAreaSeriesDynamicModule,
                        IgxRangeColumnSeriesDynamicModule,
                        IgxDataChartCategoryCoreDynamicModule,
                        IgxDataChartCategoryModule
                    ],
                    imports: [CommonModule, IgxLineSeriesDynamicModule,
                        IgxAreaSeriesDynamicModule,
                        IgxColumnSeriesDynamicModule,
                        IgxSplineSeriesDynamicModule,
                        IgxSplineAreaSeriesDynamicModule,
                        IgxStepLineSeriesDynamicModule,
                        IgxStepAreaSeriesDynamicModule,
                        IgxPointSeriesDynamicModule,
                        IgxWaterfallSeriesDynamicModule,
                        IgxRangeAreaSeriesDynamicModule,
                        IgxRangeColumnSeriesDynamicModule,
                        IgxDataChartCategoryCoreDynamicModule,
                        IgxDataChartCategoryModule
                    ],
                    entryComponents: []
                }]
        }] });
