// igniteui-angular-charts/esm2015/lib/igx-scatter-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterSeriesModule } from './igx-scatter-series-module';
import { IgxScatterSeriesComponent } from './igx-scatter-series-component';
import { ScatterSeries } from './ScatterSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxScatterSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxScatterSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxScatterSeriesComponent", IgxScatterSeriesComponent);
            TypeRegistrar.register("ScatterSeries", ScatterSeries.$type);
        }
    }
    IgxScatterSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesDynamicModule, imports: [CommonModule, IgxScatterSeriesModule], exports: [IgxScatterSeriesModule] });
    IgxScatterSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesDynamicModule, imports: [[CommonModule, IgxScatterSeriesModule], IgxScatterSeriesModule] });
    return IgxScatterSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxScatterSeriesModule],
                    exports: [IgxScatterSeriesModule],
                    entryComponents: [IgxScatterSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
