// igniteui-angular-charts/esm2015/lib/igx-time-axis-label-format-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTimeAxisLabelFormatComponent } from './igx-time-axis-label-format-component';
import * as i0 from "@angular/core";
export let IgxTimeAxisLabelFormatModule = /*@__PURE__*/ (() => {
    class IgxTimeAxisLabelFormatModule {
    }
    IgxTimeAxisLabelFormatModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisLabelFormatModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTimeAxisLabelFormatModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisLabelFormatModule, declarations: [IgxTimeAxisLabelFormatComponent], imports: [CommonModule], exports: [IgxTimeAxisLabelFormatComponent] });
    IgxTimeAxisLabelFormatModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisLabelFormatModule, imports: [[CommonModule]] });
    return IgxTimeAxisLabelFormatModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisLabelFormatModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxTimeAxisLabelFormatComponent],
                    imports: [CommonModule],
                    exports: [IgxTimeAxisLabelFormatComponent],
                    entryComponents: []
                }]
        }] });
