// igniteui-angular-charts/esm2015/lib/igx-scatter-spline-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterSplineSeriesModule } from './igx-scatter-spline-series-module';
import { IgxScatterSplineSeriesComponent } from './igx-scatter-spline-series-component';
import { ScatterSplineSeries } from './ScatterSplineSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxScatterSplineSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxScatterSplineSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxScatterSplineSeriesComponent", IgxScatterSplineSeriesComponent);
            TypeRegistrar.register("ScatterSplineSeries", ScatterSplineSeries.$type);
        }
    }
    IgxScatterSplineSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSplineSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxScatterSplineSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSplineSeriesDynamicModule, imports: [CommonModule, IgxScatterSplineSeriesModule], exports: [IgxScatterSplineSeriesModule] });
    IgxScatterSplineSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSplineSeriesDynamicModule, imports: [[CommonModule, IgxScatterSplineSeriesModule], IgxScatterSplineSeriesModule] });
    return IgxScatterSplineSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScatterSplineSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxScatterSplineSeriesModule],
                    exports: [IgxScatterSplineSeriesModule],
                    entryComponents: [IgxScatterSplineSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
