// igniteui-angular-charts/esm2015/lib/igx-average-directional-index-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxAverageDirectionalIndexIndicatorModule } from './igx-average-directional-index-indicator-module';
import { IgxAverageDirectionalIndexIndicatorComponent } from './igx-average-directional-index-indicator-component';
import { AverageDirectionalIndexIndicator } from './AverageDirectionalIndexIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxAverageDirectionalIndexIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxAverageDirectionalIndexIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxAverageDirectionalIndexIndicatorComponent", IgxAverageDirectionalIndexIndicatorComponent);
            TypeRegistrar.register("AverageDirectionalIndexIndicator", AverageDirectionalIndexIndicator.$type);
        }
    }
    IgxAverageDirectionalIndexIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAverageDirectionalIndexIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxAverageDirectionalIndexIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAverageDirectionalIndexIndicatorDynamicModule, imports: [CommonModule, IgxAverageDirectionalIndexIndicatorModule], exports: [IgxAverageDirectionalIndexIndicatorModule] });
    IgxAverageDirectionalIndexIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAverageDirectionalIndexIndicatorDynamicModule, imports: [[CommonModule, IgxAverageDirectionalIndexIndicatorModule], IgxAverageDirectionalIndexIndicatorModule] });
    return IgxAverageDirectionalIndexIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAverageDirectionalIndexIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxAverageDirectionalIndexIndicatorModule],
                    exports: [IgxAverageDirectionalIndexIndicatorModule],
                    entryComponents: [IgxAverageDirectionalIndexIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
