// igniteui-angular-charts/esm2015/lib/igx-area-fragment-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxFragmentBaseComponent } from "./igx-fragment-base-component";
import { IgxHorizontalAnchoredCategorySeriesComponent } from "./igx-horizontal-anchored-category-series-component";
import { IgxAnchoredCategorySeriesComponent } from "./igx-anchored-category-series-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { AreaFragment } from "./AreaFragment";
import * as i0 from "@angular/core";
/**
 * Represents one part of a StackedAreaSeries.
*/
export let IgxAreaFragmentComponent = /*@__PURE__*/ (() => {
    class IgxAreaFragmentComponent extends IgxFragmentBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new AreaFragment();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows an area or line shape.
        */
        get isAreaOrLine() {
            return this.i.et;
        }
        /**
         * Gets whether the current series shows an area shape.
        */
        get isArea() {
            return this.i.es;
        }
    }
    IgxAreaFragmentComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaFragmentComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxAreaFragmentComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxAreaFragmentComponent, selector: "igx-area-fragment", providers: [{ provide: IgxFragmentBaseComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }, { provide: IgxHorizontalAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }, { provide: IgxAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxAreaFragmentComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxAreaFragmentComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-area-fragment',
                    template: ``,
                    providers: [{ provide: IgxFragmentBaseComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }, { provide: IgxHorizontalAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }, { provide: IgxAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxAreaFragmentComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
