// igniteui-angular-charts/esm2015/lib/igx-stacked-100-column-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStacked100ColumnSeriesComponent } from './igx-stacked-100-column-series-component';
import * as i0 from "@angular/core";
export let IgxStacked100ColumnSeriesModule = /*@__PURE__*/ (() => {
    class IgxStacked100ColumnSeriesModule {
    }
    IgxStacked100ColumnSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100ColumnSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStacked100ColumnSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100ColumnSeriesModule, declarations: [IgxStacked100ColumnSeriesComponent], imports: [CommonModule], exports: [IgxStacked100ColumnSeriesComponent] });
    IgxStacked100ColumnSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100ColumnSeriesModule, imports: [[CommonModule]] });
    return IgxStacked100ColumnSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100ColumnSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStacked100ColumnSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxStacked100ColumnSeriesComponent],
                    entryComponents: []
                }]
        }] });
