// igniteui-angular-charts/esm2015/lib/igx-pie-chart-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPieChartCoreDynamicModule } from './igx-pie-chart-core-dynamic-module';
import { IgxDataChartInteractivityDynamicModule } from './igx-data-chart-interactivity-dynamic-module';
import { IgxPieChartModule } from './igx-pie-chart-module';
import * as i0 from "@angular/core";
export let IgxPieChartDynamicModule = /*@__PURE__*/ (() => {
    class IgxPieChartDynamicModule {
    }
    IgxPieChartDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPieChartDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartDynamicModule, imports: [CommonModule, IgxPieChartCoreDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxPieChartModule], exports: [IgxPieChartCoreDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxPieChartModule] });
    IgxPieChartDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartDynamicModule, imports: [[CommonModule, IgxPieChartCoreDynamicModule,
                IgxDataChartInteractivityDynamicModule,
                IgxPieChartModule
            ], IgxPieChartCoreDynamicModule,
            IgxDataChartInteractivityDynamicModule,
            IgxPieChartModule] });
    return IgxPieChartDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxPieChartCoreDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxPieChartModule
                    ],
                    imports: [CommonModule, IgxPieChartCoreDynamicModule,
                        IgxDataChartInteractivityDynamicModule,
                        IgxPieChartModule
                    ],
                    entryComponents: []
                }]
        }] });
