// igniteui-angular-charts/esm2015/lib/igx-financial-series-component.js
import { EventEmitter, Output, Component, Input } from '@angular/core';
import { delegateCombine } from "igniteui-angular-core";
import { IgxCategoryAxisBaseComponent } from "./igx-category-axis-base-component";
import { IgxNumericYAxisComponent } from "./igx-numeric-y-axis-component";
import { CategoryTransitionInMode_$type } from "./CategoryTransitionInMode";
import { IgxAssigningCategoryStyleEventArgs } from "./igx-assigning-category-style-event-args";
import { IgxFinancialEventArgs } from "./igx-financial-event-args";
import { IgxSeriesComponent } from "./igx-series-component";
import { ensureBool, brushToString, stringToBrush, ensureEnum, toPoint, fromPoint } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents the base class for all financial indicator and overlay series.
*
* The `FinancialSeries` represents the base class for all financial indicator and overlay series.
*
* ```html
*  <igx-data-chart
*     #chart
*         [dataSource]="data" >
*   <igx-category-x-axis
*         label="label"
*         #xAxis>
*         </igx-category-x-axis>
*  <igx-numeric-y-axis
*         minimumValue="0"
*         #yAxis>
* </igx-numeric-y-axis>
* <igx-financial-price-series
*     [xAxis]="xAxis"
*       [yAxis]="yAxis"
*       openMemberPath="open"
*       highMemberPath="high"
*       lowMemberPath="low"
*       closeMemberPath="close"
*       volumeMemberPath="volume">
* </igx-financial-price-series>
* </igx-data-chart>
* ```
*
* ```ts
* let series = new IgxFinancialPriceSeriesComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.openMemberPath = "open";
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* series.closeMemberPath = "close";
* series.volumeMemberPath="volume";
* this.chart.series.add(series);
* ```
*/
export let IgxFinancialSeriesComponent = /*@__PURE__*/ (() => {
    class IgxFinancialSeriesComponent extends IgxSeriesComponent {
        constructor() {
            super();
            this._assigningCategoryStyle = null;
            this._typical = null;
            this._typicalBasedOn = null;
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series is a financial type series.
        */
        get isFinancial() {
            return this.i.e4;
        }
        /**
         * Gets or sets the brush to use for negative portions of the series.
        *
        * The `NegativeBrush` property is used to brush the negative portions of the series.
        *
        * ```ts
        * series.negativeBrush="red";
        * ```
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume"
        * 	  negativeBrush="red">
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        */
        get negativeBrush() {
            return brushToString(this.i.aao);
        }
        set negativeBrush(v) {
            this.i.aao = stringToBrush(v);
        }
        get hasValueAxis() {
            return this.i.ei;
        }
        get isValueAxisInverted() {
            return this.i.f3;
        }
        /**
         * Gets or sets the effective x-axis for the current FinancialSeries object.
        *
        * Use `XAxis` property to set effective x-axis for the current FinancialSeries object.
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume">
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * let series = new IgxFinancialPriceSeriesComponent();
        * series.xAxis = this.xAxis;
        * series.yAxis = this.yAxis;
        * series.openMemberPath = "open";
        * series.highMemberPath = "high";
        * series.lowMemberPath = "low";
        * series.closeMemberPath = "close";
        * series.volumeMemberPath="volume";
        * this.chart.series.add(series);
        * ```
        */
        get xAxis() {
            const r = this.i.wy;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                let e = IgxCategoryAxisBaseComponent._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        }
        set xAxis(v) {
            if (v != null && this._stylingContainer && v._styling)
                v._styling(this._stylingContainer, this, this);
            v == null ? this.i.wy = null : this.i.wy = v.i;
        }
        /**
         * Gets or sets the effective y-axis for the current FinancialSeries object.
        *
        * Use `YAxis` property to set the effective y-axis for the current FinancialSeries object.
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume">
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * let series = new IgxFinancialPriceSeriesComponent();
        * series.xAxis = this.xAxis;
        * series.yAxis = this.yAxis;
        * series.openMemberPath = "open";
        * series.highMemberPath = "high";
        * series.lowMemberPath = "low";
        * series.closeMemberPath = "close";
        * series.volumeMemberPath="volume";
        * this.chart.series.add(series);
        * ```
        */
        get yAxis() {
            const r = this.i.xh;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                let e = IgxNumericYAxisComponent._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        }
        set yAxis(v) {
            if (v != null && this._stylingContainer && v._styling)
                v._styling(this._stylingContainer, this, this);
            v == null ? this.i.xh = null : this.i.xh = v.i;
        }
        /**
         * Gets or sets the open mapping property for the current series object.
        *
        * The `OpenMemberPath` property is used for open mapping of the current series object.
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume">
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * let series = new IgxFinancialPriceSeriesComponent();
        * series.xAxis = this.xAxis;
        * series.yAxis = this.yAxis;
        * series.openMemberPath = "open";
        * series.highMemberPath = "high";
        * series.lowMemberPath = "low";
        * series.closeMemberPath = "close";
        * series.volumeMemberPath="volume";
        * this.chart.series.add(series);
        * ```
        */
        get openMemberPath() {
            return this.i.zo;
        }
        set openMemberPath(v) {
            this.i.zo = v;
        }
        /**
         * Gets or sets the high mapping property for the current series object.
        *
        * Use `HighMemberPath` propert for high mapping of the current series object.
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume">
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * let series = new IgxFinancialPriceSeriesComponent();
        * series.xAxis = this.xAxis;
        * series.yAxis = this.yAxis;
        * series.openMemberPath = "open";
        * series.highMemberPath = "high";
        * series.lowMemberPath = "low";
        * series.closeMemberPath = "close";
        * series.volumeMemberPath="volume";
        * this.chart.series.add(series);
        * ```
        */
        get highMemberPath() {
            return this.i.zf;
        }
        set highMemberPath(v) {
            this.i.zf = v;
        }
        /**
         * Gets or sets the low mapping property for the current series object.
        *
        * The `LowMemberPath` is used for low mapping property of the current series object.
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume">
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * let series = new IgxFinancialPriceSeriesComponent();
        * series.xAxis = this.xAxis;
        * series.yAxis = this.yAxis;
        * series.openMemberPath = "open";
        * series.highMemberPath = "high";
        * series.lowMemberPath = "low";
        * series.closeMemberPath = "close";
        * series.volumeMemberPath="volume";
        * this.chart.series.add(series);
        * ```
        */
        get lowMemberPath() {
            return this.i.zk;
        }
        set lowMemberPath(v) {
            this.i.zk = v;
        }
        /**
         * Gets or sets the close mapping property for the current series object.
        *
        * Use `CloseMemberPath` property for the close mapping of the current series object.
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume">
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * let series = new IgxFinancialPriceSeriesComponent();
        * series.xAxis = this.xAxis;
        * series.yAxis = this.yAxis;
        * series.openMemberPath = "open";
        * series.highMemberPath = "high";
        * series.lowMemberPath = "low";
        * series.closeMemberPath = "close";
        * series.volumeMemberPath="volume";
        * this.chart.series.add(series);
        * ```
        */
        get closeMemberPath() {
            return this.i.y2;
        }
        set closeMemberPath(v) {
            this.i.y2 = v;
        }
        /**
         * Gets or sets the volume mapping property for the current series object.
        *
        * The `VolumeMemberPath` property is used for the volume mapping property of the current series object.
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume">
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * let series = new IgxFinancialPriceSeriesComponent();
        * series.xAxis = this.xAxis;
        * series.yAxis = this.yAxis;
        * series.openMemberPath = "open";
        * series.highMemberPath = "high";
        * series.lowMemberPath = "low";
        * series.closeMemberPath = "close";
        * series.volumeMemberPath="volume";
        * this.chart.series.add(series);
        * ```
        */
        get volumeMemberPath() {
            return this.i.zs;
        }
        set volumeMemberPath(v) {
            this.i.zs = v;
        }
        /**
         * Gets or sets the highlighted X value mapping property for the current series object.
        */
        get highlightedOpenMemberPath() {
            return this.i.zb;
        }
        set highlightedOpenMemberPath(v) {
            this.i.zb = v;
        }
        /**
         * Gets or sets the highlighted Y value mapping property for the current series object.
        */
        get highlightedHighMemberPath() {
            return this.i.y7;
        }
        set highlightedHighMemberPath(v) {
            this.i.y7 = v;
        }
        /**
         * Gets or sets the highlighted X value mapping property for the current series object.
        */
        get highlightedLowMemberPath() {
            return this.i.y9;
        }
        set highlightedLowMemberPath(v) {
            this.i.y9 = v;
        }
        /**
         * Gets or sets the highlighted Y value mapping property for the current series object.
        */
        get highlightedCloseMemberPath() {
            return this.i.y5;
        }
        set highlightedCloseMemberPath(v) {
            this.i.y5 = v;
        }
        /**
         * Gets or sets the highlighted Y value mapping property for the current series object.
        */
        get highlightedVolumeMemberPath() {
            return this.i.zd;
        }
        set highlightedVolumeMemberPath(v) {
            this.i.zd = v;
        }
        /**
         * Gets or sets whether this category series should allow custom style overrides of its individual visuals.
        *
        * The `IsCustomCategoryStyleAllowed` property is used to check if the category series should allow custom style.
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume"
        * 	  isCustomCategoryStyleAllowed="True" />
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        *
        * ```ts
        * series.isCustomCategoryStyleAllowed= true;
        * ```
        */
        get isCustomCategoryStyleAllowed() {
            return this.i.x4;
        }
        set isCustomCategoryStyleAllowed(v) {
            this.i.x4 = ensureBool(v);
        }
        /**
         * Gets or sets the method by which to animate the data into the chart when the chart data source is swapped.
         * Note: Transitions are not currently supported for stacked series.
        *
        * The `TransitionInMode` property is used to gets or sets the method by which to animate the data into the chart when the chart data source is swapped.
        *
        * ```ts
        * series.transitionInMode =CategoryTransitionInMode.Auto;
        * ```
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume"
        * 	  transitionInMode= "Auto">
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        */
        get transitionInMode() {
            return this.i.w9;
        }
        set transitionInMode(v) {
            this.i.w9 = ensureEnum(CategoryTransitionInMode_$type, v);
        }
        /**
         * Gets or sets whether the series should transition into the plot area when a new data source is assigned.
         * Note: Transitions are not currently supported for stacked series.
        *
        * The `IsTransitionInEnabled` property is used to check if the series should transition into the plot area when a new data source is assigned.
        *
        * ```ts
        * series.isTransitionInEnabled= true;
        * ```
        *
        * ```html
        *  <igx-data-chart
        *     #chart
        *         [dataSource]="data" >
        *   <igx-category-x-axis
        *         label="label"
        *         #xAxis>
        *         </igx-category-x-axis>
        *  <igx-numeric-y-axis
        *         minimumValue="0"
        *         #yAxis>
        * </igx-numeric-y-axis>
        * <igx-financial-price-series
        *     [xAxis]="xAxis"
        *       [yAxis]="yAxis"
        *       openMemberPath="open"
        *       highMemberPath="high"
        *       lowMemberPath="low"
        *       closeMemberPath="close"
        *       volumeMemberPath="volume"
        * 	  isTransitionInEnabled= "true">
        * </igx-financial-price-series>
        * </igx-data-chart>
        * ```
        */
        get isTransitionInEnabled() {
            return this.i.x5;
        }
        set isTransitionInEnabled(v) {
            this.i.x5 = ensureBool(v);
        }
        /**
         * Overridden by derived series classes to indicate when negative colors are supported or not.
        */
        get isNegativeColorSupported() {
            return this.i.fk;
        }
        findByName(name) {
            var baseResult = super.findByName(name);
            if (baseResult) {
                return baseResult;
            }
            if (this.xAxis && this.xAxis.name && this.xAxis.name == name) {
                return this.xAxis;
            }
            if (this.yAxis && this.yAxis.name && this.yAxis.name == name) {
                return this.yAxis;
            }
            return null;
        }
        _styling(container, component, parent) {
            super._styling(container, component, parent);
            this._inStyling = true;
            if (this.xAxis && this.xAxis._styling) {
                this.xAxis._styling(container, component, this);
            }
            if (this.yAxis && this.yAxis._styling) {
                this.yAxis._styling(container, component, this);
            }
            this._inStyling = false;
        }
        getItemValue(item, memberPathName) {
            let iv = this.i.ko(item, memberPathName);
            return (iv);
        }
        /**
         * Gets the value of a requested member path from the series.
        
        * @param memberPathName  * The property name of a valid member path for the series
        */
        getMemberPathValue(memberPathName) {
            let iv = this.i.mb(memberPathName);
            return (iv);
        }
        getPreviousOrExactIndex(world, skipUnknowns) {
            let iv = this.i.j8(toPoint(world), skipUnknowns);
            return (iv);
        }
        getNextOrExactIndex(world, skipUnknowns) {
            let iv = this.i.j6(toPoint(world), skipUnknowns);
            return (iv);
        }
        /**
         * Returns the offset value for this series if grouped on a category axis.
        
        *
        * The `GetOffsetValue` method returns the offset value for this series if grouped on a category axis.
        *
        * ```ts
        * number offset= series.getOffsetValue();
        * ```
        */
        getOffsetValue() {
            let iv = this.i.getOffsetValue();
            return (iv);
        }
        /**
         * Returns the width of the category grouping this series is in.
        
        *
        * The `GetCategoryWidth` method returns the width of the category grouping this series is in.
        *
        * ```ts
        * number width = series.getCategoryWidth();
        * ```
        */
        getCategoryWidth() {
            let iv = this.i.getCategoryWidth();
            return (iv);
        }
        getSeriesValuePosition(world, useInterpolation, skipUnknowns) {
            let iv = this.i.v6(toPoint(world), useInterpolation, skipUnknowns);
            return fromPoint(iv);
        }
        getSeriesValue(world, useInterpolation, skipUnknowns) {
            let iv = this.i.i3(toPoint(world), useInterpolation, skipUnknowns);
            return (iv);
        }
        getSeriesHighValue(world, useInterpolation, skipUnknowns) {
            let iv = this.i.iz(toPoint(world), useInterpolation, skipUnknowns);
            return (iv);
        }
        getSeriesLowValue(world, useInterpolation, skipUnknowns) {
            let iv = this.i.i1(toPoint(world), useInterpolation, skipUnknowns);
            return (iv);
        }
        getSeriesCloseValue(world, useInterpolation, skipUnknowns) {
            let iv = this.i.yq(toPoint(world), useInterpolation, skipUnknowns);
            return (iv);
        }
        getSeriesOpenValue(world, useInterpolation, skipUnknowns) {
            let iv = this.i.yr(toPoint(world), useInterpolation, skipUnknowns);
            return (iv);
        }
        getSeriesVolumeValue(world, useInterpolation, skipUnknowns) {
            let iv = this.i.ys(toPoint(world), useInterpolation, skipUnknowns);
            return (iv);
        }
        getSeriesHighValuePosition(world, useInterpolation, skipUnknowns) {
            let iv = this.i.v2(toPoint(world), useInterpolation, skipUnknowns);
            return fromPoint(iv);
        }
        getSeriesLowValuePosition(world, useInterpolation, skipUnknowns) {
            let iv = this.i.v4(toPoint(world), useInterpolation, skipUnknowns);
            return fromPoint(iv);
        }
        getSeriesOpenValuePosition(world, useInterpolation, skipUnknowns) {
            let iv = this.i.aaq(toPoint(world), useInterpolation, skipUnknowns);
            return fromPoint(iv);
        }
        getSeriesCloseValuePosition(world, useInterpolation, skipUnknowns) {
            let iv = this.i.aap(toPoint(world), useInterpolation, skipUnknowns);
            return fromPoint(iv);
        }
        getSeriesVolumeValuePosition(world, useInterpolation, skipUnknowns) {
            let iv = this.i.aar(toPoint(world), useInterpolation, skipUnknowns);
            return fromPoint(iv);
        }
        /**
         * Determine if object can be used as YAxis
        
        * @param axis  * The object to check
        *
        * Use `CanUseAsYAxis` method to determine if the object can be used as YAxis.
        *
        * ```ts
        *   boolean canY =series.canUseAsYAxis(yAxis);
        * ```
        */
        canUseAsYAxis(axis) {
            let iv = this.i.x1(axis);
            return (iv);
        }
        /**
         * Determine if object can be used as XAxis
        
        * @param axis  * The object to check
        *
        * Use `CanUseAsXAxis` method to determine if the object can be used as XAxis.
        *
        * ```ts
        *   boolean canX =series.canUseAsXAxis(xAxis);
        * ```
        */
        canUseAsXAxis(axis) {
            let iv = this.i.x0(axis);
            return (iv);
        }
        /**
         * Gets the precise item index, if possible, based on the closeness to the previous or next whole integer. If the series cannot provide this information, GetExactItemIndex will return the same integer value as GetItemIndex.
        
        * @param world  * The world position for which to return the index.
        *
        * The `GetExactItemIndex` method gets the precise item index, if possible, based on the closeness to the previous or next whole integer.
        *
        * If the series cannot provide this information, GetExactItemIndex will return the same integer value as GetItemIndex.
        *
        * ```ts
        * number index = series.getExactItemIndex(world);
        * ```
        */
        getExactItemIndex(world) {
            let iv = this.i.ip(toPoint(world));
            return (iv);
        }
        /**
         * Gets the index of the item that resides at the provided world coordinates.
        
        * @param world  * The world coordinates of the requested item.
        *
        * The `GetItemIndex` method gets the index of the item that resides at the provided world coordinates.
        *
        * ```ts
        * number itemindex= series.getItemIndex(world);
        * ```
        */
        getItemIndex(world) {
            let iv = this.i.j1(toPoint(world));
            return (iv);
        }
        /**
         * Gets the item that is the best match for the specified world coordinates.
        
        * @param world  * The world coordinates to use.
        *
        * The `GetItem` method gets the item that is the best match for the specified world coordinates.
        */
        getItem(world) {
            let iv = this.i.kl(toPoint(world));
            return (iv);
        }
        setNegativeColors(negativeBrush, negativeOutline) {
            this.i.rx(stringToBrush(negativeBrush), stringToBrush(negativeOutline));
        }
        /**
         * Event raised when Assigning Category Style
        *
        * The `AssigningCategoryStyle` event raised when assigning Category Style.
        *
        * ```ts
        * series.assigningCategoryStyle= this.chart_AssigningCategoryStyle;
        * chart_AssigningCategoryStyle(sender :any,args: AssigningCategoryStyleEventArgs )
        *   {
        *
        *
        *   }
        * ```
        */
        get assigningCategoryStyle() {
            if (this._assigningCategoryStyle == null) {
                this._assigningCategoryStyle = new EventEmitter();
                this.i.assigningCategoryStyle = delegateCombine(this.i.assigningCategoryStyle, (o, e) => {
                    let outerArgs = new IgxAssigningCategoryStyleEventArgs();
                    outerArgs._provideImplementation(e);
                    if (this.beforeAssigningCategoryStyle) {
                        this.beforeAssigningCategoryStyle(this, outerArgs);
                    }
                    this._assigningCategoryStyle.emit({
                        sender: this,
                        args: outerArgs
                    });
                });
            }
            return this._assigningCategoryStyle;
        }
        /**
         * Handle this event in order to perform a custom typical price calculation.
        *
        * Use `Typical` event to perform a custom typical price calculation.
        *
        * ```ts
        * series.typical= this.chart_typical;
        *
        *   chart_typical(sender :any,args: FinancialEventArgs )
        *   {
        *
        *
        *   }
        * ```
        */
        get typical() {
            if (this._typical == null) {
                this._typical = new EventEmitter();
                this.i.typical = delegateCombine(this.i.typical, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxFinancialEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforeTypical) {
                            this.beforeTypical(this, outerArgs);
                        }
                        this._typical.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._typical;
        }
        /**
         * Handle this event in order to specify which columns the Typical price calculation is based on.
        *
        * Use `TypicalBasedOn` event to specify which columns the Typical price calculation is based on.
        *
        * ```ts
        *    series.typicalBasedOn= this.chart_typicalBasedOn;
        *     chart_typicalBasedOn(sender :any,args: FinancialEventArgs )
        *   {
        *
        *   }
        * ```
        */
        get typicalBasedOn() {
            if (this._typicalBasedOn == null) {
                this._typicalBasedOn = new EventEmitter();
                this.i.typicalBasedOn = delegateCombine(this.i.typicalBasedOn, (o, e) => {
                    this._runInZone(() => {
                        let outerArgs = new IgxFinancialEventArgs();
                        outerArgs._provideImplementation(e);
                        if (this.beforeTypicalBasedOn) {
                            this.beforeTypicalBasedOn(this, outerArgs);
                        }
                        this._typicalBasedOn.emit({
                            sender: this,
                            args: outerArgs
                        });
                    });
                });
            }
            return this._typicalBasedOn;
        }
    }
    IgxFinancialSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxFinancialSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxFinancialSeriesComponent, selector: "ng-component", inputs: { negativeBrush: "negativeBrush", xAxis: "xAxis", yAxis: "yAxis", openMemberPath: "openMemberPath", highMemberPath: "highMemberPath", lowMemberPath: "lowMemberPath", closeMemberPath: "closeMemberPath", volumeMemberPath: "volumeMemberPath", highlightedOpenMemberPath: "highlightedOpenMemberPath", highlightedHighMemberPath: "highlightedHighMemberPath", highlightedLowMemberPath: "highlightedLowMemberPath", highlightedCloseMemberPath: "highlightedCloseMemberPath", highlightedVolumeMemberPath: "highlightedVolumeMemberPath", isCustomCategoryStyleAllowed: "isCustomCategoryStyleAllowed", transitionInMode: "transitionInMode", isTransitionInEnabled: "isTransitionInEnabled" }, outputs: { assigningCategoryStyle: "assigningCategoryStyle", typical: "typical", typicalBasedOn: "typicalBasedOn" }, usesInheritance: true, ngImport: i0, template: ``, isInline: true });
    return IgxFinancialSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    template: ``,
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { negativeBrush: [{
                type: Input
            }], xAxis: [{
                type: Input
            }], yAxis: [{
                type: Input
            }], openMemberPath: [{
                type: Input
            }], highMemberPath: [{
                type: Input
            }], lowMemberPath: [{
                type: Input
            }], closeMemberPath: [{
                type: Input
            }], volumeMemberPath: [{
                type: Input
            }], highlightedOpenMemberPath: [{
                type: Input
            }], highlightedHighMemberPath: [{
                type: Input
            }], highlightedLowMemberPath: [{
                type: Input
            }], highlightedCloseMemberPath: [{
                type: Input
            }], highlightedVolumeMemberPath: [{
                type: Input
            }], isCustomCategoryStyleAllowed: [{
                type: Input
            }], transitionInMode: [{
                type: Input
            }], isTransitionInEnabled: [{
                type: Input
            }], assigningCategoryStyle: [{
                type: Output
            }], typical: [{
                type: Output
            }], typicalBasedOn: [{
                type: Output
            }] } });
