// igniteui-angular-charts/esm2015/lib/igx-polar-scatter-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxPolarScatterSeriesComponent } from './igx-polar-scatter-series-component';
import * as i0 from "@angular/core";
export let IgxPolarScatterSeriesModule = /*@__PURE__*/ (() => {
    class IgxPolarScatterSeriesModule {
    }
    IgxPolarScatterSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarScatterSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPolarScatterSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarScatterSeriesModule, declarations: [IgxPolarScatterSeriesComponent], imports: [CommonModule], exports: [IgxPolarScatterSeriesComponent] });
    IgxPolarScatterSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarScatterSeriesModule, imports: [[CommonModule]] });
    return IgxPolarScatterSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarScatterSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxPolarScatterSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxPolarScatterSeriesComponent],
                    entryComponents: []
                }]
        }] });
