// igniteui-angular-charts/esm2015/lib/igx-line-fragment-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxFragmentBaseComponent } from "./igx-fragment-base-component";
import { IgxHorizontalAnchoredCategorySeriesComponent } from "./igx-horizontal-anchored-category-series-component";
import { IgxAnchoredCategorySeriesComponent } from "./igx-anchored-category-series-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { LineFragment } from "./LineFragment";
import * as i0 from "@angular/core";
/**
 * Represents one part of a StackedLineSeries.
*/
export let IgxLineFragmentComponent = /*@__PURE__*/ (() => {
    class IgxLineFragmentComponent extends IgxFragmentBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new LineFragment();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows only line shapes.
        */
        get isLineOnly() {
            return this.i.fi;
        }
        /**
         * Gets whether the current series shows an area or line shape.
        */
        get isAreaOrLine() {
            return this.i.et;
        }
    }
    IgxLineFragmentComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineFragmentComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxLineFragmentComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxLineFragmentComponent, selector: "igx-line-fragment", providers: [{ provide: IgxFragmentBaseComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }, { provide: IgxHorizontalAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }, { provide: IgxAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxLineFragmentComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLineFragmentComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-line-fragment',
                    template: ``,
                    providers: [{ provide: IgxFragmentBaseComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }, { provide: IgxHorizontalAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }, { provide: IgxAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxLineFragmentComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
