// igniteui-angular-charts/esm2015/lib/igx-scale-legend-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxLegendBaseComponent } from "./igx-legend-base-component";
import { ScaleLegend } from "./ScaleLegend";
import * as i0 from "@angular/core";
/**
 * Represents a legend that indicates the size and the color scale for a collection of series.
*/
export let IgxScaleLegendComponent = /*@__PURE__*/ (() => {
    class IgxScaleLegendComponent extends IgxLegendBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new ScaleLegend();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
    }
    IgxScaleLegendComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxScaleLegendComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxScaleLegendComponent, selector: "igx-scale-legend", providers: [{ provide: IgxLegendBaseComponent, useExisting: forwardRef(() => IgxScaleLegendComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxScaleLegendComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxScaleLegendComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-scale-legend',
                    template: ``,
                    providers: [{ provide: IgxLegendBaseComponent, useExisting: forwardRef(() => IgxScaleLegendComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
