// igniteui-angular-charts/esm2015/lib/igx-mass-index-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxMassIndexIndicatorModule } from './igx-mass-index-indicator-module';
import { IgxMassIndexIndicatorComponent } from './igx-mass-index-indicator-component';
import { MassIndexIndicator } from './MassIndexIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxMassIndexIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxMassIndexIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxMassIndexIndicatorComponent", IgxMassIndexIndicatorComponent);
            TypeRegistrar.register("MassIndexIndicator", MassIndexIndicator.$type);
        }
    }
    IgxMassIndexIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMassIndexIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxMassIndexIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMassIndexIndicatorDynamicModule, imports: [CommonModule, IgxMassIndexIndicatorModule], exports: [IgxMassIndexIndicatorModule] });
    IgxMassIndexIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMassIndexIndicatorDynamicModule, imports: [[CommonModule, IgxMassIndexIndicatorModule], IgxMassIndexIndicatorModule] });
    return IgxMassIndexIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxMassIndexIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxMassIndexIndicatorModule],
                    exports: [IgxMassIndexIndicatorModule],
                    entryComponents: [IgxMassIndexIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
