// igniteui-angular-charts/esm2015/lib/igx-callout-layer-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCalloutLayerModule } from './igx-callout-layer-module';
import { IgxCalloutLayerComponent } from './igx-callout-layer-component';
import { CalloutLayer } from './CalloutLayer';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxCalloutLayerDynamicModule = /*@__PURE__*/ (() => {
    class IgxCalloutLayerDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxCalloutLayerComponent", IgxCalloutLayerComponent);
            TypeRegistrar.register("CalloutLayer", CalloutLayer.$type);
        }
    }
    IgxCalloutLayerDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCalloutLayerDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCalloutLayerDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCalloutLayerDynamicModule, imports: [CommonModule, IgxCalloutLayerModule], exports: [IgxCalloutLayerModule] });
    IgxCalloutLayerDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCalloutLayerDynamicModule, imports: [[CommonModule, IgxCalloutLayerModule], IgxCalloutLayerModule] });
    return IgxCalloutLayerDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCalloutLayerDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCalloutLayerModule],
                    exports: [IgxCalloutLayerModule],
                    entryComponents: [IgxCalloutLayerComponent]
                }]
        }], ctorParameters: function () { return []; } });
