// igniteui-angular-charts/esm2015/lib/igx-radial-area-series-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { UnknownValuePlotting_$type } from "igniteui-angular-core";
import { IgxAnchoredRadialSeriesComponent } from "./igx-anchored-radial-series-component";
import { IgxRadialBaseComponent } from "./igx-radial-base-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { RadialAreaSeries } from "./RadialAreaSeries";
import { ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent radial area series.
*/
export let IgxRadialAreaSeriesComponent = /*@__PURE__*/ (() => {
    class IgxRadialAreaSeriesComponent extends IgxAnchoredRadialSeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new RadialAreaSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows an area shape.
        */
        get isArea() {
            return this.i.es;
        }
        /**
         * Determines how unknown values will be plotted on the chart.
         * Null and Double.NaN are two examples of unknown values.
        */
        get unknownValuePlotting() {
            return this.i.ade;
        }
        set unknownValuePlotting(v) {
            this.i.ade = ensureEnum(UnknownValuePlotting_$type, v);
        }
    }
    IgxRadialAreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialAreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxRadialAreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxRadialAreaSeriesComponent, selector: "igx-radial-area-series", inputs: { unknownValuePlotting: "unknownValuePlotting" }, providers: [{ provide: IgxAnchoredRadialSeriesComponent, useExisting: forwardRef(() => IgxRadialAreaSeriesComponent) }, { provide: IgxRadialBaseComponent, useExisting: forwardRef(() => IgxRadialAreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxRadialAreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxRadialAreaSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxRadialAreaSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxRadialAreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-radial-area-series',
                    template: ``,
                    providers: [{ provide: IgxAnchoredRadialSeriesComponent, useExisting: forwardRef(() => IgxRadialAreaSeriesComponent) }, { provide: IgxRadialBaseComponent, useExisting: forwardRef(() => IgxRadialAreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxRadialAreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxRadialAreaSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { unknownValuePlotting: [{
                type: Input
            }] } });
