// igniteui-angular-charts/esm2015/lib/igx-custom-contour-value-resolver-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCustomContourValueResolverModule } from './igx-custom-contour-value-resolver-module';
import { IgxCustomContourValueResolverComponent } from './igx-custom-contour-value-resolver-component';
import { CustomContourValueResolver } from './CustomContourValueResolver';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxCustomContourValueResolverDynamicModule = /*@__PURE__*/ (() => {
    class IgxCustomContourValueResolverDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxCustomContourValueResolverComponent", IgxCustomContourValueResolverComponent);
            TypeRegistrar.register("CustomContourValueResolver", CustomContourValueResolver.$type);
        }
    }
    IgxCustomContourValueResolverDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomContourValueResolverDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCustomContourValueResolverDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomContourValueResolverDynamicModule, imports: [CommonModule, IgxCustomContourValueResolverModule], exports: [IgxCustomContourValueResolverModule] });
    IgxCustomContourValueResolverDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomContourValueResolverDynamicModule, imports: [[CommonModule, IgxCustomContourValueResolverModule], IgxCustomContourValueResolverModule] });
    return IgxCustomContourValueResolverDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomContourValueResolverDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCustomContourValueResolverModule],
                    exports: [IgxCustomContourValueResolverModule],
                    entryComponents: [IgxCustomContourValueResolverComponent]
                }]
        }], ctorParameters: function () { return []; } });
