// igniteui-angular-charts/esm2015/lib/igx-data-chart-scatter-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSizeScaleModule } from './igx-size-scale-module';
import { IgxValueBrushScaleModule } from './igx-value-brush-scale-module';
import { IgxCustomPaletteColorScaleModule } from './igx-custom-palette-color-scale-module';
import { IgxBubbleSeriesModule } from './igx-bubble-series-module';
import { IgxScatterSeriesModule } from './igx-scatter-series-module';
import { IgxScatterLineSeriesModule } from './igx-scatter-line-series-module';
import { IgxScatterSplineSeriesModule } from './igx-scatter-spline-series-module';
import { IgxHighDensityScatterSeriesModule } from './igx-high-density-scatter-series-module';
import { IgxDataChartScatterCoreModule } from './igx-data-chart-scatter-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartScatterModule = /*@__PURE__*/ (() => {
    class IgxDataChartScatterModule {
    }
    IgxDataChartScatterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartScatterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterModule, imports: [CommonModule, IgxSizeScaleModule,
            IgxValueBrushScaleModule,
            IgxCustomPaletteColorScaleModule,
            IgxBubbleSeriesModule,
            IgxScatterSeriesModule,
            IgxScatterLineSeriesModule,
            IgxScatterSplineSeriesModule,
            IgxHighDensityScatterSeriesModule,
            IgxDataChartScatterCoreModule], exports: [IgxSizeScaleModule,
            IgxValueBrushScaleModule,
            IgxCustomPaletteColorScaleModule,
            IgxBubbleSeriesModule,
            IgxScatterSeriesModule,
            IgxScatterLineSeriesModule,
            IgxScatterSplineSeriesModule,
            IgxHighDensityScatterSeriesModule,
            IgxDataChartScatterCoreModule] });
    IgxDataChartScatterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterModule, imports: [[CommonModule, IgxSizeScaleModule,
                IgxValueBrushScaleModule,
                IgxCustomPaletteColorScaleModule,
                IgxBubbleSeriesModule,
                IgxScatterSeriesModule,
                IgxScatterLineSeriesModule,
                IgxScatterSplineSeriesModule,
                IgxHighDensityScatterSeriesModule,
                IgxDataChartScatterCoreModule
            ], IgxSizeScaleModule,
            IgxValueBrushScaleModule,
            IgxCustomPaletteColorScaleModule,
            IgxBubbleSeriesModule,
            IgxScatterSeriesModule,
            IgxScatterLineSeriesModule,
            IgxScatterSplineSeriesModule,
            IgxHighDensityScatterSeriesModule,
            IgxDataChartScatterCoreModule] });
    return IgxDataChartScatterModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartScatterModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxSizeScaleModule,
                        IgxValueBrushScaleModule,
                        IgxCustomPaletteColorScaleModule,
                        IgxBubbleSeriesModule,
                        IgxScatterSeriesModule,
                        IgxScatterLineSeriesModule,
                        IgxScatterSplineSeriesModule,
                        IgxHighDensityScatterSeriesModule,
                        IgxDataChartScatterCoreModule
                    ],
                    imports: [CommonModule, IgxSizeScaleModule,
                        IgxValueBrushScaleModule,
                        IgxCustomPaletteColorScaleModule,
                        IgxBubbleSeriesModule,
                        IgxScatterSeriesModule,
                        IgxScatterLineSeriesModule,
                        IgxScatterSplineSeriesModule,
                        IgxHighDensityScatterSeriesModule,
                        IgxDataChartScatterCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
