// igniteui-angular-charts/esm2015/lib/igx-bar-fragment-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxBarFragmentModule } from './igx-bar-fragment-module';
import { IgxBarFragmentComponent } from './igx-bar-fragment-component';
import { BarFragment } from './BarFragment';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxBarFragmentDynamicModule = /*@__PURE__*/ (() => {
    class IgxBarFragmentDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxBarFragmentComponent", IgxBarFragmentComponent);
            TypeRegistrar.register("BarFragment", BarFragment.$type);
        }
    }
    IgxBarFragmentDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarFragmentDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxBarFragmentDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarFragmentDynamicModule, imports: [CommonModule, IgxBarFragmentModule], exports: [IgxBarFragmentModule] });
    IgxBarFragmentDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarFragmentDynamicModule, imports: [[CommonModule, IgxBarFragmentModule], IgxBarFragmentModule] });
    return IgxBarFragmentDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxBarFragmentDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxBarFragmentModule],
                    exports: [IgxBarFragmentModule],
                    entryComponents: [IgxBarFragmentComponent]
                }]
        }], ctorParameters: function () { return []; } });
