// igniteui-angular-charts/esm2015/lib/igx-pie-chart-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumberFormatSpecifierDynamicModule } from "igniteui-angular-core";
import { IgxDateTimeFormatSpecifierDynamicModule } from "igniteui-angular-core";
import { IgxPieChartComponent } from './igx-pie-chart-component';
import { FastItemObjectColumn } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxPieChartCoreModule } from './igx-pie-chart-core-module';
import * as i0 from "@angular/core";
export let IgxPieChartCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxPieChartCoreDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('IgxPieChartComponent', IgxPieChartComponent);
            TypeRegistrar.register('FastItemObjectColumn', FastItemObjectColumn.$type);
        }
    }
    IgxPieChartCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPieChartCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartCoreDynamicModule, imports: [CommonModule, IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxPieChartCoreModule], exports: [IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxPieChartCoreModule] });
    IgxPieChartCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartCoreDynamicModule, imports: [[CommonModule, IgxNumberFormatSpecifierDynamicModule,
                IgxDateTimeFormatSpecifierDynamicModule,
                IgxPieChartCoreModule
            ], IgxNumberFormatSpecifierDynamicModule,
            IgxDateTimeFormatSpecifierDynamicModule,
            IgxPieChartCoreModule] });
    return IgxPieChartCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPieChartCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxNumberFormatSpecifierDynamicModule,
                        IgxDateTimeFormatSpecifierDynamicModule,
                        IgxPieChartCoreModule
                    ],
                    imports: [CommonModule, IgxNumberFormatSpecifierDynamicModule,
                        IgxDateTimeFormatSpecifierDynamicModule,
                        IgxPieChartCoreModule
                    ],
                    entryComponents: [IgxPieChartComponent]
                }]
        }], ctorParameters: function () { return []; } });
