// igniteui-angular-charts/esm2015/lib/igx-treemap-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTreemapComponent } from './igx-treemap-component';
import { IgxTreemapNodeStyleDynamicModule } from './igx-treemap-node-style-dynamic-module';
import { IgxTreemapNodeStyleMappingDynamicModule } from './igx-treemap-node-style-mapping-dynamic-module';
import { Treemap } from './Treemap';
import { FastItemObjectColumn } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxTreemapModule } from './igx-treemap-module';
import * as i0 from "@angular/core";
export let IgxTreemapDynamicModule = /*@__PURE__*/ (() => {
    class IgxTreemapDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('Treemap', Treemap);
            TypeRegistrar.registerCons('IgxTreemapComponent', IgxTreemapComponent);
            TypeRegistrar.register('Treemap', Treemap.$type);
            TypeRegistrar.register('FastItemObjectColumn', FastItemObjectColumn.$type);
        }
    }
    IgxTreemapDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTreemapDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapDynamicModule, imports: [CommonModule, IgxTreemapNodeStyleDynamicModule,
            IgxTreemapNodeStyleMappingDynamicModule,
            IgxTreemapModule], exports: [IgxTreemapNodeStyleDynamicModule,
            IgxTreemapNodeStyleMappingDynamicModule,
            IgxTreemapModule] });
    IgxTreemapDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapDynamicModule, imports: [[CommonModule, IgxTreemapNodeStyleDynamicModule,
                IgxTreemapNodeStyleMappingDynamicModule,
                IgxTreemapModule
            ], IgxTreemapNodeStyleDynamicModule,
            IgxTreemapNodeStyleMappingDynamicModule,
            IgxTreemapModule] });
    return IgxTreemapDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTreemapDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxTreemapNodeStyleDynamicModule,
                        IgxTreemapNodeStyleMappingDynamicModule,
                        IgxTreemapModule
                    ],
                    imports: [CommonModule, IgxTreemapNodeStyleDynamicModule,
                        IgxTreemapNodeStyleMappingDynamicModule,
                        IgxTreemapModule
                    ],
                    entryComponents: [IgxTreemapComponent]
                }]
        }], ctorParameters: function () { return []; } });
