// igniteui-angular-charts/esm2015/lib/igx-custom-palette-color-scale-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCustomPaletteColorScaleModule } from './igx-custom-palette-color-scale-module';
import { IgxCustomPaletteColorScaleComponent } from './igx-custom-palette-color-scale-component';
import { CustomPaletteColorScale } from './CustomPaletteColorScale';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxCustomPaletteColorScaleDynamicModule = /*@__PURE__*/ (() => {
    class IgxCustomPaletteColorScaleDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxCustomPaletteColorScaleComponent", IgxCustomPaletteColorScaleComponent);
            TypeRegistrar.register("CustomPaletteColorScale", CustomPaletteColorScale.$type);
        }
    }
    IgxCustomPaletteColorScaleDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteColorScaleDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCustomPaletteColorScaleDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteColorScaleDynamicModule, imports: [CommonModule, IgxCustomPaletteColorScaleModule], exports: [IgxCustomPaletteColorScaleModule] });
    IgxCustomPaletteColorScaleDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteColorScaleDynamicModule, imports: [[CommonModule, IgxCustomPaletteColorScaleModule], IgxCustomPaletteColorScaleModule] });
    return IgxCustomPaletteColorScaleDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCustomPaletteColorScaleDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCustomPaletteColorScaleModule],
                    exports: [IgxCustomPaletteColorScaleModule],
                    entryComponents: [IgxCustomPaletteColorScaleComponent]
                }]
        }], ctorParameters: function () { return []; } });
