// igniteui-angular-charts/esm2015/lib/igx-spline-area-fragment-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxSplineFragmentBaseComponent } from "./igx-spline-fragment-base-component";
import { IgxSplineSeriesBaseComponent } from "./igx-spline-series-base-component";
import { IgxHorizontalAnchoredCategorySeriesComponent } from "./igx-horizontal-anchored-category-series-component";
import { IgxAnchoredCategorySeriesComponent } from "./igx-anchored-category-series-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { SplineAreaFragment } from "./SplineAreaFragment";
import * as i0 from "@angular/core";
/**
 * Represents one part of a StackedSplineAreaSeries.
*/
export let IgxSplineAreaFragmentComponent = /*@__PURE__*/ (() => {
    class IgxSplineAreaFragmentComponent extends IgxSplineFragmentBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new SplineAreaFragment();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series shows an area or line shape.
        */
        get isAreaOrLine() {
            return this.i.et;
        }
        /**
         * Gets whether the current series shows an area shape.
        */
        get isArea() {
            return this.i.es;
        }
        /**
         * Gets whether the current series shows a spline shape.
        */
        get isSpline() {
            return this.i.fy;
        }
    }
    IgxSplineAreaFragmentComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineAreaFragmentComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxSplineAreaFragmentComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxSplineAreaFragmentComponent, selector: "igx-spline-area-fragment", providers: [{ provide: IgxSplineFragmentBaseComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxSplineSeriesBaseComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxHorizontalAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxSplineAreaFragmentComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSplineAreaFragmentComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-spline-area-fragment',
                    template: ``,
                    providers: [{ provide: IgxSplineFragmentBaseComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxSplineSeriesBaseComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxHorizontalAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxAnchoredCategorySeriesComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxSplineAreaFragmentComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
