// igniteui-angular-charts/esm2015/lib/igx-polar-line-series-base-component.js
import { Component } from '@angular/core';
import { IgxPolarBaseComponent } from "./igx-polar-base-component";
import * as i0 from "@angular/core";
/**
 * Represents the base class from which all IgxDataChartComponent polar line series are derived.
*/
export let IgxPolarLineSeriesBaseComponent = /*@__PURE__*/ (() => {
    class IgxPolarLineSeriesBaseComponent extends IgxPolarBaseComponent {
        constructor() {
            super();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
    }
    IgxPolarLineSeriesBaseComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesBaseComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxPolarLineSeriesBaseComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxPolarLineSeriesBaseComponent, selector: "ng-component", usesInheritance: true, ngImport: i0, template: ``, isInline: true });
    return IgxPolarLineSeriesBaseComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPolarLineSeriesBaseComponent, decorators: [{
            type: Component,
            args: [{
                    template: ``,
                }]
        }], ctorParameters: function () { return []; } });
