// igniteui-angular-charts/esm2015/lib/igx-numeric-angle-axis-component.js
import { Component, forwardRef, Input, ChangeDetectionStrategy } from '@angular/core';
import { AxisAngleLabelMode_$type } from "./AxisAngleLabelMode";
import { IgxNumericAxisBaseComponent } from "./igx-numeric-axis-base-component";
import { IgxAxisComponent } from "./igx-axis-component";
import { NumericAngleAxis } from "./NumericAngleAxis";
import { ensureEnum } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent angle based axis for polar series.
*/
export let IgxNumericAngleAxisComponent = /*@__PURE__*/ (() => {
    class IgxNumericAngleAxisComponent extends IgxNumericAxisBaseComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new NumericAngleAxis();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets if the current axis is of angular axis type
        */
        get isAngular() {
            return this.i.b9;
        }
        /**
         * Indicates the angle in degress that the chart's 0th angle should be offset.
        */
        get startAngleOffset() {
            return this.i.nu;
        }
        set startAngleOffset(v) {
            this.i.nu = +v;
        }
        /**
         * Indicates the mode axis labels will operate in.
        */
        get labelMode() {
            return this.i.ni;
        }
        set labelMode(v) {
            this.i.ni = ensureEnum(AxisAngleLabelMode_$type, v);
        }
        /**
         * Gets the scaled angle value in radians based on the raw input.
        
        * @param unscaledAngle  * The unscaled angle.
        */
        getScaledAngle(unscaledAngle) {
            let iv = this.i.getScaledAngle(unscaledAngle);
            return (iv);
        }
        /**
         * Gets the raw axis value back from the angle that would be used on the chart.
        
        * @param scaledAngle  * The chart angle value.
        */
        getUnscaledAngle(scaledAngle) {
            let iv = this.i.getUnscaledAngle(scaledAngle);
            return (iv);
        }
    }
    IgxNumericAngleAxisComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAngleAxisComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxNumericAngleAxisComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxNumericAngleAxisComponent, selector: "igx-numeric-angle-axis", inputs: { startAngleOffset: "startAngleOffset", labelMode: "labelMode" }, providers: [{ provide: IgxNumericAxisBaseComponent, useExisting: forwardRef(() => IgxNumericAngleAxisComponent) }, { provide: IgxAxisComponent, useExisting: forwardRef(() => IgxNumericAngleAxisComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxNumericAngleAxisComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumericAngleAxisComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-numeric-angle-axis',
                    template: ``,
                    providers: [{ provide: IgxNumericAxisBaseComponent, useExisting: forwardRef(() => IgxNumericAngleAxisComponent) }, { provide: IgxAxisComponent, useExisting: forwardRef(() => IgxNumericAngleAxisComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { startAngleOffset: [{
                type: Input
            }], labelMode: [{
                type: Input
            }] } });
