// igniteui-angular-charts/esm2015/lib/igx-size-scale-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxSizeScaleComponent } from './igx-size-scale-component';
import * as i0 from "@angular/core";
export let IgxSizeScaleModule = /*@__PURE__*/ (() => {
    class IgxSizeScaleModule {
    }
    IgxSizeScaleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxSizeScaleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleModule, declarations: [IgxSizeScaleComponent], imports: [CommonModule], exports: [IgxSizeScaleComponent] });
    IgxSizeScaleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleModule, imports: [[CommonModule]] });
    return IgxSizeScaleModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxSizeScaleModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxSizeScaleComponent],
                    imports: [CommonModule],
                    exports: [IgxSizeScaleComponent],
                    entryComponents: []
                }]
        }] });
