// igniteui-angular-charts/esm2015/lib/igx-stacked-100-spline-area-series-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStackedSplineAreaSeriesComponent } from "./igx-stacked-spline-area-series-component";
import { IgxHorizontalStackedSeriesBaseComponent } from "./igx-horizontal-stacked-series-base-component";
import { IgxStackedSeriesBaseComponent } from "./igx-stacked-series-base-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { Stacked100SplineAreaSeries } from "./Stacked100SplineAreaSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent stacked100 spline area series.
*/
export let IgxStacked100SplineAreaSeriesComponent = /*@__PURE__*/ (() => {
    class IgxStacked100SplineAreaSeriesComponent extends IgxStackedSplineAreaSeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new Stacked100SplineAreaSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        get isPercentBased() {
            return this.i.aan;
        }
    }
    IgxStacked100SplineAreaSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineAreaSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStacked100SplineAreaSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStacked100SplineAreaSeriesComponent, selector: "igx-stacked-100-spline-area-series", providers: [{ provide: IgxStackedSplineAreaSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStacked100SplineAreaSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineAreaSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-stacked-100-spline-area-series',
                    template: ``,
                    providers: [{ provide: IgxStackedSplineAreaSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineAreaSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
