// igniteui-angular-charts/esm2015/lib/igx-commodity-channel-index-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCommodityChannelIndexIndicatorModule } from './igx-commodity-channel-index-indicator-module';
import { IgxCommodityChannelIndexIndicatorComponent } from './igx-commodity-channel-index-indicator-component';
import { CommodityChannelIndexIndicator } from './CommodityChannelIndexIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxCommodityChannelIndexIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxCommodityChannelIndexIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxCommodityChannelIndexIndicatorComponent", IgxCommodityChannelIndexIndicatorComponent);
            TypeRegistrar.register("CommodityChannelIndexIndicator", CommodityChannelIndexIndicator.$type);
        }
    }
    IgxCommodityChannelIndexIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCommodityChannelIndexIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCommodityChannelIndexIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCommodityChannelIndexIndicatorDynamicModule, imports: [CommonModule, IgxCommodityChannelIndexIndicatorModule], exports: [IgxCommodityChannelIndexIndicatorModule] });
    IgxCommodityChannelIndexIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCommodityChannelIndexIndicatorDynamicModule, imports: [[CommonModule, IgxCommodityChannelIndexIndicatorModule], IgxCommodityChannelIndexIndicatorModule] });
    return IgxCommodityChannelIndexIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCommodityChannelIndexIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxCommodityChannelIndexIndicatorModule],
                    exports: [IgxCommodityChannelIndexIndicatorModule],
                    entryComponents: [IgxCommodityChannelIndexIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
