// igniteui-angular-charts/esm2015/lib/igx-financial-price-series-proxy-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FinancialPriceSeriesProxy } from './FinancialPriceSeriesProxy';
import { FinancialEventArgs as FinancialEventArgs_internal } from './FinancialEventArgs';
import { IgxFinancialEventArgs } from './igx-financial-event-args';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxFinancialPriceSeriesProxyDynamicModule = /*@__PURE__*/ (() => {
    class IgxFinancialPriceSeriesProxyDynamicModule {
        constructor() {
            TypeRegistrar.register('FinancialPriceSeriesProxy', FinancialPriceSeriesProxy.$type);
            TypeRegistrar.register('FinancialEventArgs', FinancialEventArgs_internal.$type);
            TypeRegistrar.registerCons('igx-financial-event-args', IgxFinancialEventArgs);
        }
    }
    IgxFinancialPriceSeriesProxyDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesProxyDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFinancialPriceSeriesProxyDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesProxyDynamicModule, imports: [CommonModule] });
    IgxFinancialPriceSeriesProxyDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesProxyDynamicModule, imports: [[CommonModule,
            ]] });
    return IgxFinancialPriceSeriesProxyDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialPriceSeriesProxyDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
