// igniteui-angular-charts/esm2015/lib/igx-trix-indicator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxIndicatorProxyModule } from './igx-indicator-proxy-module';
import { IgxTRIXIndicatorComponent } from './igx-trix-indicator-component';
import * as i0 from "@angular/core";
export let IgxTRIXIndicatorModule = /*@__PURE__*/ (() => {
    class IgxTRIXIndicatorModule {
    }
    IgxTRIXIndicatorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTRIXIndicatorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTRIXIndicatorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTRIXIndicatorModule, declarations: [IgxTRIXIndicatorComponent], imports: [CommonModule, IgxIndicatorProxyModule], exports: [IgxTRIXIndicatorComponent, IgxIndicatorProxyModule] });
    IgxTRIXIndicatorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTRIXIndicatorModule, imports: [[CommonModule, IgxIndicatorProxyModule], IgxIndicatorProxyModule] });
    return IgxTRIXIndicatorModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTRIXIndicatorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxTRIXIndicatorComponent],
                    imports: [CommonModule, IgxIndicatorProxyModule],
                    exports: [IgxTRIXIndicatorComponent, IgxIndicatorProxyModule],
                    entryComponents: []
                }]
        }] });
