// igniteui-angular-charts/esm2015/lib/igx-stacked-100-column-series-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStackedColumnSeriesComponent } from "./igx-stacked-column-series-component";
import { IgxHorizontalStackedSeriesBaseComponent } from "./igx-horizontal-stacked-series-base-component";
import { IgxStackedSeriesBaseComponent } from "./igx-stacked-series-base-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { Stacked100ColumnSeries } from "./Stacked100ColumnSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent stacked100 column series.
*/
export let IgxStacked100ColumnSeriesComponent = /*@__PURE__*/ (() => {
    class IgxStacked100ColumnSeriesComponent extends IgxStackedColumnSeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new Stacked100ColumnSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        get isPercentBased() {
            return this.i.aan;
        }
    }
    IgxStacked100ColumnSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100ColumnSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStacked100ColumnSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStacked100ColumnSeriesComponent, selector: "igx-stacked-100-column-series", providers: [{ provide: IgxStackedColumnSeriesComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStacked100ColumnSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100ColumnSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-stacked-100-column-series',
                    template: ``,
                    providers: [{ provide: IgxStackedColumnSeriesComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100ColumnSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
