// igniteui-angular-charts/esm2015/lib/igx-fragment-base-component.js
import { Component } from '@angular/core';
import { IgxHorizontalAnchoredCategorySeriesComponent } from "./igx-horizontal-anchored-category-series-component";
import { toPoint, fromPoint } from "igniteui-angular-core";
import * as i0 from "@angular/core";
/**
 * Represents one part of a stacked series.
*/
export let IgxFragmentBaseComponent = /*@__PURE__*/ (() => {
    class IgxFragmentBaseComponent extends IgxHorizontalAnchoredCategorySeriesComponent {
        constructor() {
            super();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        /**
         * Gets whether the current series is a stack fragment.
        */
        get isFragment() {
            return this.i.e9;
        }
        getSeriesValuePosition(world, useInterpolation, skipUnknowns) {
            let iv = this.i.v6(toPoint(world), useInterpolation, skipUnknowns);
            return fromPoint(iv);
        }
    }
    IgxFragmentBaseComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFragmentBaseComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxFragmentBaseComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxFragmentBaseComponent, selector: "ng-component", usesInheritance: true, ngImport: i0, template: ``, isInline: true });
    return IgxFragmentBaseComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFragmentBaseComponent, decorators: [{
            type: Component,
            args: [{
                    template: ``,
                }]
        }], ctorParameters: function () { return []; } });
