// igniteui-angular-charts/esm2015/lib/igx-typical-price-indicator-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStrategyBasedIndicatorComponent } from "./igx-strategy-based-indicator-component";
import { IgxFinancialIndicatorComponent } from "./igx-financial-indicator-component";
import { IgxFinancialSeriesComponent } from "./igx-financial-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { TypicalPriceIndicator } from "./TypicalPriceIndicator";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent Typical Price indicator series.
 * The typical price indicator is represented as a arithmetic average
 * of the High, Low, and Close for a day.
 * Default required members: High, Low, Close
*
* The `TypicalPriceIndicator` class represents a IgxDataChartComponent Typical Price indicator series.
*
* ```html
* <igx-data-chart
*     [dataSource]="data">
* 	<igx-category-x-axis
* 		label="label"
* 		#xAxis>
*     </igx-category-x-axis>
*     <igx-numeric-y-axis
* 		#yAxis>
*     </igx-numeric-y-axis>
* 	<igx-typical-price-indicator
*         #series
*         [xAxis]="xAxis"
*         [yAxis]="yAxis"
*         highMemberPath="high"
* 		lowMemberPath="low"
* 		volumeMemberPath="volume">
*     </igx-typical-price-indicator>
* </igx-data-chart>
* ```
*
* ```ts
* let series = new IgxTypicalPriceIndicatorComponent();
* series.xAxis = this.xAxis;
* series.yAxis = this.yAxis;
* series.openMemberPath = "open";
* series.highMemberPath = "high";
* series.lowMemberPath = "low";
* series.closeMemberPath = "close";
* this.chart.series.add(series);
* ```
*/
export let IgxTypicalPriceIndicatorComponent = /*@__PURE__*/ (() => {
    class IgxTypicalPriceIndicatorComponent extends IgxStrategyBasedIndicatorComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new TypicalPriceIndicator();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
    }
    IgxTypicalPriceIndicatorComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxTypicalPriceIndicatorComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxTypicalPriceIndicatorComponent, selector: "igx-typical-price-indicator", providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(() => IgxTypicalPriceIndicatorComponent) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(() => IgxTypicalPriceIndicatorComponent) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(() => IgxTypicalPriceIndicatorComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxTypicalPriceIndicatorComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxTypicalPriceIndicatorComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-typical-price-indicator',
                    template: ``,
                    providers: [{ provide: IgxStrategyBasedIndicatorComponent, useExisting: forwardRef(() => IgxTypicalPriceIndicatorComponent) }, { provide: IgxFinancialIndicatorComponent, useExisting: forwardRef(() => IgxTypicalPriceIndicatorComponent) }, { provide: IgxFinancialSeriesComponent, useExisting: forwardRef(() => IgxTypicalPriceIndicatorComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxTypicalPriceIndicatorComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
