// igniteui-angular-charts/esm2015/lib/igx-data-pie-chart-core-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataPieChartComponent } from './igx-data-pie-chart-component';
import { IgxProportionalCategoryAngleAxisDynamicModule } from './igx-proportional-category-angle-axis-dynamic-module';
import { IgxNumericRadiusAxisDynamicModule } from './igx-numeric-radius-axis-dynamic-module';
import { FastItemObjectColumn } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDataPieChartCoreModule } from './igx-data-pie-chart-core-module';
import * as i0 from "@angular/core";
export let IgxDataPieChartCoreDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataPieChartCoreDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('IgxDataPieChartComponent', IgxDataPieChartComponent);
            TypeRegistrar.register('FastItemObjectColumn', FastItemObjectColumn.$type);
        }
    }
    IgxDataPieChartCoreDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartCoreDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataPieChartCoreDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartCoreDynamicModule, imports: [CommonModule, IgxProportionalCategoryAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule,
            IgxDataPieChartCoreModule], exports: [IgxProportionalCategoryAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule,
            IgxDataPieChartCoreModule] });
    IgxDataPieChartCoreDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartCoreDynamicModule, imports: [[CommonModule, IgxProportionalCategoryAngleAxisDynamicModule,
                IgxNumericRadiusAxisDynamicModule,
                IgxDataPieChartCoreModule
            ], IgxProportionalCategoryAngleAxisDynamicModule,
            IgxNumericRadiusAxisDynamicModule,
            IgxDataPieChartCoreModule] });
    return IgxDataPieChartCoreDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartCoreDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxProportionalCategoryAngleAxisDynamicModule,
                        IgxNumericRadiusAxisDynamicModule,
                        IgxDataPieChartCoreModule
                    ],
                    imports: [CommonModule, IgxProportionalCategoryAngleAxisDynamicModule,
                        IgxNumericRadiusAxisDynamicModule,
                        IgxDataPieChartCoreModule
                    ],
                    entryComponents: [IgxDataPieChartComponent]
                }]
        }], ctorParameters: function () { return []; } });
