// igniteui-angular-charts/esm2015/lib/igx-overlay-proxy-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OverlayProxy } from './OverlayProxy';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxOverlayProxyModule } from './igx-overlay-proxy-module';
import * as i0 from "@angular/core";
export let IgxOverlayProxyDynamicModule = /*@__PURE__*/ (() => {
    class IgxOverlayProxyDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('OverlayProxy', OverlayProxy);
            TypeRegistrar.register('OverlayProxy', OverlayProxy.$type);
        }
    }
    IgxOverlayProxyDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOverlayProxyDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxOverlayProxyDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOverlayProxyDynamicModule, imports: [CommonModule, IgxOverlayProxyModule], exports: [IgxOverlayProxyModule] });
    IgxOverlayProxyDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOverlayProxyDynamicModule, imports: [[CommonModule, IgxOverlayProxyModule
            ], IgxOverlayProxyModule] });
    return IgxOverlayProxyDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxOverlayProxyDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxOverlayProxyModule
                    ],
                    imports: [CommonModule, IgxOverlayProxyModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
