// igniteui-angular-charts/esm2015/lib/igx-stacked-100-bar-series-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStacked100BarSeriesModule } from './igx-stacked-100-bar-series-module';
import { IgxStacked100BarSeriesComponent } from './igx-stacked-100-bar-series-component';
import { Stacked100BarSeries } from './Stacked100BarSeries';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxStacked100BarSeriesDynamicModule = /*@__PURE__*/ (() => {
    class IgxStacked100BarSeriesDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxStacked100BarSeriesComponent", IgxStacked100BarSeriesComponent);
            TypeRegistrar.register("Stacked100BarSeries", Stacked100BarSeries.$type);
        }
    }
    IgxStacked100BarSeriesDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStacked100BarSeriesDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesDynamicModule, imports: [CommonModule, IgxStacked100BarSeriesModule], exports: [IgxStacked100BarSeriesModule] });
    IgxStacked100BarSeriesDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesDynamicModule, imports: [[CommonModule, IgxStacked100BarSeriesModule], IgxStacked100BarSeriesModule] });
    return IgxStacked100BarSeriesDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100BarSeriesDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxStacked100BarSeriesModule],
                    exports: [IgxStacked100BarSeriesModule],
                    entryComponents: [IgxStacked100BarSeriesComponent]
                }]
        }], ctorParameters: function () { return []; } });
