// igniteui-angular-charts/esm2015/lib/igx-stacked-100-spline-series-component.js
import { Component, forwardRef, ChangeDetectionStrategy } from '@angular/core';
import { IgxStackedSplineSeriesComponent } from "./igx-stacked-spline-series-component";
import { IgxHorizontalStackedSeriesBaseComponent } from "./igx-horizontal-stacked-series-base-component";
import { IgxStackedSeriesBaseComponent } from "./igx-stacked-series-base-component";
import { IgxCategorySeriesComponent } from "./igx-category-series-component";
import { IgxMarkerSeriesComponent } from "./igx-marker-series-component";
import { IgxSeriesComponent } from "./igx-series-component";
import { Stacked100SplineSeries } from "./Stacked100SplineSeries";
import * as i0 from "@angular/core";
/**
 * Represents a IgxDataChartComponent stacked100 spline series.
*/
export let IgxStacked100SplineSeriesComponent = /*@__PURE__*/ (() => {
    class IgxStacked100SplineSeriesComponent extends IgxStackedSplineSeriesComponent {
        constructor() {
            super();
        }
        createImplementation() {
            return new Stacked100SplineSeries();
        }
        /**
                                     * @hidden
                                     */
        get i() {
            return this._implementation;
        }
        get isPercentBased() {
            return this.i.aan;
        }
    }
    IgxStacked100SplineSeriesComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    IgxStacked100SplineSeriesComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: IgxStacked100SplineSeriesComponent, selector: "igx-stacked-100-spline-series", providers: [{ provide: IgxStackedSplineSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }], usesInheritance: true, ngImport: i0, template: ``, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
    return IgxStacked100SplineSeriesComponent;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100SplineSeriesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-stacked-100-spline-series',
                    template: ``,
                    providers: [{ provide: IgxStackedSplineSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }, { provide: IgxHorizontalStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }, { provide: IgxStackedSeriesBaseComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }, { provide: IgxCategorySeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }, { provide: IgxMarkerSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }, { provide: IgxSeriesComponent, useExisting: forwardRef(() => IgxStacked100SplineSeriesComponent) }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], ctorParameters: function () { return []; } });
