// igniteui-angular-charts/esm2015/lib/igx-typical-price-indicator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTypicalPriceIndicatorModule } from './igx-typical-price-indicator-module';
import { IgxTypicalPriceIndicatorComponent } from './igx-typical-price-indicator-component';
import { TypicalPriceIndicator } from './TypicalPriceIndicator';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxTypicalPriceIndicatorDynamicModule = /*@__PURE__*/ (() => {
    class IgxTypicalPriceIndicatorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxTypicalPriceIndicatorComponent", IgxTypicalPriceIndicatorComponent);
            TypeRegistrar.register("TypicalPriceIndicator", TypicalPriceIndicator.$type);
        }
    }
    IgxTypicalPriceIndicatorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTypicalPriceIndicatorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorDynamicModule, imports: [CommonModule, IgxTypicalPriceIndicatorModule], exports: [IgxTypicalPriceIndicatorModule] });
    IgxTypicalPriceIndicatorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorDynamicModule, imports: [[CommonModule, IgxTypicalPriceIndicatorModule], IgxTypicalPriceIndicatorModule] });
    return IgxTypicalPriceIndicatorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTypicalPriceIndicatorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxTypicalPriceIndicatorModule],
                    exports: [IgxTypicalPriceIndicatorModule],
                    entryComponents: [IgxTypicalPriceIndicatorComponent]
                }]
        }], ctorParameters: function () { return []; } });
