// igniteui-angular-charts/esm2015/lib/igx-data-chart-vertical-category-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxBarSeriesDynamicModule } from './igx-bar-series-dynamic-module';
import { IgxDataChartVerticalCategoryCoreDynamicModule } from './igx-data-chart-vertical-category-core-dynamic-module';
import { IgxDataChartVerticalCategoryModule } from './igx-data-chart-vertical-category-module';
import * as i0 from "@angular/core";
export let IgxDataChartVerticalCategoryDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartVerticalCategoryDynamicModule {
    }
    IgxDataChartVerticalCategoryDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartVerticalCategoryDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryDynamicModule, imports: [CommonModule, IgxBarSeriesDynamicModule,
            IgxDataChartVerticalCategoryCoreDynamicModule,
            IgxDataChartVerticalCategoryModule], exports: [IgxBarSeriesDynamicModule,
            IgxDataChartVerticalCategoryCoreDynamicModule,
            IgxDataChartVerticalCategoryModule] });
    IgxDataChartVerticalCategoryDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryDynamicModule, imports: [[CommonModule, IgxBarSeriesDynamicModule,
                IgxDataChartVerticalCategoryCoreDynamicModule,
                IgxDataChartVerticalCategoryModule
            ], IgxBarSeriesDynamicModule,
            IgxDataChartVerticalCategoryCoreDynamicModule,
            IgxDataChartVerticalCategoryModule] });
    return IgxDataChartVerticalCategoryDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartVerticalCategoryDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxBarSeriesDynamicModule,
                        IgxDataChartVerticalCategoryCoreDynamicModule,
                        IgxDataChartVerticalCategoryModule
                    ],
                    imports: [CommonModule, IgxBarSeriesDynamicModule,
                        IgxDataChartVerticalCategoryCoreDynamicModule,
                        IgxDataChartVerticalCategoryModule
                    ],
                    entryComponents: []
                }]
        }] });
