// igniteui-angular-charts/esm2015/lib/igx-data-chart-shape-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxScatterAreaSeriesModule } from './igx-scatter-area-series-module';
import { IgxScatterContourSeriesModule } from './igx-scatter-contour-series-module';
import { IgxScatterPolygonSeriesModule } from './igx-scatter-polygon-series-module';
import { IgxScatterPolylineSeriesModule } from './igx-scatter-polyline-series-module';
import { IgxDataChartShapeCoreModule } from './igx-data-chart-shape-core-module';
import * as i0 from "@angular/core";
export let IgxDataChartShapeModule = /*@__PURE__*/ (() => {
    class IgxDataChartShapeModule {
    }
    IgxDataChartShapeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartShapeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartShapeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartShapeModule, imports: [CommonModule, IgxScatterAreaSeriesModule,
            IgxScatterContourSeriesModule,
            IgxScatterPolygonSeriesModule,
            IgxScatterPolylineSeriesModule,
            IgxDataChartShapeCoreModule], exports: [IgxScatterAreaSeriesModule,
            IgxScatterContourSeriesModule,
            IgxScatterPolygonSeriesModule,
            IgxScatterPolylineSeriesModule,
            IgxDataChartShapeCoreModule] });
    IgxDataChartShapeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartShapeModule, imports: [[CommonModule, IgxScatterAreaSeriesModule,
                IgxScatterContourSeriesModule,
                IgxScatterPolygonSeriesModule,
                IgxScatterPolylineSeriesModule,
                IgxDataChartShapeCoreModule
            ], IgxScatterAreaSeriesModule,
            IgxScatterContourSeriesModule,
            IgxScatterPolygonSeriesModule,
            IgxScatterPolylineSeriesModule,
            IgxDataChartShapeCoreModule] });
    return IgxDataChartShapeModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartShapeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxScatterAreaSeriesModule,
                        IgxScatterContourSeriesModule,
                        IgxScatterPolygonSeriesModule,
                        IgxScatterPolylineSeriesModule,
                        IgxDataChartShapeCoreModule
                    ],
                    imports: [CommonModule, IgxScatterAreaSeriesModule,
                        IgxScatterContourSeriesModule,
                        IgxScatterPolygonSeriesModule,
                        IgxScatterPolylineSeriesModule,
                        IgxDataChartShapeCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
