// igniteui-angular-charts/esm2015/lib/igx-data-chart-annotation-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AnnotationFactoryManager } from './AnnotationFactoryManager';
import { IgxCrosshairLayerDynamicModule } from './igx-crosshair-layer-dynamic-module';
import { IgxCategoryToolTipLayerDynamicModule } from './igx-category-tool-tip-layer-dynamic-module';
import { IgxItemToolTipLayerDynamicModule } from './igx-item-tool-tip-layer-dynamic-module';
import { IgxCategoryItemHighlightLayerDynamicModule } from './igx-category-item-highlight-layer-dynamic-module';
import { IgxCategoryHighlightLayerDynamicModule } from './igx-category-highlight-layer-dynamic-module';
import { IgxFinalValueLayerDynamicModule } from './igx-final-value-layer-dynamic-module';
import { IgxCalloutLayerDynamicModule } from './igx-callout-layer-dynamic-module';
import { IgxDataToolTipLayerDynamicModule } from './igx-data-tool-tip-layer-dynamic-module';
import { IgxValueLayerDynamicModule } from './igx-value-layer-dynamic-module';
import { TypeRegistrar } from "igniteui-angular-core";
import { IgxDataChartAnnotationModule } from './igx-data-chart-annotation-module';
import * as i0 from "@angular/core";
export let IgxDataChartAnnotationDynamicModule = /*@__PURE__*/ (() => {
    class IgxDataChartAnnotationDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('AnnotationFactoryManager', AnnotationFactoryManager);
            TypeRegistrar.register('AnnotationFactoryManager', AnnotationFactoryManager.$type);
        }
    }
    IgxDataChartAnnotationDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartAnnotationDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataChartAnnotationDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartAnnotationDynamicModule, imports: [CommonModule, IgxCrosshairLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxValueLayerDynamicModule,
            IgxDataChartAnnotationModule], exports: [IgxCrosshairLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxValueLayerDynamicModule,
            IgxDataChartAnnotationModule] });
    IgxDataChartAnnotationDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartAnnotationDynamicModule, imports: [[CommonModule, IgxCrosshairLayerDynamicModule,
                IgxCategoryToolTipLayerDynamicModule,
                IgxItemToolTipLayerDynamicModule,
                IgxCategoryItemHighlightLayerDynamicModule,
                IgxCategoryHighlightLayerDynamicModule,
                IgxFinalValueLayerDynamicModule,
                IgxCalloutLayerDynamicModule,
                IgxDataToolTipLayerDynamicModule,
                IgxValueLayerDynamicModule,
                IgxDataChartAnnotationModule
            ], IgxCrosshairLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxValueLayerDynamicModule,
            IgxDataChartAnnotationModule] });
    return IgxDataChartAnnotationDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataChartAnnotationDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxCrosshairLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxFinalValueLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxValueLayerDynamicModule,
                        IgxDataChartAnnotationModule
                    ],
                    imports: [CommonModule, IgxCrosshairLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxFinalValueLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxValueLayerDynamicModule,
                        IgxDataChartAnnotationModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
