// igniteui-angular-charts/esm2015/lib/igx-funnel-chart-core-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumberFormatSpecifierModule } from "igniteui-angular-core";
import { IgxDateTimeFormatSpecifierModule } from "igniteui-angular-core";
import { IgxFunnelChartComponent } from './igx-funnel-chart-component';
import { FastItemObjectColumn } from "igniteui-angular-core";
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxFunnelChartCoreModule = /*@__PURE__*/ (() => {
    class IgxFunnelChartCoreModule {
        constructor() {
            TypeRegistrar.register('FastItemObjectColumn', FastItemObjectColumn.$type);
        }
    }
    IgxFunnelChartCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFunnelChartCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartCoreModule, declarations: [IgxFunnelChartComponent], imports: [CommonModule, IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule], exports: [IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule,
            IgxFunnelChartComponent] });
    IgxFunnelChartCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartCoreModule, imports: [[CommonModule, IgxNumberFormatSpecifierModule,
                IgxDateTimeFormatSpecifierModule
            ], IgxNumberFormatSpecifierModule,
            IgxDateTimeFormatSpecifierModule] });
    return IgxFunnelChartCoreModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFunnelChartCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxFunnelChartComponent],
                    exports: [IgxNumberFormatSpecifierModule,
                        IgxDateTimeFormatSpecifierModule,
                        IgxFunnelChartComponent
                    ],
                    imports: [CommonModule, IgxNumberFormatSpecifierModule,
                        IgxDateTimeFormatSpecifierModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
