// igniteui-angular-charts/esm2015/lib/igx-legend-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxLegendComponent } from './igx-legend-component';
import { LegendTemplates } from './LegendTemplates';
import { TypeRegistrar } from "igniteui-angular-core";
import * as i0 from "@angular/core";
export let IgxLegendModule = /*@__PURE__*/ (() => {
    class IgxLegendModule {
        constructor() {
            TypeRegistrar.register('LegendTemplates', LegendTemplates.$type);
        }
    }
    IgxLegendModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLegendModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxLegendModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLegendModule, declarations: [IgxLegendComponent], imports: [CommonModule], exports: [IgxLegendComponent] });
    IgxLegendModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLegendModule, imports: [[CommonModule,
            ]] });
    return IgxLegendModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLegendModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxLegendComponent],
                    exports: [IgxLegendComponent
                    ],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
