// igniteui-angular-charts/esm2015/lib/igx-financial-chart-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxLineSeriesDynamicModule } from './igx-line-series-dynamic-module';
import { IgxAreaSeriesDynamicModule } from './igx-area-series-dynamic-module';
import { IgxColumnSeriesDynamicModule } from './igx-column-series-dynamic-module';
import { IgxNumericYAxisDynamicModule } from './igx-numeric-y-axis-dynamic-module';
import { IgxTimeXAxisDynamicModule } from './igx-time-x-axis-dynamic-module';
import { IgxTimeAxisBreakDynamicModule } from './igx-time-axis-break-dynamic-module';
import { IgxTimeAxisIntervalDynamicModule } from './igx-time-axis-interval-dynamic-module';
import { IgxTimeAxisLabelFormatDynamicModule } from './igx-time-axis-label-format-dynamic-module';
import { IgxCategoryXAxisDynamicModule } from './igx-category-x-axis-dynamic-module';
import { IgxOrdinalTimeXAxisDynamicModule } from './igx-ordinal-time-x-axis-dynamic-module';
import { IgxPercentChangeYAxisDynamicModule } from './igx-percent-change-y-axis-dynamic-module';
import { IgxFinancialPriceSeriesDynamicModule } from './igx-financial-price-series-dynamic-module';
import { IgxDataChartCategoryTrendLineModule } from './igx-data-chart-category-trend-line-module';
import { IgxIndicatorsDynamicModule } from './igx-indicators-dynamic-module';
import { IgxOverlaysDynamicModule } from './igx-overlays-dynamic-module';
import { IgxZoomSliderDynamicModule } from './igx-zoom-slider-dynamic-module';
import { IgxCrosshairLayerDynamicModule } from './igx-crosshair-layer-dynamic-module';
import { IgxCalloutLayerDynamicModule } from './igx-callout-layer-dynamic-module';
import { IgxItemToolTipLayerDynamicModule } from './igx-item-tool-tip-layer-dynamic-module';
import { IgxCategoryToolTipLayerDynamicModule } from './igx-category-tool-tip-layer-dynamic-module';
import { IgxCategoryHighlightLayerDynamicModule } from './igx-category-highlight-layer-dynamic-module';
import { IgxCategoryItemHighlightLayerDynamicModule } from './igx-category-item-highlight-layer-dynamic-module';
import { IgxDataToolTipLayerDynamicModule } from './igx-data-tool-tip-layer-dynamic-module';
import { IgxFinalValueLayerDynamicModule } from './igx-final-value-layer-dynamic-module';
import { IgxNumberAbbreviatorModule } from './igx-number-abbreviator-module';
import { IgxDataChartInteractivityModule } from './igx-data-chart-interactivity-module';
import { IgxDataChartAnnotationModule } from './igx-data-chart-annotation-module';
import { IgxFinancialLegendModule } from './igx-financial-legend-module';
import { IgxFinancialChartCoreModule } from './igx-financial-chart-core-module';
import * as i0 from "@angular/core";
export let IgxFinancialChartModule = /*@__PURE__*/ (() => {
    class IgxFinancialChartModule {
    }
    IgxFinancialChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxFinancialChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartModule, imports: [CommonModule, IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxTimeXAxisDynamicModule,
            IgxTimeAxisBreakDynamicModule,
            IgxTimeAxisIntervalDynamicModule,
            IgxTimeAxisLabelFormatDynamicModule,
            IgxCategoryXAxisDynamicModule,
            IgxOrdinalTimeXAxisDynamicModule,
            IgxPercentChangeYAxisDynamicModule,
            IgxFinancialPriceSeriesDynamicModule,
            IgxDataChartCategoryTrendLineModule,
            IgxIndicatorsDynamicModule,
            IgxOverlaysDynamicModule,
            IgxZoomSliderDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxNumberAbbreviatorModule,
            IgxDataChartInteractivityModule,
            IgxDataChartAnnotationModule,
            IgxFinancialLegendModule,
            IgxFinancialChartCoreModule], exports: [IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxTimeXAxisDynamicModule,
            IgxTimeAxisBreakDynamicModule,
            IgxTimeAxisIntervalDynamicModule,
            IgxTimeAxisLabelFormatDynamicModule,
            IgxCategoryXAxisDynamicModule,
            IgxOrdinalTimeXAxisDynamicModule,
            IgxPercentChangeYAxisDynamicModule,
            IgxFinancialPriceSeriesDynamicModule,
            IgxDataChartCategoryTrendLineModule,
            IgxIndicatorsDynamicModule,
            IgxOverlaysDynamicModule,
            IgxZoomSliderDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxNumberAbbreviatorModule,
            IgxDataChartInteractivityModule,
            IgxDataChartAnnotationModule,
            IgxFinancialLegendModule,
            IgxFinancialChartCoreModule] });
    IgxFinancialChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartModule, imports: [[CommonModule, IgxLineSeriesDynamicModule,
                IgxAreaSeriesDynamicModule,
                IgxColumnSeriesDynamicModule,
                IgxNumericYAxisDynamicModule,
                IgxTimeXAxisDynamicModule,
                IgxTimeAxisBreakDynamicModule,
                IgxTimeAxisIntervalDynamicModule,
                IgxTimeAxisLabelFormatDynamicModule,
                IgxCategoryXAxisDynamicModule,
                IgxOrdinalTimeXAxisDynamicModule,
                IgxPercentChangeYAxisDynamicModule,
                IgxFinancialPriceSeriesDynamicModule,
                IgxDataChartCategoryTrendLineModule,
                IgxIndicatorsDynamicModule,
                IgxOverlaysDynamicModule,
                IgxZoomSliderDynamicModule,
                IgxCrosshairLayerDynamicModule,
                IgxCalloutLayerDynamicModule,
                IgxItemToolTipLayerDynamicModule,
                IgxCategoryToolTipLayerDynamicModule,
                IgxCategoryHighlightLayerDynamicModule,
                IgxCategoryItemHighlightLayerDynamicModule,
                IgxDataToolTipLayerDynamicModule,
                IgxFinalValueLayerDynamicModule,
                IgxNumberAbbreviatorModule,
                IgxDataChartInteractivityModule,
                IgxDataChartAnnotationModule,
                IgxFinancialLegendModule,
                IgxFinancialChartCoreModule
            ], IgxLineSeriesDynamicModule,
            IgxAreaSeriesDynamicModule,
            IgxColumnSeriesDynamicModule,
            IgxNumericYAxisDynamicModule,
            IgxTimeXAxisDynamicModule,
            IgxTimeAxisBreakDynamicModule,
            IgxTimeAxisIntervalDynamicModule,
            IgxTimeAxisLabelFormatDynamicModule,
            IgxCategoryXAxisDynamicModule,
            IgxOrdinalTimeXAxisDynamicModule,
            IgxPercentChangeYAxisDynamicModule,
            IgxFinancialPriceSeriesDynamicModule,
            IgxDataChartCategoryTrendLineModule,
            IgxIndicatorsDynamicModule,
            IgxOverlaysDynamicModule,
            IgxZoomSliderDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxFinalValueLayerDynamicModule,
            IgxNumberAbbreviatorModule,
            IgxDataChartInteractivityModule,
            IgxDataChartAnnotationModule,
            IgxFinancialLegendModule,
            IgxFinancialChartCoreModule] });
    return IgxFinancialChartModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxFinancialChartModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxLineSeriesDynamicModule,
                        IgxAreaSeriesDynamicModule,
                        IgxColumnSeriesDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxTimeXAxisDynamicModule,
                        IgxTimeAxisBreakDynamicModule,
                        IgxTimeAxisIntervalDynamicModule,
                        IgxTimeAxisLabelFormatDynamicModule,
                        IgxCategoryXAxisDynamicModule,
                        IgxOrdinalTimeXAxisDynamicModule,
                        IgxPercentChangeYAxisDynamicModule,
                        IgxFinancialPriceSeriesDynamicModule,
                        IgxDataChartCategoryTrendLineModule,
                        IgxIndicatorsDynamicModule,
                        IgxOverlaysDynamicModule,
                        IgxZoomSliderDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxFinalValueLayerDynamicModule,
                        IgxNumberAbbreviatorModule,
                        IgxDataChartInteractivityModule,
                        IgxDataChartAnnotationModule,
                        IgxFinancialLegendModule,
                        IgxFinancialChartCoreModule
                    ],
                    imports: [CommonModule, IgxLineSeriesDynamicModule,
                        IgxAreaSeriesDynamicModule,
                        IgxColumnSeriesDynamicModule,
                        IgxNumericYAxisDynamicModule,
                        IgxTimeXAxisDynamicModule,
                        IgxTimeAxisBreakDynamicModule,
                        IgxTimeAxisIntervalDynamicModule,
                        IgxTimeAxisLabelFormatDynamicModule,
                        IgxCategoryXAxisDynamicModule,
                        IgxOrdinalTimeXAxisDynamicModule,
                        IgxPercentChangeYAxisDynamicModule,
                        IgxFinancialPriceSeriesDynamicModule,
                        IgxDataChartCategoryTrendLineModule,
                        IgxIndicatorsDynamicModule,
                        IgxOverlaysDynamicModule,
                        IgxZoomSliderDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxFinalValueLayerDynamicModule,
                        IgxNumberAbbreviatorModule,
                        IgxDataChartInteractivityModule,
                        IgxDataChartAnnotationModule,
                        IgxFinancialLegendModule,
                        IgxFinancialChartCoreModule
                    ],
                    entryComponents: []
                }]
        }] });
