// igniteui-angular-charts/esm2015/lib/igx-time-axis-interval-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTimeAxisIntervalComponent } from './igx-time-axis-interval-component';
import * as i0 from "@angular/core";
export let IgxTimeAxisIntervalModule = /*@__PURE__*/ (() => {
    class IgxTimeAxisIntervalModule {
    }
    IgxTimeAxisIntervalModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisIntervalModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTimeAxisIntervalModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisIntervalModule, declarations: [IgxTimeAxisIntervalComponent], imports: [CommonModule], exports: [IgxTimeAxisIntervalComponent] });
    IgxTimeAxisIntervalModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisIntervalModule, imports: [[CommonModule]] });
    return IgxTimeAxisIntervalModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTimeAxisIntervalModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxTimeAxisIntervalComponent],
                    imports: [CommonModule],
                    exports: [IgxTimeAxisIntervalComponent],
                    entryComponents: []
                }]
        }] });
