// igniteui-angular-charts/esm2015/lib/igx-stacked-100-line-series-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxStacked100LineSeriesComponent } from './igx-stacked-100-line-series-component';
import * as i0 from "@angular/core";
export let IgxStacked100LineSeriesModule = /*@__PURE__*/ (() => {
    class IgxStacked100LineSeriesModule {
    }
    IgxStacked100LineSeriesModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100LineSeriesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxStacked100LineSeriesModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100LineSeriesModule, declarations: [IgxStacked100LineSeriesComponent], imports: [CommonModule], exports: [IgxStacked100LineSeriesComponent] });
    IgxStacked100LineSeriesModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100LineSeriesModule, imports: [[CommonModule]] });
    return IgxStacked100LineSeriesModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxStacked100LineSeriesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxStacked100LineSeriesComponent],
                    imports: [CommonModule],
                    exports: [IgxStacked100LineSeriesComponent],
                    entryComponents: []
                }]
        }] });
