// igniteui-angular-charts/esm2015/lib/igx-data-pie-chart-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDataPieChartCoreModule } from './igx-data-pie-chart-core-module';
import { IgxRadialPieSeriesDynamicModule } from './igx-radial-pie-series-dynamic-module';
import { IgxCrosshairLayerDynamicModule } from './igx-crosshair-layer-dynamic-module';
import { IgxCalloutLayerDynamicModule } from './igx-callout-layer-dynamic-module';
import { IgxItemToolTipLayerDynamicModule } from './igx-item-tool-tip-layer-dynamic-module';
import { IgxCategoryToolTipLayerDynamicModule } from './igx-category-tool-tip-layer-dynamic-module';
import { IgxCategoryHighlightLayerDynamicModule } from './igx-category-highlight-layer-dynamic-module';
import { IgxCategoryItemHighlightLayerDynamicModule } from './igx-category-item-highlight-layer-dynamic-module';
import { IgxDataToolTipLayerDynamicModule } from './igx-data-tool-tip-layer-dynamic-module';
import { IgxNumberAbbreviatorModule } from './igx-number-abbreviator-module';
import { IgxDataChartInteractivityModule } from './igx-data-chart-interactivity-module';
import { IgxDataChartAnnotationModule } from './igx-data-chart-annotation-module';
import * as i0 from "@angular/core";
export let IgxDataPieChartModule = /*@__PURE__*/ (() => {
    class IgxDataPieChartModule {
    }
    IgxDataPieChartModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataPieChartModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartModule, imports: [CommonModule, IgxDataPieChartCoreModule,
            IgxRadialPieSeriesDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxNumberAbbreviatorModule,
            IgxDataChartInteractivityModule,
            IgxDataChartAnnotationModule], exports: [IgxDataPieChartCoreModule,
            IgxRadialPieSeriesDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxNumberAbbreviatorModule,
            IgxDataChartInteractivityModule,
            IgxDataChartAnnotationModule] });
    IgxDataPieChartModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartModule, imports: [[CommonModule, IgxDataPieChartCoreModule,
                IgxRadialPieSeriesDynamicModule,
                IgxCrosshairLayerDynamicModule,
                IgxCalloutLayerDynamicModule,
                IgxItemToolTipLayerDynamicModule,
                IgxCategoryToolTipLayerDynamicModule,
                IgxCategoryHighlightLayerDynamicModule,
                IgxCategoryItemHighlightLayerDynamicModule,
                IgxDataToolTipLayerDynamicModule,
                IgxNumberAbbreviatorModule,
                IgxDataChartInteractivityModule,
                IgxDataChartAnnotationModule
            ], IgxDataPieChartCoreModule,
            IgxRadialPieSeriesDynamicModule,
            IgxCrosshairLayerDynamicModule,
            IgxCalloutLayerDynamicModule,
            IgxItemToolTipLayerDynamicModule,
            IgxCategoryToolTipLayerDynamicModule,
            IgxCategoryHighlightLayerDynamicModule,
            IgxCategoryItemHighlightLayerDynamicModule,
            IgxDataToolTipLayerDynamicModule,
            IgxNumberAbbreviatorModule,
            IgxDataChartInteractivityModule,
            IgxDataChartAnnotationModule] });
    return IgxDataPieChartModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataPieChartModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxDataPieChartCoreModule,
                        IgxRadialPieSeriesDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxNumberAbbreviatorModule,
                        IgxDataChartInteractivityModule,
                        IgxDataChartAnnotationModule
                    ],
                    imports: [CommonModule, IgxDataPieChartCoreModule,
                        IgxRadialPieSeriesDynamicModule,
                        IgxCrosshairLayerDynamicModule,
                        IgxCalloutLayerDynamicModule,
                        IgxItemToolTipLayerDynamicModule,
                        IgxCategoryToolTipLayerDynamicModule,
                        IgxCategoryHighlightLayerDynamicModule,
                        IgxCategoryItemHighlightLayerDynamicModule,
                        IgxDataToolTipLayerDynamicModule,
                        IgxNumberAbbreviatorModule,
                        IgxDataChartInteractivityModule,
                        IgxDataChartAnnotationModule
                    ],
                    entryComponents: []
                }]
        }] });
