// igniteui-angular-charts/esm2015/lib/igx-category-y-axis-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxCategoryYAxisComponent } from './igx-category-y-axis-component';
import * as i0 from "@angular/core";
export let IgxCategoryYAxisModule = /*@__PURE__*/ (() => {
    class IgxCategoryYAxisModule {
    }
    IgxCategoryYAxisModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryYAxisModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxCategoryYAxisModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryYAxisModule, declarations: [IgxCategoryYAxisComponent], imports: [CommonModule], exports: [IgxCategoryYAxisComponent] });
    IgxCategoryYAxisModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryYAxisModule, imports: [[CommonModule]] });
    return IgxCategoryYAxisModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxCategoryYAxisModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxCategoryYAxisComponent],
                    imports: [CommonModule],
                    exports: [IgxCategoryYAxisComponent],
                    entryComponents: []
                }]
        }] });
