// igniteui-angular-charts/esm2015/lib/ItemwiseStrategyBasedIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { ItemwiseStrategyCalculationStrategy } from "./ItemwiseStrategyCalculationStrategy";
import { markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let ItemwiseStrategyBasedIndicator = /*@__PURE__*/ (() => {
    class ItemwiseStrategyBasedIndicator extends StrategyBasedIndicator {
        constructor() {
            super();
        }
        get_abn() {
            return new ItemwiseStrategyCalculationStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        aca() {
            return true;
        }
        ab1() {
            super.ab1();
            this.ab8 = this.ab9;
            this.abm.c = this.ab8;
        }
    }
    ItemwiseStrategyBasedIndicator.$t = markType(ItemwiseStrategyBasedIndicator, 'ItemwiseStrategyBasedIndicator', StrategyBasedIndicator.$);
    return ItemwiseStrategyBasedIndicator;
})();
