// igniteui-angular-charts/esm2015/lib/MarketFacilitationIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ItemwiseStrategyBasedIndicator } from "./ItemwiseStrategyBasedIndicator";
import { MarketFacilitationIndexIndicatorStrategy } from "./MarketFacilitationIndexIndicatorStrategy";
import { markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let MarketFacilitationIndexIndicator = /*@__PURE__*/ (() => {
    class MarketFacilitationIndexIndicator extends ItemwiseStrategyBasedIndicator {
        constructor() {
            super();
        }
        bi() {
            return new MarketFacilitationIndexIndicator();
        }
        get_aau() {
            return 1;
        }
        get_ab9() {
            return new MarketFacilitationIndexIndicatorStrategy();
        }
        get ab9() {
            return this.get_ab9();
        }
        get_ab0() {
            return MarketFacilitationIndexIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
    }
    MarketFacilitationIndexIndicator.$t = markType(MarketFacilitationIndexIndicator, 'MarketFacilitationIndexIndicator', ItemwiseStrategyBasedIndicator.$);
    return MarketFacilitationIndexIndicator;
})();
