// igniteui-angular-charts/esm2015/lib/MedianPriceIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ItemwiseStrategyBasedIndicator } from "./ItemwiseStrategyBasedIndicator";
import { MedianPriceIndicatorStrategy } from "./MedianPriceIndicatorStrategy";
import { markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let MedianPriceIndicator = /*@__PURE__*/ (() => {
    class MedianPriceIndicator extends ItemwiseStrategyBasedIndicator {
        bi() {
            return new MedianPriceIndicator();
        }
        get_ab9() {
            return new MedianPriceIndicatorStrategy();
        }
        get ab9() {
            return this.get_ab9();
        }
        get_ab0() {
            return MedianPriceIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
    }
    MedianPriceIndicator.$t = markType(MedianPriceIndicator, 'MedianPriceIndicator', ItemwiseStrategyBasedIndicator.$);
    return MedianPriceIndicator;
})();
