// igniteui-angular-charts/esm2015/lib/MoneyFlowIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { MoneyFlowIndexIndicatorStrategy } from "./MoneyFlowIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let MoneyFlowIndexIndicator = /*@__PURE__*/ (() => {
    class MoneyFlowIndexIndicator extends StrategyBasedIndicator {
        constructor() {
            super();
        }
        bi() {
            return new MoneyFlowIndexIndicator();
        }
        get_aau() {
            return 1;
        }
        get_abn() {
            return new MoneyFlowIndexIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return MoneyFlowIndexIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        get ab8() {
            return typeGetValue(this.c(MoneyFlowIndexIndicator.ab9));
        }
        set ab8(a) {
            this.h(MoneyFlowIndexIndicator.ab9, a);
        }
        abt() {
            return this.ab8;
        }
        ab2(a) {
            this.ab8 = a;
        }
    }
    MoneyFlowIndexIndicator.$t = markType(MoneyFlowIndexIndicator, 'MoneyFlowIndexIndicator', StrategyBasedIndicator.$);
    MoneyFlowIndexIndicator.ab9 = StrategyBasedIndicator.ab5(14, MoneyFlowIndexIndicator.$);
    return MoneyFlowIndexIndicator;
})();
