// igniteui-angular-charts/esm2015/lib/MovingAverageConvergenceDivergenceIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { MovingAverageConvergenceDivergenceIndicatorStrategy } from "./MovingAverageConvergenceDivergenceIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
import { Brush } from "igniteui-angular-core";
/**
 * @hidden
 */
export let MovingAverageConvergenceDivergenceIndicator = /*@__PURE__*/ (() => {
    class MovingAverageConvergenceDivergenceIndicator extends StrategyBasedIndicator {
        constructor() {
            super();
            this.trendLineBrush = ((() => {
                let $ret = new Brush();
                $ret.fill = "#565656";
                return $ret;
            })());
            this.trendLineType = 10;
            this.trendLinePeriod = 9;
        }
        bi() {
            return new MovingAverageConvergenceDivergenceIndicator();
        }
        get_aau() {
            return 1;
        }
        get_abn() {
            return new MovingAverageConvergenceDivergenceIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return MovingAverageConvergenceDivergenceIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        get shortPeriod() {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.acb));
        }
        set shortPeriod(a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.acb, a);
        }
        abv() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.aca));
        }
        set longPeriod(a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.aca, a);
        }
        abs() {
            return this.longPeriod;
        }
        get signalPeriod() {
            return typeGetValue(this.c(MovingAverageConvergenceDivergenceIndicator.acc));
        }
        set signalPeriod(a) {
            this.h(MovingAverageConvergenceDivergenceIndicator.acc, a);
        }
        aa3() {
            return this.signalPeriod;
        }
    }
    MovingAverageConvergenceDivergenceIndicator.$t = markType(MovingAverageConvergenceDivergenceIndicator, 'MovingAverageConvergenceDivergenceIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    MovingAverageConvergenceDivergenceIndicator.acb = StrategyBasedIndicator.ab7(10, MovingAverageConvergenceDivergenceIndicator.$);
    MovingAverageConvergenceDivergenceIndicator.aca = StrategyBasedIndicator.ab3(30, MovingAverageConvergenceDivergenceIndicator.$);
    MovingAverageConvergenceDivergenceIndicator.acc = StrategyBasedIndicator.ab6(9, MovingAverageConvergenceDivergenceIndicator.$, "SignalPeriod");
    return MovingAverageConvergenceDivergenceIndicator;
})();
