// igniteui-angular-charts/esm2015/lib/NegativeVolumeIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { NegativeVolumeIndexIndicatorStrategy } from "./NegativeVolumeIndexIndicatorStrategy";
import { markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let NegativeVolumeIndexIndicator = /*@__PURE__*/ (() => {
    class NegativeVolumeIndexIndicator extends StrategyBasedIndicator {
        bi() {
            return new NegativeVolumeIndexIndicator();
        }
        get_abn() {
            return new NegativeVolumeIndexIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return NegativeVolumeIndexIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
    }
    NegativeVolumeIndexIndicator.$t = markType(NegativeVolumeIndexIndicator, 'NegativeVolumeIndexIndicator', StrategyBasedIndicator.$);
    return NegativeVolumeIndexIndicator;
})();
