// igniteui-angular-charts/esm2015/lib/PercentageVolumeOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentageVolumeOscillatorIndicatorStrategy } from "./PercentageVolumeOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let PercentageVolumeOscillatorIndicator = /*@__PURE__*/ (() => {
    class PercentageVolumeOscillatorIndicator extends StrategyBasedIndicator {
        bi() {
            return new PercentageVolumeOscillatorIndicator();
        }
        get_abn() {
            return new PercentageVolumeOscillatorIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return PercentageVolumeOscillatorIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        get shortPeriod() {
            return typeGetValue(this.c(PercentageVolumeOscillatorIndicator.ab9));
        }
        set shortPeriod(a) {
            this.h(PercentageVolumeOscillatorIndicator.ab9, a);
        }
        abv() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(PercentageVolumeOscillatorIndicator.ab8));
        }
        set longPeriod(a) {
            this.h(PercentageVolumeOscillatorIndicator.ab8, a);
        }
        abs() {
            return this.longPeriod;
        }
    }
    PercentageVolumeOscillatorIndicator.$t = markType(PercentageVolumeOscillatorIndicator, 'PercentageVolumeOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    PercentageVolumeOscillatorIndicator.ab9 = StrategyBasedIndicator.ab7(10, PercentageVolumeOscillatorIndicator.$);
    PercentageVolumeOscillatorIndicator.ab8 = StrategyBasedIndicator.ab3(30, PercentageVolumeOscillatorIndicator.$);
    return PercentageVolumeOscillatorIndicator;
})();
