// igniteui-angular-charts/esm2015/lib/PercentagePriceOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { IShortPeriodLongPeriod_$type } from "./IShortPeriodLongPeriod";
import { PercentagePriceOscillatorIndicatorStrategy } from "./PercentagePriceOscillatorIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let PercentagePriceOscillatorIndicator = /*@__PURE__*/ (() => {
    class PercentagePriceOscillatorIndicator extends StrategyBasedIndicator {
        bi() {
            return new PercentagePriceOscillatorIndicator();
        }
        get_abn() {
            return new PercentagePriceOscillatorIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return PercentagePriceOscillatorIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        get shortPeriod() {
            return typeGetValue(this.c(PercentagePriceOscillatorIndicator.ab9));
        }
        set shortPeriod(a) {
            this.h(PercentagePriceOscillatorIndicator.ab9, a);
        }
        abv() {
            return this.shortPeriod;
        }
        get longPeriod() {
            return typeGetValue(this.c(PercentagePriceOscillatorIndicator.ab8));
        }
        set longPeriod(a) {
            this.h(PercentagePriceOscillatorIndicator.ab8, a);
        }
        abs() {
            return this.longPeriod;
        }
    }
    PercentagePriceOscillatorIndicator.$t = markType(PercentagePriceOscillatorIndicator, 'PercentagePriceOscillatorIndicator', StrategyBasedIndicator.$, [IShortPeriodLongPeriod_$type]);
    PercentagePriceOscillatorIndicator.ab9 = StrategyBasedIndicator.ab7(10, PercentagePriceOscillatorIndicator.$);
    PercentagePriceOscillatorIndicator.ab8 = StrategyBasedIndicator.ab3(30, PercentagePriceOscillatorIndicator.$);
    return PercentagePriceOscillatorIndicator;
})();
