// igniteui-angular-charts/esm2015/lib/PolarLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PolarLineSeriesBase } from "./PolarLineSeriesBase";
import { UnknownValuePlotting_$type } from "igniteui-angular-core";
import { DependencyProperty } from "igniteui-angular-core";
import { runOn, enumGetBox, markType, markDep } from "igniteui-angular-core";
import { Path } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { PolarLineSeriesBaseView } from "./PolarLineSeriesBaseView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let PolarLineSeries = /*@__PURE__*/ (() => {
    class PolarLineSeries extends PolarLineSeriesBase {
        constructor() {
            super();
            this.ab = PolarLineSeries.$;
        }
        bi() {
            return new PolarLineSeries();
        }
        get_fi() {
            return true;
        }
        cq() {
            return new PolarLineSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.acm = a;
        }
        ack(a) {
            let b = a;
            b.da();
        }
        acl(a, b) {
            let c = b;
            c.db(a.m, this.ib);
            this.yx.az(this, this.zp(), runOn(this, this.yf));
            let d = false;
            let e = this.yx.b;
            if (e != null) {
                d = true;
            }
            if (d) {
                this.aa4(-1, this.zb.count, b.q);
            }
            this.yx.ad(b.dd, true, false, true, false);
        }
        acc() {
            return this.acn;
        }
        q5(a, b, c, d) {
            super.q5(a, b, c, d);
            switch (b) {
                case PolarLineSeries.$$p[0]:
                    this.rn(false);
                    this.qh();
                    break;
            }
        }
    }
    PolarLineSeries.$t = markType(PolarLineSeries, 'PolarLineSeries', PolarLineSeriesBase.$);
    PolarLineSeries.$$p = markDep(DependencyProperty, PropertyMetadata, PolarLineSeries, 'raisePropertyChanged', ['UnknownValuePlotting:acn:acp', [UnknownValuePlotting_$type, enumGetBox(UnknownValuePlotting_$type, 1)]]);
    return PolarLineSeries;
})();
/**
 * @hidden
 */
export let PolarLineSeriesView = /*@__PURE__*/ (() => {
    class PolarLineSeriesView extends PolarLineSeriesBaseView {
        constructor(a) {
            super(a);
            this.c9 = null;
            this.dd = new Path();
            this.dc = new Path();
            this.c9 = a;
        }
        au() {
            super.au();
            if (!this.q) {
                this.b4.w2 = Defaults.e;
            }
        }
        da() {
            this.dd.an = null;
            this.dc.an = null;
            this.as();
        }
        db(a, b) {
            this.c9.ab9.d(this.dd, a, b);
            this.as();
        }
        bd() {
            super.bd();
        }
        be() {
            super.be();
            let a = this.br();
            this.dc.an = this.dd.an;
            this.dc._stroke = a;
            this.dc.ad = this.e.ic + 3;
        }
        a8(a, b) {
            super.a8(a, b);
            if (a.d) {
                if (b) {
                    a.t(this.dc);
                }
                else {
                    a.t(this.dd);
                }
            }
        }
        an(a) {
            super.an(a);
            this.e.bb.exportPathData(a, this.dd, "polarShape", ["Main"]);
        }
    }
    PolarLineSeriesView.$t = markType(PolarLineSeriesView, 'PolarLineSeriesView', PolarLineSeriesBaseView.$);
    return PolarLineSeriesView;
})();
