// igniteui-angular-charts/esm2015/lib/PolarScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PolarBase } from "./PolarBase";
import { markType } from "igniteui-angular-core";
import { PolarBaseView } from "./PolarBaseView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-angular-core";
/**
 * @hidden
 */
export let PolarScatterSeries = /*@__PURE__*/ (() => {
    class PolarScatterSeries extends PolarBase {
        constructor() {
            super();
            this.ab = PolarScatterSeries.$;
        }
        bi() {
            return new PolarScatterSeries();
        }
        get_eg() {
            return true;
        }
        cq() {
            return new PolarScatterSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.ab8 = a;
        }
        jj() {
            return 0.7;
        }
        wt() {
            if (this.ws == 0) {
                return 2;
            }
            return super.wt();
        }
    }
    PolarScatterSeries.$t = markType(PolarScatterSeries, 'PolarScatterSeries', PolarBase.$);
    return PolarScatterSeries;
})();
/**
 * @hidden
 */
export let PolarScatterSeriesView = /*@__PURE__*/ (() => {
    class PolarScatterSeriesView extends PolarBaseView {
        constructor(a) {
            super(a);
            this.c8 = null;
            this.c8 = a;
        }
        au() {
            super.au();
            if (!this.q) {
                this.c8.w2 = Defaults.e;
            }
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.vv = a;
            this.e.jk = 3;
            this.e.jl = 2;
            this.e.jm = 2;
            this.e.ha = false;
        }
    }
    PolarScatterSeriesView.$t = markType(PolarScatterSeriesView, 'PolarScatterSeriesView', PolarBaseView.$);
    return PolarScatterSeriesView;
})();
