// igniteui-angular-charts/esm2015/lib/PriceChannelOverlay_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FinancialOverlay } from "./FinancialOverlay";
import { DependencyProperty } from "igniteui-angular-core";
import { Series } from "./Series";
import { List$1 } from "igniteui-angular-core";
import { Number_$type, typeCast, runOn, markType, markDep, Point_$type, fromEnum } from "igniteui-angular-core";
import { CategoryFrame } from "./CategoryFrame";
import { MathUtil } from "igniteui-angular-core";
import { FinancialSeriesView } from "./FinancialSeriesView";
import { ScalerParams } from "./ScalerParams";
import { ISortingAxis_$type } from "./ISortingAxis";
import { FinancialBucketCalculator } from "./FinancialBucketCalculator";
import { Polyline } from "igniteui-angular-core";
import { Polygon } from "igniteui-angular-core";
import { RangeValueList } from "./RangeValueList";
import { PropertyMetadata } from "igniteui-angular-core";
import { truncate, isNaN_, isInfinity } from "igniteui-angular-core";
import { Flattener } from "igniteui-angular-core";
/**
 * @hidden
 */
export let PriceChannelOverlay = /*@__PURE__*/ (() => {
    class PriceChannelOverlay extends FinancialOverlay {
        constructor() {
            super();
            this.aaz = new List$1(Number_$type, 0);
            this.aay = new List$1(Number_$type, 0);
            this.ab = PriceChannelOverlay.$;
            this.w2 = new CategoryFrame(3);
            this.w4 = new CategoryFrame(3);
            this.w0 = new CategoryFrame(3);
        }
        bi() {
            return new PriceChannelOverlay();
        }
        q5(a, b, c, d) {
            switch (b) {
                case "HighColumn":
                case "LowColumn":
                    this.aas = false;
                    break;
            }
            super.q5(a, b, c, d);
            switch (b) {
                case PriceChannelOverlay.$$p[0]:
                    this.aas = false;
                    this.rn(false);
                    break;
            }
        }
        hc(a, b, c) {
            let d = super.hc(a, b, c);
            if (this.xs == null || this.xt == null) {
                d = false;
            }
            return d;
        }
        aat() {
            this.aaz.clear();
            this.aay.clear();
            let a = truncate(MathUtil.d(this.aa0, 0, this.dl.count));
            let b = Math.min(this.xs.count, this.xt.count);
            let c = this.yi(this.xs);
            let d = this.yi(this.xt);
            for (let e = 0; e < b; e++) {
                let f = Math.min(a, e);
                let g = -1.7976931348623157E+308;
                let h = 1.7976931348623157E+308;
                for (let i = 0; i < f; i++) {
                    if (!isNaN_(c.item(e - i))) {
                        g = Math.max(g, c.item(e - i));
                    }
                    if (!isNaN_(d.item(e - i))) {
                        h = Math.min(h, d.item(e - i));
                    }
                }
                if (e == 0) {
                    h = d.item(0);
                    g = c.item(0);
                }
                this.aaz.add(g);
                this.aay.add(h);
            }
            return true;
        }
        z4(a, b) {
            super.z4(a, b);
            let c = b.by;
            let d = b.bx;
            let e = this.getEffectiveViewport1(b);
            let f = new ScalerParams(0, c, d, this.wy.ch, e);
            let g = new ScalerParams(0, c, d, this.xh.ch, e);
            g.b = this.iw();
            let h = this.wy;
            let i = this.xh;
            a.f.clear();
            a.n.clear();
            a.q.clear();
            let j = 0;
            let k = typeCast(ISortingAxis_$type, this.wy);
            if (k != null && k.i$b.count != this.dl.count) {
                return;
            }
            j = this.yp(c, d, e);
            if (!this.aas) {
                this.aas = this.aat();
            }
            g.b = this.iw();
            let l = this.wy.en(2, f) - this.wy.en(1, f);
            for (let m = b.b0.c; m <= b.b0.d; ++m) {
                let n;
                if (k == null) {
                    n = b.b0.getBucket(m);
                }
                else {
                    let o = k.i$b._inner[m];
                    let p = k.i$d(o);
                    let q = this.aaz._inner[m];
                    let r = this.aay._inner[m];
                    let s = q;
                    let t = r;
                    let u = p;
                    while (m < b.b0.d) {
                        o = k.i$b._inner[m + 1];
                        u = k.i$d(o);
                        if (u - p > l) {
                            break;
                        }
                        m++;
                        s = Math.max(q, this.aaz._inner[m]);
                        t = Math.min(r, this.aay._inner[m]);
                    }
                    if (!isInfinity(t) && !isInfinity(s)) {
                        let v = NaN;
                        if (!isNaN_(p)) {
                            v = this.wy.ek(p, f);
                        }
                        n = [v, t, s];
                    }
                    else {
                        n = [NaN, NaN, NaN];
                    }
                }
                let w = Math.max(1, l);
                if (!isNaN_(n[0]) && m * w >= this.aau) {
                    if (this.wy != null && typeCast(ISortingAxis_$type, this.wy) !== null) {
                        n[0] = n[0] + j;
                    }
                    else {
                        n[0] = h.ek(n[0], f) + j;
                    }
                    n[1] = i.ek(n[1], g);
                    n[2] = i.ek(n[2], g);
                    a.f.add(n);
                }
            }
        }
        o3(a, b) {
            super.o3(a, b);
            let c = b;
            if (c != null) {
                c.b7();
            }
        }
        z6(a, b) {
            super.z6(a, b);
            let c = typeCast(PriceChannelOverlayView.$, b);
            if (c == null) {
                return;
            }
            let d = this.getEffectiveViewport1(b);
            if (b.b3(a)) {
                c.b7();
                let e = a.f.count;
                let f = (g) => a.f._inner[g][0];
                let g = (h) => a.f._inner[e - 1 - h][0];
                let h = (i) => a.f._inner[i][1];
                let i = (j) => a.f._inner[e - 1 - j][2];
                c.b8(e, f, g, h, i);
                b.b5(a);
            }
            this.w7.a2(this, this.x6(), runOn(this, this.wq), this.yv(b), this.yx(b));
            let j = false;
            let k = this.w7.b;
            let l = a.f;
            let m = this.dl.count;
            if (k != null) {
                j = true;
            }
            if (j) {
                let n = new ScalerParams(0, b.by, b.bx, this.wy.ch, d);
                this.z3(l, -1, m, this.wy, n, b.q, -1);
            }
            let o = c.ce;
            let p = c.cf;
            let q = c.ca;
            this.w7.ad(o, true, false, true, true);
            this.w7.ad(p, true, false, true, true);
            this.w7.ad(q, false, true, false, false);
            q._opacity = this.w7.j * this.h7;
        }
        cq() {
            return new PriceChannelOverlayView(this);
        }
        qx(a) {
            super.qx(a);
            this.aax = typeCast(PriceChannelOverlayView.$, a);
        }
        i3(a, b, c) {
            if (this.c9 == null) {
                return NaN;
            }
            let d = this.getEffectiveViewport1(this.ct);
            let e = new ScalerParams(0, this.c9.vl, this.ct.bx, this.wy.ch, d);
            let f = this.yp(this.c9.vl, this.ct.bx, d);
            return this.i5(new RangeValueList(this.aaz, this.aay), a, this.wy, e, f, runOn(this, this.yo), b, c);
        }
        j6(a, b) {
            if (this.aaz == null || this.aay == null) {
                return super.j6(a, b);
            }
            return this.j7(a, b, this.wy, runOn(this, this.yo), new RangeValueList(this.aaz, this.aay));
        }
        j8(a, b) {
            if (this.aaz == null || this.aay == null) {
                return super.j8(a, b);
            }
            return this.j9(a, b, this.wy, runOn(this, this.yo), new RangeValueList(this.aaz, this.aay));
        }
        iw() {
            return Series.iy(this.aay, null);
        }
    }
    PriceChannelOverlay.$t = markType(PriceChannelOverlay, 'PriceChannelOverlay', FinancialOverlay.$);
    PriceChannelOverlay.$$p = markDep(DependencyProperty, PropertyMetadata, PriceChannelOverlay, 'raisePropertyChanged', ['Period:aa0:aa2', [1, 14]]);
    return PriceChannelOverlay;
})();
/**
 * @hidden
 */
export let PriceChannelBucketCalculator = /*@__PURE__*/ (() => {
    class PriceChannelBucketCalculator extends FinancialBucketCalculator {
        constructor(a) {
            super(a);
            this.f = null;
            this.f = a;
        }
        getBucket(a) {
            let b = a * this.b;
            let c = Math.min(b + this.b - 1, this.a.b2.dl.count - 1);
            if (b <= c) {
                let d = -1.7976931348623157E+308;
                let e = 1.7976931348623157E+308;
                let f = 0;
                for (let g = b; g <= c; ++g) {
                    if (!isNaN_(this.f.b6.aaz._inner[g]) && !isNaN_(this.f.b6.aay._inner[g])) {
                        d = Math.max(d, this.f.b6.aaz._inner[g]);
                        e = Math.min(e, this.f.b6.aay._inner[g]);
                        ++f;
                    }
                }
                if (f > 0 && e != 1.7976931348623157E+308 && d != -1.7976931348623157E+308) {
                    return [0.5 * (b + c), e, d];
                }
            }
            return [NaN, NaN, NaN];
        }
    }
    PriceChannelBucketCalculator.$t = markType(PriceChannelBucketCalculator, 'PriceChannelBucketCalculator', FinancialBucketCalculator.$);
    return PriceChannelBucketCalculator;
})();
/**
 * @hidden
 */
export let PriceChannelOverlayView = /*@__PURE__*/ (() => {
    class PriceChannelOverlayView extends FinancialSeriesView {
        constructor(a) {
            super(a);
            this.cb = new Polygon();
            this.cg = new Polyline();
            this.ch = new Polyline();
            this.b6 = null;
            this.b9 = new Polygon();
            this.cc = new Polyline();
            this.cd = new Polyline();
            this.b6 = a;
        }
        get ca() {
            return this.cb;
        }
        get ce() {
            return this.cg;
        }
        get cf() {
            return this.ch;
        }
        au() {
            super.au();
            if (!this.q) {
            }
        }
        b1() {
            return new PriceChannelBucketCalculator(this);
        }
        b7() {
            this.cb.an.clear();
            this.cg.an.clear();
            this.ch.an.clear();
            this.as();
        }
        b8(a, b, c, d, e) {
            for (let f of fromEnum(Flattener.d(a, b, d, this.e.ib))) {
                this.cb.an.add({ $type: Point_$type, x: b(f), y: d(f) });
                this.cg.an.add({ $type: Point_$type, x: b(f), y: d(f) });
            }
            for (let g of fromEnum(Flattener.d(a, c, e, this.e.ib))) {
                this.cb.an.add({ $type: Point_$type, x: c(g), y: e(g) });
                this.ch.an.add({ $type: Point_$type, x: c(g), y: e(g) });
            }
            this.as();
        }
        be() {
            super.be();
            this.b9.an = this.cb.an;
            this.cc.an = this.cg.an;
            this.cd.an = this.ch.an;
            let a = this.br();
            this.b9._fill = a;
            this.b9._opacity = 1;
            this.cc._stroke = a;
            this.cc.ad = this.e.ic + 3;
            this.cd._stroke = a;
            this.cd.ad = this.e.ic + 3;
        }
        a8(a, b) {
            super.a8(a, b);
            if (a.d) {
                if (b) {
                    a.u(this.b9);
                    a.v(this.cc);
                    a.v(this.cd);
                }
                else {
                    a.u(this.cb);
                    a.v(this.cg);
                    a.v(this.ch);
                }
            }
        }
        an(a) {
            super.an(a);
            this.e.bb.exportPolygonData(a, this.cb, "fill", ["Fill"]);
            this.e.bb.exportPolylineData(a, this.cg, "bottom", ["Lower", "Main"]);
            this.e.bb.exportPolylineData(a, this.ch, "top", ["Upper"]);
        }
    }
    PriceChannelOverlayView.$t = markType(PriceChannelOverlayView, 'PriceChannelOverlayView', FinancialSeriesView.$);
    return PriceChannelOverlayView;
})();
