// igniteui-angular-charts/esm2015/lib/PriceVolumeTrendIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { PriceVolumeTrendIndicatorStrategy } from "./PriceVolumeTrendIndicatorStrategy";
import { markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let PriceVolumeTrendIndicator = /*@__PURE__*/ (() => {
    class PriceVolumeTrendIndicator extends StrategyBasedIndicator {
        bi() {
            return new PriceVolumeTrendIndicator();
        }
        get_abn() {
            return new PriceVolumeTrendIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return PriceVolumeTrendIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
    }
    PriceVolumeTrendIndicator.$t = markType(PriceVolumeTrendIndicator, 'PriceVolumeTrendIndicator', StrategyBasedIndicator.$);
    return PriceVolumeTrendIndicator;
})();
