// igniteui-angular-charts/esm2015/lib/RadialBaseChart.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DomainChart } from "./DomainChart";
import { fromEnum, enumGetBox, markType } from "igniteui-angular-core";
import { HorizontalAlignment_$type } from "igniteui-angular-core";
import { VerticalAlignment_$type } from "igniteui-angular-core";
import { Visibility_$type } from "igniteui-angular-core";
import { ValueAxisLabelLocation_$type } from "./ValueAxisLabelLocation";
import { AngleAxisLabelLocation_$type } from "./AngleAxisLabelLocation";
import { AxisLabelSettings } from "./AxisLabelSettings";
import { AxisDefaults } from "./AxisDefaults";
import { TitleSettings } from "./TitleSettings";
import { ScalerParams } from "./ScalerParams";
import { DeviceUtils } from "igniteui-angular-core";
/**
 * @hidden
 */
export let RadialBaseChart = /*@__PURE__*/ (() => {
    class RadialBaseChart extends DomainChart {
        constructor() {
            super(...arguments);
            this.zo = null;
            this.zp = null;
            this.z3 = 0;
            this.z5 = 5;
            this.z4 = 0;
            this.z2 = 5;
            this.aak = 5;
            this.aam = 0;
            this.aal = 5;
            this.aaj = 0;
            this.afd = null;
            this.afk = null;
            this.afb = null;
            this.afc = null;
            this.aae = 0;
            this.aav = 0;
            this.aad = NaN;
            this.aau = NaN;
            this.aag = NaN;
            this.aax = NaN;
            this.aaf = NaN;
            this.aaw = NaN;
            this.aac = NaN;
            this.aat = NaN;
            this.afj = null;
            this.afq = null;
            this.ab5 = null;
            this.ab9 = null;
            this.ab7 = null;
            this.acb = null;
            this.ab0 = null;
            this.ab1 = null;
            this.afe = null;
            this.afl = null;
            this.z6 = DeviceUtils.g(1);
            this.aan = DeviceUtils.g(1);
            this.z7 = DeviceUtils.g(1);
            this.aao = DeviceUtils.g(1);
            this.afg = null;
            this.afn = null;
            this.afh = null;
            this.afo = null;
            this.z8 = DeviceUtils.g(1);
            this.aap = DeviceUtils.g(1);
            this.z9 = DeviceUtils.g(8);
            this.aaq = 0;
            this.afi = null;
            this.afp = null;
            this.aaa = DeviceUtils.g(1);
            this.aar = DeviceUtils.g(1);
            this.ab6 = null;
            this.aca = null;
            this.aff = null;
            this.afm = null;
            this.z1 = 0;
            this.aai = 0;
            this.z0 = NaN;
            this.aah = NaN;
            this.aab = 0;
            this.aas = -90;
            this.zw = false;
            this.zx = false;
            this.ae6 = 1;
            this.af9 = 1;
            this.ae5 = 1;
            this.ae7 = 2;
            this.af7 = 0;
            this.af8 = 1;
            this.agd = 0;
            this.age = 0;
            this.zu = 4;
            this.zk = 4;
            this.ab4 = null;
            this.zg = null;
            this.ab8 = null;
            this.zh = null;
        }
        provideContainer(a) {
            if (a != null) {
                let b = this.zm();
                let c = this.zn();
                if (b != null) {
                    b.provideRenderer(a);
                }
                if (c != null) {
                    c.provideRenderer(a);
                }
            }
            super.provideContainer(a);
        }
        ux() {
            super.ux();
            let a = this.zm();
            let b = this.zn();
            if (a != null && a.az != null) {
                this.aft = a.az.textColor;
            }
            if (b != null && b.az != null) {
                this.af0 = b.az.textColor;
            }
            this.afy = AxisDefaults.j;
            this.af5 = AxisDefaults.j;
            this.afu = AxisDefaults.h;
            this.af1 = AxisDefaults.h;
            this.afx = AxisDefaults.g;
            this.af4 = AxisDefaults.g;
        }
        vo(a, b, c) {
            super.vo(a, b, c);
            switch (a) {
                case "AngleAxisLabelLeftMargin":
                    this.aex((d) => d.leftMargin = this.aa1);
                    break;
                case "AngleAxisLabelTopMargin":
                    this.aex((d) => d.topMargin = this.aa3);
                    break;
                case "AngleAxisLabelRightMargin":
                    this.aex((d) => d.rightMargin = this.aa2);
                    break;
                case "AngleAxisLabelBottomMargin":
                    this.aex((d) => d.bottomMargin = this.aa0);
                    break;
                case "ValueAxisLabelLeftMargin":
                    this.ae2((d) => d.leftMargin = this.abm);
                    break;
                case "ValueAxisLabelTopMargin":
                    this.ae2((d) => d.topMargin = this.abo);
                    break;
                case "ValueAxisLabelRightMargin":
                    this.ae2((d) => d.rightMargin = this.abn);
                    break;
                case "ValueAxisLabelBottomMargin":
                    this.ae2((d) => d.bottomMargin = this.abl);
                    break;
                case "ActualItemsSource":
                case "ShouldDisplayMockData":
                case "ShouldUseSkeletonStyleForMockData":
                    this.afr = this.yg(this.aft);
                    this.afs = this.yg(this.af0);
                    break;
                case "AngleAxisLabelTextColor":
                    this.afr = this.yg(this.aft);
                    break;
                case "ValueAxisLabelTextColor":
                    this.afs = this.yg(this.af0);
                    break;
                case "ActualAngleAxisLabelTextColor":
                    this.aex((d) => d.textColor = this.afr);
                    break;
                case "ActualValueAxisLabelTextColor":
                    this.ae2((d) => d.textColor = this.afs);
                    break;
                case "AngleAxisTitleMargin":
                    this.aez((d) => d.margin = this.abc);
                    break;
                case "ValueAxisTitleMargin":
                    this.ae4((d) => d.margin = this.abx);
                    break;
                case "AngleAxisTitleLeftMargin":
                    this.aez((d) => d.leftMargin = this.abb);
                    break;
                case "ValueAxisTitleLeftMargin":
                    this.ae4((d) => d.leftMargin = this.abw);
                    break;
                case "AngleAxisTitleTopMargin":
                    this.aez((d) => d.topMargin = this.abe);
                    break;
                case "ValueAxisTitleTopMargin":
                    this.ae4((d) => d.topMargin = this.abz);
                    break;
                case "AngleAxisTitleRightMargin":
                    this.aez((d) => d.rightMargin = this.abd);
                    break;
                case "ValueAxisTitleRightMargin":
                    this.ae4((d) => d.rightMargin = this.aby);
                    break;
                case "AngleAxisTitleBottomMargin":
                    this.aez((d) => d.bottomMargin = this.aba);
                    break;
                case "ValueAxisTitleBottomMargin":
                    this.ae4((d) => d.bottomMargin = this.abv);
                    break;
                case "AngleAxisTitleTextColor":
                    this.aez((d) => d.textColor = this.afz);
                    break;
                case "ValueAxisTitleTextColor":
                    this.ae4((d) => d.textColor = this.af6);
                    break;
                case "AngleAxisFormatLabel":
                    this.aev((d) => d.au = this.zq);
                    break;
                case "ValueAxisFormatLabel":
                    this.ae0((d) => d.au = this.zr);
                    break;
            }
        }
        get zq() {
            return this.zo;
        }
        set zq(a) {
            let b = this.zq;
            if (a != b) {
                this.zo = a;
                this.vl("AngleAxisFormatLabel", b, this.zq);
            }
        }
        get zr() {
            return this.zp;
        }
        set zr(a) {
            let b = this.zr;
            if (a != b) {
                this.zp = a;
                this.vl("ValueAxisFormatLabel", b, this.zr);
            }
        }
        get aa1() {
            return this.z3;
        }
        set aa1(a) {
            let b = this.aa1;
            if (a != b) {
                this.z3 = a;
                this.vl("AngleAxisLabelLeftMargin", b, this.aa1);
            }
        }
        get aa3() {
            return this.z5;
        }
        set aa3(a) {
            let b = this.aa3;
            if (a != b) {
                this.z5 = a;
                this.vl("AngleAxisLabelTopMargin", b, this.aa3);
            }
        }
        get aa2() {
            return this.z4;
        }
        set aa2(a) {
            let b = this.aa2;
            if (a != b) {
                this.z4 = a;
                this.vl("AngleAxisLabelRightMargin", b, this.aa2);
            }
        }
        get aa0() {
            return this.z2;
        }
        set aa0(a) {
            let b = this.aa0;
            if (a != b) {
                this.z2 = a;
                this.vl("AngleAxisLabelBottomMargin", b, this.aa0);
            }
        }
        get abm() {
            return this.aak;
        }
        set abm(a) {
            let b = this.abm;
            if (a != b) {
                this.aak = a;
                this.vl("ValueAxisLabelLeftMargin", b, this.abm);
            }
        }
        get abo() {
            return this.aam;
        }
        set abo(a) {
            let b = this.abo;
            if (a != b) {
                this.aam = a;
                this.vl("ValueAxisLabelTopMargin", b, this.abo);
            }
        }
        get abn() {
            return this.aal;
        }
        set abn(a) {
            let b = this.abn;
            if (a != b) {
                this.aal = a;
                this.vl("ValueAxisLabelRightMargin", b, this.abn);
            }
        }
        get abl() {
            return this.aaj;
        }
        set abl(a) {
            let b = this.abl;
            if (a != b) {
                this.aaj = a;
                this.vl("ValueAxisLabelBottomMargin", b, this.abl);
            }
        }
        get aft() {
            return this.afd;
        }
        set aft(a) {
            let b = this.aft;
            if (a != b) {
                this.afd = a;
                this.vl("AngleAxisLabelTextColor", b, this.aft);
            }
        }
        get af0() {
            return this.afk;
        }
        set af0(a) {
            let b = this.af0;
            if (a != b) {
                this.afk = a;
                this.vl("ValueAxisLabelTextColor", b, this.af0);
            }
        }
        get afr() {
            return this.afb;
        }
        set afr(a) {
            let b = this.afb;
            if (a != b) {
                this.afb = a;
                this.vl("ActualAngleAxisLabelTextColor", b, this.afr);
            }
        }
        get afs() {
            return this.afc;
        }
        set afs(a) {
            let b = this.afc;
            if (a != b) {
                this.afc = a;
                this.vl("ActualValueAxisLabelTextColor", b, this.afs);
            }
        }
        get abc() {
            return this.aae;
        }
        set abc(a) {
            let b = this.abc;
            if (a != b) {
                this.aae = a;
                this.vl("AngleAxisTitleMargin", b, this.abc);
            }
        }
        get abx() {
            return this.aav;
        }
        set abx(a) {
            let b = this.abx;
            if (a != b) {
                this.aav = a;
                this.vl("ValueAxisTitleMargin", b, this.abx);
            }
        }
        get abb() {
            return this.aad;
        }
        set abb(a) {
            let b = this.abb;
            if (a != b) {
                this.aad = a;
                this.vl("AngleAxisTitleLeftMargin", b, this.abb);
            }
        }
        get abw() {
            return this.aau;
        }
        set abw(a) {
            let b = this.abw;
            if (a != b) {
                this.aau = a;
                this.vl("ValueAxisTitleLeftMargin", b, this.abw);
            }
        }
        get abe() {
            return this.aag;
        }
        set abe(a) {
            let b = this.abe;
            if (a != b) {
                this.aag = a;
                this.vl("AngleAxisTitleTopMargin", b, this.abe);
            }
        }
        get abz() {
            return this.aax;
        }
        set abz(a) {
            let b = this.abz;
            if (a != b) {
                this.aax = a;
                this.vl("ValueAxisTitleTopMargin", b, this.abz);
            }
        }
        get abd() {
            return this.aaf;
        }
        set abd(a) {
            let b = this.abd;
            if (a != b) {
                this.aaf = a;
                this.vl("AngleAxisTitleRightMargin", b, this.abd);
            }
        }
        get aby() {
            return this.aaw;
        }
        set aby(a) {
            let b = this.aby;
            if (a != b) {
                this.aaw = a;
                this.vl("ValueAxisTitleRightMargin", b, this.aby);
            }
        }
        get aba() {
            return this.aac;
        }
        set aba(a) {
            let b = this.aba;
            if (a != b) {
                this.aac = a;
                this.vl("AngleAxisTitleBottomMargin", b, this.aba);
            }
        }
        get abv() {
            return this.aat;
        }
        set abv(a) {
            let b = this.abv;
            if (a != b) {
                this.aat = a;
                this.vl("ValueAxisTitleBottomMargin", b, this.abv);
            }
        }
        get afz() {
            return this.afj;
        }
        set afz(a) {
            let b = this.afz;
            if (a != b) {
                this.afj = a;
                this.vl("AngleAxisTitleTextColor", b, this.afz);
            }
        }
        get af6() {
            return this.afq;
        }
        set af6(a) {
            let b = this.af6;
            if (a != b) {
                this.afq = a;
                this.vl("ValueAxisTitleTextColor", b, this.af6);
            }
        }
        vp(a, b, c) {
            super.vp(a, b, c);
            switch (a) {
                case "AngleAxisLabelTextStyle":
                    this.aex((d) => d.textStyle = this.acs);
                    break;
                case "ValueAxisLabelTextStyle":
                    this.ae2((d) => d.textStyle = this.ady);
                    break;
                case "AngleAxisTitleTextStyle":
                    this.aez((d) => d.textStyle = this.adg);
                    break;
                case "ValueAxisTitleTextStyle":
                    this.ae4((d) => d.textStyle = this.aem);
                    break;
            }
        }
        get acs() {
            return this.ab5;
        }
        set acs(a) {
            let b = this.acs;
            if (a != b) {
                this.ab5 = a;
                this.vl("AngleAxisLabelTextStyle", b, this.acs);
            }
        }
        get ady() {
            return this.ab9;
        }
        set ady(a) {
            let b = this.ady;
            if (a != b) {
                this.ab9 = a;
                this.vl("ValueAxisLabelTextStyle", b, this.ady);
            }
        }
        get adg() {
            return this.ab7;
        }
        set adg(a) {
            let b = this.adg;
            if (a != b) {
                this.ab7 = a;
                this.vl("AngleAxisTitleTextStyle", b, this.adg);
            }
        }
        get aem() {
            return this.acb;
        }
        set aem(a) {
            let b = this.aem;
            if (a != b) {
                this.acb = a;
                this.vl("ValueAxisTitleTextStyle", b, this.aem);
            }
        }
        aes(a) {
            this.aep(a, (b) => b.leftMargin = this.aa1);
            this.aep(a, (b) => b.topMargin = this.aa3);
            this.aep(a, (b) => b.rightMargin = this.aa2);
            this.aep(a, (b) => b.bottomMargin = this.aa0);
            this.aep(a, (b) => b.textColor = this.afr);
            this.aeq(a, (b) => b.textColor = this.afz);
            this.aeq(a, (b) => b.leftMargin = this.abb);
            this.aeq(a, (b) => b.topMargin = this.abe);
            this.aeq(a, (b) => b.rightMargin = this.abd);
            this.aeq(a, (b) => b.bottomMargin = this.aba);
            this.aep(a, (b) => b.textStyle = this.acs);
            this.aeq(a, (b) => b.textStyle = this.adg);
        }
        aeu(a) {
            this.aep(a, (b) => b.leftMargin = this.abm);
            this.aep(a, (b) => b.topMargin = this.abo);
            this.aep(a, (b) => b.rightMargin = this.abn);
            this.aep(a, (b) => b.bottomMargin = this.abl);
            this.aep(a, (b) => b.textColor = this.afs);
            this.aeq(a, (b) => b.textColor = this.af6);
            this.aeq(a, (b) => b.leftMargin = this.abw);
            this.aeq(a, (b) => b.topMargin = this.abz);
            this.aeq(a, (b) => b.rightMargin = this.aby);
            this.aeq(a, (b) => b.bottomMargin = this.abv);
            this.aep(a, (b) => b.textStyle = this.ady);
            this.aeq(a, (b) => b.textStyle = this.aem);
        }
        wf(a) {
            let b = this.bi(a);
            let c = this.dg(a, b);
            if (c == null) {
                return;
            }
            c.itemsSource = this.h2(a);
            if (a.highlightedData != null) {
                c.highlightedItemsSource = this.h3(a);
            }
            c.name = this.qk(a);
            c.title = this.ql(a, b);
            this.dataChart.series.add(c);
        }
        aev(a) {
            this.aew(a, null);
        }
        aew(a, b) {
            a(this.zm());
        }
        ae1(a, b) {
            a(this.zn());
        }
        ae0(a) {
            this.ae1(a, true);
        }
        uh(a, b) {
            this.aev(a);
            this.ae1(a, b);
        }
        aez(a) {
            this.aeq(this.zm(), a);
        }
        ae4(a) {
            this.aeq(this.zn(), a);
        }
        ae2(a) {
            this.aep(this.zn(), a);
        }
        aex(a) {
            this.aep(this.zm(), a);
        }
        abf(a) {
            let b = this.zm();
            return b.ek(a, new ScalerParams(1, this.yp, this.yo, b.ch));
        }
        abh(a) {
            let b = this.zm();
            return b.en(a, new ScalerParams(1, this.yp, this.yo, b.ch));
        }
        abg(a) {
            let b = this.zn();
            return b.ek(a, new ScalerParams(1, this.yp, this.yo, b.ch));
        }
        abi(a) {
            let b = this.zn();
            return b.en(a, new ScalerParams(1, this.yp, this.yo, b.ch));
        }
        wc(a) {
            super.wc(a);
            if (a == null) {
                return;
            }
            if (a.axes != null) {
                let b = this.zm();
                let c = this.zn();
                if (a.axes.contains(b)) {
                    a.axes.remove(b);
                }
                if (a.axes.contains(c)) {
                    a.axes.remove(c);
                }
            }
        }
        static adj(a, b) {
            for (let c of fromEnum(b)) {
                let d = c.findMatchingHint(16);
                if (d != null) {
                    let e = d.path;
                    return e;
                }
            }
            return null;
        }
        us(a, b) {
            super.us(a, b);
            if (a == this.dataChart) {
                this.dataChart.axes.add(this.zm());
                this.dataChart.axes.add(this.zn());
            }
        }
        aer(a) {
            if (a != null && this.lu != null && a.fm == null) {
                a.provideRenderer(this.lu);
            }
            a.jx = this.afu;
            a.es = this.aa4;
            a.jy = this.afv;
            a.ev = this.aa5;
            a.j0 = this.afx;
            a.fh = this.aa6;
            a.jz = this.afw;
            a.fi = this.aa7;
            a.j1 = this.afy;
            a.fj = this.aa8;
            a.c4 = true;
            a.cu = this.hg;
            a.cw = this.hg;
            let b = a == this.zm();
            if (b) {
                a.g7 = this.ac7;
            }
            a.f1 = this.ab2;
            a.gh = this.acj;
            a.ah = this.zi;
            this.aep(a, (c) => c.angle = this.aaz);
            this.aep(a, (c) => {
                if (c.extent != this.aay) {
                    c.extent = this.aay;
                }
            });
            this.aep(a, (c) => c.horizontalAlignment = this.ae8);
            this.aep(a, (c) => c.verticalAlignment = this.aga);
            this.aep(a, (c) => c.visibility = this.agf);
            this.aep(a, (c) => c.location = this.zs(this.zl));
            this.aeq(a, (c) => c.horizontalAlignment = this.ae9);
            this.aeq(a, (c) => c.angle = this.aa9);
            this.aes(a);
        }
        aet(a) {
            if (a == null) {
                return;
            }
            if (this.lu != null && a.fm == null) {
                a.provideRenderer(this.lu);
            }
            a.cu = this.hg;
            a.cw = this.hg;
            this.aep(a, (b) => b.angle = this.abk);
            this.aep(a, (b) => {
                if (b.extent != this.abj) {
                    b.extent = this.abj;
                }
            });
            this.aep(a, (b) => b.horizontalAlignment = this.afa);
            this.aep(a, (b) => b.verticalAlignment = this.agb);
            this.aep(a, (b) => b.visibility = this.agg);
            this.aep(a, (b) => b.location = this.zt(this.zv));
            this.aeq(a, (b) => b.verticalAlignment = this.agc);
            this.aeq(a, (b) => b.angle = this.abu);
            a.jx = this.af1;
            a.es = this.abp;
            a.jy = this.af2;
            a.ev = this.abq;
            a.j0 = this.af4;
            a.fh = this.abr;
            a.jz = this.af3;
            a.fi = this.abs;
            a.j1 = this.af5;
            a.fj = this.abt;
            let b = a == this.zn();
            if (b) {
                a.g7 = this.aed;
            }
            a.gh = this.adp;
            a.ah = this.zj;
            this.aeu(a);
        }
        up() {
            super.up();
            this.aer(this.zm());
            this.aet(this.zn());
        }
        aep(a, b) {
            if (a == null) {
                return;
            }
            if (a.az == null) {
                a.az = new AxisLabelSettings();
                a.az.angle = a.d6();
                a.az.bottomMargin = a.d7();
                a.az.leftMargin = a.d8();
                a.az.rightMargin = a.d9();
                a.az.topMargin = a.ea();
                a.az.horizontalAlignment = a.js();
                a.az.location = a.a0();
                a.az.verticalAlignment = a.kl();
            }
            b(a.az);
        }
        aeq(a, b) {
            if (a == null) {
                return;
            }
            if (a.bf == null) {
                a.bf = new TitleSettings();
                a.bf.bottomMargin = a.ec();
                a.bf.leftMargin = a.ed();
                a.bf.rightMargin = a.ee();
                a.bf.topMargin = a.ef();
                a.bf.location = a.a1();
                a.bf.angle = a.eb();
                a.bf.horizontalAlignment = a.jt();
                a.bf.verticalAlignment = a.km();
            }
            b(a.bf);
        }
        vl(a, b, c) {
            super.vl(a, b, c);
            switch (a) {
                case "AngleAxisLabelAngle":
                    this.aex((d) => d.angle = this.aaz);
                    break;
                case "ValueAxisLabelAngle":
                    this.ae2((d) => d.angle = this.abk);
                    break;
                case "AngleAxisLabelHorizontalAlignment":
                    this.aex((d) => d.horizontalAlignment = this.ae8);
                    break;
                case "ValueAxisLabelHorizontalAlignment":
                    this.ae2((d) => d.horizontalAlignment = this.afa);
                    break;
                case "AngleAxisLabelVerticalAlignment":
                    this.aex((d) => d.verticalAlignment = this.aga);
                    break;
                case "ValueAxisLabelVerticalAlignment":
                    this.ae2((d) => d.verticalAlignment = this.agb);
                    break;
                case "AngleAxisLabelVisibility":
                    this.aex((d) => d.visibility = this.agf);
                    break;
                case "ValueAxisLabelVisibility":
                    this.ae2((d) => d.visibility = this.agg);
                    break;
                case "AngleAxisLabel":
                    this.aev((d) => d.f1 = this.ab2);
                    break;
                case "ValueAxisLabel":
                    this.ae0((d) => d.f1 = this.ab3);
                    break;
                case "AngleAxisLabelFormat":
                    this.aev((d) => d.gh = this.acj);
                    break;
                case "AngleAxisLabelFormatSpecifiers":
                    this.aev((d) => d.ah = this.zi);
                    break;
                case "ValueAxisLabelFormat":
                    this.ae0((d) => d.gh = this.adp);
                    break;
                case "ValueAxisLabelFormatSpecifiers":
                    this.ae0((d) => d.ah = this.zj);
                    break;
                case "AngleAxisTitleAngle":
                    this.aez((d) => d.angle = this.aa9);
                    break;
                case "ValueAxisTitleAngle":
                    this.ae4((d) => d.angle = this.abu);
                    break;
                case "AngleAxisTitleAlignment":
                    this.aez((d) => d.horizontalAlignment = this.ae9);
                    break;
                case "ValueAxisTitleAlignment":
                    this.ae4((d) => d.verticalAlignment = this.agc);
                    break;
                case "AngleAxisTitle":
                    this.aev((d) => d.g7 = this.ac7);
                    break;
                case "ValueAxisTitle":
                    this.ae1((d) => d.g7 = this.aed, false);
                    break;
                case "AngleAxisMajorStroke":
                    this.aew((d) => d.jx = this.afu, a);
                    break;
                case "ValueAxisMajorStroke":
                    this.ae0((d) => d.jx = this.af1);
                    break;
                case "AngleAxisMajorStrokeThickness":
                    this.aew((d) => d.es = this.aa4, a);
                    break;
                case "ValueAxisMajorStrokeThickness":
                    this.ae0((d) => d.es = this.abp);
                    break;
                case "AngleAxisMinorStroke":
                    this.aev((d) => d.jy = this.afv);
                    break;
                case "ValueAxisMinorStroke":
                    this.ae0((d) => d.jy = this.af2);
                    break;
                case "AngleAxisMinorStrokeThickness":
                    this.aev((d) => d.ev = this.aa5);
                    break;
                case "ValueAxisMinorStrokeThickness":
                    this.ae0((d) => d.ev = this.abq);
                    break;
                case "AngleAxisTickLength":
                    this.aev((d) => d.fi = this.aa7);
                    break;
                case "ValueAxisTickLength":
                    this.ae0((d) => d.fi = this.abs);
                    break;
                case "AngleAxisTickStroke":
                    this.aev((d) => d.j1 = this.afy);
                    break;
                case "ValueAxisTickStroke":
                    this.ae0((d) => d.j1 = this.af5);
                    break;
                case "AngleAxisTickStrokeThickness":
                    this.aev((d) => d.fj = this.aa8);
                    break;
                case "ValueAxisTickStrokeThickness":
                    this.ae0((d) => d.fj = this.abt);
                    break;
                case "AngleAxisStroke":
                    this.aev((d) => d.j0 = this.afx);
                    break;
                case "ValueAxisStroke":
                    this.ae0((d) => d.j0 = this.af4);
                    break;
                case "AngleAxisStrokeThickness":
                    this.aev((d) => d.fh = this.aa6);
                    break;
                case "ValueAxisStrokeThickness":
                    this.ae0((d) => d.fh = this.abr);
                    break;
                case "AngleAxisExtent":
                    this.aex((d) => d.extent = this.aay);
                    break;
                case "ValueAxisExtent":
                    this.ae2((d) => d.extent = this.abj);
                    break;
                case "AngleAxisStrip":
                    this.aev((d) => d.jz = this.afw);
                    break;
                case "ValueAxisStrip":
                    this.ae0((d) => d.jz = this.af3);
                    break;
                case "AngleAxisInverted":
                    this.aev((d) => d.ch = this.zy);
                    break;
                case "ValueAxisInverted":
                    this.ae0((d) => d.ch = this.zz);
                    break;
                case "ValueAxisLabelLocation":
                    this.ae3();
                    break;
                case "AngleAxisLabelLocation":
                    this.aey();
                    break;
            }
        }
        ae3() {
            let a = this.zt(this.zv);
            if (a == 2 || a == 7) {
                this.afa = 2;
            }
            else {
                this.afa = 0;
            }
            this.ae2((b) => b.location = a);
        }
        aey() {
            let a = this.zs(this.zl);
            if (a == 1 || a == 5) {
                this.aga = 0;
            }
            else {
                this.aga = 2;
            }
            this.aex((b) => b.location = a);
        }
        zt(a) {
            if (a == 0) {
                return 2;
            }
            else if (a == 1) {
                return 3;
            }
            else if (a == 3) {
                return 7;
            }
            else if (a == 2) {
                return 6;
            }
            else {
                if (this.cf == 1) {
                    return 3;
                }
                else {
                    return 2;
                }
            }
        }
        zs(a) {
            if (a == 1) {
                return 1;
            }
            else if (a == 0) {
                return 0;
            }
            else if (a == 3) {
                return 5;
            }
            else if (a == 2) {
                return 4;
            }
            else {
                return 1;
            }
        }
        get ab2() {
            return this.ab0;
        }
        set ab2(a) {
            let b = this.ab2;
            if (a != b) {
                this.ab0 = a;
                this.vl("AngleAxisLabel", b, this.ab2);
            }
        }
        get ab3() {
            return this.ab1;
        }
        set ab3(a) {
            let b = this.ab3;
            if (a != b) {
                this.ab1 = a;
                this.vl("ValueAxisLabel", b, this.ab3);
            }
        }
        get afu() {
            return this.afe;
        }
        set afu(a) {
            let b = this.afu;
            if (a != b) {
                this.afe = a;
                this.vl("AngleAxisMajorStroke", b, this.afu);
            }
        }
        get af1() {
            return this.afl;
        }
        set af1(a) {
            let b = this.af1;
            if (a != b) {
                this.afl = a;
                this.vl("ValueAxisMajorStroke", b, this.af1);
            }
        }
        get aa4() {
            return this.z6;
        }
        set aa4(a) {
            let b = this.aa4;
            if (a != b) {
                this.z6 = a;
                this.vl("AngleAxisMajorStrokeThickness", b, this.aa4);
            }
        }
        get abp() {
            return this.aan;
        }
        set abp(a) {
            let b = this.abp;
            if (a != b) {
                this.aan = a;
                this.vl("ValueAxisMajorStrokeThickness", b, this.abp);
            }
        }
        get aa5() {
            return this.z7;
        }
        set aa5(a) {
            let b = this.aa5;
            if (a != b) {
                this.z7 = a;
                this.vl("AngleAxisMinorStrokeThickness", b, this.aa5);
            }
        }
        get abq() {
            return this.aao;
        }
        set abq(a) {
            let b = this.abq;
            if (a != b) {
                this.aao = a;
                this.vl("ValueAxisMinorStrokeThickness", b, this.abq);
            }
        }
        get afw() {
            return this.afg;
        }
        set afw(a) {
            let b = this.afw;
            if (a != b) {
                this.afg = a;
                this.vl("AngleAxisStrip", b, this.afw);
            }
        }
        get af3() {
            return this.afn;
        }
        set af3(a) {
            let b = this.af3;
            if (a != b) {
                this.afn = a;
                this.vl("ValueAxisStrip", b, this.af3);
            }
        }
        get afx() {
            return this.afh;
        }
        set afx(a) {
            let b = this.afx;
            if (a != b) {
                this.afh = a;
                this.vl("AngleAxisStroke", b, this.afx);
            }
        }
        get af4() {
            return this.afo;
        }
        set af4(a) {
            let b = this.af4;
            if (a != b) {
                this.afo = a;
                this.vl("ValueAxisStroke", b, this.af4);
            }
        }
        get aa6() {
            return this.z8;
        }
        set aa6(a) {
            let b = this.aa6;
            if (a != b) {
                this.z8 = a;
                this.vl("AngleAxisStrokeThickness", b, this.aa6);
            }
        }
        get abr() {
            return this.aap;
        }
        set abr(a) {
            let b = this.abr;
            if (a != b) {
                this.aap = a;
                this.vl("ValueAxisStrokeThickness", b, this.abr);
            }
        }
        get aa7() {
            return this.z9;
        }
        set aa7(a) {
            let b = this.aa7;
            if (a != b) {
                this.z9 = a;
                this.vl("AngleAxisTickLength", b, this.aa7);
            }
        }
        get abs() {
            return this.aaq;
        }
        set abs(a) {
            let b = this.abs;
            if (a != b) {
                this.aaq = a;
                this.vl("ValueAxisTickLength", b, this.abs);
            }
        }
        get afy() {
            return this.afi;
        }
        set afy(a) {
            let b = this.afy;
            if (a != b) {
                this.afi = a;
                this.vl("AngleAxisTickStroke", b, this.afy);
            }
        }
        get af5() {
            return this.afp;
        }
        set af5(a) {
            let b = this.af5;
            if (a != b) {
                this.afp = a;
                this.vl("ValueAxisTickStroke", b, this.af5);
            }
        }
        get aa8() {
            return this.aaa;
        }
        set aa8(a) {
            let b = this.aa8;
            if (a != b) {
                this.aaa = a;
                this.vl("AngleAxisTickStrokeThickness", b, this.aa8);
            }
        }
        get abt() {
            return this.aar;
        }
        set abt(a) {
            let b = this.abt;
            if (a != b) {
                this.aar = a;
                this.vl("ValueAxisTickStrokeThickness", b, this.abt);
            }
        }
        get ac7() {
            return this.ab6;
        }
        set ac7(a) {
            let b = this.ac7;
            if (a != b) {
                this.ab6 = a;
                this.vl("AngleAxisTitle", b, this.ac7);
            }
        }
        get aed() {
            return this.aca;
        }
        set aed(a) {
            let b = this.aed;
            if (a != b) {
                this.aca = a;
                this.vl("ValueAxisTitle", b, this.aed);
            }
        }
        get afv() {
            return this.aff;
        }
        set afv(a) {
            let b = this.afv;
            if (a != b) {
                this.aff = a;
                this.vl("AngleAxisMinorStroke", b, this.afv);
            }
        }
        get af2() {
            return this.afm;
        }
        set af2(a) {
            let b = this.af2;
            if (a != b) {
                this.afm = a;
                this.vl("ValueAxisMinorStroke", b, this.af2);
            }
        }
        get aaz() {
            return this.z1;
        }
        set aaz(a) {
            let b = this.aaz;
            if (a != b) {
                this.z1 = a;
                this.vl("AngleAxisLabelAngle", b, this.aaz);
            }
        }
        get abk() {
            return this.aai;
        }
        set abk(a) {
            let b = this.abk;
            if (a != b) {
                this.aai = a;
                this.vl("ValueAxisLabelAngle", b, this.abk);
            }
        }
        get aay() {
            return this.z0;
        }
        set aay(a) {
            let b = this.aay;
            if (a != b) {
                this.z0 = a;
                this.vl("AngleAxisExtent", b, this.aay);
            }
        }
        get abj() {
            return this.aah;
        }
        set abj(a) {
            let b = this.abj;
            if (a != b) {
                this.aah = a;
                this.vl("ValueAxisExtent", b, this.abj);
            }
        }
        get aa9() {
            return this.aab;
        }
        set aa9(a) {
            let b = this.aa9;
            if (a != b) {
                this.aab = a;
                this.vl("AngleAxisTitleAngle", b, this.aa9);
            }
        }
        get abu() {
            return this.aas;
        }
        set abu(a) {
            let b = this.abu;
            if (a != b) {
                this.aas = a;
                this.vl("ValueAxisTitleAngle", b, this.abu);
            }
        }
        get zy() {
            return this.zw;
        }
        set zy(a) {
            let b = this.zy;
            if (a != b) {
                this.zw = a;
                this.vl("AngleAxisInverted", b, this.zy);
            }
        }
        get zz() {
            return this.zx;
        }
        set zz(a) {
            let b = this.zz;
            if (a != b) {
                this.zx = a;
                this.vl("ValueAxisInverted", b, this.zz);
            }
        }
        get ae9() {
            return this.ae6;
        }
        set ae9(a) {
            let b = this.ae9;
            if (a != b) {
                this.ae6 = a;
                this.vl("AngleAxisTitleAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.ae9));
            }
        }
        get agc() {
            return this.af9;
        }
        set agc(a) {
            let b = this.agc;
            if (a != b) {
                this.af9 = a;
                this.vl("ValueAxisTitleAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.agc));
            }
        }
        get ae8() {
            return this.ae5;
        }
        set ae8(a) {
            let b = this.ae8;
            if (a != b) {
                this.ae5 = a;
                this.vl("AngleAxisLabelHorizontalAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.ae8));
            }
        }
        get_afa() {
            return this.ae7;
        }
        set_afa(a) {
            let b = this.afa;
            if (a != b) {
                this.ae7 = a;
                this.vl("ValueAxisLabelHorizontalAlignment", enumGetBox(HorizontalAlignment_$type, b), enumGetBox(HorizontalAlignment_$type, this.afa));
            }
        }
        get afa() {
            return this.get_afa();
        }
        set afa(a) {
            this.set_afa(a);
        }
        get aga() {
            return this.af7;
        }
        set aga(a) {
            let b = this.aga;
            if (a != b) {
                this.af7 = a;
                this.vl("AngleAxisLabelVerticalAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.aga));
            }
        }
        get agb() {
            return this.af8;
        }
        set agb(a) {
            let b = this.agb;
            if (a != b) {
                this.af8 = a;
                this.vl("ValueAxisLabelVerticalAlignment", enumGetBox(VerticalAlignment_$type, b), enumGetBox(VerticalAlignment_$type, this.agb));
            }
        }
        get agf() {
            return this.agd;
        }
        set agf(a) {
            let b = this.agf;
            if (a != b) {
                this.agd = a;
                this.vl("AngleAxisLabelVisibility", enumGetBox(Visibility_$type, b), enumGetBox(Visibility_$type, this.agf));
            }
        }
        get agg() {
            return this.age;
        }
        set agg(a) {
            let b = this.agg;
            if (a != b) {
                this.age = a;
                this.vl("ValueAxisLabelVisibility", enumGetBox(Visibility_$type, b), enumGetBox(Visibility_$type, this.agg));
            }
        }
        get_zv() {
            return this.zu;
        }
        set_zv(a) {
            let b = this.zv;
            if (a != b) {
                this.zu = a;
                this.vl("ValueAxisLabelLocation", enumGetBox(ValueAxisLabelLocation_$type, b), enumGetBox(ValueAxisLabelLocation_$type, this.zv));
            }
        }
        get zv() {
            return this.get_zv();
        }
        set zv(a) {
            this.set_zv(a);
        }
        get_zl() {
            return this.zk;
        }
        set_zl(a) {
            let b = this.zl;
            if (a != b) {
                this.zk = a;
                this.vl("AngleAxisLabelLocation", enumGetBox(AngleAxisLabelLocation_$type, b), enumGetBox(AngleAxisLabelLocation_$type, this.zl));
            }
        }
        get zl() {
            return this.get_zl();
        }
        set zl(a) {
            this.set_zl(a);
        }
        get acj() {
            return this.ab4;
        }
        set acj(a) {
            let b = this.ab4;
            if (a != b) {
                this.ab4 = a;
                this.vl("AngleAxisLabelFormat", b, this.ab4);
            }
        }
        get zi() {
            return this.zg;
        }
        set zi(a) {
            let b = this.zg;
            if (a != b) {
                this.zg = a;
                this.vl("AngleAxisLabelFormatSpecifiers", b, this.zg);
            }
        }
        get adp() {
            return this.ab8;
        }
        set adp(a) {
            let b = this.ab8;
            if (a != b) {
                this.ab8 = a;
                this.vl("ValueAxisLabelFormat", b, this.ab8);
            }
        }
        get zj() {
            return this.zh;
        }
        set zj(a) {
            let b = this.zh;
            if (a != b) {
                this.zh = a;
                this.vl("ValueAxisLabelFormatSpecifiers", b, this.zh);
            }
        }
    }
    RadialBaseChart.$t = markType(RadialBaseChart, 'RadialBaseChart', DomainChart.$);
    return RadialBaseChart;
})();
