// igniteui-angular-charts/esm2015/lib/RadialColumnSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnchoredRadialSeries } from "./AnchoredRadialSeries";
import { DependencyProperty } from "igniteui-angular-core";
import { Defaults } from "./Defaults";
import { Point_$type, runOn, markType, markDep, fromEnum } from "igniteui-angular-core";
import { Pool$1 } from "igniteui-angular-core";
import { Path } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { ScalerParams } from "./ScalerParams";
import { PathGeometry } from "igniteui-angular-core";
import { PathFigure } from "igniteui-angular-core";
import { PolyLineSegment } from "igniteui-angular-core";
import { ViewportUtils } from "./ViewportUtils";
import { PropertyMetadata } from "igniteui-angular-core";
import { isNaN_ } from "igniteui-angular-core";
import { AnchoredRadialSeriesView } from "./AnchoredRadialSeriesView";
/**
 * @hidden
 */
export let RadialColumnSeries = /*@__PURE__*/ (() => {
    class RadialColumnSeries extends AnchoredRadialSeries {
        constructor() {
            super();
            this.ab = RadialColumnSeries.$;
        }
        bi() {
            return new RadialColumnSeries();
        }
        get_ex() {
            return true;
        }
        get_fj() {
            return true;
        }
        ji() {
            if (!isNaN_(this.jh)) {
                return super.ji();
            }
            return Defaults.h;
        }
        cq() {
            return new RadialColumnSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.add = a;
        }
        ac() {
            super.ac();
            this.rn(false);
        }
        q5(a, b, c, d) {
            super.q5(a, b, c, d);
            switch (b) {
                case RadialColumnSeries.$$p[0]:
                case RadialColumnSeries.$$p[1]:
                    this.qz();
                    break;
            }
        }
        preferredCategoryMode(a) {
            return a != null && a == this.yr ? 2 : 0;
        }
        o3(a, b) {
            super.o3(a, b);
            let c = b;
            if (a) {
                c.c6.count = 0;
            }
        }
        get_eb() {
            return true;
        }
        iu() {
            return this.yr.k5(this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct));
        }
        aa2(a, b) {
            super.aa2(a, b);
            let c = a.f;
            let d = b.by;
            let e = b.bx;
            let f = this.getEffectiveViewport1(b);
            let g = this.wd(b);
            let h = this.yw.k2(e, d, f, g);
            let i = b;
            let j = this.yw;
            let k = Math.max(0, 0.5 * j.no);
            k = Math.max(k, h.minLength * 0.9);
            let l = ViewportUtils.d(k, d, e, f);
            let m = this.yr.k5(d, e, f);
            let n = Math.max(0.1 * this.yw.np, k);
            let o = this.yz.d(0, n, d, e, f);
            let p = this.yz.e(0, n, d, e, f);
            let q = this.yz.d(m, n, d, e, f);
            let r = this.yz.e(m, n, d, e, f);
            let s = Math.sqrt((o - q) * (o - q) + (p - r) * (p - r));
            let t = { $type: Point_$type, x: ViewportUtils.e(0.5, d, e, f), y: ViewportUtils.i(0.5, d, e, f) };
            let u = this.abx.count;
            this.y7.a2(this, this.zo(), runOn(this, this.yf), this.zx(b), this.zy(b));
            this.y7.r = this.ade;
            this.y7.s = this.adf;
            this.y7.k = this.ade;
            this.y7.l = this.adf;
            let v = false;
            let w = this.y7.b;
            if (w != null) {
                v = true;
            }
            let x = this.yr;
            let y = new ScalerParams(1, b.by, b.bx, x.ch);
            y.c = this.getEffectiveViewport();
            for (let z = 0; z < c.count; ++z) {
                let aa = Math.min(a.f._inner[z][2], h.maxLength * 1.1);
                let ab = this.yz.d(a.f._inner[z][0], aa, d, e, f);
                let ac = this.yz.e(a.f._inner[z][0], aa, d, e, f);
                let ad = Math.sqrt((ab - t.x) * (ab - t.x) + (ac - t.y) * (ac - t.y));
                let ae = ad;
                let af = l;
                if (v) {
                    this.act(c, z, u, x, y, b.q);
                }
                let ag = this.y7.k;
                let ah = this.y7.l;
                let ai = a.f._inner[z][0] - (Math.PI / 2);
                let aj = i.c6.item(z);
                let ak = 0 - (s / 2);
                let al = (s / 2);
                let am = Math.max(af, ae);
                let an = Math.min(af, ae);
                let ao = Math.cos(ai);
                let ap = Math.sin(ai);
                let aq = { $type: Point_$type, x: t.x + (ak * ao - am * ap), y: t.y + (ak * ap + am * ao) };
                let ar = { $type: Point_$type, x: t.x + (al * ao - am * ap), y: t.y + (al * ap + am * ao) };
                let as = { $type: Point_$type, x: t.x + (al * ao - an * ap), y: t.y + (al * ap + an * ao) };
                let at = { $type: Point_$type, x: t.x + (ak * ao - an * ap), y: t.y + (ak * ap + an * ao) };
                let au = new PathGeometry();
                let av = new PathFigure();
                av._startPoint = aq;
                av._isClosed = true;
                let aw = new PolyLineSegment();
                aw._points.add(ar);
                aw._points.add(as);
                aw._points.add(at);
                av._segments.add(aw);
                au.c.add(av);
                aj.an = au;
                this.y7.ad(aj, false, false, false, false);
            }
            i.c6.count = c.count;
            this.add.db();
        }
    }
    RadialColumnSeries.$t = markType(RadialColumnSeries, 'RadialColumnSeries', AnchoredRadialSeries.$);
    RadialColumnSeries.$$p = markDep(DependencyProperty, PropertyMetadata, RadialColumnSeries, 'raisePropertyChanged', ['RadiusX:ade:adi', [1, 0], 'RadiusY:adf:adj', [1, 0]]);
    return RadialColumnSeries;
})();
/**
 * @hidden
 */
export let RadialColumnSeriesView = /*@__PURE__*/ (() => {
    class RadialColumnSeriesView extends AnchoredRadialSeriesView {
        constructor(a) {
            super(a);
            this.c5 = null;
            this.c6 = null;
            this.c7 = null;
            this.dd = new Path();
            this.c5 = a;
            this.c6 = ((() => {
                let $ret = new Pool$1(Path.$);
                $ret.create = runOn(this, this.de);
                $ret.activate = runOn(this, this.c8);
                $ret.disactivate = runOn(this, this.da);
                $ret.destroy = runOn(this, this.c9);
                return $ret;
            })());
        }
        au() {
            super.au();
            this.c7 = new List$1(Path.$, 0);
            if (!this.q) {
                this.cw.w2 = Defaults.c;
                this.cw.jn = Defaults.i;
            }
        }
        de() {
            let a = new Path();
            this.c7.add(a);
            a._visibility = 1;
            return a;
        }
        c8(a) {
            a._visibility = 0;
        }
        da(a) {
            a._visibility = 1;
        }
        c9(a) {
            this.c7.remove(a);
        }
        dc(a, b, c, d, e, f) {
            this.as();
        }
        bg(a, b) {
            super.bg(a, b);
            let c = a;
        }
        bh(a, b) {
            super.bh(a, b);
            let c = a;
        }
        ac(a) {
            return this.c7._inner[a];
        }
        ab(a) {
            let b = this.c7._inner[a];
            this.dd._visibility = b._visibility;
            this.dd.an = b.an;
            let c = this.bs(a);
            this.dd._fill = c;
            this.dd._stroke = c;
            this.dd.ad = this.e.ic + 3;
            return this.dd;
        }
        a8(a, b) {
            super.a8(a, b);
            if (a.d) {
                for (let c = 0; c < this.c7.count; c++) {
                    let d = this.aa(c, b);
                    this.bf(d, c, b);
                    a.t(d);
                }
            }
        }
        an(a) {
            super.an(a);
            let b = 0;
            for (let c of fromEnum(this.c6.a)) {
                this.e.bb.exportPathData(a, c, "column" + b, ["Main", "Fill"]);
                b++;
            }
        }
        db() {
            this.as();
        }
    }
    RadialColumnSeriesView.$t = markType(RadialColumnSeriesView, 'RadialColumnSeriesView', AnchoredRadialSeriesView.$);
    return RadialColumnSeriesView;
})();
