// igniteui-angular-charts/esm2015/lib/RangeColumnSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalRangeCategorySeries } from "./HorizontalRangeCategorySeries";
import { DependencyProperty } from "igniteui-angular-core";
import { Defaults } from "./Defaults";
import { typeCast, runOn, markType, markDep, fromEnum } from "igniteui-angular-core";
import { Rect } from "igniteui-angular-core";
import { ScalerParams } from "./ScalerParams";
import { List$1 } from "igniteui-angular-core";
import { Pool$1 } from "igniteui-angular-core";
import { Rectangle } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { truncate, isNaN_, isInfinity } from "igniteui-angular-core";
import { RangeCategorySeriesView } from "./RangeCategorySeriesView";
/**
 * @hidden
 */
export let RangeColumnSeries = /*@__PURE__*/ (() => {
    class RangeColumnSeries extends HorizontalRangeCategorySeries {
        constructor() {
            super();
            this.ab = RangeColumnSeries.$;
        }
        bi() {
            return new RangeColumnSeries();
        }
        get_ex() {
            return true;
        }
        get_fj() {
            return true;
        }
        ji() {
            if (!isNaN_(this.jh)) {
                return super.ji();
            }
            return Defaults.h;
        }
        cq() {
            return new RangeColumnSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.abj = a;
        }
        get_eb() {
            return true;
        }
        preferredCategoryMode(a) {
            return 2;
        }
        wh(a) {
            if (this.xAxis == null || this.yAxis == null) {
                return Rect.empty;
            }
            let b = this.v1(a);
            let c = this.ct.by;
            let d = this.ct.bx;
            let e = this.getEffectiveViewport1(this.ct);
            let f = new ScalerParams(0, c, d, this.yAxis.ch, e);
            f.b = this.iw();
            let g = this.am(this.yk(), this.yl.f, this.zm(this.ct), this.zn(this.ct), a, true);
            if (g == null) {
                return Rect.empty;
            }
            let h = g[0];
            let i = g[1];
            let j = Math.abs(b.x - h[0]);
            let k = Math.abs(b.x - i[0]);
            let l = this.xAxis.k5(c, d, e);
            if (j < k) {
                let m = h[0] - 0.5 * l;
                let n = h[1];
                let o = h[2];
                return new Rect(0, m, Math.min(n, o), l, Math.max(n, o) - Math.min(n, o));
            }
            else {
                let p = i[0] - 0.5 * l;
                let q = i[1];
                let r = i[2];
                return new Rect(0, p, Math.min(q, r), l, Math.max(q, r) - Math.min(q, r));
            }
        }
        gu(a, b) {
            if (this.gx(a, b)) {
                return true;
            }
            if (this.w9(a, b)) {
                return true;
            }
            return false;
        }
        o3(a, b) {
            super.o3(a, b);
            let c = b;
            if (a && c.c8 != null) {
                c.c8.count = 0;
            }
        }
        iu() {
            return this.xAxis.k5(this.ct.by, this.ct.bx, this.getEffectiveViewport1(this.ct));
        }
        z4(a, b) {
            super.z4(a, b);
            let c = typeCast(RangeColumnSeriesView.$, b);
            let d = a.f;
            let e = b.by;
            let f = b.bx;
            let g = this.getEffectiveViewport1(b);
            let h = this.xAxis.k5(e, f, g);
            if (isNaN_(h) || isInfinity(h)) {
                c.c8.count = 0;
                return;
            }
            this.yt.a2(this, this.zd(), runOn(this, this.ye), this.zm(b), this.zn(b));
            this.yt.r = this.abk;
            this.yt.s = this.abl;
            this.yt.k = this.abk;
            this.yt.l = this.abl;
            let i = false;
            let j = this.yt.b;
            if (j != null) {
                i = true;
            }
            let k = this.xAxis.cn;
            let l = this.lowColumn.count;
            let m = this.xAxis;
            let n = new ScalerParams(0, e, f, this.xAxis.ch, g);
            let o = 0;
            for (let p = 0; p < d.count; ++p) {
                let q = d._inner[p][0] - 0.5 * h;
                let r = Math.min(d._inner[p][1], d._inner[p][2]);
                let s = Math.max(d._inner[p][1], d._inner[p][2]);
                let t = Math.abs(s - r);
                if (isInfinity(t) || isNaN_(t)) {
                    continue;
                }
                let u = c.c8.item(o);
                o++;
                u.width = h;
                u.height = t;
                let v = this.yt.aw(l, p)[0];
                if (k && a.m != null && p >= 0 && p < a.m.count) {
                    v = truncate(Math.round(a.m._inner[p]));
                }
                if (i) {
                    this.z0(d, p, l, m, n, b.q, v);
                }
                this.yt.ad(u, false, false, false, false);
                u.ap = this.yt.k;
                u.aq = this.yt.l;
                c.dd(u, q, r);
            }
            c.c8.count = o;
            b.c3(a);
        }
        get useHighMarkerFidelity() {
            return this.zh;
        }
    }
    RangeColumnSeries.$t = markType(RangeColumnSeries, 'RangeColumnSeries', HorizontalRangeCategorySeries.$);
    RangeColumnSeries.$$p = markDep(DependencyProperty, PropertyMetadata, RangeColumnSeries, 'raisePropertyChanged', ['RadiusX:abk:abo', [1, 0], 'RadiusY:abl:abp', [1, 0]]);
    return RangeColumnSeries;
})();
/**
 * @hidden
 */
export let RangeColumnSeriesView = /*@__PURE__*/ (() => {
    class RangeColumnSeriesView extends RangeCategorySeriesView {
        constructor(a) {
            super(a);
            this.c7 = null;
            this.c8 = null;
            this.c9 = null;
            this.de = new Rectangle();
            this.c7 = a;
            this.c8 = ((() => {
                let $ret = new Pool$1(Rectangle.$);
                $ret.create = runOn(this, this.df);
                $ret.activate = runOn(this, this.da);
                $ret.disactivate = runOn(this, this.dc);
                $ret.destroy = runOn(this, this.db);
                return $ret;
            })());
        }
        au() {
            super.au();
            this.c9 = new List$1(Rectangle.$, 0);
            if (!this.q) {
                this.c7.w2 = Defaults.c;
                this.c7.jn = Defaults.i;
            }
        }
        df() {
            let a = new Rectangle();
            this.c9.add(a);
            a._visibility = 1;
            return a;
        }
        da(a) {
            a._visibility = 0;
        }
        dc(a) {
            a._visibility = 1;
        }
        db(a) {
            this.c9.remove(a);
        }
        dd(a, b, c) {
            if (!this.isDirty) {
                this.as();
            }
            a.o = c;
            a.n = b;
        }
        ac(a) {
            return this.c9._inner[a];
        }
        ab(a) {
            let b = this.c9._inner[a];
            this.de._visibility = b._visibility;
            this.de.n = b.n;
            this.de.o = b.o;
            this.de.width = b.width;
            this.de.height = b.height;
            let c = this.bs(a);
            this.de._fill = c;
            this.de._stroke = c;
            this.de.ad = this.e.ic + 3;
            return this.de;
        }
        a8(a, b) {
            super.a8(a, b);
            if (a.d) {
                for (let c = 0; c < this.c9.count; c++) {
                    let d = this.aa(c, b);
                    this.bf(d, c, b);
                    a.w(d);
                }
            }
        }
        an(a) {
            super.an(a);
            let b = 0;
            let c = new List$1(Rectangle.$, 0);
            for (let d of fromEnum(this.c8.a)) {
                c.add(d);
            }
            c.aa((e, f) => {
                if (e.n < f.n) {
                    return -1;
                }
                else if (e.n > f.n) {
                    return 1;
                }
                else {
                    return 0;
                }
            });
            for (let e of fromEnum(c)) {
                this.e.bb.exportRectangleData(a, e, "column" + b, ["Main", "Fill"]);
            }
            b++;
        }
    }
    RangeColumnSeriesView.$t = markType(RangeColumnSeriesView, 'RangeColumnSeriesView', RangeCategorySeriesView.$);
    return RangeColumnSeriesView;
})();
