// igniteui-angular-charts/esm2015/lib/RelativeStrengthIndexIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { RelativeStrengthIndexIndicatorStrategy } from "./RelativeStrengthIndexIndicatorStrategy";
import { typeGetValue, markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let RelativeStrengthIndexIndicator = /*@__PURE__*/ (() => {
    class RelativeStrengthIndexIndicator extends StrategyBasedIndicator {
        constructor() {
            super();
        }
        bi() {
            return new RelativeStrengthIndexIndicator();
        }
        get_aau() {
            return 1;
        }
        get_abn() {
            return new RelativeStrengthIndexIndicatorStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return RelativeStrengthIndexIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
        get ab8() {
            return typeGetValue(this.c(RelativeStrengthIndexIndicator.ab9));
        }
        set ab8(a) {
            this.h(RelativeStrengthIndexIndicator.ab9, a);
        }
        abt() {
            return this.ab8;
        }
        ab2(a) {
            this.ab8 = a;
        }
    }
    RelativeStrengthIndexIndicator.$t = markType(RelativeStrengthIndexIndicator, 'RelativeStrengthIndexIndicator', StrategyBasedIndicator.$);
    RelativeStrengthIndexIndicator.ab9 = StrategyBasedIndicator.ab5(14, RelativeStrengthIndexIndicator.$);
    return RelativeStrengthIndexIndicator;
})();
