// igniteui-angular-charts/esm2015/lib/ScatterSplineSeries.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterBase } from "./ScatterBase";
import { DependencyProperty } from "igniteui-angular-core";
import { Point_$type, typeCast, runOn, markType, markDep } from "igniteui-angular-core";
import { ScatterFrame } from "./ScatterFrame";
import { ScatterSplineSeriesView } from "./ScatterSplineSeriesView";
import { Defaults } from "./Defaults";
import { Clipper } from "igniteui-angular-core";
import { OwnedPoint } from "./OwnedPoint";
import { ScalerParams } from "./ScalerParams";
import { PathGeometry } from "igniteui-angular-core";
import { Numeric } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { isNaN_ } from "igniteui-angular-core";
/**
 * @hidden
 */
export let ScatterSplineSeries = /*@__PURE__*/ (() => {
    class ScatterSplineSeries extends ScatterBase {
        constructor() {
            super();
            this.ab = ScatterSplineSeries.$;
            this.yu = new ScatterFrame();
            this.yw = new ScatterFrame();
            this.ys = new ScatterFrame();
        }
        bi() {
            return new ScatterSplineSeries();
        }
        get_fy() {
            return true;
        }
        cq() {
            return new ScatterSplineSeriesView(this);
        }
        zo() {
            return true;
        }
        ji() {
            if (!isNaN_(this.jh)) {
                return super.ji();
            }
            return Defaults.x;
        }
        q5(a, b, c, d) {
            super.q5(a, b, c, d);
            switch (b) {
                case ScatterSplineSeries.$$p[0]:
                    this.rn(false);
                    this.qh();
                    break;
            }
        }
        b0() {
            if (this.b1 == 0) {
                return 2;
            }
            else {
                return super.b0();
            }
        }
        aa6(a, b) {
            super.aa6(a, b);
            a.m.clear();
            a.u.clear();
            let c = b.by;
            let d = b.bx;
            let e = this.getEffectiveViewport1(b);
            let f = this.ib * this.ib;
            let g = Math.min(this.y9 != null ? this.y9.count : 0, this.za != null ? this.za.count : 0);
            let h = new ScalerParams(0, c, d, this.yo.d, e);
            let i = ((() => {
                let $ret = new ScalerParams(0, c, d, this.yo.e, e);
                $ret.b = this.iw();
                return $ret;
            })());
            let j = (k) => this.yo.a.ek(this.y9.item(k), h);
            let k = (l) => this.yo.b.ek(this.za.item(l), i);
            let l = ((() => {
                let $ret = new Clipper(0, d, false);
                $ret.i = a.m;
                return $ret;
            })());
            for (let m = 0; m < g;) {
                let n = m;
                ++m;
                while (m < g && this.ace(j, k, n, m) < f) {
                    ++m;
                }
                if (g > this.z1) {
                    l.j(this.aci(j, k, n, m - 1));
                }
                else {
                    let o = new OwnedPoint();
                    o.c = { $type: Point_$type, x: j(n), y: k(n) };
                    o.a = this.dl.item(n);
                    if (!a.u.containsKey(o.a)) {
                        a.u.addItem(o.a, o);
                    }
                }
            }
            if (g > this.z1) {
                l.i = null;
            }
        }
        aci(a, b, c, d) {
            if (c == d) {
                return { $type: Point_$type, x: a(c), y: b(c) };
            }
            let e = 0;
            let f = 0;
            let g = (d - c + 1);
            for (let h = c; h <= d; ++h) {
                e += a(h);
                f += b(h);
            }
            return { $type: Point_$type, x: e / g, y: f / g };
        }
        ace(a, b, c, d) {
            let e = a(d) - a(c);
            let f = b(d) - b(c);
            return e * e + f * f;
        }
        get_zi() {
            return true;
        }
        o3(a, b) {
            super.o3(a, b);
            let c = typeCast(ScatterSplineSeriesView.$, b);
            c.c9(false);
        }
        abh(a, b) {
            this.o3(false, b);
            super.abh(a, b);
            this.aa9(a);
            let c = typeCast(ScatterSplineSeriesView.$, b);
            let d = c.dl;
            this.y1.az(this, this.zn(), runOn(this, this.yh));
            let e = false;
            let f = this.y1.b;
            if (f != null) {
                e = true;
            }
            if (e) {
                this.aa2(-1, this.y9.count, b.q);
            }
            this.y1.ad(d, true, false, true, false);
            c.dl.an = new PathGeometry();
            c.dl.an.c = Numeric.h(a.m.count, (g) => a.m._inner[g].x, (g) => a.m._inner[g].y, this.acf);
        }
    }
    ScatterSplineSeries.$t = markType(ScatterSplineSeries, 'ScatterSplineSeries', ScatterBase.$);
    ScatterSplineSeries.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterSplineSeries, 'raisePropertyChanged', ['Stiffness:acf:ach', [1, 0.5]]);
    return ScatterSplineSeries;
})();
