// igniteui-angular-charts/esm2015/lib/ScatterPolygonSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ShapeSeriesBase } from "./ShapeSeriesBase";
import { ShapeAxisInfoCache } from "./ShapeAxisInfoCache";
import { StyleSelector } from "./StyleSelector";
import { DependencyProperty } from "igniteui-angular-core";
import { Style } from "igniteui-angular-core";
import { MarkerType_$type } from "./MarkerType";
import { DataTemplate } from "igniteui-angular-core";
import { Brush } from "igniteui-angular-core";
import { CollisionAvoidanceType_$type } from "./CollisionAvoidanceType";
import { MarkerOutlineMode_$type } from "./MarkerOutlineMode";
import { MarkerFillMode_$type } from "./MarkerFillMode";
import { Rect } from "igniteui-angular-core";
import { ScalerParams } from "./ScalerParams";
import { Base, runOn, enumGetBox, Number_$type, typeCast, fromEnum, markType, markDep, PointUtil } from "igniteui-angular-core";
import { Dictionary$2 } from "igniteui-angular-core";
import { OwnedPoint } from "./OwnedPoint";
import { ShapeSeriesViewBase } from "./ShapeSeriesViewBase";
import { Marker } from "./Marker";
import { MarkerSeries } from "./MarkerSeries";
import { BrushUtil } from "igniteui-angular-core";
import { DataContext } from "igniteui-angular-core";
import { HashPool$2 } from "igniteui-angular-core";
import { List$1 } from "igniteui-angular-core";
import { SeriesPointOfInterest } from "./SeriesPointOfInterest";
import { RectUtil } from "igniteui-angular-core";
import { PropertyMetadata } from "igniteui-angular-core";
import { DeviceUtils } from "igniteui-angular-core";
import { isNaN_, isInfinity } from "igniteui-angular-core";
import { NumericMarkerManager } from "./NumericMarkerManager";
import { DataTemplateMeasureInfo } from "igniteui-angular-core";
import { Size } from "igniteui-angular-core";
import { Defaults } from "./Defaults";
import { DataTemplatePassInfo } from "igniteui-angular-core";
import { DataTemplateRenderInfo } from "igniteui-angular-core";
import { MarkerVisualData } from "./MarkerVisualData";
import { stringIsNullOrEmpty } from "igniteui-angular-core";
/**
 * @hidden
 */
export let ScatterPolygonSeries = /*@__PURE__*/ (() => {
    class ScatterPolygonSeries extends ShapeSeriesBase {
        constructor() {
            super();
            this.zv = null;
            this.zs = new ScalerParams(1, ScatterPolygonSeries.aa1, ScatterPolygonSeries.aa1, false);
            this.zt = new ScalerParams(1, ScatterPolygonSeries.aa1, ScatterPolygonSeries.aa1, false);
            this.aah = null;
            this.zn = 1;
            this.zj = 1;
            this.ab = ScatterPolygonSeries.$;
        }
        bi() {
            return new ScatterPolygonSeries();
        }
        get_fo() {
            return true;
        }
        get_fj() {
            return true;
        }
        xy() {
            return this.zx();
        }
        y0(a) {
            let b = a;
            b.cy(this.aa3);
            b.cz(this.zw);
            super.y0(a);
            if (this.d3(a)) {
                return;
            }
            let c, d;
            let e = a.getViewInfo(c, d);
            c = e.p0;
            d = e.p1;
            let f = this.getEffectiveViewport1(a);
            this.zs = new ScalerParams(1, d, c, this.w2.ch);
            this.zs.c = f;
            this.zt = new ScalerParams(1, d, c, this.w4.ch);
            this.zt.c = f;
            if (c.width < 1 || c.height < 1) {
                return;
            }
            this.zv = ((() => {
                let $ret = new ShapeAxisInfoCache();
                $ret.a = this.w2;
                $ret.b = this.w4;
                $ret.e = this.w2.ch;
                $ret.f = this.w4.ch;
                $ret.d = this.dl;
                $ret.c = this.xh;
                return $ret;
            })());
            if (this.zx()) {
                let g = new Dictionary$2(Base.$, OwnedPoint.$, 0);
                b.dg.l(g, 400, d, c, this.ib);
                b.dg.j(g, false);
                if (this.xw()) {
                    this.aae(a);
                }
            }
        }
        aae(a) {
            a.dk = true;
            this.w9.ay(this, this.xw(), runOn(this, this.ww));
            let b = this.w9.d;
            let c = b != null;
            let d = this.getEffectiveViewport1(a);
            let e = this.dl == null ? 0 : this.dl.count;
            a.dw((f) => {
                if (f._visibility == 0) {
                    let g = f.ak;
                    let h = f.content;
                    if (h == null) {
                        h = new DataContext();
                        f.content = h;
                    }
                    if (c) {
                        this.yv(g, e, a.q);
                    }
                    this.w9.ac(f, h);
                }
            });
        }
        zx() {
            return this.aah != null && ((this.zr != 1 && this.zr != 0) || this.aai != null);
        }
        cq() {
            return new ScatterPolygonSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.zu = a;
        }
        get_hasMarkers() {
            return true;
        }
        get_ej() {
            return this.zr != 13 && this.zr != 1;
        }
        vf() {
            return this.aax;
        }
        vg() {
            return this.aay;
        }
        st() {
            return this.aag;
        }
        l8() {
            return this.l9(this, this.aah);
        }
        il() {
            return this.zz;
        }
        get zo() {
            return this.zn;
        }
        set zo(a) {
            this.zn = a;
        }
        zq() {
            if (this.zp == 0) {
                return 1;
            }
            return this.zp;
        }
        sf(a) {
            super.sf(a);
            this.zp = a;
        }
        get zk() {
            return this.zj;
        }
        set zk(a) {
            this.zj = a;
        }
        zm() {
            if (this.zl == 0) {
                return 2;
            }
            return this.zl;
        }
        se(a) {
            super.se(a);
            this.zl = a;
        }
        sg(a) {
            super.sg(a);
            this.zz = a;
        }
        sb() {
            super.sb();
            if (this.index < 0) {
                return;
            }
            if (this.aai != null) {
                this.zu.dv();
                this.zu.ds();
            }
            else {
                let a = MarkerSeries.w3(this, this.zr);
                let b = MarkerSeries.xg(a);
                if (b == null) {
                    this.aag = MarkerSeries.xv;
                }
                else {
                    this.zu.dp(b);
                }
            }
            if (this.aaz != null && this.zr != 13) {
                this.r2("ActualMarkerBrush");
                let c = this.aax;
                this.zu.dt();
                this.zu.dq();
                let d = this.aax;
                let e = this.ff("ActualMarkerBrush", this.ht.f > 0 ? this.ht.d() : null, c, d);
                this.r4("ActualMarkerBrush", e);
            }
            else {
                if (this.zr == 13) {
                    this.aax = this.ct.bt();
                }
                else {
                    if (this.zk != 2) {
                        if (this.c9 != null) {
                            let f = this.c9.uv(this.index);
                            if (!isNaN_(this.ia) && this.ia != 1) {
                                f = BrushUtil.q(f, this.ia);
                            }
                            this.aax = f;
                        }
                        else {
                            this.aax = null;
                        }
                    }
                }
            }
            if (this.aa0 != null) {
                this.zu.du();
                this.zu.dr();
            }
            else {
                if (this.zr == 13) {
                    this.aay = this.ct.bt();
                }
                else {
                    if (this.zp == 2) {
                        this.aay = this.aax;
                    }
                    else {
                        this.aay = this.c9 == null ? null : this.c9.uw(this.index);
                    }
                }
            }
            if (this.zk == 2 && this.aaz == null && this.zr != 13) {
                if (this.aay != null) {
                    let g = this.aay;
                    if (!isNaN_(this.ia) && this.ia != 1) {
                        g = BrushUtil.q(g, this.ia);
                    }
                    this.aax = g;
                }
                else {
                    this.aax = null;
                }
            }
        }
        qo() {
            super.qo();
            this.zo = this.zq();
            this.zk = this.zm();
        }
        q5(a, b, c, d) {
            super.q5(a, b, c, d);
            switch (b) {
                case ScatterPolygonSeries.$$p[3]:
                case ScatterPolygonSeries.$$p[11]:
                case ScatterPolygonSeries.$$p[7]:
                case ScatterPolygonSeries.$$p[9]:
                case ScatterPolygonSeries.$$p[10]:
                    this.rb(b);
                    this.qz();
                    break;
                case ScatterPolygonSeries.$$p[6]:
                    this.zo = this.zq();
                    this.rb(null);
                    this.qz();
                    break;
                case ScatterPolygonSeries.$$p[5]:
                    this.zk = this.zm();
                    this.rb(null);
                    this.qz();
                    break;
                case ScatterPolygonSeries.$$p[2]:
                    this.aah = d;
                    if (c == MarkerSeries.xv || d == MarkerSeries.xv || (c == null || d != null)) {
                        this.zu.dx();
                        this.rn(false);
                    }
                    this.ct.bl();
                    if (this.toolTip != null && this.fg) {
                        this.a9.a$as(this, this.toolTip);
                    }
                    this.rn(false);
                    this.qh();
                    break;
                case ScatterPolygonSeries.$$p[12]:
                    this.zu.cy(this.aa3);
                    this.rn(false);
                    this.qh();
                    break;
                case ScatterPolygonSeries.$$p[13]:
                    this.zu.cz(this.zw);
                    this.rn(false);
                    this.qh();
                    break;
                case ScatterPolygonSeries.$$p[4]:
                    this.rn(false);
                    this.qh();
                    break;
                case ScatterPolygonSeries.$$p[0]:
                case ScatterPolygonSeries.$$p[1]:
                    if (this.zx()) {
                        this.rn(false);
                    }
                    break;
            }
        }
        a1(a, b) {
            let c = this.zu.b4(a);
            let d = null;
            if (c != null) {
                d = c.content;
            }
            if (d != null) {
                return d;
            }
            return super.a1(a, b);
        }
        aaf(a, b) {
            let c = new List$1(Base.$, 0);
            for (let d of fromEnum(b.c)) {
                if (!a.containsKey(d)) {
                    c.add1(d);
                }
            }
            for (let e of fromEnum(c)) {
                b.h(e);
            }
        }
        d9(a, b, c, d) {
            if (!this.hasMarkers) {
                return false;
            }
            let e = this.w2;
            let f = new ScalerParams(1, this.c9.vl, this.wl, e.ch);
            f.c = this.getEffectiveViewport1(this.ct);
            let g = this.w4;
            let h = new ScalerParams(1, this.c9.vl, this.wl, g.ch);
            h.c = this.getEffectiveViewport1(this.ct);
            a.clear();
            this.zu.dw((i) => {
                if (i._visibility == 0) {
                    let j = new SeriesPointOfInterest();
                    let k = i.n;
                    let l = i.o;
                    if (k < this.wl.left || k > this.wl.right || l < this.wl.top || l > this.wl.bottom) {
                        return;
                    }
                    j.d = e.en(k, f);
                    j.e = g.en(l, h);
                    j.b = i.content;
                    j.g = this.va;
                    j.c = i.ak;
                    if (this.g7) {
                        j.g = this.vh(i.ak);
                    }
                    a.add(j);
                }
            });
            return true;
        }
        zh(a, b, c, d, e) {
            if (c == null || c.length != this.zv.d.count) {
                c = new Array(this.zv.d.count);
                for (let f = 0; f < this.zv.d.count; f++) {
                    c[f] = PointUtil.create();
                }
            }
            let g = ((() => {
                let $ret = new ScalerParams(1, d, e, this.w2.ch);
                $ret.c = this.getEffectiveViewport1(a);
                return $ret;
            })());
            let h = ((() => {
                let $ret = new ScalerParams(1, d, e, this.w4.ch);
                $ret.c = this.getEffectiveViewport1(a);
                return $ret;
            })());
            let i = this.wd(a);
            let j = this.zv.a.en(i.left, g);
            let k = this.zv.a.en(i.right, g);
            let l = this.zv.b.en(i.bottom, h);
            let m = this.zv.b.en(i.top, h);
            if (this.zv.e) {
                let n = j;
                j = k;
                k = n;
            }
            if (this.zv.f) {
                let o = l;
                l = m;
                m = o;
            }
            let p = this.zv;
            let q = p.a;
            let r = p.b;
            let s;
            let t;
            let u = g.d.left;
            let v = g.d.right;
            let w = h.d.top;
            let x = h.d.bottom;
            for (let y = 0; y < this.zv.d.count; y++) {
                s = 0;
                t = 0;
                let z = null;
                let aa = a.cc.tryGetValue(y, z);
                z = aa.p1;
                if (z == null) {
                    c[y].x = NaN;
                    c[y].y = NaN;
                }
                else {
                    let ab = z.c;
                    let ac = RectUtil.l(ab);
                    s = ac.x;
                    t = ac.y;
                    if (s >= u && s <= v && t >= w && t <= x) {
                        c[y].x = s;
                        c[y].y = t;
                    }
                    else {
                        c[y].x = NaN;
                        c[y].y = NaN;
                    }
                }
            }
            return c;
        }
        zy(a, b) {
            if (b == null) {
                b = new List$1(Number_$type, 2, a.e);
            }
            else {
                b.clear();
            }
            let c = 0;
            let d = this.dl;
            for (let e of fromEnum(a.c)) {
                b.add(d.indexOf(e));
                c++;
            }
            return b;
        }
    }
    ScatterPolygonSeries.$t = markType(ScatterPolygonSeries, 'ScatterPolygonSeries', ShapeSeriesBase.$);
    ScatterPolygonSeries.aa1 = new Rect(0, 0, 0, 1, 1);
    ScatterPolygonSeries.$$p = markDep(DependencyProperty, PropertyMetadata, ScatterPolygonSeries, 'raisePropertyChanged', ['ActualMarkerBrush:aax:aaj', [Brush.$], 'ActualMarkerOutline:aay:aak', [Brush.$], 'ActualMarkerTemplate:aag:aal', [DataTemplate.$, null], 'MarkerBrush:aaz:aam', [Brush.$], 'MarkerCollisionAvoidance:zi:aan', [CollisionAvoidanceType_$type, enumGetBox(CollisionAvoidanceType_$type, 0)], 'MarkerFillMode:zl:aao', [MarkerFillMode_$type, enumGetBox(MarkerFillMode_$type, 0)], 'MarkerOutlineMode:zp:aap', [MarkerOutlineMode_$type, enumGetBox(MarkerOutlineMode_$type, 0)], 'MarkerOutline:aa0:aaq', [Brush.$], 'MarkerStyle:aa2:aar', [Style.$], 'MarkerTemplate:aai:aas', [DataTemplate.$, null], 'MarkerThickness:zz:aat', [1, DeviceUtils.g(2)], 'MarkerType:zr:aau', [MarkerType_$type, enumGetBox(MarkerType_$type, 1)], 'ShapeStyle:aa3:aav', [Style.$, null], 'ShapeStyleSelector:zw:aaw', [StyleSelector.$, null]]);
    return ScatterPolygonSeries;
})();
/**
 * @hidden
 */
export let ScatterPolygonSeriesView = /*@__PURE__*/ (() => {
    class ScatterPolygonSeriesView extends ShapeSeriesViewBase {
        constructor(a) {
            super(a);
            this.dk = false;
            this.db = null;
            this.dm = null;
            this.dh = null;
            this.df = null;
            this.dj = false;
            this.di = null;
            this.dn = null;
            this.dc = new Marker();
            this.d4 = null;
            this.dc = new Marker();
            this.dc.content = new DataContext();
            this.dh = a;
            this.di = new HashPool$2(Base.$, Marker.$);
            this.dn = new List$1(Marker.$, 0);
            this.dy(this.di);
            this.df = new NumericMarkerManager(1, (b) => this.di.item(b), (b) => this.dh.zv.d.item(b), runOn(this, this.d2), runOn(this, this.da), runOn(this, this.dl), () => this.dh.zi);
            this.df.q = runOn(this, this.d5);
        }
        au() {
            super.au();
            if (!this.q) {
                this.dh.aa0 = Defaults.am;
                this.dh.jn = Defaults.z;
                this.dh.zz = Defaults.y;
                this.dh.zr = Defaults.f;
                this.dh.zl = Defaults.b;
            }
        }
        get dg() {
            return this.df;
        }
        d2(a) {
            this.dh.aaf(a, this.di);
        }
        da() {
            this.db = this.dh.zh(this, this.di, this.db, this.by, this.bx);
            return this.db;
        }
        dl() {
            this.dm = this.dh.zy(this.di, this.dm);
            return this.dm;
        }
        c5(a, b) {
            return this.c6(a, b, true);
        }
        cm(a, b, c) {
            super.cm(a, b, c);
            ShapeSeriesViewBase.cl(this, a, b);
        }
        ck(a, b) {
            let c = a;
            if (c == null) {
                return;
            }
            c.an = b;
        }
        dw(a) {
            this.di.f(a);
        }
        dy(a) {
            a.create = runOn(this, this.de);
            a.destroy = runOn(this, this.d0);
            a.activate = runOn(this, this.dz);
            a.disactivate = runOn(this, this.d1);
        }
        de() {
            let a = new Marker();
            if (!this.dj) {
                a.content = ((() => {
                    let $ret = new DataContext();
                    $ret.series = this.e;
                    return $ret;
                })());
            }
            a.ah = this.e.aag;
            this.dn.add(a);
            return a;
        }
        b4(a) {
            let b = this.dd(a, 0);
            if (b == null) {
                b = this.dd(a, 3);
            }
            return b;
        }
        dd(a, b) {
            let c;
            let d;
            let e;
            let f;
            for (let g = this.dn.count - 1; g >= 0; g--) {
                let h = this.dn._inner[g];
                if (h._visibility == 1 || h._opacity == 0) {
                    continue;
                }
                c = (h.m / 2) + b;
                d = (h.l / 2) + b;
                e = h.ai;
                f = h.aj;
                if ((h.n + e) - c <= a.x && (h.n + e) + c >= a.x && (h.o + f) - d <= a.y && (h.o + f) + d >= a.y) {
                    return h;
                }
            }
            return null;
        }
        d0(a) {
            a.content = null;
            this.dn.remove(a);
        }
        dz(a) {
            a._visibility = 0;
        }
        d1(a) {
            if (a.content != null) {
                (typeCast(DataContext.$, a.content)).item = null;
            }
            a._visibility = 1;
        }
        d3(a) {
            this.dj = a;
        }
        bj(a, b) {
            super.bj(a, b);
            if (!this.dk) {
                let c = a;
                let d = c.content;
                if (d != null) {
                    if (this.dh.aax != null) {
                        d.actualItemBrush = this.dh.aax;
                    }
                    else {
                        d.actualItemBrush = this.b6.u3;
                    }
                    if (this.dh.aax != null) {
                        d.outline = this.dh.aay;
                    }
                    else {
                        d.outline = this.b6.u7;
                    }
                    d.thickness = this.dh.zz;
                }
            }
        }
        bk(a, b) {
            super.bk(a, b);
            let c = a;
            this.dc._visibility = c._visibility;
            this.dc.ah = c.ah;
            this.dc.width = c.width;
            this.dc.height = c.height;
            this.dc.m = c.m;
            this.dc.l = c.l;
            this.dc.n = c.n;
            this.dc.o = c.o;
            let d = this.bs(b);
            let e = this.dc.content;
            let f = c.content;
            e.item = f.item;
            e.series = f.series;
            e.thickness = f.thickness;
            if (e != null) {
                e.actualItemBrush = d;
                e.outline = d;
                e.thickness = this.dh.zz + 3;
            }
        }
        a7(a, b) {
            super.a7(a, b);
            let c = new DataTemplatePassInfo();
            c.isHitTestRender = b;
            c.context = a.h();
            c.viewportTop = this.bx.top;
            c.viewportLeft = this.bx.left;
            c.viewportWidth = this.bx.width;
            c.viewportHeight = this.bx.height;
            c.passID = "Markers";
            let d = new DataTemplateRenderInfo();
            d.passInfo = c;
            d.isHitTestRender = b;
            let e = new DataTemplateMeasureInfo();
            e.passInfo = c;
            let f = false;
            let g = a.h();
            e.context = g;
            d.context = g;
            let h = 0;
            let i = 0;
            if (this.dh.aag != null && this.dh.aag.passStarting != null) {
                this.dh.aag.passStarting(c);
            }
            for (let j = 0; j < this.dn.count; j++) {
                let k = this.dn._inner[j];
                if (k._visibility == 1) {
                    continue;
                }
                this.bi(k, j, b);
                if (b) {
                    k = this.dc;
                }
                if (!f) {
                    e.width = k.width;
                    e.height = k.height;
                    e.renderOffsetX = 0;
                    e.renderOffsetY = 0;
                    e.renderContext = a;
                    let l = k.ah;
                    if (l.measure != null) {
                        e.data = k.content;
                        l.measure(e);
                        f = e.isConstant;
                        if (f) {
                            h = e.width;
                            i = e.height;
                        }
                    }
                    d.availableWidth = e.width;
                    d.availableHeight = e.height;
                    d.renderOffsetX = e.renderOffsetX;
                    d.renderOffsetY = e.renderOffsetY;
                    d.renderContext = a;
                }
                else {
                    d.availableWidth = h;
                    d.availableHeight = i;
                }
                if (!isNaN_(k.width) && !isInfinity(k.width)) {
                    d.availableWidth = k.width;
                }
                if (!isNaN_(k.height) && !isInfinity(k.height)) {
                    d.availableHeight = k.height;
                }
                a.q(d, k);
                k.m = d.availableWidth;
                k.l = d.availableHeight;
                k.ai = d.renderOffsetX;
                k.aj = d.renderOffsetY;
            }
            if (this.dh.aag != null && this.dh.aag.passCompleted != null) {
                this.dh.aag.passCompleted(c);
            }
        }
        dv() {
            this.dh.aag = null;
        }
        ds() {
            this.dh.aag = null;
            this.dh.aag = this.dh.aai;
        }
        dt() {
            this.dh.aax = null;
        }
        dq() {
            this.dh.aax = null;
            if (!isNaN_(this.e.ia) && this.e.ia != 1) {
                let a = BrushUtil.q(this.dh.aaz, this.e.ia);
                this.dh.aax = a;
            }
            else {
                this.dh.aax = this.dh.aaz;
            }
        }
        du() {
            this.dh.aay = null;
        }
        dr() {
            this.dh.aay = null;
            this.dh.aay = this.dh.aa0;
        }
        dp(a) {
            switch (a) {
                case "CircleMarkerTemplate":
                    this.dh.aag = this.dh.c9.rn;
                    break;
                case "TriangleMarkerTemplate":
                    this.dh.aag = this.dh.c9.rw;
                    break;
                case "PyramidMarkerTemplate":
                    this.dh.aag = this.dh.c9.rt;
                    break;
                case "SquareMarkerTemplate":
                    this.dh.aag = this.dh.c9.ru;
                    break;
                case "DiamondMarkerTemplate":
                    this.dh.aag = this.dh.c9.ro;
                    break;
                case "PentagonMarkerTemplate":
                    this.dh.aag = this.dh.c9.rr;
                    break;
                case "HexagonMarkerTemplate":
                    this.dh.aag = this.dh.c9.rp;
                    break;
                case "TetragramMarkerTemplate":
                    this.dh.aag = this.dh.c9.rr;
                    break;
                case "PentagramMarkerTemplate":
                    this.dh.aag = this.dh.c9.rs;
                    break;
                case "HexagramMarkerTemplate":
                    this.dh.aag = this.dh.c9.rq;
                    break;
            }
        }
        dx() {
            for (let a of fromEnum(this.dn)) {
                a.ah = this.dh.aag;
            }
            this.as();
        }
        d5(a) {
            if (this.d4 == null) {
                this.d4 = new DataTemplateMeasureInfo();
                this.d4.context = this.k.h();
            }
            this.d4.width = a.width;
            this.d4.height = a.height;
            this.d4.data = a.content;
            let b = a.ah;
            if (b.measure != null) {
                b.measure(this.d4);
            }
            return new Size(1, this.d4.width, this.d4.height);
        }
        an(a) {
            let b = a;
            super.an(b);
            for (let c of fromEnum(this.di.c)) {
                let d = this.di.item(c);
                if (d._visibility != 1) {
                    let e = typeCast(ScatterPolygonSeries.$, this.e);
                    let f = this.e.bb.getMarkerVisualData(e, d, e.aag);
                    b.markerShapes.add(typeCast(MarkerVisualData.$, f));
                }
            }
        }
        getDefaultTooltipTemplate() {
            let a = "<div class='ui-chart-default-tooltip-content'><span";
            let b = this.e.a9.a$u(this.e);
            if (!stringIsNullOrEmpty(b)) {
                a += " style='color:" + b + "'";
            }
            a += ">" + this.dh.title + "</span><br/>";
            a += "</div>";
            return a;
        }
    }
    ScatterPolygonSeriesView.$t = markType(ScatterPolygonSeriesView, 'ScatterPolygonSeriesView', ShapeSeriesViewBase.$);
    return ScatterPolygonSeriesView;
})();
