// igniteui-angular-charts/esm2015/lib/ScatterSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterBase } from "./ScatterBase";
import { markType } from "igniteui-angular-core";
import { ScatterFrame } from "./ScatterFrame";
import { ScatterBaseView } from "./ScatterBaseView";
import { Defaults } from "./Defaults";
import { Color } from "igniteui-angular-core";
/**
 * @hidden
 */
export let ScatterSeries = /*@__PURE__*/ (() => {
    class ScatterSeries extends ScatterBase {
        constructor() {
            super();
            this.ab = ScatterSeries.$;
            this.yu = new ScatterFrame();
            this.yw = new ScatterFrame();
            this.ys = new ScatterFrame();
        }
        bi() {
            return new ScatterSeries();
        }
        get_eg() {
            return true;
        }
        cq() {
            return new ScatterSeriesView(this);
        }
        jj() {
            return 0.7;
        }
        wt() {
            if (this.ws == 0) {
                return 2;
            }
            return super.wt();
        }
        qx(a) {
            super.qx(a);
            this.ace = a;
        }
    }
    ScatterSeries.$t = markType(ScatterSeries, 'ScatterSeries', ScatterBase.$);
    return ScatterSeries;
})();
/**
 * @hidden
 */
export let ScatterSeriesView = /*@__PURE__*/ (() => {
    class ScatterSeriesView extends ScatterBaseView {
        constructor(a) {
            super(a);
            this.dk = null;
            this.dk = a;
        }
        au() {
            super.au();
            if (!this.q) {
                this.b4.w2 = Defaults.e;
            }
        }
        af() {
            let a = new Color();
            a.colorString = "rgba(95,95,95,0.5)";
            this.e.vv = a;
            this.e.jk = 3;
            this.e.jl = 2;
            this.e.jm = 2;
            this.e.ha = false;
        }
    }
    ScatterSeriesView.$t = markType(ScatterSeriesView, 'ScatterSeriesView', ScatterBaseView.$);
    return ScatterSeriesView;
})();
