// igniteui-angular-charts/esm2015/lib/Stacked100SplineAreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedSplineAreaSeries } from "./StackedSplineAreaSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-angular-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedSplineAreaSeriesView } from "./StackedSplineAreaSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let Stacked100SplineAreaSeries = /*@__PURE__*/ (() => {
    class Stacked100SplineAreaSeries extends StackedSplineAreaSeries {
        constructor() {
            super();
            this.ab = Stacked100SplineAreaSeries.$;
        }
        cq() {
            return new Stacked100SplineAreaSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.abo = a;
        }
        get_aan() {
            return true;
        }
        ac() {
            super.ac();
            this.rn(false);
        }
        aa3() {
            super.aa3();
            Stacked100DataPreparer.a(this);
        }
        get useHighMarkerFidelity() {
            return this.zh;
        }
    }
    Stacked100SplineAreaSeries.$t = markType(Stacked100SplineAreaSeries, 'Stacked100SplineAreaSeries', StackedSplineAreaSeries.$, [IStacked100Series_$type]);
    return Stacked100SplineAreaSeries;
})();
/**
 * @hidden
 */
export let Stacked100SplineAreaSeriesView = /*@__PURE__*/ (() => {
    class Stacked100SplineAreaSeriesView extends StackedSplineAreaSeriesView {
        constructor(a) {
            super(a);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.w2 = Defaults.e;
            }
        }
    }
    Stacked100SplineAreaSeriesView.$t = markType(Stacked100SplineAreaSeriesView, 'Stacked100SplineAreaSeriesView', StackedSplineAreaSeriesView.$);
    return Stacked100SplineAreaSeriesView;
})();
