// igniteui-angular-charts/esm2015/lib/Stacked100ColumnSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedColumnSeries } from "./StackedColumnSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { Defaults } from "./Defaults";
import { markType } from "igniteui-angular-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { isNaN_ } from "igniteui-angular-core";
import { StackedColumnSeriesView } from "./StackedColumnSeriesView";
import { Stacked100ColumnBucketCalculator } from "./Stacked100ColumnBucketCalculator";
/**
 * @hidden
 */
export let Stacked100ColumnSeries = /*@__PURE__*/ (() => {
    class Stacked100ColumnSeries extends StackedColumnSeries {
        constructor() {
            super();
            this.ab = Stacked100ColumnSeries.$;
        }
        get_aan() {
            return true;
        }
        ji() {
            if (!isNaN_(this.jh)) {
                return super.ji();
            }
            return Defaults.h;
        }
        ac() {
            super.ac();
            this.rn(false);
        }
        cq() {
            return new Stacked100ColumnSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.abr = a;
        }
        aae() {
            return this.abr;
        }
        aa3() {
            super.aa3();
            Stacked100DataPreparer.a(this);
        }
        get useHighMarkerFidelity() {
            return this.zh;
        }
    }
    Stacked100ColumnSeries.$t = markType(Stacked100ColumnSeries, 'Stacked100ColumnSeries', StackedColumnSeries.$, [IStacked100Series_$type]);
    return Stacked100ColumnSeries;
})();
/**
 * @hidden
 */
export let Stacked100ColumnSeriesView = /*@__PURE__*/ (() => {
    class Stacked100ColumnSeriesView extends StackedColumnSeriesView {
        constructor(a) {
            super(a);
            this.da = null;
            this.da = a;
        }
        cx() {
            return new Stacked100ColumnBucketCalculator(this);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.w2 = Defaults.c;
                this.c5.jn = Defaults.i;
            }
        }
    }
    Stacked100ColumnSeriesView.$t = markType(Stacked100ColumnSeriesView, 'Stacked100ColumnSeriesView', StackedColumnSeriesView.$);
    return Stacked100ColumnSeriesView;
})();
