// igniteui-angular-charts/esm2015/lib/StackedLineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalStackedSeriesBase } from "./HorizontalStackedSeriesBase";
import { markType } from "igniteui-angular-core";
import { AxisRange } from "./AxisRange";
import { StackedSeriesView } from "./StackedSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let StackedLineSeries = /*@__PURE__*/ (() => {
    class StackedLineSeries extends HorizontalStackedSeriesBase {
        constructor() {
            super();
            this.ab = StackedLineSeries.$;
        }
        bi() {
            return new StackedLineSeries();
        }
        get_fi() {
            return true;
        }
        ac() {
            super.ac();
            this.rn(false);
        }
        cq() {
            return new StackedLineSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.abk = a;
        }
        ax(a) {
            if (this.itemsSource == null) {
                return null;
            }
            if (a == this.xAxis) {
                return new AxisRange(0, this.dl.count - 1);
            }
            if (a == this.yAxis) {
                this.aa3();
                return new AxisRange(this.aau, this.aat);
            }
            return null;
        }
        preferredCategoryMode(a) {
            return 0;
        }
        get useHighMarkerFidelity() {
            return this.zh;
        }
    }
    StackedLineSeries.$t = markType(StackedLineSeries, 'StackedLineSeries', HorizontalStackedSeriesBase.$);
    return StackedLineSeries;
})();
/**
 * @hidden
 */
export let StackedLineSeriesView = /*@__PURE__*/ (() => {
    class StackedLineSeriesView extends StackedSeriesView {
        constructor(a) {
            super(a);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.w2 = Defaults.e;
            }
        }
    }
    StackedLineSeriesView.$t = markType(StackedLineSeriesView, 'StackedLineSeriesView', StackedSeriesView.$);
    return StackedLineSeriesView;
})();
