// igniteui-angular-charts/esm2015/lib/Stacked100SplineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedSplineSeries } from "./StackedSplineSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-angular-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedSplineSeriesView } from "./StackedSplineSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let Stacked100SplineSeries = /*@__PURE__*/ (() => {
    class Stacked100SplineSeries extends StackedSplineSeries {
        constructor() {
            super();
            this.ab = Stacked100SplineSeries.$;
        }
        cq() {
            return new Stacked100SplineSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.abo = a;
        }
        get_aan() {
            return true;
        }
        ac() {
            super.ac();
            this.rn(false);
        }
        aa3() {
            super.aa3();
            Stacked100DataPreparer.a(this);
        }
        get useHighMarkerFidelity() {
            return this.zh;
        }
    }
    Stacked100SplineSeries.$t = markType(Stacked100SplineSeries, 'Stacked100SplineSeries', StackedSplineSeries.$, [IStacked100Series_$type]);
    return Stacked100SplineSeries;
})();
/**
 * @hidden
 */
export let Stacked100SplineSeriesView = /*@__PURE__*/ (() => {
    class Stacked100SplineSeriesView extends StackedSplineSeriesView {
        constructor(a) {
            super(a);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.w2 = Defaults.e;
            }
        }
    }
    Stacked100SplineSeriesView.$t = markType(Stacked100SplineSeriesView, 'Stacked100SplineSeriesView', StackedSplineSeriesView.$);
    return Stacked100SplineSeriesView;
})();
