// igniteui-angular-charts/esm2015/lib/Stacked100BarSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedBarSeries } from "./StackedBarSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { Defaults } from "./Defaults";
import { markType } from "igniteui-angular-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { isNaN_ } from "igniteui-angular-core";
import { StackedBarSeriesView } from "./StackedBarSeriesView";
import { Stacked100BarBucketCalculator } from "./Stacked100BarBucketCalculator";
/**
 * @hidden
 */
export let Stacked100BarSeries = /*@__PURE__*/ (() => {
    class Stacked100BarSeries extends StackedBarSeries {
        constructor() {
            super();
            this.ab = Stacked100BarSeries.$;
        }
        get_aan() {
            return true;
        }
        ji() {
            if (!isNaN_(this.jh)) {
                return super.ji();
            }
            return Defaults.h;
        }
        ac() {
            super.ac();
            this.rn(false);
        }
        cq() {
            return new Stacked100BarSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.abr = a;
        }
        aae() {
            return this.abr;
        }
        aa3() {
            super.aa3();
            Stacked100DataPreparer.a(this);
        }
        get useHighMarkerFidelity() {
            return this.zh;
        }
    }
    Stacked100BarSeries.$t = markType(Stacked100BarSeries, 'Stacked100BarSeries', StackedBarSeries.$, [IStacked100Series_$type]);
    return Stacked100BarSeries;
})();
/**
 * @hidden
 */
export let Stacked100BarSeriesView = /*@__PURE__*/ (() => {
    class Stacked100BarSeriesView extends StackedBarSeriesView {
        constructor(a) {
            super(a);
            this.db = null;
            this.db = a;
        }
        cx() {
            return new Stacked100BarBucketCalculator(this);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.w2 = Defaults.c;
                this.c5.jn = Defaults.i;
            }
        }
    }
    Stacked100BarSeriesView.$t = markType(Stacked100BarSeriesView, 'Stacked100BarSeriesView', StackedBarSeriesView.$);
    return Stacked100BarSeriesView;
})();
