// igniteui-angular-charts/esm2015/lib/Stacked100AreaSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedAreaSeries } from "./StackedAreaSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-angular-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedAreaSeriesView } from "./StackedAreaSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let Stacked100AreaSeries = /*@__PURE__*/ (() => {
    class Stacked100AreaSeries extends StackedAreaSeries {
        constructor() {
            super();
            this.ab = Stacked100AreaSeries.$;
        }
        get_aan() {
            return true;
        }
        ac() {
            super.ac();
            this.rn(false);
        }
        cq() {
            return new Stacked100AreaSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.abl = a;
        }
        aa3() {
            super.aa3();
            Stacked100DataPreparer.a(this);
        }
        get useHighMarkerFidelity() {
            return this.zh;
        }
    }
    Stacked100AreaSeries.$t = markType(Stacked100AreaSeries, 'Stacked100AreaSeries', StackedAreaSeries.$, [IStacked100Series_$type]);
    return Stacked100AreaSeries;
})();
/**
 * @hidden
 */
export let Stacked100AreaSeriesView = /*@__PURE__*/ (() => {
    class Stacked100AreaSeriesView extends StackedAreaSeriesView {
        constructor(a) {
            super(a);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.w2 = Defaults.e;
            }
        }
    }
    Stacked100AreaSeriesView.$t = markType(Stacked100AreaSeriesView, 'Stacked100AreaSeriesView', StackedAreaSeriesView.$);
    return Stacked100AreaSeriesView;
})();
