// igniteui-angular-charts/esm2015/lib/Stacked100LineSeries_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StackedLineSeries } from "./StackedLineSeries";
import { IStacked100Series_$type } from "./IStacked100Series";
import { markType } from "igniteui-angular-core";
import { Stacked100DataPreparer } from "./Stacked100DataPreparer";
import { StackedLineSeriesView } from "./StackedLineSeriesView";
import { Defaults } from "./Defaults";
/**
 * @hidden
 */
export let Stacked100LineSeries = /*@__PURE__*/ (() => {
    class Stacked100LineSeries extends StackedLineSeries {
        constructor() {
            super();
            this.ab = Stacked100LineSeries.$;
        }
        cq() {
            return new Stacked100LineSeriesView(this);
        }
        qx(a) {
            super.qx(a);
            this.abl = a;
        }
        get_aan() {
            return true;
        }
        ac() {
            super.ac();
            this.rn(false);
        }
        aa3() {
            super.aa3();
            Stacked100DataPreparer.a(this);
        }
        get useHighMarkerFidelity() {
            return this.zh;
        }
    }
    Stacked100LineSeries.$t = markType(Stacked100LineSeries, 'Stacked100LineSeries', StackedLineSeries.$, [IStacked100Series_$type]);
    return Stacked100LineSeries;
})();
/**
 * @hidden
 */
export let Stacked100LineSeriesView = /*@__PURE__*/ (() => {
    class Stacked100LineSeriesView extends StackedLineSeriesView {
        constructor(a) {
            super(a);
        }
        au() {
            super.au();
            if (!this.q) {
                this.c5.w2 = Defaults.e;
            }
        }
    }
    Stacked100LineSeriesView.$t = markType(Stacked100LineSeriesView, 'Stacked100LineSeriesView', StackedLineSeriesView.$);
    return Stacked100LineSeriesView;
})();
