// igniteui-angular-charts/esm2015/lib/UltimateOscillatorIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicator } from "./StrategyBasedIndicator";
import { UltimateOscillatorIndicatorCalculationStrategy } from "./UltimateOscillatorIndicatorCalculationStrategy";
import { markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let UltimateOscillatorIndicator = /*@__PURE__*/ (() => {
    class UltimateOscillatorIndicator extends StrategyBasedIndicator {
        bi() {
            return new UltimateOscillatorIndicator();
        }
        get_abn() {
            return new UltimateOscillatorIndicatorCalculationStrategy();
        }
        get abn() {
            return this.get_abn();
        }
        get_ab0() {
            return UltimateOscillatorIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
    }
    UltimateOscillatorIndicator.$t = markType(UltimateOscillatorIndicator, 'UltimateOscillatorIndicator', StrategyBasedIndicator.$);
    return UltimateOscillatorIndicator;
})();
