// igniteui-angular-charts/esm2015/lib/WeightedCloseIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ItemwiseStrategyBasedIndicator } from "./ItemwiseStrategyBasedIndicator";
import { WeightedCloseIndicatorStrategy } from "./WeightedCloseIndicatorStrategy";
import { markType } from "igniteui-angular-core";
/**
 * @hidden
 */
export let WeightedCloseIndicator = /*@__PURE__*/ (() => {
    class WeightedCloseIndicator extends ItemwiseStrategyBasedIndicator {
        bi() {
            return new WeightedCloseIndicator();
        }
        get_ab9() {
            return new WeightedCloseIndicatorStrategy();
        }
        get ab9() {
            return this.get_ab9();
        }
        get_ab0() {
            return WeightedCloseIndicator.$;
        }
        get ab0() {
            return this.get_ab0();
        }
    }
    WeightedCloseIndicator.$t = markType(WeightedCloseIndicator, 'WeightedCloseIndicator', ItemwiseStrategyBasedIndicator.$);
    return WeightedCloseIndicator;
})();
