// igniteui-angular-core/esm2015/lib/AnchoredRadialSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RadialBaseDescription } from "./RadialBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let AnchoredRadialSeriesDescription = /*@__PURE__*/ (() => {
    class AnchoredRadialSeriesDescription extends RadialBaseDescription {
        constructor() {
            super();
            this.g1 = false;
            this.hf = null;
            this.ha = null;
            this.hd = null;
            this.he = null;
            this.hc = null;
            this.hb = null;
            this.g9 = null;
            this.g3 = 0;
            this.gz = null;
            this.g5 = 0;
            this.g6 = 0;
        }
        get_type() {
            return "AnchoredRadialSeries";
        }
        get useCategoryNormalizedValues() {
            return this.g1;
        }
        set useCategoryNormalizedValues(a) {
            this.g1 = a;
            this.g("UseCategoryNormalizedValues");
        }
        get valueMemberPath() {
            return this.hf;
        }
        set valueMemberPath(a) {
            this.hf = a;
            this.g("ValueMemberPath");
        }
        get highlightedValueMemberPath() {
            return this.ha;
        }
        set highlightedValueMemberPath(a) {
            this.ha = a;
            this.g("HighlightedValueMemberPath");
        }
        get valueMemberAsLegendLabel() {
            return this.hd;
        }
        set valueMemberAsLegendLabel(a) {
            this.hd = a;
            this.g("ValueMemberAsLegendLabel");
        }
        get valueMemberAsLegendUnit() {
            return this.he;
        }
        set valueMemberAsLegendUnit(a) {
            this.he = a;
            this.g("ValueMemberAsLegendUnit");
        }
        get trendLineType() {
            return this.hc;
        }
        set trendLineType(a) {
            this.hc = a;
            this.g("TrendLineType");
        }
        get trendLineBrush() {
            return this.hb;
        }
        set trendLineBrush(a) {
            this.hb = a;
            this.g("TrendLineBrush");
        }
        get actualTrendLineBrush() {
            return this.g9;
        }
        set actualTrendLineBrush(a) {
            this.g9 = a;
            this.g("ActualTrendLineBrush");
        }
        get trendLineThickness() {
            return this.g3;
        }
        set trendLineThickness(a) {
            this.g3 = a;
            this.g("TrendLineThickness");
        }
        get trendLineDashArray() {
            return this.gz;
        }
        set trendLineDashArray(a) {
            this.gz = a;
            this.g("TrendLineDashArray");
        }
        get trendLinePeriod() {
            return this.g5;
        }
        set trendLinePeriod(a) {
            this.g5 = a;
            this.g("TrendLinePeriod");
        }
        get trendLineZIndex() {
            return this.g6;
        }
        set trendLineZIndex(a) {
            this.g6 = a;
            this.g("TrendLineZIndex");
        }
    }
    AnchoredRadialSeriesDescription.$t = markType(AnchoredRadialSeriesDescription, 'AnchoredRadialSeriesDescription', RadialBaseDescription.$);
    return AnchoredRadialSeriesDescription;
})();
