// igniteui-angular-core/esm2015/lib/AnchoredCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let AnchoredCategorySeriesDescription = /*@__PURE__*/ (() => {
    class AnchoredCategorySeriesDescription extends CategorySeriesDescription {
        constructor() {
            super();
            this.gf = null;
            this.f9 = null;
            this.gd = null;
            this.ge = null;
            this.gc = null;
            this.gb = null;
            this.f7 = null;
            this.f3 = 0;
            this.f1 = null;
            this.f5 = 0;
            this.ga = null;
            this.f8 = null;
        }
        get_type() {
            return "AnchoredCategorySeries";
        }
        get valueMemberPath() {
            return this.gf;
        }
        set valueMemberPath(a) {
            this.gf = a;
            this.g("ValueMemberPath");
        }
        get highlightedValueMemberPath() {
            return this.f9;
        }
        set highlightedValueMemberPath(a) {
            this.f9 = a;
            this.g("HighlightedValueMemberPath");
        }
        get valueMemberAsLegendLabel() {
            return this.gd;
        }
        set valueMemberAsLegendLabel(a) {
            this.gd = a;
            this.g("ValueMemberAsLegendLabel");
        }
        get valueMemberAsLegendUnit() {
            return this.ge;
        }
        set valueMemberAsLegendUnit(a) {
            this.ge = a;
            this.g("ValueMemberAsLegendUnit");
        }
        get trendLineType() {
            return this.gc;
        }
        set trendLineType(a) {
            this.gc = a;
            this.g("TrendLineType");
        }
        get trendLineBrush() {
            return this.gb;
        }
        set trendLineBrush(a) {
            this.gb = a;
            this.g("TrendLineBrush");
        }
        get actualTrendLineBrush() {
            return this.f7;
        }
        set actualTrendLineBrush(a) {
            this.f7 = a;
            this.g("ActualTrendLineBrush");
        }
        get trendLineThickness() {
            return this.f3;
        }
        set trendLineThickness(a) {
            this.f3 = a;
            this.g("TrendLineThickness");
        }
        get trendLineDashArray() {
            return this.f1;
        }
        set trendLineDashArray(a) {
            this.f1 = a;
            this.g("TrendLineDashArray");
        }
        get trendLinePeriod() {
            return this.f5;
        }
        set trendLinePeriod(a) {
            this.f5 = a;
            this.g("TrendLinePeriod");
        }
        get markerCollisionAvoidance() {
            return this.ga;
        }
        set markerCollisionAvoidance(a) {
            this.ga = a;
            this.g("MarkerCollisionAvoidance");
        }
        get consolidatedItemHitTestBehavior() {
            return this.f8;
        }
        set consolidatedItemHitTestBehavior(a) {
            this.f8 = a;
            this.g("ConsolidatedItemHitTestBehavior");
        }
    }
    AnchoredCategorySeriesDescription.$t = markType(AnchoredCategorySeriesDescription, 'AnchoredCategorySeriesDescription', CategorySeriesDescription.$);
    return AnchoredCategorySeriesDescription;
})();
