// igniteui-angular-core/esm2015/lib/AnnotationLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let AnnotationLayerDescription = /*@__PURE__*/ (() => {
    class AnnotationLayerDescription extends SeriesDescription {
        constructor() {
            super();
            this.fd = false;
            this.fe = false;
            this.e9 = null;
            this.fb = false;
            this.fy = null;
            this.fv = null;
            this.fn = 0;
            this.fk = 0;
            this.fz = null;
            this.fw = null;
            this.f0 = null;
            this.fx = null;
            this.fm = 0;
            this.fj = 0;
            this.fo = 0;
            this.fl = 0;
            this.e3 = null;
            this.e1 = null;
            this.e4 = null;
            this.e2 = null;
            this.fc = false;
        }
        get_type() {
            return "AnnotationLayer";
        }
        get useIndex() {
            return this.fd;
        }
        set useIndex(a) {
            this.fd = a;
            this.g("UseIndex");
        }
        get useLegend() {
            return this.fe;
        }
        set useLegend(a) {
            this.fe = a;
            this.g("UseLegend");
        }
        get cursorPosition() {
            return this.e9;
        }
        set cursorPosition(a) {
            this.e9 = a;
            this.g("CursorPosition");
        }
        get isDefaultCrosshairDisabled() {
            return this.fb;
        }
        set isDefaultCrosshairDisabled(a) {
            this.fb = a;
            this.g("IsDefaultCrosshairDisabled");
        }
        get appearanceMode() {
            return this.fy;
        }
        set appearanceMode(a) {
            this.fy = a;
            this.g("AppearanceMode");
        }
        get actualAppearanceMode() {
            return this.fv;
        }
        set actualAppearanceMode(a) {
            this.fv = a;
            this.g("ActualAppearanceMode");
        }
        get shiftAmount() {
            return this.fn;
        }
        set shiftAmount(a) {
            this.fn = a;
            this.g("ShiftAmount");
        }
        get actualShiftAmount() {
            return this.fk;
        }
        set actualShiftAmount(a) {
            this.fk = a;
            this.g("ActualShiftAmount");
        }
        get horizontalAppearanceMode() {
            return this.fz;
        }
        set horizontalAppearanceMode(a) {
            this.fz = a;
            this.g("HorizontalAppearanceMode");
        }
        get actualHorizontalAppearanceMode() {
            return this.fw;
        }
        set actualHorizontalAppearanceMode(a) {
            this.fw = a;
            this.g("ActualHorizontalAppearanceMode");
        }
        get verticalAppearanceMode() {
            return this.f0;
        }
        set verticalAppearanceMode(a) {
            this.f0 = a;
            this.g("VerticalAppearanceMode");
        }
        get actualVerticalAppearanceMode() {
            return this.fx;
        }
        set actualVerticalAppearanceMode(a) {
            this.fx = a;
            this.g("ActualVerticalAppearanceMode");
        }
        get horizontalShiftAmount() {
            return this.fm;
        }
        set horizontalShiftAmount(a) {
            this.fm = a;
            this.g("HorizontalShiftAmount");
        }
        get actualHorizontalShiftAmount() {
            return this.fj;
        }
        set actualHorizontalShiftAmount(a) {
            this.fj = a;
            this.g("ActualHorizontalShiftAmount");
        }
        get verticalShiftAmount() {
            return this.fo;
        }
        set verticalShiftAmount(a) {
            this.fo = a;
            this.g("VerticalShiftAmount");
        }
        get actualVerticalShiftAmount() {
            return this.fl;
        }
        set actualVerticalShiftAmount(a) {
            this.fl = a;
            this.g("ActualVerticalShiftAmount");
        }
        get horizontalDashArray() {
            return this.e3;
        }
        set horizontalDashArray(a) {
            this.e3 = a;
            this.g("HorizontalDashArray");
        }
        get actualHorizontalDashArray() {
            return this.e1;
        }
        set actualHorizontalDashArray(a) {
            this.e1 = a;
            this.g("ActualHorizontalDashArray");
        }
        get verticalDashArray() {
            return this.e4;
        }
        set verticalDashArray(a) {
            this.e4 = a;
            this.g("VerticalDashArray");
        }
        get actualVerticalDashArray() {
            return this.e2;
        }
        set actualVerticalDashArray(a) {
            this.e2 = a;
            this.g("ActualVerticalDashArray");
        }
        get shouldRenderAsOverlay() {
            return this.fc;
        }
        set shouldRenderAsOverlay(a) {
            this.fc = a;
            this.g("ShouldRenderAsOverlay");
        }
    }
    AnnotationLayerDescription.$t = markType(AnnotationLayerDescription, 'AnnotationLayerDescription', SeriesDescription.$);
    return AnnotationLayerDescription;
})();
