// igniteui-angular-core/esm2015/lib/BarSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { VerticalAnchoredCategorySeriesDescription } from "./VerticalAnchoredCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let BarSeriesDescription = /*@__PURE__*/ (() => {
    class BarSeriesDescription extends VerticalAnchoredCategorySeriesDescription {
        constructor() {
            super();
            this.gt = 0;
            this.gu = 0;
        }
        get_type() {
            return "BarSeries";
        }
        get radiusX() {
            return this.gt;
        }
        set radiusX(a) {
            this.gt = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.gu;
        }
        set radiusY(a) {
            this.gu = a;
            this.g("RadiusY");
        }
    }
    BarSeriesDescription.$t = markType(BarSeriesDescription, 'BarSeriesDescription', VerticalAnchoredCategorySeriesDescription.$);
    return BarSeriesDescription;
})();
