// igniteui-angular-core/esm2015/lib/BollingerBandsOverlayDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FinancialOverlayDescription } from "./FinancialOverlayDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let BollingerBandsOverlayDescription = /*@__PURE__*/ (() => {
    class BollingerBandsOverlayDescription extends FinancialOverlayDescription {
        constructor() {
            super();
            this.f7 = 0;
            this.f5 = 0;
        }
        get_type() {
            return "BollingerBandsOverlay";
        }
        get period() {
            return this.f7;
        }
        set period(a) {
            this.f7 = a;
            this.g("Period");
        }
        get multiplier() {
            return this.f5;
        }
        set multiplier(a) {
            this.f5 = a;
            this.g("Multiplier");
        }
    }
    BollingerBandsOverlayDescription.$t = markType(BollingerBandsOverlayDescription, 'BollingerBandsOverlayDescription', FinancialOverlayDescription.$);
    return BollingerBandsOverlayDescription;
})();
