// igniteui-angular-core/esm2015/lib/BubbleSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterBaseDescription } from "./ScatterBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let BubbleSeriesDescription = /*@__PURE__*/ (() => {
    class BubbleSeriesDescription extends ScatterBaseDescription {
        constructor() {
            super();
            this.hn = null;
            this.g5 = null;
            this.hk = null;
            this.hj = null;
            this.g3 = null;
            this.g7 = false;
            this.g8 = false;
            this.he = 0;
            this.hd = 0;
            this.g9 = false;
            this.hl = null;
            this.hh = null;
            this.hm = null;
            this.hi = null;
        }
        get_type() {
            return "BubbleSeries";
        }
        get radiusMemberPath() {
            return this.hn;
        }
        set radiusMemberPath(a) {
            this.hn = a;
            this.g("RadiusMemberPath");
        }
        get radiusScale() {
            return this.g5;
        }
        set radiusScale(a) {
            this.g5 = a;
            this.g("RadiusScale");
        }
        get labelMemberPath() {
            return this.hk;
        }
        set labelMemberPath(a) {
            this.hk = a;
            this.g("LabelMemberPath");
        }
        get fillMemberPath() {
            return this.hj;
        }
        set fillMemberPath(a) {
            this.hj = a;
            this.g("FillMemberPath");
        }
        get fillScale() {
            return this.g3;
        }
        set fillScale(a) {
            this.g3 = a;
            this.g("FillScale");
        }
        get fillScaleUseGlobalValues() {
            return this.g7;
        }
        set fillScaleUseGlobalValues(a) {
            this.g7 = a;
            this.g("FillScaleUseGlobalValues");
        }
        get markerOutlineUsesFillScale() {
            return this.g8;
        }
        set markerOutlineUsesFillScale(a) {
            this.g8 = a;
            this.g("MarkerOutlineUsesFillScale");
        }
        get markerOutlineBrightness() {
            return this.he;
        }
        set markerOutlineBrightness(a) {
            this.he = a;
            this.g("MarkerOutlineBrightness");
        }
        get markerBrushBrightness() {
            return this.hd;
        }
        set markerBrushBrightness(a) {
            this.hd = a;
            this.g("MarkerBrushBrightness");
        }
        get radiusScaleUseGlobalValues() {
            return this.g9;
        }
        set radiusScaleUseGlobalValues(a) {
            this.g9 = a;
            this.g("RadiusScaleUseGlobalValues");
        }
        get radiusMemberAsLegendLabel() {
            return this.hl;
        }
        set radiusMemberAsLegendLabel(a) {
            this.hl = a;
            this.g("RadiusMemberAsLegendLabel");
        }
        get fillMemberAsLegendLabel() {
            return this.hh;
        }
        set fillMemberAsLegendLabel(a) {
            this.hh = a;
            this.g("FillMemberAsLegendLabel");
        }
        get radiusMemberAsLegendUnit() {
            return this.hm;
        }
        set radiusMemberAsLegendUnit(a) {
            this.hm = a;
            this.g("RadiusMemberAsLegendUnit");
        }
        get fillMemberAsLegendUnit() {
            return this.hi;
        }
        set fillMemberAsLegendUnit(a) {
            this.hi = a;
            this.g("FillMemberAsLegendUnit");
        }
    }
    BubbleSeriesDescription.$t = markType(BubbleSeriesDescription, 'BubbleSeriesDescription', ScatterBaseDescription.$);
    return BubbleSeriesDescription;
})();
