// igniteui-angular-core/esm2015/lib/CalloutLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CalloutLayerDescription = /*@__PURE__*/ (() => {
    class CalloutLayerDescription extends AnnotationLayerDescription {
        constructor() {
            super();
            this.hv = null;
            this.hz = null;
            this.h0 = null;
            this.ht = null;
            this.gc = false;
            this.hu = null;
            this.f8 = null;
            this.hr = null;
            this.hw = null;
            this.hx = null;
            this.hq = null;
            this.f7 = null;
            this.gb = false;
            this.gg = false;
            this.gj = false;
            this.gf = false;
            this.gh = false;
            this.gi = false;
            this.gk = false;
            this.hs = null;
            this.hd = null;
            this.gv = 0;
            this.ge = false;
            this.gd = false;
            this.g9 = 0;
            this.hp = null;
            this.hk = null;
            this.hh = null;
            this.he = null;
            this.ha = 0;
            this.hl = null;
            this.hj = null;
            this.gx = 0;
            this.gz = 0;
            this.gy = 0;
            this.gw = 0;
            this.hf = null;
            this.g0 = 0;
            this.g1 = 0;
            this.hy = null;
            this.ho = null;
            this.hm = null;
            this.hg = null;
            this.hi = null;
            this.hn = null;
        }
        get_type() {
            return "CalloutLayer";
        }
        get labelMemberPath() {
            return this.hv;
        }
        set labelMemberPath(a) {
            this.hv = a;
            this.g("LabelMemberPath");
        }
        get xMemberPath() {
            return this.hz;
        }
        set xMemberPath(a) {
            this.hz = a;
            this.g("XMemberPath");
        }
        get yMemberPath() {
            return this.h0;
        }
        set yMemberPath(a) {
            this.h0 = a;
            this.g("YMemberPath");
        }
        get keyMemberPath() {
            return this.ht;
        }
        set keyMemberPath(a) {
            this.ht = a;
            this.g("KeyMemberPath");
        }
        get isCalloutOffsettingEnabled() {
            return this.gc;
        }
        set isCalloutOffsettingEnabled(a) {
            this.gc = a;
            this.g("IsCalloutOffsettingEnabled");
        }
        get labelFormat() {
            return this.hu;
        }
        set labelFormat(a) {
            this.hu = a;
            this.g("LabelFormat");
        }
        get labelFormatSpecifiers() {
            return this.f8;
        }
        set labelFormatSpecifiers(a) {
            this.f8 = a;
            this.g("LabelFormatSpecifiers");
        }
        get contentMemberPath() {
            return this.hr;
        }
        set contentMemberPath(a) {
            this.hr = a;
            this.g("ContentMemberPath");
        }
        get targetSeriesName() {
            return this.hw;
        }
        set targetSeriesName(a) {
            this.hw = a;
            this.g("TargetSeriesName");
        }
        get targetSeriesRef() {
            return this.hx;
        }
        set targetSeriesRef(a) {
            this.hx = a;
            this.g("TargetSeriesRef");
        }
        get collisionChannel() {
            return this.hq;
        }
        set collisionChannel(a) {
            this.hq = a;
            this.g("CollisionChannel");
        }
        get allowedPositions() {
            return this.f7;
        }
        set allowedPositions(a) {
            this.f7 = a;
            this.g("AllowedPositions");
        }
        get isAutoCalloutBehaviorEnabled() {
            return this.gb;
        }
        set isAutoCalloutBehaviorEnabled(a) {
            this.gb = a;
            this.g("IsAutoCalloutBehaviorEnabled");
        }
        get useInterpolatedValueForAutoCalloutLabels() {
            return this.gg;
        }
        set useInterpolatedValueForAutoCalloutLabels(a) {
            this.gg = a;
            this.g("UseInterpolatedValueForAutoCalloutLabels");
        }
        get useSeriesColorForOutline() {
            return this.gj;
        }
        set useSeriesColorForOutline(a) {
            this.gj = a;
            this.g("UseSeriesColorForOutline");
        }
        get useAutoContrastingLabelColors() {
            return this.gf;
        }
        set useAutoContrastingLabelColors(a) {
            this.gf = a;
            this.g("UseAutoContrastingLabelColors");
        }
        get useItemColorForFill() {
            return this.gh;
        }
        set useItemColorForFill(a) {
            this.gh = a;
            this.g("UseItemColorForFill");
        }
        get useItemColorForOutline() {
            return this.gi;
        }
        set useItemColorForOutline(a) {
            this.gi = a;
            this.g("UseItemColorForOutline");
        }
        get useValueForAutoCalloutLabels() {
            return this.gk;
        }
        set useValueForAutoCalloutLabels(a) {
            this.gk = a;
            this.g("UseValueForAutoCalloutLabels");
        }
        get highlightedValueLabelMode() {
            return this.hs;
        }
        set highlightedValueLabelMode(a) {
            this.hs = a;
            this.g("HighlightedValueLabelMode");
        }
        get autoCalloutVisibilityMode() {
            return this.hd;
        }
        set autoCalloutVisibilityMode(a) {
            this.hd = a;
            this.g("AutoCalloutVisibilityMode");
        }
        get calloutCornerRadius() {
            return this.gv;
        }
        set calloutCornerRadius(a) {
            this.gv = a;
            this.g("CalloutCornerRadius");
        }
        get isCustomCalloutStyleEnabled() {
            return this.ge;
        }
        set isCustomCalloutStyleEnabled(a) {
            this.ge = a;
            this.g("IsCustomCalloutStyleEnabled");
        }
        get isCustomCalloutRenderStyleEnabled() {
            return this.gd;
        }
        set isCustomCalloutRenderStyleEnabled(a) {
            this.gd = a;
            this.g("IsCustomCalloutRenderStyleEnabled");
        }
        get autoCalloutLabelPrecision() {
            return this.g9;
        }
        set autoCalloutLabelPrecision(a) {
            this.g9 = a;
            this.g("AutoCalloutLabelPrecision");
        }
        get calloutTextColor() {
            return this.hp;
        }
        set calloutTextColor(a) {
            this.hp = a;
            this.g("CalloutTextColor");
        }
        get calloutLightTextColor() {
            return this.hk;
        }
        set calloutLightTextColor(a) {
            this.hk = a;
            this.g("CalloutLightTextColor");
        }
        get calloutDarkTextColor() {
            return this.hh;
        }
        set calloutDarkTextColor(a) {
            this.hh = a;
            this.g("CalloutDarkTextColor");
        }
        get calloutBackground() {
            return this.he;
        }
        set calloutBackground(a) {
            this.he = a;
            this.g("CalloutBackground");
        }
        get calloutInterpolatedValuePrecision() {
            return this.ha;
        }
        set calloutInterpolatedValuePrecision(a) {
            this.ha = a;
            this.g("CalloutInterpolatedValuePrecision");
        }
        get calloutOutline() {
            return this.hl;
        }
        set calloutOutline(a) {
            this.hl = a;
            this.g("CalloutOutline");
        }
        get calloutLeaderBrush() {
            return this.hj;
        }
        set calloutLeaderBrush(a) {
            this.hj = a;
            this.g("CalloutLeaderBrush");
        }
        get calloutPaddingLeft() {
            return this.gx;
        }
        set calloutPaddingLeft(a) {
            this.gx = a;
            this.g("CalloutPaddingLeft");
        }
        get calloutPaddingTop() {
            return this.gz;
        }
        set calloutPaddingTop(a) {
            this.gz = a;
            this.g("CalloutPaddingTop");
        }
        get calloutPaddingRight() {
            return this.gy;
        }
        set calloutPaddingRight(a) {
            this.gy = a;
            this.g("CalloutPaddingRight");
        }
        get calloutPaddingBottom() {
            return this.gw;
        }
        set calloutPaddingBottom(a) {
            this.gw = a;
            this.g("CalloutPaddingBottom");
        }
        get calloutCollisionMode() {
            return this.hf;
        }
        set calloutCollisionMode(a) {
            this.hf = a;
            this.g("CalloutCollisionMode");
        }
        get calloutPositionPadding() {
            return this.g0;
        }
        set calloutPositionPadding(a) {
            this.g0 = a;
            this.g("CalloutPositionPadding");
        }
        get calloutStrokeThickness() {
            return this.g1;
        }
        set calloutStrokeThickness(a) {
            this.g1 = a;
            this.g("CalloutStrokeThickness");
        }
        get textStyle() {
            return this.hy;
        }
        set textStyle(a) {
            this.hy = a;
            this.g("TextStyle");
        }
        get calloutStyleUpdatingRef() {
            return this.ho;
        }
        set calloutStyleUpdatingRef(a) {
            this.ho = a;
            this.g("CalloutStyleUpdatingRef");
        }
        get calloutRenderStyleUpdatingRef() {
            return this.hm;
        }
        set calloutRenderStyleUpdatingRef(a) {
            this.hm = a;
            this.g("CalloutRenderStyleUpdatingRef");
        }
        get calloutContentUpdatingRef() {
            return this.hg;
        }
        set calloutContentUpdatingRef(a) {
            this.hg = a;
            this.g("CalloutContentUpdatingRef");
        }
        get calloutLabelUpdatingRef() {
            return this.hi;
        }
        set calloutLabelUpdatingRef(a) {
            this.hi = a;
            this.g("CalloutLabelUpdatingRef");
        }
        get calloutSeriesSelectingRef() {
            return this.hn;
        }
        set calloutSeriesSelectingRef(a) {
            this.hn = a;
            this.g("CalloutSeriesSelectingRef");
        }
    }
    CalloutLayerDescription.$t = markType(CalloutLayerDescription, 'CalloutLayerDescription', AnnotationLayerDescription.$);
    return CalloutLayerDescription;
})();
