// igniteui-angular-core/esm2015/lib/CategoryToolTipLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CategoryToolTipLayerDescription = /*@__PURE__*/ (() => {
    class CategoryToolTipLayerDescription extends AnnotationLayerDescription {
        constructor() {
            super();
            this.f9 = null;
            this.f7 = false;
        }
        get_type() {
            return "CategoryToolTipLayer";
        }
        get targetAxisRef() {
            return this.f9;
        }
        set targetAxisRef(a) {
            this.f9 = a;
            this.g("TargetAxisRef");
        }
        get useInterpolation() {
            return this.f7;
        }
        set useInterpolation(a) {
            this.f7 = a;
            this.g("UseInterpolation");
        }
    }
    CategoryToolTipLayerDescription.$t = markType(CategoryToolTipLayerDescription, 'CategoryToolTipLayerDescription', AnnotationLayerDescription.$);
    return CategoryToolTipLayerDescription;
})();
