// igniteui-angular-core/esm2015/lib/CategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CategorySeriesDescription = /*@__PURE__*/ (() => {
    class CategorySeriesDescription extends MarkerSeriesDescription {
        constructor() {
            super();
            this.fo = false;
            this.fn = false;
            this.fq = false;
            this.fx = null;
            this.fp = false;
            this.fw = null;
            this.fv = null;
        }
        get_type() {
            return "CategorySeries";
        }
        get isCustomCategoryStyleAllowed() {
            return this.fo;
        }
        set isCustomCategoryStyleAllowed(a) {
            this.fo = a;
            this.g("IsCustomCategoryStyleAllowed");
        }
        get isCustomCategoryMarkerStyleAllowed() {
            return this.fn;
        }
        set isCustomCategoryMarkerStyleAllowed(a) {
            this.fn = a;
            this.g("IsCustomCategoryMarkerStyleAllowed");
        }
        get useHighMarkerFidelity() {
            return this.fq;
        }
        set useHighMarkerFidelity(a) {
            this.fq = a;
            this.g("UseHighMarkerFidelity");
        }
        get transitionInMode() {
            return this.fx;
        }
        set transitionInMode(a) {
            this.fx = a;
            this.g("TransitionInMode");
        }
        get isTransitionInEnabled() {
            return this.fp;
        }
        set isTransitionInEnabled(a) {
            this.fp = a;
            this.g("IsTransitionInEnabled");
        }
        get assigningCategoryStyleRef() {
            return this.fw;
        }
        set assigningCategoryStyleRef(a) {
            this.fw = a;
            this.g("AssigningCategoryStyleRef");
        }
        get assigningCategoryMarkerStyleRef() {
            return this.fv;
        }
        set assigningCategoryMarkerStyleRef(a) {
            this.fv = a;
            this.g("AssigningCategoryMarkerStyleRef");
        }
    }
    CategorySeriesDescription.$t = markType(CategorySeriesDescription, 'CategorySeriesDescription', MarkerSeriesDescription.$);
    return CategorySeriesDescription;
})();
