// igniteui-angular-core/esm2015/lib/CategoryHighlightLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let CategoryHighlightLayerDescription = /*@__PURE__*/ (() => {
    class CategoryHighlightLayerDescription extends AnnotationLayerDescription {
        constructor() {
            super();
            this.gb = null;
            this.f7 = false;
            this.f9 = 0;
        }
        get_type() {
            return "CategoryHighlightLayer";
        }
        get targetAxisRef() {
            return this.gb;
        }
        set targetAxisRef(a) {
            this.gb = a;
            this.g("TargetAxisRef");
        }
        get useInterpolation() {
            return this.f7;
        }
        set useInterpolation(a) {
            this.f7 = a;
            this.g("UseInterpolation");
        }
        get bandHighlightWidth() {
            return this.f9;
        }
        set bandHighlightWidth(a) {
            this.f9 = a;
            this.g("BandHighlightWidth");
        }
    }
    CategoryHighlightLayerDescription.$t = markType(CategoryHighlightLayerDescription, 'CategoryHighlightLayerDescription', AnnotationLayerDescription.$);
    return CategoryHighlightLayerDescription;
})();
