// igniteui-angular-core/esm2015/lib/CellKeyDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let CellKeyDescription = /*@__PURE__*/ (() => {
    class CellKeyDescription extends Description {
        constructor() {
            super();
            this.h = null;
            this.k = null;
            this.n = null;
        }
        get_type() {
            return "CellKey";
        }
        get type() {
            return this.get_type();
        }
        get primaryKey() {
            return this.h;
        }
        set primaryKey(a) {
            this.h = a;
            this.g("PrimaryKey");
        }
        get rowItem() {
            return this.k;
        }
        set rowItem(a) {
            this.k = a;
            this.g("RowItem");
        }
        get columnUniqueKey() {
            return this.n;
        }
        set columnUniqueKey(a) {
            this.n = a;
            this.g("ColumnUniqueKey");
        }
    }
    CellKeyDescription.$t = markType(CellKeyDescription, 'CellKeyDescription', Description.$);
    CellKeyDescription.__marshalByValue = true;
    CellKeyDescription.__marshalByValueAlias = "CellKey";
    return CellKeyDescription;
})();
