// igniteui-angular-core/esm2015/lib/ChaikinOscillatorIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicatorDescription } from "./StrategyBasedIndicatorDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ChaikinOscillatorIndicatorDescription = /*@__PURE__*/ (() => {
    class ChaikinOscillatorIndicatorDescription extends StrategyBasedIndicatorDescription {
        constructor() {
            super();
            this.gk = 0;
            this.gj = 0;
        }
        get_type() {
            return "ChaikinOscillatorIndicator";
        }
        get shortPeriod() {
            return this.gk;
        }
        set shortPeriod(a) {
            this.gk = a;
            this.g("ShortPeriod");
        }
        get longPeriod() {
            return this.gj;
        }
        set longPeriod(a) {
            this.gj = a;
            this.g("LongPeriod");
        }
    }
    ChaikinOscillatorIndicatorDescription.$t = markType(ChaikinOscillatorIndicatorDescription, 'ChaikinOscillatorIndicatorDescription', StrategyBasedIndicatorDescription.$);
    return ChaikinOscillatorIndicatorDescription;
})();
