// igniteui-angular-core/esm2015/lib/ChartSortDescriptionDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ChartSortDescriptionDescription = /*@__PURE__*/ (() => {
    class ChartSortDescriptionDescription extends Description {
        constructor() {
            super();
            this.j = null;
            this.k = null;
        }
        get_type() {
            return "ChartSortDescription";
        }
        get type() {
            return this.get_type();
        }
        get field() {
            return this.j;
        }
        set field(a) {
            this.j = a;
            this.g("Field");
        }
        get sortDirection() {
            return this.k;
        }
        set sortDirection(a) {
            this.k = a;
            this.g("SortDirection");
        }
    }
    ChartSortDescriptionDescription.$t = markType(ChartSortDescriptionDescription, 'ChartSortDescriptionDescription', Description.$);
    ChartSortDescriptionDescription.__marshalByValue = true;
    ChartSortDescriptionDescription.__marshalByValueAlias = "ChartSortDescription";
    return ChartSortDescriptionDescription;
})();
