// igniteui-angular-core/esm2015/lib/ColumnSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalAnchoredCategorySeriesDescription } from "./HorizontalAnchoredCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ColumnSeriesDescription = /*@__PURE__*/ (() => {
    class ColumnSeriesDescription extends HorizontalAnchoredCategorySeriesDescription {
        constructor() {
            super();
            this.gt = 0;
            this.gu = 0;
            this.gx = null;
        }
        get_type() {
            return "ColumnSeries";
        }
        get radiusX() {
            return this.gt;
        }
        set radiusX(a) {
            this.gt = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.gu;
        }
        set radiusY(a) {
            this.gu = a;
            this.g("RadiusY");
        }
        get consolidatedColumnVerticalPosition() {
            return this.gx;
        }
        set consolidatedColumnVerticalPosition(a) {
            this.gx = a;
            this.g("ConsolidatedColumnVerticalPosition");
        }
    }
    ColumnSeriesDescription.$t = markType(ColumnSeriesDescription, 'ColumnSeriesDescription', HorizontalAnchoredCategorySeriesDescription.$);
    return ColumnSeriesDescription;
})();
