// igniteui-angular-core/esm2015/lib/DataLegendDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataLegendDescription = /*@__PURE__*/ (() => {
    class DataLegendDescription extends Description {
        constructor() {
            super();
            this.t = false;
            this.ah = 0;
            this.l = null;
            this.i = null;
            this.k = null;
            this.h = null;
            this.w = false;
            this.fo = null;
            this.dy = 0;
            this.dz = 0;
            this.fr = null;
            this.fp = null;
            this.fq = null;
            this.x = false;
            this.fs = null;
            this.m = null;
            this.by = 0;
            this.bz = 0;
            this.b0 = 0;
            this.b1 = 0;
            this.y = false;
            this.fx = null;
            this.z = false;
            this.b3 = 0;
            this.b4 = 0;
            this.b5 = 0;
            this.b6 = 0;
            this.ft = null;
            this.fu = null;
            this.b2 = 0;
            this.fv = null;
            this.fw = null;
            this.ea = null;
            this.j = null;
            this.d8 = null;
            this.d9 = null;
            this.eb = null;
            this.ec = null;
            this.ed = null;
            this.a0 = 0;
            this.a1 = 0;
            this.a2 = 0;
            this.a3 = 0;
            this.av = 0;
            this.aw = 0;
            this.ax = 0;
            this.ay = 0;
            this.v = false;
            this.ee = null;
            this.az = 0;
            this.ef = null;
            this.eg = null;
            this.u = false;
            this.d4 = null;
            this.ar = 0;
            this.as = 0;
            this.at = 0;
            this.au = 0;
            this.am = 0;
            this.an = 0;
            this.ao = 0;
            this.ap = 0;
            this.d5 = null;
            this.aq = 0;
            this.d6 = null;
            this.d7 = null;
            this.ez = null;
            this.e0 = null;
            this.bf = 0;
            this.e1 = null;
            this.e2 = null;
            this.e3 = null;
            this.ey = null;
            this.bg = 0;
            this.bh = 0;
            this.bi = 0;
            this.bj = 0;
            this.bb = 0;
            this.bc = 0;
            this.bd = 0;
            this.be = 0;
            this.e9 = null;
            this.fa = null;
            this.bl = 0;
            this.fb = null;
            this.fc = null;
            this.et = null;
            this.eu = null;
            this.ev = null;
            this.ba = 0;
            this.ew = null;
            this.ex = null;
            this.e4 = null;
            this.e5 = null;
            this.e6 = null;
            this.bk = 0;
            this.e7 = null;
            this.e8 = null;
            this.ai = 0;
            this.aj = 0;
            this.ak = 0;
            this.al = 0;
            this.d2 = null;
            this.bm = 0;
            this.bn = 0;
            this.fd = null;
            this.fi = null;
            this.fj = null;
            this.fk = null;
            this.bu = 0;
            this.bv = 0;
            this.bw = 0;
            this.bx = 0;
            this.fl = null;
            this.bt = 0;
            this.fm = null;
            this.fn = null;
            this.bp = 0;
            this.bq = 0;
            this.br = 0;
            this.bs = 0;
            this.fe = null;
            this.ff = null;
            this.bo = 0;
            this.fg = null;
            this.fh = null;
            this.ei = null;
            this.ej = null;
            this.a5 = 0;
            this.a6 = 0;
            this.a7 = 0;
            this.a8 = 0;
            this.ek = null;
            this.a4 = 0;
            this.el = null;
            this.em = null;
            this.a9 = 0;
            this.d3 = null;
            this.eo = null;
            this.en = null;
            this.eq = null;
            this.es = null;
            this.ep = null;
            this.er = null;
            this.fy = null;
            this.eh = null;
        }
        get_type() {
            return "DataLegend";
        }
        get type() {
            return this.get_type();
        }
        get actualBadgesVisible() {
            return this.t;
        }
        set actualBadgesVisible(a) {
            this.t = a;
            this.g("ActualBadgesVisible");
        }
        get actualPixelScalingRatio() {
            return this.ah;
        }
        set actualPixelScalingRatio(a) {
            this.ah = a;
            this.g("ActualPixelScalingRatio");
        }
        get includedSeries() {
            return this.l;
        }
        set includedSeries(a) {
            this.l = a;
            this.g("IncludedSeries");
        }
        get excludedSeries() {
            return this.i;
        }
        set excludedSeries(a) {
            this.i = a;
            this.g("ExcludedSeries");
        }
        get includedColumns() {
            return this.k;
        }
        set includedColumns(a) {
            this.k = a;
            this.g("IncludedColumns");
        }
        get excludedColumns() {
            return this.h;
        }
        set excludedColumns(a) {
            this.h = a;
            this.g("ExcludedColumns");
        }
        get shouldUpdateWhenSeriesDataChanges() {
            return this.w;
        }
        set shouldUpdateWhenSeriesDataChanges(a) {
            this.w = a;
            this.g("ShouldUpdateWhenSeriesDataChanges");
        }
        get valueFormatAbbreviation() {
            return this.fo;
        }
        set valueFormatAbbreviation(a) {
            this.fo = a;
            this.g("ValueFormatAbbreviation");
        }
        get valueFormatMaxFractions() {
            return this.dy;
        }
        set valueFormatMaxFractions(a) {
            this.dy = a;
            this.g("ValueFormatMaxFractions");
        }
        get valueFormatMinFractions() {
            return this.dz;
        }
        set valueFormatMinFractions(a) {
            this.dz = a;
            this.g("ValueFormatMinFractions");
        }
        get valueFormatMode() {
            return this.fr;
        }
        set valueFormatMode(a) {
            this.fr = a;
            this.g("ValueFormatMode");
        }
        get valueFormatCulture() {
            return this.fp;
        }
        set valueFormatCulture(a) {
            this.fp = a;
            this.g("ValueFormatCulture");
        }
        get valueFormatCurrencyCode() {
            return this.fq;
        }
        set valueFormatCurrencyCode(a) {
            this.fq = a;
            this.g("ValueFormatCurrencyCode");
        }
        get valueFormatUseGrouping() {
            return this.x;
        }
        set valueFormatUseGrouping(a) {
            this.x = a;
            this.g("ValueFormatUseGrouping");
        }
        get valueFormatString() {
            return this.fs;
        }
        set valueFormatString(a) {
            this.fs = a;
            this.g("ValueFormatString");
        }
        get valueFormatSpecifiers() {
            return this.m;
        }
        set valueFormatSpecifiers(a) {
            this.m = a;
            this.g("ValueFormatSpecifiers");
        }
        get valueRowMarginBottom() {
            return this.by;
        }
        set valueRowMarginBottom(a) {
            this.by = a;
            this.g("ValueRowMarginBottom");
        }
        get valueRowMarginLeft() {
            return this.bz;
        }
        set valueRowMarginLeft(a) {
            this.bz = a;
            this.g("ValueRowMarginLeft");
        }
        get valueRowMarginRight() {
            return this.b0;
        }
        set valueRowMarginRight(a) {
            this.b0 = a;
            this.g("ValueRowMarginRight");
        }
        get valueRowMarginTop() {
            return this.b1;
        }
        set valueRowMarginTop(a) {
            this.b1 = a;
            this.g("ValueRowMarginTop");
        }
        get valueRowVisible() {
            return this.y;
        }
        set valueRowVisible(a) {
            this.y = a;
            this.g("ValueRowVisible");
        }
        get valueTextWhenMissingData() {
            return this.fx;
        }
        set valueTextWhenMissingData(a) {
            this.fx = a;
            this.g("ValueTextWhenMissingData");
        }
        get valueTextUseSeriesColors() {
            return this.z;
        }
        set valueTextUseSeriesColors(a) {
            this.z = a;
            this.g("ValueTextUseSeriesColors");
        }
        get valueTextMarginBottom() {
            return this.b3;
        }
        set valueTextMarginBottom(a) {
            this.b3 = a;
            this.g("ValueTextMarginBottom");
        }
        get valueTextMarginLeft() {
            return this.b4;
        }
        set valueTextMarginLeft(a) {
            this.b4 = a;
            this.g("ValueTextMarginLeft");
        }
        get valueTextMarginRight() {
            return this.b5;
        }
        set valueTextMarginRight(a) {
            this.b5 = a;
            this.g("ValueTextMarginRight");
        }
        get valueTextMarginTop() {
            return this.b6;
        }
        set valueTextMarginTop(a) {
            this.b6 = a;
            this.g("ValueTextMarginTop");
        }
        get valueTextColor() {
            return this.ft;
        }
        set valueTextColor(a) {
            this.ft = a;
            this.g("ValueTextColor");
        }
        get valueTextFontFamily() {
            return this.fu;
        }
        set valueTextFontFamily(a) {
            this.fu = a;
            this.g("ValueTextFontFamily");
        }
        get valueTextFontSize() {
            return this.b2;
        }
        set valueTextFontSize(a) {
            this.b2 = a;
            this.g("ValueTextFontSize");
        }
        get valueTextFontStyle() {
            return this.fv;
        }
        set valueTextFontStyle(a) {
            this.fv = a;
            this.g("ValueTextFontStyle");
        }
        get valueTextFontWeight() {
            return this.fw;
        }
        set valueTextFontWeight(a) {
            this.fw = a;
            this.g("ValueTextFontWeight");
        }
        get headerFormatString() {
            return this.ea;
        }
        set headerFormatString(a) {
            this.ea = a;
            this.g("HeaderFormatString");
        }
        get headerFormatSpecifiers() {
            return this.j;
        }
        set headerFormatSpecifiers(a) {
            this.j = a;
            this.g("HeaderFormatSpecifiers");
        }
        get headerFormatCulture() {
            return this.d8;
        }
        set headerFormatCulture(a) {
            this.d8 = a;
            this.g("HeaderFormatCulture");
        }
        get headerFormatDate() {
            return this.d9;
        }
        set headerFormatDate(a) {
            this.d9 = a;
            this.g("HeaderFormatDate");
        }
        get headerFormatTime() {
            return this.eb;
        }
        set headerFormatTime(a) {
            this.eb = a;
            this.g("HeaderFormatTime");
        }
        get headerText() {
            return this.ec;
        }
        set headerText(a) {
            this.ec = a;
            this.g("HeaderText");
        }
        get headerTextColor() {
            return this.ed;
        }
        set headerTextColor(a) {
            this.ed = a;
            this.g("HeaderTextColor");
        }
        get headerTextMarginBottom() {
            return this.a0;
        }
        set headerTextMarginBottom(a) {
            this.a0 = a;
            this.g("HeaderTextMarginBottom");
        }
        get headerTextMarginLeft() {
            return this.a1;
        }
        set headerTextMarginLeft(a) {
            this.a1 = a;
            this.g("HeaderTextMarginLeft");
        }
        get headerTextMarginRight() {
            return this.a2;
        }
        set headerTextMarginRight(a) {
            this.a2 = a;
            this.g("HeaderTextMarginRight");
        }
        get headerTextMarginTop() {
            return this.a3;
        }
        set headerTextMarginTop(a) {
            this.a3 = a;
            this.g("HeaderTextMarginTop");
        }
        get headerRowMarginBottom() {
            return this.av;
        }
        set headerRowMarginBottom(a) {
            this.av = a;
            this.g("HeaderRowMarginBottom");
        }
        get headerRowMarginLeft() {
            return this.aw;
        }
        set headerRowMarginLeft(a) {
            this.aw = a;
            this.g("HeaderRowMarginLeft");
        }
        get headerRowMarginRight() {
            return this.ax;
        }
        set headerRowMarginRight(a) {
            this.ax = a;
            this.g("HeaderRowMarginRight");
        }
        get headerRowMarginTop() {
            return this.ay;
        }
        set headerRowMarginTop(a) {
            this.ay = a;
            this.g("HeaderRowMarginTop");
        }
        get headerRowVisible() {
            return this.v;
        }
        set headerRowVisible(a) {
            this.v = a;
            this.g("HeaderRowVisible");
        }
        get headerTextFontFamily() {
            return this.ee;
        }
        set headerTextFontFamily(a) {
            this.ee = a;
            this.g("HeaderTextFontFamily");
        }
        get headerTextFontSize() {
            return this.az;
        }
        set headerTextFontSize(a) {
            this.az = a;
            this.g("HeaderTextFontSize");
        }
        get headerTextFontStyle() {
            return this.ef;
        }
        set headerTextFontStyle(a) {
            this.ef = a;
            this.g("HeaderTextFontStyle");
        }
        get headerTextFontWeight() {
            return this.eg;
        }
        set headerTextFontWeight(a) {
            this.eg = a;
            this.g("HeaderTextFontWeight");
        }
        get groupRowVisible() {
            return this.u;
        }
        set groupRowVisible(a) {
            this.u = a;
            this.g("GroupRowVisible");
        }
        get groupTextColor() {
            return this.d4;
        }
        set groupTextColor(a) {
            this.d4 = a;
            this.g("GroupTextColor");
        }
        get groupTextMarginBottom() {
            return this.ar;
        }
        set groupTextMarginBottom(a) {
            this.ar = a;
            this.g("GroupTextMarginBottom");
        }
        get groupTextMarginLeft() {
            return this.as;
        }
        set groupTextMarginLeft(a) {
            this.as = a;
            this.g("GroupTextMarginLeft");
        }
        get groupTextMarginRight() {
            return this.at;
        }
        set groupTextMarginRight(a) {
            this.at = a;
            this.g("GroupTextMarginRight");
        }
        get groupTextMarginTop() {
            return this.au;
        }
        set groupTextMarginTop(a) {
            this.au = a;
            this.g("GroupTextMarginTop");
        }
        get groupRowMarginBottom() {
            return this.am;
        }
        set groupRowMarginBottom(a) {
            this.am = a;
            this.g("GroupRowMarginBottom");
        }
        get groupRowMarginLeft() {
            return this.an;
        }
        set groupRowMarginLeft(a) {
            this.an = a;
            this.g("GroupRowMarginLeft");
        }
        get groupRowMarginRight() {
            return this.ao;
        }
        set groupRowMarginRight(a) {
            this.ao = a;
            this.g("GroupRowMarginRight");
        }
        get groupRowMarginTop() {
            return this.ap;
        }
        set groupRowMarginTop(a) {
            this.ap = a;
            this.g("GroupRowMarginTop");
        }
        get groupTextFontFamily() {
            return this.d5;
        }
        set groupTextFontFamily(a) {
            this.d5 = a;
            this.g("GroupTextFontFamily");
        }
        get groupTextFontSize() {
            return this.aq;
        }
        set groupTextFontSize(a) {
            this.aq = a;
            this.g("GroupTextFontSize");
        }
        get groupTextFontStyle() {
            return this.d6;
        }
        set groupTextFontStyle(a) {
            this.d6 = a;
            this.g("GroupTextFontStyle");
        }
        get groupTextFontWeight() {
            return this.d7;
        }
        set groupTextFontWeight(a) {
            this.d7 = a;
            this.g("GroupTextFontWeight");
        }
        get summaryTitleTextColor() {
            return this.ez;
        }
        set summaryTitleTextColor(a) {
            this.ez = a;
            this.g("SummaryTitleTextColor");
        }
        get summaryTitleTextFontFamily() {
            return this.e0;
        }
        set summaryTitleTextFontFamily(a) {
            this.e0 = a;
            this.g("SummaryTitleTextFontFamily");
        }
        get summaryTitleTextFontSize() {
            return this.bf;
        }
        set summaryTitleTextFontSize(a) {
            this.bf = a;
            this.g("SummaryTitleTextFontSize");
        }
        get summaryTitleTextFontStyle() {
            return this.e1;
        }
        set summaryTitleTextFontStyle(a) {
            this.e1 = a;
            this.g("SummaryTitleTextFontStyle");
        }
        get summaryTitleTextFontWeight() {
            return this.e2;
        }
        set summaryTitleTextFontWeight(a) {
            this.e2 = a;
            this.g("SummaryTitleTextFontWeight");
        }
        get summaryType() {
            return this.e3;
        }
        set summaryType(a) {
            this.e3 = a;
            this.g("SummaryType");
        }
        get summaryTitleText() {
            return this.ey;
        }
        set summaryTitleText(a) {
            this.ey = a;
            this.g("SummaryTitleText");
        }
        get summaryTitleTextMarginBottom() {
            return this.bg;
        }
        set summaryTitleTextMarginBottom(a) {
            this.bg = a;
            this.g("SummaryTitleTextMarginBottom");
        }
        get summaryTitleTextMarginLeft() {
            return this.bh;
        }
        set summaryTitleTextMarginLeft(a) {
            this.bh = a;
            this.g("SummaryTitleTextMarginLeft");
        }
        get summaryTitleTextMarginRight() {
            return this.bi;
        }
        set summaryTitleTextMarginRight(a) {
            this.bi = a;
            this.g("SummaryTitleTextMarginRight");
        }
        get summaryTitleTextMarginTop() {
            return this.bj;
        }
        set summaryTitleTextMarginTop(a) {
            this.bj = a;
            this.g("SummaryTitleTextMarginTop");
        }
        get summaryRowMarginBottom() {
            return this.bb;
        }
        set summaryRowMarginBottom(a) {
            this.bb = a;
            this.g("SummaryRowMarginBottom");
        }
        get summaryRowMarginLeft() {
            return this.bc;
        }
        set summaryRowMarginLeft(a) {
            this.bc = a;
            this.g("SummaryRowMarginLeft");
        }
        get summaryRowMarginRight() {
            return this.bd;
        }
        set summaryRowMarginRight(a) {
            this.bd = a;
            this.g("SummaryRowMarginRight");
        }
        get summaryRowMarginTop() {
            return this.be;
        }
        set summaryRowMarginTop(a) {
            this.be = a;
            this.g("SummaryRowMarginTop");
        }
        get summaryValueTextColor() {
            return this.e9;
        }
        set summaryValueTextColor(a) {
            this.e9 = a;
            this.g("SummaryValueTextColor");
        }
        get summaryValueTextFontFamily() {
            return this.fa;
        }
        set summaryValueTextFontFamily(a) {
            this.fa = a;
            this.g("SummaryValueTextFontFamily");
        }
        get summaryValueTextFontSize() {
            return this.bl;
        }
        set summaryValueTextFontSize(a) {
            this.bl = a;
            this.g("SummaryValueTextFontSize");
        }
        get summaryValueTextFontStyle() {
            return this.fb;
        }
        set summaryValueTextFontStyle(a) {
            this.fb = a;
            this.g("SummaryValueTextFontStyle");
        }
        get summaryValueTextFontWeight() {
            return this.fc;
        }
        set summaryValueTextFontWeight(a) {
            this.fc = a;
            this.g("SummaryValueTextFontWeight");
        }
        get summaryLabelText() {
            return this.et;
        }
        set summaryLabelText(a) {
            this.et = a;
            this.g("SummaryLabelText");
        }
        get summaryLabelTextColor() {
            return this.eu;
        }
        set summaryLabelTextColor(a) {
            this.eu = a;
            this.g("SummaryLabelTextColor");
        }
        get summaryLabelTextFontFamily() {
            return this.ev;
        }
        set summaryLabelTextFontFamily(a) {
            this.ev = a;
            this.g("SummaryLabelTextFontFamily");
        }
        get summaryLabelTextFontSize() {
            return this.ba;
        }
        set summaryLabelTextFontSize(a) {
            this.ba = a;
            this.g("SummaryLabelTextFontSize");
        }
        get summaryLabelTextFontStyle() {
            return this.ew;
        }
        set summaryLabelTextFontStyle(a) {
            this.ew = a;
            this.g("SummaryLabelTextFontStyle");
        }
        get summaryLabelTextFontWeight() {
            return this.ex;
        }
        set summaryLabelTextFontWeight(a) {
            this.ex = a;
            this.g("SummaryLabelTextFontWeight");
        }
        get summaryUnitsText() {
            return this.e4;
        }
        set summaryUnitsText(a) {
            this.e4 = a;
            this.g("SummaryUnitsText");
        }
        get summaryUnitsTextColor() {
            return this.e5;
        }
        set summaryUnitsTextColor(a) {
            this.e5 = a;
            this.g("SummaryUnitsTextColor");
        }
        get summaryUnitsTextFontFamily() {
            return this.e6;
        }
        set summaryUnitsTextFontFamily(a) {
            this.e6 = a;
            this.g("SummaryUnitsTextFontFamily");
        }
        get summaryUnitsTextFontSize() {
            return this.bk;
        }
        set summaryUnitsTextFontSize(a) {
            this.bk = a;
            this.g("SummaryUnitsTextFontSize");
        }
        get summaryUnitsTextFontStyle() {
            return this.e7;
        }
        set summaryUnitsTextFontStyle(a) {
            this.e7 = a;
            this.g("SummaryUnitsTextFontStyle");
        }
        get summaryUnitsTextFontWeight() {
            return this.e8;
        }
        set summaryUnitsTextFontWeight(a) {
            this.e8 = a;
            this.g("SummaryUnitsTextFontWeight");
        }
        get badgeMarginBottom() {
            return this.ai;
        }
        set badgeMarginBottom(a) {
            this.ai = a;
            this.g("BadgeMarginBottom");
        }
        get badgeMarginLeft() {
            return this.aj;
        }
        set badgeMarginLeft(a) {
            this.aj = a;
            this.g("BadgeMarginLeft");
        }
        get badgeMarginRight() {
            return this.ak;
        }
        set badgeMarginRight(a) {
            this.ak = a;
            this.g("BadgeMarginRight");
        }
        get badgeMarginTop() {
            return this.al;
        }
        set badgeMarginTop(a) {
            this.al = a;
            this.g("BadgeMarginTop");
        }
        get badgeShape() {
            return this.d2;
        }
        set badgeShape(a) {
            this.d2 = a;
            this.g("BadgeShape");
        }
        get targetCursorPositionX() {
            return this.bm;
        }
        set targetCursorPositionX(a) {
            this.bm = a;
            this.g("TargetCursorPositionX");
        }
        get targetCursorPositionY() {
            return this.bn;
        }
        set targetCursorPositionY(a) {
            this.bn = a;
            this.g("TargetCursorPositionY");
        }
        get targetRef() {
            return this.fd;
        }
        set targetRef(a) {
            this.fd = a;
            this.g("TargetRef");
        }
        get unitsDisplayMode() {
            return this.fi;
        }
        set unitsDisplayMode(a) {
            this.fi = a;
            this.g("UnitsDisplayMode");
        }
        get unitsText() {
            return this.fj;
        }
        set unitsText(a) {
            this.fj = a;
            this.g("UnitsText");
        }
        get unitsTextColor() {
            return this.fk;
        }
        set unitsTextColor(a) {
            this.fk = a;
            this.g("UnitsTextColor");
        }
        get unitsTextMarginBottom() {
            return this.bu;
        }
        set unitsTextMarginBottom(a) {
            this.bu = a;
            this.g("UnitsTextMarginBottom");
        }
        get unitsTextMarginLeft() {
            return this.bv;
        }
        set unitsTextMarginLeft(a) {
            this.bv = a;
            this.g("UnitsTextMarginLeft");
        }
        get unitsTextMarginRight() {
            return this.bw;
        }
        set unitsTextMarginRight(a) {
            this.bw = a;
            this.g("UnitsTextMarginRight");
        }
        get unitsTextMarginTop() {
            return this.bx;
        }
        set unitsTextMarginTop(a) {
            this.bx = a;
            this.g("UnitsTextMarginTop");
        }
        get unitsTextFontFamily() {
            return this.fl;
        }
        set unitsTextFontFamily(a) {
            this.fl = a;
            this.g("UnitsTextFontFamily");
        }
        get unitsTextFontSize() {
            return this.bt;
        }
        set unitsTextFontSize(a) {
            this.bt = a;
            this.g("UnitsTextFontSize");
        }
        get unitsTextFontStyle() {
            return this.fm;
        }
        set unitsTextFontStyle(a) {
            this.fm = a;
            this.g("UnitsTextFontStyle");
        }
        get unitsTextFontWeight() {
            return this.fn;
        }
        set unitsTextFontWeight(a) {
            this.fn = a;
            this.g("UnitsTextFontWeight");
        }
        get titleTextMarginBottom() {
            return this.bp;
        }
        set titleTextMarginBottom(a) {
            this.bp = a;
            this.g("TitleTextMarginBottom");
        }
        get titleTextMarginLeft() {
            return this.bq;
        }
        set titleTextMarginLeft(a) {
            this.bq = a;
            this.g("TitleTextMarginLeft");
        }
        get titleTextMarginRight() {
            return this.br;
        }
        set titleTextMarginRight(a) {
            this.br = a;
            this.g("TitleTextMarginRight");
        }
        get titleTextMarginTop() {
            return this.bs;
        }
        set titleTextMarginTop(a) {
            this.bs = a;
            this.g("TitleTextMarginTop");
        }
        get titleTextColor() {
            return this.fe;
        }
        set titleTextColor(a) {
            this.fe = a;
            this.g("TitleTextColor");
        }
        get titleTextFontFamily() {
            return this.ff;
        }
        set titleTextFontFamily(a) {
            this.ff = a;
            this.g("TitleTextFontFamily");
        }
        get titleTextFontSize() {
            return this.bo;
        }
        set titleTextFontSize(a) {
            this.bo = a;
            this.g("TitleTextFontSize");
        }
        get titleTextFontStyle() {
            return this.fg;
        }
        set titleTextFontStyle(a) {
            this.fg = a;
            this.g("TitleTextFontStyle");
        }
        get titleTextFontWeight() {
            return this.fh;
        }
        set titleTextFontWeight(a) {
            this.fh = a;
            this.g("TitleTextFontWeight");
        }
        get labelDisplayMode() {
            return this.ei;
        }
        set labelDisplayMode(a) {
            this.ei = a;
            this.g("LabelDisplayMode");
        }
        get labelTextColor() {
            return this.ej;
        }
        set labelTextColor(a) {
            this.ej = a;
            this.g("LabelTextColor");
        }
        get labelTextMarginBottom() {
            return this.a5;
        }
        set labelTextMarginBottom(a) {
            this.a5 = a;
            this.g("LabelTextMarginBottom");
        }
        get labelTextMarginLeft() {
            return this.a6;
        }
        set labelTextMarginLeft(a) {
            this.a6 = a;
            this.g("LabelTextMarginLeft");
        }
        get labelTextMarginRight() {
            return this.a7;
        }
        set labelTextMarginRight(a) {
            this.a7 = a;
            this.g("LabelTextMarginRight");
        }
        get labelTextMarginTop() {
            return this.a8;
        }
        set labelTextMarginTop(a) {
            this.a8 = a;
            this.g("LabelTextMarginTop");
        }
        get labelTextFontFamily() {
            return this.ek;
        }
        set labelTextFontFamily(a) {
            this.ek = a;
            this.g("LabelTextFontFamily");
        }
        get labelTextFontSize() {
            return this.a4;
        }
        set labelTextFontSize(a) {
            this.a4 = a;
            this.g("LabelTextFontSize");
        }
        get labelTextFontStyle() {
            return this.el;
        }
        set labelTextFontStyle(a) {
            this.el = a;
            this.g("LabelTextFontStyle");
        }
        get labelTextFontWeight() {
            return this.em;
        }
        set labelTextFontWeight(a) {
            this.em = a;
            this.g("LabelTextFontWeight");
        }
        get pixelScalingRatio() {
            return this.a9;
        }
        set pixelScalingRatio(a) {
            this.a9 = a;
            this.g("PixelScalingRatio");
        }
        get calculateColumnSummaryRef() {
            return this.d3;
        }
        set calculateColumnSummaryRef(a) {
            this.d3 = a;
            this.g("CalculateColumnSummaryRef");
        }
        get styleHeaderRowRef() {
            return this.eo;
        }
        set styleHeaderRowRef(a) {
            this.eo = a;
            this.g("StyleHeaderRowRef");
        }
        get styleGroupRowRef() {
            return this.en;
        }
        set styleGroupRowRef(a) {
            this.en = a;
            this.g("StyleGroupRowRef");
        }
        get styleSeriesRowRef() {
            return this.eq;
        }
        set styleSeriesRowRef(a) {
            this.eq = a;
            this.g("StyleSeriesRowRef");
        }
        get styleSummaryRowRef() {
            return this.es;
        }
        set styleSummaryRowRef(a) {
            this.es = a;
            this.g("StyleSummaryRowRef");
        }
        get styleSeriesColumnRef() {
            return this.ep;
        }
        set styleSeriesColumnRef(a) {
            this.ep = a;
            this.g("StyleSeriesColumnRef");
        }
        get styleSummaryColumnRef() {
            return this.er;
        }
        set styleSummaryColumnRef(a) {
            this.er = a;
            this.g("StyleSummaryColumnRef");
        }
        get width() {
            return this.fy;
        }
        set width(a) {
            this.fy = a;
            this.g("Width");
        }
        get height() {
            return this.eh;
        }
        set height(a) {
            this.eh = a;
            this.g("Height");
        }
    }
    DataLegendDescription.$t = markType(DataLegendDescription, 'DataLegendDescription', Description.$);
    return DataLegendDescription;
})();
