// igniteui-angular-core/esm2015/lib/DataToolTipLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataToolTipLayerDescription = /*@__PURE__*/ (() => {
    class DataToolTipLayerDescription extends AnnotationLayerDescription {
        constructor() {
            super();
            this.ls = null;
            this.gm = false;
            this.kp = null;
            this.kl = null;
            this.hq = 0;
            this.hr = 0;
            this.g1 = 0;
            this.g2 = 0;
            this.kn = null;
            this.kj = null;
            this.ko = null;
            this.kk = null;
            this.gl = false;
            this.gb = null;
            this.f8 = null;
            this.ga = null;
            this.f7 = null;
            this.l3 = null;
            this.kf = 0;
            this.kg = 0;
            this.l5 = null;
            this.l4 = null;
            this.gn = false;
            this.l6 = null;
            this.gc = null;
            this.ie = 0;
            this.ig = 0;
            this.ih = 0;
            this.ii = 0;
            this.go = false;
            this.mb = null;
            this.gp = false;
            this.ik = 0;
            this.il = 0;
            this.im = 0;
            this.io = 0;
            this.l7 = null;
            this.l8 = null;
            this.ij = 0;
            this.l9 = null;
            this.ma = null;
            this.kw = null;
            this.f9 = null;
            this.ku = null;
            this.kv = null;
            this.kx = null;
            this.ky = null;
            this.kz = null;
            this.hh = 0;
            this.hi = 0;
            this.hj = 0;
            this.hk = 0;
            this.hc = 0;
            this.hd = 0;
            this.he = 0;
            this.hf = 0;
            this.gk = false;
            this.k0 = null;
            this.hg = 0;
            this.k1 = null;
            this.k2 = null;
            this.gj = false;
            this.kq = null;
            this.g8 = 0;
            this.g9 = 0;
            this.ha = 0;
            this.hb = 0;
            this.g3 = 0;
            this.g4 = 0;
            this.g5 = 0;
            this.g6 = 0;
            this.kr = null;
            this.g7 = 0;
            this.ks = null;
            this.kt = null;
            this.le = null;
            this.lf = null;
            this.hx = 0;
            this.lg = null;
            this.lh = null;
            this.li = null;
            this.ld = null;
            this.hy = 0;
            this.hz = 0;
            this.h0 = 0;
            this.h1 = 0;
            this.ht = 0;
            this.hu = 0;
            this.hv = 0;
            this.hw = 0;
            this.lo = null;
            this.lp = null;
            this.h3 = 0;
            this.lq = null;
            this.lr = null;
            this.k8 = null;
            this.k9 = null;
            this.la = null;
            this.hs = 0;
            this.lb = null;
            this.lc = null;
            this.lj = null;
            this.lk = null;
            this.ll = null;
            this.h2 = 0;
            this.lm = null;
            this.ln = null;
            this.gx = 0;
            this.gy = 0;
            this.gz = 0;
            this.g0 = 0;
            this.km = null;
            this.lx = null;
            this.ly = null;
            this.lz = null;
            this.ia = 0;
            this.ib = 0;
            this.ic = 0;
            this.id = 0;
            this.l0 = null;
            this.h9 = 0;
            this.l1 = null;
            this.l2 = null;
            this.h5 = 0;
            this.h6 = 0;
            this.h7 = 0;
            this.h8 = 0;
            this.lt = null;
            this.lu = null;
            this.h4 = 0;
            this.lv = null;
            this.lw = null;
            this.k3 = null;
            this.k4 = null;
            this.hm = 0;
            this.hn = 0;
            this.ho = 0;
            this.hp = 0;
            this.k5 = null;
            this.hl = 0;
            this.k6 = null;
            this.k7 = null;
        }
        get_type() {
            return "DataToolTipLayer";
        }
        get targetAxisRef() {
            return this.ls;
        }
        set targetAxisRef(a) {
            this.ls = a;
            this.g("TargetAxisRef");
        }
        get useInterpolation() {
            return this.gm;
        }
        set useInterpolation(a) {
            this.gm = a;
            this.g("UseInterpolation");
        }
        get groupingMode() {
            return this.kp;
        }
        set groupingMode(a) {
            this.kp = a;
            this.g("GroupingMode");
        }
        get actualGroupingMode() {
            return this.kl;
        }
        set actualGroupingMode(a) {
            this.kl = a;
            this.g("ActualGroupingMode");
        }
        get positionOffsetX() {
            return this.hq;
        }
        set positionOffsetX(a) {
            this.hq = a;
            this.g("PositionOffsetX");
        }
        get positionOffsetY() {
            return this.hr;
        }
        set positionOffsetY(a) {
            this.hr = a;
            this.g("PositionOffsetY");
        }
        get defaultPositionOffsetX() {
            return this.g1;
        }
        set defaultPositionOffsetX(a) {
            this.g1 = a;
            this.g("DefaultPositionOffsetX");
        }
        get defaultPositionOffsetY() {
            return this.g2;
        }
        set defaultPositionOffsetY(a) {
            this.g2 = a;
            this.g("DefaultPositionOffsetY");
        }
        get groupedPositionModeX() {
            return this.kn;
        }
        set groupedPositionModeX(a) {
            this.kn = a;
            this.g("GroupedPositionModeX");
        }
        get actualGroupedPositionModeX() {
            return this.kj;
        }
        set actualGroupedPositionModeX(a) {
            this.kj = a;
            this.g("ActualGroupedPositionModeX");
        }
        get groupedPositionModeY() {
            return this.ko;
        }
        set groupedPositionModeY(a) {
            this.ko = a;
            this.g("GroupedPositionModeY");
        }
        get actualGroupedPositionModeY() {
            return this.kk;
        }
        set actualGroupedPositionModeY(a) {
            this.kk = a;
            this.g("ActualGroupedPositionModeY");
        }
        get shouldUpdateWhenSeriesDataChanges() {
            return this.gl;
        }
        set shouldUpdateWhenSeriesDataChanges(a) {
            this.gl = a;
            this.g("ShouldUpdateWhenSeriesDataChanges");
        }
        get includedSeries() {
            return this.gb;
        }
        set includedSeries(a) {
            this.gb = a;
            this.g("IncludedSeries");
        }
        get excludedSeries() {
            return this.f8;
        }
        set excludedSeries(a) {
            this.f8 = a;
            this.g("ExcludedSeries");
        }
        get includedColumns() {
            return this.ga;
        }
        set includedColumns(a) {
            this.ga = a;
            this.g("IncludedColumns");
        }
        get excludedColumns() {
            return this.f7;
        }
        set excludedColumns(a) {
            this.f7 = a;
            this.g("ExcludedColumns");
        }
        get valueFormatAbbreviation() {
            return this.l3;
        }
        set valueFormatAbbreviation(a) {
            this.l3 = a;
            this.g("ValueFormatAbbreviation");
        }
        get valueFormatMaxFractions() {
            return this.kf;
        }
        set valueFormatMaxFractions(a) {
            this.kf = a;
            this.g("ValueFormatMaxFractions");
        }
        get valueFormatMinFractions() {
            return this.kg;
        }
        set valueFormatMinFractions(a) {
            this.kg = a;
            this.g("ValueFormatMinFractions");
        }
        get valueFormatMode() {
            return this.l5;
        }
        set valueFormatMode(a) {
            this.l5 = a;
            this.g("ValueFormatMode");
        }
        get valueFormatCulture() {
            return this.l4;
        }
        set valueFormatCulture(a) {
            this.l4 = a;
            this.g("ValueFormatCulture");
        }
        get valueFormatUseGrouping() {
            return this.gn;
        }
        set valueFormatUseGrouping(a) {
            this.gn = a;
            this.g("ValueFormatUseGrouping");
        }
        get valueFormatString() {
            return this.l6;
        }
        set valueFormatString(a) {
            this.l6 = a;
            this.g("ValueFormatString");
        }
        get valueFormatSpecifiers() {
            return this.gc;
        }
        set valueFormatSpecifiers(a) {
            this.gc = a;
            this.g("ValueFormatSpecifiers");
        }
        get valueRowMarginBottom() {
            return this.ie;
        }
        set valueRowMarginBottom(a) {
            this.ie = a;
            this.g("ValueRowMarginBottom");
        }
        get valueRowMarginLeft() {
            return this.ig;
        }
        set valueRowMarginLeft(a) {
            this.ig = a;
            this.g("ValueRowMarginLeft");
        }
        get valueRowMarginRight() {
            return this.ih;
        }
        set valueRowMarginRight(a) {
            this.ih = a;
            this.g("ValueRowMarginRight");
        }
        get valueRowMarginTop() {
            return this.ii;
        }
        set valueRowMarginTop(a) {
            this.ii = a;
            this.g("ValueRowMarginTop");
        }
        get valueRowVisible() {
            return this.go;
        }
        set valueRowVisible(a) {
            this.go = a;
            this.g("ValueRowVisible");
        }
        get valueTextWhenMissingData() {
            return this.mb;
        }
        set valueTextWhenMissingData(a) {
            this.mb = a;
            this.g("ValueTextWhenMissingData");
        }
        get valueTextUseSeriesColors() {
            return this.gp;
        }
        set valueTextUseSeriesColors(a) {
            this.gp = a;
            this.g("ValueTextUseSeriesColors");
        }
        get valueTextMarginBottom() {
            return this.ik;
        }
        set valueTextMarginBottom(a) {
            this.ik = a;
            this.g("ValueTextMarginBottom");
        }
        get valueTextMarginLeft() {
            return this.il;
        }
        set valueTextMarginLeft(a) {
            this.il = a;
            this.g("ValueTextMarginLeft");
        }
        get valueTextMarginRight() {
            return this.im;
        }
        set valueTextMarginRight(a) {
            this.im = a;
            this.g("ValueTextMarginRight");
        }
        get valueTextMarginTop() {
            return this.io;
        }
        set valueTextMarginTop(a) {
            this.io = a;
            this.g("ValueTextMarginTop");
        }
        get valueTextColor() {
            return this.l7;
        }
        set valueTextColor(a) {
            this.l7 = a;
            this.g("ValueTextColor");
        }
        get valueTextFontFamily() {
            return this.l8;
        }
        set valueTextFontFamily(a) {
            this.l8 = a;
            this.g("ValueTextFontFamily");
        }
        get valueTextFontSize() {
            return this.ij;
        }
        set valueTextFontSize(a) {
            this.ij = a;
            this.g("ValueTextFontSize");
        }
        get valueTextFontStyle() {
            return this.l9;
        }
        set valueTextFontStyle(a) {
            this.l9 = a;
            this.g("ValueTextFontStyle");
        }
        get valueTextFontWeight() {
            return this.ma;
        }
        set valueTextFontWeight(a) {
            this.ma = a;
            this.g("ValueTextFontWeight");
        }
        get headerFormatString() {
            return this.kw;
        }
        set headerFormatString(a) {
            this.kw = a;
            this.g("HeaderFormatString");
        }
        get headerFormatSpecifiers() {
            return this.f9;
        }
        set headerFormatSpecifiers(a) {
            this.f9 = a;
            this.g("HeaderFormatSpecifiers");
        }
        get headerFormatCulture() {
            return this.ku;
        }
        set headerFormatCulture(a) {
            this.ku = a;
            this.g("HeaderFormatCulture");
        }
        get headerFormatDate() {
            return this.kv;
        }
        set headerFormatDate(a) {
            this.kv = a;
            this.g("HeaderFormatDate");
        }
        get headerFormatTime() {
            return this.kx;
        }
        set headerFormatTime(a) {
            this.kx = a;
            this.g("HeaderFormatTime");
        }
        get headerText() {
            return this.ky;
        }
        set headerText(a) {
            this.ky = a;
            this.g("HeaderText");
        }
        get headerTextColor() {
            return this.kz;
        }
        set headerTextColor(a) {
            this.kz = a;
            this.g("HeaderTextColor");
        }
        get headerTextMarginBottom() {
            return this.hh;
        }
        set headerTextMarginBottom(a) {
            this.hh = a;
            this.g("HeaderTextMarginBottom");
        }
        get headerTextMarginLeft() {
            return this.hi;
        }
        set headerTextMarginLeft(a) {
            this.hi = a;
            this.g("HeaderTextMarginLeft");
        }
        get headerTextMarginRight() {
            return this.hj;
        }
        set headerTextMarginRight(a) {
            this.hj = a;
            this.g("HeaderTextMarginRight");
        }
        get headerTextMarginTop() {
            return this.hk;
        }
        set headerTextMarginTop(a) {
            this.hk = a;
            this.g("HeaderTextMarginTop");
        }
        get headerRowMarginBottom() {
            return this.hc;
        }
        set headerRowMarginBottom(a) {
            this.hc = a;
            this.g("HeaderRowMarginBottom");
        }
        get headerRowMarginLeft() {
            return this.hd;
        }
        set headerRowMarginLeft(a) {
            this.hd = a;
            this.g("HeaderRowMarginLeft");
        }
        get headerRowMarginRight() {
            return this.he;
        }
        set headerRowMarginRight(a) {
            this.he = a;
            this.g("HeaderRowMarginRight");
        }
        get headerRowMarginTop() {
            return this.hf;
        }
        set headerRowMarginTop(a) {
            this.hf = a;
            this.g("HeaderRowMarginTop");
        }
        get headerRowVisible() {
            return this.gk;
        }
        set headerRowVisible(a) {
            this.gk = a;
            this.g("HeaderRowVisible");
        }
        get headerTextFontFamily() {
            return this.k0;
        }
        set headerTextFontFamily(a) {
            this.k0 = a;
            this.g("HeaderTextFontFamily");
        }
        get headerTextFontSize() {
            return this.hg;
        }
        set headerTextFontSize(a) {
            this.hg = a;
            this.g("HeaderTextFontSize");
        }
        get headerTextFontStyle() {
            return this.k1;
        }
        set headerTextFontStyle(a) {
            this.k1 = a;
            this.g("HeaderTextFontStyle");
        }
        get headerTextFontWeight() {
            return this.k2;
        }
        set headerTextFontWeight(a) {
            this.k2 = a;
            this.g("HeaderTextFontWeight");
        }
        get groupRowVisible() {
            return this.gj;
        }
        set groupRowVisible(a) {
            this.gj = a;
            this.g("GroupRowVisible");
        }
        get groupTextColor() {
            return this.kq;
        }
        set groupTextColor(a) {
            this.kq = a;
            this.g("GroupTextColor");
        }
        get groupTextMarginBottom() {
            return this.g8;
        }
        set groupTextMarginBottom(a) {
            this.g8 = a;
            this.g("GroupTextMarginBottom");
        }
        get groupTextMarginLeft() {
            return this.g9;
        }
        set groupTextMarginLeft(a) {
            this.g9 = a;
            this.g("GroupTextMarginLeft");
        }
        get groupTextMarginRight() {
            return this.ha;
        }
        set groupTextMarginRight(a) {
            this.ha = a;
            this.g("GroupTextMarginRight");
        }
        get groupTextMarginTop() {
            return this.hb;
        }
        set groupTextMarginTop(a) {
            this.hb = a;
            this.g("GroupTextMarginTop");
        }
        get groupRowMarginBottom() {
            return this.g3;
        }
        set groupRowMarginBottom(a) {
            this.g3 = a;
            this.g("GroupRowMarginBottom");
        }
        get groupRowMarginLeft() {
            return this.g4;
        }
        set groupRowMarginLeft(a) {
            this.g4 = a;
            this.g("GroupRowMarginLeft");
        }
        get groupRowMarginRight() {
            return this.g5;
        }
        set groupRowMarginRight(a) {
            this.g5 = a;
            this.g("GroupRowMarginRight");
        }
        get groupRowMarginTop() {
            return this.g6;
        }
        set groupRowMarginTop(a) {
            this.g6 = a;
            this.g("GroupRowMarginTop");
        }
        get groupTextFontFamily() {
            return this.kr;
        }
        set groupTextFontFamily(a) {
            this.kr = a;
            this.g("GroupTextFontFamily");
        }
        get groupTextFontSize() {
            return this.g7;
        }
        set groupTextFontSize(a) {
            this.g7 = a;
            this.g("GroupTextFontSize");
        }
        get groupTextFontStyle() {
            return this.ks;
        }
        set groupTextFontStyle(a) {
            this.ks = a;
            this.g("GroupTextFontStyle");
        }
        get groupTextFontWeight() {
            return this.kt;
        }
        set groupTextFontWeight(a) {
            this.kt = a;
            this.g("GroupTextFontWeight");
        }
        get summaryTitleTextColor() {
            return this.le;
        }
        set summaryTitleTextColor(a) {
            this.le = a;
            this.g("SummaryTitleTextColor");
        }
        get summaryTitleTextFontFamily() {
            return this.lf;
        }
        set summaryTitleTextFontFamily(a) {
            this.lf = a;
            this.g("SummaryTitleTextFontFamily");
        }
        get summaryTitleTextFontSize() {
            return this.hx;
        }
        set summaryTitleTextFontSize(a) {
            this.hx = a;
            this.g("SummaryTitleTextFontSize");
        }
        get summaryTitleTextFontStyle() {
            return this.lg;
        }
        set summaryTitleTextFontStyle(a) {
            this.lg = a;
            this.g("SummaryTitleTextFontStyle");
        }
        get summaryTitleTextFontWeight() {
            return this.lh;
        }
        set summaryTitleTextFontWeight(a) {
            this.lh = a;
            this.g("SummaryTitleTextFontWeight");
        }
        get summaryType() {
            return this.li;
        }
        set summaryType(a) {
            this.li = a;
            this.g("SummaryType");
        }
        get summaryTitleText() {
            return this.ld;
        }
        set summaryTitleText(a) {
            this.ld = a;
            this.g("SummaryTitleText");
        }
        get summaryTitleTextMarginBottom() {
            return this.hy;
        }
        set summaryTitleTextMarginBottom(a) {
            this.hy = a;
            this.g("SummaryTitleTextMarginBottom");
        }
        get summaryTitleTextMarginLeft() {
            return this.hz;
        }
        set summaryTitleTextMarginLeft(a) {
            this.hz = a;
            this.g("SummaryTitleTextMarginLeft");
        }
        get summaryTitleTextMarginRight() {
            return this.h0;
        }
        set summaryTitleTextMarginRight(a) {
            this.h0 = a;
            this.g("SummaryTitleTextMarginRight");
        }
        get summaryTitleTextMarginTop() {
            return this.h1;
        }
        set summaryTitleTextMarginTop(a) {
            this.h1 = a;
            this.g("SummaryTitleTextMarginTop");
        }
        get summaryRowMarginBottom() {
            return this.ht;
        }
        set summaryRowMarginBottom(a) {
            this.ht = a;
            this.g("SummaryRowMarginBottom");
        }
        get summaryRowMarginLeft() {
            return this.hu;
        }
        set summaryRowMarginLeft(a) {
            this.hu = a;
            this.g("SummaryRowMarginLeft");
        }
        get summaryRowMarginRight() {
            return this.hv;
        }
        set summaryRowMarginRight(a) {
            this.hv = a;
            this.g("SummaryRowMarginRight");
        }
        get summaryRowMarginTop() {
            return this.hw;
        }
        set summaryRowMarginTop(a) {
            this.hw = a;
            this.g("SummaryRowMarginTop");
        }
        get summaryValueTextColor() {
            return this.lo;
        }
        set summaryValueTextColor(a) {
            this.lo = a;
            this.g("SummaryValueTextColor");
        }
        get summaryValueTextFontFamily() {
            return this.lp;
        }
        set summaryValueTextFontFamily(a) {
            this.lp = a;
            this.g("SummaryValueTextFontFamily");
        }
        get summaryValueTextFontSize() {
            return this.h3;
        }
        set summaryValueTextFontSize(a) {
            this.h3 = a;
            this.g("SummaryValueTextFontSize");
        }
        get summaryValueTextFontStyle() {
            return this.lq;
        }
        set summaryValueTextFontStyle(a) {
            this.lq = a;
            this.g("SummaryValueTextFontStyle");
        }
        get summaryValueTextFontWeight() {
            return this.lr;
        }
        set summaryValueTextFontWeight(a) {
            this.lr = a;
            this.g("SummaryValueTextFontWeight");
        }
        get summaryLabelText() {
            return this.k8;
        }
        set summaryLabelText(a) {
            this.k8 = a;
            this.g("SummaryLabelText");
        }
        get summaryLabelTextColor() {
            return this.k9;
        }
        set summaryLabelTextColor(a) {
            this.k9 = a;
            this.g("SummaryLabelTextColor");
        }
        get summaryLabelTextFontFamily() {
            return this.la;
        }
        set summaryLabelTextFontFamily(a) {
            this.la = a;
            this.g("SummaryLabelTextFontFamily");
        }
        get summaryLabelTextFontSize() {
            return this.hs;
        }
        set summaryLabelTextFontSize(a) {
            this.hs = a;
            this.g("SummaryLabelTextFontSize");
        }
        get summaryLabelTextFontStyle() {
            return this.lb;
        }
        set summaryLabelTextFontStyle(a) {
            this.lb = a;
            this.g("SummaryLabelTextFontStyle");
        }
        get summaryLabelTextFontWeight() {
            return this.lc;
        }
        set summaryLabelTextFontWeight(a) {
            this.lc = a;
            this.g("SummaryLabelTextFontWeight");
        }
        get summaryUnitsText() {
            return this.lj;
        }
        set summaryUnitsText(a) {
            this.lj = a;
            this.g("SummaryUnitsText");
        }
        get summaryUnitsTextColor() {
            return this.lk;
        }
        set summaryUnitsTextColor(a) {
            this.lk = a;
            this.g("SummaryUnitsTextColor");
        }
        get summaryUnitsTextFontFamily() {
            return this.ll;
        }
        set summaryUnitsTextFontFamily(a) {
            this.ll = a;
            this.g("SummaryUnitsTextFontFamily");
        }
        get summaryUnitsTextFontSize() {
            return this.h2;
        }
        set summaryUnitsTextFontSize(a) {
            this.h2 = a;
            this.g("SummaryUnitsTextFontSize");
        }
        get summaryUnitsTextFontStyle() {
            return this.lm;
        }
        set summaryUnitsTextFontStyle(a) {
            this.lm = a;
            this.g("SummaryUnitsTextFontStyle");
        }
        get summaryUnitsTextFontWeight() {
            return this.ln;
        }
        set summaryUnitsTextFontWeight(a) {
            this.ln = a;
            this.g("SummaryUnitsTextFontWeight");
        }
        get badgeMarginBottom() {
            return this.gx;
        }
        set badgeMarginBottom(a) {
            this.gx = a;
            this.g("BadgeMarginBottom");
        }
        get badgeMarginLeft() {
            return this.gy;
        }
        set badgeMarginLeft(a) {
            this.gy = a;
            this.g("BadgeMarginLeft");
        }
        get badgeMarginRight() {
            return this.gz;
        }
        set badgeMarginRight(a) {
            this.gz = a;
            this.g("BadgeMarginRight");
        }
        get badgeMarginTop() {
            return this.g0;
        }
        set badgeMarginTop(a) {
            this.g0 = a;
            this.g("BadgeMarginTop");
        }
        get badgeShape() {
            return this.km;
        }
        set badgeShape(a) {
            this.km = a;
            this.g("BadgeShape");
        }
        get unitsDisplayMode() {
            return this.lx;
        }
        set unitsDisplayMode(a) {
            this.lx = a;
            this.g("UnitsDisplayMode");
        }
        get unitsText() {
            return this.ly;
        }
        set unitsText(a) {
            this.ly = a;
            this.g("UnitsText");
        }
        get unitsTextColor() {
            return this.lz;
        }
        set unitsTextColor(a) {
            this.lz = a;
            this.g("UnitsTextColor");
        }
        get unitsTextMarginBottom() {
            return this.ia;
        }
        set unitsTextMarginBottom(a) {
            this.ia = a;
            this.g("UnitsTextMarginBottom");
        }
        get unitsTextMarginLeft() {
            return this.ib;
        }
        set unitsTextMarginLeft(a) {
            this.ib = a;
            this.g("UnitsTextMarginLeft");
        }
        get unitsTextMarginRight() {
            return this.ic;
        }
        set unitsTextMarginRight(a) {
            this.ic = a;
            this.g("UnitsTextMarginRight");
        }
        get unitsTextMarginTop() {
            return this.id;
        }
        set unitsTextMarginTop(a) {
            this.id = a;
            this.g("UnitsTextMarginTop");
        }
        get unitsTextFontFamily() {
            return this.l0;
        }
        set unitsTextFontFamily(a) {
            this.l0 = a;
            this.g("UnitsTextFontFamily");
        }
        get unitsTextFontSize() {
            return this.h9;
        }
        set unitsTextFontSize(a) {
            this.h9 = a;
            this.g("UnitsTextFontSize");
        }
        get unitsTextFontStyle() {
            return this.l1;
        }
        set unitsTextFontStyle(a) {
            this.l1 = a;
            this.g("UnitsTextFontStyle");
        }
        get unitsTextFontWeight() {
            return this.l2;
        }
        set unitsTextFontWeight(a) {
            this.l2 = a;
            this.g("UnitsTextFontWeight");
        }
        get titleTextMarginBottom() {
            return this.h5;
        }
        set titleTextMarginBottom(a) {
            this.h5 = a;
            this.g("TitleTextMarginBottom");
        }
        get titleTextMarginLeft() {
            return this.h6;
        }
        set titleTextMarginLeft(a) {
            this.h6 = a;
            this.g("TitleTextMarginLeft");
        }
        get titleTextMarginRight() {
            return this.h7;
        }
        set titleTextMarginRight(a) {
            this.h7 = a;
            this.g("TitleTextMarginRight");
        }
        get titleTextMarginTop() {
            return this.h8;
        }
        set titleTextMarginTop(a) {
            this.h8 = a;
            this.g("TitleTextMarginTop");
        }
        get titleTextColor() {
            return this.lt;
        }
        set titleTextColor(a) {
            this.lt = a;
            this.g("TitleTextColor");
        }
        get titleTextFontFamily() {
            return this.lu;
        }
        set titleTextFontFamily(a) {
            this.lu = a;
            this.g("TitleTextFontFamily");
        }
        get titleTextFontSize() {
            return this.h4;
        }
        set titleTextFontSize(a) {
            this.h4 = a;
            this.g("TitleTextFontSize");
        }
        get titleTextFontStyle() {
            return this.lv;
        }
        set titleTextFontStyle(a) {
            this.lv = a;
            this.g("TitleTextFontStyle");
        }
        get titleTextFontWeight() {
            return this.lw;
        }
        set titleTextFontWeight(a) {
            this.lw = a;
            this.g("TitleTextFontWeight");
        }
        get labelDisplayMode() {
            return this.k3;
        }
        set labelDisplayMode(a) {
            this.k3 = a;
            this.g("LabelDisplayMode");
        }
        get labelTextColor() {
            return this.k4;
        }
        set labelTextColor(a) {
            this.k4 = a;
            this.g("LabelTextColor");
        }
        get labelTextMarginBottom() {
            return this.hm;
        }
        set labelTextMarginBottom(a) {
            this.hm = a;
            this.g("LabelTextMarginBottom");
        }
        get labelTextMarginLeft() {
            return this.hn;
        }
        set labelTextMarginLeft(a) {
            this.hn = a;
            this.g("LabelTextMarginLeft");
        }
        get labelTextMarginRight() {
            return this.ho;
        }
        set labelTextMarginRight(a) {
            this.ho = a;
            this.g("LabelTextMarginRight");
        }
        get labelTextMarginTop() {
            return this.hp;
        }
        set labelTextMarginTop(a) {
            this.hp = a;
            this.g("LabelTextMarginTop");
        }
        get labelTextFontFamily() {
            return this.k5;
        }
        set labelTextFontFamily(a) {
            this.k5 = a;
            this.g("LabelTextFontFamily");
        }
        get labelTextFontSize() {
            return this.hl;
        }
        set labelTextFontSize(a) {
            this.hl = a;
            this.g("LabelTextFontSize");
        }
        get labelTextFontStyle() {
            return this.k6;
        }
        set labelTextFontStyle(a) {
            this.k6 = a;
            this.g("LabelTextFontStyle");
        }
        get labelTextFontWeight() {
            return this.k7;
        }
        set labelTextFontWeight(a) {
            this.k7 = a;
            this.g("LabelTextFontWeight");
        }
    }
    DataToolTipLayerDescription.$t = markType(DataToolTipLayerDescription, 'DataToolTipLayerDescription', AnnotationLayerDescription.$);
    return DataToolTipLayerDescription;
})();
