// igniteui-angular-core/esm2015/lib/DataPieChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DataPieBaseChartDescription } from "./DataPieBaseChartDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let DataPieChartDescription = /*@__PURE__*/ (() => {
    class DataPieChartDescription extends DataPieBaseChartDescription {
        constructor() {
            super();
            this.x7 = 0;
            this.x8 = 0;
            this.yf = null;
            this.yi = null;
            this.yc = null;
            this.x5 = false;
            this.yg = null;
            this.yh = null;
            this.ye = null;
            this.yj = null;
            this.yd = null;
            this.yb = null;
        }
        get_type() {
            return "DataPieChart";
        }
        get transitionInDuration() {
            return this.x7;
        }
        set transitionInDuration(a) {
            this.x7 = a;
            this.g("TransitionInDuration");
        }
        get transitionOutDuration() {
            return this.x8;
        }
        set transitionOutDuration(a) {
            this.x8 = a;
            this.g("TransitionOutDuration");
        }
        get transitionInEasingFunctionRef() {
            return this.yf;
        }
        set transitionInEasingFunctionRef(a) {
            this.yf = a;
            this.g("TransitionInEasingFunctionRef");
        }
        get transitionOutEasingFunctionRef() {
            return this.yi;
        }
        set transitionOutEasingFunctionRef(a) {
            this.yi = a;
            this.g("TransitionOutEasingFunctionRef");
        }
        get chartType() {
            return this.yc;
        }
        set chartType(a) {
            this.yc = a;
            this.g("ChartType");
        }
        get isTransitionInEnabled() {
            return this.x5;
        }
        set isTransitionInEnabled(a) {
            this.x5 = a;
            this.g("IsTransitionInEnabled");
        }
        get transitionInMode() {
            return this.yg;
        }
        set transitionInMode(a) {
            this.yg = a;
            this.g("TransitionInMode");
        }
        get transitionInSpeedType() {
            return this.yh;
        }
        set transitionInSpeedType(a) {
            this.yh = a;
            this.g("TransitionInSpeedType");
        }
        get tooltipTemplateRef() {
            return this.ye;
        }
        set tooltipTemplateRef(a) {
            this.ye = a;
            this.g("TooltipTemplateRef");
        }
        get width() {
            return this.yj;
        }
        set width(a) {
            this.yj = a;
            this.g("Width");
        }
        get height() {
            return this.yd;
        }
        set height(a) {
            this.yd = a;
            this.g("Height");
        }
        get background() {
            return this.yb;
        }
        set background(a) {
            this.yb = a;
            this.g("Background");
        }
    }
    DataPieChartDescription.$t = markType(DataPieChartDescription, 'DataPieChartDescription', DataPieBaseChartDescription.$);
    return DataPieChartDescription;
})();
