// igniteui-angular-core/esm2015/lib/DomainChartSeriesPointerEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let DomainChartSeriesPointerEventArgsDescription = /*@__PURE__*/ (() => {
    class DomainChartSeriesPointerEventArgsDescription extends Description {
        constructor() {
            super();
            this.j = null;
            this.l = null;
            this.h = null;
        }
        get_type() {
            return "DomainChartSeriesPointerEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get series() {
            return this.j;
        }
        set series(a) {
            this.j = a;
            this.g("Series");
        }
        get itemRef() {
            return this.l;
        }
        set itemRef(a) {
            this.l = a;
            this.g("ItemRef");
        }
        get plotAreaPosition() {
            return this.h;
        }
        set plotAreaPosition(a) {
            this.h = a;
            this.g("PlotAreaPosition");
        }
    }
    DomainChartSeriesPointerEventArgsDescription.$t = markType(DomainChartSeriesPointerEventArgsDescription, 'DomainChartSeriesPointerEventArgsDescription', Description.$);
    return DomainChartSeriesPointerEventArgsDescription;
})();
