// igniteui-angular-core/esm2015/lib/FinancialSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let FinancialSeriesDescription = /*@__PURE__*/ (() => {
    class FinancialSeriesDescription extends SeriesDescription {
        constructor() {
            super();
            this.fe = null;
            this.fk = null;
            this.fl = null;
            this.ff = null;
            this.fc = null;
            this.fd = null;
            this.e6 = null;
            this.fj = null;
            this.fa = null;
            this.e8 = null;
            this.e9 = null;
            this.e7 = null;
            this.fb = null;
            this.e1 = false;
            this.fg = null;
            this.e2 = false;
            this.e5 = null;
            this.fh = null;
            this.fi = null;
        }
        get_type() {
            return "FinancialSeries";
        }
        get negativeBrush() {
            return this.fe;
        }
        set negativeBrush(a) {
            this.fe = a;
            this.g("NegativeBrush");
        }
        get xAxisRef() {
            return this.fk;
        }
        set xAxisRef(a) {
            this.fk = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.fl;
        }
        set yAxisRef(a) {
            this.fl = a;
            this.g("YAxisRef");
        }
        get openMemberPath() {
            return this.ff;
        }
        set openMemberPath(a) {
            this.ff = a;
            this.g("OpenMemberPath");
        }
        get highMemberPath() {
            return this.fc;
        }
        set highMemberPath(a) {
            this.fc = a;
            this.g("HighMemberPath");
        }
        get lowMemberPath() {
            return this.fd;
        }
        set lowMemberPath(a) {
            this.fd = a;
            this.g("LowMemberPath");
        }
        get closeMemberPath() {
            return this.e6;
        }
        set closeMemberPath(a) {
            this.e6 = a;
            this.g("CloseMemberPath");
        }
        get volumeMemberPath() {
            return this.fj;
        }
        set volumeMemberPath(a) {
            this.fj = a;
            this.g("VolumeMemberPath");
        }
        get highlightedOpenMemberPath() {
            return this.fa;
        }
        set highlightedOpenMemberPath(a) {
            this.fa = a;
            this.g("HighlightedOpenMemberPath");
        }
        get highlightedHighMemberPath() {
            return this.e8;
        }
        set highlightedHighMemberPath(a) {
            this.e8 = a;
            this.g("HighlightedHighMemberPath");
        }
        get highlightedLowMemberPath() {
            return this.e9;
        }
        set highlightedLowMemberPath(a) {
            this.e9 = a;
            this.g("HighlightedLowMemberPath");
        }
        get highlightedCloseMemberPath() {
            return this.e7;
        }
        set highlightedCloseMemberPath(a) {
            this.e7 = a;
            this.g("HighlightedCloseMemberPath");
        }
        get highlightedVolumeMemberPath() {
            return this.fb;
        }
        set highlightedVolumeMemberPath(a) {
            this.fb = a;
            this.g("HighlightedVolumeMemberPath");
        }
        get isCustomCategoryStyleAllowed() {
            return this.e1;
        }
        set isCustomCategoryStyleAllowed(a) {
            this.e1 = a;
            this.g("IsCustomCategoryStyleAllowed");
        }
        get transitionInMode() {
            return this.fg;
        }
        set transitionInMode(a) {
            this.fg = a;
            this.g("TransitionInMode");
        }
        get isTransitionInEnabled() {
            return this.e2;
        }
        set isTransitionInEnabled(a) {
            this.e2 = a;
            this.g("IsTransitionInEnabled");
        }
        get assigningCategoryStyleRef() {
            return this.e5;
        }
        set assigningCategoryStyleRef(a) {
            this.e5 = a;
            this.g("AssigningCategoryStyleRef");
        }
        get typicalRef() {
            return this.fh;
        }
        set typicalRef(a) {
            this.fh = a;
            this.g("TypicalRef");
        }
        get typicalBasedOnRef() {
            return this.fi;
        }
        set typicalBasedOnRef(a) {
            this.fi = a;
            this.g("TypicalBasedOnRef");
        }
    }
    FinancialSeriesDescription.$t = markType(FinancialSeriesDescription, 'FinancialSeriesDescription', SeriesDescription.$);
    return FinancialSeriesDescription;
})();
