// igniteui-angular-core/esm2015/lib/FinalValueLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let FinalValueLayerDescription = /*@__PURE__*/ (() => {
    class FinalValueLayerDescription extends AnnotationLayerDescription {
        constructor() {
            super();
            this.gq = null;
            this.gr = null;
            this.gp = null;
            this.go = null;
            this.gl = null;
            this.f7 = 0;
            this.gj = 0;
            this.gn = null;
            this.f9 = 0;
            this.gb = 0;
            this.ga = 0;
            this.f8 = 0;
            this.gc = 0;
            this.gm = null;
        }
        get_type() {
            return "FinalValueLayer";
        }
        get targetSeriesName() {
            return this.gq;
        }
        set targetSeriesName(a) {
            this.gq = a;
            this.g("TargetSeriesName");
        }
        get targetSeriesRef() {
            return this.gr;
        }
        set targetSeriesRef(a) {
            this.gr = a;
            this.g("TargetSeriesRef");
        }
        get finalValueSelectionMode() {
            return this.gp;
        }
        set finalValueSelectionMode(a) {
            this.gp = a;
            this.g("FinalValueSelectionMode");
        }
        get axisAnnotationTextColor() {
            return this.go;
        }
        set axisAnnotationTextColor(a) {
            this.go = a;
            this.g("AxisAnnotationTextColor");
        }
        get axisAnnotationBackground() {
            return this.gl;
        }
        set axisAnnotationBackground(a) {
            this.gl = a;
            this.g("AxisAnnotationBackground");
        }
        get axisAnnotationBackgroundCornerRadius() {
            return this.f7;
        }
        set axisAnnotationBackgroundCornerRadius(a) {
            this.f7 = a;
            this.g("AxisAnnotationBackgroundCornerRadius");
        }
        get axisAnnotationInterpolatedValuePrecision() {
            return this.gj;
        }
        set axisAnnotationInterpolatedValuePrecision(a) {
            this.gj = a;
            this.g("AxisAnnotationInterpolatedValuePrecision");
        }
        get axisAnnotationOutline() {
            return this.gn;
        }
        set axisAnnotationOutline(a) {
            this.gn = a;
            this.g("AxisAnnotationOutline");
        }
        get axisAnnotationPaddingLeft() {
            return this.f9;
        }
        set axisAnnotationPaddingLeft(a) {
            this.f9 = a;
            this.g("AxisAnnotationPaddingLeft");
        }
        get axisAnnotationPaddingTop() {
            return this.gb;
        }
        set axisAnnotationPaddingTop(a) {
            this.gb = a;
            this.g("AxisAnnotationPaddingTop");
        }
        get axisAnnotationPaddingRight() {
            return this.ga;
        }
        set axisAnnotationPaddingRight(a) {
            this.ga = a;
            this.g("AxisAnnotationPaddingRight");
        }
        get axisAnnotationPaddingBottom() {
            return this.f8;
        }
        set axisAnnotationPaddingBottom(a) {
            this.f8 = a;
            this.g("AxisAnnotationPaddingBottom");
        }
        get axisAnnotationStrokeThickness() {
            return this.gc;
        }
        set axisAnnotationStrokeThickness(a) {
            this.gc = a;
            this.g("AxisAnnotationStrokeThickness");
        }
        get axisAnnotationFormatLabelRef() {
            return this.gm;
        }
        set axisAnnotationFormatLabelRef(a) {
            this.gm = a;
            this.g("AxisAnnotationFormatLabelRef");
        }
    }
    FinalValueLayerDescription.$t = markType(FinalValueLayerDescription, 'FinalValueLayerDescription', AnnotationLayerDescription.$);
    return FinalValueLayerDescription;
})();
