// igniteui-angular-core/esm2015/lib/FinancialPriceSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FinancialSeriesDescription } from "./FinancialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let FinancialPriceSeriesDescription = /*@__PURE__*/ (() => {
    class FinancialPriceSeriesDescription extends FinancialSeriesDescription {
        constructor() {
            super();
            this.gl = null;
            this.gk = null;
            this.f9 = null;
            this.f5 = 0;
            this.f3 = null;
            this.f7 = 0;
            this.gc = null;
            this.gh = null;
            this.gf = null;
            this.gd = null;
            this.ga = null;
            this.gi = null;
            this.gg = null;
            this.ge = null;
            this.gb = null;
            this.gj = null;
        }
        get_type() {
            return "FinancialPriceSeries";
        }
        get trendLineType() {
            return this.gl;
        }
        set trendLineType(a) {
            this.gl = a;
            this.g("TrendLineType");
        }
        get trendLineBrush() {
            return this.gk;
        }
        set trendLineBrush(a) {
            this.gk = a;
            this.g("TrendLineBrush");
        }
        get actualTrendLineBrush() {
            return this.f9;
        }
        set actualTrendLineBrush(a) {
            this.f9 = a;
            this.g("ActualTrendLineBrush");
        }
        get trendLineThickness() {
            return this.f5;
        }
        set trendLineThickness(a) {
            this.f5 = a;
            this.g("TrendLineThickness");
        }
        get trendLineDashArray() {
            return this.f3;
        }
        set trendLineDashArray(a) {
            this.f3 = a;
            this.g("TrendLineDashArray");
        }
        get trendLinePeriod() {
            return this.f7;
        }
        set trendLinePeriod(a) {
            this.f7 = a;
            this.g("TrendLinePeriod");
        }
        get displayType() {
            return this.gc;
        }
        set displayType(a) {
            this.gc = a;
            this.g("DisplayType");
        }
        get negativeOutline() {
            return this.gh;
        }
        set negativeOutline(a) {
            this.gh = a;
            this.g("NegativeOutline");
        }
        get lowMemberAsLegendLabel() {
            return this.gf;
        }
        set lowMemberAsLegendLabel(a) {
            this.gf = a;
            this.g("LowMemberAsLegendLabel");
        }
        get highMemberAsLegendLabel() {
            return this.gd;
        }
        set highMemberAsLegendLabel(a) {
            this.gd = a;
            this.g("HighMemberAsLegendLabel");
        }
        get closeMemberAsLegendLabel() {
            return this.ga;
        }
        set closeMemberAsLegendLabel(a) {
            this.ga = a;
            this.g("CloseMemberAsLegendLabel");
        }
        get openMemberAsLegendLabel() {
            return this.gi;
        }
        set openMemberAsLegendLabel(a) {
            this.gi = a;
            this.g("OpenMemberAsLegendLabel");
        }
        get lowMemberAsLegendUnit() {
            return this.gg;
        }
        set lowMemberAsLegendUnit(a) {
            this.gg = a;
            this.g("LowMemberAsLegendUnit");
        }
        get highMemberAsLegendUnit() {
            return this.ge;
        }
        set highMemberAsLegendUnit(a) {
            this.ge = a;
            this.g("HighMemberAsLegendUnit");
        }
        get closeMemberAsLegendUnit() {
            return this.gb;
        }
        set closeMemberAsLegendUnit(a) {
            this.gb = a;
            this.g("CloseMemberAsLegendUnit");
        }
        get openMemberAsLegendUnit() {
            return this.gj;
        }
        set openMemberAsLegendUnit(a) {
            this.gj = a;
            this.g("OpenMemberAsLegendUnit");
        }
    }
    FinancialPriceSeriesDescription.$t = markType(FinancialPriceSeriesDescription, 'FinancialPriceSeriesDescription', FinancialSeriesDescription.$);
    return FinancialPriceSeriesDescription;
})();
