// igniteui-angular-core/esm2015/lib/GeographicSymbolSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMarkerSeriesDescription } from "./GeographicMarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicSymbolSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicSymbolSeriesDescription extends GeographicMarkerSeriesDescription {
        constructor() {
            super();
            this.f0 = null;
            this.f1 = null;
            this.f2 = null;
            this.fs = false;
            this.fr = false;
            this.fz = null;
            this.fv = 0;
            this.fy = null;
            this.fx = null;
        }
        get_type() {
            return "GeographicSymbolSeries";
        }
        get latitudeMemberPath() {
            return this.f0;
        }
        set latitudeMemberPath(a) {
            this.f0 = a;
            this.g("LatitudeMemberPath");
        }
        get longitudeMemberPath() {
            return this.f1;
        }
        set longitudeMemberPath(a) {
            this.f1 = a;
            this.g("LongitudeMemberPath");
        }
        get markerCollisionAvoidance() {
            return this.f2;
        }
        set markerCollisionAvoidance(a) {
            this.f2 = a;
            this.g("MarkerCollisionAvoidance");
        }
        get isCustomScatterStyleAllowed() {
            return this.fs;
        }
        set isCustomScatterStyleAllowed(a) {
            this.fs = a;
            this.g("IsCustomScatterStyleAllowed");
        }
        get isCustomScatterMarkerStyleAllowed() {
            return this.fr;
        }
        set isCustomScatterMarkerStyleAllowed(a) {
            this.fr = a;
            this.g("IsCustomScatterMarkerStyleAllowed");
        }
        get itemSearchMode() {
            return this.fz;
        }
        set itemSearchMode(a) {
            this.fz = a;
            this.g("ItemSearchMode");
        }
        get itemSearchThreshold() {
            return this.fv;
        }
        set itemSearchThreshold(a) {
            this.fv = a;
            this.g("ItemSearchThreshold");
        }
        get assigningScatterStyleRef() {
            return this.fy;
        }
        set assigningScatterStyleRef(a) {
            this.fy = a;
            this.g("AssigningScatterStyleRef");
        }
        get assigningScatterMarkerStyleRef() {
            return this.fx;
        }
        set assigningScatterMarkerStyleRef(a) {
            this.fx = a;
            this.g("AssigningScatterMarkerStyleRef");
        }
    }
    GeographicSymbolSeriesDescription.$t = markType(GeographicSymbolSeriesDescription, 'GeographicSymbolSeriesDescription', GeographicMarkerSeriesDescription.$);
    return GeographicSymbolSeriesDescription;
})();
