// igniteui-angular-core/esm2015/lib/GeographicContourLineSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicXYTriangulatingSeriesDescription } from "./GeographicXYTriangulatingSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicContourLineSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicContourLineSeriesDescription extends GeographicXYTriangulatingSeriesDescription {
        constructor() {
            super();
            this.fo = null;
            this.fk = null;
            this.fj = null;
            this.fh = null;
            this.fn = null;
        }
        get_type() {
            return "GeographicContourLineSeries";
        }
        get valueMemberPath() {
            return this.fo;
        }
        set valueMemberPath(a) {
            this.fo = a;
            this.g("ValueMemberPath");
        }
        get fillScale() {
            return this.fk;
        }
        set fillScale(a) {
            this.fk = a;
            this.g("FillScale");
        }
        get actualFillScale() {
            return this.fj;
        }
        set actualFillScale(a) {
            this.fj = a;
            this.g("ActualFillScale");
        }
        get valueResolver() {
            return this.fh;
        }
        set valueResolver(a) {
            this.fh = a;
            this.g("ValueResolver");
        }
        get triangulationStatusChangedRef() {
            return this.fn;
        }
        set triangulationStatusChangedRef(a) {
            this.fn = a;
            this.g("TriangulationStatusChangedRef");
        }
    }
    GeographicContourLineSeriesDescription.$t = markType(GeographicContourLineSeriesDescription, 'GeographicContourLineSeriesDescription', GeographicXYTriangulatingSeriesDescription.$);
    return GeographicContourLineSeriesDescription;
})();
