// igniteui-angular-core/esm2015/lib/GeographicShapeSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicShapeSeriesBaseDescription = /*@__PURE__*/ (() => {
    class GeographicShapeSeriesBaseDescription extends GeographicMapSeriesHostDescription {
        constructor() {
            super();
            this.fk = null;
            this.e8 = false;
            this.e7 = false;
            this.fj = null;
            this.fe = 0;
            this.fd = 0;
            this.e5 = null;
            this.fb = 0;
            this.fi = null;
            this.fh = null;
        }
        get_type() {
            return "GeographicShapeSeriesBase";
        }
        get shapeMemberPath() {
            return this.fk;
        }
        set shapeMemberPath(a) {
            this.fk = a;
            this.g("ShapeMemberPath");
        }
        get isCustomShapeStyleAllowed() {
            return this.e8;
        }
        set isCustomShapeStyleAllowed(a) {
            this.e8 = a;
            this.g("IsCustomShapeStyleAllowed");
        }
        get isCustomShapeMarkerStyleAllowed() {
            return this.e7;
        }
        set isCustomShapeMarkerStyleAllowed(a) {
            this.e7 = a;
            this.g("IsCustomShapeMarkerStyleAllowed");
        }
        get itemSearchMode() {
            return this.fj;
        }
        set itemSearchMode(a) {
            this.fj = a;
            this.g("ItemSearchMode");
        }
        get itemSearchThreshold() {
            return this.fe;
        }
        set itemSearchThreshold(a) {
            this.fe = a;
            this.g("ItemSearchThreshold");
        }
        get itemSearchPointsThreshold() {
            return this.fd;
        }
        set itemSearchPointsThreshold(a) {
            this.fd = a;
            this.g("ItemSearchPointsThreshold");
        }
        get shapefileDataSource() {
            return this.e5;
        }
        set shapefileDataSource(a) {
            this.e5 = a;
            this.g("ShapefileDataSource");
        }
        get shapeFilterResolution() {
            return this.fb;
        }
        set shapeFilterResolution(a) {
            this.fb = a;
            this.g("ShapeFilterResolution");
        }
        get assigningShapeStyleRef() {
            return this.fi;
        }
        set assigningShapeStyleRef(a) {
            this.fi = a;
            this.g("AssigningShapeStyleRef");
        }
        get assigningShapeMarkerStyleRef() {
            return this.fh;
        }
        set assigningShapeMarkerStyleRef(a) {
            this.fh = a;
            this.g("AssigningShapeMarkerStyleRef");
        }
    }
    GeographicShapeSeriesBaseDescription.$t = markType(GeographicShapeSeriesBaseDescription, 'GeographicShapeSeriesBaseDescription', GeographicMapSeriesHostDescription.$);
    return GeographicShapeSeriesBaseDescription;
})();
