// igniteui-angular-core/esm2015/lib/GeographicShapeSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicShapeSeriesBaseDescription } from "./GeographicShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicShapeSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicShapeSeriesDescription extends GeographicShapeSeriesBaseDescription {
        constructor() {
            super();
            this.f7 = null;
            this.f5 = null;
            this.f6 = null;
            this.fr = 0;
            this.fq = 0;
            this.f4 = null;
            this.f2 = null;
            this.f0 = null;
            this.f3 = null;
            this.fy = null;
            this.f1 = null;
            this.fp = 0;
            this.fz = null;
            this.fx = null;
            this.fv = null;
            this.fw = null;
            this.f8 = null;
        }
        get_type() {
            return "GeographicShapeSeries";
        }
        get shapeStyleSelectorRef() {
            return this.f7;
        }
        set shapeStyleSelectorRef(a) {
            this.f7 = a;
            this.g("ShapeStyleSelectorRef");
        }
        get shapeFill() {
            return this.f5;
        }
        set shapeFill(a) {
            this.f5 = a;
            this.g("ShapeFill");
        }
        get shapeStroke() {
            return this.f6;
        }
        set shapeStroke(a) {
            this.f6 = a;
            this.g("ShapeStroke");
        }
        get shapeStrokeThickness() {
            return this.fr;
        }
        set shapeStrokeThickness(a) {
            this.fr = a;
            this.g("ShapeStrokeThickness");
        }
        get shapeOpacity() {
            return this.fq;
        }
        set shapeOpacity(a) {
            this.fq = a;
            this.g("ShapeOpacity");
        }
        get markerType() {
            return this.f4;
        }
        set markerType(a) {
            this.f4 = a;
            this.g("MarkerType");
        }
        get markerOutlineMode() {
            return this.f2;
        }
        set markerOutlineMode(a) {
            this.f2 = a;
            this.g("MarkerOutlineMode");
        }
        get markerFillMode() {
            return this.f0;
        }
        set markerFillMode(a) {
            this.f0 = a;
            this.g("MarkerFillMode");
        }
        get markerTemplateRef() {
            return this.f3;
        }
        set markerTemplateRef(a) {
            this.f3 = a;
            this.g("MarkerTemplateRef");
        }
        get markerBrush() {
            return this.fy;
        }
        set markerBrush(a) {
            this.fy = a;
            this.g("MarkerBrush");
        }
        get markerOutline() {
            return this.f1;
        }
        set markerOutline(a) {
            this.f1 = a;
            this.g("MarkerOutline");
        }
        get markerThickness() {
            return this.fp;
        }
        set markerThickness(a) {
            this.fp = a;
            this.g("MarkerThickness");
        }
        get markerCollisionAvoidance() {
            return this.fz;
        }
        set markerCollisionAvoidance(a) {
            this.fz = a;
            this.g("MarkerCollisionAvoidance");
        }
        get actualMarkerTemplateRef() {
            return this.fx;
        }
        set actualMarkerTemplateRef(a) {
            this.fx = a;
            this.g("ActualMarkerTemplateRef");
        }
        get actualMarkerBrush() {
            return this.fv;
        }
        set actualMarkerBrush(a) {
            this.fv = a;
            this.g("ActualMarkerBrush");
        }
        get actualMarkerOutline() {
            return this.fw;
        }
        set actualMarkerOutline(a) {
            this.fw = a;
            this.g("ActualMarkerOutline");
        }
        get styleShapeRef() {
            return this.f8;
        }
        set styleShapeRef(a) {
            this.f8 = a;
            this.g("StyleShapeRef");
        }
    }
    GeographicShapeSeriesDescription.$t = markType(GeographicShapeSeriesDescription, 'GeographicShapeSeriesDescription', GeographicShapeSeriesBaseDescription.$);
    return GeographicShapeSeriesDescription;
})();
