// igniteui-angular-core/esm2015/lib/GeographicPolylineSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicShapeSeriesBaseDescription } from "./GeographicShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicPolylineSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicPolylineSeriesDescription extends GeographicShapeSeriesBaseDescription {
        constructor() {
            super();
            this.fv = null;
            this.ft = null;
            this.fu = null;
            this.fq = 0;
            this.fp = 0;
            this.fw = null;
        }
        get_type() {
            return "GeographicPolylineSeries";
        }
        get shapeStyleSelectorRef() {
            return this.fv;
        }
        set shapeStyleSelectorRef(a) {
            this.fv = a;
            this.g("ShapeStyleSelectorRef");
        }
        get shapeFill() {
            return this.ft;
        }
        set shapeFill(a) {
            this.ft = a;
            this.g("ShapeFill");
        }
        get shapeStroke() {
            return this.fu;
        }
        set shapeStroke(a) {
            this.fu = a;
            this.g("ShapeStroke");
        }
        get shapeStrokeThickness() {
            return this.fq;
        }
        set shapeStrokeThickness(a) {
            this.fq = a;
            this.g("ShapeStrokeThickness");
        }
        get shapeOpacity() {
            return this.fp;
        }
        set shapeOpacity(a) {
            this.fp = a;
            this.g("ShapeOpacity");
        }
        get styleShapeRef() {
            return this.fw;
        }
        set styleShapeRef(a) {
            this.fw = a;
            this.g("StyleShapeRef");
        }
    }
    GeographicPolylineSeriesDescription.$t = markType(GeographicPolylineSeriesDescription, 'GeographicPolylineSeriesDescription', GeographicShapeSeriesBaseDescription.$);
    return GeographicPolylineSeriesDescription;
})();
