// igniteui-angular-core/esm2015/lib/GeographicTileSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicShapeSeriesBaseDescription } from "./GeographicShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicTileSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicTileSeriesDescription extends GeographicShapeSeriesBaseDescription {
        constructor() {
            super();
            this.fp = null;
            this.fr = null;
        }
        get_type() {
            return "GeographicTileSeries";
        }
        get tileImagery() {
            return this.fp;
        }
        set tileImagery(a) {
            this.fp = a;
            this.g("TileImagery");
        }
        get imageTilesReadyRef() {
            return this.fr;
        }
        set imageTilesReadyRef(a) {
            this.fr = a;
            this.g("ImageTilesReadyRef");
        }
    }
    GeographicTileSeriesDescription.$t = markType(GeographicTileSeriesDescription, 'GeographicTileSeriesDescription', GeographicShapeSeriesBaseDescription.$);
    return GeographicTileSeriesDescription;
})();
