// igniteui-angular-core/esm2015/lib/GeographicMapImageryDescription_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
import { SeriesViewerDescription } from "./SeriesViewerDescription";
/**
 * @hidden
 */
export let GeographicMapImageryDescription = /*@__PURE__*/ (() => {
    class GeographicMapImageryDescription extends Description {
        constructor() {
            super();
            this.j = null;
            this.h = null;
            this.t = null;
            this.l = false;
            this.u = null;
            this.n = 0;
            this.s = null;
            this.r = null;
            this.p = null;
            this.q = null;
        }
        get_type() {
            return "GeographicMapImagery";
        }
        get type() {
            return this.get_type();
        }
        get windowRect() {
            return this.j;
        }
        set windowRect(a) {
            this.j = a;
            this.g("WindowRect");
        }
        get geographicMap() {
            return this.h;
        }
        set geographicMap(a) {
            this.h = a;
            this.g("GeographicMap");
        }
        get referer() {
            return this.t;
        }
        set referer(a) {
            this.t = a;
            this.g("Referer");
        }
        get isHorizontalWrappingEnabled() {
            return this.l;
        }
        set isHorizontalWrappingEnabled(a) {
            this.l = a;
            this.g("IsHorizontalWrappingEnabled");
        }
        get userAgent() {
            return this.u;
        }
        set userAgent(a) {
            this.u = a;
            this.g("UserAgent");
        }
        get opacity() {
            return this.n;
        }
        set opacity(a) {
            this.n = a;
            this.g("Opacity");
        }
        get imageTilesReadyRef() {
            return this.s;
        }
        set imageTilesReadyRef(a) {
            this.s = a;
            this.g("ImageTilesReadyRef");
        }
        get imagesChangedRef() {
            return this.r;
        }
        set imagesChangedRef(a) {
            this.r = a;
            this.g("ImagesChangedRef");
        }
        get cancellingImageRef() {
            return this.p;
        }
        set cancellingImageRef(a) {
            this.p = a;
            this.g("CancellingImageRef");
        }
        get downloadingImageRef() {
            return this.q;
        }
        set downloadingImageRef(a) {
            this.q = a;
            this.g("DownloadingImageRef");
        }
    }
    GeographicMapImageryDescription.$t = markType(GeographicMapImageryDescription, 'GeographicMapImageryDescription', Description.$);
    return GeographicMapImageryDescription;
})();
/**
 * @hidden
 */
export let GeographicMapDescription = /*@__PURE__*/ (() => {
    class GeographicMapDescription extends SeriesViewerDescription {
        constructor() {
            super();
            this.h6 = false;
            this.ij = null;
            this.h3 = false;
            this.ig = null;
            this.h5 = false;
            this.h4 = false;
            this.h0 = null;
            this.hz = null;
            this.hx = null;
            this.ic = 0;
            this.ib = 0;
            this.ii = null;
            this.ik = null;
            this.ih = null;
        }
        get_type() {
            return "GeographicMap";
        }
        get zoomable() {
            return this.h6;
        }
        set zoomable(a) {
            this.h6 = a;
            this.g("Zoomable");
        }
        get resizeBehavior() {
            return this.ij;
        }
        set resizeBehavior(a) {
            this.ij = a;
            this.g("ResizeBehavior");
        }
        get isHorizontalWrappingEnabled() {
            return this.h3;
        }
        set isHorizontalWrappingEnabled(a) {
            this.h3 = a;
            this.g("IsHorizontalWrappingEnabled");
        }
        get backgroundTilingMode() {
            return this.ig;
        }
        set backgroundTilingMode(a) {
            this.ig = a;
            this.g("BackgroundTilingMode");
        }
        get useWorldRectForZoomBounds() {
            return this.h5;
        }
        set useWorldRectForZoomBounds(a) {
            this.h5 = a;
            this.g("UseWorldRectForZoomBounds");
        }
        get suppressZoomResetOnWorldRectChange() {
            return this.h4;
        }
        set suppressZoomResetOnWorldRectChange(a) {
            this.h4 = a;
            this.g("SuppressZoomResetOnWorldRectChange");
        }
        get worldRect() {
            return this.h0;
        }
        set worldRect(a) {
            this.h0 = a;
            this.g("WorldRect");
        }
        get actualWorldRect() {
            return this.hz;
        }
        set actualWorldRect(a) {
            this.hz = a;
            this.g("ActualWorldRect");
        }
        get backgroundContent() {
            return this.hx;
        }
        set backgroundContent(a) {
            this.hx = a;
            this.g("BackgroundContent");
        }
        get windowScale() {
            return this.ic;
        }
        set windowScale(a) {
            this.ic = a;
            this.g("WindowScale");
        }
        get actualWindowScale() {
            return this.ib;
        }
        set actualWindowScale(a) {
            this.ib = a;
            this.g("ActualWindowScale");
        }
        get imageTilesReadyRef() {
            return this.ii;
        }
        set imageTilesReadyRef(a) {
            this.ii = a;
            this.g("ImageTilesReadyRef");
        }
        get width() {
            return this.ik;
        }
        set width(a) {
            this.ik = a;
            this.g("Width");
        }
        get height() {
            return this.ih;
        }
        set height(a) {
            this.ih = a;
            this.g("Height");
        }
    }
    GeographicMapDescription.$t = markType(GeographicMapDescription, 'GeographicMapDescription', SeriesViewerDescription.$);
    return GeographicMapDescription;
})();
