// igniteui-angular-core/esm2015/lib/GeographicHighDensityScatterSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicHighDensityScatterSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicHighDensityScatterSeriesDescription extends GeographicMapSeriesHostDescription {
        constructor() {
            super();
            this.fj = null;
            this.fk = null;
            this.e6 = false;
            this.e5 = false;
            this.fa = 0;
            this.e9 = 0;
            this.fi = null;
            this.fh = null;
            this.fd = 0;
            this.fe = 0;
            this.fl = null;
        }
        get_type() {
            return "GeographicHighDensityScatterSeries";
        }
        get latitudeMemberPath() {
            return this.fj;
        }
        set latitudeMemberPath(a) {
            this.fj = a;
            this.g("LatitudeMemberPath");
        }
        get longitudeMemberPath() {
            return this.fk;
        }
        set longitudeMemberPath(a) {
            this.fk = a;
            this.g("LongitudeMemberPath");
        }
        get useBruteForce() {
            return this.e6;
        }
        set useBruteForce(a) {
            this.e6 = a;
            this.g("UseBruteForce");
        }
        get progressiveLoad() {
            return this.e5;
        }
        set progressiveLoad(a) {
            this.e5 = a;
            this.g("ProgressiveLoad");
        }
        get heatMinimum() {
            return this.fa;
        }
        set heatMinimum(a) {
            this.fa = a;
            this.g("HeatMinimum");
        }
        get heatMaximum() {
            return this.e9;
        }
        set heatMaximum(a) {
            this.e9 = a;
            this.g("HeatMaximum");
        }
        get heatMinimumColor() {
            return this.fi;
        }
        set heatMinimumColor(a) {
            this.fi = a;
            this.g("HeatMinimumColor");
        }
        get heatMaximumColor() {
            return this.fh;
        }
        set heatMaximumColor(a) {
            this.fh = a;
            this.g("HeatMaximumColor");
        }
        get pointExtent() {
            return this.fd;
        }
        set pointExtent(a) {
            this.fd = a;
            this.g("PointExtent");
        }
        get progressiveStatus() {
            return this.fe;
        }
        set progressiveStatus(a) {
            this.fe = a;
            this.g("ProgressiveStatus");
        }
        get progressiveLoadStatusChangedRef() {
            return this.fl;
        }
        set progressiveLoadStatusChangedRef(a) {
            this.fl = a;
            this.g("ProgressiveLoadStatusChangedRef");
        }
    }
    GeographicHighDensityScatterSeriesDescription.$t = markType(GeographicHighDensityScatterSeriesDescription, 'GeographicHighDensityScatterSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicHighDensityScatterSeriesDescription;
})();
