// igniteui-angular-core/esm2015/lib/GeographicXYTriangulatingSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicXYTriangulatingSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicXYTriangulatingSeriesDescription extends GeographicMapSeriesHostDescription {
        constructor() {
            super();
            this.e6 = null;
            this.e5 = null;
            this.e7 = null;
            this.e8 = null;
            this.e9 = null;
            this.fa = null;
        }
        get_type() {
            return "GeographicXYTriangulatingSeries";
        }
        get longitudeMemberPath() {
            return this.e6;
        }
        set longitudeMemberPath(a) {
            this.e6 = a;
            this.g("LongitudeMemberPath");
        }
        get latitudeMemberPath() {
            return this.e5;
        }
        set latitudeMemberPath(a) {
            this.e5 = a;
            this.g("LatitudeMemberPath");
        }
        get trianglesSourceRef() {
            return this.e7;
        }
        set trianglesSourceRef(a) {
            this.e7 = a;
            this.g("TrianglesSourceRef");
        }
        get triangleVertexMemberPath1() {
            return this.e8;
        }
        set triangleVertexMemberPath1(a) {
            this.e8 = a;
            this.g("TriangleVertexMemberPath1");
        }
        get triangleVertexMemberPath2() {
            return this.e9;
        }
        set triangleVertexMemberPath2(a) {
            this.e9 = a;
            this.g("TriangleVertexMemberPath2");
        }
        get triangleVertexMemberPath3() {
            return this.fa;
        }
        set triangleVertexMemberPath3(a) {
            this.fa = a;
            this.g("TriangleVertexMemberPath3");
        }
    }
    GeographicXYTriangulatingSeriesDescription.$t = markType(GeographicXYTriangulatingSeriesDescription, 'GeographicXYTriangulatingSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicXYTriangulatingSeriesDescription;
})();
