// igniteui-angular-core/esm2015/lib/GeographicScatterAreaSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { GeographicXYTriangulatingSeriesDescription } from "./GeographicXYTriangulatingSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let GeographicScatterAreaSeriesDescription = /*@__PURE__*/ (() => {
    class GeographicScatterAreaSeriesDescription extends GeographicXYTriangulatingSeriesDescription {
        constructor() {
            super();
            this.fl = null;
            this.fi = null;
            this.fh = null;
            this.fm = null;
        }
        get_type() {
            return "GeographicScatterAreaSeries";
        }
        get colorMemberPath() {
            return this.fl;
        }
        set colorMemberPath(a) {
            this.fl = a;
            this.g("ColorMemberPath");
        }
        get colorScale() {
            return this.fi;
        }
        set colorScale(a) {
            this.fi = a;
            this.g("ColorScale");
        }
        get actualColorScale() {
            return this.fh;
        }
        set actualColorScale(a) {
            this.fh = a;
            this.g("ActualColorScale");
        }
        get triangulationStatusChangedRef() {
            return this.fm;
        }
        set triangulationStatusChangedRef(a) {
            this.fm = a;
            this.g("TriangulationStatusChangedRef");
        }
    }
    GeographicScatterAreaSeriesDescription.$t = markType(GeographicScatterAreaSeriesDescription, 'GeographicScatterAreaSeriesDescription', GeographicXYTriangulatingSeriesDescription.$);
    return GeographicScatterAreaSeriesDescription;
})();
