// igniteui-angular-core/esm2015/lib/HorizontalRangeCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { RangeCategorySeriesDescription } from "./RangeCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let HorizontalRangeCategorySeriesDescription = /*@__PURE__*/ (() => {
    class HorizontalRangeCategorySeriesDescription extends RangeCategorySeriesDescription {
        constructor() {
            super();
            this.gd = null;
            this.ge = null;
            this.gb = null;
            this.f9 = null;
            this.gc = null;
            this.ga = null;
        }
        get_type() {
            return "HorizontalRangeCategorySeries";
        }
        get xAxisRef() {
            return this.gd;
        }
        set xAxisRef(a) {
            this.gd = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.ge;
        }
        set yAxisRef(a) {
            this.ge = a;
            this.g("YAxisRef");
        }
        get lowMemberAsLegendLabel() {
            return this.gb;
        }
        set lowMemberAsLegendLabel(a) {
            this.gb = a;
            this.g("LowMemberAsLegendLabel");
        }
        get highMemberAsLegendLabel() {
            return this.f9;
        }
        set highMemberAsLegendLabel(a) {
            this.f9 = a;
            this.g("HighMemberAsLegendLabel");
        }
        get lowMemberAsLegendUnit() {
            return this.gc;
        }
        set lowMemberAsLegendUnit(a) {
            this.gc = a;
            this.g("LowMemberAsLegendUnit");
        }
        get highMemberAsLegendUnit() {
            return this.ga;
        }
        set highMemberAsLegendUnit(a) {
            this.ga = a;
            this.g("HighMemberAsLegendUnit");
        }
    }
    HorizontalRangeCategorySeriesDescription.$t = markType(HorizontalRangeCategorySeriesDescription, 'HorizontalRangeCategorySeriesDescription', RangeCategorySeriesDescription.$);
    return HorizontalRangeCategorySeriesDescription;
})();
