// igniteui-angular-core/esm2015/lib/IDataSeriesAdapterRule_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Type, Base, String_$type, fromEnum, markType, IEnumerable_$type, typeCast, Array_$type, runOn, EnumUtil, Number_$type, typeGetValue, INotifyPropertyChanged_$type, PropertyChangedEventArgs, delegateCombine, delegateRemove } from "./type";
import { DataSeriesCollection } from "./DataSeriesCollection";
import { Stack$1 } from "./Stack$1";
import { IDataSourceLocalDataProvider_$type } from "./IDataSourceLocalDataProvider";
import { HashSet$1 } from "./HashSet$1";
import { DataSeriesType_$type } from "./DataSeriesType";
import { DataSeriesAdapterProviderInfo } from "./DataSeriesAdapterProviderInfo";
import { DataSeries } from "./DataSeries";
import { List$1 } from "./List$1";
import { Matcher } from "./Matcher";
import { DataSeriesIntent_$type } from "./DataSeriesIntent";
import { stringIsNullOrEmpty, stringReplace, stringEmpty } from "./string";
import { dateMinValue } from "./date";
import { IDataSourceDataProviderUpdateNotifier_$type } from "./IDataSourceDataProviderUpdateNotifier";
import { ObservableCollection$1 } from "./ObservableCollection$1";
import { SortDescriptionCollection } from "./SortDescriptionCollection";
import { FilterExpressionCollection } from "./FilterExpressionCollection";
import { SummaryDescriptionCollection } from "./SummaryDescriptionCollection";
import { Dictionary$2 } from "./Dictionary$2";
import { Tuple$2 } from "./Tuple$2";
import { IDataSource_$type } from "./IDataSource";
import { AggregatedDataProvider } from "./AggregatedDataProvider";
import { FlatDataProvider } from "./FlatDataProvider";
/**
 * @hidden
 */
export let IDataSeriesAdapterRule_$type = new Type(null, 'IDataSeriesAdapterRule');
/**
 * @hidden
 */
export let DataSeriesAdapterRunContext = /*@__PURE__*/ (() => {
    class DataSeriesAdapterRunContext extends Base {
        constructor(a) {
            super();
            this.g = null;
            this._analyzer = null;
            this.h = null;
            this.n = null;
            this.o = null;
            this.q = null;
            this.p = null;
            this.b = null;
            this.a = null;
            this.m = null;
            this.g = a;
            this.analyzer = new DataSeriesDataProviderAnalyzer(this);
            this.analyzer.includedProperties = this.g.includedProperties;
            this.analyzer.excludedProperties = this.g.excludedProperties;
            this.h = new DataSeriesCollection();
            this.n = new Stack$1(IDataSourceLocalDataProvider_$type);
            this.o = new Stack$1(IDataSourceLocalDataProvider_$type);
            this.q = new Stack$1(String_$type);
            this.p = new Stack$1(String_$type);
            this.b = this.g.includedProperties;
            this.a = this.g.excludedProperties;
            this.m = null;
            if (this.g.supportedSeriesTypes != null) {
                this.m = new HashSet$1(DataSeriesType_$type, 0);
                for (let b = 0; b < this.g.supportedSeriesTypes.length; b++) {
                    this.m.add_1(this.g.supportedSeriesTypes[b]);
                }
            }
        }
        get analyzer() {
            return this._analyzer;
        }
        set analyzer(a) {
            this._analyzer = a;
        }
        getSubProvider(a, b) {
            let c = this.g.h(a, false);
            this.pushDataSource(c.c);
            if (c.b != null) {
                this.pushHighlightingDataSource(c.b);
            }
            else {
                this.pushHighlightingDataSource(null);
            }
            this.pushPath(b);
            return c.c;
        }
        popSubProvider() {
            this.popDataSource();
            this.popHighlightingDataSource();
            this.popPath();
        }
        z(a, b) {
            return this.g.ap(a, b);
        }
        recurseRules(a, b, c) {
            this.g.bb(this, a, b, c);
        }
        get includedProperties() {
            return this.b;
        }
        get excludedProperties() {
            return this.a;
        }
        getCurrentDataSource() {
            return this.n.d();
        }
        getCurrentHighlightingDataSource() {
            return this.o.d();
        }
        pushDataSource(a) {
            this.n.h(a);
        }
        pushHighlightingDataSource(a) {
            this.o.h(a);
        }
        hasValidType(a) {
            if (this.m == null) {
                return true;
            }
            if (this.m.contains(a.suggestedSeries)) {
                return true;
            }
            return false;
        }
        popDataSource() {
            return this.n.e();
        }
        popHighlightingDataSource() {
            return this.o.e();
        }
        getCurrentPathSegment() {
            return this.q.d();
        }
        getCurrentPath() {
            let a = "";
            let b = true;
            for (let c of fromEnum(this.q)) {
                if (b) {
                    b = false;
                }
                else {
                    a += ".";
                }
                a += c;
            }
            return a;
        }
        getParentTitle() {
            if (this.p.f == 0) {
                return null;
            }
            return this.p.d();
        }
        pushPath(a) {
            this.q.h(a);
        }
        popPath() {
            return this.q.e();
        }
        pushParentTitle(a) {
            this.p.h(a);
        }
        popParentTitle() {
            return this.p.e();
        }
        ae(a, b) {
            if (a.priority == -1) {
                a.priority = b.priority;
            }
            if (a.dataPath == null) {
                a.dataPath = this.getCurrentPath();
            }
            if (a.data == null) {
                a.data = this.getCurrentDataSource().dataSource;
            }
            if (a.highlightedData == null) {
                let c = this.getCurrentHighlightingDataSource();
                if (c != null) {
                    a.highlightedData = c.dataSource;
                }
            }
        }
        addDataSeries(a, b) {
            this.ae(a, b);
            this.h.add(a);
        }
        removeDataSeries(a) {
            this.h.remove(a);
        }
        insertDataSeries(a, b, c) {
            this.ae(b, c);
            this.h.insert(a, b);
        }
        clearDataSeries() {
            this.h.clear();
        }
        getDataSeriesCount() {
            return this.h.count;
        }
        getDataSeriesAt(a) {
            return this.h._inner[a];
        }
    }
    DataSeriesAdapterRunContext.$t = markType(DataSeriesAdapterRunContext, 'DataSeriesAdapterRunContext');
    return DataSeriesAdapterRunContext;
})();
/**
 * @hidden
 */
export let DataSeriesDataProviderAnalyzer = /*@__PURE__*/ (() => {
    class DataSeriesDataProviderAnalyzer extends Base {
        constructor(a) {
            super();
            this.u = null;
            this.b = null;
            this.a = null;
            this.x = false;
            this.w = false;
            this.ap = new HashSet$1(String_$type, 0);
            this.ar = new List$1(Matcher.$, 0);
            this.ao = new HashSet$1(String_$type, 0);
            this.aq = new List$1(Matcher.$, 0);
            this.u = a;
        }
        get includedProperties() {
            return this.b;
        }
        set includedProperties(a) {
            this.b = a;
            this.ap.clear();
            if (this.b == null) {
                this.x = false;
            }
            else {
                this.x = true;
                let d = this.b;
                for (let c = 0; c < d.length; c++) {
                    let b = d[c];
                    if (Matcher.a(b)) {
                        this.ar.add(new Matcher(b));
                    }
                    else {
                        this.ap.add_1(b);
                    }
                }
            }
        }
        get excludedProperties() {
            return this.a;
        }
        set excludedProperties(a) {
            this.a = a;
            this.ao.clear();
            if (this.a == null) {
                this.w = false;
            }
            else {
                this.w = true;
                let d = this.a;
                for (let c = 0; c < d.length; c++) {
                    let b = d[c];
                    if (Matcher.a(b)) {
                        this.aq.add(new Matcher(b));
                    }
                    else {
                        this.ao.add_1(b);
                    }
                }
            }
        }
        getParentTitle() {
            return this.u.getParentTitle();
        }
        shouldIncludeProperty(a, b) {
            if (!stringIsNullOrEmpty(b)) {
                a = b + "." + a;
            }
            if (this.x && !this.ap.contains(a) && !this.ad(a)) {
                return false;
            }
            if (this.w && (this.ao.contains(a) || this.z(a))) {
                return false;
            }
            return true;
        }
        z(a) {
            for (let b = 0; b < this.aq.count; b++) {
                if (this.aq._inner[b].b(a)) {
                    return true;
                }
            }
            return false;
        }
        ad(a) {
            for (let b = 0; b < this.ar.count; b++) {
                if (this.ar._inner[b].b(a)) {
                    return true;
                }
            }
            return false;
        }
        expandCamelCasedWords(a) {
            if (a == null) {
                return null;
            }
            a = stringReplace(stringReplace(a, "[", ""), "]", "");
            let b = new List$1(String_$type, 0);
            let c = 0;
            for (let d = 0; d < a.length; d++) {
                let e = a.charAt(d);
                let f = e.toUpperCase();
                let g = e.toLowerCase();
                if (d == 0) {
                    e = f;
                }
                let h = (c == 0 && b.count > 0);
                let i = (c > 1);
                if (f == e) {
                    if (h) {
                        b.add(' ');
                    }
                    c++;
                }
                else if (g == e) {
                    if (i) {
                        b.insert(b.count - 1, ' ');
                    }
                    c = 0;
                }
                else {
                    c = 0;
                }
                b.add(e);
            }
            let j = stringEmpty();
            for (let k = 0; k < b.count; k++) {
                j += b._inner[k];
            }
            return j;
        }
        hasStringProperties() {
            return this.getAllStringProperties().length > 0;
        }
        getAllStringProperties() {
            return this.s((a) => a == 0, null);
        }
        getAllExcludedStringProperties() {
            return this.r((a) => a == 0, null);
        }
        isCollection(a) {
            if (typeCast(IEnumerable_$type, a) !== null || typeCast(Array_$type, a) !== null) {
                return true;
            }
            return false;
        }
        getAllPropertiesWithName(a) {
            return this.s(null, (b) => b == a);
        }
        getAllNumericProperties() {
            return this.s(runOn(this, this.aj), null);
        }
        getAllObjectProperties() {
            return this.s(runOn(this, this.ak), null);
        }
        getAllCollectionProperties() {
            let a = this.getAllObjectProperties();
            if (a.length == 0) {
                return new Array(0);
            }
            let b = this.u.getCurrentDataSource();
            let c = new List$1(String_$type, 0);
            for (let d = 0; d < b.actualCount; d++) {
                if (this.isCollection(b.getItemAtIndex(d))) {
                    c.add(b.actualSchema.propertyNames[d]);
                }
            }
            return c.toArray();
        }
        getFirstNumericProperty() {
            let a = this.getAllNumericProperties();
            if (a.length == 0) {
                return null;
            }
            return a[0];
        }
        hasNumericProperties() {
            return this.getAllNumericProperties().length > 0;
        }
        ag(a) {
            if (a == 8 || a == 9) {
                return true;
            }
            return false;
        }
        v(a) {
            let b = this.u.getCurrentDataSource();
            if (b == null) {
                return 11;
            }
            let c = b.actualSchema;
            if (c == null) {
                return 11;
            }
            let d = -1;
            for (let e = 0; e < c.propertyNames.length; e++) {
                if (Base.equalsStatic(c.propertyNames[e], a)) {
                    d = e;
                    break;
                }
            }
            if (d == -1) {
                return 11;
            }
            return c.propertyTypes[d];
        }
        s(a, b) {
            let c = this.u.getCurrentDataSource();
            let d = c.actualSchema;
            if (d == null) {
                return new Array(0);
            }
            let e = this.u.getCurrentPath();
            let f = 0;
            for (let g = 0; g < d.propertyTypes.length; g++) {
                if ((a == null || a(d.propertyTypes[g])) && (b == null || b(d.propertyNames[g])) && this.an(d.propertyDataIntents, g) && this.shouldIncludeProperty(d.propertyNames[g], e)) {
                    f++;
                }
            }
            let h = new Array(f);
            f = 0;
            for (let i = 0; i < d.propertyTypes.length; i++) {
                if ((a == null || a(d.propertyTypes[i])) && (b == null || b(d.propertyNames[i])) && this.an(d.propertyDataIntents, i) && this.shouldIncludeProperty(d.propertyNames[i], e)) {
                    h[f] = d.propertyNames[i];
                    f++;
                }
            }
            return h;
        }
        r(a, b) {
            let c = this.u.getCurrentDataSource();
            let d = c.actualSchema;
            if (d == null) {
                return new Array(0);
            }
            let e = this.u.getCurrentPath();
            let f = 0;
            for (let g = 0; g < d.propertyTypes.length; g++) {
                if ((a == null || a(d.propertyTypes[g])) && (b == null || b(d.propertyNames[g])) && this.an(d.propertyDataIntents, g) && !this.shouldIncludeProperty(d.propertyNames[g], e)) {
                    f++;
                }
            }
            let h = new Array(f);
            f = 0;
            for (let i = 0; i < d.propertyTypes.length; i++) {
                if ((a == null || a(d.propertyTypes[i])) && (b == null || b(d.propertyNames[i])) && this.an(d.propertyDataIntents, i) && !this.shouldIncludeProperty(d.propertyNames[i], e)) {
                    h[f] = d.propertyNames[i];
                    f++;
                }
            }
            return h;
        }
        an(a, b) {
            if (a == null) {
                return true;
            }
            return !this.y(a, b, "DontPlot");
        }
        y(a, b, c) {
            if (a == null) {
                return false;
            }
            for (let d = 0; d < a[b].length; d++) {
                let e = a[b][d];
                if (Base.equalsStatic(e.toLowerCase(), c.toLowerCase())) {
                    return true;
                }
            }
            return false;
        }
        getAllPropertiesWithIntent(a) {
            return this.m(a, false, false);
        }
        getAllPropertiesWithValuelessIntent(a) {
            return this.m(a, true, false);
        }
        getPropertyIntentValue(a, b) {
            let c = this.u.getCurrentDataSource();
            if (c == null) {
                return null;
            }
            let d = c.actualSchema;
            if (d == null) {
                return null;
            }
            let e = d.propertyDataIntents;
            if (e == null || d.propertyNames == null) {
                return null;
            }
            for (let f = 0; f < d.propertyNames.length; f++) {
                if (d.propertyNames[f] != null && Base.equalsStatic(d.propertyNames[f], b)) {
                    for (let g = 0; g < e[f].length; g++) {
                        let h = e[f][g];
                        let i = h.split('/');
                        if (i.length == 2 && i[0] != null && Base.equalsStatic(i[0], a)) {
                            return i[1];
                        }
                    }
                }
            }
            return null;
        }
        getAllPropertiesWithValuedIntent(a) {
            return this.m(a, false, true);
        }
        m(a, b, c) {
            let d = this.u.getCurrentDataSource();
            if (d == null) {
                return new Array(0);
            }
            let e = d.actualSchema;
            if (e == null) {
                return new Array(0);
            }
            let f = e.propertyDataIntents;
            if (f == null || e.propertyNames == null) {
                return new Array(0);
            }
            let g = 0;
            for (let h = 0; h < e.propertyNames.length; h++) {
                for (let i = 0; i < f[h].length; i++) {
                    let j = f[h][i];
                    let k = j.split('/');
                    if (b && k.length > 1) {
                        continue;
                    }
                    if (c && k.length < 2) {
                        continue;
                    }
                    if (Base.equalsStatic(a.toLowerCase(), k[0].toLowerCase())) {
                        g++;
                    }
                }
            }
            let l = new Array(g);
            g = 0;
            for (let m = 0; m < e.propertyNames.length; m++) {
                for (let n = 0; n < f[m].length; n++) {
                    let o = f[m][n];
                    let p = o.split('/');
                    if (b && p.length > 1) {
                        continue;
                    }
                    if (c && p.length < 2) {
                        continue;
                    }
                    if (Base.equalsStatic(a.toLowerCase(), o.toLowerCase())) {
                        l[g] = e.propertyNames[m];
                        g++;
                    }
                }
            }
            return l;
        }
        getAllDateTimeProperties() {
            return this.s(runOn(this, this.ag), null);
        }
        getAllExcludedDateTimeProperties() {
            return this.r(runOn(this, this.ag), null);
        }
        getFirstDateTimeProperty() {
            let a = this.getAllDateTimeProperties();
            if (a.length == 0) {
                return null;
            }
            return a[0];
        }
        getFirstStringProperty() {
            let a = this.getAllStringProperties();
            if (a.length == 0) {
                return null;
            }
            return a[0];
        }
        hasDateTimeProperties() {
            return this.getAllDateTimeProperties().length > 0;
        }
        aj(a) {
            return a == 1 || a == 3 || a == 4 || a == 7 || a == 5 || a == 6 || a == 10;
        }
        ak(a) {
            return a == 11;
        }
        isMonotonic(a, b, c) {
            if (a == null) {
                return false;
            }
            let d = false;
            let e = false;
            let f = a.actualCount;
            for (let g = 0; g < Math.min(f, c); g++) {
                if (g == 0) {
                    continue;
                }
                let h = this.as(a.getItemValue(a.getItemAtIndex(g - 1), b));
                let i = this.as(a.getItemValue(a.getItemAtIndex(g), b));
                if (!e && h != i) {
                    d = i > h;
                    e = true;
                }
                if (h == i) {
                    continue;
                }
                if (e) {
                    let j = i > h;
                    if (j != d) {
                        return false;
                    }
                }
            }
            return true;
        }
        a1(a, b, c) {
            let d = this.u.getCurrentDataSource();
            let e = d.actualSchema;
            if (e == null) {
                return null;
            }
            let f = this.s(b, null);
            if (f.length == 0) {
                return null;
            }
            if (d.actualCount > a) {
                return null;
            }
            for (let g = 0; g < f.length; g++) {
                if (c(f[g], this.v(f[g]))) {
                    return f[g];
                }
            }
            return null;
        }
        getFirstDistinctStringProperty(a) {
            return this.a1(a, (b) => b == 0, runOn(this, this.al));
        }
        getFirstDistinctNumericProperty(a) {
            return this.a1(a, runOn(this, this.aj), runOn(this, this.ai));
        }
        getFirstDistinctMonotonicNumericProperty(a) {
            let b = this.u.getCurrentDataSource();
            return this.a1(a, runOn(this, this.aj), (c, d) => this.ai(c, d) && this.isMonotonic(b, c, a));
        }
        getAllDistinctMonotonicNumericProperties(a) {
            let b = this.u.getCurrentDataSource();
            return this.s(runOn(this, this.aj), (c) => this.ai(c, this.v(c)) && this.isMonotonic(b, c, a));
        }
        getAllExcludedDistinctMonotonicNumericProperties(a) {
            let b = this.u.getCurrentDataSource();
            return this.r(runOn(this, this.aj), (c) => this.ai(c, this.v(c)) && this.isMonotonic(b, c, a));
        }
        getFirstDistinctDateTimeProperty(a) {
            return this.a1(a, runOn(this, this.ag), runOn(this, this.af));
        }
        getFirstStringPropertyPreferringDistinct(a) {
            let b = this.getFirstDistinctStringProperty(a);
            if (b != null) {
                return b;
            }
            let c = this.getAllStringProperties();
            if (c.length == 0) {
                return null;
            }
            return c[0];
        }
        getTitleString(a, b) {
            return this.getTitleStringWithSeparator(a, b, " - ");
        }
        getTitleStringWithSeparator(a, b, c) {
            let d = this.u.getCurrentDataSource();
            let e = null;
            if (e == null && b != null) {
                let f = null;
                for (let g = 0; g < b.length; g++) {
                    let h = this.getPropertyIntentValue(EnumUtil.getName(DataSeriesIntent_$type, (10)), b[g]);
                    if (e == null) {
                        e = h;
                        f = h;
                    }
                    else {
                        if (h != null) {
                            if (h.trim().length > 0 && h != f) {
                                e += c + h;
                            }
                            f = h;
                        }
                    }
                }
            }
            if (e == null) {
                let i = this.getAllPropertiesWithValuelessIntent(EnumUtil.getName(DataSeriesIntent_$type, (10)));
                if (i != null && i.length > 0) {
                    e = i[0];
                    e = this.av(a, d, e);
                }
            }
            if (e == null) {
                if (this.getParentTitle() != null) {
                    e = this.getParentTitle();
                }
            }
            if (e == null && a != null) {
                let j = this.getAllStringProperties();
                for (let l = 0; l < j.length; l++) {
                    let k = j[l];
                    if (k != null && (Base.equalsStatic(k.toLowerCase(), "label") || Base.equalsStatic(k.toLowerCase(), "title"))) {
                        e = k;
                    }
                }
                if (e != null) {
                    e = this.av(a, d, e);
                }
            }
            return e;
        }
        av(a, b, c) {
            if (a != null) {
                let d = b.getItemValue(a, c);
                if (d != null) {
                    c = d.toString();
                }
            }
            else {
                c = this.getNonNullPropertyValue(c).toString();
            }
            return c;
        }
        getNonNullPropertyValue(a) {
            let b = this.u.getCurrentDataSource();
            for (let c = 0; c < b.actualCount; c++) {
                let d = b.getItemValue(b.getItemAtIndex(c), a);
                if (d != null) {
                    return d;
                }
            }
            return null;
        }
        al(a, b) {
            let c = this.u.getCurrentDataSource();
            let d = new HashSet$1(String_$type, 0);
            for (let e = 0; e < c.actualCount; e++) {
                let f = c.getItemValue(c.getItemAtIndex(e), a);
                let g = f;
                if (d.contains(g)) {
                    return false;
                }
                d.add_1(g);
            }
            return true;
        }
        af(a, b) {
            let c = this.u.getCurrentDataSource();
            let d = new HashSet$1(Number_$type, 0);
            for (let e = 0; e < c.actualCount; e++) {
                let f;
                let g = c.getItemValue(c.getItemAtIndex(e), a);
                if (g == null) {
                    f = dateMinValue();
                }
                else {
                    f = g;
                }
                if (d.contains(f.getTime())) {
                    return false;
                }
                d.add_1(f.getTime());
            }
            return true;
        }
        as(value_) {
            return (+value_);
        }
        ai(a, b) {
            let c = this.u.getCurrentDataSource();
            if (b == 4 || b == 1 || b == 3) {
                let d = new HashSet$1(Number_$type, 0);
                for (let e = 0; e < c.actualCount; e++) {
                    let f;
                    let g = c.getItemValue(c.getItemAtIndex(e), a);
                    if (g == null) {
                        f = -9.2233720368547758E+18;
                    }
                    else {
                        f = typeGetValue(g);
                    }
                    if (d.contains(f)) {
                        return false;
                    }
                    d.add_1(f);
                }
                return true;
            }
            if (b == 7 || b == 5 || b == 6) {
                let h = new HashSet$1(Number_$type, 0);
                for (let i = 0; i < c.actualCount; i++) {
                    let j;
                    let k = c.getItemValue(c.getItemAtIndex(i), a);
                    if (k == null) {
                        j = NaN;
                    }
                    else {
                        j = this.as(k);
                    }
                    if (h.contains(j)) {
                        return false;
                    }
                    h.add_1(j);
                }
                return true;
            }
            return false;
        }
    }
    DataSeriesDataProviderAnalyzer.$t = markType(DataSeriesDataProviderAnalyzer, 'DataSeriesDataProviderAnalyzer');
    return DataSeriesDataProviderAnalyzer;
})();
/**
 * @hidden
 */
export let DataSeriesAdapterUpdateNotifier = /*@__PURE__*/ (() => {
    class DataSeriesAdapterUpdateNotifier extends Base {
        constructor(a, b) {
            super();
            this.a = null;
            this.b = null;
            this.a = a;
            this.b = b;
        }
        notifyClearItems() {
            this.a.a4(this.b);
        }
        notifyInsertItem(a, b) {
            this.a.a5(this.b, a, b);
        }
        notifyRemoveItem(a, b) {
            this.a.a6(this.b, a, b);
        }
        notifySetItem(a, b, c) {
            this.a.a7(this.b, a, b, c);
        }
    }
    DataSeriesAdapterUpdateNotifier.$t = markType(DataSeriesAdapterUpdateNotifier, 'DataSeriesAdapterUpdateNotifier', Base.$, [IDataSourceDataProviderUpdateNotifier_$type]);
    return DataSeriesAdapterUpdateNotifier;
})();
/**
 * @hidden
 */
export let DataSeriesAdapterRuleCollection = /*@__PURE__*/ (() => {
    class DataSeriesAdapterRuleCollection extends ObservableCollection$1 {
        constructor() {
            super(IDataSeriesAdapterRule_$type, 0);
        }
    }
    DataSeriesAdapterRuleCollection.$t = markType(DataSeriesAdapterRuleCollection, 'DataSeriesAdapterRuleCollection', ObservableCollection$1.$.specialize(IDataSeriesAdapterRule_$type));
    return DataSeriesAdapterRuleCollection;
})();
/**
 * @hidden
 */
export let DataSeriesAdapterRemovingDuplicatesEventArgs = /*@__PURE__*/ (() => {
    class DataSeriesAdapterRemovingDuplicatesEventArgs extends Base {
        constructor() {
            super(...arguments);
            this.b = null;
            this.d = null;
            this.e = null;
            this.a = null;
            this.c = false;
        }
    }
    DataSeriesAdapterRemovingDuplicatesEventArgs.$t = markType(DataSeriesAdapterRemovingDuplicatesEventArgs, 'DataSeriesAdapterRemovingDuplicatesEventArgs');
    return DataSeriesAdapterRemovingDuplicatesEventArgs;
})();
/**
 * @hidden
 */
export let DataSeriesAdapterRunEventArgs = /*@__PURE__*/ (() => {
    class DataSeriesAdapterRunEventArgs extends Base {
        constructor() {
            super(...arguments);
            this.a = null;
            this.b = false;
        }
    }
    DataSeriesAdapterRunEventArgs.$t = markType(DataSeriesAdapterRunEventArgs, 'DataSeriesAdapterRunEventArgs');
    return DataSeriesAdapterRunEventArgs;
})();
/**
 * @hidden
 */
export let DataSeriesAdapterImplementation = /*@__PURE__*/ (() => {
    class DataSeriesAdapterImplementation extends Base {
        constructor() {
            super();
            this.ab = false;
            this.z = false;
            this.aa = false;
            this.ai = null;
            this.ak = null;
            this.ah = new Dictionary$2(Base.$, DataSeriesAdapterProviderInfo.$, 0);
            this.ag = new Dictionary$2(Base.$, DataSeriesAdapterProviderInfo.$, 0);
            this.aj = null;
            this.propertyChanged = null;
            this.c = null;
            this.b = null;
            this.a = null;
            this.t = null;
            this.s = null;
            this.r = null;
            this.m = null;
            this.n = null;
            this.x = null;
            this._rules = null;
            this._dataSeries = null;
            this.ac = false;
            this.removingUnsupported = null;
            this.removingDuplicates = null;
            this.t = new SortDescriptionCollection();
            this.r = new SortDescriptionCollection();
            this.x = new SummaryDescriptionCollection();
            this.m = new FilterExpressionCollection();
            this.n = new FilterExpressionCollection();
            this.s = new SortDescriptionCollection();
            this.t.n(runOn(this, this.as));
            this.r.n(runOn(this, this.as));
            this.x.n(runOn(this, this.as));
            this.m.n(runOn(this, this.as));
            this.n.n(runOn(this, this.as));
            this.s.n(runOn(this, this.as));
            this.rules = new DataSeriesAdapterRuleCollection();
            this.dataSeries = new DataSeriesCollection();
            this.a8(this.rules);
            let a = this.rules;
            a.collectionChanged = delegateCombine(a.collectionChanged, runOn(this, this.bg));
        }
        as(a, b) {
            let c = this.ad;
            let d = this.af();
            let e = this.ae();
            if (c != this.z || d != this.ab || e != this.aa) {
                this.z = c;
                this.ab = d;
                this.aa = e;
                this.refresh();
            }
        }
        ap(a, b) {
            let c = stringEmpty();
            if (this.ai != null) {
                c = this.ai(a);
            }
            if (stringIsNullOrEmpty(c)) {
                return b;
            }
            return c;
        }
        a8(a) {
        }
        bg(a, b) {
            this.refresh();
        }
        get g() {
            return this.ak;
        }
        get itemsSource() {
            return this.aj;
        }
        set itemsSource(a) {
            let b = this.aj;
            this.aj = a;
            if (b != this.aj) {
                this.a3("ItemsSource", b, this.aj);
            }
        }
        get supportedSeriesTypes() {
            return this.c;
        }
        set supportedSeriesTypes(a) {
            let b = this.c;
            this.c = a;
            if (b != this.c) {
                this.a3("SupportedSeriesTypes", b, this.c);
            }
        }
        get includedProperties() {
            return this.b;
        }
        set includedProperties(a) {
            let b = this.b;
            this.b = a;
            if (b != this.b) {
                this.a3("IncludedProperties", b, this.b);
            }
        }
        get excludedProperties() {
            return this.a;
        }
        set excludedProperties(a) {
            let b = this.a;
            this.a = a;
            if (b != this.a) {
                this.a3("ExcludedProperties", b, this.a);
            }
        }
        get sortDescriptions() {
            return this.t;
        }
        get postSortDescriptions() {
            return this.s;
        }
        get groupDescriptions() {
            return this.r;
        }
        get filterExpressions() {
            return this.m;
        }
        get highlightFilterExpressions() {
            return this.n;
        }
        get summaryDescriptions() {
            return this.x;
        }
        get rules() {
            return this._rules;
        }
        set rules(a) {
            this._rules = a;
        }
        get dataSeries() {
            return this._dataSeries;
        }
        set dataSeries(a) {
            this._dataSeries = a;
        }
        a3(a, b, c) {
            this.a9(a, b, c);
            if (this.propertyChanged != null) {
                this.propertyChanged(this, new PropertyChangedEventArgs(a));
            }
        }
        a9(a, b, c) {
            switch (a) {
                case "ItemsSource":
                    this.refresh();
                    break;
                case "SupportedSeriesTypes":
                    this.refresh();
                    break;
                case "IncludedProperties":
                    this.refresh();
                    break;
                case "ExcludedProperties":
                    this.refresh();
                    break;
            }
        }
        refresh() {
            this.ac = true;
            this.bf();
            let a = new DataSeriesAdapterRunContext(this);
            let b = this.h(this.itemsSource, true);
            this.at(this.itemsSource, b, null, false);
            a.pushDataSource(b.c);
            if (b.b != null) {
                a.pushHighlightingDataSource(b.b);
            }
            else {
                a.pushHighlightingDataSource(null);
            }
            this.bi(a);
            this.bk(a);
            this.au();
            this.ac = false;
        }
        ay(a, b) {
            let c = new DataSeriesAdapterRunContext(this);
            let d = this.h(this.itemsSource, true);
            c.pushDataSource(d.c);
            a.clear();
            let e = new HashSet$1(String_$type, 0);
            for (let f = 0; f < this.rules.count; f++) {
                let g = this.rules._inner[f];
                let h = g.getPrimaryAxisLabelsStrings(c);
                if (h == null) {
                    continue;
                }
                for (let j = 0; j < h.length; j++) {
                    let i = h[j];
                    if (i == null) {
                        continue;
                    }
                    e.add_1(i);
                    a.add_1(new Tuple$2(String_$type, String_$type, i, c.analyzer.expandCamelCasedWords(i)));
                }
            }
            b.clear();
            let k = d.c.actualSchema.propertyNames;
            for (let m = 0; m < k.length; m++) {
                let l = k[m];
                if (e.contains(l)) {
                    continue;
                }
                b.add_1(new Tuple$2(String_$type, String_$type, l, c.analyzer.expandCamelCasedWords(l)));
            }
        }
        bf() {
            for (let a of fromEnum(this.ah.keys)) {
                let b = this.ah.item(a);
                b.d = false;
            }
        }
        au() {
            let a = new List$1(Base.$, 0);
            for (let b of fromEnum(this.ah.keys)) {
                let c = this.ah.item(b);
                if (!c.d) {
                    a.add1(b);
                    this.av(c);
                }
            }
            for (let d = 0; d < a.count; d++) {
                let e = a._inner[d];
                this.ah.removeItem(e);
            }
        }
        bk(a) {
            this.be(a);
            this.bd(a);
            this.bj(a);
        }
        be(a) {
            let b = true;
            if (this.removingUnsupported != null) {
                let c = ((() => {
                    let $ret = new DataSeriesAdapterRunEventArgs();
                    $ret.a = a;
                    return $ret;
                })());
                this.removingUnsupported(this, c);
                b = !c.b;
            }
            if (b) {
                if (this.supportedSeriesTypes == null) {
                    return;
                }
                let d = new HashSet$1(DataSeriesType_$type, 0);
                for (let e = 0; e < this.supportedSeriesTypes.length; e++) {
                    d.add_1(this.supportedSeriesTypes[e]);
                }
                for (let f = a.getDataSeriesCount() - 1; f >= 0; f--) {
                    let g = a.getDataSeriesAt(f);
                    if (!d.contains(g.suggestedSeries)) {
                        a.removeDataSeries(g);
                    }
                }
            }
        }
        bj(a) {
            let b = 0;
            for (; b < Math.min(this.dataSeries.count, a.getDataSeriesCount()); b++) {
                if (!this.dataSeries._inner[b].equals(a.getDataSeriesAt(b))) {
                    this.dataSeries.removeAt(b);
                    b--;
                }
            }
            for (; b < a.getDataSeriesCount(); b++) {
                this.dataSeries.insert(b, a.getDataSeriesAt(b));
            }
            for (; b < this.dataSeries.count; b++) {
                this.dataSeries.removeAt(b);
                b--;
            }
        }
        bd(a) {
            let b = new Dictionary$2(String_$type, List$1.$.specialize(DataSeries.$), 0);
            for (let c = 0; c < a.getDataSeriesCount(); c++) {
                let d = a.getDataSeriesAt(c);
                let e = d.j();
                for (let f = 0; f < e.count; f++) {
                    if (e.item(f).intent == 16 || e.item(f).intent == 17) {
                        continue;
                    }
                    let g = (d.dataPath != null ? (d.dataPath + ".") : "") + e.item(f).path;
                    if (e.item(f).intent == 25) {
                        g = e.item(f).path;
                    }
                    if (!b.containsKey(g)) {
                        b.item(g, new List$1(DataSeries.$, 0));
                    }
                    b.item(g).add(d);
                }
            }
            for (let h of fromEnum(b.keys)) {
                if (b.item(h).count > 1) {
                    let i = new Array(b.item(h).count);
                    let j = b.item(h);
                    let k = j.count;
                    for (let l = 0; l < k; l++) {
                        i[l] = j._inner[l];
                    }
                    let m = true;
                    if (this.removingDuplicates != null) {
                        let n = ((() => {
                            let $ret = new DataSeriesAdapterRemovingDuplicatesEventArgs();
                            $ret.e = h;
                            $ret.b = a;
                            $ret.a = i;
                            return $ret;
                        })());
                        m = !n.c;
                    }
                    if (m) {
                        let o = -1;
                        let p = null;
                        j = b.item(h);
                        k = j.count;
                        for (let q = 0; q < k; q++) {
                            let r = j._inner[q];
                            if (r.priority > o) {
                                o = r.priority;
                                p = r;
                            }
                        }
                        for (let s = 0; s < k; s++) {
                            let t = j._inner[s];
                            if (t != p) {
                                a.removeDataSeries(t);
                            }
                        }
                    }
                }
            }
        }
        bi(a) {
            for (let b = 0; b < this.rules.count; b++) {
                this.bh(a, this.rules._inner[b]);
            }
        }
        bh(a, b) {
            b.evaluate(a);
        }
        get ad() {
            return this.sortDescriptions.k.count > 0 || this.groupDescriptions.k.count > 0 || this.filterExpressions.k.count > 0 || this.highlightFilterExpressions.k.count > 0 || this.summaryDescriptions.k.count > 0;
        }
        af() {
            return this.groupDescriptions.k.count > 0 && this.postSortDescriptions.k.count > 0;
        }
        ae() {
            return this.highlightFilterExpressions.k.count > 0;
        }
        h(a, b) {
            let c = new DataSeriesAdapterProviderInfo();
            let d = this.af();
            let e = this.ae();
            let f = 0;
            if (typeCast(IDataSource_$type, a) !== null || (b && this.ad)) {
                if (d || e) {
                    f = 2;
                }
                else {
                    f = 1;
                }
            }
            if (a != null && this.ah.containsKey(a)) {
                let g = this.ah.item(a);
                g.d = true;
                if (g.a != f) {
                    this.aw(this.ah, a, g);
                }
                else {
                    return g;
                }
            }
            if (typeCast(IDataSourceLocalDataProvider_$type, a) !== null) {
                c.c = a;
                return c;
            }
            if (typeCast(IDataSource_$type, a) !== null) {
                let h = new AggregatedDataProvider();
                h.targetDataSource = a;
                this.sortDescriptions.o(h.targetDataSource.sortDescriptions);
                this.groupDescriptions.o(h.targetDataSource.groupDescriptions);
                this.filterExpressions.o(h.targetDataSource.filterExpressions);
                this.summaryDescriptions.o(h.targetDataSource.summaryDescriptions);
                c.e.add(() => {
                    this.sortDescriptions.s(h.targetDataSource.sortDescriptions);
                    this.groupDescriptions.s(h.targetDataSource.groupDescriptions);
                    this.filterExpressions.s(h.targetDataSource.filterExpressions);
                    this.summaryDescriptions.s(h.targetDataSource.summaryDescriptions);
                });
                if (f == 2) {
                    let i = this.q(c, h);
                    c.c = i;
                    if (e) {
                        this.ax(c);
                    }
                    return c;
                }
                c.c = h;
                return c;
            }
            if (b && this.ad) {
                let j = this.g.getLocalDataSource(a);
                j.deferAutoRefresh = true;
                let k = new AggregatedDataProvider();
                k.targetDataSource = j;
                this.sortDescriptions.o(k.targetDataSource.sortDescriptions);
                this.groupDescriptions.o(k.targetDataSource.groupDescriptions);
                this.filterExpressions.o(k.targetDataSource.filterExpressions);
                this.summaryDescriptions.o(k.targetDataSource.summaryDescriptions);
                c.e.add(() => {
                    this.sortDescriptions.s(k.targetDataSource.sortDescriptions);
                    this.groupDescriptions.s(k.targetDataSource.groupDescriptions);
                    this.filterExpressions.s(k.targetDataSource.filterExpressions);
                    this.summaryDescriptions.s(k.targetDataSource.summaryDescriptions);
                });
                j.deferAutoRefresh = false;
                if (f == 2) {
                    let l = this.q(c, k);
                    c.c = l;
                    if (e) {
                        this.ax(c);
                    }
                    return c;
                }
                c.c = k;
                return c;
            }
            let m = this.g.getDataProvider(a);
            c.c = m;
            return c;
        }
        ax(a) {
            let b = new DataSeriesAdapterRunContext(this);
            a.c.flushAutoRefresh();
            b.pushDataSource(a.c);
            let c = this.aq(b, a);
            if (c == null) {
                return;
            }
            if (typeCast(FlatDataProvider.$, a.c) !== null) {
                let d = a.c;
                let e = d.targetDataSource;
                let f = e.clone();
                let g = f.dataProvider;
                let h = g.targetDataSource;
                this.highlightFilterExpressions.o(h.filterExpressions);
                this.groupDescriptions.o(h.groupDescriptions);
                this.sortDescriptions.o(h.sortDescriptions);
                this.summaryDescriptions.o(h.summaryDescriptions);
                a.e.add(() => {
                    this.highlightFilterExpressions.s(h.filterExpressions);
                    this.groupDescriptions.s(h.groupDescriptions);
                    this.sortDescriptions.s(h.sortDescriptions);
                    this.summaryDescriptions.s(h.summaryDescriptions);
                });
                f.flushAutoRefresh();
                let i = new FlatDataProvider();
                i.targetDataSource = e;
                i.rightDataSource = f;
                d.joinType = 1;
                d.rightDataSource = f;
                d.leftJoinProperty = c;
                d.rightJoinProperty = c;
                d.joinCollisionType = 1;
                i.joinType = 1;
                i.rightDataSource = f;
                i.leftJoinProperty = c;
                i.rightJoinProperty = c;
                i.joinCollisionType = 2;
                d.broadcastChangesToRightDataSource = true;
                a.c = d;
                i.flushAutoRefresh();
                a.b = i;
            }
        }
        aq(a, b) {
            let c = -1;
            let d = null;
            for (let e = 0; e < this.rules.count; e++) {
                let f = this.rules._inner[e].priority;
                let g = this.rules._inner[e].getPrimaryAxisLabelsString(a);
                if (g != null && f > c) {
                    d = g;
                    c = f;
                }
            }
            return d;
        }
        q(a, b) {
            let c = this.g.getLocalDataSource(null);
            c.dataProvider = b;
            let d = new FlatDataProvider();
            d.targetDataSource = c;
            this.postSortDescriptions.o(d.targetDataSource.sortDescriptions);
            a.e.add(() => this.postSortDescriptions.s(d.targetDataSource.sortDescriptions));
            return d;
        }
        aw(a, b, c) {
            this.av(c);
            a.removeItem(b);
        }
        av(a) {
            if (a.e.count > 0) {
                for (let b = 0; b < a.e.count; b++) {
                    let c = a.e._inner[b];
                    c();
                }
                a.e.clear();
            }
            let d = a.c;
            d.updateNotifier = null;
            d.schemaChanged = delegateRemove(d.schemaChanged, runOn(this, this.ba));
            let e = a.b;
            if (e != null) {
                e.updateNotifier = null;
                e.schemaChanged = delegateRemove(e.schemaChanged, runOn(this, this.ba));
            }
        }
        bb(a, b, c, d) {
            let e = this.h(b, false);
            this.at(b, e, a.getCurrentDataSource(), d);
            a.pushDataSource(e.c);
            if (e.b != null) {
                a.pushHighlightingDataSource(e.b);
            }
            else {
                a.pushHighlightingDataSource(null);
            }
            a.pushPath(c);
            this.bi(a);
            a.popPath();
            a.popDataSource();
            a.popHighlightingDataSource();
        }
        at(a, b, c, d) {
            if (a != null && !this.ah.containsKey(a)) {
                b.a = 0;
                if (typeCast(AggregatedDataProvider.$, b.c) !== null) {
                    b.a = 1;
                }
                else if (typeCast(FlatDataProvider.$, b.c) !== null) {
                    b.a = 2;
                }
                b.d = true;
                this.ah.addItem(a, b);
                let e = b.c;
                e.schemaChanged = delegateCombine(e.schemaChanged, runOn(this, this.ba));
                if (b.b != null) {
                    let f = b.b;
                    f.schemaChanged = delegateCombine(f.schemaChanged, runOn(this, this.ba));
                }
                if (c != null) {
                    let g = new DataSeriesAdapterUpdateNotifier(this, b.c);
                    c.updateNotifier = g;
                }
                if (d) {
                    let h = new DataSeriesAdapterUpdateNotifier(this, b.c);
                    b.c.updateNotifier = h;
                    if (b.b != null) {
                        let i = new DataSeriesAdapterUpdateNotifier(this, b.b);
                        b.b.updateNotifier = h;
                    }
                }
            }
        }
        ba(a, b) {
            if (!this.ac) {
                this.refresh();
            }
        }
        a4(a) {
            this.refresh();
        }
        a5(a, b, c) {
            this.refresh();
        }
        a6(a, b, c) {
            this.refresh();
        }
        a7(a, b, c, d) {
            this.refresh();
        }
        notifyClearItems(a) {
            if (this.ah.containsKey(a)) {
                this.ah.item(a).c.notifyClearItems();
            }
        }
        notifyInsertItem(a, b, c) {
            if (this.ah.containsKey(a)) {
                this.ah.item(a).c.notifyInsertItem(b, c);
            }
        }
        notifyRemoveItem(a, b, c) {
            if (this.ah.containsKey(a)) {
                this.ah.item(a).c.notifyRemoveItem(b, c);
            }
        }
        notifySetItem(a, b, c, d) {
            if (this.ah.containsKey(a)) {
                this.ah.item(a).c.notifySetItem(b, c, d);
            }
        }
    }
    DataSeriesAdapterImplementation.$t = markType(DataSeriesAdapterImplementation, 'DataSeriesAdapterImplementation', Base.$, [INotifyPropertyChanged_$type]);
    return DataSeriesAdapterImplementation;
})();
