// igniteui-angular-core/esm2015/lib/igx-capture-image-settings.js
import { CaptureImageFormat_$type } from "./CaptureImageFormat";
import { CaptureImageSettings as CaptureImageSettings_internal } from "./CaptureImageSettings";
import { ensureEnum, ensureBool } from "./componentUtil";
export class IgxCaptureImageSettings {
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    createImplementation() {
        return new CaptureImageSettings_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    get format() {
        return this.i.a;
    }
    set format(v) {
        this.i.a = ensureEnum(CaptureImageFormat_$type, v);
    }
    get addToClipboard() {
        return this.i.b;
    }
    set addToClipboard(v) {
        this.i.b = ensureBool(v);
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
