// igniteui-angular-core/esm2015/lib/igx-format-specifier.js
import { FormatSpecifier as FormatSpecifier_internal } from "./FormatSpecifier";
export class IgxFormatSpecifier {
    constructor() {
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    createImplementation() {
        return new FormatSpecifier_internal();
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
    getLocalCulture() {
        let iv = this.i.getLocalCulture();
        return (iv);
    }
}
