// igniteui-angular-core/esm2015/lib/igx-child-content-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxChildContentComponent } from './igx-child-content-component';
import { TypeRegistrar } from './type';
import * as i0 from "@angular/core";
export let IgxChildContentModule = /*@__PURE__*/ (() => {
    class IgxChildContentModule {
        constructor() {
            TypeRegistrar.registerCons("IgxChildContentComponent", IgxChildContentComponent);
        }
    }
    IgxChildContentModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChildContentModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxChildContentModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChildContentModule, declarations: [IgxChildContentComponent], imports: [CommonModule], exports: [IgxChildContentComponent] });
    IgxChildContentModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChildContentModule, imports: [[CommonModule]] });
    return IgxChildContentModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChildContentModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxChildContentComponent],
                    exports: [IgxChildContentComponent],
                    imports: [CommonModule],
                    entryComponents: [IgxChildContentComponent]
                }]
        }], ctorParameters: function () { return []; } });
