// igniteui-angular-core/esm2015/lib/igx-popup-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxChildContentDynamicModule } from "./igx-child-content-dynamic-module";
import { IgxPopupComponent } from './igx-popup-component';
import { Popup } from './Popup';
import { TypeRegistrar } from "./type";
import { IgxPopupModule } from './igx-popup-module';
import * as i0 from "@angular/core";
export let IgxPopupDynamicModule = /*@__PURE__*/ (() => {
    class IgxPopupDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('Popup', Popup);
            TypeRegistrar.registerCons('IgxPopupComponent', IgxPopupComponent);
            TypeRegistrar.register('Popup', Popup.$type);
        }
    }
    IgxPopupDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPopupDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPopupDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPopupDynamicModule, imports: [CommonModule, IgxChildContentDynamicModule,
            IgxPopupModule], exports: [IgxChildContentDynamicModule,
            IgxPopupModule] });
    IgxPopupDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPopupDynamicModule, imports: [[CommonModule, IgxChildContentDynamicModule,
                IgxPopupModule
            ], IgxChildContentDynamicModule,
            IgxPopupModule] });
    return IgxPopupDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPopupDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxChildContentDynamicModule,
                        IgxPopupModule
                    ],
                    imports: [CommonModule, IgxChildContentDynamicModule,
                        IgxPopupModule
                    ],
                    entryComponents: [IgxPopupComponent]
                }]
        }], ctorParameters: function () { return []; } });
