// igniteui-angular-core/esm2015/lib/igx-heat-tile-generator-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxHeatTileGenerator } from './igx-heat-tile-generator';
import { TypeRegistrar } from "./type";
import { IgxHeatTileGeneratorModule } from './igx-heat-tile-generator-module';
import * as i0 from "@angular/core";
export let IgxHeatTileGeneratorDynamicModule = /*@__PURE__*/ (() => {
    class IgxHeatTileGeneratorDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('IgxHeatTileGenerator', IgxHeatTileGenerator);
        }
    }
    IgxHeatTileGeneratorDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHeatTileGeneratorDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxHeatTileGeneratorDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHeatTileGeneratorDynamicModule, imports: [CommonModule, IgxHeatTileGeneratorModule], exports: [IgxHeatTileGeneratorModule] });
    IgxHeatTileGeneratorDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHeatTileGeneratorDynamicModule, imports: [[CommonModule, IgxHeatTileGeneratorModule
            ], IgxHeatTileGeneratorModule] });
    return IgxHeatTileGeneratorDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHeatTileGeneratorDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxHeatTileGeneratorModule
                    ],
                    imports: [CommonModule, IgxHeatTileGeneratorModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
