// igniteui-angular-core/esm2015/lib/igx-heat-tile-generator-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxHeatTileGenerator } from './igx-heat-tile-generator';
import { TypeRegistrar } from "./type";
import * as i0 from "@angular/core";
export let IgxHeatTileGeneratorModule = /*@__PURE__*/ (() => {
    class IgxHeatTileGeneratorModule {
        constructor() {
            TypeRegistrar.registerCons('IgxHeatTileGenerator', IgxHeatTileGenerator);
        }
    }
    IgxHeatTileGeneratorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHeatTileGeneratorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxHeatTileGeneratorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHeatTileGeneratorModule, imports: [CommonModule] });
    IgxHeatTileGeneratorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHeatTileGeneratorModule, imports: [[CommonModule,
            ]] });
    return IgxHeatTileGeneratorModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxHeatTileGeneratorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
