// igniteui-angular-core/esm2015/lib/igx-data-visualization-locale-zh-hant-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Localization } from "./Localization";
import { DataVisualizationLocaleZhHant } from './DataVisualizationLocaleZhHant';
import * as i0 from "@angular/core";
export let IgxDataVisualizationLocaleZhHantModule = /*@__PURE__*/ (() => {
    class IgxDataVisualizationLocaleZhHantModule {
        constructor() {
            Localization.register("DataVisualization-zh-Hant", new DataVisualizationLocaleZhHant());
        }
    }
    IgxDataVisualizationLocaleZhHantModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleZhHantModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataVisualizationLocaleZhHantModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleZhHantModule, imports: [CommonModule] });
    IgxDataVisualizationLocaleZhHantModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleZhHantModule, imports: [[CommonModule]] });
    return IgxDataVisualizationLocaleZhHantModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleZhHantModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule],
                    exports: [],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
