// igniteui-angular-core/esm2015/lib/igx-popup-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxChildContentModule } from "./igx-child-content-module";
import { IgxPopupComponent } from './igx-popup-component';
import { Popup } from './Popup';
import { TypeRegistrar } from "./type";
import * as i0 from "@angular/core";
export let IgxPopupModule = /*@__PURE__*/ (() => {
    class IgxPopupModule {
        constructor() {
            TypeRegistrar.registerCons('IgxPopupComponent', IgxPopupComponent);
            TypeRegistrar.register('Popup', Popup.$type);
        }
    }
    IgxPopupModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPopupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxPopupModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPopupModule, declarations: [IgxPopupComponent], imports: [CommonModule, IgxChildContentModule], exports: [IgxChildContentModule,
            IgxPopupComponent] });
    IgxPopupModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPopupModule, imports: [[CommonModule, IgxChildContentModule
            ], IgxChildContentModule] });
    return IgxPopupModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxPopupModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxPopupComponent],
                    exports: [IgxChildContentModule,
                        IgxPopupComponent
                    ],
                    imports: [CommonModule, IgxChildContentModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
