// igniteui-angular-core/esm2015/lib/igx-number-format-specifier-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxNumberFormatSpecifierModule } from './igx-number-format-specifier-module';
import { IgxNumberFormatSpecifier } from './igx-number-format-specifier';
import { NumberFormatSpecifier } from './NumberFormatSpecifier';
import { TypeRegistrar } from './type';
import * as i0 from "@angular/core";
export let IgxNumberFormatSpecifierDynamicModule = /*@__PURE__*/ (() => {
    class IgxNumberFormatSpecifierDynamicModule {
        constructor() {
            TypeRegistrar.registerCons("IgxNumberFormatSpecifier", IgxNumberFormatSpecifier);
            TypeRegistrar.register("NumberFormatSpecifier", NumberFormatSpecifier.$type);
        }
    }
    IgxNumberFormatSpecifierDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumberFormatSpecifierDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxNumberFormatSpecifierDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumberFormatSpecifierDynamicModule, imports: [CommonModule, IgxNumberFormatSpecifierModule], exports: [IgxNumberFormatSpecifierModule] });
    IgxNumberFormatSpecifierDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumberFormatSpecifierDynamicModule, imports: [[CommonModule, IgxNumberFormatSpecifierModule], IgxNumberFormatSpecifierModule] });
    return IgxNumberFormatSpecifierDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxNumberFormatSpecifierDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxNumberFormatSpecifierModule],
                    exports: [IgxNumberFormatSpecifierModule],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
