// igniteui-angular-core/esm2015/lib/igx-shape-data-source-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxShapeDataSource } from './igx-shape-data-source';
import { TypeRegistrar } from "./type";
import { IgxShapeDataSourceModule } from './igx-shape-data-source-module';
import * as i0 from "@angular/core";
export let IgxShapeDataSourceDynamicModule = /*@__PURE__*/ (() => {
    class IgxShapeDataSourceDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('IgxShapeDataSource', IgxShapeDataSource);
        }
    }
    IgxShapeDataSourceDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxShapeDataSourceDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxShapeDataSourceDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxShapeDataSourceDynamicModule, imports: [CommonModule, IgxShapeDataSourceModule], exports: [IgxShapeDataSourceModule] });
    IgxShapeDataSourceDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxShapeDataSourceDynamicModule, imports: [[CommonModule, IgxShapeDataSourceModule
            ], IgxShapeDataSourceModule] });
    return IgxShapeDataSourceDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxShapeDataSourceDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxShapeDataSourceModule
                    ],
                    imports: [CommonModule, IgxShapeDataSourceModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
