// igniteui-angular-core/esm2015/lib/igx-dv-interactivity-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CanvasGestureDOMEventProxy } from './CanvasGestureDOMEventProxy';
import { IgxTooltipContainerDynamicModule } from './igx-tooltip-container-dynamic-module';
import { IgxSimpleDefaultTooltipComponent } from './igx-simple-default-tooltip-component';
import { TypeRegistrar } from "./type";
import { IgxDVInteractivityModule } from './igx-dv-interactivity-module';
import * as i0 from "@angular/core";
export let IgxDVInteractivityDynamicModule = /*@__PURE__*/ (() => {
    class IgxDVInteractivityDynamicModule {
        constructor() {
            TypeRegistrar.registerCons('CanvasGestureDOMEventProxy', CanvasGestureDOMEventProxy);
            TypeRegistrar.register('CanvasGestureDOMEventProxy', CanvasGestureDOMEventProxy.$type);
            IgxSimpleDefaultTooltipComponent.register();
        }
    }
    IgxDVInteractivityDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDVInteractivityDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDVInteractivityDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDVInteractivityDynamicModule, imports: [CommonModule, IgxTooltipContainerDynamicModule,
            IgxDVInteractivityModule], exports: [IgxTooltipContainerDynamicModule,
            IgxDVInteractivityModule] });
    IgxDVInteractivityDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDVInteractivityDynamicModule, imports: [[CommonModule, IgxTooltipContainerDynamicModule,
                IgxDVInteractivityModule
            ], IgxTooltipContainerDynamicModule,
            IgxDVInteractivityModule] });
    return IgxDVInteractivityDynamicModule;
})();
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDVInteractivityDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxTooltipContainerDynamicModule,
                        IgxDVInteractivityModule
                    ],
                    imports: [CommonModule, IgxTooltipContainerDynamicModule,
                        IgxDVInteractivityModule
                    ],
                    entryComponents: [IgxSimpleDefaultTooltipComponent]
                }]
        }], ctorParameters: function () { return []; } });
