// igniteui-angular-core/esm2015/lib/INativeUIElementBridge_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, Type, enumGetBox, EnumUtil, markType } from "./type";
import { BaseControlTheme_$type } from "./BaseControlTheme";
import { NativeUICursor_$type } from "./NativeUICursor";
import { Visibility_$type } from "./Visibility";
import { INativeUIElementFactory_$type } from "./INativeUIElementFactory";
import { List$1 } from "./List$1";
import { NativeUI } from "./NativeUI";
/**
 * @hidden
 */
export let INativeUIElementBridge_$type = new Type(null, 'INativeUIElementBridge');
/**
 * @hidden
 */
export let INativeUIContainerBridge_$type = new Type(null, 'INativeUIContainerBridge', null, [INativeUIElementBridge_$type]);
/**
 * @hidden
 */
export let NativeUIComponent = /*@__PURE__*/ (() => {
    class NativeUIComponent extends Base {
        constructor() {
            super(...arguments);
            this.j = null;
            this.ao = null;
            this.d = null;
            this.an = null;
            this.p = null;
            this.r = null;
            this.l = null;
            this.m = null;
        }
        az() {
            if (this.j != null) {
                this.j.bn(this);
            }
        }
        bd(a) {
            return this.d.getDesiredSize(this.an, a);
        }
        bc() {
            return this.d.getBoundingRect(this.an);
        }
        t(a) {
            return this.d.contains(this.an, a);
        }
        ay() {
            this.d.focus(this.an);
        }
        get_e() {
            return 8;
        }
        get e() {
            return this.get_e();
        }
        get am() {
            return this.d.getMethodTarget(this.an);
        }
        static ar(a) {
            if (!NativeUIComponent.w.contains(a)) {
                NativeUIComponent.w.add(a);
            }
        }
        static a(a, b) {
            let c = new List$1(INativeUIElementFactory_$type, 0);
            let d = new List$1(INativeUIElementFactory_$type, 0);
            for (let e = 0; e < a.length; e++) {
                let f = a[e];
                let g = b[e];
                for (let h = 0; h < NativeUIComponent.w.count; h++) {
                    if (NativeUIComponent.w._inner[h].flavor == g) {
                        d.add(NativeUIComponent.w._inner[h]);
                    }
                }
                for (let i = 0; i < d.count; i++) {
                    if (d._inner[i].supportsComponent(f)) {
                        c.add(d._inner[i]);
                        break;
                    }
                }
                if (c.count - 1 == e) {
                    d.clear();
                    continue;
                }
                for (let j = 0; j < NativeUIComponent.w.count; j++) {
                    if (NativeUIComponent.w._inner[j].supportsComponent(f)) {
                        c.add(NativeUIComponent.w._inner[j]);
                        break;
                    }
                }
                if (c.count - 1 != e) {
                    c.add(null);
                }
                d.clear();
            }
            return c.toArray();
        }
        static v(a, b) {
            let c = NativeUIComponent.a([a], [b]);
            return c != null && c.length > 0 && c[0] != null;
        }
        static as(a, b, c, d) {
            let e = NativeUIComponent.a([b], [c]);
            if (e.length > 0 && e[0] != null) {
                e[0].createComponent(a, b, d);
            }
        }
        static at(a, b, c) {
            NativeUIComponent.as(a, b, 0, c);
        }
        static au(a, b, c, d) {
            let e = 0;
            let f = new Array(b.length);
            let g = NativeUIComponent.a(b, c);
            for (let h = 0; h < g.length; h++) {
                let i = g[h];
                let j = b[h];
                if (i != null) {
                    i.createComponent(a, j, (k) => {
                        f[h] = k;
                        e++;
                        if (e == b.length) {
                            d(f);
                        }
                    });
                }
                else {
                    f[h] = null;
                    e++;
                    if (e == b.length) {
                        d(f);
                    }
                }
            }
        }
        static av(a, b, c) {
            let d = new Array(b.length);
            for (let e = 0; e < d.length; e++) {
                d[e] = 0;
            }
            NativeUIComponent.au(a, b, d, c);
        }
        static aw(a, b, c, d) {
            let e = NativeUIComponent.a([b], [c]);
            if (e.length > 0 && e[0] != null) {
                e[0].createComponentSync(a, b, d);
            }
        }
        static ax(a, b, c) {
            NativeUIComponent.aw(a, b, 0, c);
        }
        al(a) {
            return this.d.getTargetForProperty(a, this.an);
        }
        ai(a) {
            let b = this.al(a);
            return NativeUI.l(a, b);
        }
        a4(a, b) {
            let c = this.al(a);
            this.d.beforeSetNativeValue(a, b);
            NativeUI.ag(a, c, b);
            this.d.afterSetNativeValue(a, b);
        }
        ag(a) {
            let b = this.ak(a);
            return this.ah(b, a);
        }
        a2(a, b) {
            let c = this.ak(a);
            this.a3(c, a, b);
        }
        ap(a, b) {
            let c = this.aj(a);
            this.aq(c, this, a, b);
        }
        aq(a, b, c, d) {
            this.d.addHandler(a, b, c, d);
        }
        a0(a, b) {
            let c = this.aj(a);
            this.a1(c, this, a, b);
        }
        a1(a, b, c, d) {
            this.d.removeHandler(a, b, c, d);
        }
        ah(a, b) {
            return this.d.getValue(a, b);
        }
        a3(a, b, c) {
            this.d.setValue(a, b, c);
        }
        ak(a) {
            let b = this.an;
            b = this.d.getTargetForBuildInProperty(b, a);
            return b;
        }
        aj(a) {
            let b = this.an;
            b = this.d.getTargetForBuildInEvent(b, a);
            return b;
        }
        get ae() {
            return this.ai(9);
        }
        set ae(a) {
            this.a4(9, a);
        }
        get y() {
            return this.ai(10);
        }
        set y(a) {
            this.a4(10, a);
        }
        get aa() {
            return this.ai(1);
        }
        set aa(a) {
            this.a4(1, a);
        }
        get c() {
            let a = this.ag(18);
            return EnumUtil.getEnumValue(BaseControlTheme_$type, a);
        }
        set c(a) {
            this.a2(18, enumGetBox(BaseControlTheme_$type, a));
        }
        get a6() {
            return this.ai(2);
        }
        set a6(a) {
            this.a4(2, a);
        }
        get a7() {
            return this.ai(3);
        }
        set a7(a) {
            this.a4(3, a);
        }
        get be() {
            return this.ai(4);
        }
        set be(a) {
            this.a4(4, a);
        }
        get a5() {
            return this.ai(5);
        }
        set a5(a) {
            this.a4(5, a);
        }
        get bg() {
            return this.ai(6);
        }
        set bg(a) {
            this.a4(6, a);
        }
        get a8() {
            return this.ai(7);
        }
        set a8(a) {
            this.a4(7, a);
        }
        get x() {
            return this.ai(8);
        }
        set x(a) {
            this.a4(8, a);
        }
        get z() {
            return this.ai(11);
        }
        set z(a) {
            this.a4(11, a);
        }
        get ad() {
            return this.ai(12);
        }
        set ad(a) {
            this.a4(12, a);
        }
        get bf() {
            return this.ai(13);
        }
        set bf(a) {
            this.a4(13, a);
        }
        get bb() {
            return this.ai(14);
        }
        set bb(a) {
            this.a4(14, a);
        }
        get ba() {
            return this.ai(15);
        }
        set ba(a) {
            this.a4(15, a);
        }
        get a9() {
            return this.ai(16);
        }
        set a9(a) {
            this.a4(16, a);
        }
        get ac() {
            return this.ai(17);
        }
        set ac(a) {
            this.a4(17, a);
        }
        get b() {
            return this.ai(18);
        }
        set b(a) {
            this.a4(18, a);
        }
        get ab() {
            return this.ai(19);
        }
        set ab(a) {
            this.a4(19, a);
        }
        get f() {
            return this.d.getHorizontalAlignment(this.an);
        }
        set f(a) {
            this.d.setHorizontalAlignment(this.an, a);
        }
        get h() {
            return this.d.getVerticalAlignment(this.an);
        }
        set h(a) {
            this.d.setVerticalAlignment(this.an, a);
        }
        get g() {
            return this.d.getHorizontalContentAlignment(this.an);
        }
        set g(a) {
            this.d.setHorizontalContentAlignment(this.an, a);
        }
        get i() {
            return this.d.getVerticalContentAlignment(this.an);
        }
        set i(a) {
            this.d.setVerticalContentAlignment(this.an, a);
        }
        get k() {
            return EnumUtil.getEnumValue(NativeUICursor_$type, this.ai(20));
        }
        set k(a) {
            this.a4(20, enumGetBox(NativeUICursor_$type, a));
        }
        get_u() {
            return this.ai(21);
        }
        set_u(a) {
            this.a4(21, a);
        }
        get u() {
            return this.get_u();
        }
        set u(a) {
            this.set_u(a);
        }
        get bh() {
            return EnumUtil.getEnumValue(Visibility_$type, this.ai(24));
        }
        set bh(a) {
            this.a4(24, enumGetBox(Visibility_$type, a));
        }
        get q() {
            return this.p;
        }
        set q(a) {
            let b = this.p;
            this.p = a;
            if (this.p != b) {
                if (b != null) {
                    this.a0(7, b);
                }
                if (this.p != null) {
                    this.ap(7, this.p);
                }
            }
        }
        get s() {
            return this.r;
        }
        set s(a) {
            let b = this.r;
            this.r = a;
            if (this.r != b) {
                if (b != null) {
                    this.a0(8, b);
                }
                if (this.p != null) {
                    this.ap(8, this.r);
                }
            }
        }
        get n() {
            return this.l;
        }
        set n(a) {
            let b = this.l;
            this.l = a;
            if (this.l != b) {
                if (b != null) {
                    this.a0(13, b);
                }
                if (this.l != null) {
                    this.ap(13, this.l);
                }
            }
        }
        get o() {
            return this.m;
        }
        set o(a) {
            let b = this.m;
            this.m = a;
            if (this.m != b) {
                if (b != null) {
                    this.a0(14, b);
                }
                if (this.m != null) {
                    this.ap(14, this.m);
                }
            }
        }
        af() {
            return this.d.exportVisualData(this.am);
        }
    }
    NativeUIComponent.$t = markType(NativeUIComponent, 'NativeUIComponent');
    NativeUIComponent.w = new List$1(INativeUIElementFactory_$type, 0);
    return NativeUIComponent;
})();
/**
 * @hidden
 */
export let NativeUIContainer = /*@__PURE__*/ (() => {
    class NativeUIContainer extends NativeUIComponent {
        get bk() {
            return this.bi.getChildrenCount(this.am);
        }
        get bi() {
            return this.d;
        }
        bj(a) {
            return this.bi.getChildAt(this.am, a);
        }
        bo(a) {
            let b = this.bj(a);
            if (b != null) {
                b.j = null;
            }
            this.bi.removeChildAt(this.am, a);
        }
        bl(a) {
            a.j = this;
            this.bi.addChild(this.am, a);
        }
        bm(a, b) {
            b.j = this;
            this.bi.insertChild(this.am, a, b);
        }
        bn(a) {
            a.j = null;
            this.bi.removeChild(this.am, a);
        }
    }
    NativeUIContainer.$t = markType(NativeUIContainer, 'NativeUIContainer', NativeUIComponent.$);
    return NativeUIContainer;
})();
