// igniteui-angular-core/esm2015/lib/ItemToolTipLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ItemToolTipLayerDescription = /*@__PURE__*/ (() => {
    class ItemToolTipLayerDescription extends AnnotationLayerDescription {
        constructor() {
            super();
            this.gb = null;
            this.gc = null;
            this.f8 = false;
            this.f7 = false;
        }
        get_type() {
            return "ItemToolTipLayer";
        }
        get targetSeriesName() {
            return this.gb;
        }
        set targetSeriesName(a) {
            this.gb = a;
            this.g("TargetSeriesName");
        }
        get targetSeriesRef() {
            return this.gc;
        }
        set targetSeriesRef(a) {
            this.gc = a;
            this.g("TargetSeriesRef");
        }
        get useInterpolation() {
            return this.f8;
        }
        set useInterpolation(a) {
            this.f8 = a;
            this.g("UseInterpolation");
        }
        get skipUnknownValues() {
            return this.f7;
        }
        set skipUnknownValues(a) {
            this.f7 = a;
            this.g("SkipUnknownValues");
        }
    }
    ItemToolTipLayerDescription.$t = markType(ItemToolTipLayerDescription, 'ItemToolTipLayerDescription', AnnotationLayerDescription.$);
    return ItemToolTipLayerDescription;
})();
