// igniteui-angular-core/esm2015/lib/LabelCollisionManager.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, fromEnum, runOn, Point_$type, markType } from "./type";
import { List$1 } from "./List$1";
import { LabelCollisionPlacementPositions_$type } from "./LabelCollisionPlacementPositions";
import { QuadTree } from "./QuadTree";
import { Rect } from "./Rect";
import { LinkedList$1 } from "./LinkedList$1";
import { LabelCollisionInfo } from "./LabelCollisionInfo";
import { QuadTreeBoundingBox } from "./QuadTreeBoundingBox";
import { Random } from "./Random";
import { HashSet$1 } from "./HashSet$1";
import { MathUtil } from "./MathUtil";
import { GeometryUtil } from "./GeometryUtil";
import { truncate, isNaN_ } from "./number";
/**
 * @hidden
 */
export let LabelCollisionManager = /*@__PURE__*/ (() => {
    class LabelCollisionManager extends Base {
        constructor(a, b, c, d, e, f, g, h, i, j, k) {
            super();
            this.o = false;
            this.a = new Array(8);
            this.aj = null;
            this.labelsMoved = null;
            this.m = null;
            this.h = null;
            this.g = 0;
            this.e = null;
            this.ai = null;
            this.au = 0;
            this.a2 = 0;
            this.a1 = 0;
            this.bu = null;
            this.aw = 0;
            this.at = 0;
            this.ag = new LinkedList$1(LabelCollisionInfo.$);
            this.ay = 0;
            this.ah = new List$1(LabelCollisionInfo.$, 0);
            this.q = false;
            this.p = false;
            this.a0 = 0;
            this.ax = 0;
            this.ak = 0;
            this.al = 0;
            this.az = 0;
            this.r = false;
            this.n = false;
            this.b = null;
            this.a7 = null;
            this.a5 = null;
            this.am = 0;
            this.an = 0;
            this.o = k != null;
            this.bu = c;
            this.aw = a;
            this.at = b;
            this.h = new QuadTree(a, b, 8, 8);
            this.g = f;
            this.e = d;
            this.ai = new List$1(LabelCollisionPlacementPositions_$type, 0);
            this.ai.add(0);
            this.ai.add(5);
            this.ai.add(7);
            this.ai.add(2);
            this.ai.add(3);
            this.ai.add(6);
            this.ai.add(1);
            this.ai.add(4);
            let l = 0;
            for (let m = 0; m < g.length; m++) {
                if (this.ai.contains(g[m])) {
                    this.ai.remove(g[m]);
                    this.ai.insert(l, g[m]);
                    l++;
                }
            }
            if (k != null) {
                this.aj = new List$1(LabelCollisionPlacementPositions_$type, 0);
                for (let n = 0; n < this.a.length; n++) {
                    this.a[n] = false;
                }
                for (let o = 0; o < k.length; o++) {
                    this.a[k[o]] = true;
                    this.aj.add(k[o]);
                }
                this.b = k;
            }
            else {
                this.aj = new List$1(LabelCollisionPlacementPositions_$type, 0);
                this.aj.add(0);
                this.aj.add(5);
                this.aj.add(7);
                this.aj.add(2);
                this.aj.add(3);
                this.aj.add(6);
                this.aj.add(1);
                this.aj.add(4);
                if (this.ad(this.g)) {
                    this.aj = new List$1(LabelCollisionPlacementPositions_$type, 0);
                    this.aj.add(10);
                    this.aj.add(11);
                    this.aj.add(8);
                    this.aj.add(9);
                }
                let p = 0;
                for (let q = 0; q < g.length; q++) {
                    if (this.aj.contains(g[q])) {
                        this.aj.remove(g[q]);
                        this.aj.insert(p, g[q]);
                        p++;
                    }
                }
            }
            this.au = e;
            this.a2 = h;
            this.a1 = i;
            this.n = j;
        }
        l(a, b, c, d, e) {
            let f = 0;
            let g = 0;
            switch (b) {
                case 0:
                    f = a.s;
                    g = (((a.t - a.q) - a.o) - this.au - d);
                    break;
                case 4:
                    f = (((a.s - a.r) - a.p) - this.au) - c;
                    g = (((a.t - a.q) - a.o) - this.au) - d;
                    break;
                case 5:
                    f = (((a.s + a.r) + a.p) + this.au) + c;
                    g = (((a.t - a.q) - a.o) - this.au) - d;
                    break;
                case 1:
                    f = (((a.s - a.r) - a.p) - this.au) - c;
                    g = a.t;
                    break;
                case 2:
                    f = (((a.s + a.r) + a.p) + this.au) + c;
                    g = a.t;
                    break;
                case 3:
                    f = a.s;
                    g = (((a.t + a.q) + a.o) + this.au) + d;
                    break;
                case 6:
                    f = (((a.s - a.r) - a.p) - this.au) - c;
                    g = (((a.t + a.q) + a.o) + this.au) + d;
                    break;
                case 7:
                    f = (((a.s + a.r) + a.p) + this.au) + c;
                    g = (((a.t + a.q) + a.o) + this.au) - d;
                    break;
                case 8:
                    {
                        let h = this.i(a);
                        if (h != null) {
                            a.b = 8;
                            f = h.d;
                            g = h.e;
                        }
                        else {
                            return null;
                        }
                        break;
                    }
                case 9:
                    {
                        let i = this.j(a);
                        if (i != null) {
                            a.b = 9;
                            f = i.d;
                            g = i.e;
                        }
                        else {
                            return null;
                        }
                        break;
                    }
                case 10:
                    {
                        let j = this.k(a);
                        if (j != null) {
                            a.b = 10;
                            f = j.d;
                            g = j.e;
                        }
                        else {
                            return null;
                        }
                        break;
                    }
                case 11:
                    {
                        let k = this.i(a);
                        if (k != null) {
                            a.b = 8;
                            f = k.d;
                            g = k.e;
                        }
                        else {
                            let l = this.j(a);
                            if (l != null) {
                                f = l.d;
                                g = l.e;
                            }
                            else {
                                let m = this.j(a);
                                if (m != null) {
                                    f = m.d;
                                    g = m.e;
                                }
                                else {
                                    return null;
                                }
                            }
                        }
                        break;
                    }
            }
            let n = new QuadTreeBoundingBox();
            n.d = f;
            n.e = g;
            n.g = a.p;
            n.f = a.o;
            n.i = a;
            if (this.x(n) && !e) {
                return null;
            }
            return n;
        }
        x(a) {
            let b = a.d - a.g;
            let c = a.d + a.g;
            let d = a.e - a.f;
            let e = a.e + a.f;
            let f = this.bu;
            if (b < f.left || c > f.right || d < f.top || e > f.bottom) {
                return true;
            }
            return false;
        }
        a8(a) {
            this.ay++;
            this.ag.f(a);
            a.i = this.ag.d;
            if (a.e == null) {
                a.e = new QuadTreeBoundingBox();
                a.e.d = a.s;
                a.e.e = a.t;
                a.e.g = a.r;
                a.e.f = a.q;
                a.e.i = a;
                this.h.c(a.e);
            }
            if (a.a != null) {
                if (a.d == null) {
                    a.d = QuadTreeBoundingBox.a(a.a.boundingBox);
                    a.d.i = a;
                    this.h.c(a.d);
                }
            }
        }
        bl(a) {
            if (a.e != null) {
                this.h.d(a.e);
                a.e = null;
            }
            if (a.d != null) {
                this.h.d(a.d);
                a.d = null;
            }
            if (a.c != null) {
                this.h.d(a.c);
                a.c = null;
            }
            if (a.i != null) {
                this.ag.h(a.i);
                a.i = null;
            }
            this.ay--;
        }
        bi(a) {
            if (a.e != null && a.e.d == a.s && a.e.e == a.t) {
                return;
            }
            if (a.e != null) {
                this.h.d(a.e);
            }
            if (a.d != null) {
                this.h.d(a.d);
                a.d = null;
            }
            let b = NaN;
            let c = NaN;
            if (a.e == null) {
                a.e = new QuadTreeBoundingBox();
                a.e.i = a;
            }
            else {
                if (a.c != null) {
                    b = a.c.d - a.e.d;
                    c = a.c.e - a.e.e;
                }
            }
            a.e.d = a.s;
            a.e.e = a.t;
            a.e.g = a.r;
            a.e.f = a.q;
            this.h.c(a.e);
            if (a.a != null) {
                if (a.d == null) {
                    a.d = QuadTreeBoundingBox.a(a.a.boundingBox);
                    a.d.i = a;
                    this.h.c(a.d);
                }
            }
            if (a.c != null) {
                this.h.d(a.c);
                if (a.b == 8 || a.b == 9) {
                    b = NaN;
                }
                if (!isNaN_(b)) {
                    a.c = new QuadTreeBoundingBox();
                    a.c.d = a.e.d + b;
                    a.c.e = a.e.e + c;
                    a.c.i = a;
                    a.c.g = a.p;
                    a.c.f = a.o;
                    a.k = a.c.d;
                    a.l = a.c.e;
                    this.h.c(a.c);
                }
                else {
                    a.c = this.l(a, a.b, 0, 0, true);
                    if (a.c != null) {
                        this.bg(a);
                        a.k = a.c.d;
                        a.l = a.c.e;
                        this.h.c(a.c);
                    }
                    else {
                        this.bf(a);
                    }
                }
            }
        }
        bh(a) {
            this.h.d(a.c);
            a.c.d = a.k;
            a.c.e = a.l;
            a.c.g = a.p;
            a.c.f = a.o;
            this.h.c(a.c);
        }
        bp() {
            this.p = false;
            {
                let a = this.ag.c;
                while (a != null) {
                    a.c.h = false;
                    if (a.c.c != null) {
                        this.h.d(a.c.c);
                        a.c.c = null;
                        if (a.c.j != null) {
                            a.c.j.clear();
                        }
                    }
                    a = a.a;
                }
                this.az = 0;
            }
            this.ax = 0;
            this.ak = 0.4;
            this.al = 0.4;
            let b = truncate(Math.ceil((this.a1 / 150) * this.a2));
            this.a0 = b;
            this.bc();
        }
        ba() {
            this.ah.clear();
        }
        a9() {
            this.p = true;
        }
        be() {
            this.q = false;
            if (this.p) {
                return;
            }
            this.bc();
        }
        bc() {
            this.ah.clear();
            this.bj(false);
            if (this.g == 0) {
                for (let a = 0; a < this.a2; a++) {
                    this.al = this.ak * (this.a0 - this.ax) / this.a0;
                    this.bd();
                    this.ax++;
                    if (this.ax > this.a0) {
                        if (this.m != null) {
                            this.m();
                        }
                        return;
                    }
                }
                if (this.labelsMoved != null) {
                    for (let b of fromEnum(this.ah)) {
                    }
                    this.labelsMoved(this.ah);
                }
                if (this.az > 0) {
                    if (!this.q) {
                        this.q = true;
                        this.e.executeDelayed(runOn(this, this.be), 150);
                    }
                }
                else {
                    if (this.m != null) {
                        this.m();
                    }
                }
            }
            else {
                if (this.g == 2) {
                    let c = this.ag.c;
                    this.h.k();
                    while (c != null) {
                        let d = c.c;
                        if (d.c != null) {
                            this.h.d(d.c);
                            d.c = null;
                            d.f = false;
                            d.j = null;
                        }
                        c = c.a;
                    }
                    this.az = 0;
                    this.bj(true);
                }
                if (this.labelsMoved != null) {
                    this.labelsMoved(this.ah);
                }
            }
        }
        bd() {
            let a = truncate(Math.round((this.ay - 1) * LabelCollisionManager.a6.nextDouble()));
            let b = 0;
            let c = this.ag.c;
            let d = this.as(null);
            let e = 30;
            let f = 60;
            while (c != null) {
                let g = c.c;
                let h = 0;
                let i = 0;
                if (this.n) {
                    h = LabelCollisionManager.a6.nextDouble() * g.p * 2;
                    i = LabelCollisionManager.a6.nextDouble() * g.o * 2;
                }
                if (b == a) {
                    let j = truncate(Math.round((this.aj.count - 1) * LabelCollisionManager.a6.nextDouble()));
                    while (this.aj._inner[j] == g.b) {
                        j = truncate(Math.round((this.aj.count - 1) * LabelCollisionManager.a6.nextDouble()));
                    }
                    let k = g.c;
                    let l = this.l(g, this.aj._inner[j], h, i, false);
                    let m = 0;
                    while (l == null) {
                        if (m >= e) {
                            h = 0;
                            i = 0;
                        }
                        if (m >= f) {
                            break;
                        }
                        j = truncate(Math.round((this.aj.count - 1) * LabelCollisionManager.a6.nextDouble()));
                        l = this.l(g, this.aj._inner[j], h, i, false);
                        m++;
                    }
                    g.c = l;
                    if (k != null) {
                        this.h.d(k);
                    }
                    if (l != null) {
                        this.h.c(l);
                        let n = this.as(g);
                        let o = d > n;
                        if (!o) {
                            let p = Math.exp(-(n - d) / this.al);
                            if (LabelCollisionManager.a6.nextDouble() < p) {
                                o = true;
                            }
                        }
                        if (!o && k != null) {
                            this.h.d(l);
                            this.h.c(k);
                            g.c = k;
                            if (!this.r) {
                                this.as(g);
                            }
                        }
                        else {
                            g.b = this.aj._inner[j];
                            g.k = g.c.d;
                            g.l = g.c.e;
                            this.ah.add(g);
                        }
                    }
                    else {
                        if (!this.r) {
                            this.as(g);
                        }
                        g.f = false;
                        this.ah.add(g);
                    }
                    break;
                }
                b++;
                c = c.a;
            }
        }
        v(a, b) {
            return this.w(a, b);
        }
        w(a, b) {
            if (!a.c(b)) {
                return false;
            }
            if (a.i == null) {
                return true;
            }
            let c = a.i;
            if (c.a == null) {
                if (b.i != null) {
                    let d = b.i;
                    if (d.d == b) {
                        let e = d.a;
                        return e.e(a.j());
                    }
                    else {
                        return true;
                    }
                }
            }
            let f = c.a;
            if (c.d == a) {
                return f.e(b.j());
            }
            if (b.i != null) {
                let g = b.i;
                if (g.d == b) {
                    if (this.ad(this.g)) {
                        return false;
                    }
                    let h = g.a;
                    return h.e(a.j());
                }
                else {
                    return true;
                }
            }
            return true;
        }
        as(a) {
            if (this.r) {
                let b = 0;
                let c = this.ag.c;
                while (c != null) {
                    let d = c.c;
                    let e = d.f;
                    d.f = false;
                    if (d.c != null) {
                        let f = this.h.e(d.c);
                        for (let g = 0; g < f.count; g++) {
                            let h = f._inner[g].l.c;
                            while (h != null) {
                                let i = h.c;
                                if (this.ae(i, d)) {
                                    h = h.a;
                                    continue;
                                }
                                if (this.v(d.c, i)) {
                                    d.f = true;
                                    b += d.c.h(i);
                                }
                                h = h.a;
                            }
                        }
                        if (d.f != e) {
                            this.ah.add(d);
                        }
                    }
                    c = c.a;
                }
                return b;
            }
            else {
                if (a != null) {
                    let j = new HashSet$1(LabelCollisionInfo.$, 0);
                    let k = new HashSet$1(LabelCollisionInfo.$, 0);
                    let l = new HashSet$1(LabelCollisionInfo.$, 0);
                    let m = new List$1(LabelCollisionInfo.$, 0);
                    if (a.j != null) {
                        for (let n = 0; n < a.j.count; n++) {
                            j.add_1(a.j._inner[n]);
                            k.add_1(a.j._inner[n]);
                        }
                    }
                    let o = a.f;
                    a.f = false;
                    if (a.c != null) {
                        let p = this.h.e(a.c);
                        for (let q = 0; q < p.count; q++) {
                            let r = p._inner[q].l.c;
                            while (r != null) {
                                let s = r.c;
                                if (this.ae(s, a)) {
                                    r = r.a;
                                    continue;
                                }
                                if (this.v(a.c, s) && !l.contains(s.i)) {
                                    a.f = true;
                                    l.add_1(s.i);
                                    m.add(s.i);
                                }
                                r = r.a;
                            }
                        }
                        p = this.h.e(a.e);
                        for (let t = 0; t < p.count; t++) {
                            let u = p._inner[t].l.c;
                            while (u != null) {
                                let v = u.c;
                                if (this.ae(v, a)) {
                                    u = u.a;
                                    continue;
                                }
                                if (this.s(a.e, v)) {
                                    u = u.a;
                                    continue;
                                }
                                if (this.v(a.e, v) && !l.contains(v.i)) {
                                    a.f = true;
                                    l.add_1(v.i);
                                    m.add(v.i);
                                }
                                u = u.a;
                            }
                        }
                        if (a.f != o) {
                            this.ah.add(a);
                        }
                        if (a.j != null) {
                            for (let w = a.j.count - 1; w >= 0; w--) {
                                if (!l.contains(a.j._inner[w])) {
                                    let x = a.j._inner[w];
                                    let y = false;
                                    if (x != a) {
                                        y = x.j.remove(a);
                                    }
                                    else {
                                    }
                                    a.j.removeAt(w);
                                    if (x.j.count == 0 && y) {
                                        this.az--;
                                    }
                                    let z = j.remove(x);
                                    if (j.count == 0 && z) {
                                        this.az--;
                                    }
                                }
                            }
                        }
                        for (let aa = 0; aa < m.count; aa++) {
                            let ab = m._inner[aa];
                            if (!j.contains(ab)) {
                                if (a.j == null) {
                                    a.j = new List$1(LabelCollisionInfo.$, 0);
                                }
                                a.j.add(ab);
                                if (j.count == 0) {
                                    this.az++;
                                }
                                j.add_1(ab);
                                if (ab != a) {
                                    if (ab.j == null) {
                                        ab.j = new List$1(LabelCollisionInfo.$, 0);
                                    }
                                    if (ab.j.count == 0) {
                                        this.az++;
                                    }
                                    ab.j.add(a);
                                }
                            }
                        }
                    }
                }
                return this.az;
            }
        }
        s(a, b) {
            if ((((a.i != null) && a.i.e == a) || ((a.i != null) && a.i.d == a)) && (((a.i != null) && a.i.e == b) || ((a.i != null) && a.i.d == b))) {
                return true;
            }
            return false;
        }
        get a3() {
            return this.az;
        }
        bm(a, b, c) {
            this.bu = c;
            let d = new QuadTree(a, b, this.h.i, this.h.f);
            let e = this.ag.c;
            while (e != null) {
                d.c(e.c.e);
                if (e.c.d != null) {
                    d.c(e.c.d);
                }
                if (e.c.c != null) {
                    d.c(e.c.c);
                }
                e = e.a;
            }
            this.h = d;
            this.aw = a;
            this.at = b;
        }
        ab() {
            let a = this.ag.c;
            while (a != null) {
                a.c.f = false;
                if (a.c.c != null) {
                    if (this.x(a.c.c)) {
                        return true;
                    }
                }
                a = a.a;
            }
            return false;
        }
        t() {
            let a = this.ag.c;
            let b = this.aj;
            while (a != null) {
                if (a.c.c != null) {
                    if (this.x(a.c.c)) {
                        let c = a.c;
                        for (let d = 0; d < b.count; d++) {
                            let e = this.l(c, b._inner[d], 0, 0, false);
                            if (e != null) {
                                return true;
                            }
                        }
                    }
                }
                a = a.a;
            }
            return false;
        }
        a4() {
            let a = this.ag.c;
            while (a != null) {
                a.c.f = false;
                a = a.a;
            }
            a = this.ag.c;
            while (a != null) {
                let b = a.c.c;
                let c = a.c.e;
                let d = this.h.e(c);
                for (let e = 0; e < d.count; e++) {
                    let f = d._inner[e].l.c;
                    while (f != null) {
                        let g = f.c;
                        if (this.s(c, g)) {
                            f = f.a;
                            continue;
                        }
                        if (this.v(c, g)) {
                            let h = f.c.i;
                            if (!this.ae(f.c, a.c)) {
                                h.f = true;
                                a.c.f = true;
                            }
                        }
                        f = f.a;
                    }
                }
                if (b == null) {
                    return -1;
                }
                d = this.h.e(b);
                for (let i = 0; i < d.count; i++) {
                    let j = d._inner[i].l.c;
                    while (j != null) {
                        let k = j.c;
                        if (this.v(b, k)) {
                            let l = j.c.i;
                            if (!this.ae(j.c, a.c)) {
                                l.f = true;
                                a.c.f = true;
                            }
                        }
                        j = j.a;
                    }
                }
                a = a.a;
            }
            let m = 0;
            let n = 1.7976931348623157E+308;
            let o = -1.7976931348623157E+308;
            let p = 1.7976931348623157E+308;
            let q = -1.7976931348623157E+308;
            a = this.ag.c;
            while (a != null) {
                if (a.c.f) {
                    m++;
                    if (!isNaN_(a.c.s) && !isNaN_(a.c.t)) {
                        n = Math.min(n, a.c.s);
                        o = Math.max(o, a.c.s);
                        p = Math.min(p, a.c.t);
                        q = Math.max(q, a.c.t);
                    }
                }
                a = a.a;
            }
            this.am = Math.abs(o - n);
            this.an = Math.abs(q - p);
            return m;
        }
        bj(a) {
            let b = this.ag.c;
            if (this.ad(this.g)) {
                this.bk();
                return;
            }
            while (b != null) {
                let c = b.c;
                if (c.c == null || a) {
                    let d = false;
                    if (a) {
                        let e = this.h.a(c.s, c.t, 1.5);
                        let f = this.c(e);
                        if (f != null) {
                            for (let g = 0; g < f.length; g++) {
                                let h = f[g];
                                let i = this.l(c, h, 0, 0, false);
                                if (i == null) {
                                    continue;
                                }
                                let j = this.h.e(i);
                                d = true;
                                for (let k = 0; k < j.count; k++) {
                                    let l = j._inner[k].l.c;
                                    while (l != null) {
                                        let m = l.c;
                                        if (this.v(i, m)) {
                                            d = false;
                                            break;
                                        }
                                        l = l.a;
                                    }
                                }
                                if (d) {
                                    c.c = i;
                                    c.b = h;
                                    c.k = c.c.d;
                                    c.l = c.c.e;
                                    this.h.c(c.c);
                                    this.ah.add(c);
                                    break;
                                }
                            }
                        }
                    }
                    if (!d) {
                        d = this.z(c, false);
                    }
                    if (!d) {
                        if (c.c == null) {
                            this.z(c, true);
                        }
                    }
                }
                b = b.a;
            }
        }
        bk() {
            let a = this.ag.c;
            let b = new List$1(LabelCollisionInfo.$, 0);
            let c = new List$1(LabelCollisionInfo.$, 0);
            let d = new List$1(LabelCollisionInfo.$, 0);
            let e = new List$1(LabelCollisionInfo.$, 0);
            let f = this.g;
            while (a != null) {
                let g = a.c;
                let h = g.a;
                if (g.a == null || h.type != 2) {
                    g.h = true;
                    this.ah.add(g);
                    a = a.a;
                    continue;
                }
                let i = h;
                let j = this.bt(i);
                let k = g;
                let l = k.p + k.p;
                let m = k.o + k.o;
                let n = (i.startAngle + i.endAngle) / 2;
                n = MathUtil.f(n);
                let o = Rect.empty;
                if (f == 3 || f == 7) {
                    let p = this.i(k);
                    if (p != null) {
                        if (this.u(p, g)) {
                            if (f == 7) {
                                this.bf(k);
                                a = a.a;
                                continue;
                            }
                        }
                        else {
                            k.b = 8;
                            this.bq(k, p);
                            d.add(k);
                            this.ah.add(k);
                            this.h.c(p);
                            a = a.a;
                            continue;
                        }
                    }
                    if (k.h) {
                        a = a.a;
                        continue;
                    }
                }
                if (f == 5 || f == 3) {
                    let q = this.j(k);
                    if (q != null) {
                        if (this.u(q, g)) {
                            if (f == 5) {
                                this.bf(k);
                                a = a.a;
                                continue;
                            }
                        }
                        else {
                            k.b = 9;
                            e.add(k);
                            this.bq(k, q);
                            this.ah.add(k);
                            this.h.c(q);
                            a = a.a;
                            continue;
                        }
                    }
                    if (k.h) {
                        a = a.a;
                        continue;
                    }
                }
                let r = this.k(k);
                if (r != null) {
                    k.b = 10;
                    this.bq(k, r);
                    this.ah.add(k);
                    this.h.c(k.c);
                    if (k.c.d < j.x) {
                        c.add(k);
                    }
                    else {
                        b.add(k);
                    }
                }
                else {
                    this.bf(k);
                }
                a = a.a;
            }
            this.bo(b);
            this.bn(b);
            this.bo(c);
            this.bn(c);
        }
        bq(a, b) {
            let c = new List$1(LabelCollisionInfo.$, 0);
            let d = new HashSet$1(LabelCollisionInfo.$, 0);
            if (!this.r) {
                c = new List$1(LabelCollisionInfo.$, 0);
            }
            this.bb(a, b, c, d);
            let e = a.e;
            if (e == null) {
                e = new QuadTreeBoundingBox();
                e.i = a;
                e.d = a.s;
                e.e = a.t;
                e.g = a.r;
                e.f = a.q;
                this.h.c(e);
                a.e = e;
            }
            this.bb(a, e, c, d);
            this.af(a, c);
            a.c = b;
            a.k = a.c.d;
            a.l = a.c.e;
        }
        bb(a, b, c, d) {
            if (this.u(b, a)) {
                let e = this.h.e(b);
                for (let f = 0; f < e.count; f++) {
                    let g = e._inner[f].l.c;
                    while (g != null) {
                        let h = g.c;
                        if (this.v(b, h)) {
                            if (!this.r) {
                                if (!d.contains(h.i)) {
                                    c.add(h.i);
                                    d.add_1(h.i);
                                }
                            }
                        }
                        g = g.a;
                    }
                }
            }
        }
        k(a) {
            let b = null;
            let c = a.a;
            if (c == null) {
                return null;
            }
            let d = this.bt(c);
            let e = a.p + a.p;
            let f = a.o + a.o;
            let g = (c.startAngle + c.endAngle) / 2;
            g = MathUtil.f(g);
            let h = GeometryUtil.q(d, g, c.radius + this.au);
            a.b = 10;
            this.bg(a);
            let i = LabelCollisionManager.ap(d, c.radius, c, a);
            let j = GeometryUtil.q(d, g, i);
            let k = new Rect(0, j.x - e / 2, j.y - f / 2, e, f);
            if (g < 90 && g >= 0) {
                k = new Rect(0, h.x, h.y, e, f);
            }
            else if (g < 180 && g >= 90) {
                k = new Rect(0, h.x - e, h.y, e, f);
            }
            else if (g < 270 && g >= 180) {
                k = new Rect(0, h.x - e, h.y - f, e, f);
            }
            else {
                k = new Rect(0, h.x, h.y - f, e, f);
            }
            if (k.y < 0) {
                k = new Rect(0, k.x, 0, k.width, k.height);
            }
            if (k.bottom > this.bu.bottom) {
                k = new Rect(0, k.x, this.bu.bottom - k.height, k.width, k.height);
            }
            b = ((() => {
                let $ret = new QuadTreeBoundingBox();
                $ret.d = k.left + k.width / 2;
                $ret.e = k.top + k.height / 2;
                $ret.g = k.width / 2;
                $ret.f = k.height / 2;
                $ret.i = a;
                return $ret;
            })());
            return b;
        }
        bg(a) {
            if (a == null) {
                return;
            }
            let b = !a.h;
            if (b) {
                return;
            }
            a.h = false;
            this.ah.add(a);
        }
        bf(a) {
            if (a == null) {
                return;
            }
            let b = !a.h;
            if (!b) {
                return;
            }
            a.h = true;
            this.ah.add(a);
        }
        j(a) {
            let b = null;
            let c = a.a;
            if (c == null) {
                return null;
            }
            let d = this.bt(c);
            let e = a.p + a.p;
            let f = a.o + a.o;
            let g = (c.startAngle + c.endAngle) / 2;
            g = MathUtil.f(g);
            let h = LabelCollisionManager.ap(d, c.radius, c, a);
            let i = GeometryUtil.q(d, g, h);
            let j = new Rect(0, i.x - e / 2, i.y - f / 2, e, f);
            let k = this.aa(c, j, a, d);
            if (k || this.g == 5) {
                a.b = 9;
                if (!k && this.g == 5) {
                    this.bf(a);
                }
                else {
                    b = ((() => {
                        let $ret = new QuadTreeBoundingBox();
                        $ret.d = j.left + j.width / 2;
                        $ret.e = j.top + j.height / 2;
                        $ret.g = j.width / 2;
                        $ret.f = j.height / 2;
                        $ret.i = a;
                        return $ret;
                    })());
                    this.bg(a);
                    return b;
                }
            }
            return b;
        }
        i(a) {
            let b = null;
            let c = a.a;
            if (c == null) {
                return null;
            }
            let d = this.bt(c);
            let e = a.p + a.p;
            let f = a.o + a.o;
            let g = (c.startAngle + c.endAngle) / 2;
            g = MathUtil.f(g);
            let h = c.innerRadius;
            let i = GeometryUtil.q(d, g, c.radius - (c.radius - h) / 2);
            let j = new Rect(0, i.x - a.p, i.y - a.o, e, f);
            let k = this.aa(c, j, a, i);
            if (k || this.g == 7) {
                a.b = 8;
                if (!k && this.g == 7) {
                    this.bf(a);
                }
                else {
                    this.bg(a);
                    b = ((() => {
                        let $ret = new QuadTreeBoundingBox();
                        $ret.d = j.left + j.width / 2;
                        $ret.e = j.top + j.height / 2;
                        $ret.g = j.width / 2;
                        $ret.f = j.height / 2;
                        $ret.i = a;
                        return $ret;
                    })());
                    return b;
                }
            }
            return b;
        }
        static ap(a, b, c, d) {
            let e = (c.startAngle + c.endAngle / 2);
            e = MathUtil.f(e);
            let f = e * Math.PI / 180;
            let g = 1;
            let h = GeometryUtil.q({ $type: Point_$type, x: 0, y: 0 }, e, g);
            let i = LabelCollisionManager.bs(c, d);
            let j = Math.sqrt(i.x * i.x + i.y * i.y);
            let k = { $type: Point_$type, x: i.x / j, y: i.y / j };
            let l = LabelCollisionManager.aq(h, k);
            let m = j / b;
            let n = l * m;
            let o = n;
            let p = GeometryUtil.i({ $type: Point_$type, x: 0, y: 0 }, h);
            let q = GeometryUtil.i({ $type: Point_$type, x: 0, y: 0 }, k);
            if (q < p) {
                o = -o;
            }
            let r = GeometryUtil.q(a, (f + o) * 180 / Math.PI, b);
            let s = LabelCollisionManager.br(c, d, r);
            let t = Math.sqrt((s.x - a.x) * (s.x - a.x) + (s.y - a.y) * (s.y - a.y));
            let u = t - 5;
            return u;
        }
        static br(a, b, c) {
            let d = (a.startAngle + a.endAngle / 2);
            d = MathUtil.f(d);
            let e = GeometryUtil.m(d);
            let f = b.p;
            let g = b.o;
            if (e >= 0 && e <= 90) {
                return { $type: Point_$type, x: c.x - f, y: c.y - g };
            }
            if (e >= 90 && e <= 180) {
                return { $type: Point_$type, x: c.x + f, y: c.y - g };
            }
            if (e >= 180 && e <= 270) {
                return { $type: Point_$type, x: c.x + f, y: c.y + g };
            }
            return { $type: Point_$type, x: c.x - f, y: c.y + g };
        }
        static aq(a, b) {
            let c = a.x * b.x + a.y * b.y;
            let d = Math.sqrt(a.x * a.x + a.y * a.y);
            let e = Math.sqrt(b.x * b.x + b.y * b.y);
            let f = Math.acos(c / (d * e));
            return f;
        }
        static bs(a, b) {
            let c = (a.startAngle + a.endAngle / 2);
            c = MathUtil.f(c);
            let d = GeometryUtil.m(c);
            let e = b.p;
            let f = b.o;
            if (d >= 0 && d <= 90) {
                return { $type: Point_$type, x: e, y: f };
            }
            if (d >= 90 && d <= 180) {
                return { $type: Point_$type, x: -e, y: f };
            }
            if (d >= 180 && d <= 270) {
                return { $type: Point_$type, x: -e, y: -f };
            }
            return { $type: Point_$type, x: e, y: -f };
        }
        static ao(a, b, c, d) {
            let e = MathUtil.g(a - b, c - d);
            let f = Math.asin((c - d) / e) * 180 / Math.PI;
            if (a < b) {
                f = 180 - f;
            }
            if (a > b) {
                f = 360 + f;
            }
            if (f == 360) {
                f = 0;
            }
            return GeometryUtil.m(f);
        }
        bo(a) {
            let b = new Array(a.count);
            for (let c = 0; c < a.count; c++) {
                let d = a._inner[c];
                let e = d.a;
                let f = this.bt(e);
                let g = (e.startAngle + e.endAngle) / 2;
                g = MathUtil.f(g);
                let h = GeometryUtil.q(f, g, e.radius);
                b[c] = h.y;
            }
            for (let i = 0; i < a.count; i++) {
                for (let j = i + 1; j < a.count; j++) {
                    if (b[i] >= b[j]) {
                        let k = a._inner[i];
                        a._inner[i] = a._inner[j];
                        a._inner[j] = k;
                        let l = b[i];
                        b[i] = b[j];
                        b[j] = l;
                    }
                }
            }
        }
        bn(a) {
            if (a.count == 0) {
                return;
            }
            let b = this.bu.width;
            let c = this.bu.height;
            let d = a.count;
            let e = a._inner[0].a;
            let f = e.radius;
            let g = this.bt(e);
            let h = true;
            let i = 0;
            let j = Number.POSITIVE_INFINITY;
            let k = Number.NEGATIVE_INFINITY;
            i = this.a4();
            let l = 0;
            for (let m of fromEnum(a)) {
                j = Math.min(j, m.o + m.o);
                k = Math.max(k, m.o + m.o);
                l += m.o + m.o;
            }
            if (l > c) {
                h = false;
            }
            if (h && i > 0) {
                for (let n = 0; n < d - 1; n++) {
                    for (let o = n + 1; o < d; o++) {
                        let p = a._inner[n];
                        let q = a._inner[o];
                        if (p.c.c(q.c)) {
                            let r = q.c;
                            let s = r.e - r.f;
                            let t = r.g + r.g;
                            let u = MathUtil.f(this.ar(q));
                            let v = Math.min(p.c.e + p.c.f + 0.01, c - j);
                            let w = this.au + f;
                            let x = Math.abs(g.y - (s + j / 2));
                            let y = Math.sqrt(Math.abs(w * w - x * x));
                            let z = GeometryUtil.m(u);
                            if (z > 90 && z < 270) {
                                y = (t + y) * -1;
                            }
                            let aa = g.x + y;
                            let ab = ((() => {
                                let $ret = new QuadTreeBoundingBox();
                                $ret.d = aa + r.g;
                                $ret.e = v + r.f;
                                $ret.g = r.g;
                                $ret.f = r.f;
                                $ret.i = q;
                                return $ret;
                            })());
                            if (q.c != null) {
                                this.h.d(q.c);
                            }
                            this.bq(q, ab);
                            this.h.c(q.c);
                        }
                    }
                }
                for (let ac = d - 1; ac > 0; ac--) {
                    for (let ad = ac - 1; ad >= 0; ad--) {
                        let ae = a._inner[ac];
                        let af = a._inner[ad];
                        if (ae.c.c(af.c)) {
                            let ag = af.c;
                            let ah = ag.e - ag.f;
                            let ai = ag.g + ag.g;
                            let aj = Math.max((ae.c.e - ae.c.f) - j - 0.01, 0);
                            let ak = this.au + f;
                            let al = Math.abs(g.y - (ah + j / 2));
                            let am = Math.sqrt(Math.abs(ak * ak - al * al));
                            let an = GeometryUtil.m(MathUtil.f(this.ar(af)));
                            if (an > 90 && an < 270) {
                                am = (ai + am) * -1;
                            }
                            let ao = g.x + am;
                            let ap = ((() => {
                                let $ret = new QuadTreeBoundingBox();
                                $ret.d = ao + ag.g;
                                $ret.e = aj + ag.f;
                                $ret.g = ag.g;
                                $ret.f = ag.f;
                                $ret.i = af;
                                return $ret;
                            })());
                            if (af.c != null) {
                                this.h.d(af.c);
                            }
                            this.bq(af, ap);
                            this.h.c(af.c);
                        }
                    }
                }
            }
            for (let aq of fromEnum(a)) {
                let ar = aq.c.j();
                if (ar.left > b || ar.right < 0) {
                    this.bf(aq);
                    if (aq.c != null) {
                        this.h.d(aq.c);
                    }
                    this.af(aq, null);
                    aq.c = null;
                }
                else if (ar.left < 0) {
                    this.bf(aq);
                    if (aq.c != null) {
                        this.h.d(aq.c);
                    }
                    this.af(aq, null);
                    aq.c = null;
                }
                else if (ar.right > b) {
                    this.bf(aq);
                    if (aq.c != null) {
                        this.h.d(aq.c);
                    }
                    this.af(aq, null);
                    aq.c = null;
                }
            }
        }
        ar(a) {
            let b = a.a;
            return (b.startAngle + b.endAngle) / 2;
        }
        static av(a) {
            let b = Math.round(a * Math.pow(10, 5)) / Math.pow(10, 5);
            return b;
        }
        ac(a) {
            return LabelCollisionManager.av(MathUtil.f(Math.abs(a.endAngle - a.startAngle))) == 360;
        }
        aa(a, b, c, d) {
            if (a == null) {
                return false;
            }
            let e = this.bt(a);
            let f = Math.min(a.startAngle, a.endAngle);
            let g = Math.max(a.startAngle, a.endAngle);
            f = MathUtil.f(f);
            g = MathUtil.f(g);
            let h = false;
            let i = false;
            let j = GeometryUtil.q(e, f, a.radius);
            let k = GeometryUtil.q(e, g, a.radius);
            f = LabelCollisionManager.ao(j.x, e.x, j.y, e.y);
            g = LabelCollisionManager.ao(k.x, e.x, k.y, e.y);
            if (this.ac(a)) {
                i = true;
            }
            let l;
            l = MathUtil.g(b.right - e.x, b.top - e.y);
            if (l > a.radius) {
                return false;
            }
            l = MathUtil.g(b.right - e.x, b.bottom - e.y);
            if (l > a.radius) {
                return false;
            }
            l = MathUtil.g(b.left - e.x, b.top - e.y);
            if (l > a.radius) {
                return false;
            }
            l = MathUtil.g(b.left - e.x, b.bottom - e.y);
            if (l > a.radius) {
                return false;
            }
            if (i) {
                return true;
            }
            if (f > g) {
                f = f - 360;
                h = true;
            }
            let m;
            m = LabelCollisionManager.ao(b.right, e.x, b.top, e.y);
            if (h && m > 180 && m < 360) {
                m = m - 360;
            }
            if (m < f || m > g) {
                return false;
            }
            m = LabelCollisionManager.ao(b.right, e.x, b.bottom, e.y);
            if (h && m > 180 && m < 360) {
                m = m - 360;
            }
            if (m < f || m > g) {
                return false;
            }
            m = LabelCollisionManager.ao(b.left, e.x, b.top, e.y);
            if (h && m > 180 && m < 360) {
                m = m - 360;
            }
            if (m < f || m > g) {
                return false;
            }
            m = LabelCollisionManager.ao(b.left, e.x, b.bottom, e.y);
            if (h && m > 180 && m < 360) {
                m = m - 360;
            }
            if (m < f || m > g) {
                return false;
            }
            return true;
        }
        bt(a) {
            return { $type: Point_$type, x: a.centerX, y: a.centerY };
        }
        ad(a) {
            return a == 7 || a == 3 || a == 6 || a == 4 || a == 5;
        }
        z(a, b) {
            let c = this.aj;
            a.h = false;
            let d = false;
            if (!d) {
                for (let e = 0; e < c.count; e++) {
                    let f = this.l(a, c._inner[e], 0, 0, b);
                    if (a.h) {
                        break;
                    }
                    if (f == null) {
                        continue;
                    }
                    d = !this.u(f, a);
                    let g = a.e;
                    if (g == null) {
                        g = new QuadTreeBoundingBox();
                        g.i = a;
                        g.d = a.s;
                        g.e = a.t;
                        g.g = a.r;
                        g.f = a.q;
                        this.h.c(g);
                        a.e = g;
                    }
                    if (a.a != null) {
                        if (a.d == null) {
                            a.d = QuadTreeBoundingBox.a(a.a.boundingBox);
                            a.d.i = a;
                            this.h.c(a.d);
                        }
                    }
                    let h = this.u(g, a);
                    d = d && !h;
                    if (d) {
                        a.c = f;
                        if (c._inner[e] != 11) {
                            a.b = c._inner[e];
                        }
                        a.k = a.c.d;
                        a.l = a.c.e;
                        this.h.c(a.c);
                        this.ah.add(a);
                        break;
                    }
                }
            }
            if (!d && !a.h) {
                let i = Number.POSITIVE_INFINITY;
                let j = 0;
                let k = null;
                for (let l = 0; l < c.count; l++) {
                    let m = 0;
                    let n = null;
                    let o = new HashSet$1(LabelCollisionInfo.$, 0);
                    if (!this.r) {
                        n = new List$1(LabelCollisionInfo.$, 0);
                    }
                    let p = this.l(a, c._inner[l], 0, 0, b);
                    if (p == null) {
                        continue;
                    }
                    let q = this.h.e(p);
                    for (let r = 0; r < q.count; r++) {
                        let s = q._inner[r].l.c;
                        while (s != null) {
                            let t = s.c;
                            if (this.v(p, t)) {
                                m += p.h(t);
                                if (!this.r) {
                                    if (!o.contains(t.i)) {
                                        n.add(t.i);
                                        o.add_1(t.i);
                                    }
                                }
                            }
                            s = s.a;
                        }
                    }
                    let u = a.e;
                    if (u == null) {
                        u = new QuadTreeBoundingBox();
                        u.i = a;
                        u.d = a.s;
                        u.e = a.t;
                        u.g = a.r;
                        u.f = a.q;
                    }
                    q = this.h.e(u);
                    for (let v = 0; v < q.count; v++) {
                        let w = q._inner[v].l.c;
                        while (w != null) {
                            let x = w.c;
                            if (this.ae(w.c, a)) {
                                w = w.a;
                                continue;
                            }
                            if (this.s(u, x)) {
                                w = w.a;
                                continue;
                            }
                            if (this.v(u, x)) {
                                if (!this.r) {
                                    if (!o.contains(x.i)) {
                                        n.add(x.i);
                                        o.add_1(x.i);
                                    }
                                }
                            }
                            w = w.a;
                        }
                    }
                    if (m < i) {
                        let y = this.l(a, c._inner[l], 0, 0, b);
                        if (y != null) {
                            i = m;
                            j = l;
                            if (!this.r) {
                                k = n;
                            }
                        }
                    }
                }
                let z = this.l(a, c._inner[j], 0, 0, b);
                a.f = true;
                if (!this.af(a, k)) {
                    return false;
                }
                a.h = false;
                a.c = z;
                a.b = c._inner[j];
                a.k = a.c.d;
                a.l = a.c.e;
                this.h.c(a.c);
                this.ah.add(a);
            }
            return d;
        }
        af(a, b) {
            if (!this.r) {
                if (a.j != null && a.j.count > 0) {
                    this.az--;
                    for (let c = a.j.count - 1; c >= 0; c--) {
                        let d = a.j._inner[c];
                        if (d.j != null) {
                            d.j.remove(a);
                            if (d.j.count == 0) {
                                this.az--;
                            }
                        }
                    }
                }
                if (b == null) {
                    a.f = false;
                    this.ah.add(a);
                    return false;
                }
                a.j = b;
                if (b.count > 0) {
                    this.az++;
                }
                for (let e = 0; e < b.count; e++) {
                    let f = b._inner[e];
                    if (f.j == null) {
                        f.j = new List$1(LabelCollisionInfo.$, 0);
                    }
                    if (!f.j.contains(a)) {
                        if (f.j.count == 0) {
                            this.az++;
                        }
                        f.j.add(a);
                    }
                }
            }
            return true;
        }
        ae(a, b) {
            if (a.i == b && (a.i == null || a.i.d != a)) {
                return true;
            }
            return false;
        }
        u(a, b) {
            let c = false;
            let d = this.h.e(a);
            for (let e = 0; e < d.count; e++) {
                let f = d._inner[e].l.c;
                while (f != null) {
                    if (this.ae(f.c, b)) {
                        f = f.a;
                        continue;
                    }
                    let g = f.c;
                    if (this.s(a, g)) {
                        f = f.a;
                        continue;
                    }
                    if (this.v(a, g)) {
                        c = true;
                        break;
                    }
                    f = f.a;
                }
            }
            return c;
        }
        c(a) {
            let b = new Array(3);
            let c = GeometryUtil.i({ $type: Point_$type, x: 0, y: 0 }, { $type: Point_$type, x: a.a, y: a.b });
            c = c * 180 / Math.PI;
            c = GeometryUtil.m(c);
            if ((c >= 0 && c <= 22.5) || (c >= 337.5 && c <= 360)) {
                b[0] = 2;
                b[1] = 5;
                b[2] = 7;
            }
            if (c >= 22.5 && c <= 67.5) {
                b[0] = 2;
                b[1] = 7;
                b[2] = 3;
            }
            if (c >= 67.5 && c <= 112.5) {
                b[0] = 7;
                b[1] = 3;
                b[2] = 6;
            }
            if (c >= 112.5 && c <= 157.5) {
                b[0] = 3;
                b[1] = 6;
                b[2] = 1;
            }
            if (c >= 157.5 && c <= 202.5) {
                b[0] = 6;
                b[1] = 1;
                b[2] = 4;
            }
            if (c >= 202.5 && c <= 247.5) {
                b[0] = 1;
                b[1] = 4;
                b[2] = 0;
            }
            if (c >= 247.5 && c <= 292.5) {
                b[0] = 4;
                b[1] = 0;
                b[2] = 5;
            }
            if (c >= 292.5 && c <= 337.5) {
                b[0] = 0;
                b[1] = 5;
                b[2] = 2;
            }
            if (this.o) {
                let d = 0;
                for (let e = 0; e < b.length; e++) {
                    if (this.a[b[e]]) {
                        d++;
                    }
                }
                if (d == 0) {
                    return null;
                }
                let f = 0;
                let g = new Array(d);
                for (let h = 0; h < b.length; h++) {
                    if (this.a[b[h]]) {
                        g[f] = b[h];
                        f++;
                    }
                }
                b = g;
            }
            return b;
        }
        f(a, b) {
            let c = false;
            let d = null;
            let e = new QuadTreeBoundingBox();
            e.d = a.x;
            e.e = a.y;
            e.g = 5;
            e.f = 5;
            let f = this.h.e(e);
            for (let g = 0; g < f.count; g++) {
                let h = f._inner[g].l.c;
                while (h != null) {
                    let i = h.c;
                    if (i.i.e == i) {
                        h = h.a;
                        continue;
                    }
                    if (this.v(e, i)) {
                        d = i;
                        c = true;
                        break;
                    }
                    h = h.a;
                }
            }
            if (c) {
                return d.i;
            }
            else {
                return null;
            }
        }
        y() {
            return this.am < 10 && this.an < 10;
        }
    }
    LabelCollisionManager.$t = markType(LabelCollisionManager, 'LabelCollisionManager');
    LabelCollisionManager.a6 = new Random(0);
    return LabelCollisionManager;
})();
