// igniteui-angular-core/esm2015/lib/ListPanelTemplateItemUpdatingEventArgsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ListPanelTemplateItemUpdatingEventArgsDescription = /*@__PURE__*/ (() => {
    class ListPanelTemplateItemUpdatingEventArgsDescription extends Description {
        constructor() {
            super();
            this.h = null;
            this.j = null;
        }
        get_type() {
            return "ListPanelTemplateItemUpdatingEventArgs";
        }
        get type() {
            return this.get_type();
        }
        get model() {
            return this.h;
        }
        set model(a) {
            this.h = a;
            this.g("Model");
        }
        get contentRef() {
            return this.j;
        }
        set contentRef(a) {
            this.j = a;
            this.g("ContentRef");
        }
    }
    ListPanelTemplateItemUpdatingEventArgsDescription.$t = markType(ListPanelTemplateItemUpdatingEventArgsDescription, 'ListPanelTemplateItemUpdatingEventArgsDescription', Description.$);
    return ListPanelTemplateItemUpdatingEventArgsDescription;
})();
