// igniteui-angular-core/esm2015/lib/NativeUIGrid.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIContainer } from "./NativeUIContainer";
import { NativeUIBoxSizing_$type } from "./NativeUIBoxSizing";
import { enumGetBox, EnumUtil, markType } from "./type";
import { NativeUIPlaceContent_$type } from "./NativeUIPlaceContent";
/**
 * @hidden
 */
export let NativeUIGrid = /*@__PURE__*/ (() => {
    class NativeUIGrid extends NativeUIContainer {
        constructor() {
            super(...arguments);
            this.br = null;
        }
        get_e() {
            return 7;
        }
        get by() {
            return this.bp.getRowDefinitionsCount(this.am);
        }
        get bv() {
            return this.bp.getColumnDefinitionsCount(this.am);
        }
        get bp() {
            return this.d;
        }
        get bq() {
            let a = this.ag(38);
            return EnumUtil.getEnumValue(NativeUIBoxSizing_$type, a);
        }
        set bq(a) {
            this.a2(38, enumGetBox(NativeUIBoxSizing_$type, a));
        }
        get bt() {
            let a = this.ag(72);
            return EnumUtil.getEnumValue(NativeUIPlaceContent_$type, a);
        }
        set bt(a) {
            this.a2(72, enumGetBox(NativeUIPlaceContent_$type, a));
        }
        get bu() {
            return this.ag(37);
        }
        set bu(a) {
            this.a2(37, a);
        }
        b1(a, b, c) {
            this.bp.addRowDefinition(this.am, a, b, c);
        }
        b9(a, b, c) {
            this.bp.updateRowDefinition(this.am, a, b, c);
        }
        bz(a, b, c) {
            this.bp.addColumnDefinition(this.am, a, b, c);
        }
        b0(a, b, c, d) {
            this.bp.addColumnDefinitionWithMin(this.am, a, b, c, d);
        }
        b8(a, b, c) {
            this.bp.updateColumnDefinition(this.am, a, b, c);
        }
        b2(a, b, c) {
            this.bp.insertColumnDefinition(this.am, a, b, c);
        }
        b3(a, b, c, d) {
            this.bp.insertColumnDefinitionWithMin(this.am, a, b, c, d);
        }
        b7(a, b) {
            this.bp.setRow(this.am, a, b);
        }
        b6(a, b) {
            this.bp.setColumn(this.am, a, b);
        }
        bx(a) {
            return this.bp.getRow(this.am, a);
        }
        bw(a) {
            return this.bp.getColumn(this.am, a);
        }
        b5(a) {
            this.bp.removeRowDefinition(this.am, a);
        }
        b4(a) {
            this.bp.removeColumnDefinition(this.am, a);
        }
        get bs() {
            return this.br;
        }
        set bs(a) {
            let b = this.br;
            this.br = a;
            if (this.br != b) {
                if (b != null) {
                    this.a0(1, b);
                }
                if (this.br != null) {
                    this.ap(1, this.br);
                }
            }
        }
    }
    NativeUIGrid.$t = markType(NativeUIGrid, 'NativeUIGrid', NativeUIContainer.$);
    return NativeUIGrid;
})();
