// igniteui-angular-core/esm2015/lib/NativeUIPopup.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIContainer } from "./NativeUIContainer";
import { NativeUIPopupAnimationType_$type } from "./NativeUIPopupAnimationType";
import { NativeUIPopupPointerPosition_$type } from "./NativeUIPopupPointerPosition";
import { typeGetValue, enumGetBox, EnumUtil, markType } from "./type";
/**
 * @hidden
 */
export let NativeUIPopup = /*@__PURE__*/ (() => {
    class NativeUIPopup extends NativeUIContainer {
        constructor() {
            super(...arguments);
            this.bt = null;
            this.br = null;
        }
        get bp() {
            return this.d;
        }
        get b2() {
            return this.ag(22);
        }
        get b1() {
            return this.ag(59);
        }
        get by() {
            return this.ag(60);
        }
        get bw() {
            return this.ag(25);
        }
        set bw(a) {
            this.a2(25, a);
        }
        get b5() {
            return typeGetValue(this.ag(26));
        }
        set b5(a) {
            this.a2(26, a);
        }
        get bx() {
            return this.ag(39);
        }
        set bx(a) {
            this.a2(39, a);
        }
        get bq() {
            return EnumUtil.getEnumValue(NativeUIPopupAnimationType_$type, this.ag(35));
        }
        set bq(a) {
            this.a2(35, enumGetBox(NativeUIPopupAnimationType_$type, a));
        }
        get cb() {
            return this.ag(23);
        }
        set cb(a) {
            this.a2(23, a);
        }
        get b3() {
            return this.ag(17);
        }
        set b3(a) {
            this.a2(17, a);
        }
        get_u() {
            return this.ag(36);
        }
        set_u(a) {
            this.a2(36, a);
        }
        get bz() {
            return this.ag(37);
        }
        set bz(a) {
            this.a2(37, a);
        }
        get b6() {
            return typeGetValue(this.ag(40));
        }
        set b6(a) {
            this.a2(40, a);
        }
        get b0() {
            return this.ag(41);
        }
        set b0(a) {
            this.a2(41, a);
        }
        get bv() {
            return EnumUtil.getEnumValue(NativeUIPopupPointerPosition_$type, this.ag(43));
        }
        set bv(a) {
            this.a2(43, enumGetBox(NativeUIPopupPointerPosition_$type, a));
        }
        get cc() {
            return this.ag(42);
        }
        set cc(a) {
            this.a2(42, a);
        }
        get b4() {
            return this.ag(61);
        }
        set b4(a) {
            this.a2(61, a);
        }
        get bu() {
            return this.bt;
        }
        set bu(a) {
            let b = this.bt;
            this.bt = a;
            if (this.bt != b) {
                if (b != null) {
                    this.a0(5, b);
                }
                if (this.bt != null) {
                    this.ap(5, this.bt);
                }
            }
        }
        get bs() {
            return this.br;
        }
        set bs(a) {
            let b = this.br;
            this.br = a;
            if (this.br != b) {
                if (b != null) {
                    this.a0(6, b);
                }
                if (this.br != null) {
                    this.ap(6, this.br);
                }
            }
        }
        ca(a, b, c) {
            this.bp.showRelativeToExclusionRect(this.am, a, b, c);
        }
        b7() {
            this.bp.close(this.am);
        }
        b8() {
            this.bp.notifyContentSizeChanged(this.am);
        }
        b9(a) {
            this.bp.notifyExclusionRectChanged(this.am, a);
        }
    }
    NativeUIPopup.$t = markType(NativeUIPopup, 'NativeUIPopup', NativeUIContainer.$);
    return NativeUIPopup;
})();
