// igniteui-angular-core/esm2015/lib/NativeUIAbsolutePanel.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIContainer } from "./NativeUIContainer";
import { markType } from "./type";
/**
 * @hidden
 */
export let NativeUIAbsolutePanel = /*@__PURE__*/ (() => {
    class NativeUIAbsolutePanel extends NativeUIContainer {
        get_e() {
            return 9;
        }
        get bp() {
            return this.d;
        }
        bs(a, b) {
            this.bp.setX(this.am, a, b);
        }
        bt(a, b) {
            this.bp.setY(this.am, a, b);
        }
        bq(a) {
            return this.bp.getX(this.am, a);
        }
        br(a) {
            return this.bp.getY(this.am, a);
        }
    }
    NativeUIAbsolutePanel.$t = markType(NativeUIAbsolutePanel, 'NativeUIAbsolutePanel', NativeUIContainer.$);
    return NativeUIAbsolutePanel;
})();
