// igniteui-angular-core/esm2015/lib/NativeUIDropDown.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { enumGetBox, EnumUtil, markType } from "./type";
/**
 * @hidden
 */
export let NativeUIDropDown = /*@__PURE__*/ (() => {
    class NativeUIDropDown extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bj = null;
        }
        get_e() {
            return 4;
        }
        get bi() {
            let a = this.ag(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        }
        set bi(a) {
            this.a2(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        }
        get bk() {
            return this.bj;
        }
        set bk(a) {
            let b = this.bj;
            this.bj = a;
            if (this.bj != b) {
                if (b != null) {
                    this.a0(3, b);
                }
                if (this.bj != null) {
                    this.ap(3, this.bj);
                }
            }
        }
        get bm() {
            let a = this.ag(9);
            return a;
        }
        set bm(a) {
            this.a2(9, a);
        }
        get bl() {
            let a = this.ag(8);
            return a;
        }
        set bl(a) {
            this.a2(8, a);
        }
    }
    NativeUIDropDown.$t = markType(NativeUIDropDown, 'NativeUIDropDown', NativeUIComponent.$);
    return NativeUIDropDown;
})();
