// igniteui-angular-core/esm2015/lib/NativeUIButton.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { enumGetBox, EnumUtil, markType } from "./type";
import { NativeUIButtonDisplayType_$type } from "./NativeUIButtonDisplayType";
/**
 * @hidden
 */
export let NativeUIButton = /*@__PURE__*/ (() => {
    class NativeUIButton extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bj = null;
        }
        get_e() {
            return 1;
        }
        get bn() {
            let a = this.ag(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        }
        set bn(a) {
            this.a2(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        }
        get bl() {
            let a = this.ag(0);
            return a == null ? 0 : EnumUtil.getEnumValue(NativeUIButtonDisplayType_$type, a);
        }
        set bl(a) {
            this.a2(0, enumGetBox(NativeUIButtonDisplayType_$type, a));
        }
        get bw() {
            let a = this.ag(3);
            return a;
        }
        set bw(a) {
            this.a2(3, a);
        }
        get bm() {
            return this.ag(54);
        }
        set bm(a) {
            this.a2(54, a);
        }
        get bs() {
            return this.ag(4);
        }
        set bs(a) {
            this.a2(4, a);
        }
        get bq() {
            return this.ag(5);
        }
        set bq(a) {
            this.a2(5, a);
        }
        get bp() {
            return this.ag(6);
        }
        set bp(a) {
            this.a2(6, a);
        }
        get bu() {
            return this.ag(32);
        }
        set bu(a) {
            this.a2(32, a);
        }
        get by() {
            return this.ag(24);
        }
        set by(a) {
            this.a2(24, a);
        }
        get bz() {
            return this.ag(55);
        }
        set bz(a) {
            this.a2(55, a);
        }
        get bt() {
            return this.ag(57);
        }
        set bt(a) {
            this.a2(57, a);
        }
        get br() {
            return this.ag(56);
        }
        set br(a) {
            this.a2(56, a);
        }
        get bo() {
            return this.ag(58);
        }
        set bo(a) {
            this.a2(58, a);
        }
        get bx() {
            return this.ag(17);
        }
        set bx(a) {
            this.a2(17, a);
        }
        get bi() {
            return this.ag(11);
        }
        set bi(a) {
            this.a2(11, a);
        }
        get bv() {
            return this.ag(83);
        }
        set bv(a) {
            this.a2(83, a);
        }
        get bk() {
            return this.bj;
        }
        set bk(a) {
            let b = this.bj;
            this.bj = a;
            if (this.bj != b) {
                if (b != null) {
                    this.a0(1, b);
                }
                if (this.bj != null) {
                    this.ap(1, this.bj);
                }
            }
        }
    }
    NativeUIButton.$t = markType(NativeUIButton, 'NativeUIButton', NativeUIComponent.$);
    return NativeUIButton;
})();
