// igniteui-angular-core/esm2015/lib/NativeUICheckboxList.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { enumGetBox, EnumUtil, markType } from "./type";
import { ControlDisplayDensity_$type } from "./ControlDisplayDensity";
/**
 * @hidden
 */
export let NativeUICheckboxList = /*@__PURE__*/ (() => {
    class NativeUICheckboxList extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bk = null;
            this.bl = null;
            this.bm = null;
        }
        get_e() {
            return 16;
        }
        get bn() {
            return this.bk;
        }
        set bn(a) {
            let b = this.bk;
            this.bk = a;
            if (this.bk != b) {
                if (b != null) {
                    this.a0(2, b);
                }
                if (this.bk != null) {
                    this.ap(2, this.bk);
                }
            }
        }
        get bo() {
            return this.bl;
        }
        set bo(a) {
            let b = this.bl;
            this.bl = a;
            if (this.bl != b) {
                if (b != null) {
                    this.a0(18, b);
                }
                if (this.bl != null) {
                    this.ap(18, this.bl);
                }
            }
        }
        get bp() {
            return this.bm;
        }
        set bp(a) {
            let b = this.bm;
            this.bm = a;
            if (this.bm != b) {
                if (b != null) {
                    this.a0(19, b);
                }
                if (this.bm != null) {
                    this.ap(19, this.bm);
                }
            }
        }
        get br() {
            let a = this.ag(8);
            return a;
        }
        set br(a) {
            this.a2(8, a);
        }
        get bu() {
            let a = this.ag(23);
            return a;
        }
        set bu(a) {
            this.a2(23, a);
        }
        get bs() {
            let a = this.ag(75);
            return a;
        }
        set bs(a) {
            this.a2(75, a);
        }
        get bt() {
            let a = this.ag(74);
            return a;
        }
        set bt(a) {
            this.a2(74, a);
        }
        get bj() {
            let a = this.ag(1);
            return EnumUtil.getEnumValue(ControlDisplayDensity_$type, a);
        }
        set bj(a) {
            this.a2(1, enumGetBox(ControlDisplayDensity_$type, a));
        }
        get bq() {
            let a = this.ag(76);
            return a;
        }
        set bq(a) {
            this.a2(76, a);
        }
        get bi() {
            let a = this.ag(77);
            return a;
        }
        set bi(a) {
            this.a2(77, a);
        }
        get by() {
            let a = this.ag(33);
            return a;
        }
        set by(a) {
            this.a2(33, a);
        }
        get bv() {
            let a = this.ag(24);
            return a;
        }
        set bv(a) {
            this.a2(24, a);
        }
        get bw() {
            let a = this.ag(78);
            return a;
        }
        set bw(a) {
            this.a2(78, a);
        }
        get bx() {
            let a = this.ag(79);
            return a;
        }
        set bx(a) {
            this.a2(79, a);
        }
    }
    NativeUICheckboxList.$t = markType(NativeUICheckboxList, 'NativeUICheckboxList', NativeUIComponent.$);
    return NativeUICheckboxList;
})();
