// igniteui-angular-core/esm2015/lib/NativeUICheckbox.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { markType } from "./type";
/**
 * @hidden
 */
export let NativeUICheckbox = /*@__PURE__*/ (() => {
    class NativeUICheckbox extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.bi = null;
        }
        get_e() {
            return 6;
        }
        get bk() {
            let a = this.ag(4);
            return a;
        }
        set bk(a) {
            this.a2(4, a);
        }
        get bl() {
            let a = this.ag(7);
            return a;
        }
        set bl(a) {
            this.a2(7, a);
        }
        get bm() {
            let a = this.ag(17);
            return a;
        }
        set bm(a) {
            this.a2(17, a);
        }
        get bp() {
            return this.ag(65);
        }
        set bp(a) {
            this.a2(65, a);
        }
        get bn() {
            return this.ag(66);
        }
        set bn(a) {
            this.a2(66, a);
        }
        get bq() {
            return this.ag(67);
        }
        set bq(a) {
            this.a2(67, a);
        }
        get bo() {
            return this.ag(68);
        }
        set bo(a) {
            this.a2(68, a);
        }
        get bj() {
            return this.bi;
        }
        set bj(a) {
            let b = this.bi;
            this.bi = a;
            if (this.bi != b) {
                if (b != null) {
                    this.a0(2, b);
                }
                if (this.bi != null) {
                    this.ap(2, this.bi);
                }
            }
        }
    }
    NativeUICheckbox.$t = markType(NativeUICheckbox, 'NativeUICheckbox', NativeUIComponent.$);
    return NativeUICheckbox;
})();
