// igniteui-angular-core/esm2015/lib/NumericAngleAxisDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NumericAxisBaseDescription } from "./NumericAxisBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let NumericAngleAxisDescription = /*@__PURE__*/ (() => {
    class NumericAngleAxisDescription extends NumericAxisBaseDescription {
        constructor() {
            super();
            this.e1 = 0;
            this.e3 = null;
        }
        get_type() {
            return "NumericAngleAxis";
        }
        get startAngleOffset() {
            return this.e1;
        }
        set startAngleOffset(a) {
            this.e1 = a;
            this.g("StartAngleOffset");
        }
        get labelMode() {
            return this.e3;
        }
        set labelMode(a) {
            this.e3 = a;
            this.g("LabelMode");
        }
    }
    NumericAngleAxisDescription.$t = markType(NumericAngleAxisDescription, 'NumericAngleAxisDescription', NumericAxisBaseDescription.$);
    return NumericAngleAxisDescription;
})();
