// igniteui-angular-core/esm2015/lib/PieSliceCollisionGeometry.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CollisionGeometry } from "./CollisionGeometry";
import { CollisionRect } from "./CollisionRect";
import { Point_$type, markType } from "./type";
import { GeometryUtil } from "./GeometryUtil";
/**
 * @hidden
 */
export let PieSliceCollisionGeometry = /*@__PURE__*/ (() => {
    class PieSliceCollisionGeometry extends CollisionGeometry {
        constructor(a, b, c, d, e, f) {
            super();
            this.h = 0;
            this.i = 0;
            this.l = 0;
            this.m = 0;
            this.j = 0;
            this.g = CollisionRect.empty;
            this.k = 0;
            this.h = a;
            this.i = b;
            this.m = e;
            this.j = f;
            this.l = c;
            this.k = d;
        }
        get_type() {
            return 2;
        }
        get type() {
            return this.get_type();
        }
        get centerX() {
            return this.h;
        }
        get centerY() {
            return this.i;
        }
        get radius() {
            return this.l;
        }
        get startAngle() {
            return this.m;
        }
        get endAngle() {
            return this.j;
        }
        get_boundingBox() {
            if (this.g.isEmpty) {
                let a = GeometryUtil.u({ $type: Point_$type, x: this.h, y: this.i }, this.m, this.j, this.l);
                this.g = new CollisionRect(0, a);
            }
            return this.g;
        }
        get boundingBox() {
            return this.get_boundingBox();
        }
        get innerRadius() {
            return this.k;
        }
        d(a) {
            let b = { $type: Point_$type, x: a.left, y: a.top };
            let c = { $type: Point_$type, x: a.right, y: a.top };
            let d = { $type: Point_$type, x: a.right, y: a.bottom };
            let e = { $type: Point_$type, x: a.left, y: a.bottom };
            let f = { $type: Point_$type, x: this.h, y: this.i };
            let g = this.l;
            let h = this.k;
            if (GeometryUtil.e(b, f, this.m, this.j, g, h) || GeometryUtil.e(c, f, this.m, this.j, g, h) || GeometryUtil.e(d, f, this.m, this.j, g, h) || GeometryUtil.e(e, f, this.m, this.j, g, h)) {
                return true;
            }
            return false;
        }
    }
    PieSliceCollisionGeometry.$t = markType(PieSliceCollisionGeometry, 'PieSliceCollisionGeometry', CollisionGeometry.$);
    return PieSliceCollisionGeometry;
})();
