// igniteui-angular-core/esm2015/lib/PopupView_combined.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, runOn, markType, INotifyPropertyChanged_$type, PropertyChangedEventArgs, enumGetBox, fromEnum } from "./type";
import { Size } from "./Size";
import { ElevationHelper } from "./ElevationHelper";
import { CornerRadius } from "./CornerRadius";
import { NativeUI } from "./NativeUI";
import { PopupPointerPosition_$type } from "./PopupPointerPosition";
import { Popup_OuterAnimationInfo } from "./Popup_OuterAnimationInfo";
import { Popup_InnerAnimationInfo } from "./Popup_InnerAnimationInfo";
import { AnimationKeyFrameEffect } from "./AnimationKeyFrameEffect";
import { AnimationKeyFrameEffectTiming } from "./AnimationKeyFrameEffectTiming";
import { KeyFrameAnimationFactory } from "./KeyFrameAnimationFactory";
import { Rect } from "./Rect";
import { PopupAnimationType_$type } from "./PopupAnimationType";
import { stringFormat } from "./stringExtended";
import { isNaN_, truncate } from "./number";
import { stringEmpty } from "./string";
import { NativeUIInputsFactory } from "./NativeUIInputsFactory";
import { OnClosedEventArgs } from "./OnClosedEventArgs";
import { MouseEventArgs } from "./MouseEventArgs";
import { FocusEventArgs } from "./FocusEventArgs";
import { PopupAnimation } from "./PopupAnimation";
import { IKeyFrameAnimation_$type } from "./IKeyFrameAnimation";
import { List$1 } from "./List$1";
import { AnimationKeyFrameProperty } from "./AnimationKeyFrameProperty";
import { TranslateTransform } from "./TranslateTransform";
import { OnPopupEventArgs } from "./OnPopupEventArgs";
import { DeviceUtils } from "./DeviceUtils";
import { PopupVisualModelExport } from "./PopupVisualModelExport";
import { AppearanceHelper } from "./AppearanceHelper";
import { BrushUtil } from "./BrushUtil";
/**
 * @hidden
 */
export let PopupView = /*@__PURE__*/ (() => {
    class PopupView extends Base {
        constructor() {
            super(...arguments);
            this.ag = null;
            this.f = null;
            this.ao = null;
            this.an = null;
            this.am = null;
            this.aj = null;
            this.al = null;
            this.ak = null;
            this.aq = null;
            this.v = 0;
            this.w = 0;
            this.u = 0;
            this.m = false;
            this.x = 0;
            this.s = 0;
            this.t = 0;
            this.q = 0;
            this.r = 0;
            this.ai = null;
            this.ap = null;
            this.br = new Size(1, 0, 0);
            this.k = true;
            this.p = false;
        }
        get af() {
            return this.g.a3;
        }
        get ac() {
            return this.af / 2;
        }
        get ab() {
            return this.af * 1.414;
        }
        get aa() {
            return this.ab / 2;
        }
        get ad() {
            return (this.bs.width / 2 - this.ac);
        }
        get ae() {
            return (this.bs.height / 2 - this.ac);
        }
        get g() {
            return this.f;
        }
        get ah() {
            return this.ag;
        }
        a6() {
            this.ao.setStyleProperty("height", "0%");
            this.ao.setStyleProperty("width", "0%");
            this.aj = this.ag.createElement("div");
            this.be(this.g.aq);
            this.aj.setStyleProperty("z-index", "10000");
            this.aj.setStyleProperty("inset", "unset");
            this.aj.setStyleProperty("padding", "0px");
            this.aj.setStyleProperty("border-width", "0px");
            this.al = this.ag.createElement("div");
            this.al.setStyleProperty("height", 20 * 1.414 / 2 + "px");
            this.ak = this.ag.createElement("div");
            this.ak.setStyleProperty("position", "relative");
            this.ak.setStyleProperty("width", this.af + "px");
            this.ak.setStyleProperty("height", this.af + "px");
            this.ak.setStyleProperty("z-index", "1");
            this.ak.setStyleProperty("clip-path", stringFormat("path(\"M0,0l{0},{0}l0,-{0}l-{0},0Z\")", this.af));
            this.al.append(this.ak);
            this.an = this.ag.createElement("div");
            this.an.setStyleProperty("position", "relative");
            this.an.setStyleProperty("overflow", "hidden");
            this.an.setAttribute("key", "popupContentAreaOuter");
            this.bf(this.g.ar);
            this.aj.listen("mouseenter", runOn(this, this.a7));
            this.aj.listen("mouseleave", runOn(this, this.a8));
            this.aj.listen("focusin", runOn(this, this.a4));
            this.aj.listen("focusout", runOn(this, this.a5));
            this.am = this.ag.createElement("div");
            this.am.setAttribute("key", "popupContentAreaInner");
            this.an.append(this.am);
            this.aj.append(this.an);
        }
        bn() {
            let a = this.g;
            let b = ElevationHelper.e(a.a6, a.cx, a.cw, a.cv);
            if (this.l()) {
                this.aj.setStyleProperty("box-shadow", b);
            }
            else {
                this.an.setStyleProperty("box-shadow", b);
            }
        }
        bl() {
            if (this.g.cy != null) {
                this.an.setStyleProperty("background-color", this.g.cy._fill);
                this.ak.setStyleProperty("background-color", this.g.cy._fill);
            }
            if (this.g.c2 != null) {
                this.ak.setStyleProperty("background-color", this.g.c2._fill);
            }
        }
        bm() {
            if (!isNaN_(this.g.a0)) {
                NativeUI.o(this.an, new CornerRadius(0, this.g.a0));
            }
        }
        bo() {
            if (null == this.ao) {
                return;
            }
            this.be(this.g.aq);
            this.bn();
            this.bl();
            this.bm();
        }
        a7(a) {
            this.g.cd();
        }
        a8(a) {
            this.g.ce();
        }
        a4(a) {
            this.g.cb();
        }
        a5(a) {
            this.g.cc(a.originalEvent.relatedTarget);
        }
        o() {
            return (HTMLElement.prototype.hasOwnProperty("popover"));
        }
        l() {
            return this.o() && this.g.ax;
        }
        at() {
            if (null != this.aj) {
                this.bo();
                this.an.setStyleProperty("height", "0px");
                let a = this.aj.getNativeElement();
                if (false == this.m) {
                    this.al.remove();
                    if (this.g.at) {
                        let b = this.g.t();
                        if (b == 1) {
                            this.aj.setStyleProperty("display", "block");
                            let c = this.aj.getNativeElement();
                            let d = this.al.getNativeElement();
                            let e = this.an.getNativeElement();
                            c.insertBefore(d, e);
                        }
                        else if (b == 2) {
                            this.aj.setStyleProperty("display", "block");
                            this.aj.append(this.al);
                        }
                        else if (b == 3) {
                            this.aj.setStyleProperty("display", "flex");
                            let f = this.aj.getNativeElement();
                            let g = this.al.getNativeElement();
                            let h = this.an.getNativeElement();
                            f.insertBefore(g, h);
                            this.al.setStyleProperty("height", "unset");
                            this.al.setStyleProperty("width", 20 * 1.414 / 2 + "px");
                        }
                        else if (b == 4) {
                            this.aj.setStyleProperty("display", "flex");
                            this.aj.append(this.al);
                            this.al.setStyleProperty("height", "unset");
                            this.al.setStyleProperty("width", 20 * 1.414 / 2 + "px");
                        }
                    }
                    if (this.g.bd != null) {
                        this.g.bd.appendChild(a);
                    }
                    else {
                        document.body.appendChild(a);
                    }
                    this.bh(this.g.ax);
                    if (this.l()) {
                        a.showPopover();
                    }
                    this.m = true;
                    if (this.k) {
                        this.a1();
                    }
                }
            }
        }
        au(a, b) {
            this.aj.setStyleProperty("top", PopupView.ar(a.f));
            this.an.setStyleProperty("height", PopupView.as(a.h));
            this.am.setStyleProperty("top", PopupView.as(b.b));
            this.am.setStyleProperty("opacity", "1");
            this.am.setStyleProperty("transform", stringEmpty());
            this.x = 0;
            if (this.p) {
                this.bk();
            }
            if (this.g.aw) {
                this.ap.observe(this.ai.getNativeElement());
            }
            else {
                this.ap.disconnect();
            }
            if (this.g.an) {
                this.bg(this.g.as);
            }
            return {
                p0: a,
                p1: b
            };
        }
        av(a) {
            this.x = a;
            if (this.g.an) {
                this.bg(false);
            }
            this.bi(this.bs.width, a);
        }
        bc() {
            if (null != this.aj && this.m) {
                let a = this.aj.getNativeElement();
                if (this.g.bd != null) {
                    this.g.bd.removeChild(a);
                }
                else {
                    document.body.removeChild(a);
                }
                this.m = false;
            }
        }
        aw(a) {
            this.ai = a;
            this.am.append(this.ai);
            this.ap = new ResizeObserver(runOn(this, this.a3));
        }
        a3(a) {
            if (this.m) {
                if (this.g.ao() && this.g.p == 0) {
                    return;
                }
                this.g.b8();
            }
        }
        j(a) {
            let b = this.aj.getNativeElement();
            return b.contains(a);
        }
        bb(a) {
            this.ai = null;
            this.ap.disconnect();
            this.ap = null;
            a.remove();
        }
        d(a) {
            return new AnimationKeyFrameEffect(0, this.am, 0, null, a);
        }
        e(a) {
            return new AnimationKeyFrameEffect(0, this.an, 0, null, a);
        }
        ay() {
            if (KeyFrameAnimationFactory.h == null) {
                KeyFrameAnimationFactory.h = this.ah;
            }
        }
        c(a) {
            return new AnimationKeyFrameEffect(0, this.aj, 0, null, a);
        }
        b(a) {
            return new AnimationKeyFrameEffect(0, this.an, 0, null, a);
        }
        a(a) {
            return new AnimationKeyFrameEffect(0, this.aj, 0, null, a);
        }
        i() {
            return 12;
        }
        h() {
            return 11;
        }
        n() {
            return this.g.bd != null && !this.g.aq && !this.l();
        }
        get bs() {
            if (this.k) {
                this.a1();
            }
            return this.br;
        }
        y(a, b, c) {
            let d;
            switch (b) {
                case 3:
                    d = (!this.n() ? a.left : 0) - (this.bs.width + (this.g.at ? this.aa : 0));
                    break;
                case 4:
                    d = !this.n() ? a.right : a.width;
                    break;
                case 0:
                case 2:
                case 1:
                default:
                    switch (c) {
                        case 2:
                            d = (!this.n() ? a.right : a.width) - this.bs.width;
                            break;
                        case 3:
                            d = (!this.n() ? a.right : a.width) - (a.width / 2) - (this.bs.width / 2);
                            break;
                        case 0:
                        case 1:
                        default:
                            d = !this.n() ? a.left : 0;
                            break;
                    }
                    break;
            }
            return d;
        }
        z(a, b, c) {
            let d;
            switch (b) {
                case 3:
                case 4:
                    switch (c) {
                        case 2:
                            d = (!this.n() ? a.bottom : 0) - this.bs.height;
                            break;
                        case 3:
                            d = (!this.n() ? a.top : -a.height) - (this.bs.height / 2) + (a.height / 2);
                            break;
                        case 0:
                        case 1:
                        default:
                            d = !this.n() ? a.top : -a.height;
                            break;
                    }
                    break;
                case 2:
                    d = (!this.n() ? a.top : -a.height) - (this.bs.height + (this.g.at ? this.aa : 0));
                    break;
                case 0:
                case 1:
                default:
                    d = !this.n() ? a.bottom : 0;
                    break;
            }
            return d;
        }
        static as(a) {
            return stringFormat("{0}px", a);
        }
        static ar(a) {
            return stringFormat("{0}px", a);
        }
        bj() {
            if (this.g.r == 1 || this.g.r == 2) {
                switch (this.g.j) {
                    case 3:
                        this.v = this.ad + (this.s - this.q);
                        break;
                }
            }
            else if (this.g.r == 3 || this.g.r == 4) {
                switch (this.g.j) {
                    case 3:
                        this.w = this.ae + (this.t - this.r);
                        break;
                }
            }
            this.ak.setStyleProperty("transform", "translate(" + this.v + "px, " + this.w + "px) rotate(" + this.u + "deg)");
        }
        ax(a, b, c, d) {
            this.s = a;
            this.t = b;
            if (this.g.r == 1 || this.g.r == 2) {
                let e = (window.innerWidth);
                let f = a;
                let g = f + c;
                if (f < 5) {
                    a = 5;
                }
                else if (g > e - 5) {
                    a = f - (g - (e - 5));
                }
            }
            else if (this.g.r == 3 || this.g.r == 4) {
                let h = (window.innerHeight);
                let i = b;
                let j = i + d;
                if (i < 5) {
                    b = 5;
                }
                else if (j > h - 5) {
                    b = i - (j - (h - 5));
                }
            }
            return {
                p0: a,
                p1: b
            };
        }
        bq(a, b, c) {
            let d = this.z(a, b, c);
            let e = this.y(a, b, c);
            if (this.g.bd == null) {
                let f = this.ax(e, d, this.bs.width, this.bs.height);
                e = f.p0;
                d = f.p1;
            }
            this.bd(e, d);
            if (b == 2 || c == 2) {
                let g = 10;
                while (g > 0) {
                    let h = this.z(a, b, c);
                    let i = this.y(a, b, c);
                    if (this.g.bd == null) {
                        let j = this.ax(i, h, this.bs.width, this.bs.height);
                        i = j.p0;
                        h = j.p1;
                    }
                    if (h == d && i == e) {
                        g = 0;
                        break;
                    }
                    d = h;
                    e = i;
                    this.bd(e, d);
                    g--;
                }
            }
            this.bj();
            return new Rect(0, e, d, this.bs.width, this.bs.height);
        }
        bp(a, b, c) {
            let d = new Rect(4);
            let target_ = a;
            if ((target_.getBoundingClientRect)) {
                let bRect_ = target_.getBoundingClientRect();
                d.left = (bRect_.left);
                d.top = (bRect_.top);
                d.width = (bRect_.width);
                d.height = (bRect_.height);
            }
            let e = this.bq(d, b, c);
            return e;
        }
        a1() {
            if (this.m) {
                this.an.setStyleProperty("width", stringEmpty());
                this.an.setStyleProperty("height", stringEmpty());
                this.aj.setStyleProperty("width", stringEmpty());
                this.aj.setStyleProperty("height", stringEmpty());
                this.aj.setStyleProperty("position", "absolute");
                this.br = new Size(1, this.am.outerWidth(), this.am.outerHeight());
                this.k = false;
                this.bk();
            }
            else {
                this.k = true;
            }
        }
        bd(a, b) {
            let c = stringFormat("{0}px", a);
            let d = stringFormat("{0}px", b);
            this.aj.setStyleProperty("left", c);
            this.aj.setStyleProperty("top", d);
            this.q = a;
            this.r = b;
        }
        a9(a) {
            if (a == null) {
                this.ag = null;
                if (this.aj != null) {
                    this.aj.unlistenAll();
                }
                return;
            }
            this.ag = a;
            let b = a.rootWrapper;
            this.az(b);
            this.ag.getPortal(this.am, "ChildContent", (c) => {
                let comp_ = c.componentRef;
                this.aq = comp_;
            }, true);
        }
        bk() {
            let a = new Size(1, this.bs.width, this.bs.height);
            if (!this.m || (!this.g.aw && !this.g.ap)) {
                a.height = 0;
            }
            let b = a.width;
            let c = a.height;
            let d = stringFormat("{0}px", b);
            let e = stringFormat("{0}px", c);
            this.bi(b, c);
            if (this.g.at) {
                switch (this.g.t()) {
                    case 2:
                        d = stringFormat("{0}px", b);
                        e = stringFormat("{0}px", this.aa + c);
                        break;
                    case 1:
                        d = stringFormat("{0}px", b);
                        e = stringFormat("{0}px", this.aa + c);
                        break;
                    case 3:
                        d = stringFormat("{0}px", this.aa + b);
                        e = stringFormat("{0}px", c);
                        break;
                    case 4:
                        d = stringFormat("{0}px", this.aa + b);
                        e = stringFormat("{0}px", c);
                        break;
                }
            }
            if (this.f.ao()) {
                if (this.g.ap || this.x == c) {
                    return;
                }
                this.f.b2();
                this.p = true;
                return;
            }
            this.p = false;
            let f = stringFormat("{0}px", b);
            let g = stringFormat("{0}px", c);
            this.an.setStyleProperty("width", f);
            this.an.setStyleProperty("height", g);
            if (!this.l()) {
                if (this.g.bd != null) {
                    e = "0px";
                }
                this.aj.setStyleProperty("width", d);
                this.aj.setStyleProperty("height", e);
            }
            this.be(this.g.aq);
        }
        bi(a, b) {
            if (this.g.at) {
                let c = 0;
                let d = 0;
                switch (this.g.t()) {
                    case 2:
                        c = 135;
                        d = -this.ac;
                        this.v = a / 2 - this.ac;
                        this.w = d;
                        break;
                    case 1:
                        c = -45;
                        d = this.aa - this.ac;
                        this.v = a / 2 - this.ac;
                        this.w = d;
                        break;
                    case 3:
                        c = -135;
                        d = this.aa - this.ac;
                        this.v = d;
                        this.w = b / 2 - this.ac;
                        break;
                    case 4:
                        c = 45;
                        d = -this.ac;
                        this.v = d;
                        this.w = b / 2 - this.ac;
                        break;
                }
                this.u = c;
                this.ak.setStyleProperty("transform", "translate(" + this.v + "px, " + this.w + "px) rotate(" + this.u + "deg)");
            }
        }
        az(a) {
            this.ao = a;
            this.a6();
        }
        a0(a) {
            this.f = a;
        }
        a2() {
        }
        ba() {
            this.at();
            if (this.g.aw) {
                return;
            }
            this.bk();
        }
        bg(a) {
            if (this.aj != null) {
                this.aj.setStyleProperty("pointer-events", a ? "auto" : "none");
            }
        }
        bf(a) {
            if (this.aj != null) {
                this.aj.setAttribute("tabindex", a ? "-1" : "");
            }
        }
        be(a) {
            if (this.aj != null) {
                this.aj.setStyleProperty("position", a ? "fixed" : this.g.bd != null ? "relative" : "absolute");
            }
        }
        bh(a) {
            if (this.aj != null && this.o()) {
                let b = this.aj.getNativeElement();
                if (a) {
                    b.popover = "manual";
                }
                else {
                    b.popover = undefined;
                }
            }
        }
    }
    PopupView.$t = markType(PopupView, 'PopupView');
    return PopupView;
})();
/**
 * @hidden
 */
export let Popup = /*@__PURE__*/ (() => {
    class Popup extends Base {
        constructor() {
            super();
            this.q = 0;
            this.i = 0;
            this.a1 = 0;
            this.c5 = null;
            this.c3 = null;
            this.bb = null;
            this.onClosed = null;
            this.onPopup = null;
            this.propertyChanged = null;
            this.popupMouseEnter = null;
            this.popupMouseLeave = null;
            this.popupGotFocus = null;
            this.popupLostFocus = null;
            this.w = null;
            this.af = false;
            this.a7 = 50;
            this.ak = true;
            this.o = 0;
            this.a5 = -1;
            this.a4 = Popup.a9;
            this.av = false;
            this.cs = Popup.c1;
            this.cr = Popup.c0;
            this.cq = Popup.cz;
            this.ct = null;
            this.ay = 0;
            this.ag = true;
            this.ac = false;
            this.ah = false;
            this.s = 0;
            this.cu = null;
            this.az = 20;
            this.ai = false;
            this.ad = false;
            this.ae = false;
            this.aj = false;
            this.k = null;
            NativeUIInputsFactory.n();
            let a = this.v();
            this.w = a;
            a.a0(this);
        }
        get c6() {
            return this.c5;
        }
        get c4() {
            return this.c3;
        }
        get r() {
            return this.q;
        }
        set r(a) {
            this.q = a;
        }
        get j() {
            return this.i;
        }
        get a2() {
            return this.a1;
        }
        get bd() {
            return this.bb;
        }
        cf(a, b, c) {
            if (this.propertyChanged != null) {
                this.propertyChanged(this, new PropertyChangedEventArgs(a));
            }
            this.ci(a, b, c);
        }
        cd() {
            if (this.popupMouseEnter != null) {
                this.popupMouseEnter(this, new MouseEventArgs());
            }
        }
        ce() {
            if (this.popupMouseLeave != null) {
                this.popupMouseLeave(this, new MouseEventArgs());
            }
        }
        cb() {
            if (this.popupGotFocus != null) {
                this.popupGotFocus(this, new FocusEventArgs());
            }
        }
        cc(a) {
            if (this.popupLostFocus != null) {
                this.popupLostFocus(this, ((() => {
                    let $ret = new FocusEventArgs();
                    $ret.a = a;
                    return $ret;
                })()));
            }
        }
        get ar() {
            return this.af;
        }
        set ar(a) {
            let b = this.af;
            this.af = a;
            if (b != this.af) {
                this.cf("IsFocusable", b, this.af);
            }
        }
        get a8() {
            return this.a7;
        }
        set a8(a) {
            let b = this.a7;
            this.a7 = a;
            if (b != this.a7) {
                this.cf("AnimationDuration", b, a);
            }
        }
        get al() {
            return this.ak;
        }
        set al(a) {
            let b = this.ak;
            this.ak = a;
            if (b != this.ak) {
                this.cf("AnimationEnabled", b, a);
            }
        }
        get p() {
            return this.o;
        }
        set p(a) {
            let b = this.o;
            this.o = a;
            if (b != this.o) {
                this.cf("AnimationType", enumGetBox(PopupAnimationType_$type, b), enumGetBox(PopupAnimationType_$type, this.o));
            }
        }
        get ba() {
            return this.a5;
        }
        set ba(a) {
            let b = this.a5;
            this.a5 = a;
            if (b != this.a5) {
                this.cf("Elevation", b, a);
            }
        }
        get a6() {
            return this.a4;
        }
        set a6(a) {
            let b = this.a4;
            this.a4 = a;
            if (b != this.a4) {
                this.cf("ActualElevation", b, a);
            }
        }
        get aw() {
            return this.av;
        }
        get cx() {
            return this.cs;
        }
        set cx(a) {
            let b = this.cs;
            this.cs = a;
            if (b != this.cs) {
                this.cf("ActualUmbraShadowColor", b, a);
            }
        }
        get cw() {
            return this.cr;
        }
        set cw(a) {
            let b = this.cr;
            this.cr = a;
            if (b != this.cr) {
                this.cf("ActualPenumbraShadowColor", b, a);
            }
        }
        get cv() {
            return this.cq;
        }
        set cv(a) {
            let b = this.cq;
            this.cq = a;
            if (b != this.cq) {
                this.cf("ActualAmbientShadowColor", b, a);
            }
        }
        get cy() {
            return this.ct;
        }
        set cy(a) {
            let b = this.ct;
            this.ct = a;
            if (b != this.ct) {
                this.cf("Background", b, this.ct);
            }
        }
        get a0() {
            return this.ay;
        }
        set a0(a) {
            let b = this.ay;
            this.ay = a;
            if (b != this.ay) {
                this.cf("CornerRadius", b, this.ay);
            }
        }
        get as() {
            return this.ag;
        }
        set as(a) {
            let b = this.ag;
            this.ag = a;
            if (b != this.ag) {
                this.cf("IsHitTestVisible", b, this.ag);
            }
        }
        get an() {
            return this.ac;
        }
        set an(a) {
            let b = this.ac;
            this.ac = a;
            if (b != this.ac) {
                this.cf("DisableHitTestDuringAnimation", b, this.ac);
            }
        }
        get at() {
            return this.ah;
        }
        set at(a) {
            let b = this.ah;
            this.ah = a;
            if (b != this.ah) {
                this.cf("IsPointerEnabled", b, this.ah);
            }
        }
        get u() {
            return this.s;
        }
        set u(a) {
            let b = this.s;
            this.s = a;
            if (b != this.s) {
                this.cf("PointerPosition", enumGetBox(PopupPointerPosition_$type, b), enumGetBox(PopupPointerPosition_$type, this.s));
            }
        }
        get c2() {
            return this.cu;
        }
        set c2(a) {
            let b = this.cu;
            this.cu = a;
            if (b != this.cu) {
                this.cf("PointerBackground", b, this.cu);
            }
        }
        get a3() {
            return this.az;
        }
        set a3(a) {
            let b = this.az;
            this.az = a;
            if (b != this.az) {
                this.cf("PointerSize", b, this.az);
            }
        }
        get au() {
            return this.ai;
        }
        set au(a) {
            this.ai = a;
        }
        get ap() {
            return this.ad;
        }
        set ap(a) {
            this.ad = a;
        }
        get aq() {
            return this.ae;
        }
        set aq(a) {
            let b = this.ae;
            this.ae = a;
            if (b != this.ae) {
                this.cf("IsFixed", b, this.ae);
            }
        }
        get ax() {
            return this.aj;
        }
        set ax(a) {
            let b = this.aj;
            this.aj = a;
            if (b != this.aj) {
                this.cf("UseTopLayer", b, this.aj);
            }
        }
        v() {
            return new PopupView();
        }
        ci(a, b, c) {
            switch (a) {
                case "IsShown":
                    if (false == this.aw) {
                        this.w.bc();
                        this.bb = null;
                    }
                    break;
                case "Elevation":
                    this.a6 = this.ba == -1 ? Popup.a9 : this.ba;
                    break;
                case "ActualElevation":
                case "ActualUmbraShadowColor":
                case "ActualPenumbraShadowColor":
                case "ActualAmbientShadowColor":
                case "Background":
                case "CornerRadius":
                    this.w.bo();
                    break;
                case "AnimationEnabled":
                    if (false == this.al) {
                        this.b2();
                    }
                    break;
                case "IsHitTestVisible":
                    this.w.bg(this.as);
                    break;
                case "IsFocusable":
                    this.w.bf(this.ar);
                    break;
                case "IsFixed":
                    this.w.be(this.aq);
                    break;
                case "UseTopLayer":
                    this.w.bh(this.ax);
                    break;
            }
        }
        b2() {
            if (this.ao()) {
                this.k.e();
            }
        }
        z(a) {
            let b = this.g(a);
            let c = a ? this.c() : this.b();
            let d = this.w.d(c);
            let e = new List$1(AnimationKeyFrameProperty.$, 0);
            let f = new List$1(AnimationKeyFrameProperty.$, 0);
            let g = new TranslateTransform();
            g.k = b.a;
            let h = new TranslateTransform();
            h.k = b.b;
            let i = new AnimationKeyFrameProperty(14, g);
            let j = new AnimationKeyFrameProperty(14, h);
            e.add(i);
            f.add(j);
            d.m(...e.toArray());
            d.m(...f.toArray());
            return KeyFrameAnimationFactory.f.c(d);
        }
        a(a) {
            let b = new List$1(IKeyFrameAnimation_$type, 0);
            let c = this.h(a);
            this.w.av(c.h);
            let d = a ? this.c() : this.b();
            let e = this.w.e(d);
            let f = new List$1(AnimationKeyFrameProperty.$, 0);
            let g = new List$1(AnimationKeyFrameProperty.$, 0);
            let h = new AnimationKeyFrameProperty(10, c.g);
            let i = new AnimationKeyFrameProperty(10, c.h);
            f.add(h);
            g.add(i);
            e.m(...f.toArray());
            e.m(...g.toArray());
            b.add(KeyFrameAnimationFactory.f.c(e));
            if (this.q == 2) {
                let j = c.c;
                let k = c.f;
                let l = this.w.c(d);
                let m = new List$1(AnimationKeyFrameProperty.$, 0);
                let n = new List$1(AnimationKeyFrameProperty.$, 0);
                let o = new AnimationKeyFrameProperty(12, j);
                let p = new AnimationKeyFrameProperty(12, k);
                m.add(o);
                n.add(p);
                l.m(...m.toArray());
                l.m(...n.toArray());
                b.add(KeyFrameAnimationFactory.f.c(l));
            }
            return b.toArray();
        }
        b() {
            let a = new AnimationKeyFrameEffectTiming();
            a.g = this.a8;
            a.b = 3;
            return a;
        }
        c() {
            let a = new AnimationKeyFrameEffectTiming();
            a.g = this.a8;
            a.b = 3;
            return a;
        }
        b7() {
            this.w.ay();
        }
        g(a) {
            let b = truncate(this.w.bs.height);
            if (b <= 0) {
                b = 500;
            }
            let c = a ? -b : 0;
            let d = a ? 0 : -b;
            if (this.q == 2) {
                c = 0;
                d = 0;
            }
            let e = ((() => {
                let $ret = new Popup_InnerAnimationInfo();
                $ret.a = c;
                $ret.b = d;
                return $ret;
            })());
            return e;
        }
        h(a) {
            let b = truncate(this.w.bs.height);
            if (b <= 0) {
                b = 500;
            }
            let c = a ? 0 : b;
            let d = a ? b : 0;
            let e = this.c5.top;
            let f = this.c5.top;
            let g = this.c5.left;
            let h = g;
            switch (this.q) {
                case 2:
                    e = a ? this.c5.bottom : this.c5.top;
                    f = a ? this.c5.top : this.c5.bottom;
                    g = this.c5.left;
                    h = g;
                    break;
                case 1:
                    e = this.c5.top;
                    f = this.c5.top;
                    g = this.c5.left;
                    h = g;
                    break;
                case 3:
                case 4:
                    e = this.c5.top;
                    f = e;
                    g = this.c5.left;
                    h = g;
                    break;
            }
            let i = ((() => {
                let $ret = new Popup_OuterAnimationInfo();
                $ret.g = c;
                $ret.h = d;
                $ret.c = e;
                $ret.f = f;
                $ret.a = g;
                $ret.d = h;
                return $ret;
            })());
            return i;
        }
        ao() {
            if (this.k == null) {
                return false;
            }
            return this.k.b;
        }
        b8() {
            this.w.a1();
            if (this.aw) {
                if (this.bd != null) {
                    this.c5 = this.w.bp(this.bd, this.q, this.i);
                }
                else {
                    this.c5 = this.w.bq(this.c3, this.q, this.i);
                }
            }
        }
        ch(a, b, c) {
            this.c3 = a;
            this.q = b;
            this.i = c;
            this.a1 = 0;
            this.w.ba();
            this.c5 = this.w.bq(a, b, c);
        }
        cg(a, b, c) {
            this.bb = a;
            this.q = b;
            this.i = c;
            this.a1 = 0;
            this.w.ba();
            this.c5 = this.w.bp(a, b, c);
        }
        cm(a, b, c) {
            if (this.av) {
                this.av = false;
                this.cf("IsShown", true, this.av);
            }
            this.ch(a, b, c);
            this.cn(true);
        }
        cl(a, b, c) {
            if (this.av) {
                this.av = false;
                this.cf("IsShown", true, this.av);
            }
            this.cg(a, b, c);
            this.cn(true);
        }
        cn(a) {
            let b = this.av;
            this.av = a;
            this.au = a;
            this.ap = !a;
            if (false == this.al) {
                let c = this.h(a);
                let d = this.g(a);
                let e = this.w.au(c, d);
                c = e.p0;
                d = e.p1;
                return;
            }
            if (this.ao()) {
                this.k.i();
                return;
            }
            this.b7();
            this.k = this.l(this.p, this.av);
            this.k.h((f) => {
                this.k.g();
                if (f) {
                    switch (this.p) {
                        case 0:
                            this.b4(this.av);
                            break;
                        case 1:
                            this.b3(this.av);
                            break;
                    }
                }
                this.cf("IsShown", b, this.av);
                if (this.av) {
                    if (this.onPopup != null) {
                        this.onPopup(this, new OnPopupEventArgs());
                    }
                }
                else {
                    if (this.onClosed != null) {
                        this.onClosed(this, new OnClosedEventArgs());
                    }
                }
                this.au = false;
                this.ap = false;
                let g = this.h(this.av);
                let h = this.g(this.av);
                let i = this.w.au(g, h);
                g = i.p0;
                h = i.p1;
            });
        }
        l(a, b) {
            switch (a) {
                case 0: return this.n(b);
                case 1: return this.m(b);
                default: return this.n(b);
            }
        }
        n(a) {
            let b = this.z(a);
            let c = this.a(a);
            let d = new PopupAnimation();
            d.c.add(b);
            for (let f = 0; f < c.length; f++) {
                let e = c[f];
                d.c.add(e);
            }
            return d;
        }
        b4(a) {
            let b = this.h(a);
            let c = this.g(a);
            for (let d of fromEnum(this.k.c)) {
                for (let e = 0; e < d.effect.f.length; e++) {
                    switch (d.effect.f[e]) {
                        case 14:
                            d.effect.g[e].c(d.effect.a4, ((() => {
                                let $ret = new TranslateTransform();
                                $ret.k = c.b;
                                return $ret;
                            })()));
                            break;
                        case 10:
                            d.effect.g[e].c(d.effect.a4, b.h);
                            break;
                        case 12:
                            d.effect.g[e].c(d.effect.a4, b.f);
                            break;
                    }
                }
            }
        }
        m(a) {
            let b = this.h(a);
            this.w.av(a ? b.h : b.g);
            let c = this.c();
            let d = this.w.c(c);
            let e = DeviceUtils.g(30);
            let f = this.w.i();
            let g = b.f - e;
            let h = b.f;
            switch (this.q) {
                case 2:
                    g = a ? b.f + e : b.c + e;
                    h = a ? b.f : b.c;
                    f = this.w.i();
                    break;
                case 1:
                    g = b.f - e;
                    h = b.f;
                    f = this.w.i();
                    break;
                case 3:
                    g = b.d + e;
                    h = b.d;
                    f = this.w.h();
                    break;
                case 4:
                    g = b.d - e;
                    h = b.d;
                    f = this.w.h();
                    break;
            }
            d.m(new AnimationKeyFrameProperty(f, a ? g : h));
            d.m(new AnimationKeyFrameProperty(f, a ? h : g));
            let i = this.w.a(c);
            i.m(new AnimationKeyFrameProperty(1, a ? 0 : 1));
            i.m(new AnimationKeyFrameProperty(1, a ? 1 : 0));
            let j = this.w.b(c);
            j.m(new AnimationKeyFrameProperty(10, (a ? b.h : b.g)));
            j.m(new AnimationKeyFrameProperty(10, (a ? b.h : b.g)));
            let k = new PopupAnimation();
            k.c.add(KeyFrameAnimationFactory.f.c(d));
            k.c.add(KeyFrameAnimationFactory.f.c(i));
            k.c.add(KeyFrameAnimationFactory.f.c(j));
            return k;
        }
        b3(a) {
            let b = this.h(a);
            let c = this.w.i();
            let d = NaN;
            let e = a ? 1 : 0;
            let f = a ? b.h : b.g;
            switch (this.q) {
                case 2:
                    c = this.w.i();
                    d = a ? b.f : b.c;
                    break;
                case 1:
                    c = this.w.i();
                    d = b.f;
                    break;
                case 3:
                    c = this.w.h();
                    d = b.d;
                    break;
                case 4:
                    c = this.w.h();
                    d = b.d;
                    break;
            }
            for (let g of fromEnum(this.k.c)) {
                for (let h = 0; h < g.effect.f.length; h++) {
                    if (g.effect.f[h] == c) {
                        g.effect.g[h].c(g.effect.a4, d);
                    }
                    else if (g.effect.f[h] == 1) {
                        g.effect.g[h].c(g.effect.a4, e);
                    }
                    else if (g.effect.f[h] == 10) {
                        g.effect.g[h].c(g.effect.a4, f);
                    }
                }
            }
        }
        contains(a) {
            return this.w.j(a);
        }
        appendPopupContent(a) {
            this.w.aw(a);
        }
        removePopupContent(a) {
            this.w.bb(a);
        }
        b5() {
            if (!this.av) {
                return;
            }
            this.cn(false);
        }
        destroy() {
            this.provideRenderer(null);
        }
        showRelativeToExclusionRect(a, b, c) {
            this.cm(a, b, c);
        }
        cp(a, b, c) {
            this.cl(a, b, c);
        }
        b9(a) {
            this.c3 = a;
            this.c5 = this.w.bq(a, this.q, this.i);
        }
        notifySizeChanged() {
        }
        provideRenderer(a) {
            this.w.a9(a);
        }
        bc() {
            return this.d();
        }
        bp() {
            let a = this.d();
            return a.f();
        }
        d() {
            let a = new PopupVisualModelExport();
            a.e = this.a6;
            a.d = this.aw;
            a.c = AppearanceHelper.a(this.cx);
            a.b = AppearanceHelper.a(this.cw);
            a.a = AppearanceHelper.a(this.cv);
            return a;
        }
        t() {
            let a = this.u;
            if (a == 0) {
                switch (this.r) {
                    case 4:
                        a = 3;
                        break;
                    case 3:
                        a = 4;
                        break;
                    case 2:
                        a = 2;
                        break;
                    case 1:
                        a = 1;
                        break;
                }
            }
            return a;
        }
    }
    Popup.$t = markType(Popup, 'Popup', Base.$, [INotifyPropertyChanged_$type]);
    Popup.a9 = 2;
    Popup.c1 = BrushUtil.h(66, 0, 0, 0);
    Popup.c0 = BrushUtil.h(30, 0, 0, 0);
    Popup.cz = BrushUtil.h(20, 0, 0, 0);
    return Popup;
})();
