// igniteui-angular-core/esm2015/lib/RadialPieSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnchoredRadialSeriesDescription } from "./AnchoredRadialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let RadialPieSeriesDescription = /*@__PURE__*/ (() => {
    class RadialPieSeriesDescription extends AnchoredRadialSeriesDescription {
        constructor() {
            super();
            this.hn = 0;
            this.ho = 0;
            this.hr = null;
        }
        get_type() {
            return "RadialPieSeries";
        }
        get radiusX() {
            return this.hn;
        }
        set radiusX(a) {
            this.hn = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.ho;
        }
        set radiusY(a) {
            this.ho = a;
            this.g("RadiusY");
        }
        get legendLabelMemberPath() {
            return this.hr;
        }
        set legendLabelMemberPath(a) {
            this.hr = a;
            this.g("LegendLabelMemberPath");
        }
    }
    RadialPieSeriesDescription.$t = markType(RadialPieSeriesDescription, 'RadialPieSeriesDescription', AnchoredRadialSeriesDescription.$);
    return RadialPieSeriesDescription;
})();
