// igniteui-angular-core/esm2015/lib/RadialBaseChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DomainChartDescription } from "./DomainChartDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let RadialBaseChartDescription = /*@__PURE__*/ (() => {
    class RadialBaseChartDescription extends DomainChartDescription {
        constructor() {
            super();
            this.t1 = null;
            this.uj = null;
            this.r6 = 0;
            this.r8 = 0;
            this.r7 = 0;
            this.r5 = 0;
            this.sn = 0;
            this.sp = 0;
            this.so = 0;
            this.sm = 0;
            this.t6 = null;
            this.uo = null;
            this.tz = null;
            this.t0 = null;
            this.sh = 0;
            this.sy = 0;
            this.sg = 0;
            this.sx = 0;
            this.sj = 0;
            this.s0 = 0;
            this.si = 0;
            this.sz = 0;
            this.sf = 0;
            this.sw = 0;
            this.uh = null;
            this.uz = null;
            this.t7 = null;
            this.up = null;
            this.ui = null;
            this.u0 = null;
            this.t5 = null;
            this.un = null;
            this.ua = null;
            this.us = null;
            this.r9 = 0;
            this.sq = 0;
            this.sa = 0;
            this.sr = 0;
            this.uc = null;
            this.uu = null;
            this.ud = null;
            this.uv = null;
            this.sb = 0;
            this.ss = 0;
            this.sc = 0;
            this.st = 0;
            this.ue = null;
            this.uw = null;
            this.sd = 0;
            this.su = 0;
            this.uf = null;
            this.ux = null;
            this.ub = null;
            this.ut = null;
            this.r4 = 0;
            this.sl = 0;
            this.r3 = 0;
            this.sk = 0;
            this.se = 0;
            this.sv = 0;
            this.rz = false;
            this.r0 = false;
            this.ug = null;
            this.uy = null;
            this.t3 = null;
            this.ul = null;
            this.t8 = null;
            this.uq = null;
            this.t9 = null;
            this.ur = null;
            this.um = null;
            this.t4 = null;
            this.t2 = null;
            this.rv = null;
            this.uk = null;
            this.rw = null;
        }
        get_type() {
            return "RadialBaseChart";
        }
        get angleAxisFormatLabelRef() {
            return this.t1;
        }
        set angleAxisFormatLabelRef(a) {
            this.t1 = a;
            this.g("AngleAxisFormatLabelRef");
        }
        get valueAxisFormatLabelRef() {
            return this.uj;
        }
        set valueAxisFormatLabelRef(a) {
            this.uj = a;
            this.g("ValueAxisFormatLabelRef");
        }
        get angleAxisLabelLeftMargin() {
            return this.r6;
        }
        set angleAxisLabelLeftMargin(a) {
            this.r6 = a;
            this.g("AngleAxisLabelLeftMargin");
        }
        get angleAxisLabelTopMargin() {
            return this.r8;
        }
        set angleAxisLabelTopMargin(a) {
            this.r8 = a;
            this.g("AngleAxisLabelTopMargin");
        }
        get angleAxisLabelRightMargin() {
            return this.r7;
        }
        set angleAxisLabelRightMargin(a) {
            this.r7 = a;
            this.g("AngleAxisLabelRightMargin");
        }
        get angleAxisLabelBottomMargin() {
            return this.r5;
        }
        set angleAxisLabelBottomMargin(a) {
            this.r5 = a;
            this.g("AngleAxisLabelBottomMargin");
        }
        get valueAxisLabelLeftMargin() {
            return this.sn;
        }
        set valueAxisLabelLeftMargin(a) {
            this.sn = a;
            this.g("ValueAxisLabelLeftMargin");
        }
        get valueAxisLabelTopMargin() {
            return this.sp;
        }
        set valueAxisLabelTopMargin(a) {
            this.sp = a;
            this.g("ValueAxisLabelTopMargin");
        }
        get valueAxisLabelRightMargin() {
            return this.so;
        }
        set valueAxisLabelRightMargin(a) {
            this.so = a;
            this.g("ValueAxisLabelRightMargin");
        }
        get valueAxisLabelBottomMargin() {
            return this.sm;
        }
        set valueAxisLabelBottomMargin(a) {
            this.sm = a;
            this.g("ValueAxisLabelBottomMargin");
        }
        get angleAxisLabelTextColor() {
            return this.t6;
        }
        set angleAxisLabelTextColor(a) {
            this.t6 = a;
            this.g("AngleAxisLabelTextColor");
        }
        get valueAxisLabelTextColor() {
            return this.uo;
        }
        set valueAxisLabelTextColor(a) {
            this.uo = a;
            this.g("ValueAxisLabelTextColor");
        }
        get actualAngleAxisLabelTextColor() {
            return this.tz;
        }
        set actualAngleAxisLabelTextColor(a) {
            this.tz = a;
            this.g("ActualAngleAxisLabelTextColor");
        }
        get actualValueAxisLabelTextColor() {
            return this.t0;
        }
        set actualValueAxisLabelTextColor(a) {
            this.t0 = a;
            this.g("ActualValueAxisLabelTextColor");
        }
        get angleAxisTitleMargin() {
            return this.sh;
        }
        set angleAxisTitleMargin(a) {
            this.sh = a;
            this.g("AngleAxisTitleMargin");
        }
        get valueAxisTitleMargin() {
            return this.sy;
        }
        set valueAxisTitleMargin(a) {
            this.sy = a;
            this.g("ValueAxisTitleMargin");
        }
        get angleAxisTitleLeftMargin() {
            return this.sg;
        }
        set angleAxisTitleLeftMargin(a) {
            this.sg = a;
            this.g("AngleAxisTitleLeftMargin");
        }
        get valueAxisTitleLeftMargin() {
            return this.sx;
        }
        set valueAxisTitleLeftMargin(a) {
            this.sx = a;
            this.g("ValueAxisTitleLeftMargin");
        }
        get angleAxisTitleTopMargin() {
            return this.sj;
        }
        set angleAxisTitleTopMargin(a) {
            this.sj = a;
            this.g("AngleAxisTitleTopMargin");
        }
        get valueAxisTitleTopMargin() {
            return this.s0;
        }
        set valueAxisTitleTopMargin(a) {
            this.s0 = a;
            this.g("ValueAxisTitleTopMargin");
        }
        get angleAxisTitleRightMargin() {
            return this.si;
        }
        set angleAxisTitleRightMargin(a) {
            this.si = a;
            this.g("AngleAxisTitleRightMargin");
        }
        get valueAxisTitleRightMargin() {
            return this.sz;
        }
        set valueAxisTitleRightMargin(a) {
            this.sz = a;
            this.g("ValueAxisTitleRightMargin");
        }
        get angleAxisTitleBottomMargin() {
            return this.sf;
        }
        set angleAxisTitleBottomMargin(a) {
            this.sf = a;
            this.g("AngleAxisTitleBottomMargin");
        }
        get valueAxisTitleBottomMargin() {
            return this.sw;
        }
        set valueAxisTitleBottomMargin(a) {
            this.sw = a;
            this.g("ValueAxisTitleBottomMargin");
        }
        get angleAxisTitleTextColor() {
            return this.uh;
        }
        set angleAxisTitleTextColor(a) {
            this.uh = a;
            this.g("AngleAxisTitleTextColor");
        }
        get valueAxisTitleTextColor() {
            return this.uz;
        }
        set valueAxisTitleTextColor(a) {
            this.uz = a;
            this.g("ValueAxisTitleTextColor");
        }
        get angleAxisLabelTextStyle() {
            return this.t7;
        }
        set angleAxisLabelTextStyle(a) {
            this.t7 = a;
            this.g("AngleAxisLabelTextStyle");
        }
        get valueAxisLabelTextStyle() {
            return this.up;
        }
        set valueAxisLabelTextStyle(a) {
            this.up = a;
            this.g("ValueAxisLabelTextStyle");
        }
        get angleAxisTitleTextStyle() {
            return this.ui;
        }
        set angleAxisTitleTextStyle(a) {
            this.ui = a;
            this.g("AngleAxisTitleTextStyle");
        }
        get valueAxisTitleTextStyle() {
            return this.u0;
        }
        set valueAxisTitleTextStyle(a) {
            this.u0 = a;
            this.g("ValueAxisTitleTextStyle");
        }
        get angleAxisLabelRef() {
            return this.t5;
        }
        set angleAxisLabelRef(a) {
            this.t5 = a;
            this.g("AngleAxisLabelRef");
        }
        get valueAxisLabelRef() {
            return this.un;
        }
        set valueAxisLabelRef(a) {
            this.un = a;
            this.g("ValueAxisLabelRef");
        }
        get angleAxisMajorStroke() {
            return this.ua;
        }
        set angleAxisMajorStroke(a) {
            this.ua = a;
            this.g("AngleAxisMajorStroke");
        }
        get valueAxisMajorStroke() {
            return this.us;
        }
        set valueAxisMajorStroke(a) {
            this.us = a;
            this.g("ValueAxisMajorStroke");
        }
        get angleAxisMajorStrokeThickness() {
            return this.r9;
        }
        set angleAxisMajorStrokeThickness(a) {
            this.r9 = a;
            this.g("AngleAxisMajorStrokeThickness");
        }
        get valueAxisMajorStrokeThickness() {
            return this.sq;
        }
        set valueAxisMajorStrokeThickness(a) {
            this.sq = a;
            this.g("ValueAxisMajorStrokeThickness");
        }
        get angleAxisMinorStrokeThickness() {
            return this.sa;
        }
        set angleAxisMinorStrokeThickness(a) {
            this.sa = a;
            this.g("AngleAxisMinorStrokeThickness");
        }
        get valueAxisMinorStrokeThickness() {
            return this.sr;
        }
        set valueAxisMinorStrokeThickness(a) {
            this.sr = a;
            this.g("ValueAxisMinorStrokeThickness");
        }
        get angleAxisStrip() {
            return this.uc;
        }
        set angleAxisStrip(a) {
            this.uc = a;
            this.g("AngleAxisStrip");
        }
        get valueAxisStrip() {
            return this.uu;
        }
        set valueAxisStrip(a) {
            this.uu = a;
            this.g("ValueAxisStrip");
        }
        get angleAxisStroke() {
            return this.ud;
        }
        set angleAxisStroke(a) {
            this.ud = a;
            this.g("AngleAxisStroke");
        }
        get valueAxisStroke() {
            return this.uv;
        }
        set valueAxisStroke(a) {
            this.uv = a;
            this.g("ValueAxisStroke");
        }
        get angleAxisStrokeThickness() {
            return this.sb;
        }
        set angleAxisStrokeThickness(a) {
            this.sb = a;
            this.g("AngleAxisStrokeThickness");
        }
        get valueAxisStrokeThickness() {
            return this.ss;
        }
        set valueAxisStrokeThickness(a) {
            this.ss = a;
            this.g("ValueAxisStrokeThickness");
        }
        get angleAxisTickLength() {
            return this.sc;
        }
        set angleAxisTickLength(a) {
            this.sc = a;
            this.g("AngleAxisTickLength");
        }
        get valueAxisTickLength() {
            return this.st;
        }
        set valueAxisTickLength(a) {
            this.st = a;
            this.g("ValueAxisTickLength");
        }
        get angleAxisTickStroke() {
            return this.ue;
        }
        set angleAxisTickStroke(a) {
            this.ue = a;
            this.g("AngleAxisTickStroke");
        }
        get valueAxisTickStroke() {
            return this.uw;
        }
        set valueAxisTickStroke(a) {
            this.uw = a;
            this.g("ValueAxisTickStroke");
        }
        get angleAxisTickStrokeThickness() {
            return this.sd;
        }
        set angleAxisTickStrokeThickness(a) {
            this.sd = a;
            this.g("AngleAxisTickStrokeThickness");
        }
        get valueAxisTickStrokeThickness() {
            return this.su;
        }
        set valueAxisTickStrokeThickness(a) {
            this.su = a;
            this.g("ValueAxisTickStrokeThickness");
        }
        get angleAxisTitle() {
            return this.uf;
        }
        set angleAxisTitle(a) {
            this.uf = a;
            this.g("AngleAxisTitle");
        }
        get valueAxisTitle() {
            return this.ux;
        }
        set valueAxisTitle(a) {
            this.ux = a;
            this.g("ValueAxisTitle");
        }
        get angleAxisMinorStroke() {
            return this.ub;
        }
        set angleAxisMinorStroke(a) {
            this.ub = a;
            this.g("AngleAxisMinorStroke");
        }
        get valueAxisMinorStroke() {
            return this.ut;
        }
        set valueAxisMinorStroke(a) {
            this.ut = a;
            this.g("ValueAxisMinorStroke");
        }
        get angleAxisLabelAngle() {
            return this.r4;
        }
        set angleAxisLabelAngle(a) {
            this.r4 = a;
            this.g("AngleAxisLabelAngle");
        }
        get valueAxisLabelAngle() {
            return this.sl;
        }
        set valueAxisLabelAngle(a) {
            this.sl = a;
            this.g("ValueAxisLabelAngle");
        }
        get angleAxisExtent() {
            return this.r3;
        }
        set angleAxisExtent(a) {
            this.r3 = a;
            this.g("AngleAxisExtent");
        }
        get valueAxisExtent() {
            return this.sk;
        }
        set valueAxisExtent(a) {
            this.sk = a;
            this.g("ValueAxisExtent");
        }
        get angleAxisTitleAngle() {
            return this.se;
        }
        set angleAxisTitleAngle(a) {
            this.se = a;
            this.g("AngleAxisTitleAngle");
        }
        get valueAxisTitleAngle() {
            return this.sv;
        }
        set valueAxisTitleAngle(a) {
            this.sv = a;
            this.g("ValueAxisTitleAngle");
        }
        get angleAxisInverted() {
            return this.rz;
        }
        set angleAxisInverted(a) {
            this.rz = a;
            this.g("AngleAxisInverted");
        }
        get valueAxisInverted() {
            return this.r0;
        }
        set valueAxisInverted(a) {
            this.r0 = a;
            this.g("ValueAxisInverted");
        }
        get angleAxisTitleAlignment() {
            return this.ug;
        }
        set angleAxisTitleAlignment(a) {
            this.ug = a;
            this.g("AngleAxisTitleAlignment");
        }
        get valueAxisTitleAlignment() {
            return this.uy;
        }
        set valueAxisTitleAlignment(a) {
            this.uy = a;
            this.g("ValueAxisTitleAlignment");
        }
        get angleAxisLabelHorizontalAlignment() {
            return this.t3;
        }
        set angleAxisLabelHorizontalAlignment(a) {
            this.t3 = a;
            this.g("AngleAxisLabelHorizontalAlignment");
        }
        get valueAxisLabelHorizontalAlignment() {
            return this.ul;
        }
        set valueAxisLabelHorizontalAlignment(a) {
            this.ul = a;
            this.g("ValueAxisLabelHorizontalAlignment");
        }
        get angleAxisLabelVerticalAlignment() {
            return this.t8;
        }
        set angleAxisLabelVerticalAlignment(a) {
            this.t8 = a;
            this.g("AngleAxisLabelVerticalAlignment");
        }
        get valueAxisLabelVerticalAlignment() {
            return this.uq;
        }
        set valueAxisLabelVerticalAlignment(a) {
            this.uq = a;
            this.g("ValueAxisLabelVerticalAlignment");
        }
        get angleAxisLabelVisibility() {
            return this.t9;
        }
        set angleAxisLabelVisibility(a) {
            this.t9 = a;
            this.g("AngleAxisLabelVisibility");
        }
        get valueAxisLabelVisibility() {
            return this.ur;
        }
        set valueAxisLabelVisibility(a) {
            this.ur = a;
            this.g("ValueAxisLabelVisibility");
        }
        get valueAxisLabelLocation() {
            return this.um;
        }
        set valueAxisLabelLocation(a) {
            this.um = a;
            this.g("ValueAxisLabelLocation");
        }
        get angleAxisLabelLocation() {
            return this.t4;
        }
        set angleAxisLabelLocation(a) {
            this.t4 = a;
            this.g("AngleAxisLabelLocation");
        }
        get angleAxisLabelFormat() {
            return this.t2;
        }
        set angleAxisLabelFormat(a) {
            this.t2 = a;
            this.g("AngleAxisLabelFormat");
        }
        get angleAxisLabelFormatSpecifiers() {
            return this.rv;
        }
        set angleAxisLabelFormatSpecifiers(a) {
            this.rv = a;
            this.g("AngleAxisLabelFormatSpecifiers");
        }
        get valueAxisLabelFormat() {
            return this.uk;
        }
        set valueAxisLabelFormat(a) {
            this.uk = a;
            this.g("ValueAxisLabelFormat");
        }
        get valueAxisLabelFormatSpecifiers() {
            return this.rw;
        }
        set valueAxisLabelFormatSpecifiers(a) {
            this.rw = a;
            this.g("ValueAxisLabelFormatSpecifiers");
        }
    }
    RadialBaseChartDescription.$t = markType(RadialBaseChartDescription, 'RadialBaseChartDescription', DomainChartDescription.$);
    return RadialBaseChartDescription;
})();
