// igniteui-angular-core/esm2015/lib/RangeCategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let RangeCategorySeriesDescription = /*@__PURE__*/ (() => {
    class RangeCategorySeriesDescription extends CategorySeriesDescription {
        constructor() {
            super();
            this.f4 = null;
            this.f3 = null;
            this.f2 = null;
            this.f1 = null;
        }
        get_type() {
            return "RangeCategorySeries";
        }
        get lowMemberPath() {
            return this.f4;
        }
        set lowMemberPath(a) {
            this.f4 = a;
            this.g("LowMemberPath");
        }
        get highMemberPath() {
            return this.f3;
        }
        set highMemberPath(a) {
            this.f3 = a;
            this.g("HighMemberPath");
        }
        get highlightedLowMemberPath() {
            return this.f2;
        }
        set highlightedLowMemberPath(a) {
            this.f2 = a;
            this.g("HighlightedLowMemberPath");
        }
        get highlightedHighMemberPath() {
            return this.f1;
        }
        set highlightedHighMemberPath(a) {
            this.f1 = a;
            this.g("HighlightedHighMemberPath");
        }
    }
    RangeCategorySeriesDescription.$t = markType(RangeCategorySeriesDescription, 'RangeCategorySeriesDescription', CategorySeriesDescription.$);
    return RangeCategorySeriesDescription;
})();
