// igniteui-angular-core/esm2015/lib/ScatterBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ScatterBaseDescription = /*@__PURE__*/ (() => {
    class ScatterBaseDescription extends MarkerSeriesDescription {
        constructor() {
            super();
            this.gd = null;
            this.gh = null;
            this.gg = null;
            this.gk = null;
            this.f7 = null;
            this.f8 = null;
            this.ge = null;
            this.gi = null;
            this.gf = null;
            this.gj = null;
            this.gc = null;
            this.gb = null;
            this.f4 = null;
            this.ft = 0;
            this.fn = null;
            this.fx = 0;
            this.ga = null;
            this.fy = 0;
            this.fw = 0;
            this.f9 = null;
            this.fv = 0;
            this.f3 = null;
            this.fq = false;
            this.fp = false;
            this.f6 = null;
            this.f5 = null;
        }
        get_type() {
            return "ScatterBase";
        }
        get xAxisRef() {
            return this.gd;
        }
        set xAxisRef(a) {
            this.gd = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.gh;
        }
        set yAxisRef(a) {
            this.gh = a;
            this.g("YAxisRef");
        }
        get xMemberPath() {
            return this.gg;
        }
        set xMemberPath(a) {
            this.gg = a;
            this.g("XMemberPath");
        }
        get yMemberPath() {
            return this.gk;
        }
        set yMemberPath(a) {
            this.gk = a;
            this.g("YMemberPath");
        }
        get highlightedXMemberPath() {
            return this.f7;
        }
        set highlightedXMemberPath(a) {
            this.f7 = a;
            this.g("HighlightedXMemberPath");
        }
        get highlightedYMemberPath() {
            return this.f8;
        }
        set highlightedYMemberPath(a) {
            this.f8 = a;
            this.g("HighlightedYMemberPath");
        }
        get xMemberAsLegendLabel() {
            return this.ge;
        }
        set xMemberAsLegendLabel(a) {
            this.ge = a;
            this.g("XMemberAsLegendLabel");
        }
        get yMemberAsLegendLabel() {
            return this.gi;
        }
        set yMemberAsLegendLabel(a) {
            this.gi = a;
            this.g("YMemberAsLegendLabel");
        }
        get xMemberAsLegendUnit() {
            return this.gf;
        }
        set xMemberAsLegendUnit(a) {
            this.gf = a;
            this.g("XMemberAsLegendUnit");
        }
        get yMemberAsLegendUnit() {
            return this.gj;
        }
        set yMemberAsLegendUnit(a) {
            this.gj = a;
            this.g("YMemberAsLegendUnit");
        }
        get trendLineType() {
            return this.gc;
        }
        set trendLineType(a) {
            this.gc = a;
            this.g("TrendLineType");
        }
        get trendLineBrush() {
            return this.gb;
        }
        set trendLineBrush(a) {
            this.gb = a;
            this.g("TrendLineBrush");
        }
        get actualTrendLineBrush() {
            return this.f4;
        }
        set actualTrendLineBrush(a) {
            this.f4 = a;
            this.g("ActualTrendLineBrush");
        }
        get trendLineThickness() {
            return this.ft;
        }
        set trendLineThickness(a) {
            this.ft = a;
            this.g("TrendLineThickness");
        }
        get trendLineDashArray() {
            return this.fn;
        }
        set trendLineDashArray(a) {
            this.fn = a;
            this.g("TrendLineDashArray");
        }
        get trendLinePeriod() {
            return this.fx;
        }
        set trendLinePeriod(a) {
            this.fx = a;
            this.g("TrendLinePeriod");
        }
        get markerCollisionAvoidance() {
            return this.ga;
        }
        set markerCollisionAvoidance(a) {
            this.ga = a;
            this.g("MarkerCollisionAvoidance");
        }
        get trendLineZIndex() {
            return this.fy;
        }
        set trendLineZIndex(a) {
            this.fy = a;
            this.g("TrendLineZIndex");
        }
        get maximumMarkers() {
            return this.fw;
        }
        set maximumMarkers(a) {
            this.fw = a;
            this.g("MaximumMarkers");
        }
        get itemSearchMode() {
            return this.f9;
        }
        set itemSearchMode(a) {
            this.f9 = a;
            this.g("ItemSearchMode");
        }
        get itemSearchThreshold() {
            return this.fv;
        }
        set itemSearchThreshold(a) {
            this.fv = a;
            this.g("ItemSearchThreshold");
        }
        get actualItemSearchMode() {
            return this.f3;
        }
        set actualItemSearchMode(a) {
            this.f3 = a;
            this.g("ActualItemSearchMode");
        }
        get isCustomScatterStyleAllowed() {
            return this.fq;
        }
        set isCustomScatterStyleAllowed(a) {
            this.fq = a;
            this.g("IsCustomScatterStyleAllowed");
        }
        get isCustomScatterMarkerStyleAllowed() {
            return this.fp;
        }
        set isCustomScatterMarkerStyleAllowed(a) {
            this.fp = a;
            this.g("IsCustomScatterMarkerStyleAllowed");
        }
        get assigningScatterStyleRef() {
            return this.f6;
        }
        set assigningScatterStyleRef(a) {
            this.f6 = a;
            this.g("AssigningScatterStyleRef");
        }
        get assigningScatterMarkerStyleRef() {
            return this.f5;
        }
        set assigningScatterMarkerStyleRef(a) {
            this.f5 = a;
            this.g("AssigningScatterMarkerStyleRef");
        }
    }
    ScatterBaseDescription.$t = markType(ScatterBaseDescription, 'ScatterBaseDescription', MarkerSeriesDescription.$);
    return ScatterBaseDescription;
})();
