// igniteui-angular-core/esm2015/lib/ScatterPolygonSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ShapeSeriesBaseDescription } from "./ShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ScatterPolygonSeriesDescription = /*@__PURE__*/ (() => {
    class ScatterPolygonSeriesDescription extends ShapeSeriesBaseDescription {
        constructor() {
            super();
            this.ft = null;
            this.f6 = null;
            this.f5 = null;
            this.fv = 0;
            this.fz = null;
            this.f0 = null;
            this.fx = null;
            this.f3 = null;
            this.fy = null;
            this.f1 = null;
            this.f4 = null;
            this.f2 = null;
        }
        get_type() {
            return "ScatterPolygonSeries";
        }
        get shapeStyle() {
            return this.ft;
        }
        set shapeStyle(a) {
            this.ft = a;
            this.g("ShapeStyle");
        }
        get markerType() {
            return this.f6;
        }
        set markerType(a) {
            this.f6 = a;
            this.g("MarkerType");
        }
        get markerTemplateRef() {
            return this.f5;
        }
        set markerTemplateRef(a) {
            this.f5 = a;
            this.g("MarkerTemplateRef");
        }
        get markerThickness() {
            return this.fv;
        }
        set markerThickness(a) {
            this.fv = a;
            this.g("MarkerThickness");
        }
        get actualMarkerTemplateRef() {
            return this.fz;
        }
        set actualMarkerTemplateRef(a) {
            this.fz = a;
            this.g("ActualMarkerTemplateRef");
        }
        get markerBrush() {
            return this.f0;
        }
        set markerBrush(a) {
            this.f0 = a;
            this.g("MarkerBrush");
        }
        get actualMarkerBrush() {
            return this.fx;
        }
        set actualMarkerBrush(a) {
            this.fx = a;
            this.g("ActualMarkerBrush");
        }
        get markerOutline() {
            return this.f3;
        }
        set markerOutline(a) {
            this.f3 = a;
            this.g("MarkerOutline");
        }
        get actualMarkerOutline() {
            return this.fy;
        }
        set actualMarkerOutline(a) {
            this.fy = a;
            this.g("ActualMarkerOutline");
        }
        get markerCollisionAvoidance() {
            return this.f1;
        }
        set markerCollisionAvoidance(a) {
            this.f1 = a;
            this.g("MarkerCollisionAvoidance");
        }
        get markerOutlineMode() {
            return this.f4;
        }
        set markerOutlineMode(a) {
            this.f4 = a;
            this.g("MarkerOutlineMode");
        }
        get markerFillMode() {
            return this.f2;
        }
        set markerFillMode(a) {
            this.f2 = a;
            this.g("MarkerFillMode");
        }
    }
    ScatterPolygonSeriesDescription.$t = markType(ScatterPolygonSeriesDescription, 'ScatterPolygonSeriesDescription', ShapeSeriesBaseDescription.$);
    return ScatterPolygonSeriesDescription;
})();
