// igniteui-angular-core/esm2015/lib/ScatterAreaSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterTriangulationSeriesDescription } from "./ScatterTriangulationSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ScatterAreaSeriesDescription = /*@__PURE__*/ (() => {
    class ScatterAreaSeriesDescription extends ScatterTriangulationSeriesDescription {
        constructor() {
            super();
            this.fs = null;
            this.fx = null;
            this.fr = null;
            this.fv = null;
            this.fw = null;
        }
        get_type() {
            return "ScatterAreaSeries";
        }
        get colorScale() {
            return this.fs;
        }
        set colorScale(a) {
            this.fs = a;
            this.g("ColorScale");
        }
        get colorMemberPath() {
            return this.fx;
        }
        set colorMemberPath(a) {
            this.fx = a;
            this.g("ColorMemberPath");
        }
        get actualColorScale() {
            return this.fr;
        }
        set actualColorScale(a) {
            this.fr = a;
            this.g("ActualColorScale");
        }
        get colorMemberAsLegendLabel() {
            return this.fv;
        }
        set colorMemberAsLegendLabel(a) {
            this.fv = a;
            this.g("ColorMemberAsLegendLabel");
        }
        get colorMemberAsLegendUnit() {
            return this.fw;
        }
        set colorMemberAsLegendUnit(a) {
            this.fw = a;
            this.g("ColorMemberAsLegendUnit");
        }
    }
    ScatterAreaSeriesDescription.$t = markType(ScatterAreaSeriesDescription, 'ScatterAreaSeriesDescription', ScatterTriangulationSeriesDescription.$);
    return ScatterAreaSeriesDescription;
})();
