// igniteui-angular-core/esm2015/lib/ScatterContourSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ScatterTriangulationSeriesDescription } from "./ScatterTriangulationSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ScatterContourSeriesDescription = /*@__PURE__*/ (() => {
    class ScatterContourSeriesDescription extends ScatterTriangulationSeriesDescription {
        constructor() {
            super();
            this.fz = null;
            this.fx = null;
            this.fy = null;
            this.fu = null;
            this.ft = null;
            this.fr = null;
        }
        get_type() {
            return "ScatterContourSeries";
        }
        get valueMemberPath() {
            return this.fz;
        }
        set valueMemberPath(a) {
            this.fz = a;
            this.g("ValueMemberPath");
        }
        get valueMemberAsLegendLabel() {
            return this.fx;
        }
        set valueMemberAsLegendLabel(a) {
            this.fx = a;
            this.g("ValueMemberAsLegendLabel");
        }
        get valueMemberAsLegendUnit() {
            return this.fy;
        }
        set valueMemberAsLegendUnit(a) {
            this.fy = a;
            this.g("ValueMemberAsLegendUnit");
        }
        get fillScale() {
            return this.fu;
        }
        set fillScale(a) {
            this.fu = a;
            this.g("FillScale");
        }
        get actualFillScale() {
            return this.ft;
        }
        set actualFillScale(a) {
            this.ft = a;
            this.g("ActualFillScale");
        }
        get valueResolver() {
            return this.fr;
        }
        set valueResolver(a) {
            this.fr = a;
            this.g("ValueResolver");
        }
    }
    ScatterContourSeriesDescription.$t = markType(ScatterContourSeriesDescription, 'ScatterContourSeriesDescription', ScatterTriangulationSeriesDescription.$);
    return ScatterContourSeriesDescription;
})();
