// igniteui-angular-core/esm2015/lib/SeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let SeriesDescription = /*@__PURE__*/ (() => {
    class SeriesDescription extends Description {
        constructor() {
            super();
            this.co = null;
            this.cv = null;
            this.l = null;
            this.cp = null;
            this.cs = null;
            this.c3 = null;
            this.c2 = null;
            this.c0 = null;
            this.cg = null;
            this.cf = null;
            this.ce = null;
            this.cz = null;
            this.cy = null;
            this.c1 = null;
            this.cq = null;
            this.de = null;
            this.df = null;
            this.di = null;
            this.b3 = 0;
            this.az = 0;
            this.a8 = 0;
            this.bg = 0;
            this.bd = 0;
            this.be = 0;
            this.bf = 0;
            this.cu = null;
            this.ct = null;
            this.cm = null;
            this.b9 = null;
            this.da = null;
            this.cr = null;
            this.ci = null;
            this.ca = null;
            this.c9 = null;
            this.c6 = null;
            this.ch = null;
            this.c5 = null;
            this.cw = null;
            this.aa = false;
            this.bc = 0;
            this.a0 = 0;
            this.k = null;
            this.cc = null;
            this.cj = null;
            this.cb = null;
            this.aw = 0;
            this.a3 = 0;
            this.ax = 0;
            this.a4 = 0;
            this.u = false;
            this.v = false;
            this.y = false;
            this.af = false;
            this.ad = false;
            this.ac = false;
            this.ab = false;
            this.x = false;
            this.a9 = 0;
            this.db = null;
            this.ag = false;
            this.ba = 0;
            this.bb = 0;
            this.a1 = 0;
            this.av = 0;
            this.a5 = 0;
            this.ay = 0;
            this.w = false;
            this.ae = false;
            this.dl = null;
            this.c7 = null;
            this.b4 = 0;
            this.b5 = 0;
            this.dg = null;
            this.dj = null;
            this.c4 = null;
            this.ck = null;
            this.i = null;
            this.cl = null;
            this.j = null;
            this.z = false;
            this.cn = null;
            this.t = false;
            this.cx = null;
            this.cd = null;
            this.a2 = 0;
            this.a7 = 0;
            this.m = null;
            this.h = null;
            this.a6 = 0;
            this.dk = null;
            this.dh = null;
            this.c8 = null;
            this.dd = null;
            this.dc = null;
        }
        get_type() {
            return "Series";
        }
        get type() {
            return this.get_type();
        }
        get dataLegendGroup() {
            return this.co;
        }
        set dataLegendGroup(a) {
            this.co = a;
            this.g("DataLegendGroup");
        }
        get highlightedValuesDataLegendGroup() {
            return this.cv;
        }
        set highlightedValuesDataLegendGroup(a) {
            this.cv = a;
            this.g("HighlightedValuesDataLegendGroup");
        }
        get highlightedValuesExtraPropertyOverlays() {
            return this.l;
        }
        set highlightedValuesExtraPropertyOverlays(a) {
            this.l = a;
            this.g("HighlightedValuesExtraPropertyOverlays");
        }
        get dataSourceRef() {
            return this.cp;
        }
        set dataSourceRef(a) {
            this.cp = a;
            this.g("DataSourceRef");
        }
        get highlightedDataSourceRef() {
            return this.cs;
        }
        set highlightedDataSourceRef(a) {
            this.cs = a;
            this.g("HighlightedDataSourceRef");
        }
        get legendRef() {
            return this.c3;
        }
        set legendRef(a) {
            this.c3 = a;
            this.g("LegendRef");
        }
        get legendItemVisibility() {
            return this.c2;
        }
        set legendItemVisibility(a) {
            this.c2 = a;
            this.g("LegendItemVisibility");
        }
        get legendItemBadgeTemplateRef() {
            return this.c0;
        }
        set legendItemBadgeTemplateRef(a) {
            this.c0 = a;
            this.g("LegendItemBadgeTemplateRef");
        }
        get actualLegendItemBadgeTemplateRef() {
            return this.cg;
        }
        set actualLegendItemBadgeTemplateRef(a) {
            this.cg = a;
            this.g("ActualLegendItemBadgeTemplateRef");
        }
        get actualLegendItemBadgeOutline() {
            return this.cf;
        }
        set actualLegendItemBadgeOutline(a) {
            this.cf = a;
            this.g("ActualLegendItemBadgeOutline");
        }
        get actualLegendItemBadgeBrush() {
            return this.ce;
        }
        set actualLegendItemBadgeBrush(a) {
            this.ce = a;
            this.g("ActualLegendItemBadgeBrush");
        }
        get legendItemBadgeShape() {
            return this.cz;
        }
        set legendItemBadgeShape(a) {
            this.cz = a;
            this.g("LegendItemBadgeShape");
        }
        get legendItemBadgeMode() {
            return this.cy;
        }
        set legendItemBadgeMode(a) {
            this.cy = a;
            this.g("LegendItemBadgeMode");
        }
        get legendItemTemplateRef() {
            return this.c1;
        }
        set legendItemTemplateRef(a) {
            this.c1 = a;
            this.g("LegendItemTemplateRef");
        }
        get discreteLegendItemTemplateRef() {
            return this.cq;
        }
        set discreteLegendItemTemplateRef(a) {
            this.cq = a;
            this.g("DiscreteLegendItemTemplateRef");
        }
        get transitionEasingFunctionRef() {
            return this.de;
        }
        set transitionEasingFunctionRef(a) {
            this.de = a;
            this.g("TransitionEasingFunctionRef");
        }
        get transitionInEasingFunctionRef() {
            return this.df;
        }
        set transitionInEasingFunctionRef(a) {
            this.df = a;
            this.g("TransitionInEasingFunctionRef");
        }
        get transitionOutEasingFunctionRef() {
            return this.di;
        }
        set transitionOutEasingFunctionRef(a) {
            this.di = a;
            this.g("TransitionOutEasingFunctionRef");
        }
        get transitionDuration() {
            return this.b3;
        }
        set transitionDuration(a) {
            this.b3 = a;
            this.g("TransitionDuration");
        }
        get actualResolution() {
            return this.az;
        }
        set actualResolution(a) {
            this.az = a;
            this.g("ActualResolution");
        }
        get resolution() {
            return this.a8;
        }
        set resolution(a) {
            this.a8 = a;
            this.g("Resolution");
        }
        get visibleRangeMarginTop() {
            return this.bg;
        }
        set visibleRangeMarginTop(a) {
            this.bg = a;
            this.g("VisibleRangeMarginTop");
        }
        get visibleRangeMarginBottom() {
            return this.bd;
        }
        set visibleRangeMarginBottom(a) {
            this.bd = a;
            this.g("VisibleRangeMarginBottom");
        }
        get visibleRangeMarginLeft() {
            return this.be;
        }
        set visibleRangeMarginLeft(a) {
            this.be = a;
            this.g("VisibleRangeMarginLeft");
        }
        get visibleRangeMarginRight() {
            return this.bf;
        }
        set visibleRangeMarginRight(a) {
            this.bf = a;
            this.g("VisibleRangeMarginRight");
        }
        get highlightedTitleSuffix() {
            return this.cu;
        }
        set highlightedTitleSuffix(a) {
            this.cu = a;
            this.g("HighlightedTitleSuffix");
        }
        get highlightedLegendItemVisibility() {
            return this.ct;
        }
        set highlightedLegendItemVisibility(a) {
            this.ct = a;
            this.g("HighlightedLegendItemVisibility");
        }
        get brush() {
            return this.cm;
        }
        set brush(a) {
            this.cm = a;
            this.g("Brush");
        }
        get actualBrush() {
            return this.b9;
        }
        set actualBrush(a) {
            this.b9 = a;
            this.g("ActualBrush");
        }
        get selectionBrush() {
            return this.da;
        }
        set selectionBrush(a) {
            this.da = a;
            this.g("SelectionBrush");
        }
        get focusBrush() {
            return this.cr;
        }
        set focusBrush(a) {
            this.cr = a;
            this.g("FocusBrush");
        }
        get actualSelectionBrush() {
            return this.ci;
        }
        set actualSelectionBrush(a) {
            this.ci = a;
            this.g("ActualSelectionBrush");
        }
        get actualFocusBrush() {
            return this.ca;
        }
        set actualFocusBrush(a) {
            this.ca = a;
            this.g("ActualFocusBrush");
        }
        get safeActualBrush() {
            return this.c9;
        }
        set safeActualBrush(a) {
            this.c9 = a;
            this.g("SafeActualBrush");
        }
        get outline() {
            return this.c6;
        }
        set outline(a) {
            this.c6 = a;
            this.g("Outline");
        }
        get actualOutline() {
            return this.ch;
        }
        set actualOutline(a) {
            this.ch = a;
            this.g("ActualOutline");
        }
        get lineJoin() {
            return this.c5;
        }
        set lineJoin(a) {
            this.c5 = a;
            this.g("LineJoin");
        }
        get highlightedValuesDisplayMode() {
            return this.cw;
        }
        set highlightedValuesDisplayMode(a) {
            this.cw = a;
            this.g("HighlightedValuesDisplayMode");
        }
        get shouldAnimateOnDataSourceSwap() {
            return this.aa;
        }
        set shouldAnimateOnDataSourceSwap(a) {
            this.aa = a;
            this.g("ShouldAnimateOnDataSourceSwap");
        }
        get thickness() {
            return this.bc;
        }
        set thickness(a) {
            this.bc = a;
            this.g("Thickness");
        }
        get actualThickness() {
            return this.a0;
        }
        set actualThickness(a) {
            this.a0 = a;
            this.g("ActualThickness");
        }
        get dashArray() {
            return this.k;
        }
        set dashArray(a) {
            this.k = a;
            this.g("DashArray");
        }
        get actualHighlightingMode() {
            return this.cc;
        }
        set actualHighlightingMode(a) {
            this.cc = a;
            this.g("ActualHighlightingMode");
        }
        get actualSelectionMode() {
            return this.cj;
        }
        set actualSelectionMode(a) {
            this.cj = a;
            this.g("ActualSelectionMode");
        }
        get actualFocusMode() {
            return this.cb;
        }
        set actualFocusMode(a) {
            this.cb = a;
            this.g("ActualFocusMode");
        }
        get actualHighlightedValuesFadeOpacity() {
            return this.aw;
        }
        set actualHighlightedValuesFadeOpacity(a) {
            this.aw = a;
            this.g("ActualHighlightedValuesFadeOpacity");
        }
        get highlightedValuesFadeOpacity() {
            return this.a3;
        }
        set highlightedValuesFadeOpacity(a) {
            this.a3 = a;
            this.g("HighlightedValuesFadeOpacity");
        }
        get actualHighlightingFadeOpacity() {
            return this.ax;
        }
        set actualHighlightingFadeOpacity(a) {
            this.ax = a;
            this.g("ActualHighlightingFadeOpacity");
        }
        get highlightingFadeOpacity() {
            return this.a4;
        }
        set highlightingFadeOpacity(a) {
            this.a4 = a;
            this.g("HighlightingFadeOpacity");
        }
        get isActualLegendFinancial() {
            return this.u;
        }
        set isActualLegendFinancial(a) {
            this.u = a;
            this.g("IsActualLegendFinancial");
        }
        get isComponentHighlightingModeIgnored() {
            return this.v;
        }
        set isComponentHighlightingModeIgnored(a) {
            this.v = a;
            this.g("IsComponentHighlightingModeIgnored");
        }
        get isHighlightingEnabled() {
            return this.y;
        }
        set isHighlightingEnabled(a) {
            this.y = a;
            this.g("IsHighlightingEnabled");
        }
        get useItemWiseColors() {
            return this.af;
        }
        set useItemWiseColors(a) {
            this.af = a;
            this.g("UseItemWiseColors");
        }
        get shouldShiftOpacityForSafeActualBrush() {
            return this.ad;
        }
        set shouldShiftOpacityForSafeActualBrush(a) {
            this.ad = a;
            this.g("ShouldShiftOpacityForSafeActualBrush");
        }
        get shouldRemoveHighlightedDataOnLayerHidden() {
            return this.ac;
        }
        set shouldRemoveHighlightedDataOnLayerHidden(a) {
            this.ac = a;
            this.g("ShouldRemoveHighlightedDataOnLayerHidden");
        }
        get shouldHideAutoCallouts() {
            return this.ab;
        }
        set shouldHideAutoCallouts(a) {
            this.ab = a;
            this.g("ShouldHideAutoCallouts");
        }
        get isDropShadowEnabled() {
            return this.x;
        }
        set isDropShadowEnabled(a) {
            this.x = a;
            this.g("IsDropShadowEnabled");
        }
        get shadowBlur() {
            return this.a9;
        }
        set shadowBlur(a) {
            this.a9 = a;
            this.g("ShadowBlur");
        }
        get shadowColor() {
            return this.db;
        }
        set shadowColor(a) {
            this.db = a;
            this.g("ShadowColor");
        }
        get useSingleShadow() {
            return this.ag;
        }
        set useSingleShadow(a) {
            this.ag = a;
            this.g("UseSingleShadow");
        }
        get shadowOffsetX() {
            return this.ba;
        }
        set shadowOffsetX(a) {
            this.ba = a;
            this.g("ShadowOffsetX");
        }
        get shadowOffsetY() {
            return this.bb;
        }
        set shadowOffsetY(a) {
            this.bb = a;
            this.g("ShadowOffsetY");
        }
        get areaFillOpacity() {
            return this.a1;
        }
        set areaFillOpacity(a) {
            this.a1 = a;
            this.g("AreaFillOpacity");
        }
        get actualAreaFillOpacity() {
            return this.av;
        }
        set actualAreaFillOpacity(a) {
            this.av = a;
            this.g("ActualAreaFillOpacity");
        }
        get markerFillOpacity() {
            return this.a5;
        }
        set markerFillOpacity(a) {
            this.a5 = a;
            this.g("MarkerFillOpacity");
        }
        get actualMarkerFillOpacity() {
            return this.ay;
        }
        set actualMarkerFillOpacity(a) {
            this.ay = a;
            this.g("ActualMarkerFillOpacity");
        }
        get isDefaultToolTipSelected() {
            return this.w;
        }
        set isDefaultToolTipSelected(a) {
            this.w = a;
            this.g("IsDefaultToolTipSelected");
        }
        get showDefaultTooltip() {
            return this.ae;
        }
        set showDefaultTooltip(a) {
            this.ae = a;
            this.g("ShowDefaultTooltip");
        }
        get visibleRangeMode() {
            return this.dl;
        }
        set visibleRangeMode(a) {
            this.dl = a;
            this.g("VisibleRangeMode");
        }
        get outlineMode() {
            return this.c7;
        }
        set outlineMode(a) {
            this.c7 = a;
            this.g("OutlineMode");
        }
        get transitionInDuration() {
            return this.b4;
        }
        set transitionInDuration(a) {
            this.b4 = a;
            this.g("TransitionInDuration");
        }
        get transitionOutDuration() {
            return this.b5;
        }
        set transitionOutDuration(a) {
            this.b5 = a;
            this.g("TransitionOutDuration");
        }
        get transitionInSpeedType() {
            return this.dg;
        }
        set transitionInSpeedType(a) {
            this.dg = a;
            this.g("TransitionInSpeedType");
        }
        get transitionOutSpeedType() {
            return this.dj;
        }
        set transitionOutSpeedType(a) {
            this.dj = a;
            this.g("TransitionOutSpeedType");
        }
        get lineCap() {
            return this.c4;
        }
        set lineCap(a) {
            this.c4 = a;
            this.g("LineCap");
        }
        get autoCalloutLabelFormat() {
            return this.ck;
        }
        set autoCalloutLabelFormat(a) {
            this.ck = a;
            this.g("AutoCalloutLabelFormat");
        }
        get autoCalloutLabelFormatSpecifiers() {
            return this.i;
        }
        set autoCalloutLabelFormatSpecifiers(a) {
            this.i = a;
            this.g("AutoCalloutLabelFormatSpecifiers");
        }
        get autoCalloutValueLabelFormat() {
            return this.cl;
        }
        set autoCalloutValueLabelFormat(a) {
            this.cl = a;
            this.g("AutoCalloutValueLabelFormat");
        }
        get autoCalloutValueLabelFormatSpecifiers() {
            return this.j;
        }
        set autoCalloutValueLabelFormatSpecifiers(a) {
            this.j = a;
            this.g("AutoCalloutValueLabelFormatSpecifiers");
        }
        get mouseOverEnabled() {
            return this.z;
        }
        set mouseOverEnabled(a) {
            this.z = a;
            this.g("MouseOverEnabled");
        }
        get coercionMethodsRef() {
            return this.cn;
        }
        set coercionMethodsRef(a) {
            this.cn = a;
            this.g("CoercionMethodsRef");
        }
        get expectFunctions() {
            return this.t;
        }
        set expectFunctions(a) {
            this.t = a;
            this.g("ExpectFunctions");
        }
        get hitTestMode() {
            return this.cx;
        }
        set hitTestMode(a) {
            this.cx = a;
            this.g("HitTestMode");
        }
        get actualHitTestMode() {
            return this.cd;
        }
        set actualHitTestMode(a) {
            this.cd = a;
            this.g("ActualHitTestMode");
        }
        get finalValue() {
            return this.a2;
        }
        set finalValue(a) {
            this.a2 = a;
            this.g("FinalValue");
        }
        get percentChange() {
            return this.a7;
        }
        set percentChange(a) {
            this.a7 = a;
            this.g("PercentChange");
        }
        get layers() {
            return this.m;
        }
        set layers(a) {
            this.m = a;
            this.g("Layers");
        }
        get actualLayers() {
            return this.h;
        }
        set actualLayers(a) {
            this.h = a;
            this.g("ActualLayers");
        }
        get opacity() {
            return this.a6;
        }
        set opacity(a) {
            this.a6 = a;
            this.g("Opacity");
        }
        get visibility() {
            return this.dk;
        }
        set visibility(a) {
            this.dk = a;
            this.g("Visibility");
        }
        get transitionOutCompletedRef() {
            return this.dh;
        }
        set transitionOutCompletedRef(a) {
            this.dh = a;
            this.g("TransitionOutCompletedRef");
        }
        get renderRequestedRef() {
            return this.c8;
        }
        set renderRequestedRef(a) {
            this.c8 = a;
            this.g("RenderRequestedRef");
        }
        get tooltipTemplateRef() {
            return this.dd;
        }
        set tooltipTemplateRef(a) {
            this.dd = a;
            this.g("TooltipTemplateRef");
        }
        get title() {
            return this.dc;
        }
        set title(a) {
            this.dc = a;
            this.g("Title");
        }
    }
    SeriesDescription.$t = markType(SeriesDescription, 'SeriesDescription', Description.$);
    return SeriesDescription;
})();
