// igniteui-angular-core/esm2015/lib/ShapeSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ShapeSeriesBaseDescription = /*@__PURE__*/ (() => {
    class ShapeSeriesBaseDescription extends SeriesDescription {
        constructor() {
            super();
            this.e2 = false;
            this.e1 = false;
            this.fb = null;
            this.ff = null;
            this.e8 = 0;
            this.e7 = 0;
            this.fg = null;
            this.fe = null;
            this.fi = null;
            this.fj = null;
            this.e5 = 0;
            this.fd = null;
            this.fc = null;
            this.fh = null;
        }
        get_type() {
            return "ShapeSeriesBase";
        }
        get isCustomShapeStyleAllowed() {
            return this.e2;
        }
        set isCustomShapeStyleAllowed(a) {
            this.e2 = a;
            this.g("IsCustomShapeStyleAllowed");
        }
        get isCustomShapeMarkerStyleAllowed() {
            return this.e1;
        }
        set isCustomShapeMarkerStyleAllowed(a) {
            this.e1 = a;
            this.g("IsCustomShapeMarkerStyleAllowed");
        }
        get actualItemSearchMode() {
            return this.fb;
        }
        set actualItemSearchMode(a) {
            this.fb = a;
            this.g("ActualItemSearchMode");
        }
        get itemSearchMode() {
            return this.ff;
        }
        set itemSearchMode(a) {
            this.ff = a;
            this.g("ItemSearchMode");
        }
        get itemSearchThreshold() {
            return this.e8;
        }
        set itemSearchThreshold(a) {
            this.e8 = a;
            this.g("ItemSearchThreshold");
        }
        get itemSearchPointsThreshold() {
            return this.e7;
        }
        set itemSearchPointsThreshold(a) {
            this.e7 = a;
            this.g("ItemSearchPointsThreshold");
        }
        get shapeMemberPath() {
            return this.fg;
        }
        set shapeMemberPath(a) {
            this.fg = a;
            this.g("ShapeMemberPath");
        }
        get highlightedShapeMemberPath() {
            return this.fe;
        }
        set highlightedShapeMemberPath(a) {
            this.fe = a;
            this.g("HighlightedShapeMemberPath");
        }
        get xAxisRef() {
            return this.fi;
        }
        set xAxisRef(a) {
            this.fi = a;
            this.g("XAxisRef");
        }
        get yAxisRef() {
            return this.fj;
        }
        set yAxisRef(a) {
            this.fj = a;
            this.g("YAxisRef");
        }
        get shapeFilterResolution() {
            return this.e5;
        }
        set shapeFilterResolution(a) {
            this.e5 = a;
            this.g("ShapeFilterResolution");
        }
        get assigningShapeStyleRef() {
            return this.fd;
        }
        set assigningShapeStyleRef(a) {
            this.fd = a;
            this.g("AssigningShapeStyleRef");
        }
        get assigningShapeMarkerStyleRef() {
            return this.fc;
        }
        set assigningShapeMarkerStyleRef(a) {
            this.fc = a;
            this.g("AssigningShapeMarkerStyleRef");
        }
        get styleShapeRef() {
            return this.fh;
        }
        set styleShapeRef(a) {
            this.fh = a;
            this.g("StyleShapeRef");
        }
    }
    ShapeSeriesBaseDescription.$t = markType(ShapeSeriesBaseDescription, 'ShapeSeriesBaseDescription', SeriesDescription.$);
    return ShapeSeriesBaseDescription;
})();
