// igniteui-angular-core/esm2015/lib/StackedSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let StackedSeriesBaseDescription = /*@__PURE__*/ (() => {
    class StackedSeriesBaseDescription extends CategorySeriesDescription {
        constructor() {
            super();
            this.f1 = null;
            this.f3 = false;
            this.f4 = false;
            this.f7 = null;
        }
        get_type() {
            return "StackedSeriesBase";
        }
        get series() {
            return this.f1;
        }
        set series(a) {
            this.f1 = a;
            this.g("Series");
        }
        get autoGenerateSeries() {
            return this.f3;
        }
        set autoGenerateSeries(a) {
            this.f3 = a;
            this.g("AutoGenerateSeries");
        }
        get reverseLegendOrder() {
            return this.f4;
        }
        set reverseLegendOrder(a) {
            this.f4 = a;
            this.g("ReverseLegendOrder");
        }
        get seriesCreatedRef() {
            return this.f7;
        }
        set seriesCreatedRef(a) {
            this.f7 = a;
            this.g("SeriesCreatedRef");
        }
    }
    StackedSeriesBaseDescription.$t = markType(StackedSeriesBaseDescription, 'StackedSeriesBaseDescription', CategorySeriesDescription.$);
    return StackedSeriesBaseDescription;
})();
