// igniteui-angular-core/esm2015/lib/StackedBarSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { VerticalStackedSeriesBaseDescription } from "./VerticalStackedSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let StackedBarSeriesDescription = /*@__PURE__*/ (() => {
    class StackedBarSeriesDescription extends VerticalStackedSeriesBaseDescription {
        constructor() {
            super();
            this.gd = 0;
            this.ge = 0;
        }
        get_type() {
            return "StackedBarSeries";
        }
        get radiusX() {
            return this.gd;
        }
        set radiusX(a) {
            this.gd = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.ge;
        }
        set radiusY(a) {
            this.ge = a;
            this.g("RadiusY");
        }
    }
    StackedBarSeriesDescription.$t = markType(StackedBarSeriesDescription, 'StackedBarSeriesDescription', VerticalStackedSeriesBaseDescription.$);
    return StackedBarSeriesDescription;
})();
