// igniteui-angular-core/esm2015/lib/ToolActionRadioDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionRadioDescription = /*@__PURE__*/ (() => {
    class ToolActionRadioDescription extends ToolActionDescription {
        constructor() {
            super();
            this.dr = null;
            this.di = false;
            this.dj = false;
            this.du = null;
            this.dp = null;
            this.ds = null;
            this.dm = null;
            this.dv = null;
            this.dq = null;
            this.dt = null;
            this.dn = null;
        }
        get_type() {
            return "ToolActionRadio";
        }
        get channel() {
            return this.dr;
        }
        set channel(a) {
            this.dr = a;
            this.g("Channel");
        }
        get isChecked() {
            return this.di;
        }
        set isChecked(a) {
            this.di = a;
            this.g("IsChecked");
        }
        get isManual() {
            return this.dj;
        }
        set isManual(a) {
            this.dj = a;
            this.g("IsManual");
        }
        get uncheckedBackgroundColor() {
            return this.du;
        }
        set uncheckedBackgroundColor(a) {
            this.du = a;
            this.g("UncheckedBackgroundColor");
        }
        get actualUncheckedBackgroundColor() {
            return this.dp;
        }
        set actualUncheckedBackgroundColor(a) {
            this.dp = a;
            this.g("ActualUncheckedBackgroundColor");
        }
        get checkedBackgroundColor() {
            return this.ds;
        }
        set checkedBackgroundColor(a) {
            this.ds = a;
            this.g("CheckedBackgroundColor");
        }
        get actualCheckedBackgroundColor() {
            return this.dm;
        }
        set actualCheckedBackgroundColor(a) {
            this.dm = a;
            this.g("ActualCheckedBackgroundColor");
        }
        get uncheckedBorderColor() {
            return this.dv;
        }
        set uncheckedBorderColor(a) {
            this.dv = a;
            this.g("UncheckedBorderColor");
        }
        get actualUncheckedBorderColor() {
            return this.dq;
        }
        set actualUncheckedBorderColor(a) {
            this.dq = a;
            this.g("ActualUncheckedBorderColor");
        }
        get checkedBorderColor() {
            return this.dt;
        }
        set checkedBorderColor(a) {
            this.dt = a;
            this.g("CheckedBorderColor");
        }
        get actualCheckedBorderColor() {
            return this.dn;
        }
        set actualCheckedBorderColor(a) {
            this.dn = a;
            this.g("ActualCheckedBorderColor");
        }
    }
    ToolActionRadioDescription.$t = markType(ToolActionRadioDescription, 'ToolActionRadioDescription', ToolActionDescription.$);
    return ToolActionRadioDescription;
})();
