// igniteui-angular-core/esm2015/lib/ToolActionCheckboxDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionCheckboxDescription = /*@__PURE__*/ (() => {
    class ToolActionCheckboxDescription extends ToolActionDescription {
        constructor() {
            super();
            this.di = false;
            this.dr = null;
            this.dm = null;
            this.dp = null;
            this.dk = null;
            this.ds = null;
            this.dn = null;
            this.dq = null;
            this.dl = null;
        }
        get_type() {
            return "ToolActionCheckbox";
        }
        get isChecked() {
            return this.di;
        }
        set isChecked(a) {
            this.di = a;
            this.g("IsChecked");
        }
        get uncheckedBackgroundColor() {
            return this.dr;
        }
        set uncheckedBackgroundColor(a) {
            this.dr = a;
            this.g("UncheckedBackgroundColor");
        }
        get actualUncheckedBackgroundColor() {
            return this.dm;
        }
        set actualUncheckedBackgroundColor(a) {
            this.dm = a;
            this.g("ActualUncheckedBackgroundColor");
        }
        get checkedBackgroundColor() {
            return this.dp;
        }
        set checkedBackgroundColor(a) {
            this.dp = a;
            this.g("CheckedBackgroundColor");
        }
        get actualCheckedBackgroundColor() {
            return this.dk;
        }
        set actualCheckedBackgroundColor(a) {
            this.dk = a;
            this.g("ActualCheckedBackgroundColor");
        }
        get uncheckedBorderColor() {
            return this.ds;
        }
        set uncheckedBorderColor(a) {
            this.ds = a;
            this.g("UncheckedBorderColor");
        }
        get actualUncheckedBorderColor() {
            return this.dn;
        }
        set actualUncheckedBorderColor(a) {
            this.dn = a;
            this.g("ActualUncheckedBorderColor");
        }
        get checkedBorderColor() {
            return this.dq;
        }
        set checkedBorderColor(a) {
            this.dq = a;
            this.g("CheckedBorderColor");
        }
        get actualCheckedBorderColor() {
            return this.dl;
        }
        set actualCheckedBorderColor(a) {
            this.dl = a;
            this.g("ActualCheckedBorderColor");
        }
    }
    ToolActionCheckboxDescription.$t = markType(ToolActionCheckboxDescription, 'ToolActionCheckboxDescription', ToolActionDescription.$);
    return ToolActionCheckboxDescription;
})();
