// igniteui-angular-core/esm2015/lib/ToolActionInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionInfoDescription = /*@__PURE__*/ (() => {
    class ToolActionInfoDescription extends Description {
        constructor() {
            super();
            this.as = null;
            this.an = null;
            this.al = null;
            this.am = null;
            this.x = 0;
            this.y = 0;
            this.w = 0;
            this.z = 0;
            this.ar = null;
            this.av = null;
            this.at = null;
            this.aq = null;
            this.ap = null;
            this.s = 0;
            this.t = 0;
            this.u = 0;
            this.r = 0;
            this.v = 0;
            this.q = 0;
            this.l = false;
            this.m = false;
            this.au = null;
            this.ao = null;
            this.k = false;
            this.h = null;
        }
        get_type() {
            return "ToolActionInfo";
        }
        get type() {
            return this.get_type();
        }
        get parentId() {
            return this.as;
        }
        set parentId(a) {
            this.as = a;
            this.g("ParentId");
        }
        get commandId() {
            return this.an;
        }
        set commandId(a) {
            this.an = a;
            this.g("CommandId");
        }
        get commandArgument() {
            return this.al;
        }
        set commandArgument(a) {
            this.al = a;
            this.g("CommandArgument");
        }
        get commandArgumentValueRef() {
            return this.am;
        }
        set commandArgumentValueRef(a) {
            this.am = a;
            this.g("CommandArgumentValueRef");
        }
        get paddingLeft() {
            return this.x;
        }
        set paddingLeft(a) {
            this.x = a;
            this.g("PaddingLeft");
        }
        get paddingRight() {
            return this.y;
        }
        set paddingRight(a) {
            this.y = a;
            this.g("PaddingRight");
        }
        get paddingBottom() {
            return this.w;
        }
        set paddingBottom(a) {
            this.w = a;
            this.g("PaddingBottom");
        }
        get paddingTop() {
            return this.z;
        }
        set paddingTop(a) {
            this.z = a;
            this.g("PaddingTop");
        }
        get name() {
            return this.ar;
        }
        set name(a) {
            this.ar = a;
            this.g("Name");
        }
        get title() {
            return this.av;
        }
        set title(a) {
            this.av = a;
            this.g("Title");
        }
        get subtitle() {
            return this.at;
        }
        set subtitle(a) {
            this.at = a;
            this.g("Subtitle");
        }
        get iconName() {
            return this.aq;
        }
        set iconName(a) {
            this.aq = a;
            this.g("IconName");
        }
        get iconCollectionName() {
            return this.ap;
        }
        set iconCollectionName(a) {
            this.ap = a;
            this.g("IconCollectionName");
        }
        get iconViewBoxLeft() {
            return this.s;
        }
        set iconViewBoxLeft(a) {
            this.s = a;
            this.g("IconViewBoxLeft");
        }
        get iconViewBoxTop() {
            return this.t;
        }
        set iconViewBoxTop(a) {
            this.t = a;
            this.g("IconViewBoxTop");
        }
        get iconViewBoxWidth() {
            return this.u;
        }
        set iconViewBoxWidth(a) {
            this.u = a;
            this.g("IconViewBoxWidth");
        }
        get iconViewBoxHeight() {
            return this.r;
        }
        set iconViewBoxHeight(a) {
            this.r = a;
            this.g("IconViewBoxHeight");
        }
        get iconWidth() {
            return this.v;
        }
        set iconWidth(a) {
            this.v = a;
            this.g("IconWidth");
        }
        get iconHeight() {
            return this.q;
        }
        set iconHeight(a) {
            this.q = a;
            this.g("IconHeight");
        }
        get isDisabled() {
            return this.l;
        }
        set isDisabled(a) {
            this.l = a;
            this.g("IsDisabled");
        }
        get isHighlighted() {
            return this.m;
        }
        set isHighlighted(a) {
            this.m = a;
            this.g("IsHighlighted");
        }
        get textColor() {
            return this.au;
        }
        set textColor(a) {
            this.au = a;
            this.g("TextColor");
        }
        get disabledTextColor() {
            return this.ao;
        }
        set disabledTextColor(a) {
            this.ao = a;
            this.g("DisabledTextColor");
        }
        get closeOnExecute() {
            return this.k;
        }
        set closeOnExecute(a) {
            this.k = a;
            this.g("CloseOnExecute");
        }
        get actions() {
            return this.h;
        }
        set actions(a) {
            this.h = a;
            this.g("Actions");
        }
    }
    ToolActionInfoDescription.$t = markType(ToolActionInfoDescription, 'ToolActionInfoDescription', Description.$);
    ToolActionInfoDescription.__marshalByValue = true;
    ToolActionInfoDescription.__marshalByValueAlias = "ToolActionInfo";
    return ToolActionInfoDescription;
})();
