// igniteui-angular-core/esm2015/lib/ToolActionButtonDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionButtonDescription = /*@__PURE__*/ (() => {
    class ToolActionButtonDescription extends ToolActionDescription {
        constructor() {
            super();
            this.dk = null;
            this.di = 0;
        }
        get_type() {
            return "ToolActionButton";
        }
        get displayType() {
            return this.dk;
        }
        set displayType(a) {
            this.dk = a;
            this.g("DisplayType");
        }
        get cornerRadius() {
            return this.di;
        }
        set cornerRadius(a) {
            this.di = a;
            this.g("CornerRadius");
        }
    }
    ToolActionButtonDescription.$t = markType(ToolActionButtonDescription, 'ToolActionButtonDescription', ToolActionDescription.$);
    return ToolActionButtonDescription;
})();
