// igniteui-angular-core/esm2015/lib/ToolActionCheckboxListDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionCheckboxListDescription = /*@__PURE__*/ (() => {
    class ToolActionCheckboxListDescription extends ToolActionDescription {
        constructor() {
            super();
            this.dn = null;
            this.dp = null;
            this.dm = null;
            this.di = null;
            this.dj = null;
        }
        get_type() {
            return "ToolActionCheckboxList";
        }
        get dataSourceRef() {
            return this.dn;
        }
        set dataSourceRef(a) {
            this.dn = a;
            this.g("DataSourceRef");
        }
        get selectedMemberPath() {
            return this.dp;
        }
        set selectedMemberPath(a) {
            this.dp = a;
            this.g("SelectedMemberPath");
        }
        get dataMemberPath() {
            return this.dm;
        }
        set dataMemberPath(a) {
            this.dm = a;
            this.g("DataMemberPath");
        }
        get primaryKey() {
            return this.di;
        }
        set primaryKey(a) {
            this.di = a;
            this.g("PrimaryKey");
        }
        get selectedKeys() {
            return this.dj;
        }
        set selectedKeys(a) {
            this.dj = a;
            this.g("SelectedKeys");
        }
    }
    ToolActionCheckboxListDescription.$t = markType(ToolActionCheckboxListDescription, 'ToolActionCheckboxListDescription', ToolActionDescription.$);
    return ToolActionCheckboxListDescription;
})();
