// igniteui-angular-core/esm2015/lib/ToolActionCheckboxListInfoDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionInfoDescription } from "./ToolActionInfoDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionCheckboxListInfoDescription = /*@__PURE__*/ (() => {
    class ToolActionCheckboxListInfoDescription extends ToolActionInfoDescription {
        constructor() {
            super();
            this.bd = null;
            this.bc = null;
            this.be = null;
            this.a8 = null;
        }
        get_type() {
            return "ToolActionCheckboxListInfo";
        }
        get dataSourceRef() {
            return this.bd;
        }
        set dataSourceRef(a) {
            this.bd = a;
            this.g("DataSourceRef");
        }
        get dataMemberPath() {
            return this.bc;
        }
        set dataMemberPath(a) {
            this.bc = a;
            this.g("DataMemberPath");
        }
        get selectedMemberPath() {
            return this.be;
        }
        set selectedMemberPath(a) {
            this.be = a;
            this.g("SelectedMemberPath");
        }
        get primaryKey() {
            return this.a8;
        }
        set primaryKey(a) {
            this.a8 = a;
            this.g("PrimaryKey");
        }
    }
    ToolActionCheckboxListInfoDescription.$t = markType(ToolActionCheckboxListInfoDescription, 'ToolActionCheckboxListInfoDescription', ToolActionInfoDescription.$);
    ToolActionCheckboxListInfoDescription.__marshalByValue1 = true;
    ToolActionCheckboxListInfoDescription.__marshalByValueAlias1 = "ToolActionCheckboxListInfo";
    return ToolActionCheckboxListInfoDescription;
})();
