// igniteui-angular-core/esm2015/lib/ToolActionGroupHeaderDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ToolActionGroupHeaderDescription = /*@__PURE__*/ (() => {
    class ToolActionGroupHeaderDescription extends ToolActionDescription {
        constructor() {
            super();
            this.di = null;
            this.dj = null;
        }
        get_type() {
            return "ToolActionGroupHeader";
        }
        get actualBackIconColor() {
            return this.di;
        }
        set actualBackIconColor(a) {
            this.di = a;
            this.g("ActualBackIconColor");
        }
        get backIconColor() {
            return this.dj;
        }
        set backIconColor(a) {
            this.dj = a;
            this.g("BackIconColor");
        }
    }
    ToolActionGroupHeaderDescription.$t = markType(ToolActionGroupHeaderDescription, 'ToolActionGroupHeaderDescription', ToolActionDescription.$);
    return ToolActionGroupHeaderDescription;
})();
