// igniteui-angular-core/esm2015/lib/ValueOverlayDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ValueOverlayDescription = /*@__PURE__*/ (() => {
    class ValueOverlayDescription extends SeriesDescription {
        constructor() {
            super();
            this.fn = null;
            this.e9 = 0;
            this.e1 = false;
            this.fk = null;
            this.fm = null;
            this.fj = null;
            this.e3 = 0;
            this.fh = 0;
            this.fl = null;
            this.e5 = 0;
            this.e7 = 0;
            this.e6 = 0;
            this.e4 = 0;
            this.e8 = 0;
        }
        get_type() {
            return "ValueOverlay";
        }
        get axisRef() {
            return this.fn;
        }
        set axisRef(a) {
            this.fn = a;
            this.g("AxisRef");
        }
        get value() {
            return this.e9;
        }
        set value(a) {
            this.e9 = a;
            this.g("Value");
        }
        get isAxisAnnotationEnabled() {
            return this.e1;
        }
        set isAxisAnnotationEnabled(a) {
            this.e1 = a;
            this.g("IsAxisAnnotationEnabled");
        }
        get axisAnnotationFormatLabelRef() {
            return this.fk;
        }
        set axisAnnotationFormatLabelRef(a) {
            this.fk = a;
            this.g("AxisAnnotationFormatLabelRef");
        }
        get axisAnnotationTextColor() {
            return this.fm;
        }
        set axisAnnotationTextColor(a) {
            this.fm = a;
            this.g("AxisAnnotationTextColor");
        }
        get axisAnnotationBackground() {
            return this.fj;
        }
        set axisAnnotationBackground(a) {
            this.fj = a;
            this.g("AxisAnnotationBackground");
        }
        get axisAnnotationBackgroundCornerRadius() {
            return this.e3;
        }
        set axisAnnotationBackgroundCornerRadius(a) {
            this.e3 = a;
            this.g("AxisAnnotationBackgroundCornerRadius");
        }
        get axisAnnotationInterpolatedValuePrecision() {
            return this.fh;
        }
        set axisAnnotationInterpolatedValuePrecision(a) {
            this.fh = a;
            this.g("AxisAnnotationInterpolatedValuePrecision");
        }
        get axisAnnotationOutline() {
            return this.fl;
        }
        set axisAnnotationOutline(a) {
            this.fl = a;
            this.g("AxisAnnotationOutline");
        }
        get axisAnnotationPaddingLeft() {
            return this.e5;
        }
        set axisAnnotationPaddingLeft(a) {
            this.e5 = a;
            this.g("AxisAnnotationPaddingLeft");
        }
        get axisAnnotationPaddingTop() {
            return this.e7;
        }
        set axisAnnotationPaddingTop(a) {
            this.e7 = a;
            this.g("AxisAnnotationPaddingTop");
        }
        get axisAnnotationPaddingRight() {
            return this.e6;
        }
        set axisAnnotationPaddingRight(a) {
            this.e6 = a;
            this.g("AxisAnnotationPaddingRight");
        }
        get axisAnnotationPaddingBottom() {
            return this.e4;
        }
        set axisAnnotationPaddingBottom(a) {
            this.e4 = a;
            this.g("AxisAnnotationPaddingBottom");
        }
        get axisAnnotationStrokeThickness() {
            return this.e8;
        }
        set axisAnnotationStrokeThickness(a) {
            this.e8 = a;
            this.g("AxisAnnotationStrokeThickness");
        }
    }
    ValueOverlayDescription.$t = markType(ValueOverlayDescription, 'ValueOverlayDescription', SeriesDescription.$);
    return ValueOverlayDescription;
})();
