// igniteui-angular-core/esm2015/lib/ValueLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ValueLayerDescription = /*@__PURE__*/ (() => {
    class ValueLayerDescription extends AnnotationLayerDescription {
        constructor() {
            super();
            this.ha = null;
            this.g5 = null;
            this.hb = null;
            this.g8 = null;
            this.g9 = null;
            this.g6 = null;
            this.g7 = null;
            this.f9 = false;
            this.f7 = false;
            this.hd = null;
            this.hh = null;
            this.hf = null;
            this.hc = null;
            this.gd = 0;
            this.g1 = 0;
            this.he = null;
            this.gf = 0;
            this.gh = 0;
            this.gg = 0;
            this.ge = 0;
            this.gl = 0;
            this.gn = 0;
            this.gm = 0;
            this.gk = 0;
            this.gi = 0;
            this.hj = null;
            this.hg = null;
            this.gj = 0;
            this.g2 = 0;
            this.hi = null;
            this.go = 0;
            this.f8 = false;
        }
        get_type() {
            return "ValueLayer";
        }
        get valueMode() {
            return this.ha;
        }
        set valueMode(a) {
            this.ha = a;
            this.g("ValueMode");
        }
        get horizontalLineStroke() {
            return this.g5;
        }
        set horizontalLineStroke(a) {
            this.g5 = a;
            this.g("HorizontalLineStroke");
        }
        get verticalLineStroke() {
            return this.hb;
        }
        set verticalLineStroke(a) {
            this.hb = a;
            this.g("VerticalLineStroke");
        }
        get targetSeriesName() {
            return this.g8;
        }
        set targetSeriesName(a) {
            this.g8 = a;
            this.g("TargetSeriesName");
        }
        get targetSeriesRef() {
            return this.g9;
        }
        set targetSeriesRef(a) {
            this.g9 = a;
            this.g("TargetSeriesRef");
        }
        get targetAxisName() {
            return this.g6;
        }
        set targetAxisName(a) {
            this.g6 = a;
            this.g("TargetAxisName");
        }
        get targetAxisRef() {
            return this.g7;
        }
        set targetAxisRef(a) {
            this.g7 = a;
            this.g("TargetAxisRef");
        }
        get useInterpolation() {
            return this.f9;
        }
        set useInterpolation(a) {
            this.f9 = a;
            this.g("UseInterpolation");
        }
        get isAxisAnnotationEnabled() {
            return this.f7;
        }
        set isAxisAnnotationEnabled(a) {
            this.f7 = a;
            this.g("IsAxisAnnotationEnabled");
        }
        get xAxisAnnotationFormatLabelRef() {
            return this.hd;
        }
        set xAxisAnnotationFormatLabelRef(a) {
            this.hd = a;
            this.g("XAxisAnnotationFormatLabelRef");
        }
        get yAxisAnnotationFormatLabelRef() {
            return this.hh;
        }
        set yAxisAnnotationFormatLabelRef(a) {
            this.hh = a;
            this.g("YAxisAnnotationFormatLabelRef");
        }
        get xAxisAnnotationTextColor() {
            return this.hf;
        }
        set xAxisAnnotationTextColor(a) {
            this.hf = a;
            this.g("XAxisAnnotationTextColor");
        }
        get xAxisAnnotationBackground() {
            return this.hc;
        }
        set xAxisAnnotationBackground(a) {
            this.hc = a;
            this.g("XAxisAnnotationBackground");
        }
        get xAxisAnnotationBackgroundCornerRadius() {
            return this.gd;
        }
        set xAxisAnnotationBackgroundCornerRadius(a) {
            this.gd = a;
            this.g("XAxisAnnotationBackgroundCornerRadius");
        }
        get xAxisAnnotationInterpolatedValuePrecision() {
            return this.g1;
        }
        set xAxisAnnotationInterpolatedValuePrecision(a) {
            this.g1 = a;
            this.g("XAxisAnnotationInterpolatedValuePrecision");
        }
        get xAxisAnnotationOutline() {
            return this.he;
        }
        set xAxisAnnotationOutline(a) {
            this.he = a;
            this.g("XAxisAnnotationOutline");
        }
        get xAxisAnnotationPaddingLeft() {
            return this.gf;
        }
        set xAxisAnnotationPaddingLeft(a) {
            this.gf = a;
            this.g("XAxisAnnotationPaddingLeft");
        }
        get xAxisAnnotationPaddingTop() {
            return this.gh;
        }
        set xAxisAnnotationPaddingTop(a) {
            this.gh = a;
            this.g("XAxisAnnotationPaddingTop");
        }
        get xAxisAnnotationPaddingRight() {
            return this.gg;
        }
        set xAxisAnnotationPaddingRight(a) {
            this.gg = a;
            this.g("XAxisAnnotationPaddingRight");
        }
        get xAxisAnnotationPaddingBottom() {
            return this.ge;
        }
        set xAxisAnnotationPaddingBottom(a) {
            this.ge = a;
            this.g("XAxisAnnotationPaddingBottom");
        }
        get yAxisAnnotationPaddingLeft() {
            return this.gl;
        }
        set yAxisAnnotationPaddingLeft(a) {
            this.gl = a;
            this.g("YAxisAnnotationPaddingLeft");
        }
        get yAxisAnnotationPaddingTop() {
            return this.gn;
        }
        set yAxisAnnotationPaddingTop(a) {
            this.gn = a;
            this.g("YAxisAnnotationPaddingTop");
        }
        get yAxisAnnotationPaddingRight() {
            return this.gm;
        }
        set yAxisAnnotationPaddingRight(a) {
            this.gm = a;
            this.g("YAxisAnnotationPaddingRight");
        }
        get yAxisAnnotationPaddingBottom() {
            return this.gk;
        }
        set yAxisAnnotationPaddingBottom(a) {
            this.gk = a;
            this.g("YAxisAnnotationPaddingBottom");
        }
        get xAxisAnnotationStrokeThickness() {
            return this.gi;
        }
        set xAxisAnnotationStrokeThickness(a) {
            this.gi = a;
            this.g("XAxisAnnotationStrokeThickness");
        }
        get yAxisAnnotationTextColor() {
            return this.hj;
        }
        set yAxisAnnotationTextColor(a) {
            this.hj = a;
            this.g("YAxisAnnotationTextColor");
        }
        get yAxisAnnotationBackground() {
            return this.hg;
        }
        set yAxisAnnotationBackground(a) {
            this.hg = a;
            this.g("YAxisAnnotationBackground");
        }
        get yAxisAnnotationBackgroundCornerRadius() {
            return this.gj;
        }
        set yAxisAnnotationBackgroundCornerRadius(a) {
            this.gj = a;
            this.g("YAxisAnnotationBackgroundCornerRadius");
        }
        get yAxisAnnotationInterpolatedValuePrecision() {
            return this.g2;
        }
        set yAxisAnnotationInterpolatedValuePrecision(a) {
            this.g2 = a;
            this.g("YAxisAnnotationInterpolatedValuePrecision");
        }
        get yAxisAnnotationOutline() {
            return this.hi;
        }
        set yAxisAnnotationOutline(a) {
            this.hi = a;
            this.g("YAxisAnnotationOutline");
        }
        get yAxisAnnotationStrokeThickness() {
            return this.go;
        }
        set yAxisAnnotationStrokeThickness(a) {
            this.go = a;
            this.g("YAxisAnnotationStrokeThickness");
        }
        get skipUnknownValues() {
            return this.f8;
        }
        set skipUnknownValues(a) {
            this.f8 = a;
            this.g("SkipUnknownValues");
        }
    }
    ValueLayerDescription.$t = markType(ValueLayerDescription, 'ValueLayerDescription', AnnotationLayerDescription.$);
    return ValueLayerDescription;
})();
