// igniteui-angular-core/esm2015/lib/WaterfallSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalAnchoredCategorySeriesDescription } from "./HorizontalAnchoredCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WaterfallSeriesDescription = /*@__PURE__*/ (() => {
    class WaterfallSeriesDescription extends HorizontalAnchoredCategorySeriesDescription {
        constructor() {
            super();
            this.gx = null;
            this.gt = 0;
            this.gu = 0;
            this.gy = null;
        }
        get_type() {
            return "WaterfallSeries";
        }
        get negativeBrush() {
            return this.gx;
        }
        set negativeBrush(a) {
            this.gx = a;
            this.g("NegativeBrush");
        }
        get radiusX() {
            return this.gt;
        }
        set radiusX(a) {
            this.gt = a;
            this.g("RadiusX");
        }
        get radiusY() {
            return this.gu;
        }
        set radiusY(a) {
            this.gu = a;
            this.g("RadiusY");
        }
        get negativeOutline() {
            return this.gy;
        }
        set negativeOutline(a) {
            this.gy = a;
            this.g("NegativeOutline");
        }
    }
    WaterfallSeriesDescription.$t = markType(WaterfallSeriesDescription, 'WaterfallSeriesDescription', HorizontalAnchoredCategorySeriesDescription.$);
    return WaterfallSeriesDescription;
})();
