// igniteui-angular-core/esm2015/lib/WebCheckboxChangeEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebCheckboxChangeEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebCheckboxChangeEventArgsDetailDescription extends Description {
        constructor() {
            super();
            this.i = false;
            this.l = null;
        }
        get_type() {
            return "WebCheckboxChangeEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        get checked() {
            return this.i;
        }
        set checked(a) {
            this.i = a;
            this.g("Checked");
        }
        get value() {
            return this.l;
        }
        set value(a) {
            this.l = a;
            this.g("Value");
        }
    }
    WebCheckboxChangeEventArgsDetailDescription.$t = markType(WebCheckboxChangeEventArgsDetailDescription, 'WebCheckboxChangeEventArgsDetailDescription', Description.$);
    WebCheckboxChangeEventArgsDetailDescription.__marshalByValue = true;
    WebCheckboxChangeEventArgsDetailDescription.__marshalByValueAlias = "CheckboxChangeEventArgsDetail";
    return WebCheckboxChangeEventArgsDetailDescription;
})();
