// igniteui-angular-core/esm2015/lib/WebColumnVisibilityChangingEventArgsDetailDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnVisibilityChangingEventArgsDetailDescription = /*@__PURE__*/ (() => {
    class WebColumnVisibilityChangingEventArgsDetailDescription extends Description {
        constructor() {
            super();
            this.m = null;
            this.j = false;
            this.i = false;
        }
        get_type() {
            return "WebColumnVisibilityChangingEventArgsDetail";
        }
        get type() {
            return this.get_type();
        }
        get column() {
            return this.m;
        }
        set column(a) {
            this.m = a;
            this.g("Column");
        }
        get newValue() {
            return this.j;
        }
        set newValue(a) {
            this.j = a;
            this.g("NewValue");
        }
        get cancel() {
            return this.i;
        }
        set cancel(a) {
            this.i = a;
            this.g("Cancel");
        }
    }
    WebColumnVisibilityChangingEventArgsDetailDescription.$t = markType(WebColumnVisibilityChangingEventArgsDetailDescription, 'WebColumnVisibilityChangingEventArgsDetailDescription', Description.$);
    WebColumnVisibilityChangingEventArgsDetailDescription.__marshalByValue = true;
    WebColumnVisibilityChangingEventArgsDetailDescription.__marshalByValueAlias = "ColumnVisibilityChangingEventArgsDetail";
    return WebColumnVisibilityChangingEventArgsDetailDescription;
})();
