// igniteui-angular-core/esm2015/lib/WebColumnGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebColumnDescription } from "./WebColumnDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebColumnGroupDescription = /*@__PURE__*/ (() => {
    class WebColumnGroupDescription extends WebColumnDescription {
        constructor() {
            super();
            this.ck = null;
            this.cn = false;
            this.co = false;
            this.cs = null;
        }
        get_type() {
            return "WebColumnGroup";
        }
        get children() {
            return this.ck;
        }
        set children(a) {
            this.ck = a;
            this.g("Children");
        }
        get collapsible() {
            return this.cn;
        }
        set collapsible(a) {
            this.cn = a;
            this.g("Collapsible");
        }
        get expanded() {
            return this.co;
        }
        set expanded(a) {
            this.co = a;
            this.g("Expanded");
        }
        get collapsibleIndicatorTemplateRef() {
            return this.cs;
        }
        set collapsibleIndicatorTemplateRef(a) {
            this.cs = a;
            this.g("CollapsibleIndicatorTemplateRef");
        }
    }
    WebColumnGroupDescription.$t = markType(WebColumnGroupDescription, 'WebColumnGroupDescription', WebColumnDescription.$);
    WebColumnGroupDescription.__canMarshalByValue1 = true;
    WebColumnGroupDescription.__marshalByValueAlias1 = "ColumnGroup";
    return WebColumnGroupDescription;
})();
