// igniteui-angular-core/esm2015/lib/WebHeadSelectorTemplateDetailsDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebHeadSelectorTemplateDetailsDescription = /*@__PURE__*/ (() => {
    class WebHeadSelectorTemplateDetailsDescription extends Description {
        constructor() {
            super();
            this.h = 0;
            this.i = 0;
        }
        get_type() {
            return "WebHeadSelectorTemplateDetails";
        }
        get type() {
            return this.get_type();
        }
        get selectedCount() {
            return this.h;
        }
        set selectedCount(a) {
            this.h = a;
            this.g("SelectedCount");
        }
        get totalCount() {
            return this.i;
        }
        set totalCount(a) {
            this.i = a;
            this.g("TotalCount");
        }
    }
    WebHeadSelectorTemplateDetailsDescription.$t = markType(WebHeadSelectorTemplateDetailsDescription, 'WebHeadSelectorTemplateDetailsDescription', Description.$);
    return WebHeadSelectorTemplateDetailsDescription;
})();
