// igniteui-angular-core/esm2015/lib/WebInputBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebInputBaseDescription = /*@__PURE__*/ (() => {
    class WebInputBaseDescription extends Description {
        constructor() {
            super();
            this.j = false;
            this.k = false;
            this.w = null;
            this.u = null;
            this.h = false;
            this.l = false;
            this.v = null;
            this.i = false;
            this.t = null;
            this.s = null;
            this.r = null;
        }
        get_type() {
            return "WebInputBase";
        }
        get type() {
            return this.get_type();
        }
        get outlined() {
            return this.j;
        }
        set outlined(a) {
            this.j = a;
            this.g("Outlined");
        }
        get readOnly() {
            return this.k;
        }
        set readOnly(a) {
            this.k = a;
            this.g("ReadOnly");
        }
        get placeholder() {
            return this.w;
        }
        set placeholder(a) {
            this.w = a;
            this.g("Placeholder");
        }
        get label() {
            return this.u;
        }
        set label(a) {
            this.u = a;
            this.g("Label");
        }
        get disabled() {
            return this.h;
        }
        set disabled(a) {
            this.h = a;
            this.g("Disabled");
        }
        get required() {
            return this.l;
        }
        set required(a) {
            this.l = a;
            this.g("Required");
        }
        get name() {
            return this.v;
        }
        set name(a) {
            this.v = a;
            this.g("Name");
        }
        get invalid() {
            return this.i;
        }
        set invalid(a) {
            this.i = a;
            this.g("Invalid");
        }
        get inputOcurredRef() {
            return this.t;
        }
        set inputOcurredRef(a) {
            this.t = a;
            this.g("InputOcurredRef");
        }
        get focusRef() {
            return this.s;
        }
        set focusRef(a) {
            this.s = a;
            this.g("FocusRef");
        }
        get blurRef() {
            return this.r;
        }
        set blurRef(a) {
            this.r = a;
            this.g("BlurRef");
        }
    }
    WebInputBaseDescription.$t = markType(WebInputBaseDescription, 'WebInputBaseDescription', Description.$);
    return WebInputBaseDescription;
})();
