// igniteui-angular-core/esm2015/lib/WebInputDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebInputBaseDescription } from "./WebInputBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebInputDescription = /*@__PURE__*/ (() => {
    class WebInputDescription extends WebInputBaseDescription {
        constructor() {
            super();
            this.av = null;
            this.aq = null;
            this.ar = null;
            this.au = null;
            this.aj = 0;
            this.ai = 0;
            this.at = null;
            this.as = null;
            this.ak = 0;
            this.ae = false;
            this.ao = null;
            this.af = false;
            this.ap = null;
        }
        get_type() {
            return "WebInput";
        }
        get valueRef() {
            return this.av;
        }
        set valueRef(a) {
            this.av = a;
            this.g("ValueRef");
        }
        get displayType() {
            return this.aq;
        }
        set displayType(a) {
            this.aq = a;
            this.g("DisplayType");
        }
        get inputMode() {
            return this.ar;
        }
        set inputMode(a) {
            this.ar = a;
            this.g("InputMode");
        }
        get pattern() {
            return this.au;
        }
        set pattern(a) {
            this.au = a;
            this.g("Pattern");
        }
        get minLength() {
            return this.aj;
        }
        set minLength(a) {
            this.aj = a;
            this.g("MinLength");
        }
        get maxLength() {
            return this.ai;
        }
        set maxLength(a) {
            this.ai = a;
            this.g("MaxLength");
        }
        get min() {
            return this.at;
        }
        set min(a) {
            this.at = a;
            this.g("Min");
        }
        get max() {
            return this.as;
        }
        set max(a) {
            this.as = a;
            this.g("Max");
        }
        get step() {
            return this.ak;
        }
        set step(a) {
            this.ak = a;
            this.g("Step");
        }
        get autofocus() {
            return this.ae;
        }
        set autofocus(a) {
            this.ae = a;
            this.g("Autofocus");
        }
        get autocomplete() {
            return this.ao;
        }
        set autocomplete(a) {
            this.ao = a;
            this.g("Autocomplete");
        }
        get validateOnly() {
            return this.af;
        }
        set validateOnly(a) {
            this.af = a;
            this.g("ValidateOnly");
        }
        get changeRef() {
            return this.ap;
        }
        set changeRef(a) {
            this.ap = a;
            this.g("ChangeRef");
        }
    }
    WebInputDescription.$t = markType(WebInputDescription, 'WebInputDescription', WebInputBaseDescription.$);
    return WebInputDescription;
})();
