// igniteui-angular-core/esm2015/lib/WebRadioGroupDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRadioGroupDescription = /*@__PURE__*/ (() => {
    class WebRadioGroupDescription extends Description {
        constructor() {
            super();
            this.h = null;
            this.j = null;
            this.k = null;
            this.i = null;
        }
        get_type() {
            return "WebRadioGroup";
        }
        get type() {
            return this.get_type();
        }
        get alignment() {
            return this.h;
        }
        set alignment(a) {
            this.h = a;
            this.g("Alignment");
        }
        get name() {
            return this.j;
        }
        set name(a) {
            this.j = a;
            this.g("Name");
        }
        get value() {
            return this.k;
        }
        set value(a) {
            this.k = a;
            this.g("Value");
        }
        get changeRef() {
            return this.i;
        }
        set changeRef(a) {
            this.i = a;
            this.g("ChangeRef");
        }
    }
    WebRadioGroupDescription.$t = markType(WebRadioGroupDescription, 'WebRadioGroupDescription', Description.$);
    return WebRadioGroupDescription;
})();
