// igniteui-angular-core/esm2015/lib/WebRowIslandDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebRowIslandDescription = /*@__PURE__*/ (() => {
    class WebRowIslandDescription extends Description {
        constructor() {
            super();
            this.bw = null;
            this.j = null;
            this.h = null;
            this.ar = false;
            this.cw = null;
            this.az = false;
            this.aq = false;
            this.bg = 0;
            this.ap = false;
            this.i = null;
            this.au = false;
            this.cj = null;
            this.bo = null;
            this.c1 = null;
            this.bh = 0;
            this.x = null;
            this.v = null;
            this.db = null;
            this.dv = null;
            this.c5 = null;
            this.k = null;
            this.cf = null;
            this.dn = null;
            this.da = null;
            this.dj = null;
            this.dp = null;
            this.dd = null;
            this.cy = null;
            this.cx = null;
            this.ck = null;
            this.dz = null;
            this.d0 = null;
            this.d1 = null;
            this.n = null;
            this.l = null;
            this.af = null;
            this.co = null;
            this.aa = null;
            this.z = null;
            this.c2 = null;
            this.c4 = null;
            this.as = false;
            this.av = false;
            this.d9 = null;
            this.aw = false;
            this.c0 = null;
            this.ea = null;
            this.bf = 0;
            this.b9 = null;
            this.ci = null;
            this.at = false;
            this.ay = false;
            this.ch = null;
            this.aj = null;
            this.ao = false;
            this.an = false;
            this.cp = null;
            this.d6 = null;
            this.d5 = null;
            this.a0 = false;
            this.ad = null;
            this.ah = null;
            this.al = null;
            this.dy = null;
            this.cz = null;
            this.du = null;
            this.cg = null;
            this.m = null;
            this.bv = null;
            this.ds = null;
            this.b5 = null;
            this.bi = 0;
            this.ax = false;
            this.c3 = null;
            this.cs = null;
            this.ct = null;
            this.cd = null;
            this.cn = null;
            this.bp = null;
            this.cv = null;
            this.bq = null;
            this.dc = null;
            this.cq = null;
            this.d8 = null;
            this.dx = null;
            this.dt = null;
            this.b6 = null;
            this.b1 = null;
            this.b2 = null;
            this.bt = null;
            this.bu = null;
            this.br = null;
            this.bs = null;
            this.dl = null;
            this.di = null;
            this.dk = null;
            this.dm = null;
            this.bx = null;
            this.b4 = null;
            this.d2 = null;
            this.d3 = null;
            this.cl = null;
            this.cm = null;
            this.c9 = null;
            this.df = null;
            this.de = null;
            this.c8 = null;
            this.b3 = null;
            this.ca = null;
            this.ce = null;
            this.b8 = null;
            this.b7 = null;
            this.b0 = null;
            this.by = null;
            this.bz = null;
            this.cu = null;
            this.dh = null;
            this.dg = null;
            this.cr = null;
            this.dw = null;
            this.dr = null;
            this.dq = null;
            this.bn = null;
            this.d4 = null;
            this.d7 = null;
            this.c6 = null;
            this.c7 = null;
            this.cc = null;
            this.cb = null;
        }
        get_type() {
            return "WebRowIsland";
        }
        get type() {
            return this.get_type();
        }
        get childDataKey() {
            return this.bw;
        }
        set childDataKey(a) {
            this.bw = a;
            this.g("ChildDataKey");
        }
        get childLayoutList() {
            return this.j;
        }
        set childLayoutList(a) {
            this.j = a;
            this.g("ChildLayoutList");
        }
        get actionStripComponents() {
            return this.h;
        }
        set actionStripComponents(a) {
            this.h = a;
            this.g("ActionStripComponents");
        }
        get expandChildren() {
            return this.ar;
        }
        set expandChildren(a) {
            this.ar = a;
            this.g("ExpandChildren");
        }
        get hasChildrenKey() {
            return this.cw;
        }
        set hasChildrenKey(a) {
            this.cw = a;
            this.g("HasChildrenKey");
        }
        get showExpandAll() {
            return this.az;
        }
        set showExpandAll(a) {
            this.az = a;
            this.g("ShowExpandAll");
        }
        get batchEditing() {
            return this.aq;
        }
        set batchEditing(a) {
            this.aq = a;
            this.g("BatchEditing");
        }
        get snackbarDisplayTime() {
            return this.bg;
        }
        set snackbarDisplayTime(a) {
            this.bg = a;
            this.g("SnackbarDisplayTime");
        }
        get autoGenerate() {
            return this.ap;
        }
        set autoGenerate(a) {
            this.ap = a;
            this.g("AutoGenerate");
        }
        get autoGenerateExclude() {
            return this.i;
        }
        set autoGenerateExclude(a) {
            this.i = a;
            this.g("AutoGenerateExclude");
        }
        get moving() {
            return this.au;
        }
        set moving(a) {
            this.au = a;
            this.g("Moving");
        }
        get emptyGridTemplateRef() {
            return this.cj;
        }
        set emptyGridTemplateRef(a) {
            this.cj = a;
            this.g("EmptyGridTemplateRef");
        }
        get addRowEmptyTemplateRef() {
            return this.bo;
        }
        set addRowEmptyTemplateRef(a) {
            this.bo = a;
            this.g("AddRowEmptyTemplateRef");
        }
        get loadingGridTemplateRef() {
            return this.c1;
        }
        set loadingGridTemplateRef(a) {
            this.c1 = a;
            this.g("LoadingGridTemplateRef");
        }
        get summaryRowHeight() {
            return this.bh;
        }
        set summaryRowHeight(a) {
            this.bh = a;
            this.g("SummaryRowHeight");
        }
        get dataCloneStrategy() {
            return this.x;
        }
        set dataCloneStrategy(a) {
            this.x = a;
            this.g("DataCloneStrategy");
        }
        get clipboardOptions() {
            return this.v;
        }
        set clipboardOptions(a) {
            this.v = a;
            this.g("ClipboardOptions");
        }
        get rowClassesRef() {
            return this.db;
        }
        set rowClassesRef(a) {
            this.db = a;
            this.g("RowClassesRef");
        }
        get rowStylesRef() {
            return this.dv;
        }
        set rowStylesRef(a) {
            this.dv = a;
            this.g("RowStylesRef");
        }
        get primaryKey() {
            return this.c5;
        }
        set primaryKey(a) {
            this.c5 = a;
            this.g("PrimaryKey");
        }
        get columnList() {
            return this.k;
        }
        set columnList(a) {
            this.k = a;
            this.g("ColumnList");
        }
        get dragGhostCustomTemplateRef() {
            return this.cf;
        }
        set dragGhostCustomTemplateRef(a) {
            this.cf = a;
            this.g("DragGhostCustomTemplateRef");
        }
        get rowEditTextTemplateRef() {
            return this.dn;
        }
        set rowEditTextTemplateRef(a) {
            this.dn = a;
            this.g("RowEditTextTemplateRef");
        }
        get rowAddTextTemplateRef() {
            return this.da;
        }
        set rowAddTextTemplateRef(a) {
            this.da = a;
            this.g("RowAddTextTemplateRef");
        }
        get rowEditActionsTemplateRef() {
            return this.dj;
        }
        set rowEditActionsTemplateRef(a) {
            this.dj = a;
            this.g("RowEditActionsTemplateRef");
        }
        get rowExpandedIndicatorTemplateRef() {
            return this.dp;
        }
        set rowExpandedIndicatorTemplateRef(a) {
            this.dp = a;
            this.g("RowExpandedIndicatorTemplateRef");
        }
        get rowCollapsedIndicatorTemplateRef() {
            return this.dd;
        }
        set rowCollapsedIndicatorTemplateRef(a) {
            this.dd = a;
            this.g("RowCollapsedIndicatorTemplateRef");
        }
        get headerExpandedIndicatorTemplateRef() {
            return this.cy;
        }
        set headerExpandedIndicatorTemplateRef(a) {
            this.cy = a;
            this.g("HeaderExpandedIndicatorTemplateRef");
        }
        get headerCollapsedIndicatorTemplateRef() {
            return this.cx;
        }
        set headerCollapsedIndicatorTemplateRef(a) {
            this.cx = a;
            this.g("HeaderCollapsedIndicatorTemplateRef");
        }
        get excelStyleHeaderIconTemplateRef() {
            return this.ck;
        }
        set excelStyleHeaderIconTemplateRef(a) {
            this.ck = a;
            this.g("ExcelStyleHeaderIconTemplateRef");
        }
        get sortAscendingHeaderIconTemplateRef() {
            return this.dz;
        }
        set sortAscendingHeaderIconTemplateRef(a) {
            this.dz = a;
            this.g("SortAscendingHeaderIconTemplateRef");
        }
        get sortDescendingHeaderIconTemplateRef() {
            return this.d0;
        }
        set sortDescendingHeaderIconTemplateRef(a) {
            this.d0 = a;
            this.g("SortDescendingHeaderIconTemplateRef");
        }
        get sortHeaderIconTemplateRef() {
            return this.d1;
        }
        set sortHeaderIconTemplateRef(a) {
            this.d1 = a;
            this.g("SortHeaderIconTemplateRef");
        }
        get toolbar() {
            return this.n;
        }
        set toolbar(a) {
            this.n = a;
            this.g("Toolbar");
        }
        get paginationComponents() {
            return this.l;
        }
        set paginationComponents(a) {
            this.l = a;
            this.g("PaginationComponents");
        }
        get resourceStrings() {
            return this.af;
        }
        set resourceStrings(a) {
            this.af = a;
            this.g("ResourceStrings");
        }
        get filteringLogic() {
            return this.co;
        }
        set filteringLogic(a) {
            this.co = a;
            this.g("FilteringLogic");
        }
        get filteringExpressionsTree() {
            return this.aa;
        }
        set filteringExpressionsTree(a) {
            this.aa = a;
            this.g("FilteringExpressionsTree");
        }
        get advancedFilteringExpressionsTree() {
            return this.z;
        }
        set advancedFilteringExpressionsTree(a) {
            this.z = a;
            this.g("AdvancedFilteringExpressionsTree");
        }
        get locale() {
            return this.c2;
        }
        set locale(a) {
            this.c2 = a;
            this.g("Locale");
        }
        get pagingMode() {
            return this.c4;
        }
        set pagingMode(a) {
            this.c4 = a;
            this.g("PagingMode");
        }
        get hideRowSelectors() {
            return this.as;
        }
        set hideRowSelectors(a) {
            this.as = a;
            this.g("HideRowSelectors");
        }
        get rowDraggable() {
            return this.av;
        }
        set rowDraggable(a) {
            this.av = a;
            this.g("RowDraggable");
        }
        get validationTrigger() {
            return this.d9;
        }
        set validationTrigger(a) {
            this.d9 = a;
            this.g("ValidationTrigger");
        }
        get rowEditable() {
            return this.aw;
        }
        set rowEditable(a) {
            this.aw = a;
            this.g("RowEditable");
        }
        get height() {
            return this.c0;
        }
        set height(a) {
            this.c0 = a;
            this.g("Height");
        }
        get width() {
            return this.ea;
        }
        set width(a) {
            this.ea = a;
            this.g("Width");
        }
        get rowHeight() {
            return this.bf;
        }
        set rowHeight(a) {
            this.bf = a;
            this.g("RowHeight");
        }
        get columnWidth() {
            return this.b9;
        }
        set columnWidth(a) {
            this.b9 = a;
            this.g("ColumnWidth");
        }
        get emptyGridMessage() {
            return this.ci;
        }
        set emptyGridMessage(a) {
            this.ci = a;
            this.g("EmptyGridMessage");
        }
        get isLoading() {
            return this.at;
        }
        set isLoading(a) {
            this.at = a;
            this.g("IsLoading");
        }
        get shouldGenerate() {
            return this.ay;
        }
        set shouldGenerate(a) {
            this.ay = a;
            this.g("ShouldGenerate");
        }
        get emptyFilteredGridMessage() {
            return this.ch;
        }
        set emptyFilteredGridMessage(a) {
            this.ch = a;
            this.g("EmptyFilteredGridMessage");
        }
        get pinning() {
            return this.aj;
        }
        set pinning(a) {
            this.aj = a;
            this.g("Pinning");
        }
        get allowFiltering() {
            return this.ao;
        }
        set allowFiltering(a) {
            this.ao = a;
            this.g("AllowFiltering");
        }
        get allowAdvancedFiltering() {
            return this.an;
        }
        set allowAdvancedFiltering(a) {
            this.an = a;
            this.g("AllowAdvancedFiltering");
        }
        get filterMode() {
            return this.cp;
        }
        set filterMode(a) {
            this.cp = a;
            this.g("FilterMode");
        }
        get summaryPosition() {
            return this.d6;
        }
        set summaryPosition(a) {
            this.d6 = a;
            this.g("SummaryPosition");
        }
        get summaryCalculationMode() {
            return this.d5;
        }
        set summaryCalculationMode(a) {
            this.d5 = a;
            this.g("SummaryCalculationMode");
        }
        get showSummaryOnCollapse() {
            return this.a0;
        }
        set showSummaryOnCollapse(a) {
            this.a0 = a;
            this.g("ShowSummaryOnCollapse");
        }
        get filterStrategy() {
            return this.ad;
        }
        set filterStrategy(a) {
            this.ad = a;
            this.g("FilterStrategy");
        }
        get sortStrategy() {
            return this.ah;
        }
        set sortStrategy(a) {
            this.ah = a;
            this.g("SortStrategy");
        }
        get sortingOptions() {
            return this.al;
        }
        set sortingOptions(a) {
            this.al = a;
            this.g("SortingOptions");
        }
        get selectedRowsRef() {
            return this.dy;
        }
        set selectedRowsRef(a) {
            this.dy = a;
            this.g("SelectedRowsRef");
        }
        get headSelectorTemplateRef() {
            return this.cz;
        }
        set headSelectorTemplateRef(a) {
            this.cz = a;
            this.g("HeadSelectorTemplateRef");
        }
        get rowSelectorTemplateRef() {
            return this.du;
        }
        set rowSelectorTemplateRef(a) {
            this.du = a;
            this.g("RowSelectorTemplateRef");
        }
        get dragIndicatorIconTemplateRef() {
            return this.cg;
        }
        set dragIndicatorIconTemplateRef(a) {
            this.cg = a;
            this.g("DragIndicatorIconTemplateRef");
        }
        get sortingExpressions() {
            return this.m;
        }
        set sortingExpressions(a) {
            this.m = a;
            this.g("SortingExpressions");
        }
        get cellSelection() {
            return this.bv;
        }
        set cellSelection(a) {
            this.bv = a;
            this.g("CellSelection");
        }
        get rowSelection() {
            return this.ds;
        }
        set rowSelection(a) {
            this.ds = a;
            this.g("RowSelection");
        }
        get columnSelection() {
            return this.b5;
        }
        set columnSelection(a) {
            this.b5 = a;
            this.g("ColumnSelection");
        }
        get totalRecords() {
            return this.bi;
        }
        set totalRecords(a) {
            this.bi = a;
            this.g("TotalRecords");
        }
        get selectRowOnClick() {
            return this.ax;
        }
        set selectRowOnClick(a) {
            this.ax = a;
            this.g("SelectRowOnClick");
        }
        get name() {
            return this.c3;
        }
        set name(a) {
            this.c3 = a;
            this.g("Name");
        }
        get gridCreatedRef() {
            return this.cs;
        }
        set gridCreatedRef(a) {
            this.cs = a;
            this.g("GridCreatedRef");
        }
        get gridInitializedRef() {
            return this.ct;
        }
        set gridInitializedRef(a) {
            this.ct = a;
            this.g("GridInitializedRef");
        }
        get dataPreLoadRef() {
            return this.cd;
        }
        set dataPreLoadRef(a) {
            this.cd = a;
            this.g("DataPreLoadRef");
        }
        get filteringExpressionsTreeChangeRef() {
            return this.cn;
        }
        set filteringExpressionsTreeChangeRef(a) {
            this.cn = a;
            this.g("FilteringExpressionsTreeChangeRef");
        }
        get advancedFilteringExpressionsTreeChangeRef() {
            return this.bp;
        }
        set advancedFilteringExpressionsTreeChangeRef(a) {
            this.bp = a;
            this.g("AdvancedFilteringExpressionsTreeChangeRef");
        }
        get gridScrollRef() {
            return this.cv;
        }
        set gridScrollRef(a) {
            this.cv = a;
            this.g("GridScrollRef");
        }
        get cellClickRef() {
            return this.bq;
        }
        set cellClickRef(a) {
            this.bq = a;
            this.g("CellClickRef");
        }
        get rowClickRef() {
            return this.dc;
        }
        set rowClickRef(a) {
            this.dc = a;
            this.g("RowClickRef");
        }
        get formGroupCreatedRef() {
            return this.cq;
        }
        set formGroupCreatedRef(a) {
            this.cq = a;
            this.g("FormGroupCreatedRef");
        }
        get validationStatusChangeRef() {
            return this.d8;
        }
        set validationStatusChangeRef(a) {
            this.d8 = a;
            this.g("ValidationStatusChangeRef");
        }
        get selectedRef() {
            return this.dx;
        }
        set selectedRef(a) {
            this.dx = a;
            this.g("SelectedRef");
        }
        get rowSelectionChangingRef() {
            return this.dt;
        }
        set rowSelectionChangingRef(a) {
            this.dt = a;
            this.g("RowSelectionChangingRef");
        }
        get columnSelectionChangingRef() {
            return this.b6;
        }
        set columnSelectionChangingRef(a) {
            this.b6 = a;
            this.g("ColumnSelectionChangingRef");
        }
        get columnPinRef() {
            return this.b1;
        }
        set columnPinRef(a) {
            this.b1 = a;
            this.g("ColumnPinRef");
        }
        get columnPinnedRef() {
            return this.b2;
        }
        set columnPinnedRef(a) {
            this.b2 = a;
            this.g("ColumnPinnedRef");
        }
        get cellEditEnterRef() {
            return this.bt;
        }
        set cellEditEnterRef(a) {
            this.bt = a;
            this.g("CellEditEnterRef");
        }
        get cellEditExitRef() {
            return this.bu;
        }
        set cellEditExitRef(a) {
            this.bu = a;
            this.g("CellEditExitRef");
        }
        get cellEditRef() {
            return this.br;
        }
        set cellEditRef(a) {
            this.br = a;
            this.g("CellEditRef");
        }
        get cellEditDoneRef() {
            return this.bs;
        }
        set cellEditDoneRef(a) {
            this.bs = a;
            this.g("CellEditDoneRef");
        }
        get rowEditEnterRef() {
            return this.dl;
        }
        set rowEditEnterRef(a) {
            this.dl = a;
            this.g("RowEditEnterRef");
        }
        get rowEditRef() {
            return this.di;
        }
        set rowEditRef(a) {
            this.di = a;
            this.g("RowEditRef");
        }
        get rowEditDoneRef() {
            return this.dk;
        }
        set rowEditDoneRef(a) {
            this.dk = a;
            this.g("RowEditDoneRef");
        }
        get rowEditExitRef() {
            return this.dm;
        }
        set rowEditExitRef(a) {
            this.dm = a;
            this.g("RowEditExitRef");
        }
        get columnInitRef() {
            return this.bx;
        }
        set columnInitRef(a) {
            this.bx = a;
            this.g("ColumnInitRef");
        }
        get columnsAutogeneratedRef() {
            return this.b4;
        }
        set columnsAutogeneratedRef(a) {
            this.b4 = a;
            this.g("ColumnsAutogeneratedRef");
        }
        get sortingRef() {
            return this.d2;
        }
        set sortingRef(a) {
            this.d2 = a;
            this.g("SortingRef");
        }
        get sortingDoneRef() {
            return this.d3;
        }
        set sortingDoneRef(a) {
            this.d3 = a;
            this.g("SortingDoneRef");
        }
        get filteringRef() {
            return this.cl;
        }
        set filteringRef(a) {
            this.cl = a;
            this.g("FilteringRef");
        }
        get filteringDoneRef() {
            return this.cm;
        }
        set filteringDoneRef(a) {
            this.cm = a;
            this.g("FilteringDoneRef");
        }
        get rowAddedRef() {
            return this.c9;
        }
        set rowAddedRef(a) {
            this.c9 = a;
            this.g("RowAddedRef");
        }
        get rowDeletedRef() {
            return this.df;
        }
        set rowDeletedRef(a) {
            this.df = a;
            this.g("RowDeletedRef");
        }
        get rowDeleteRef() {
            return this.de;
        }
        set rowDeleteRef(a) {
            this.de = a;
            this.g("RowDeleteRef");
        }
        get rowAddRef() {
            return this.c8;
        }
        set rowAddRef(a) {
            this.c8 = a;
            this.g("RowAddRef");
        }
        get columnResizedRef() {
            return this.b3;
        }
        set columnResizedRef(a) {
            this.b3 = a;
            this.g("ColumnResizedRef");
        }
        get contextMenuRef() {
            return this.ca;
        }
        set contextMenuRef(a) {
            this.ca = a;
            this.g("ContextMenuRef");
        }
        get doubleClickRef() {
            return this.ce;
        }
        set doubleClickRef(a) {
            this.ce = a;
            this.g("DoubleClickRef");
        }
        get columnVisibilityChangingRef() {
            return this.b8;
        }
        set columnVisibilityChangingRef(a) {
            this.b8 = a;
            this.g("ColumnVisibilityChangingRef");
        }
        get columnVisibilityChangedRef() {
            return this.b7;
        }
        set columnVisibilityChangedRef(a) {
            this.b7 = a;
            this.g("ColumnVisibilityChangedRef");
        }
        get columnMovingStartRef() {
            return this.b0;
        }
        set columnMovingStartRef(a) {
            this.b0 = a;
            this.g("ColumnMovingStartRef");
        }
        get columnMovingRef() {
            return this.by;
        }
        set columnMovingRef(a) {
            this.by = a;
            this.g("ColumnMovingRef");
        }
        get columnMovingEndRef() {
            return this.bz;
        }
        set columnMovingEndRef(a) {
            this.bz = a;
            this.g("ColumnMovingEndRef");
        }
        get gridKeydownRef() {
            return this.cu;
        }
        set gridKeydownRef(a) {
            this.cu = a;
            this.g("GridKeydownRef");
        }
        get rowDragStartRef() {
            return this.dh;
        }
        set rowDragStartRef(a) {
            this.dh = a;
            this.g("RowDragStartRef");
        }
        get rowDragEndRef() {
            return this.dg;
        }
        set rowDragEndRef(a) {
            this.dg = a;
            this.g("RowDragEndRef");
        }
        get gridCopyRef() {
            return this.cr;
        }
        set gridCopyRef(a) {
            this.cr = a;
            this.g("GridCopyRef");
        }
        get rowToggleRef() {
            return this.dw;
        }
        set rowToggleRef(a) {
            this.dw = a;
            this.g("RowToggleRef");
        }
        get rowPinningRef() {
            return this.dr;
        }
        set rowPinningRef(a) {
            this.dr = a;
            this.g("RowPinningRef");
        }
        get rowPinnedRef() {
            return this.dq;
        }
        set rowPinnedRef(a) {
            this.dq = a;
            this.g("RowPinnedRef");
        }
        get activeNodeChangeRef() {
            return this.bn;
        }
        set activeNodeChangeRef(a) {
            this.bn = a;
            this.g("ActiveNodeChangeRef");
        }
        get sortingExpressionsChangeRef() {
            return this.d4;
        }
        set sortingExpressionsChangeRef(a) {
            this.d4 = a;
            this.g("SortingExpressionsChangeRef");
        }
        get toolbarExportingRef() {
            return this.d7;
        }
        set toolbarExportingRef(a) {
            this.d7 = a;
            this.g("ToolbarExportingRef");
        }
        get rangeSelectedRef() {
            return this.c6;
        }
        set rangeSelectedRef(a) {
            this.c6 = a;
            this.g("RangeSelectedRef");
        }
        get renderedRef() {
            return this.c7;
        }
        set renderedRef(a) {
            this.c7 = a;
            this.g("RenderedRef");
        }
        get dataChangingRef() {
            return this.cc;
        }
        set dataChangingRef(a) {
            this.cc = a;
            this.g("DataChangingRef");
        }
        get dataChangedRef() {
            return this.cb;
        }
        set dataChangedRef(a) {
            this.cb = a;
            this.g("DataChangedRef");
        }
    }
    WebRowIslandDescription.$t = markType(WebRowIslandDescription, 'WebRowIslandDescription', Description.$);
    return WebRowIslandDescription;
})();
