// igniteui-angular-core/esm2015/lib/WebSelectDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { WebBaseComboBoxLikeDescription } from "./WebBaseComboBoxLikeDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let WebSelectDescription = /*@__PURE__*/ (() => {
    class WebSelectDescription extends WebBaseComboBoxLikeDescription {
        constructor() {
            super();
            this.ao = null;
            this.t = false;
            this.q = false;
            this.aa = 0;
            this.ah = null;
            this.al = null;
            this.am = null;
            this.an = null;
            this.r = false;
            this.u = false;
            this.ai = null;
            this.s = false;
            this.ad = null;
            this.ag = null;
            this.ac = null;
            this.ak = null;
            this.aj = null;
            this.af = null;
            this.ae = null;
        }
        get_type() {
            return "WebSelect";
        }
        get value() {
            return this.ao;
        }
        set value(a) {
            this.ao = a;
            this.g("Value");
        }
        get outlined() {
            return this.t;
        }
        set outlined(a) {
            this.t = a;
            this.g("Outlined");
        }
        get autofocus() {
            return this.q;
        }
        set autofocus(a) {
            this.q = a;
            this.g("Autofocus");
        }
        get distance() {
            return this.aa;
        }
        set distance(a) {
            this.aa = a;
            this.g("Distance");
        }
        get label() {
            return this.ah;
        }
        set label(a) {
            this.ah = a;
            this.g("Label");
        }
        get placeholder() {
            return this.al;
        }
        set placeholder(a) {
            this.al = a;
            this.g("Placeholder");
        }
        get placement() {
            return this.am;
        }
        set placement(a) {
            this.am = a;
            this.g("Placement");
        }
        get scrollStrategy() {
            return this.an;
        }
        set scrollStrategy(a) {
            this.an = a;
            this.g("ScrollStrategy");
        }
        get disabled() {
            return this.r;
        }
        set disabled(a) {
            this.r = a;
            this.g("Disabled");
        }
        get required() {
            return this.u;
        }
        set required(a) {
            this.u = a;
            this.g("Required");
        }
        get name() {
            return this.ai;
        }
        set name(a) {
            this.ai = a;
            this.g("Name");
        }
        get invalid() {
            return this.s;
        }
        set invalid(a) {
            this.s = a;
            this.g("Invalid");
        }
        get changeRef() {
            return this.ad;
        }
        set changeRef(a) {
            this.ad = a;
            this.g("ChangeRef");
        }
        get focusRef() {
            return this.ag;
        }
        set focusRef(a) {
            this.ag = a;
            this.g("FocusRef");
        }
        get blurRef() {
            return this.ac;
        }
        set blurRef(a) {
            this.ac = a;
            this.g("BlurRef");
        }
        get openingRef() {
            return this.ak;
        }
        set openingRef(a) {
            this.ak = a;
            this.g("OpeningRef");
        }
        get openedRef() {
            return this.aj;
        }
        set openedRef(a) {
            this.aj = a;
            this.g("OpenedRef");
        }
        get closingRef() {
            return this.af;
        }
        set closingRef(a) {
            this.af = a;
            this.g("ClosingRef");
        }
        get closedRef() {
            return this.ae;
        }
        set closedRef(a) {
            this.ae = a;
            this.g("ClosedRef");
        }
    }
    WebSelectDescription.$t = markType(WebSelectDescription, 'WebSelectDescription', WebBaseComboBoxLikeDescription.$);
    return WebSelectDescription;
})();
