// igniteui-angular-core/esm2015/lib/XYChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { DomainChartDescription } from "./DomainChartDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let XYChartDescription = /*@__PURE__*/ (() => {
    class XYChartDescription extends DomainChartDescription {
        constructor() {
            super();
            this.t1 = null;
            this.uj = null;
            this.r6 = 0;
            this.r8 = 0;
            this.r7 = 0;
            this.r5 = 0;
            this.sn = 0;
            this.sp = 0;
            this.so = 0;
            this.sm = 0;
            this.t6 = null;
            this.uo = null;
            this.tz = null;
            this.t0 = null;
            this.sh = 0;
            this.sy = 0;
            this.sg = 0;
            this.sx = 0;
            this.sj = 0;
            this.s0 = 0;
            this.si = 0;
            this.sz = 0;
            this.sf = 0;
            this.sw = 0;
            this.uh = null;
            this.uz = null;
            this.t7 = null;
            this.up = null;
            this.ui = null;
            this.u0 = null;
            this.t5 = null;
            this.un = null;
            this.ua = null;
            this.us = null;
            this.r9 = 0;
            this.sq = 0;
            this.sa = 0;
            this.sr = 0;
            this.uc = null;
            this.uu = null;
            this.ud = null;
            this.uv = null;
            this.sb = 0;
            this.ss = 0;
            this.sc = 0;
            this.st = 0;
            this.ue = null;
            this.uw = null;
            this.sd = 0;
            this.su = 0;
            this.uf = null;
            this.ux = null;
            this.ub = null;
            this.ut = null;
            this.r4 = 0;
            this.sl = 0;
            this.r3 = 0;
            this.sk = 0;
            this.se = 0;
            this.sv = 0;
            this.rz = false;
            this.r0 = false;
            this.ug = null;
            this.uy = null;
            this.t3 = null;
            this.ul = null;
            this.t8 = null;
            this.uq = null;
            this.t9 = null;
            this.ur = null;
            this.um = null;
            this.t4 = null;
            this.t2 = null;
            this.rv = null;
            this.uk = null;
            this.rw = null;
        }
        get_type() {
            return "XYChart";
        }
        get xAxisFormatLabelRef() {
            return this.t1;
        }
        set xAxisFormatLabelRef(a) {
            this.t1 = a;
            this.g("XAxisFormatLabelRef");
        }
        get yAxisFormatLabelRef() {
            return this.uj;
        }
        set yAxisFormatLabelRef(a) {
            this.uj = a;
            this.g("YAxisFormatLabelRef");
        }
        get xAxisLabelLeftMargin() {
            return this.r6;
        }
        set xAxisLabelLeftMargin(a) {
            this.r6 = a;
            this.g("XAxisLabelLeftMargin");
        }
        get xAxisLabelTopMargin() {
            return this.r8;
        }
        set xAxisLabelTopMargin(a) {
            this.r8 = a;
            this.g("XAxisLabelTopMargin");
        }
        get xAxisLabelRightMargin() {
            return this.r7;
        }
        set xAxisLabelRightMargin(a) {
            this.r7 = a;
            this.g("XAxisLabelRightMargin");
        }
        get xAxisLabelBottomMargin() {
            return this.r5;
        }
        set xAxisLabelBottomMargin(a) {
            this.r5 = a;
            this.g("XAxisLabelBottomMargin");
        }
        get yAxisLabelLeftMargin() {
            return this.sn;
        }
        set yAxisLabelLeftMargin(a) {
            this.sn = a;
            this.g("YAxisLabelLeftMargin");
        }
        get yAxisLabelTopMargin() {
            return this.sp;
        }
        set yAxisLabelTopMargin(a) {
            this.sp = a;
            this.g("YAxisLabelTopMargin");
        }
        get yAxisLabelRightMargin() {
            return this.so;
        }
        set yAxisLabelRightMargin(a) {
            this.so = a;
            this.g("YAxisLabelRightMargin");
        }
        get yAxisLabelBottomMargin() {
            return this.sm;
        }
        set yAxisLabelBottomMargin(a) {
            this.sm = a;
            this.g("YAxisLabelBottomMargin");
        }
        get xAxisLabelTextColor() {
            return this.t6;
        }
        set xAxisLabelTextColor(a) {
            this.t6 = a;
            this.g("XAxisLabelTextColor");
        }
        get yAxisLabelTextColor() {
            return this.uo;
        }
        set yAxisLabelTextColor(a) {
            this.uo = a;
            this.g("YAxisLabelTextColor");
        }
        get actualXAxisLabelTextColor() {
            return this.tz;
        }
        set actualXAxisLabelTextColor(a) {
            this.tz = a;
            this.g("ActualXAxisLabelTextColor");
        }
        get actualYAxisLabelTextColor() {
            return this.t0;
        }
        set actualYAxisLabelTextColor(a) {
            this.t0 = a;
            this.g("ActualYAxisLabelTextColor");
        }
        get xAxisTitleMargin() {
            return this.sh;
        }
        set xAxisTitleMargin(a) {
            this.sh = a;
            this.g("XAxisTitleMargin");
        }
        get yAxisTitleMargin() {
            return this.sy;
        }
        set yAxisTitleMargin(a) {
            this.sy = a;
            this.g("YAxisTitleMargin");
        }
        get xAxisTitleLeftMargin() {
            return this.sg;
        }
        set xAxisTitleLeftMargin(a) {
            this.sg = a;
            this.g("XAxisTitleLeftMargin");
        }
        get yAxisTitleLeftMargin() {
            return this.sx;
        }
        set yAxisTitleLeftMargin(a) {
            this.sx = a;
            this.g("YAxisTitleLeftMargin");
        }
        get xAxisTitleTopMargin() {
            return this.sj;
        }
        set xAxisTitleTopMargin(a) {
            this.sj = a;
            this.g("XAxisTitleTopMargin");
        }
        get yAxisTitleTopMargin() {
            return this.s0;
        }
        set yAxisTitleTopMargin(a) {
            this.s0 = a;
            this.g("YAxisTitleTopMargin");
        }
        get xAxisTitleRightMargin() {
            return this.si;
        }
        set xAxisTitleRightMargin(a) {
            this.si = a;
            this.g("XAxisTitleRightMargin");
        }
        get yAxisTitleRightMargin() {
            return this.sz;
        }
        set yAxisTitleRightMargin(a) {
            this.sz = a;
            this.g("YAxisTitleRightMargin");
        }
        get xAxisTitleBottomMargin() {
            return this.sf;
        }
        set xAxisTitleBottomMargin(a) {
            this.sf = a;
            this.g("XAxisTitleBottomMargin");
        }
        get yAxisTitleBottomMargin() {
            return this.sw;
        }
        set yAxisTitleBottomMargin(a) {
            this.sw = a;
            this.g("YAxisTitleBottomMargin");
        }
        get xAxisTitleTextColor() {
            return this.uh;
        }
        set xAxisTitleTextColor(a) {
            this.uh = a;
            this.g("XAxisTitleTextColor");
        }
        get yAxisTitleTextColor() {
            return this.uz;
        }
        set yAxisTitleTextColor(a) {
            this.uz = a;
            this.g("YAxisTitleTextColor");
        }
        get xAxisLabelTextStyle() {
            return this.t7;
        }
        set xAxisLabelTextStyle(a) {
            this.t7 = a;
            this.g("XAxisLabelTextStyle");
        }
        get yAxisLabelTextStyle() {
            return this.up;
        }
        set yAxisLabelTextStyle(a) {
            this.up = a;
            this.g("YAxisLabelTextStyle");
        }
        get xAxisTitleTextStyle() {
            return this.ui;
        }
        set xAxisTitleTextStyle(a) {
            this.ui = a;
            this.g("XAxisTitleTextStyle");
        }
        get yAxisTitleTextStyle() {
            return this.u0;
        }
        set yAxisTitleTextStyle(a) {
            this.u0 = a;
            this.g("YAxisTitleTextStyle");
        }
        get xAxisLabelRef() {
            return this.t5;
        }
        set xAxisLabelRef(a) {
            this.t5 = a;
            this.g("XAxisLabelRef");
        }
        get yAxisLabelRef() {
            return this.un;
        }
        set yAxisLabelRef(a) {
            this.un = a;
            this.g("YAxisLabelRef");
        }
        get xAxisMajorStroke() {
            return this.ua;
        }
        set xAxisMajorStroke(a) {
            this.ua = a;
            this.g("XAxisMajorStroke");
        }
        get yAxisMajorStroke() {
            return this.us;
        }
        set yAxisMajorStroke(a) {
            this.us = a;
            this.g("YAxisMajorStroke");
        }
        get xAxisMajorStrokeThickness() {
            return this.r9;
        }
        set xAxisMajorStrokeThickness(a) {
            this.r9 = a;
            this.g("XAxisMajorStrokeThickness");
        }
        get yAxisMajorStrokeThickness() {
            return this.sq;
        }
        set yAxisMajorStrokeThickness(a) {
            this.sq = a;
            this.g("YAxisMajorStrokeThickness");
        }
        get xAxisMinorStrokeThickness() {
            return this.sa;
        }
        set xAxisMinorStrokeThickness(a) {
            this.sa = a;
            this.g("XAxisMinorStrokeThickness");
        }
        get yAxisMinorStrokeThickness() {
            return this.sr;
        }
        set yAxisMinorStrokeThickness(a) {
            this.sr = a;
            this.g("YAxisMinorStrokeThickness");
        }
        get xAxisStrip() {
            return this.uc;
        }
        set xAxisStrip(a) {
            this.uc = a;
            this.g("XAxisStrip");
        }
        get yAxisStrip() {
            return this.uu;
        }
        set yAxisStrip(a) {
            this.uu = a;
            this.g("YAxisStrip");
        }
        get xAxisStroke() {
            return this.ud;
        }
        set xAxisStroke(a) {
            this.ud = a;
            this.g("XAxisStroke");
        }
        get yAxisStroke() {
            return this.uv;
        }
        set yAxisStroke(a) {
            this.uv = a;
            this.g("YAxisStroke");
        }
        get xAxisStrokeThickness() {
            return this.sb;
        }
        set xAxisStrokeThickness(a) {
            this.sb = a;
            this.g("XAxisStrokeThickness");
        }
        get yAxisStrokeThickness() {
            return this.ss;
        }
        set yAxisStrokeThickness(a) {
            this.ss = a;
            this.g("YAxisStrokeThickness");
        }
        get xAxisTickLength() {
            return this.sc;
        }
        set xAxisTickLength(a) {
            this.sc = a;
            this.g("XAxisTickLength");
        }
        get yAxisTickLength() {
            return this.st;
        }
        set yAxisTickLength(a) {
            this.st = a;
            this.g("YAxisTickLength");
        }
        get xAxisTickStroke() {
            return this.ue;
        }
        set xAxisTickStroke(a) {
            this.ue = a;
            this.g("XAxisTickStroke");
        }
        get yAxisTickStroke() {
            return this.uw;
        }
        set yAxisTickStroke(a) {
            this.uw = a;
            this.g("YAxisTickStroke");
        }
        get xAxisTickStrokeThickness() {
            return this.sd;
        }
        set xAxisTickStrokeThickness(a) {
            this.sd = a;
            this.g("XAxisTickStrokeThickness");
        }
        get yAxisTickStrokeThickness() {
            return this.su;
        }
        set yAxisTickStrokeThickness(a) {
            this.su = a;
            this.g("YAxisTickStrokeThickness");
        }
        get xAxisTitle() {
            return this.uf;
        }
        set xAxisTitle(a) {
            this.uf = a;
            this.g("XAxisTitle");
        }
        get yAxisTitle() {
            return this.ux;
        }
        set yAxisTitle(a) {
            this.ux = a;
            this.g("YAxisTitle");
        }
        get xAxisMinorStroke() {
            return this.ub;
        }
        set xAxisMinorStroke(a) {
            this.ub = a;
            this.g("XAxisMinorStroke");
        }
        get yAxisMinorStroke() {
            return this.ut;
        }
        set yAxisMinorStroke(a) {
            this.ut = a;
            this.g("YAxisMinorStroke");
        }
        get xAxisLabelAngle() {
            return this.r4;
        }
        set xAxisLabelAngle(a) {
            this.r4 = a;
            this.g("XAxisLabelAngle");
        }
        get yAxisLabelAngle() {
            return this.sl;
        }
        set yAxisLabelAngle(a) {
            this.sl = a;
            this.g("YAxisLabelAngle");
        }
        get xAxisExtent() {
            return this.r3;
        }
        set xAxisExtent(a) {
            this.r3 = a;
            this.g("XAxisExtent");
        }
        get yAxisExtent() {
            return this.sk;
        }
        set yAxisExtent(a) {
            this.sk = a;
            this.g("YAxisExtent");
        }
        get xAxisTitleAngle() {
            return this.se;
        }
        set xAxisTitleAngle(a) {
            this.se = a;
            this.g("XAxisTitleAngle");
        }
        get yAxisTitleAngle() {
            return this.sv;
        }
        set yAxisTitleAngle(a) {
            this.sv = a;
            this.g("YAxisTitleAngle");
        }
        get xAxisInverted() {
            return this.rz;
        }
        set xAxisInverted(a) {
            this.rz = a;
            this.g("XAxisInverted");
        }
        get yAxisInverted() {
            return this.r0;
        }
        set yAxisInverted(a) {
            this.r0 = a;
            this.g("YAxisInverted");
        }
        get xAxisTitleAlignment() {
            return this.ug;
        }
        set xAxisTitleAlignment(a) {
            this.ug = a;
            this.g("XAxisTitleAlignment");
        }
        get yAxisTitleAlignment() {
            return this.uy;
        }
        set yAxisTitleAlignment(a) {
            this.uy = a;
            this.g("YAxisTitleAlignment");
        }
        get xAxisLabelHorizontalAlignment() {
            return this.t3;
        }
        set xAxisLabelHorizontalAlignment(a) {
            this.t3 = a;
            this.g("XAxisLabelHorizontalAlignment");
        }
        get yAxisLabelHorizontalAlignment() {
            return this.ul;
        }
        set yAxisLabelHorizontalAlignment(a) {
            this.ul = a;
            this.g("YAxisLabelHorizontalAlignment");
        }
        get xAxisLabelVerticalAlignment() {
            return this.t8;
        }
        set xAxisLabelVerticalAlignment(a) {
            this.t8 = a;
            this.g("XAxisLabelVerticalAlignment");
        }
        get yAxisLabelVerticalAlignment() {
            return this.uq;
        }
        set yAxisLabelVerticalAlignment(a) {
            this.uq = a;
            this.g("YAxisLabelVerticalAlignment");
        }
        get xAxisLabelVisibility() {
            return this.t9;
        }
        set xAxisLabelVisibility(a) {
            this.t9 = a;
            this.g("XAxisLabelVisibility");
        }
        get yAxisLabelVisibility() {
            return this.ur;
        }
        set yAxisLabelVisibility(a) {
            this.ur = a;
            this.g("YAxisLabelVisibility");
        }
        get yAxisLabelLocation() {
            return this.um;
        }
        set yAxisLabelLocation(a) {
            this.um = a;
            this.g("YAxisLabelLocation");
        }
        get xAxisLabelLocation() {
            return this.t4;
        }
        set xAxisLabelLocation(a) {
            this.t4 = a;
            this.g("XAxisLabelLocation");
        }
        get xAxisLabelFormat() {
            return this.t2;
        }
        set xAxisLabelFormat(a) {
            this.t2 = a;
            this.g("XAxisLabelFormat");
        }
        get xAxisLabelFormatSpecifiers() {
            return this.rv;
        }
        set xAxisLabelFormatSpecifiers(a) {
            this.rv = a;
            this.g("XAxisLabelFormatSpecifiers");
        }
        get yAxisLabelFormat() {
            return this.uk;
        }
        set yAxisLabelFormat(a) {
            this.uk = a;
            this.g("YAxisLabelFormat");
        }
        get yAxisLabelFormatSpecifiers() {
            return this.rw;
        }
        set yAxisLabelFormatSpecifiers(a) {
            this.rw = a;
            this.g("YAxisLabelFormatSpecifiers");
        }
    }
    XYChartDescription.$t = markType(XYChartDescription, 'XYChartDescription', DomainChartDescription.$);
    return XYChartDescription;
})();
