// igniteui-angular-core/esm5/lib/AnchoredRadialSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { RadialBaseDescription } from "./RadialBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var AnchoredRadialSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(AnchoredRadialSeriesDescription, _super);
    function AnchoredRadialSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.g1 = false;
        _this.hf = null;
        _this.ha = null;
        _this.hd = null;
        _this.he = null;
        _this.hc = null;
        _this.hb = null;
        _this.g9 = null;
        _this.g3 = 0;
        _this.gz = null;
        _this.g5 = 0;
        _this.g6 = 0;
        return _this;
    }
    AnchoredRadialSeriesDescription.prototype.get_type = function () {
        return "AnchoredRadialSeries";
    };
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "useCategoryNormalizedValues", {
        get: function () {
            return this.g1;
        },
        set: function (a) {
            this.g1 = a;
            this.g("UseCategoryNormalizedValues");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.hf;
        },
        set: function (a) {
            this.hf = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "highlightedValueMemberPath", {
        get: function () {
            return this.ha;
        },
        set: function (a) {
            this.ha = a;
            this.g("HighlightedValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "valueMemberAsLegendLabel", {
        get: function () {
            return this.hd;
        },
        set: function (a) {
            this.hd = a;
            this.g("ValueMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "valueMemberAsLegendUnit", {
        get: function () {
            return this.he;
        },
        set: function (a) {
            this.he = a;
            this.g("ValueMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineType", {
        get: function () {
            return this.hc;
        },
        set: function (a) {
            this.hc = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.hb;
        },
        set: function (a) {
            this.hb = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.g9;
        },
        set: function (a) {
            this.g9 = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.g3;
        },
        set: function (a) {
            this.g3 = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.gz;
        },
        set: function (a) {
            this.gz = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.g5;
        },
        set: function (a) {
            this.g5 = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(AnchoredRadialSeriesDescription.prototype, "trendLineZIndex", {
        get: function () {
            return this.g6;
        },
        set: function (a) {
            this.g6 = a;
            this.g("TrendLineZIndex");
        },
        enumerable: false,
        configurable: true
    });
    AnchoredRadialSeriesDescription.$t = markType(AnchoredRadialSeriesDescription, 'AnchoredRadialSeriesDescription', RadialBaseDescription.$);
    return AnchoredRadialSeriesDescription;
}(RadialBaseDescription));
export { AnchoredRadialSeriesDescription };
