// igniteui-angular-core/esm5/lib/BubbleSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ScatterBaseDescription } from "./ScatterBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var BubbleSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BubbleSeriesDescription, _super);
    function BubbleSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.hn = null;
        _this.g5 = null;
        _this.hk = null;
        _this.hj = null;
        _this.g3 = null;
        _this.g7 = false;
        _this.g8 = false;
        _this.he = 0;
        _this.hd = 0;
        _this.g9 = false;
        _this.hl = null;
        _this.hh = null;
        _this.hm = null;
        _this.hi = null;
        return _this;
    }
    BubbleSeriesDescription.prototype.get_type = function () {
        return "BubbleSeries";
    };
    Object.defineProperty(BubbleSeriesDescription.prototype, "radiusMemberPath", {
        get: function () {
            return this.hn;
        },
        set: function (a) {
            this.hn = a;
            this.g("RadiusMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "radiusScale", {
        get: function () {
            return this.g5;
        },
        set: function (a) {
            this.g5 = a;
            this.g("RadiusScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "labelMemberPath", {
        get: function () {
            return this.hk;
        },
        set: function (a) {
            this.hk = a;
            this.g("LabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "fillMemberPath", {
        get: function () {
            return this.hj;
        },
        set: function (a) {
            this.hj = a;
            this.g("FillMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "fillScale", {
        get: function () {
            return this.g3;
        },
        set: function (a) {
            this.g3 = a;
            this.g("FillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "fillScaleUseGlobalValues", {
        get: function () {
            return this.g7;
        },
        set: function (a) {
            this.g7 = a;
            this.g("FillScaleUseGlobalValues");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "markerOutlineUsesFillScale", {
        get: function () {
            return this.g8;
        },
        set: function (a) {
            this.g8 = a;
            this.g("MarkerOutlineUsesFillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "markerOutlineBrightness", {
        get: function () {
            return this.he;
        },
        set: function (a) {
            this.he = a;
            this.g("MarkerOutlineBrightness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "markerBrushBrightness", {
        get: function () {
            return this.hd;
        },
        set: function (a) {
            this.hd = a;
            this.g("MarkerBrushBrightness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "radiusScaleUseGlobalValues", {
        get: function () {
            return this.g9;
        },
        set: function (a) {
            this.g9 = a;
            this.g("RadiusScaleUseGlobalValues");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "radiusMemberAsLegendLabel", {
        get: function () {
            return this.hl;
        },
        set: function (a) {
            this.hl = a;
            this.g("RadiusMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "fillMemberAsLegendLabel", {
        get: function () {
            return this.hh;
        },
        set: function (a) {
            this.hh = a;
            this.g("FillMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "radiusMemberAsLegendUnit", {
        get: function () {
            return this.hm;
        },
        set: function (a) {
            this.hm = a;
            this.g("RadiusMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BubbleSeriesDescription.prototype, "fillMemberAsLegendUnit", {
        get: function () {
            return this.hi;
        },
        set: function (a) {
            this.hi = a;
            this.g("FillMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    BubbleSeriesDescription.$t = markType(BubbleSeriesDescription, 'BubbleSeriesDescription', ScatterBaseDescription.$);
    return BubbleSeriesDescription;
}(ScatterBaseDescription));
export { BubbleSeriesDescription };
