// igniteui-angular-core/esm5/lib/CalloutLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var CalloutLayerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CalloutLayerDescription, _super);
    function CalloutLayerDescription() {
        var _this = _super.call(this) || this;
        _this.hv = null;
        _this.hz = null;
        _this.h0 = null;
        _this.ht = null;
        _this.gc = false;
        _this.hu = null;
        _this.f8 = null;
        _this.hr = null;
        _this.hw = null;
        _this.hx = null;
        _this.hq = null;
        _this.f7 = null;
        _this.gb = false;
        _this.gg = false;
        _this.gj = false;
        _this.gf = false;
        _this.gh = false;
        _this.gi = false;
        _this.gk = false;
        _this.hs = null;
        _this.hd = null;
        _this.gv = 0;
        _this.ge = false;
        _this.gd = false;
        _this.g9 = 0;
        _this.hp = null;
        _this.hk = null;
        _this.hh = null;
        _this.he = null;
        _this.ha = 0;
        _this.hl = null;
        _this.hj = null;
        _this.gx = 0;
        _this.gz = 0;
        _this.gy = 0;
        _this.gw = 0;
        _this.hf = null;
        _this.g0 = 0;
        _this.g1 = 0;
        _this.hy = null;
        _this.ho = null;
        _this.hm = null;
        _this.hg = null;
        _this.hi = null;
        _this.hn = null;
        return _this;
    }
    CalloutLayerDescription.prototype.get_type = function () {
        return "CalloutLayer";
    };
    Object.defineProperty(CalloutLayerDescription.prototype, "labelMemberPath", {
        get: function () {
            return this.hv;
        },
        set: function (a) {
            this.hv = a;
            this.g("LabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "xMemberPath", {
        get: function () {
            return this.hz;
        },
        set: function (a) {
            this.hz = a;
            this.g("XMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "yMemberPath", {
        get: function () {
            return this.h0;
        },
        set: function (a) {
            this.h0 = a;
            this.g("YMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "keyMemberPath", {
        get: function () {
            return this.ht;
        },
        set: function (a) {
            this.ht = a;
            this.g("KeyMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "isCalloutOffsettingEnabled", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("IsCalloutOffsettingEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "labelFormat", {
        get: function () {
            return this.hu;
        },
        set: function (a) {
            this.hu = a;
            this.g("LabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "labelFormatSpecifiers", {
        get: function () {
            return this.f8;
        },
        set: function (a) {
            this.f8 = a;
            this.g("LabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "contentMemberPath", {
        get: function () {
            return this.hr;
        },
        set: function (a) {
            this.hr = a;
            this.g("ContentMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "targetSeriesName", {
        get: function () {
            return this.hw;
        },
        set: function (a) {
            this.hw = a;
            this.g("TargetSeriesName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "targetSeriesRef", {
        get: function () {
            return this.hx;
        },
        set: function (a) {
            this.hx = a;
            this.g("TargetSeriesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "collisionChannel", {
        get: function () {
            return this.hq;
        },
        set: function (a) {
            this.hq = a;
            this.g("CollisionChannel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "allowedPositions", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("AllowedPositions");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "isAutoCalloutBehaviorEnabled", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("IsAutoCalloutBehaviorEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useInterpolatedValueForAutoCalloutLabels", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("UseInterpolatedValueForAutoCalloutLabels");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useSeriesColorForOutline", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("UseSeriesColorForOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useAutoContrastingLabelColors", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("UseAutoContrastingLabelColors");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useItemColorForFill", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("UseItemColorForFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useItemColorForOutline", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("UseItemColorForOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "useValueForAutoCalloutLabels", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("UseValueForAutoCalloutLabels");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "highlightedValueLabelMode", {
        get: function () {
            return this.hs;
        },
        set: function (a) {
            this.hs = a;
            this.g("HighlightedValueLabelMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "autoCalloutVisibilityMode", {
        get: function () {
            return this.hd;
        },
        set: function (a) {
            this.hd = a;
            this.g("AutoCalloutVisibilityMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutCornerRadius", {
        get: function () {
            return this.gv;
        },
        set: function (a) {
            this.gv = a;
            this.g("CalloutCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "isCustomCalloutStyleEnabled", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("IsCustomCalloutStyleEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "isCustomCalloutRenderStyleEnabled", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("IsCustomCalloutRenderStyleEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "autoCalloutLabelPrecision", {
        get: function () {
            return this.g9;
        },
        set: function (a) {
            this.g9 = a;
            this.g("AutoCalloutLabelPrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutTextColor", {
        get: function () {
            return this.hp;
        },
        set: function (a) {
            this.hp = a;
            this.g("CalloutTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutLightTextColor", {
        get: function () {
            return this.hk;
        },
        set: function (a) {
            this.hk = a;
            this.g("CalloutLightTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutDarkTextColor", {
        get: function () {
            return this.hh;
        },
        set: function (a) {
            this.hh = a;
            this.g("CalloutDarkTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutBackground", {
        get: function () {
            return this.he;
        },
        set: function (a) {
            this.he = a;
            this.g("CalloutBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutInterpolatedValuePrecision", {
        get: function () {
            return this.ha;
        },
        set: function (a) {
            this.ha = a;
            this.g("CalloutInterpolatedValuePrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutOutline", {
        get: function () {
            return this.hl;
        },
        set: function (a) {
            this.hl = a;
            this.g("CalloutOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutLeaderBrush", {
        get: function () {
            return this.hj;
        },
        set: function (a) {
            this.hj = a;
            this.g("CalloutLeaderBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutPaddingLeft", {
        get: function () {
            return this.gx;
        },
        set: function (a) {
            this.gx = a;
            this.g("CalloutPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutPaddingTop", {
        get: function () {
            return this.gz;
        },
        set: function (a) {
            this.gz = a;
            this.g("CalloutPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutPaddingRight", {
        get: function () {
            return this.gy;
        },
        set: function (a) {
            this.gy = a;
            this.g("CalloutPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutPaddingBottom", {
        get: function () {
            return this.gw;
        },
        set: function (a) {
            this.gw = a;
            this.g("CalloutPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutCollisionMode", {
        get: function () {
            return this.hf;
        },
        set: function (a) {
            this.hf = a;
            this.g("CalloutCollisionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutPositionPadding", {
        get: function () {
            return this.g0;
        },
        set: function (a) {
            this.g0 = a;
            this.g("CalloutPositionPadding");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutStrokeThickness", {
        get: function () {
            return this.g1;
        },
        set: function (a) {
            this.g1 = a;
            this.g("CalloutStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "textStyle", {
        get: function () {
            return this.hy;
        },
        set: function (a) {
            this.hy = a;
            this.g("TextStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutStyleUpdatingRef", {
        get: function () {
            return this.ho;
        },
        set: function (a) {
            this.ho = a;
            this.g("CalloutStyleUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutRenderStyleUpdatingRef", {
        get: function () {
            return this.hm;
        },
        set: function (a) {
            this.hm = a;
            this.g("CalloutRenderStyleUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutContentUpdatingRef", {
        get: function () {
            return this.hg;
        },
        set: function (a) {
            this.hg = a;
            this.g("CalloutContentUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutLabelUpdatingRef", {
        get: function () {
            return this.hi;
        },
        set: function (a) {
            this.hi = a;
            this.g("CalloutLabelUpdatingRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CalloutLayerDescription.prototype, "calloutSeriesSelectingRef", {
        get: function () {
            return this.hn;
        },
        set: function (a) {
            this.hn = a;
            this.g("CalloutSeriesSelectingRef");
        },
        enumerable: false,
        configurable: true
    });
    CalloutLayerDescription.$t = markType(CalloutLayerDescription, 'CalloutLayerDescription', AnnotationLayerDescription.$);
    return CalloutLayerDescription;
}(AnnotationLayerDescription));
export { CalloutLayerDescription };
