// igniteui-angular-core/esm5/lib/CategoryItemHighlightLayerDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { AnnotationLayerDescription } from "./AnnotationLayerDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var CategoryItemHighlightLayerDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategoryItemHighlightLayerDescription, _super);
    function CategoryItemHighlightLayerDescription() {
        var _this = _super.call(this) || this;
        _this.gj = null;
        _this.gi = null;
        _this.f8 = false;
        _this.gd = null;
        _this.gh = null;
        _this.ge = null;
        _this.gf = null;
        _this.gg = null;
        _this.gb = 0;
        _this.f7 = false;
        return _this;
    }
    CategoryItemHighlightLayerDescription.prototype.get_type = function () {
        return "CategoryItemHighlightLayer";
    };
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "targetSeriesRef", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("TargetSeriesRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "targetSeriesName", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("TargetSeriesName");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "useInterpolation", {
        get: function () {
            return this.f8;
        },
        set: function (a) {
            this.f8 = a;
            this.g("UseInterpolation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "highlightType", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("HighlightType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "markerType", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("MarkerType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "markerBrush", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("MarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "markerOutline", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("MarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "markerTemplateRef", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("MarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "bandHighlightWidth", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("BandHighlightWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategoryItemHighlightLayerDescription.prototype, "skipUnknownValues", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("SkipUnknownValues");
        },
        enumerable: false,
        configurable: true
    });
    CategoryItemHighlightLayerDescription.$t = markType(CategoryItemHighlightLayerDescription, 'CategoryItemHighlightLayerDescription', AnnotationLayerDescription.$);
    return CategoryItemHighlightLayerDescription;
}(AnnotationLayerDescription));
export { CategoryItemHighlightLayerDescription };
