// igniteui-angular-core/esm5/lib/CategorySeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var CategorySeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(CategorySeriesDescription, _super);
    function CategorySeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fo = false;
        _this.fn = false;
        _this.fq = false;
        _this.fx = null;
        _this.fp = false;
        _this.fw = null;
        _this.fv = null;
        return _this;
    }
    CategorySeriesDescription.prototype.get_type = function () {
        return "CategorySeries";
    };
    Object.defineProperty(CategorySeriesDescription.prototype, "isCustomCategoryStyleAllowed", {
        get: function () {
            return this.fo;
        },
        set: function (a) {
            this.fo = a;
            this.g("IsCustomCategoryStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "isCustomCategoryMarkerStyleAllowed", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("IsCustomCategoryMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "useHighMarkerFidelity", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("UseHighMarkerFidelity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "transitionInMode", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("TransitionInMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "isTransitionInEnabled", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("IsTransitionInEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "assigningCategoryStyleRef", {
        get: function () {
            return this.fw;
        },
        set: function (a) {
            this.fw = a;
            this.g("AssigningCategoryStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(CategorySeriesDescription.prototype, "assigningCategoryMarkerStyleRef", {
        get: function () {
            return this.fv;
        },
        set: function (a) {
            this.fv = a;
            this.g("AssigningCategoryMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    CategorySeriesDescription.$t = markType(CategorySeriesDescription, 'CategorySeriesDescription', MarkerSeriesDescription.$);
    return CategorySeriesDescription;
}(MarkerSeriesDescription));
export { CategorySeriesDescription };
