// igniteui-angular-core/esm5/lib/DataPieChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { DataPieBaseChartDescription } from "./DataPieBaseChartDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var DataPieChartDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataPieChartDescription, _super);
    function DataPieChartDescription() {
        var _this = _super.call(this) || this;
        _this.x7 = 0;
        _this.x8 = 0;
        _this.yf = null;
        _this.yi = null;
        _this.yc = null;
        _this.x5 = false;
        _this.yg = null;
        _this.yh = null;
        _this.ye = null;
        _this.yj = null;
        _this.yd = null;
        _this.yb = null;
        return _this;
    }
    DataPieChartDescription.prototype.get_type = function () {
        return "DataPieChart";
    };
    Object.defineProperty(DataPieChartDescription.prototype, "transitionInDuration", {
        get: function () {
            return this.x7;
        },
        set: function (a) {
            this.x7 = a;
            this.g("TransitionInDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "transitionOutDuration", {
        get: function () {
            return this.x8;
        },
        set: function (a) {
            this.x8 = a;
            this.g("TransitionOutDuration");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "transitionInEasingFunctionRef", {
        get: function () {
            return this.yf;
        },
        set: function (a) {
            this.yf = a;
            this.g("TransitionInEasingFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "transitionOutEasingFunctionRef", {
        get: function () {
            return this.yi;
        },
        set: function (a) {
            this.yi = a;
            this.g("TransitionOutEasingFunctionRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "chartType", {
        get: function () {
            return this.yc;
        },
        set: function (a) {
            this.yc = a;
            this.g("ChartType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "isTransitionInEnabled", {
        get: function () {
            return this.x5;
        },
        set: function (a) {
            this.x5 = a;
            this.g("IsTransitionInEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "transitionInMode", {
        get: function () {
            return this.yg;
        },
        set: function (a) {
            this.yg = a;
            this.g("TransitionInMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "transitionInSpeedType", {
        get: function () {
            return this.yh;
        },
        set: function (a) {
            this.yh = a;
            this.g("TransitionInSpeedType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "tooltipTemplateRef", {
        get: function () {
            return this.ye;
        },
        set: function (a) {
            this.ye = a;
            this.g("TooltipTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "width", {
        get: function () {
            return this.yj;
        },
        set: function (a) {
            this.yj = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "height", {
        get: function () {
            return this.yd;
        },
        set: function (a) {
            this.yd = a;
            this.g("Height");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieChartDescription.prototype, "background", {
        get: function () {
            return this.yb;
        },
        set: function (a) {
            this.yb = a;
            this.g("Background");
        },
        enumerable: false,
        configurable: true
    });
    DataPieChartDescription.$t = markType(DataPieChartDescription, 'DataPieChartDescription', DataPieBaseChartDescription.$);
    return DataPieChartDescription;
}(DataPieBaseChartDescription));
export { DataPieChartDescription };
