// igniteui-angular-core/esm5/lib/DataPieBaseChartDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { RadialBaseChartDescription } from "./RadialBaseChartDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var DataPieBaseChartDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(DataPieBaseChartDescription, _super);
    function DataPieBaseChartDescription() {
        var _this = _super.call(this) || this;
        _this.xi = null;
        _this.v6 = null;
        _this.xa = null;
        _this.v4 = null;
        _this.xf = null;
        _this.v5 = null;
        _this.w8 = null;
        _this.v3 = null;
        _this.wl = 0;
        _this.xk = null;
        _this.xj = null;
        _this.xb = null;
        _this.w5 = null;
        _this.xm = null;
        _this.w7 = null;
        _this.xe = null;
        _this.wm = 0;
        _this.xd = null;
        _this.w6 = null;
        _this.wd = false;
        _this.xl = null;
        _this.wj = 0;
        _this.wk = 0;
        _this.wp = 0;
        _this.we = false;
        _this.w3 = 0;
        _this.wr = 0;
        _this.wq = 0;
        _this.ws = 0;
        _this.wn = 0;
        _this.wo = 0;
        _this.xh = null;
        _this.xg = null;
        _this.w9 = null;
        _this.wb = false;
        _this.wc = false;
        _this.xc = null;
        return _this;
    }
    DataPieBaseChartDescription.prototype.get_type = function () {
        return "DataPieBaseChart";
    };
    Object.defineProperty(DataPieBaseChartDescription.prototype, "sliceLabelFormat", {
        get: function () {
            return this.xi;
        },
        set: function (a) {
            this.xi = a;
            this.g("SliceLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "sliceLabelFormatSpecifiers", {
        get: function () {
            return this.v6;
        },
        set: function (a) {
            this.v6 = a;
            this.g("SliceLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "legendSliceLabelFormat", {
        get: function () {
            return this.xa;
        },
        set: function (a) {
            this.xa = a;
            this.g("LegendSliceLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "legendSliceLabelFormatSpecifiers", {
        get: function () {
            return this.v4;
        },
        set: function (a) {
            this.v4 = a;
            this.g("LegendSliceLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "othersSliceLabelFormat", {
        get: function () {
            return this.xf;
        },
        set: function (a) {
            this.xf = a;
            this.g("OthersSliceLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "othersSliceLabelFormatSpecifiers", {
        get: function () {
            return this.v5;
        },
        set: function (a) {
            this.v5 = a;
            this.g("OthersSliceLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "legendOthersSliceLabelFormat", {
        get: function () {
            return this.w8;
        },
        set: function (a) {
            this.w8 = a;
            this.g("LegendOthersSliceLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "legendOthersSliceLabelFormatSpecifiers", {
        get: function () {
            return this.v3;
        },
        set: function (a) {
            this.v3 = a;
            this.g("LegendOthersSliceLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "innerExtent", {
        get: function () {
            return this.wl;
        },
        set: function (a) {
            this.wl = a;
            this.g("InnerExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "sweepDirection", {
        get: function () {
            return this.xk;
        },
        set: function (a) {
            this.xk = a;
            this.g("SweepDirection");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "sliceLabelPositionMode", {
        get: function () {
            return this.xj;
        },
        set: function (a) {
            this.xj = a;
            this.g("SliceLabelPositionMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "lightSliceLabelColor", {
        get: function () {
            return this.xb;
        },
        set: function (a) {
            this.xb = a;
            this.g("LightSliceLabelColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "darkSliceLabelColor", {
        get: function () {
            return this.w5;
        },
        set: function (a) {
            this.w5 = a;
            this.g("DarkSliceLabelColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.xm;
        },
        set: function (a) {
            this.xm = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "legendLabelMemberPath", {
        get: function () {
            return this.w7;
        },
        set: function (a) {
            this.w7 = a;
            this.g("LegendLabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "othersCategoryType", {
        get: function () {
            return this.xe;
        },
        set: function (a) {
            this.xe = a;
            this.g("OthersCategoryType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "othersCategoryThreshold", {
        get: function () {
            return this.wm;
        },
        set: function (a) {
            this.wm = a;
            this.g("OthersCategoryThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "othersCategoryText", {
        get: function () {
            return this.xd;
        },
        set: function (a) {
            this.xd = a;
            this.g("OthersCategoryText");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "labelMemberPath", {
        get: function () {
            return this.w6;
        },
        set: function (a) {
            this.w6 = a;
            this.g("LabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "valueAxisFavorLabellingScaleEnd", {
        get: function () {
            return this.wd;
        },
        set: function (a) {
            this.wd = a;
            this.g("ValueAxisFavorLabellingScaleEnd");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "valueAxisAutoRangeBufferMode", {
        get: function () {
            return this.xl;
        },
        set: function (a) {
            this.xl = a;
            this.g("ValueAxisAutoRangeBufferMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "angleAxisInterval", {
        get: function () {
            return this.wj;
        },
        set: function (a) {
            this.wj = a;
            this.g("AngleAxisInterval");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "angleAxisMinorInterval", {
        get: function () {
            return this.wk;
        },
        set: function (a) {
            this.wk = a;
            this.g("AngleAxisMinorInterval");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "valueAxisInterval", {
        get: function () {
            return this.wp;
        },
        set: function (a) {
            this.wp = a;
            this.g("ValueAxisInterval");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "valueAxisIsLogarithmic", {
        get: function () {
            return this.we;
        },
        set: function (a) {
            this.we = a;
            this.g("ValueAxisIsLogarithmic");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "valueAxisLogarithmBase", {
        get: function () {
            return this.w3;
        },
        set: function (a) {
            this.w3 = a;
            this.g("ValueAxisLogarithmBase");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "valueAxisMinimumValue", {
        get: function () {
            return this.wr;
        },
        set: function (a) {
            this.wr = a;
            this.g("ValueAxisMinimumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "valueAxisMaximumValue", {
        get: function () {
            return this.wq;
        },
        set: function (a) {
            this.wq = a;
            this.g("ValueAxisMaximumValue");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "valueAxisMinorInterval", {
        get: function () {
            return this.ws;
        },
        set: function (a) {
            this.ws = a;
            this.g("ValueAxisMinorInterval");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "radiusExtent", {
        get: function () {
            return this.wn;
        },
        set: function (a) {
            this.wn = a;
            this.g("RadiusExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "startAngle", {
        get: function () {
            return this.wo;
        },
        set: function (a) {
            this.wo = a;
            this.g("StartAngle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "sliceLabelContentSeparator", {
        get: function () {
            return this.xh;
        },
        set: function (a) {
            this.xh = a;
            this.g("SliceLabelContentSeparator");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "sliceLabelContentMode", {
        get: function () {
            return this.xg;
        },
        set: function (a) {
            this.xg = a;
            this.g("SliceLabelContentMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "legendSliceLabelContentMode", {
        get: function () {
            return this.w9;
        },
        set: function (a) {
            this.w9 = a;
            this.g("LegendSliceLabelContentMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "angleAxisFavorLabellingScaleEnd", {
        get: function () {
            return this.wb;
        },
        set: function (a) {
            this.wb = a;
            this.g("AngleAxisFavorLabellingScaleEnd");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "valueAxisAbbreviateLargeNumbers", {
        get: function () {
            return this.wc;
        },
        set: function (a) {
            this.wc = a;
            this.g("ValueAxisAbbreviateLargeNumbers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DataPieBaseChartDescription.prototype, "markerCollision", {
        get: function () {
            return this.xc;
        },
        set: function (a) {
            this.xc = a;
            this.g("MarkerCollision");
        },
        enumerable: false,
        configurable: true
    });
    DataPieBaseChartDescription.$t = markType(DataPieBaseChartDescription, 'DataPieBaseChartDescription', RadialBaseChartDescription.$);
    return DataPieBaseChartDescription;
}(RadialBaseChartDescription));
export { DataPieBaseChartDescription };
