// igniteui-angular-core/esm5/lib/GeographicProportionalSymbolSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicMarkerSeriesDescription } from "./GeographicMarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicProportionalSymbolSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicProportionalSymbolSeriesDescription, _super);
    function GeographicProportionalSymbolSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.gg = null;
        _this.gh = null;
        _this.gi = null;
        _this.ft = null;
        _this.gf = null;
        _this.gd = null;
        _this.fr = null;
        _this.fv = false;
        _this.fy = false;
        _this.f6 = 0;
        _this.f5 = 0;
        _this.fz = false;
        _this.fx = false;
        _this.fw = false;
        _this.ge = null;
        _this.f9 = 0;
        _this.gc = null;
        _this.gb = null;
        return _this;
    }
    GeographicProportionalSymbolSeriesDescription.prototype.get_type = function () {
        return "GeographicProportionalSymbolSeries";
    };
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "latitudeMemberPath", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("LatitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "longitudeMemberPath", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("LongitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "radiusMemberPath", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("RadiusMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "radiusScale", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("RadiusScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "labelMemberPath", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("LabelMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "fillMemberPath", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("FillMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "fillScale", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("FillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "fillScaleUseGlobalValues", {
        get: function () {
            return this.fv;
        },
        set: function (a) {
            this.fv = a;
            this.g("FillScaleUseGlobalValues");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "markerOutlineUsesFillScale", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            this.fy = a;
            this.g("MarkerOutlineUsesFillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "markerOutlineBrightness", {
        get: function () {
            return this.f6;
        },
        set: function (a) {
            this.f6 = a;
            this.g("MarkerOutlineBrightness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "markerBrushBrightness", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("MarkerBrushBrightness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "radiusScaleUseGlobalValues", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("RadiusScaleUseGlobalValues");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "isCustomScatterStyleAllowed", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("IsCustomScatterStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "isCustomScatterMarkerStyleAllowed", {
        get: function () {
            return this.fw;
        },
        set: function (a) {
            this.fw = a;
            this.g("IsCustomScatterMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "itemSearchMode", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("ItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "itemSearchThreshold", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("ItemSearchThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "assigningScatterStyleRef", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("AssigningScatterStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicProportionalSymbolSeriesDescription.prototype, "assigningScatterMarkerStyleRef", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("AssigningScatterMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicProportionalSymbolSeriesDescription.$t = markType(GeographicProportionalSymbolSeriesDescription, 'GeographicProportionalSymbolSeriesDescription', GeographicMarkerSeriesDescription.$);
    return GeographicProportionalSymbolSeriesDescription;
}(GeographicMarkerSeriesDescription));
export { GeographicProportionalSymbolSeriesDescription };
