// igniteui-angular-core/esm5/lib/GeographicMarkerSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicMarkerSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicMarkerSeriesDescription, _super);
    function GeographicMarkerSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fh = null;
        _this.e5 = 0;
        _this.ff = null;
        _this.fd = null;
        _this.fg = null;
        _this.fc = null;
        _this.fe = null;
        _this.e7 = 0;
        _this.fb = null;
        _this.e9 = null;
        _this.fa = null;
        return _this;
    }
    GeographicMarkerSeriesDescription.prototype.get_type = function () {
        return "GeographicMarkerSeries";
    };
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "markerType", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("MarkerType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "markerThickness", {
        get: function () {
            return this.e5;
        },
        set: function (a) {
            this.e5 = a;
            this.g("MarkerThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "markerOutlineMode", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            this.ff = a;
            this.g("MarkerOutlineMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "markerFillMode", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("MarkerFillMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "markerTemplateRef", {
        get: function () {
            return this.fg;
        },
        set: function (a) {
            this.fg = a;
            this.g("MarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "markerBrush", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("MarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "markerOutline", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("MarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "maximumMarkers", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("MaximumMarkers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "actualMarkerTemplateRef", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("ActualMarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "actualMarkerBrush", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("ActualMarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicMarkerSeriesDescription.prototype, "actualMarkerOutline", {
        get: function () {
            return this.fa;
        },
        set: function (a) {
            this.fa = a;
            this.g("ActualMarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    GeographicMarkerSeriesDescription.$t = markType(GeographicMarkerSeriesDescription, 'GeographicMarkerSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicMarkerSeriesDescription;
}(GeographicMapSeriesHostDescription));
export { GeographicMarkerSeriesDescription };
