// igniteui-angular-core/esm5/lib/GeographicHighDensityScatterSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicHighDensityScatterSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicHighDensityScatterSeriesDescription, _super);
    function GeographicHighDensityScatterSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fj = null;
        _this.fk = null;
        _this.e6 = false;
        _this.e5 = false;
        _this.fa = 0;
        _this.e9 = 0;
        _this.fi = null;
        _this.fh = null;
        _this.fd = 0;
        _this.fe = 0;
        _this.fl = null;
        return _this;
    }
    GeographicHighDensityScatterSeriesDescription.prototype.get_type = function () {
        return "GeographicHighDensityScatterSeries";
    };
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "latitudeMemberPath", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("LatitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "longitudeMemberPath", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("LongitudeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "useBruteForce", {
        get: function () {
            return this.e6;
        },
        set: function (a) {
            this.e6 = a;
            this.g("UseBruteForce");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "progressiveLoad", {
        get: function () {
            return this.e5;
        },
        set: function (a) {
            this.e5 = a;
            this.g("ProgressiveLoad");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMinimum", {
        get: function () {
            return this.fa;
        },
        set: function (a) {
            this.fa = a;
            this.g("HeatMinimum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMaximum", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("HeatMaximum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMinimumColor", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("HeatMinimumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "heatMaximumColor", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("HeatMaximumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "pointExtent", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("PointExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "progressiveStatus", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("ProgressiveStatus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicHighDensityScatterSeriesDescription.prototype, "progressiveLoadStatusChangedRef", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("ProgressiveLoadStatusChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicHighDensityScatterSeriesDescription.$t = markType(GeographicHighDensityScatterSeriesDescription, 'GeographicHighDensityScatterSeriesDescription', GeographicMapSeriesHostDescription.$);
    return GeographicHighDensityScatterSeriesDescription;
}(GeographicMapSeriesHostDescription));
export { GeographicHighDensityScatterSeriesDescription };
