// igniteui-angular-core/esm5/lib/GeographicShapeSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicMapSeriesHostDescription } from "./GeographicMapSeriesHostDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicShapeSeriesBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicShapeSeriesBaseDescription, _super);
    function GeographicShapeSeriesBaseDescription() {
        var _this = _super.call(this) || this;
        _this.fk = null;
        _this.e8 = false;
        _this.e7 = false;
        _this.fj = null;
        _this.fe = 0;
        _this.fd = 0;
        _this.e5 = null;
        _this.fb = 0;
        _this.fi = null;
        _this.fh = null;
        return _this;
    }
    GeographicShapeSeriesBaseDescription.prototype.get_type = function () {
        return "GeographicShapeSeriesBase";
    };
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "shapeMemberPath", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("ShapeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "isCustomShapeStyleAllowed", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("IsCustomShapeStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "isCustomShapeMarkerStyleAllowed", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("IsCustomShapeMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "itemSearchMode", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("ItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "itemSearchThreshold", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("ItemSearchThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "itemSearchPointsThreshold", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("ItemSearchPointsThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "shapefileDataSource", {
        get: function () {
            return this.e5;
        },
        set: function (a) {
            this.e5 = a;
            this.g("ShapefileDataSource");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "shapeFilterResolution", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("ShapeFilterResolution");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "assigningShapeStyleRef", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("AssigningShapeStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicShapeSeriesBaseDescription.prototype, "assigningShapeMarkerStyleRef", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("AssigningShapeMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicShapeSeriesBaseDescription.$t = markType(GeographicShapeSeriesBaseDescription, 'GeographicShapeSeriesBaseDescription', GeographicMapSeriesHostDescription.$);
    return GeographicShapeSeriesBaseDescription;
}(GeographicMapSeriesHostDescription));
export { GeographicShapeSeriesBaseDescription };
