// igniteui-angular-core/esm5/lib/GeographicContourLineSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicXYTriangulatingSeriesDescription } from "./GeographicXYTriangulatingSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicContourLineSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicContourLineSeriesDescription, _super);
    function GeographicContourLineSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fo = null;
        _this.fk = null;
        _this.fj = null;
        _this.fh = null;
        _this.fn = null;
        return _this;
    }
    GeographicContourLineSeriesDescription.prototype.get_type = function () {
        return "GeographicContourLineSeries";
    };
    Object.defineProperty(GeographicContourLineSeriesDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.fo;
        },
        set: function (a) {
            this.fo = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicContourLineSeriesDescription.prototype, "fillScale", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("FillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicContourLineSeriesDescription.prototype, "actualFillScale", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("ActualFillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicContourLineSeriesDescription.prototype, "valueResolver", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("ValueResolver");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicContourLineSeriesDescription.prototype, "triangulationStatusChangedRef", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("TriangulationStatusChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicContourLineSeriesDescription.$t = markType(GeographicContourLineSeriesDescription, 'GeographicContourLineSeriesDescription', GeographicXYTriangulatingSeriesDescription.$);
    return GeographicContourLineSeriesDescription;
}(GeographicXYTriangulatingSeriesDescription));
export { GeographicContourLineSeriesDescription };
