// igniteui-angular-core/esm5/lib/GeographicScatterAreaSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicXYTriangulatingSeriesDescription } from "./GeographicXYTriangulatingSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicScatterAreaSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicScatterAreaSeriesDescription, _super);
    function GeographicScatterAreaSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fl = null;
        _this.fi = null;
        _this.fh = null;
        _this.fm = null;
        return _this;
    }
    GeographicScatterAreaSeriesDescription.prototype.get_type = function () {
        return "GeographicScatterAreaSeries";
    };
    Object.defineProperty(GeographicScatterAreaSeriesDescription.prototype, "colorMemberPath", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("ColorMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicScatterAreaSeriesDescription.prototype, "colorScale", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("ColorScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicScatterAreaSeriesDescription.prototype, "actualColorScale", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("ActualColorScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicScatterAreaSeriesDescription.prototype, "triangulationStatusChangedRef", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("TriangulationStatusChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicScatterAreaSeriesDescription.$t = markType(GeographicScatterAreaSeriesDescription, 'GeographicScatterAreaSeriesDescription', GeographicXYTriangulatingSeriesDescription.$);
    return GeographicScatterAreaSeriesDescription;
}(GeographicXYTriangulatingSeriesDescription));
export { GeographicScatterAreaSeriesDescription };
