// igniteui-angular-core/esm5/lib/GeographicPolylineSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { GeographicShapeSeriesBaseDescription } from "./GeographicShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var GeographicPolylineSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(GeographicPolylineSeriesDescription, _super);
    function GeographicPolylineSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fv = null;
        _this.ft = null;
        _this.fu = null;
        _this.fq = 0;
        _this.fp = 0;
        _this.fw = null;
        return _this;
    }
    GeographicPolylineSeriesDescription.prototype.get_type = function () {
        return "GeographicPolylineSeries";
    };
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeStyleSelectorRef", {
        get: function () {
            return this.fv;
        },
        set: function (a) {
            this.fv = a;
            this.g("ShapeStyleSelectorRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeFill", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("ShapeFill");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeStroke", {
        get: function () {
            return this.fu;
        },
        set: function (a) {
            this.fu = a;
            this.g("ShapeStroke");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeStrokeThickness", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("ShapeStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "shapeOpacity", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("ShapeOpacity");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GeographicPolylineSeriesDescription.prototype, "styleShapeRef", {
        get: function () {
            return this.fw;
        },
        set: function (a) {
            this.fw = a;
            this.g("StyleShapeRef");
        },
        enumerable: false,
        configurable: true
    });
    GeographicPolylineSeriesDescription.$t = markType(GeographicPolylineSeriesDescription, 'GeographicPolylineSeriesDescription', GeographicShapeSeriesBaseDescription.$);
    return GeographicPolylineSeriesDescription;
}(GeographicShapeSeriesBaseDescription));
export { GeographicPolylineSeriesDescription };
