// igniteui-angular-core/esm5/lib/HighDensityScatterSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var HighDensityScatterSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(HighDensityScatterSeriesDescription, _super);
    function HighDensityScatterSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fg = null;
        _this.fk = null;
        _this.fj = null;
        _this.fh = null;
        _this.fl = null;
        _this.fi = null;
        _this.fm = null;
        _this.fn = null;
        _this.e2 = false;
        _this.e1 = false;
        _this.e6 = 0;
        _this.e5 = 0;
        _this.fe = null;
        _this.fd = null;
        _this.e9 = 0;
        _this.fa = 0;
        _this.ff = null;
        return _this;
    }
    HighDensityScatterSeriesDescription.prototype.get_type = function () {
        return "HighDensityScatterSeries";
    };
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xAxisRef", {
        get: function () {
            return this.fg;
        },
        set: function (a) {
            this.fg = a;
            this.g("XAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yAxisRef", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("YAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xMemberPath", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("XMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xMemberAsLegendLabel", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("XMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yMemberAsLegendLabel", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("YMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "xMemberAsLegendUnit", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("XMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yMemberAsLegendUnit", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("YMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "yMemberPath", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("YMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "useBruteForce", {
        get: function () {
            return this.e2;
        },
        set: function (a) {
            this.e2 = a;
            this.g("UseBruteForce");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "progressiveLoad", {
        get: function () {
            return this.e1;
        },
        set: function (a) {
            this.e1 = a;
            this.g("ProgressiveLoad");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMinimum", {
        get: function () {
            return this.e6;
        },
        set: function (a) {
            this.e6 = a;
            this.g("HeatMinimum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMaximum", {
        get: function () {
            return this.e5;
        },
        set: function (a) {
            this.e5 = a;
            this.g("HeatMaximum");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMinimumColor", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("HeatMinimumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "heatMaximumColor", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("HeatMaximumColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "pointExtent", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("PointExtent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "progressiveStatus", {
        get: function () {
            return this.fa;
        },
        set: function (a) {
            this.fa = a;
            this.g("ProgressiveStatus");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HighDensityScatterSeriesDescription.prototype, "progressiveLoadStatusChangedRef", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            this.ff = a;
            this.g("ProgressiveLoadStatusChangedRef");
        },
        enumerable: false,
        configurable: true
    });
    HighDensityScatterSeriesDescription.$t = markType(HighDensityScatterSeriesDescription, 'HighDensityScatterSeriesDescription', SeriesDescription.$);
    return HighDensityScatterSeriesDescription;
}(SeriesDescription));
export { HighDensityScatterSeriesDescription };
