// igniteui-angular-core/esm5/lib/igx-data-visualization-locale-pt-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Localization } from "./Localization";
import { DataVisualizationLocalePt } from './DataVisualizationLocalePt';
import * as i0 from "@angular/core";
var IgxDataVisualizationLocalePtModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataVisualizationLocalePtModule() {
        Localization.register("DataVisualization-pt", new DataVisualizationLocalePt());
    }
    IgxDataVisualizationLocalePtModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocalePtModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataVisualizationLocalePtModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocalePtModule, imports: [CommonModule] });
    IgxDataVisualizationLocalePtModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocalePtModule, imports: [[CommonModule]] });
    return IgxDataVisualizationLocalePtModule;
}());
export { IgxDataVisualizationLocalePtModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocalePtModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule],
                    exports: [],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
