// igniteui-angular-core/esm5/lib/igx-local-data-source-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LocalDataSource } from './LocalDataSource';
import { LiteralFilterExpression } from './LiteralFilterExpression';
import { PropertyReferenceFilterExpression } from './PropertyReferenceFilterExpression';
import { OperationFilterExpression } from './OperationFilterExpression';
import { FunctionFilterExpression } from './FunctionFilterExpression';
import { TypeRegistrar } from "./type";
import { IgxLocalDataSourceModule } from './igx-local-data-source-module';
import * as i0 from "@angular/core";
var IgxLocalDataSourceDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxLocalDataSourceDynamicModule() {
        TypeRegistrar.register('LocalDataSource', LocalDataSource.$type);
        TypeRegistrar.register('LocalDataSource', LocalDataSource.$type);
        TypeRegistrar.register('LiteralFilterExpression', LiteralFilterExpression.$type);
        TypeRegistrar.register('LiteralFilterExpression', LiteralFilterExpression.$type);
        TypeRegistrar.register('PropertyReferenceFilterExpression', PropertyReferenceFilterExpression.$type);
        TypeRegistrar.register('PropertyReferenceFilterExpression', PropertyReferenceFilterExpression.$type);
        TypeRegistrar.register('OperationFilterExpression', OperationFilterExpression.$type);
        TypeRegistrar.register('OperationFilterExpression', OperationFilterExpression.$type);
        TypeRegistrar.register('FunctionFilterExpression', FunctionFilterExpression.$type);
        TypeRegistrar.register('FunctionFilterExpression', FunctionFilterExpression.$type);
    }
    IgxLocalDataSourceDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLocalDataSourceDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxLocalDataSourceDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLocalDataSourceDynamicModule, imports: [CommonModule, IgxLocalDataSourceModule], exports: [IgxLocalDataSourceModule] });
    IgxLocalDataSourceDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLocalDataSourceDynamicModule, imports: [[CommonModule, IgxLocalDataSourceModule
            ], IgxLocalDataSourceModule] });
    return IgxLocalDataSourceDynamicModule;
}());
export { IgxLocalDataSourceDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLocalDataSourceDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxLocalDataSourceModule
                    ],
                    imports: [CommonModule, IgxLocalDataSourceModule
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
