// igniteui-angular-core/esm5/lib/igx-dv-interactivity-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CanvasGestureDOMEventProxy } from './CanvasGestureDOMEventProxy';
import { IgxTooltipContainerModule } from './igx-tooltip-container-module';
import { IgxSimpleDefaultTooltipComponent } from './igx-simple-default-tooltip-component';
import { TypeRegistrar } from "./type";
import * as i0 from "@angular/core";
var IgxDVInteractivityModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDVInteractivityModule() {
        TypeRegistrar.register('CanvasGestureDOMEventProxy', CanvasGestureDOMEventProxy.$type);
        IgxSimpleDefaultTooltipComponent.register();
    }
    IgxDVInteractivityModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDVInteractivityModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDVInteractivityModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDVInteractivityModule, declarations: [IgxSimpleDefaultTooltipComponent], imports: [CommonModule, IgxTooltipContainerModule], exports: [IgxTooltipContainerModule,
            IgxSimpleDefaultTooltipComponent] });
    IgxDVInteractivityModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDVInteractivityModule, imports: [[CommonModule, IgxTooltipContainerModule
            ], IgxTooltipContainerModule] });
    return IgxDVInteractivityModule;
}());
export { IgxDVInteractivityModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDVInteractivityModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxSimpleDefaultTooltipComponent],
                    exports: [IgxTooltipContainerModule,
                        IgxSimpleDefaultTooltipComponent
                    ],
                    imports: [CommonModule, IgxTooltipContainerModule
                    ],
                    entryComponents: [IgxSimpleDefaultTooltipComponent]
                }]
        }], ctorParameters: function () { return []; } });
