// igniteui-angular-core/esm5/lib/igx-local-data-source-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LocalDataSource } from './LocalDataSource';
import { LiteralFilterExpression } from './LiteralFilterExpression';
import { PropertyReferenceFilterExpression } from './PropertyReferenceFilterExpression';
import { OperationFilterExpression } from './OperationFilterExpression';
import { FunctionFilterExpression } from './FunctionFilterExpression';
import { TypeRegistrar } from "./type";
import * as i0 from "@angular/core";
var IgxLocalDataSourceModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxLocalDataSourceModule() {
        TypeRegistrar.register('LocalDataSource', LocalDataSource.$type);
        TypeRegistrar.register('LocalDataSource', LocalDataSource.$type);
        TypeRegistrar.register('LiteralFilterExpression', LiteralFilterExpression.$type);
        TypeRegistrar.register('LiteralFilterExpression', LiteralFilterExpression.$type);
        TypeRegistrar.register('PropertyReferenceFilterExpression', PropertyReferenceFilterExpression.$type);
        TypeRegistrar.register('PropertyReferenceFilterExpression', PropertyReferenceFilterExpression.$type);
        TypeRegistrar.register('OperationFilterExpression', OperationFilterExpression.$type);
        TypeRegistrar.register('OperationFilterExpression', OperationFilterExpression.$type);
        TypeRegistrar.register('FunctionFilterExpression', FunctionFilterExpression.$type);
        TypeRegistrar.register('FunctionFilterExpression', FunctionFilterExpression.$type);
    }
    IgxLocalDataSourceModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLocalDataSourceModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxLocalDataSourceModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLocalDataSourceModule, imports: [CommonModule] });
    IgxLocalDataSourceModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLocalDataSourceModule, imports: [[CommonModule,
            ]] });
    return IgxLocalDataSourceModule;
}());
export { IgxLocalDataSourceModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxLocalDataSourceModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [],
                    imports: [CommonModule,
                    ],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
