// igniteui-angular-core/esm5/lib/igx-template-content-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTemplateContentComponent } from './igx-template-content-component';
import { IgxTemplateContentModule } from './igx-template-content-module';
import { TypeRegistrar } from './type';
import * as i0 from "@angular/core";
var IgxTemplateContentDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxTemplateContentDynamicModule() {
        TypeRegistrar.registerCons("IgxTemplateContentComponent", IgxTemplateContentComponent);
    }
    IgxTemplateContentDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTemplateContentDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTemplateContentDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTemplateContentDynamicModule, imports: [CommonModule, IgxTemplateContentModule], exports: [IgxTemplateContentModule] });
    IgxTemplateContentDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTemplateContentDynamicModule, imports: [[CommonModule, IgxTemplateContentModule], IgxTemplateContentModule] });
    return IgxTemplateContentDynamicModule;
}());
export { IgxTemplateContentDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTemplateContentDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxTemplateContentModule],
                    imports: [CommonModule, IgxTemplateContentModule],
                    entryComponents: [IgxTemplateContentComponent]
                }]
        }], ctorParameters: function () { return []; } });
