// igniteui-angular-core/esm5/lib/igx-data-visualization-locale-tr-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Localization } from "./Localization";
import { DataVisualizationLocaleTr } from './DataVisualizationLocaleTr';
import * as i0 from "@angular/core";
var IgxDataVisualizationLocaleTrModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataVisualizationLocaleTrModule() {
        Localization.register("DataVisualization-tr", new DataVisualizationLocaleTr());
    }
    IgxDataVisualizationLocaleTrModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleTrModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataVisualizationLocaleTrModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleTrModule, imports: [CommonModule] });
    IgxDataVisualizationLocaleTrModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleTrModule, imports: [[CommonModule]] });
    return IgxDataVisualizationLocaleTrModule;
}());
export { IgxDataVisualizationLocaleTrModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleTrModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule],
                    exports: [],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
