// igniteui-angular-core/esm5/lib/igx-date-time-format-specifier-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE: you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxDateTimeFormatSpecifierModule } from './igx-date-time-format-specifier-module';
import { IgxDateTimeFormatSpecifier } from './igx-date-time-format-specifier';
import { DateTimeFormatSpecifier } from './DateTimeFormatSpecifier';
import { TypeRegistrar } from './type';
import * as i0 from "@angular/core";
var IgxDateTimeFormatSpecifierDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDateTimeFormatSpecifierDynamicModule() {
        TypeRegistrar.registerCons("IgxDateTimeFormatSpecifier", IgxDateTimeFormatSpecifier);
        TypeRegistrar.register("DateTimeFormatSpecifier", DateTimeFormatSpecifier.$type);
    }
    IgxDateTimeFormatSpecifierDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDateTimeFormatSpecifierDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDateTimeFormatSpecifierDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDateTimeFormatSpecifierDynamicModule, imports: [CommonModule, IgxDateTimeFormatSpecifierModule], exports: [IgxDateTimeFormatSpecifierModule] });
    IgxDateTimeFormatSpecifierDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDateTimeFormatSpecifierDynamicModule, imports: [[CommonModule, IgxDateTimeFormatSpecifierModule], IgxDateTimeFormatSpecifierModule] });
    return IgxDateTimeFormatSpecifierDynamicModule;
}());
export { IgxDateTimeFormatSpecifierDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDateTimeFormatSpecifierDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule, IgxDateTimeFormatSpecifierModule],
                    exports: [IgxDateTimeFormatSpecifierModule],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
