// igniteui-angular-core/esm5/lib/igx-tooltip-container-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTooltipContainerComponent } from './igx-tooltip-container-component';
import { IgxTemplateContentModule } from './igx-template-content-module';
import * as i0 from "@angular/core";
var IgxTooltipContainerModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxTooltipContainerModule() {
        IgxTooltipContainerComponent.register();
    }
    IgxTooltipContainerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTooltipContainerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTooltipContainerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTooltipContainerModule, declarations: [IgxTooltipContainerComponent], imports: [CommonModule, IgxTemplateContentModule], exports: [IgxTooltipContainerComponent, IgxTemplateContentModule] });
    IgxTooltipContainerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTooltipContainerModule, imports: [[CommonModule, IgxTemplateContentModule], IgxTemplateContentModule] });
    return IgxTooltipContainerModule;
}());
export { IgxTooltipContainerModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTooltipContainerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxTooltipContainerComponent],
                    exports: [IgxTooltipContainerComponent, IgxTemplateContentModule],
                    imports: [CommonModule, IgxTemplateContentModule],
                    entryComponents: [IgxTooltipContainerComponent]
                }]
        }], ctorParameters: function () { return []; } });
