// igniteui-angular-core/esm5/lib/igx-data-visualization-locale-en-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Localization } from "./Localization";
import { DataVisualizationLocaleEn } from './DataVisualizationLocaleEn';
import * as i0 from "@angular/core";
var IgxDataVisualizationLocaleEnModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxDataVisualizationLocaleEnModule() {
        Localization.register("DataVisualization-en", new DataVisualizationLocaleEn());
    }
    IgxDataVisualizationLocaleEnModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleEnModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxDataVisualizationLocaleEnModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleEnModule, imports: [CommonModule] });
    IgxDataVisualizationLocaleEnModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleEnModule, imports: [[CommonModule]] });
    return IgxDataVisualizationLocaleEnModule;
}());
export { IgxDataVisualizationLocaleEnModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxDataVisualizationLocaleEnModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [CommonModule],
                    exports: [],
                    entryComponents: []
                }]
        }], ctorParameters: function () { return []; } });
