// igniteui-angular-core/esm5/lib/igx-child-content-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxChildContentComponent } from './igx-child-content-component';
import { IgxChildContentModule } from './igx-child-content-module';
import { TypeRegistrar } from './type';
import * as i0 from "@angular/core";
var IgxChildContentDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxChildContentDynamicModule() {
        TypeRegistrar.registerCons("IgxChildContentComponent", IgxChildContentComponent);
    }
    IgxChildContentDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChildContentDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxChildContentDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChildContentDynamicModule, imports: [CommonModule, IgxChildContentModule], exports: [IgxChildContentModule] });
    IgxChildContentDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChildContentDynamicModule, imports: [[CommonModule, IgxChildContentModule], IgxChildContentModule] });
    return IgxChildContentDynamicModule;
}());
export { IgxChildContentDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxChildContentDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxChildContentModule],
                    imports: [CommonModule, IgxChildContentModule],
                    entryComponents: [IgxChildContentComponent]
                }]
        }], ctorParameters: function () { return []; } });
