// igniteui-angular-core/esm5/lib/igx-component-renderer-container-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxComponentRendererContainerComponent } from './igx-component-renderer-container-component';
import { TypeRegistrar } from './type';
import * as i0 from "@angular/core";
var IgxComponentRendererContainerModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxComponentRendererContainerModule() {
        TypeRegistrar.registerCons("IgxComponentRendererContainerComponent", IgxComponentRendererContainerComponent);
    }
    IgxComponentRendererContainerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxComponentRendererContainerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxComponentRendererContainerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxComponentRendererContainerModule, declarations: [IgxComponentRendererContainerComponent], imports: [CommonModule], exports: [IgxComponentRendererContainerComponent] });
    IgxComponentRendererContainerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxComponentRendererContainerModule, imports: [[CommonModule]] });
    return IgxComponentRendererContainerModule;
}());
export { IgxComponentRendererContainerModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxComponentRendererContainerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [IgxComponentRendererContainerComponent],
                    exports: [IgxComponentRendererContainerComponent],
                    imports: [CommonModule],
                    entryComponents: [IgxComponentRendererContainerComponent]
                }]
        }], ctorParameters: function () { return []; } });
