// igniteui-angular-core/esm5/lib/igx-tooltip-container-dynamic-module.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IgxTooltipContainerComponent } from './igx-tooltip-container-component';
import { IgxTooltipContainerModule } from './igx-tooltip-container-module';
import { IgxTemplateContentModule } from './igx-template-content-module';
import * as i0 from "@angular/core";
var IgxTooltipContainerDynamicModule = /** @class */ /*@__PURE__*/ (function () {
    function IgxTooltipContainerDynamicModule() {
        IgxTooltipContainerComponent.register();
    }
    IgxTooltipContainerDynamicModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTooltipContainerDynamicModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    IgxTooltipContainerDynamicModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTooltipContainerDynamicModule, imports: [CommonModule, IgxTemplateContentModule, IgxTooltipContainerModule], exports: [IgxTooltipContainerModule, IgxTemplateContentModule] });
    IgxTooltipContainerDynamicModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTooltipContainerDynamicModule, imports: [[CommonModule, IgxTemplateContentModule, IgxTooltipContainerModule], IgxTooltipContainerModule, IgxTemplateContentModule] });
    return IgxTooltipContainerDynamicModule;
}());
export { IgxTooltipContainerDynamicModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: IgxTooltipContainerDynamicModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    exports: [IgxTooltipContainerModule, IgxTemplateContentModule],
                    imports: [CommonModule, IgxTemplateContentModule, IgxTooltipContainerModule],
                    entryComponents: [IgxTooltipContainerComponent]
                }]
        }], ctorParameters: function () { return []; } });
