// igniteui-angular-core/esm5/lib/LabelCollisionManager.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends, __values } from "tslib";
import { Base, fromEnum, runOn, Point_$type, markType } from "./type";
import { List$1 } from "./List$1";
import { LabelCollisionPlacementPositions_$type } from "./LabelCollisionPlacementPositions";
import { QuadTree } from "./QuadTree";
import { Rect } from "./Rect";
import { LinkedList$1 } from "./LinkedList$1";
import { LabelCollisionInfo } from "./LabelCollisionInfo";
import { QuadTreeBoundingBox } from "./QuadTreeBoundingBox";
import { Random } from "./Random";
import { HashSet$1 } from "./HashSet$1";
import { MathUtil } from "./MathUtil";
import { GeometryUtil } from "./GeometryUtil";
import { truncate, isNaN_ } from "./number";
/**
 * @hidden
 */
var LabelCollisionManager = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(LabelCollisionManager, _super);
    function LabelCollisionManager(a, b, c, d, e, f, g, h, i, j, k) {
        var _this = _super.call(this) || this;
        _this.o = false;
        _this.a = new Array(8);
        _this.aj = null;
        _this.labelsMoved = null;
        _this.m = null;
        _this.h = null;
        _this.g = 0;
        _this.e = null;
        _this.ai = null;
        _this.au = 0;
        _this.a2 = 0;
        _this.a1 = 0;
        _this.bu = null;
        _this.aw = 0;
        _this.at = 0;
        _this.ag = new LinkedList$1(LabelCollisionInfo.$);
        _this.ay = 0;
        _this.ah = new List$1(LabelCollisionInfo.$, 0);
        _this.q = false;
        _this.p = false;
        _this.a0 = 0;
        _this.ax = 0;
        _this.ak = 0;
        _this.al = 0;
        _this.az = 0;
        _this.r = false;
        _this.n = false;
        _this.b = null;
        _this.a7 = null;
        _this.a5 = null;
        _this.am = 0;
        _this.an = 0;
        _this.o = k != null;
        _this.bu = c;
        _this.aw = a;
        _this.at = b;
        _this.h = new QuadTree(a, b, 8, 8);
        _this.g = f;
        _this.e = d;
        _this.ai = new List$1(LabelCollisionPlacementPositions_$type, 0);
        _this.ai.add(0);
        _this.ai.add(5);
        _this.ai.add(7);
        _this.ai.add(2);
        _this.ai.add(3);
        _this.ai.add(6);
        _this.ai.add(1);
        _this.ai.add(4);
        var l = 0;
        for (var m = 0; m < g.length; m++) {
            if (_this.ai.contains(g[m])) {
                _this.ai.remove(g[m]);
                _this.ai.insert(l, g[m]);
                l++;
            }
        }
        if (k != null) {
            _this.aj = new List$1(LabelCollisionPlacementPositions_$type, 0);
            for (var n = 0; n < _this.a.length; n++) {
                _this.a[n] = false;
            }
            for (var o = 0; o < k.length; o++) {
                _this.a[k[o]] = true;
                _this.aj.add(k[o]);
            }
            _this.b = k;
        }
        else {
            _this.aj = new List$1(LabelCollisionPlacementPositions_$type, 0);
            _this.aj.add(0);
            _this.aj.add(5);
            _this.aj.add(7);
            _this.aj.add(2);
            _this.aj.add(3);
            _this.aj.add(6);
            _this.aj.add(1);
            _this.aj.add(4);
            if (_this.ad(_this.g)) {
                _this.aj = new List$1(LabelCollisionPlacementPositions_$type, 0);
                _this.aj.add(10);
                _this.aj.add(11);
                _this.aj.add(8);
                _this.aj.add(9);
            }
            var p = 0;
            for (var q = 0; q < g.length; q++) {
                if (_this.aj.contains(g[q])) {
                    _this.aj.remove(g[q]);
                    _this.aj.insert(p, g[q]);
                    p++;
                }
            }
        }
        _this.au = e;
        _this.a2 = h;
        _this.a1 = i;
        _this.n = j;
        return _this;
    }
    LabelCollisionManager.prototype.l = function (a, b, c, d, e) {
        var f = 0;
        var g = 0;
        switch (b) {
            case 0:
                f = a.s;
                g = (((a.t - a.q) - a.o) - this.au - d);
                break;
            case 4:
                f = (((a.s - a.r) - a.p) - this.au) - c;
                g = (((a.t - a.q) - a.o) - this.au) - d;
                break;
            case 5:
                f = (((a.s + a.r) + a.p) + this.au) + c;
                g = (((a.t - a.q) - a.o) - this.au) - d;
                break;
            case 1:
                f = (((a.s - a.r) - a.p) - this.au) - c;
                g = a.t;
                break;
            case 2:
                f = (((a.s + a.r) + a.p) + this.au) + c;
                g = a.t;
                break;
            case 3:
                f = a.s;
                g = (((a.t + a.q) + a.o) + this.au) + d;
                break;
            case 6:
                f = (((a.s - a.r) - a.p) - this.au) - c;
                g = (((a.t + a.q) + a.o) + this.au) + d;
                break;
            case 7:
                f = (((a.s + a.r) + a.p) + this.au) + c;
                g = (((a.t + a.q) + a.o) + this.au) - d;
                break;
            case 8:
                {
                    var h = this.i(a);
                    if (h != null) {
                        a.b = 8;
                        f = h.d;
                        g = h.e;
                    }
                    else {
                        return null;
                    }
                    break;
                }
            case 9:
                {
                    var i = this.j(a);
                    if (i != null) {
                        a.b = 9;
                        f = i.d;
                        g = i.e;
                    }
                    else {
                        return null;
                    }
                    break;
                }
            case 10:
                {
                    var j = this.k(a);
                    if (j != null) {
                        a.b = 10;
                        f = j.d;
                        g = j.e;
                    }
                    else {
                        return null;
                    }
                    break;
                }
            case 11:
                {
                    var k = this.i(a);
                    if (k != null) {
                        a.b = 8;
                        f = k.d;
                        g = k.e;
                    }
                    else {
                        var l = this.j(a);
                        if (l != null) {
                            f = l.d;
                            g = l.e;
                        }
                        else {
                            var m = this.j(a);
                            if (m != null) {
                                f = m.d;
                                g = m.e;
                            }
                            else {
                                return null;
                            }
                        }
                    }
                    break;
                }
        }
        var n = new QuadTreeBoundingBox();
        n.d = f;
        n.e = g;
        n.g = a.p;
        n.f = a.o;
        n.i = a;
        if (this.x(n) && !e) {
            return null;
        }
        return n;
    };
    LabelCollisionManager.prototype.x = function (a) {
        var b = a.d - a.g;
        var c = a.d + a.g;
        var d = a.e - a.f;
        var e = a.e + a.f;
        var f = this.bu;
        if (b < f.left || c > f.right || d < f.top || e > f.bottom) {
            return true;
        }
        return false;
    };
    LabelCollisionManager.prototype.a8 = function (a) {
        this.ay++;
        this.ag.f(a);
        a.i = this.ag.d;
        if (a.e == null) {
            a.e = new QuadTreeBoundingBox();
            a.e.d = a.s;
            a.e.e = a.t;
            a.e.g = a.r;
            a.e.f = a.q;
            a.e.i = a;
            this.h.c(a.e);
        }
        if (a.a != null) {
            if (a.d == null) {
                a.d = QuadTreeBoundingBox.a(a.a.boundingBox);
                a.d.i = a;
                this.h.c(a.d);
            }
        }
    };
    LabelCollisionManager.prototype.bl = function (a) {
        if (a.e != null) {
            this.h.d(a.e);
            a.e = null;
        }
        if (a.d != null) {
            this.h.d(a.d);
            a.d = null;
        }
        if (a.c != null) {
            this.h.d(a.c);
            a.c = null;
        }
        if (a.i != null) {
            this.ag.h(a.i);
            a.i = null;
        }
        this.ay--;
    };
    LabelCollisionManager.prototype.bi = function (a) {
        if (a.e != null && a.e.d == a.s && a.e.e == a.t) {
            return;
        }
        if (a.e != null) {
            this.h.d(a.e);
        }
        if (a.d != null) {
            this.h.d(a.d);
            a.d = null;
        }
        var b = NaN;
        var c = NaN;
        if (a.e == null) {
            a.e = new QuadTreeBoundingBox();
            a.e.i = a;
        }
        else {
            if (a.c != null) {
                b = a.c.d - a.e.d;
                c = a.c.e - a.e.e;
            }
        }
        a.e.d = a.s;
        a.e.e = a.t;
        a.e.g = a.r;
        a.e.f = a.q;
        this.h.c(a.e);
        if (a.a != null) {
            if (a.d == null) {
                a.d = QuadTreeBoundingBox.a(a.a.boundingBox);
                a.d.i = a;
                this.h.c(a.d);
            }
        }
        if (a.c != null) {
            this.h.d(a.c);
            if (a.b == 8 || a.b == 9) {
                b = NaN;
            }
            if (!isNaN_(b)) {
                a.c = new QuadTreeBoundingBox();
                a.c.d = a.e.d + b;
                a.c.e = a.e.e + c;
                a.c.i = a;
                a.c.g = a.p;
                a.c.f = a.o;
                a.k = a.c.d;
                a.l = a.c.e;
                this.h.c(a.c);
            }
            else {
                a.c = this.l(a, a.b, 0, 0, true);
                if (a.c != null) {
                    this.bg(a);
                    a.k = a.c.d;
                    a.l = a.c.e;
                    this.h.c(a.c);
                }
                else {
                    this.bf(a);
                }
            }
        }
    };
    LabelCollisionManager.prototype.bh = function (a) {
        this.h.d(a.c);
        a.c.d = a.k;
        a.c.e = a.l;
        a.c.g = a.p;
        a.c.f = a.o;
        this.h.c(a.c);
    };
    LabelCollisionManager.prototype.bp = function () {
        this.p = false;
        {
            var a = this.ag.c;
            while (a != null) {
                a.c.h = false;
                if (a.c.c != null) {
                    this.h.d(a.c.c);
                    a.c.c = null;
                    if (a.c.j != null) {
                        a.c.j.clear();
                    }
                }
                a = a.a;
            }
            this.az = 0;
        }
        this.ax = 0;
        this.ak = 0.4;
        this.al = 0.4;
        var b = truncate(Math.ceil((this.a1 / 150) * this.a2));
        this.a0 = b;
        this.bc();
    };
    LabelCollisionManager.prototype.ba = function () {
        this.ah.clear();
    };
    LabelCollisionManager.prototype.a9 = function () {
        this.p = true;
    };
    LabelCollisionManager.prototype.be = function () {
        this.q = false;
        if (this.p) {
            return;
        }
        this.bc();
    };
    LabelCollisionManager.prototype.bc = function () {
        var e_1, _a;
        this.ah.clear();
        this.bj(false);
        if (this.g == 0) {
            for (var a = 0; a < this.a2; a++) {
                this.al = this.ak * (this.a0 - this.ax) / this.a0;
                this.bd();
                this.ax++;
                if (this.ax > this.a0) {
                    if (this.m != null) {
                        this.m();
                    }
                    return;
                }
            }
            if (this.labelsMoved != null) {
                try {
                    for (var _b = __values(fromEnum(this.ah)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var b = _c.value;
                    }
                }
                catch (e_1_1) {
                    e_1 = { error: e_1_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_1)
                            throw e_1.error;
                    }
                }
                this.labelsMoved(this.ah);
            }
            if (this.az > 0) {
                if (!this.q) {
                    this.q = true;
                    this.e.executeDelayed(runOn(this, this.be), 150);
                }
            }
            else {
                if (this.m != null) {
                    this.m();
                }
            }
        }
        else {
            if (this.g == 2) {
                var c = this.ag.c;
                this.h.k();
                while (c != null) {
                    var d = c.c;
                    if (d.c != null) {
                        this.h.d(d.c);
                        d.c = null;
                        d.f = false;
                        d.j = null;
                    }
                    c = c.a;
                }
                this.az = 0;
                this.bj(true);
            }
            if (this.labelsMoved != null) {
                this.labelsMoved(this.ah);
            }
        }
    };
    LabelCollisionManager.prototype.bd = function () {
        var a = truncate(Math.round((this.ay - 1) * LabelCollisionManager.a6.nextDouble()));
        var b = 0;
        var c = this.ag.c;
        var d = this.as(null);
        var e = 30;
        var f = 60;
        while (c != null) {
            var g = c.c;
            var h = 0;
            var i = 0;
            if (this.n) {
                h = LabelCollisionManager.a6.nextDouble() * g.p * 2;
                i = LabelCollisionManager.a6.nextDouble() * g.o * 2;
            }
            if (b == a) {
                var j = truncate(Math.round((this.aj.count - 1) * LabelCollisionManager.a6.nextDouble()));
                while (this.aj._inner[j] == g.b) {
                    j = truncate(Math.round((this.aj.count - 1) * LabelCollisionManager.a6.nextDouble()));
                }
                var k = g.c;
                var l = this.l(g, this.aj._inner[j], h, i, false);
                var m = 0;
                while (l == null) {
                    if (m >= e) {
                        h = 0;
                        i = 0;
                    }
                    if (m >= f) {
                        break;
                    }
                    j = truncate(Math.round((this.aj.count - 1) * LabelCollisionManager.a6.nextDouble()));
                    l = this.l(g, this.aj._inner[j], h, i, false);
                    m++;
                }
                g.c = l;
                if (k != null) {
                    this.h.d(k);
                }
                if (l != null) {
                    this.h.c(l);
                    var n = this.as(g);
                    var o = d > n;
                    if (!o) {
                        var p = Math.exp(-(n - d) / this.al);
                        if (LabelCollisionManager.a6.nextDouble() < p) {
                            o = true;
                        }
                    }
                    if (!o && k != null) {
                        this.h.d(l);
                        this.h.c(k);
                        g.c = k;
                        if (!this.r) {
                            this.as(g);
                        }
                    }
                    else {
                        g.b = this.aj._inner[j];
                        g.k = g.c.d;
                        g.l = g.c.e;
                        this.ah.add(g);
                    }
                }
                else {
                    if (!this.r) {
                        this.as(g);
                    }
                    g.f = false;
                    this.ah.add(g);
                }
                break;
            }
            b++;
            c = c.a;
        }
    };
    LabelCollisionManager.prototype.v = function (a, b) {
        return this.w(a, b);
    };
    LabelCollisionManager.prototype.w = function (a, b) {
        if (!a.c(b)) {
            return false;
        }
        if (a.i == null) {
            return true;
        }
        var c = a.i;
        if (c.a == null) {
            if (b.i != null) {
                var d = b.i;
                if (d.d == b) {
                    var e = d.a;
                    return e.e(a.j());
                }
                else {
                    return true;
                }
            }
        }
        var f = c.a;
        if (c.d == a) {
            return f.e(b.j());
        }
        if (b.i != null) {
            var g = b.i;
            if (g.d == b) {
                if (this.ad(this.g)) {
                    return false;
                }
                var h = g.a;
                return h.e(a.j());
            }
            else {
                return true;
            }
        }
        return true;
    };
    LabelCollisionManager.prototype.as = function (a) {
        if (this.r) {
            var b = 0;
            var c = this.ag.c;
            while (c != null) {
                var d = c.c;
                var e = d.f;
                d.f = false;
                if (d.c != null) {
                    var f = this.h.e(d.c);
                    for (var g = 0; g < f.count; g++) {
                        var h = f._inner[g].l.c;
                        while (h != null) {
                            var i = h.c;
                            if (this.ae(i, d)) {
                                h = h.a;
                                continue;
                            }
                            if (this.v(d.c, i)) {
                                d.f = true;
                                b += d.c.h(i);
                            }
                            h = h.a;
                        }
                    }
                    if (d.f != e) {
                        this.ah.add(d);
                    }
                }
                c = c.a;
            }
            return b;
        }
        else {
            if (a != null) {
                var j = new HashSet$1(LabelCollisionInfo.$, 0);
                var k = new HashSet$1(LabelCollisionInfo.$, 0);
                var l = new HashSet$1(LabelCollisionInfo.$, 0);
                var m = new List$1(LabelCollisionInfo.$, 0);
                if (a.j != null) {
                    for (var n = 0; n < a.j.count; n++) {
                        j.add_1(a.j._inner[n]);
                        k.add_1(a.j._inner[n]);
                    }
                }
                var o = a.f;
                a.f = false;
                if (a.c != null) {
                    var p = this.h.e(a.c);
                    for (var q = 0; q < p.count; q++) {
                        var r = p._inner[q].l.c;
                        while (r != null) {
                            var s = r.c;
                            if (this.ae(s, a)) {
                                r = r.a;
                                continue;
                            }
                            if (this.v(a.c, s) && !l.contains(s.i)) {
                                a.f = true;
                                l.add_1(s.i);
                                m.add(s.i);
                            }
                            r = r.a;
                        }
                    }
                    p = this.h.e(a.e);
                    for (var t = 0; t < p.count; t++) {
                        var u = p._inner[t].l.c;
                        while (u != null) {
                            var v = u.c;
                            if (this.ae(v, a)) {
                                u = u.a;
                                continue;
                            }
                            if (this.s(a.e, v)) {
                                u = u.a;
                                continue;
                            }
                            if (this.v(a.e, v) && !l.contains(v.i)) {
                                a.f = true;
                                l.add_1(v.i);
                                m.add(v.i);
                            }
                            u = u.a;
                        }
                    }
                    if (a.f != o) {
                        this.ah.add(a);
                    }
                    if (a.j != null) {
                        for (var w = a.j.count - 1; w >= 0; w--) {
                            if (!l.contains(a.j._inner[w])) {
                                var x = a.j._inner[w];
                                var y = false;
                                if (x != a) {
                                    y = x.j.remove(a);
                                }
                                else {
                                }
                                a.j.removeAt(w);
                                if (x.j.count == 0 && y) {
                                    this.az--;
                                }
                                var z = j.remove(x);
                                if (j.count == 0 && z) {
                                    this.az--;
                                }
                            }
                        }
                    }
                    for (var aa = 0; aa < m.count; aa++) {
                        var ab = m._inner[aa];
                        if (!j.contains(ab)) {
                            if (a.j == null) {
                                a.j = new List$1(LabelCollisionInfo.$, 0);
                            }
                            a.j.add(ab);
                            if (j.count == 0) {
                                this.az++;
                            }
                            j.add_1(ab);
                            if (ab != a) {
                                if (ab.j == null) {
                                    ab.j = new List$1(LabelCollisionInfo.$, 0);
                                }
                                if (ab.j.count == 0) {
                                    this.az++;
                                }
                                ab.j.add(a);
                            }
                        }
                    }
                }
            }
            return this.az;
        }
    };
    LabelCollisionManager.prototype.s = function (a, b) {
        if ((((a.i != null) && a.i.e == a) || ((a.i != null) && a.i.d == a)) && (((a.i != null) && a.i.e == b) || ((a.i != null) && a.i.d == b))) {
            return true;
        }
        return false;
    };
    Object.defineProperty(LabelCollisionManager.prototype, "a3", {
        get: function () {
            return this.az;
        },
        enumerable: false,
        configurable: true
    });
    LabelCollisionManager.prototype.bm = function (a, b, c) {
        this.bu = c;
        var d = new QuadTree(a, b, this.h.i, this.h.f);
        var e = this.ag.c;
        while (e != null) {
            d.c(e.c.e);
            if (e.c.d != null) {
                d.c(e.c.d);
            }
            if (e.c.c != null) {
                d.c(e.c.c);
            }
            e = e.a;
        }
        this.h = d;
        this.aw = a;
        this.at = b;
    };
    LabelCollisionManager.prototype.ab = function () {
        var a = this.ag.c;
        while (a != null) {
            a.c.f = false;
            if (a.c.c != null) {
                if (this.x(a.c.c)) {
                    return true;
                }
            }
            a = a.a;
        }
        return false;
    };
    LabelCollisionManager.prototype.t = function () {
        var a = this.ag.c;
        var b = this.aj;
        while (a != null) {
            if (a.c.c != null) {
                if (this.x(a.c.c)) {
                    var c = a.c;
                    for (var d = 0; d < b.count; d++) {
                        var e = this.l(c, b._inner[d], 0, 0, false);
                        if (e != null) {
                            return true;
                        }
                    }
                }
            }
            a = a.a;
        }
        return false;
    };
    LabelCollisionManager.prototype.a4 = function () {
        var a = this.ag.c;
        while (a != null) {
            a.c.f = false;
            a = a.a;
        }
        a = this.ag.c;
        while (a != null) {
            var b = a.c.c;
            var c = a.c.e;
            var d = this.h.e(c);
            for (var e = 0; e < d.count; e++) {
                var f = d._inner[e].l.c;
                while (f != null) {
                    var g = f.c;
                    if (this.s(c, g)) {
                        f = f.a;
                        continue;
                    }
                    if (this.v(c, g)) {
                        var h = f.c.i;
                        if (!this.ae(f.c, a.c)) {
                            h.f = true;
                            a.c.f = true;
                        }
                    }
                    f = f.a;
                }
            }
            if (b == null) {
                return -1;
            }
            d = this.h.e(b);
            for (var i = 0; i < d.count; i++) {
                var j = d._inner[i].l.c;
                while (j != null) {
                    var k = j.c;
                    if (this.v(b, k)) {
                        var l = j.c.i;
                        if (!this.ae(j.c, a.c)) {
                            l.f = true;
                            a.c.f = true;
                        }
                    }
                    j = j.a;
                }
            }
            a = a.a;
        }
        var m = 0;
        var n = 1.7976931348623157E+308;
        var o = -1.7976931348623157E+308;
        var p = 1.7976931348623157E+308;
        var q = -1.7976931348623157E+308;
        a = this.ag.c;
        while (a != null) {
            if (a.c.f) {
                m++;
                if (!isNaN_(a.c.s) && !isNaN_(a.c.t)) {
                    n = Math.min(n, a.c.s);
                    o = Math.max(o, a.c.s);
                    p = Math.min(p, a.c.t);
                    q = Math.max(q, a.c.t);
                }
            }
            a = a.a;
        }
        this.am = Math.abs(o - n);
        this.an = Math.abs(q - p);
        return m;
    };
    LabelCollisionManager.prototype.bj = function (a) {
        var b = this.ag.c;
        if (this.ad(this.g)) {
            this.bk();
            return;
        }
        while (b != null) {
            var c = b.c;
            if (c.c == null || a) {
                var d = false;
                if (a) {
                    var e = this.h.a(c.s, c.t, 1.5);
                    var f = this.c(e);
                    if (f != null) {
                        for (var g = 0; g < f.length; g++) {
                            var h = f[g];
                            var i = this.l(c, h, 0, 0, false);
                            if (i == null) {
                                continue;
                            }
                            var j = this.h.e(i);
                            d = true;
                            for (var k = 0; k < j.count; k++) {
                                var l = j._inner[k].l.c;
                                while (l != null) {
                                    var m = l.c;
                                    if (this.v(i, m)) {
                                        d = false;
                                        break;
                                    }
                                    l = l.a;
                                }
                            }
                            if (d) {
                                c.c = i;
                                c.b = h;
                                c.k = c.c.d;
                                c.l = c.c.e;
                                this.h.c(c.c);
                                this.ah.add(c);
                                break;
                            }
                        }
                    }
                }
                if (!d) {
                    d = this.z(c, false);
                }
                if (!d) {
                    if (c.c == null) {
                        this.z(c, true);
                    }
                }
            }
            b = b.a;
        }
    };
    LabelCollisionManager.prototype.bk = function () {
        var a = this.ag.c;
        var b = new List$1(LabelCollisionInfo.$, 0);
        var c = new List$1(LabelCollisionInfo.$, 0);
        var d = new List$1(LabelCollisionInfo.$, 0);
        var e = new List$1(LabelCollisionInfo.$, 0);
        var f = this.g;
        while (a != null) {
            var g = a.c;
            var h = g.a;
            if (g.a == null || h.type != 2) {
                g.h = true;
                this.ah.add(g);
                a = a.a;
                continue;
            }
            var i = h;
            var j = this.bt(i);
            var k = g;
            var l = k.p + k.p;
            var m = k.o + k.o;
            var n = (i.startAngle + i.endAngle) / 2;
            n = MathUtil.f(n);
            var o = Rect.empty;
            if (f == 3 || f == 7) {
                var p = this.i(k);
                if (p != null) {
                    if (this.u(p, g)) {
                        if (f == 7) {
                            this.bf(k);
                            a = a.a;
                            continue;
                        }
                    }
                    else {
                        k.b = 8;
                        this.bq(k, p);
                        d.add(k);
                        this.ah.add(k);
                        this.h.c(p);
                        a = a.a;
                        continue;
                    }
                }
                if (k.h) {
                    a = a.a;
                    continue;
                }
            }
            if (f == 5 || f == 3) {
                var q = this.j(k);
                if (q != null) {
                    if (this.u(q, g)) {
                        if (f == 5) {
                            this.bf(k);
                            a = a.a;
                            continue;
                        }
                    }
                    else {
                        k.b = 9;
                        e.add(k);
                        this.bq(k, q);
                        this.ah.add(k);
                        this.h.c(q);
                        a = a.a;
                        continue;
                    }
                }
                if (k.h) {
                    a = a.a;
                    continue;
                }
            }
            var r = this.k(k);
            if (r != null) {
                k.b = 10;
                this.bq(k, r);
                this.ah.add(k);
                this.h.c(k.c);
                if (k.c.d < j.x) {
                    c.add(k);
                }
                else {
                    b.add(k);
                }
            }
            else {
                this.bf(k);
            }
            a = a.a;
        }
        this.bo(b);
        this.bn(b);
        this.bo(c);
        this.bn(c);
    };
    LabelCollisionManager.prototype.bq = function (a, b) {
        var c = new List$1(LabelCollisionInfo.$, 0);
        var d = new HashSet$1(LabelCollisionInfo.$, 0);
        if (!this.r) {
            c = new List$1(LabelCollisionInfo.$, 0);
        }
        this.bb(a, b, c, d);
        var e = a.e;
        if (e == null) {
            e = new QuadTreeBoundingBox();
            e.i = a;
            e.d = a.s;
            e.e = a.t;
            e.g = a.r;
            e.f = a.q;
            this.h.c(e);
            a.e = e;
        }
        this.bb(a, e, c, d);
        this.af(a, c);
        a.c = b;
        a.k = a.c.d;
        a.l = a.c.e;
    };
    LabelCollisionManager.prototype.bb = function (a, b, c, d) {
        if (this.u(b, a)) {
            var e = this.h.e(b);
            for (var f = 0; f < e.count; f++) {
                var g = e._inner[f].l.c;
                while (g != null) {
                    var h = g.c;
                    if (this.v(b, h)) {
                        if (!this.r) {
                            if (!d.contains(h.i)) {
                                c.add(h.i);
                                d.add_1(h.i);
                            }
                        }
                    }
                    g = g.a;
                }
            }
        }
    };
    LabelCollisionManager.prototype.k = function (a) {
        var b = null;
        var c = a.a;
        if (c == null) {
            return null;
        }
        var d = this.bt(c);
        var e = a.p + a.p;
        var f = a.o + a.o;
        var g = (c.startAngle + c.endAngle) / 2;
        g = MathUtil.f(g);
        var h = GeometryUtil.q(d, g, c.radius + this.au);
        a.b = 10;
        this.bg(a);
        var i = LabelCollisionManager.ap(d, c.radius, c, a);
        var j = GeometryUtil.q(d, g, i);
        var k = new Rect(0, j.x - e / 2, j.y - f / 2, e, f);
        if (g < 90 && g >= 0) {
            k = new Rect(0, h.x, h.y, e, f);
        }
        else if (g < 180 && g >= 90) {
            k = new Rect(0, h.x - e, h.y, e, f);
        }
        else if (g < 270 && g >= 180) {
            k = new Rect(0, h.x - e, h.y - f, e, f);
        }
        else {
            k = new Rect(0, h.x, h.y - f, e, f);
        }
        if (k.y < 0) {
            k = new Rect(0, k.x, 0, k.width, k.height);
        }
        if (k.bottom > this.bu.bottom) {
            k = new Rect(0, k.x, this.bu.bottom - k.height, k.width, k.height);
        }
        b = ((function () {
            var $ret = new QuadTreeBoundingBox();
            $ret.d = k.left + k.width / 2;
            $ret.e = k.top + k.height / 2;
            $ret.g = k.width / 2;
            $ret.f = k.height / 2;
            $ret.i = a;
            return $ret;
        })());
        return b;
    };
    LabelCollisionManager.prototype.bg = function (a) {
        if (a == null) {
            return;
        }
        var b = !a.h;
        if (b) {
            return;
        }
        a.h = false;
        this.ah.add(a);
    };
    LabelCollisionManager.prototype.bf = function (a) {
        if (a == null) {
            return;
        }
        var b = !a.h;
        if (!b) {
            return;
        }
        a.h = true;
        this.ah.add(a);
    };
    LabelCollisionManager.prototype.j = function (a) {
        var b = null;
        var c = a.a;
        if (c == null) {
            return null;
        }
        var d = this.bt(c);
        var e = a.p + a.p;
        var f = a.o + a.o;
        var g = (c.startAngle + c.endAngle) / 2;
        g = MathUtil.f(g);
        var h = LabelCollisionManager.ap(d, c.radius, c, a);
        var i = GeometryUtil.q(d, g, h);
        var j = new Rect(0, i.x - e / 2, i.y - f / 2, e, f);
        var k = this.aa(c, j, a, d);
        if (k || this.g == 5) {
            a.b = 9;
            if (!k && this.g == 5) {
                this.bf(a);
            }
            else {
                b = ((function () {
                    var $ret = new QuadTreeBoundingBox();
                    $ret.d = j.left + j.width / 2;
                    $ret.e = j.top + j.height / 2;
                    $ret.g = j.width / 2;
                    $ret.f = j.height / 2;
                    $ret.i = a;
                    return $ret;
                })());
                this.bg(a);
                return b;
            }
        }
        return b;
    };
    LabelCollisionManager.prototype.i = function (a) {
        var b = null;
        var c = a.a;
        if (c == null) {
            return null;
        }
        var d = this.bt(c);
        var e = a.p + a.p;
        var f = a.o + a.o;
        var g = (c.startAngle + c.endAngle) / 2;
        g = MathUtil.f(g);
        var h = c.innerRadius;
        var i = GeometryUtil.q(d, g, c.radius - (c.radius - h) / 2);
        var j = new Rect(0, i.x - a.p, i.y - a.o, e, f);
        var k = this.aa(c, j, a, i);
        if (k || this.g == 7) {
            a.b = 8;
            if (!k && this.g == 7) {
                this.bf(a);
            }
            else {
                this.bg(a);
                b = ((function () {
                    var $ret = new QuadTreeBoundingBox();
                    $ret.d = j.left + j.width / 2;
                    $ret.e = j.top + j.height / 2;
                    $ret.g = j.width / 2;
                    $ret.f = j.height / 2;
                    $ret.i = a;
                    return $ret;
                })());
                return b;
            }
        }
        return b;
    };
    LabelCollisionManager.ap = function (a, b, c, d) {
        var e = (c.startAngle + c.endAngle / 2);
        e = MathUtil.f(e);
        var f = e * Math.PI / 180;
        var g = 1;
        var h = GeometryUtil.q({ $type: Point_$type, x: 0, y: 0 }, e, g);
        var i = LabelCollisionManager.bs(c, d);
        var j = Math.sqrt(i.x * i.x + i.y * i.y);
        var k = { $type: Point_$type, x: i.x / j, y: i.y / j };
        var l = LabelCollisionManager.aq(h, k);
        var m = j / b;
        var n = l * m;
        var o = n;
        var p = GeometryUtil.i({ $type: Point_$type, x: 0, y: 0 }, h);
        var q = GeometryUtil.i({ $type: Point_$type, x: 0, y: 0 }, k);
        if (q < p) {
            o = -o;
        }
        var r = GeometryUtil.q(a, (f + o) * 180 / Math.PI, b);
        var s = LabelCollisionManager.br(c, d, r);
        var t = Math.sqrt((s.x - a.x) * (s.x - a.x) + (s.y - a.y) * (s.y - a.y));
        var u = t - 5;
        return u;
    };
    LabelCollisionManager.br = function (a, b, c) {
        var d = (a.startAngle + a.endAngle / 2);
        d = MathUtil.f(d);
        var e = GeometryUtil.m(d);
        var f = b.p;
        var g = b.o;
        if (e >= 0 && e <= 90) {
            return { $type: Point_$type, x: c.x - f, y: c.y - g };
        }
        if (e >= 90 && e <= 180) {
            return { $type: Point_$type, x: c.x + f, y: c.y - g };
        }
        if (e >= 180 && e <= 270) {
            return { $type: Point_$type, x: c.x + f, y: c.y + g };
        }
        return { $type: Point_$type, x: c.x - f, y: c.y + g };
    };
    LabelCollisionManager.aq = function (a, b) {
        var c = a.x * b.x + a.y * b.y;
        var d = Math.sqrt(a.x * a.x + a.y * a.y);
        var e = Math.sqrt(b.x * b.x + b.y * b.y);
        var f = Math.acos(c / (d * e));
        return f;
    };
    LabelCollisionManager.bs = function (a, b) {
        var c = (a.startAngle + a.endAngle / 2);
        c = MathUtil.f(c);
        var d = GeometryUtil.m(c);
        var e = b.p;
        var f = b.o;
        if (d >= 0 && d <= 90) {
            return { $type: Point_$type, x: e, y: f };
        }
        if (d >= 90 && d <= 180) {
            return { $type: Point_$type, x: -e, y: f };
        }
        if (d >= 180 && d <= 270) {
            return { $type: Point_$type, x: -e, y: -f };
        }
        return { $type: Point_$type, x: e, y: -f };
    };
    LabelCollisionManager.ao = function (a, b, c, d) {
        var e = MathUtil.g(a - b, c - d);
        var f = Math.asin((c - d) / e) * 180 / Math.PI;
        if (a < b) {
            f = 180 - f;
        }
        if (a > b) {
            f = 360 + f;
        }
        if (f == 360) {
            f = 0;
        }
        return GeometryUtil.m(f);
    };
    LabelCollisionManager.prototype.bo = function (a) {
        var b = new Array(a.count);
        for (var c = 0; c < a.count; c++) {
            var d = a._inner[c];
            var e = d.a;
            var f = this.bt(e);
            var g = (e.startAngle + e.endAngle) / 2;
            g = MathUtil.f(g);
            var h = GeometryUtil.q(f, g, e.radius);
            b[c] = h.y;
        }
        for (var i = 0; i < a.count; i++) {
            for (var j = i + 1; j < a.count; j++) {
                if (b[i] >= b[j]) {
                    var k = a._inner[i];
                    a._inner[i] = a._inner[j];
                    a._inner[j] = k;
                    var l = b[i];
                    b[i] = b[j];
                    b[j] = l;
                }
            }
        }
    };
    LabelCollisionManager.prototype.bn = function (a) {
        var e_2, _a, e_3, _b;
        if (a.count == 0) {
            return;
        }
        var b = this.bu.width;
        var c = this.bu.height;
        var d = a.count;
        var e = a._inner[0].a;
        var f = e.radius;
        var g = this.bt(e);
        var h = true;
        var i = 0;
        var j = Number.POSITIVE_INFINITY;
        var k = Number.NEGATIVE_INFINITY;
        i = this.a4();
        var l = 0;
        try {
            for (var _c = __values(fromEnum(a)), _d = _c.next(); !_d.done; _d = _c.next()) {
                var m = _d.value;
                j = Math.min(j, m.o + m.o);
                k = Math.max(k, m.o + m.o);
                l += m.o + m.o;
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return))
                    _a.call(_c);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
        if (l > c) {
            h = false;
        }
        if (h && i > 0) {
            for (var n = 0; n < d - 1; n++) {
                var _loop_1 = function (o) {
                    var p = a._inner[n];
                    var q = a._inner[o];
                    if (p.c.c(q.c)) {
                        var r_1 = q.c;
                        var s = r_1.e - r_1.f;
                        var t = r_1.g + r_1.g;
                        var u = MathUtil.f(this_1.ar(q));
                        var v_1 = Math.min(p.c.e + p.c.f + 0.01, c - j);
                        var w = this_1.au + f;
                        var x = Math.abs(g.y - (s + j / 2));
                        var y = Math.sqrt(Math.abs(w * w - x * x));
                        var z = GeometryUtil.m(u);
                        if (z > 90 && z < 270) {
                            y = (t + y) * -1;
                        }
                        var aa_1 = g.x + y;
                        var ab = ((function () {
                            var $ret = new QuadTreeBoundingBox();
                            $ret.d = aa_1 + r_1.g;
                            $ret.e = v_1 + r_1.f;
                            $ret.g = r_1.g;
                            $ret.f = r_1.f;
                            $ret.i = q;
                            return $ret;
                        })());
                        if (q.c != null) {
                            this_1.h.d(q.c);
                        }
                        this_1.bq(q, ab);
                        this_1.h.c(q.c);
                    }
                };
                var this_1 = this;
                for (var o = n + 1; o < d; o++) {
                    _loop_1(o);
                }
            }
            for (var ac = d - 1; ac > 0; ac--) {
                var _loop_2 = function (ad) {
                    var ae = a._inner[ac];
                    var af = a._inner[ad];
                    if (ae.c.c(af.c)) {
                        var ag_1 = af.c;
                        var ah = ag_1.e - ag_1.f;
                        var ai = ag_1.g + ag_1.g;
                        var aj_1 = Math.max((ae.c.e - ae.c.f) - j - 0.01, 0);
                        var ak = this_2.au + f;
                        var al = Math.abs(g.y - (ah + j / 2));
                        var am = Math.sqrt(Math.abs(ak * ak - al * al));
                        var an = GeometryUtil.m(MathUtil.f(this_2.ar(af)));
                        if (an > 90 && an < 270) {
                            am = (ai + am) * -1;
                        }
                        var ao_1 = g.x + am;
                        var ap = ((function () {
                            var $ret = new QuadTreeBoundingBox();
                            $ret.d = ao_1 + ag_1.g;
                            $ret.e = aj_1 + ag_1.f;
                            $ret.g = ag_1.g;
                            $ret.f = ag_1.f;
                            $ret.i = af;
                            return $ret;
                        })());
                        if (af.c != null) {
                            this_2.h.d(af.c);
                        }
                        this_2.bq(af, ap);
                        this_2.h.c(af.c);
                    }
                };
                var this_2 = this;
                for (var ad = ac - 1; ad >= 0; ad--) {
                    _loop_2(ad);
                }
            }
        }
        try {
            for (var _e = __values(fromEnum(a)), _f = _e.next(); !_f.done; _f = _e.next()) {
                var aq = _f.value;
                var ar = aq.c.j();
                if (ar.left > b || ar.right < 0) {
                    this.bf(aq);
                    if (aq.c != null) {
                        this.h.d(aq.c);
                    }
                    this.af(aq, null);
                    aq.c = null;
                }
                else if (ar.left < 0) {
                    this.bf(aq);
                    if (aq.c != null) {
                        this.h.d(aq.c);
                    }
                    this.af(aq, null);
                    aq.c = null;
                }
                else if (ar.right > b) {
                    this.bf(aq);
                    if (aq.c != null) {
                        this.h.d(aq.c);
                    }
                    this.af(aq, null);
                    aq.c = null;
                }
            }
        }
        catch (e_3_1) {
            e_3 = { error: e_3_1 };
        }
        finally {
            try {
                if (_f && !_f.done && (_b = _e.return))
                    _b.call(_e);
            }
            finally {
                if (e_3)
                    throw e_3.error;
            }
        }
    };
    LabelCollisionManager.prototype.ar = function (a) {
        var b = a.a;
        return (b.startAngle + b.endAngle) / 2;
    };
    LabelCollisionManager.av = function (a) {
        var b = Math.round(a * Math.pow(10, 5)) / Math.pow(10, 5);
        return b;
    };
    LabelCollisionManager.prototype.ac = function (a) {
        return LabelCollisionManager.av(MathUtil.f(Math.abs(a.endAngle - a.startAngle))) == 360;
    };
    LabelCollisionManager.prototype.aa = function (a, b, c, d) {
        if (a == null) {
            return false;
        }
        var e = this.bt(a);
        var f = Math.min(a.startAngle, a.endAngle);
        var g = Math.max(a.startAngle, a.endAngle);
        f = MathUtil.f(f);
        g = MathUtil.f(g);
        var h = false;
        var i = false;
        var j = GeometryUtil.q(e, f, a.radius);
        var k = GeometryUtil.q(e, g, a.radius);
        f = LabelCollisionManager.ao(j.x, e.x, j.y, e.y);
        g = LabelCollisionManager.ao(k.x, e.x, k.y, e.y);
        if (this.ac(a)) {
            i = true;
        }
        var l;
        l = MathUtil.g(b.right - e.x, b.top - e.y);
        if (l > a.radius) {
            return false;
        }
        l = MathUtil.g(b.right - e.x, b.bottom - e.y);
        if (l > a.radius) {
            return false;
        }
        l = MathUtil.g(b.left - e.x, b.top - e.y);
        if (l > a.radius) {
            return false;
        }
        l = MathUtil.g(b.left - e.x, b.bottom - e.y);
        if (l > a.radius) {
            return false;
        }
        if (i) {
            return true;
        }
        if (f > g) {
            f = f - 360;
            h = true;
        }
        var m;
        m = LabelCollisionManager.ao(b.right, e.x, b.top, e.y);
        if (h && m > 180 && m < 360) {
            m = m - 360;
        }
        if (m < f || m > g) {
            return false;
        }
        m = LabelCollisionManager.ao(b.right, e.x, b.bottom, e.y);
        if (h && m > 180 && m < 360) {
            m = m - 360;
        }
        if (m < f || m > g) {
            return false;
        }
        m = LabelCollisionManager.ao(b.left, e.x, b.top, e.y);
        if (h && m > 180 && m < 360) {
            m = m - 360;
        }
        if (m < f || m > g) {
            return false;
        }
        m = LabelCollisionManager.ao(b.left, e.x, b.bottom, e.y);
        if (h && m > 180 && m < 360) {
            m = m - 360;
        }
        if (m < f || m > g) {
            return false;
        }
        return true;
    };
    LabelCollisionManager.prototype.bt = function (a) {
        return { $type: Point_$type, x: a.centerX, y: a.centerY };
    };
    LabelCollisionManager.prototype.ad = function (a) {
        return a == 7 || a == 3 || a == 6 || a == 4 || a == 5;
    };
    LabelCollisionManager.prototype.z = function (a, b) {
        var c = this.aj;
        a.h = false;
        var d = false;
        if (!d) {
            for (var e = 0; e < c.count; e++) {
                var f = this.l(a, c._inner[e], 0, 0, b);
                if (a.h) {
                    break;
                }
                if (f == null) {
                    continue;
                }
                d = !this.u(f, a);
                var g = a.e;
                if (g == null) {
                    g = new QuadTreeBoundingBox();
                    g.i = a;
                    g.d = a.s;
                    g.e = a.t;
                    g.g = a.r;
                    g.f = a.q;
                    this.h.c(g);
                    a.e = g;
                }
                if (a.a != null) {
                    if (a.d == null) {
                        a.d = QuadTreeBoundingBox.a(a.a.boundingBox);
                        a.d.i = a;
                        this.h.c(a.d);
                    }
                }
                var h = this.u(g, a);
                d = d && !h;
                if (d) {
                    a.c = f;
                    if (c._inner[e] != 11) {
                        a.b = c._inner[e];
                    }
                    a.k = a.c.d;
                    a.l = a.c.e;
                    this.h.c(a.c);
                    this.ah.add(a);
                    break;
                }
            }
        }
        if (!d && !a.h) {
            var i = Number.POSITIVE_INFINITY;
            var j = 0;
            var k = null;
            for (var l = 0; l < c.count; l++) {
                var m = 0;
                var n = null;
                var o = new HashSet$1(LabelCollisionInfo.$, 0);
                if (!this.r) {
                    n = new List$1(LabelCollisionInfo.$, 0);
                }
                var p = this.l(a, c._inner[l], 0, 0, b);
                if (p == null) {
                    continue;
                }
                var q = this.h.e(p);
                for (var r = 0; r < q.count; r++) {
                    var s = q._inner[r].l.c;
                    while (s != null) {
                        var t = s.c;
                        if (this.v(p, t)) {
                            m += p.h(t);
                            if (!this.r) {
                                if (!o.contains(t.i)) {
                                    n.add(t.i);
                                    o.add_1(t.i);
                                }
                            }
                        }
                        s = s.a;
                    }
                }
                var u = a.e;
                if (u == null) {
                    u = new QuadTreeBoundingBox();
                    u.i = a;
                    u.d = a.s;
                    u.e = a.t;
                    u.g = a.r;
                    u.f = a.q;
                }
                q = this.h.e(u);
                for (var v = 0; v < q.count; v++) {
                    var w = q._inner[v].l.c;
                    while (w != null) {
                        var x = w.c;
                        if (this.ae(w.c, a)) {
                            w = w.a;
                            continue;
                        }
                        if (this.s(u, x)) {
                            w = w.a;
                            continue;
                        }
                        if (this.v(u, x)) {
                            if (!this.r) {
                                if (!o.contains(x.i)) {
                                    n.add(x.i);
                                    o.add_1(x.i);
                                }
                            }
                        }
                        w = w.a;
                    }
                }
                if (m < i) {
                    var y = this.l(a, c._inner[l], 0, 0, b);
                    if (y != null) {
                        i = m;
                        j = l;
                        if (!this.r) {
                            k = n;
                        }
                    }
                }
            }
            var z = this.l(a, c._inner[j], 0, 0, b);
            a.f = true;
            if (!this.af(a, k)) {
                return false;
            }
            a.h = false;
            a.c = z;
            a.b = c._inner[j];
            a.k = a.c.d;
            a.l = a.c.e;
            this.h.c(a.c);
            this.ah.add(a);
        }
        return d;
    };
    LabelCollisionManager.prototype.af = function (a, b) {
        if (!this.r) {
            if (a.j != null && a.j.count > 0) {
                this.az--;
                for (var c = a.j.count - 1; c >= 0; c--) {
                    var d = a.j._inner[c];
                    if (d.j != null) {
                        d.j.remove(a);
                        if (d.j.count == 0) {
                            this.az--;
                        }
                    }
                }
            }
            if (b == null) {
                a.f = false;
                this.ah.add(a);
                return false;
            }
            a.j = b;
            if (b.count > 0) {
                this.az++;
            }
            for (var e = 0; e < b.count; e++) {
                var f = b._inner[e];
                if (f.j == null) {
                    f.j = new List$1(LabelCollisionInfo.$, 0);
                }
                if (!f.j.contains(a)) {
                    if (f.j.count == 0) {
                        this.az++;
                    }
                    f.j.add(a);
                }
            }
        }
        return true;
    };
    LabelCollisionManager.prototype.ae = function (a, b) {
        if (a.i == b && (a.i == null || a.i.d != a)) {
            return true;
        }
        return false;
    };
    LabelCollisionManager.prototype.u = function (a, b) {
        var c = false;
        var d = this.h.e(a);
        for (var e = 0; e < d.count; e++) {
            var f = d._inner[e].l.c;
            while (f != null) {
                if (this.ae(f.c, b)) {
                    f = f.a;
                    continue;
                }
                var g = f.c;
                if (this.s(a, g)) {
                    f = f.a;
                    continue;
                }
                if (this.v(a, g)) {
                    c = true;
                    break;
                }
                f = f.a;
            }
        }
        return c;
    };
    LabelCollisionManager.prototype.c = function (a) {
        var b = new Array(3);
        var c = GeometryUtil.i({ $type: Point_$type, x: 0, y: 0 }, { $type: Point_$type, x: a.a, y: a.b });
        c = c * 180 / Math.PI;
        c = GeometryUtil.m(c);
        if ((c >= 0 && c <= 22.5) || (c >= 337.5 && c <= 360)) {
            b[0] = 2;
            b[1] = 5;
            b[2] = 7;
        }
        if (c >= 22.5 && c <= 67.5) {
            b[0] = 2;
            b[1] = 7;
            b[2] = 3;
        }
        if (c >= 67.5 && c <= 112.5) {
            b[0] = 7;
            b[1] = 3;
            b[2] = 6;
        }
        if (c >= 112.5 && c <= 157.5) {
            b[0] = 3;
            b[1] = 6;
            b[2] = 1;
        }
        if (c >= 157.5 && c <= 202.5) {
            b[0] = 6;
            b[1] = 1;
            b[2] = 4;
        }
        if (c >= 202.5 && c <= 247.5) {
            b[0] = 1;
            b[1] = 4;
            b[2] = 0;
        }
        if (c >= 247.5 && c <= 292.5) {
            b[0] = 4;
            b[1] = 0;
            b[2] = 5;
        }
        if (c >= 292.5 && c <= 337.5) {
            b[0] = 0;
            b[1] = 5;
            b[2] = 2;
        }
        if (this.o) {
            var d = 0;
            for (var e = 0; e < b.length; e++) {
                if (this.a[b[e]]) {
                    d++;
                }
            }
            if (d == 0) {
                return null;
            }
            var f = 0;
            var g = new Array(d);
            for (var h = 0; h < b.length; h++) {
                if (this.a[b[h]]) {
                    g[f] = b[h];
                    f++;
                }
            }
            b = g;
        }
        return b;
    };
    LabelCollisionManager.prototype.f = function (a, b) {
        var c = false;
        var d = null;
        var e = new QuadTreeBoundingBox();
        e.d = a.x;
        e.e = a.y;
        e.g = 5;
        e.f = 5;
        var f = this.h.e(e);
        for (var g = 0; g < f.count; g++) {
            var h = f._inner[g].l.c;
            while (h != null) {
                var i = h.c;
                if (i.i.e == i) {
                    h = h.a;
                    continue;
                }
                if (this.v(e, i)) {
                    d = i;
                    c = true;
                    break;
                }
                h = h.a;
            }
        }
        if (c) {
            return d.i;
        }
        else {
            return null;
        }
    };
    LabelCollisionManager.prototype.y = function () {
        return this.am < 10 && this.an < 10;
    };
    LabelCollisionManager.$t = markType(LabelCollisionManager, 'LabelCollisionManager');
    LabelCollisionManager.a6 = new Random(0);
    return LabelCollisionManager;
}(Base));
export { LabelCollisionManager };
