// igniteui-angular-core/esm5/lib/NativeUIPopup.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { NativeUIContainer } from "./NativeUIContainer";
import { NativeUIPopupAnimationType_$type } from "./NativeUIPopupAnimationType";
import { NativeUIPopupPointerPosition_$type } from "./NativeUIPopupPointerPosition";
import { typeGetValue, enumGetBox, EnumUtil, markType } from "./type";
/**
 * @hidden
 */
var NativeUIPopup = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NativeUIPopup, _super);
    function NativeUIPopup() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.bt = null;
        _this.br = null;
        return _this;
    }
    Object.defineProperty(NativeUIPopup.prototype, "bp", {
        get: function () {
            return this.d;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "b2", {
        get: function () {
            return this.ag(22);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "b1", {
        get: function () {
            return this.ag(59);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "by", {
        get: function () {
            return this.ag(60);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "bw", {
        get: function () {
            return this.ag(25);
        },
        set: function (a) {
            this.a2(25, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "b5", {
        get: function () {
            return typeGetValue(this.ag(26));
        },
        set: function (a) {
            this.a2(26, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "bx", {
        get: function () {
            return this.ag(39);
        },
        set: function (a) {
            this.a2(39, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "bq", {
        get: function () {
            return EnumUtil.getEnumValue(NativeUIPopupAnimationType_$type, this.ag(35));
        },
        set: function (a) {
            this.a2(35, enumGetBox(NativeUIPopupAnimationType_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "cb", {
        get: function () {
            return this.ag(23);
        },
        set: function (a) {
            this.a2(23, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "b3", {
        get: function () {
            return this.ag(17);
        },
        set: function (a) {
            this.a2(17, a);
        },
        enumerable: false,
        configurable: true
    });
    NativeUIPopup.prototype.get_u = function () {
        return this.ag(36);
    };
    NativeUIPopup.prototype.set_u = function (a) {
        this.a2(36, a);
    };
    Object.defineProperty(NativeUIPopup.prototype, "bz", {
        get: function () {
            return this.ag(37);
        },
        set: function (a) {
            this.a2(37, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "b6", {
        get: function () {
            return typeGetValue(this.ag(40));
        },
        set: function (a) {
            this.a2(40, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "b0", {
        get: function () {
            return this.ag(41);
        },
        set: function (a) {
            this.a2(41, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "bv", {
        get: function () {
            return EnumUtil.getEnumValue(NativeUIPopupPointerPosition_$type, this.ag(43));
        },
        set: function (a) {
            this.a2(43, enumGetBox(NativeUIPopupPointerPosition_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "cc", {
        get: function () {
            return this.ag(42);
        },
        set: function (a) {
            this.a2(42, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "b4", {
        get: function () {
            return this.ag(61);
        },
        set: function (a) {
            this.a2(61, a);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "bu", {
        get: function () {
            return this.bt;
        },
        set: function (a) {
            var b = this.bt;
            this.bt = a;
            if (this.bt != b) {
                if (b != null) {
                    this.a0(5, b);
                }
                if (this.bt != null) {
                    this.ap(5, this.bt);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIPopup.prototype, "bs", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            var b = this.br;
            this.br = a;
            if (this.br != b) {
                if (b != null) {
                    this.a0(6, b);
                }
                if (this.br != null) {
                    this.ap(6, this.br);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    NativeUIPopup.prototype.ca = function (a, b, c) {
        this.bp.showRelativeToExclusionRect(this.am, a, b, c);
    };
    NativeUIPopup.prototype.b7 = function () {
        this.bp.close(this.am);
    };
    NativeUIPopup.prototype.b8 = function () {
        this.bp.notifyContentSizeChanged(this.am);
    };
    NativeUIPopup.prototype.b9 = function (a) {
        this.bp.notifyExclusionRectChanged(this.am, a);
    };
    NativeUIPopup.$t = markType(NativeUIPopup, 'NativeUIPopup', NativeUIContainer.$);
    return NativeUIPopup;
}(NativeUIContainer));
export { NativeUIPopup };
