// igniteui-angular-core/esm5/lib/NativeUIGrid.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { NativeUIContainer } from "./NativeUIContainer";
import { NativeUIBoxSizing_$type } from "./NativeUIBoxSizing";
import { enumGetBox, EnumUtil, markType } from "./type";
import { NativeUIPlaceContent_$type } from "./NativeUIPlaceContent";
/**
 * @hidden
 */
var NativeUIGrid = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NativeUIGrid, _super);
    function NativeUIGrid() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.br = null;
        return _this;
    }
    NativeUIGrid.prototype.get_e = function () {
        return 7;
    };
    Object.defineProperty(NativeUIGrid.prototype, "by", {
        get: function () {
            return this.bp.getRowDefinitionsCount(this.am);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIGrid.prototype, "bv", {
        get: function () {
            return this.bp.getColumnDefinitionsCount(this.am);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIGrid.prototype, "bp", {
        get: function () {
            return this.d;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIGrid.prototype, "bq", {
        get: function () {
            var a = this.ag(38);
            return EnumUtil.getEnumValue(NativeUIBoxSizing_$type, a);
        },
        set: function (a) {
            this.a2(38, enumGetBox(NativeUIBoxSizing_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIGrid.prototype, "bt", {
        get: function () {
            var a = this.ag(72);
            return EnumUtil.getEnumValue(NativeUIPlaceContent_$type, a);
        },
        set: function (a) {
            this.a2(72, enumGetBox(NativeUIPlaceContent_$type, a));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NativeUIGrid.prototype, "bu", {
        get: function () {
            return this.ag(37);
        },
        set: function (a) {
            this.a2(37, a);
        },
        enumerable: false,
        configurable: true
    });
    NativeUIGrid.prototype.b1 = function (a, b, c) {
        this.bp.addRowDefinition(this.am, a, b, c);
    };
    NativeUIGrid.prototype.b9 = function (a, b, c) {
        this.bp.updateRowDefinition(this.am, a, b, c);
    };
    NativeUIGrid.prototype.bz = function (a, b, c) {
        this.bp.addColumnDefinition(this.am, a, b, c);
    };
    NativeUIGrid.prototype.b0 = function (a, b, c, d) {
        this.bp.addColumnDefinitionWithMin(this.am, a, b, c, d);
    };
    NativeUIGrid.prototype.b8 = function (a, b, c) {
        this.bp.updateColumnDefinition(this.am, a, b, c);
    };
    NativeUIGrid.prototype.b2 = function (a, b, c) {
        this.bp.insertColumnDefinition(this.am, a, b, c);
    };
    NativeUIGrid.prototype.b3 = function (a, b, c, d) {
        this.bp.insertColumnDefinitionWithMin(this.am, a, b, c, d);
    };
    NativeUIGrid.prototype.b7 = function (a, b) {
        this.bp.setRow(this.am, a, b);
    };
    NativeUIGrid.prototype.b6 = function (a, b) {
        this.bp.setColumn(this.am, a, b);
    };
    NativeUIGrid.prototype.bx = function (a) {
        return this.bp.getRow(this.am, a);
    };
    NativeUIGrid.prototype.bw = function (a) {
        return this.bp.getColumn(this.am, a);
    };
    NativeUIGrid.prototype.b5 = function (a) {
        this.bp.removeRowDefinition(this.am, a);
    };
    NativeUIGrid.prototype.b4 = function (a) {
        this.bp.removeColumnDefinition(this.am, a);
    };
    Object.defineProperty(NativeUIGrid.prototype, "bs", {
        get: function () {
            return this.br;
        },
        set: function (a) {
            var b = this.br;
            this.br = a;
            if (this.br != b) {
                if (b != null) {
                    this.a0(1, b);
                }
                if (this.br != null) {
                    this.ap(1, this.br);
                }
            }
        },
        enumerable: false,
        configurable: true
    });
    NativeUIGrid.$t = markType(NativeUIGrid, 'NativeUIGrid', NativeUIContainer.$);
    return NativeUIGrid;
}(NativeUIContainer));
export { NativeUIGrid };
