// igniteui-angular-core/esm5/lib/NativeUIAbsolutePanel.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { NativeUIContainer } from "./NativeUIContainer";
import { markType } from "./type";
/**
 * @hidden
 */
var NativeUIAbsolutePanel = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NativeUIAbsolutePanel, _super);
    function NativeUIAbsolutePanel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NativeUIAbsolutePanel.prototype.get_e = function () {
        return 9;
    };
    Object.defineProperty(NativeUIAbsolutePanel.prototype, "bp", {
        get: function () {
            return this.d;
        },
        enumerable: false,
        configurable: true
    });
    NativeUIAbsolutePanel.prototype.bs = function (a, b) {
        this.bp.setX(this.am, a, b);
    };
    NativeUIAbsolutePanel.prototype.bt = function (a, b) {
        this.bp.setY(this.am, a, b);
    };
    NativeUIAbsolutePanel.prototype.bq = function (a) {
        return this.bp.getX(this.am, a);
    };
    NativeUIAbsolutePanel.prototype.br = function (a) {
        return this.bp.getY(this.am, a);
    };
    NativeUIAbsolutePanel.$t = markType(NativeUIAbsolutePanel, 'NativeUIAbsolutePanel', NativeUIContainer.$);
    return NativeUIAbsolutePanel;
}(NativeUIContainer));
export { NativeUIAbsolutePanel };
