// igniteui-angular-core/esm5/lib/PolarBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var PolarBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(PolarBaseDescription, _super);
    function PolarBaseDescription() {
        var _this = _super.call(this) || this;
        _this.gc = null;
        _this.gl = null;
        _this.gg = null;
        _this.gf = null;
        _this.f9 = null;
        _this.gi = null;
        _this.fs = false;
        _this.f0 = 0;
        _this.gn = null;
        _this.gm = null;
        _this.f8 = null;
        _this.fx = 0;
        _this.fn = null;
        _this.f1 = 0;
        _this.f2 = 0;
        _this.fp = false;
        _this.f7 = null;
        _this.gh = null;
        _this.fz = 0;
        _this.fr = false;
        _this.fq = false;
        _this.gj = null;
        _this.ga = null;
        _this.gk = null;
        _this.gb = null;
        _this.ge = null;
        _this.gd = null;
        return _this;
    }
    PolarBaseDescription.prototype.get_type = function () {
        return "PolarBase";
    };
    Object.defineProperty(PolarBaseDescription.prototype, "angleMemberPath", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("AngleMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusMemberPath", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("RadiusMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "highlightedRadiusMemberPath", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("HighlightedRadiusMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "highlightedAngleMemberPath", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("HighlightedAngleMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "angleAxisRef", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("AngleAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusAxisRef", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("RadiusAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "useCartesianInterpolation", {
        get: function () {
            return this.fs;
        },
        set: function (a) {
            this.fs = a;
            this.g("UseCartesianInterpolation");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "maximumMarkers", {
        get: function () {
            return this.f0;
        },
        set: function (a) {
            this.f0 = a;
            this.g("MaximumMarkers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineType", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("TrendLineType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineBrush", {
        get: function () {
            return this.gm;
        },
        set: function (a) {
            this.gm = a;
            this.g("TrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "actualTrendLineBrush", {
        get: function () {
            return this.f8;
        },
        set: function (a) {
            this.f8 = a;
            this.g("ActualTrendLineBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineThickness", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("TrendLineThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineDashArray", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("TrendLineDashArray");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLinePeriod", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("TrendLinePeriod");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "trendLineZIndex", {
        get: function () {
            return this.f2;
        },
        set: function (a) {
            this.f2 = a;
            this.g("TrendLineZIndex");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "clipSeriesToBounds", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("ClipSeriesToBounds");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "actualItemSearchMode", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("ActualItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "itemSearchMode", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("ItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "itemSearchThreshold", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("ItemSearchThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "isCustomPolarStyleAllowed", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("IsCustomPolarStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "isCustomPolarMarkerStyleAllowed", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("IsCustomPolarMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusMemberAsLegendLabel", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("RadiusMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "angleMemberAsLegendLabel", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("AngleMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "radiusMemberAsLegendUnit", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("RadiusMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "angleMemberAsLegendUnit", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("AngleMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "assigningPolarStyleRef", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("AssigningPolarStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(PolarBaseDescription.prototype, "assigningPolarMarkerStyleRef", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("AssigningPolarMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    PolarBaseDescription.$t = markType(PolarBaseDescription, 'PolarBaseDescription', MarkerSeriesDescription.$);
    return PolarBaseDescription;
}(MarkerSeriesDescription));
export { PolarBaseDescription };
