// igniteui-angular-core/esm5/lib/RadialBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { MarkerSeriesDescription } from "./MarkerSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var RadialBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(RadialBaseDescription, _super);
    function RadialBaseDescription() {
        var _this = _super.call(this) || this;
        _this.f5 = 0;
        _this.f6 = 0;
        _this.gc = null;
        _this.gg = null;
        _this.ge = null;
        _this.gd = null;
        _this.fn = null;
        _this.gj = null;
        _this.fr = null;
        _this.gf = null;
        _this.fo = null;
        _this.gi = null;
        _this.fq = null;
        _this.gh = null;
        _this.fp = null;
        _this.fz = false;
        _this.fy = false;
        _this.gk = null;
        _this.f0 = false;
        _this.f9 = null;
        _this.gl = null;
        _this.fx = false;
        _this.gb = null;
        _this.ga = null;
        return _this;
    }
    RadialBaseDescription.prototype.get_type = function () {
        return "RadialBase";
    };
    Object.defineProperty(RadialBaseDescription.prototype, "autoCalloutLabelPrecision", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("AutoCalloutLabelPrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "autoCalloutPercentagePrecision", {
        get: function () {
            return this.f6;
        },
        set: function (a) {
            this.f6 = a;
            this.g("AutoCalloutPercentagePrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "autoCalloutLabelValueSeparator", {
        get: function () {
            return this.gc;
        },
        set: function (a) {
            this.gc = a;
            this.g("AutoCalloutLabelValueSeparator");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "legendRadialLabelMode", {
        get: function () {
            return this.gg;
        },
        set: function (a) {
            this.gg = a;
            this.g("LegendRadialLabelMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "autoCalloutRadialLabelMode", {
        get: function () {
            return this.ge;
        },
        set: function (a) {
            this.ge = a;
            this.g("AutoCalloutRadialLabelMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "autoCalloutOthersLabelFormat", {
        get: function () {
            return this.gd;
        },
        set: function (a) {
            this.gd = a;
            this.g("AutoCalloutOthersLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "autoCalloutOthersLabelFormatSpecifiers", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("AutoCalloutOthersLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "proportionalRadialLabelFormat", {
        get: function () {
            return this.gj;
        },
        set: function (a) {
            this.gj = a;
            this.g("ProportionalRadialLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "proportionalRadialLabelFormatSpecifiers", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("ProportionalRadialLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "legendProportionalRadialLabelFormat", {
        get: function () {
            return this.gf;
        },
        set: function (a) {
            this.gf = a;
            this.g("LegendProportionalRadialLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "legendProportionalRadialLabelFormatSpecifiers", {
        get: function () {
            return this.fo;
        },
        set: function (a) {
            this.fo = a;
            this.g("LegendProportionalRadialLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "othersProportionalRadialLabelFormat", {
        get: function () {
            return this.gi;
        },
        set: function (a) {
            this.gi = a;
            this.g("OthersProportionalRadialLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "othersProportionalRadialLabelFormatSpecifiers", {
        get: function () {
            return this.fq;
        },
        set: function (a) {
            this.fq = a;
            this.g("OthersProportionalRadialLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "othersLegendProportionalRadialLabelFormat", {
        get: function () {
            return this.gh;
        },
        set: function (a) {
            this.gh = a;
            this.g("OthersLegendProportionalRadialLabelFormat");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "othersLegendProportionalRadialLabelFormatSpecifiers", {
        get: function () {
            return this.fp;
        },
        set: function (a) {
            this.fp = a;
            this.g("OthersLegendProportionalRadialLabelFormatSpecifiers");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "isCustomRadialStyleAllowed", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("IsCustomRadialStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "isCustomRadialMarkerStyleAllowed", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            this.fy = a;
            this.g("IsCustomRadialMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "transitionInMode", {
        get: function () {
            return this.gk;
        },
        set: function (a) {
            this.gk = a;
            this.g("TransitionInMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "isTransitionInEnabled", {
        get: function () {
            return this.f0;
        },
        set: function (a) {
            this.f0 = a;
            this.g("IsTransitionInEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "angleAxisRef", {
        get: function () {
            return this.f9;
        },
        set: function (a) {
            this.f9 = a;
            this.g("AngleAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "valueAxisRef", {
        get: function () {
            return this.gl;
        },
        set: function (a) {
            this.gl = a;
            this.g("ValueAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "clipSeriesToBounds", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("ClipSeriesToBounds");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "assigningRadialStyleRef", {
        get: function () {
            return this.gb;
        },
        set: function (a) {
            this.gb = a;
            this.g("AssigningRadialStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(RadialBaseDescription.prototype, "assigningRadialMarkerStyleRef", {
        get: function () {
            return this.ga;
        },
        set: function (a) {
            this.ga = a;
            this.g("AssigningRadialMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    RadialBaseDescription.$t = markType(RadialBaseDescription, 'RadialBaseDescription', MarkerSeriesDescription.$);
    return RadialBaseDescription;
}(MarkerSeriesDescription));
export { RadialBaseDescription };
