// igniteui-angular-core/esm5/lib/ScatterPolygonSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ShapeSeriesBaseDescription } from "./ShapeSeriesBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ScatterPolygonSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterPolygonSeriesDescription, _super);
    function ScatterPolygonSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.ft = null;
        _this.f6 = null;
        _this.f5 = null;
        _this.fv = 0;
        _this.fz = null;
        _this.f0 = null;
        _this.fx = null;
        _this.f3 = null;
        _this.fy = null;
        _this.f1 = null;
        _this.f4 = null;
        _this.f2 = null;
        return _this;
    }
    ScatterPolygonSeriesDescription.prototype.get_type = function () {
        return "ScatterPolygonSeries";
    };
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "shapeStyle", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("ShapeStyle");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerType", {
        get: function () {
            return this.f6;
        },
        set: function (a) {
            this.f6 = a;
            this.g("MarkerType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerTemplateRef", {
        get: function () {
            return this.f5;
        },
        set: function (a) {
            this.f5 = a;
            this.g("MarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerThickness", {
        get: function () {
            return this.fv;
        },
        set: function (a) {
            this.fv = a;
            this.g("MarkerThickness");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "actualMarkerTemplateRef", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("ActualMarkerTemplateRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerBrush", {
        get: function () {
            return this.f0;
        },
        set: function (a) {
            this.f0 = a;
            this.g("MarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "actualMarkerBrush", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("ActualMarkerBrush");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerOutline", {
        get: function () {
            return this.f3;
        },
        set: function (a) {
            this.f3 = a;
            this.g("MarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "actualMarkerOutline", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            this.fy = a;
            this.g("ActualMarkerOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerCollisionAvoidance", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("MarkerCollisionAvoidance");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerOutlineMode", {
        get: function () {
            return this.f4;
        },
        set: function (a) {
            this.f4 = a;
            this.g("MarkerOutlineMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterPolygonSeriesDescription.prototype, "markerFillMode", {
        get: function () {
            return this.f2;
        },
        set: function (a) {
            this.f2 = a;
            this.g("MarkerFillMode");
        },
        enumerable: false,
        configurable: true
    });
    ScatterPolygonSeriesDescription.$t = markType(ScatterPolygonSeriesDescription, 'ScatterPolygonSeriesDescription', ShapeSeriesBaseDescription.$);
    return ScatterPolygonSeriesDescription;
}(ShapeSeriesBaseDescription));
export { ScatterPolygonSeriesDescription };
