// igniteui-angular-core/esm5/lib/ScatterContourSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ScatterTriangulationSeriesDescription } from "./ScatterTriangulationSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ScatterContourSeriesDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ScatterContourSeriesDescription, _super);
    function ScatterContourSeriesDescription() {
        var _this = _super.call(this) || this;
        _this.fz = null;
        _this.fx = null;
        _this.fy = null;
        _this.fu = null;
        _this.ft = null;
        _this.fr = null;
        return _this;
    }
    ScatterContourSeriesDescription.prototype.get_type = function () {
        return "ScatterContourSeries";
    };
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "valueMemberPath", {
        get: function () {
            return this.fz;
        },
        set: function (a) {
            this.fz = a;
            this.g("ValueMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "valueMemberAsLegendLabel", {
        get: function () {
            return this.fx;
        },
        set: function (a) {
            this.fx = a;
            this.g("ValueMemberAsLegendLabel");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "valueMemberAsLegendUnit", {
        get: function () {
            return this.fy;
        },
        set: function (a) {
            this.fy = a;
            this.g("ValueMemberAsLegendUnit");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "fillScale", {
        get: function () {
            return this.fu;
        },
        set: function (a) {
            this.fu = a;
            this.g("FillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "actualFillScale", {
        get: function () {
            return this.ft;
        },
        set: function (a) {
            this.ft = a;
            this.g("ActualFillScale");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ScatterContourSeriesDescription.prototype, "valueResolver", {
        get: function () {
            return this.fr;
        },
        set: function (a) {
            this.fr = a;
            this.g("ValueResolver");
        },
        enumerable: false,
        configurable: true
    });
    ScatterContourSeriesDescription.$t = markType(ScatterContourSeriesDescription, 'ScatterContourSeriesDescription', ScatterTriangulationSeriesDescription.$);
    return ScatterContourSeriesDescription;
}(ScatterTriangulationSeriesDescription));
export { ScatterContourSeriesDescription };
