// igniteui-angular-core/esm5/lib/ShapeSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ShapeSeriesBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ShapeSeriesBaseDescription, _super);
    function ShapeSeriesBaseDescription() {
        var _this = _super.call(this) || this;
        _this.e2 = false;
        _this.e1 = false;
        _this.fb = null;
        _this.ff = null;
        _this.e8 = 0;
        _this.e7 = 0;
        _this.fg = null;
        _this.fe = null;
        _this.fi = null;
        _this.fj = null;
        _this.e5 = 0;
        _this.fd = null;
        _this.fc = null;
        _this.fh = null;
        return _this;
    }
    ShapeSeriesBaseDescription.prototype.get_type = function () {
        return "ShapeSeriesBase";
    };
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "isCustomShapeStyleAllowed", {
        get: function () {
            return this.e2;
        },
        set: function (a) {
            this.e2 = a;
            this.g("IsCustomShapeStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "isCustomShapeMarkerStyleAllowed", {
        get: function () {
            return this.e1;
        },
        set: function (a) {
            this.e1 = a;
            this.g("IsCustomShapeMarkerStyleAllowed");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "actualItemSearchMode", {
        get: function () {
            return this.fb;
        },
        set: function (a) {
            this.fb = a;
            this.g("ActualItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "itemSearchMode", {
        get: function () {
            return this.ff;
        },
        set: function (a) {
            this.ff = a;
            this.g("ItemSearchMode");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "itemSearchThreshold", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("ItemSearchThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "itemSearchPointsThreshold", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("ItemSearchPointsThreshold");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "shapeMemberPath", {
        get: function () {
            return this.fg;
        },
        set: function (a) {
            this.fg = a;
            this.g("ShapeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "highlightedShapeMemberPath", {
        get: function () {
            return this.fe;
        },
        set: function (a) {
            this.fe = a;
            this.g("HighlightedShapeMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "xAxisRef", {
        get: function () {
            return this.fi;
        },
        set: function (a) {
            this.fi = a;
            this.g("XAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "yAxisRef", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("YAxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "shapeFilterResolution", {
        get: function () {
            return this.e5;
        },
        set: function (a) {
            this.e5 = a;
            this.g("ShapeFilterResolution");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "assigningShapeStyleRef", {
        get: function () {
            return this.fd;
        },
        set: function (a) {
            this.fd = a;
            this.g("AssigningShapeStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "assigningShapeMarkerStyleRef", {
        get: function () {
            return this.fc;
        },
        set: function (a) {
            this.fc = a;
            this.g("AssigningShapeMarkerStyleRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ShapeSeriesBaseDescription.prototype, "styleShapeRef", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("StyleShapeRef");
        },
        enumerable: false,
        configurable: true
    });
    ShapeSeriesBaseDescription.$t = markType(ShapeSeriesBaseDescription, 'ShapeSeriesBaseDescription', SeriesDescription.$);
    return ShapeSeriesBaseDescription;
}(SeriesDescription));
export { ShapeSeriesBaseDescription };
