// igniteui-angular-core/esm5/lib/StackedSeriesBaseDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { CategorySeriesDescription } from "./CategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var StackedSeriesBaseDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(StackedSeriesBaseDescription, _super);
    function StackedSeriesBaseDescription() {
        var _this = _super.call(this) || this;
        _this.f1 = null;
        _this.f3 = false;
        _this.f4 = false;
        _this.f7 = null;
        return _this;
    }
    StackedSeriesBaseDescription.prototype.get_type = function () {
        return "StackedSeriesBase";
    };
    Object.defineProperty(StackedSeriesBaseDescription.prototype, "series", {
        get: function () {
            return this.f1;
        },
        set: function (a) {
            this.f1 = a;
            this.g("Series");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesBaseDescription.prototype, "autoGenerateSeries", {
        get: function () {
            return this.f3;
        },
        set: function (a) {
            this.f3 = a;
            this.g("AutoGenerateSeries");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesBaseDescription.prototype, "reverseLegendOrder", {
        get: function () {
            return this.f4;
        },
        set: function (a) {
            this.f4 = a;
            this.g("ReverseLegendOrder");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(StackedSeriesBaseDescription.prototype, "seriesCreatedRef", {
        get: function () {
            return this.f7;
        },
        set: function (a) {
            this.f7 = a;
            this.g("SeriesCreatedRef");
        },
        enumerable: false,
        configurable: true
    });
    StackedSeriesBaseDescription.$t = markType(StackedSeriesBaseDescription, 'StackedSeriesBaseDescription', CategorySeriesDescription.$);
    return StackedSeriesBaseDescription;
}(CategorySeriesDescription));
export { StackedSeriesBaseDescription };
