// igniteui-angular-core/esm5/lib/ToolActionCheckboxDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolActionCheckboxDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolActionCheckboxDescription, _super);
    function ToolActionCheckboxDescription() {
        var _this = _super.call(this) || this;
        _this.di = false;
        _this.dr = null;
        _this.dm = null;
        _this.dp = null;
        _this.dk = null;
        _this.ds = null;
        _this.dn = null;
        _this.dq = null;
        _this.dl = null;
        return _this;
    }
    ToolActionCheckboxDescription.prototype.get_type = function () {
        return "ToolActionCheckbox";
    };
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "isChecked", {
        get: function () {
            return this.di;
        },
        set: function (a) {
            this.di = a;
            this.g("IsChecked");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "uncheckedBackgroundColor", {
        get: function () {
            return this.dr;
        },
        set: function (a) {
            this.dr = a;
            this.g("UncheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "actualUncheckedBackgroundColor", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            this.dm = a;
            this.g("ActualUncheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "checkedBackgroundColor", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("CheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "actualCheckedBackgroundColor", {
        get: function () {
            return this.dk;
        },
        set: function (a) {
            this.dk = a;
            this.g("ActualCheckedBackgroundColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "uncheckedBorderColor", {
        get: function () {
            return this.ds;
        },
        set: function (a) {
            this.ds = a;
            this.g("UncheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "actualUncheckedBorderColor", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.g("ActualUncheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "checkedBorderColor", {
        get: function () {
            return this.dq;
        },
        set: function (a) {
            this.dq = a;
            this.g("CheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxDescription.prototype, "actualCheckedBorderColor", {
        get: function () {
            return this.dl;
        },
        set: function (a) {
            this.dl = a;
            this.g("ActualCheckedBorderColor");
        },
        enumerable: false,
        configurable: true
    });
    ToolActionCheckboxDescription.$t = markType(ToolActionCheckboxDescription, 'ToolActionCheckboxDescription', ToolActionDescription.$);
    return ToolActionCheckboxDescription;
}(ToolActionDescription));
export { ToolActionCheckboxDescription };
