// igniteui-angular-core/esm5/lib/ToolActionCheckboxListDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { ToolActionDescription } from "./ToolActionDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ToolActionCheckboxListDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ToolActionCheckboxListDescription, _super);
    function ToolActionCheckboxListDescription() {
        var _this = _super.call(this) || this;
        _this.dn = null;
        _this.dp = null;
        _this.dm = null;
        _this.di = null;
        _this.dj = null;
        return _this;
    }
    ToolActionCheckboxListDescription.prototype.get_type = function () {
        return "ToolActionCheckboxList";
    };
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "dataSourceRef", {
        get: function () {
            return this.dn;
        },
        set: function (a) {
            this.dn = a;
            this.g("DataSourceRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "selectedMemberPath", {
        get: function () {
            return this.dp;
        },
        set: function (a) {
            this.dp = a;
            this.g("SelectedMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "dataMemberPath", {
        get: function () {
            return this.dm;
        },
        set: function (a) {
            this.dm = a;
            this.g("DataMemberPath");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "primaryKey", {
        get: function () {
            return this.di;
        },
        set: function (a) {
            this.di = a;
            this.g("PrimaryKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ToolActionCheckboxListDescription.prototype, "selectedKeys", {
        get: function () {
            return this.dj;
        },
        set: function (a) {
            this.dj = a;
            this.g("SelectedKeys");
        },
        enumerable: false,
        configurable: true
    });
    ToolActionCheckboxListDescription.$t = markType(ToolActionCheckboxListDescription, 'ToolActionCheckboxListDescription', ToolActionDescription.$);
    return ToolActionCheckboxListDescription;
}(ToolActionDescription));
export { ToolActionCheckboxListDescription };
