// igniteui-angular-core/esm5/lib/ValueOverlayDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { SeriesDescription } from "./SeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var ValueOverlayDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(ValueOverlayDescription, _super);
    function ValueOverlayDescription() {
        var _this = _super.call(this) || this;
        _this.fn = null;
        _this.e9 = 0;
        _this.e1 = false;
        _this.fk = null;
        _this.fm = null;
        _this.fj = null;
        _this.e3 = 0;
        _this.fh = 0;
        _this.fl = null;
        _this.e5 = 0;
        _this.e7 = 0;
        _this.e6 = 0;
        _this.e4 = 0;
        _this.e8 = 0;
        return _this;
    }
    ValueOverlayDescription.prototype.get_type = function () {
        return "ValueOverlay";
    };
    Object.defineProperty(ValueOverlayDescription.prototype, "axisRef", {
        get: function () {
            return this.fn;
        },
        set: function (a) {
            this.fn = a;
            this.g("AxisRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "value", {
        get: function () {
            return this.e9;
        },
        set: function (a) {
            this.e9 = a;
            this.g("Value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "isAxisAnnotationEnabled", {
        get: function () {
            return this.e1;
        },
        set: function (a) {
            this.e1 = a;
            this.g("IsAxisAnnotationEnabled");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationFormatLabelRef", {
        get: function () {
            return this.fk;
        },
        set: function (a) {
            this.fk = a;
            this.g("AxisAnnotationFormatLabelRef");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationTextColor", {
        get: function () {
            return this.fm;
        },
        set: function (a) {
            this.fm = a;
            this.g("AxisAnnotationTextColor");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationBackground", {
        get: function () {
            return this.fj;
        },
        set: function (a) {
            this.fj = a;
            this.g("AxisAnnotationBackground");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationBackgroundCornerRadius", {
        get: function () {
            return this.e3;
        },
        set: function (a) {
            this.e3 = a;
            this.g("AxisAnnotationBackgroundCornerRadius");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationInterpolatedValuePrecision", {
        get: function () {
            return this.fh;
        },
        set: function (a) {
            this.fh = a;
            this.g("AxisAnnotationInterpolatedValuePrecision");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationOutline", {
        get: function () {
            return this.fl;
        },
        set: function (a) {
            this.fl = a;
            this.g("AxisAnnotationOutline");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationPaddingLeft", {
        get: function () {
            return this.e5;
        },
        set: function (a) {
            this.e5 = a;
            this.g("AxisAnnotationPaddingLeft");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationPaddingTop", {
        get: function () {
            return this.e7;
        },
        set: function (a) {
            this.e7 = a;
            this.g("AxisAnnotationPaddingTop");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationPaddingRight", {
        get: function () {
            return this.e6;
        },
        set: function (a) {
            this.e6 = a;
            this.g("AxisAnnotationPaddingRight");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationPaddingBottom", {
        get: function () {
            return this.e4;
        },
        set: function (a) {
            this.e4 = a;
            this.g("AxisAnnotationPaddingBottom");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ValueOverlayDescription.prototype, "axisAnnotationStrokeThickness", {
        get: function () {
            return this.e8;
        },
        set: function (a) {
            this.e8 = a;
            this.g("AxisAnnotationStrokeThickness");
        },
        enumerable: false,
        configurable: true
    });
    ValueOverlayDescription.$t = markType(ValueOverlayDescription, 'ValueOverlayDescription', SeriesDescription.$);
    return ValueOverlayDescription;
}(SeriesDescription));
export { ValueOverlayDescription };
