// igniteui-angular-core/esm5/lib/WebColumnStateDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { Description } from "./Description";
import { markType } from "./type";
/**
 * @hidden
 */
var WebColumnStateDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebColumnStateDescription, _super);
    function WebColumnStateDescription() {
        var _this = _super.call(this) || this;
        _this.t = false;
        _this.w = false;
        _this.o = false;
        _this.m = false;
        _this.x = false;
        _this.p = false;
        _this.ay = null;
        _this.az = null;
        _this.a0 = null;
        _this.q = false;
        _this.s = false;
        _this.av = null;
        _this.r = false;
        _this.aw = null;
        _this.ar = null;
        _this.ax = null;
        _this.u = false;
        _this.v = false;
        _this.j = false;
        _this.aq = null;
        _this.k = false;
        _this.l = false;
        _this.i = false;
        _this.n = false;
        _this.y = false;
        return _this;
    }
    WebColumnStateDescription.prototype.get_type = function () {
        return "WebColumnState";
    };
    Object.defineProperty(WebColumnStateDescription.prototype, "type", {
        get: function () {
            return this.get_type();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "pinned", {
        get: function () {
            return this.t;
        },
        set: function (a) {
            this.t = a;
            this.g("Pinned");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "sortable", {
        get: function () {
            return this.w;
        },
        set: function (a) {
            this.w = a;
            this.g("Sortable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "filterable", {
        get: function () {
            return this.o;
        },
        set: function (a) {
            this.o = a;
            this.g("Filterable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "editable", {
        get: function () {
            return this.m;
        },
        set: function (a) {
            this.m = a;
            this.g("Editable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "sortingIgnoreCase", {
        get: function () {
            return this.x;
        },
        set: function (a) {
            this.x = a;
            this.g("SortingIgnoreCase");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "filteringIgnoreCase", {
        get: function () {
            return this.p;
        },
        set: function (a) {
            this.p = a;
            this.g("FilteringIgnoreCase");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "headerClasses", {
        get: function () {
            return this.ay;
        },
        set: function (a) {
            this.ay = a;
            this.g("HeaderClasses");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "headerGroupClasses", {
        get: function () {
            return this.az;
        },
        set: function (a) {
            this.az = a;
            this.g("HeaderGroupClasses");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "maxWidth", {
        get: function () {
            return this.a0;
        },
        set: function (a) {
            this.a0 = a;
            this.g("MaxWidth");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "groupable", {
        get: function () {
            return this.q;
        },
        set: function (a) {
            this.q = a;
            this.g("Groupable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "hidden", {
        get: function () {
            return this.s;
        },
        set: function (a) {
            this.s = a;
            this.g("Hidden");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "dataType", {
        get: function () {
            return this.av;
        },
        set: function (a) {
            this.av = a;
            this.g("DataType");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "hasSummary", {
        get: function () {
            return this.r;
        },
        set: function (a) {
            this.r = a;
            this.g("HasSummary");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "field", {
        get: function () {
            return this.aw;
        },
        set: function (a) {
            this.aw = a;
            this.g("Field");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "width", {
        get: function () {
            return this.ar;
        },
        set: function (a) {
            this.ar = a;
            this.g("Width");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "header", {
        get: function () {
            return this.ax;
        },
        set: function (a) {
            this.ax = a;
            this.g("Header");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "resizable", {
        get: function () {
            return this.u;
        },
        set: function (a) {
            this.u = a;
            this.g("Resizable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "searchable", {
        get: function () {
            return this.v;
        },
        set: function (a) {
            this.v = a;
            this.g("Searchable");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "columnGroup", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.g("ColumnGroup");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "parent", {
        get: function () {
            return this.aq;
        },
        set: function (a) {
            this.aq = a;
            this.g("Parent");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "disableHiding", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.g("DisableHiding");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "disablePinning", {
        get: function () {
            return this.l;
        },
        set: function (a) {
            this.l = a;
            this.g("DisablePinning");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "collapsible", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.g("Collapsible");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "expanded", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.g("Expanded");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebColumnStateDescription.prototype, "visibleWhenCollapsed", {
        get: function () {
            return this.y;
        },
        set: function (a) {
            this.y = a;
            this.g("VisibleWhenCollapsed");
        },
        enumerable: false,
        configurable: true
    });
    WebColumnStateDescription.$t = markType(WebColumnStateDescription, 'WebColumnStateDescription', Description.$);
    WebColumnStateDescription.__marshalByValue = true;
    WebColumnStateDescription.__marshalByValueAlias = "ColumnState";
    return WebColumnStateDescription;
}(Description));
export { WebColumnStateDescription };
