// igniteui-angular-core/esm5/lib/WebHierarchicalGridBaseDirectiveDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { WebGridBaseDirectiveDescription } from "./WebGridBaseDirectiveDescription";
import { markType } from "./type";
/**
 * @hidden
 */
var WebHierarchicalGridBaseDirectiveDescription = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(WebHierarchicalGridBaseDirectiveDescription, _super);
    function WebHierarchicalGridBaseDirectiveDescription() {
        var _this = _super.call(this) || this;
        _this.gs = null;
        _this.gp = false;
        _this.gn = null;
        _this.gr = null;
        return _this;
    }
    WebHierarchicalGridBaseDirectiveDescription.prototype.get_type = function () {
        return "WebHierarchicalGridBaseDirective";
    };
    Object.defineProperty(WebHierarchicalGridBaseDirectiveDescription.prototype, "hasChildrenKey", {
        get: function () {
            return this.gs;
        },
        set: function (a) {
            this.gs = a;
            this.g("HasChildrenKey");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHierarchicalGridBaseDirectiveDescription.prototype, "showExpandAll", {
        get: function () {
            return this.gp;
        },
        set: function (a) {
            this.gp = a;
            this.g("ShowExpandAll");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHierarchicalGridBaseDirectiveDescription.prototype, "rootGrid", {
        get: function () {
            return this.gn;
        },
        set: function (a) {
            this.gn = a;
            this.g("RootGrid");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebHierarchicalGridBaseDirectiveDescription.prototype, "dataPreLoadRef", {
        get: function () {
            return this.gr;
        },
        set: function (a) {
            this.gr = a;
            this.g("DataPreLoadRef");
        },
        enumerable: false,
        configurable: true
    });
    WebHierarchicalGridBaseDirectiveDescription.$t = markType(WebHierarchicalGridBaseDirectiveDescription, 'WebHierarchicalGridBaseDirectiveDescription', WebGridBaseDirectiveDescription.$);
    return WebHierarchicalGridBaseDirectiveDescription;
}(WebGridBaseDirectiveDescription));
export { WebHierarchicalGridBaseDirectiveDescription };
